---
title: "{{< fa circle-info >}} About ConnectomeDB"
---

---

**Understanding how ligand-receptor (LR) pairs mediate cell-cell communication is essential for advancing both biological and medical research. ConnectomeDB, launched in 2015, is an ongoing project designed to create a comprehensive and highly-accurate database of interacting LR pairs and analysis tools to advance the understanding of cell-cell communication in human and other species. Developers:** [Perkins Systems Biology & Genomics Lab](https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/) & [YCU Bioinformatics Lab](https://www-user.yokohama-cu.ac.jp/~bioinfo/hp/).


**If you are using ConnectomeDB or our related tools, please** [cite our work](citeus.qmd).  
<span class="half-break"></span>

::: {.panel-tabset}

## **What is ConnectomeDB2025?**

#### **ConnectomeDB2025 - Database & Online Resource for Cell-to-Cell Communication Predictions**
<span class="half-break"></span>

- **5,000+ manually curated human LR pairs**: most comprehensive and most accurate with primary literature support\
- **2,500+ ligands** and **2,500+ receptors**: extensive gene metadata & annotations\
- **Biologically relevant content**: ontologies, cancers and other diseases\
- **Homologs in vertebrate species**: ~93% LR pairs in mouse and in other species\

<span class="half-break"></span>
**Publication**: [ ConnectomeDB2025, a high quality manually curated ligand-receptor database for cell-to-cell communication prediction (Journal Title, 2025)](about.qmd) *(TBA)*
**GitHub repo**: <https://github.com/bioinfo-YCU/ConnectomeDB>  *(Currently open to developpers only)*

## **How to use ConnectomeDB?**

#### **Using ConnectomeDB2025**
<span class="half-break"></span>

The online implementation of ConnectomeDB2025 is a free, no-code, and user-friendly resource. Only a web browser and an internet connection are needed. Please read our [**Terms & Conditions**](policy.qmd) and start exploring ligand-receptor communication in human, mouse, or other available species by accessing [**ConnectomeDB 2025**](database-human.qmd).

<span class="half-break"></span>
**Not familiar with ConnectomeDB 2025 or want to learn more?** Please check out [**our tutorials**](tutorials.qmd).

## **Contribute to ConnectomeDB**

#### **Your Feedback**
<span class="half-break"></span>

We welcome your feedback and suggestions for improving ConnectomeDB content and would be glad to hear any ideas for further development.  
If you would like to contribute to the ConnectomeDB project, please visit the [GitHub repository](https://github.com/bioinfo-YCU/ConnectomeDB) *(Currently open to developpers only)* or [email us](mailto:need_to_think@com).

## **Try NATMI**

#### **NATMI (Network Analysis Toolkit for Multicellular Interactions)**
<span class="half-break"></span>

**NATMI** is our Python-based toolkit designed for constructing and analyzing multi-cellular communication networks in multiple species. NATMI can use both single-cell and bulk gene expression, as well as proteomic data, to predict cell-to-cell communication at the levels of niches, tissues, and the entire organism.

<span class="half-break"></span>

**Publication**: [Predicting cell-to-cell communication networks using NATMI (Nat commun, 2020)](https://doi.org/10.1038/s41467-020-18873-z)   
**GitHub repo**: <https://github.com/forrest-lab/NATMI> 


## **Questions?**

#### **Let Us Know**
<span class="half-break"></span>

For questions related to the **ConnectomeDB2025 online resource**, please fill out [this form](this form) or contact us via this [email](mailto:bioinfo@yokohama-cu.ac.jp).  
For general inquires, please send Alistair, Jordan and Rui an [email](mailto:alistair.forrest@gmail.com; jordanr@yokohama-cu.ac.jp; rui.hou@research.uwa.edu.au).

:::
