---
title: "{{< fa circle-info >}} About ConnectomeDB"
---

---

**Understanding how ligand-receptor (LR) pairs mediate cell-cell communication is essential for advancing biological and medical discoveries. ConnectomeDB, launched in 2015, is an ongoing project designed to create a comprehensive and highly-accurate database of interacting LR pairs and analysis tools to enhance the understanding of cell-cell communication in human and other species. Developers:** [Perkins Systems Biology & Genomics Lab](https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/) & [YCU Bioinformatics Lab](https://www-user.yokohama-cu.ac.jp/~bioinfo/hp/).


::: {.panel-tabset}

## **What is ConnectomeDB2025?**

#### **ConnectomeDB2025 - Database & Online Resource for Cell-to-Cell Communication Predictions**
<span class="half-break"></span>

- **5,000+ Manually Curated Human LR pairs**: most comprehensive and most accurate with primary literature support\
- **2,500+ Ligands** & **2,500+ Receptors**: gene cards with extensive metadata and annotations\
- **Biologically Relevant Content**: ontologies, cancers and other diseases annotations\
- **Homologous Interactions**: ~93% LR pairs in mouse and in other vertebrate species\

<span class="half-break"></span>
**Publication**: [ ConnectomeDB2025, a high quality manually curated ligand-receptor database for cell-to-cell communication prediction (Journal Title, 2025)](about.qmd) *(TBA)*  
**GitHub repo**: <https://github.com/bioinfo-YCU/ConnectomeDB>  *(Currently open to developers only)*

## **How to use ConnectomeDB?**

#### **Using ConnectomeDB2025**
<span class="half-break"></span>

The online implementation of ConnectomeDB2025 is a free, no-code, and user-friendly resource developped using <a href='https://quarto.org/' target='_blank'>Quarto 1.6+</a> and <a href='https://www.biorender.com/' target='_blank'> BioRender</a> images. To use this resource, only a web browser and an internet connection are needed. Please be sure to read our [**Terms & Conditions**](policy.qmd) before using this website. 

<span class="half-break"></span>
**Not familiar with ConnectomeDB 2025 or want to learn more?** Please check out [our tutorials](tutorials.qmd).

## **Contribute to ConnectomeDB**

#### **Your Feedback**
<span class="half-break"></span>

We welcome your feedback and suggestions for improving ConnectomeDB content and would be glad to hear any ideas for further development.  
If you would like to contribute to the ConnectomeDB project, please visit the [GitHub repository](https://github.com/bioinfo-YCU/ConnectomeDB) *(Currently open to developpers only)* or [email us](mailto:need_to_think@com).

## **Try NATMI**

#### **NATMI (Network Analysis Toolkit for Multicellular Interactions)**
<span class="half-break"></span>

**NATMI** is our Python-based toolkit designed for constructing and analyzing multi-cellular communication networks in multiple species. NATMI can use both single-cell and bulk gene expression, as well as proteomic data, to predict cell-to-cell communication at the levels of niches, tissues, and the entire organism.

<span class="half-break"></span>

**Publication**: [Predicting cell-to-cell communication networks using NATMI (Nat commun, 2020)](https://doi.org/10.1038/s41467-020-18873-z)   
**GitHub repo**: <https://github.com/forrest-lab/NATMI> 


## **Questions?**

#### **Let Us Know**
<span class="half-break"></span>

For questions related to the **ConnectomeDB2025 online resource**, please fill out the following form below or contact us via this [email](mailto:bioinfo@yokohama-cu.ac.jp).  

<iframe src="https://docs.google.com/forms/d/e/1FAIpQLSemqP51oOCW5P5nWTBEyyK4x889RZMB4ir5pL_b0lAA9EeYzA/viewform?embedded=true" width="800" height="400" frameborder="0" marginheight="0" marginwidth="0">Loadingâ€¦</iframe>

***
For general inquires, please send Alistair, Jordan and Rui an [email](mailto:alistair.forrest@gmail.com; jordanr@yokohama-cu.ac.jp; rui.hou@research.uwa.edu.au).

:::

**If you are using ConnectomeDB or our related tools, please** [cite our work](citeus.qmd).  
<span class="half-break"></span>
