---
title: "About ConnectomeDB"
---

<span class="half-break"></span>

**Understanding ligand-receptor (LR)-mediated cell-cell communication is essential for advancing biological and medical research. ConnectomeDB, initially released in 2015, is a long term-project that aims to provide comprehensive and highly-accuarte database of interacting LR pairs and analysis tools for studying cell-cell communication in human and other species.** **The resource is developped by**: [Perkins Systems Biology & Genomics Lab](https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/) & [YCU Bioinformatics Lab](https://www-user.yokohama-cu.ac.jp/~bioinfo/hp/).

<span class="half-break"></span>

::: {.panel-tabset}

## **What is ConnectomeDB2025?**

### ConnectomeDB2025

- our newest database and user-friendly free online resource for predicting cell-to-cell communication in multiple species
 - 2,361 manually-curated human LR Interacting Pairs with extensive experimental support from primary literature
 - homologues LR interactions in mouse (~93%) and other selected vertebrate species (chimpanzee, rat, etc.)
 - exploration and prioritization of LR pairs based on existing evidence and biological relevance
 - easy access to extensive metadata and annotations for 854 Ligand & 700 Receptor genes
 - ideal for discovery of novel therapeutic targets, biomarkers, signaling pathways, and more

**This work**: [ ConnectomeDB2025, a high quality manually curated ligand-receptor database for cell-to-cell communication prediction (Journal Ttile, 2025)](https://doi.org/10.1038/s41467-020-18873-z)  
**GitHub**: <https://github.com/bioinfo-YCU/ConnectomeDB>  
---
**All papers**: [see here](citeus.qmd)  

## **How to use ConnectomeDB2025?**

The online implementation of ConnectomeDB2025 is a user-friendly, no-code resource that only requires a browser and an internet connection. To explore ligand-receptor communication in humans, mice, or other available species, please read our [**Terms & Conditions**](link)(needs to be made) and access [**ConnectomeDB 2025**](database.qmd).

**Not familiar with ConnectomeDB 2025 or want to learn more?** Please check out [**our tutorial**](link). (Needs to be made)

## **Terms & Conditions**

Might need to add **Terms & Conditions**; 
see CDB as an example <https://www.cellphonedb.org/legal.html>

They also have Cookie Policy. All are a part of the footer, so it is accessible when the website is loaded.

## **Contribute to ConnectomeDB**

We welcome your suggestions for **ConnectomeDB** content and would be happy to hear your ideas for future development.

Please visit the [GitHub repository](https://github.com/bioinfo-YCU/ConnectomeDB) or [email us](mailto:need_to_think@com).

## **Try NATMI**

**NATMI**  (Network Analysis Toolkit for Multicellular Interactions) is our Python-based toolkit designed for constructing and analyzing multi-cellular communication networks in multiple species. NATMI can use both single-cell and bulk gene expression, as well as proteomic data, to predict cell-to-cell communication at the levels of niches, tissues, and the entire organism.

**Publication**: [Predicting cell-to-cell communication networks using NATMI (Nat commun, 2020)](https://doi.org/10.1038/s41467-020-18873-z)   
**GitHub Repo**: <https://github.com/forrest-lab/NATMI> 

## **Questions?**

For question related to **connectomeDB2025 online resource** please fill out [this form](this form) or use [this email](mailto:bioinfo@yokohama-cu.ac.jp).

For question related to bla bla please [email](mailto:xxx@.com;mailto:xxx@.com;).

:::
