---
title: "{{< fa circle-info >}} About ConnectomeDB"
execute:
  echo: false
  warnings: false  # Add this line to suppress warnings globally
---

<i class="far fa-link"></i>
```{python}
import os
import sys
import pandas as pd
import warnings
warnings.filterwarnings("ignore", category=pd.errors.DtypeWarning)
sys.path.append(os.path.abspath("src"))  # Add src directory to path
from createDataTable import (
    lrPairsCount,
    ligandCount,
    receptorCount,
    sourceCount
)
```

<hr>

<div style="border-left: 4px solid #C28840; background-color: #F5F5F5; padding: 0px 0.3em; margin: 1em 0; font-family: Arial, sans-serif; font-size: 110%;">
  <span class="half-break"></span>
  <strong>Before using this resource</strong>, please read our 
  <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/policy.html" target="_blank" style="color: #1E90FF; text-decoration: underline; font-weight: bold;">Terms &amp; Conditions</a>.
  <span class="half-break"></span>
</div>

<div style="border: 1.5px solid #F5F5F5; border-radius: 10px; padding: 8px 10px 0px 10px; background-color: #F5F5F5; font-family: Arial, sans-serif; font-size: 110%; width: 100%; box-sizing: border-box; margin: 0; line-height: 1.25;">
  <strong>ConnectomeDB</strong> is a comprehensive and ongoing project that provides a high-quality database of interacting ligand-receptor pairs and analysis tools. Launched in 2015, it aims to enhance understanding of cell-cell communication in humans and other mammals, supporting biological and medical research.<span class="half-break"></span>
  <strong>Developers</strong>: 
  <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/hp/" target="_blank" style="color:#1E90FF; text-decoration:underline;">YCU Bioinformatics Lab</a>
    &amp; 
  <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/" target="_blank" style="color:#1E90FF; text-decoration:underline;">Perkins Systems Biology Genomics Lab</a>.
</div>

<span class="half-break"></span>

::: {.panel-tabset}

## **What is ConnectomeDB2025?**

#### **ConnectomeDB2025 - Database & Online Resource for Cell-to-Cell Communication Predictions**
<span class="half-break"></span>

- **`{python} f"{lrPairsCount:,}"`  Manually Curated Human LR pairs**: comprehensive and most accurate with primary literature support\
- **`{python} f"{ligandCount:,}"` Ligands** & **`{python} f"{receptorCount:,}"` Receptors**: gene cards with extensive metadata and annotations\
- **Biologically Relevant Content**: ontologies, cancers and other diseases annotations\
- **Homologous Interactions**: ~98% <variable> LR pairs in mouse and other vertebrate species - [Al: see details here - might be link to another page?]\
 - **Free, user-friendly resource**: developed with <a href='https://quarto.org/' target='_blank'>Quarto 1.7+</a> and <a href='https://www.biorender.com/' target='_blank'> BioRender</a> 

<hr>

<span class="half-break"></span>
**Publication**: [ConnectomeDB2025, a high quality manually curated ligand-receptor database for cell-to-cell communication prediction (Journal Title, 2025)](about.qmd)  
**GitHub repo**: <https://github.com/bioinfo-YCU/ConnectomeDB> *(open upon publication)*

## **How to use ConnectomeDB?**

#### **Access ConnectomeDB2025 Online**
<span class="half-break"></span>

An internet connection and a web browser are required. The following major operating systems and web browsers have been tested and are quaranteed to work:

- **Windows 10/11**: Google Chrome, Microsoft Edge
- **macOS 15+**: Google Chrome, Safari

<hr>

<span class="half-break"></span>
**Not familiar with ConnectomeDB 2025 or want to learn more?** Please check out [our tutorials](tutorials.qmd) *(currently under construction)*.

## **Contribute to ConnectomeDB**

#### **Your Feedback**
<span class="half-break"></span>

We welcome your feedback and suggestions for improving ConnectomeDB content and would be glad to hear any ideas for further development.  
If you would like to contribute to the ConnectomeDB project, please visit the [GitHub repository](https://github.com/bioinfo-YCU/ConnectomeDB) *(open upon publication)* or [email us](mailto:alistair.forrest@gmail.com; jordanr@yokohama-cu.ac.jp; rui.hou@research.uwa.edu.au).

## **Try NATMI**

#### **Network Analysis Toolkit for Multicellular Interactions (NATMI)**
<span class="half-break"></span>

- our *Python-based toolkit* designed for constructing and analyzing multi-cellular communication networks in multiple species
- predicts cell-to-cell communication at the levels of niches, tissues, and the entire organism
- can use both single-cell and bulk gene expression and proteomic data

<hr>

<span class="half-break"></span>

**Publication**: [Predicting cell-to-cell communication networks using NATMI (Nat commun, 2020)](https://doi.org/10.1038/s41467-020-18873-z)   
**GitHub repo**: <https://github.com/forrest-lab/NATMI> 

## **Cite us**
<span class="half-break"></span>

#### **Please cite the following papers if you use ConnectomeDB or NATMI in your publication:**
<span class="half-break"></span>

1. [ConnectomeDB2025, a high quality manually curated ligand-receptor database for cell-to-cell communication prediction](), P Liu, et al. *Journal Title* **XX (YY)**, ZZZ (2025)  
This work describes ConnectomeDB2025, which is our most current ligand-receptor (LR) pair database and online resource.

2. [Predicting cell-to-cell communication networks using NATMI](https://doi.org/10.1038/s41467-020-18873-z), R Hou, et al. *Nature communications* **11 (1)**, 5011 (2020)  
This study introduces NATMI (Network Analysis Toolkit for Multicellular Interactions) and the ConnectomeDB2020 release.

3. [A draft network of ligandâ€“receptor-mediated multicellular signalling in human](https://doi.org/10.1038/ncomms8866), JA Ramilowski, et al. *Nature communications* **6 (1)**, 7866 (2015)  
This is the original resource upon which subsequent ConnectomeDB releases are based, and is widely used by many other LR databases.

## **Inquiries**

#### **Let us Help You**
<span class="half-break"></span>

**General inquiries:** send Alistair, Jordan, and Rui an [email](mailto:alistair.forrest@gmail.com; jordanr@yokohama-cu.ac.jp; rui.hou@research.uwa.edu.au).

<hr>

**Online resource issues**: fill out this [Inquiry form](https://forms.gle/m7k5vygRPdTRdaTh6) or contact us via this [email](mailto:bioinfo@yokohama-cu.ac.jp; alistair.forrest@gmail.com).  
