---
title: "{{< fa circle-info >}} Introduction"
execute:
  echo: false
  warnings: false  # Add this line to suppress warnings globally
header-includes: <script src='/js/keepDropdownMenuGold.js'></script>
---

<hr>

```{python}
import os
import sys
import pandas as pd
import warnings
warnings.filterwarnings("ignore", category=pd.errors.DtypeWarning)
sys.path.append(os.path.abspath("src"))  # Add src directory to path
from createDataTable import (
  lrPairsCount,
  ligandCount,
  receptorCount,
  numOfMouseOrth
)
```

<div style="border: 1.5px solid #F5F5F5; border-radius: 10px; padding: 8px 10px 0px 10px; background-color: #F5F5F5; width: 100%; box-sizing: border-box; margin: 0; line-height: 1.25;">
  <strong>ConnectomeDB</strong> is a comprehensive and ongoing project that provides a high-quality manually curated database of interacting ligand-receptor pairs for use in cell-cell communication analysis. First released in 2015 ([Ramilowski, *et al.*](https://doi.org/10.1038/ncomms8866){target="_blank"}), and subsequently updated in 2020 ([Hou, *et al.*](https://www.nature.com/articles/s41467-020-18873-z){target="_blank"}) and 2025 (Liu, Maezono *et al.*), it aims to enhance the understanding of cell-cell communication in humans and other mammals, supporting biological and medical research.<span class="half-break"></span></div>

<span class="half-break"></span>

### **ConnectomeDB2025**

<div style="margin-top: 1em;"></div>

- **`{python} f"{lrPairsCount:,}"` Manually Curated Ligand–Receptor Pairs** (Primarily Human; includes `{python} f"{numOfMouseOrth:,}"` Mouse-specific): comprehensive and most accurate with primary literature support\
- **`{python} f"{ligandCount:,}"` Ligands** & **`{python} f"{receptorCount:,}"` Receptors**: metadata and functional annotations\
- **Homologous Interactions**: ~98% LR pairs in mouse and in 11 other vertebrate species\

<hr>

### **Online Resource**

Accesing all parts of the online resource requires no username and no password.

**Software**: ConnectomeDB2025 is developed with <a href='https://quarto.org/' target='_blank'>Quarto 1.7+</a> and <a href='https://www.biorender.com/' target='_blank'> BioRender</a>\
**OS & Browsers**: Windows 10/11 (Google Chrome, Microsoft Edge), macOS 15+ (Google Chrome, Safari), mobile devices (Google Chrome, Safari).

<hr>

<div style="margin-top: 1em;"></div>

### **Developpers**

This resource was developed and is maintained by <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/hp/" target="_blank" >YCU Bioinformatics Lab</a>
and 
<a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/" target="_blank">Perkins Systems Biology &amp; Genomics Lab</a>\
**GitHub**: 

<hr>

<div style="margin-top: 1em;"></div>

### **Cite Us**

1. *ConnectomeDB2025, a high quality manually curated ligand-receptor database for cell-to-cell communication prediction*, P Liu, S Maezono, *et al.* (TBA)\
Short description\
PMID: [(TBA)](); DOI: [(TBA)]()
    
2. *Predicting cell-to-cell communication networks using NATMI*, R Hou, *et al.* *Nature communications* **11(1)**, 5011, **2020**\
Short description\
PMID: [33024107](https://pubmed.ncbi.nlm.nih.gov/33024107/){target="_blank"}; DOI: [10.1038/s41467-020-18873-z](https://www.nature.com/articles/s41467-020-18873-z){target="_blank"}
    
3. *A draft network of ligand–receptor-mediated multicellular signalling in human*, JA Ramilowski, *et al.* *Nature communications* **6(1)**, 7866, **2015**\
Short description\
PMID: [26198319](https://pubmed.ncbi.nlm.nih.gov/26198319/){target="_blank"}; DOI: [10.1038/ncomms8866](https://doi.org/10.1038/ncomms8866){target="_blank"}

