{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fcc5ddc1-519e-439c-a130-3dd8bcd181bd",
   "metadata": {},
   "source": [
    "# Python Notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "023d8273-becd-447e-98d2-949b29e23aba",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'os' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Change working directory to ConnectomeDB\u001b[39;00m\n\u001b[1;32m      2\u001b[0m project_root \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m/Users/sakuramaezono/Library/CloudStorage/OneDrive-YokohamaCityUniversity/Personal/05_Python_repositories/ConnectomeDB\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m----> 3\u001b[0m \u001b[43mos\u001b[49m\u001b[38;5;241m.\u001b[39mchdir(project_root)\n\u001b[1;32m      4\u001b[0m sys\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mappend(os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mabspath(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msrc\u001b[39m\u001b[38;5;124m\"\u001b[39m))\n",
      "\u001b[0;31mNameError\u001b[0m: name 'os' is not defined"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "79f99584-74aa-4467-8cfc-da4979c15eb4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PMID</th>\n",
       "      <th>original source</th>\n",
       "      <th>Title</th>\n",
       "      <th>Journal</th>\n",
       "      <th>Year</th>\n",
       "      <th>Species</th>\n",
       "      <th>&lt;span title=\"Double-click header of Interaction ID to ensure all values are shown\"&gt;Interaction ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\" Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\"&gt;Human LR Pair&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Database Source to ensure all values are shown\"&gt;Database Source&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\" Official Gene Symbol; Hover on symbols below to show gene names\"&gt;Ligand&amp;nbsp;&amp;nbsp;&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>...</th>\n",
       "      <th>&lt;span title=\"Double-click header of Dog Receptor Ensembl ID to ensure all values are shown\"&gt;Dog Receptor Ensembl ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Dog Receptor to ensure all values are shown\"&gt;Dog Receptor&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Horse Ligand Ensembl ID to ensure all values are shown\"&gt;Horse Ligand Ensembl ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Horse Ligand to ensure all values are shown\"&gt;Horse Ligand&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Horse Receptor Ensembl ID to ensure all values are shown\"&gt;Horse Receptor Ensembl ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Horse Receptor to ensure all values are shown\"&gt;Horse Receptor&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Sheep Ligand Ensembl ID to ensure all values are shown\"&gt;Sheep Ligand Ensembl ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Sheep Ligand to ensure all values are shown\"&gt;Sheep Ligand&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Sheep Receptor Ensembl ID to ensure all values are shown\"&gt;Sheep Receptor Ensembl ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Sheep Receptor to ensure all values are shown\"&gt;Sheep Receptor&amp;nbsp;&lt;/span&gt;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12194978</td>\n",
       "      <td>extras from chatGPT on connectomeDB2020</td>\n",
       "      <td>The role of Grp 78 in alpha 2-macroglobulin-in...</td>\n",
       "      <td>The Journal of biological chemistry</td>\n",
       "      <td>2002</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CDB0001</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>connectomeDB2020 x GPT,connectomeDB2025 ðŸ†•</td>\n",
       "      <td>&lt;span title=\"alpha-2-macroglobulin\"&gt;A2M&lt;/span&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSCAFG00845009518</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>ENSECAG00000012558</td>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSECAG00000024205</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>ENSOARG00020005617</td>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSOARG00020007714</td>\n",
       "      <td>HSPA5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>32541810</td>\n",
       "      <td>ConnectomeDB2025 (this publication)</td>\n",
       "      <td>Activated Î±</td>\n",
       "      <td>Scientific reports</td>\n",
       "      <td>2020</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CDB0001</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>connectomeDB2020 x GPT,connectomeDB2025 ðŸ†•</td>\n",
       "      <td>&lt;span title=\"alpha-2-macroglobulin\"&gt;A2M&lt;/span&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSCAFG00845009518</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>ENSECAG00000012558</td>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSECAG00000024205</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>ENSOARG00020005617</td>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSOARG00020007714</td>\n",
       "      <td>HSPA5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16239146</td>\n",
       "      <td>ConnectomeDB2025 (this publication)</td>\n",
       "      <td>Adam meets Eph: an ADAM substrate recognition ...</td>\n",
       "      <td>Cell</td>\n",
       "      <td>2005</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CDB0006</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>connectomeDB2025 ðŸ†•</td>\n",
       "      <td>&lt;span title=\"ADAM metallopeptidase domain 10\"&gt;...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSCAFG00845006153</td>\n",
       "      <td>EPHA5</td>\n",
       "      <td>ENSECAG00000007963</td>\n",
       "      <td>ADAM10</td>\n",
       "      <td>ENSECAG00000000530</td>\n",
       "      <td>EPHA5</td>\n",
       "      <td>ENSOARG00020021312</td>\n",
       "      <td>ADAM10</td>\n",
       "      <td>ENSOARG00020000230</td>\n",
       "      <td>EPHA5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10944520</td>\n",
       "      <td>Ramilowski_2015_Literature_supported</td>\n",
       "      <td>RGD-independent binding of integrin alpha9beta...</td>\n",
       "      <td>The Journal of biological chemistry</td>\n",
       "      <td>2000</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CDB0009</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>connectomeDB2015</td>\n",
       "      <td>&lt;span title=\"ADAM metallopeptidase domain 12\"&gt;...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSCAFG00845002947</td>\n",
       "      <td>ITGB1</td>\n",
       "      <td>ENSECAG00000012744</td>\n",
       "      <td>ADAM12</td>\n",
       "      <td>ENSECAG00000022498</td>\n",
       "      <td>ITGB1</td>\n",
       "      <td>ENSOARG00020023949</td>\n",
       "      <td>ADAM12</td>\n",
       "      <td>ENSOARG00020022585</td>\n",
       "      <td>ITGB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9914169</td>\n",
       "      <td>extras from chatGPT on connectomeDB2020</td>\n",
       "      <td>Interaction of metargidin (ADAM-15) with alpha...</td>\n",
       "      <td>Journal of cell science</td>\n",
       "      <td>1999</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CDB0011</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>connectomeDB2020 x GPT</td>\n",
       "      <td>&lt;span title=\"ADAM metallopeptidase domain 15\"&gt;...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSCAFG00845016791</td>\n",
       "      <td>ITGA5</td>\n",
       "      <td>ENSECAG00000008411</td>\n",
       "      <td>ADAM15</td>\n",
       "      <td>ENSECAG00000024979</td>\n",
       "      <td>ITGA5</td>\n",
       "      <td>ENSOARG00020007047</td>\n",
       "      <td>ADAM15</td>\n",
       "      <td>ENSOARG00020016394</td>\n",
       "      <td>ITGA5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5002</th>\n",
       "      <td>25171406</td>\n",
       "      <td>ConnectomeDB2025 (this publication)</td>\n",
       "      <td>Single-cell identity generated by combinatoria...</td>\n",
       "      <td>Cell</td>\n",
       "      <td>2014</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CDB2563</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>connectomeDB2025 ðŸ†•</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhb20&lt;/span&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5003</th>\n",
       "      <td>25171406</td>\n",
       "      <td>ConnectomeDB2025 (this publication)</td>\n",
       "      <td>Single-cell identity generated by combinatoria...</td>\n",
       "      <td>Cell</td>\n",
       "      <td>2014</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CDB2564</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>connectomeDB2025 ðŸ†•</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhb21&lt;/span&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5004</th>\n",
       "      <td>25171406</td>\n",
       "      <td>ConnectomeDB2025 (this publication)</td>\n",
       "      <td>Single-cell identity generated by combinatoria...</td>\n",
       "      <td>Cell</td>\n",
       "      <td>2014</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CDB2565</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>connectomeDB2025 ðŸ†•</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhb22&lt;/span&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5005</th>\n",
       "      <td>25171406</td>\n",
       "      <td>ConnectomeDB2025 (this publication)</td>\n",
       "      <td>Single-cell identity generated by combinatoria...</td>\n",
       "      <td>Cell</td>\n",
       "      <td>2014</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CDB2566</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>connectomeDB2025 ðŸ†•</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhgb8&lt;/span&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5006</th>\n",
       "      <td>27647835</td>\n",
       "      <td>ConnectomeDB2025 (this publication)</td>\n",
       "      <td>Scavenger Receptor SCARA5 Acts as an HMGB1 Rec...</td>\n",
       "      <td>Journal of immunology</td>\n",
       "      <td>2016</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CDB1533</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>connectomeDB2025 ðŸ†•</td>\n",
       "      <td>&lt;span title=\"high mobility group box 1\"&gt;HMGB1&lt;...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSCAFG00845009520</td>\n",
       "      <td>SCARA5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSECAG00000013893</td>\n",
       "      <td>SCARA5</td>\n",
       "      <td>ENSOARG00020020082</td>\n",
       "      <td></td>\n",
       "      <td>ENSOARG00020010482</td>\n",
       "      <td>SCARA5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5007 rows Ã— 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          PMID                          original source  \\\n",
       "0     12194978  extras from chatGPT on connectomeDB2020   \n",
       "1     32541810      ConnectomeDB2025 (this publication)   \n",
       "2     16239146      ConnectomeDB2025 (this publication)   \n",
       "3     10944520     Ramilowski_2015_Literature_supported   \n",
       "4      9914169  extras from chatGPT on connectomeDB2020   \n",
       "...        ...                                      ...   \n",
       "5002  25171406      ConnectomeDB2025 (this publication)   \n",
       "5003  25171406      ConnectomeDB2025 (this publication)   \n",
       "5004  25171406      ConnectomeDB2025 (this publication)   \n",
       "5005  25171406      ConnectomeDB2025 (this publication)   \n",
       "5006  27647835      ConnectomeDB2025 (this publication)   \n",
       "\n",
       "                                                  Title  \\\n",
       "0     The role of Grp 78 in alpha 2-macroglobulin-in...   \n",
       "1                                           Activated Î±   \n",
       "2     Adam meets Eph: an ADAM substrate recognition ...   \n",
       "3     RGD-independent binding of integrin alpha9beta...   \n",
       "4     Interaction of metargidin (ADAM-15) with alpha...   \n",
       "...                                                 ...   \n",
       "5002  Single-cell identity generated by combinatoria...   \n",
       "5003  Single-cell identity generated by combinatoria...   \n",
       "5004  Single-cell identity generated by combinatoria...   \n",
       "5005  Single-cell identity generated by combinatoria...   \n",
       "5006  Scavenger Receptor SCARA5 Acts as an HMGB1 Rec...   \n",
       "\n",
       "                                  Journal  Year       Species  \\\n",
       "0     The Journal of biological chemistry  2002  Homo sapiens   \n",
       "1                      Scientific reports  2020  Homo sapiens   \n",
       "2                                    Cell  2005  Homo sapiens   \n",
       "3     The Journal of biological chemistry  2000  Homo sapiens   \n",
       "4                 Journal of cell science  1999  Homo sapiens   \n",
       "...                                   ...   ...           ...   \n",
       "5002                                 Cell  2014  Homo sapiens   \n",
       "5003                                 Cell  2014  Homo sapiens   \n",
       "5004                                 Cell  2014  Homo sapiens   \n",
       "5005                                 Cell  2014  Homo sapiens   \n",
       "5006                Journal of immunology  2016  Homo sapiens   \n",
       "\n",
       "     <span title=\"Double-click header of Interaction ID to ensure all values are shown\">Interaction ID&nbsp;</span>  \\\n",
       "0                                               CDB0001                                                               \n",
       "1                                               CDB0001                                                               \n",
       "2                                               CDB0006                                                               \n",
       "3                                               CDB0009                                                               \n",
       "4                                               CDB0011                                                               \n",
       "...                                                 ...                                                               \n",
       "5002                                            CDB2563                                                               \n",
       "5003                                            CDB2564                                                               \n",
       "5004                                            CDB2565                                                               \n",
       "5005                                            CDB2566                                                               \n",
       "5006                                            CDB1533                                                               \n",
       "\n",
       "     <span title=\" Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\">Human LR Pair</span>  \\\n",
       "0     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                                \n",
       "1     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                                \n",
       "2     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                                \n",
       "3     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                                \n",
       "4     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                                \n",
       "...                                                 ...                                                                \n",
       "5002  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                                \n",
       "5003  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                                \n",
       "5004  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                                \n",
       "5005  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                                \n",
       "5006  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                                \n",
       "\n",
       "     <span title=\"Double-click header of Database Source to ensure all values are shown\">Database Source&nbsp;</span>  \\\n",
       "0             connectomeDB2020 x GPT,connectomeDB2025 ðŸ†•                                                                 \n",
       "1             connectomeDB2020 x GPT,connectomeDB2025 ðŸ†•                                                                 \n",
       "2                                    connectomeDB2025 ðŸ†•                                                                 \n",
       "3                                      connectomeDB2015                                                                 \n",
       "4                                connectomeDB2020 x GPT                                                                 \n",
       "...                                                 ...                                                                 \n",
       "5002                                 connectomeDB2025 ðŸ†•                                                                 \n",
       "5003                                 connectomeDB2025 ðŸ†•                                                                 \n",
       "5004                                 connectomeDB2025 ðŸ†•                                                                 \n",
       "5005                                 connectomeDB2025 ðŸ†•                                                                 \n",
       "5006                                 connectomeDB2025 ðŸ†•                                                                 \n",
       "\n",
       "     <span title=\" Official Gene Symbol; Hover on symbols below to show gene names\">Ligand&nbsp;&nbsp;&nbsp;</span>  \\\n",
       "0        <span title=\"alpha-2-macroglobulin\">A2M</span>                                                               \n",
       "1        <span title=\"alpha-2-macroglobulin\">A2M</span>                                                               \n",
       "2     <span title=\"ADAM metallopeptidase domain 10\">...                                                               \n",
       "3     <span title=\"ADAM metallopeptidase domain 12\">...                                                               \n",
       "4     <span title=\"ADAM metallopeptidase domain 15\">...                                                               \n",
       "...                                                 ...                                                               \n",
       "5002                     <span title=\" \">Pcdhb20</span>                                                               \n",
       "5003                     <span title=\" \">Pcdhb21</span>                                                               \n",
       "5004                     <span title=\" \">Pcdhb22</span>                                                               \n",
       "5005                     <span title=\" \">Pcdhgb8</span>                                                               \n",
       "5006  <span title=\"high mobility group box 1\">HMGB1<...                                                               \n",
       "\n",
       "      ...  \\\n",
       "0     ...   \n",
       "1     ...   \n",
       "2     ...   \n",
       "3     ...   \n",
       "4     ...   \n",
       "...   ...   \n",
       "5002  ...   \n",
       "5003  ...   \n",
       "5004  ...   \n",
       "5005  ...   \n",
       "5006  ...   \n",
       "\n",
       "     <span title=\"Double-click header of Dog Receptor Ensembl ID to ensure all values are shown\">Dog Receptor Ensembl ID&nbsp;</span>  \\\n",
       "0                                    ENSCAFG00845009518                                                                                 \n",
       "1                                    ENSCAFG00845009518                                                                                 \n",
       "2                                    ENSCAFG00845006153                                                                                 \n",
       "3                                    ENSCAFG00845002947                                                                                 \n",
       "4                                    ENSCAFG00845016791                                                                                 \n",
       "...                                                 ...                                                                                 \n",
       "5002                                                                                                                                    \n",
       "5003                                                                                                                                    \n",
       "5004                                                                                                                                    \n",
       "5005                                                                                                                                    \n",
       "5006                                 ENSCAFG00845009520                                                                                 \n",
       "\n",
       "     <span title=\"Double-click header of Dog Receptor to ensure all values are shown\">Dog Receptor&nbsp;</span>  \\\n",
       "0                                                 HSPA5                                                           \n",
       "1                                                 HSPA5                                                           \n",
       "2                                                 EPHA5                                                           \n",
       "3                                                 ITGB1                                                           \n",
       "4                                                 ITGA5                                                           \n",
       "...                                                 ...                                                           \n",
       "5002                                                                                                              \n",
       "5003                                                                                                              \n",
       "5004                                                                                                              \n",
       "5005                                                                                                              \n",
       "5006                                             SCARA5                                                           \n",
       "\n",
       "     <span title=\"Double-click header of Horse Ligand Ensembl ID to ensure all values are shown\">Horse Ligand Ensembl ID&nbsp;</span>  \\\n",
       "0                                    ENSECAG00000012558                                                                                 \n",
       "1                                    ENSECAG00000012558                                                                                 \n",
       "2                                    ENSECAG00000007963                                                                                 \n",
       "3                                    ENSECAG00000012744                                                                                 \n",
       "4                                    ENSECAG00000008411                                                                                 \n",
       "...                                                 ...                                                                                 \n",
       "5002                                                                                                                                    \n",
       "5003                                                                                                                                    \n",
       "5004                                                                                                                                    \n",
       "5005                                                                                                                                    \n",
       "5006                                                                                                                                    \n",
       "\n",
       "     <span title=\"Double-click header of Horse Ligand to ensure all values are shown\">Horse Ligand&nbsp;</span>  \\\n",
       "0                                                   A2M                                                           \n",
       "1                                                   A2M                                                           \n",
       "2                                                ADAM10                                                           \n",
       "3                                                ADAM12                                                           \n",
       "4                                                ADAM15                                                           \n",
       "...                                                 ...                                                           \n",
       "5002                                                                                                              \n",
       "5003                                                                                                              \n",
       "5004                                                                                                              \n",
       "5005                                                                                                              \n",
       "5006                                                                                                              \n",
       "\n",
       "     <span title=\"Double-click header of Horse Receptor Ensembl ID to ensure all values are shown\">Horse Receptor Ensembl ID&nbsp;</span>  \\\n",
       "0                                    ENSECAG00000024205                                                                                     \n",
       "1                                    ENSECAG00000024205                                                                                     \n",
       "2                                    ENSECAG00000000530                                                                                     \n",
       "3                                    ENSECAG00000022498                                                                                     \n",
       "4                                    ENSECAG00000024979                                                                                     \n",
       "...                                                 ...                                                                                     \n",
       "5002                                                                                                                                        \n",
       "5003                                                                                                                                        \n",
       "5004                                                                                                                                        \n",
       "5005                                                                                                                                        \n",
       "5006                                 ENSECAG00000013893                                                                                     \n",
       "\n",
       "     <span title=\"Double-click header of Horse Receptor to ensure all values are shown\">Horse Receptor&nbsp;</span>  \\\n",
       "0                                                 HSPA5                                                               \n",
       "1                                                 HSPA5                                                               \n",
       "2                                                 EPHA5                                                               \n",
       "3                                                 ITGB1                                                               \n",
       "4                                                 ITGA5                                                               \n",
       "...                                                 ...                                                               \n",
       "5002                                                                                                                  \n",
       "5003                                                                                                                  \n",
       "5004                                                                                                                  \n",
       "5005                                                                                                                  \n",
       "5006                                             SCARA5                                                               \n",
       "\n",
       "     <span title=\"Double-click header of Sheep Ligand Ensembl ID to ensure all values are shown\">Sheep Ligand Ensembl ID&nbsp;</span>  \\\n",
       "0                                    ENSOARG00020005617                                                                                 \n",
       "1                                    ENSOARG00020005617                                                                                 \n",
       "2                                    ENSOARG00020021312                                                                                 \n",
       "3                                    ENSOARG00020023949                                                                                 \n",
       "4                                    ENSOARG00020007047                                                                                 \n",
       "...                                                 ...                                                                                 \n",
       "5002                                                                                                                                    \n",
       "5003                                                                                                                                    \n",
       "5004                                                                                                                                    \n",
       "5005                                                                                                                                    \n",
       "5006                                 ENSOARG00020020082                                                                                 \n",
       "\n",
       "     <span title=\"Double-click header of Sheep Ligand to ensure all values are shown\">Sheep Ligand&nbsp;</span>  \\\n",
       "0                                                   A2M                                                           \n",
       "1                                                   A2M                                                           \n",
       "2                                                ADAM10                                                           \n",
       "3                                                ADAM12                                                           \n",
       "4                                                ADAM15                                                           \n",
       "...                                                 ...                                                           \n",
       "5002                                                                                                              \n",
       "5003                                                                                                              \n",
       "5004                                                                                                              \n",
       "5005                                                                                                              \n",
       "5006                                                                                                              \n",
       "\n",
       "     <span title=\"Double-click header of Sheep Receptor Ensembl ID to ensure all values are shown\">Sheep Receptor Ensembl ID&nbsp;</span>  \\\n",
       "0                                    ENSOARG00020007714                                                                                     \n",
       "1                                    ENSOARG00020007714                                                                                     \n",
       "2                                    ENSOARG00020000230                                                                                     \n",
       "3                                    ENSOARG00020022585                                                                                     \n",
       "4                                    ENSOARG00020016394                                                                                     \n",
       "...                                                 ...                                                                                     \n",
       "5002                                                                                                                                        \n",
       "5003                                                                                                                                        \n",
       "5004                                                                                                                                        \n",
       "5005                                                                                                                                        \n",
       "5006                                 ENSOARG00020010482                                                                                     \n",
       "\n",
       "     <span title=\"Double-click header of Sheep Receptor to ensure all values are shown\">Sheep Receptor&nbsp;</span>  \n",
       "0                                                 HSPA5                                                              \n",
       "1                                                 HSPA5                                                              \n",
       "2                                                 EPHA5                                                              \n",
       "3                                                 ITGB1                                                              \n",
       "4                                                 ITGA5                                                              \n",
       "...                                                 ...                                                              \n",
       "5002                                                                                                                 \n",
       "5003                                                                                                                 \n",
       "5004                                                                                                                 \n",
       "5005                                                                                                                 \n",
       "5006                                             SCARA5                                                              \n",
       "\n",
       "[5007 rows x 64 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Function to prepare functional annotation datatable\n",
    "import sys, os\n",
    "# Change working directory to ConnectomeDB\n",
    "project_root = \"/Users/sakuramaezono/Library/CloudStorage/OneDrive-YokohamaCityUniversity/Personal/05_Python_repositories/ConnectomeDB\"\n",
    "os.chdir(project_root)\n",
    "sys.path.append(os.path.abspath(\"src\"))\n",
    "from itables import init_notebook_mode\n",
    "import pandas as pd\n",
    "from itables import show\n",
    "from itables import options\n",
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "from bs4 import BeautifulSoup\n",
    "from createDataTable import gene_pair, gene_pair0\n",
    "import warnings\n",
    "import fetchGSheet \n",
    "\n",
    "mapping_ID = dict(zip(gene_pair0['Human LR Pair'], gene_pair0['Interaction ID']))\n",
    "gene_pair_PMID = fetchGSheet.gene_pair.dropna(axis=1, how='all')\n",
    "gene_pair_PMID= gene_pair_PMID[[\"LR pair\", \"PMID\", \"original source\"]]\n",
    "# Replace values in the column based on the mapping\n",
    "gene_pair_PMID[\"Interaction ID\"] = gene_pair_PMID['LR pair'].replace(mapping_ID)\n",
    "df_pub = pd.read_csv(\"data/pubmed_results.csv\", usecols=[0,1,3,4,5])\n",
    "gene_pair_PMID[\"PMID\"] = gene_pair_PMID[\"PMID\"].astype(str)\n",
    "df_pub[\"PMID\"] = df_pub[\"PMID\"].astype(str)\n",
    "gene_pair_trip = pd.merge(gene_pair_PMID, df_pub, how='left', on='PMID')\n",
    "gene_pair_trip[\"Year\"] = pd.to_numeric(gene_pair_trip[\"Year\"], errors=\"coerce\").astype(\"Int64\")\n",
    "gene_pair_trip = gene_pair_trip.merge(gene_pair, how='left', left_on='Interaction ID', right_on=gene_pair.columns[0])\n",
    "gene_pair_trip = gene_pair_trip.drop(columns=[\"Interaction ID\", \"LR pair\"])\n",
    "gene_pair_trip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "dc2b28ea-8da1-4e82-9a99-8b716f87cd11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'<span title=\"Double-click header of Database Source to ensure all values are shown\">Database Source&nbsp;</span>'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair.columns[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f184143d-7411-4c08-a24e-e290711ec397",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LR pair</th>\n",
       "      <th>PMID</th>\n",
       "      <th>Interaction ID</th>\n",
       "      <th>Title</th>\n",
       "      <th>Journal</th>\n",
       "      <th>Year</th>\n",
       "      <th>Species</th>\n",
       "      <th>&lt;span title=\"Double-click header of Interaction ID to ensure all values are shown\"&gt;Interaction ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\" Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\"&gt;Human LR Pair&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Database Source to ensure all values are shown\"&gt;Database Source&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>...</th>\n",
       "      <th>&lt;span title=\"Double-click header of Dog Receptor Ensembl ID to ensure all values are shown\"&gt;Dog Receptor Ensembl ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Dog Receptor to ensure all values are shown\"&gt;Dog Receptor&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Horse Ligand Ensembl ID to ensure all values are shown\"&gt;Horse Ligand Ensembl ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Horse Ligand to ensure all values are shown\"&gt;Horse Ligand&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Horse Receptor Ensembl ID to ensure all values are shown\"&gt;Horse Receptor Ensembl ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Horse Receptor to ensure all values are shown\"&gt;Horse Receptor&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Sheep Ligand Ensembl ID to ensure all values are shown\"&gt;Sheep Ligand Ensembl ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Sheep Ligand to ensure all values are shown\"&gt;Sheep Ligand&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Sheep Receptor Ensembl ID to ensure all values are shown\"&gt;Sheep Receptor Ensembl ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Sheep Receptor to ensure all values are shown\"&gt;Sheep Receptor&amp;nbsp;&lt;/span&gt;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>12194978</td>\n",
       "      <td>CDB0001</td>\n",
       "      <td>The role of Grp 78 in alpha 2-macroglobulin-in...</td>\n",
       "      <td>The Journal of biological chemistry</td>\n",
       "      <td>2002</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CDB0001</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>connectomeDB2020 x GPT,connectomeDB2025 ðŸ†•</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSCAFG00845009518</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>ENSECAG00000012558</td>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSECAG00000024205</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>ENSOARG00020005617</td>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSOARG00020007714</td>\n",
       "      <td>HSPA5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>32541810</td>\n",
       "      <td>CDB0001</td>\n",
       "      <td>Activated Î±</td>\n",
       "      <td>Scientific reports</td>\n",
       "      <td>2020</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CDB0001</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>connectomeDB2020 x GPT,connectomeDB2025 ðŸ†•</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSCAFG00845009518</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>ENSECAG00000012558</td>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSECAG00000024205</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>ENSOARG00020005617</td>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSOARG00020007714</td>\n",
       "      <td>HSPA5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADAM10 EPHA5</td>\n",
       "      <td>16239146</td>\n",
       "      <td>CDB0006</td>\n",
       "      <td>Adam meets Eph: an ADAM substrate recognition ...</td>\n",
       "      <td>Cell</td>\n",
       "      <td>2005</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CDB0006</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>connectomeDB2025 ðŸ†•</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSCAFG00845006153</td>\n",
       "      <td>EPHA5</td>\n",
       "      <td>ENSECAG00000007963</td>\n",
       "      <td>ADAM10</td>\n",
       "      <td>ENSECAG00000000530</td>\n",
       "      <td>EPHA5</td>\n",
       "      <td>ENSOARG00020021312</td>\n",
       "      <td>ADAM10</td>\n",
       "      <td>ENSOARG00020000230</td>\n",
       "      <td>EPHA5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADAM12 ITGB1</td>\n",
       "      <td>10944520</td>\n",
       "      <td>CDB0009</td>\n",
       "      <td>RGD-independent binding of integrin alpha9beta...</td>\n",
       "      <td>The Journal of biological chemistry</td>\n",
       "      <td>2000</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CDB0009</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>connectomeDB2015</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSCAFG00845002947</td>\n",
       "      <td>ITGB1</td>\n",
       "      <td>ENSECAG00000012744</td>\n",
       "      <td>ADAM12</td>\n",
       "      <td>ENSECAG00000022498</td>\n",
       "      <td>ITGB1</td>\n",
       "      <td>ENSOARG00020023949</td>\n",
       "      <td>ADAM12</td>\n",
       "      <td>ENSOARG00020022585</td>\n",
       "      <td>ITGB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADAM15 ITGA5</td>\n",
       "      <td>9914169</td>\n",
       "      <td>CDB0011</td>\n",
       "      <td>Interaction of metargidin (ADAM-15) with alpha...</td>\n",
       "      <td>Journal of cell science</td>\n",
       "      <td>1999</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CDB0011</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>connectomeDB2020 x GPT</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSCAFG00845016791</td>\n",
       "      <td>ITGA5</td>\n",
       "      <td>ENSECAG00000008411</td>\n",
       "      <td>ADAM15</td>\n",
       "      <td>ENSECAG00000024979</td>\n",
       "      <td>ITGA5</td>\n",
       "      <td>ENSOARG00020007047</td>\n",
       "      <td>ADAM15</td>\n",
       "      <td>ENSOARG00020016394</td>\n",
       "      <td>ITGA5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5002</th>\n",
       "      <td>Pcdhb20 Pcdhb20</td>\n",
       "      <td>25171406</td>\n",
       "      <td>CDB2563</td>\n",
       "      <td>Single-cell identity generated by combinatoria...</td>\n",
       "      <td>Cell</td>\n",
       "      <td>2014</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CDB2563</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>connectomeDB2025 ðŸ†•</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5003</th>\n",
       "      <td>Pcdhb21 Pcdhb21</td>\n",
       "      <td>25171406</td>\n",
       "      <td>CDB2564</td>\n",
       "      <td>Single-cell identity generated by combinatoria...</td>\n",
       "      <td>Cell</td>\n",
       "      <td>2014</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CDB2564</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>connectomeDB2025 ðŸ†•</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5004</th>\n",
       "      <td>Pcdhb22 Pcdhb22</td>\n",
       "      <td>25171406</td>\n",
       "      <td>CDB2565</td>\n",
       "      <td>Single-cell identity generated by combinatoria...</td>\n",
       "      <td>Cell</td>\n",
       "      <td>2014</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CDB2565</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>connectomeDB2025 ðŸ†•</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5005</th>\n",
       "      <td>Pcdhgb8 Pcdhgb8</td>\n",
       "      <td>25171406</td>\n",
       "      <td>CDB2566</td>\n",
       "      <td>Single-cell identity generated by combinatoria...</td>\n",
       "      <td>Cell</td>\n",
       "      <td>2014</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CDB2566</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>connectomeDB2025 ðŸ†•</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5006</th>\n",
       "      <td>HMGB1 SCARA5</td>\n",
       "      <td>27647835</td>\n",
       "      <td>CDB1533</td>\n",
       "      <td>Scavenger Receptor SCARA5 Acts as an HMGB1 Rec...</td>\n",
       "      <td>Journal of immunology</td>\n",
       "      <td>2016</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CDB1533</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>connectomeDB2025 ðŸ†•</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSCAFG00845009520</td>\n",
       "      <td>SCARA5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSECAG00000013893</td>\n",
       "      <td>SCARA5</td>\n",
       "      <td>ENSOARG00020020082</td>\n",
       "      <td></td>\n",
       "      <td>ENSOARG00020010482</td>\n",
       "      <td>SCARA5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5007 rows Ã— 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              LR pair      PMID Interaction ID  \\\n",
       "0           A2M HSPA5  12194978        CDB0001   \n",
       "1           A2M HSPA5  32541810        CDB0001   \n",
       "2        ADAM10 EPHA5  16239146        CDB0006   \n",
       "3        ADAM12 ITGB1  10944520        CDB0009   \n",
       "4        ADAM15 ITGA5   9914169        CDB0011   \n",
       "...               ...       ...            ...   \n",
       "5002  Pcdhb20 Pcdhb20  25171406        CDB2563   \n",
       "5003  Pcdhb21 Pcdhb21  25171406        CDB2564   \n",
       "5004  Pcdhb22 Pcdhb22  25171406        CDB2565   \n",
       "5005  Pcdhgb8 Pcdhgb8  25171406        CDB2566   \n",
       "5006     HMGB1 SCARA5  27647835        CDB1533   \n",
       "\n",
       "                                                  Title  \\\n",
       "0     The role of Grp 78 in alpha 2-macroglobulin-in...   \n",
       "1                                           Activated Î±   \n",
       "2     Adam meets Eph: an ADAM substrate recognition ...   \n",
       "3     RGD-independent binding of integrin alpha9beta...   \n",
       "4     Interaction of metargidin (ADAM-15) with alpha...   \n",
       "...                                                 ...   \n",
       "5002  Single-cell identity generated by combinatoria...   \n",
       "5003  Single-cell identity generated by combinatoria...   \n",
       "5004  Single-cell identity generated by combinatoria...   \n",
       "5005  Single-cell identity generated by combinatoria...   \n",
       "5006  Scavenger Receptor SCARA5 Acts as an HMGB1 Rec...   \n",
       "\n",
       "                                  Journal  Year       Species  \\\n",
       "0     The Journal of biological chemistry  2002  Homo sapiens   \n",
       "1                      Scientific reports  2020  Homo sapiens   \n",
       "2                                    Cell  2005  Homo sapiens   \n",
       "3     The Journal of biological chemistry  2000  Homo sapiens   \n",
       "4                 Journal of cell science  1999  Homo sapiens   \n",
       "...                                   ...   ...           ...   \n",
       "5002                                 Cell  2014  Homo sapiens   \n",
       "5003                                 Cell  2014  Homo sapiens   \n",
       "5004                                 Cell  2014  Homo sapiens   \n",
       "5005                                 Cell  2014  Homo sapiens   \n",
       "5006                Journal of immunology  2016  Homo sapiens   \n",
       "\n",
       "     <span title=\"Double-click header of Interaction ID to ensure all values are shown\">Interaction ID&nbsp;</span>  \\\n",
       "0                                               CDB0001                                                               \n",
       "1                                               CDB0001                                                               \n",
       "2                                               CDB0006                                                               \n",
       "3                                               CDB0009                                                               \n",
       "4                                               CDB0011                                                               \n",
       "...                                                 ...                                                               \n",
       "5002                                            CDB2563                                                               \n",
       "5003                                            CDB2564                                                               \n",
       "5004                                            CDB2565                                                               \n",
       "5005                                            CDB2566                                                               \n",
       "5006                                            CDB1533                                                               \n",
       "\n",
       "     <span title=\" Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\">Human LR Pair</span>  \\\n",
       "0     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                                \n",
       "1     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                                \n",
       "2     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                                \n",
       "3     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                                \n",
       "4     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                                \n",
       "...                                                 ...                                                                \n",
       "5002  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                                \n",
       "5003  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                                \n",
       "5004  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                                \n",
       "5005  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                                \n",
       "5006  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                                \n",
       "\n",
       "     <span title=\"Double-click header of Database Source to ensure all values are shown\">Database Source&nbsp;</span>  \\\n",
       "0             connectomeDB2020 x GPT,connectomeDB2025 ðŸ†•                                                                 \n",
       "1             connectomeDB2020 x GPT,connectomeDB2025 ðŸ†•                                                                 \n",
       "2                                    connectomeDB2025 ðŸ†•                                                                 \n",
       "3                                      connectomeDB2015                                                                 \n",
       "4                                connectomeDB2020 x GPT                                                                 \n",
       "...                                                 ...                                                                 \n",
       "5002                                 connectomeDB2025 ðŸ†•                                                                 \n",
       "5003                                 connectomeDB2025 ðŸ†•                                                                 \n",
       "5004                                 connectomeDB2025 ðŸ†•                                                                 \n",
       "5005                                 connectomeDB2025 ðŸ†•                                                                 \n",
       "5006                                 connectomeDB2025 ðŸ†•                                                                 \n",
       "\n",
       "      ...  \\\n",
       "0     ...   \n",
       "1     ...   \n",
       "2     ...   \n",
       "3     ...   \n",
       "4     ...   \n",
       "...   ...   \n",
       "5002  ...   \n",
       "5003  ...   \n",
       "5004  ...   \n",
       "5005  ...   \n",
       "5006  ...   \n",
       "\n",
       "     <span title=\"Double-click header of Dog Receptor Ensembl ID to ensure all values are shown\">Dog Receptor Ensembl ID&nbsp;</span>  \\\n",
       "0                                    ENSCAFG00845009518                                                                                 \n",
       "1                                    ENSCAFG00845009518                                                                                 \n",
       "2                                    ENSCAFG00845006153                                                                                 \n",
       "3                                    ENSCAFG00845002947                                                                                 \n",
       "4                                    ENSCAFG00845016791                                                                                 \n",
       "...                                                 ...                                                                                 \n",
       "5002                                                                                                                                    \n",
       "5003                                                                                                                                    \n",
       "5004                                                                                                                                    \n",
       "5005                                                                                                                                    \n",
       "5006                                 ENSCAFG00845009520                                                                                 \n",
       "\n",
       "     <span title=\"Double-click header of Dog Receptor to ensure all values are shown\">Dog Receptor&nbsp;</span>  \\\n",
       "0                                                 HSPA5                                                           \n",
       "1                                                 HSPA5                                                           \n",
       "2                                                 EPHA5                                                           \n",
       "3                                                 ITGB1                                                           \n",
       "4                                                 ITGA5                                                           \n",
       "...                                                 ...                                                           \n",
       "5002                                                                                                              \n",
       "5003                                                                                                              \n",
       "5004                                                                                                              \n",
       "5005                                                                                                              \n",
       "5006                                             SCARA5                                                           \n",
       "\n",
       "     <span title=\"Double-click header of Horse Ligand Ensembl ID to ensure all values are shown\">Horse Ligand Ensembl ID&nbsp;</span>  \\\n",
       "0                                    ENSECAG00000012558                                                                                 \n",
       "1                                    ENSECAG00000012558                                                                                 \n",
       "2                                    ENSECAG00000007963                                                                                 \n",
       "3                                    ENSECAG00000012744                                                                                 \n",
       "4                                    ENSECAG00000008411                                                                                 \n",
       "...                                                 ...                                                                                 \n",
       "5002                                                                                                                                    \n",
       "5003                                                                                                                                    \n",
       "5004                                                                                                                                    \n",
       "5005                                                                                                                                    \n",
       "5006                                                                                                                                    \n",
       "\n",
       "     <span title=\"Double-click header of Horse Ligand to ensure all values are shown\">Horse Ligand&nbsp;</span>  \\\n",
       "0                                                   A2M                                                           \n",
       "1                                                   A2M                                                           \n",
       "2                                                ADAM10                                                           \n",
       "3                                                ADAM12                                                           \n",
       "4                                                ADAM15                                                           \n",
       "...                                                 ...                                                           \n",
       "5002                                                                                                              \n",
       "5003                                                                                                              \n",
       "5004                                                                                                              \n",
       "5005                                                                                                              \n",
       "5006                                                                                                              \n",
       "\n",
       "     <span title=\"Double-click header of Horse Receptor Ensembl ID to ensure all values are shown\">Horse Receptor Ensembl ID&nbsp;</span>  \\\n",
       "0                                    ENSECAG00000024205                                                                                     \n",
       "1                                    ENSECAG00000024205                                                                                     \n",
       "2                                    ENSECAG00000000530                                                                                     \n",
       "3                                    ENSECAG00000022498                                                                                     \n",
       "4                                    ENSECAG00000024979                                                                                     \n",
       "...                                                 ...                                                                                     \n",
       "5002                                                                                                                                        \n",
       "5003                                                                                                                                        \n",
       "5004                                                                                                                                        \n",
       "5005                                                                                                                                        \n",
       "5006                                 ENSECAG00000013893                                                                                     \n",
       "\n",
       "     <span title=\"Double-click header of Horse Receptor to ensure all values are shown\">Horse Receptor&nbsp;</span>  \\\n",
       "0                                                 HSPA5                                                               \n",
       "1                                                 HSPA5                                                               \n",
       "2                                                 EPHA5                                                               \n",
       "3                                                 ITGB1                                                               \n",
       "4                                                 ITGA5                                                               \n",
       "...                                                 ...                                                               \n",
       "5002                                                                                                                  \n",
       "5003                                                                                                                  \n",
       "5004                                                                                                                  \n",
       "5005                                                                                                                  \n",
       "5006                                             SCARA5                                                               \n",
       "\n",
       "     <span title=\"Double-click header of Sheep Ligand Ensembl ID to ensure all values are shown\">Sheep Ligand Ensembl ID&nbsp;</span>  \\\n",
       "0                                    ENSOARG00020005617                                                                                 \n",
       "1                                    ENSOARG00020005617                                                                                 \n",
       "2                                    ENSOARG00020021312                                                                                 \n",
       "3                                    ENSOARG00020023949                                                                                 \n",
       "4                                    ENSOARG00020007047                                                                                 \n",
       "...                                                 ...                                                                                 \n",
       "5002                                                                                                                                    \n",
       "5003                                                                                                                                    \n",
       "5004                                                                                                                                    \n",
       "5005                                                                                                                                    \n",
       "5006                                 ENSOARG00020020082                                                                                 \n",
       "\n",
       "     <span title=\"Double-click header of Sheep Ligand to ensure all values are shown\">Sheep Ligand&nbsp;</span>  \\\n",
       "0                                                   A2M                                                           \n",
       "1                                                   A2M                                                           \n",
       "2                                                ADAM10                                                           \n",
       "3                                                ADAM12                                                           \n",
       "4                                                ADAM15                                                           \n",
       "...                                                 ...                                                           \n",
       "5002                                                                                                              \n",
       "5003                                                                                                              \n",
       "5004                                                                                                              \n",
       "5005                                                                                                              \n",
       "5006                                                                                                              \n",
       "\n",
       "     <span title=\"Double-click header of Sheep Receptor Ensembl ID to ensure all values are shown\">Sheep Receptor Ensembl ID&nbsp;</span>  \\\n",
       "0                                    ENSOARG00020007714                                                                                     \n",
       "1                                    ENSOARG00020007714                                                                                     \n",
       "2                                    ENSOARG00020000230                                                                                     \n",
       "3                                    ENSOARG00020022585                                                                                     \n",
       "4                                    ENSOARG00020016394                                                                                     \n",
       "...                                                 ...                                                                                     \n",
       "5002                                                                                                                                        \n",
       "5003                                                                                                                                        \n",
       "5004                                                                                                                                        \n",
       "5005                                                                                                                                        \n",
       "5006                                 ENSOARG00020010482                                                                                     \n",
       "\n",
       "     <span title=\"Double-click header of Sheep Receptor to ensure all values are shown\">Sheep Receptor&nbsp;</span>  \n",
       "0                                                 HSPA5                                                              \n",
       "1                                                 HSPA5                                                              \n",
       "2                                                 EPHA5                                                              \n",
       "3                                                 ITGB1                                                              \n",
       "4                                                 ITGA5                                                              \n",
       "...                                                 ...                                                              \n",
       "5002                                                                                                                 \n",
       "5003                                                                                                                 \n",
       "5004                                                                                                                 \n",
       "5005                                                                                                                 \n",
       "5006                                             SCARA5                                                              \n",
       "\n",
       "[5007 rows x 65 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6836abc-dbf3-4e60-94f7-73d1993ef857",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_annot = gene_pair0[[\"Human LR Pair\", \"Cancer-related\", \"Top Pathway\"]]\n",
    "df= pd.read_csv(\"data/disease_annotations_per_pair.csv\") # Liana Diseases\n",
    "gene_pair_annot = gene_pair_annot.merge(df, how='left', left_on='Human LR Pair', right_on='interaction')\n",
    "gene_pair_annot = gene_pair_annot.drop(columns=[\"interaction\"])\n",
    "df= pd.read_csv(\"data/pathway_annotations_per_pair.csv\") # Liana Pathway\n",
    "gene_pair_annot = gene_pair_annot.merge(df, how='left', left_on='Human LR Pair', right_on='interaction')\n",
    "gene_pair_annot = gene_pair_annot.drop(columns=[\"interaction\", \"weight\"])\n",
    "\n",
    "gene_pair_annot = gene_pair_annot.rename(columns={\n",
    "                                     \"disease\": \"Disease\", \n",
    "                                     \"source\": \"Related Pathway\"}\n",
    "                            )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a440a212-90a4-4c6a-9764-d1c4422f1da4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'<span title=\"Double-click header of Interaction ID to ensure all values are shown\">Interaction ID&nbsp;</span>'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e0ae701-0b2a-43ce-968a-680ae9efb382",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Main execution\n",
    "if __name__ == \"__main__\":\n",
    "    template = load_template(TEMPLATE_PATH)\n",
    "    interaction_card, ligand_card_1, receptor_card_1, ligand_card_2, receptor_card_2 = prepare_dataframes(gene_pair0)\n",
    "    generate_html_files(template, interaction_card, ligand_card_1, receptor_card_1, ligand_card_2, receptor_card_2, OUTPUT_DIR)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3c01bac-c9eb-4a90-982d-66df4c546c80",
   "metadata": {},
   "outputs": [],
   "source": [
    "fetchGSheet.gene_pair.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e43d24e7-09e2-4794-854e-ff099b87aad0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to prepare datatables (cleaning and hyperlinking, adding tool tips, etc) input for the database qmds\n",
    "import sys, os\n",
    "from itables import init_notebook_mode\n",
    "import pandas as pd\n",
    "from itables import show\n",
    "from itables import options\n",
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "import fetchGSheet \n",
    "import warnings\n",
    "\n",
    "# Suppress SettingWithCopyWarning\n",
    "warnings.simplefilter(\"ignore\", category=UserWarning)\n",
    "\n",
    "\n",
    "# Other vertebrates\n",
    "species_list = [\n",
    "    \"ptroglodytes\", \"ggallus\", \"sscrofa\", \"btaurus\", \n",
    "    \"clfamiliaris\", \"ecaballus\", \"oarambouillet\"\n",
    "]\n",
    "\n",
    "# Select only the relevant columns from pop_up_info\n",
    "pop_up_info = pd.read_table(\"data/HGNC_gene_info_full.tsv\")\n",
    "pop_up_info = pop_up_info.rename(columns={\"hgnc_id\": \"HGNC ID\", \n",
    "                                          \"name\": \"Approved name\",\n",
    "                                          \"symbol\": \"Approved symbol\",\n",
    "                                          \"rgd_id\": \"RGD ID\",\n",
    "                                          \"mgd_id\": \"MGI ID\", \n",
    "                                          \"rgd_id\": \"RGD ID\",\n",
    "                                          \"alias_symbol\": \"Alias symbol\",\n",
    "                                          \"prev_symbol\": \"Previous symbol\",\n",
    "                                          \"date_symbol_changed\": \"Date symbol changed\"\n",
    "                                          \n",
    "                                         })\n",
    "\n",
    "pop_up_info_lim = pop_up_info[[\"HGNC ID\", \"Approved name\", \"MGI ID\", \"RGD ID\"]] # rm \"Approved symbol\" for now\n",
    "pop_up_info_lim = pop_up_info_lim.drop_duplicates(subset=\"HGNC ID\", keep=\"first\")\n",
    "\n",
    "# Drop columns where all values are NA in gene_pair\n",
    "gene_pair = fetchGSheet.gene_pair.dropna(axis=1, how='all')\n",
    "\n",
    "# for now, rm some columns\n",
    "gene_pair = gene_pair[['LR pair', 'Ligand', 'Ligand.HGNC', 'Receptor', 'Receptor.HGNC',\n",
    "                       'perplexity link', 'PMID', 'binding location', \n",
    "                       'bind in trans?', 'bidirectional signalling?',\n",
    "                       'interaction type', 'original source']]\n",
    "\n",
    "# Mapping for replacements\n",
    "mapping = dict(zip(fetchGSheet.src_info['original source'], fetchGSheet.src_info['shortname']))\n",
    "# Replace values in the column based on the mapping\n",
    "gene_pair['original source'] = gene_pair['original source'].replace(mapping)\n",
    "\n",
    "## add Ligand/Receptor Location\n",
    "mapping_loc = dict(zip(fetchGSheet.loc_info['ApprovedSymbol'], fetchGSheet.loc_info['Localization']))\n",
    "gene_pair['Ligand location'] = gene_pair['Ligand'].replace(mapping_loc)\n",
    "gene_pair['Receptor location'] = gene_pair['Receptor'].replace(mapping_loc)\n",
    "\n",
    "# Fetch species IDs from the dataset\n",
    "hgnc_id = [col for col in gene_pair.columns if \"HGNC\" in col]\n",
    "hgnc_id = pd.concat([gene_pair[col] for col in hgnc_id]).unique()\n",
    "\n",
    "# Rename columns for better clarity\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "    \"LR pair\": \"Human LR Pair\",\n",
    "    \"Ligand.HGNC\": \"Ligand HGNC ID\",\n",
    "    \"Receptor.HGNC\": \"Receptor HGNC ID\",\n",
    "    \"perplexity link\": \"Perplexity\", # will be replaced with actual link later\n",
    "    \"original source\": \"Database Source\",\n",
    "    \"PMID\": \"PMID support\"\n",
    "})\n",
    "\n",
    "# Recreate Perplexity link\n",
    "# Function to generate Perplexity search link\n",
    "def create_url_basic(gene_name):\n",
    "    query = f\"What is the primary evidence that {gene_name} bind-each-other-as-a-ligand-and-receptor-pair. Exclude reviews, uniprot, wiki, genecards, PIPS, iuphar as sources.\"\n",
    "    encoded_query = query.replace(\" \", \"%20\")\n",
    "    return f\"https://www.perplexity.ai/search?q={encoded_query}\"\n",
    "\n",
    "# Apply function to the DataFrame\n",
    "gene_pair[\"Perplexity\"] = gene_pair[\"Perplexity\"].apply(create_url_basic)\n",
    "\n",
    "# Merge gene_pair with pop_up_info_lim for Ligand(L)\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Ligand HGNC ID', right_on='HGNC ID')\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Ligand name\", \n",
    "                                     \"MGI ID\": \"Ligand MGI ID\",\n",
    "                                     \"RGD ID\": \"Ligand RGD ID\"},\n",
    "                            )\n",
    "gene_pair = gene_pair.drop(columns=[\"HGNC ID\"])\n",
    "# Add top pathway per pair\n",
    "LR_pairs = gene_pair[\"Human LR Pair\"].unique()\n",
    "df= pd.read_csv(\"data/pathway_annotations_per_pair.csv\")\n",
    "df = df[df[\"interaction\"].isin(LR_pairs)]\n",
    "# Sort by absolute value of 'weight', descending (larger abs(weight) first)\n",
    "df_sorted = df.reindex(df['weight'].abs().sort_values(ascending=False).index)\n",
    "# Keep only the first occurrence for each unique 'interaction'\n",
    "df_unique = df_sorted.drop_duplicates(subset='interaction', keep='first')\n",
    "df = df_unique.reset_index(drop=True)\n",
    "top_pathway_df = df[[\"interaction\", \"source\"]]\n",
    "top_pathway_df = top_pathway_df.rename(columns={\n",
    "                                      \"source\": \"Top Pathway\"\n",
    "})\n",
    "gene_pair = gene_pair.merge(top_pathway_df, how='left', left_on='Human LR Pair', right_on='interaction')\n",
    "\n",
    "# Add Disease Category per pair\n",
    "df= pd.read_csv(\"data/diseaseType_per_pair.csv\")\n",
    "disease_df = df[df[\"interaction_x\"].isin(LR_pairs)]\n",
    "\n",
    "gene_pair = gene_pair.merge(disease_df, how='left', left_on='Human LR Pair', right_on='interaction_x')\n",
    "\n",
    "# Add MGI annotation\n",
    "MGI_info = pd.read_csv(\"data/MGI_ID_biomart.csv\")\n",
    "gene_pair = gene_pair.merge(MGI_info, how='left', left_on='Ligand MGI ID', right_on='MGI ID')\n",
    "\n",
    "# Find rows where Ligand HGNC ID is missing & copy Ligand to MGI name for those rows\n",
    "mask = gene_pair['Ligand HGNC ID'].astype(str).str.strip() == ''\n",
    "gene_pair.loc[mask, 'MGI name'] = gene_pair.loc[mask, 'Ligand']\n",
    "# Map MGI ID using the MGI_info table\n",
    "gene_pair = gene_pair.merge(MGI_info, left_on='MGI name', right_on='MGI name', how='left', suffixes=('', '_from_info'))\n",
    "# Fill missing 'MGI ID' only where it was previously missing\n",
    "gene_pair['Ligand MGI ID'] = gene_pair['Ligand MGI ID'].combine_first(gene_pair['MGI ID_from_info'])\n",
    "gene_pair = gene_pair.drop(columns=['MGI ID_from_info'])\n",
    "\n",
    "# Add RGD annotation\n",
    "RGD_info = pd.read_csv(\"data/RGD_ID_biomart.csv\")\n",
    "RGD_info['RGD ID'] = \"RGD:\" + RGD_info['RGD ID'].astype(str)\n",
    "gene_pair = gene_pair.merge(RGD_info, how='left', left_on='Ligand RGD ID', right_on='RGD ID')\n",
    "\n",
    "# Add ZFIN id and symbol\n",
    "ZFIN_info = pd.read_csv(\"data/ZFIN_ID_human_orthos.txt\", sep=\"\\t\", skiprows=1)\n",
    "ZFIN_info = ZFIN_info[['ZFIN ID', 'ZFIN Symbol', 'ZFIN Name', 'HGNC ID']]\n",
    "\n",
    "ZFIN_info = ZFIN_info.dropna(subset=['HGNC ID'])\n",
    "ZFIN_info = ZFIN_info.drop_duplicates(subset=['HGNC ID'])\n",
    "ZFIN_info['HGNC ID'] = ZFIN_info['HGNC ID'].apply(lambda x: f'HGNC:{int(x)}')\n",
    "gene_pair = gene_pair.merge(ZFIN_info, how='left', left_on='Ligand HGNC ID', right_on='HGNC ID')\n",
    "\n",
    "gene_pair = gene_pair.drop(columns=[\"RGD ID\", \"MGI ID\", \"HGNC ID\", \"interaction\", \"interaction_x\"])\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "                                     \"MGI name\": \"Mouse Ligand\", \n",
    "                                     \"RGD name\": \"Rat Ligand\",\n",
    "                                     \"ZFIN ID\": \"Ligand ZFIN ID\",\n",
    "                                     \"ZFIN Symbol\": \"Zebrafish Ligand\",\n",
    "                                     \"ZFIN Name\": \"Zebrafish Ligand name\"}\n",
    "                            )\n",
    "\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Receptor HGNC ID', right_on='HGNC ID')\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Receptor name\",\n",
    "                                      \"MGI ID\": \"Receptor MGI ID\",\n",
    "                                      \"RGD ID\": \"Receptor RGD ID\"}\n",
    "                            )\n",
    "\n",
    "\n",
    "gene_pair = gene_pair.drop(columns=[\"HGNC ID\"])\n",
    "\n",
    "# Add MGI name\n",
    "gene_pair = gene_pair.merge(MGI_info, how='left', left_on='Receptor MGI ID', right_on='MGI ID')\n",
    "# Find rows where Receptor HGNC ID is missing & copy Receptor to MGI name for those rows\n",
    "mask = gene_pair['Ligand HGNC ID'].astype(str).str.strip() == ''\n",
    "gene_pair.loc[mask, 'MGI name'] = gene_pair.loc[mask, 'Receptor']\n",
    "# Map MGI ID using the MGI_info table\n",
    "gene_pair = gene_pair.merge(MGI_info, left_on='MGI name', right_on='MGI name', how='left', suffixes=('', '_from_info'))\n",
    "# Fill missing 'MGI ID' only where it was previously missing\n",
    "gene_pair['Receptor MGI ID'] = gene_pair['Receptor MGI ID'].combine_first(gene_pair['MGI ID_from_info'])\n",
    "gene_pair = gene_pair.drop(columns=['MGI ID_from_info'])\n",
    "\n",
    "gene_pair = gene_pair.merge(RGD_info, how='left', left_on='Receptor RGD ID', right_on='RGD ID')\n",
    "gene_pair = gene_pair.merge(ZFIN_info, how='left', left_on='Receptor HGNC ID', right_on='HGNC ID')\n",
    "gene_pair = gene_pair.drop(columns=[\"RGD ID\", \"MGI ID\", \"HGNC ID\"])\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "                                     \"MGI name\": \"Mouse Receptor\", \n",
    "                                     \"RGD name\": \"Rat Receptor\",\n",
    "                                     \"ZFIN ID\": \"Receptor ZFIN ID\",\n",
    "                                     \"ZFIN Symbol\": \"Zebrafish Receptor\",\n",
    "                                     \"ZFIN Name\": \"Zebrafish Receptor name\"})\n",
    "\n",
    "gene_pair.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4fd1949-2d83-4a09-a623-0394e71575e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "LR_pairs = gene_pair[\"Human LR Pair\"].unique()\n",
    "df= pd.read_csv(\"data/pathway_annotations_per_pair.csv\")\n",
    "#df = df[df[\"interaction\"].isin(LR_pairs)]\n",
    "# Sort by absolute value of 'weight', descending (larger abs(weight) first)\n",
    "df_sorted = df.reindex(df['weight'].abs().sort_values(ascending=False).index)\n",
    "# Keep only the first occurrence for each unique 'interaction'\n",
    "df_unique = df_sorted.drop_duplicates(subset='interaction', keep='first')\n",
    "df = df_unique.reset_index(drop=True)\n",
    "top_pathway_df = df[[\"interaction\", \"source\"]]\n",
    "top_pathway_df = top_pathway_df.rename(columns={\n",
    "                                      \"source\": \"Top Pathway\"\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c424589-94c8-44e8-8e7f-8cd45ffabf40",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_pathway_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1052948f-e7c4-4201-a592-d2704480b56c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d0e4b4d-6505-4df1-a919-c8fc00ee65d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find rows where Ligand HGNC ID is missing & copy Ligand to MGI name for those rows\n",
    "mask = gene_pair['Ligand HGNC ID'].astype(str).str.strip() == ''\n",
    "gene_pair.loc[mask, 'Ligand MGI name'] = gene_pair.loc[mask, 'Ligand']\n",
    "mask\n",
    "gene_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e69bb7e7-0ffd-4abb-971a-7ca3a7eac5d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map MGI ID using the MGI_info table\n",
    "gene_pair = gene_pair.merge(MGI_info, left_on='Ligand MGI name', right_on='MGI name', how='left', suffixes=('', '_from_info'))\n",
    "gene_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1de2e65-1112-466d-ab06-85494acfd69e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fill missing 'MGI ID' only where it was previously missing\n",
    "gene_pair['Ligand MGI ID'] = gene_pair['Ligand MGI ID'].combine_first(gene_pair['MGI ID_from_info'])\n",
    "gene_pair = gene_pair.drop(columns=['MGI ID_from_info'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1653ce55-ef4e-4c1a-8fa5-4ccd26fde47f",
   "metadata": {},
   "outputs": [],
   "source": [
    "MGI_info[MGI_info[\"MGI name\"] == \"Pcdhb17\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6583164-0b40-4bcc-b1be-e9ba177a9792",
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59d93ccc-bc81-48ae-8c5c-89dcf74c46eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair[[\"PMID support\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "133caa02-bc7a-4f79-9b31-41b643c6fc7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add ZFIN id and symbol\n",
    "ZFIN_info = pd.read_csv(\"data/ZFIN_ID_human_orthos.txt\", sep=\"\\t\", skiprows=1)\n",
    "ZFIN_info = ZFIN_info[['ZFIN ID', 'ZFIN Symbol', 'ZFIN Name', 'HGNC ID']]\n",
    "\n",
    "ZFIN_info = ZFIN_info.dropna(subset=['HGNC ID'])\n",
    "ZFIN_info = ZFIN_info.drop_duplicates(subset=['HGNC ID'])\n",
    "ZFIN_info['HGNC ID'] = ZFIN_info['HGNC ID'].apply(lambda x: f'HGNC:{int(x)}')\n",
    "ZFIN_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb370708-2a65-4447-99d4-828bd34f4fa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair = gene_pair.merge(ZFIN_info, how='left', left_on='Ligand HGNC ID', right_on='HGNC ID')\n",
    "gene_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bca4aabf-e1da-475c-8adb-84903cc8bcae",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28cda5c0-e043-48e7-babc-dffe1a5ef1eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_up_info_lim = pop_up_info[[\"HGNC ID\", \"Approved symbol\", \"Approved name\", \"MGI ID\", \"RGD ID\"]]\n",
    "pop_up_info_lim = pop_up_info_lim.drop_duplicates(subset=\"HGNC ID\", keep=\"first\")\n",
    "pop_up_info_lim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "804d7f23-9cf9-4e52-b98c-9aecb9de59bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import pandas as pd\n",
    "from itables import init_notebook_mode, show\n",
    "from IPython.display import display, Javascript\n",
    "import itables.options as opt\n",
    "# Change working directory to ConnectomeDB\n",
    "project_root = os.path.dirname(os.getcwd())\n",
    "os.chdir(project_root)\n",
    "sys.path.append(os.path.abspath(\"src\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e79f620b-2f96-4ad1-9cb0-f9042626236b",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de98cadd-0785-445b-84b3-057295ca5c8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/Users/sakuramaezono/Library/CloudStorage/OneDrive-YokohamaCityUniversity/Personal/05_Python_repositories/ConnectomeDB')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c3330b8-0552-4b57-8264-fe447fe25deb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import liana as li\n",
    "import omnipath as op\n",
    "import decoupler as dc\n",
    "import pandas as pd\n",
    "\n",
    "import sys\n",
    "import os\n",
    "sys.path.append(os.path.abspath(\"src\"))  # Add src directory to path\n",
    "from createDataTable import gene_pair0\n",
    "gene_pair0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77f9494f-c8a9-47cd-82f9-25c01e289519",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8589fb2c-6d6d-44f0-b58d-3785862b6764",
   "metadata": {},
   "outputs": [],
   "source": [
    "import liana as li\n",
    "import omnipath as op\n",
    "import decoupler as dc\n",
    "import pandas as pd\n",
    "\n",
    "import sys\n",
    "import os\n",
    "sys.path.append(os.path.abspath(\"src\"))  # Add src directory to path\n",
    "from createDataTable import gene_pair0\n",
    "\n",
    "gene_pair_annot = gene_pair0[[\"Human LR Pair\", \"Cancer-related\", \"Top Pathway\"]]\n",
    "df= pd.read_csv(\"data/disease_annotations_per_pair.csv\") # Liana Diseases\n",
    "gene_pair_annot = gene_pair_annot.merge(df, how='left', left_on='Human LR Pair', right_on='interaction_x')\n",
    "df= pd.read_csv(\"data/pathway_annotations_per_pair.csv\") # Liana Pathway\n",
    "gene_pair_annot = gene_pair_annot.merge(df, how='left', left_on='Human LR Pair', right_on='interaction')\n",
    "gene_pair_annot = gene_pair_annot.drop(columns=[\"interaction_x\", \"interaction\", \"weight\"])\n",
    "\n",
    "gene_pair_annot = gene_pair_annot.rename(columns={\n",
    "                                     \"disease\": \"Disease\", \n",
    "                                     \"source\": \"Related Pathway\"}\n",
    "                            )\n",
    "# Create the links to the HTML cards\n",
    "gene_pair_annot[\"Human LR Pair\"] = [\n",
    "    f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/cards/{lrPairOrig}.html\">{lrPair}</a>'\n",
    "    for lrPairOrig, lrPair in zip(gene_pair_annot[\"Human LR Pair\"], gene_pair_annot[\"Human LR Pair\"])\n",
    "]\n",
    "\n",
    "# reorder\n",
    "gene_pair_annot = gene_pair_annot[[\"Human LR Pair\", \"Disease\", \"Disease Type\", \"Cancer-related\",  \"Related Pathway\", \"Top Pathway\"]]\n",
    "gene_pair_annot[\"Disease\"] = gene_pair_annot[\"Disease\"].apply(\n",
    "    lambda x: \"unknown\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\"] else x)\n",
    "gene_pair_annot[\"Disease Type\"] = gene_pair_annot[\"Disease Type\"].apply(\n",
    "    lambda x: \"unknown\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\"] else x)\n",
    "gene_pair_annot[\"Related Pathway\"] = gene_pair_annot[\"Related Pathway\"].apply(\n",
    "    lambda x: \"unknown\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\"] else x)\n",
    "gene_pair_annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d2a3162-2a6c-4637-a0a0-a8274dd11500",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be791e46-e64d-4d1c-8538-3ac153aea722",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop columns where all values are NA in gene_pair\n",
    "gene_pair = gene_pair.dropna(axis=1, how='all')\n",
    "\n",
    "gene_pair = gene_pair.fillna(\" \")\n",
    "gene_pair = gene_pair[gene_pair['Human LR Pair'] != ' ']\n",
    "\n",
    "if \"PMID link\" in gene_pair.columns:\n",
    "    gene_pair = gene_pair.drop(columns=[\"PMID link\"])\n",
    "\n",
    "# Add\n",
    "first_columns=['Human LR Pair', 'Ligand', 'Receptor', 'Interaction Source']\n",
    "\n",
    "end_columns=['HGNC L R', 'sanity check', 'curator', 'secondary source?']\n",
    "gene_pair = gene_pair[first_columns + [col for col in gene_pair.columns if col not in first_columns + end_columns] + end_columns]\n",
    "\n",
    "\n",
    "# number of unique vars\n",
    "\n",
    "lrPairsCount = len(gene_pair[\"Human LR Pair\"].unique())\n",
    "\n",
    "ligandCount = len(gene_pair[\"Ligand\"].unique())\n",
    "\n",
    "receptorCount = len(gene_pair[\"Receptor\"].unique())\n",
    "\n",
    "# Mouse Orthologue\n",
    "MouseLigandCount = len(gene_pair[\"Ligand MGI ID\"].unique())\n",
    "\n",
    "MouseReceptorCount = len(gene_pair[\"Receptor MGI ID\"].unique())\n",
    "\n",
    "# Rat Orthologue\n",
    "RatLigandCount = len(gene_pair[\"Ligand RGD ID\"].unique())\n",
    "\n",
    "RatReceptorCount = len(gene_pair[\"Receptor RGD ID\"].unique())\n",
    "\n",
    "gene_pair[\"PMID support\"] = [value.replace(\" \", \"\") for value in gene_pair[\"PMID support\"]]\n",
    "\n",
    "source = np.array(gene_pair[\"PMID support\"].unique())\n",
    "source = source.astype(str)\n",
    "source = \",\".join(sorted(set(filter(lambda x: x.lower() != 'nan', source))))\n",
    "\n",
    "# Split the string into individual elements, filter out empty strings, and get unique values\n",
    "source = sorted(\n",
    "    set(filter(lambda x: x.strip() and x.strip().lower() != 'nan', source.split(',')))\n",
    ")\n",
    "source = [value.replace(\" \", \"\") for value in source]\n",
    "sourceCount = len(source)\n",
    "\n",
    "# for creating PMIDs\n",
    "gene_pair00 = gene_pair[['Human LR Pair', 'PMID support']]\n",
    "\n",
    "# create URLs for the HGNC IDs\n",
    "\n",
    "# ligand\n",
    "gene_pair[\"Ligand HGNC ID\"] = [\n",
    "    '<a href=\"https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/{}\" target=\"_blank\">{}</a>'.format(ligand, ligand)\n",
    "    for ligand in gene_pair[\"Ligand HGNC ID\"]\n",
    "]\n",
    "\n",
    "# receptor\n",
    "gene_pair[\"Receptor HGNC ID\"] = [\n",
    "    '<a href=\"https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/{}\" target=\"_blank\">{}</a>'.format(receptor, receptor)\n",
    "    for receptor in gene_pair[\"Receptor HGNC ID\"]\n",
    "]\n",
    "\n",
    "# Perplexity\n",
    "gene_pair[\"Perplexity\"] = [\n",
    "    '<a href=\"{}\" target=\"_blank\"> <img src=\"https://img.icons8.com/?size=30&id=0NbBuNOxUwps&format=png&color=000000\" alt=\"Perplexity AI\" /></a>'.format(url)\n",
    "    for url in gene_pair[\"Perplexity\"]\n",
    "]\n",
    "\n",
    "# Function to generate hyperlinks for the \"PMID support\" column\n",
    "# Function to generate hyperlinks for the \"PMID support\" column\n",
    "def generate_links_with_doi(df, gene_column, pmid_column):\n",
    "    def create_link(gene, sources):\n",
    "        # Replace spaces with \"â€”â€”\" in the gene name for the link\n",
    "        gene_name = gene.replace(\" \", \"â€”â€”\")\n",
    "        \n",
    "        if len(sources) == 1:\n",
    "            source = sources[0]\n",
    "            if source.startswith(\"https://www.biorxiv.org/content/\"):\n",
    "                # If the value starts with \"https://doi.org/\", use it as the hyperlink\n",
    "                return f'<a href=\"{source}\" target=\"_blank\">BioRxiv preprint</a>'\n",
    "            else:\n",
    "                # If it's a single PMID, hyperlink the PMID text\n",
    "                return f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/pubmed/{gene_name}_pmid_details.html\">{source}</a>'\n",
    "        else:\n",
    "            # If multiple PMIDs, show the count and hyperlink to the page\n",
    "            return f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/pubmed/{gene_name}_pmid_details.html\" target=\"_blank\">{len(sources)} PMIDs</a>'\n",
    "\n",
    "    # Process each row to generate the \"PMID support\" column\n",
    "    df[\"PMID support\"] = [\n",
    "        create_link(\n",
    "            gene=row[gene_column], \n",
    "            sources=[s.strip() for s in row[pmid_column].split(',') if s.strip()]\n",
    "        )\n",
    "        for _, row in df.iterrows()\n",
    "    ]\n",
    "    return df\n",
    "\n",
    "\n",
    "# Generate the links for the \"PMID support\" column\n",
    "gene_pair = generate_links_with_doi(gene_pair, gene_column=\"Human LR Pair\", pmid_column=\"PMID support\")\n",
    "\n",
    "gene_pair[\"Ligand MGI ID\"] = [\n",
    "        f'<a href=\"https://www.informatics.jax.org/marker/{mouseOrth}\" target=\"_blank\">{mouseOrth}</a>' \n",
    "        if pd.notna(mouseOrth) and mouseOrth.strip() else \"\" \n",
    "        for mouseOrth in gene_pair[\"Ligand MGI ID\"]\n",
    "    ]\n",
    "\n",
    "gene_pair[\"Receptor MGI ID\"] = [\n",
    "        f'<a href=\"https://www.informatics.jax.org/marker/{mouseOrth}\" target=\"_blank\">{mouseOrth}</a>' \n",
    "        if pd.notna(mouseOrth) and mouseOrth.strip() else \"\" \n",
    "        for mouseOrth in gene_pair[\"Receptor MGI ID\"]\n",
    "    ]\n",
    "\n",
    "gene_pair[\"Ligand RGD ID\"] = [\n",
    "        f'<a href=\"https://rgd.mcw.edu/rgdweb/report/gene/main.html?id={ratOrth.replace(\"RGD:\", \"\")}\" target=\"_blank\">{ratOrth}</a>' \n",
    "        if pd.notna(ratOrth) and ratOrth.strip() else \"\" \n",
    "        for ratOrth in gene_pair[\"Ligand RGD ID\"]\n",
    "    ]\n",
    "\n",
    "gene_pair[\"Receptor RGD ID\"] = [\n",
    "        f'<a href=\"https://rgd.mcw.edu/rgdweb/report/gene/main.html?id={ratOrth.replace(\"RGD:\", \"\")}\" target=\"_blank\">{ratOrth}</a>' \n",
    "        if pd.notna(ratOrth) and ratOrth.strip() else \"\" \n",
    "        for ratOrth in gene_pair[\"Receptor RGD ID\"]\n",
    "    ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "503bcddf-1547-4959-91fe-2faa0eaac2ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair[\"Source\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "126a784e-b5c6-45cc-913b-105533b74457",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to prepare datatables (cleaning and hyperlinking, adding tool tips, etc) input for the database qmds\n",
    "import sys, os\n",
    "from itables import init_notebook_mode\n",
    "import pandas as pd\n",
    "from itables import show\n",
    "from itables import options\n",
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "import fetchGSheet \n",
    "import warnings\n",
    "\n",
    "# Suppress SettingWithCopyWarning\n",
    "warnings.simplefilter(\"ignore\", category=UserWarning)\n",
    "\n",
    "\n",
    "# Other vertebrates\n",
    "species_list = [\n",
    "    \"ptroglodytes\", \"ggallus\", \"sscrofa\", \"btaurus\", \n",
    "    \"clfamiliaris\", \"ecaballus\", \"oarambouillet\"\n",
    "]\n",
    "\n",
    "# Select only the relevant columns from pop_up_info\n",
    "pop_up_info = pd.read_table(\"data/HGNC_gene_info_full.tsv\")\n",
    "pop_up_info = pop_up_info.rename(columns={\"hgnc_id\": \"HGNC ID\", \n",
    "                                          \"name\": \"Approved name\",\n",
    "                                          \"symbol\": \"Approved symbol\",\n",
    "                                          \"rgd_id\": \"RGD ID\",\n",
    "                                          \"mgd_id\": \"MGI ID\", \n",
    "                                          \"rgd_id\": \"RGD ID\",\n",
    "                                          \"alias_symbol\": \"Alias symbol\",\n",
    "                                          \"prev_symbol\": \"Previous symbol\",\n",
    "                                          \"date_symbol_changed\": \"Date symbol changed\"\n",
    "                                          \n",
    "                                         })\n",
    "\n",
    "pop_up_info_lim = pop_up_info[[\"HGNC ID\", \"Approved name\", \"MGI ID\", \"RGD ID\"]] # rm \"Approved symbol\" for now\n",
    "pop_up_info_lim = pop_up_info_lim.drop_duplicates(subset=\"HGNC ID\", keep=\"first\")\n",
    "\n",
    "# Drop columns where all values are NA in gene_pair\n",
    "gene_pair = fetchGSheet.gene_pair.dropna(axis=1, how='all')\n",
    "\n",
    "# for now, rm some columns\n",
    "gene_pair = gene_pair[['LR pair', 'Ligand', 'Ligand.HGNC', 'Receptor', 'Receptor.HGNC',\n",
    "                       'perplexity link', 'PMID', 'binding location', \n",
    "                       'bind in trans?', 'bidirectional signalling?',\n",
    "                       'interaction type', 'original source']]\n",
    "\n",
    "# Mapping for replacements\n",
    "mapping = dict(zip(fetchGSheet.src_info['original source'], fetchGSheet.src_info['shortname']))\n",
    "# Replace values in the column based on the mapping\n",
    "gene_pair['original source'] = gene_pair['original source'].replace(mapping)\n",
    "\n",
    "## add Ligand/Receptor Location\n",
    "mapping_loc = dict(zip(fetchGSheet.loc_info['ApprovedSymbol'], fetchGSheet.loc_info['Localization']))\n",
    "gene_pair['Ligand location'] = gene_pair['Ligand'].replace(mapping_loc)\n",
    "gene_pair['Receptor location'] = gene_pair['Receptor'].replace(mapping_loc)\n",
    "\n",
    "# Fetch species IDs from the dataset\n",
    "hgnc_id = [col for col in gene_pair.columns if \"HGNC\" in col]\n",
    "hgnc_id = pd.concat([gene_pair[col] for col in hgnc_id]).unique()\n",
    "\n",
    "# Rename columns for better clarity\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "    \"LR pair\": \"Human LR Pair\",\n",
    "    \"Ligand.HGNC\": \"Ligand HGNC ID\",\n",
    "    \"Receptor.HGNC\": \"Receptor HGNC ID\",\n",
    "    \"perplexity link\": \"Perplexity\", # will be replaced with actual link later\n",
    "    \"original source\": \"Database Source\",\n",
    "    \"PMID\": \"PMID support\"\n",
    "})\n",
    "\n",
    "# Recreate Perplexity link\n",
    "# Function to generate Perplexity search link\n",
    "def create_url_basic(gene_name):\n",
    "    query = f\"What is the primary evidence that {gene_name} bind-each-other-as-a-ligand-and-receptor-pair. Exclude reviews, uniprot, wiki, genecards, PIPS, iuphar as sources.\"\n",
    "    encoded_query = query.replace(\" \", \"%20\")\n",
    "    return f\"https://www.perplexity.ai/search?q={encoded_query}\"\n",
    "\n",
    "# Apply function to the DataFrame\n",
    "gene_pair[\"Perplexity\"] = gene_pair[\"Perplexity\"].apply(create_url_basic)\n",
    "\n",
    "# Merge gene_pair with pop_up_info_lim for Ligand(L)\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Ligand HGNC ID', right_on='HGNC ID')\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Ligand name\", \n",
    "                                     \"MGI ID\": \"Ligand MGI ID\",\n",
    "                                     \"RGD ID\": \"Ligand RGD ID\"},\n",
    "                            )\n",
    "gene_pair = gene_pair.drop(columns=[\"HGNC ID\"])\n",
    "# Add top pathway per pair\n",
    "LR_pairs = gene_pair[\"Human LR Pair\"].unique()\n",
    "df= pd.read_csv(\"data/pathway_annotations_per_pair.csv\")\n",
    "#df = df[df[\"interaction\"].isin(LR_pairs)]\n",
    "# Sort by absolute value of 'weight', descending (larger abs(weight) first)\n",
    "df_sorted = df.reindex(df['weight'].abs().sort_values(ascending=False).index)\n",
    "# Keep only the first occurrence for each unique 'interaction'\n",
    "df_unique = df_sorted.drop_duplicates(subset='interaction', keep='first')\n",
    "df = df_unique.reset_index(drop=True)\n",
    "top_pathway_df = df[[\"interaction\", \"source\"]]\n",
    "top_pathway_df = top_pathway_df.rename(columns={\n",
    "                                      \"source\": \"Top Pathway\"\n",
    "})\n",
    "top_pathway_df[\"interaction\"] = [value.replace(\"^\", \" \") for value in top_pathway_df[\"interaction\"]]\n",
    "gene_pair = gene_pair.merge(top_pathway_df, how='left', left_on='Human LR Pair', right_on='interaction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89856101-d7eb-41d5-9f01-ba2d34d3076b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df= pd.read_csv(\"data/diseaseType_per_pair.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fe51be8-249c-4ffe-b8b6-aeace66fda33",
   "metadata": {},
   "outputs": [],
   "source": [
    "disease_df = df[df[\"interaction_x\"].isin(LR_pairs)]\n",
    "\n",
    "gene_pair = gene_pair.merge(disease_df, how='left', left_on='Human LR Pair', right_on='interaction_x')\n",
    "gene_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87cfc3c8-311f-40a6-b014-febcd42f7c40",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os, sys\n",
    "import json\n",
    "# Add the src directory to the path for importing modules\n",
    "sys.path.append(os.path.abspath(\"src\"))\n",
    "from createPMIDpages import gene_pair00\n",
    "\n",
    "# Load the files\n",
    "file1 = pd.read_csv(\"data/pubmed_results.csv\") \n",
    "file2 = gene_pair00\n",
    "\n",
    "# Convert the PMIDs column in file2 to lists for easy comparison\n",
    "file1['PMID'] = file1['PMID'].astype(str)\n",
    "file2['PMID_List'] = file2['PMID support'].apply(lambda x: x.split(','))\n",
    "\n",
    "# Create a dictionary for quick PMID to Abstract mapping\n",
    "pmid_to_abstract = dict(zip(file1['PMID'], file1['Abstract']))\n",
    "pmid_to_abstract"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8c97139-ccc9-4e87-9e17-d164834051ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_for_llm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6295d052-8d3a-419c-9379-42aa5adfead1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to get all abstracts for a list of PMIDs\n",
    "def get_abstracts(pmids):\n",
    "    return [pmid_to_abstract[pmid] for pmid in pmids if pmid in pmid_to_abstract]\n",
    "\n",
    "# Map abstracts to LR pairs\n",
    "file2['Abstracts'] = file2['PMID_List'].apply(get_abstracts)\n",
    "\n",
    "# Convert to a list of dictionaries\n",
    "data_for_llm = file2[['Human LR Pair', 'Abstracts']].to_dict(orient='records')\n",
    "\n",
    "# Save as JSON\n",
    "with open(\"data/data_for_llm.json\", \"w\") as f:\n",
    "    json.dump(data_for_llm, f, indent=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2400ee4-b091-4bac-9924-3ab1bad98f81",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save as JSON\n",
    "with open(\"data/data_for_llm.json\", \"w\") as f:\n",
    "    json.dump(data_for_llm, f, indent=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88c4c3a3-c33b-49ba-91fc-b39ee3cb0837",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to add species-specific species Enseml ID and symbol for all other species except for mouse, rat, and zebrafish\n",
    "def appendOtherSpeciesInfo(species, origDF):\n",
    "    # Load species-specific data\n",
    "    species_info = pd.read_csv(f\"data/{species}_ID_biomart.csv\")\n",
    "\n",
    "    # Keep relevant columns\n",
    "    species_info = species_info[[f\"{species}_homolog_ensembl_gene\", \n",
    "                                 f\"{species}_homolog_associated_gene_name\", \n",
    "                                 'hgnc_id']]\n",
    "\n",
    "    # Remove rows where 'hgnc_id' is NaN and drop duplicates\n",
    "    species_info = species_info.dropna(subset=['hgnc_id'])\n",
    "    species_info = species_info.drop_duplicates(subset=['hgnc_id'])\n",
    "\n",
    "    # Merge with ligand data\n",
    "    origDF = origDF.merge(species_info, how='left', \n",
    "                           left_on='Ligand HGNC ID', right_on='hgnc_id')\n",
    "    \n",
    "    # Rename columns for ligand info\n",
    "    origDF = origDF.rename(columns={\n",
    "        f\"{species}_homolog_associated_gene_name\": f\"{species} Ligand\", \n",
    "        f\"{species}_homolog_ensembl_gene\": f\"{species} Ligand Ensembl ID\"\n",
    "    })\n",
    "\n",
    "    # Drop duplicate 'hgnc_id' column\n",
    "    origDF = origDF.drop(columns=['hgnc_id'])\n",
    "\n",
    "    # Merge with receptor data\n",
    "    origDF = origDF.merge(species_info, how='left', \n",
    "                           left_on='Receptor HGNC ID', right_on='hgnc_id')\n",
    "\n",
    "    # Rename columns for receptor info\n",
    "    origDF = origDF.rename(columns={\n",
    "        f\"{species}_homolog_associated_gene_name\": f\"{species} Receptor\", \n",
    "        f\"{species}_homolog_ensembl_gene\": f\"{species} Receptor Ensembl ID\"\n",
    "    })\n",
    "\n",
    "        # Drop duplicate 'hgnc_id' column\n",
    "    origDF = origDF.drop(columns=['hgnc_id'])\n",
    "\n",
    "    # Drop columns where all values are NaN\n",
    "    origDF = origDF.dropna(axis=1, how='all')\n",
    "\n",
    "    return origDF\n",
    "\n",
    "species_list = [\n",
    "    \"ptroglodytes\", \"ggallus\", \"sscrofa\", \"btaurus\", \n",
    "    \"clfamiliaris\", \"ecaballus\", \"oarambouillet\"\n",
    "]\n",
    "\n",
    "# Loop through each species and update gene_pair\n",
    "for species in species_list:\n",
    "    gene_pair = appendOtherSpeciesInfo(species, gene_pair)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0628a664-321a-46fe-a1c9-6aca3bffd7fd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e54bef91-fa38-4a01-a39d-993b7fd6adf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b443872e-66e7-427a-a031-75a16471e8a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_row(change):\n",
    "    global gene_pair\n",
    "    # Add a new row at the top with None values\n",
    "    new_row = {col: None for col in gene_pair.columns}\n",
    "    gene_pair = pd.DataFrame([new_row] + gene_pair.to_dict(orient=\"records\"))\n",
    "    update_table()\n",
    "\n",
    "# Function to remove the last row of the dataframe\n",
    "def remove_row(change):\n",
    "    global gene_pair\n",
    "    if len(gene_pair) > 0:\n",
    "        gene_pair = gene_pair[:-1]  # Remove the last row\n",
    "        update_table()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "228f917d-b5b9-409e-bc04-7b95afd9a725",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19161b56-8989-458f-80ef-d1e3eb97e029",
   "metadata": {},
   "outputs": [],
   "source": [
    "duplicates = gene_pair00[gene_pair00[\"Human LR Pair\"].duplicated()]\n",
    "print(duplicates[\"Human LR Pair\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9d79fce-3cfa-469a-abdb-cd05c49d221b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to create horizontal bar plots of each gene in Human Taxon --expression log(x+1) transformed with cell types as y-axis\n",
    "\n",
    "import requests\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "sys.path.append(os.path.abspath(\"src\"))  # Add src directory to path\n",
    "from createDataTable import gene_pair0\n",
    "\n",
    "# Input file\n",
    "input_file=\"data/connectome_j.tsv\" #\"data/connectome_j.tsv\" # data/ExpressionGenes.txt\n",
    "# Get all unique genes\n",
    "ligand_list = gene_pair0[\"Ligand\"].tolist()\n",
    "receptor_list = gene_pair0[\"Receptor\"].tolist()\n",
    "unique_genes = list(set(ligand_list + receptor_list))  # Combine and remove duplicates\n",
    "\n",
    "connectomeDB = pd.read_table(input_file, sep=\"\\t\")\n",
    "# All Taxon for now\n",
    "#connectomeDB = connectomeDB[connectomeDB[\"Taxon\"]== \"Human\"]\n",
    "if \"Taxon\" in connectomeDB.columns:\n",
    "    connectomeDB = connectomeDB.drop(columns=[\"Localization\", \"Taxon\"] + [col for col in connectomeDB.columns if col.startswith(\"F5_\")])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd5564ed-51a3-4f4b-8dc0-57f7a31358e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "column_sums = connectomeDB.iloc[:, 1:].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56b32bc0-ebe0-4f37-ac0b-ac81d07ccae5",
   "metadata": {},
   "outputs": [],
   "source": [
    "connectomeDB.iloc[:, 1:].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7c62af8-646e-4a62-8f19-adfc56766c93",
   "metadata": {},
   "outputs": [],
   "source": [
    "intersection = pd.Series(list(set(connectomeDB['ApprovedSymbol']).intersection(unique_genes)))\n",
    "intersection\n",
    "\n",
    "connectomeDB = connectomeDB[connectomeDB[\"ApprovedSymbol\"].isin(intersection)]\n",
    "connectomeDB\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7707cc4d-426a-41a1-8f1a-fe68ffb62330",
   "metadata": {},
   "outputs": [],
   "source": [
    "# log(x+1) transform\n",
    "connectomeDB.iloc[:, 1:] = np.log1p(connectomeDB.iloc[:, 1:])\n",
    "# Reshape \n",
    "connectomeDB_long = connectomeDB.melt(id_vars=[\"ApprovedSymbol\"], \n",
    "                                      var_name=\"cellTypes\", value_name=\"expr_val\")\n",
    "cellCat = pd.read_csv(\"data/cell_categories.csv\")\n",
    "connectomeDB_long = connectomeDB_long.merge(cellCat, how='left', left_on='cellTypes', right_on='cellType')\n",
    "connectomeDB_long = connectomeDB_long.drop(columns=[\"cellType\"])\n",
    "\n",
    "intersection = pd.Series(list(set(connectomeDB_long['cellTypes']).intersection(set(cellCat['cellType']))))\n",
    "intersection\n",
    "\n",
    "diff_df = pd.Series(list(set(connectomeDB_long['cellTypes']).difference(set(cellCat['cellType']))))\n",
    "diff_df\n",
    "\n",
    "def plot_gene_expression(df):\n",
    "    # Define the colors for each cell category\n",
    "    colors = {\n",
    "        \"missing\": \"#B0B0B0\",  # Neutral gray\n",
    "        \"other\": \"#D4A76A\",  # Warm gold\n",
    "        \"mesenchymal\": \"#377EB8\",  # Vibrant blue\n",
    "        \"epithelial\": \"#E41A1C\",  # Bold red\n",
    "        \"hematopoietic\": \"#4DAF4A\",  # Fresh green\n",
    "        \"endothelial\": \"#984EA3\",  # Deep purple\n",
    "        \"nervous system\": \"#FF7F00\",  # Bright orange\n",
    "    }\n",
    "\n",
    "    # Define sorting order for cell categories\n",
    "    category_order = {cat: i for i, cat in enumerate(colors.keys())}\n",
    "\n",
    "    for gene, sub_df in df.groupby(\"ApprovedSymbol\"):\n",
    "        # Sort by category first, then by expression value (highest first)\n",
    "        sub_df = sub_df.copy()\n",
    "        sub_df[\"category_order\"] = sub_df[\"cellCategory\"].map(category_order).fillna(len(category_order))\n",
    "        sub_df = sub_df.sort_values([\"category_order\", \"expr_val\"], ascending=[True, False])\n",
    "\n",
    "        num_bars = len(sub_df)\n",
    "\n",
    "        # Plotly Figure setup\n",
    "        fig = go.Figure()\n",
    "\n",
    "        # Loop through each category and create a trace for it\n",
    "        for category, color in colors.items():\n",
    "            # Filter data for the current category\n",
    "            category_data = sub_df[sub_df[\"cellCategory\"] == category]\n",
    "\n",
    "            # Add the trace for the current category\n",
    "            fig.add_trace(go.Bar(\n",
    "                y=category_data[\"cellTypes\"],  # Categories for y-axis\n",
    "                x=category_data[\"expr_val\"],  # Expression values for x-axis\n",
    "                orientation='h',  # Horizontal bars\n",
    "                marker=dict(color=color),\n",
    "                hovertemplate=\n",
    "                    '<b>%{y}</b><br>' +  # Cell type (y-axis value)\n",
    "                    'Expression Value: %{x}',  # Expression value (x-axis value)\n",
    "                    #'Category: %{text}',  # Custom text (cell category)\n",
    "                #text=category_data[\"cellCategory\"],  # Pass the cell category as custom text\n",
    "                name=category,  # Use the category name for the legend\n",
    "                showlegend=True,  # Ensure the legend is shown for this trace\n",
    "            ))\n",
    "\n",
    "        # Update layout settings\n",
    "        fig.update_layout(\n",
    "            title=\"\",\n",
    "            xaxis_title=\"log(x+1) Expression value\",\n",
    "            yaxis_title=\"Cell Types\",\n",
    "            yaxis=dict(\n",
    "                tickmode='array',\n",
    "                tickvals=np.arange(num_bars),\n",
    "                ticktext=sub_df[\"cellTypes\"],\n",
    "                tickangle=0,  # Avoid overlapping labels by setting the angle to 0\n",
    "                tickfont=dict(size=6),  # Set font size for the labels\n",
    "            ),\n",
    "            showlegend=True,\n",
    "            legend_title=\"Cell Category\",\n",
    "            legend=dict(\n",
    "                orientation=\"v\",  # Vertical legend\n",
    "                yanchor=\"top\",\n",
    "                y=1,\n",
    "                xanchor=\"\",\n",
    "                x=1.05,  # Position the legend outside of the plot area\n",
    "                font=dict(size=10)\n",
    "            ),\n",
    "            margin=dict(t=50, b=50, l=150, r=50),\n",
    "            height=min(1000, max(500, num_bars * 30)),  # Adjust plot height\n",
    "            plot_bgcolor='rgba(0,0,0,0)',  # Transparent background\n",
    "            paper_bgcolor='rgba(0,0,0,0)',  # Transparent paper background\n",
    "        )\n",
    "\n",
    "        # Save to HTML file\n",
    "        fig.write_html(f\"data/gene_expr_plots/{gene}.html\")\n",
    "\n",
    "\n",
    "plot_gene_expression(connectomeDB_long)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00571f7a-50f1-4901-b0f8-1e76d14e43c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "connectomeDB_long"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7f4b069-0370-424e-9b83-4caaefce586d",
   "metadata": {},
   "source": [
    "## Testing Liana+"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c78bfa0-43b3-4857-9ca4-41dcd8a230e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import liana as li\n",
    "import omnipath as op\n",
    "import decoupler as dc\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "766bd3fd-d53d-4586-8d1e-09befbaa9675",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "sys.path.append(os.path.abspath(\"src\"))  # Add src directory to path\n",
    "from createDataTable import gene_pair0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f681087-38ba-4cfa-a105-8707ff70ac5f",
   "metadata": {},
   "source": [
    "### Pathway Annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a67bd553-c8df-4970-9811-435c46e2c51c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load PROGENy pathways, we use decoupler as a proxy as it formats the data in a more convenient way\n",
    "progeny = dc.get_progeny(top=10000)\n",
    "progeny"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95b2a08f-d44b-48ce-8a26-4f74ec640a51",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_pairs = gene_pair0[[\"Ligand\", \"Receptor\"]]\n",
    "lr_pairs.columns = lr_pairs.columns.str.lower()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7f4b57a-700a-4e04-9c95-f33719a4f877",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "548ba05e-8929-4434-a765-0d77b67d0c4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate ligand-receptor geneset\n",
    "lr_progeny = li.rs.generate_lr_geneset(lr_pairs, progeny, lr_sep=\"^\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "091fba45-e3f2-4e26-b1fb-6bd3ed8ebacb",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_progeny"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9beaedff-b8b8-4c41-b8c7-7ead6e5403aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# some of the pairs are missing\n",
    "len(lr_progeny[\"interaction\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23e12dce-6c86-4006-b9ba-01a6924aa0f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_file=\"data/pathway_annotations_per_pair.csv\"\n",
    "lr_progeny.to_csv(output_file, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e477d7b-ced3-41ea-b412-65d9936b6644",
   "metadata": {},
   "outputs": [],
   "source": [
    "whichDB= 'DisGeNet'\n",
    "# A database of expression profiles related to human diseases, including cancer\n",
    "diseases = op.requests.Annotations.get(\n",
    "    resources = [whichDB]\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb551b9f-add0-45ea-98f8-58c7f3fe8a54",
   "metadata": {},
   "outputs": [],
   "source": [
    "diseases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4fd5feb-feee-4193-aabc-6da6257b0dc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "diseases.to_csv(\"data/\" + whichDB + \".csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf34da1b-e119-4c62-bfd4-41770dd2d010",
   "metadata": {},
   "source": [
    "### Disease Annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1056987-e2ca-495c-a03f-78d002b63302",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DisGeNet\n",
    "diseases = op.requests.Annotations.get(\n",
    "    resources = ['DisGeNet']\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79d3b1e6-9ceb-4782-9c00-ef11379f0ea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "diseases = diseases[['genesymbol', 'label', 'value']]\n",
    "diseases = diseases.pivot_table(index='genesymbol',\n",
    "                                columns='label', values='value',\n",
    "                                aggfunc=lambda x: '; '.join(x)).reset_index()\n",
    "diseases = diseases[['genesymbol', 'disease']]\n",
    "diseases['disease'] = diseases['disease'].str.split('; ')\n",
    "diseases = diseases.explode('disease')\n",
    "lr_diseases = li.rs.generate_lr_geneset(lr_pairs, diseases, source='disease', target='genesymbol', weight=None, lr_sep=\"^\")\n",
    "lr_diseases.sort_values(\"interaction\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "115a5201-dd41-4519-ae48-68fc7e004828",
   "metadata": {},
   "outputs": [],
   "source": [
    "# some of the pairs are missing\n",
    "len(lr_diseases[\"interaction\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fd738f5-cebd-49e2-be19-673450d33efb",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_file=\"data/disease_annotations_per_pair.csv\"\n",
    "lr_diseases.to_csv(output_file, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de5cef07-a5eb-483d-8f28-ba9b65a5cda2",
   "metadata": {},
   "outputs": [],
   "source": [
    "op.requests.Annotations.resources()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2afd8af-fa7d-4a3a-b4a5-9f67d1b42e86",
   "metadata": {},
   "source": [
    "### Get FASTA sequences for each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a7dd4db-bf07-4131-b3b5-f9611116cb7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "import sys\n",
    "import os\n",
    "\n",
    "sys.path.append(os.path.abspath(\"src\"))  # Add src directory to path/\n",
    "from createDataTable import gene_pair0\n",
    "\n",
    "\n",
    "# Get all unique genes\n",
    "ligand_list = gene_pair0[\"Ligand\"].tolist()\n",
    "receptor_list = gene_pair0[\"Receptor\"].tolist()\n",
    "unique_genes = list(set(ligand_list + receptor_list)) \n",
    "LR_pairs=gene_pair0['Human LR Pair'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78afcc47-8c40-45e1-8e7a-1ee3adaf1e74",
   "metadata": {},
   "outputs": [],
   "source": [
    "df= pd.read_csv(\"data/pathway_annotations_per_pair.csv\")\n",
    "df = df[df[\"interaction\"].isin(LR_pairs)]\n",
    "# Step 1: Sort by absolute value of 'weight', descending (larger abs(weight) first)\n",
    "df_sorted = df.reindex(df['weight'].abs().sort_values(ascending=False).index)\n",
    "# Keep only the first occurrence for each unique 'interaction'\n",
    "df_unique = df_sorted.drop_duplicates(subset='interaction', keep='first')\n",
    "df = df_unique.reset_index(drop=True)\n",
    "top_pathway_df = df[[\"interaction\", \"source\"]]\n",
    "top_pathway_df = top_pathway_df.rename(columns={\n",
    "                                      \"source\": \"Top Pathway\"\n",
    "}\n",
    "                            )\n",
    "top_pathway_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "7f6e98fa-2643-44ad-9042-44f70a291598",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease</th>\n",
       "      <th>interaction</th>\n",
       "      <th>Disease Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Liver carcinoma</td>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bipolar Disorder</td>\n",
       "      <td>ACE BDKRB2</td>\n",
       "      <td>Mental Health</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Hypertensive disease</td>\n",
       "      <td>ACE BDKRB2</td>\n",
       "      <td>Cardiovascular Diseases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Hypotension</td>\n",
       "      <td>ACE BDKRB2</td>\n",
       "      <td>Cardiovascular Diseases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Major Depressive Disorder</td>\n",
       "      <td>ACE BDKRB2</td>\n",
       "      <td>Mental Health</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3506</th>\n",
       "      <td>Breast Carcinoma</td>\n",
       "      <td>YBX1 NOTCH3</td>\n",
       "      <td>Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3507</th>\n",
       "      <td>Malignant neoplasm of breast</td>\n",
       "      <td>YBX1 NOTCH3</td>\n",
       "      <td>Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3508</th>\n",
       "      <td>Mammary Carcinoma, Human</td>\n",
       "      <td>YBX1 NOTCH3</td>\n",
       "      <td>Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3509</th>\n",
       "      <td>Mammary Neoplasms</td>\n",
       "      <td>YBX1 NOTCH3</td>\n",
       "      <td>Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3510</th>\n",
       "      <td>Mammary Neoplasms, Human</td>\n",
       "      <td>YBX1 NOTCH3</td>\n",
       "      <td>Cancer</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3511 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           disease  interaction             Disease Type\n",
       "0                  Liver carcinoma    A2M HSPA5                   Cancer\n",
       "1                 Bipolar Disorder   ACE BDKRB2            Mental Health\n",
       "2             Hypertensive disease   ACE BDKRB2  Cardiovascular Diseases\n",
       "3                      Hypotension   ACE BDKRB2  Cardiovascular Diseases\n",
       "4        Major Depressive Disorder   ACE BDKRB2            Mental Health\n",
       "...                            ...          ...                      ...\n",
       "3506              Breast Carcinoma  YBX1 NOTCH3                   Cancer\n",
       "3507  Malignant neoplasm of breast  YBX1 NOTCH3                   Cancer\n",
       "3508      Mammary Carcinoma, Human  YBX1 NOTCH3                   Cancer\n",
       "3509             Mammary Neoplasms  YBX1 NOTCH3                   Cancer\n",
       "3510      Mammary Neoplasms, Human  YBX1 NOTCH3                   Cancer\n",
       "\n",
       "[3511 rows x 3 columns]"
      ]
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df= pd.read_csv(\"data/disease_annotations_per_pair.csv\")\n",
    "df_cat=pd.read_csv(\"data/disease_categories.csv\")\n",
    "mapping = dict(zip(df_cat['Disease Name'], df_cat['Category']))\n",
    "# Replace values in the column based on the mapping\n",
    "df[\"Disease Type\"] = df['disease'].replace(mapping)\n",
    "df = df[[\"interaction\", \"Disease Type\"]].drop_duplicates()\n",
    "df['Disease Type'] = df['Disease Type'].astype(str)\n",
    "# Group by 'col1' and combine 'col2' values with ', '\n",
    "df = df.groupby('interaction')['Disease Type'].apply(', '.join).reset_index()\n",
    "# Create \"Cancer-related\" column based on whether \"Cancers & Neoplasms\" is in col2\n",
    "df['Cancer-related'] = df['Disease Type'].apply(lambda x: 'Yes' if 'Cancer' in x else 'No')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "8df80779-3fe2-409a-a3d3-55353c103955",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interaction</th>\n",
       "      <th>Disease Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACE BDKRB2</td>\n",
       "      <td>Mental Health, Cardiovascular Diseases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADAM12 ITGA9</td>\n",
       "      <td>Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADAM17 ITGB1</td>\n",
       "      <td>Liver Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADAMTS13 CD163</td>\n",
       "      <td>Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>745</th>\n",
       "      <td>WNT5A FZD2</td>\n",
       "      <td>Dental Disorders</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>746</th>\n",
       "      <td>WNT5A ROR2</td>\n",
       "      <td>Metabolic, Congenital Syndromes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747</th>\n",
       "      <td>WNT5A TLR4</td>\n",
       "      <td>Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>748</th>\n",
       "      <td>WNT8B FZD4</td>\n",
       "      <td>Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>749</th>\n",
       "      <td>YBX1 NOTCH3</td>\n",
       "      <td>Cancer</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>750 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        interaction                            Disease Type\n",
       "0         A2M HSPA5                                  Cancer\n",
       "1        ACE BDKRB2  Mental Health, Cardiovascular Diseases\n",
       "2      ADAM12 ITGA9                                  Cancer\n",
       "3      ADAM17 ITGB1                           Liver Disease\n",
       "4    ADAMTS13 CD163                                  Cancer\n",
       "..              ...                                     ...\n",
       "745      WNT5A FZD2                        Dental Disorders\n",
       "746      WNT5A ROR2         Metabolic, Congenital Syndromes\n",
       "747      WNT5A TLR4                                  Cancer\n",
       "748      WNT8B FZD4                                  Cancer\n",
       "749     YBX1 NOTCH3                                  Cancer\n",
       "\n",
       "[750 rows x 2 columns]"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "1a8e0819-44c2-42e6-a173-9ec9d788773a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create \"Cancer-related\" column based on whether \"Cancers & Neoplasms\" is in col2\n",
    "df_combined['Cancer-related'] = df_combined['Disease Type'].apply(lambda x: 'Yes' if 'Cancer' in x else 'No')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "id": "29273681-19aa-4ea9-a7f0-d6a58792f9ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interaction</th>\n",
       "      <th>Disease Type</th>\n",
       "      <th>Cancer-related</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACE BDKRB2</td>\n",
       "      <td>Mental Health, Cardiovascular Diseases</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADAM12 ITGA9</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADAM17 ITGB1</td>\n",
       "      <td>Liver Disease</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADAMTS13 CD163</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>745</th>\n",
       "      <td>WNT5A FZD2</td>\n",
       "      <td>Dental Disorders</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>746</th>\n",
       "      <td>WNT5A ROR2</td>\n",
       "      <td>Metabolic, Congenital Syndromes</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747</th>\n",
       "      <td>WNT5A TLR4</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>748</th>\n",
       "      <td>WNT8B FZD4</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>749</th>\n",
       "      <td>YBX1 NOTCH3</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>750 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        interaction                            Disease Type Cancer-related\n",
       "0         A2M HSPA5                                  Cancer            Yes\n",
       "1        ACE BDKRB2  Mental Health, Cardiovascular Diseases             No\n",
       "2      ADAM12 ITGA9                                  Cancer            Yes\n",
       "3      ADAM17 ITGB1                           Liver Disease             No\n",
       "4    ADAMTS13 CD163                                  Cancer            Yes\n",
       "..              ...                                     ...            ...\n",
       "745      WNT5A FZD2                        Dental Disorders             No\n",
       "746      WNT5A ROR2         Metabolic, Congenital Syndromes             No\n",
       "747      WNT5A TLR4                                  Cancer            Yes\n",
       "748      WNT8B FZD4                                  Cancer            Yes\n",
       "749     YBX1 NOTCH3                                  Cancer            Yes\n",
       "\n",
       "[750 rows x 3 columns]"
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d57c5e50-7fef-4a11-a8ba-89461725bf90",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_combined.to_csv(\"data/diseaseType_per_pair.csv\", index =False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b01534d-9167-42e0-a5dc-5c674c7cc871",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathway_list = df['source'].unique()\n",
    "len(disease_list)\n",
    "pair_list = df['interaction'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff356205-9ae3-4697-ba72-011b412b545c",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(pair_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "591228d8-5e7f-4a2f-8eb8-63619f232642",
   "metadata": {},
   "outputs": [],
   "source": [
    "uniquepairs = list(set(pair_list) & set(LR_pairs)) \n",
    "len(uniquepairs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04b78ab8-83a3-471c-ba42-8cdbad34e50c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get pair count per Disease\n",
    "pairPerDisease = disease_df.groupby('source')['interaction'].nunique().reset_index()\n",
    "pairPerDisease.to_csv(\"data/pairPerPathwayCount.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7adeb15e-7a45-4b57-af02-de7c9c1b6af3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get Disease count per LRPair\n",
    "pairPerDisease = disease_df.groupby('interaction')['source'].nunique().reset_index()\n",
    "pairPerDisease.to_csv(\"data/PathwayPerLRPair.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21b854e6-20a7-4c04-9268-e7df424c9bf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df[\"Gene Symbol\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46458bdf-7f37-4f86-9d03-27b894f15faa",
   "metadata": {},
   "outputs": [],
   "source": [
    "df= pd.read_table(\"data/human_uniprot_isoforms.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8602f29d-65b0-4cda-9f04-0b8ef3159a13",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb7e7d70-7f61-4786-bae3-759a8d509344",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[['UniProt ID', 'Gene Symbol', 'Isoform Type', 'FASTA Sequence']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d7aeca2-f763-4b85-8503-65921fc934aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88a97936-a9d2-471d-b287-3482fcfd357b",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df = gene_pair0[[\"Human LR Pair\", \"Ligand\", \"Receptor\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd3f4d8c-b251-4174-84de-727cd77ea189",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbb66bb9-0275-442d-b5bb-de269e41642d",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df = lim_df.merge(df, how='left', left_on='Ligand', right_on='Gene Symbol')\n",
    "lim_df = lim_df.drop(columns=[\"Gene Symbol\"])\n",
    "lim_df = lim_df.rename(columns={\"FASTA Sequence\": \"Ligand Sequence\",\n",
    "                                \"UniProt ID\": \"Ligand Isoform Uniprot ID\",\n",
    "                                \"Isoform Type\": \"Ligand Isoform Type\"})\n",
    "lim_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "336f774d-e37e-4d24-b415-d0eccb1eccb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df = lim_df.merge(df, how='left', left_on='Receptor', right_on='Gene Symbol')\n",
    "lim_df = lim_df.drop(columns=[\"Gene Symbol\"])\n",
    "lim_df = lim_df.rename(columns={\"FASTA Sequence\": \"Receptor Sequence\",\n",
    "                                \"UniProt ID\": \"Receptor Isoform Uniprot ID\",\n",
    "                                \"Isoform Type\": \"Receptor Isoform Type\"})\n",
    "lim_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af8e9511-4026-4ef3-866a-d52aabd55e18",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df.to_csv(\"data/LRpair_uniprot_sequences.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef772ddd-9d77-4557-a1e0-84605e6e084d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import re\n",
    "import pandas as pd\n",
    "\n",
    "# Step 1: Extract Gene Symbol Mapping from GTF\n",
    "gtf_file = \"data/gencode.v47.annotation.gtf.gz\"\n",
    "gene_map = {}\n",
    "\n",
    "# Read GTF file and extract gene_id -> gene_name mapping\n",
    "with gzip.open(gtf_file, \"rt\") as f:\n",
    "    for line in f:\n",
    "        if line.startswith(\"#\"):  # Skip comments\n",
    "            continue\n",
    "        \n",
    "        fields = line.strip().split(\"\\t\")\n",
    "        if fields[2] == \"gene\":  # Only extract gene entries\n",
    "            info = {key.strip(): value.strip('\"') for key, value in re.findall(r'(\\S+) \"([^\"]+)\"', fields[8])}\n",
    "            if \"gene_id\" in info and \"gene_name\" in info:\n",
    "                gene_map[info[\"gene_id\"]] = info[\"gene_name\"]\n",
    "\n",
    "print(f\"âœ… Extracted {len(gene_map)} gene mappings from GTF.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5251cc7d-871d-4d23-936a-20703a223ce4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 2: Parse GENCODE Protein FASTA and Add Gene Symbols\n",
    "fasta_file = \"data/gencode.v47.pc_translations.fa.gz\"\n",
    "\n",
    "# Store extracted data\n",
    "records = []\n",
    "\n",
    "# Open the GENCODE FASTA file and parse sequences\n",
    "with gzip.open(fasta_file, \"rt\") as f:\n",
    "    header = None\n",
    "    sequence = []\n",
    "    \n",
    "    for line in f:\n",
    "        line = line.strip()\n",
    "        \n",
    "        if line.startswith(\">\"):\n",
    "            # Store previous sequence if exists\n",
    "            if header and sequence:\n",
    "                # Extract the Gene Symbol using the Gene ID\n",
    "                gene_symbol = gene_map.get(header[\"gene_id\"], \"Unknown\")\n",
    "                isoform_type = \"Canonical\" if \"-1\" in header[\"protein_id\"] else \"Alternative Isoform\"\n",
    "                \n",
    "                # Append the parsed data to records\n",
    "                records.append([header[\"protein_id\"], header[\"transcript_id\"], header[\"gene_id\"], gene_symbol, isoform_type, \"\".join(sequence)])\n",
    "            \n",
    "            # Split header by '|' and extract necessary fields\n",
    "            fields = line[1:].split(\"|\")  # Skip the '>' symbol and split by '|'\n",
    "            if len(fields) >= 6:\n",
    "                header = {\n",
    "                    \"protein_id\": fields[0], \n",
    "                    \"transcript_id\": fields[1], \n",
    "                    \"gene_id\": fields[2]  \n",
    "                }\n",
    "                sequence = []\n",
    "            else:\n",
    "                header = None\n",
    "        \n",
    "        elif header:\n",
    "            sequence.append(line)\n",
    "\n",
    "    # Add the last record if needed\n",
    "    if header and sequence:\n",
    "        gene_symbol = gene_map.get(header[\"gene_id\"], \"Unknown\")\n",
    "        isoform_type = \"Canonical\" if \"-1\" in header[\"protein_id\"] else \"Alternative Isoform\"\n",
    "        records.append([header[\"protein_id\"], header[\"transcript_id\"], header[\"gene_id\"], gene_symbol, isoform_type, \"\".join(sequence)])\n",
    "\n",
    "# Step 3: Convert to pandas DataFrame and Save to TSV\n",
    "df = pd.DataFrame(records, columns=[\"Ensembl Protein ID\", \"Ensembl Transcript ID\", \"Ensembl Gene ID\", \"Gene Symbol\", \"Isoform Type\", \"FASTA Sequence\"])\n",
    "\n",
    "# Save to TSV\n",
    "df.to_csv(\"data/gencode_protein_isoforms_with_symbols.tsv\", sep=\"\\t\", index=False)\n",
    "\n",
    "# Print completion message\n",
    "print(f\"âœ… Extracted {len(df)} protein sequences with Gene Symbols and saved to 'gencode_protein_isoforms_with_symbols.tsv'.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e721088-e99d-4909-b523-8f334f3c8e5e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f343d57b-9a88-4ade-89f2-9cdf0ce2ef74",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df = gene_pair0[[\"Human LR Pair\", \"Ligand\", \"Receptor\"]]\n",
    "lim_df = lim_df.merge(df, how='left', left_on='Ligand', right_on='Gene Symbol')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e760dccb-b0f6-41e5-a8fc-554eb1909cd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dc1f428-00f5-4f40-8906-499ac5d8a459",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df = lim_df.drop(columns=[\"Gene Symbol\"])\n",
    "lim_df = lim_df.rename(columns={\"FASTA Sequence\": \"Ligand Sequence\",\n",
    "                                \"Ensembl Protein ID\": \"Ligand Ensembl Protein ID\",\n",
    "                                \"Ensembl Transcript ID\": \"Ligand Ensembl Transcript ID\",\n",
    "                                \"Ensembl Gene ID\": \"Ligand Ensembl Gene ID\",\n",
    "                                \"Isoform Type\": \"Ligand Isoform Type\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bc88ddd-5476-4f3d-8fa8-815c0827f101",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13dd1035-ecfd-462c-a6d5-61b59da565e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df = lim_df.merge(df, how='left', left_on='Receptor', right_on='Gene Symbol')\n",
    "lim_df = lim_df.drop(columns=[\"Gene Symbol\"])\n",
    "lim_df = lim_df.rename(columns={\"FASTA Sequence\": \"Receptor Sequence\",\n",
    "                                \"Ensembl Protein ID\": \"Receptor Ensembl Protein ID\",\n",
    "                                \"Ensembl Transcript ID\": \"Receptor Ensembl Transcript ID\",\n",
    "                                \"Ensembl Gene ID\": \"Receptor Ensembl Gene ID\",\n",
    "                                \"Isoform Type\": \"Receptor Isoform Type\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7028f462-74cc-40d7-a7d3-003689e862f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf876e97-f9fa-4ca2-bf21-f82861e132ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df.to_csv(\"data/LRpair_gencode_sequences.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c995394a-d081-4971-be36-38240103bc32",
   "metadata": {},
   "source": [
    "####################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2253500-50d6-4759-9ca2-c282eb8e2fc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to scrape data from Pubmed for Title, Abstract, Journal, and Year\n",
    "### IMPORTANT: TURN OFF VPN and make sure you have the data directory (from Sakura)\n",
    "\n",
    "import sys\n",
    "import requests\n",
    "import pandas as pd\n",
    "import time\n",
    "import os\n",
    "import xml.etree.ElementTree as ET\n",
    "\n",
    "sys.path.append(os.path.abspath(\"src\"))  \n",
    "import fetchGSheet\n",
    "\n",
    "# Read the API key from a file\n",
    "with open(\"data/ncbi_api_key.txt\", \"r\") as file:\n",
    "    ncbi_api_key = file.read().strip()\n",
    "\n",
    "# File to save the results\n",
    "output_file = \"data/pubmed_results.csv\"\n",
    "\n",
    "# Example of fetching HGNC gene symbols (you should have the `fetchGSheet.pop_up_info` dataframe ready)\n",
    "def extract_hgnc_symbols(fetchGSheet):\n",
    "    # Concatenate Approved, Alias, and Previous symbols, then extract unique symbols\n",
    "    hgnc_symbols = pd.concat([\n",
    "        fetchGSheet['Approved symbol'],\n",
    "        fetchGSheet['Alias symbol'],\n",
    "        fetchGSheet['Previous symbol']\n",
    "    ], axis=0).dropna().str.upper().unique()  # Remove NaNs and make uppercase for matching\n",
    "     # Remove any empty strings from the list\n",
    "    hgnc_symbols = [symbol for symbol in hgnc_symbols if symbol != \"\"]\n",
    "    return set(hgnc_symbols)  # Return as a set for fast lookup\n",
    "    \n",
    "hgnc_symbols = extract_hgnc_symbols(fetchGSheet.pop_up_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a012dcd5-7fe8-4289-ad32-6a5197f65484",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(hgnc_symbols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d48fc01-b7e3-4698-b8e8-a287dc069243",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Official species names and their corresponding terms (scientific names)\n",
    "# Load your list of PMIDs\n",
    "pmid_list = source\n",
    "species_dict = {\n",
    "    \"human\": \"Homo sapiens\",\n",
    "    \"mouse\": \"Mus musculus\",\n",
    "    \"rat\": \"Rattus norvegicus\",\n",
    "    \"rabbit\": \"Oryctolagus cuniculus\",\n",
    "    \"monkey\": \"Macaca spp.\",\n",
    "    \"dog\": \"Canis lupus familiaris\",\n",
    "    \"pig\": \"Sus scrofa\",\n",
    "    \"zebra fish\": \"Danio rerio\",\n",
    "    \"chicken\": \"Gallus gallus\",\n",
    "    \"horse\": \"Equus ferus caballus\",\n",
    "    \"cat\": \"Felis catus\",\n",
    "    \"sheep\": \"Ovis aries\",\n",
    "    \"cow\": \"Bos taurus\",\n",
    "    \"fruit fly\": \"Drosophila melanogaster\",\n",
    "    \"c. elegans\": \"Caenorhabditis elegans\",\n",
    "}\n",
    "\n",
    "def fetch_pubmed_data(pmid_list, hgnc_symbols):\n",
    "    base_url = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\"\n",
    "    results = []\n",
    "\n",
    "    # Load existing data if output file exists\n",
    "    if os.path.exists(output_file):\n",
    "        existing_data = pd.read_csv(output_file)\n",
    "    else:\n",
    "        existing_data = pd.DataFrame(columns=[\"PMID\", \"Title\", \"Abstract\", \"Journal\", \"Year\", \"Species\"])\n",
    "\n",
    "    # Split PMIDs into batches\n",
    "    batch_size = 50\n",
    "    pmid_batches = [pmid_list[i:i + batch_size] for i in range(0, len(pmid_list), batch_size)]\n",
    "\n",
    "    # Iterate over the batches\n",
    "    for batch in pmid_batches:\n",
    "        params = {\n",
    "            \"db\": \"pubmed\",\n",
    "            \"id\": \",\".join(batch),  # Join PMIDs as comma-separated\n",
    "            \"retmode\": \"xml\",\n",
    "            \"api_key\": ncbi_api_key\n",
    "        }\n",
    "\n",
    "        try:\n",
    "            response = requests.get(base_url, params=params)\n",
    "            response.raise_for_status()\n",
    "\n",
    "            # Parse the XML response\n",
    "            root = ET.fromstring(response.text)\n",
    "            for article in root.findall(\".//PubmedArticle\"):\n",
    "                # Extract Title and Abstract\n",
    "                title = article.findtext(\".//ArticleTitle\", default=\"N/A\")\n",
    "                abstract = article.findtext(\".//AbstractText\", default=\"No abstract available\")\n",
    "\n",
    "                # Extract Journal Title\n",
    "                journal_tag = article.find(\".//Journal/Title\")\n",
    "                journal = journal_tag.text.strip() if journal_tag is not None and journal_tag.text else \"N/A\"\n",
    "\n",
    "                # Extract Publication Year\n",
    "                pub_date = article.find(\".//PubDate\")\n",
    "                if pub_date is not None:\n",
    "                    year_tag = pub_date.find(\"Year\")\n",
    "                    year = year_tag.text if year_tag is not None else \"N/A\"\n",
    "\n",
    "                    # Fallback to MedlineDate if Year is missing\n",
    "                    if year == \"N/A\":\n",
    "                        medline_date_tag = pub_date.find(\"MedlineDate\")\n",
    "                        year = medline_date_tag.text.split()[0] if medline_date_tag is not None else \"N/A\"\n",
    "                else:\n",
    "                    year = \"N/A\"  # PubDate is completely missing\n",
    "\n",
    "                # Initialize species as N/A\n",
    "                species = \"N/A\"\n",
    "\n",
    "                # Check if the word \"patient\" is detected in title or abstract (assume human)\n",
    "                if \"patient\" in title.lower() or \"patient\" in abstract.lower():\n",
    "                    species = \"Homo sapiens\"\n",
    "                elif \"human\" in title.lower() or \"human\" in abstract.lower():\n",
    "                    species = \"Homo sapiens\"\n",
    "                else:\n",
    "                    # Look for HGNC gene symbols in title or abstract (assume human if found)\n",
    "                    for gene in hgnc_symbols:\n",
    "                        if gene in title or gene in abstract:\n",
    "                            species = \"Homo sapiens\"\n",
    "                            break\n",
    "                    else:\n",
    "                        # Look for MeSH terms related to species\n",
    "                        for mesh_heading in article.findall(\".//MeshHeadingList/MeshHeading\"):\n",
    "                            descriptor_name = mesh_heading.findtext(\"DescriptorName\")\n",
    "                            if descriptor_name:\n",
    "                                # Match official species names using the species_dict\n",
    "                                for species_term, scientific_name in species_dict.items():\n",
    "                                    if species_term in descriptor_name.lower():\n",
    "                                        species = scientific_name\n",
    "                                        break  # Stop after finding the first match\n",
    "\n",
    "                # Append the result\n",
    "                results.append({\n",
    "                    \"PMID\": article.findtext(\".//MedlineCitation/PMID\"),\n",
    "                    \"Title\": title,\n",
    "                    \"Abstract\": abstract,\n",
    "                    \"Journal\": journal,\n",
    "                    \"Year\": year,\n",
    "                    \"Species\": species\n",
    "                })\n",
    "\n",
    "        except Exception as e:\n",
    "            print(f\"Error fetching batch {batch}: {e}\")\n",
    "            # Optionally save the response for debugging\n",
    "            with open(f\"error_batch_{batch[0]}_{batch[-1]}.xml\", \"w\") as f:\n",
    "                f.write(response.text)\n",
    "\n",
    "        # Rate limiting to avoid API overload\n",
    "        time.sleep(1)  # Increase delay for better API compliance\n",
    "\n",
    "    # Save results\n",
    "    new_data = pd.DataFrame(results)\n",
    "    if not new_data.empty:\n",
    "        # Merge existing and new data, updating missing values\n",
    "        updated_data = pd.concat([existing_data, new_data])\n",
    "\n",
    "        # Ensure all PMIDs are strings\n",
    "        updated_data[\"PMID\"] = updated_data[\"PMID\"].astype(str)\n",
    "\n",
    "        # Drop rows with missing PMIDs\n",
    "        updated_data = updated_data.dropna(subset=[\"PMID\"])\n",
    "\n",
    "        # Ensure rows are ordered and remove duplicates\n",
    "        updated_data = (\n",
    "            updated_data.sort_values(by=\"PMID\")  # Ensure rows are ordered\n",
    "            .drop_duplicates(subset=\"PMID\", keep=\"last\")  # Keep the latest data\n",
    "        )\n",
    "        updated_data[\"Journal\"] = updated_data[\"Journal\"].str.split(\" (\", n=1, expand=False, regex=False).str[0]\n",
    "        updated_data.to_csv(output_file, index=False)\n",
    "    else:\n",
    "        print(\"No new data fetched.\")\n",
    "\n",
    "    return results\n",
    "\n",
    "# Fetch PubMed data with your list of PMIDs, output file path, and NCBI API key\n",
    "fetch_pubmed_data(pmid_list, hgnc_symbols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25f83505-39a9-42ef-8878-89f5044b1ad8",
   "metadata": {},
   "outputs": [],
   "source": [
    "from createDataTable import gene_pair, gene_pair000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a58bc65f-88c6-4af5-a3e0-577708ee43d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc6e12a4-6104-4c01-a652-632619e0ffac",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a9daffb-0dcc-4c57-92fd-a2857aaf7704",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from bs4 import BeautifulSoup\n",
    "\n",
    "# Example DataFrame with HTML tags in column names\n",
    "data = {\n",
    "    \"<span title='Double-click header of Sheep Ligand Ensembl ID to ensure all values are shown'>Sheep Ligand Ensembl ID&nbsp;</span>\": [1, 2, 3],\n",
    "    \"<span title='Double-click header of Cow Ligand Ensembl ID to ensure all values are shown'>Cow Ligand Ensembl ID&nbsp;</span>\": [4, 5, 6],\n",
    "    \"<span title='Double-click header of Dog Ligand Ensembl ID to ensure all values are shown'>Dog Ligand Ensembl ID&nbsp;</span>\": [7, 8, 9],\n",
    "}\n",
    "\n",
    "species_gene_pair = pd.DataFrame(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "774057f5-b0a0-4e17-92cf-770c2b01a089",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_gene_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57fdbc6c-074a-4050-9409-13714b24f62a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import liana as li\n",
    "import omnipath as op\n",
    "import decoupler as dc\n",
    "import pandas as pd\n",
    "\n",
    "import sys\n",
    "import os\n",
    "sys.path.append(os.path.abspath(\"src\"))  # Add src directory to path\n",
    "from createDataTable import gene_pair0\n",
    "\n",
    "### Parameters\n",
    "topN = 10000 #Number of top pathways to be included\n",
    "pathway_output_file=\"data/pathway_annotations_per_pair.csv\"\n",
    "output_file=\"data/disease_annotations_per_pair.csv\"\n",
    "\n",
    "### Pathway Annotations\n",
    "\n",
    "# load PROGENy pathways, we use decoupler as a proxy as it formats the data in a more convenient way\n",
    "progeny = dc.get_progeny(top=topN)\n",
    "# import connectomeDB database ligands and receptors\n",
    "lr_pairs = gene_pair0[[\"Ligand\", \"Receptor\"]]\n",
    "lr_pairs.columns = lr_pairs.columns.str.lower()\n",
    "\n",
    "# generate ligand-receptor geneset\n",
    "lr_progeny = li.rs.generate_lr_geneset(lr_pairs, progeny, lr_sep=\"^\")\n",
    "# some of the pairs are missing\n",
    "len(lr_progeny[\"interaction\"].unique())\n",
    "# Replace '^' with ' ' in the 2nd column\n",
    "lr_progeny.iloc[:, 1] = lr_progeny.iloc[:, 1].str.replace(r'\\^', ' ', regex=True)\n",
    "lr_progeny"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9ad6d33-3944-4243-b929-e35aea2afdb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_progeny.iloc[:, 1] = lr_progeny.iloc[:, 1].str.replace(r'\\^', ' ', regex=True)\n",
    "lr_progeny.iloc[:, 1] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6550957-f516-4b19-85dc-c51e76bf3991",
   "metadata": {},
   "outputs": [],
   "source": [
    "ligand_index =4\n",
    "ligand_col = [col for col in species_gene_pair.columns if \"Ligand&nbsp;\" in col][ligand_index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "439a3787-117b-4c21-8f2a-6ec4a314b116",
   "metadata": {},
   "outputs": [],
   "source": [
    "ligand_col"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97688da5-bdbd-4ded-89d6-1827eac7e8ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "    # Rename columns to remove species name\n",
    "    species_gene_pair.columns = [\n",
    "        col.replace(f\"{species} \", \"\").strip() if \"Ligand\" in col or \"Receptor\" in col else col\n",
    "        for col in species_gene_pair.columns\n",
    "    ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "649c7110-588c-4b51-8581-dababcf0dfce",
   "metadata": {},
   "outputs": [],
   "source": [
    "    ligand_col = [col for col in species_gene_pair.columns if \"Ligand&nbsp;\" in col][ligand_index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78335dcc-c5f1-4c7d-a489-1a2f302c4ddd",
   "metadata": {},
   "outputs": [],
   "source": [
    "ligand_col"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70e12f3d-805c-4160-a980-cafed9cfeccf",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_gene_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29823718-1aa5-4904-a441-d63c5de70e6e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "quarto_llm_env",
   "language": "python",
   "name": "quarto_llm_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
