{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fcc5ddc1-519e-439c-a130-3dd8bcd181bd",
   "metadata": {},
   "source": [
    "# Python Notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b443872e-66e7-427a-a031-75a16471e8a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_row(change):\n",
    "    global gene_pair\n",
    "    # Add a new row at the top with None values\n",
    "    new_row = {col: None for col in gene_pair.columns}\n",
    "    gene_pair = pd.DataFrame([new_row] + gene_pair.to_dict(orient=\"records\"))\n",
    "    update_table()\n",
    "\n",
    "# Function to remove the last row of the dataframe\n",
    "def remove_row(change):\n",
    "    global gene_pair\n",
    "    if len(gene_pair) > 0:\n",
    "        gene_pair = gene_pair[:-1]  # Remove the last row\n",
    "        update_table()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8374473-0394-42ee-80ea-d457bc41c8a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_5280/1947918405.py:63: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  interaction_card[\"Perplexity\"] = interaction_card[\"Perplexity\"].str.replace('size=30', 'size=80')\n",
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_5280/1947918405.py:76: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ligand_card_2[\"HGNC gene card\"] = ligand_card_2[\"Ligand HGNC ID\"].apply(convert_hgnc_url)\n",
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_5280/1947918405.py:77: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ligand_card_2[\"Disease relevance\"] = ligand_card_2[\"Ligand HGNC ID\"].apply(convert_hgnc_url_disease)\n",
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_5280/1947918405.py:78: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ligand_card_2[\"Expression Profile\"] = ligand_card_2[\"Ligand HGNC ID\"].apply(convert_hgnc_url_exp)\n",
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_5280/1947918405.py:88: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  receptor_card_2[\"HGNC gene card\"] = receptor_card_2[\"Receptor HGNC ID\"].apply(convert_hgnc_url)\n",
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_5280/1947918405.py:89: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  receptor_card_2[\"Disease relevance\"] = receptor_card_2[\"Receptor HGNC ID\"].apply(convert_hgnc_url_disease)\n",
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_5280/1947918405.py:90: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  receptor_card_2[\"Expression Profile\"] = receptor_card_2[\"Receptor HGNC ID\"].apply(convert_hgnc_url_exp)\n"
     ]
    }
   ],
   "source": [
    "## Function to create Ligand-Receptor pair cards\n",
    "\n",
    "import os\n",
    "import jinja2\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "sys.path.append(os.path.abspath(\"src\"))  \n",
    "import fetchGSheet\n",
    "from createDataTable import gene_pair0\n",
    "\n",
    "# Paths\n",
    "TEMPLATE_PATH = 'HTML/cardTemplate.html'\n",
    "OUTPUT_DIR = 'data/cards/'\n",
    "\n",
    "def load_template(template_path):\n",
    "    \"\"\"Load Jinja2 template from a file.\"\"\"\n",
    "    with open(template_path, 'r') as file:\n",
    "        return jinja2.Template(file.read())\n",
    "\n",
    "# Function to convert the links\n",
    "def convert_hgnc_url(col):\n",
    "   # Extract the HGNC ID from the original URL\n",
    "    hgnc_id = col.split(\"HGNC:\")[1].split('\"')[0]  # Extract the ID number (e.g., \"31702\")\n",
    "    # Extract the visible text inside the <a> tag\n",
    "    visible_text = \"genecard.org\" # simplify col.split(\">\")[1].split(\"<\")[0]  # Extract \"HGNC:31702\"\n",
    "    # Construct the new link, keeping the text intact\n",
    "    new_link = f'<a href=\"https://www.genecards.org/cgi-bin/carddisp.pl?id_type=hgnc&id={hgnc_id}\" target=\"_blank\">{visible_text}</a>'\n",
    "    return new_link\n",
    "\n",
    "\n",
    "def convert_hgnc_url_disease(col):\n",
    "   # Extract the HGNC ID from the original URL\n",
    "    hgnc_id = col.split(\"HGNC:\")[1].split('\"')[0]  # Extract the ID number (e.g., \"31702\")\n",
    "    # Extract the visible text inside the <a> tag\n",
    "    visible_text = \"see here\" #col.split(\">\")[1].split(\"<\")[0]  # Extract \"HGNC:31702\"\n",
    "    # Construct the new link, keeping the text intact\n",
    "    new_link = f'<a href=\"https://www.genecards.org/cgi-bin/carddisp.pl?id_type=hgnc&id={hgnc_id}#diseases\" target=\"_blank\">{visible_text}</a>'\n",
    "    return new_link\n",
    "    \n",
    "def convert_hgnc_url_exp(col):\n",
    "   # Extract the HGNC ID from the original URL\n",
    "    hgnc_id = col.split(\"HGNC:\")[1].split('\"')[0]  # Extract the ID number (e.g., \"31702\")\n",
    "    # Extract the visible text inside the <a> tag\n",
    "    visible_text = \"see here\" #col.split(\">\")[1].split(\"<\")[0]  # Extract \"HGNC:31702\"\n",
    "    # Construct the new link, keeping the text intact\n",
    "    new_link = f'<a href=\"https://www.genecards.org/cgi-bin/carddisp.pl?id_type=hgnc&id={hgnc_id}#expression\" target=\"_blank\">{visible_text}</a>'\n",
    "    return new_link\n",
    "\n",
    "def prepare_dataframes(gene_pair0):\n",
    "    \"\"\"Prepare interaction, ligand, and receptor dataframes.\"\"\"\n",
    "    DBlength = len(gene_pair0)\n",
    "    gene_pair0[\"Interaction ID\"] = [f\"CDB{str(i).zfill(4)}\" for i in range(1, DBlength + 1)]\n",
    "    gene_pair0[\"Interaction Type\"] = [\n",
    "        f'{ligand} {ligandLocation} ligand binds to {receptor} {receptorLocation} receptor'\n",
    "        for ligand, ligandLocation, receptor, receptorLocation in zip(\n",
    "            gene_pair0[\"Ligand\"], gene_pair0[\"Ligand location\"],\n",
    "            gene_pair0[\"Receptor\"], gene_pair0[\"Receptor location\"]\n",
    "        )\n",
    "    ]\n",
    "    interaction_card = gene_pair0[[\"Interaction ID\", \"Human LR Pair\", \"Interaction Type\", \"Perplexity\", \"PMID support\"]]\n",
    "    interaction_card[\"Perplexity\"] = interaction_card[\"Perplexity\"].str.replace('size=30', 'size=80')\n",
    "\n",
    "    pop_up_info_lim = fetchGSheet.pop_up_info[\n",
    "        [\"Approved symbol\", \"Alias symbol\", \"Previous symbol\", \"Date symbol changed\"]\n",
    "    ].drop_duplicates(subset=\"Approved symbol\", keep=\"first\")\n",
    "    \n",
    "    ligand_card = gene_pair0[[\"Human LR Pair\", \"Ligand\", \"Ligand name\", \"Ligand HGNC ID\", \"Ligand MGI ID\", \"Ligand RGD ID\", \"Ligand location\"]].merge(\n",
    "        pop_up_info_lim, how='left', left_on='Ligand', right_on='Approved symbol'\n",
    "    ).drop_duplicates(subset='Human LR Pair', keep=\"first\").drop(columns=[\"Ligand\", \"Approved symbol\"])\n",
    "\n",
    "    ligand_card_1 = ligand_card[[\"Human LR Pair\", \"Alias symbol\", \"Date symbol changed\", \"Ligand name\"]] \n",
    "    ligand_card_2 = ligand_card[[\"Human LR Pair\", \"Ligand HGNC ID\", \"Ligand location\"]] \n",
    "    # convert links\n",
    "    ligand_card_2[\"HGNC gene card\"] = ligand_card_2[\"Ligand HGNC ID\"].apply(convert_hgnc_url)\n",
    "    ligand_card_2[\"Disease relevance\"] = ligand_card_2[\"Ligand HGNC ID\"].apply(convert_hgnc_url_disease)\n",
    "    ligand_card_2[\"Expression Profile\"] = ligand_card_2[\"Ligand HGNC ID\"].apply(convert_hgnc_url_exp)\n",
    "    ligand_card_2 = ligand_card_2[[\"Human LR Pair\", \"Ligand HGNC ID\", \"HGNC gene card\", \"Disease relevance\", \"Expression Profile\", \"Ligand location\"]]       \n",
    "\n",
    "          \n",
    "    receptor_card = gene_pair0[[\"Human LR Pair\", \"Receptor\", \"Receptor name\", \"Receptor HGNC ID\", \"Receptor MGI ID\", \"Receptor RGD ID\", \"Receptor location\"]].merge(\n",
    "        pop_up_info_lim, how='left', left_on='Receptor', right_on='Approved symbol'\n",
    "    ).drop_duplicates(subset='Human LR Pair', keep=\"first\").drop(columns=[\"Receptor\", \"Approved symbol\"])\n",
    "    \n",
    "    receptor_card_1 = receptor_card[[\"Human LR Pair\", \"Alias symbol\", \"Date symbol changed\", \"Receptor name\"]] \n",
    "    receptor_card_2 = receptor_card[[\"Human LR Pair\", \"Receptor HGNC ID\", \"Receptor location\"]] \n",
    "    receptor_card_2[\"HGNC gene card\"] = receptor_card_2[\"Receptor HGNC ID\"].apply(convert_hgnc_url)\n",
    "    receptor_card_2[\"Disease relevance\"] = receptor_card_2[\"Receptor HGNC ID\"].apply(convert_hgnc_url_disease)\n",
    "    receptor_card_2[\"Expression Profile\"] = receptor_card_2[\"Receptor HGNC ID\"].apply(convert_hgnc_url_exp)\n",
    "    receptor_card_2 = receptor_card_2[[\"Human LR Pair\", \"Receptor HGNC ID\", \"HGNC gene card\", \"Disease relevance\", \"Expression Profile\", \"Receptor location\"]]       \n",
    "\n",
    "    return interaction_card, ligand_card_1, ligand_card_2,receptor_card_1, receptor_card_2\n",
    "\n",
    "def generate_html_files(template, interaction_card, ligand_card_1, receptor_card_1, ligand_card_2, receptor_card_2, output_dir):\n",
    "    \"\"\"Generate HTML files for each Human LR Pair.\"\"\"\n",
    "    column_values = interaction_card[\"Human LR Pair\"].dropna().unique()\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "    \n",
    "    for value in column_values:\n",
    "        value1, value2 = value.split()\n",
    "        row0 = interaction_card[interaction_card['Human LR Pair'] == value]\n",
    "        row1 = ligand_card_1[ligand_card_1['Human LR Pair'] == value]\n",
    "        row2 = receptor_card_1[receptor_card_1['Human LR Pair'] == value]\n",
    "        row3 = ligand_card_2[ligand_card_2['Human LR Pair'] == value]\n",
    "        row4 = receptor_card_2[receptor_card_2['Human LR Pair'] == value]\n",
    "        \n",
    "\n",
    "        table0_data = row0.drop('Human LR Pair', axis=1).to_dict(orient='records')[0] if not row0.empty else {}\n",
    "        table1_data = row1.drop('Human LR Pair', axis=1).to_dict(orient='records')[0] if not row1.empty else {}\n",
    "        table2_data = row2.drop('Human LR Pair', axis=1).to_dict(orient='records')[0] if not row2.empty else {}\n",
    "        table3_data = row3.drop('Human LR Pair', axis=1).to_dict(orient='records')[0] if not row3.empty else {}\n",
    "        table4_data = row4.drop('Human LR Pair', axis=1).to_dict(orient='records')[0] if not row4.empty else {}\n",
    "\n",
    "        rendered_content = template.render(\n",
    "            value1=value1,\n",
    "            value2=value2,\n",
    "            table0_data=table0_data,\n",
    "            table1_data=table1_data,\n",
    "            table2_data=table2_data,\n",
    "            table3_data=table3_data,\n",
    "            table4_data=table4_data\n",
    "        )\n",
    "        \n",
    "        output_file = os.path.join(output_dir, f\"{value1} {value2}.html\")\n",
    "        with open(output_file, 'w') as file:\n",
    "            file.write(rendered_content)\n",
    "\n",
    "# Main execution\n",
    "if __name__ == \"__main__\":\n",
    "    template = load_template(TEMPLATE_PATH)\n",
    "    interaction_card, ligand_card_1, receptor_card_1, ligand_card_2, receptor_card_2 = prepare_dataframes(gene_pair0)\n",
    "    generate_html_files(template, interaction_card, ligand_card_1, receptor_card_1, ligand_card_2, receptor_card_2, OUTPUT_DIR)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "327783b9-5ddf-44d0-8fee-51819b75e4d0",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"['ZFIN ID'] not in index\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[11], line 22\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[38;5;66;03m# Select only the relevant columns from pop_up_info\u001b[39;00m\n\u001b[1;32m     19\u001b[0m pop_up_info \u001b[38;5;241m=\u001b[39m fetchGSheet\u001b[38;5;241m.\u001b[39mpop_up_info\u001b[38;5;241m.\u001b[39mrename(columns\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMouse genome informatics (MGI) ID\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMGI ID\u001b[39m\u001b[38;5;124m\"\u001b[39m, \n\u001b[1;32m     20\u001b[0m                                                       \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mZebrafish genome database (ZFIN) ID\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mZFIN ID\u001b[39m\u001b[38;5;124m\"\u001b[39m})\n\u001b[0;32m---> 22\u001b[0m pop_up_info_lim \u001b[38;5;241m=\u001b[39m \u001b[43mpop_up_info\u001b[49m\u001b[43m[\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mApproved symbol\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mApproved name\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mMGI ID\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mZFIN ID\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m     23\u001b[0m pop_up_info_lim \u001b[38;5;241m=\u001b[39m pop_up_info_lim\u001b[38;5;241m.\u001b[39mdrop_duplicates(subset\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mApproved symbol\u001b[39m\u001b[38;5;124m\"\u001b[39m, keep\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfirst\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     25\u001b[0m \u001b[38;5;66;03m# Drop columns where all values are NA in gene_pair\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/frame.py:4108\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   4106\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m is_iterator(key):\n\u001b[1;32m   4107\u001b[0m         key \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(key)\n\u001b[0;32m-> 4108\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_indexer_strict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcolumns\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m[\u001b[38;5;241m1\u001b[39m]\n\u001b[1;32m   4110\u001b[0m \u001b[38;5;66;03m# take() does not accept boolean indexers\u001b[39;00m\n\u001b[1;32m   4111\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(indexer, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdtype\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m) \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mbool\u001b[39m:\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/indexes/base.py:6200\u001b[0m, in \u001b[0;36mIndex._get_indexer_strict\u001b[0;34m(self, key, axis_name)\u001b[0m\n\u001b[1;32m   6197\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   6198\u001b[0m     keyarr, indexer, new_indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reindex_non_unique(keyarr)\n\u001b[0;32m-> 6200\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_raise_if_missing\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkeyarr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis_name\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   6202\u001b[0m keyarr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtake(indexer)\n\u001b[1;32m   6203\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(key, Index):\n\u001b[1;32m   6204\u001b[0m     \u001b[38;5;66;03m# GH 42790 - Preserve name from an Index\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/indexes/base.py:6252\u001b[0m, in \u001b[0;36mIndex._raise_if_missing\u001b[0;34m(self, key, indexer, axis_name)\u001b[0m\n\u001b[1;32m   6249\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNone of [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mkey\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m] are in the [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00maxis_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m]\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   6251\u001b[0m not_found \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(ensure_index(key)[missing_mask\u001b[38;5;241m.\u001b[39mnonzero()[\u001b[38;5;241m0\u001b[39m]]\u001b[38;5;241m.\u001b[39munique())\n\u001b[0;32m-> 6252\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mnot_found\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m not in index\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mKeyError\u001b[0m: \"['ZFIN ID'] not in index\""
     ]
    }
   ],
   "source": [
    "## Function to prepare datatables (cleaning and hyperlinking, adding tool tips, etc) input for the database qmds\n",
    "import sys, os\n",
    "sys.path.append(os.path.abspath(\"src\"))  \n",
    "from itables import init_notebook_mode\n",
    "import pandas as pd\n",
    "from itables import show\n",
    "from itables import options\n",
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "import fetchGSheet \n",
    "import warnings\n",
    "\n",
    "# Suppress SettingWithCopyWarning\n",
    "warnings.simplefilter(\"ignore\", category=UserWarning)\n",
    "\n",
    "\n",
    "# Select only the relevant columns from pop_up_info\n",
    "\n",
    "pop_up_info = fetchGSheet.pop_up_info.rename(columns={\"Mouse genome informatics (MGI) ID\": \"MGI ID\", \n",
    "                                                      \"Zebrafish genome database (ZFIN) ID\": \"ZFIN ID\"})\n",
    "\n",
    "pop_up_info_lim = pop_up_info[[\"Approved symbol\", \"Approved name\", \"MGI ID\", \"ZFIN ID\"]]\n",
    "pop_up_info_lim = pop_up_info_lim.drop_duplicates(subset=\"Approved symbol\", keep=\"first\")\n",
    "\n",
    "# Drop columns where all values are NA in gene_pair\n",
    "gene_pair = fetchGSheet.gene_pair.dropna(axis=1, how='all')\n",
    "# Fetch species IDs from the dataset\n",
    "hgnc_id = [col for col in gene_pair.columns if \"HGNC ID\" in col]\n",
    "hgnc_id = pd.concat([gene_pair[col] for col in hgnc_id]).unique()\n",
    "\n",
    "# Rename columns for better clarity\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "    \"Ligand receptor pair\": \"Human LR Pair\",\n",
    "    \"Ligand gene symbol\": \"Ligand\",\n",
    "    \"Receptor gene symbol\": \"Receptor\",\n",
    "    \"Perplexity link\": \"Perplexity\",\n",
    "    \"Source\": \"Interaction Source\"\n",
    "})\n",
    "\n",
    "# Merge gene_pair with pop_up_info_lim for Ligand(L)\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Ligand', right_on='Approved symbol')\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Ligand name\", \n",
    "                                     \"MGI ID\": \"Ligand MGI ID\",\n",
    "                                     \"ZFIN ID\": \"Ligand ZFIN ID\"},\n",
    "                            )\n",
    "\n",
    "# Add MGI annotation\n",
    "MGI_info = pd.read_csv(\"data/MGI_ID_biomart.csv\")\n",
    "gene_pair = gene_pair.merge(MGI_info, how='left', left_on='Ligand MGI ID', right_on='MGI ID')\n",
    "\n",
    "# Add ZFIN annotation\n",
    "ZFIN_info = pd.read_csv(\"data/ZFIN_ID_biomart.csv\")\n",
    "ZFIN_info['ZFIN ID'] = \"ZFIN:\" + ZFIN_info['ZFIN ID'].astype(str)\n",
    "gene_pair = gene_pair.merge(ZFIN_info, how='left', left_on='Ligand ZFIN ID', right_on='ZFIN ID')\n",
    "\n",
    "# Add ZFIN id and symbol\n",
    "ZFIN_info = pd.read_csv(\"data/ZFIN_ID_human_orthos.txt\", sep=\"\\t\", skiprows=1)\n",
    "ZFIN_info = ZFIN_info[['ZFIN ID', 'ZFIN Symbol', 'ZFIN Name', 'HGNC ID']]\n",
    "ZFIN_info = ZFIN_info.dropna(subset=['HGNC ID'])\n",
    "ZFIN_info = ZFIN_info.drop_duplicates(subset=['HGNC ID'])\n",
    "ZFIN_info['HGNC ID'] = ZFIN_info['HGNC ID'].apply(lambda x: f'HGNC:{int(x)}')\n",
    "\n",
    "\n",
    "gene_pair = gene_pair.drop(columns=[\"ZFIN ID\", \"MGI ID\", \"HGNC ID\"])\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "                                     \"MGI name\": \"Mouse Ligand\", \n",
    "                                     \"ZFIN name\": \"Zebrafish Ligand\",\n",
    "                                     \"ZFIN ID\": \"Ligand ZFIN ID\",\n",
    "                                     \"ZFIN Symbol\": \"Zebrafish Ligand\",\n",
    "                                     \"ZFIN Name\": \"Zebrafish Ligand name\"}\n",
    "                            )\n",
    "\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Receptor', right_on='Approved symbol')\n",
    "gene_pair = gene_pair.merge(ZFIN_info, how='left', left_on='Ligand HGNC ID', right_on='HGNC ID')\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Receptor name\",\n",
    "                                      \"MGI ID\": \"Receptor MGI ID\",\n",
    "                                      \"ZFIN ID\": \"Receptor ZFIN ID\"}\n",
    "                            )\n",
    "\n",
    "# Add MGI name\n",
    "gene_pair = gene_pair.merge(MGI_info, how='left', left_on='Receptor MGI ID', right_on='MGI ID')\n",
    "gene_pair = gene_pair.merge(ZFIN_info, how='left', left_on='Receptor ZFIN ID', right_on='ZFIN ID')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "33d4d56d-e5dc-4729-bd87-6b1326312565",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to prepare datatables (cleaning and hyperlinking, adding tool tips, etc) input for the database qmds\n",
    "import sys, os\n",
    "sys.path.append(os.path.abspath(\"src\"))  \n",
    "from itables import init_notebook_mode\n",
    "import pandas as pd\n",
    "from itables import show\n",
    "from itables import options\n",
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "import fetchGSheet \n",
    "import warnings\n",
    "\n",
    "# Suppress SettingWithCopyWarning\n",
    "warnings.simplefilter(\"ignore\", category=UserWarning)\n",
    "\n",
    "\n",
    "# Select only the relevant columns from pop_up_info\n",
    "\n",
    "pop_up_info = fetchGSheet.pop_up_info.rename(columns={\"Mouse genome informatics (MGI) ID\": \"MGI ID\", \n",
    "                                                      \"Rat genome database (RGD) ID\": \"RGD ID\"})\n",
    "\n",
    "pop_up_info_lim = pop_up_info[[\"Approved symbol\", \"Approved name\", \"MGI ID\", \"RGD ID\"]]\n",
    "pop_up_info_lim = pop_up_info_lim.drop_duplicates(subset=\"Approved symbol\", keep=\"first\")\n",
    "\n",
    "# Drop columns where all values are NA in gene_pair\n",
    "gene_pair = fetchGSheet.gene_pair.dropna(axis=1, how='all')\n",
    "# Fetch species IDs from the dataset\n",
    "hgnc_id = [col for col in gene_pair.columns if \"HGNC ID\" in col]\n",
    "hgnc_id = pd.concat([gene_pair[col] for col in hgnc_id]).unique()\n",
    "\n",
    "# Rename columns for better clarity\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "    \"Ligand receptor pair\": \"Human LR Pair\",\n",
    "    \"Ligand gene symbol\": \"Ligand\",\n",
    "    \"Receptor gene symbol\": \"Receptor\",\n",
    "    \"Perplexity link\": \"Perplexity\",\n",
    "    \"Source\": \"Interaction Source\"\n",
    "})\n",
    "\n",
    "# Merge gene_pair with pop_up_info_lim for Ligand(L)\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Ligand', right_on='Approved symbol')\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Ligand name\", \n",
    "                                     \"MGI ID\": \"Ligand MGI ID\",\n",
    "                                     \"RGD ID\": \"Ligand RGD ID\"},\n",
    "                            )\n",
    "\n",
    "# Add MGI annotation\n",
    "MGI_info = pd.read_csv(\"data/MGI_ID_biomart.csv\")\n",
    "gene_pair = gene_pair.merge(MGI_info, how='left', left_on='Ligand MGI ID', right_on='MGI ID')\n",
    "\n",
    "# Add RGD annotation\n",
    "RGD_info = pd.read_csv(\"data/RGD_ID_biomart.csv\")\n",
    "RGD_info['RGD ID'] = \"RGD:\" + RGD_info['RGD ID'].astype(str)\n",
    "gene_pair = gene_pair.merge(RGD_info, how='left', left_on='Ligand RGD ID', right_on='RGD ID')\n",
    "\n",
    "# Add ZFIN id and symbol\n",
    "ZFIN_info = pd.read_csv(\"data/ZFIN_ID_human_orthos.txt\", sep=\"\\t\", skiprows=1)\n",
    "ZFIN_info = ZFIN_info[['ZFIN ID', 'ZFIN Symbol', 'ZFIN Name', 'HGNC ID']]\n",
    "\n",
    "ZFIN_info = ZFIN_info.dropna(subset=['HGNC ID'])\n",
    "ZFIN_info = ZFIN_info.drop_duplicates(subset=['HGNC ID'])\n",
    "ZFIN_info['HGNC ID'] = ZFIN_info['HGNC ID'].apply(lambda x: f'HGNC:{int(x)}')\n",
    "gene_pair = gene_pair.merge(ZFIN_info, how='left', left_on='Ligand HGNC ID', right_on='HGNC ID')\n",
    "\n",
    "gene_pair = gene_pair.drop(columns=[\"RGD ID\", \"MGI ID\", \"HGNC ID\"])\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "                                     \"MGI name\": \"Mouse Ligand\", \n",
    "                                     \"RGD name\": \"Rat Ligand\",\n",
    "                                     \"ZFIN ID\": \"Ligand ZFIN ID\",\n",
    "                                     \"ZFIN Symbol\": \"Zebrafish Ligand\",\n",
    "                                     \"ZFIN Name\": \"Zebrafish Ligand name\"}\n",
    "                            )\n",
    "\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Receptor', right_on='Approved symbol')\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Receptor name\",\n",
    "                                      \"MGI ID\": \"Receptor MGI ID\",\n",
    "                                      \"RGD ID\": \"Receptor RGD ID\"}\n",
    "                            )\n",
    "\n",
    "# Add MGI name\n",
    "gene_pair = gene_pair.merge(MGI_info, how='left', left_on='Receptor MGI ID', right_on='MGI ID')\n",
    "gene_pair = gene_pair.merge(RGD_info, how='left', left_on='Receptor RGD ID', right_on='RGD ID')\n",
    "gene_pair = gene_pair.merge(ZFIN_info, how='left', left_on='Receptor HGNC ID', right_on='HGNC ID')\n",
    "gene_pair = gene_pair.drop(columns=[\"RGD ID\", \"MGI ID\", \"HGNC ID\"])\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "                                     \"MGI name\": \"Mouse Receptor\", \n",
    "                                     \"RGD name\": \"Rat Receptor\",\n",
    "                                     \"ZFIN ID\": \"Receptor ZFIN ID\",\n",
    "                                     \"ZFIN Symbol\": \"Zebrafish Receptor\",\n",
    "                                     \"ZFIN Name\": \"Zebrafish Receptor name\"}\n",
    "                            )\n",
    "\n",
    "gene_pair = gene_pair.drop(columns=[\"Approved symbol_x\", \"Approved symbol_y\"])\n",
    "\n",
    "# Drop columns where all values are NA in gene_pair\n",
    "gene_pair = gene_pair.dropna(axis=1, how='all')\n",
    "\n",
    "gene_pair = gene_pair.fillna(\" \")\n",
    "gene_pair = gene_pair[gene_pair['Human LR Pair'] != ' ']\n",
    "\n",
    "if \"PMID link\" in gene_pair.columns:\n",
    "    gene_pair = gene_pair.drop(columns=[\"PMID link\"])\n",
    "\n",
    "# Add\n",
    "first_columns=['Human LR Pair', 'Ligand', 'Receptor', 'Interaction Source']\n",
    "\n",
    "end_columns=['HGNC L R', 'sanity check', 'curator', 'secondary source?']\n",
    "gene_pair = gene_pair[first_columns + [col for col in gene_pair.columns if col not in first_columns + end_columns] + end_columns]\n",
    "\n",
    "\n",
    "# number of unique vars\n",
    "\n",
    "lrPairsCount = len(gene_pair[\"Human LR Pair\"].unique())\n",
    "\n",
    "ligandCount = len(gene_pair[\"Ligand\"].unique())\n",
    "\n",
    "receptorCount = len(gene_pair[\"Receptor\"].unique())\n",
    "\n",
    "# Mouse Orthologue\n",
    "MouseLigandCount = len(gene_pair[\"Ligand MGI ID\"].unique())\n",
    "\n",
    "MouseReceptorCount = len(gene_pair[\"Receptor MGI ID\"].unique())\n",
    "\n",
    "# Rat Orthologue\n",
    "RatLigandCount = len(gene_pair[\"Ligand RGD ID\"].unique())\n",
    "\n",
    "RatReceptorCount = len(gene_pair[\"Receptor RGD ID\"].unique())\n",
    "\n",
    "gene_pair[\"PMID support\"] = [value.replace(\" \", \"\") for value in gene_pair[\"PMID support\"]]\n",
    "\n",
    "source = np.array(gene_pair[\"PMID support\"].unique())\n",
    "source = source.astype(str)\n",
    "source = \",\".join(sorted(set(filter(lambda x: x.lower() != 'nan', source))))\n",
    "\n",
    "# Split the string into individual elements, filter out empty strings, and get unique values\n",
    "source = sorted(\n",
    "    set(filter(lambda x: x.strip() and x.strip().lower() != 'nan', source.split(',')))\n",
    ")\n",
    "source = [value.replace(\" \", \"\") for value in source]\n",
    "sourceCount = len(source)\n",
    "\n",
    "# for creating PMIDs\n",
    "gene_pair00 = gene_pair[['Human LR Pair', 'PMID support']]\n",
    "\n",
    "# create URLs for the HGNC IDs\n",
    "\n",
    "# ligand\n",
    "gene_pair[\"Ligand HGNC ID\"] = [\n",
    "    '<a href=\"https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/{}\" target=\"_blank\">{}</a>'.format(ligand, ligand)\n",
    "    for ligand in gene_pair[\"Ligand HGNC ID\"]\n",
    "]\n",
    "\n",
    "# receptor\n",
    "gene_pair[\"Receptor HGNC ID\"] = [\n",
    "    '<a href=\"https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/{}\" target=\"_blank\">{}</a>'.format(receptor, receptor)\n",
    "    for receptor in gene_pair[\"Receptor HGNC ID\"]\n",
    "]\n",
    "\n",
    "# Perplexity\n",
    "gene_pair[\"Perplexity\"] = [\n",
    "    '<a href=\"{}\" target=\"_blank\"> <img src=\"https://img.icons8.com/?size=30&id=0NbBuNOxUwps&format=png&color=000000\" alt=\"Perplexity AI\" /></a>'.format(url)\n",
    "    for url in gene_pair[\"Perplexity\"]\n",
    "]\n",
    "\n",
    "# Function to generate hyperlinks for the \"PMID support\" column\n",
    "# Function to generate hyperlinks for the \"PMID support\" column\n",
    "def generate_links_with_doi(df, gene_column, pmid_column):\n",
    "    def create_link(gene, sources):\n",
    "        # Replace spaces with \"——\" in the gene name for the link\n",
    "        gene_name = gene.replace(\" \", \"——\")\n",
    "        \n",
    "        if len(sources) == 1:\n",
    "            source = sources[0]\n",
    "            if source.startswith(\"https://www.biorxiv.org/content/\"):\n",
    "                # If the value starts with \"https://doi.org/\", use it as the hyperlink\n",
    "                return f'<a href=\"{source}\" target=\"_blank\">BioRxiv preprint</a>'\n",
    "            else:\n",
    "                # If it's a single PMID, hyperlink the PMID text\n",
    "                return f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/pubmed/{gene_name}_pmid_details.html\">{source}</a>'\n",
    "        else:\n",
    "            # If multiple PMIDs, show the count and hyperlink to the page\n",
    "            return f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/pubmed/{gene_name}_pmid_details.html\" target=\"_blank\">{len(sources)} PMIDs</a>'\n",
    "\n",
    "    # Process each row to generate the \"PMID support\" column\n",
    "    df[\"PMID support\"] = [\n",
    "        create_link(\n",
    "            gene=row[gene_column], \n",
    "            sources=[s.strip() for s in row[pmid_column].split(',') if s.strip()]\n",
    "        )\n",
    "        for _, row in df.iterrows()\n",
    "    ]\n",
    "    return df\n",
    "\n",
    "\n",
    "# Generate the links for the \"PMID support\" column\n",
    "gene_pair = generate_links_with_doi(gene_pair, gene_column=\"Human LR Pair\", pmid_column=\"PMID support\")\n",
    "\n",
    "gene_pair[\"Ligand MGI ID\"] = [\n",
    "        f'<a href=\"https://www.informatics.jax.org/marker/{mouseOrth}\" target=\"_blank\">{mouseOrth}</a>' \n",
    "        if pd.notna(mouseOrth) and mouseOrth.strip() else \"\" \n",
    "        for mouseOrth in gene_pair[\"Ligand MGI ID\"]\n",
    "    ]\n",
    "\n",
    "gene_pair[\"Receptor MGI ID\"] = [\n",
    "        f'<a href=\"https://www.informatics.jax.org/marker/{mouseOrth}\" target=\"_blank\">{mouseOrth}</a>' \n",
    "        if pd.notna(mouseOrth) and mouseOrth.strip() else \"\" \n",
    "        for mouseOrth in gene_pair[\"Receptor MGI ID\"]\n",
    "    ]\n",
    "\n",
    "gene_pair[\"Ligand RGD ID\"] = [\n",
    "        f'<a href=\"https://rgd.mcw.edu/rgdweb/report/gene/main.html?id={ratOrth.replace(\"RGD:\", \"\")}\" target=\"_blank\">{ratOrth}</a>' \n",
    "        if pd.notna(ratOrth) and ratOrth.strip() else \"\" \n",
    "        for ratOrth in gene_pair[\"Ligand RGD ID\"]\n",
    "    ]\n",
    "\n",
    "gene_pair[\"Receptor RGD ID\"] = [\n",
    "        f'<a href=\"https://rgd.mcw.edu/rgdweb/report/gene/main.html?id={ratOrth.replace(\"RGD:\", \"\")}\" target=\"_blank\">{ratOrth}</a>' \n",
    "        if pd.notna(ratOrth) and ratOrth.strip() else \"\" \n",
    "        for ratOrth in gene_pair[\"Receptor RGD ID\"]\n",
    "    ]\n",
    "\n",
    "gene_pair[\"Zebrafish Ligand\"] = [\n",
    "    f'<span title=\"{ligand_name}\">{ligand_symbol}</span>'\n",
    "    for ligand_name, ligand_symbol in zip(gene_pair[\"Zebrafish Ligand name\"], \n",
    "                                          gene_pair[\"Zebrafish Ligand\"])\n",
    "]\n",
    "gene_pair[\"Zebrafish Receptor\"] = [\n",
    "    f'<span title=\"{ligand_name}\">{ligand_symbol}</span>'\n",
    "    for ligand_name, ligand_symbol in zip(gene_pair[\"Zebrafish Receptor name\"], \n",
    "                                          gene_pair[\"Zebrafish Receptor\"])\n",
    "]\n",
    "\n",
    "mouse_columns = ['Mouse Ligand', 'Mouse Receptor','Ligand MGI ID','Receptor MGI ID'] \n",
    "rat_columns = ['Rat Ligand','Rat Receptor','Ligand RGD ID','Receptor RGD ID']\n",
    "zebrafish = ['Zebrafish Ligand','Zebrafish Receptor','Ligand ZFIN ID','Receptor ZFIN ID']\n",
    "\n",
    "gene_pair0 = gene_pair[['Human LR Pair', 'Ligand', 'Receptor', 'Perplexity', 'PMID support',\n",
    "       'Ligand HGNC ID', 'Ligand location', 'Receptor HGNC ID',\n",
    "       'Receptor location', 'Ligand name', 'Receptor name'] + mouse_columns + rat_columns]\n",
    "\n",
    "gene_pair = gene_pair[['Human LR Pair', 'Ligand', 'Receptor', 'Interaction Source', 'Perplexity', 'PMID support',\n",
    "        'Ligand HGNC ID', 'Receptor HGNC ID', 'Ligand location', 'Receptor location',\n",
    "        'Ligand name', 'Receptor name'] + mouse_columns + rat_columns + zebrafish + end_columns]\n",
    "\n",
    "\n",
    "# gene symbol\n",
    "gene_pair[\"Ligand\"] = [\n",
    "    f'<span title=\"{ligand_name}\">{ligand_symbol}</span>'\n",
    "    for ligand_name, ligand_symbol in zip(gene_pair[\"Ligand name\"], \n",
    "                                          gene_pair[\"Ligand\"])\n",
    "]\n",
    "\n",
    "# gene symbol\n",
    "gene_pair[\"Receptor\"] = [\n",
    "    f'<span title=\"{receptor_name}\">{receptor_symbol}</span>'\n",
    "    for receptor_name, receptor_symbol in zip(gene_pair[\"Receptor name\"], \n",
    "                                              gene_pair[\"Receptor\"])\n",
    "]\n",
    "\n",
    "def replace_spaces(row):\n",
    "    if row['Ligand location'] == 'secreted':\n",
    "        return row['Human LR Pair'].replace(\" \", \" <span style='font-size: 15px;'>○</span> <span style='font-size: 24px;'>⤚</span> \")\n",
    "    elif row['Ligand location'] == '':\n",
    "        return row['Human LR Pair'].replace(\" \", \" <span style='font-size: 15px;'>○</span> <span style='font-size: 24px;'>⤚</span> \")\n",
    "    elif row['Ligand location'] == 'plasma membrane':\n",
    "        return row['Human LR Pair'].replace(\" \", \" <span style='font-size: 24px;'>⤙</span> <span style='font-size: 24px;'>⤚</span> \")\n",
    "    else:\n",
    "        return row['Human LR Pair'].replace(\" \", \" \\u2192 \")\n",
    "\n",
    "# Apply the function to the 'LR Pair' column\n",
    "gene_pair['Human LR Pair'] = gene_pair.apply(replace_spaces, axis=1)\n",
    "\n",
    "gene_pair = gene_pair.drop(columns=[\"Ligand name\", \"Receptor name\"])\n",
    "\n",
    "\n",
    "# Create the links to the HTML cards\n",
    "gene_pair[\"Human LR Pair\"] = [\n",
    "    f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/cards/{lrPairOrig}.html\">{lrPair}</a>'\n",
    "    for lrPairOrig, lrPair in zip(gene_pair0[\"Human LR Pair\"], gene_pair[\"Human LR Pair\"])\n",
    "]\n",
    "\n",
    "\n",
    "# Add tooltips to the column headers\n",
    "gene_pair.columns = [\n",
    "    f'<span title=\" Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\">{col}</span>' if col == \"Human LR Pair\" else\n",
    "    f'<span title=\"Click the logo below to run Perplexity on the Human LR pair\">{col}&nbsp;</span>' if col == \"Perplexity\" else\n",
    "    f'<span title=\" Official Gene Symbol; Hover on symbols below to show gene names\">{col}&nbsp;&nbsp;&nbsp;</span>' if col in [\"Ligand\", \"Receptor\"] else\n",
    "    f'<span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">{col}&nbsp;&nbsp;</span>' if col in [\"Ligand HGNC ID\", \"Receptor HGNC ID\"] else\n",
    "    f'<span title=\" PubMed IDs (PMID) with Literature Evidence for LR Interaction. Click on the link for more details\">{col}</span>' if col == \"PMID support\" else\n",
    "    f'<span title=\"Rat Genome Database (RGD) ID. Click on the link for more details\">{col}</span>' if col in [\"Ligand RGD ID\", \"Receptor RGD ID\"] else\n",
    "    f'<span title=\"Mouse Genome Informatics (MGI) ID. Click on the link for more details\">{col}</span>' if col in [\"Ligand MGI ID\", \"Receptor MGI ID\"]else\n",
    "    f'<span title=\"Zebrafish Information Network (ZFIN) ID. Click on the link for more details\">{col}</span>' if col in [\"Ligand ZFIN ID\", \"Receptor ZFIN ID\"] else\n",
    "    f'<span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\">{col}</span>' if col in [\"Ligand location\", \"Receptor location\"] else\n",
    "    f'<span title=\"Double-click header of {col} to ensure all values are shown\">{col}&nbsp;</span>'\n",
    "    for col in gene_pair.columns\n",
    "]\n",
    "\n",
    "gene_pair = gene_pair.reset_index(drop=True)  # Remove the index\n",
    "gene_pair000 = gene_pair.copy()\n",
    "\n",
    "keywords_to_modify = [\"Ligand\", \"Receptor\"]\n",
    "exclude_keywords = [\"HGNC ID\", \"Location\", \"Human\"]  # Columns containing this will not be modified\n",
    "\n",
    "# Copy the original columns so we can modify only the first 10\n",
    "new_columns = gene_pair000.columns.tolist()\n",
    "\n",
    "# Modify only the first 10 columns\n",
    "new_columns[:10] = [\n",
    "    f'{col.split(\">\")[0]}\">Human {col.split(\">\")[1]}</span>'\n",
    "    if any(keyword in col for keyword in keywords_to_modify) and not any(exclude in col for exclude in exclude_keywords)\n",
    "    else col\n",
    "    for col in new_columns[:10]\n",
    "]\n",
    "\n",
    "# Assign the modified column names back to the DataFrame\n",
    "gene_pair000.columns = new_columns\n",
    "human_columns = [col for col in gene_pair000.columns][:10]\n",
    "\n",
    "\n",
    "### MOUSE ###\n",
    "# Find columns with \"Mouse\" in the name\n",
    "mouse_columns = [col for col in gene_pair.columns if \"MGI\" in col or \"Mouse\" in col]\n",
    "\n",
    "# Filter rows where all \"Mouse\" columns are not \" \"\n",
    "mouse_gene_pair = gene_pair000[(gene_pair000[mouse_columns].map(str.strip) != \"\").all(axis=1)]\n",
    "# Dynamically identify columns containing \"Ligand\" and \"Receptor\" in their names \n",
    "# since it is now in span format\n",
    "\n",
    "new_columns = mouse_gene_pair.columns.tolist()\n",
    "\n",
    "new_columns = [\n",
    "    col.replace(\"Mouse \", \"\").strip()\n",
    "    if \"Mouse Ligand\" in col or \"Mouse Receptor\" in col\n",
    "    else col\n",
    "    for col in new_columns\n",
    "]\n",
    "mouse_gene_pair.columns = new_columns\n",
    "        \n",
    "ligand_col = [col for col in mouse_gene_pair.columns if \"Ligand&nbsp;\" in col][1]\n",
    "receptor_col = [col for col in mouse_gene_pair.columns if \"Receptor&nbsp;\" in col][1]\n",
    "ligand_location = [col for col in mouse_gene_pair.columns if \"Ligand location\" in col][0]\n",
    "receptor_location = [col for col in mouse_gene_pair.columns if \"Receptor location\" in col][0]\n",
    "\n",
    "\n",
    "# Combine columns into \"Mouse LR Pair\" with appropriate replacements\n",
    "def format_lr_pair(row):\n",
    "    if row[ligand_location] == 'secreted':\n",
    "        return f\"{row[ligand_col]} <span style='font-size: 15px;'>○</span> <span style='font-size: 24px;'>⤚</span> {row[receptor_col]}\"\n",
    "    elif row[ligand_location] == '':\n",
    "        return f\"{row[ligand_col]} <span style='font-size: 15px;'>○</span> <span style='font-size: 24px;'>⤚</span> {row[receptor_col]}\"\n",
    "    elif row[receptor_location] == 'plasma membrane':\n",
    "        return f\"{row[ligand_col]} <span style='font-size: 24px;'>⤙</span> <span style='font-size: 24px;'>⤚</span> {row[receptor_col]}\"\n",
    "    else:\n",
    "        return f\"{row[ligand_col]} \\u2192 {row[receptor_col]}\"\n",
    "\n",
    "# Apply the function row-wise and assign to the new column using .loc\n",
    "mouse_gene_pair1 = mouse_gene_pair.copy() \n",
    "mouse_gene_pair1.loc[:, \"Mouse LR Pair\"] = mouse_gene_pair1.apply(format_lr_pair, axis=1)\n",
    "mouse_columns = [col for col in mouse_gene_pair1.columns if \"MGI\" in col]\n",
    "# Reorder the DataFrame\n",
    "new_order = [\"Mouse LR Pair\", ligand_col, receptor_col] + mouse_columns + human_columns\n",
    "mouse_gene_pair1 = mouse_gene_pair1[new_order]\n",
    "MouselrPairsCount = len(mouse_gene_pair1[\"Mouse LR Pair\"].unique())\n",
    "HumanMouseLRPairsPer = (MouselrPairsCount/lrPairsCount)*100\n",
    "HumanMouseLRPairsPer = round(HumanMouseLRPairsPer, 2)\n",
    "\n",
    "mouse_gene_pair1 = mouse_gene_pair1.reset_index(drop=True)  \n",
    "\n",
    "\n",
    "### RAT ###\n",
    "\n",
    "## Limit to those with either Rat Ligand or Receptor\n",
    "rat_columns = [col for col in gene_pair.columns if \"RGD\" in col or \"Rat\" in col]\n",
    "# Filter rows where all \"Rat\" columns are not \" \"\n",
    "rat_gene_pair = gene_pair000[(gene_pair000[rat_columns].map(str.strip) != \"\").all(axis=1)]\n",
    "\n",
    "\n",
    "new_columns = rat_gene_pair.columns.tolist()\n",
    "\n",
    "new_columns = [\n",
    "    col.replace(\"Rat \", \"\").strip()\n",
    "    if \"Ligand\" in col or \"Receptor\" in col\n",
    "    else col\n",
    "    for col in new_columns\n",
    "]\n",
    "rat_gene_pair.columns = new_columns\n",
    "\n",
    "# Dynamically identify columns containing \"Ligand\" and \"Receptor\" in their names \n",
    "# since it is now in span format\n",
    "ligand_col = [col for col in rat_gene_pair.columns if \"Ligand&nbsp;\" in col][2]\n",
    "receptor_col = [col for col in rat_gene_pair.columns if \"Receptor&nbsp;\" in col][2]\n",
    "ligand_location = [col for col in rat_gene_pair.columns if \"Ligand location\" in col][0]\n",
    "receptor_location = [col for col in rat_gene_pair.columns if \"Receptor location\" in col][0]\n",
    "\n",
    "def format_lr_pair(row):\n",
    "    if row[ligand_location] == 'secreted':\n",
    "        return f\"{row[ligand_col]} <span style='font-size: 15px;'>○</span> <span style='font-size: 24px;'>⤚</span> {row[receptor_col]}\"\n",
    "    elif row[ligand_location] == '':\n",
    "        return f\"{row[ligand_col]} <span style='font-size: 15px;'>○</span> <span style='font-size: 24px;'>⤚</span> {row[receptor_col]}\"\n",
    "    elif row[receptor_location] == 'plasma membrane':\n",
    "        return f\"{row[ligand_col]} <span style='font-size: 24px;'>⤙</span> <span style='font-size: 24px;'>⤚</span> {row[receptor_col]}\"\n",
    "    else:\n",
    "        return f\"{row[ligand_col]} \\u2192 {row[receptor_col]}\"\n",
    "\n",
    "rat_gene_pair1 = rat_gene_pair.copy() \n",
    "rat_gene_pair1.loc[:, \"Rat LR Pair\"] = rat_gene_pair1.apply(format_lr_pair, axis=1)\n",
    "rat_columns = [col for col in rat_gene_pair1.columns if \"RGD\" in col]\n",
    "# Reorder the DataFrame\n",
    "new_order = [\"Rat LR Pair\", ligand_col, receptor_col] + rat_columns + human_columns\n",
    "rat_gene_pair1 = rat_gene_pair1[new_order]\n",
    "rat_gene_pair1 = rat_gene_pair1.reset_index(drop=True)  \n",
    "\n",
    "### ZEBRAFISH ###\n",
    "\n",
    "## Limit to those with either Zebrafish Ligand or Receptor\n",
    "Zebrafish_columns = [col for col in gene_pair.columns if \"ZFIN\" in col or \"Zebrafish\" in col]\n",
    "# Filter rows where all \"Zebrafish\" columns are not \" \"\n",
    "Zebrafish_gene_pair = gene_pair000[(gene_pair000[Zebrafish_columns].map(str.strip) != \"\").all(axis=1)]\n",
    "\n",
    "\n",
    "new_columns = Zebrafish_gene_pair.columns.tolist()\n",
    "\n",
    "new_columns = [\n",
    "    col.replace(\"Zebrafish \", \"\").strip()\n",
    "    if \"Ligand\" in col or \"Receptor\" in col\n",
    "    else col\n",
    "    for col in new_columns\n",
    "]\n",
    "Zebrafish_gene_pair.columns = new_columns\n",
    "\n",
    "# Dynamically identify columns containing \"Ligand\" and \"Receptor\" in their names \n",
    "# since it is now in span format\n",
    "ligand_col = [col for col in Zebrafish_gene_pair.columns if \"Ligand&nbsp;\" in col][3]\n",
    "receptor_col = [col for col in Zebrafish_gene_pair.columns if \"Receptor&nbsp;\" in col][3]\n",
    "ligand_location = [col for col in Zebrafish_gene_pair.columns if \"Ligand location\" in col][0]\n",
    "receptor_location = [col for col in Zebrafish_gene_pair.columns if \"Receptor location\" in col][0]\n",
    "\n",
    "def format_lr_pair(row):\n",
    "    if row[ligand_location] == 'secreted':\n",
    "        return f\"{row[ligand_col]} <span style='font-size: 15px;'>○</span> <span style='font-size: 24px;'>⤚</span> {row[receptor_col]}\"\n",
    "    elif row[ligand_location] == '':\n",
    "        return f\"{row[ligand_col]} <span style='font-size: 15px;'>○</span> <span style='font-size: 24px;'>⤚</span> {row[receptor_col]}\"\n",
    "    elif row[receptor_location] == 'plasma membrane':\n",
    "        return f\"{row[ligand_col]} <span style='font-size: 24px;'>⤙</span> <span style='font-size: 24px;'>⤚</span> {row[receptor_col]}\"\n",
    "    else:\n",
    "        return f\"{row[ligand_col]} \\u2192 {row[receptor_col]}\"\n",
    "\n",
    "Zebrafish_gene_pair1 = Zebrafish_gene_pair.copy() \n",
    "Zebrafish_gene_pair1.loc[:, \"Zebrafish LR Pair\"] = Zebrafish_gene_pair1.apply(format_lr_pair, axis=1)\n",
    "Zebrafish_columns = [col for col in Zebrafish_gene_pair1.columns if \"ZFIN\" in col]\n",
    "# Reorder the DataFrame\n",
    "new_order = [\"Zebrafish LR Pair\", ligand_col, receptor_col] + Zebrafish_columns + human_columns\n",
    "Zebrafish_gene_pair1 = Zebrafish_gene_pair1[new_order]\n",
    "Zebrafish_gene_pair1 = Zebrafish_gene_pair1.reset_index(drop=True)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "93a95969-73e1-42c1-b98b-dfdd8e10daf8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "228f917d-b5b9-409e-bc04-7b95afd9a725",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Human LR Pair', 'Ligand', 'Receptor', 'Interaction Source',\n",
       "       'PMID support', 'Ligand HGNC ID', 'Ligand location', 'Receptor HGNC ID',\n",
       "       'Receptor location', 'Perplexity', 'Ligand name', 'Ligand MGI ID',\n",
       "       'Ligand RGD ID', 'Mouse Ligand', 'Rat Ligand', 'Ligand ZFIN ID',\n",
       "       'Ligand ZFIN ID', 'Zebrafish Ligand', 'Zebrafish Ligand name',\n",
       "       'Receptor name', 'Receptor MGI ID', 'Receptor RGD ID', 'Mouse Receptor',\n",
       "       'Rat Receptor', 'Ligand ZFIN ID', 'Ligand ZFIN ID',\n",
       "       'Zebrafish Receptor', 'Zebrafish Receptor name', 'HGNC L R',\n",
       "       'sanity check', 'curator', 'secondary source?'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "19161b56-8989-458f-80ef-d1e3eb97e029",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2312       TAFA4 FPR1\n",
      "2333    SLAMF1 SLAMF1\n",
      "2342     FLRT3 ADGRL3\n",
      "2344     TENM2 ADGRL1\n",
      "Name: Human LR Pair, dtype: object\n"
     ]
    }
   ],
   "source": [
    "duplicates = gene_pair00[gene_pair00[\"Human LR Pair\"].duplicated()]\n",
    "print(duplicates[\"Human LR Pair\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fad95ba-e2c9-43a8-be98-cce3d1cbdf5c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "quarto_llm_env",
   "language": "python",
   "name": "quarto_llm_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
