{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fcc5ddc1-519e-439c-a130-3dd8bcd181bd",
   "metadata": {},
   "source": [
    "# Python Notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b443872e-66e7-427a-a031-75a16471e8a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_row(change):\n",
    "    global gene_pair\n",
    "    # Add a new row at the top with None values\n",
    "    new_row = {col: None for col in gene_pair.columns}\n",
    "    gene_pair = pd.DataFrame([new_row] + gene_pair.to_dict(orient=\"records\"))\n",
    "    update_table()\n",
    "\n",
    "# Function to remove the last row of the dataframe\n",
    "def remove_row(change):\n",
    "    global gene_pair\n",
    "    if len(gene_pair) > 0:\n",
    "        gene_pair = gene_pair[:-1]  # Remove the last row\n",
    "        update_table()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "327783b9-5ddf-44d0-8fee-51819b75e4d0",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"['HGNC ID'] not found in axis\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[61], line 65\u001b[0m\n\u001b[1;32m     61\u001b[0m ZFIN_info \u001b[38;5;241m=\u001b[39m ZFIN_info\u001b[38;5;241m.\u001b[39mdrop_duplicates(subset\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mHGNC ID\u001b[39m\u001b[38;5;124m'\u001b[39m])\n\u001b[1;32m     62\u001b[0m ZFIN_info[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mHGNC ID\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m ZFIN_info[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mHGNC ID\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mapply(\u001b[38;5;28;01mlambda\u001b[39;00m x: \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mHGNC:\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mint\u001b[39m(x)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m---> 65\u001b[0m gene_pair \u001b[38;5;241m=\u001b[39m \u001b[43mgene_pair\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdrop\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mRGD ID\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mMGI ID\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mHGNC ID\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     67\u001b[0m gene_pair \u001b[38;5;241m=\u001b[39m gene_pair\u001b[38;5;241m.\u001b[39mrename(columns\u001b[38;5;241m=\u001b[39m{\n\u001b[1;32m     68\u001b[0m                                      \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMGI name\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMouse Ligand\u001b[39m\u001b[38;5;124m\"\u001b[39m, \n\u001b[1;32m     69\u001b[0m                                      \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mRGD name\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mRat Ligand\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     72\u001b[0m                                      \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mZFIN Name\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mZebrafish Ligand name\u001b[39m\u001b[38;5;124m\"\u001b[39m}\n\u001b[1;32m     73\u001b[0m                             )\n\u001b[1;32m     75\u001b[0m gene_pair \u001b[38;5;241m=\u001b[39m gene_pair\u001b[38;5;241m.\u001b[39mmerge(pop_up_info_lim, how\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mleft\u001b[39m\u001b[38;5;124m'\u001b[39m, left_on\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mReceptor\u001b[39m\u001b[38;5;124m'\u001b[39m, right_on\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mApproved symbol\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/frame.py:5581\u001b[0m, in \u001b[0;36mDataFrame.drop\u001b[0;34m(self, labels, axis, index, columns, level, inplace, errors)\u001b[0m\n\u001b[1;32m   5433\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mdrop\u001b[39m(\n\u001b[1;32m   5434\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   5435\u001b[0m     labels: IndexLabel \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   5442\u001b[0m     errors: IgnoreRaise \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mraise\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m   5443\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m DataFrame \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   5444\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   5445\u001b[0m \u001b[38;5;124;03m    Drop specified labels from rows or columns.\u001b[39;00m\n\u001b[1;32m   5446\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   5579\u001b[0m \u001b[38;5;124;03m            weight  1.0     0.8\u001b[39;00m\n\u001b[1;32m   5580\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 5581\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdrop\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   5582\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlabels\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlabels\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   5583\u001b[0m \u001b[43m        \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   5584\u001b[0m \u001b[43m        \u001b[49m\u001b[43mindex\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mindex\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   5585\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   5586\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlevel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlevel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   5587\u001b[0m \u001b[43m        \u001b[49m\u001b[43minplace\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minplace\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   5588\u001b[0m \u001b[43m        \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   5589\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/generic.py:4788\u001b[0m, in \u001b[0;36mNDFrame.drop\u001b[0;34m(self, labels, axis, index, columns, level, inplace, errors)\u001b[0m\n\u001b[1;32m   4786\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m axis, labels \u001b[38;5;129;01min\u001b[39;00m axes\u001b[38;5;241m.\u001b[39mitems():\n\u001b[1;32m   4787\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m labels \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 4788\u001b[0m         obj \u001b[38;5;241m=\u001b[39m \u001b[43mobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_drop_axis\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlabels\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlevel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlevel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4790\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m inplace:\n\u001b[1;32m   4791\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_update_inplace(obj)\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/generic.py:4830\u001b[0m, in \u001b[0;36mNDFrame._drop_axis\u001b[0;34m(self, labels, axis, level, errors, only_slice)\u001b[0m\n\u001b[1;32m   4828\u001b[0m         new_axis \u001b[38;5;241m=\u001b[39m axis\u001b[38;5;241m.\u001b[39mdrop(labels, level\u001b[38;5;241m=\u001b[39mlevel, errors\u001b[38;5;241m=\u001b[39merrors)\n\u001b[1;32m   4829\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 4830\u001b[0m         new_axis \u001b[38;5;241m=\u001b[39m \u001b[43maxis\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdrop\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlabels\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4831\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m axis\u001b[38;5;241m.\u001b[39mget_indexer(new_axis)\n\u001b[1;32m   4833\u001b[0m \u001b[38;5;66;03m# Case for non-unique axis\u001b[39;00m\n\u001b[1;32m   4834\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/indexes/base.py:7070\u001b[0m, in \u001b[0;36mIndex.drop\u001b[0;34m(self, labels, errors)\u001b[0m\n\u001b[1;32m   7068\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m mask\u001b[38;5;241m.\u001b[39many():\n\u001b[1;32m   7069\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m errors \u001b[38;5;241m!=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mignore\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m-> 7070\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mlabels[mask]\u001b[38;5;241m.\u001b[39mtolist()\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m not found in axis\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   7071\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m indexer[\u001b[38;5;241m~\u001b[39mmask]\n\u001b[1;32m   7072\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdelete(indexer)\n",
      "\u001b[0;31mKeyError\u001b[0m: \"['HGNC ID'] not found in axis\""
     ]
    }
   ],
   "source": [
    "## Function to prepare datatables (cleaning and hyperlinking, adding tool tips, etc) input for the database qmds\n",
    "import sys, os\n",
    "sys.path.append(os.path.abspath(\"src\"))  \n",
    "from itables import init_notebook_mode\n",
    "import pandas as pd\n",
    "from itables import show\n",
    "from itables import options\n",
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "import fetchGSheet \n",
    "import warnings\n",
    "\n",
    "# Suppress SettingWithCopyWarning\n",
    "warnings.simplefilter(\"ignore\", category=UserWarning)\n",
    "\n",
    "\n",
    "# Select only the relevant columns from pop_up_info\n",
    "\n",
    "pop_up_info = fetchGSheet.pop_up_info.rename(columns={\"Mouse genome informatics (MGI) ID\": \"MGI ID\", \n",
    "                                                      \"Zebrafish genome database (ZFIN) ID\": \"ZFIN ID\"})\n",
    "\n",
    "pop_up_info_lim = pop_up_info[[\"Approved symbol\", \"Approved name\", \"MGI ID\", \"ZFIN ID\"]]\n",
    "pop_up_info_lim = pop_up_info_lim.drop_duplicates(subset=\"Approved symbol\", keep=\"first\")\n",
    "\n",
    "# Drop columns where all values are NA in gene_pair\n",
    "gene_pair = fetchGSheet.gene_pair.dropna(axis=1, how='all')\n",
    "# Fetch species IDs from the dataset\n",
    "hgnc_id = [col for col in gene_pair.columns if \"HGNC ID\" in col]\n",
    "hgnc_id = pd.concat([gene_pair[col] for col in hgnc_id]).unique()\n",
    "\n",
    "# Rename columns for better clarity\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "    \"Ligand receptor pair\": \"Human LR Pair\",\n",
    "    \"Ligand gene symbol\": \"Ligand\",\n",
    "    \"Receptor gene symbol\": \"Receptor\",\n",
    "    \"Perplexity link\": \"Perplexity\",\n",
    "    \"Source\": \"Interaction Source\"\n",
    "})\n",
    "\n",
    "# Merge gene_pair with pop_up_info_lim for Ligand(L)\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Ligand', right_on='Approved symbol')\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Ligand name\", \n",
    "                                     \"MGI ID\": \"Ligand MGI ID\",\n",
    "                                     \"ZFIN ID\": \"Ligand ZFIN ID\"},\n",
    "                            )\n",
    "\n",
    "# Add MGI annotation\n",
    "MGI_info = pd.read_csv(\"data/MGI_ID_biomart.csv\")\n",
    "gene_pair = gene_pair.merge(MGI_info, how='left', left_on='Ligand MGI ID', right_on='MGI ID')\n",
    "\n",
    "# Add ZFIN annotation\n",
    "ZFIN_info = pd.read_csv(\"data/ZFIN_ID_biomart.csv\")\n",
    "ZFIN_info['ZFIN ID'] = \"ZFIN:\" + ZFIN_info['ZFIN ID'].astype(str)\n",
    "gene_pair = gene_pair.merge(ZFIN_info, how='left', left_on='Ligand ZFIN ID', right_on='ZFIN ID')\n",
    "\n",
    "# Add ZFIN id and symbol\n",
    "ZFIN_info = pd.read_csv(\"data/ZFIN_ID_human_orthos.txt\", sep=\"\\t\", skiprows=1)\n",
    "ZFIN_info = ZFIN_info[['ZFIN ID', 'ZFIN Symbol', 'ZFIN Name', 'HGNC ID']]\n",
    "ZFIN_info = ZFIN_info.dropna(subset=['HGNC ID'])\n",
    "ZFIN_info = ZFIN_info.drop_duplicates(subset=['HGNC ID'])\n",
    "ZFIN_info['HGNC ID'] = ZFIN_info['HGNC ID'].apply(lambda x: f'HGNC:{int(x)}')\n",
    "\n",
    "\n",
    "gene_pair = gene_pair.drop(columns=[\"ZFIN ID\", \"MGI ID\", \"HGNC ID\"])\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "                                     \"MGI name\": \"Mouse Ligand\", \n",
    "                                     \"ZFIN name\": \"Zebrafish Ligand\",\n",
    "                                     \"ZFIN ID\": \"Ligand ZFIN ID\",\n",
    "                                     \"ZFIN Symbol\": \"Zebrafish Ligand\",\n",
    "                                     \"ZFIN Name\": \"Zebrafish Ligand name\"}\n",
    "                            )\n",
    "\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Receptor', right_on='Approved symbol')\n",
    "gene_pair = gene_pair.merge(ZFIN_info, how='left', left_on='Ligand HGNC ID', right_on='HGNC ID')\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Receptor name\",\n",
    "                                      \"MGI ID\": \"Receptor MGI ID\",\n",
    "                                      \"ZFIN ID\": \"Receptor ZFIN ID\"}\n",
    "                            )\n",
    "\n",
    "# Add MGI name\n",
    "gene_pair = gene_pair.merge(MGI_info, how='left', left_on='Receptor MGI ID', right_on='MGI ID')\n",
    "gene_pair = gene_pair.merge(ZFIN_info, how='left', left_on='Receptor ZFIN ID', right_on='ZFIN ID')\n",
    "ZFIN_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "33d4d56d-e5dc-4729-bd87-6b1326312565",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Human LR Pair</th>\n",
       "      <th>Interaction Source</th>\n",
       "      <th>PMID support</th>\n",
       "      <th>Ligand</th>\n",
       "      <th>Ligand HGNC ID</th>\n",
       "      <th>Ligand location</th>\n",
       "      <th>Receptor</th>\n",
       "      <th>Receptor HGNC ID</th>\n",
       "      <th>Receptor location</th>\n",
       "      <th>HGNC L R</th>\n",
       "      <th>...</th>\n",
       "      <th>Ligand RGD ID</th>\n",
       "      <th>Mouse Ligand</th>\n",
       "      <th>Rat Ligand</th>\n",
       "      <th>Ligand ZFIN ID</th>\n",
       "      <th>Zebrafish Ligand</th>\n",
       "      <th>Zebrafish Ligand name</th>\n",
       "      <th>Approved symbol_y</th>\n",
       "      <th>Receptor name</th>\n",
       "      <th>Receptor MGI ID</th>\n",
       "      <th>Receptor RGD ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCL3L3 ACKR2</td>\n",
       "      <td>Ramilowski_2015_Literature_supported</td>\n",
       "      <td>10364178,</td>\n",
       "      <td>CCL3L3</td>\n",
       "      <td>HGNC:30554</td>\n",
       "      <td>secreted</td>\n",
       "      <td>ACKR2</td>\n",
       "      <td>HGNC:1565</td>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>HGNC:30554 HGNC:1565</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>Ccl3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACKR2</td>\n",
       "      <td>atypical chemokine receptor 2</td>\n",
       "      <td>MGI:1891697</td>\n",
       "      <td>RGD:620323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DEFB103B CCR2</td>\n",
       "      <td>Ramilowski_2015_Literature_supported</td>\n",
       "      <td>23390582,</td>\n",
       "      <td>DEFB103B</td>\n",
       "      <td>HGNC:31702</td>\n",
       "      <td>secreted</td>\n",
       "      <td>CCR2</td>\n",
       "      <td>HGNC:1603</td>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>HGNC:31702 HGNC:1603</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>Defb14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCR2</td>\n",
       "      <td>C-C motif chemokine receptor 2</td>\n",
       "      <td>MGI:106185</td>\n",
       "      <td>RGD:620876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CCL3L3 CCR5</td>\n",
       "      <td>Ramilowski_2015_Literature_supported</td>\n",
       "      <td>11734558, 10364178,</td>\n",
       "      <td>CCL3L3</td>\n",
       "      <td>HGNC:30554</td>\n",
       "      <td>secreted</td>\n",
       "      <td>CCR5</td>\n",
       "      <td>HGNC:1606</td>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>HGNC:30554 HGNC:1606</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>Ccl3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCR5</td>\n",
       "      <td>C-C motif chemokine receptor 5</td>\n",
       "      <td>MGI:107182</td>\n",
       "      <td>RGD:620596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DEFB103B CCR6</td>\n",
       "      <td>Ramilowski_2015_Literature_supported</td>\n",
       "      <td>23390582,</td>\n",
       "      <td>DEFB103B</td>\n",
       "      <td>HGNC:31702</td>\n",
       "      <td>secreted</td>\n",
       "      <td>CCR6</td>\n",
       "      <td>HGNC:1607</td>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>HGNC:31702 HGNC:1607</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>Defb14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCR6</td>\n",
       "      <td>C-C motif chemokine receptor 6</td>\n",
       "      <td>MGI:1333797</td>\n",
       "      <td>RGD:1308562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DEFB4A CCR6</td>\n",
       "      <td>Ramilowski_2015_Literature_supported</td>\n",
       "      <td>10521347,11714836,</td>\n",
       "      <td>DEFB4A</td>\n",
       "      <td>HGNC:2767</td>\n",
       "      <td>secreted</td>\n",
       "      <td>CCR6</td>\n",
       "      <td>HGNC:1607</td>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>HGNC:2767 HGNC:1607</td>\n",
       "      <td>...</td>\n",
       "      <td>RGD:1563630</td>\n",
       "      <td>Defb3</td>\n",
       "      <td>Defb5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCR6</td>\n",
       "      <td>C-C motif chemokine receptor 6</td>\n",
       "      <td>MGI:1333797</td>\n",
       "      <td>RGD:1308562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2361</th>\n",
       "      <td>KIR2DL5A PVR</td>\n",
       "      <td>ConnectomeDB2025 (this publication)</td>\n",
       "      <td>36377656,</td>\n",
       "      <td>KIR2DL5A</td>\n",
       "      <td>HGNC:16345</td>\n",
       "      <td></td>\n",
       "      <td>PVR</td>\n",
       "      <td>HGNC:9705</td>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>HGNC:16345 HGNC:9705</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>Kir3dl1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PVR</td>\n",
       "      <td>PVR cell adhesion molecule</td>\n",
       "      <td>MGI:107741</td>\n",
       "      <td>RGD:3813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2362</th>\n",
       "      <td>SAA1 SCARB1</td>\n",
       "      <td>ConnectomeDB2025 (this publication)</td>\n",
       "      <td>15561721,</td>\n",
       "      <td>SAA1</td>\n",
       "      <td>HGNC:10513</td>\n",
       "      <td></td>\n",
       "      <td>SCARB1</td>\n",
       "      <td>HGNC:1664</td>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>HGNC:10513 HGNC:1664</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>Saa1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZDB-GENE-040927-15</td>\n",
       "      <td>saa</td>\n",
       "      <td>serum amyloid A</td>\n",
       "      <td>SCARB1</td>\n",
       "      <td>scavenger receptor class B member 1</td>\n",
       "      <td>MGI:893578</td>\n",
       "      <td>RGD:2302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2363</th>\n",
       "      <td>SAA1 TLR2</td>\n",
       "      <td>ConnectomeDB2025 (this publication)</td>\n",
       "      <td>18566366,</td>\n",
       "      <td>SAA1</td>\n",
       "      <td>HGNC:10513</td>\n",
       "      <td></td>\n",
       "      <td>TLR2</td>\n",
       "      <td>HGNC:11848</td>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>HGNC:10513 HGNC:11848</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>Saa1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZDB-GENE-040927-15</td>\n",
       "      <td>saa</td>\n",
       "      <td>serum amyloid A</td>\n",
       "      <td>TLR2</td>\n",
       "      <td>toll like receptor 2</td>\n",
       "      <td>MGI:1346060</td>\n",
       "      <td>RGD:735138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2364</th>\n",
       "      <td>SAA1 TLR4</td>\n",
       "      <td>ConnectomeDB2025 (this publication)</td>\n",
       "      <td>35247611,</td>\n",
       "      <td>SAA1</td>\n",
       "      <td>HGNC:10513</td>\n",
       "      <td></td>\n",
       "      <td>TLR4</td>\n",
       "      <td>HGNC:11850</td>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>HGNC:10513 HGNC:11850</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>Saa1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZDB-GENE-040927-15</td>\n",
       "      <td>saa</td>\n",
       "      <td>serum amyloid A</td>\n",
       "      <td>TLR4</td>\n",
       "      <td>toll like receptor 4</td>\n",
       "      <td>MGI:96824</td>\n",
       "      <td>RGD:3870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2365</th>\n",
       "      <td>SAA3P LY96</td>\n",
       "      <td>ConnectomeDB2025 (this publication)</td>\n",
       "      <td>23858030,</td>\n",
       "      <td>SAA3P</td>\n",
       "      <td>HGNC:10515</td>\n",
       "      <td></td>\n",
       "      <td>LY96</td>\n",
       "      <td>HGNC:17156</td>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>HGNC:10515 HGNC:17156</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>Saa3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LY96</td>\n",
       "      <td>lymphocyte antigen 96</td>\n",
       "      <td>MGI:1341909</td>\n",
       "      <td>RGD:1302993</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2366 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Human LR Pair                    Interaction Source  \\\n",
       "0      CCL3L3 ACKR2  Ramilowski_2015_Literature_supported   \n",
       "1     DEFB103B CCR2  Ramilowski_2015_Literature_supported   \n",
       "2       CCL3L3 CCR5  Ramilowski_2015_Literature_supported   \n",
       "3     DEFB103B CCR6  Ramilowski_2015_Literature_supported   \n",
       "4       DEFB4A CCR6  Ramilowski_2015_Literature_supported   \n",
       "...             ...                                   ...   \n",
       "2361   KIR2DL5A PVR   ConnectomeDB2025 (this publication)   \n",
       "2362    SAA1 SCARB1   ConnectomeDB2025 (this publication)   \n",
       "2363      SAA1 TLR2   ConnectomeDB2025 (this publication)   \n",
       "2364      SAA1 TLR4   ConnectomeDB2025 (this publication)   \n",
       "2365     SAA3P LY96   ConnectomeDB2025 (this publication)   \n",
       "\n",
       "             PMID support    Ligand Ligand HGNC ID Ligand location Receptor  \\\n",
       "0               10364178,    CCL3L3     HGNC:30554        secreted    ACKR2   \n",
       "1               23390582,  DEFB103B     HGNC:31702        secreted     CCR2   \n",
       "2     11734558, 10364178,    CCL3L3     HGNC:30554        secreted     CCR5   \n",
       "3               23390582,  DEFB103B     HGNC:31702        secreted     CCR6   \n",
       "4      10521347,11714836,    DEFB4A      HGNC:2767        secreted     CCR6   \n",
       "...                   ...       ...            ...             ...      ...   \n",
       "2361            36377656,  KIR2DL5A     HGNC:16345                      PVR   \n",
       "2362            15561721,      SAA1     HGNC:10513                   SCARB1   \n",
       "2363            18566366,      SAA1     HGNC:10513                     TLR2   \n",
       "2364            35247611,      SAA1     HGNC:10513                     TLR4   \n",
       "2365            23858030,     SAA3P     HGNC:10515                     LY96   \n",
       "\n",
       "     Receptor HGNC ID Receptor location               HGNC L R  ...  \\\n",
       "0           HGNC:1565   plasma membrane   HGNC:30554 HGNC:1565  ...   \n",
       "1           HGNC:1603   plasma membrane   HGNC:31702 HGNC:1603  ...   \n",
       "2           HGNC:1606   plasma membrane   HGNC:30554 HGNC:1606  ...   \n",
       "3           HGNC:1607   plasma membrane   HGNC:31702 HGNC:1607  ...   \n",
       "4           HGNC:1607   plasma membrane    HGNC:2767 HGNC:1607  ...   \n",
       "...               ...               ...                    ...  ...   \n",
       "2361        HGNC:9705   plasma membrane   HGNC:16345 HGNC:9705  ...   \n",
       "2362        HGNC:1664   plasma membrane   HGNC:10513 HGNC:1664  ...   \n",
       "2363       HGNC:11848   plasma membrane  HGNC:10513 HGNC:11848  ...   \n",
       "2364       HGNC:11850   plasma membrane  HGNC:10513 HGNC:11850  ...   \n",
       "2365       HGNC:17156   plasma membrane  HGNC:10515 HGNC:17156  ...   \n",
       "\n",
       "     Ligand RGD ID Mouse Ligand Rat Ligand      Ligand ZFIN ID  \\\n",
       "0                          Ccl3        NaN                 NaN   \n",
       "1                        Defb14        NaN                 NaN   \n",
       "2                          Ccl3        NaN                 NaN   \n",
       "3                        Defb14        NaN                 NaN   \n",
       "4      RGD:1563630        Defb3      Defb5                 NaN   \n",
       "...            ...          ...        ...                 ...   \n",
       "2361                    Kir3dl1        NaN                 NaN   \n",
       "2362                       Saa1        NaN  ZDB-GENE-040927-15   \n",
       "2363                       Saa1        NaN  ZDB-GENE-040927-15   \n",
       "2364                       Saa1        NaN  ZDB-GENE-040927-15   \n",
       "2365                       Saa3        NaN                 NaN   \n",
       "\n",
       "     Zebrafish Ligand Zebrafish Ligand name Approved symbol_y  \\\n",
       "0                 NaN                   NaN             ACKR2   \n",
       "1                 NaN                   NaN              CCR2   \n",
       "2                 NaN                   NaN              CCR5   \n",
       "3                 NaN                   NaN              CCR6   \n",
       "4                 NaN                   NaN              CCR6   \n",
       "...               ...                   ...               ...   \n",
       "2361              NaN                   NaN               PVR   \n",
       "2362              saa       serum amyloid A            SCARB1   \n",
       "2363              saa       serum amyloid A              TLR2   \n",
       "2364              saa       serum amyloid A              TLR4   \n",
       "2365              NaN                   NaN              LY96   \n",
       "\n",
       "                            Receptor name Receptor MGI ID Receptor RGD ID  \n",
       "0           atypical chemokine receptor 2     MGI:1891697      RGD:620323  \n",
       "1          C-C motif chemokine receptor 2      MGI:106185      RGD:620876  \n",
       "2          C-C motif chemokine receptor 5      MGI:107182      RGD:620596  \n",
       "3          C-C motif chemokine receptor 6     MGI:1333797     RGD:1308562  \n",
       "4          C-C motif chemokine receptor 6     MGI:1333797     RGD:1308562  \n",
       "...                                   ...             ...             ...  \n",
       "2361           PVR cell adhesion molecule      MGI:107741        RGD:3813  \n",
       "2362  scavenger receptor class B member 1      MGI:893578        RGD:2302  \n",
       "2363                 toll like receptor 2     MGI:1346060      RGD:735138  \n",
       "2364                 toll like receptor 4       MGI:96824        RGD:3870  \n",
       "2365                lymphocyte antigen 96     MGI:1341909     RGD:1302993  \n",
       "\n",
       "[2366 rows x 28 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Function to prepare datatables (cleaning and hyperlinking, adding tool tips, etc) input for the database qmds\n",
    "import sys, os\n",
    "sys.path.append(os.path.abspath(\"src\"))  \n",
    "from itables import init_notebook_mode\n",
    "import pandas as pd\n",
    "from itables import show\n",
    "from itables import options\n",
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "import fetchGSheet \n",
    "import warnings\n",
    "\n",
    "# Suppress SettingWithCopyWarning\n",
    "warnings.simplefilter(\"ignore\", category=UserWarning)\n",
    "\n",
    "\n",
    "# Select only the relevant columns from pop_up_info\n",
    "\n",
    "pop_up_info = fetchGSheet.pop_up_info.rename(columns={\"Mouse genome informatics (MGI) ID\": \"MGI ID\", \n",
    "                                                      \"Zebrafish genome database (ZFIN) ID\": \"ZFIN ID\"})\n",
    "\n",
    "pop_up_info_lim = pop_up_info[[\"Approved symbol\", \"Approved name\", \"MGI ID\", \"ZFIN ID\"]]\n",
    "pop_up_info_lim = pop_up_info_lim.drop_duplicates(subset=\"Approved symbol\", keep=\"first\")\n",
    "\n",
    "# Drop columns where all values are NA in gene_pair\n",
    "gene_pair = fetchGSheet.gene_pair.dropna(axis=1, how='all')\n",
    "# Fetch species IDs from the dataset\n",
    "hgnc_id = [col for col in gene_pair.columns if \"HGNC ID\" in col]\n",
    "hgnc_id = pd.concat([gene_pair[col] for col in hgnc_id]).unique()\n",
    "\n",
    "# Rename columns for better clarity\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "    \"Ligand receptor pair\": \"Human LR Pair\",\n",
    "    \"Ligand gene symbol\": \"Ligand\",\n",
    "    \"Receptor gene symbol\": \"Receptor\",\n",
    "    \"Perplexity link\": \"Perplexity\",\n",
    "    \"Source\": \"Interaction Source\"\n",
    "})\n",
    "\n",
    "# Merge gene_pair with pop_up_info_lim for Ligand(L)\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Ligand', right_on='Approved symbol')\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Ligand name\", \n",
    "                                     \"MGI ID\": \"Ligand MGI ID\",\n",
    "                                     \"ZFIN ID\": \"Ligand ZFIN ID\"},\n",
    "                            )\n",
    "\n",
    "# Add MGI annotation\n",
    "MGI_info = pd.read_csv(\"data/MGI_ID_biomart.csv\")\n",
    "gene_pair = gene_pair.merge(MGI_info, how='left', left_on='Ligand MGI ID', right_on='MGI ID')\n",
    "\n",
    "# Add ZFIN annotation\n",
    "ZFIN_info = pd.read_csv(\"data/ZFIN_ID_biomart.csv\")\n",
    "ZFIN_info['ZFIN ID'] = \"ZFIN:\" + ZFIN_info['ZFIN ID'].astype(str)\n",
    "gene_pair = gene_pair.merge(ZFIN_info, how='left', left_on='Ligand ZFIN ID', right_on='ZFIN ID')\n",
    "\n",
    "# Add ZFIN id and symbol\n",
    "ZFIN_info = pd.read_csv(\"data/ZFIN_ID_human_orthos.txt\", sep=\"\\t\", skiprows=1)\n",
    "ZFIN_info = ZFIN_info[['ZFIN ID', 'ZFIN Symbol', 'ZFIN Name', 'HGNC ID']]\n",
    "\n",
    "ZFIN_info = ZFIN_info.dropna(subset=['HGNC ID'])\n",
    "ZFIN_info = ZFIN_info.drop_duplicates(subset=['HGNC ID'])\n",
    "ZFIN_info['HGNC ID'] = ZFIN_info['HGNC ID'].apply(lambda x: f'HGNC:{int(x)}')\n",
    "gene_pair = gene_pair.merge(ZFIN_info, how='left', left_on='Ligand HGNC ID', right_on='HGNC ID')\n",
    "\n",
    "gene_pair = gene_pair.drop(columns=[\"ZFIN ID\", \"MGI ID\", \"HGNC ID\"])\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "                                     \"MGI name\": \"Mouse Ligand\", \n",
    "                                     \"ZFIN name\": \"Zebrafish Ligand\",\n",
    "                                     \"ZFIN ID\": \"Ligand ZFIN ID\",\n",
    "                                     \"ZFIN Symbol\": \"Zebrafish Ligand\",\n",
    "                                     \"ZFIN Name\": \"Zebrafish Ligand name\"}\n",
    "                            )\n",
    "\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Receptor', right_on='Approved symbol')\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Receptor name\",\n",
    "                                      \"MGI ID\": \"Receptor MGI ID\",\n",
    "                                      \"ZFIN ID\": \"Receptor ZFIN ID\"}\n",
    "                            )\n",
    "gene_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "93a95969-73e1-42c1-b98b-dfdd8e10daf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "### ZEBRAFISH ###\n",
    "\n",
    "## Limit to those with either Zebrafish Ligand or Receptor\n",
    "Zebrafish_columns = [col for col in gene_pair.columns if \"ZFIN\" in col or \"Zebrafish\" in col]\n",
    "# Filter rows where all \"Zebrafish\" columns are not \" \"\n",
    "Zebrafish_gene_pair = gene_pair000[(gene_pair000[Zebrafish_columns].map(str.strip) != \"\").all(axis=1)]\n",
    "\n",
    "\n",
    "new_columns = Zebrafish_gene_pair.columns.tolist()\n",
    "\n",
    "new_columns = [\n",
    "    col.replace(\"Zebrafish \", \"\").strip()\n",
    "    if \"Ligand\" in col or \"Receptor\" in col\n",
    "    else col\n",
    "    for col in new_columns\n",
    "]\n",
    "Zebrafish_gene_pair.columns = new_columns\n",
    "\n",
    "# Dynamically identify columns containing \"Ligand\" and \"Receptor\" in their names \n",
    "# since it is now in span format\n",
    "ligand_col = [col for col in Zebrafish_gene_pair.columns if \"Ligand&nbsp;\" in col][3]\n",
    "receptor_col = [col for col in Zebrafish_gene_pair.columns if \"Receptor&nbsp;\" in col][3]\n",
    "ligand_location = [col for col in Zebrafish_gene_pair.columns if \"Ligand location\" in col][0]\n",
    "receptor_location = [col for col in Zebrafish_gene_pair.columns if \"Receptor location\" in col][0]\n",
    "\n",
    "def format_lr_pair(row):\n",
    "    if row[ligand_location] == 'secreted':\n",
    "        return f\"{row[ligand_col]} <span style='font-size: 15px;'>○</span> <span style='font-size: 24px;'>⤚</span> {row[receptor_col]}\"\n",
    "    elif row[ligand_location] == '':\n",
    "        return f\"{row[ligand_col]} <span style='font-size: 15px;'>○</span> <span style='font-size: 24px;'>⤚</span> {row[receptor_col]}\"\n",
    "    elif row[receptor_location] == 'plasma membrane':\n",
    "        return f\"{row[ligand_col]} <span style='font-size: 24px;'>⤙</span> <span style='font-size: 24px;'>⤚</span> {row[receptor_col]}\"\n",
    "    else:\n",
    "        return f\"{row[ligand_col]} \\u2192 {row[receptor_col]}\"\n",
    "\n",
    "Zebrafish_gene_pair1 = Zebrafish_gene_pair.copy() \n",
    "Zebrafish_gene_pair1.loc[:, \"Zebrafish LR Pair\"] = Zebrafish_gene_pair1.apply(format_lr_pair, axis=1)\n",
    "Zebrafish_columns = [col for col in Zebrafish_gene_pair1.columns if \"ZFIN\" in col]\n",
    "# Reorder the DataFrame\n",
    "new_order = [\"Zebrafish LR Pair\", ligand_col, receptor_col] + Zebrafish_columns + human_columns\n",
    "Zebrafish_gene_pair1 = Zebrafish_gene_pair1[new_order]\n",
    "Zebrafish_gene_pair1 = Zebrafish_gene_pair1.reset_index(drop=True)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "228f917d-b5b9-409e-bc04-7b95afd9a725",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Human LR Pair', 'Ligand', 'Receptor', 'Interaction Source',\n",
       "       'PMID support', 'Ligand HGNC ID', 'Ligand location', 'Receptor HGNC ID',\n",
       "       'Receptor location', 'Perplexity', 'Ligand name', 'Ligand MGI ID',\n",
       "       'Ligand RGD ID', 'Mouse Ligand', 'Rat Ligand', 'Ligand ZFIN ID',\n",
       "       'Ligand ZFIN ID', 'Zebrafish Ligand', 'Zebrafish Ligand name',\n",
       "       'Receptor name', 'Receptor MGI ID', 'Receptor RGD ID', 'Mouse Receptor',\n",
       "       'Rat Receptor', 'Ligand ZFIN ID', 'Ligand ZFIN ID',\n",
       "       'Zebrafish Receptor', 'Zebrafish Receptor name', 'HGNC L R',\n",
       "       'sanity check', 'curator', 'secondary source?'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19161b56-8989-458f-80ef-d1e3eb97e029",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "quarto_llm_env",
   "language": "python",
   "name": "quarto_llm_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
