{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fcc5ddc1-519e-439c-a130-3dd8bcd181bd",
   "metadata": {},
   "source": [
    "# Python Notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b443872e-66e7-427a-a031-75a16471e8a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_row(change):\n",
    "    global gene_pair\n",
    "    # Add a new row at the top with None values\n",
    "    new_row = {col: None for col in gene_pair.columns}\n",
    "    gene_pair = pd.DataFrame([new_row] + gene_pair.to_dict(orient=\"records\"))\n",
    "    update_table()\n",
    "\n",
    "# Function to remove the last row of the dataframe\n",
    "def remove_row(change):\n",
    "    global gene_pair\n",
    "    if len(gene_pair) > 0:\n",
    "        gene_pair = gene_pair[:-1]  # Remove the last row\n",
    "        update_table()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "327783b9-5ddf-44d0-8fee-51819b75e4d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to convert MGI/RGD IDs/Other Species to gene symbols via biomart\n",
    "import os\n",
    "import sys\n",
    "from biomart import BiomartServer\n",
    "import pandas as pd\n",
    "\n",
    "# Add source directory to the path\n",
    "sys.path.append(os.path.abspath(\"src\"))\n",
    "from createDataTable import pop_up_info_lim\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e1255a12-5942-4909-b947-e57884d3fe03",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Rat LR Pair</th>\n",
       "      <th>&lt;span title=\"Click on the Rat Genome Database(RGD) IDs below for more details\"&gt;Ligand RGD ID&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Ligand RGD name to ensure all values are shown\"&gt;Ligand RGD name&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Click on the Rat Genome Database(RGD) IDs below for more details\"&gt;Receptor RGD ID&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Receptor RGD name to ensure all values are shown\"&gt;Receptor RGD name&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Ligand Receptor Pair\"&gt;LR Pair&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Source to ensure all values are shown\"&gt;Source&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Hover on symbols below to show gene names\"&gt;Ligand&amp;nbsp;&amp;nbsp;&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Hover on symbols below to show gene names\"&gt;Receptor&amp;nbsp;&amp;nbsp;&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Click the logo below to run Perplexity on the LR pair\"&gt;Perplexity&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>...</th>\n",
       "      <th>&lt;span title=\"Location is defined as subcellular location\"&gt;Receptor location&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Click on the Mouse Genome Informatics(MGI) IDs below for more details\"&gt;Ligand MGI ID&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Ligand MGI name to ensure all values are shown\"&gt;Ligand MGI name&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Click on the Mouse Genome Informatics(MGI) IDs below for more details\"&gt;Receptor MGI ID&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Receptor MGI name to ensure all values are shown\"&gt;Receptor MGI name&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of HGNC L R to ensure all values are shown\"&gt;HGNC L R&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of sanity check to ensure all values are shown\"&gt;sanity check&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of curator to ensure all values are shown\"&gt;curator&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of secondary source? to ensure all values are shown\"&gt;secondary source?&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>Rat LR Pair</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Defb5 ○ &lt;span style='font-size: 30px;'&gt;⤚&lt;/span...</td>\n",
       "      <td>&lt;a href=\"https://rgd.mcw.edu/rgdweb/report/gen...</td>\n",
       "      <td>Defb5</td>\n",
       "      <td>&lt;a href=\"https://rgd.mcw.edu/rgdweb/report/gen...</td>\n",
       "      <td>Ccr6</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Ramilowski_2015_Literature_supported</td>\n",
       "      <td>&lt;span title=\"defensin beta 4A\"&gt;DEFB4A&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"C-C motif chemokine receptor 6\"&gt;C...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Do...</td>\n",
       "      <td>...</td>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>Defb3</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>Ccr6</td>\n",
       "      <td>HGNC:2767 HGNC:1607</td>\n",
       "      <td>ok</td>\n",
       "      <td>Al</td>\n",
       "      <td></td>\n",
       "      <td>Defb5 ○ &lt;span style='font-size: 30px;'&gt;⤚&lt;/span...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Il17a ○ &lt;span style='font-size: 30px;'&gt;⤚&lt;/span...</td>\n",
       "      <td>&lt;a href=\"https://rgd.mcw.edu/rgdweb/report/gen...</td>\n",
       "      <td>Il17a</td>\n",
       "      <td>&lt;a href=\"https://rgd.mcw.edu/rgdweb/report/gen...</td>\n",
       "      <td>Il17rc</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Ramilowski_2015_Literature_supported</td>\n",
       "      <td>&lt;span title=\"interleukin 17A\"&gt;IL17A&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"interleukin 17 receptor C\"&gt;IL17RC...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Do...</td>\n",
       "      <td>...</td>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>Il17a</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>Il17rc</td>\n",
       "      <td>HGNC:5981 HGNC:18358</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Il17a ○ &lt;span style='font-size: 30px;'&gt;⤚&lt;/span...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Npnt ○ &lt;span style='font-size: 30px;'&gt;⤚&lt;/span&gt;...</td>\n",
       "      <td>&lt;a href=\"https://rgd.mcw.edu/rgdweb/report/gen...</td>\n",
       "      <td>Npnt</td>\n",
       "      <td>&lt;a href=\"https://rgd.mcw.edu/rgdweb/report/gen...</td>\n",
       "      <td>Itgb1</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Ramilowski_2015_Literature_supported</td>\n",
       "      <td>&lt;span title=\"nephronectin\"&gt;NPNT&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"integrin subunit beta 1\"&gt;ITGB1&lt;/s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Do...</td>\n",
       "      <td>...</td>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>Npnt</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>Itgb1</td>\n",
       "      <td>HGNC:27405 HGNC:6153</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>CellphoneDB</td>\n",
       "      <td>Npnt ○ &lt;span style='font-size: 30px;'&gt;⤚&lt;/span&gt;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Rph3a ○ &lt;span style='font-size: 30px;'&gt;⤚&lt;/span...</td>\n",
       "      <td>&lt;a href=\"https://rgd.mcw.edu/rgdweb/report/gen...</td>\n",
       "      <td>Rph3a</td>\n",
       "      <td>&lt;a href=\"https://rgd.mcw.edu/rgdweb/report/gen...</td>\n",
       "      <td>Nrxn1</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Ramilowski_2015_Literature_supported</td>\n",
       "      <td>&lt;span title=\"rabphilin 3A\"&gt;RPH3A&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"neurexin 1\"&gt;NRXN1&lt;/span&gt;</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Do...</td>\n",
       "      <td>...</td>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>Rph3a</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>Nrxn1</td>\n",
       "      <td>HGNC:17056 HGNC:8008</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>CellphoneDB</td>\n",
       "      <td>Rph3a ○ &lt;span style='font-size: 30px;'&gt;⤚&lt;/span...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Defb5 ○ &lt;span style='font-size: 30px;'&gt;⤚&lt;/span...</td>\n",
       "      <td>&lt;a href=\"https://rgd.mcw.edu/rgdweb/report/gen...</td>\n",
       "      <td>Defb5</td>\n",
       "      <td>&lt;a href=\"https://rgd.mcw.edu/rgdweb/report/gen...</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Ramilowski_2015_Literature_supported</td>\n",
       "      <td>&lt;span title=\"defensin beta 4A\"&gt;DEFB4A&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"toll like receptor 4\"&gt;TLR4&lt;/span&gt;</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Do...</td>\n",
       "      <td>...</td>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>Defb3</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>HGNC:2767 HGNC:11850</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Defb5 ○ &lt;span style='font-size: 30px;'&gt;⤚&lt;/span...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2144</th>\n",
       "      <td>Lrfn4 &lt;span style='font-size: 30px;'&gt;⤙&lt;/span&gt; ...</td>\n",
       "      <td>&lt;a href=\"https://rgd.mcw.edu/rgdweb/report/gen...</td>\n",
       "      <td>Lrfn4</td>\n",
       "      <td>&lt;a href=\"https://rgd.mcw.edu/rgdweb/report/gen...</td>\n",
       "      <td>Ptprf</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>ConnectomeDB2025 (this publication)</td>\n",
       "      <td>&lt;span title=\"leucine rich repeat and fibronect...</td>\n",
       "      <td>&lt;span title=\"protein tyrosine phosphatase rece...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Do...</td>\n",
       "      <td>...</td>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>Lrfn4</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>Ptprf</td>\n",
       "      <td>HGNC:28456 HGNC:9670</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Lrfn4 &lt;span style='font-size: 30px;'&gt;⤙&lt;/span&gt; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2145</th>\n",
       "      <td>Lrfn4 &lt;span style='font-size: 30px;'&gt;⤙&lt;/span&gt; ...</td>\n",
       "      <td>&lt;a href=\"https://rgd.mcw.edu/rgdweb/report/gen...</td>\n",
       "      <td>Lrfn4</td>\n",
       "      <td>&lt;a href=\"https://rgd.mcw.edu/rgdweb/report/gen...</td>\n",
       "      <td>Ptprs</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>ConnectomeDB2025 (this publication)</td>\n",
       "      <td>&lt;span title=\"leucine rich repeat and fibronect...</td>\n",
       "      <td>&lt;span title=\"protein tyrosine phosphatase rece...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Do...</td>\n",
       "      <td>...</td>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>Lrfn4</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>Ptprs</td>\n",
       "      <td>HGNC:28456 HGNC:9681</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Lrfn4 &lt;span style='font-size: 30px;'&gt;⤙&lt;/span&gt; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2146</th>\n",
       "      <td>Lrfn5 &lt;span style='font-size: 30px;'&gt;⤙&lt;/span&gt; ...</td>\n",
       "      <td>&lt;a href=\"https://rgd.mcw.edu/rgdweb/report/gen...</td>\n",
       "      <td>Lrfn5</td>\n",
       "      <td>&lt;a href=\"https://rgd.mcw.edu/rgdweb/report/gen...</td>\n",
       "      <td>Ptprd</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>ConnectomeDB2025 (this publication)</td>\n",
       "      <td>&lt;span title=\"leucine rich repeat and fibronect...</td>\n",
       "      <td>&lt;span title=\"protein tyrosine phosphatase rece...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Do...</td>\n",
       "      <td>...</td>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>Lrfn5</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>Ptprd</td>\n",
       "      <td>HGNC:20360 HGNC:9668</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Lrfn5 &lt;span style='font-size: 30px;'&gt;⤙&lt;/span&gt; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2147</th>\n",
       "      <td>Lrfn5 &lt;span style='font-size: 30px;'&gt;⤙&lt;/span&gt; ...</td>\n",
       "      <td>&lt;a href=\"https://rgd.mcw.edu/rgdweb/report/gen...</td>\n",
       "      <td>Lrfn5</td>\n",
       "      <td>&lt;a href=\"https://rgd.mcw.edu/rgdweb/report/gen...</td>\n",
       "      <td>Ptprf</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>ConnectomeDB2025 (this publication)</td>\n",
       "      <td>&lt;span title=\"leucine rich repeat and fibronect...</td>\n",
       "      <td>&lt;span title=\"protein tyrosine phosphatase rece...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Do...</td>\n",
       "      <td>...</td>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>Lrfn5</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>Ptprf</td>\n",
       "      <td>HGNC:20360 HGNC:9670</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Lrfn5 &lt;span style='font-size: 30px;'&gt;⤙&lt;/span&gt; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2148</th>\n",
       "      <td>Lrfn5 &lt;span style='font-size: 30px;'&gt;⤙&lt;/span&gt; ...</td>\n",
       "      <td>&lt;a href=\"https://rgd.mcw.edu/rgdweb/report/gen...</td>\n",
       "      <td>Lrfn5</td>\n",
       "      <td>&lt;a href=\"https://rgd.mcw.edu/rgdweb/report/gen...</td>\n",
       "      <td>Ptprs</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>ConnectomeDB2025 (this publication)</td>\n",
       "      <td>&lt;span title=\"leucine rich repeat and fibronect...</td>\n",
       "      <td>&lt;span title=\"protein tyrosine phosphatase rece...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Do...</td>\n",
       "      <td>...</td>\n",
       "      <td>plasma membrane</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>Lrfn5</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>Ptprs</td>\n",
       "      <td>HGNC:20360 HGNC:9681</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Lrfn5 &lt;span style='font-size: 30px;'&gt;⤙&lt;/span&gt; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2149 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            Rat LR Pair  \\\n",
       "0     Defb5 ○ <span style='font-size: 30px;'>⤚</span...   \n",
       "1     Il17a ○ <span style='font-size: 30px;'>⤚</span...   \n",
       "2     Npnt ○ <span style='font-size: 30px;'>⤚</span>...   \n",
       "3     Rph3a ○ <span style='font-size: 30px;'>⤚</span...   \n",
       "4     Defb5 ○ <span style='font-size: 30px;'>⤚</span...   \n",
       "...                                                 ...   \n",
       "2144  Lrfn4 <span style='font-size: 30px;'>⤙</span> ...   \n",
       "2145  Lrfn4 <span style='font-size: 30px;'>⤙</span> ...   \n",
       "2146  Lrfn5 <span style='font-size: 30px;'>⤙</span> ...   \n",
       "2147  Lrfn5 <span style='font-size: 30px;'>⤙</span> ...   \n",
       "2148  Lrfn5 <span style='font-size: 30px;'>⤙</span> ...   \n",
       "\n",
       "     <span title=\"Click on the Rat Genome Database(RGD) IDs below for more details\">Ligand RGD ID</span>  \\\n",
       "0     <a href=\"https://rgd.mcw.edu/rgdweb/report/gen...                                                    \n",
       "1     <a href=\"https://rgd.mcw.edu/rgdweb/report/gen...                                                    \n",
       "2     <a href=\"https://rgd.mcw.edu/rgdweb/report/gen...                                                    \n",
       "3     <a href=\"https://rgd.mcw.edu/rgdweb/report/gen...                                                    \n",
       "4     <a href=\"https://rgd.mcw.edu/rgdweb/report/gen...                                                    \n",
       "...                                                 ...                                                    \n",
       "2144  <a href=\"https://rgd.mcw.edu/rgdweb/report/gen...                                                    \n",
       "2145  <a href=\"https://rgd.mcw.edu/rgdweb/report/gen...                                                    \n",
       "2146  <a href=\"https://rgd.mcw.edu/rgdweb/report/gen...                                                    \n",
       "2147  <a href=\"https://rgd.mcw.edu/rgdweb/report/gen...                                                    \n",
       "2148  <a href=\"https://rgd.mcw.edu/rgdweb/report/gen...                                                    \n",
       "\n",
       "     <span title=\"Double-click header of Ligand RGD name to ensure all values are shown\">Ligand RGD name&nbsp;</span>  \\\n",
       "0                                                 Defb5                                                                 \n",
       "1                                                 Il17a                                                                 \n",
       "2                                                  Npnt                                                                 \n",
       "3                                                 Rph3a                                                                 \n",
       "4                                                 Defb5                                                                 \n",
       "...                                                 ...                                                                 \n",
       "2144                                              Lrfn4                                                                 \n",
       "2145                                              Lrfn4                                                                 \n",
       "2146                                              Lrfn5                                                                 \n",
       "2147                                              Lrfn5                                                                 \n",
       "2148                                              Lrfn5                                                                 \n",
       "\n",
       "     <span title=\"Click on the Rat Genome Database(RGD) IDs below for more details\">Receptor RGD ID</span>  \\\n",
       "0     <a href=\"https://rgd.mcw.edu/rgdweb/report/gen...                                                      \n",
       "1     <a href=\"https://rgd.mcw.edu/rgdweb/report/gen...                                                      \n",
       "2     <a href=\"https://rgd.mcw.edu/rgdweb/report/gen...                                                      \n",
       "3     <a href=\"https://rgd.mcw.edu/rgdweb/report/gen...                                                      \n",
       "4     <a href=\"https://rgd.mcw.edu/rgdweb/report/gen...                                                      \n",
       "...                                                 ...                                                      \n",
       "2144  <a href=\"https://rgd.mcw.edu/rgdweb/report/gen...                                                      \n",
       "2145  <a href=\"https://rgd.mcw.edu/rgdweb/report/gen...                                                      \n",
       "2146  <a href=\"https://rgd.mcw.edu/rgdweb/report/gen...                                                      \n",
       "2147  <a href=\"https://rgd.mcw.edu/rgdweb/report/gen...                                                      \n",
       "2148  <a href=\"https://rgd.mcw.edu/rgdweb/report/gen...                                                      \n",
       "\n",
       "     <span title=\"Double-click header of Receptor RGD name to ensure all values are shown\">Receptor RGD name&nbsp;</span>  \\\n",
       "0                                                  Ccr6                                                                     \n",
       "1                                                Il17rc                                                                     \n",
       "2                                                 Itgb1                                                                     \n",
       "3                                                 Nrxn1                                                                     \n",
       "4                                                  Tlr4                                                                     \n",
       "...                                                 ...                                                                     \n",
       "2144                                              Ptprf                                                                     \n",
       "2145                                              Ptprs                                                                     \n",
       "2146                                              Ptprd                                                                     \n",
       "2147                                              Ptprf                                                                     \n",
       "2148                                              Ptprs                                                                     \n",
       "\n",
       "      <span title=\"Ligand Receptor Pair\">LR Pair</span>  \\\n",
       "0     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "1     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "2     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "3     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "...                                                 ...   \n",
       "2144  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "2145  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "2146  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "2147  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "2148  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "\n",
       "     <span title=\"Double-click header of Source to ensure all values are shown\">Source&nbsp;</span>  \\\n",
       "0                  Ramilowski_2015_Literature_supported                                               \n",
       "1                  Ramilowski_2015_Literature_supported                                               \n",
       "2                  Ramilowski_2015_Literature_supported                                               \n",
       "3                  Ramilowski_2015_Literature_supported                                               \n",
       "4                  Ramilowski_2015_Literature_supported                                               \n",
       "...                                                 ...                                               \n",
       "2144                ConnectomeDB2025 (this publication)                                               \n",
       "2145                ConnectomeDB2025 (this publication)                                               \n",
       "2146                ConnectomeDB2025 (this publication)                                               \n",
       "2147                ConnectomeDB2025 (this publication)                                               \n",
       "2148                ConnectomeDB2025 (this publication)                                               \n",
       "\n",
       "     <span title=\"Hover on symbols below to show gene names\">Ligand&nbsp;&nbsp;&nbsp;</span>  \\\n",
       "0          <span title=\"defensin beta 4A\">DEFB4A</span>                                        \n",
       "1            <span title=\"interleukin 17A\">IL17A</span>                                        \n",
       "2                <span title=\"nephronectin\">NPNT</span>                                        \n",
       "3               <span title=\"rabphilin 3A\">RPH3A</span>                                        \n",
       "4          <span title=\"defensin beta 4A\">DEFB4A</span>                                        \n",
       "...                                                 ...                                        \n",
       "2144  <span title=\"leucine rich repeat and fibronect...                                        \n",
       "2145  <span title=\"leucine rich repeat and fibronect...                                        \n",
       "2146  <span title=\"leucine rich repeat and fibronect...                                        \n",
       "2147  <span title=\"leucine rich repeat and fibronect...                                        \n",
       "2148  <span title=\"leucine rich repeat and fibronect...                                        \n",
       "\n",
       "     <span title=\"Hover on symbols below to show gene names\">Receptor&nbsp;&nbsp;&nbsp;</span>  \\\n",
       "0     <span title=\"C-C motif chemokine receptor 6\">C...                                          \n",
       "1     <span title=\"interleukin 17 receptor C\">IL17RC...                                          \n",
       "2     <span title=\"integrin subunit beta 1\">ITGB1</s...                                          \n",
       "3                 <span title=\"neurexin 1\">NRXN1</span>                                          \n",
       "4        <span title=\"toll like receptor 4\">TLR4</span>                                          \n",
       "...                                                 ...                                          \n",
       "2144  <span title=\"protein tyrosine phosphatase rece...                                          \n",
       "2145  <span title=\"protein tyrosine phosphatase rece...                                          \n",
       "2146  <span title=\"protein tyrosine phosphatase rece...                                          \n",
       "2147  <span title=\"protein tyrosine phosphatase rece...                                          \n",
       "2148  <span title=\"protein tyrosine phosphatase rece...                                          \n",
       "\n",
       "     <span title=\"Click the logo below to run Perplexity on the LR pair\">Perplexity&nbsp;</span>  \\\n",
       "0     <a href=\"https://www.perplexity.ai/search?q=Do...                                            \n",
       "1     <a href=\"https://www.perplexity.ai/search?q=Do...                                            \n",
       "2     <a href=\"https://www.perplexity.ai/search?q=Do...                                            \n",
       "3     <a href=\"https://www.perplexity.ai/search?q=Do...                                            \n",
       "4     <a href=\"https://www.perplexity.ai/search?q=Do...                                            \n",
       "...                                                 ...                                            \n",
       "2144  <a href=\"https://www.perplexity.ai/search?q=Do...                                            \n",
       "2145  <a href=\"https://www.perplexity.ai/search?q=Do...                                            \n",
       "2146  <a href=\"https://www.perplexity.ai/search?q=Do...                                            \n",
       "2147  <a href=\"https://www.perplexity.ai/search?q=Do...                                            \n",
       "2148  <a href=\"https://www.perplexity.ai/search?q=Do...                                            \n",
       "\n",
       "      ...  \\\n",
       "0     ...   \n",
       "1     ...   \n",
       "2     ...   \n",
       "3     ...   \n",
       "4     ...   \n",
       "...   ...   \n",
       "2144  ...   \n",
       "2145  ...   \n",
       "2146  ...   \n",
       "2147  ...   \n",
       "2148  ...   \n",
       "\n",
       "     <span title=\"Location is defined as subcellular location\">Receptor location</span>  \\\n",
       "0                                       plasma membrane                                   \n",
       "1                                       plasma membrane                                   \n",
       "2                                       plasma membrane                                   \n",
       "3                                       plasma membrane                                   \n",
       "4                                       plasma membrane                                   \n",
       "...                                                 ...                                   \n",
       "2144                                    plasma membrane                                   \n",
       "2145                                    plasma membrane                                   \n",
       "2146                                    plasma membrane                                   \n",
       "2147                                    plasma membrane                                   \n",
       "2148                                    plasma membrane                                   \n",
       "\n",
       "     <span title=\"Click on the Mouse Genome Informatics(MGI) IDs below for more details\">Ligand MGI ID</span>  \\\n",
       "0     <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "1     <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "2     <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "3     <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "4     <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "...                                                 ...                                                         \n",
       "2144  <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "2145  <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "2146  <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "2147  <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "2148  <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "\n",
       "     <span title=\"Double-click header of Ligand MGI name to ensure all values are shown\">Ligand MGI name&nbsp;</span>  \\\n",
       "0                                                 Defb3                                                                 \n",
       "1                                                 Il17a                                                                 \n",
       "2                                                  Npnt                                                                 \n",
       "3                                                 Rph3a                                                                 \n",
       "4                                                 Defb3                                                                 \n",
       "...                                                 ...                                                                 \n",
       "2144                                              Lrfn4                                                                 \n",
       "2145                                              Lrfn4                                                                 \n",
       "2146                                              Lrfn5                                                                 \n",
       "2147                                              Lrfn5                                                                 \n",
       "2148                                              Lrfn5                                                                 \n",
       "\n",
       "     <span title=\"Click on the Mouse Genome Informatics(MGI) IDs below for more details\">Receptor MGI ID</span>  \\\n",
       "0     <a href=\"https://www.informatics.jax.org/marke...                                                           \n",
       "1     <a href=\"https://www.informatics.jax.org/marke...                                                           \n",
       "2     <a href=\"https://www.informatics.jax.org/marke...                                                           \n",
       "3     <a href=\"https://www.informatics.jax.org/marke...                                                           \n",
       "4     <a href=\"https://www.informatics.jax.org/marke...                                                           \n",
       "...                                                 ...                                                           \n",
       "2144  <a href=\"https://www.informatics.jax.org/marke...                                                           \n",
       "2145  <a href=\"https://www.informatics.jax.org/marke...                                                           \n",
       "2146  <a href=\"https://www.informatics.jax.org/marke...                                                           \n",
       "2147  <a href=\"https://www.informatics.jax.org/marke...                                                           \n",
       "2148  <a href=\"https://www.informatics.jax.org/marke...                                                           \n",
       "\n",
       "     <span title=\"Double-click header of Receptor MGI name to ensure all values are shown\">Receptor MGI name&nbsp;</span>  \\\n",
       "0                                                  Ccr6                                                                     \n",
       "1                                                Il17rc                                                                     \n",
       "2                                                 Itgb1                                                                     \n",
       "3                                                 Nrxn1                                                                     \n",
       "4                                                  Tlr4                                                                     \n",
       "...                                                 ...                                                                     \n",
       "2144                                              Ptprf                                                                     \n",
       "2145                                              Ptprs                                                                     \n",
       "2146                                              Ptprd                                                                     \n",
       "2147                                              Ptprf                                                                     \n",
       "2148                                              Ptprs                                                                     \n",
       "\n",
       "     <span title=\"Double-click header of HGNC L R to ensure all values are shown\">HGNC L R&nbsp;</span>  \\\n",
       "0                                   HGNC:2767 HGNC:1607                                                   \n",
       "1                                  HGNC:5981 HGNC:18358                                                   \n",
       "2                                  HGNC:27405 HGNC:6153                                                   \n",
       "3                                  HGNC:17056 HGNC:8008                                                   \n",
       "4                                  HGNC:2767 HGNC:11850                                                   \n",
       "...                                                 ...                                                   \n",
       "2144                               HGNC:28456 HGNC:9670                                                   \n",
       "2145                               HGNC:28456 HGNC:9681                                                   \n",
       "2146                               HGNC:20360 HGNC:9668                                                   \n",
       "2147                               HGNC:20360 HGNC:9670                                                   \n",
       "2148                               HGNC:20360 HGNC:9681                                                   \n",
       "\n",
       "     <span title=\"Double-click header of sanity check to ensure all values are shown\">sanity check&nbsp;</span>  \\\n",
       "0                                                    ok                                                           \n",
       "1                                                                                                                 \n",
       "2                                                                                                                 \n",
       "3                                                                                                                 \n",
       "4                                                                                                                 \n",
       "...                                                 ...                                                           \n",
       "2144                                                                                                              \n",
       "2145                                                                                                              \n",
       "2146                                                                                                              \n",
       "2147                                                                                                              \n",
       "2148                                                                                                              \n",
       "\n",
       "     <span title=\"Double-click header of curator to ensure all values are shown\">curator&nbsp;</span>  \\\n",
       "0                                                    Al                                                 \n",
       "1                                                                                                       \n",
       "2                                                                                                       \n",
       "3                                                                                                       \n",
       "4                                                                                                       \n",
       "...                                                 ...                                                 \n",
       "2144                                                                                                    \n",
       "2145                                                                                                    \n",
       "2146                                                                                                    \n",
       "2147                                                                                                    \n",
       "2148                                                                                                    \n",
       "\n",
       "     <span title=\"Double-click header of secondary source? to ensure all values are shown\">secondary source?&nbsp;</span>  \\\n",
       "0                                                                                                                           \n",
       "1                                                                                                                           \n",
       "2                                           CellphoneDB                                                                     \n",
       "3                                           CellphoneDB                                                                     \n",
       "4                                                                                                                           \n",
       "...                                                 ...                                                                     \n",
       "2144                                                                                                                        \n",
       "2145                                                                                                                        \n",
       "2146                                                                                                                        \n",
       "2147                                                                                                                        \n",
       "2148                                                                                                                        \n",
       "\n",
       "                                            Rat LR Pair  \n",
       "0     Defb5 ○ <span style='font-size: 30px;'>⤚</span...  \n",
       "1     Il17a ○ <span style='font-size: 30px;'>⤚</span...  \n",
       "2     Npnt ○ <span style='font-size: 30px;'>⤚</span>...  \n",
       "3     Rph3a ○ <span style='font-size: 30px;'>⤚</span...  \n",
       "4     Defb5 ○ <span style='font-size: 30px;'>⤚</span...  \n",
       "...                                                 ...  \n",
       "2144  Lrfn4 <span style='font-size: 30px;'>⤙</span> ...  \n",
       "2145  Lrfn4 <span style='font-size: 30px;'>⤙</span> ...  \n",
       "2146  Lrfn5 <span style='font-size: 30px;'>⤙</span> ...  \n",
       "2147  Lrfn5 <span style='font-size: 30px;'>⤙</span> ...  \n",
       "2148  Lrfn5 <span style='font-size: 30px;'>⤙</span> ...  \n",
       "\n",
       "[2149 rows x 24 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Species-specific parameters\n",
    "species_id_prefix = \"ZFIN\"  # \"RGD\" for rat, \"MGI\" for mouse\n",
    "dataset_name = \"drerio_gene_ensembl\" # \"drerio_gene_ensembl\" for zebra fish, \"rnorvegicus_gene_ensembl\" for rat, \"mmusculus_gene_ensembl\" for mouse\n",
    "gene_symbol_field = \"zfin_symbol\"  # \"external_gene_name\" for mouse\n",
    "\n",
    "# Access Biomart server and dataset\n",
    "biomart_server_url = \"http://www.ensembl.org/biomart\"\n",
    "server = BiomartServer(biomart_server_url)\n",
    "dataset = server.datasets[dataset_name]\n",
    "\n",
    "# Fetch species IDs from the dataset\n",
    "species_ids = pop_up_info_lim[species_id_prefix + ' ID'].unique()\n",
    "species_ids = [id for id in species_ids if id.startswith(species_id_prefix + \":\")]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "33240925-c4f8-4f9c-acf8-c0a7464ba23c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['<span title=\"Click on the Mouse Genome Informatics(MGI) IDs below for more details\">Ligand MGI ID</span>',\n",
       " '<span title=\"Double-click header of Mouse Ligand to ensure all values are shown\">Mouse Ligand&nbsp;</span>',\n",
       " '<span title=\"Click on the Mouse Genome Informatics(MGI) IDs below for more details\">Receptor MGI ID</span>',\n",
       " '<span title=\"Double-click header of Mouse Receptor to ensure all values are shown\">Mouse Receptor&nbsp;</span>']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Function to prepare datatables (cleaning and hyperlinking, adding tool tips, etc) input for the database qmds\n",
    "import sys, os\n",
    "sys.path.append(os.path.abspath(\"src\"))  \n",
    "from itables import init_notebook_mode\n",
    "import pandas as pd\n",
    "from itables import show\n",
    "from itables import options\n",
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "import fetchGSheet \n",
    "import warnings\n",
    "\n",
    "# Suppress SettingWithCopyWarning\n",
    "warnings.simplefilter(\"ignore\", category=UserWarning)\n",
    "\n",
    "\n",
    "# Select only the relevant columns from pop_up_info\n",
    "\n",
    "pop_up_info = fetchGSheet.pop_up_info.rename(columns={\"Mouse genome informatics (MGI) ID\": \"MGI ID\",\"Rat genome database (RGD) ID\": \"RGD ID\"})\n",
    "\n",
    "pop_up_info_lim = pop_up_info[[\"Approved symbol\", \"Approved name\", \"MGI ID\", \"RGD ID\"]]\n",
    "pop_up_info_lim = pop_up_info_lim.drop_duplicates(subset=\"Approved symbol\", keep=\"first\")\n",
    "\n",
    "# Drop columns where all values are NA in gene_pair\n",
    "gene_pair = fetchGSheet.gene_pair.dropna(axis=1, how='all')\n",
    "\n",
    "# Rename columns for better clarity\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "    \"Ligand receptor pair\": \"LR Pair\",\n",
    "    \"Ligand gene symbol\": \"Ligand\",\n",
    "    \"Receptor gene symbol\": \"Receptor\",\n",
    "    \"Perplexity link\": \"Perplexity\"\n",
    "})\n",
    "\n",
    "# Merge gene_pair with pop_up_info_lim for Ligand(L)\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Ligand', right_on='Approved symbol')\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Ligand name\", \n",
    "                                     \"MGI ID\": \"Ligand MGI ID\",\n",
    "                                     \"RGD ID\": \"Ligand RGD ID\"},\n",
    "                            )\n",
    "\n",
    "# Add MGI name\n",
    "MGI_info = pd.read_csv(\"data/MGI_ID_biomart.csv\")\n",
    "gene_pair = gene_pair.merge(MGI_info, how='left', left_on='Ligand MGI ID', right_on='MGI ID')\n",
    "\n",
    "# Add RGD name\n",
    "RGD_info = pd.read_csv(\"data/RGD_ID_biomart.csv\")\n",
    "RGD_info['RGD ID'] = \"RGD:\" + RGD_info['RGD ID'].astype(str)\n",
    "gene_pair = gene_pair.merge(RGD_info, how='left', left_on='Ligand RGD ID', right_on='RGD ID')\n",
    "\n",
    "gene_pair = gene_pair.drop(columns=[\"RGD ID\", \"MGI ID\"])\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "                                     \"MGI name\": \"Mouse Ligand\", \n",
    "                                     \"RGD name\": \"Rat Ligand\"}\n",
    "                            )\n",
    "\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Receptor', right_on='Approved symbol')\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Receptor name\",\n",
    "                                      \"MGI ID\": \"Receptor MGI ID\",\n",
    "                                      \"RGD ID\": \"Receptor RGD ID\"}\n",
    "                            )\n",
    "\n",
    "# Add MGI name\n",
    "gene_pair = gene_pair.merge(MGI_info, how='left', left_on='Receptor MGI ID', right_on='MGI ID')\n",
    "gene_pair = gene_pair.merge(RGD_info, how='left', left_on='Receptor RGD ID', right_on='RGD ID')\n",
    "gene_pair = gene_pair.drop(columns=[\"RGD ID\", \"MGI ID\"])\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "                                     \"MGI name\": \"Mouse Receptor\", \n",
    "                                     \"RGD name\": \"Rat Receptor\"}\n",
    "                            )\n",
    "gene_pair = gene_pair.drop(columns=[\"Approved symbol_x\", \"Approved symbol_y\"])\n",
    "\n",
    "# Drop columns where all values are NA in gene_pair\n",
    "gene_pair = gene_pair.dropna(axis=1, how='all')\n",
    "\n",
    "gene_pair = gene_pair.fillna(\" \")\n",
    "gene_pair = gene_pair[gene_pair['LR Pair'] != ' ']\n",
    "\n",
    "if \"PMID link\" in gene_pair.columns:\n",
    "    gene_pair = gene_pair.drop(columns=[\"PMID link\"])\n",
    "\n",
    "# Add\n",
    "first_columns=['LR Pair','Source', 'Ligand','Receptor', 'Perplexity']\n",
    "end_columns=['HGNC L R','sanity check', 'curator','secondary source?']\n",
    "gene_pair = gene_pair[first_columns + [col for col in gene_pair.columns if col not in first_columns + end_columns] + end_columns]\n",
    "\n",
    "# number of unique vars\n",
    "\n",
    "lrPairsCount = len(gene_pair[\"LR Pair\"].unique())\n",
    "\n",
    "ligandCount = len(gene_pair[\"Ligand\"].unique())\n",
    "\n",
    "receptorCount = len(gene_pair[\"Receptor\"].unique())\n",
    "\n",
    "# Mouse Orthologue\n",
    "MouseLigandCount = len(gene_pair[\"Ligand MGI ID\"].unique())\n",
    "\n",
    "MouseReceptorCount = len(gene_pair[\"Receptor MGI ID\"].unique())\n",
    "\n",
    "# Rat Orthologue\n",
    "RatLigandCount = len(gene_pair[\"Ligand RGD ID\"].unique())\n",
    "\n",
    "RatReceptorCount = len(gene_pair[\"Receptor RGD ID\"].unique())\n",
    "\n",
    "gene_pair[\"PMID support\"] = [value.replace(\" \", \"\") for value in gene_pair[\"PMID support\"]]\n",
    "\n",
    "source = np.array(gene_pair[\"PMID support\"].unique())\n",
    "source = source.astype(str)\n",
    "source = \",\".join(sorted(set(filter(lambda x: x.lower() != 'nan', source))))\n",
    "\n",
    "# Split the string into individual elements, filter out empty strings, and get unique values\n",
    "source = sorted(\n",
    "    set(filter(lambda x: x.strip() and x.strip().lower() != 'nan', source.split(',')))\n",
    ")\n",
    "source = [value.replace(\" \", \"\") for value in source]\n",
    "sourceCount = len(source)\n",
    "\n",
    "# for creating PMIDs\n",
    "gene_pair00 = gene_pair[['LR Pair', 'PMID support']]\n",
    "\n",
    "# create URLs for the HGNC IDs\n",
    "\n",
    "# ligand\n",
    "gene_pair[\"Ligand HGNC ID\"] = [\n",
    "    '<a href=\"https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/{}\" target=\"_blank\">{}</a>'.format(ligand, ligand)\n",
    "    for ligand in gene_pair[\"Ligand HGNC ID\"]\n",
    "]\n",
    "\n",
    "# receptor\n",
    "gene_pair[\"Receptor HGNC ID\"] = [\n",
    "    '<a href=\"https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/{}\" target=\"_blank\">{}</a>'.format(receptor, receptor)\n",
    "    for receptor in gene_pair[\"Receptor HGNC ID\"]\n",
    "]\n",
    "\n",
    "# Perplexity\n",
    "gene_pair[\"Perplexity\"] = [\n",
    "    '<a href=\"{}\" target=\"_blank\"> <img src=\"https://img.icons8.com/?size=35&id=0NbBuNOxUwps&format=png&color=000000\" alt=\"Perplexity AI\" /></a>'.format(url)\n",
    "    for url in gene_pair[\"Perplexity\"]\n",
    "]\n",
    "\n",
    "# Function to generate hyperlinks for the \"PMID support\" column\n",
    "# Function to generate hyperlinks for the \"PMID support\" column\n",
    "def generate_links_with_doi(df, gene_column, pmid_column):\n",
    "    def create_link(gene, sources):\n",
    "        # Replace spaces with \"——\" in the gene name for the link\n",
    "        gene_name = gene.replace(\" \", \"——\")\n",
    "        \n",
    "        if len(sources) == 1:\n",
    "            source = sources[0]\n",
    "            if source.startswith(\"https://www.biorxiv.org/content/\"):\n",
    "                # If the value starts with \"https://doi.org/\", use it as the hyperlink\n",
    "                return f'<a href=\"{source}\" target=\"_blank\">BioRxiv preprint</a>'\n",
    "            else:\n",
    "                # If it's a single PMID, hyperlink the PMID text\n",
    "                return f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/pubmed/{gene_name}_pmid_details.html\">{source}</a>'\n",
    "        else:\n",
    "            # If multiple PMIDs, show the count and hyperlink to the page\n",
    "            return f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/pubmed/{gene_name}_pmid_details.html\" target=\"_blank\">{len(sources)} PMIDs</a>'\n",
    "\n",
    "    # Process each row to generate the \"PMID support\" column\n",
    "    df[\"PMID support\"] = [\n",
    "        create_link(\n",
    "            gene=row[gene_column], \n",
    "            sources=[s.strip() for s in row[pmid_column].split(',') if s.strip()]\n",
    "        )\n",
    "        for _, row in df.iterrows()\n",
    "    ]\n",
    "    return df\n",
    "\n",
    "\n",
    "# Generate the links for the \"PMID support\" column\n",
    "gene_pair = generate_links_with_doi(gene_pair, gene_column=\"LR Pair\", pmid_column=\"PMID support\")\n",
    "\n",
    "gene_pair[\"Ligand MGI ID\"] = [\n",
    "        f'<a href=\"https://www.informatics.jax.org/marker/{mouseOrth}\" target=\"_blank\">{mouseOrth}</a>' \n",
    "        if pd.notna(mouseOrth) and mouseOrth.strip() else \"\" \n",
    "        for mouseOrth in gene_pair[\"Ligand MGI ID\"]\n",
    "    ]\n",
    "\n",
    "gene_pair[\"Receptor MGI ID\"] = [\n",
    "        f'<a href=\"https://www.informatics.jax.org/marker/{mouseOrth}\" target=\"_blank\">{mouseOrth}</a>' \n",
    "        if pd.notna(mouseOrth) and mouseOrth.strip() else \"\" \n",
    "        for mouseOrth in gene_pair[\"Receptor MGI ID\"]\n",
    "    ]\n",
    "\n",
    "gene_pair[\"Ligand RGD ID\"] = [\n",
    "        f'<a href=\"https://rgd.mcw.edu/rgdweb/report/gene/main.html?id={ratOrth.replace(\"RGD:\", \"\")}\" target=\"_blank\">{ratOrth}</a>' \n",
    "        if pd.notna(ratOrth) and ratOrth.strip() else \"\" \n",
    "        for ratOrth in gene_pair[\"Ligand RGD ID\"]\n",
    "    ]\n",
    "\n",
    "gene_pair[\"Receptor RGD ID\"] = [\n",
    "        f'<a href=\"https://rgd.mcw.edu/rgdweb/report/gene/main.html?id={ratOrth.replace(\"RGD:\", \"\")}\" target=\"_blank\">{ratOrth}</a>' \n",
    "        if pd.notna(ratOrth) and ratOrth.strip() else \"\" \n",
    "        for ratOrth in gene_pair[\"Receptor RGD ID\"]\n",
    "    ]\n",
    "\n",
    "mouse_columns = [col for col in gene_pair.columns if \"MGI\" in col]\n",
    "rat_columns = [col for col in gene_pair.columns if \"RGD\" in col]\n",
    "\n",
    "gene_pair0 = gene_pair[['LR Pair', 'Ligand', 'Receptor', 'Perplexity', 'PMID support',\n",
    "       'Ligand HGNC ID', 'Ligand location', 'Receptor HGNC ID',\n",
    "       'Receptor location', 'Ligand name', 'Receptor name'] + mouse_columns + rat_columns]\n",
    "\n",
    "\n",
    "# gene symbol\n",
    "gene_pair[\"Ligand\"] = [\n",
    "    f'<span title=\"{ligand_name}\">{ligand_symbol}</span>'\n",
    "    for ligand_name, ligand_symbol in zip(gene_pair[\"Ligand name\"], gene_pair[\"Ligand\"])\n",
    "]\n",
    "\n",
    "# gene symbol\n",
    "gene_pair[\"Receptor\"] = [\n",
    "    f'<span title=\"{receptor_name}\">{receptor_symbol}</span>'\n",
    "    for receptor_name, receptor_symbol in zip(gene_pair[\"Receptor name\"], gene_pair[\"Receptor\"])\n",
    "]\n",
    "\n",
    "def replace_spaces(row):\n",
    "    if row['Ligand location'] == 'secreted':\n",
    "        return row['LR Pair'].replace(\" \", \" ○ <span style='font-size: 30px;'>⤚</span> \")\n",
    "    elif row['Ligand location'] == 'plasma membrane':\n",
    "        return row['LR Pair'].replace(\" \", \" <span style='font-size: 30px;'>⤙</span> <span style='font-size: 30px;'>⤚</span> \")\n",
    "    else:\n",
    "        return row['LR Pair'].replace(\" \", \" \\u2192 \")\n",
    "\n",
    "# Apply the function to the 'LR Pair' column\n",
    "gene_pair['LR Pair'] = gene_pair.apply(replace_spaces, axis=1)\n",
    "\n",
    "gene_pair = gene_pair.drop(columns=[\"Ligand name\", \"Receptor name\"])\n",
    "\n",
    "\n",
    "# Create the links to the HTML cards\n",
    "gene_pair[\"LR Pair\"] = [\n",
    "    f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/cards/{lrPairOrig}.html\">{lrPair}</a>'\n",
    "    for lrPairOrig, lrPair in zip(gene_pair0[\"LR Pair\"], gene_pair[\"LR Pair\"])\n",
    "]\n",
    "\n",
    "\n",
    "# Add tooltips to the column headers\n",
    "gene_pair.columns = [\n",
    "    f'<span title=\"Ligand Receptor Pair\">{col}</span>' if col == \"LR Pair\" else\n",
    "    f'<span title=\"Click the logo below to run Perplexity on the LR pair\">{col}&nbsp;</span>' if col == \"Perplexity\" else\n",
    "    f'<span title=\"Hover on symbols below to show gene names\">{col}&nbsp;&nbsp;&nbsp;</span>' if col in [\"Ligand\", \"Receptor\"] else\n",
    "    f'<span title=\"Click on HGNC IDs below for more details\">{col}&nbsp;&nbsp;</span>' if col in [\"Ligand HGNC ID\", \"Receptor HGNC ID\"] else\n",
    "    f'<span title=\"Click on the Pubmed IDs (PMID) below for more details\">{col}</span>' if col == \"PMID support\" else\n",
    "    f'<span title=\"Click on the Rat Genome Database(RGD) IDs below for more details\">{col}</span>' if col in [\"Ligand RGD ID\", \"Receptor RGD ID\"] else\n",
    "    f'<span title=\"Click on the Mouse Genome Informatics(MGI) IDs below for more details\">{col}</span>' if col in [\"Ligand MGI ID\", \"Receptor MGI ID\"] else\n",
    "    f'<span title=\"Location is defined as subcellular location\">{col}</span>' if col in [\"Ligand location\", \"Receptor location\"] else\n",
    "    f'<span title=\"Double-click header of {col} to ensure all values are shown\">{col}&nbsp;</span>'\n",
    "    for col in gene_pair.columns\n",
    "]\n",
    "\n",
    "gene_pair = gene_pair.reset_index(drop=True)  # Remove the index\n",
    "\n",
    "## Limit to those with either Mouse Ligand or Receptor\n",
    "# Find columns with \"Mouse\" in the name\n",
    "mouse_columns = [col for col in gene_pair.columns if \"MGI\" or \"Mouse\" in col]\n",
    "mouse_columns = [col for col in gene_pair.columns if \"MGI\" in col or \"Mouse\" in col]\n",
    "mouse_columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "760ed86c-6dd4-4a67-9040-dfab8352ac1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['<span title=\"Ligand Receptor Pair\">LR Pair</span>',\n",
       " '<span title=\"Double-click header of Source to ensure all values are shown\">Source&nbsp;</span>',\n",
       " '<span title=\"Hover on symbols below to show gene names\">Ligand&nbsp;&nbsp;&nbsp;</span>',\n",
       " '<span title=\"Hover on symbols below to show gene names\">Receptor&nbsp;&nbsp;&nbsp;</span>',\n",
       " '<span title=\"Click the logo below to run Perplexity on the LR pair\">Perplexity&nbsp;</span>',\n",
       " '<span title=\"Click on the Pubmed IDs (PMID) below for more details\">PMID support</span>',\n",
       " '<span title=\"Click on HGNC IDs below for more details\">Ligand HGNC ID&nbsp;&nbsp;</span>',\n",
       " '<span title=\"Location is defined as subcellular location\">Ligand location</span>',\n",
       " '<span title=\"Click on HGNC IDs below for more details\">Receptor HGNC ID&nbsp;&nbsp;</span>',\n",
       " '<span title=\"Location is defined as subcellular location\">Receptor location</span>']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter rows where all \"Mouse\" columns are not \" \"\n",
    "mouse_gene_pair = gene_pair[(gene_pair[mouse_columns].map(str.strip) != \"\").all(axis=1)]\n",
    "# Dynamically identify columns containing \"Ligand\" and \"Receptor\" in their names \n",
    "# since it is now in span format\n",
    "        \n",
    "ligand_col = [col for col in mouse_gene_pair.columns if \"Mouse Ligand\" in col][0]\n",
    "receptor_col = [col for col in mouse_gene_pair.columns if \"Mouse Receptor\" in col][0]\n",
    "ligand_location = [col for col in mouse_gene_pair.columns if \"Ligand location\" in col][0]\n",
    "receptor_location = [col for col in mouse_gene_pair.columns if \"Receptor location\" in col][0]\n",
    "\n",
    "# Combine columns into \"Mouse LR Pair\" with appropriate replacements\n",
    "def format_lr_pair(row):\n",
    "    if row[ligand_location] == 'secreted':\n",
    "        return f\"{row[ligand_col]} ○ <span style='font-size: 30px;'>⤚</span> {row[receptor_col]}\"\n",
    "    elif row[receptor_location] == 'plasma membrane':\n",
    "        return f\"{row[ligand_col]} <span style='font-size: 30px;'>⤙</span> <span style='font-size: 30px;'>⤚</span> {row[receptor_col]}\"\n",
    "    else:\n",
    "        return f\"{row[ligand_col]} \\u2192 {row[receptor_col]}\"\n",
    "\n",
    "\n",
    "# Apply the function row-wise and assign to the new column using .loc\n",
    "mouse_gene_pair1 = mouse_gene_pair.copy() \n",
    "mouse_gene_pair1.loc[:, \"Mouse LR Pair\"] = mouse_gene_pair1.apply(format_lr_pair, axis=1)\n",
    "# Now, construct new_order\n",
    "# Ensure you are working with a subset of columns that are not in mouse_columns\n",
    "human_columns = [col for col in mouse_gene_pair1.columns if col not in mouse_columns][:10]\n",
    "\n",
    "new_order = [\"Mouse LR Pair\"] + mouse_columns + first_10_non_mouse + [col for col in mouse_gene_pair1.columns if col not in mouse_columns and col not in first_10_non_mouse]\n",
    "first_10_non_mouse\n",
    "\n",
    "\n",
    "keywords_to_modify = [\"LR Pair\", \"Ligand\", \"Receptor\"]\n",
    "exclude_keywords = [\"HGNC ID\"]  # Columns containing this will not be modified\n",
    "\n",
    "# Modify the column names conditionally by adding \"Human \" prefix\n",
    "modified_columns = [\n",
    "    f'Human {col}' if any(keyword in col for keyword in keywords_to_modify) and not any(exclude in col for exclude in exclude_keywords) else col\n",
    "    for col in column_names\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a3670e02-a849-4eb1-b707-eb34ed4aafa9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['<span title=\"Ligand Receptor Pair\">LR Pair</span>',\n",
       " '<span title=\"Double-click header of Source to ensure all values are shown\">Source&nbsp;</span>',\n",
       " '<span title=\"Hover on symbols below to show gene names\">Ligand&nbsp;&nbsp;&nbsp;</span>',\n",
       " '<span title=\"Hover on symbols below to show gene names\">Receptor&nbsp;&nbsp;&nbsp;</span>',\n",
       " '<span title=\"Click the logo below to run Perplexity on the LR pair\">Perplexity&nbsp;</span>',\n",
       " '<span title=\"Click on the Pubmed IDs (PMID) below for more details\">PMID support</span>',\n",
       " '<span title=\"Click on HGNC IDs below for more details\">Ligand HGNC ID&nbsp;&nbsp;</span>',\n",
       " '<span title=\"Location is defined as subcellular location\">Ligand location</span>',\n",
       " '<span title=\"Click on HGNC IDs below for more details\">Receptor HGNC ID&nbsp;&nbsp;</span>',\n",
       " '<span title=\"Location is defined as subcellular location\">Receptor location</span>']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_columns = [col for col in gene_pair.columns][:10]\n",
    "human_columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "dac5333e-1b03-4347-a44a-a048eed03e91",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['<span title=\"<span title=\"Ligand Receptor Pair\"\">Human LR Pair</span</span>',\n",
       " '<span title=\"Double-click header of Source to ensure all values are shown\">Source&nbsp;</span>',\n",
       " '<span title=\"<span title=\"Hover on symbols below to show gene names\"\">Human Ligand&nbsp;&nbsp;&nbsp;</span</span>',\n",
       " '<span title=\"<span title=\"Hover on symbols below to show gene names\"\">Human Receptor&nbsp;&nbsp;&nbsp;</span</span>',\n",
       " '<span title=\"Click the logo below to run Perplexity on the LR pair\">Perplexity&nbsp;</span>',\n",
       " '<span title=\"Click on the Pubmed IDs (PMID) below for more details\">PMID support</span>',\n",
       " '<span title=\"Click on HGNC IDs below for more details\">Ligand HGNC ID&nbsp;&nbsp;</span>',\n",
       " '<span title=\"<span title=\"Location is defined as subcellular location\"\">Human Ligand location</span</span>',\n",
       " '<span title=\"Click on HGNC IDs below for more details\">Receptor HGNC ID&nbsp;&nbsp;</span>',\n",
       " '<span title=\"<span title=\"Location is defined as subcellular location\"\">Human Receptor location</span</span>']"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Function to prepare datatables (cleaning and hyperlinking, adding tool tips, etc) input for the database qmds\n",
    "import sys, os\n",
    "sys.path.append(os.path.abspath(\"src\"))  \n",
    "from itables import init_notebook_mode\n",
    "import pandas as pd\n",
    "from itables import show\n",
    "from itables import options\n",
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "import fetchGSheet \n",
    "import warnings\n",
    "\n",
    "# Suppress SettingWithCopyWarning\n",
    "warnings.simplefilter(\"ignore\", category=UserWarning)\n",
    "\n",
    "\n",
    "# Select only the relevant columns from pop_up_info\n",
    "\n",
    "pop_up_info = fetchGSheet.pop_up_info.rename(columns={\"Mouse genome informatics (MGI) ID\": \"MGI ID\", \"Rat genome database (RGD) ID\": \"RGD ID\"})\n",
    "\n",
    "pop_up_info_lim = pop_up_info[[\"Approved symbol\", \"Approved name\", \"MGI ID\", \"RGD ID\"]]\n",
    "pop_up_info_lim = pop_up_info_lim.drop_duplicates(subset=\"Approved symbol\", keep=\"first\")\n",
    "\n",
    "# Drop columns where all values are NA in gene_pair\n",
    "gene_pair = fetchGSheet.gene_pair.dropna(axis=1, how='all')\n",
    "\n",
    "# Rename columns for better clarity\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "    \"Ligand receptor pair\": \"LR Pair\",\n",
    "    \"Ligand gene symbol\": \"Ligand\",\n",
    "    \"Receptor gene symbol\": \"Receptor\",\n",
    "    \"Perplexity link\": \"Perplexity\"\n",
    "})\n",
    "\n",
    "# Merge gene_pair with pop_up_info_lim for Ligand(L)\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Ligand', right_on='Approved symbol')\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Ligand name\", \n",
    "                                     \"MGI ID\": \"Ligand MGI ID\",\n",
    "                                     \"RGD ID\": \"Ligand RGD ID\"},\n",
    "                            )\n",
    "\n",
    "# Add MGI name\n",
    "MGI_info = pd.read_csv(\"data/MGI_ID_biomart.csv\")\n",
    "gene_pair = gene_pair.merge(MGI_info, how='left', left_on='Ligand MGI ID', right_on='MGI ID')\n",
    "\n",
    "# Add RGD name\n",
    "RGD_info = pd.read_csv(\"data/RGD_ID_biomart.csv\")\n",
    "RGD_info['RGD ID'] = \"RGD:\" + RGD_info['RGD ID'].astype(str)\n",
    "gene_pair = gene_pair.merge(RGD_info, how='left', left_on='Ligand RGD ID', right_on='RGD ID')\n",
    "\n",
    "gene_pair = gene_pair.drop(columns=[\"RGD ID\", \"MGI ID\"])\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "                                     \"MGI name\": \"Mouse Ligand\", \n",
    "                                     \"RGD name\": \"Rat Ligand\"}\n",
    "                            )\n",
    "\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Receptor', right_on='Approved symbol')\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Receptor name\",\n",
    "                                      \"MGI ID\": \"Receptor MGI ID\",\n",
    "                                      \"RGD ID\": \"Receptor RGD ID\"}\n",
    "                            )\n",
    "\n",
    "# Add MGI name\n",
    "gene_pair = gene_pair.merge(MGI_info, how='left', left_on='Receptor MGI ID', right_on='MGI ID')\n",
    "gene_pair = gene_pair.merge(RGD_info, how='left', left_on='Receptor RGD ID', right_on='RGD ID')\n",
    "gene_pair = gene_pair.drop(columns=[\"RGD ID\", \"MGI ID\"])\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "                                     \"MGI name\": \"Mouse Receptor\", \n",
    "                                     \"RGD name\": \"Rat Receptor\"}\n",
    "                            )\n",
    "gene_pair = gene_pair.drop(columns=[\"Approved symbol_x\", \"Approved symbol_y\"])\n",
    "\n",
    "# Drop columns where all values are NA in gene_pair\n",
    "gene_pair = gene_pair.dropna(axis=1, how='all')\n",
    "\n",
    "gene_pair = gene_pair.fillna(\" \")\n",
    "gene_pair = gene_pair[gene_pair['LR Pair'] != ' ']\n",
    "\n",
    "if \"PMID link\" in gene_pair.columns:\n",
    "    gene_pair = gene_pair.drop(columns=[\"PMID link\"])\n",
    "\n",
    "# Add\n",
    "first_columns=['LR Pair','Source', 'Ligand','Receptor', 'Perplexity']\n",
    "end_columns=['HGNC L R','sanity check', 'curator','secondary source?']\n",
    "gene_pair = gene_pair[first_columns + [col for col in gene_pair.columns if col not in first_columns + end_columns] + end_columns]\n",
    "\n",
    "# number of unique vars\n",
    "\n",
    "lrPairsCount = len(gene_pair[\"LR Pair\"].unique())\n",
    "\n",
    "ligandCount = len(gene_pair[\"Ligand\"].unique())\n",
    "\n",
    "receptorCount = len(gene_pair[\"Receptor\"].unique())\n",
    "\n",
    "# Mouse Orthologue\n",
    "MouseLigandCount = len(gene_pair[\"Ligand MGI ID\"].unique())\n",
    "\n",
    "MouseReceptorCount = len(gene_pair[\"Receptor MGI ID\"].unique())\n",
    "\n",
    "# Rat Orthologue\n",
    "RatLigandCount = len(gene_pair[\"Ligand RGD ID\"].unique())\n",
    "\n",
    "RatReceptorCount = len(gene_pair[\"Receptor RGD ID\"].unique())\n",
    "\n",
    "gene_pair[\"PMID support\"] = [value.replace(\" \", \"\") for value in gene_pair[\"PMID support\"]]\n",
    "\n",
    "source = np.array(gene_pair[\"PMID support\"].unique())\n",
    "source = source.astype(str)\n",
    "source = \",\".join(sorted(set(filter(lambda x: x.lower() != 'nan', source))))\n",
    "\n",
    "# Split the string into individual elements, filter out empty strings, and get unique values\n",
    "source = sorted(\n",
    "    set(filter(lambda x: x.strip() and x.strip().lower() != 'nan', source.split(',')))\n",
    ")\n",
    "source = [value.replace(\" \", \"\") for value in source]\n",
    "sourceCount = len(source)\n",
    "\n",
    "# for creating PMIDs\n",
    "gene_pair00 = gene_pair[['LR Pair', 'PMID support']]\n",
    "\n",
    "# create URLs for the HGNC IDs\n",
    "\n",
    "# ligand\n",
    "gene_pair[\"Ligand HGNC ID\"] = [\n",
    "    '<a href=\"https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/{}\" target=\"_blank\">{}</a>'.format(ligand, ligand)\n",
    "    for ligand in gene_pair[\"Ligand HGNC ID\"]\n",
    "]\n",
    "\n",
    "# receptor\n",
    "gene_pair[\"Receptor HGNC ID\"] = [\n",
    "    '<a href=\"https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/{}\" target=\"_blank\">{}</a>'.format(receptor, receptor)\n",
    "    for receptor in gene_pair[\"Receptor HGNC ID\"]\n",
    "]\n",
    "\n",
    "# Perplexity\n",
    "gene_pair[\"Perplexity\"] = [\n",
    "    '<a href=\"{}\" target=\"_blank\"> <img src=\"https://img.icons8.com/?size=35&id=0NbBuNOxUwps&format=png&color=000000\" alt=\"Perplexity AI\" /></a>'.format(url)\n",
    "    for url in gene_pair[\"Perplexity\"]\n",
    "]\n",
    "\n",
    "# Function to generate hyperlinks for the \"PMID support\" column\n",
    "# Function to generate hyperlinks for the \"PMID support\" column\n",
    "def generate_links_with_doi(df, gene_column, pmid_column):\n",
    "    def create_link(gene, sources):\n",
    "        # Replace spaces with \"——\" in the gene name for the link\n",
    "        gene_name = gene.replace(\" \", \"——\")\n",
    "        \n",
    "        if len(sources) == 1:\n",
    "            source = sources[0]\n",
    "            if source.startswith(\"https://www.biorxiv.org/content/\"):\n",
    "                # If the value starts with \"https://doi.org/\", use it as the hyperlink\n",
    "                return f'<a href=\"{source}\" target=\"_blank\">BioRxiv preprint</a>'\n",
    "            else:\n",
    "                # If it's a single PMID, hyperlink the PMID text\n",
    "                return f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/pubmed/{gene_name}_pmid_details.html\">{source}</a>'\n",
    "        else:\n",
    "            # If multiple PMIDs, show the count and hyperlink to the page\n",
    "            return f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/pubmed/{gene_name}_pmid_details.html\" target=\"_blank\">{len(sources)} PMIDs</a>'\n",
    "\n",
    "    # Process each row to generate the \"PMID support\" column\n",
    "    df[\"PMID support\"] = [\n",
    "        create_link(\n",
    "            gene=row[gene_column], \n",
    "            sources=[s.strip() for s in row[pmid_column].split(',') if s.strip()]\n",
    "        )\n",
    "        for _, row in df.iterrows()\n",
    "    ]\n",
    "    return df\n",
    "\n",
    "\n",
    "# Generate the links for the \"PMID support\" column\n",
    "gene_pair = generate_links_with_doi(gene_pair, gene_column=\"LR Pair\", pmid_column=\"PMID support\")\n",
    "\n",
    "gene_pair[\"Ligand MGI ID\"] = [\n",
    "        f'<a href=\"https://www.informatics.jax.org/marker/{mouseOrth}\" target=\"_blank\">{mouseOrth}</a>' \n",
    "        if pd.notna(mouseOrth) and mouseOrth.strip() else \"\" \n",
    "        for mouseOrth in gene_pair[\"Ligand MGI ID\"]\n",
    "    ]\n",
    "\n",
    "gene_pair[\"Receptor MGI ID\"] = [\n",
    "        f'<a href=\"https://www.informatics.jax.org/marker/{mouseOrth}\" target=\"_blank\">{mouseOrth}</a>' \n",
    "        if pd.notna(mouseOrth) and mouseOrth.strip() else \"\" \n",
    "        for mouseOrth in gene_pair[\"Receptor MGI ID\"]\n",
    "    ]\n",
    "\n",
    "gene_pair[\"Ligand RGD ID\"] = [\n",
    "        f'<a href=\"https://rgd.mcw.edu/rgdweb/report/gene/main.html?id={ratOrth.replace(\"RGD:\", \"\")}\" target=\"_blank\">{ratOrth}</a>' \n",
    "        if pd.notna(ratOrth) and ratOrth.strip() else \"\" \n",
    "        for ratOrth in gene_pair[\"Ligand RGD ID\"]\n",
    "    ]\n",
    "\n",
    "gene_pair[\"Receptor RGD ID\"] = [\n",
    "        f'<a href=\"https://rgd.mcw.edu/rgdweb/report/gene/main.html?id={ratOrth.replace(\"RGD:\", \"\")}\" target=\"_blank\">{ratOrth}</a>' \n",
    "        if pd.notna(ratOrth) and ratOrth.strip() else \"\" \n",
    "        for ratOrth in gene_pair[\"Receptor RGD ID\"]\n",
    "    ]\n",
    "\n",
    "mouse_columns = [col for col in gene_pair.columns if \"MGI\" in col]\n",
    "rat_columns = [col for col in gene_pair.columns if \"RGD\" in col]\n",
    "\n",
    "gene_pair0 = gene_pair[['LR Pair', 'Ligand', 'Receptor', 'Perplexity', 'PMID support',\n",
    "       'Ligand HGNC ID', 'Ligand location', 'Receptor HGNC ID',\n",
    "       'Receptor location', 'Ligand name', 'Receptor name'] + mouse_columns + rat_columns]\n",
    "\n",
    "\n",
    "# gene symbol\n",
    "gene_pair[\"Ligand\"] = [\n",
    "    f'<span title=\"{ligand_name}\">{ligand_symbol}</span>'\n",
    "    for ligand_name, ligand_symbol in zip(gene_pair[\"Ligand name\"], gene_pair[\"Ligand\"])\n",
    "]\n",
    "\n",
    "# gene symbol\n",
    "gene_pair[\"Receptor\"] = [\n",
    "    f'<span title=\"{receptor_name}\">{receptor_symbol}</span>'\n",
    "    for receptor_name, receptor_symbol in zip(gene_pair[\"Receptor name\"], gene_pair[\"Receptor\"])\n",
    "]\n",
    "\n",
    "def replace_spaces(row):\n",
    "    if row['Ligand location'] == 'secreted':\n",
    "        return row['LR Pair'].replace(\" \", \" ○ <span style='font-size: 30px;'>⤚</span> \")\n",
    "    elif row['Ligand location'] == 'plasma membrane':\n",
    "        return row['LR Pair'].replace(\" \", \" <span style='font-size: 30px;'>⤙</span> <span style='font-size: 30px;'>⤚</span> \")\n",
    "    else:\n",
    "        return row['LR Pair'].replace(\" \", \" \\u2192 \")\n",
    "\n",
    "# Apply the function to the 'LR Pair' column\n",
    "gene_pair['LR Pair'] = gene_pair.apply(replace_spaces, axis=1)\n",
    "\n",
    "gene_pair = gene_pair.drop(columns=[\"Ligand name\", \"Receptor name\"])\n",
    "\n",
    "\n",
    "# Create the links to the HTML cards\n",
    "gene_pair[\"LR Pair\"] = [\n",
    "    f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/cards/{lrPairOrig}.html\">{lrPair}</a>'\n",
    "    for lrPairOrig, lrPair in zip(gene_pair0[\"LR Pair\"], gene_pair[\"LR Pair\"])\n",
    "]\n",
    "\n",
    "\n",
    "# Add tooltips to the column headers\n",
    "gene_pair.columns = [\n",
    "    f'<span title=\"Ligand Receptor Pair\">{col}</span>' if col == \"LR Pair\" else\n",
    "    f'<span title=\"Click the logo below to run Perplexity on the LR pair\">{col}&nbsp;</span>' if col == \"Perplexity\" else\n",
    "    f'<span title=\"Hover on symbols below to show gene names\">{col}&nbsp;&nbsp;&nbsp;</span>' if col in [\"Ligand\", \"Receptor\"] else\n",
    "    f'<span title=\"Click on HGNC IDs below for more details\">{col}&nbsp;&nbsp;</span>' if col in [\"Ligand HGNC ID\", \"Receptor HGNC ID\"] else\n",
    "    f'<span title=\"Click on the Pubmed IDs (PMID) below for more details\">{col}</span>' if col == \"PMID support\" else\n",
    "    f'<span title=\"Click on the Rat Genome Database(RGD) IDs below for more details\">{col}</span>' if col in [\"Ligand RGD ID\", \"Receptor RGD ID\"] else\n",
    "    f'<span title=\"Click on the Mouse Genome Informatics(MGI) IDs below for more details\">{col}</span>' if col in [\"Ligand MGI ID\", \"Receptor MGI ID\"] else\n",
    "    f'<span title=\"Location is defined as subcellular location\">{col}</span>' if col in [\"Ligand location\", \"Receptor location\"] else\n",
    "    f'<span title=\"Double-click header of {col} to ensure all values are shown\">{col}&nbsp;</span>'\n",
    "    for col in gene_pair.columns\n",
    "]\n",
    "\n",
    "gene_pair = gene_pair.reset_index(drop=True)  # Remove the index\n",
    "gene_pair000 = gene_pair.copy()\n",
    "\n",
    "keywords_to_modify = [\"LR Pair\", \"Ligand\", \"Receptor\"]\n",
    "exclude_keywords = [\"HGNC ID\", \"Location\"]  # Columns containing this will not be modified\n",
    "\n",
    "# Copy the original columns so we can modify only the first 10\n",
    "new_columns = gene_pair000.columns.tolist()\n",
    "\n",
    "# Modify only the first 10 columns\n",
    "new_columns[:10] = [\n",
    "    f'<span title=\"{col.split(\">\")[0]}\">Human {col.split(\">\")[1]}</span>'\n",
    "    if any(keyword in col for keyword in keywords_to_modify) and not any(exclude in col for exclude in exclude_keywords)\n",
    "    else col\n",
    "    for col in new_columns[:10]\n",
    "]\n",
    "\n",
    "# Assign the modified column names back to the DataFrame\n",
    "gene_pair000.columns = new_columns\n",
    "human_columns = [col for col in gene_pair000.columns][:10]\n",
    "human_columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "3e85f9c2-c753-4f09-9f53-7e7f4f8822a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['<span title=\"Ligand Receptor Pair\">LR Pair</span>',\n",
       " '<span title=\"Double-click header of Source to ensure all values are shown\">Source&nbsp;</span>',\n",
       " '<span title=\"Hover on symbols below to show gene names\">Ligand&nbsp;&nbsp;&nbsp;</span>',\n",
       " '<span title=\"Hover on symbols below to show gene names\">Receptor&nbsp;&nbsp;&nbsp;</span>',\n",
       " '<span title=\"Click the logo below to run Perplexity on the LR pair\">Perplexity&nbsp;</span>',\n",
       " '<span title=\"Click on the Pubmed IDs (PMID) below for more details\">PMID support</span>',\n",
       " '<span title=\"Click on HGNC IDs below for more details\">Ligand HGNC ID&nbsp;&nbsp;</span>',\n",
       " '<span title=\"Location is defined as subcellular location\">Ligand location</span>',\n",
       " '<span title=\"Click on HGNC IDs below for more details\">Receptor HGNC ID&nbsp;&nbsp;</span>',\n",
       " '<span title=\"Location is defined as subcellular location\">Receptor location</span>']"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Modify only the first 10 columns conditionally by adding \"Human \" prefix\n",
    "# Apply renaming to the columns based on conditions\n",
    "human_columns = [col for col in gene_pair.columns][:10]\n",
    "human_columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6c6ea636-dcf3-46a2-8782-286618a230dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['<span title=\"Ligand Receptor Pair\">LR Pair</span>',\n",
       " '<span title=\"Double-click header of Source to ensure all values are shown\">Source&nbsp;</span>',\n",
       " '<span title=\"Hover on symbols below to show gene names\">Ligand&nbsp;&nbsp;&nbsp;</span>',\n",
       " '<span title=\"Hover on symbols below to show gene names\">Receptor&nbsp;&nbsp;&nbsp;</span>',\n",
       " '<span title=\"Click the logo below to run Perplexity on the LR pair\">Perplexity&nbsp;</span>',\n",
       " '<span title=\"Click on the Pubmed IDs (PMID) below for more details\">PMID support</span>',\n",
       " '<span title=\"Click on HGNC IDs below for more details\">Ligand HGNC ID&nbsp;&nbsp;</span>',\n",
       " '<span title=\"Location is defined as subcellular location\">Ligand location</span>',\n",
       " '<span title=\"Click on HGNC IDs below for more details\">Receptor HGNC ID&nbsp;&nbsp;</span>',\n",
       " '<span title=\"Location is defined as subcellular location\">Receptor location</span>',\n",
       " '<span title=\"<span title=\"Click on the Mouse Genome Informatics(MGI) IDs below for more details\"\">Human Ligand MGI ID</span</span>',\n",
       " '<span title=\"<span title=\"Click on the Rat Genome Database(RGD) IDs below for more details\"\">Human Ligand RGD ID</span</span>',\n",
       " '<span title=\"<span title=\"Double-click header of Mouse Ligand to ensure all values are shown\"\">Human Mouse Ligand&nbsp;</span</span>',\n",
       " '<span title=\"<span title=\"Double-click header of Rat Ligand to ensure all values are shown\"\">Human Rat Ligand&nbsp;</span</span>',\n",
       " '<span title=\"<span title=\"Click on the Mouse Genome Informatics(MGI) IDs below for more details\"\">Human Receptor MGI ID</span</span>',\n",
       " '<span title=\"<span title=\"Click on the Rat Genome Database(RGD) IDs below for more details\"\">Human Receptor RGD ID</span</span>',\n",
       " '<span title=\"<span title=\"Double-click header of Mouse Receptor to ensure all values are shown\"\">Human Mouse Receptor&nbsp;</span</span>',\n",
       " '<span title=\"<span title=\"Double-click header of Rat Receptor to ensure all values are shown\"\">Human Rat Receptor&nbsp;</span</span>',\n",
       " '<span title=\"Double-click header of HGNC L R to ensure all values are shown\">HGNC L R&nbsp;</span>',\n",
       " '<span title=\"Double-click header of sanity check to ensure all values are shown\">sanity check&nbsp;</span>',\n",
       " '<span title=\"Double-click header of curator to ensure all values are shown\">curator&nbsp;</span>',\n",
       " '<span title=\"Double-click header of secondary source? to ensure all values are shown\">secondary source?&nbsp;</span>']"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef05b729-f13f-4701-b135-c641c9b0c7a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to convert MGI/RGD IDs/Other Species to gene symbols via biomart\n",
    "import os\n",
    "import sys\n",
    "from biomart import BiomartServer\n",
    "import pandas as pd\n",
    "\n",
    "# Add source directory to the path\n",
    "sys.path.append(os.path.abspath(\"src\"))\n",
    "from createDataTable import pop_up_info_lim\n",
    "\n",
    "# Function to dynamically query BioMart\n",
    "def fetch_gene_mappings(server_url, dataset_name, id_type, id_list, id_filter, attributes, chunk_size=100):\n",
    "    \"\"\"\n",
    "    Fetch gene mappings from BioMart based on the provided IDs.\n",
    "\n",
    "    Args:\n",
    "        server_url (str): URL of the BioMart server.\n",
    "        dataset_name (str): Dataset name (e.g., 'mmusculus_gene_ensembl' for mouse, 'rnorvegicus_gene_ensembl' for rat).\n",
    "        id_type (str): Type of IDs ('MGI' or 'RGD').\n",
    "        id_list (list): List of IDs to query.\n",
    "        id_filter (str): BioMart filter for the ID type (e.g., 'mgi_id', 'rgd_id').\n",
    "        attributes (list): List of attributes to fetch (e.g., ['mgi_id', 'external_gene_name']).\n",
    "        chunk_size (int): Number of IDs to query in each chunk.\n",
    "\n",
    "    Returns:\n",
    "        dict: Mapping of input IDs to external gene names.\n",
    "    \"\"\"\n",
    "    # Connect to the BioMart server\n",
    "    server = BiomartServer(server_url)\n",
    "    dataset = server.datasets[dataset_name]\n",
    "\n",
    "    # Filter unique and valid IDs\n",
    "    id_list = [id.strip() for id in id_list if id.startswith(id_type)]\n",
    "    id_list = list(set(id_list))\n",
    "\n",
    "    # Query in chunks\n",
    "    gene_mapping = {}\n",
    "    for i in range(0, len(id_list), chunk_size):\n",
    "        chunk = id_list[i:i + chunk_size]\n",
    "        try:\n",
    "            response = dataset.search({\n",
    "                'filters': {id_filter: chunk},\n",
    "                'attributes': attributes\n",
    "            })\n",
    "            # Parse the response and add to gene_mapping\n",
    "            for line in response.iter_lines(decode_unicode=True):\n",
    "                gene_id, gene_name = line.split(\"\\t\")\n",
    "                gene_mapping[gene_id] = gene_name\n",
    "        except Exception as e:\n",
    "            print(f\"Error processing chunk {i // chunk_size + 1}: {e}\")\n",
    "    return gene_mapping\n",
    "\n",
    "# Fetch MGI to gene name mappings for mouse\n",
    "mgi_ids = pop_up_info_lim['MGI ID'].unique()\n",
    "gene_mapping_mouse = fetch_gene_mappings(\n",
    "    server_url=\"http://www.ensembl.org/biomart\",\n",
    "    dataset_name='mmusculus_gene_ensembl',\n",
    "    id_type=\"MGI\",\n",
    "    id_list=mgi_ids,\n",
    "    id_filter='mgi_id',\n",
    "    attributes=['mgi_id', 'external_gene_name'],\n",
    "    chunk_size=100\n",
    ")\n",
    "pd.DataFrame.from_dict(gene_mapping_mouse, orient='index', columns=['Gene Name']).to_csv(\"data/MGI_ID_biomart.csv\")\n",
    "\n",
    "\n",
    "# Fetch RGD to gene name mappings for rat\n",
    "rgd_ids = pop_up_info_lim['RGD ID'].unique()\n",
    "gene_mapping_rat = fetch_gene_mappings(\n",
    "    server_url=\"http://www.ensembl.org/biomart\",\n",
    "    dataset_name='rnorvegicus_gene_ensembl',\n",
    "    id_type=\"RGD\",\n",
    "    id_list=rgd_ids,\n",
    "    id_filter='rgd_id',\n",
    "    attributes=['rgd_id', 'external_gene_name'],\n",
    "    chunk_size=100\n",
    ")\n",
    "pd.DataFrame.from_dict(gene_mapping_rat, orient='index', columns=['Gene Name']).to_csv(\"data/RGD_ID_biomart.csv\")\n",
    "\n",
    "# Print the results\n",
    "# print(\"Mouse gene mappings:\")\n",
    "# for mgi, gene_name in gene_mapping_mouse.items():\n",
    "#     print(f\"{mgi}: {gene_name}\")\n",
    "\n",
    "# print(\"\\nRat gene mappings:\")\n",
    "# for rgd, gene_name in gene_mapping_rat.items():\n",
    "#     print(f\"{rgd}: {gene_name}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "116dc0cf-4ac7-407c-873e-6387f561a9fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to create Ligand-Receptor pair cards\n",
    "\n",
    "import os\n",
    "import jinja2\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "sys.path.append(os.path.abspath(\"src\"))  \n",
    "import fetchGSheet\n",
    "from createDataTable import gene_pair0\n",
    "\n",
    "# Paths\n",
    "TEMPLATE_PATH = 'HTML/cardTemplate.html'\n",
    "OUTPUT_DIR = 'data/cards/'\n",
    "\n",
    "def load_template(template_path):\n",
    "    \"\"\"Load Jinja2 template from a file.\"\"\"\n",
    "    with open(template_path, 'r') as file:\n",
    "        return jinja2.Template(file.read())\n",
    "\n",
    "def prepare_dataframes(gene_pair0):\n",
    "    \"\"\"Prepare interaction, ligand, and receptor dataframes.\"\"\"\n",
    "    DBlength = len(gene_pair0)\n",
    "    gene_pair0[\"Interaction ID\"] = [f\"CDB{str(i).zfill(4)}\" for i in range(1, DBlength + 1)]\n",
    "    gene_pair0[\"Interaction Type\"] = [\n",
    "        f'{ligandLocation} {ligand} binds with {receptor} in {receptorLocation}'\n",
    "        for ligand, ligandLocation, receptor, receptorLocation in zip(\n",
    "            gene_pair0[\"Ligand\"], gene_pair0[\"Ligand location\"],\n",
    "            gene_pair0[\"Receptor\"], gene_pair0[\"Receptor location\"]\n",
    "        )\n",
    "    ]\n",
    "    interaction_card = gene_pair0[[\"Interaction ID\", \"LR Pair\", \"Interaction Type\", \"Perplexity\", \"PMID support\"]]\n",
    "    interaction_card[\"Perplexity\"] = interaction_card[\"Perplexity\"].str.replace('size=35', 'size=80')\n",
    "\n",
    "    pop_up_info_lim = fetchGSheet.pop_up_info[\n",
    "        [\"Approved symbol\", \"Alias symbol\", \"Previous symbol\", \"Date symbol changed\"]\n",
    "    ].drop_duplicates(subset=\"Approved symbol\", keep=\"first\")\n",
    "    \n",
    "    ligand_card = gene_pair0[[\"LR Pair\", \"Ligand\", \"Ligand name\", \"Ligand HGNC ID\", \"Ligand MGI ID\", \"Ligand RGD ID\", \"Ligand location\"]].merge(\n",
    "        pop_up_info_lim, how='left', left_on='Ligand', right_on='Approved symbol'\n",
    "    ).drop_duplicates(subset='LR Pair', keep=\"first\").drop(columns=[\"Ligand\", \"Approved symbol\"])\n",
    "\n",
    "    ligand_card_1 = ligand_card[[\"LR Pair\", \"Alias symbol\", \"Date symbol changed\", \"Ligand name\"]] \n",
    "    ligand_card_2 = ligand_card[[\"LR Pair\", \"Ligand HGNC ID\", \"Ligand MGI ID\", \"Ligand RGD ID\", \"Ligand location\"]] \n",
    "    \n",
    "    receptor_card = gene_pair0[[\"LR Pair\", \"Receptor\", \"Receptor name\", \"Receptor HGNC ID\", \"Receptor MGI ID\", \"Receptor RGD ID\", \"Receptor location\"]].merge(\n",
    "        pop_up_info_lim, how='left', left_on='Receptor', right_on='Approved symbol'\n",
    "    ).drop_duplicates(subset='LR Pair', keep=\"first\").drop(columns=[\"Receptor\", \"Approved symbol\"])\n",
    "    \n",
    "    receptor_card_1 = receptor_card[[\"LR Pair\", \"Alias symbol\", \"Date symbol changed\", \"Receptor name\"]] \n",
    "    receptor_card_2 = receptor_card[[\"LR Pair\", \"Receptor HGNC ID\", \"Receptor MGI ID\", \"Receptor RGD ID\", \"Receptor location\"]] \n",
    "\n",
    "    return interaction_card, ligand_card_1, ligand_card_2, receptor_card_1, receptor_card_2\n",
    "\n",
    "def generate_html_files(template, interaction_card, ligand_card_1, receptor_card_1, ligand_card_2, receptor_card_2, output_dir):\n",
    "    \"\"\"Generate HTML files for each LR Pair.\"\"\"\n",
    "    column_values = interaction_card[\"LR Pair\"].dropna().unique()\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "    \n",
    "    for value in column_values:\n",
    "        value1, value2 = value.split()\n",
    "        row0 = interaction_card[interaction_card['LR Pair'] == value]\n",
    "        row1 = ligand_card_1[ligand_card_1['LR Pair'] == value]\n",
    "        row2 = receptor_card_1[receptor_card_1['LR Pair'] == value]\n",
    "        row3 = ligand_card_2[ligand_card_2['LR Pair'] == value]\n",
    "        row4 = receptor_card_2[receptor_card_2['LR Pair'] == value]\n",
    "        \n",
    "\n",
    "        table0_data = row0.drop('LR Pair', axis=1).to_dict(orient='records')[0] if not row0.empty else {}\n",
    "        table1_data = row1.drop('LR Pair', axis=1).to_dict(orient='records')[0] if not row1.empty else {}\n",
    "        table2_data = row2.drop('LR Pair', axis=1).to_dict(orient='records')[0] if not row2.empty else {}\n",
    "        table3_data = row3.drop('LR Pair', axis=1).to_dict(orient='records')[0] if not row3.empty else {}\n",
    "        table4_data = row4.drop('LR Pair', axis=1).to_dict(orient='records')[0] if not row4.empty else {}\n",
    "\n",
    "        rendered_content = template.render(\n",
    "            value1=value1,\n",
    "            value2=value2,\n",
    "            table0_data=table0_data,\n",
    "            table1_data=table1_data,\n",
    "            table2_data=table2_data,\n",
    "            table3_data=table3_data,\n",
    "            table4_data=table4_data\n",
    "        )\n",
    "        \n",
    "        output_file = os.path.join(output_dir, f\"{value1} {value2}.html\")\n",
    "        with open(output_file, 'w') as file:\n",
    "            file.write(rendered_content)\n",
    "\n",
    "# Main execution\n",
    "if __name__ == \"__main__\":\n",
    "    template = load_template(TEMPLATE_PATH)\n",
    "    interaction_card, ligand_card_1, receptor_card_1, ligand_card_2, receptor_card_2 = prepare_dataframes(gene_pair0)\n",
    "    generate_html_files(template, interaction_card, ligand_card_1, receptor_card_1, ligand_card_2, receptor_card_2, OUTPUT_DIR)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "386c564f-f89b-4814-a5b9-c1799a87afc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "rat_gene_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "367639a9-ab68-46ae-bf50-2bd4cfbff3f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to prepare datatables (cleaning and hyperlinking, adding tool tips, etc) input for the database qmds\n",
    "import sys, os\n",
    "sys.path.append(os.path.abspath(\"src\"))  \n",
    "from itables import init_notebook_mode\n",
    "import pandas as pd\n",
    "from itables import show\n",
    "from itables import options\n",
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "import fetchGSheet \n",
    "\n",
    "# Select only the relevant columns from pop_up_info\n",
    "\n",
    "pop_up_info = fetchGSheet.pop_up_info.rename(columns={\"Mouse genome informatics (MGI) ID\": \"MGI ID\",\"Rat genome database (RGD) ID\": \"RGD ID\"})\n",
    "\n",
    "pop_up_info_lim = pop_up_info[[\"Approved symbol\", \"Approved name\", \"MGI ID\", \"RGD ID\"]]\n",
    "pop_up_info_lim = pop_up_info_lim.drop_duplicates(subset=\"Approved symbol\", keep=\"first\")\n",
    "\n",
    "# Drop columns where all values are NA in gene_pair\n",
    "gene_pair = fetchGSheet.gene_pair.dropna(axis=1, how='all')\n",
    "\n",
    "# Rename columns for better clarity\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "    \"Ligand receptor pair\": \"LR Pair\",\n",
    "    \"Ligand gene symbol\": \"Ligand\",\n",
    "    \"Receptor gene symbol\": \"Receptor\",\n",
    "    \"Perplexity link\": \"Perplexity\"\n",
    "})\n",
    "\n",
    "# Merge gene_pair with pop_up_info_lim for Ligand(L)\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Ligand', right_on='Approved symbol')\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Ligand name\", \n",
    "                                     \"MGI ID\": \"Ligand MGI ID\",\n",
    "                                     \"RGD ID\": \"Ligand RGD ID\"},\n",
    "                            )\n",
    "\n",
    "# Add MGI name\n",
    "MGI_info = pd.read_csv(\"data/MGI_ID_biomart.csv\")\n",
    "gene_pair = gene_pair.merge(MGI_info, how='left', left_on='Ligand MGI ID', right_on='MGI ID')\n",
    "gene_pair = gene_pair.drop(columns=[\"MGI ID\"])\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "                                     \"MGI name\": \"Ligand MGI name\"}\n",
    "                            )\n",
    "\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Receptor', right_on='Approved symbol')\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Receptor name\",\n",
    "                                      \"MGI ID\": \"Receptor MGI ID\",\n",
    "                                      \"RGD ID\": \"Receptor RGD ID\"}\n",
    "                            )\n",
    "\n",
    "# Add MGI name\n",
    "gene_pair = gene_pair.merge(MGI_info, how='left', left_on='Receptor MGI ID', right_on='MGI ID')\n",
    "gene_pair = gene_pair.drop(columns=[\"MGI ID\"])\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "                                     \"MGI name\": \"Receptor MGI name\"}\n",
    "                            )\n",
    "gene_pair.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94dd3cc8-7853-4453-9698-29a7bd511b62",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81c7a762-5b40-46dc-8a56-d8f7c5680360",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b6a25aa-f15a-4209-9fb5-e5ab52dd5d30",
   "metadata": {},
   "outputs": [],
   "source": [
    "MGI_info = pd.read_csv(\"data/MGI_ID_biomart.csv\")\n",
    "MGI_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bee1b45-6798-4359-9fef-7957f0c5f4fc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "quarto_env",
   "language": "python",
   "name": "quarto_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
