{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fcc5ddc1-519e-439c-a130-3dd8bcd181bd",
   "metadata": {},
   "source": [
    "# Python Notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b443872e-66e7-427a-a031-75a16471e8a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_row(change):\n",
    "    global gene_pair\n",
    "    # Add a new row at the top with None values\n",
    "    new_row = {col: None for col in gene_pair.columns}\n",
    "    gene_pair = pd.DataFrame([new_row] + gene_pair.to_dict(orient=\"records\"))\n",
    "    update_table()\n",
    "\n",
    "# Function to remove the last row of the dataframe\n",
    "def remove_row(change):\n",
    "    global gene_pair\n",
    "    if len(gene_pair) > 0:\n",
    "        gene_pair = gene_pair[:-1]  # Remove the last row\n",
    "        update_table()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "228f917d-b5b9-409e-bc04-7b95afd9a725",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19161b56-8989-458f-80ef-d1e3eb97e029",
   "metadata": {},
   "outputs": [],
   "source": [
    "duplicates = gene_pair00[gene_pair00[\"Human LR Pair\"].duplicated()]\n",
    "print(duplicates[\"Human LR Pair\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "a9d79fce-3cfa-469a-abdb-cd05c49d221b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1450"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys\n",
    "import os\n",
    "\n",
    "sys.path.append(os.path.abspath(\"src\"))  # Add src directory to path\n",
    "from createDataTable import gene_pair0\n",
    "\n",
    "# Get all unique genes\n",
    "ligand_list = gene_pair0[\"Ligand\"].tolist()\n",
    "receptor_list = gene_pair0[\"Receptor\"].tolist()\n",
    "unique_genes = list(set(ligand_list + receptor_list))  # Combine and remove duplicates\n",
    "len(unique_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "7707cc4d-426a-41a1-8f1a-fe68ffb62330",
   "metadata": {},
   "outputs": [],
   "source": [
    "connectomeDB = pd.read_table(\"data/connectome_j.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "484a55e5-4c5d-4d8f-9bb7-4138af267ee2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Eutheria', 'Deuterostomia', 'Euteleostomi', 'Tetrapoda', 'LCA',\n",
       "       'Opisthokonta', 'Theria', 'Eukaryota', 'Euarchontoglires',\n",
       "       'Bilateria', 'Chordata', nan, 'Mammalia', 'Catarrhini', 'Amniota',\n",
       "       'Human', 'Homininae'], dtype=object)"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "connectomeDB['Taxon'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "14cc7b9c-5cda-49d7-8a5b-efc16555df48",
   "metadata": {},
   "outputs": [],
   "source": [
    "connectomeDB = connectomeDB[connectomeDB[\"Taxon\"]== \"Human\"]\n",
    "connectomeDB = connectomeDB.drop(columns=[\"Localization\", \"Taxon\"] + [col for col in connectomeDB.columns if col.startswith(\"F5_\")])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "063a9b88-083f-4cf4-a6b2-d9b90de36941",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ApprovedSymbol</th>\n",
       "      <th>Adipocyte Breast</th>\n",
       "      <th>Adipocyte Omental</th>\n",
       "      <th>Adipocyte Perirenal</th>\n",
       "      <th>Adipocyte Subcutaneous</th>\n",
       "      <th>Alveolar Epithelial</th>\n",
       "      <th>Amniotic Epithelial</th>\n",
       "      <th>Amniotic Membrane</th>\n",
       "      <th>Annulus Pulposus</th>\n",
       "      <th>Astrocyte Cerebellum</th>\n",
       "      <th>...</th>\n",
       "      <th>Smooth Muscle Subclavian Artery</th>\n",
       "      <th>Smooth Muscle Tracheal</th>\n",
       "      <th>Smooth Muscle Umbilical Artery</th>\n",
       "      <th>Smooth Muscle Umbilical Vein</th>\n",
       "      <th>Smooth Muscle Uterine</th>\n",
       "      <th>Synoviocyte</th>\n",
       "      <th>Tenocyte</th>\n",
       "      <th>Trabecular Meshwork</th>\n",
       "      <th>Tracheal Epithelial</th>\n",
       "      <th>Urothelial</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>ACTR3C</td>\n",
       "      <td>0.503</td>\n",
       "      <td>1.023</td>\n",
       "      <td>0.872</td>\n",
       "      <td>0.000</td>\n",
       "      <td>3.239</td>\n",
       "      <td>3.443</td>\n",
       "      <td>3.871</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.873</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.617</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.313</td>\n",
       "      <td>0.355</td>\n",
       "      <td>0.376</td>\n",
       "      <td>0.586</td>\n",
       "      <td>0.525</td>\n",
       "      <td>2.317</td>\n",
       "      <td>1.758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>ADAT2</td>\n",
       "      <td>3.231</td>\n",
       "      <td>3.508</td>\n",
       "      <td>5.234</td>\n",
       "      <td>6.398</td>\n",
       "      <td>8.762</td>\n",
       "      <td>6.664</td>\n",
       "      <td>3.304</td>\n",
       "      <td>2.563</td>\n",
       "      <td>6.817</td>\n",
       "      <td>...</td>\n",
       "      <td>6.714</td>\n",
       "      <td>3.665</td>\n",
       "      <td>4.726</td>\n",
       "      <td>4.293</td>\n",
       "      <td>4.257</td>\n",
       "      <td>3.394</td>\n",
       "      <td>5.543</td>\n",
       "      <td>5.995</td>\n",
       "      <td>18.675</td>\n",
       "      <td>8.048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>ADC</td>\n",
       "      <td>0.577</td>\n",
       "      <td>0.773</td>\n",
       "      <td>0.872</td>\n",
       "      <td>0.769</td>\n",
       "      <td>0.246</td>\n",
       "      <td>2.170</td>\n",
       "      <td>2.017</td>\n",
       "      <td>4.636</td>\n",
       "      <td>3.679</td>\n",
       "      <td>...</td>\n",
       "      <td>2.351</td>\n",
       "      <td>1.273</td>\n",
       "      <td>1.195</td>\n",
       "      <td>2.451</td>\n",
       "      <td>1.064</td>\n",
       "      <td>1.575</td>\n",
       "      <td>1.277</td>\n",
       "      <td>3.428</td>\n",
       "      <td>2.343</td>\n",
       "      <td>2.246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>ADIG</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326</th>\n",
       "      <td>ADORA2A</td>\n",
       "      <td>0.091</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.298</td>\n",
       "      <td>0.349</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.297</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.109</td>\n",
       "      <td>...</td>\n",
       "      <td>0.772</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2.155</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.705</td>\n",
       "      <td>0.278</td>\n",
       "      <td>0.379</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17184</th>\n",
       "      <td>ZNF843</td>\n",
       "      <td>1.155</td>\n",
       "      <td>4.577</td>\n",
       "      <td>1.745</td>\n",
       "      <td>2.335</td>\n",
       "      <td>0.246</td>\n",
       "      <td>0.505</td>\n",
       "      <td>0.297</td>\n",
       "      <td>1.304</td>\n",
       "      <td>1.423</td>\n",
       "      <td>...</td>\n",
       "      <td>0.184</td>\n",
       "      <td>0.617</td>\n",
       "      <td>0.332</td>\n",
       "      <td>0.532</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.516</td>\n",
       "      <td>1.485</td>\n",
       "      <td>0.175</td>\n",
       "      <td>0.395</td>\n",
       "      <td>0.218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17198</th>\n",
       "      <td>ZNF891</td>\n",
       "      <td>1.674</td>\n",
       "      <td>0.823</td>\n",
       "      <td>1.745</td>\n",
       "      <td>0.682</td>\n",
       "      <td>0.246</td>\n",
       "      <td>1.028</td>\n",
       "      <td>1.910</td>\n",
       "      <td>2.108</td>\n",
       "      <td>1.301</td>\n",
       "      <td>...</td>\n",
       "      <td>0.330</td>\n",
       "      <td>0.328</td>\n",
       "      <td>0.665</td>\n",
       "      <td>0.823</td>\n",
       "      <td>1.774</td>\n",
       "      <td>1.845</td>\n",
       "      <td>1.019</td>\n",
       "      <td>0.554</td>\n",
       "      <td>0.640</td>\n",
       "      <td>0.873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17202</th>\n",
       "      <td>ZNF93</td>\n",
       "      <td>3.903</td>\n",
       "      <td>3.736</td>\n",
       "      <td>1.745</td>\n",
       "      <td>2.371</td>\n",
       "      <td>4.531</td>\n",
       "      <td>2.889</td>\n",
       "      <td>4.261</td>\n",
       "      <td>1.571</td>\n",
       "      <td>6.253</td>\n",
       "      <td>...</td>\n",
       "      <td>1.177</td>\n",
       "      <td>2.797</td>\n",
       "      <td>2.059</td>\n",
       "      <td>9.404</td>\n",
       "      <td>6.740</td>\n",
       "      <td>1.832</td>\n",
       "      <td>2.206</td>\n",
       "      <td>5.471</td>\n",
       "      <td>2.335</td>\n",
       "      <td>3.269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17238</th>\n",
       "      <td>ZSCAN30</td>\n",
       "      <td>6.787</td>\n",
       "      <td>7.138</td>\n",
       "      <td>9.596</td>\n",
       "      <td>3.835</td>\n",
       "      <td>3.425</td>\n",
       "      <td>4.031</td>\n",
       "      <td>4.497</td>\n",
       "      <td>8.077</td>\n",
       "      <td>3.281</td>\n",
       "      <td>...</td>\n",
       "      <td>4.667</td>\n",
       "      <td>4.360</td>\n",
       "      <td>3.494</td>\n",
       "      <td>4.157</td>\n",
       "      <td>6.031</td>\n",
       "      <td>8.258</td>\n",
       "      <td>6.964</td>\n",
       "      <td>2.713</td>\n",
       "      <td>3.747</td>\n",
       "      <td>4.549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17243</th>\n",
       "      <td>ZSWIM1</td>\n",
       "      <td>5.236</td>\n",
       "      <td>6.757</td>\n",
       "      <td>4.362</td>\n",
       "      <td>8.878</td>\n",
       "      <td>8.797</td>\n",
       "      <td>9.162</td>\n",
       "      <td>5.602</td>\n",
       "      <td>4.009</td>\n",
       "      <td>9.795</td>\n",
       "      <td>...</td>\n",
       "      <td>5.032</td>\n",
       "      <td>6.714</td>\n",
       "      <td>5.949</td>\n",
       "      <td>5.378</td>\n",
       "      <td>5.321</td>\n",
       "      <td>4.053</td>\n",
       "      <td>6.180</td>\n",
       "      <td>12.619</td>\n",
       "      <td>5.643</td>\n",
       "      <td>11.211</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>466 rows × 145 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      ApprovedSymbol  Adipocyte Breast  Adipocyte Omental  \\\n",
       "210           ACTR3C             0.503              1.023   \n",
       "280            ADAT2             3.231              3.508   \n",
       "282              ADC             0.577              0.773   \n",
       "313             ADIG             0.000              0.000   \n",
       "326          ADORA2A             0.091              0.000   \n",
       "...              ...               ...                ...   \n",
       "17184         ZNF843             1.155              4.577   \n",
       "17198         ZNF891             1.674              0.823   \n",
       "17202          ZNF93             3.903              3.736   \n",
       "17238        ZSCAN30             6.787              7.138   \n",
       "17243         ZSWIM1             5.236              6.757   \n",
       "\n",
       "       Adipocyte Perirenal  Adipocyte Subcutaneous  Alveolar Epithelial  \\\n",
       "210                  0.872                   0.000                3.239   \n",
       "280                  5.234                   6.398                8.762   \n",
       "282                  0.872                   0.769                0.246   \n",
       "313                  0.000                   0.000                0.000   \n",
       "326                  0.000                   0.298                0.349   \n",
       "...                    ...                     ...                  ...   \n",
       "17184                1.745                   2.335                0.246   \n",
       "17198                1.745                   0.682                0.246   \n",
       "17202                1.745                   2.371                4.531   \n",
       "17238                9.596                   3.835                3.425   \n",
       "17243                4.362                   8.878                8.797   \n",
       "\n",
       "       Amniotic Epithelial  Amniotic Membrane  Annulus Pulposus  \\\n",
       "210                  3.443              3.871             0.062   \n",
       "280                  6.664              3.304             2.563   \n",
       "282                  2.170              2.017             4.636   \n",
       "313                  0.000              0.000             0.000   \n",
       "326                  0.000              0.297             0.000   \n",
       "...                    ...                ...               ...   \n",
       "17184                0.505              0.297             1.304   \n",
       "17198                1.028              1.910             2.108   \n",
       "17202                2.889              4.261             1.571   \n",
       "17238                4.031              4.497             8.077   \n",
       "17243                9.162              5.602             4.009   \n",
       "\n",
       "       Astrocyte Cerebellum  ...  Smooth Muscle Subclavian Artery  \\\n",
       "210                   0.873  ...                            0.000   \n",
       "280                   6.817  ...                            6.714   \n",
       "282                   3.679  ...                            2.351   \n",
       "313                   0.000  ...                            0.000   \n",
       "326                   0.109  ...                            0.772   \n",
       "...                     ...  ...                              ...   \n",
       "17184                 1.423  ...                            0.184   \n",
       "17198                 1.301  ...                            0.330   \n",
       "17202                 6.253  ...                            1.177   \n",
       "17238                 3.281  ...                            4.667   \n",
       "17243                 9.795  ...                            5.032   \n",
       "\n",
       "       Smooth Muscle Tracheal  Smooth Muscle Umbilical Artery  \\\n",
       "210                     0.617                           0.000   \n",
       "280                     3.665                           4.726   \n",
       "282                     1.273                           1.195   \n",
       "313                     0.000                           0.000   \n",
       "326                     0.000                           2.155   \n",
       "...                       ...                             ...   \n",
       "17184                   0.617                           0.332   \n",
       "17198                   0.328                           0.665   \n",
       "17202                   2.797                           2.059   \n",
       "17238                   4.360                           3.494   \n",
       "17243                   6.714                           5.949   \n",
       "\n",
       "       Smooth Muscle Umbilical Vein  Smooth Muscle Uterine  Synoviocyte  \\\n",
       "210                           0.313                  0.355        0.376   \n",
       "280                           4.293                  4.257        3.394   \n",
       "282                           2.451                  1.064        1.575   \n",
       "313                           0.000                  0.000        0.000   \n",
       "326                           0.000                  0.000        0.705   \n",
       "...                             ...                    ...          ...   \n",
       "17184                         0.532                  0.000        1.516   \n",
       "17198                         0.823                  1.774        1.845   \n",
       "17202                         9.404                  6.740        1.832   \n",
       "17238                         4.157                  6.031        8.258   \n",
       "17243                         5.378                  5.321        4.053   \n",
       "\n",
       "       Tenocyte  Trabecular Meshwork  Tracheal Epithelial  Urothelial  \n",
       "210       0.586                0.525                2.317       1.758  \n",
       "280       5.543                5.995               18.675       8.048  \n",
       "282       1.277                3.428                2.343       2.246  \n",
       "313       0.000                0.000                0.000       0.000  \n",
       "326       0.278                0.379                0.000       0.000  \n",
       "...         ...                  ...                  ...         ...  \n",
       "17184     1.485                0.175                0.395       0.218  \n",
       "17198     1.019                0.554                0.640       0.873  \n",
       "17202     2.206                5.471                2.335       3.269  \n",
       "17238     6.964                2.713                3.747       4.549  \n",
       "17243     6.180               12.619                5.643      11.211  \n",
       "\n",
       "[466 rows x 145 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "connectomeDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "e33aa7dc-ccf5-4afa-8350-c69f01b67f68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ApprovedSymbol</th>\n",
       "      <th>Adipocyte Breast</th>\n",
       "      <th>Adipocyte Omental</th>\n",
       "      <th>Adipocyte Perirenal</th>\n",
       "      <th>Adipocyte Subcutaneous</th>\n",
       "      <th>Alveolar Epithelial</th>\n",
       "      <th>Amniotic Epithelial</th>\n",
       "      <th>Amniotic Membrane</th>\n",
       "      <th>Annulus Pulposus</th>\n",
       "      <th>Astrocyte Cerebellum</th>\n",
       "      <th>...</th>\n",
       "      <th>Smooth Muscle Subclavian Artery</th>\n",
       "      <th>Smooth Muscle Tracheal</th>\n",
       "      <th>Smooth Muscle Umbilical Artery</th>\n",
       "      <th>Smooth Muscle Umbilical Vein</th>\n",
       "      <th>Smooth Muscle Uterine</th>\n",
       "      <th>Synoviocyte</th>\n",
       "      <th>Tenocyte</th>\n",
       "      <th>Trabecular Meshwork</th>\n",
       "      <th>Tracheal Epithelial</th>\n",
       "      <th>Urothelial</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>ACTR3C</td>\n",
       "      <td>0.407463</td>\n",
       "      <td>0.704582</td>\n",
       "      <td>0.627007</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.444327</td>\n",
       "      <td>1.491330</td>\n",
       "      <td>1.583299</td>\n",
       "      <td>0.060154</td>\n",
       "      <td>0.627541</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.480573</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.272315</td>\n",
       "      <td>0.303801</td>\n",
       "      <td>0.319181</td>\n",
       "      <td>0.461215</td>\n",
       "      <td>0.421994</td>\n",
       "      <td>1.199061</td>\n",
       "      <td>1.014506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>ADAT2</td>\n",
       "      <td>1.442438</td>\n",
       "      <td>1.505854</td>\n",
       "      <td>1.830018</td>\n",
       "      <td>2.001210</td>\n",
       "      <td>2.278497</td>\n",
       "      <td>2.036534</td>\n",
       "      <td>1.459545</td>\n",
       "      <td>1.270603</td>\n",
       "      <td>2.056301</td>\n",
       "      <td>...</td>\n",
       "      <td>2.043037</td>\n",
       "      <td>1.540088</td>\n",
       "      <td>1.745017</td>\n",
       "      <td>1.666385</td>\n",
       "      <td>1.659561</td>\n",
       "      <td>1.480240</td>\n",
       "      <td>1.878396</td>\n",
       "      <td>1.945196</td>\n",
       "      <td>2.979349</td>\n",
       "      <td>2.202544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>ADC</td>\n",
       "      <td>0.455524</td>\n",
       "      <td>0.572673</td>\n",
       "      <td>0.627007</td>\n",
       "      <td>0.570414</td>\n",
       "      <td>0.219938</td>\n",
       "      <td>1.153732</td>\n",
       "      <td>1.104263</td>\n",
       "      <td>1.729175</td>\n",
       "      <td>1.543084</td>\n",
       "      <td>...</td>\n",
       "      <td>1.209259</td>\n",
       "      <td>0.821101</td>\n",
       "      <td>0.786182</td>\n",
       "      <td>1.238664</td>\n",
       "      <td>0.724646</td>\n",
       "      <td>0.945850</td>\n",
       "      <td>0.822859</td>\n",
       "      <td>1.487948</td>\n",
       "      <td>1.206869</td>\n",
       "      <td>1.177423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>ADIG</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326</th>\n",
       "      <td>ADORA2A</td>\n",
       "      <td>0.087095</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.260825</td>\n",
       "      <td>0.299364</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.260054</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.103459</td>\n",
       "      <td>...</td>\n",
       "      <td>0.572109</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.148988</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.533565</td>\n",
       "      <td>0.245296</td>\n",
       "      <td>0.321359</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17184</th>\n",
       "      <td>ZNF843</td>\n",
       "      <td>0.767791</td>\n",
       "      <td>1.718651</td>\n",
       "      <td>1.009781</td>\n",
       "      <td>1.204473</td>\n",
       "      <td>0.219938</td>\n",
       "      <td>0.408793</td>\n",
       "      <td>0.260054</td>\n",
       "      <td>0.834647</td>\n",
       "      <td>0.885006</td>\n",
       "      <td>...</td>\n",
       "      <td>0.168899</td>\n",
       "      <td>0.480573</td>\n",
       "      <td>0.286682</td>\n",
       "      <td>0.426574</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.922670</td>\n",
       "      <td>0.910273</td>\n",
       "      <td>0.161268</td>\n",
       "      <td>0.332894</td>\n",
       "      <td>0.197210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17198</th>\n",
       "      <td>ZNF891</td>\n",
       "      <td>0.983575</td>\n",
       "      <td>0.600483</td>\n",
       "      <td>1.009781</td>\n",
       "      <td>0.519984</td>\n",
       "      <td>0.219938</td>\n",
       "      <td>0.707050</td>\n",
       "      <td>1.068153</td>\n",
       "      <td>1.133979</td>\n",
       "      <td>0.833344</td>\n",
       "      <td>...</td>\n",
       "      <td>0.285179</td>\n",
       "      <td>0.283674</td>\n",
       "      <td>0.509825</td>\n",
       "      <td>0.600483</td>\n",
       "      <td>1.020290</td>\n",
       "      <td>1.045563</td>\n",
       "      <td>0.702602</td>\n",
       "      <td>0.440832</td>\n",
       "      <td>0.494696</td>\n",
       "      <td>0.627541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17202</th>\n",
       "      <td>ZNF93</td>\n",
       "      <td>1.589847</td>\n",
       "      <td>1.555193</td>\n",
       "      <td>1.009781</td>\n",
       "      <td>1.215209</td>\n",
       "      <td>1.710369</td>\n",
       "      <td>1.358152</td>\n",
       "      <td>1.660321</td>\n",
       "      <td>0.944295</td>\n",
       "      <td>1.981415</td>\n",
       "      <td>...</td>\n",
       "      <td>0.777948</td>\n",
       "      <td>1.334211</td>\n",
       "      <td>1.118088</td>\n",
       "      <td>2.342190</td>\n",
       "      <td>2.046402</td>\n",
       "      <td>1.040983</td>\n",
       "      <td>1.165024</td>\n",
       "      <td>1.867331</td>\n",
       "      <td>1.204473</td>\n",
       "      <td>1.451380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17238</th>\n",
       "      <td>ZSCAN30</td>\n",
       "      <td>2.052456</td>\n",
       "      <td>2.096544</td>\n",
       "      <td>2.360477</td>\n",
       "      <td>1.575881</td>\n",
       "      <td>1.487270</td>\n",
       "      <td>1.615619</td>\n",
       "      <td>1.704202</td>\n",
       "      <td>2.205744</td>\n",
       "      <td>1.454187</td>\n",
       "      <td>...</td>\n",
       "      <td>1.734660</td>\n",
       "      <td>1.678964</td>\n",
       "      <td>1.502743</td>\n",
       "      <td>1.640355</td>\n",
       "      <td>1.950329</td>\n",
       "      <td>2.225488</td>\n",
       "      <td>2.074931</td>\n",
       "      <td>1.311840</td>\n",
       "      <td>1.557513</td>\n",
       "      <td>1.713618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17243</th>\n",
       "      <td>ZSWIM1</td>\n",
       "      <td>1.830339</td>\n",
       "      <td>2.048596</td>\n",
       "      <td>1.679337</td>\n",
       "      <td>2.290310</td>\n",
       "      <td>2.282076</td>\n",
       "      <td>2.318655</td>\n",
       "      <td>1.887373</td>\n",
       "      <td>1.611236</td>\n",
       "      <td>2.379083</td>\n",
       "      <td>...</td>\n",
       "      <td>1.797079</td>\n",
       "      <td>2.043037</td>\n",
       "      <td>1.938598</td>\n",
       "      <td>1.852855</td>\n",
       "      <td>1.843877</td>\n",
       "      <td>1.619982</td>\n",
       "      <td>1.971299</td>\n",
       "      <td>2.611466</td>\n",
       "      <td>1.893564</td>\n",
       "      <td>2.502337</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>466 rows × 145 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      ApprovedSymbol  Adipocyte Breast  Adipocyte Omental  \\\n",
       "210           ACTR3C          0.407463           0.704582   \n",
       "280            ADAT2          1.442438           1.505854   \n",
       "282              ADC          0.455524           0.572673   \n",
       "313             ADIG          0.000000           0.000000   \n",
       "326          ADORA2A          0.087095           0.000000   \n",
       "...              ...               ...                ...   \n",
       "17184         ZNF843          0.767791           1.718651   \n",
       "17198         ZNF891          0.983575           0.600483   \n",
       "17202          ZNF93          1.589847           1.555193   \n",
       "17238        ZSCAN30          2.052456           2.096544   \n",
       "17243         ZSWIM1          1.830339           2.048596   \n",
       "\n",
       "       Adipocyte Perirenal  Adipocyte Subcutaneous  Alveolar Epithelial  \\\n",
       "210               0.627007                0.000000             1.444327   \n",
       "280               1.830018                2.001210             2.278497   \n",
       "282               0.627007                0.570414             0.219938   \n",
       "313               0.000000                0.000000             0.000000   \n",
       "326               0.000000                0.260825             0.299364   \n",
       "...                    ...                     ...                  ...   \n",
       "17184             1.009781                1.204473             0.219938   \n",
       "17198             1.009781                0.519984             0.219938   \n",
       "17202             1.009781                1.215209             1.710369   \n",
       "17238             2.360477                1.575881             1.487270   \n",
       "17243             1.679337                2.290310             2.282076   \n",
       "\n",
       "       Amniotic Epithelial  Amniotic Membrane  Annulus Pulposus  \\\n",
       "210               1.491330           1.583299          0.060154   \n",
       "280               2.036534           1.459545          1.270603   \n",
       "282               1.153732           1.104263          1.729175   \n",
       "313               0.000000           0.000000          0.000000   \n",
       "326               0.000000           0.260054          0.000000   \n",
       "...                    ...                ...               ...   \n",
       "17184             0.408793           0.260054          0.834647   \n",
       "17198             0.707050           1.068153          1.133979   \n",
       "17202             1.358152           1.660321          0.944295   \n",
       "17238             1.615619           1.704202          2.205744   \n",
       "17243             2.318655           1.887373          1.611236   \n",
       "\n",
       "       Astrocyte Cerebellum  ...  Smooth Muscle Subclavian Artery  \\\n",
       "210                0.627541  ...                         0.000000   \n",
       "280                2.056301  ...                         2.043037   \n",
       "282                1.543084  ...                         1.209259   \n",
       "313                0.000000  ...                         0.000000   \n",
       "326                0.103459  ...                         0.572109   \n",
       "...                     ...  ...                              ...   \n",
       "17184              0.885006  ...                         0.168899   \n",
       "17198              0.833344  ...                         0.285179   \n",
       "17202              1.981415  ...                         0.777948   \n",
       "17238              1.454187  ...                         1.734660   \n",
       "17243              2.379083  ...                         1.797079   \n",
       "\n",
       "       Smooth Muscle Tracheal  Smooth Muscle Umbilical Artery  \\\n",
       "210                  0.480573                        0.000000   \n",
       "280                  1.540088                        1.745017   \n",
       "282                  0.821101                        0.786182   \n",
       "313                  0.000000                        0.000000   \n",
       "326                  0.000000                        1.148988   \n",
       "...                       ...                             ...   \n",
       "17184                0.480573                        0.286682   \n",
       "17198                0.283674                        0.509825   \n",
       "17202                1.334211                        1.118088   \n",
       "17238                1.678964                        1.502743   \n",
       "17243                2.043037                        1.938598   \n",
       "\n",
       "       Smooth Muscle Umbilical Vein  Smooth Muscle Uterine  Synoviocyte  \\\n",
       "210                        0.272315               0.303801     0.319181   \n",
       "280                        1.666385               1.659561     1.480240   \n",
       "282                        1.238664               0.724646     0.945850   \n",
       "313                        0.000000               0.000000     0.000000   \n",
       "326                        0.000000               0.000000     0.533565   \n",
       "...                             ...                    ...          ...   \n",
       "17184                      0.426574               0.000000     0.922670   \n",
       "17198                      0.600483               1.020290     1.045563   \n",
       "17202                      2.342190               2.046402     1.040983   \n",
       "17238                      1.640355               1.950329     2.225488   \n",
       "17243                      1.852855               1.843877     1.619982   \n",
       "\n",
       "       Tenocyte  Trabecular Meshwork  Tracheal Epithelial  Urothelial  \n",
       "210    0.461215             0.421994             1.199061    1.014506  \n",
       "280    1.878396             1.945196             2.979349    2.202544  \n",
       "282    0.822859             1.487948             1.206869    1.177423  \n",
       "313    0.000000             0.000000             0.000000    0.000000  \n",
       "326    0.245296             0.321359             0.000000    0.000000  \n",
       "...         ...                  ...                  ...         ...  \n",
       "17184  0.910273             0.161268             0.332894    0.197210  \n",
       "17198  0.702602             0.440832             0.494696    0.627541  \n",
       "17202  1.165024             1.867331             1.204473    1.451380  \n",
       "17238  2.074931             1.311840             1.557513    1.713618  \n",
       "17243  1.971299             2.611466             1.893564    2.502337  \n",
       "\n",
       "[466 rows x 145 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# log(x+1) transform\n",
    "connectomeDB.iloc[:, 1:] = np.log1p(connectomeDB.iloc[:, 1:])\n",
    "connectomeDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "52913d1f-6671-4daf-a9bc-1c9d6fa32a47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ApprovedSymbol</th>\n",
       "      <th>cellTypes</th>\n",
       "      <th>expr_val</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACTR3C</td>\n",
       "      <td>Adipocyte Breast</td>\n",
       "      <td>0.407463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADAT2</td>\n",
       "      <td>Adipocyte Breast</td>\n",
       "      <td>1.442438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADC</td>\n",
       "      <td>Adipocyte Breast</td>\n",
       "      <td>0.455524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADIG</td>\n",
       "      <td>Adipocyte Breast</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADORA2A</td>\n",
       "      <td>Adipocyte Breast</td>\n",
       "      <td>0.087095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67099</th>\n",
       "      <td>ZNF843</td>\n",
       "      <td>Urothelial</td>\n",
       "      <td>0.197210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67100</th>\n",
       "      <td>ZNF891</td>\n",
       "      <td>Urothelial</td>\n",
       "      <td>0.627541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67101</th>\n",
       "      <td>ZNF93</td>\n",
       "      <td>Urothelial</td>\n",
       "      <td>1.451380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67102</th>\n",
       "      <td>ZSCAN30</td>\n",
       "      <td>Urothelial</td>\n",
       "      <td>1.713618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67103</th>\n",
       "      <td>ZSWIM1</td>\n",
       "      <td>Urothelial</td>\n",
       "      <td>2.502337</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67104 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      ApprovedSymbol         cellTypes  expr_val\n",
       "0             ACTR3C  Adipocyte Breast  0.407463\n",
       "1              ADAT2  Adipocyte Breast  1.442438\n",
       "2                ADC  Adipocyte Breast  0.455524\n",
       "3               ADIG  Adipocyte Breast  0.000000\n",
       "4            ADORA2A  Adipocyte Breast  0.087095\n",
       "...              ...               ...       ...\n",
       "67099         ZNF843        Urothelial  0.197210\n",
       "67100         ZNF891        Urothelial  0.627541\n",
       "67101          ZNF93        Urothelial  1.451380\n",
       "67102        ZSCAN30        Urothelial  1.713618\n",
       "67103         ZSWIM1        Urothelial  2.502337\n",
       "\n",
       "[67104 rows x 3 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reshape \n",
    "connectomeDB_long = connectomeDB.melt(id_vars=[\"ApprovedSymbol\"], \n",
    "                                      var_name=\"cellTypes\", value_name=\"expr_val\")\n",
    "connectomeDB_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "0e9af35d-fa08-49e3-98fa-b1892023cd76",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "144"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(connectomeDB_long[\"cellTypes\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "085d46de-c47c-4d00-944f-517ce6c67729",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ApprovedSymbol</th>\n",
       "      <th>cellTypes</th>\n",
       "      <th>expr_val</th>\n",
       "      <th>cellCategory</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACTR3C</td>\n",
       "      <td>Adipocyte Breast</td>\n",
       "      <td>0.407463</td>\n",
       "      <td>mesenchymal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADAT2</td>\n",
       "      <td>Adipocyte Breast</td>\n",
       "      <td>1.442438</td>\n",
       "      <td>mesenchymal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADC</td>\n",
       "      <td>Adipocyte Breast</td>\n",
       "      <td>0.455524</td>\n",
       "      <td>mesenchymal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADIG</td>\n",
       "      <td>Adipocyte Breast</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mesenchymal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADORA2A</td>\n",
       "      <td>Adipocyte Breast</td>\n",
       "      <td>0.087095</td>\n",
       "      <td>mesenchymal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68031</th>\n",
       "      <td>ZNF843</td>\n",
       "      <td>Urothelial</td>\n",
       "      <td>0.197210</td>\n",
       "      <td>epithelial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68032</th>\n",
       "      <td>ZNF891</td>\n",
       "      <td>Urothelial</td>\n",
       "      <td>0.627541</td>\n",
       "      <td>epithelial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68033</th>\n",
       "      <td>ZNF93</td>\n",
       "      <td>Urothelial</td>\n",
       "      <td>1.451380</td>\n",
       "      <td>epithelial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68034</th>\n",
       "      <td>ZSCAN30</td>\n",
       "      <td>Urothelial</td>\n",
       "      <td>1.713618</td>\n",
       "      <td>epithelial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68035</th>\n",
       "      <td>ZSWIM1</td>\n",
       "      <td>Urothelial</td>\n",
       "      <td>2.502337</td>\n",
       "      <td>epithelial</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68036 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      ApprovedSymbol         cellTypes  expr_val cellCategory\n",
       "0             ACTR3C  Adipocyte Breast  0.407463  mesenchymal\n",
       "1              ADAT2  Adipocyte Breast  1.442438  mesenchymal\n",
       "2                ADC  Adipocyte Breast  0.455524  mesenchymal\n",
       "3               ADIG  Adipocyte Breast  0.000000  mesenchymal\n",
       "4            ADORA2A  Adipocyte Breast  0.087095  mesenchymal\n",
       "...              ...               ...       ...          ...\n",
       "68031         ZNF843        Urothelial  0.197210   epithelial\n",
       "68032         ZNF891        Urothelial  0.627541   epithelial\n",
       "68033          ZNF93        Urothelial  1.451380   epithelial\n",
       "68034        ZSCAN30        Urothelial  1.713618   epithelial\n",
       "68035         ZSWIM1        Urothelial  2.502337   epithelial\n",
       "\n",
       "[68036 rows x 4 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cellCat = pd.read_csv(\"data/cell_categories.csv\")\n",
    "connectomeDB_long = connectomeDB_long.merge(cellCat, how='left', left_on='cellTypes', right_on='cellType')\n",
    "connectomeDB_long = connectomeDB_long.drop(columns=[\"cellType\"])\n",
    "connectomeDB_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "1ee1d4a9-bd80-4a11-88c4-f3dea6a03335",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                                   Mature Adipocyte\n",
       "1                          Renal Cortical Epithelial\n",
       "2                                      CD19+ B cells\n",
       "3                                Adipocyte Perirenal\n",
       "4                                  Keratinocyte Oral\n",
       "                           ...                      \n",
       "139                                         ES cells\n",
       "140                    Mesenchymal stem cell Adipose\n",
       "141    CD14+ Monocyte derived Endothelial progenitor\n",
       "142                             Bronchial Epithelial\n",
       "143                                       Urothelial\n",
       "Length: 144, dtype: object"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intersection = pd.Series(list(set(connectomeDB_long['cellTypes']).intersection(set(cellCat['cellType']))))\n",
    "intersection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "77d40068-c428-4eb3-bc5e-c06c66aa39b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], dtype: object)"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff_df = pd.Series(list(set(connectomeDB_long['cellTypes']).difference(set(cellCat['cellType']))))\n",
    "diff_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "d1ce1ef8-1c72-43ed-9c29-ee3e9dca3a0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                       Embryonic Stem Cell\n",
       "1      CD4+CD25-CD45RA+ Naive Conv. T Cells\n",
       "2                            Natural Killer\n",
       "3       CD4+CD25+CD45RA+ Naive Reg. T Cells\n",
       "4                        Mast Cells stimul.\n",
       "5     CD14+ Monocyte der. Endothelial prog.\n",
       "6             CD34+ Hematopoietic Stem Cell\n",
       "7                               CD34+ prog.\n",
       "8       CD4+CD25-CD45RA- Mem. Conv. T Cells\n",
       "9          Dendritic Monocyte Immature der.\n",
       "10                           Mallassez der.\n",
       "11             Granulocyte Macrophage prog.\n",
       "12                    Intestinal Epithelial\n",
       "13                                  MCBUSSC\n",
       "14       CD4+CD25+CD45RA- Mem. Reg. T Cells\n",
       "15                 Macrophage Monocyte der.\n",
       "16           CD133+ Hematopoietic Stem Cell\n",
       "dtype: object"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff_df2 = pd.Series(list(set(cellCat['cellType']).difference(set(connectomeDB_long['cellTypes']))))\n",
    "diff_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "ed3d29fe-85d4-4f01-94b7-e989a512bb80",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['mesenchymal', 'epithelial', 'other', 'nervous system',\n",
       "       'hematopoietic', 'endothelial'], dtype=object)"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "connectomeDB_long[\"cellCategory\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "6b6d5b6b-456f-416d-a709-c3aa929ed1e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "def plot_gene_expression(df):\n",
    "    colors = {\n",
    "        \"mesenchymal\": \"#1f77b4\",  # Blue  \n",
    "        \"epithelial\": \"#ff7f0e\",   # Orange  \n",
    "        \"hematopoietic\": \"#2ca02c\",  # Green  \n",
    "        \"endothelial\": \"#d62728\",  # Red  \n",
    "        \"nervous system\": \"#9467bd\",  # Purple  \n",
    "        \"other\": \"#8c564b\",  # Brown  \n",
    "        \"missing\": \"#7f7f7f\",  # Gray  \n",
    "        np.nan: \"#e377c2\"  # Pink for NaN values\n",
    "    }\n",
    "\n",
    "    # Define sorting order for cell categories\n",
    "    category_order = {cat: i for i, cat in enumerate(colors.keys())}\n",
    "\n",
    "    for gene, sub_df in df.groupby(\"ApprovedSymbol\"):\n",
    "        # Sort by category first, then by expression value (highest first)\n",
    "        sub_df = sub_df.copy()\n",
    "        sub_df[\"category_order\"] = sub_df[\"cellCategory\"].map(category_order).fillna(len(category_order))\n",
    "        sub_df = sub_df.sort_values([\"category_order\", \"expr_val\"], ascending=[True, False])\n",
    "\n",
    "        fig, ax = plt.subplots(figsize=(8, min(10, max(4, len(sub_df) * 0.3))))  # Limit vertical size\n",
    "\n",
    "        # Assign colors\n",
    "        bar_colors = sub_df[\"cellCategory\"].map(colors).fillna(\"gray\")\n",
    "\n",
    "        # Plot\n",
    "        ax.barh(sub_df[\"cellTypes\"], sub_df[\"expr_val\"], color=bar_colors)\n",
    "\n",
    "        ax.set_title(gene, fontsize=14)\n",
    "        ax.set_xlabel(\"Expression Value (log(x+1) transformed\", fontsize=12)\n",
    "        ax.set_ylabel(\"\")\n",
    "        ax.tick_params(axis=\"x\", labelsize=10)\n",
    "        ax.tick_params(axis=\"y\", labelsize=6)\n",
    "\n",
    "        # Legend: Only show categories that exist in the plot\n",
    "        unique_categories = sub_df[\"cellCategory\"].dropna().unique()\n",
    "        legend_handles = [plt.Rectangle((0, 0), 1, 1, color=colors[cat]) for cat in unique_categories]\n",
    "        ax.legend(legend_handles, unique_categories, title=\"Cell Category\", fontsize=9, title_fontsize=11, loc=\"upper right\")\n",
    "\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(f\"data/gene_expr_plots/{gene}.png\", dpi=150, bbox_inches=\"tight\")\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "8945b0e6-3e2f-40d3-b02a-459414257694",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_gene_expression(connectomeDB_long)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7f4b069-0370-424e-9b83-4caaefce586d",
   "metadata": {},
   "source": [
    "## Testing Liana+"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c78bfa0-43b3-4857-9ca4-41dcd8a230e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import liana as li\n",
    "import omnipath as op\n",
    "import decoupler as dc\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "766bd3fd-d53d-4586-8d1e-09befbaa9675",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "sys.path.append(os.path.abspath(\"src\"))  # Add src directory to path\n",
    "from createDataTable import gene_pair0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f681087-38ba-4cfa-a105-8707ff70ac5f",
   "metadata": {},
   "source": [
    "### Pathway Annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a67bd553-c8df-4970-9811-435c46e2c51c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load PROGENy pathways, we use decoupler as a proxy as it formats the data in a more convenient way\n",
    "progeny = dc.get_progeny(top=10000)\n",
    "progeny"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95b2a08f-d44b-48ce-8a26-4f74ec640a51",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_pairs = gene_pair0[[\"Ligand\", \"Receptor\"]]\n",
    "lr_pairs.columns = lr_pairs.columns.str.lower()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7f4b57a-700a-4e04-9c95-f33719a4f877",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "548ba05e-8929-4434-a765-0d77b67d0c4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate ligand-receptor geneset\n",
    "lr_progeny = li.rs.generate_lr_geneset(lr_pairs, progeny, lr_sep=\"^\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "091fba45-e3f2-4e26-b1fb-6bd3ed8ebacb",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_progeny"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9beaedff-b8b8-4c41-b8c7-7ead6e5403aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# some of the pairs are missing\n",
    "len(lr_progeny[\"interaction\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23e12dce-6c86-4006-b9ba-01a6924aa0f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_file=\"data/pathway_annotations_per_pair.csv\"\n",
    "lr_progeny.to_csv(output_file, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e477d7b-ced3-41ea-b412-65d9936b6644",
   "metadata": {},
   "outputs": [],
   "source": [
    "whichDB= 'DisGeNet'\n",
    "# A database of expression profiles related to human diseases, including cancer\n",
    "diseases = op.requests.Annotations.get(\n",
    "    resources = [whichDB]\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb551b9f-add0-45ea-98f8-58c7f3fe8a54",
   "metadata": {},
   "outputs": [],
   "source": [
    "diseases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4fd5feb-feee-4193-aabc-6da6257b0dc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "diseases.to_csv(\"data/\" + whichDB + \".csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf34da1b-e119-4c62-bfd4-41770dd2d010",
   "metadata": {},
   "source": [
    "### Disease Annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1056987-e2ca-495c-a03f-78d002b63302",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DisGeNet\n",
    "diseases = op.requests.Annotations.get(\n",
    "    resources = ['DisGeNet']\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79d3b1e6-9ceb-4782-9c00-ef11379f0ea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "diseases = diseases[['genesymbol', 'label', 'value']]\n",
    "diseases = diseases.pivot_table(index='genesymbol',\n",
    "                                columns='label', values='value',\n",
    "                                aggfunc=lambda x: '; '.join(x)).reset_index()\n",
    "diseases = diseases[['genesymbol', 'disease']]\n",
    "diseases['disease'] = diseases['disease'].str.split('; ')\n",
    "diseases = diseases.explode('disease')\n",
    "lr_diseases = li.rs.generate_lr_geneset(lr_pairs, diseases, source='disease', target='genesymbol', weight=None, lr_sep=\"^\")\n",
    "lr_diseases.sort_values(\"interaction\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "115a5201-dd41-4519-ae48-68fc7e004828",
   "metadata": {},
   "outputs": [],
   "source": [
    "# some of the pairs are missing\n",
    "len(lr_diseases[\"interaction\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fd738f5-cebd-49e2-be19-673450d33efb",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_file=\"data/disease_annotations_per_pair.csv\"\n",
    "lr_diseases.to_csv(output_file, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de5cef07-a5eb-483d-8f28-ba9b65a5cda2",
   "metadata": {},
   "outputs": [],
   "source": [
    "op.requests.Annotations.resources()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2afd8af-fa7d-4a3a-b4a5-9f67d1b42e86",
   "metadata": {},
   "source": [
    "### Get FASTA sequences for each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a7dd4db-bf07-4131-b3b5-f9611116cb7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "import sys\n",
    "import os\n",
    "\n",
    "sys.path.append(os.path.abspath(\"src\"))  # Add src directory to path\n",
    "from createDataTable import gene_pair0\n",
    "\n",
    "# Get all unique genes\n",
    "ligand_list = gene_pair0[\"Ligand\"].tolist()\n",
    "receptor_list = gene_pair0[\"Receptor\"].tolist()\n",
    "unique_genes = list(set(ligand_list + receptor_list))  # Combine and remove duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78afcc47-8c40-45e1-8e7a-1ee3adaf1e74",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "import gzip\n",
    "\n",
    "# Input FASTA file\n",
    "fasta_file = \"data/uniprotkb_proteome_UP000005640_AND_revi_2025_03_27.fasta.gz\"\n",
    "\n",
    "# Regex pattern to extract details from header\n",
    "header_pattern = re.compile(r\"^>sp\\|(?P<uniprot_id>[A-Z0-9]+(?:-\\d+)?)\\|(?P<protein_name>.+?) OS=Homo sapiens OX=9606 GN=(?P<gene_name>[A-Za-z0-9-]+)\")\n",
    "\n",
    "# Store extracted data\n",
    "records = []\n",
    "\n",
    "# Read and parse the file\n",
    "with gzip.open(fasta_file, \"rt\") as f:\n",
    "    header = None\n",
    "    sequence = []\n",
    "    \n",
    "    for line in f:\n",
    "        line = line.strip()\n",
    "        \n",
    "        if line.startswith(\">\"):\n",
    "            # Store previous sequence if exists\n",
    "            if header and sequence:\n",
    "                isoform_type = \"Canonical\" if \"-\" not in header[\"uniprot_id\"] else \"Alternative Isoform\"\n",
    "                records.append([header[\"uniprot_id\"], header[\"gene_name\"], header[\"protein_name\"], isoform_type, \"\".join(sequence)])\n",
    "            \n",
    "            # Match new header\n",
    "            match = header_pattern.match(line)\n",
    "            if match:\n",
    "                header = match.groupdict()\n",
    "                sequence = []\n",
    "            else:\n",
    "                header = None\n",
    "        \n",
    "        elif header:\n",
    "            sequence.append(line)\n",
    "\n",
    "    # Add the last record\n",
    "    if header and sequence:\n",
    "        isoform_type = \"Canonical\" if \"-\" not in header[\"uniprot_id\"] else \"Alternative Isoform\"\n",
    "        records.append([header[\"uniprot_id\"], header[\"gene_name\"], header[\"protein_name\"], isoform_type, \"\".join(sequence)])\n",
    "\n",
    "# Convert to pandas DataFrame\n",
    "df = pd.DataFrame(records, columns=[\"UniProt ID\", \"Gene Symbol\", \"Protein Name\", \"Isoform Type\", \"FASTA Sequence\"])\n",
    "\n",
    "# Save as TSV\n",
    "df.to_csv(\"data/human_uniprot_isoforms.tsv\", sep=\"\\t\", index=False)\n",
    "\n",
    "print(f\"✅ Extracted {len(df)} Homo sapiens protein sequences and saved to 'human_uniprot_isoforms.tsv'.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21b854e6-20a7-4c04-9268-e7df424c9bf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df[\"Gene Symbol\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46458bdf-7f37-4f86-9d03-27b894f15faa",
   "metadata": {},
   "outputs": [],
   "source": [
    "df= pd.read_table(\"data/human_uniprot_isoforms.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8602f29d-65b0-4cda-9f04-0b8ef3159a13",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb7e7d70-7f61-4786-bae3-759a8d509344",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[['UniProt ID', 'Gene Symbol', 'Isoform Type', 'FASTA Sequence']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d7aeca2-f763-4b85-8503-65921fc934aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88a97936-a9d2-471d-b287-3482fcfd357b",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df = gene_pair0[[\"Human LR Pair\", \"Ligand\", \"Receptor\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd3f4d8c-b251-4174-84de-727cd77ea189",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbb66bb9-0275-442d-b5bb-de269e41642d",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df = lim_df.merge(df, how='left', left_on='Ligand', right_on='Gene Symbol')\n",
    "lim_df = lim_df.drop(columns=[\"Gene Symbol\"])\n",
    "lim_df = lim_df.rename(columns={\"FASTA Sequence\": \"Ligand Sequence\",\n",
    "                                \"UniProt ID\": \"Ligand Isoform Uniprot ID\",\n",
    "                                \"Isoform Type\": \"Ligand Isoform Type\"})\n",
    "lim_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "336f774d-e37e-4d24-b415-d0eccb1eccb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df = lim_df.merge(df, how='left', left_on='Receptor', right_on='Gene Symbol')\n",
    "lim_df = lim_df.drop(columns=[\"Gene Symbol\"])\n",
    "lim_df = lim_df.rename(columns={\"FASTA Sequence\": \"Receptor Sequence\",\n",
    "                                \"UniProt ID\": \"Receptor Isoform Uniprot ID\",\n",
    "                                \"Isoform Type\": \"Receptor Isoform Type\"})\n",
    "lim_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af8e9511-4026-4ef3-866a-d52aabd55e18",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df.to_csv(\"data/LRpair_uniprot_sequences.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef772ddd-9d77-4557-a1e0-84605e6e084d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import re\n",
    "import pandas as pd\n",
    "\n",
    "# Step 1: Extract Gene Symbol Mapping from GTF\n",
    "gtf_file = \"data/gencode.v47.annotation.gtf.gz\"\n",
    "gene_map = {}\n",
    "\n",
    "# Read GTF file and extract gene_id -> gene_name mapping\n",
    "with gzip.open(gtf_file, \"rt\") as f:\n",
    "    for line in f:\n",
    "        if line.startswith(\"#\"):  # Skip comments\n",
    "            continue\n",
    "        \n",
    "        fields = line.strip().split(\"\\t\")\n",
    "        if fields[2] == \"gene\":  # Only extract gene entries\n",
    "            info = {key.strip(): value.strip('\"') for key, value in re.findall(r'(\\S+) \"([^\"]+)\"', fields[8])}\n",
    "            if \"gene_id\" in info and \"gene_name\" in info:\n",
    "                gene_map[info[\"gene_id\"]] = info[\"gene_name\"]\n",
    "\n",
    "print(f\"✅ Extracted {len(gene_map)} gene mappings from GTF.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5251cc7d-871d-4d23-936a-20703a223ce4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 2: Parse GENCODE Protein FASTA and Add Gene Symbols\n",
    "fasta_file = \"data/gencode.v47.pc_translations.fa.gz\"\n",
    "\n",
    "# Store extracted data\n",
    "records = []\n",
    "\n",
    "# Open the GENCODE FASTA file and parse sequences\n",
    "with gzip.open(fasta_file, \"rt\") as f:\n",
    "    header = None\n",
    "    sequence = []\n",
    "    \n",
    "    for line in f:\n",
    "        line = line.strip()\n",
    "        \n",
    "        if line.startswith(\">\"):\n",
    "            # Store previous sequence if exists\n",
    "            if header and sequence:\n",
    "                # Extract the Gene Symbol using the Gene ID\n",
    "                gene_symbol = gene_map.get(header[\"gene_id\"], \"Unknown\")\n",
    "                isoform_type = \"Canonical\" if \"-1\" in header[\"protein_id\"] else \"Alternative Isoform\"\n",
    "                \n",
    "                # Append the parsed data to records\n",
    "                records.append([header[\"protein_id\"], header[\"transcript_id\"], header[\"gene_id\"], gene_symbol, isoform_type, \"\".join(sequence)])\n",
    "            \n",
    "            # Split header by '|' and extract necessary fields\n",
    "            fields = line[1:].split(\"|\")  # Skip the '>' symbol and split by '|'\n",
    "            if len(fields) >= 6:\n",
    "                header = {\n",
    "                    \"protein_id\": fields[0], \n",
    "                    \"transcript_id\": fields[1], \n",
    "                    \"gene_id\": fields[2]  \n",
    "                }\n",
    "                sequence = []\n",
    "            else:\n",
    "                header = None\n",
    "        \n",
    "        elif header:\n",
    "            sequence.append(line)\n",
    "\n",
    "    # Add the last record if needed\n",
    "    if header and sequence:\n",
    "        gene_symbol = gene_map.get(header[\"gene_id\"], \"Unknown\")\n",
    "        isoform_type = \"Canonical\" if \"-1\" in header[\"protein_id\"] else \"Alternative Isoform\"\n",
    "        records.append([header[\"protein_id\"], header[\"transcript_id\"], header[\"gene_id\"], gene_symbol, isoform_type, \"\".join(sequence)])\n",
    "\n",
    "# Step 3: Convert to pandas DataFrame and Save to TSV\n",
    "df = pd.DataFrame(records, columns=[\"Ensembl Protein ID\", \"Ensembl Transcript ID\", \"Ensembl Gene ID\", \"Gene Symbol\", \"Isoform Type\", \"FASTA Sequence\"])\n",
    "\n",
    "# Save to TSV\n",
    "df.to_csv(\"data/gencode_protein_isoforms_with_symbols.tsv\", sep=\"\\t\", index=False)\n",
    "\n",
    "# Print completion message\n",
    "print(f\"✅ Extracted {len(df)} protein sequences with Gene Symbols and saved to 'gencode_protein_isoforms_with_symbols.tsv'.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e721088-e99d-4909-b523-8f334f3c8e5e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f343d57b-9a88-4ade-89f2-9cdf0ce2ef74",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df = gene_pair0[[\"Human LR Pair\", \"Ligand\", \"Receptor\"]]\n",
    "lim_df = lim_df.merge(df, how='left', left_on='Ligand', right_on='Gene Symbol')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e760dccb-b0f6-41e5-a8fc-554eb1909cd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dc1f428-00f5-4f40-8906-499ac5d8a459",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df = lim_df.drop(columns=[\"Gene Symbol\"])\n",
    "lim_df = lim_df.rename(columns={\"FASTA Sequence\": \"Ligand Sequence\",\n",
    "                                \"Ensembl Protein ID\": \"Ligand Ensembl Protein ID\",\n",
    "                                \"Ensembl Transcript ID\": \"Ligand Ensembl Transcript ID\",\n",
    "                                \"Ensembl Gene ID\": \"Ligand Ensembl Gene ID\",\n",
    "                                \"Isoform Type\": \"Ligand Isoform Type\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bc88ddd-5476-4f3d-8fa8-815c0827f101",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13dd1035-ecfd-462c-a6d5-61b59da565e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df = lim_df.merge(df, how='left', left_on='Receptor', right_on='Gene Symbol')\n",
    "lim_df = lim_df.drop(columns=[\"Gene Symbol\"])\n",
    "lim_df = lim_df.rename(columns={\"FASTA Sequence\": \"Receptor Sequence\",\n",
    "                                \"Ensembl Protein ID\": \"Receptor Ensembl Protein ID\",\n",
    "                                \"Ensembl Transcript ID\": \"Receptor Ensembl Transcript ID\",\n",
    "                                \"Ensembl Gene ID\": \"Receptor Ensembl Gene ID\",\n",
    "                                \"Isoform Type\": \"Receptor Isoform Type\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7028f462-74cc-40d7-a7d3-003689e862f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf876e97-f9fa-4ca2-bf21-f82861e132ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "lim_df.to_csv(\"data/LRpair_gencode_sequences.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c995394a-d081-4971-be36-38240103bc32",
   "metadata": {},
   "source": [
    "####################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2253500-50d6-4759-9ca2-c282eb8e2fc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to scrape data from Pubmed for Title, Abstract, Journal, and Year\n",
    "### IMPORTANT: TURN OFF VPN and make sure you have the data directory (from Sakura)\n",
    "\n",
    "import sys\n",
    "import requests\n",
    "import pandas as pd\n",
    "import time\n",
    "import os\n",
    "import xml.etree.ElementTree as ET\n",
    "\n",
    "sys.path.append(os.path.abspath(\"src\"))  \n",
    "import fetchGSheet\n",
    "\n",
    "# Read the API key from a file\n",
    "with open(\"data/ncbi_api_key.txt\", \"r\") as file:\n",
    "    ncbi_api_key = file.read().strip()\n",
    "\n",
    "# File to save the results\n",
    "output_file = \"data/pubmed_results.csv\"\n",
    "\n",
    "# Load your list of PMIDs\n",
    "pmid_list = source\n",
    "\n",
    "# Example of fetching HGNC gene symbols (you should have the `fetchGSheet.pop_up_info` dataframe ready)\n",
    "def extract_hgnc_symbols(fetchGSheet):\n",
    "    # Concatenate Approved, Alias, and Previous symbols, then extract unique symbols\n",
    "    hgnc_symbols = pd.concat([\n",
    "        fetchGSheet['Approved symbol'],\n",
    "        fetchGSheet['Alias symbol'],\n",
    "        fetchGSheet['Previous symbol']\n",
    "    ], axis=0).dropna().str.upper().unique()  # Remove NaNs and make uppercase for matching\n",
    "     # Remove any empty strings from the list\n",
    "    hgnc_symbols = [symbol for symbol in hgnc_symbols if symbol != \"\"]\n",
    "    return set(hgnc_symbols)  # Return as a set for fast lookup\n",
    "    \n",
    "hgnc_symbols = extract_hgnc_symbols(fetchGSheet.pop_up_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00aa64ed-aa94-4a51-8ffb-cafadb2623dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example of fetching HGNC gene symbols (you should have the `fetchGSheet.pop_up_info` dataframe ready)\n",
    "def extract_hgnc_symbols(fetchGSheet):\n",
    "    # Concatenate Approved, Alias, and Previous symbols, then extract unique symbols\n",
    "    hgnc_symbols = pd.concat([\n",
    "        fetchGSheet['Approved symbol'],\n",
    "        fetchGSheet['Alias symbol'],\n",
    "        fetchGSheet['Previous symbol']\n",
    "    ], axis=0).dropna().str.upper().unique()  # Remove NaNs and make uppercase for matching\n",
    "     # Remove any empty strings from the list\n",
    "    hgnc_symbols = [symbol for symbol in hgnc_symbols if symbol != \"\"]\n",
    "    return set(hgnc_symbols)  # Return as a set for fast lookup\n",
    "hgnc_symbols = extract_hgnc_symbols(fetchGSheet.pop_up_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d48fc01-b7e3-4698-b8e8-a287dc069243",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Official species names and their corresponding terms (scientific names)\n",
    "species_dict = {\n",
    "    \"human\": \"Homo sapiens\",\n",
    "    \"mouse\": \"Mus musculus\",\n",
    "    \"rat\": \"Rattus norvegicus\",\n",
    "    \"rabbit\": \"Oryctolagus cuniculus\",\n",
    "    \"monkey\": \"Macaca spp.\",\n",
    "    \"dog\": \"Canis lupus familiaris\",\n",
    "    \"pig\": \"Sus scrofa\",\n",
    "    \"zebra fish\": \"Danio rerio\",\n",
    "    \"chicken\": \"Gallus gallus\",\n",
    "    \"horse\": \"Equus ferus caballus\",\n",
    "    \"cat\": \"Felis catus\",\n",
    "    \"sheep\": \"Ovis aries\",\n",
    "    \"cow\": \"Bos taurus\",\n",
    "    \"fruit fly\": \"Drosophila melanogaster\",\n",
    "    \"c. elegans\": \"Caenorhabditis elegans\",\n",
    "}\n",
    "\n",
    "def fetch_pubmed_data(pmid_list, hgnc_symbols):\n",
    "    base_url = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\"\n",
    "    results = []\n",
    "\n",
    "    # Load existing data if output file exists\n",
    "    if os.path.exists(output_file):\n",
    "        existing_data = pd.read_csv(output_file)\n",
    "    else:\n",
    "        existing_data = pd.DataFrame(columns=[\"PMID\", \"Title\", \"Abstract\", \"Journal\", \"Year\", \"Species\"])\n",
    "\n",
    "    # Split PMIDs into batches\n",
    "    batch_size = 50\n",
    "    pmid_batches = [pmid_list[i:i + batch_size] for i in range(0, len(pmid_list), batch_size)]\n",
    "\n",
    "    # Iterate over the batches\n",
    "    for batch in pmid_batches:\n",
    "        params = {\n",
    "            \"db\": \"pubmed\",\n",
    "            \"id\": \",\".join(batch),  # Join PMIDs as comma-separated\n",
    "            \"retmode\": \"xml\",\n",
    "            \"api_key\": ncbi_api_key\n",
    "        }\n",
    "\n",
    "        try:\n",
    "            response = requests.get(base_url, params=params)\n",
    "            response.raise_for_status()\n",
    "\n",
    "            # Parse the XML response\n",
    "            root = ET.fromstring(response.text)\n",
    "            for article in root.findall(\".//PubmedArticle\"):\n",
    "                # Extract Title and Abstract\n",
    "                title = article.findtext(\".//ArticleTitle\", default=\"N/A\")\n",
    "                abstract = article.findtext(\".//AbstractText\", default=\"No abstract available\")\n",
    "\n",
    "                # Extract Journal Title\n",
    "                journal_tag = article.find(\".//Journal/Title\")\n",
    "                journal = journal_tag.text.strip() if journal_tag is not None and journal_tag.text else \"N/A\"\n",
    "\n",
    "                # Extract Publication Year\n",
    "                pub_date = article.find(\".//PubDate\")\n",
    "                if pub_date is not None:\n",
    "                    year_tag = pub_date.find(\"Year\")\n",
    "                    year = year_tag.text if year_tag is not None else \"N/A\"\n",
    "\n",
    "                    # Fallback to MedlineDate if Year is missing\n",
    "                    if year == \"N/A\":\n",
    "                        medline_date_tag = pub_date.find(\"MedlineDate\")\n",
    "                        year = medline_date_tag.text.split()[0] if medline_date_tag is not None else \"N/A\"\n",
    "                else:\n",
    "                    year = \"N/A\"  # PubDate is completely missing\n",
    "\n",
    "                # Initialize species as N/A\n",
    "                species = \"N/A\"\n",
    "\n",
    "                # Check if the word \"patient\" is detected in title or abstract (assume human)\n",
    "                if \"patient\" in title.lower() or \"patient\" in abstract.lower():\n",
    "                    species = \"Homo sapiens\"\n",
    "                elif \"human\" in title.lower() or \"human\" in abstract.lower():\n",
    "                    species = \"Homo sapiens\"\n",
    "                else:\n",
    "                    # Look for HGNC gene symbols in title or abstract (assume human if found)\n",
    "                    for gene in hgnc_symbols:\n",
    "                        if gene in title or gene in abstract:\n",
    "                            species = \"Homo sapiens\"\n",
    "                            break\n",
    "                    else:\n",
    "                        # Look for MeSH terms related to species\n",
    "                        for mesh_heading in article.findall(\".//MeshHeadingList/MeshHeading\"):\n",
    "                            descriptor_name = mesh_heading.findtext(\"DescriptorName\")\n",
    "                            if descriptor_name:\n",
    "                                # Match official species names using the species_dict\n",
    "                                for species_term, scientific_name in species_dict.items():\n",
    "                                    if species_term in descriptor_name.lower():\n",
    "                                        species = scientific_name\n",
    "                                        break  # Stop after finding the first match\n",
    "\n",
    "                # Append the result\n",
    "                results.append({\n",
    "                    \"PMID\": article.findtext(\".//MedlineCitation/PMID\"),\n",
    "                    \"Title\": title,\n",
    "                    \"Abstract\": abstract,\n",
    "                    \"Journal\": journal,\n",
    "                    \"Year\": year,\n",
    "                    \"Species\": species\n",
    "                })\n",
    "\n",
    "        except Exception as e:\n",
    "            print(f\"Error fetching batch {batch}: {e}\")\n",
    "            # Optionally save the response for debugging\n",
    "            with open(f\"error_batch_{batch[0]}_{batch[-1]}.xml\", \"w\") as f:\n",
    "                f.write(response.text)\n",
    "\n",
    "        # Rate limiting to avoid API overload\n",
    "        time.sleep(1)  # Increase delay for better API compliance\n",
    "\n",
    "    # Save results\n",
    "    new_data = pd.DataFrame(results)\n",
    "    if not new_data.empty:\n",
    "        # Merge existing and new data, updating missing values\n",
    "        updated_data = pd.concat([existing_data, new_data])\n",
    "\n",
    "        # Ensure all PMIDs are strings\n",
    "        updated_data[\"PMID\"] = updated_data[\"PMID\"].astype(str)\n",
    "\n",
    "        # Drop rows with missing PMIDs\n",
    "        updated_data = updated_data.dropna(subset=[\"PMID\"])\n",
    "\n",
    "        # Ensure rows are ordered and remove duplicates\n",
    "        updated_data = (\n",
    "            updated_data.sort_values(by=\"PMID\")  # Ensure rows are ordered\n",
    "            .drop_duplicates(subset=\"PMID\", keep=\"last\")  # Keep the latest data\n",
    "        )\n",
    "        updated_data[\"Journal\"] = updated_data[\"Journal\"].str.split(\" (\", n=1, expand=False, regex=False).str[0]\n",
    "        updated_data.to_csv(output_file, index=False)\n",
    "    else:\n",
    "        print(\"No new data fetched.\")\n",
    "\n",
    "    return results\n",
    "\n",
    "# Fetch PubMed data with your list of PMIDs, output file path, and NCBI API key\n",
    "fetch_pubmed_data(pmid_list, hgnc_symbols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25f83505-39a9-42ef-8878-89f5044b1ad8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "quarto_llm_env",
   "language": "python",
   "name": "quarto_llm_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
