{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fcc5ddc1-519e-439c-a130-3dd8bcd181bd",
   "metadata": {},
   "source": [
    "# Python Notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "d4b960ac-0404-4f97-b786-01c850c43e7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to prepare datatables (cleaning and hyperlinking, adding tool tips, etc) input for the database qmds\n",
    "import sys, os\n",
    "import pandas as pd\n",
    "# Change working directory to ConnectomeDB\n",
    "project_root = \"/Users/sakuramaezono/Library/CloudStorage/OneDrive-YokohamaCityUniversity/Personal/05_Python_repositories/ConnectomeDB\"\n",
    "os.chdir(project_root)\n",
    "sys.path.append(os.path.abspath(\"src\"))\n",
    "\n",
    "import fetchGSheet "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "15bad265-2ab8-4069-82d3-b53f44dfd31e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_1532/2847567185.py:29: DtypeWarning: Columns (32,34,38,40,50) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  pop_up_info = pd.read_table(\"data/HGNC_gene_info_full.tsv\")\n"
     ]
    }
   ],
   "source": [
    "## Function to prepare datatables (cleaning and hyperlinking, adding tool tips, etc) input for the database qmds\n",
    "import sys, os\n",
    "from itables import init_notebook_mode\n",
    "import pandas as pd\n",
    "from itables import show\n",
    "from itables import options\n",
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "import fetchGSheet \n",
    "import warnings\n",
    "import urllib.parse\n",
    "\n",
    "# Suppress SettingWithCopyWarning\n",
    "warnings.simplefilter(\"ignore\", category=UserWarning)\n",
    "\n",
    "#https://comp.med.yokohama-cu.ac.jp/reviewer/connectomedb/\n",
    "site_url = \"https://connectomedb.org/\"\n",
    "# Other vertebrates\n",
    "species_list = [\n",
    "    \"mmusculus\", \"rnorvegicus\", \"drerio\", \"ptroglodytes\", \"ggallus\", \"sscrofa\", \"btaurus\", \n",
    "    \"clfamiliaris\", \"ecaballus\", \"oarambouillet\",\n",
    "    \"cjacchus\", \"mmulatta\", \"xtropicalis\"\n",
    "]\n",
    "\n",
    "# Select only the relevant columns from pop_up_info\n",
    "cols_to_keep = cols_to_keep = list(range(0, 30)) \n",
    "# Step 3: Load file using only the desired columns\n",
    "df = pd.read_table(\"data/HGNC_gene_info_full.tsv\", usecols=cols_to_keep)\n",
    "pop_up_info = pd.read_table(\"data/HGNC_gene_info_full.tsv\")\n",
    "pop_up_info = pop_up_info.rename(columns={\"hgnc_id\": \"HGNC ID\", \n",
    "                                          \"name\": \"Approved name\",\n",
    "                                          \"symbol\": \"Approved symbol\",\n",
    "                                          #\"rgd_id\": \"RGD ID\",\n",
    "                                          #\"mgd_id\": \"MGI ID\", \n",
    "                                          \"alias_symbol\": \"Alias symbol\", # add to table\n",
    "                                          \"prev_symbol\": \"Previous symbol\", # add to table\n",
    "                                          \"date_symbol_changed\": \"Date symbol changed\"\n",
    "                                         })\n",
    "\n",
    "# Keep only first MGI/RGD ID\n",
    "#pop_up_info[\"MGI ID\"] = pop_up_info[\"MGI ID\"].str.split(\"|\").str[0]\n",
    "#pop_up_info[\"RGD ID\"] = pop_up_info[\"RGD ID\"].str.split(\"|\").str[0]\n",
    "\n",
    "pop_up_info[\"Alias symbol\"] = pop_up_info[\"Alias symbol\"].apply(\n",
    "    lambda x: \"N/A\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\"] else x\n",
    ")\n",
    "\n",
    "pop_up_info[\"Previous symbol\"] = pop_up_info[\"Previous symbol\"].apply(\n",
    "    lambda x: \"N/A\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\"] else x\n",
    ")\n",
    "\n",
    "# Replace \"|\" with \", \"\n",
    "pop_up_info[\"Alias symbol\"] = [value.replace(\"|\", \", \") for value in pop_up_info[\"Alias symbol\"]]\n",
    "pop_up_info[\"Previous symbol\"] = [value.replace(\"|\", \", \") for value in pop_up_info[\"Previous symbol\"]]\n",
    "\n",
    "pop_up_info[\"Date symbol changed\"] = pop_up_info[\"Date symbol changed\"].apply(\n",
    "    lambda x: \"N/A\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\"] else x\n",
    ")\n",
    "\n",
    "\n",
    "pop_up_info_lim = pop_up_info[[\"HGNC ID\", \"Approved name\", \"Alias symbol\", # \"MGI ID\", \"RGD ID\"\n",
    "                               \"Approved symbol\", \"Previous symbol\"]] # rm \"Approved symbol\" for now\n",
    "pop_up_info_lim = pop_up_info_lim.drop_duplicates(subset=\"HGNC ID\", keep=\"first\")\n",
    "\n",
    "# Drop columns where all values are NA in gene_pair\n",
    "gene_pair = fetchGSheet.gene_pair_human.dropna(axis=1, how='all')\n",
    "gene_pair = gene_pair[gene_pair['LR_pair_orig'] != '']\n",
    "# for now set source count as triplicates\n",
    "sourceCount = len(gene_pair[['LR_pair_orig']])\n",
    "\n",
    "### KEEP ALL AS OF LATEST input datatable\n",
    "# for now, keep only the following columns\n",
    "# gene_pair = gene_pair[['LR pair', 'Ligand', 'Ligand.HGNC', 'Receptor', 'Receptor.HGNC',\n",
    "#                        'perplexity link', 'PMID', 'binding location', \n",
    "#                        'bind in trans?', 'bidirectional signalling?',\n",
    "#                        'interaction type', 'original source']]\n",
    "\n",
    "gene_pair = gene_pair.dropna(subset=['LR_pair_orig'])\n",
    "\n",
    "# some PMIDs kick in with \",\" so replace\n",
    "gene_pair[\"PMID\"] = [value.replace(\",\", \"\") for value in gene_pair[\"PMID\"]]\n",
    "gene_pair = gene_pair.dropna(subset=['PMID'])\n",
    "\n",
    "### NO NEED FOR MAPPING AS OF LATEST input datatable\n",
    "# Mapping for replacements\n",
    "# mapping = dict(zip(fetchGSheet.src_info['original source'], fetchGSheet.src_info['shortname']))\n",
    "# # Replace values in the column based on the mapping\n",
    "# gene_pair['original source'] = gene_pair['original source'].replace(mapping)\n",
    "\n",
    "gene_pair.columns = gene_pair.columns.str.strip()\n",
    "gene_pair[['Ligand', 'Receptor']] = gene_pair['LR Pair Card'].str.split(' ', n=1, expand=True)\n",
    "\n",
    "## add Ligand/Receptor Location\n",
    "def dedup_locations(loc_str):\n",
    "    # Split, strip, deduplicate, and sort\n",
    "    parts = [loc.strip() for loc in loc_str.split(',') if loc.strip()]\n",
    "    unique_sorted = sorted(set(parts), key=str.lower)  # case-insensitive sort\n",
    "    return unique_sorted\n",
    "\n",
    "def generate_LocToolTip(row, geneloc, loc_col):\n",
    "    ligand = row[loc_col]\n",
    "    original_locations = [loc.strip() for loc in row[\"location\"].split(',')]\n",
    "    original_sources = [src.strip() for src in row[\"source\"].split(',')]\n",
    "\n",
    "    # Get deduplicated locations\n",
    "    unique_locations = dedup_locations(row[\"location\"])\n",
    "\n",
    "    if len(unique_locations) == 1:\n",
    "        # Single tooltip case\n",
    "        location = unique_locations[0]\n",
    "        matching_rows = geneloc[(geneloc[loc_col] == ligand) & (geneloc[\"location\"].str.contains(location))]\n",
    "        all_sources = matching_rows[\"source\"].unique()\n",
    "        sources_str = \", \".join(sorted(set(all_sources)))\n",
    "        return f'<span title=\"based on {sources_str}\">{location}</span>'\n",
    "    else:\n",
    "        # Multiple tooltips — find each (ligand, location) match in original df\n",
    "        spans = []\n",
    "        for loc in unique_locations:\n",
    "            matching_rows = geneloc[\n",
    "                (geneloc[loc_col] == ligand) &\n",
    "                (geneloc[\"location\"].str.contains(loc))\n",
    "            ]\n",
    "            all_sources = matching_rows[\"source\"].unique()\n",
    "            sources_str = \", \".join(sorted(set(all_sources)))\n",
    "            spans.append(f'<span title=\"based on {sources_str}\">{loc}</span>')\n",
    "        return \", \".join(spans)\n",
    "\n",
    "\n",
    "# Group the original loc_info by Ligand\n",
    "ligand_loc = fetchGSheet.ligand_loc.dropna(axis=1, how='all')\n",
    "grouped = ligand_loc.groupby(\"Ligand\").agg({\n",
    "    \"location\": lambda x: ', '.join(x),\n",
    "    \"source\": lambda x: ', '.join(x)\n",
    "}).reset_index()\n",
    "\n",
    "# Generate tooltips\n",
    "grouped[\"Ligand location\"] = grouped.apply(lambda row: generate_LocToolTip(row, ligand_loc,loc_col=\"Ligand\"), axis=1)\n",
    "# create dict\n",
    "mapping_loc = dict(zip(grouped['Ligand'], grouped['Ligand location'])) \n",
    "gene_pair['Ligand location'] = gene_pair['Ligand'].replace(mapping_loc)\n",
    "\n",
    "\n",
    "# Group the original loc_info by Receptor\n",
    "receptor_loc = fetchGSheet.receptor_loc.dropna(axis=1, how='all')\n",
    "grouped = receptor_loc.groupby(\"Receptor\").agg({\n",
    "    \"location\": lambda x: ', '.join(x),\n",
    "    \"source\": lambda x: ', '.join(x)\n",
    "}).reset_index()\n",
    "\n",
    "# Generate tooltips\n",
    "grouped[\"Receptor location\"] = grouped.apply(lambda row: generate_LocToolTip(row, receptor_loc,loc_col=\"Receptor\"), axis=1)\n",
    "# create dict\n",
    "mapping_loc = dict(zip(grouped['Receptor'], grouped['Receptor location'])) \n",
    "gene_pair['Receptor location'] = gene_pair['Receptor'].replace(mapping_loc)\n",
    "\n",
    "\n",
    "# Set missing mappings to 'unknown'\n",
    "gene_pair.loc[gene_pair['Ligand location'] == gene_pair['Ligand'], 'Ligand location'] = 'unknown'\n",
    "gene_pair.loc[gene_pair['Receptor location'] == gene_pair['Receptor'], 'Receptor location'] = 'unknown'\n",
    "# Set \"n/a\" to unknown\n",
    "gene_pair['Ligand location'] = [value.replace(\"n/a\", \"unknown\") for value in gene_pair['Ligand location']]\n",
    "gene_pair['Receptor location'] = [value.replace(\"n/a\", \"unknown\") for value in gene_pair['Receptor location']]\n",
    "\n",
    "# Fetch HGNC IDs from the dataset\n",
    "hgnc_id = [col for col in gene_pair.columns if \"HGNC\" in col]\n",
    "hgnc_id = pd.concat([gene_pair[col] for col in hgnc_id]).unique()\n",
    "\n",
    "gene_pair['Human LR Pair'] = np.where(\n",
    "    gene_pair['Human evidence'] == \"not conserved\", \n",
    "    \"no human ortholog\",                                  \n",
    "    gene_pair['Homo sapiens_ligand'] + \" \" + gene_pair['Homo sapiens_receptor'] \n",
    ")\n",
    "\n",
    "\n",
    "# Rename columns for better clarity\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "    \"LR_pair_orig\": \"LR Pair\",\n",
    "    \"HGNC ligand\": \"Ligand HGNC ID\",\n",
    "    \"HGNC receptor\": \"Receptor HGNC ID\",\n",
    "    \"ENSEMBL ligand\": \"Ligand ENSEMBL ID\",\n",
    "    \"ENSEMBL receptor\": \"Receptor ENSEMBL ID\",\n",
    "    # \"perplexity link\": \"Perplexity\", # will be replaced with actual link later\n",
    "    # \"original source\": \"Database Source\",\n",
    "    \"Ligand location\": \"Ligand Location\",\n",
    "    \"Receptor location\": \"Receptor Location\",\n",
    "    # \"binding location\": \"Binding Location\",\n",
    "    # \"bind in trans?\" : \"Trans-binding\", \n",
    "    # \"bidirectional signalling?\": \"Bidirectional Signalling\",\n",
    "    # \"interaction type\" : \"Interaction Type\"\n",
    "    #\"PMID\": \"PMID support\" # was PMID support\n",
    "})\n",
    "gene_pair = gene_pair.drop(columns=[\"Ligand\", \"Receptor\"])\n",
    "gene_pair = gene_pair.rename(columns={\"Homo sapiens_ligand\": \"Ligand\", \"Homo sapiens_receptor\": \"Receptor\"})\n",
    "# Merge gene_pair with pop_up_info_lim for Ligand(L)\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Ligand HGNC ID', right_on='HGNC ID')\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Ligand Name\", \n",
    "                                     #\"MGI ID\": \"Ligand MGI ID\", # NOT APPLIED YET BUT should be taken from Ensembl BioMart\n",
    "                                     #\"RGD ID\": \"Ligand RGD ID\", # NOT APPLIED YET BUT should be taken from Ensembl BioMart\n",
    "                                      \"Alias symbol\": \"Ligand Aliases\",\n",
    "                                      \"Previous symbol\": \"Ligand Old symbol\",\n",
    "                                     },\n",
    "                            )\n",
    "gene_pair = gene_pair.drop(columns=[\"HGNC ID\", \"Approved symbol\"])\n",
    "\n",
    "\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Receptor HGNC ID', right_on='HGNC ID')\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Receptor Name\",\n",
    "             #\"MGI ID\": \"Receptor MGI ID\",\n",
    "             #\"RGD ID\": \"Receptor RGD ID\",\n",
    "                                      \"Alias symbol\": \"Receptor Aliases\",\n",
    "                                      \"Previous symbol\": \"Receptor Old symbol\",}\n",
    "                            )\n",
    "\n",
    "### AS OF LATEST DB skip pathways for now (code in addPathwayDiseaseAnnot_temp.py)\n",
    "\n",
    "# Add new columns where all Ligand Symbol & Aliases and Receptor Symbol & Aliases merged in one column\n",
    "def format_symbol_aliases(symbol, old_symbol, aliases):\n",
    "    \"\"\"\n",
    "    Formats symbol, old symbols, and aliases.\n",
    "    If the final formatted string would be empty after considering N/A values\n",
    "    and empty inputs, it returns \"mouse-specific\".\n",
    "    Otherwise, it formats based on the presence of old_symbol and aliases,\n",
    "    removing unnecessary parentheses or commas, following the structure:\n",
    "    \"Symbol (Old Symbol, Aliases)\" if both exist.\n",
    "    \"\"\"\n",
    "    # Normalize inputs to empty strings if they are None/NaN or just whitespace\n",
    "    symbol_str = str(symbol).strip()\n",
    "    old_symbol_str = str(old_symbol).strip()\n",
    "    aliases_str = str(aliases).strip()\n",
    "\n",
    "    # Filter out values that are empty strings or \"N/A\" for old_symbol and aliases\n",
    "    parts_for_join = []\n",
    "    if old_symbol_str and old_symbol_str != \"N/A\":\n",
    "        parts_for_join.append(old_symbol_str)\n",
    "    if aliases_str and aliases_str != \"N/A\":\n",
    "        parts_for_join.append(aliases_str)\n",
    "\n",
    "    # Construct the preliminary result based on your original logic:\n",
    "    # \"symbol (old_symbol, aliases)\" if parts_for_join is not empty, else \"symbol\"\n",
    "    if parts_for_join:\n",
    "        prelim_result = f\"{symbol_str} ({', '.join(parts_for_join)})\"\n",
    "    else:\n",
    "        prelim_result = symbol_str # Just the symbol if no old_symbol or aliases\n",
    "\n",
    "    return prelim_result\n",
    "\n",
    "# This is crucial for consistent handling by the function before processing \"N/A\".\n",
    "gene_pair['Ligand'] = gene_pair['Ligand'].fillna('')\n",
    "gene_pair['Ligand Old symbol'] = gene_pair['Ligand Old symbol'].fillna('')\n",
    "gene_pair['Ligand Aliases'] = gene_pair['Ligand Aliases'].fillna('')\n",
    "\n",
    "\n",
    "# to later check which ligand-receptor pairs are non-human\n",
    "def is_mouse_specific(name):\n",
    "    if not isinstance(name, str):\n",
    "        return False\n",
    "    name = name.strip()  # remove leading/trailing spaces\n",
    "    return any(c.islower() for c in name[1:])\n",
    "\n",
    "gene_pair = gene_pair.drop(columns=[\"HGNC ID\"])\n",
    "\n",
    "gene_pair['Ligand Symbols'] = gene_pair.apply(\n",
    "    lambda row: \"no human ortholog\" if is_mouse_specific(row['Ligand']) \n",
    "               else format_symbol_aliases(row['Ligand'], row['Ligand Old symbol'], row['Ligand Aliases']),\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "\n",
    "# This is crucial for consistent handling by the function before processing \"N/A\".\n",
    "gene_pair['Receptor'] = gene_pair['Receptor'].fillna('')\n",
    "gene_pair['Receptor Old symbol'] = gene_pair['Receptor Old symbol'].fillna('')\n",
    "gene_pair['Receptor Aliases'] = gene_pair['Receptor Aliases'].fillna('')\n",
    "\n",
    "gene_pair['Receptor Symbols'] = gene_pair.apply(\n",
    "    lambda row: \"no human ortholog\" if is_mouse_specific(row['Receptor']) \n",
    "                else format_symbol_aliases(row['Receptor'], row['Receptor Old symbol'], row['Receptor Aliases']),\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "### tooltips \n",
    "gene_pair[\"Ligand Symbols\"] = [\n",
    "    f'<span title=\"{aliases}\">{aliases}</span>'\n",
    "    for aliases in gene_pair[\"Ligand Symbols\"]\n",
    "]\n",
    "gene_pair[\"Receptor Symbols\"] = [\n",
    "    f'<span title=\"{aliases}\">{aliases}</span>'\n",
    "    for aliases in gene_pair[\"Receptor Symbols\"]\n",
    "]\n",
    "\n",
    "# might be used later just save info for now (for mouse cards)\n",
    "grab_mouse_info = gene_pair[\"LR Pair Card\"][gene_pair[\"Human evidence\"].isin([\"absent in human\", \"not conserved\"])]\n",
    "grab_mouse_info = grab_mouse_info.unique()\n",
    "grab_mouse_info\n",
    "\n",
    "# Add an empty Perplexity column filled with None (just to save it's order\n",
    "gene_pair['Perplexity'] = None\n",
    "#gene_pair = gene_pair.drop(columns=[\"Approved symbol_x\", \"Approved symbol_y\"])\n",
    "\n",
    "### For latest DB, skip (code saved as addOrth_temp.py)\n",
    "\n",
    "# Add\n",
    "first_columns=['LR Pair Card', 'Human LR Pair', 'Ligand', 'Receptor', 'Ligand Symbols', 'Receptor Symbols', 'Ligand Location', 'Receptor Location',\t'Ligand HGNC ID', 'Receptor HGNC ID', 'Perplexity', 'Human evidence'] # 'Database Source'\n",
    "\n",
    "end_columns=['PMID', 'Pair_species', 'lig_species', 'rec_species', 'ligand_orig', 'receptor_orig']\n",
    "gene_pair = gene_pair[first_columns + [col for col in gene_pair.columns if col not in first_columns + end_columns] + end_columns]\n",
    "# gene_pair = gene_pair[first_columns + [col for col in gene_pair.columns if col not in first_columns]]\n",
    "\n",
    "# number of unique vars (Human and Mouse both counted)\n",
    "\n",
    "lrPairsCount = len(gene_pair[\"LR Pair Card\"].unique())\n",
    "\n",
    "ligandCount = len(gene_pair[\"Ligand\"].unique())\n",
    "\n",
    "receptorCount = len(gene_pair[\"Receptor\"].unique())\n",
    "\n",
    "\n",
    "### Remove from here for latest DB\n",
    "# # Mouse Orthologue\n",
    "# MouseLigandCount = len(gene_pair[\"Ligand MGI ID\"].unique())\n",
    "\n",
    "# MouseReceptorCount = len(gene_pair[\"Receptor MGI ID\"].unique())\n",
    "\n",
    "# # Rat Orthologue\n",
    "# RatLigandCount = len(gene_pair[\"Ligand RGD ID\"].unique())\n",
    "\n",
    "# RatReceptorCount = len(gene_pair[\"Receptor RGD ID\"].unique())\n",
    "\n",
    "# gene_pair[\"PMID\"] = [value.replace(\" \", \"\") for value in gene_pair[\"PMID\"]] # was'PMID support'\n",
    "\n",
    "\n",
    "source = np.array(gene_pair[\"PMID\"].unique())\n",
    "source = source.astype(str)\n",
    "source = \",\".join(sorted(set(filter(lambda x: x.lower() != 'nan', source))))\n",
    "# Split the string into individual elements, filter out empty strings, and get unique values\n",
    "source = sorted(\n",
    "    set(filter(lambda x: x.strip() and x.strip().lower() != 'nan', source.split(',')))\n",
    ")\n",
    "source = [value.replace(\" \", \"\") for value in source]\n",
    "\n",
    "# Function to join unique sorted values\n",
    "agg_func = lambda x: ', '.join(sorted(set(map(str, x))))\n",
    "\n",
    "# Group and aggregate all columns except 'LR Pair Card'\n",
    "gene_pair = gene_pair.groupby('LR Pair Card').agg(agg_func).reset_index()\n",
    "gene_pair = gene_pair[gene_pair['LR Pair Card'] != '']\n",
    "# Identify rows where BOTH 'Ligand HGNC ID' and 'Receptor HGNC ID' are empty\n",
    "# We check if the stripped string is empty, as fillna('') converts None/NaN to empty strings\n",
    "has_hgnc_id = (gene_pair['Ligand HGNC ID'].astype(str).str.strip() != '') | \\\n",
    "              (gene_pair['Receptor HGNC ID'].astype(str).str.strip() != '')\n",
    "\n",
    "# Separate the DataFrame into two parts (human-based cards and mouse based)\n",
    "human_rows = gene_pair[~(gene_pair[\"Human evidence\"].isin([\"absent in human\", \"not conserved\"]))]\n",
    "mouse_rows = gene_pair[gene_pair[\"Human evidence\"].isin([\"absent in human\", \"not conserved\"])]\n",
    "\n",
    "# Concatenate the DataFrames: rows with IDs first, then rows without IDs\n",
    "gene_pair = pd.concat([human_rows, mouse_rows]).reset_index(drop=True)\n",
    "DBlength = len(gene_pair)\n",
    "gene_pair[\"Interaction ID\"] = [f\"CDB{str(i).zfill(5)}\" for i in range(1, DBlength + 1)]\n",
    "\n",
    "# for creating PMIDs\n",
    "gene_pair00 = gene_pair[['LR Pair Card', 'PMID']] # was \"PMID support\"\n",
    "\n",
    "# create Perplexity link\n",
    "def create_url_basic(perplexity_col):\n",
    "    query = f\"What is the primary evidence that {perplexity_col} bind-each-other-as-a-ligand-and-receptor-pair. Exclude reviews, uniprot, wiki, genecards, PIPS, iuphar as sources.\"\n",
    "    encoded_query = query.replace(\" \", \"%20\")\n",
    "    return f\"https://www.perplexity.ai/search?q={encoded_query}\"\n",
    "    \n",
    "# Option 2 -- new query all together\n",
    "\n",
    "# def generate_perplexity_link_pmid(row): \n",
    "#     query = f\"What-is-the-biological-relevance-of-the-ligand-and-receptor-pair-{row['Human LR Pair']}-based-on-Pubmed-ID-{row['PMID']}\"\n",
    "#     return (\n",
    "#          f'<a href=\"https://www.perplexity.ai/search?q={query}\" target=\"_blank\">'\n",
    "#         f'<img src=\"https://img.icons8.com/?size=30&id=0NbBuNOxUwps&format=png&color=000000\" alt=\"Perplexity AI\" /></a>'\n",
    "#     )\n",
    "\n",
    "# cannot use perplexity logo\n",
    "def generate_perplexity_link_pmid(row): \n",
    "    query = f\"What-is-the-biological-relevance-of-the-ligand-and-receptor-pair-{row['Human LR Pair']}-based-on-Pubmed-ID-{row['PMID']}\"\n",
    "    return (\n",
    "         f'<a href=\"https://www.perplexity.ai/search?q={query}\" target=\"_blank\" style=\"text-decoration: none;\">&#128269;</a>'\n",
    "    )\n",
    "\n",
    "# Apply function to the DataFrame\n",
    "gene_pair[\"Perplexity\"] = gene_pair.apply(generate_perplexity_link_pmid, axis=1)\n",
    "\n",
    "# create URLs for the HGNC IDs\n",
    "\n",
    "# ligand\n",
    "gene_pair[\"Ligand HGNC ID\"] = [\n",
    "    '<a href=\"https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/{}\" target=\"_blank\">{}</a>'.format(ligand, ligand)\n",
    "    for ligand in gene_pair[\"Ligand HGNC ID\"]\n",
    "]\n",
    "\n",
    "# receptor\n",
    "gene_pair[\"Receptor HGNC ID\"] = [\n",
    "    '<a href=\"https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/{}\" target=\"_blank\">{}</a>'.format(receptor, receptor)\n",
    "    for receptor in gene_pair[\"Receptor HGNC ID\"]\n",
    "]\n",
    "\n",
    "\n",
    "# Function to generate hyperlinks for the \"PMID support\" column\n",
    "def generate_links_with_doi(df, gene_column, pmid_column, id_column):\n",
    "    def create_link(gene, id_col, sources):\n",
    "        # Replace spaces with \"——\" in the gene name for the link\n",
    "        gene_name = gene.replace(\" \", \"—\")\n",
    "        \n",
    "        if len(sources) == 1:\n",
    "            source = sources[0]\n",
    "            if source.startswith(\"https://www.biorxiv.org/content/\"):\n",
    "                # If the value starts with \"https://doi.org/\", use it as the hyperlink\n",
    "                return f'<a href=\"{source}\" target=\"_blank\">BioRxiv</a>'\n",
    "            else:\n",
    "                # If it's a single PMID, hyperlink the PMID text\n",
    "                return f'<a href=\"{site_url}cards/{gene}.html\">{source}</a>'\n",
    "        else:\n",
    "            # If multiple PMIDs, show the count and hyperlink to the page\n",
    "            return f'<a href=\"https://comp.med.yokohama-cu.ac.jp/reviewer/connectomedb/cards/{gene}.html\" target=\"_blank\">{len(sources)} PMIDs</a>'\n",
    "\n",
    "    # Process each row to generate the \"PMID\" column # was \"PMID support\"\n",
    "    df[\"PMID\"] = [\n",
    "        create_link(\n",
    "            gene=row[gene_column], \n",
    "            id_col = row[id_column],\n",
    "            sources=[s.strip() for s in row[pmid_column].split(',') if s.strip()]\n",
    "        )\n",
    "        for _, row in df.iterrows()\n",
    "    ]\n",
    "    return df\n",
    "\n",
    "\n",
    "# Generate the links for the \"PMID\" column # was \"PMID support\"\n",
    "gene_pair = generate_links_with_doi(gene_pair, gene_column=\"Human LR Pair\", \n",
    "                                    pmid_column=\"PMID\", id_column= \"Interaction ID\")\n",
    "\n",
    "# for disease type, cancer-related and top pathways, when missing say \"ask Perplexity\"\n",
    "\n",
    "\n",
    "def generate_perplexity_kegglinks(\n",
    "    df,\n",
    "    pathway_col=\"KEGG Pathway\",\n",
    "    default_query_template=\"What-biological or other functional-pathways-is-the-ligand-receptor-{pair}-associated-with\"\n",
    "):\n",
    "    def create_link(row):\n",
    "        value = row.get(pathway_col, \"\")\n",
    "        \n",
    "        if pd.isna(value) or str(value).strip().lower() in [\"nan\", \"none\", \"\", \"unknown\"]:\n",
    "            pair = row[\"Human LR Pair\"]\n",
    "            label = \"ask Perplexity\"\n",
    "            query = default_query_template.format(pair=pair)\n",
    "            encoded_query = urllib.parse.quote(query)\n",
    "            return f'<a href=\"https://www.perplexity.ai/search?q={encoded_query}\" target=\"_blank\">{label}</a>'\n",
    "        else:\n",
    "            return value\n",
    "\n",
    "    df[pathway_col] = df.apply(create_link, axis=1)\n",
    "    return df\n",
    "\n",
    "gene_pair = generate_perplexity_kegglinks(gene_pair, pathway_col=\"KEGG Pathway\")\n",
    "    \n",
    "def generate_perplexity_links(df, pathway_col, default_query_template):\n",
    "    def create_link(row):\n",
    "        pathway_value = str(row[pathway_col]).strip().lower()\n",
    "        pair = row[\"Human LR Pair\"]\n",
    "        \n",
    "        if pd.isna(row[pathway_col]) or pathway_value in [\"nan\", \"none\", \"\", \"unknown\"]:\n",
    "            label = \"ask Perplexity\"\n",
    "            query = default_query_template.format(pair=pair)\n",
    "            output =  f'<a href=\"https://www.perplexity.ai/search?q={query}\" target=\"_blank\">{label}</a>'\n",
    "        else:\n",
    "            label = row[pathway_col]\n",
    "            query = f\"What-is-the-role-of-the-ligand-and-receptor-pair-{pair}-in-{label}\"\n",
    "            output = f'{label} (see <a href=\"https://www.perplexity.ai/search?q={query}\" target=\"_blank\">evidence in Perplexity</a>)'\n",
    "        \n",
    "        return output\n",
    "    \n",
    "    df[pathway_col] = df.apply(create_link, axis=1)\n",
    "    return df\n",
    "\n",
    "### SKIP for latest DB\n",
    "\n",
    "# gene_pair = generate_perplexity_links(\n",
    "#     gene_pair,\n",
    "#     pathway_col=\"PROGENy Pathway\",\n",
    "#     default_query_template=\"What-major signalling pathways-is-the-ligand-receptor-pair-{pair}-associated-with\"\n",
    "# )\n",
    "\n",
    "# gene_pair = generate_perplexity_links(\n",
    "#     gene_pair,\n",
    "#     pathway_col=\"Disease Type\",\n",
    "#     default_query_template=\"What-disease types-is-the-ligand-receptor-pair-{pair}-associated-with\"\n",
    "# )\n",
    "# # if it is a yes or no question\n",
    "# def generate_perplexity_links_yesno(\n",
    "#     df,\n",
    "#     pathway_col=\"Cancer-related\",\n",
    "#     default_query_template=\"Is-the-{pair}-associated-with-cancer-and-if-so-which-ones\"\n",
    "# ):\n",
    "#     def create_link(row):\n",
    "#         pathway_value = str(row[pathway_col]).strip().lower()\n",
    "#         pair = row[\"Human LR Pair\"]\n",
    "        \n",
    "#         if pd.isna(row[pathway_col]) or pathway_value in [\"nan\", \"none\", \"\", \"unknown\"]:\n",
    "#             label = \"ask Perplexity\"\n",
    "#             query = default_query_template.format(pair=pair)\n",
    "#             output = f'<a href=\"https://www.perplexity.ai/search?q={query}\" target=\"_blank\">{label}</a>'\n",
    "#         else:\n",
    "#             label = row[pathway_col]\n",
    "#             query = f\"Provide evidence to support the statement-{pair}-is-related-to-cancer-answer-is-{label}\"\n",
    "#             output = f'{label} (see <a href=\"https://www.perplexity.ai/search?q={query}\" target=\"_blank\">evidence in Perplexity</a>)'\n",
    "#         return output\n",
    "    \n",
    "#     df[pathway_col] = df.apply(create_link, axis=1)\n",
    "#     return df\n",
    "\n",
    "# gene_pair = generate_perplexity_links_yesno(\n",
    "#     gene_pair,\n",
    "#     pathway_col=\"Cancer-related\"\n",
    "# )\n",
    "\n",
    "# Add tooltip to name\n",
    "\n",
    "def add_geneToolTip(species):\n",
    "    def tooltip_html(symbol, name):\n",
    "        return (\n",
    "            f'<span class=\"tooltip\">{symbol}'\n",
    "            f'<span class=\"tooltiptext\">{name}</span></span>'\n",
    "        )\n",
    "\n",
    "    gene_pair[species + \" Ligand\"] = [\n",
    "        tooltip_html(ligand_symbol, ligand_name)\n",
    "        for ligand_name, ligand_symbol in zip(gene_pair[species + \" Ligand Name\"], gene_pair[species + \" Ligand\"])\n",
    "    ]\n",
    "    gene_pair[species + \" Receptor\"] = [\n",
    "        tooltip_html(receptor_symbol, receptor_name)\n",
    "        for receptor_name, receptor_symbol in zip(gene_pair[species + \" Receptor Name\"], gene_pair[species + \" Receptor\"])\n",
    "    ]\n",
    "\n",
    "## Make the Human evidence consistent\n",
    "gene_pair[\"Human evidence\"] = gene_pair[\"Human evidence\"].replace(\"CONSERVATION, DIRECT\", \"DIRECT, CONSERVATION\")\n",
    "\n",
    "### Remove tooltip for name for each species for now as only zebrafish has the proper names ###     \n",
    "# speciesPrime_list = [\"Zebrafish\"]\n",
    "# # Loop through each species and update gene_pair\n",
    "# for species in speciesPrime_list:\n",
    "#    gene_pair = add_geneToolTip(species)\n",
    "\n",
    "mouse_columns = ['Mouse Ligand', 'Mouse Receptor','Ligand MGI ID','Receptor MGI ID'] \n",
    "rat_columns = ['Rat Ligand','Rat Receptor','Ligand RGD ID','Receptor RGD ID']\n",
    "zebrafish_columns = ['Zebrafish Ligand','Zebrafish Receptor','Ligand ZFIN ID','Receptor ZFIN ID']\n",
    "\n",
    "# List of prefixes\n",
    "prefixes = (\"Chimpanzee\", \"Chicken\", \"Pig\", \"Cow\", \"Dog\", \"Horse\", \"Sheep\", \"Marmoset\", \"Macaque\", \"Frog\")\n",
    "\n",
    "# Get column names that start with any of the given prefixes\n",
    "selected_columns = [col for col in gene_pair.columns if col.startswith(prefixes)]\n",
    "# was \"PMID support\"\n",
    "gene_pair0 = gene_pair[[\"Interaction ID\"]+ first_columns+[\"PMID\"]+[\"Ligand Name\", \"Receptor Name\"]]\n",
    "# gene_pair0 = gene_pair[[\"Interaction ID\", \"Human LR Pair\", \"Ligand\", \"Receptor\",\n",
    "#                        \"Ligand Symbols\", \"Receptor Symbols\", \n",
    "#                        \"Ligand Location\", \"Receptor Location\",\n",
    "#                        \"Ligand HGNC ID\", \"Receptor HGNC ID\",\n",
    "#                        \"Perplexity\", \"PMID\", 'Ligand Name','Receptor Name', 'KEGG Pathway', 'Cancer-related', 'Disease Type', 'Binding Location', 'Trans-binding', 'Bidirectional Signalling', 'Interaction Type', \"PROGENy Pathway\"] + mouse_columns + rat_columns]\n",
    "\n",
    "gene_pair = gene_pair[[\"Interaction ID\"]+ first_columns+[\"Ligand Name\", \"Receptor Name\"]]\n",
    "\n",
    "# gene_pair = gene_pair[[\"Interaction ID\", \"Human LR Pair\", \"Ligand\", \"Receptor\",\n",
    "#                        \"Ligand Symbols\", \"Receptor Symbols\", \n",
    "#                        \"Ligand Location\", \"Receptor Location\",\n",
    "#                        \"Ligand HGNC ID\", \"Receptor HGNC ID\",\n",
    "#                        \"Perplexity\", \"PMID\", \n",
    "#                        \"Database Source\", \"Binding Location\",\n",
    "#                        \"Trans-binding\", \"Bidirectional Signalling\",\n",
    "#                        \"Interaction Type\",'Ligand Name','Receptor Name'] + mouse_columns + rat_columns + zebrafish_columns + selected_columns]\n",
    "# rm  \"KEGG Pathway\", \"PROGENy Pathway\", \"Cancer-related\", \"Disease Type\" for now\n",
    "\n",
    "# Quick check if there is mouse-specific\n",
    "gene_pair['Ligand'] = gene_pair.apply(\n",
    "    lambda row: \"no human ortholog\" if is_mouse_specific(row['Ligand']) else row['Ligand'],\n",
    "    axis=1\n",
    ")\n",
    "gene_pair['Receptor'] = gene_pair.apply(\n",
    "    lambda row: \"no human ortholog\" if is_mouse_specific(row['Receptor']) else row['Receptor'],\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "# gene symbol\n",
    "gene_pair[\"Ligand\"] = [\n",
    "    f'<span title=\"{ligand_name}\">{ligand_symbol}</span>'\n",
    "    for ligand_name, ligand_symbol in zip(gene_pair[\"Ligand Name\"], \n",
    "                                              gene_pair[\"Ligand\"])\n",
    "]\n",
    "# gene symbol\n",
    "gene_pair[\"Receptor\"] = [\n",
    "    f'<span title=\"{receptor_name}\">{receptor_symbol}</span>'\n",
    "    for receptor_name, receptor_symbol in zip(gene_pair[\"Receptor Name\"], \n",
    "                                              gene_pair[\"Receptor\"])\n",
    "]\n",
    "\n",
    "### SKIP for latest DB\n",
    "# tweak Database Source so when multiple should show multiple with tool tip\n",
    "# gene_pair[\"Database Source\"] = [\n",
    "#     f'<span title=\"{orig_dbSource}\">{(\"multiple (CDB2025 included)\" if \"connectomeDB2025\" in orig_dbSource and \",\" in orig_dbSource else \"multiple\" if \",\" in orig_dbSource else orig_dbSource)}</span>'\n",
    "#     for orig_dbSource in gene_pair[\"Database Source\"]\n",
    "# ]\n",
    "\n",
    "\n",
    "# FOR NOW just make it as simple as em-dash/arrow\n",
    "# def replace_spaces(row):\n",
    "#     if 'secreted' in row['Ligand Location'].lower():\n",
    "#         return row['Human LR Pair'].replace(\" \", \" <span style='font-size: 14px;'>○</span> <span style='font-size: 24px;'>⤚</span> \")\n",
    "#     elif row['Ligand Location'] == 'unknown':\n",
    "#         return row['Human LR Pair'].replace(\" \", \" <span style='font-size: 14px;'>○</span> <span style='font-size: 24px;'>⤚</span> \")\n",
    "#     elif 'membrane' in row['Ligand Location'].lower():\n",
    "#         return row['Human LR Pair'].replace(\" \", \" <span style='font-size: 24px;'>⤙</span> <span style='font-size: 24px;'>⤚</span> \")\n",
    "#     else:\n",
    "#         return row['Human LR Pair'].replace(\" \", \" \\u2192 \")\n",
    "\n",
    "# # Apply the function to the 'LR Pair' column\n",
    "# gene_pair['Human LR Pair'] = gene_pair.apply(replace_spaces, axis=1)\n",
    "# gene_pair[\"Human LR Pair\"] = gene_pair[\"Human LR Pair\"].str.replace(\" \", \"-\")\n",
    "gene_pair = gene_pair.drop(columns=[\"Ligand Name\", \"Receptor Name\"])\n",
    "\n",
    "\n",
    "# Create the links to the HTML cards\n",
    "gene_pair[\"LR Pair Card\"] = [\n",
    "    f'<a href=\"https://comp.med.yokohama-cu.ac.jp/reviewer/connectomedb/cards/{ \"mouse\" if evidence == \"not conserved\" else \"human\" }/{lrPairOrig.replace(\" \",\"-\")}.html\">{lrPair}</a>'\n",
    "    for lrPairOrig, lrPair, evidence in zip(gene_pair0[\"LR Pair Card\"], gene_pair[\"LR Pair Card\"], gene_pair[\"Human evidence\"])\n",
    "]\n",
    "\n",
    "\n",
    "\n",
    "# Add tooltips to the column headers\n",
    "gene_pair.columns = [\n",
    "    f'<span title=\"Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\">{col}</span>' if col == \"Human LR Pair\" else\n",
    "    f'<span title=\"Click the logo below to run Perplexity on the Human LR pair\">{col}&nbsp;</span>' if col == \"Perplexity\" else\n",
    "    f'<span title=\"Official Gene Symbol; Hover on symbols below to show gene names\">{col}&nbsp;&nbsp;&nbsp;</span>' if col in [\"Ligand\", \"Receptor\"] else\n",
    "    f'<span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">{col}&nbsp;&nbsp;</span>' if col in [\"Ligand HGNC ID\", \"Receptor HGNC ID\"] else\n",
    "    f'<span title=\" PubMed IDs (PMID) with Literature Evidence for LR Interaction. Click on the link for more details\">{col}</span>' if col == \"PMID\" else\n",
    "    f'<span title=\"Rat Genome Database (RGD) ID. Click on the link for more details\">{col}</span>' if col in [\"Ligand RGD ID\", \"Receptor RGD ID\"] else\n",
    "    f'<span title=\"Mouse Genome Informatics (MGI) ID. Click on the link for more details\">{col}</span>' if col in [\"Ligand MGI ID\", \"Receptor MGI ID\"]else\n",
    "    f'<span title=\"Zebrafish Information Network (ZFIN) ID. Click on the link for more details\">{col}</span>' if col in [\"Ligand ZFIN ID\", \"Receptor ZFIN ID\"] else\n",
    "    f'<span title=\"Location based on the predicted subcellular localization of the human proteome\">{col}</span>' if col in [\"Ligand Location\", \"Receptor Location\"] else\n",
    "    f'<span title=\"Double-click header of {col} to reverse sort\">{col}&nbsp;</span>'\n",
    "    for col in gene_pair.columns\n",
    "]\n",
    "\n",
    "gene_pair = gene_pair.reset_index(drop=True)  # Remove the index\n",
    "\n",
    "#######################################################################\n",
    "# Identify the column(s) that contain '(PMID)' and temporarily remove for presubmission\n",
    "pmid_cols = [col for col in gene_pair.columns if '(PMID)' in col]\n",
    "gene_pair = gene_pair.drop(columns=pmid_cols)\n",
    "#######################################################################\n",
    "\n",
    "gene_pair000 = gene_pair.copy()\n",
    "\n",
    "keywords_to_modify = [\"Ligand\", \"Receptor\"]\n",
    "exclude_keywords = [\"HGNC ID\", \"Location\", \"Human\"]  # Columns containing this will not be modified\n",
    "\n",
    "# Copy the original columns so we can modify only the first 10\n",
    "new_columns = gene_pair000.columns.tolist()\n",
    "\n",
    "# Modify only the first 10 columns\n",
    "new_columns[:10] = [\n",
    "    f'{col.split(\">\")[0]}\">Human {col.split(\">\")[1]}</span>'\n",
    "    if any(keyword in col for keyword in keywords_to_modify) and not any(exclude in col for exclude in exclude_keywords)\n",
    "    else col\n",
    "    for col in new_columns[:10]\n",
    "]\n",
    "\n",
    "# Assign the modified column names back to the DataFrame\n",
    "gene_pair000.columns = new_columns\n",
    "#######################################################################\n",
    "### For latest DB, no need to limit columns\n",
    "human_columns = [col for col in gene_pair000.columns][:16]\n",
    "#######################################################################\n",
    "#human_gene_pair = gene_pair.iloc[:, :-36]\n",
    "# remove mouse specific ones from the datatable\n",
    "evidence_cols = [col for col in gene_pair.columns if 'Human evidence' in col]\n",
    "human_gene_pair = gene_pair[~(gene_pair[evidence_cols[0]] == \"not conserved\")]\n",
    "# add number of mouse pair cards\n",
    "numOfMouseOrth = len(gene_pair[evidence_cols][(gene_pair[evidence_cols[0]] == \"not conserved\")])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "0fd36c9b-cf81-48c0-a9d2-62be5aff9236",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;span title=\"Double-click header of Interaction ID to reverse sort\"&gt;Interaction ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of LR Pair Card to reverse sort\"&gt;LR Pair Card&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\"&gt;Human LR Pair&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Official Gene Symbol; Hover on symbols below to show gene names\"&gt;Ligand&amp;nbsp;&amp;nbsp;&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Official Gene Symbol; Hover on symbols below to show gene names\"&gt;Receptor&amp;nbsp;&amp;nbsp;&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Ligand Symbols to reverse sort\"&gt;Ligand Symbols&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Receptor Symbols to reverse sort\"&gt;Receptor Symbols&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Location based on the predicted subcellular localization of the human proteome\"&gt;Ligand Location&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Location based on the predicted subcellular localization of the human proteome\"&gt;Receptor Location&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\"&gt;Ligand HGNC ID&amp;nbsp;&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\"&gt;Receptor HGNC ID&amp;nbsp;&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Click the logo below to run Perplexity on the Human LR pair\"&gt;Perplexity&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Human evidence to reverse sort\"&gt;Human evidence&amp;nbsp;&lt;/span&gt;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDB00001</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/re...</td>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>&lt;span title=\"alpha-2-macroglobulin\"&gt;A2M&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"heat shock protein family A (Hsp7...</td>\n",
       "      <td>&lt;span title=\"A2M (FWP007, S863-7, CPAMD5)\"&gt;A2M...</td>\n",
       "      <td>&lt;span title=\"HSPA5 (GRP78, BiP)\"&gt;HSPA5 (GRP78,...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;span title=\"based on PMID: 12194978, PMID: 32...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>DIRECT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDB00002</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/re...</td>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>&lt;span title=\"alpha-2-macroglobulin\"&gt;A2M&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"LDL receptor related protein 1\"&gt;L...</td>\n",
       "      <td>&lt;span title=\"A2M (FWP007, S863-7, CPAMD5)\"&gt;A2M...</td>\n",
       "      <td>&lt;span title=\"LRP1 (APR, A2MR, LRP, CD91, LRP1A...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>DIRECT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDB00003</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/re...</td>\n",
       "      <td>ACE BDKRB2</td>\n",
       "      <td>&lt;span title=\"angiotensin I converting enzyme\"&gt;...</td>\n",
       "      <td>&lt;span title=\"bradykinin receptor B2\"&gt;BDKRB2&lt;/s...</td>\n",
       "      <td>&lt;span title=\"ACE (DCP1, ACE1, CD143)\"&gt;ACE (DCP...</td>\n",
       "      <td>&lt;span title=\"BDKRB2 (BK-2)\"&gt;BDKRB2 (BK-2)&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>DIRECT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDB00004</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/re...</td>\n",
       "      <td>ADA DPP4</td>\n",
       "      <td>&lt;span title=\"adenosine deaminase\"&gt;ADA&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"dipeptidyl peptidase 4\"&gt;DPP4&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"ADA (ADA1)\"&gt;ADA (ADA1)&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"DPP4 (CD26, ADCP2, DPPIV)\"&gt;DPP4 (...</td>\n",
       "      <td>&lt;span title=\"based on hpa, uniprot\"&gt;cell membr...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>CONSERVATION</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDB00005</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/re...</td>\n",
       "      <td>ADAM10 EFNA5</td>\n",
       "      <td>&lt;span title=\"ADAM metallopeptidase domain 10\"&gt;...</td>\n",
       "      <td>&lt;span title=\"ephrin A5\"&gt;EFNA5&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"ADAM10 (kuz, MADM, HsT18717, CD15...</td>\n",
       "      <td>&lt;span title=\"EFNA5 (EPLG7, AF1, LERK7)\"&gt;EFNA5 ...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>CONSERVATION</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3545</th>\n",
       "      <td>CDB03546</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/re...</td>\n",
       "      <td>ZG16B TLR2</td>\n",
       "      <td>&lt;span title=\"zymogen granule protein 16B\"&gt;ZG16...</td>\n",
       "      <td>&lt;span title=\"toll like receptor 2\"&gt;TLR2&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"ZG16B (HRPE773, PRO1567, JCLN2)\"&gt;...</td>\n",
       "      <td>&lt;span title=\"TLR2 (TIL4, CD282)\"&gt;TLR2 (TIL4, C...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity\"&gt;cell me...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>DIRECT, CONSERVATION</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3546</th>\n",
       "      <td>CDB03547</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/re...</td>\n",
       "      <td>ZG16B TLR4</td>\n",
       "      <td>&lt;span title=\"zymogen granule protein 16B\"&gt;ZG16...</td>\n",
       "      <td>&lt;span title=\"toll like receptor 4\"&gt;TLR4&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"ZG16B (HRPE773, PRO1567, JCLN2)\"&gt;...</td>\n",
       "      <td>&lt;span title=\"TLR4 (hToll, CD284, TLR-4, ARMD10...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>DIRECT, CONSERVATION</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3547</th>\n",
       "      <td>CDB03548</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/re...</td>\n",
       "      <td>ZG16B TLR5</td>\n",
       "      <td>&lt;span title=\"zymogen granule protein 16B\"&gt;ZG16...</td>\n",
       "      <td>&lt;span title=\"toll like receptor 5\"&gt;TLR5&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"ZG16B (HRPE773, PRO1567, JCLN2)\"&gt;...</td>\n",
       "      <td>&lt;span title=\"TLR5 (SLEB1, TIL3, FLJ10052, MGC1...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>DIRECT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3548</th>\n",
       "      <td>CDB03549</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/re...</td>\n",
       "      <td>ZG16B TLR6</td>\n",
       "      <td>&lt;span title=\"zymogen granule protein 16B\"&gt;ZG16...</td>\n",
       "      <td>&lt;span title=\"toll like receptor 6\"&gt;TLR6&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"ZG16B (HRPE773, PRO1567, JCLN2)\"&gt;...</td>\n",
       "      <td>&lt;span title=\"TLR6 (CD286)\"&gt;TLR6 (CD286)&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>DIRECT, CONSERVATION</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3549</th>\n",
       "      <td>CDB03550</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/re...</td>\n",
       "      <td>ZP3 CHRNA7</td>\n",
       "      <td>&lt;span title=\"zona pellucida glycoprotein 3\"&gt;ZP...</td>\n",
       "      <td>&lt;span title=\"cholinergic receptor nicotinic al...</td>\n",
       "      <td>&lt;span title=\"ZP3 (ZP3A, ZP3B, ZP3-424, ZP3-372...</td>\n",
       "      <td>&lt;span title=\"CHRNA7\"&gt;CHRNA7&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>CONSERVATION</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3550 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     <span title=\"Double-click header of Interaction ID to reverse sort\">Interaction ID&nbsp;</span>  \\\n",
       "0                                              CDB00001                                                \n",
       "1                                              CDB00002                                                \n",
       "2                                              CDB00003                                                \n",
       "3                                              CDB00004                                                \n",
       "4                                              CDB00005                                                \n",
       "...                                                 ...                                                \n",
       "3545                                           CDB03546                                                \n",
       "3546                                           CDB03547                                                \n",
       "3547                                           CDB03548                                                \n",
       "3548                                           CDB03549                                                \n",
       "3549                                           CDB03550                                                \n",
       "\n",
       "     <span title=\"Double-click header of LR Pair Card to reverse sort\">LR Pair Card&nbsp;</span>  \\\n",
       "0     <a href=\"https://comp.med.yokohama-cu.ac.jp/re...                                            \n",
       "1     <a href=\"https://comp.med.yokohama-cu.ac.jp/re...                                            \n",
       "2     <a href=\"https://comp.med.yokohama-cu.ac.jp/re...                                            \n",
       "3     <a href=\"https://comp.med.yokohama-cu.ac.jp/re...                                            \n",
       "4     <a href=\"https://comp.med.yokohama-cu.ac.jp/re...                                            \n",
       "...                                                 ...                                            \n",
       "3545  <a href=\"https://comp.med.yokohama-cu.ac.jp/re...                                            \n",
       "3546  <a href=\"https://comp.med.yokohama-cu.ac.jp/re...                                            \n",
       "3547  <a href=\"https://comp.med.yokohama-cu.ac.jp/re...                                            \n",
       "3548  <a href=\"https://comp.med.yokohama-cu.ac.jp/re...                                            \n",
       "3549  <a href=\"https://comp.med.yokohama-cu.ac.jp/re...                                            \n",
       "\n",
       "     <span title=\"Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\">Human LR Pair</span>  \\\n",
       "0                                             A2M HSPA5                                                               \n",
       "1                                              A2M LRP1                                                               \n",
       "2                                            ACE BDKRB2                                                               \n",
       "3                                              ADA DPP4                                                               \n",
       "4                                          ADAM10 EFNA5                                                               \n",
       "...                                                 ...                                                               \n",
       "3545                                         ZG16B TLR2                                                               \n",
       "3546                                         ZG16B TLR4                                                               \n",
       "3547                                         ZG16B TLR5                                                               \n",
       "3548                                         ZG16B TLR6                                                               \n",
       "3549                                         ZP3 CHRNA7                                                               \n",
       "\n",
       "     <span title=\"Official Gene Symbol; Hover on symbols below to show gene names\">Ligand&nbsp;&nbsp;&nbsp;</span>  \\\n",
       "0        <span title=\"alpha-2-macroglobulin\">A2M</span>                                                              \n",
       "1        <span title=\"alpha-2-macroglobulin\">A2M</span>                                                              \n",
       "2     <span title=\"angiotensin I converting enzyme\">...                                                              \n",
       "3          <span title=\"adenosine deaminase\">ADA</span>                                                              \n",
       "4     <span title=\"ADAM metallopeptidase domain 10\">...                                                              \n",
       "...                                                 ...                                                              \n",
       "3545  <span title=\"zymogen granule protein 16B\">ZG16...                                                              \n",
       "3546  <span title=\"zymogen granule protein 16B\">ZG16...                                                              \n",
       "3547  <span title=\"zymogen granule protein 16B\">ZG16...                                                              \n",
       "3548  <span title=\"zymogen granule protein 16B\">ZG16...                                                              \n",
       "3549  <span title=\"zona pellucida glycoprotein 3\">ZP...                                                              \n",
       "\n",
       "     <span title=\"Official Gene Symbol; Hover on symbols below to show gene names\">Receptor&nbsp;&nbsp;&nbsp;</span>  \\\n",
       "0     <span title=\"heat shock protein family A (Hsp7...                                                                \n",
       "1     <span title=\"LDL receptor related protein 1\">L...                                                                \n",
       "2     <span title=\"bradykinin receptor B2\">BDKRB2</s...                                                                \n",
       "3      <span title=\"dipeptidyl peptidase 4\">DPP4</span>                                                                \n",
       "4                  <span title=\"ephrin A5\">EFNA5</span>                                                                \n",
       "...                                                 ...                                                                \n",
       "3545     <span title=\"toll like receptor 2\">TLR2</span>                                                                \n",
       "3546     <span title=\"toll like receptor 4\">TLR4</span>                                                                \n",
       "3547     <span title=\"toll like receptor 5\">TLR5</span>                                                                \n",
       "3548     <span title=\"toll like receptor 6\">TLR6</span>                                                                \n",
       "3549  <span title=\"cholinergic receptor nicotinic al...                                                                \n",
       "\n",
       "     <span title=\"Double-click header of Ligand Symbols to reverse sort\">Ligand Symbols&nbsp;</span>  \\\n",
       "0     <span title=\"A2M (FWP007, S863-7, CPAMD5)\">A2M...                                                \n",
       "1     <span title=\"A2M (FWP007, S863-7, CPAMD5)\">A2M...                                                \n",
       "2     <span title=\"ACE (DCP1, ACE1, CD143)\">ACE (DCP...                                                \n",
       "3            <span title=\"ADA (ADA1)\">ADA (ADA1)</span>                                                \n",
       "4     <span title=\"ADAM10 (kuz, MADM, HsT18717, CD15...                                                \n",
       "...                                                 ...                                                \n",
       "3545  <span title=\"ZG16B (HRPE773, PRO1567, JCLN2)\">...                                                \n",
       "3546  <span title=\"ZG16B (HRPE773, PRO1567, JCLN2)\">...                                                \n",
       "3547  <span title=\"ZG16B (HRPE773, PRO1567, JCLN2)\">...                                                \n",
       "3548  <span title=\"ZG16B (HRPE773, PRO1567, JCLN2)\">...                                                \n",
       "3549  <span title=\"ZP3 (ZP3A, ZP3B, ZP3-424, ZP3-372...                                                \n",
       "\n",
       "     <span title=\"Double-click header of Receptor Symbols to reverse sort\">Receptor Symbols&nbsp;</span>  \\\n",
       "0     <span title=\"HSPA5 (GRP78, BiP)\">HSPA5 (GRP78,...                                                    \n",
       "1     <span title=\"LRP1 (APR, A2MR, LRP, CD91, LRP1A...                                                    \n",
       "2      <span title=\"BDKRB2 (BK-2)\">BDKRB2 (BK-2)</span>                                                    \n",
       "3     <span title=\"DPP4 (CD26, ADCP2, DPPIV)\">DPP4 (...                                                    \n",
       "4     <span title=\"EFNA5 (EPLG7, AF1, LERK7)\">EFNA5 ...                                                    \n",
       "...                                                 ...                                                    \n",
       "3545  <span title=\"TLR2 (TIL4, CD282)\">TLR2 (TIL4, C...                                                    \n",
       "3546  <span title=\"TLR4 (hToll, CD284, TLR-4, ARMD10...                                                    \n",
       "3547  <span title=\"TLR5 (SLEB1, TIL3, FLJ10052, MGC1...                                                    \n",
       "3548     <span title=\"TLR6 (CD286)\">TLR6 (CD286)</span>                                                    \n",
       "3549                 <span title=\"CHRNA7\">CHRNA7</span>                                                    \n",
       "\n",
       "     <span title=\"Location based on the predicted subcellular localization of the human proteome\">Ligand Location</span>  \\\n",
       "0     <span title=\"based on perplexity, uniprot\">sec...                                                                    \n",
       "1     <span title=\"based on perplexity, uniprot\">sec...                                                                    \n",
       "2     <span title=\"based on perplexity, uniprot\">cel...                                                                    \n",
       "3     <span title=\"based on hpa, uniprot\">cell membr...                                                                    \n",
       "4     <span title=\"based on hpa, perplexity, uniprot...                                                                    \n",
       "...                                                 ...                                                                    \n",
       "3545  <span title=\"based on perplexity, uniprot\">cel...                                                                    \n",
       "3546  <span title=\"based on perplexity, uniprot\">cel...                                                                    \n",
       "3547  <span title=\"based on perplexity, uniprot\">cel...                                                                    \n",
       "3548  <span title=\"based on perplexity, uniprot\">cel...                                                                    \n",
       "3549  <span title=\"based on uniprot\">cell membrane</...                                                                    \n",
       "\n",
       "     <span title=\"Location based on the predicted subcellular localization of the human proteome\">Receptor Location</span>  \\\n",
       "0     <span title=\"based on PMID: 12194978, PMID: 32...                                                                      \n",
       "1     <span title=\"based on hpa, perplexity, uniprot...                                                                      \n",
       "2     <span title=\"based on hpa, perplexity, uniprot...                                                                      \n",
       "3     <span title=\"based on perplexity, uniprot\">cel...                                                                      \n",
       "4     <span title=\"based on perplexity, uniprot\">cel...                                                                      \n",
       "...                                                 ...                                                                      \n",
       "3545  <span title=\"based on hpa, perplexity\">cell me...                                                                      \n",
       "3546  <span title=\"based on hpa, perplexity, uniprot...                                                                      \n",
       "3547  <span title=\"based on perplexity, uniprot\">cel...                                                                      \n",
       "3548  <span title=\"based on perplexity, uniprot\">cel...                                                                      \n",
       "3549  <span title=\"based on perplexity, uniprot\">cel...                                                                      \n",
       "\n",
       "     <span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Ligand HGNC ID&nbsp;&nbsp;</span>  \\\n",
       "0     <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "1     <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "2     <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "3     <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "4     <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "...                                                 ...                                                                               \n",
       "3545  <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "3546  <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "3547  <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "3548  <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "3549  <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "\n",
       "     <span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Receptor HGNC ID&nbsp;&nbsp;</span>  \\\n",
       "0     <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "1     <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "2     <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "3     <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "4     <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "...                                                 ...                                                                                 \n",
       "3545  <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "3546  <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "3547  <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "3548  <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "3549  <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "\n",
       "     <span title=\"Click the logo below to run Perplexity on the Human LR pair\">Perplexity&nbsp;</span>  \\\n",
       "0     <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "1     <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "2     <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "3     <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "4     <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "...                                                 ...                                                  \n",
       "3545  <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "3546  <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "3547  <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "3548  <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "3549  <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "\n",
       "     <span title=\"Double-click header of Human evidence to reverse sort\">Human evidence&nbsp;</span>  \n",
       "0                                                DIRECT                                               \n",
       "1                                                DIRECT                                               \n",
       "2                                                DIRECT                                               \n",
       "3                                          CONSERVATION                                               \n",
       "4                                          CONSERVATION                                               \n",
       "...                                                 ...                                               \n",
       "3545                               DIRECT, CONSERVATION                                               \n",
       "3546                               DIRECT, CONSERVATION                                               \n",
       "3547                                             DIRECT                                               \n",
       "3548                               DIRECT, CONSERVATION                                               \n",
       "3549                                       CONSERVATION                                               \n",
       "\n",
       "[3550 rows x 13 columns]"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_gene_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "9d362bda-bdc6-46a2-b7cc-568aa0c62364",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_orth_symbol = dict(zip(ref['ensembl_gene_id'], ref['external_gene_name']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "3148e71d-6c6c-472f-a5f4-94c4c9ccc11d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_orth = fetchGSheet.safe_fetch(fetchGSheet.sheet_ID, f\"FROZEN_{species_lower}\", fetchGSheet.credentials_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "f7c574fa-1f25-4dbc-95ad-1facdf1a27e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Sorting order', 'LR_pair_orig', 'PMID', 'lig_species', 'rec_species',\n",
       "       'ligand_orig', 'receptor_orig', 'Pair_species', 'LR Pair Card',\n",
       "       'Sheep evidence', 'Sheep_ligand', 'Sheep_receptor', 'ENSEMBL ligand',\n",
       "       'ENSEMBL receptor'],\n",
       "      dtype='object', name=0)"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair_orth.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "8cc66b5a-d614-4636-a0b5-2f30ca3903f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_orth['Ligand Official Symbol'] = 'NA'\n",
    "gene_pair_orth['Ligand Official Symbol'] =  gene_pair_orth.apply(\n",
    "    lambda row: mapping_orth_symbol.get(row[f'{id} ligand'], row['Ligand Official Symbol'])\n",
    "    if pd.notna(row[f'{id} ligand']) else row['Ligand Official Symbol'],\n",
    "    axis=1\n",
    ")\n",
    "gene_pair_orth['Receptor Official Symbol'] = 'NA'\n",
    "gene_pair_orth['Receptor Official Symbol'] =  gene_pair_orth.apply(\n",
    "    lambda row: mapping_orth_symbol.get(row[f'{id} receptor'], row['Receptor Official Symbol'])\n",
    "    if pd.notna(row[f'{id} receptor']) else row['Receptor Official Symbol'],\n",
    "    axis=1\n",
    ")\n",
    "gene_pair_orth[\"same_as_off_lig\"] = (\n",
    "    gene_pair_orth[f\"{species}_ligand\"] == gene_pair_orth[\"Ligand Official Symbol\"]\n",
    ")\n",
    "\n",
    "gene_pair_orth[\"same_as_off_rec\"] = (\n",
    "    gene_pair_orth[f\"{species}_receptor\"] == gene_pair_orth[\"Receptor Official Symbol\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "a237e4d7-1154-4e2b-a609-a538f39f2d95",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_orth.to_csv(f\"data/{species}_{id}_ID_check.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "a67fc9a2-4b0d-453a-9a6a-7656f85e9403",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'species_info' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[121], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mspecies_info\u001b[49m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'species_info' is not defined"
     ]
    }
   ],
   "source": [
    "species_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "a4a6879a-c927-4b98-858f-e7b72249cd37",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'eid' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[69], line 370\u001b[0m\n\u001b[1;32m    368\u001b[0m zebrafish_gene_pair1 \u001b[38;5;241m=\u001b[39m process_species_gene_pair(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mZebrafish\u001b[39m\u001b[38;5;124m\"\u001b[39m, fetchGSheet, gene_pair)\n\u001b[1;32m    369\u001b[0m frog_gene_pair1 \u001b[38;5;241m=\u001b[39m process_species_gene_pair(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFrog\u001b[39m\u001b[38;5;124m\"\u001b[39m, fetchGSheet, gene_pair)\n\u001b[0;32m--> 370\u001b[0m chicken_gene_pair1 \u001b[38;5;241m=\u001b[39m \u001b[43mprocess_species_gene_pair\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mChicken\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfetchGSheet\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene_pair\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[69], line 350\u001b[0m, in \u001b[0;36mprocess_species_gene_pair\u001b[0;34m(species, fetchGSheet, gene_pair)\u001b[0m\n\u001b[1;32m    347\u001b[0m                 links\u001b[38;5;241m.\u001b[39mappend(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m<a href=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00murl\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m target=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_blank\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m>\u001b[39m\u001b[38;5;132;01m{\u001b[39;00meid\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m</a>\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m    348\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m, \u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(links)\n\u001b[0;32m--> 350\u001b[0m     gene_pair[\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mLigand \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mspecies_id\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m ID\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43mgene_pair\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mLigand ENSEMBL ID\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmake_ens_link\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    351\u001b[0m     gene_pair[\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mReceptor \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mspecies_id\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m ID\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m gene_pair[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mReceptor ENSEMBL ID\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mapply(make_ens_link)\n\u001b[1;32m    354\u001b[0m \u001b[38;5;66;03m### tooltips \u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/series.py:4924\u001b[0m, in \u001b[0;36mSeries.apply\u001b[0;34m(self, func, convert_dtype, args, by_row, **kwargs)\u001b[0m\n\u001b[1;32m   4789\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mapply\u001b[39m(\n\u001b[1;32m   4790\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   4791\u001b[0m     func: AggFuncType,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   4796\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m   4797\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m DataFrame \u001b[38;5;241m|\u001b[39m Series:\n\u001b[1;32m   4798\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   4799\u001b[0m \u001b[38;5;124;03m    Invoke function on values of Series.\u001b[39;00m\n\u001b[1;32m   4800\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   4915\u001b[0m \u001b[38;5;124;03m    dtype: float64\u001b[39;00m\n\u001b[1;32m   4916\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m   4917\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mSeriesApply\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   4918\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4919\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfunc\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4920\u001b[0m \u001b[43m        \u001b[49m\u001b[43mconvert_dtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconvert_dtype\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4921\u001b[0m \u001b[43m        \u001b[49m\u001b[43mby_row\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mby_row\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4922\u001b[0m \u001b[43m        \u001b[49m\u001b[43margs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4923\u001b[0m \u001b[43m        \u001b[49m\u001b[43mkwargs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m-> 4924\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/apply.py:1427\u001b[0m, in \u001b[0;36mSeriesApply.apply\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1424\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mapply_compat()\n\u001b[1;32m   1426\u001b[0m \u001b[38;5;66;03m# self.func is Callable\u001b[39;00m\n\u001b[0;32m-> 1427\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply_standard\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/apply.py:1507\u001b[0m, in \u001b[0;36mSeriesApply.apply_standard\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1501\u001b[0m \u001b[38;5;66;03m# row-wise access\u001b[39;00m\n\u001b[1;32m   1502\u001b[0m \u001b[38;5;66;03m# apply doesn't have a `na_action` keyword and for backward compat reasons\u001b[39;00m\n\u001b[1;32m   1503\u001b[0m \u001b[38;5;66;03m# we need to give `na_action=\"ignore\"` for categorical data.\u001b[39;00m\n\u001b[1;32m   1504\u001b[0m \u001b[38;5;66;03m# TODO: remove the `na_action=\"ignore\"` when that default has been changed in\u001b[39;00m\n\u001b[1;32m   1505\u001b[0m \u001b[38;5;66;03m#  Categorical (GH51645).\u001b[39;00m\n\u001b[1;32m   1506\u001b[0m action \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mignore\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(obj\u001b[38;5;241m.\u001b[39mdtype, CategoricalDtype) \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1507\u001b[0m mapped \u001b[38;5;241m=\u001b[39m \u001b[43mobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_map_values\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1508\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmapper\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcurried\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mna_action\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maction\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconvert\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconvert_dtype\u001b[49m\n\u001b[1;32m   1509\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1511\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(mapped) \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(mapped[\u001b[38;5;241m0\u001b[39m], ABCSeries):\n\u001b[1;32m   1512\u001b[0m     \u001b[38;5;66;03m# GH#43986 Need to do list(mapped) in order to get treated as nested\u001b[39;00m\n\u001b[1;32m   1513\u001b[0m     \u001b[38;5;66;03m#  See also GH#25959 regarding EA support\u001b[39;00m\n\u001b[1;32m   1514\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m obj\u001b[38;5;241m.\u001b[39m_constructor_expanddim(\u001b[38;5;28mlist\u001b[39m(mapped), index\u001b[38;5;241m=\u001b[39mobj\u001b[38;5;241m.\u001b[39mindex)\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/base.py:921\u001b[0m, in \u001b[0;36mIndexOpsMixin._map_values\u001b[0;34m(self, mapper, na_action, convert)\u001b[0m\n\u001b[1;32m    918\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(arr, ExtensionArray):\n\u001b[1;32m    919\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m arr\u001b[38;5;241m.\u001b[39mmap(mapper, na_action\u001b[38;5;241m=\u001b[39mna_action)\n\u001b[0;32m--> 921\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43malgorithms\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmap_array\u001b[49m\u001b[43m(\u001b[49m\u001b[43marr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmapper\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mna_action\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mna_action\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconvert\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconvert\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/algorithms.py:1743\u001b[0m, in \u001b[0;36mmap_array\u001b[0;34m(arr, mapper, na_action, convert)\u001b[0m\n\u001b[1;32m   1741\u001b[0m values \u001b[38;5;241m=\u001b[39m arr\u001b[38;5;241m.\u001b[39mastype(\u001b[38;5;28mobject\u001b[39m, copy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m   1742\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m na_action \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 1743\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mlib\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmap_infer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmapper\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconvert\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconvert\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1744\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1745\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m lib\u001b[38;5;241m.\u001b[39mmap_infer_mask(\n\u001b[1;32m   1746\u001b[0m         values, mapper, mask\u001b[38;5;241m=\u001b[39misna(values)\u001b[38;5;241m.\u001b[39mview(np\u001b[38;5;241m.\u001b[39muint8), convert\u001b[38;5;241m=\u001b[39mconvert\n\u001b[1;32m   1747\u001b[0m     )\n",
      "File \u001b[0;32mlib.pyx:2972\u001b[0m, in \u001b[0;36mpandas._libs.lib.map_infer\u001b[0;34m()\u001b[0m\n",
      "Cell \u001b[0;32mIn[69], line 344\u001b[0m, in \u001b[0;36mprocess_species_gene_pair.<locals>.make_ens_link\u001b[0;34m(cell)\u001b[0m\n\u001b[1;32m    342\u001b[0m links \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m    343\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m xid \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mstr\u001b[39m(cell)\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[0;32m--> 344\u001b[0m     xid \u001b[38;5;241m=\u001b[39m \u001b[43meid\u001b[49m\u001b[38;5;241m.\u001b[39mstrip()\n\u001b[1;32m    345\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m eid\u001b[38;5;241m.\u001b[39mstartswith(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mENS\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m    346\u001b[0m         url \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m http://www.ensembl.org/id/\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mxid\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'eid' is not defined"
     ]
    }
   ],
   "source": [
    "## Function to prepare datatables (cleaning and hyperlinking, adding tool tips, etc) input for the database qmds\n",
    "import sys, os\n",
    "from itables import init_notebook_mode\n",
    "import pandas as pd\n",
    "from itables import show\n",
    "from itables import options\n",
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "import re\n",
    "from bs4 import BeautifulSoup\n",
    "from createDataTable import gene_pair, gene_pair000, human_columns, lrPairsCount\n",
    "import warnings\n",
    "import fetchGSheet \n",
    "\n",
    "# Suppress SettingWithCopyWarning\n",
    "warnings.simplefilter(\"ignore\", category=UserWarning)\n",
    "\n",
    "def process_species_gene_pair(species, fetchGSheet, gene_pair):\n",
    "    species_name = {\n",
    "        \"Mouse\": \"mmusculus\",\n",
    "        \"Rat\": \"rnorvegicus\",\n",
    "        \"Zebrafish\":\"drerio\" ,\n",
    "        \"Chimpanzee\":\"ptroglodytes\",\n",
    "        \"Chicken\":\"ggallus\",\n",
    "        \"Pig\":\"sscrofa\",\n",
    "        \"Cow\":\"btaurus\",\n",
    "        \"Dog\":\"clfamiliaris\",\n",
    "        \"Horse\":\"ecaballus\",\n",
    "        \"Sheep\":\"oarambouillet\",\n",
    "        \"Marmoset\": \"cjacchus\" ,\n",
    "        \"Macaque\": \"mmulatta\",\n",
    "        \"Frog\": \"xtropicalis\"\n",
    "    }.get(species, \"Unknown species\")\n",
    "\n",
    "    if species == \"Mouse\":\n",
    "        species_id = \"MGI\"\n",
    "        species_info = pd.read_csv(f\"data/MRK_Merged_{species_id}_DB.tsv\", sep=\"\\t\", dtype=str)\n",
    "    elif species == \"Rat\":\n",
    "        species_id = \"RGD\"\n",
    "        species_info = pd.read_csv(f\"data/GENES_RAT_{species_id}_DB.tsv\", sep=\"\\t\", dtype=str)\n",
    "        # combine all known aliases also old ones\n",
    "        species_info[\"ALIASES\"] = species_info[[\"MARKER_SYMBOL\", \"OLD_SYMBOL\"]].apply(\n",
    "            lambda row: \";\".join(pd.unique(row.dropna().astype(str))), axis=1\n",
    "        )\n",
    "    elif species == \"Frog\":\n",
    "        species_id = \"XEN\"\n",
    "        species_info = pd.read_csv(f\"data/GenePageGeneralInfo_{species_id.capitalize()}base_DB.tsv\", sep=\"\\t\", dtype=str)\n",
    "    elif species == \"Zebrafish\":\n",
    "        species_id = \"ZFIN\"\n",
    "        species_info = pd.read_csv(f\"data/Zebrafish_merged_{species_id}_DB.tsv\", sep=\"\\t\", dtype=str)\n",
    "        # combine all known aliases also old ones\n",
    "        species_info[\"Aliases\"] = species_info[[\"Current Name\", \"Previous Name\"]].apply(\n",
    "            lambda row: \", \".join(pd.unique(row.dropna().astype(str))), axis=1\n",
    "        )\n",
    "    else: \n",
    "        species_id = \"ENSEMBL\"\n",
    "        species_info = pd.read_csv(f\"data/hsapiens_ID_biomart_{species_name}_centric.csv\", dtype=str)\n",
    "    \n",
    "    \n",
    "    species_lower = species.lower()\n",
    "    \n",
    "    if species == \"Mouse\":\n",
    "        gene_pair_species = getattr(fetchGSheet, f\"gene_pair_{species_lower}\")  \n",
    "    else:\n",
    "        gene_pair_species = fetchGSheet.safe_fetch(fetchGSheet.sheet_ID, f\"FROZEN_{species_lower}\", fetchGSheet.credentials_file)\n",
    "    \n",
    "    def extract_visible_text(col):\n",
    "        \"\"\"Extract visible text between '>' and '</a>'.\"\"\"\n",
    "        match = re.search(r'>([^<]+)</a>', col)\n",
    "        if match:\n",
    "            return match.group(1).strip()\n",
    "        return None\n",
    "    \n",
    "    # Grab the Interaction ID and \n",
    "    gene_pair = gene_pair.iloc[:, :9]\n",
    "    # Next, drop columns at index positions 3 and 4 (\"Ligand and Receptor\" since we already have ligand symbols and receptor symbols)\n",
    "    gene_pair = gene_pair.drop(gene_pair.columns[[3, 4]], axis=1)\n",
    "    exclude_keywords = [\"HGNC ID\", \"Location\", \"Human\"]  # Columns containing this will not be modified\n",
    "    keywords_to_modify = [\"Ligand Symbols\", \"Receptor Symbols\"]\n",
    "    # Copy the original columns so we can modify only the first 10\n",
    "    new_columns = gene_pair.columns.tolist()\n",
    "    \n",
    "    # Modify only the first 10 columns\n",
    "    new_columns = [\n",
    "        f'{col.split(\">\")[0]}\">Human {col.split(\">\")[1]}</span>'\n",
    "        if any(keyword in col for keyword in keywords_to_modify) and not any(exclude in col for exclude in exclude_keywords)\n",
    "        else col\n",
    "        for col in new_columns\n",
    "    ]\n",
    "    # Assign the modified column names back to the DataFrame\n",
    "    gene_pair.columns = new_columns\n",
    "    \n",
    "    gene_pair[\"LR Pair Card\"] = gene_pair.iloc[:, 1].apply(extract_visible_text)\n",
    "    ligand_symbols_col = [col for col in gene_pair.columns if \"Ligand Symbols\" in col][0]\n",
    "    receptor_symbols_col = [col for col in gene_pair.columns if \"Receptor Symbols\" in col][0]\n",
    "    gene_pair = gene_pair.rename(columns={ligand_symbols_col: \"Human Ligand Symbols\",\n",
    "                                          receptor_symbols_col: \"Human Receptor Symbols\"})\n",
    "    \n",
    "    \n",
    "    gene_pair_species = gene_pair_species[[\n",
    "        \"LR Pair Card\",\n",
    "        f\"{species}_ligand\",\n",
    "        f\"{species}_receptor\",\n",
    "        f\"{species} evidence\",\n",
    "        f\"{species_id} ligand\",\n",
    "        f\"{species_id} receptor\",\n",
    "        \"PMID\"\n",
    "    ]]\n",
    "    \n",
    "    \n",
    "    gene_pair =gene_pair_species.merge(gene_pair,how=\"left\", on=\"LR Pair Card\")\n",
    "    \n",
    "    if species == \"Mouse\":\n",
    "        spec_id = f\"{species_id} Marker Accession ID\"\n",
    "        spec_name = \"Marker Name\"\n",
    "        spec_alias = \"Aliases\"\n",
    "    elif species == \"Rat\":\n",
    "        spec_id = f\"GENE_{species_id}_ID\"\n",
    "        spec_name = \"NAME\"\n",
    "        spec_alias = \"ALIASES\"\n",
    "    elif species == \"Zebrafish\":\n",
    "        spec_id = f\"{species_id}_ID\"\n",
    "        spec_name = \"Current Name\"\n",
    "        spec_alias = \"Aliases\"\n",
    "    elif species == \"Frog\":\n",
    "        spec_id = \"tropicalis gene ID\"\n",
    "        spec_name = \"gene name\"\n",
    "        spec_alias = \"gene synonyms\"\n",
    "    else:\n",
    "        spec_id = \"ensembl_gene_id\"\n",
    "        spec_name = \"external_gene_name\"\n",
    "        spec_alias = \"external_synonym\"\n",
    "    \n",
    "    \n",
    "    species_info = species_info[[spec_id, spec_name,spec_alias]]\n",
    "    \n",
    "    if species in [\"Mouse\", \"Frog\"]:\n",
    "        species_info[spec_alias] = species_info[spec_alias].str.replace(\"|\", \", \", regex=False)\n",
    "    elif species == \"Rat\":\n",
    "        species_info[spec_alias] = species_info[spec_alias].str.replace(\";\", \", \", regex=False)\n",
    "\n",
    "    \n",
    "    gene_pair = gene_pair.merge(species_info,how=\"left\", left_on = f\"{species_id} ligand\",right_on=spec_id)\n",
    "    gene_pair = gene_pair.drop(columns=[spec_id])\n",
    "    gene_pair = gene_pair.rename(columns={\n",
    "                                          spec_name: \"Ligand Name\",\n",
    "                                          spec_alias: \"Ligand Symbols\"\n",
    "                                         }\n",
    "                                )\n",
    "    gene_pair = gene_pair.merge(\n",
    "        species_info,\n",
    "        how=\"left\",\n",
    "        left_on=f\"{species_id} receptor\",\n",
    "        right_on=spec_id\n",
    "    )\n",
    "    \n",
    "    gene_pair = gene_pair.drop(columns=[spec_id])\n",
    "    gene_pair = gene_pair.rename(columns={\n",
    "                                          spec_name: \"Receptor Name\",\n",
    "                                          spec_alias: \"Receptor Symbols\"\n",
    "                                         }\n",
    "                                )\n",
    "    \n",
    "    \n",
    "    \n",
    "    gene_pair[f\"{species} LR Pair\"] = np.where(\n",
    "        gene_pair[f\"{species} evidence\"] == \"not conserved\", \n",
    "        f\"no {species_lower} ortholog\",                                  \n",
    "        gene_pair[f\"{species}_ligand\"] + \" \" + gene_pair[f\"{species}_receptor\"] \n",
    "    )\n",
    "    \n",
    "    gene_pair = gene_pair[~(gene_pair[f\"{species} evidence\"] == \"not conserved\")]\n",
    "    def format_symbol_aliases(symbol, aliases):\n",
    "        \"\"\"\n",
    "        Formats symbol, old symbols, and aliases.\n",
    "        If the final formatted string would be empty after considering N/A values\n",
    "        and empty inputs, it returns \"species-specific\".\n",
    "        Otherwise, it formats based on the presence of old_symbol and aliases,\n",
    "        removing unnecessary parentheses or commas, following the structure:\n",
    "        \"Symbol (Old Symbol, Aliases)\" if both exist.\n",
    "        \"\"\"\n",
    "        # Normalize inputs to empty strings if they are None/NaN or just whitespace\n",
    "        symbol_str = str(symbol).strip()\n",
    "        # old_symbol_str = str(old_symbol).strip()\n",
    "        aliases_str = str(aliases).strip()\n",
    "    \n",
    "        # Filter out values that are empty strings or \"N/A\" for old_symbol and aliases\n",
    "        parts_for_join = []\n",
    "        # if old_symbol_str and old_symbol_str != \"N/A\":\n",
    "        #     parts_for_join.append(old_symbol_str)\n",
    "        if aliases_str and aliases_str != \"N/A\":\n",
    "            parts_for_join.append(aliases_str)\n",
    "    \n",
    "        # Construct the preliminary result based on your original logic:\n",
    "        # \"symbol (old_symbol, aliases)\" if parts_for_join is not empty, else \"symbol\"\n",
    "        if parts_for_join:\n",
    "            prelim_result = f\"{symbol_str} ({', '.join(parts_for_join)})\"\n",
    "        else:\n",
    "            prelim_result = symbol_str # Just the symbol if no old_symbol or aliases\n",
    "    \n",
    "        return prelim_result\n",
    "    \n",
    "    # This is crucial for consistent handling by the function before processing \"N/A\".\n",
    "    gene_pair[f\"{species}_ligand\"] = gene_pair[f\"{species}_ligand\"].fillna('')\n",
    "    gene_pair['Ligand Symbols'] = gene_pair['Ligand Symbols'].fillna('')\n",
    "    \n",
    "    gene_pair['Ligand Symbols'] = gene_pair.apply(\n",
    "        lambda row: format_symbol_aliases(row[f\"{species}_ligand\"], row['Ligand Symbols']),\n",
    "        axis=1\n",
    "    )\n",
    "    \n",
    "    gene_pair[f\"{species}_receptor\"] = gene_pair[f\"{species}_receptor\"].fillna('')\n",
    "    gene_pair['Receptor Symbols'] = gene_pair['Receptor Symbols'].fillna('')\n",
    "    \n",
    "    gene_pair['Receptor Symbols'] = gene_pair.apply(\n",
    "        lambda row: format_symbol_aliases(row[f\"{species}_receptor\"], row['Receptor Symbols']),\n",
    "        axis=1\n",
    "    )\n",
    "    # The list of columns to group by\n",
    "    interaction_id_col = [col for col in gene_pair.columns if \"Interaction ID\" in col][0]\n",
    "    \n",
    "    grouping_cols = [\n",
    "        interaction_id_col, f\"{species} LR Pair\" #, \"Ligand Symbols\", \"Receptor Symbols\" \n",
    "    ]\n",
    "    \n",
    "    aggregation_cols = [\n",
    "        col for col in gene_pair.columns if col not in grouping_cols\n",
    "    ]\n",
    "    \n",
    "    # 3. Create a dictionary mapping each aggregation column to the joining function\n",
    "    agg_dict = {\n",
    "        col: lambda x: ', '.join(x.astype(str).unique()) for col in aggregation_cols\n",
    "    }\n",
    "    \n",
    "    # 4. Perform the groupby and aggregation\n",
    "    gene_pair = gene_pair.groupby(grouping_cols).agg(agg_dict).reset_index()\n",
    "    \n",
    "    # make direct, conservation and conservation, direct the same\n",
    "    gene_pair[f\"{species} evidence\"] = gene_pair[f\"{species} evidence\"].replace(\"CONSERVATION, DIRECT\", \"DIRECT, CONSERVATION\")\n",
    "    \n",
    "    \n",
    "    def generate_perplexity_link_pmid(row, species, species_lower): \n",
    "        query = (\n",
    "            f\"What-is-the-biological-relevance-of-the-ligand-and-receptor-pair-\"\n",
    "            f\"{row[f'{species} LR Pair']}-based-on-Pubmed-ID-\"\n",
    "            f\"{row['PMID']}-in-{species_lower}\"\n",
    "        )\n",
    "        return (\n",
    "             f'<a href=\"https://www.perplexity.ai/search?q={query}\" target=\"_blank\" style=\"text-decoration: none;\">&#128269;</a>'\n",
    "        )\n",
    "    \n",
    "    \n",
    "    # Apply function to the DataFrame\n",
    "    gene_pair[\"Perplexity\"] = gene_pair.apply(\n",
    "        generate_perplexity_link_pmid, axis=1, args=(species, species_lower)\n",
    "    )\n",
    "    \n",
    "    \n",
    "    gene_pair = gene_pair.rename(columns={\n",
    "                                          f\"{species_id} ligand\": f\"Ligand {species_id} ID\",\n",
    "                                          f\"{species_id} receptor\": f\"Receptor {species_id} ID\"\n",
    "                                         }\n",
    "                                )\n",
    "    ligand_loc_col = [col for col in gene_pair.columns if \"Ligand Location\" in col][0]\n",
    "    receptor_loc_col = [col for col in gene_pair.columns if \"Receptor Location\" in col][0]\n",
    "    lr_pair_card = [col for col in gene_pair.columns if \">LR Pair Card\" in col][0]\n",
    "    # gene_pair.columns\n",
    "    gene_pair[\"LR Pair Card\"] = gene_pair[lr_pair_card]\n",
    "    gene_pair = gene_pair[[interaction_id_col, \"LR Pair Card\",f\"{species} LR Pair\", 'Ligand Symbols', 'Receptor Symbols', ligand_loc_col, receptor_loc_col, f\"Ligand {species_id} ID\", f\"Receptor {species_id} ID\", \"Perplexity\", f\"{species} evidence\", \"Human Ligand Symbols\", \"Human Receptor Symbols\"]]\n",
    "    if species == \"Mouse\":\n",
    "        # Linkify multiple species IDs in Ligand column\n",
    "        gene_pair[f\"Ligand {species_id} ID\"] = gene_pair[f\"Ligand {species_id} ID\"].apply(\n",
    "            lambda cell: \", \".join(\n",
    "                f'<a href=\"https://www.informatics.jax.org/marker/{mgi.strip()}\" target=\"_blank\">{mgi.strip()}</a>'\n",
    "                for mgi in str(cell).split(\", \")\n",
    "                if mgi.strip()\n",
    "            ) if pd.notna(cell) else \"\"\n",
    "        )\n",
    "        \n",
    "        # Linkify multiple MGI IDs in Receptor column\n",
    "        gene_pair[f\"Receptor {species_id} ID\"] = gene_pair[f\"Receptor {species_id} ID\"].apply(\n",
    "            lambda cell: \", \".join(\n",
    "                f'<a href=\"https://www.informatics.jax.org/marker/{mgi.strip()}\" target=\"_blank\">{mgi.strip()}</a>'\n",
    "                for mgi in str(cell).split(\", \")\n",
    "                if mgi.strip()\n",
    "            ) if pd.notna(cell) else \"\"\n",
    "        )\n",
    "        \n",
    "    elif species == \"Rat\":\n",
    "        # Linkify multiple RGD IDs in Receptor column\n",
    "        gene_pair[f\"Ligand {species_id} ID\"] = gene_pair[\"Ligand RGD ID\"].apply(\n",
    "            lambda cell: \", \".join(\n",
    "                f'<a href=\"https://rgd.mcw.edu/rgdweb/report/gene/main.html?id={rgd.strip()}\" target=\"_blank\">{\"RGD:\"+rgd.strip()}</a>'\n",
    "                for rgd in str(cell).split(\", \")\n",
    "                if rgd.strip()\n",
    "            ) if pd.notna(cell) else \"\"\n",
    "        )\n",
    "        \n",
    "        \n",
    "        # Linkify multiple RGD IDs in Receptor column\n",
    "        gene_pair[f\"Receptor {species_id} ID\"] = gene_pair[\"Receptor RGD ID\"].apply(\n",
    "            lambda cell: \", \".join(\n",
    "                f'<a href=\"https://rgd.mcw.edu/rgdweb/report/gene/main.html?id={rgd.strip()}\" target=\"_blank\">{\"RGD:\"+rgd.strip()}</a>'\n",
    "                for rgd in str(cell).split(\", \")\n",
    "                if rgd.strip()\n",
    "            ) if pd.notna(cell) else \"\"\n",
    "        )\n",
    "        \n",
    "    elif species == \"Zebrafish\":\n",
    "        # Linkify multiple Zebrafish Receptor IDs\n",
    "        gene_pair[f\"Ligand {species_id} ID\"] = gene_pair['Ligand ZFIN ID'].apply(\n",
    "            lambda cell: \", \".join(\n",
    "                f'<a href=\"https://zfin.org/{zfin.strip()}\" target=\"_blank\">{zfin.strip()}</a>'\n",
    "                for zfin in str(cell).split(\", \")\n",
    "                if zfin.strip()\n",
    "            ) if pd.notna(cell) else \"\"\n",
    "        )\n",
    "        \n",
    "        # Linkify multiple Zebrafish Ligand IDs\n",
    "        gene_pair[f\"Receptor {species_id} ID\"] = gene_pair['Receptor ZFIN ID'].apply(\n",
    "            lambda cell: \", \".join(\n",
    "                f'<a href=\"https://zfin.org/{zfin.strip()}\" target=\"_blank\">{zfin.strip()}</a>'\n",
    "                for zfin in str(cell).split(\", \")\n",
    "                if zfin.strip()\n",
    "            ) if pd.notna(cell) else \"\"\n",
    "        )\n",
    "    elif species == \"Frog\":\n",
    "        def make_xenbase_link(cell):\n",
    "            links = []\n",
    "            for xid in str(cell).split(\",\"):\n",
    "                xid = xid.strip()\n",
    "                if xid.startswith(\"XB-GENE-\"):\n",
    "                    url = f\"https://www.xenbase.org/xenbase/gene/showgene.do?method=display&geneId={xid}\"\n",
    "                    links.append(f'<a href=\"{url}\" target=\"_blank\">{xid}</a>')\n",
    "            return \", \".join(links)\n",
    "\n",
    "        gene_pair[f\"Ligand {species_id} ID\"] = gene_pair['Ligand XEN ID'].apply(make_xenbase_link)\n",
    "        gene_pair[f\"Receptor {species_id} ID\"] = gene_pair['Receptor XEN ID'].apply(make_xenbase_link)\n",
    "        \n",
    "    else:\n",
    "        def make_ens_link(cell):\n",
    "            links = []\n",
    "            for xid in str(cell).split(\",\"):\n",
    "                xid = eid.strip()\n",
    "                if eid.startswith(\"ENS\"):\n",
    "                    url = f\" http://www.ensembl.org/id/{xid}\"\n",
    "                    links.append(f'<a href=\"{url}\" target=\"_blank\">{eid}</a>')\n",
    "            return \", \".join(links)\n",
    "\n",
    "        gene_pair[f\"Ligand {species_id} ID\"] = gene_pair['Ligand ENSEMBL ID'].apply(make_ens_link)\n",
    "        gene_pair[f\"Receptor {species_id} ID\"] = gene_pair['Receptor ENSEMBL ID'].apply(make_ens_link)\n",
    "\n",
    "    \n",
    "    ### tooltips \n",
    "    gene_pair[\"Ligand Symbols\"] = [\n",
    "        f'<span title=\"{aliases}\">{aliases}</span>'\n",
    "        for aliases in gene_pair[\"Ligand Symbols\"]\n",
    "    ]\n",
    "    gene_pair[\"Receptor Symbols\"] = [\n",
    "        f'<span title=\"{aliases}\">{aliases}</span>'\n",
    "        for aliases in gene_pair[\"Receptor Symbols\"]\n",
    "    ]\n",
    "    return gene_pair\n",
    "\n",
    "\n",
    "mouse_gene_pair1 = process_species_gene_pair(\"Mouse\", fetchGSheet, gene_pair)\n",
    "rat_gene_pair1 = process_species_gene_pair(\"Rat\", fetchGSheet, gene_pair)\n",
    "zebrafish_gene_pair1 = process_species_gene_pair(\"Zebrafish\", fetchGSheet, gene_pair)\n",
    "frog_gene_pair1 = process_species_gene_pair(\"Frog\", fetchGSheet, gene_pair)\n",
    "chicken_gene_pair1 = process_species_gene_pair(\"Chicken\", fetchGSheet, gene_pair)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "cbd15e3b-da9b-451c-a9c8-8d65c46d731f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair0_copy = gene_pair0_copy.merge(conservation, how= 'left', on = \"LR Pair Card\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4e2b1166-16a4-4f9b-9462-0a793f44268b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Interaction ID</th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>Human LR Pair</th>\n",
       "      <th>Ligand</th>\n",
       "      <th>Receptor</th>\n",
       "      <th>Ligand Symbols</th>\n",
       "      <th>Receptor Symbols</th>\n",
       "      <th>Ligand Location</th>\n",
       "      <th>Receptor Location</th>\n",
       "      <th>Ligand HGNC ID</th>\n",
       "      <th>...</th>\n",
       "      <th>Human evidence</th>\n",
       "      <th>PMID</th>\n",
       "      <th>Ligand Name</th>\n",
       "      <th>Receptor Name</th>\n",
       "      <th>Mouse_ligand</th>\n",
       "      <th>Mouse_receptor</th>\n",
       "      <th>Ligand MGI ID</th>\n",
       "      <th>Receptor MGI ID</th>\n",
       "      <th>Direct</th>\n",
       "      <th>Conserved</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDB00001</td>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>A2M</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>&lt;span title=\"A2M (FWP007, S863-7, CPAMD5)\"&gt;A2M...</td>\n",
       "      <td>&lt;span title=\"HSPA5 (GRP78, BiP)\"&gt;HSPA5 (GRP78,...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;span title=\"based on PMID: 12194978, PMID: 32...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>...</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/re...</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>heat shock protein family A (Hsp70) member 5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>Mouse, Rat, Frog, Zebrafish, Chicken, Macaque,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDB00002</td>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>A2M</td>\n",
       "      <td>LRP1</td>\n",
       "      <td>&lt;span title=\"A2M (FWP007, S863-7, CPAMD5)\"&gt;A2M...</td>\n",
       "      <td>&lt;span title=\"LRP1 (APR, A2MR, LRP, CD91, LRP1A...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>...</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/re...</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>LDL receptor related protein 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>Mouse, Rat, Frog, Zebrafish, Chicken, Macaque,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDB00003</td>\n",
       "      <td>ACE BDKRB2</td>\n",
       "      <td>ACE BDKRB2</td>\n",
       "      <td>ACE</td>\n",
       "      <td>BDKRB2</td>\n",
       "      <td>&lt;span title=\"ACE (DCP1, ACE1, CD143)\"&gt;ACE (DCP...</td>\n",
       "      <td>&lt;span title=\"BDKRB2 (BK-2)\"&gt;BDKRB2 (BK-2)&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>...</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/re...</td>\n",
       "      <td>angiotensin I converting enzyme</td>\n",
       "      <td>bradykinin receptor B2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human</td>\n",
       "      <td>Mouse, Rat, Frog, Zebrafish, Chicken, Macaque,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDB00004</td>\n",
       "      <td>ADA DPP4</td>\n",
       "      <td>ADA DPP4</td>\n",
       "      <td>ADA</td>\n",
       "      <td>DPP4</td>\n",
       "      <td>&lt;span title=\"ADA (ADA1)\"&gt;ADA (ADA1)&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"DPP4 (CD26, ADCP2, DPPIV)\"&gt;DPP4 (...</td>\n",
       "      <td>&lt;span title=\"based on hpa, uniprot\"&gt;cell membr...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>...</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/re...</td>\n",
       "      <td>adenosine deaminase</td>\n",
       "      <td>dipeptidyl peptidase 4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>Human, Mouse, Rat, Frog, Zebrafish, Chicken, M...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDB00005</td>\n",
       "      <td>ADAM10 EFNA5</td>\n",
       "      <td>ADAM10 EFNA5</td>\n",
       "      <td>ADAM10</td>\n",
       "      <td>EFNA5</td>\n",
       "      <td>&lt;span title=\"ADAM10 (kuz, MADM, HsT18717, CD15...</td>\n",
       "      <td>&lt;span title=\"EFNA5 (EPLG7, AF1, LERK7)\"&gt;EFNA5 ...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>...</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/re...</td>\n",
       "      <td>ADAM metallopeptidase domain 10</td>\n",
       "      <td>ephrin A5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>Human, Mouse, Rat, Frog, Zebrafish, Chicken, M...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3584</th>\n",
       "      <td>CDB03577</td>\n",
       "      <td>Pcdhb9 Pcdhb9</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>...</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/re...</td>\n",
       "      <td>protocadherin beta 9</td>\n",
       "      <td>protocadherin beta 9</td>\n",
       "      <td>Pcdhb9</td>\n",
       "      <td>Pcdhb9</td>\n",
       "      <td>MGI:2136744</td>\n",
       "      <td>MGI:2136744</td>\n",
       "      <td>Mouse</td>\n",
       "      <td>Rat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3585</th>\n",
       "      <td>CDB03578</td>\n",
       "      <td>Pcdhgb8 Pcdhgb8</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>...</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/re...</td>\n",
       "      <td>protocadherin gamma subfamily B, 8</td>\n",
       "      <td>protocadherin gamma subfamily B, 8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>MGI:1935200</td>\n",
       "      <td>MGI:1935200</td>\n",
       "      <td>Mouse</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3586</th>\n",
       "      <td>CDB03579</td>\n",
       "      <td>Saa3 Tlr4</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>TLR4</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"TLR4 (hToll, CD284, TLR-4, ARMD10...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>...</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/re...</td>\n",
       "      <td>serum amyloid A 3</td>\n",
       "      <td>toll-like receptor 4</td>\n",
       "      <td>Saa3</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>MGI:98223</td>\n",
       "      <td>MGI:96824</td>\n",
       "      <td>Mouse</td>\n",
       "      <td>Zebrafish</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3587</th>\n",
       "      <td>CDB03580</td>\n",
       "      <td>Sema4a Timd2</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>SEMA4A</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;span title=\"SEMA4A (SEMAB, SemB, FLJ12287, CO...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>...</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/re...</td>\n",
       "      <td>sema domain, immunoglobulin domain (Ig), trans...</td>\n",
       "      <td>T cell immunoglobulin and mucin domain contain...</td>\n",
       "      <td>Sema4a</td>\n",
       "      <td>Timd2</td>\n",
       "      <td>MGI:107560</td>\n",
       "      <td>MGI:2159681</td>\n",
       "      <td>Mouse</td>\n",
       "      <td>Rat, Frog, Horse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3588</th>\n",
       "      <td>CDB03581</td>\n",
       "      <td>Zp3 Zp3r</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>ZP3</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;span title=\"ZP3 (ZP3A, ZP3B, ZP3-424, ZP3-372...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;span title=\"based on PMID: 8188752\"&gt;cell memb...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>...</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/re...</td>\n",
       "      <td>zona pellucida glycoprotein 3</td>\n",
       "      <td>zona pellucida 3 receptor</td>\n",
       "      <td>Zp3</td>\n",
       "      <td>Zp3r</td>\n",
       "      <td>MGI:99215</td>\n",
       "      <td>MGI:104965</td>\n",
       "      <td>Mouse</td>\n",
       "      <td>Rat, Zebrafish, Cow</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3589 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Interaction ID     LR Pair Card      Human LR Pair         Ligand  \\\n",
       "0          CDB00001        A2M HSPA5          A2M HSPA5            A2M   \n",
       "1          CDB00002         A2M LRP1           A2M LRP1            A2M   \n",
       "2          CDB00003       ACE BDKRB2         ACE BDKRB2            ACE   \n",
       "3          CDB00004         ADA DPP4           ADA DPP4            ADA   \n",
       "4          CDB00005     ADAM10 EFNA5       ADAM10 EFNA5         ADAM10   \n",
       "...             ...              ...                ...            ...   \n",
       "3584       CDB03577    Pcdhb9 Pcdhb9  no human ortholog  not conserved   \n",
       "3585       CDB03578  Pcdhgb8 Pcdhgb8  no human ortholog  not conserved   \n",
       "3586       CDB03579        Saa3 Tlr4  no human ortholog  not conserved   \n",
       "3587       CDB03580     Sema4a Timd2  no human ortholog         SEMA4A   \n",
       "3588       CDB03581         Zp3 Zp3r  no human ortholog            ZP3   \n",
       "\n",
       "           Receptor                                     Ligand Symbols  \\\n",
       "0             HSPA5  <span title=\"A2M (FWP007, S863-7, CPAMD5)\">A2M...   \n",
       "1              LRP1  <span title=\"A2M (FWP007, S863-7, CPAMD5)\">A2M...   \n",
       "2            BDKRB2  <span title=\"ACE (DCP1, ACE1, CD143)\">ACE (DCP...   \n",
       "3              DPP4         <span title=\"ADA (ADA1)\">ADA (ADA1)</span>   \n",
       "4             EFNA5  <span title=\"ADAM10 (kuz, MADM, HsT18717, CD15...   \n",
       "...             ...                                                ...   \n",
       "3584  not conserved  <span title=\"no human ortholog\">no human ortho...   \n",
       "3585  not conserved  <span title=\"no human ortholog\">no human ortho...   \n",
       "3586           TLR4  <span title=\"no human ortholog\">no human ortho...   \n",
       "3587  not conserved  <span title=\"SEMA4A (SEMAB, SemB, FLJ12287, CO...   \n",
       "3588  not conserved  <span title=\"ZP3 (ZP3A, ZP3B, ZP3-424, ZP3-372...   \n",
       "\n",
       "                                       Receptor Symbols  \\\n",
       "0     <span title=\"HSPA5 (GRP78, BiP)\">HSPA5 (GRP78,...   \n",
       "1     <span title=\"LRP1 (APR, A2MR, LRP, CD91, LRP1A...   \n",
       "2      <span title=\"BDKRB2 (BK-2)\">BDKRB2 (BK-2)</span>   \n",
       "3     <span title=\"DPP4 (CD26, ADCP2, DPPIV)\">DPP4 (...   \n",
       "4     <span title=\"EFNA5 (EPLG7, AF1, LERK7)\">EFNA5 ...   \n",
       "...                                                 ...   \n",
       "3584  <span title=\"no human ortholog\">no human ortho...   \n",
       "3585  <span title=\"no human ortholog\">no human ortho...   \n",
       "3586  <span title=\"TLR4 (hToll, CD284, TLR-4, ARMD10...   \n",
       "3587  <span title=\"no human ortholog\">no human ortho...   \n",
       "3588  <span title=\"no human ortholog\">no human ortho...   \n",
       "\n",
       "                                        Ligand Location  \\\n",
       "0     <span title=\"based on perplexity, uniprot\">sec...   \n",
       "1     <span title=\"based on perplexity, uniprot\">sec...   \n",
       "2     <span title=\"based on perplexity, uniprot\">cel...   \n",
       "3     <span title=\"based on hpa, uniprot\">cell membr...   \n",
       "4     <span title=\"based on hpa, perplexity, uniprot...   \n",
       "...                                                 ...   \n",
       "3584  <span title=\"based on uniprot\">cell membrane</...   \n",
       "3585  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "3586  <span title=\"based on perplexity, uniprot\">sec...   \n",
       "3587  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "3588  <span title=\"based on uniprot\">cell membrane</...   \n",
       "\n",
       "                                      Receptor Location  \\\n",
       "0     <span title=\"based on PMID: 12194978, PMID: 32...   \n",
       "1     <span title=\"based on hpa, perplexity, uniprot...   \n",
       "2     <span title=\"based on hpa, perplexity, uniprot...   \n",
       "3     <span title=\"based on perplexity, uniprot\">cel...   \n",
       "4     <span title=\"based on perplexity, uniprot\">cel...   \n",
       "...                                                 ...   \n",
       "3584  <span title=\"based on uniprot\">cell membrane</...   \n",
       "3585  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "3586  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "3587  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "3588  <span title=\"based on PMID: 8188752\">cell memb...   \n",
       "\n",
       "                                         Ligand HGNC ID  ... Human evidence  \\\n",
       "0     <a href=\"https://www.genenames.org/data/gene-s...  ...         DIRECT   \n",
       "1     <a href=\"https://www.genenames.org/data/gene-s...  ...         DIRECT   \n",
       "2     <a href=\"https://www.genenames.org/data/gene-s...  ...         DIRECT   \n",
       "3     <a href=\"https://www.genenames.org/data/gene-s...  ...   CONSERVATION   \n",
       "4     <a href=\"https://www.genenames.org/data/gene-s...  ...   CONSERVATION   \n",
       "...                                                 ...  ...            ...   \n",
       "3584  <a href=\"https://www.genenames.org/data/gene-s...  ...  not conserved   \n",
       "3585  <a href=\"https://www.genenames.org/data/gene-s...  ...  not conserved   \n",
       "3586  <a href=\"https://www.genenames.org/data/gene-s...  ...  not conserved   \n",
       "3587  <a href=\"https://www.genenames.org/data/gene-s...  ...  not conserved   \n",
       "3588  <a href=\"https://www.genenames.org/data/gene-s...  ...  not conserved   \n",
       "\n",
       "                                                   PMID  \\\n",
       "0     <a href=\"https://comp.med.yokohama-cu.ac.jp/re...   \n",
       "1     <a href=\"https://comp.med.yokohama-cu.ac.jp/re...   \n",
       "2     <a href=\"https://comp.med.yokohama-cu.ac.jp/re...   \n",
       "3     <a href=\"https://comp.med.yokohama-cu.ac.jp/re...   \n",
       "4     <a href=\"https://comp.med.yokohama-cu.ac.jp/re...   \n",
       "...                                                 ...   \n",
       "3584  <a href=\"https://comp.med.yokohama-cu.ac.jp/re...   \n",
       "3585  <a href=\"https://comp.med.yokohama-cu.ac.jp/re...   \n",
       "3586  <a href=\"https://comp.med.yokohama-cu.ac.jp/re...   \n",
       "3587  <a href=\"https://comp.med.yokohama-cu.ac.jp/re...   \n",
       "3588  <a href=\"https://comp.med.yokohama-cu.ac.jp/re...   \n",
       "\n",
       "                                            Ligand Name  \\\n",
       "0                                 alpha-2-macroglobulin   \n",
       "1                                 alpha-2-macroglobulin   \n",
       "2                       angiotensin I converting enzyme   \n",
       "3                                   adenosine deaminase   \n",
       "4                       ADAM metallopeptidase domain 10   \n",
       "...                                                 ...   \n",
       "3584                               protocadherin beta 9   \n",
       "3585                 protocadherin gamma subfamily B, 8   \n",
       "3586                                  serum amyloid A 3   \n",
       "3587  sema domain, immunoglobulin domain (Ig), trans...   \n",
       "3588                      zona pellucida glycoprotein 3   \n",
       "\n",
       "                                          Receptor Name Mouse_ligand  \\\n",
       "0          heat shock protein family A (Hsp70) member 5          NaN   \n",
       "1                        LDL receptor related protein 1          NaN   \n",
       "2                                bradykinin receptor B2          NaN   \n",
       "3                                dipeptidyl peptidase 4          NaN   \n",
       "4                                             ephrin A5          NaN   \n",
       "...                                                 ...          ...   \n",
       "3584                               protocadherin beta 9       Pcdhb9   \n",
       "3585                 protocadherin gamma subfamily B, 8      Pcdhgb8   \n",
       "3586                               toll-like receptor 4         Saa3   \n",
       "3587  T cell immunoglobulin and mucin domain contain...       Sema4a   \n",
       "3588                          zona pellucida 3 receptor          Zp3   \n",
       "\n",
       "     Mouse_receptor Ligand MGI ID Receptor MGI ID Direct  \\\n",
       "0               NaN           NaN             NaN  Human   \n",
       "1               NaN           NaN             NaN  Human   \n",
       "2               NaN           NaN             NaN  Human   \n",
       "3               NaN           NaN             NaN          \n",
       "4               NaN           NaN             NaN          \n",
       "...             ...           ...             ...    ...   \n",
       "3584         Pcdhb9   MGI:2136744     MGI:2136744  Mouse   \n",
       "3585        Pcdhgb8   MGI:1935200     MGI:1935200  Mouse   \n",
       "3586           Tlr4     MGI:98223       MGI:96824  Mouse   \n",
       "3587          Timd2    MGI:107560     MGI:2159681  Mouse   \n",
       "3588           Zp3r     MGI:99215      MGI:104965  Mouse   \n",
       "\n",
       "                                              Conserved  \n",
       "0     Mouse, Rat, Frog, Zebrafish, Chicken, Macaque,...  \n",
       "1     Mouse, Rat, Frog, Zebrafish, Chicken, Macaque,...  \n",
       "2     Mouse, Rat, Frog, Zebrafish, Chicken, Macaque,...  \n",
       "3     Human, Mouse, Rat, Frog, Zebrafish, Chicken, M...  \n",
       "4     Human, Mouse, Rat, Frog, Zebrafish, Chicken, M...  \n",
       "...                                                 ...  \n",
       "3584                                                Rat  \n",
       "3585                                                     \n",
       "3586                                          Zebrafish  \n",
       "3587                                   Rat, Frog, Horse  \n",
       "3588                                Rat, Zebrafish, Cow  \n",
       "\n",
       "[3589 rows x 22 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair0_copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "38a835f6-9268-428b-9d0c-34d73eaea769",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>Direct</th>\n",
       "      <th>Conserved</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>Human</td>\n",
       "      <td>Mouse, Rat, Frog, Zebrafish, Chicken, Macaque,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>Human</td>\n",
       "      <td>Mouse, Rat, Frog, Zebrafish, Chicken, Macaque,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACE BDKRB2</td>\n",
       "      <td>Human</td>\n",
       "      <td>Mouse, Rat, Frog, Zebrafish, Chicken, Macaque,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADA DPP4</td>\n",
       "      <td></td>\n",
       "      <td>Human, Mouse, Rat, Frog, Zebrafish, Chicken, M...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADAM10 EFNA5</td>\n",
       "      <td></td>\n",
       "      <td>Human, Mouse, Rat, Frog, Zebrafish, Chicken, M...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3576</th>\n",
       "      <td>ZG16B TLR4</td>\n",
       "      <td>Human</td>\n",
       "      <td>Human, Mouse, Rat, Frog, Macaque, Pig, Dog, Co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3577</th>\n",
       "      <td>ZG16B TLR5</td>\n",
       "      <td>Human</td>\n",
       "      <td>Mouse, Rat, Frog, Macaque, Pig, Dog, Cow, Chim...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3578</th>\n",
       "      <td>ZG16B TLR6</td>\n",
       "      <td>Human</td>\n",
       "      <td>Human, Mouse, Rat, Frog, Macaque, Pig, Dog, Co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3579</th>\n",
       "      <td>ZP3 CHRNA7</td>\n",
       "      <td>Mouse</td>\n",
       "      <td>Human, Rat, Frog, Zebrafish, Chicken, Macaque,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3580</th>\n",
       "      <td>Zp3 Zp3r</td>\n",
       "      <td>Mouse</td>\n",
       "      <td>Rat, Zebrafish, Cow</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3581 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0     LR Pair Card Direct                                          Conserved\n",
       "0        A2M HSPA5  Human  Mouse, Rat, Frog, Zebrafish, Chicken, Macaque,...\n",
       "1         A2M LRP1  Human  Mouse, Rat, Frog, Zebrafish, Chicken, Macaque,...\n",
       "2       ACE BDKRB2  Human  Mouse, Rat, Frog, Zebrafish, Chicken, Macaque,...\n",
       "3         ADA DPP4         Human, Mouse, Rat, Frog, Zebrafish, Chicken, M...\n",
       "4     ADAM10 EFNA5         Human, Mouse, Rat, Frog, Zebrafish, Chicken, M...\n",
       "...            ...    ...                                                ...\n",
       "3576    ZG16B TLR4  Human  Human, Mouse, Rat, Frog, Macaque, Pig, Dog, Co...\n",
       "3577    ZG16B TLR5  Human  Mouse, Rat, Frog, Macaque, Pig, Dog, Cow, Chim...\n",
       "3578    ZG16B TLR6  Human  Human, Mouse, Rat, Frog, Macaque, Pig, Dog, Co...\n",
       "3579    ZP3 CHRNA7  Mouse  Human, Rat, Frog, Zebrafish, Chicken, Macaque,...\n",
       "3580      Zp3 Zp3r  Mouse                                Rat, Zebrafish, Cow\n",
       "\n",
       "[3581 rows x 3 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conservation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "81018417-640d-465b-9409-171a9d0f8709",
   "metadata": {},
   "outputs": [],
   "source": [
    "conservation.to_csv(\"data/test.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "39ad6527-d486-4c5b-a671-f0658b2a6757",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to prepare datatables (cleaning and hyperlinking, adding tool tips, etc) input for the database qmds\n",
    "import sys, os\n",
    "from itables import init_notebook_mode\n",
    "import pandas as pd\n",
    "from itables import show\n",
    "from itables import options\n",
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "import re\n",
    "from bs4 import BeautifulSoup\n",
    "from createDataTable import gene_pair, gene_pair000, human_columns, lrPairsCount\n",
    "import warnings\n",
    "import fetchGSheet \n",
    "\n",
    "# Suppress SettingWithCopyWarning\n",
    "warnings.simplefilter(\"ignore\", category=UserWarning)\n",
    "\n",
    "\n",
    "species = \"Mouse\"\n",
    "species_id = \"MGI\"\n",
    "species_lower = species.lower()\n",
    "\n",
    "def extract_visible_text(col):\n",
    "    \"\"\"Extract visible text between '>' and '</a>'.\"\"\"\n",
    "    match = re.search(r'>([^<]+)</a>', col)\n",
    "    if match:\n",
    "        return match.group(1).strip()\n",
    "    return None\n",
    "\n",
    "# Grab the Interaction ID and \n",
    "gene_pair = gene_pair.iloc[:, :9]\n",
    "# Next, drop columns at index positions 3 and 4 (\"Ligand and Receptor\" since we already have ligand symbols and receptor symbols)\n",
    "gene_pair = gene_pair.drop(gene_pair.columns[[3, 4]], axis=1)\n",
    "exclude_keywords = [\"HGNC ID\", \"Location\", \"Human\"]  # Columns containing this will not be modified\n",
    "keywords_to_modify = [\"Ligand Symbols\", \"Receptor Symbols\"]\n",
    "# Copy the original columns so we can modify only the first 10\n",
    "new_columns = gene_pair.columns.tolist()\n",
    "\n",
    "# Modify only the first 10 columns\n",
    "new_columns = [\n",
    "    f'{col.split(\">\")[0]}\">Human {col.split(\">\")[1]}</span>'\n",
    "    if any(keyword in col for keyword in keywords_to_modify) and not any(exclude in col for exclude in exclude_keywords)\n",
    "    else col\n",
    "    for col in new_columns\n",
    "]\n",
    "# Assign the modified column names back to the DataFrame\n",
    "gene_pair.columns = new_columns\n",
    "\n",
    "gene_pair[\"LR Pair Card\"] = gene_pair.iloc[:, 1].apply(extract_visible_text)\n",
    "ligand_symbols_col = [col for col in gene_pair.columns if \"Ligand Symbols\" in col][0]\n",
    "receptor_symbols_col = [col for col in gene_pair.columns if \"Receptor Symbols\" in col][0]\n",
    "gene_pair = gene_pair.rename(columns={ligand_symbols_col: \"Human Ligand Symbols\",\n",
    "                                      receptor_symbols_col: \"Human Receptor Symbols\"})\n",
    "\n",
    "gene_pair_species = getattr(fetchGSheet, f\"gene_pair_{species_lower}\")  \n",
    "gene_pair_species = gene_pair_species[[\n",
    "    \"LR Pair Card\",\n",
    "    f\"{species}_ligand\",\n",
    "    f\"{species}_receptor\",\n",
    "    f\"{species} evidence\",\n",
    "    f\"{species_id} ligand\",\n",
    "    f\"{species_id} receptor\",\n",
    "    \"PMID\"\n",
    "]]\n",
    "\n",
    "\n",
    "gene_pair =gene_pair_species.merge(gene_pair,how=\"left\", on=\"LR Pair Card\")\n",
    "species_info = pd.read_csv(f\"data/MRK_Merged_{species_id}_DB.tsv\", sep=\"\\t\", dtype=str)\n",
    "species_info = species_info[[f\"{species_id} Marker Accession ID\", 'Marker Name','Aliases']]\n",
    "species_info[\"Aliases\"] = species_info[\"Aliases\"].str.replace(\"|\", \", \", regex=False)\n",
    "gene_pair = gene_pair.merge(species_info,how=\"left\", left_on = f\"{species_id} ligand\",right_on=f\"{species_id} Marker Accession ID\")\n",
    "gene_pair = gene_pair.drop(columns=[f\"{species_id} Marker Accession ID\"])\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "                                      \"Marker Name\": \"Ligand Name\",\n",
    "                                      \"Aliases\": \"Ligand Symbols\"\n",
    "                                     }\n",
    "                            )\n",
    "gene_pair = gene_pair.merge(\n",
    "    species_info,\n",
    "    how=\"left\",\n",
    "    left_on=f\"{species_id} receptor\",\n",
    "    right_on=f\"{species_id} Marker Accession ID\"\n",
    ")\n",
    "\n",
    "gene_pair = gene_pair.drop(columns=[f\"{species_id} Marker Accession ID\"])\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "                                      \"Marker Name\": \"Receptor Name\",\n",
    "                                      \"Aliases\": \"Receptor Symbols\"\n",
    "                                     }\n",
    "                            )\n",
    "\n",
    "gene_pair[f\"{species} LR Pair\"] = np.where(\n",
    "    gene_pair[f\"{species} evidence\"] == \"not conserved\", \n",
    "    f\"no {species_lower} ortholog\",                                  \n",
    "    gene_pair[f\"{species}_ligand\"] + \" \" + gene_pair[f\"{species}_receptor\"] \n",
    ")\n",
    "\n",
    "gene_pair = gene_pair[~(gene_pair[f\"{species} evidence\"] == \"not conserved\")]\n",
    "def format_symbol_aliases(symbol, aliases):\n",
    "    \"\"\"\n",
    "    Formats symbol, old symbols, and aliases.\n",
    "    If the final formatted string would be empty after considering N/A values\n",
    "    and empty inputs, it returns \"species-specific\".\n",
    "    Otherwise, it formats based on the presence of old_symbol and aliases,\n",
    "    removing unnecessary parentheses or commas, following the structure:\n",
    "    \"Symbol (Old Symbol, Aliases)\" if both exist.\n",
    "    \"\"\"\n",
    "    # Normalize inputs to empty strings if they are None/NaN or just whitespace\n",
    "    symbol_str = str(symbol).strip()\n",
    "    # old_symbol_str = str(old_symbol).strip()\n",
    "    aliases_str = str(aliases).strip()\n",
    "\n",
    "    # Filter out values that are empty strings or \"N/A\" for old_symbol and aliases\n",
    "    parts_for_join = []\n",
    "    # if old_symbol_str and old_symbol_str != \"N/A\":\n",
    "    #     parts_for_join.append(old_symbol_str)\n",
    "    if aliases_str and aliases_str != \"N/A\":\n",
    "        parts_for_join.append(aliases_str)\n",
    "\n",
    "    # Construct the preliminary result based on your original logic:\n",
    "    # \"symbol (old_symbol, aliases)\" if parts_for_join is not empty, else \"symbol\"\n",
    "    if parts_for_join:\n",
    "        prelim_result = f\"{symbol_str} ({', '.join(parts_for_join)})\"\n",
    "    else:\n",
    "        prelim_result = symbol_str # Just the symbol if no old_symbol or aliases\n",
    "\n",
    "    return prelim_result\n",
    "\n",
    "# This is crucial for consistent handling by the function before processing \"N/A\".\n",
    "gene_pair[f\"{species}_ligand\"] = gene_pair[f\"{species}_ligand\"].fillna('')\n",
    "gene_pair['Ligand Symbols'] = gene_pair['Ligand Symbols'].fillna('')\n",
    "\n",
    "gene_pair['Ligand Symbols'] = gene_pair.apply(\n",
    "    lambda row: format_symbol_aliases(row[f\"{species}_ligand\"], row['Ligand Symbols']),\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "gene_pair[f\"{species}_receptor\"] = gene_pair[f\"{species}_receptor\"].fillna('')\n",
    "gene_pair['Receptor Symbols'] = gene_pair['Receptor Symbols'].fillna('')\n",
    "\n",
    "gene_pair['Receptor Symbols'] = gene_pair.apply(\n",
    "    lambda row: format_symbol_aliases(row[f\"{species}_receptor\"], row['Receptor Symbols']),\n",
    "    axis=1\n",
    ")\n",
    "# The list of columns to group by\n",
    "interaction_id_col = [col for col in gene_pair.columns if \"Interaction ID\" in col][0]\n",
    "\n",
    "grouping_cols = [\n",
    "    interaction_id_col, f\"{species} LR Pair\", \"Ligand Symbols\", \"Receptor Symbols\" \n",
    "]\n",
    "\n",
    "aggregation_cols = [\n",
    "    col for col in gene_pair.columns if col not in grouping_cols\n",
    "]\n",
    "\n",
    "# 3. Create a dictionary mapping each aggregation column to the joining function\n",
    "agg_dict = {\n",
    "    col: lambda x: ', '.join(x.astype(str).unique()) for col in aggregation_cols\n",
    "}\n",
    "\n",
    "# 4. Perform the groupby and aggregation\n",
    "gene_pair = gene_pair.groupby(grouping_cols).agg(agg_dict).reset_index()\n",
    "\n",
    "def generate_perplexity_link_pmid(row, species, species_lower): \n",
    "    query = (\n",
    "        f\"What-is-the-biological-relevance-of-the-ligand-and-receptor-pair-\"\n",
    "        f\"{row[f'{species} LR Pair']}-based-on-Pubmed-ID-\"\n",
    "        f\"{row['PMID']}-in-{species_lower}\"\n",
    "    )\n",
    "    return (\n",
    "         f'<a href=\"https://www.perplexity.ai/search?q={query}\" target=\"_blank\" style=\"text-decoration: none;\">&#128269;</a>'\n",
    "    )\n",
    "\n",
    "\n",
    "# Apply function to the DataFrame\n",
    "gene_pair[\"Perplexity\"] = gene_pair.apply(\n",
    "    generate_perplexity_link_pmid, axis=1, args=(species, species_lower)\n",
    ")\n",
    "\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "                                      f\"{species_id} ligand\": f\"Ligand {species_id} ID\",\n",
    "                                      f\"{species_id} receptor\": f\"Receptor {species_id} ID\"\n",
    "                                     }\n",
    "                            )\n",
    "ligand_loc_col = [col for col in gene_pair.columns if \"Ligand Location\" in col][0]\n",
    "receptor_loc_col = [col for col in gene_pair.columns if \"Receptor Location\" in col][0]\n",
    "lr_pair_card = [col for col in gene_pair.columns if \">LR Pair Card\" in col][0]\n",
    "# gene_pair.columns\n",
    "gene_pair[\"LR Pair Card\"] = gene_pair[lr_pair_card]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "196f40a9-83cd-411e-af44-05ee2371686a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;span title=\"Double-click header of Interaction ID to ensure all values are shown\"&gt;Interaction ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>Mouse LR Pair</th>\n",
       "      <th>Ligand Symbols</th>\n",
       "      <th>Receptor Symbols</th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>Mouse_ligand</th>\n",
       "      <th>Mouse_receptor</th>\n",
       "      <th>Mouse evidence</th>\n",
       "      <th>Ligand MGI ID</th>\n",
       "      <th>Receptor MGI ID</th>\n",
       "      <th>PMID</th>\n",
       "      <th>&lt;span title=\"Double-click header of LR Pair Card to ensure all values are shown\"&gt;LR Pair Card&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\"&gt;Human LR Pair&lt;/span&gt;</th>\n",
       "      <th>Human Ligand Symbols</th>\n",
       "      <th>Human Receptor Symbols</th>\n",
       "      <th>&lt;span title=\"Location based on the predicted subcellular localization of the human proteome\"&gt;Ligand Location&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Location based on the predicted subcellular localization of the human proteome\"&gt;Receptor Location&lt;/span&gt;</th>\n",
       "      <th>Ligand Name</th>\n",
       "      <th>Receptor Name</th>\n",
       "      <th>Perplexity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDB00001</td>\n",
       "      <td>A2m Hspa5</td>\n",
       "      <td>A2m (A2mp)</td>\n",
       "      <td>Hspa5 (Hsce70, Bip, Grp78, Sez7, D2Wsu141e, D2...</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>A2m</td>\n",
       "      <td>Hspa5</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>MGI:2449119</td>\n",
       "      <td>MGI:95835</td>\n",
       "      <td>12194978, 32541810</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>&lt;span title=\"A2M (FWP007, S863-7, CPAMD5)\"&gt;A2M...</td>\n",
       "      <td>&lt;span title=\"HSPA5 (GRP78, BiP)\"&gt;HSPA5 (GRP78,...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;span title=\"based on PMID: 12194978, PMID: 32...</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>heat shock protein 5</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDB00002</td>\n",
       "      <td>A2m Lrp1</td>\n",
       "      <td>A2m (A2mp)</td>\n",
       "      <td>Lrp1 (CD91, A2mr, b2b1554Clo)</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>A2m</td>\n",
       "      <td>Lrp1</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>MGI:2449119</td>\n",
       "      <td>MGI:96828</td>\n",
       "      <td>10652313, 12194978, 1702392</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>&lt;span title=\"A2M (FWP007, S863-7, CPAMD5)\"&gt;A2M...</td>\n",
       "      <td>&lt;span title=\"LRP1 (APR, A2MR, LRP, CD91, LRP1A...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>low density lipoprotein receptor-related prote...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDB00003</td>\n",
       "      <td>Ace Bdkrb2</td>\n",
       "      <td>Ace (CD143)</td>\n",
       "      <td>Bdkrb2 (B2, B(2), kinin B2, BK2R, B2R)</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Ace</td>\n",
       "      <td>Bdkrb2</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>MGI:87874</td>\n",
       "      <td>MGI:102845</td>\n",
       "      <td>10748135</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>ACE BDKRB2</td>\n",
       "      <td>&lt;span title=\"ACE (DCP1, ACE1, CD143)\"&gt;ACE (DCP...</td>\n",
       "      <td>&lt;span title=\"BDKRB2 (BK-2)\"&gt;BDKRB2 (BK-2)&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>angiotensin I converting enzyme</td>\n",
       "      <td>bradykinin receptor, beta 2</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDB00004</td>\n",
       "      <td>Ada Dpp4</td>\n",
       "      <td>Ada</td>\n",
       "      <td>Dpp4 (THAM, Dpp-4, Cd26)</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Ada</td>\n",
       "      <td>Dpp4</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>MGI:87916</td>\n",
       "      <td>MGI:94919</td>\n",
       "      <td>15213224</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>ADA DPP4</td>\n",
       "      <td>&lt;span title=\"ADA (ADA1)\"&gt;ADA (ADA1)&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"DPP4 (CD26, ADCP2, DPPIV)\"&gt;DPP4 (...</td>\n",
       "      <td>&lt;span title=\"based on hpa, uniprot\"&gt;cell membr...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>adenosine deaminase</td>\n",
       "      <td>dipeptidylpeptidase 4</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDB00005</td>\n",
       "      <td>Adam10 Efna5</td>\n",
       "      <td>Adam10 (kuz, kuzbanian, 1700031C13Rik)</td>\n",
       "      <td>Efna5 (AL-1, RAGS, Ephrin-A5, Epl7, EFL-5, LER...</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Adam10</td>\n",
       "      <td>Efna5</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>MGI:109548</td>\n",
       "      <td>MGI:107444</td>\n",
       "      <td>16239146</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>ADAM10 EFNA5</td>\n",
       "      <td>&lt;span title=\"ADAM10 (kuz, MADM, HsT18717, CD15...</td>\n",
       "      <td>&lt;span title=\"EFNA5 (EPLG7, AF1, LERK7)\"&gt;EFNA5 ...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>a disintegrin and metallopeptidase domain 10</td>\n",
       "      <td>ephrin A5</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3251</th>\n",
       "      <td>CDB03577</td>\n",
       "      <td>Pcdhb9 Pcdhb9</td>\n",
       "      <td>Pcdhb9 (Pcdhb4C, PcdhbI)</td>\n",
       "      <td>Pcdhb9 (Pcdhb4C, PcdhbI)</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Pcdhb9</td>\n",
       "      <td>Pcdhb9</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>MGI:2136744</td>\n",
       "      <td>MGI:2136744</td>\n",
       "      <td>25171406</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>protocadherin beta 9</td>\n",
       "      <td>protocadherin beta 9</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3252</th>\n",
       "      <td>CDB03578</td>\n",
       "      <td>Pcdhgb8 Pcdhgb8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>MGI:1935200</td>\n",
       "      <td>MGI:1935200</td>\n",
       "      <td>25171406</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>protocadherin gamma subfamily B, 8</td>\n",
       "      <td>protocadherin gamma subfamily B, 8</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3253</th>\n",
       "      <td>CDB03579</td>\n",
       "      <td>Saa3 Tlr4</td>\n",
       "      <td>Saa3 (Saa-3, l7R3)</td>\n",
       "      <td>Tlr4 (Lps, Rasl2-8, lipopolysaccharide response)</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Saa3</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>MGI:98223</td>\n",
       "      <td>MGI:96824</td>\n",
       "      <td>23858030</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"TLR4 (hToll, CD284, TLR-4, ARMD10...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>serum amyloid A 3</td>\n",
       "      <td>toll-like receptor 4</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3254</th>\n",
       "      <td>CDB03580</td>\n",
       "      <td>Sema4a Timd2</td>\n",
       "      <td>Sema4a (SemB, SemB, Semab)</td>\n",
       "      <td>Timd2 (Tim2, TIM-2)</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Sema4a</td>\n",
       "      <td>Timd2</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>MGI:107560</td>\n",
       "      <td>MGI:2159681</td>\n",
       "      <td>12374982</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>&lt;span title=\"SEMA4A (SEMAB, SemB, FLJ12287, CO...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>sema domain, immunoglobulin domain (Ig), trans...</td>\n",
       "      <td>T cell immunoglobulin and mucin domain contain...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3255</th>\n",
       "      <td>CDB03581</td>\n",
       "      <td>Zp3 Zp3r</td>\n",
       "      <td>Zp3 (Zp-3)</td>\n",
       "      <td>Zp3r (SP56)</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Zp3</td>\n",
       "      <td>Zp3r</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>MGI:99215</td>\n",
       "      <td>MGI:104965</td>\n",
       "      <td>8188752</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>&lt;span title=\"ZP3 (ZP3A, ZP3B, ZP3-424, ZP3-372...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;span title=\"based on PMID: 8188752\"&gt;cell memb...</td>\n",
       "      <td>zona pellucida glycoprotein 3</td>\n",
       "      <td>zona pellucida 3 receptor</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3256 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     <span title=\"Double-click header of Interaction ID to ensure all values are shown\">Interaction ID&nbsp;</span>  \\\n",
       "0                                              CDB00001                                                               \n",
       "1                                              CDB00002                                                               \n",
       "2                                              CDB00003                                                               \n",
       "3                                              CDB00004                                                               \n",
       "4                                              CDB00005                                                               \n",
       "...                                                 ...                                                               \n",
       "3251                                           CDB03577                                                               \n",
       "3252                                           CDB03578                                                               \n",
       "3253                                           CDB03579                                                               \n",
       "3254                                           CDB03580                                                               \n",
       "3255                                           CDB03581                                                               \n",
       "\n",
       "        Mouse LR Pair                          Ligand Symbols  \\\n",
       "0           A2m Hspa5                              A2m (A2mp)   \n",
       "1            A2m Lrp1                              A2m (A2mp)   \n",
       "2          Ace Bdkrb2                             Ace (CD143)   \n",
       "3            Ada Dpp4                                     Ada   \n",
       "4        Adam10 Efna5  Adam10 (kuz, kuzbanian, 1700031C13Rik)   \n",
       "...               ...                                     ...   \n",
       "3251    Pcdhb9 Pcdhb9                Pcdhb9 (Pcdhb4C, PcdhbI)   \n",
       "3252  Pcdhgb8 Pcdhgb8                                 Pcdhgb8   \n",
       "3253        Saa3 Tlr4                      Saa3 (Saa-3, l7R3)   \n",
       "3254     Sema4a Timd2              Sema4a (SemB, SemB, Semab)   \n",
       "3255         Zp3 Zp3r                              Zp3 (Zp-3)   \n",
       "\n",
       "                                       Receptor Symbols  \\\n",
       "0     Hspa5 (Hsce70, Bip, Grp78, Sez7, D2Wsu141e, D2...   \n",
       "1                         Lrp1 (CD91, A2mr, b2b1554Clo)   \n",
       "2                Bdkrb2 (B2, B(2), kinin B2, BK2R, B2R)   \n",
       "3                              Dpp4 (THAM, Dpp-4, Cd26)   \n",
       "4     Efna5 (AL-1, RAGS, Ephrin-A5, Epl7, EFL-5, LER...   \n",
       "...                                                 ...   \n",
       "3251                           Pcdhb9 (Pcdhb4C, PcdhbI)   \n",
       "3252                                            Pcdhgb8   \n",
       "3253   Tlr4 (Lps, Rasl2-8, lipopolysaccharide response)   \n",
       "3254                                Timd2 (Tim2, TIM-2)   \n",
       "3255                                        Zp3r (SP56)   \n",
       "\n",
       "                                           LR Pair Card Mouse_ligand  \\\n",
       "0     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...          A2m   \n",
       "1     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...          A2m   \n",
       "2     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...          Ace   \n",
       "3     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...          Ada   \n",
       "4     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...       Adam10   \n",
       "...                                                 ...          ...   \n",
       "3251  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...       Pcdhb9   \n",
       "3252  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...      Pcdhgb8   \n",
       "3253  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...         Saa3   \n",
       "3254  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...       Sema4a   \n",
       "3255  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...          Zp3   \n",
       "\n",
       "     Mouse_receptor Mouse evidence Ligand MGI ID Receptor MGI ID  \\\n",
       "0             Hspa5   CONSERVATION   MGI:2449119       MGI:95835   \n",
       "1              Lrp1   CONSERVATION   MGI:2449119       MGI:96828   \n",
       "2            Bdkrb2   CONSERVATION     MGI:87874      MGI:102845   \n",
       "3              Dpp4   CONSERVATION     MGI:87916       MGI:94919   \n",
       "4             Efna5   CONSERVATION    MGI:109548      MGI:107444   \n",
       "...             ...            ...           ...             ...   \n",
       "3251         Pcdhb9         DIRECT   MGI:2136744     MGI:2136744   \n",
       "3252        Pcdhgb8         DIRECT   MGI:1935200     MGI:1935200   \n",
       "3253           Tlr4         DIRECT     MGI:98223       MGI:96824   \n",
       "3254          Timd2         DIRECT    MGI:107560     MGI:2159681   \n",
       "3255           Zp3r         DIRECT     MGI:99215      MGI:104965   \n",
       "\n",
       "                             PMID  \\\n",
       "0              12194978, 32541810   \n",
       "1     10652313, 12194978, 1702392   \n",
       "2                        10748135   \n",
       "3                        15213224   \n",
       "4                        16239146   \n",
       "...                           ...   \n",
       "3251                     25171406   \n",
       "3252                     25171406   \n",
       "3253                     23858030   \n",
       "3254                     12374982   \n",
       "3255                      8188752   \n",
       "\n",
       "     <span title=\"Double-click header of LR Pair Card to ensure all values are shown\">LR Pair Card&nbsp;</span>  \\\n",
       "0     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                           \n",
       "1     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                           \n",
       "2     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                           \n",
       "3     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                           \n",
       "4     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                           \n",
       "...                                                 ...                                                           \n",
       "3251  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                           \n",
       "3252  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                           \n",
       "3253  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                           \n",
       "3254  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                           \n",
       "3255  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                           \n",
       "\n",
       "     <span title=\"Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\">Human LR Pair</span>  \\\n",
       "0                                             A2M HSPA5                                                               \n",
       "1                                              A2M LRP1                                                               \n",
       "2                                            ACE BDKRB2                                                               \n",
       "3                                              ADA DPP4                                                               \n",
       "4                                          ADAM10 EFNA5                                                               \n",
       "...                                                 ...                                                               \n",
       "3251                                  no human ortholog                                                               \n",
       "3252                                  no human ortholog                                                               \n",
       "3253                                  no human ortholog                                                               \n",
       "3254                                  no human ortholog                                                               \n",
       "3255                                  no human ortholog                                                               \n",
       "\n",
       "                                   Human Ligand Symbols  \\\n",
       "0     <span title=\"A2M (FWP007, S863-7, CPAMD5)\">A2M...   \n",
       "1     <span title=\"A2M (FWP007, S863-7, CPAMD5)\">A2M...   \n",
       "2     <span title=\"ACE (DCP1, ACE1, CD143)\">ACE (DCP...   \n",
       "3            <span title=\"ADA (ADA1)\">ADA (ADA1)</span>   \n",
       "4     <span title=\"ADAM10 (kuz, MADM, HsT18717, CD15...   \n",
       "...                                                 ...   \n",
       "3251  <span title=\"no human ortholog\">no human ortho...   \n",
       "3252  <span title=\"no human ortholog\">no human ortho...   \n",
       "3253  <span title=\"no human ortholog\">no human ortho...   \n",
       "3254  <span title=\"SEMA4A (SEMAB, SemB, FLJ12287, CO...   \n",
       "3255  <span title=\"ZP3 (ZP3A, ZP3B, ZP3-424, ZP3-372...   \n",
       "\n",
       "                                 Human Receptor Symbols  \\\n",
       "0     <span title=\"HSPA5 (GRP78, BiP)\">HSPA5 (GRP78,...   \n",
       "1     <span title=\"LRP1 (APR, A2MR, LRP, CD91, LRP1A...   \n",
       "2      <span title=\"BDKRB2 (BK-2)\">BDKRB2 (BK-2)</span>   \n",
       "3     <span title=\"DPP4 (CD26, ADCP2, DPPIV)\">DPP4 (...   \n",
       "4     <span title=\"EFNA5 (EPLG7, AF1, LERK7)\">EFNA5 ...   \n",
       "...                                                 ...   \n",
       "3251  <span title=\"no human ortholog\">no human ortho...   \n",
       "3252  <span title=\"no human ortholog\">no human ortho...   \n",
       "3253  <span title=\"TLR4 (hToll, CD284, TLR-4, ARMD10...   \n",
       "3254  <span title=\"no human ortholog\">no human ortho...   \n",
       "3255  <span title=\"no human ortholog\">no human ortho...   \n",
       "\n",
       "     <span title=\"Location based on the predicted subcellular localization of the human proteome\">Ligand Location</span>  \\\n",
       "0     <span title=\"based on perplexity, uniprot\">sec...                                                                    \n",
       "1     <span title=\"based on perplexity, uniprot\">sec...                                                                    \n",
       "2     <span title=\"based on perplexity, uniprot\">cel...                                                                    \n",
       "3     <span title=\"based on hpa, uniprot\">cell membr...                                                                    \n",
       "4     <span title=\"based on hpa, perplexity, uniprot...                                                                    \n",
       "...                                                 ...                                                                    \n",
       "3251  <span title=\"based on uniprot\">cell membrane</...                                                                    \n",
       "3252  <span title=\"based on perplexity, uniprot\">cel...                                                                    \n",
       "3253  <span title=\"based on perplexity, uniprot\">sec...                                                                    \n",
       "3254  <span title=\"based on perplexity, uniprot\">cel...                                                                    \n",
       "3255  <span title=\"based on uniprot\">cell membrane</...                                                                    \n",
       "\n",
       "     <span title=\"Location based on the predicted subcellular localization of the human proteome\">Receptor Location</span>  \\\n",
       "0     <span title=\"based on PMID: 12194978, PMID: 32...                                                                      \n",
       "1     <span title=\"based on hpa, perplexity, uniprot...                                                                      \n",
       "2     <span title=\"based on hpa, perplexity, uniprot...                                                                      \n",
       "3     <span title=\"based on perplexity, uniprot\">cel...                                                                      \n",
       "4     <span title=\"based on perplexity, uniprot\">cel...                                                                      \n",
       "...                                                 ...                                                                      \n",
       "3251  <span title=\"based on uniprot\">cell membrane</...                                                                      \n",
       "3252  <span title=\"based on perplexity, uniprot\">cel...                                                                      \n",
       "3253  <span title=\"based on perplexity, uniprot\">cel...                                                                      \n",
       "3254  <span title=\"based on perplexity, uniprot\">cel...                                                                      \n",
       "3255  <span title=\"based on PMID: 8188752\">cell memb...                                                                      \n",
       "\n",
       "                                            Ligand Name  \\\n",
       "0                                 alpha-2-macroglobulin   \n",
       "1                                 alpha-2-macroglobulin   \n",
       "2                       angiotensin I converting enzyme   \n",
       "3                                   adenosine deaminase   \n",
       "4          a disintegrin and metallopeptidase domain 10   \n",
       "...                                                 ...   \n",
       "3251                               protocadherin beta 9   \n",
       "3252                 protocadherin gamma subfamily B, 8   \n",
       "3253                                  serum amyloid A 3   \n",
       "3254  sema domain, immunoglobulin domain (Ig), trans...   \n",
       "3255                      zona pellucida glycoprotein 3   \n",
       "\n",
       "                                          Receptor Name  \\\n",
       "0                                  heat shock protein 5   \n",
       "1     low density lipoprotein receptor-related prote...   \n",
       "2                           bradykinin receptor, beta 2   \n",
       "3                                 dipeptidylpeptidase 4   \n",
       "4                                             ephrin A5   \n",
       "...                                                 ...   \n",
       "3251                               protocadherin beta 9   \n",
       "3252                 protocadherin gamma subfamily B, 8   \n",
       "3253                               toll-like receptor 4   \n",
       "3254  T cell immunoglobulin and mucin domain contain...   \n",
       "3255                          zona pellucida 3 receptor   \n",
       "\n",
       "                                             Perplexity  \n",
       "0     <a href=\"https://www.perplexity.ai/search?q=Wh...  \n",
       "1     <a href=\"https://www.perplexity.ai/search?q=Wh...  \n",
       "2     <a href=\"https://www.perplexity.ai/search?q=Wh...  \n",
       "3     <a href=\"https://www.perplexity.ai/search?q=Wh...  \n",
       "4     <a href=\"https://www.perplexity.ai/search?q=Wh...  \n",
       "...                                                 ...  \n",
       "3251  <a href=\"https://www.perplexity.ai/search?q=Wh...  \n",
       "3252  <a href=\"https://www.perplexity.ai/search?q=Wh...  \n",
       "3253  <a href=\"https://www.perplexity.ai/search?q=Wh...  \n",
       "3254  <a href=\"https://www.perplexity.ai/search?q=Wh...  \n",
       "3255  <a href=\"https://www.perplexity.ai/search?q=Wh...  \n",
       "\n",
       "[3256 rows x 20 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "facd8ef3-2b9d-46e3-aecd-ecda8afb7e72",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_89560/2699435364.py:28: DtypeWarning: Columns (32,34,38,40,50) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  pop_up_info = pd.read_table(\"data/HGNC_gene_info_full.tsv\")\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['Sorting order', 'LR_pair_orig', 'PMID', 'lig_species', 'rec_species',\n",
       "       'ligand_orig', 'receptor_orig', 'Pair_species', 'LR Pair Card',\n",
       "       'Human evidence', 'Homo sapiens_ligand', 'Homo sapiens_receptor',\n",
       "       'ENSEMBL ligand', 'ENSEMBL receptor', 'HGNC ligand', 'HGNC receptor'],\n",
       "      dtype='object', name=0)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Function to prepare datatables (cleaning and hyperlinking, adding tool tips, etc) input for the database qmds\n",
    "import sys, os\n",
    "from itables import init_notebook_mode\n",
    "import pandas as pd\n",
    "from itables import show\n",
    "from itables import options\n",
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "import fetchGSheet \n",
    "import warnings\n",
    "import urllib.parse\n",
    "\n",
    "# Suppress SettingWithCopyWarning\n",
    "warnings.simplefilter(\"ignore\", category=UserWarning)\n",
    "\n",
    "\n",
    "# Other vertebrates\n",
    "species_list = [\n",
    "    \"mmusculus\", \"rnorvegicus\", \"drerio\", \"ptroglodytes\", \"ggallus\", \"sscrofa\", \"btaurus\", \n",
    "    \"clfamiliaris\", \"ecaballus\", \"oarambouillet\",\n",
    "    \"cjacchus\", \"mmulatta\", \"xtropicalis\"\n",
    "]\n",
    "\n",
    "# Select only the relevant columns from pop_up_info\n",
    "cols_to_keep = cols_to_keep = list(range(0, 30)) \n",
    "# Step 3: Load file using only the desired columns\n",
    "df = pd.read_table(\"data/HGNC_gene_info_full.tsv\", usecols=cols_to_keep)\n",
    "pop_up_info = pd.read_table(\"data/HGNC_gene_info_full.tsv\")\n",
    "pop_up_info = pop_up_info.rename(columns={\"hgnc_id\": \"HGNC ID\", \n",
    "                                          \"name\": \"Approved name\",\n",
    "                                          \"symbol\": \"Approved symbol\",\n",
    "                                          #\"rgd_id\": \"RGD ID\",\n",
    "                                          #\"mgd_id\": \"MGI ID\", \n",
    "                                          \"alias_symbol\": \"Alias symbol\", # add to table\n",
    "                                          \"prev_symbol\": \"Previous symbol\", # add to table\n",
    "                                          \"date_symbol_changed\": \"Date symbol changed\"\n",
    "                                         })\n",
    "\n",
    "# Keep only first MGI/RGD ID\n",
    "#pop_up_info[\"MGI ID\"] = pop_up_info[\"MGI ID\"].str.split(\"|\").str[0]\n",
    "#pop_up_info[\"RGD ID\"] = pop_up_info[\"RGD ID\"].str.split(\"|\").str[0]\n",
    "\n",
    "pop_up_info[\"Alias symbol\"] = pop_up_info[\"Alias symbol\"].apply(\n",
    "    lambda x: \"N/A\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\"] else x\n",
    ")\n",
    "\n",
    "pop_up_info[\"Previous symbol\"] = pop_up_info[\"Previous symbol\"].apply(\n",
    "    lambda x: \"N/A\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\"] else x\n",
    ")\n",
    "\n",
    "# Replace \"|\" with \", \"\n",
    "pop_up_info[\"Alias symbol\"] = [value.replace(\"|\", \", \") for value in pop_up_info[\"Alias symbol\"]]\n",
    "pop_up_info[\"Previous symbol\"] = [value.replace(\"|\", \", \") for value in pop_up_info[\"Previous symbol\"]]\n",
    "\n",
    "pop_up_info[\"Date symbol changed\"] = pop_up_info[\"Date symbol changed\"].apply(\n",
    "    lambda x: \"N/A\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\"] else x\n",
    ")\n",
    "\n",
    "\n",
    "pop_up_info_lim = pop_up_info[[\"HGNC ID\", \"Approved name\", \"Alias symbol\", # \"MGI ID\", \"RGD ID\"\n",
    "                               \"Approved symbol\", \"Previous symbol\"]] # rm \"Approved symbol\" for now\n",
    "pop_up_info_lim = pop_up_info_lim.drop_duplicates(subset=\"HGNC ID\", keep=\"first\")\n",
    "\n",
    "# Drop columns where all values are NA in gene_pair\n",
    "gene_pair = fetchGSheet.gene_pair_human.dropna(axis=1, how='all')\n",
    "gene_pair.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2394acfb-8709-4eb2-b7ff-3d476881460d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair = gene_pair[gene_pair['LR_pair_orig'] != '']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "f0c0b876-6759-46e4-8c71-d12acb07abdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import sys\n",
    "import pandas as pd\n",
    "import re\n",
    "import html\n",
    "import json\n",
    "from jinja2 import Environment, FileSystemLoader\n",
    "\n",
    "# === Import from createDataTable.py\n",
    "sys.path.append(\"src\")\n",
    "import createDataTable_perSpecies\n",
    "\n",
    "# === Paths (relative to project/)\n",
    "species = \"mouse\"\n",
    "# additional placeholder for search\n",
    "species_addl_search = \"Epha3, 5430401F13Rik, no human ortholog\"\n",
    "output_json = Path(f\"JSON/{species}_gene_pair.json\")  \n",
    "qmd_template = Path(f\"database/qmd_template/{species}Orth_template.qmd\") \n",
    "qmd_output = Path(f\"database/{species}Orth.qmd\")  \n",
    "template_dir = \"HTML\"\n",
    "template_name = \"datatableOrth_template.html\"\n",
    "\n",
    "# === Create output directories if needed\n",
    "output_json.parent.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "# === Clean column names and Save DataFrame to JSON\n",
    "def clean_column_names_and_generate_metadata(df):\n",
    "    def visible_text(html_string):\n",
    "        return re.sub(r'<[^>]*>', '', html.unescape(html_string)).strip()\n",
    "    \n",
    "    raw_columns = df.columns.tolist()\n",
    "    visible_columns = [visible_text(col) for col in raw_columns]\n",
    "    \n",
    "    df_cleaned = df.copy()\n",
    "    df_cleaned.columns = visible_columns\n",
    "    \n",
    "    column_metadata = [\n",
    "        {\n",
    "            \"data\": visible,\n",
    "            \"title\": html_col\n",
    "        }\n",
    "        for visible, html_col in zip(visible_columns, raw_columns)\n",
    "    ]\n",
    "    return df_cleaned, column_metadata\n",
    "\n",
    "# === Generate DataTables column definitions\n",
    "# Use getattr to dynamically access the attribute\n",
    "gene_pair = getattr(createDataTable_perSpecies, f\"{species}_gene_pair1\")  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "699a9fe0-0fc9-4443-855f-580cb9450d47",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cleaned, columns_metadata = clean_column_names_and_generate_metadata(gene_pair)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "c656e75b-12bc-4d8d-8b95-5fd247b7c0e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Interaction ID', 'LR Pair Card', 'Mouse LR Pair', 'Ligand Symbols',\n",
       "       'Receptor Symbols', 'Ligand Location', 'Receptor Location',\n",
       "       'Ligand MGI ID', 'Receptor MGI ID', 'Perplexity', 'Mouse evidence',\n",
       "       'Human Ligand Symbols', 'Human Receptor Symbols'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cleaned.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "acc4988c-6d81-463a-8912-7863ccd26627",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;span title=\"Double-click header of Interaction ID to ensure all values are shown\"&gt;Interaction ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>Mouse LR Pair</th>\n",
       "      <th>Ligand Symbols</th>\n",
       "      <th>Receptor Symbols</th>\n",
       "      <th>&lt;span title=\"Location based on the predicted subcellular localization of the human proteome\"&gt;Ligand Location&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Location based on the predicted subcellular localization of the human proteome\"&gt;Receptor Location&lt;/span&gt;</th>\n",
       "      <th>Ligand MGI ID</th>\n",
       "      <th>Receptor MGI ID</th>\n",
       "      <th>Perplexity</th>\n",
       "      <th>Mouse evidence</th>\n",
       "      <th>Human Ligand Symbols</th>\n",
       "      <th>Human Receptor Symbols</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDB00001</td>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>A2m Hspa5</td>\n",
       "      <td>A2m (A2mp)</td>\n",
       "      <td>Hspa5 (Hsce70, Bip, Grp78, Sez7, D2Wsu141e, D2...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;span title=\"based on PMID: 12194978, PMID: 32...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>&lt;span title=\"A2M (FWP007, S863-7, CPAMD5)\"&gt;A2M...</td>\n",
       "      <td>&lt;span title=\"HSPA5 (GRP78, BiP)\"&gt;HSPA5 (GRP78,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDB00002</td>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>A2m Lrp1</td>\n",
       "      <td>A2m (A2mp)</td>\n",
       "      <td>Lrp1 (CD91, A2mr, b2b1554Clo)</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>&lt;span title=\"A2M (FWP007, S863-7, CPAMD5)\"&gt;A2M...</td>\n",
       "      <td>&lt;span title=\"LRP1 (APR, A2MR, LRP, CD91, LRP1A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDB00003</td>\n",
       "      <td>ACE BDKRB2</td>\n",
       "      <td>Ace Bdkrb2</td>\n",
       "      <td>Ace (CD143)</td>\n",
       "      <td>Bdkrb2 (B2, B(2), kinin B2, BK2R, B2R)</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>&lt;span title=\"ACE (DCP1, ACE1, CD143)\"&gt;ACE (DCP...</td>\n",
       "      <td>&lt;span title=\"BDKRB2 (BK-2)\"&gt;BDKRB2 (BK-2)&lt;/span&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDB00004</td>\n",
       "      <td>ADA DPP4</td>\n",
       "      <td>Ada Dpp4</td>\n",
       "      <td>Ada</td>\n",
       "      <td>Dpp4 (THAM, Dpp-4, Cd26)</td>\n",
       "      <td>&lt;span title=\"based on hpa, uniprot\"&gt;cell membr...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>&lt;span title=\"ADA (ADA1)\"&gt;ADA (ADA1)&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"DPP4 (CD26, ADCP2, DPPIV)\"&gt;DPP4 (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDB00005</td>\n",
       "      <td>ADAM10 EFNA5</td>\n",
       "      <td>Adam10 Efna5</td>\n",
       "      <td>Adam10 (kuz, kuzbanian, 1700031C13Rik)</td>\n",
       "      <td>Efna5 (AL-1, RAGS, Ephrin-A5, Epl7, EFL-5, LER...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>&lt;span title=\"ADAM10 (kuz, MADM, HsT18717, CD15...</td>\n",
       "      <td>&lt;span title=\"EFNA5 (EPLG7, AF1, LERK7)\"&gt;EFNA5 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3251</th>\n",
       "      <td>CDB03577</td>\n",
       "      <td>Pcdhb9 Pcdhb9</td>\n",
       "      <td>Pcdhb9 Pcdhb9</td>\n",
       "      <td>Pcdhb9 (Pcdhb4C, PcdhbI)</td>\n",
       "      <td>Pcdhb9 (Pcdhb4C, PcdhbI)</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3252</th>\n",
       "      <td>CDB03578</td>\n",
       "      <td>Pcdhgb8 Pcdhgb8</td>\n",
       "      <td>Pcdhgb8 Pcdhgb8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3253</th>\n",
       "      <td>CDB03579</td>\n",
       "      <td>Saa3 Tlr4</td>\n",
       "      <td>Saa3 Tlr4</td>\n",
       "      <td>Saa3 (Saa-3, l7R3)</td>\n",
       "      <td>Tlr4 (Lps, Rasl2-8, lipopolysaccharide response)</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"TLR4 (hToll, CD284, TLR-4, ARMD10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3254</th>\n",
       "      <td>CDB03580</td>\n",
       "      <td>Sema4a Timd2</td>\n",
       "      <td>Sema4a Timd2</td>\n",
       "      <td>Sema4a (SemB, SemB, Semab)</td>\n",
       "      <td>Timd2 (Tim2, TIM-2)</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>&lt;span title=\"SEMA4A (SEMAB, SemB, FLJ12287, CO...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3255</th>\n",
       "      <td>CDB03581</td>\n",
       "      <td>Zp3 Zp3r</td>\n",
       "      <td>Zp3 Zp3r</td>\n",
       "      <td>Zp3 (Zp-3)</td>\n",
       "      <td>Zp3r (SP56)</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;span title=\"based on PMID: 8188752\"&gt;cell memb...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>&lt;span title=\"ZP3 (ZP3A, ZP3B, ZP3-424, ZP3-372...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3256 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     <span title=\"Double-click header of Interaction ID to ensure all values are shown\">Interaction ID&nbsp;</span>  \\\n",
       "0                                              CDB00001                                                               \n",
       "1                                              CDB00002                                                               \n",
       "2                                              CDB00003                                                               \n",
       "3                                              CDB00004                                                               \n",
       "4                                              CDB00005                                                               \n",
       "...                                                 ...                                                               \n",
       "3251                                           CDB03577                                                               \n",
       "3252                                           CDB03578                                                               \n",
       "3253                                           CDB03579                                                               \n",
       "3254                                           CDB03580                                                               \n",
       "3255                                           CDB03581                                                               \n",
       "\n",
       "         LR Pair Card    Mouse LR Pair  \\\n",
       "0           A2M HSPA5        A2m Hspa5   \n",
       "1            A2M LRP1         A2m Lrp1   \n",
       "2          ACE BDKRB2       Ace Bdkrb2   \n",
       "3            ADA DPP4         Ada Dpp4   \n",
       "4        ADAM10 EFNA5     Adam10 Efna5   \n",
       "...               ...              ...   \n",
       "3251    Pcdhb9 Pcdhb9    Pcdhb9 Pcdhb9   \n",
       "3252  Pcdhgb8 Pcdhgb8  Pcdhgb8 Pcdhgb8   \n",
       "3253        Saa3 Tlr4        Saa3 Tlr4   \n",
       "3254     Sema4a Timd2     Sema4a Timd2   \n",
       "3255         Zp3 Zp3r         Zp3 Zp3r   \n",
       "\n",
       "                              Ligand Symbols  \\\n",
       "0                                 A2m (A2mp)   \n",
       "1                                 A2m (A2mp)   \n",
       "2                                Ace (CD143)   \n",
       "3                                        Ada   \n",
       "4     Adam10 (kuz, kuzbanian, 1700031C13Rik)   \n",
       "...                                      ...   \n",
       "3251                Pcdhb9 (Pcdhb4C, PcdhbI)   \n",
       "3252                                 Pcdhgb8   \n",
       "3253                      Saa3 (Saa-3, l7R3)   \n",
       "3254              Sema4a (SemB, SemB, Semab)   \n",
       "3255                              Zp3 (Zp-3)   \n",
       "\n",
       "                                       Receptor Symbols  \\\n",
       "0     Hspa5 (Hsce70, Bip, Grp78, Sez7, D2Wsu141e, D2...   \n",
       "1                         Lrp1 (CD91, A2mr, b2b1554Clo)   \n",
       "2                Bdkrb2 (B2, B(2), kinin B2, BK2R, B2R)   \n",
       "3                              Dpp4 (THAM, Dpp-4, Cd26)   \n",
       "4     Efna5 (AL-1, RAGS, Ephrin-A5, Epl7, EFL-5, LER...   \n",
       "...                                                 ...   \n",
       "3251                           Pcdhb9 (Pcdhb4C, PcdhbI)   \n",
       "3252                                            Pcdhgb8   \n",
       "3253   Tlr4 (Lps, Rasl2-8, lipopolysaccharide response)   \n",
       "3254                                Timd2 (Tim2, TIM-2)   \n",
       "3255                                        Zp3r (SP56)   \n",
       "\n",
       "     <span title=\"Location based on the predicted subcellular localization of the human proteome\">Ligand Location</span>  \\\n",
       "0     <span title=\"based on perplexity, uniprot\">sec...                                                                    \n",
       "1     <span title=\"based on perplexity, uniprot\">sec...                                                                    \n",
       "2     <span title=\"based on perplexity, uniprot\">cel...                                                                    \n",
       "3     <span title=\"based on hpa, uniprot\">cell membr...                                                                    \n",
       "4     <span title=\"based on hpa, perplexity, uniprot...                                                                    \n",
       "...                                                 ...                                                                    \n",
       "3251  <span title=\"based on uniprot\">cell membrane</...                                                                    \n",
       "3252  <span title=\"based on perplexity, uniprot\">cel...                                                                    \n",
       "3253  <span title=\"based on perplexity, uniprot\">sec...                                                                    \n",
       "3254  <span title=\"based on perplexity, uniprot\">cel...                                                                    \n",
       "3255  <span title=\"based on uniprot\">cell membrane</...                                                                    \n",
       "\n",
       "     <span title=\"Location based on the predicted subcellular localization of the human proteome\">Receptor Location</span>  \\\n",
       "0     <span title=\"based on PMID: 12194978, PMID: 32...                                                                      \n",
       "1     <span title=\"based on hpa, perplexity, uniprot...                                                                      \n",
       "2     <span title=\"based on hpa, perplexity, uniprot...                                                                      \n",
       "3     <span title=\"based on perplexity, uniprot\">cel...                                                                      \n",
       "4     <span title=\"based on perplexity, uniprot\">cel...                                                                      \n",
       "...                                                 ...                                                                      \n",
       "3251  <span title=\"based on uniprot\">cell membrane</...                                                                      \n",
       "3252  <span title=\"based on perplexity, uniprot\">cel...                                                                      \n",
       "3253  <span title=\"based on perplexity, uniprot\">cel...                                                                      \n",
       "3254  <span title=\"based on perplexity, uniprot\">cel...                                                                      \n",
       "3255  <span title=\"based on PMID: 8188752\">cell memb...                                                                      \n",
       "\n",
       "                                          Ligand MGI ID  \\\n",
       "0     <a href=\"https://www.informatics.jax.org/marke...   \n",
       "1     <a href=\"https://www.informatics.jax.org/marke...   \n",
       "2     <a href=\"https://www.informatics.jax.org/marke...   \n",
       "3     <a href=\"https://www.informatics.jax.org/marke...   \n",
       "4     <a href=\"https://www.informatics.jax.org/marke...   \n",
       "...                                                 ...   \n",
       "3251  <a href=\"https://www.informatics.jax.org/marke...   \n",
       "3252  <a href=\"https://www.informatics.jax.org/marke...   \n",
       "3253  <a href=\"https://www.informatics.jax.org/marke...   \n",
       "3254  <a href=\"https://www.informatics.jax.org/marke...   \n",
       "3255  <a href=\"https://www.informatics.jax.org/marke...   \n",
       "\n",
       "                                        Receptor MGI ID  \\\n",
       "0     <a href=\"https://www.informatics.jax.org/marke...   \n",
       "1     <a href=\"https://www.informatics.jax.org/marke...   \n",
       "2     <a href=\"https://www.informatics.jax.org/marke...   \n",
       "3     <a href=\"https://www.informatics.jax.org/marke...   \n",
       "4     <a href=\"https://www.informatics.jax.org/marke...   \n",
       "...                                                 ...   \n",
       "3251  <a href=\"https://www.informatics.jax.org/marke...   \n",
       "3252  <a href=\"https://www.informatics.jax.org/marke...   \n",
       "3253  <a href=\"https://www.informatics.jax.org/marke...   \n",
       "3254  <a href=\"https://www.informatics.jax.org/marke...   \n",
       "3255  <a href=\"https://www.informatics.jax.org/marke...   \n",
       "\n",
       "                                             Perplexity Mouse evidence  \\\n",
       "0     <a href=\"https://www.perplexity.ai/search?q=Wh...   CONSERVATION   \n",
       "1     <a href=\"https://www.perplexity.ai/search?q=Wh...   CONSERVATION   \n",
       "2     <a href=\"https://www.perplexity.ai/search?q=Wh...   CONSERVATION   \n",
       "3     <a href=\"https://www.perplexity.ai/search?q=Wh...   CONSERVATION   \n",
       "4     <a href=\"https://www.perplexity.ai/search?q=Wh...   CONSERVATION   \n",
       "...                                                 ...            ...   \n",
       "3251  <a href=\"https://www.perplexity.ai/search?q=Wh...         DIRECT   \n",
       "3252  <a href=\"https://www.perplexity.ai/search?q=Wh...         DIRECT   \n",
       "3253  <a href=\"https://www.perplexity.ai/search?q=Wh...         DIRECT   \n",
       "3254  <a href=\"https://www.perplexity.ai/search?q=Wh...         DIRECT   \n",
       "3255  <a href=\"https://www.perplexity.ai/search?q=Wh...         DIRECT   \n",
       "\n",
       "                                   Human Ligand Symbols  \\\n",
       "0     <span title=\"A2M (FWP007, S863-7, CPAMD5)\">A2M...   \n",
       "1     <span title=\"A2M (FWP007, S863-7, CPAMD5)\">A2M...   \n",
       "2     <span title=\"ACE (DCP1, ACE1, CD143)\">ACE (DCP...   \n",
       "3            <span title=\"ADA (ADA1)\">ADA (ADA1)</span>   \n",
       "4     <span title=\"ADAM10 (kuz, MADM, HsT18717, CD15...   \n",
       "...                                                 ...   \n",
       "3251  <span title=\"no human ortholog\">no human ortho...   \n",
       "3252  <span title=\"no human ortholog\">no human ortho...   \n",
       "3253  <span title=\"no human ortholog\">no human ortho...   \n",
       "3254  <span title=\"SEMA4A (SEMAB, SemB, FLJ12287, CO...   \n",
       "3255  <span title=\"ZP3 (ZP3A, ZP3B, ZP3-424, ZP3-372...   \n",
       "\n",
       "                                 Human Receptor Symbols  \n",
       "0     <span title=\"HSPA5 (GRP78, BiP)\">HSPA5 (GRP78,...  \n",
       "1     <span title=\"LRP1 (APR, A2MR, LRP, CD91, LRP1A...  \n",
       "2      <span title=\"BDKRB2 (BK-2)\">BDKRB2 (BK-2)</span>  \n",
       "3     <span title=\"DPP4 (CD26, ADCP2, DPPIV)\">DPP4 (...  \n",
       "4     <span title=\"EFNA5 (EPLG7, AF1, LERK7)\">EFNA5 ...  \n",
       "...                                                 ...  \n",
       "3251  <span title=\"no human ortholog\">no human ortho...  \n",
       "3252  <span title=\"no human ortholog\">no human ortho...  \n",
       "3253  <span title=\"TLR4 (hToll, CD284, TLR-4, ARMD10...  \n",
       "3254  <span title=\"no human ortholog\">no human ortho...  \n",
       "3255  <span title=\"no human ortholog\">no human ortho...  \n",
       "\n",
       "[3256 rows x 13 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Linkify multiple MGI IDs in Ligand column\n",
    "gene_pair[\"Ligand MGI ID\"] = gene_pair[\"Ligand MGI ID\"].apply(\n",
    "    lambda cell: \", \".join(\n",
    "        f'<a href=\"https://www.informatics.jax.org/marker/{mgi.strip()}\" target=\"_blank\">{mgi.strip()}</a>'\n",
    "        for mgi in str(cell).split(\", \")\n",
    "        if mgi.strip()\n",
    "    ) if pd.notna(cell) else \"\"\n",
    ")\n",
    "\n",
    "# Linkify multiple MGI IDs in Receptor column\n",
    "gene_pair[\"Receptor MGI ID\"] = gene_pair[\"Receptor MGI ID\"].apply(\n",
    "    lambda cell: \", \".join(\n",
    "        f'<a href=\"https://www.informatics.jax.org/marker/{mgi.strip()}\" target=\"_blank\">{mgi.strip()}</a>'\n",
    "        for mgi in str(cell).split(\", \")\n",
    "        if mgi.strip()\n",
    "    ) if pd.notna(cell) else \"\"\n",
    ")\n",
    "gene_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "19ac8189-acac-4a24-b749-d5b58e381db8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "c4383ee3-3b60-4df7-9587-63ae6045ef19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['LR Pair Card', 'Mouse_ligand', 'Mouse_receptor', 'Mouse evidence',\n",
       "       'MGI ligand', 'MGI receptor', 'PMID',\n",
       "       '<span title=\"Double-click header of Interaction ID to ensure all values are shown\">Interaction ID&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of LR Pair Card to ensure all values are shown\">LR Pair Card&nbsp;</span>',\n",
       "       '<span title=\"Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\">Human LR Pair</span>',\n",
       "       'Human Ligand Symbols', 'Human Receptor Symbols',\n",
       "       '<span title=\"Location based on the predicted subcellular localization of the human proteome\">Ligand Location</span>',\n",
       "       '<span title=\"Location based on the predicted subcellular localization of the human proteome\">Receptor Location</span>',\n",
       "       'Ligand Name', 'Ligand Symbols', 'Receptor Name', 'Receptor Symbols',\n",
       "       'Mouse LR Pair'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "e29ae6b5-c654-4b36-b5c8-4ece61eb53f4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "c1cbe14d-cc35-4ef7-b595-483950fb44b9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "908d618f-282f-4459-9d7e-5bdbce44b0b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['<span title=\"Double-click header of Interaction ID to ensure all values are shown\">Interaction ID&nbsp;</span>',\n",
       "       'Mouse LR Pair', 'Ligand Symbols', 'Receptor Symbols', 'LR Pair Card',\n",
       "       'Mouse_ligand', 'Mouse_receptor', 'Mouse evidence', 'Ligand MGI ID',\n",
       "       'Receptor MGI ID', 'PMID',\n",
       "       '<span title=\"Double-click header of LR Pair Card to ensure all values are shown\">LR Pair Card&nbsp;</span>',\n",
       "       '<span title=\"Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\">Human LR Pair</span>',\n",
       "       'Human Ligand Symbols', 'Human Receptor Symbols',\n",
       "       '<span title=\"Location based on the predicted subcellular localization of the human proteome\">Ligand Location</span>',\n",
       "       '<span title=\"Location based on the predicted subcellular localization of the human proteome\">Receptor Location</span>',\n",
       "       'Ligand Name', 'Receptor Name', 'Perplexity'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36d9aa0b-9846-4ea4-ac18-6c765d86d631",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "da937b5a-dbf5-41ae-93b2-196fe7ae65b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair = gene_pair[[interaction_id_col, \"LR Pair Card\",'Mouse LR Pair', 'Ligand Symbols', 'Receptor Symbols', ligand_loc_col, receptor_loc_col, \"Ligand MGI ID\", \"Receptor MGI ID\", \"Perplexity\", \"Mouse evidence\", \"Human Ligand Symbols\", \"Human Receptor Symbols\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "801b2e2e-1897-42ac-a390-011daea6b9fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair.to_csv(\"data/test.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "83ffc720-8f13-4f9e-bf1d-7b1fdaee1954",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair = gene_pair[~(gene_pair[\"Mouse evidence\"] == \"not conserved\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "1e476d4a-818c-4835-a967-e83fdea478c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>Mouse_ligand</th>\n",
       "      <th>Mouse_receptor</th>\n",
       "      <th>Mouse evidence</th>\n",
       "      <th>MGI ligand</th>\n",
       "      <th>MGI receptor</th>\n",
       "      <th>&lt;span title=\"Double-click header of Interaction ID to ensure all values are shown\"&gt;Interaction ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of LR Pair Card to ensure all values are shown\"&gt;LR Pair Card&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\"&gt;Human LR Pair&lt;/span&gt;</th>\n",
       "      <th>Human Ligand Symbols</th>\n",
       "      <th>Human Receptor Symbols</th>\n",
       "      <th>&lt;span title=\"Location based on the predicted subcellular localization of the human proteome\"&gt;Ligand Location&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Location based on the predicted subcellular localization of the human proteome\"&gt;Receptor Location&lt;/span&gt;</th>\n",
       "      <th>MGI Marker Accession ID</th>\n",
       "      <th>Marker Name</th>\n",
       "      <th>Aliases</th>\n",
       "      <th>Mouse LR Pair</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ASIP ATRN</td>\n",
       "      <td>a</td>\n",
       "      <td>Atrn</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>MGI:87853</td>\n",
       "      <td>MGI:1341628</td>\n",
       "      <td>CDB00229</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>ASIP ATRN</td>\n",
       "      <td>&lt;span title=\"ASIP (AGTIL, ASP)\"&gt;ASIP (AGTIL, A...</td>\n",
       "      <td>&lt;span title=\"ATRN (DPPT-L, MGCA)\"&gt;ATRN (DPPT-L...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>MGI:87853</td>\n",
       "      <td>nonagouti</td>\n",
       "      <td>agouti, As, agouti signal protein, ASP, Asip</td>\n",
       "      <td>a Atrn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>A2m</td>\n",
       "      <td>Hspa5</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>MGI:2449119</td>\n",
       "      <td>MGI:95835</td>\n",
       "      <td>CDB00001</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>&lt;span title=\"A2M (FWP007, S863-7, CPAMD5)\"&gt;A2M...</td>\n",
       "      <td>&lt;span title=\"HSPA5 (GRP78, BiP)\"&gt;HSPA5 (GRP78,...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;span title=\"based on PMID: 12194978, PMID: 32...</td>\n",
       "      <td>MGI:2449119</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>A2mp</td>\n",
       "      <td>A2m Hspa5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>A2m</td>\n",
       "      <td>Hspa5</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>MGI:2449119</td>\n",
       "      <td>MGI:95835</td>\n",
       "      <td>CDB00001</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>&lt;span title=\"A2M (FWP007, S863-7, CPAMD5)\"&gt;A2M...</td>\n",
       "      <td>&lt;span title=\"HSPA5 (GRP78, BiP)\"&gt;HSPA5 (GRP78,...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;span title=\"based on PMID: 12194978, PMID: 32...</td>\n",
       "      <td>MGI:2449119</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>A2mp</td>\n",
       "      <td>A2m Hspa5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>A2m</td>\n",
       "      <td>Lrp1</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>MGI:2449119</td>\n",
       "      <td>MGI:96828</td>\n",
       "      <td>CDB00002</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>&lt;span title=\"A2M (FWP007, S863-7, CPAMD5)\"&gt;A2M...</td>\n",
       "      <td>&lt;span title=\"LRP1 (APR, A2MR, LRP, CD91, LRP1A...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>MGI:2449119</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>A2mp</td>\n",
       "      <td>A2m Lrp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>A2m</td>\n",
       "      <td>Lrp1</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>MGI:2449119</td>\n",
       "      <td>MGI:96828</td>\n",
       "      <td>CDB00002</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>&lt;span title=\"A2M (FWP007, S863-7, CPAMD5)\"&gt;A2M...</td>\n",
       "      <td>&lt;span title=\"LRP1 (APR, A2MR, LRP, CD91, LRP1A...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>MGI:2449119</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>A2mp</td>\n",
       "      <td>A2m Lrp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6089</th>\n",
       "      <td>ZG16B TLR5</td>\n",
       "      <td>Sbpl</td>\n",
       "      <td>Tlr5</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>MGI:3694550</td>\n",
       "      <td>MGI:1858171</td>\n",
       "      <td>CDB03548</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>ZG16B TLR5</td>\n",
       "      <td>&lt;span title=\"ZG16B (HRPE773, PRO1567, JCLN2)\"&gt;...</td>\n",
       "      <td>&lt;span title=\"TLR5 (SLEB1, TIL3, FLJ10052, MGC1...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>MGI:3694550</td>\n",
       "      <td>spermine binding protein-like</td>\n",
       "      <td>2310068J22Rik</td>\n",
       "      <td>Sbpl Tlr5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6090</th>\n",
       "      <td>ZG16B TLR6</td>\n",
       "      <td>Sbpl</td>\n",
       "      <td>Tlr6</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>MGI:3694550</td>\n",
       "      <td>MGI:1341296</td>\n",
       "      <td>CDB03549</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>ZG16B TLR6</td>\n",
       "      <td>&lt;span title=\"ZG16B (HRPE773, PRO1567, JCLN2)\"&gt;...</td>\n",
       "      <td>&lt;span title=\"TLR6 (CD286)\"&gt;TLR6 (CD286)&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>MGI:3694550</td>\n",
       "      <td>spermine binding protein-like</td>\n",
       "      <td>2310068J22Rik</td>\n",
       "      <td>Sbpl Tlr6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6091</th>\n",
       "      <td>ZG16B TLR6</td>\n",
       "      <td>Sbpl</td>\n",
       "      <td>Tlr6</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>MGI:3694550</td>\n",
       "      <td>MGI:1341296</td>\n",
       "      <td>CDB03549</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>ZG16B TLR6</td>\n",
       "      <td>&lt;span title=\"ZG16B (HRPE773, PRO1567, JCLN2)\"&gt;...</td>\n",
       "      <td>&lt;span title=\"TLR6 (CD286)\"&gt;TLR6 (CD286)&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>MGI:3694550</td>\n",
       "      <td>spermine binding protein-like</td>\n",
       "      <td>2310068J22Rik</td>\n",
       "      <td>Sbpl Tlr6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6092</th>\n",
       "      <td>ZP3 CHRNA7</td>\n",
       "      <td>Zp3</td>\n",
       "      <td>Chrna7</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>MGI:99215</td>\n",
       "      <td>MGI:99779</td>\n",
       "      <td>CDB03550</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>ZP3 CHRNA7</td>\n",
       "      <td>&lt;span title=\"ZP3 (ZP3A, ZP3B, ZP3-424, ZP3-372...</td>\n",
       "      <td>&lt;span title=\"CHRNA7\"&gt;CHRNA7&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>MGI:99215</td>\n",
       "      <td>zona pellucida glycoprotein 3</td>\n",
       "      <td>Zp-3</td>\n",
       "      <td>Zp3 Chrna7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6093</th>\n",
       "      <td>Zp3 Zp3R</td>\n",
       "      <td>Zp3</td>\n",
       "      <td>Zp3r</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>MGI:99215</td>\n",
       "      <td>MGI:104965</td>\n",
       "      <td>CDB03581</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>&lt;span title=\"ZP3 (ZP3A, ZP3B, ZP3-424, ZP3-372...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>MGI:99215</td>\n",
       "      <td>zona pellucida glycoprotein 3</td>\n",
       "      <td>Zp-3</td>\n",
       "      <td>Zp3 Zp3r</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6094 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     LR Pair Card Mouse_ligand Mouse_receptor Mouse evidence   MGI ligand  \\\n",
       "0       ASIP ATRN            a           Atrn         DIRECT    MGI:87853   \n",
       "1       A2M HSPA5          A2m          Hspa5   CONSERVATION  MGI:2449119   \n",
       "2       A2M HSPA5          A2m          Hspa5   CONSERVATION  MGI:2449119   \n",
       "3        A2M LRP1          A2m           Lrp1   CONSERVATION  MGI:2449119   \n",
       "4        A2M LRP1          A2m           Lrp1   CONSERVATION  MGI:2449119   \n",
       "...           ...          ...            ...            ...          ...   \n",
       "6089   ZG16B TLR5         Sbpl           Tlr5   CONSERVATION  MGI:3694550   \n",
       "6090   ZG16B TLR6         Sbpl           Tlr6   CONSERVATION  MGI:3694550   \n",
       "6091   ZG16B TLR6         Sbpl           Tlr6   CONSERVATION  MGI:3694550   \n",
       "6092   ZP3 CHRNA7          Zp3         Chrna7         DIRECT    MGI:99215   \n",
       "6093     Zp3 Zp3R          Zp3           Zp3r         DIRECT    MGI:99215   \n",
       "\n",
       "     MGI receptor  \\\n",
       "0     MGI:1341628   \n",
       "1       MGI:95835   \n",
       "2       MGI:95835   \n",
       "3       MGI:96828   \n",
       "4       MGI:96828   \n",
       "...           ...   \n",
       "6089  MGI:1858171   \n",
       "6090  MGI:1341296   \n",
       "6091  MGI:1341296   \n",
       "6092    MGI:99779   \n",
       "6093   MGI:104965   \n",
       "\n",
       "     <span title=\"Double-click header of Interaction ID to ensure all values are shown\">Interaction ID&nbsp;</span>  \\\n",
       "0                                              CDB00229                                                               \n",
       "1                                              CDB00001                                                               \n",
       "2                                              CDB00001                                                               \n",
       "3                                              CDB00002                                                               \n",
       "4                                              CDB00002                                                               \n",
       "...                                                 ...                                                               \n",
       "6089                                           CDB03548                                                               \n",
       "6090                                           CDB03549                                                               \n",
       "6091                                           CDB03549                                                               \n",
       "6092                                           CDB03550                                                               \n",
       "6093                                           CDB03581                                                               \n",
       "\n",
       "     <span title=\"Double-click header of LR Pair Card to ensure all values are shown\">LR Pair Card&nbsp;</span>  \\\n",
       "0     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                           \n",
       "1     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                           \n",
       "2     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                           \n",
       "3     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                           \n",
       "4     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                           \n",
       "...                                                 ...                                                           \n",
       "6089  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                           \n",
       "6090  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                           \n",
       "6091  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                           \n",
       "6092  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                           \n",
       "6093  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                           \n",
       "\n",
       "     <span title=\"Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\">Human LR Pair</span>  \\\n",
       "0                                             ASIP ATRN                                                               \n",
       "1                                             A2M HSPA5                                                               \n",
       "2                                             A2M HSPA5                                                               \n",
       "3                                              A2M LRP1                                                               \n",
       "4                                              A2M LRP1                                                               \n",
       "...                                                 ...                                                               \n",
       "6089                                         ZG16B TLR5                                                               \n",
       "6090                                         ZG16B TLR6                                                               \n",
       "6091                                         ZG16B TLR6                                                               \n",
       "6092                                         ZP3 CHRNA7                                                               \n",
       "6093                                  no human ortholog                                                               \n",
       "\n",
       "                                   Human Ligand Symbols  \\\n",
       "0     <span title=\"ASIP (AGTIL, ASP)\">ASIP (AGTIL, A...   \n",
       "1     <span title=\"A2M (FWP007, S863-7, CPAMD5)\">A2M...   \n",
       "2     <span title=\"A2M (FWP007, S863-7, CPAMD5)\">A2M...   \n",
       "3     <span title=\"A2M (FWP007, S863-7, CPAMD5)\">A2M...   \n",
       "4     <span title=\"A2M (FWP007, S863-7, CPAMD5)\">A2M...   \n",
       "...                                                 ...   \n",
       "6089  <span title=\"ZG16B (HRPE773, PRO1567, JCLN2)\">...   \n",
       "6090  <span title=\"ZG16B (HRPE773, PRO1567, JCLN2)\">...   \n",
       "6091  <span title=\"ZG16B (HRPE773, PRO1567, JCLN2)\">...   \n",
       "6092  <span title=\"ZP3 (ZP3A, ZP3B, ZP3-424, ZP3-372...   \n",
       "6093  <span title=\"ZP3 (ZP3A, ZP3B, ZP3-424, ZP3-372...   \n",
       "\n",
       "                                 Human Receptor Symbols  \\\n",
       "0     <span title=\"ATRN (DPPT-L, MGCA)\">ATRN (DPPT-L...   \n",
       "1     <span title=\"HSPA5 (GRP78, BiP)\">HSPA5 (GRP78,...   \n",
       "2     <span title=\"HSPA5 (GRP78, BiP)\">HSPA5 (GRP78,...   \n",
       "3     <span title=\"LRP1 (APR, A2MR, LRP, CD91, LRP1A...   \n",
       "4     <span title=\"LRP1 (APR, A2MR, LRP, CD91, LRP1A...   \n",
       "...                                                 ...   \n",
       "6089  <span title=\"TLR5 (SLEB1, TIL3, FLJ10052, MGC1...   \n",
       "6090     <span title=\"TLR6 (CD286)\">TLR6 (CD286)</span>   \n",
       "6091     <span title=\"TLR6 (CD286)\">TLR6 (CD286)</span>   \n",
       "6092                 <span title=\"CHRNA7\">CHRNA7</span>   \n",
       "6093  <span title=\"no human ortholog\">no human ortho...   \n",
       "\n",
       "     <span title=\"Location based on the predicted subcellular localization of the human proteome\">Ligand Location</span>  \\\n",
       "0     <span title=\"based on hpa, perplexity, uniprot...                                                                    \n",
       "1     <span title=\"based on perplexity, uniprot\">sec...                                                                    \n",
       "2     <span title=\"based on perplexity, uniprot\">sec...                                                                    \n",
       "3     <span title=\"based on perplexity, uniprot\">sec...                                                                    \n",
       "4     <span title=\"based on perplexity, uniprot\">sec...                                                                    \n",
       "...                                                 ...                                                                    \n",
       "6089  <span title=\"based on perplexity, uniprot\">cel...                                                                    \n",
       "6090  <span title=\"based on perplexity, uniprot\">cel...                                                                    \n",
       "6091  <span title=\"based on perplexity, uniprot\">cel...                                                                    \n",
       "6092  <span title=\"based on uniprot\">cell membrane</...                                                                    \n",
       "6093  <span title=\"based on uniprot\">cell membrane</...                                                                    \n",
       "\n",
       "     <span title=\"Location based on the predicted subcellular localization of the human proteome\">Receptor Location</span>  \\\n",
       "0     <span title=\"based on perplexity, uniprot\">cel...                                                                      \n",
       "1     <span title=\"based on PMID: 12194978, PMID: 32...                                                                      \n",
       "2     <span title=\"based on PMID: 12194978, PMID: 32...                                                                      \n",
       "3     <span title=\"based on hpa, perplexity, uniprot...                                                                      \n",
       "4     <span title=\"based on hpa, perplexity, uniprot...                                                                      \n",
       "...                                                 ...                                                                      \n",
       "6089  <span title=\"based on perplexity, uniprot\">cel...                                                                      \n",
       "6090  <span title=\"based on perplexity, uniprot\">cel...                                                                      \n",
       "6091  <span title=\"based on perplexity, uniprot\">cel...                                                                      \n",
       "6092  <span title=\"based on perplexity, uniprot\">cel...                                                                      \n",
       "6093                                            unknown                                                                      \n",
       "\n",
       "     MGI Marker Accession ID                    Marker Name  \\\n",
       "0                  MGI:87853                      nonagouti   \n",
       "1                MGI:2449119          alpha-2-macroglobulin   \n",
       "2                MGI:2449119          alpha-2-macroglobulin   \n",
       "3                MGI:2449119          alpha-2-macroglobulin   \n",
       "4                MGI:2449119          alpha-2-macroglobulin   \n",
       "...                      ...                            ...   \n",
       "6089             MGI:3694550  spermine binding protein-like   \n",
       "6090             MGI:3694550  spermine binding protein-like   \n",
       "6091             MGI:3694550  spermine binding protein-like   \n",
       "6092               MGI:99215  zona pellucida glycoprotein 3   \n",
       "6093               MGI:99215  zona pellucida glycoprotein 3   \n",
       "\n",
       "                                           Aliases Mouse LR Pair  \n",
       "0     agouti, As, agouti signal protein, ASP, Asip        a Atrn  \n",
       "1                                             A2mp     A2m Hspa5  \n",
       "2                                             A2mp     A2m Hspa5  \n",
       "3                                             A2mp      A2m Lrp1  \n",
       "4                                             A2mp      A2m Lrp1  \n",
       "...                                            ...           ...  \n",
       "6089                                 2310068J22Rik     Sbpl Tlr5  \n",
       "6090                                 2310068J22Rik     Sbpl Tlr6  \n",
       "6091                                 2310068J22Rik     Sbpl Tlr6  \n",
       "6092                                          Zp-3    Zp3 Chrna7  \n",
       "6093                                          Zp-3      Zp3 Zp3r  \n",
       "\n",
       "[6094 rows x 17 columns]"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "766dbe97-76b6-4c48-b7dc-d7fa03f1baf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_species = fetchGSheet.gene_pair_mouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "38212998-4a50-43d6-8b0a-27cd611667ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_species= gene_pair_species[[\"LR Pair Card\", \"Mouse_ligand\",\n",
    "\"Mouse_receptor\", \"MGI ligand\", \"MGI receptor\"]]\n",
    "gene_pair =gene_pair_species.merge(gene_pair,how=\"left\", on=\"LR Pair Card\")\n",
    "mouse_info = pd.read_csv(\"data/MRK_Merged_MGI_DB.tsv\", sep=\"\\t\", dtype=str)\n",
    "mouse_info = mouse_info[['MGI Marker Accession ID', 'Marker Name','Aliases']]\n",
    "gene_pair = gene_pair.merge(mouse_info,how=\"left\", left_on = \"MGI ligand\",right_on=\"MGI Marker Accession ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "2827ace6-ce1c-499d-99f5-5e6ef9876b15",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "ac52c44e-c4c6-4f09-b89d-f044d686c0b8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "2646138e-06e0-4fa1-8577-cc4569198de1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGI Marker Accession ID</th>\n",
       "      <th>Marker Name</th>\n",
       "      <th>Aliases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGI:1341858</td>\n",
       "      <td>DNA segment, 03B03F (Research Genetics)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MGI:1341869</td>\n",
       "      <td>DNA segment, 03B03R (Research Genetics)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:1918911</td>\n",
       "      <td>RIKEN cDNA 0610005C13 gene</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MGI:1923503</td>\n",
       "      <td>RIKEN cDNA 0610006L08 gene</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MGI:1925547</td>\n",
       "      <td>RIKEN cDNA 0610008J02 gene</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127817</th>\n",
       "      <td>MGI:2446208</td>\n",
       "      <td>zyg-11 family member A, cell cycle regulator</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127818</th>\n",
       "      <td>MGI:2685277</td>\n",
       "      <td>zyg-ll family member B, cell cycle regulator</td>\n",
       "      <td>LOC242610|1110046I03Rik|D4Mgi23|2810482G21Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127819</th>\n",
       "      <td>MGI:103072</td>\n",
       "      <td>zyxin</td>\n",
       "      <td>R75157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127820</th>\n",
       "      <td>MGI:2444286</td>\n",
       "      <td>zinc finger, ZZ-type with EF hand domain 1</td>\n",
       "      <td>8430405D05Rik|C130099L13Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127821</th>\n",
       "      <td>MGI:1920453</td>\n",
       "      <td>zinc finger, ZZ domain containing 3</td>\n",
       "      <td>3110065C23Rik|6430567E01Rik</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>127822 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       MGI Marker Accession ID                                   Marker Name  \\\n",
       "0                  MGI:1341858       DNA segment, 03B03F (Research Genetics)   \n",
       "1                  MGI:1341869       DNA segment, 03B03R (Research Genetics)   \n",
       "2                  MGI:1918911                    RIKEN cDNA 0610005C13 gene   \n",
       "3                  MGI:1923503                    RIKEN cDNA 0610006L08 gene   \n",
       "4                  MGI:1925547                    RIKEN cDNA 0610008J02 gene   \n",
       "...                        ...                                           ...   \n",
       "127817             MGI:2446208  zyg-11 family member A, cell cycle regulator   \n",
       "127818             MGI:2685277  zyg-ll family member B, cell cycle regulator   \n",
       "127819              MGI:103072                                         zyxin   \n",
       "127820             MGI:2444286    zinc finger, ZZ-type with EF hand domain 1   \n",
       "127821             MGI:1920453           zinc finger, ZZ domain containing 3   \n",
       "\n",
       "                                              Aliases  \n",
       "0                                                 NaN  \n",
       "1                                                 NaN  \n",
       "2                                                 NaN  \n",
       "3                                                 NaN  \n",
       "4                                                 NaN  \n",
       "...                                               ...  \n",
       "127817                                            NaN  \n",
       "127818  LOC242610|1110046I03Rik|D4Mgi23|2810482G21Rik  \n",
       "127819                                         R75157  \n",
       "127820                    8430405D05Rik|C130099L13Rik  \n",
       "127821                    3110065C23Rik|6430567E01Rik  \n",
       "\n",
       "[127822 rows x 3 columns]"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "76564640-1240-43d8-abec-ca1078ecddd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair.to_csv(\"data/test.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "3060fd20-2eb7-4f76-b7a8-31d578a528c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "orth_info = pd.read_csv(f\"data/GenePageGeneralInfo_{id}base_DB.tsv\", sep=\"\\t\", dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "5916fa76-4abd-4fc3-ac14-22b9e2a53f6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_orth_symbol = dict(zip(orth_info['tropicalis gene ID'], orth_info['tropicalis gene symbol']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "0cdcda36-0338-4dd2-af55-7a63b6750e17",
   "metadata": {},
   "outputs": [],
   "source": [
    "species = \"Frog\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "fce34276-e11c-4aad-9de2-99c9e639bb37",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_orth_symbol = dict(zip(orth_info['tropicalis gene ID'], orth_info['tropicalis gene symbol']))\n",
    "gene_pair_orth['Ligand Official Symbol'] = 'NA'\n",
    "gene_pair_orth['Ligand Official Symbol'] =  gene_pair_orth.apply(\n",
    "    lambda row: mapping_orth_symbol.get(row[f'{id} ligand'], row['Ligand Official Symbol'])\n",
    "    if pd.notna(row[f'{id} ligand']) else row['Ligand Official Symbol'],\n",
    "    axis=1\n",
    ")\n",
    "gene_pair_orth['Receptor Official Symbol'] = 'NA'\n",
    "gene_pair_orth['Receptor Official Symbol'] =  gene_pair_orth.apply(\n",
    "    lambda row: mapping_orth_symbol.get(row[f'{id} receptor'], row['Receptor Official Symbol'])\n",
    "    if pd.notna(row[f'{id} receptor']) else row['Receptor Official Symbol'],\n",
    "    axis=1\n",
    ")\n",
    "gene_pair_orth[\"same_as_off_lig\"] = (\n",
    "    gene_pair_orth[f\"{species.lower()}_ligand\"] == gene_pair_orth[\"Ligand Official Symbol\"]\n",
    ")\n",
    "\n",
    "gene_pair_orth[\"same_as_off_rec\"] = (\n",
    "    gene_pair_orth[f\"{species.lower()}_receptor\"] == gene_pair_orth[\"Receptor Official Symbol\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "e6e1df2a-ecba-4c3c-b67d-644a27a492b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_orth.to_csv(f\"data/{id}_ID_check.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "8cae86be-ca8e-4ce7-9a20-3f70aa39a103",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sorting order</th>\n",
       "      <th>LR_pair_orig</th>\n",
       "      <th>PMID</th>\n",
       "      <th>lig_species</th>\n",
       "      <th>rec_species</th>\n",
       "      <th>ligand_orig</th>\n",
       "      <th>receptor_orig</th>\n",
       "      <th>Pair_species</th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>rat evidence</th>\n",
       "      <th>rat_ligand</th>\n",
       "      <th>rat_receptor</th>\n",
       "      <th>ENSEMBL ligand</th>\n",
       "      <th>ENSEMBL receptor</th>\n",
       "      <th>RGD ligand</th>\n",
       "      <th>RGD receptor</th>\n",
       "      <th>Ligand Official Symbol</th>\n",
       "      <th>Receptor Official Symbol</th>\n",
       "      <th>same_as_off_lig</th>\n",
       "      <th>same_as_off_rec</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A Atrn</td>\n",
       "      <td>11137996</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>a</td>\n",
       "      <td>Atrn</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>ASIP ATRN</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>Asip</td>\n",
       "      <td>Atrn</td>\n",
       "      <td>ENSRNOG00000017701</td>\n",
       "      <td>ENSRNOG00000021240</td>\n",
       "      <td>2003</td>\n",
       "      <td>69063</td>\n",
       "      <td>Asip</td>\n",
       "      <td>Atrn</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>12194978</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>A2M</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>A2m</td>\n",
       "      <td>Hspa5</td>\n",
       "      <td>ENSRNOG00000028896</td>\n",
       "      <td>ENSRNOG00000018294</td>\n",
       "      <td>2004</td>\n",
       "      <td>2843</td>\n",
       "      <td>A2m</td>\n",
       "      <td>Hspa5</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>32541810</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>A2M</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>A2m</td>\n",
       "      <td>Hspa5</td>\n",
       "      <td>ENSRNOG00000028896</td>\n",
       "      <td>ENSRNOG00000018294</td>\n",
       "      <td>2004</td>\n",
       "      <td>2843</td>\n",
       "      <td>A2m</td>\n",
       "      <td>Hspa5</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>10652313</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>A2M</td>\n",
       "      <td>LRP1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>A2m</td>\n",
       "      <td>Lrp1</td>\n",
       "      <td>ENSRNOG00000028896</td>\n",
       "      <td>ENSRNOG00000025053</td>\n",
       "      <td>2004</td>\n",
       "      <td>1307535</td>\n",
       "      <td>A2m</td>\n",
       "      <td>Lrp1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>12194978</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>A2M</td>\n",
       "      <td>LRP1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>A2m</td>\n",
       "      <td>Lrp1</td>\n",
       "      <td>ENSRNOG00000028896</td>\n",
       "      <td>ENSRNOG00000025053</td>\n",
       "      <td>2004</td>\n",
       "      <td>1307535</td>\n",
       "      <td>A2m</td>\n",
       "      <td>Lrp1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6089</th>\n",
       "      <td>6090</td>\n",
       "      <td>ZG16B TLR5</td>\n",
       "      <td>20802527</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ZG16B</td>\n",
       "      <td>TLR5</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ZG16B TLR5</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>Zg16b</td>\n",
       "      <td>Tlr5</td>\n",
       "      <td>ENSRNOG00000057127</td>\n",
       "      <td>ENSRNOG00000022067</td>\n",
       "      <td>1562673</td>\n",
       "      <td>631351</td>\n",
       "      <td>Zg16b</td>\n",
       "      <td>Tlr5</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6090</th>\n",
       "      <td>6091</td>\n",
       "      <td>ZG16B TLR6</td>\n",
       "      <td>20802527</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ZG16B</td>\n",
       "      <td>TLR6</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ZG16B TLR6</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>Zg16b</td>\n",
       "      <td>Tlr6</td>\n",
       "      <td>ENSRNOG00000057127</td>\n",
       "      <td>ENSRNOG00000002161</td>\n",
       "      <td>1562673</td>\n",
       "      <td>1303030</td>\n",
       "      <td>Zg16b</td>\n",
       "      <td>Tlr6</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6091</th>\n",
       "      <td>6092</td>\n",
       "      <td>ZG16B Tlr6</td>\n",
       "      <td>20802527</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>ZG16B</td>\n",
       "      <td>Tlr6</td>\n",
       "      <td>mixed</td>\n",
       "      <td>ZG16B TLR6</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>Zg16b</td>\n",
       "      <td>Tlr6</td>\n",
       "      <td>ENSRNOG00000057127</td>\n",
       "      <td>ENSRNOG00000002161</td>\n",
       "      <td>1562673</td>\n",
       "      <td>1303030</td>\n",
       "      <td>Zg16b</td>\n",
       "      <td>Tlr6</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6092</th>\n",
       "      <td>6093</td>\n",
       "      <td>Zp3 Chrna7</td>\n",
       "      <td>22577141</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Zp3</td>\n",
       "      <td>Chrna7</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>ZP3 CHRNA7</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>Zp3</td>\n",
       "      <td>Chrna7</td>\n",
       "      <td>ENSRNOG00000001434</td>\n",
       "      <td>ENSRNOG00000010853</td>\n",
       "      <td>620606</td>\n",
       "      <td>2348</td>\n",
       "      <td>Zp3</td>\n",
       "      <td>Chrna7</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6093</th>\n",
       "      <td>6094</td>\n",
       "      <td>Zp3 Zp3R</td>\n",
       "      <td>8188752</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Zp3</td>\n",
       "      <td>Zp3r</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Zp3 Zp3r</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>Zp3</td>\n",
       "      <td>Zp3r</td>\n",
       "      <td>ENSRNOG00000001434</td>\n",
       "      <td>ENSRNOG00000025862</td>\n",
       "      <td>620606</td>\n",
       "      <td>727846</td>\n",
       "      <td>Zp3</td>\n",
       "      <td>Zp3r</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6094 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0    Sorting order LR_pair_orig      PMID   lig_species   rec_species  \\\n",
       "0                1       A Atrn  11137996  Mus Musculus  Mus Musculus   \n",
       "1                2    A2M HSPA5  12194978  Homo sapiens  Homo sapiens   \n",
       "2                3    A2M HSPA5  32541810  Homo sapiens  Homo sapiens   \n",
       "3                4     A2M LRP1  10652313  Homo sapiens  Homo sapiens   \n",
       "4                5     A2M LRP1  12194978  Homo sapiens  Homo sapiens   \n",
       "...            ...          ...       ...           ...           ...   \n",
       "6089          6090   ZG16B TLR5  20802527  Homo sapiens  Homo sapiens   \n",
       "6090          6091   ZG16B TLR6  20802527  Homo sapiens  Homo sapiens   \n",
       "6091          6092   ZG16B Tlr6  20802527  Homo sapiens  Mus Musculus   \n",
       "6092          6093   Zp3 Chrna7  22577141  Mus Musculus  Mus Musculus   \n",
       "6093          6094     Zp3 Zp3R   8188752  Mus Musculus  Mus Musculus   \n",
       "\n",
       "0    ligand_orig receptor_orig  Pair_species LR Pair Card  rat evidence  \\\n",
       "0              a          Atrn  Mus Musculus    ASIP ATRN  CONSERVATION   \n",
       "1            A2M         HSPA5  Homo sapiens    A2M HSPA5  CONSERVATION   \n",
       "2            A2M         HSPA5  Homo sapiens    A2M HSPA5  CONSERVATION   \n",
       "3            A2M          LRP1  Homo sapiens     A2M LRP1  CONSERVATION   \n",
       "4            A2M          LRP1  Homo sapiens     A2M LRP1  CONSERVATION   \n",
       "...          ...           ...           ...          ...           ...   \n",
       "6089       ZG16B          TLR5  Homo sapiens   ZG16B TLR5  CONSERVATION   \n",
       "6090       ZG16B          TLR6  Homo sapiens   ZG16B TLR6  CONSERVATION   \n",
       "6091       ZG16B          Tlr6         mixed   ZG16B TLR6  CONSERVATION   \n",
       "6092         Zp3        Chrna7  Mus Musculus   ZP3 CHRNA7  CONSERVATION   \n",
       "6093         Zp3          Zp3r  Mus Musculus     Zp3 Zp3r  CONSERVATION   \n",
       "\n",
       "0    rat_ligand rat_receptor      ENSEMBL ligand    ENSEMBL receptor  \\\n",
       "0          Asip         Atrn  ENSRNOG00000017701  ENSRNOG00000021240   \n",
       "1           A2m        Hspa5  ENSRNOG00000028896  ENSRNOG00000018294   \n",
       "2           A2m        Hspa5  ENSRNOG00000028896  ENSRNOG00000018294   \n",
       "3           A2m         Lrp1  ENSRNOG00000028896  ENSRNOG00000025053   \n",
       "4           A2m         Lrp1  ENSRNOG00000028896  ENSRNOG00000025053   \n",
       "...         ...          ...                 ...                 ...   \n",
       "6089      Zg16b         Tlr5  ENSRNOG00000057127  ENSRNOG00000022067   \n",
       "6090      Zg16b         Tlr6  ENSRNOG00000057127  ENSRNOG00000002161   \n",
       "6091      Zg16b         Tlr6  ENSRNOG00000057127  ENSRNOG00000002161   \n",
       "6092        Zp3       Chrna7  ENSRNOG00000001434  ENSRNOG00000010853   \n",
       "6093        Zp3         Zp3r  ENSRNOG00000001434  ENSRNOG00000025862   \n",
       "\n",
       "0    RGD ligand RGD receptor Ligand Official Symbol Receptor Official Symbol  \\\n",
       "0          2003        69063                   Asip                     Atrn   \n",
       "1          2004         2843                    A2m                    Hspa5   \n",
       "2          2004         2843                    A2m                    Hspa5   \n",
       "3          2004      1307535                    A2m                     Lrp1   \n",
       "4          2004      1307535                    A2m                     Lrp1   \n",
       "...         ...          ...                    ...                      ...   \n",
       "6089    1562673       631351                  Zg16b                     Tlr5   \n",
       "6090    1562673      1303030                  Zg16b                     Tlr6   \n",
       "6091    1562673      1303030                  Zg16b                     Tlr6   \n",
       "6092     620606         2348                    Zp3                   Chrna7   \n",
       "6093     620606       727846                    Zp3                     Zp3r   \n",
       "\n",
       "0     same_as_off_lig  same_as_off_rec  \n",
       "0                True             True  \n",
       "1                True             True  \n",
       "2                True             True  \n",
       "3                True             True  \n",
       "4                True             True  \n",
       "...               ...              ...  \n",
       "6089             True             True  \n",
       "6090             True             True  \n",
       "6091             True             True  \n",
       "6092             True             True  \n",
       "6093             True             True  \n",
       "\n",
       "[6094 rows x 20 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair_orth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "6d9ff6c1-1d1b-48fa-bcd1-0c0f9e8b4545",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_orth[gene_pair_orth[\"same_as_off_lig\"] == False].to_csv(\"data/RGD_ID_check_lig_false.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "95eaa696-aa8c-4487-b64a-6643efc1c0dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Xenbase gene info saved: data/GenePageGeneralInfo_20250731_Xenbase_DB.tsv — shape: (22133, 0)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Filter to only needed columns (if they exist)\n",
    "columns_to_keep = [\n",
    "    \"Xenbase gene page ID\",\n",
    "    \"Gene Symbol\",\n",
    "    \"Gene Name\",\n",
    "    \"Gene Function\",\n",
    "    \"Gene Synonyms\",\n",
    "    \"JGI ID\"\n",
    "]\n",
    "\n",
    "# Rename columns to match your naming style\n",
    "renamed_columns = {\n",
    "    \"Xenbase gene page ID\": \"Xenbase genepage ID\",\n",
    "    \"Gene Symbol\": \"gene symbol\",\n",
    "    \"Gene Name\": \"gene name\",\n",
    "    \"Gene Function\": \"gene function\",\n",
    "    \"Gene Synonyms\": \"gene synonyms\",\n",
    "    \"JGI ID\": \"JGI ID\"\n",
    "}\n",
    "\n",
    "# Sanity-check before selection\n",
    "existing_cols = [col for col in columns_to_keep if col in df_xenbase.columns]\n",
    "df_xenbase = df_xenbase[existing_cols].rename(columns=renamed_columns)\n",
    "\n",
    "# Save to file\n",
    "df_xenbase.to_csv(xenbase_file, sep=\"\\t\", index=False, encoding=\"utf-8\")\n",
    "df_xenbase.to_csv(xenbase_file_today, sep=\"\\t\", index=False, encoding=\"utf-8\")\n",
    "\n",
    "print(f\"✅ Xenbase gene info saved: {xenbase_file_today} — shape: {df_xenbase.shape}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e0245588-4bb9-4dcf-9efe-3ee01a533b02",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_81047/2648667490.py:28: DtypeWarning: Columns (32,34,38,40,50) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  pop_up_info = pd.read_table(\"data/HGNC_gene_info_full.tsv\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3581\n",
      "1097\n",
      "857\n"
     ]
    }
   ],
   "source": [
    "## Function to prepare datatables (cleaning and hyperlinking, adding tool tips, etc) input for the database qmds\n",
    "import sys, os\n",
    "from itables import init_notebook_mode\n",
    "import pandas as pd\n",
    "from itables import show\n",
    "from itables import options\n",
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "import fetchGSheet \n",
    "import warnings\n",
    "import urllib.parse\n",
    "\n",
    "# Suppress SettingWithCopyWarning\n",
    "warnings.simplefilter(\"ignore\", category=UserWarning)\n",
    "\n",
    "\n",
    "# Other vertebrates\n",
    "species_list = [\n",
    "    \"mmusculus\", \"rnorvegicus\", \"drerio\", \"ptroglodytes\", \"ggallus\", \"sscrofa\", \"btaurus\", \n",
    "    \"clfamiliaris\", \"ecaballus\", \"oarambouillet\",\n",
    "    \"cjacchus\", \"mmulatta\", \"xtropicalis\"\n",
    "]\n",
    "\n",
    "# Select only the relevant columns from pop_up_info\n",
    "cols_to_keep = cols_to_keep = list(range(0, 30)) \n",
    "# Step 3: Load file using only the desired columns\n",
    "df = pd.read_table(\"data/HGNC_gene_info_full.tsv\", usecols=cols_to_keep)\n",
    "pop_up_info = pd.read_table(\"data/HGNC_gene_info_full.tsv\")\n",
    "pop_up_info = pop_up_info.rename(columns={\"hgnc_id\": \"HGNC ID\", \n",
    "                                          \"name\": \"Approved name\",\n",
    "                                          \"symbol\": \"Approved symbol\",\n",
    "                                          #\"rgd_id\": \"RGD ID\",\n",
    "                                          #\"mgd_id\": \"MGI ID\", \n",
    "                                          \"alias_symbol\": \"Alias symbol\", # add to table\n",
    "                                          \"prev_symbol\": \"Previous symbol\", # add to table\n",
    "                                          \"date_symbol_changed\": \"Date symbol changed\"\n",
    "                                         })\n",
    "\n",
    "# Keep only first MGI/RGD ID\n",
    "#pop_up_info[\"MGI ID\"] = pop_up_info[\"MGI ID\"].str.split(\"|\").str[0]\n",
    "#pop_up_info[\"RGD ID\"] = pop_up_info[\"RGD ID\"].str.split(\"|\").str[0]\n",
    "\n",
    "pop_up_info[\"Alias symbol\"] = pop_up_info[\"Alias symbol\"].apply(\n",
    "    lambda x: \"N/A\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\"] else x\n",
    ")\n",
    "\n",
    "pop_up_info[\"Previous symbol\"] = pop_up_info[\"Previous symbol\"].apply(\n",
    "    lambda x: \"N/A\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\"] else x\n",
    ")\n",
    "\n",
    "# Replace \"|\" with \", \"\n",
    "pop_up_info[\"Alias symbol\"] = [value.replace(\"|\", \", \") for value in pop_up_info[\"Alias symbol\"]]\n",
    "pop_up_info[\"Previous symbol\"] = [value.replace(\"|\", \", \") for value in pop_up_info[\"Previous symbol\"]]\n",
    "\n",
    "pop_up_info[\"Date symbol changed\"] = pop_up_info[\"Date symbol changed\"].apply(\n",
    "    lambda x: \"N/A\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\"] else x\n",
    ")\n",
    "\n",
    "\n",
    "pop_up_info_lim = pop_up_info[[\"HGNC ID\", \"Approved name\", \"Alias symbol\", # \"MGI ID\", \"RGD ID\"\n",
    "                               \"Approved symbol\", \"Previous symbol\"]] # rm \"Approved symbol\" for now\n",
    "pop_up_info_lim = pop_up_info_lim.drop_duplicates(subset=\"HGNC ID\", keep=\"first\")\n",
    "\n",
    "# Drop columns where all values are NA in gene_pair\n",
    "gene_pair = fetchGSheet.gene_pair_human.dropna(axis=1, how='all')\n",
    "gene_pair = gene_pair[gene_pair['LR_pair_orig'] != '']\n",
    "# for now set source count as triplicates\n",
    "sourceCount = len(gene_pair[['LR_pair_orig']])\n",
    "\n",
    "### KEEP ALL AS OF LATEST input datatable\n",
    "# for now, keep only the following columns\n",
    "# gene_pair = gene_pair[['LR pair', 'Ligand', 'Ligand.HGNC', 'Receptor', 'Receptor.HGNC',\n",
    "#                        'perplexity link', 'PMID', 'binding location', \n",
    "#                        'bind in trans?', 'bidirectional signalling?',\n",
    "#                        'interaction type', 'original source']]\n",
    "\n",
    "gene_pair = gene_pair.dropna(subset=['LR_pair_orig'])\n",
    "\n",
    "# some PMIDs kick in with \",\" so replace\n",
    "gene_pair[\"PMID\"] = [value.replace(\",\", \"\") for value in gene_pair[\"PMID\"]]\n",
    "gene_pair = gene_pair.dropna(subset=['PMID'])\n",
    "\n",
    "### NO NEED FOR MAPPING AS OF LATEST input datatable\n",
    "# Mapping for replacements\n",
    "# mapping = dict(zip(fetchGSheet.src_info['original source'], fetchGSheet.src_info['shortname']))\n",
    "# # Replace values in the column based on the mapping\n",
    "# gene_pair['original source'] = gene_pair['original source'].replace(mapping)\n",
    "\n",
    "gene_pair.columns = gene_pair.columns.str.strip()\n",
    "gene_pair[['Ligand', 'Receptor']] = gene_pair['LR Pair Card'].str.split(' ', n=1, expand=True)\n",
    "\n",
    "## add Ligand/Receptor Location\n",
    "def dedup_locations(loc_str):\n",
    "    # Split, strip, deduplicate, and sort\n",
    "    parts = [loc.strip() for loc in loc_str.split(',') if loc.strip()]\n",
    "    unique_sorted = sorted(set(parts), key=str.lower)  # case-insensitive sort\n",
    "    return unique_sorted\n",
    "\n",
    "def generate_LocToolTip(row, geneloc, loc_col):\n",
    "    ligand = row[loc_col]\n",
    "    original_locations = [loc.strip() for loc in row[\"location\"].split(',')]\n",
    "    original_sources = [src.strip() for src in row[\"source\"].split(',')]\n",
    "\n",
    "    # Get deduplicated locations\n",
    "    unique_locations = dedup_locations(row[\"location\"])\n",
    "\n",
    "    if len(unique_locations) == 1:\n",
    "        # Single tooltip case\n",
    "        location = unique_locations[0]\n",
    "        matching_rows = geneloc[(geneloc[loc_col] == ligand) & (geneloc[\"location\"].str.contains(location))]\n",
    "        all_sources = matching_rows[\"source\"].unique()\n",
    "        sources_str = \", \".join(sorted(set(all_sources)))\n",
    "        return f'<span title=\"based on {sources_str}\">{location}</span>'\n",
    "    else:\n",
    "        # Multiple tooltips — find each (ligand, location) match in original df\n",
    "        spans = []\n",
    "        for loc in unique_locations:\n",
    "            matching_rows = geneloc[\n",
    "                (geneloc[loc_col] == ligand) &\n",
    "                (geneloc[\"location\"].str.contains(loc))\n",
    "            ]\n",
    "            all_sources = matching_rows[\"source\"].unique()\n",
    "            sources_str = \", \".join(sorted(set(all_sources)))\n",
    "            spans.append(f'<span title=\"based on {sources_str}\">{loc}</span>')\n",
    "        return \", \".join(spans)\n",
    "\n",
    "\n",
    "# Group the original loc_info by Ligand\n",
    "ligand_loc = fetchGSheet.ligand_loc.dropna(axis=1, how='all')\n",
    "grouped = ligand_loc.groupby(\"Ligand\").agg({\n",
    "    \"location\": lambda x: ', '.join(x),\n",
    "    \"source\": lambda x: ', '.join(x)\n",
    "}).reset_index()\n",
    "\n",
    "# Generate tooltips\n",
    "grouped[\"Ligand location\"] = grouped.apply(lambda row: generate_LocToolTip(row, ligand_loc,loc_col=\"Ligand\"), axis=1)\n",
    "# create dict\n",
    "mapping_loc = dict(zip(grouped['Ligand'], grouped['Ligand location'])) \n",
    "gene_pair['Ligand location'] = gene_pair['Ligand'].replace(mapping_loc)\n",
    "\n",
    "\n",
    "# Group the original loc_info by Receptor\n",
    "receptor_loc = fetchGSheet.receptor_loc.dropna(axis=1, how='all')\n",
    "grouped = receptor_loc.groupby(\"Receptor\").agg({\n",
    "    \"location\": lambda x: ', '.join(x),\n",
    "    \"source\": lambda x: ', '.join(x)\n",
    "}).reset_index()\n",
    "\n",
    "# Generate tooltips\n",
    "grouped[\"Receptor location\"] = grouped.apply(lambda row: generate_LocToolTip(row, receptor_loc,loc_col=\"Receptor\"), axis=1)\n",
    "# create dict\n",
    "mapping_loc = dict(zip(grouped['Receptor'], grouped['Receptor location'])) \n",
    "gene_pair['Receptor location'] = gene_pair['Receptor'].replace(mapping_loc)\n",
    "\n",
    "\n",
    "# Set missing mappings to 'unknown'\n",
    "gene_pair.loc[gene_pair['Ligand location'] == gene_pair['Ligand'], 'Ligand location'] = 'unknown'\n",
    "gene_pair.loc[gene_pair['Receptor location'] == gene_pair['Receptor'], 'Receptor location'] = 'unknown'\n",
    "# Set \"n/a\" to unknown\n",
    "gene_pair['Ligand location'] = [value.replace(\"n/a\", \"unknown\") for value in gene_pair['Ligand location']]\n",
    "gene_pair['Receptor location'] = [value.replace(\"n/a\", \"unknown\") for value in gene_pair['Receptor location']]\n",
    "\n",
    "# Fetch HGNC IDs from the dataset\n",
    "hgnc_id = [col for col in gene_pair.columns if \"HGNC\" in col]\n",
    "hgnc_id = pd.concat([gene_pair[col] for col in hgnc_id]).unique()\n",
    "\n",
    "gene_pair['Human LR Pair'] = np.where(\n",
    "    gene_pair['Human evidence'] == \"absent in human\", \n",
    "    \"no human ortholog\",                                  \n",
    "    gene_pair['Homo sapiens_ligand'] + \" \" + gene_pair['Homo sapiens_receptor'] \n",
    ")\n",
    "\n",
    "\n",
    "# Rename columns for better clarity\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "    \"LR_pair_orig\": \"LR Pair\",\n",
    "    \"HGNC ligand\": \"Ligand HGNC ID\",\n",
    "    \"HGNC receptor\": \"Receptor HGNC ID\",\n",
    "    \"ENSEMBL ligand\": \"Ligand ENSEMBL ID\",\n",
    "    \"ENSEMBL receptor\": \"Receptor ENSEMBL ID\",\n",
    "    # \"perplexity link\": \"Perplexity\", # will be replaced with actual link later\n",
    "    # \"original source\": \"Database Source\",\n",
    "    \"Ligand location\": \"Ligand Location\",\n",
    "    \"Receptor location\": \"Receptor Location\",\n",
    "    # \"binding location\": \"Binding Location\",\n",
    "    # \"bind in trans?\" : \"Trans-binding\", \n",
    "    # \"bidirectional signalling?\": \"Bidirectional Signalling\",\n",
    "    # \"interaction type\" : \"Interaction Type\"\n",
    "    #\"PMID\": \"PMID support\" # was PMID support\n",
    "})\n",
    "gene_pair = gene_pair.drop(columns=[\"Ligand\", \"Receptor\"])\n",
    "gene_pair = gene_pair.rename(columns={\"Homo sapiens_ligand\": \"Ligand\", \"Homo sapiens_receptor\": \"Receptor\"})\n",
    "# Merge gene_pair with pop_up_info_lim for Ligand(L)\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Ligand HGNC ID', right_on='HGNC ID')\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Ligand Name\", \n",
    "                                     #\"MGI ID\": \"Ligand MGI ID\", # NOT APPLIED YET BUT should be taken from Ensembl BioMart\n",
    "                                     #\"RGD ID\": \"Ligand RGD ID\", # NOT APPLIED YET BUT should be taken from Ensembl BioMart\n",
    "                                      \"Alias symbol\": \"Ligand Aliases\",\n",
    "                                      \"Previous symbol\": \"Ligand Old symbol\",\n",
    "                                     },\n",
    "                            )\n",
    "gene_pair = gene_pair.drop(columns=[\"HGNC ID\", \"Approved symbol\"])\n",
    "\n",
    "\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Receptor HGNC ID', right_on='HGNC ID')\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Receptor Name\",\n",
    "             #\"MGI ID\": \"Receptor MGI ID\",\n",
    "             #\"RGD ID\": \"Receptor RGD ID\",\n",
    "                                      \"Alias symbol\": \"Receptor Aliases\",\n",
    "                                      \"Previous symbol\": \"Receptor Old symbol\",}\n",
    "                            )\n",
    "\n",
    "### AS OF LATEST DB skip pathways for now (code in addPathwayDiseaseAnnot_temp.py)\n",
    "\n",
    "# Add new columns where all Ligand Symbol & Aliases and Receptor Symbol & Aliases merged in one column\n",
    "def format_symbol_aliases(symbol, old_symbol, aliases):\n",
    "    \"\"\"\n",
    "    Formats symbol, old symbols, and aliases.\n",
    "    If the final formatted string would be empty after considering N/A values\n",
    "    and empty inputs, it returns \"mouse-specific\".\n",
    "    Otherwise, it formats based on the presence of old_symbol and aliases,\n",
    "    removing unnecessary parentheses or commas, following the structure:\n",
    "    \"Symbol (Old Symbol, Aliases)\" if both exist.\n",
    "    \"\"\"\n",
    "    # Normalize inputs to empty strings if they are None/NaN or just whitespace\n",
    "    symbol_str = str(symbol).strip()\n",
    "    old_symbol_str = str(old_symbol).strip()\n",
    "    aliases_str = str(aliases).strip()\n",
    "\n",
    "    # Filter out values that are empty strings or \"N/A\" for old_symbol and aliases\n",
    "    parts_for_join = []\n",
    "    if old_symbol_str and old_symbol_str != \"N/A\":\n",
    "        parts_for_join.append(old_symbol_str)\n",
    "    if aliases_str and aliases_str != \"N/A\":\n",
    "        parts_for_join.append(aliases_str)\n",
    "\n",
    "    # Construct the preliminary result based on your original logic:\n",
    "    # \"symbol (old_symbol, aliases)\" if parts_for_join is not empty, else \"symbol\"\n",
    "    if parts_for_join:\n",
    "        prelim_result = f\"{symbol_str} ({', '.join(parts_for_join)})\"\n",
    "    else:\n",
    "        prelim_result = symbol_str # Just the symbol if no old_symbol or aliases\n",
    "\n",
    "    return prelim_result\n",
    "\n",
    "# This is crucial for consistent handling by the function before processing \"N/A\".\n",
    "gene_pair['Ligand'] = gene_pair['Ligand'].fillna('')\n",
    "gene_pair['Ligand Old symbol'] = gene_pair['Ligand Old symbol'].fillna('')\n",
    "gene_pair['Ligand Aliases'] = gene_pair['Ligand Aliases'].fillna('')\n",
    "\n",
    "\n",
    "# to later check which ligand-receptor pairs are non-human\n",
    "def is_mouse_specific(name):\n",
    "    if not isinstance(name, str):\n",
    "        return False\n",
    "    name = name.strip()  # remove leading/trailing spaces\n",
    "    return any(c.islower() for c in name[1:])\n",
    "\n",
    "gene_pair = gene_pair.drop(columns=[\"HGNC ID\"])\n",
    "\n",
    "gene_pair['Ligand Symbols'] = gene_pair.apply(\n",
    "    lambda row: \"no human ortholog\" if is_mouse_specific(row['Ligand']) \n",
    "               else format_symbol_aliases(row['Ligand'], row['Ligand Old symbol'], row['Ligand Aliases']),\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "\n",
    "# This is crucial for consistent handling by the function before processing \"N/A\".\n",
    "gene_pair['Receptor'] = gene_pair['Receptor'].fillna('')\n",
    "gene_pair['Receptor Old symbol'] = gene_pair['Receptor Old symbol'].fillna('')\n",
    "gene_pair['Receptor Aliases'] = gene_pair['Receptor Aliases'].fillna('')\n",
    "\n",
    "gene_pair['Receptor Symbols'] = gene_pair.apply(\n",
    "    lambda row: \"no human ortholog\" if is_mouse_specific(row['Receptor']) \n",
    "                else format_symbol_aliases(row['Receptor'], row['Receptor Old symbol'], row['Receptor Aliases']),\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "### tooltips \n",
    "gene_pair[\"Ligand Symbols\"] = [\n",
    "    f'<span title=\"{aliases}\">{aliases}</span>'\n",
    "    for aliases in gene_pair[\"Ligand Symbols\"]\n",
    "]\n",
    "gene_pair[\"Receptor Symbols\"] = [\n",
    "    f'<span title=\"{aliases}\">{aliases}</span>'\n",
    "    for aliases in gene_pair[\"Receptor Symbols\"]\n",
    "]\n",
    "\n",
    "# might be used later just save info for now (for mouse cards)\n",
    "grab_mouse_info = gene_pair[\"LR Pair Card\"][gene_pair[\"Human evidence\"].isin([\"absent in human\", \"not conserved\"])]\n",
    "grab_mouse_info = grab_mouse_info.unique()\n",
    "grab_mouse_info\n",
    "\n",
    "# Add an empty Perplexity column filled with None (just to save it's order\n",
    "gene_pair['Perplexity'] = None\n",
    "#gene_pair = gene_pair.drop(columns=[\"Approved symbol_x\", \"Approved symbol_y\"])\n",
    "\n",
    "### For latest DB, skip (code saved as addOrth_temp.py)\n",
    "\n",
    "# Add\n",
    "first_columns=['LR Pair Card', 'Human LR Pair', 'Ligand', 'Receptor', 'Ligand Symbols', 'Receptor Symbols', 'Ligand Location', 'Receptor Location',\t'Ligand HGNC ID', 'Receptor HGNC ID', 'Perplexity', 'Human evidence'] # 'Database Source'\n",
    "\n",
    "end_columns=['PMID', 'Pair_species', 'lig_species', 'rec_species', 'ligand_orig', 'receptor_orig']\n",
    "gene_pair = gene_pair[first_columns + [col for col in gene_pair.columns if col not in first_columns + end_columns] + end_columns]\n",
    "# gene_pair = gene_pair[first_columns + [col for col in gene_pair.columns if col not in first_columns]]\n",
    "\n",
    "# number of unique vars (Human and Mouse both counted)\n",
    "\n",
    "lrPairsCount = len(gene_pair[\"LR Pair Card\"].unique())\n",
    "print(lrPairsCount)\n",
    "\n",
    "ligandCount = len(gene_pair[\"Ligand\"].unique())\n",
    "print(ligandCount)\n",
    "\n",
    "receptorCount = len(gene_pair[\"Receptor\"].unique())\n",
    "print(receptorCount)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "5a63bb3f-40a2-48ec-ad0e-10078e0305c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_mouse_card= fetchGSheet.gene_pair_mouse\n",
    "gene_pair_mouse_card['Ligand Official Symbol'] = 'NA'\n",
    "gene_pair_mouse_card['Ligand Official Symbol'] =  gene_pair_mouse_card.apply(\n",
    "    lambda row: mapping_mouse_symbol.get(row['MGI ligand'], row['Ligand Official Symbol'])\n",
    "    if pd.notna(row['MGI ligand']) else row['Ligand Official Symbol'],\n",
    "    axis=1\n",
    ")\n",
    "gene_pair_mouse_card['Receptor Official Symbol'] = 'NA'\n",
    "gene_pair_mouse_card['Receptor Official Symbol'] =  gene_pair_mouse_card.apply(\n",
    "    lambda row: mapping_mouse_symbol.get(row['MGI receptor'], row['Receptor Official Symbol'])\n",
    "    if pd.notna(row['MGI receptor']) else row['Receptor Official Symbol'],\n",
    "    axis=1\n",
    ")\n",
    "gene_pair_mouse_card[\"same_as_off_lig\"] = (\n",
    "    gene_pair_mouse_card[\"Mouse_ligand\"] == gene_pair_mouse_card[\"Ligand Official Symbol\"]\n",
    ")\n",
    "\n",
    "gene_pair_mouse_card[\"same_as_off_rec\"] = (\n",
    "    gene_pair_mouse_card[\"Mouse_receptor\"] == gene_pair_mouse_card[\"Receptor Official Symbol\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "aed99b5f-4152-4398-9f90-51b17853671f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_mouse_card.to_csv(\"data/MGI_ID_check.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "4d38929f-5c31-4920-86ba-cfc819ff31e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_cards = gene_pair[gene_pair[\"LR Pair Card\"].isin(grab_mouse_info)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "6fbbd822-c5a3-44f8-8916-ab1c7f27aa2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import jinja2\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "import base64\n",
    "import re\n",
    "# Add the src directory to the path for importing modules\n",
    "sys.path.append(os.path.abspath(\"src\"))\n",
    "\n",
    "# Import necessary modules from your existing src files\n",
    "# Ensure createDataTable and createFunctionalAnnotTable are in your 'src' directory\n",
    "from fetchGSheet import gene_pair_mouse\n",
    "from createDataTable import pop_up_info, gene_pair0, generate_perplexity_links, gene_pair00, is_mouse_specific, grab_mouse_info\n",
    "from createFunctionalAnnotTable import gene_pair_annot_ligand, gene_pair_annot_receptor\n",
    "\n",
    "# Test or all\n",
    "test = True\n",
    "test_genes = [\"H60A Klrk1\", \"H60B Klrk1\", \"VEGFA NRP1\", \"THPO MPL\", \"FGF1 FGFR3\", \"Lair1 Lilrb4A\"] # Example genes\n",
    "# --- Paths ---\n",
    "MERGED_TEMPLATE_PATH = 'HTML/mergedCard_tabs.html'\n",
    "OUTPUT_DIR = 'data/cards/' # New output directory for combined files\n",
    "\n",
    "# function for replacing visible text:\n",
    "def update_link_text_with_symbol(html_str, new_symbol):\n",
    "    \"\"\"\n",
    "    Replace the visible text in an anchor tag with the provided symbol.\n",
    "    The HGNC ID is extracted from the href and left unchanged.\n",
    "    \"\"\"\n",
    "    # Only proceed if input is valid\n",
    "    if not isinstance(html_str, str) or not html_str.strip():\n",
    "        return html_str\n",
    "    \n",
    "    return re.sub(r\">(.*?)</a>\", f\">{new_symbol}</a>\", html_str) #<i class='fa-solid fa-arrow-up-right-from-square' style='margin-left:4px;'></i></a>\n",
    "\n",
    "# Create Perplexity link\n",
    "\n",
    "# Recreate Perplexity link\n",
    "def create_url_basic(symbol):\n",
    "          label = \"Perplexity (LLM)\"\n",
    "          query = f\"What diseases is {symbol} implicated in?\"\n",
    "          encoded_query = query.replace(\" \", \"%20\")\n",
    "          output = f'<a href=\"https://www.perplexity.ai/search?q={encoded_query}\" target=\"_blank\">{label}</a>' #<i class=\"fa-solid fa-arrow-up-right-from-square\" style=\"margin-left:4px;\"></i>\n",
    "          return output\n",
    "\n",
    "# --- Load and Preprocess Data (Combined from both scripts) ---\n",
    "\n",
    "# Load PubMed data (from createPMIDpages.py)\n",
    "pubmed_data = pd.read_csv(\"data/pubmed_results.csv\")\n",
    "pubmed_data[\"Year\"] = pubmed_data[\"Year\"].astype(str).str.replace(\".0\", \"\", regex=False).astype(int)\n",
    "pubmed_data[\"PMID\"] = pubmed_data[\"PMID\"].astype(str)\n",
    "pubmed_data = pubmed_data.reset_index(drop=True)\n",
    "\n",
    "### For latest DB, skip\n",
    "# # Load LLM results (from createPMIDpages.py)\n",
    "# bio_keywords = pd.read_csv(\"data/llm_results.csv\")\n",
    "\n",
    "# --- Prepare gene_pair00 for PMID section (from createPMIDpages.py) ---\n",
    "# gene_pair00 is used for PMID and Keywords, so it needs the '—' placeholder\n",
    "# Ensure gene_pair00 is a copy to avoid SettingWithCopyWarning later\n",
    "gene_pair00_copy = gene_pair00.copy()\n",
    "#gene_pair00_copy[\"Human LR Pair\"] = gene_pair00_copy[\"Human LR Pair\"].str.replace(\" \", \"—\")\n",
    "\n",
    "# Merge with LLM results\n",
    "# gene_pair000 = gene_pair00_copy.merge(bio_keywords, how='left', left_on=\"Human LR Pair\", right_on='Human LR Pair')\n",
    "# gene_pair000[\"Relevance Keywords\"] = gene_pair000[\"Relevance Keywords\"].astype(str)\n",
    " # Ensure string type\n",
    "# gene_pair000[\"Human LR Pair\"] = gene_pair000[\"Human LR Pair\"].astype(str)\n",
    "\n",
    "### For latest DB,\n",
    "gene_pair000 = gene_pair00.copy()\n",
    "gene_pair000[\"LR Pair Card\"] = gene_pair000[\"LR Pair Card\"].astype(str)\n",
    "# --- Prepare gene_pair0 for Card section (from createCards.py) ---\n",
    "# gene_pair0 is used for card details, it should retain spaces for splitting gene names\n",
    "# Ensure gene_pair0 is a copy to avoid SettingWithCopyWarning later\n",
    "gene_pair0_copy = gene_pair0.copy()\n",
    "\n",
    "# grab the pairs by interaction id that has human evidence that is absent in human\n",
    "# mouse_interaction_ids = gene_pair0_copy[gene_pair0_copy['Ligand'].apply(is_mouse_specific)]['Interaction ID'].tolist()\n",
    "mouse_interaction_ids = gene_pair0_copy[\"Interaction ID\"][gene_pair0_copy[\"Human evidence\"].isin([\"absent in human\", \"not conserved\"])]\n",
    "\n",
    "### For for latest DB, remove for now\n",
    "# # Add Disease (specific) to cards\n",
    "# df_disease = pd.read_csv(\"data/disease_annotations_per_pair.csv\")\n",
    "# df_disease = df_disease.groupby('interaction')['disease'].apply(', '.join).reset_index()\n",
    "# mapping_disease = dict(zip(df_disease['interaction'], df_disease['disease']))\n",
    "# gene_pair0_copy[\"Disease\"] = gene_pair0_copy['Human LR Pair'].map(mapping_disease).fillna(\"unknown\")\n",
    "\n",
    "# gene_pair0_copy = generate_perplexity_links(\n",
    "#     gene_pair0_copy,\n",
    "#     pathway_col=\"Disease\",\n",
    "#     default_query_template=\"What-diseases-is-the-ligand-receptor-pair-{pair}-associated-with\"\n",
    "# )\n",
    "\n",
    "### SHOULD BE ACTIVATED ONCE WE DECIDE TO OPEN DB \n",
    "# # Hide for now (linking to actual PMID database\n",
    "# gene_pair0_copy[\"Interaction ID\"] = gene_pair0_copy[\"Interaction ID\"].apply(\n",
    "#     lambda x: f\"<a href='https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database/filter/{x}.html'>{x}</a>\"\n",
    "# )\n",
    "\n",
    "# Add external link icon\n",
    "# icon_html = '<i class=\"fa-solid fa-arrow-up-right-from-square\" style=\"margin-left:4px;\"></i></a>'\n",
    "# columns_to_update = [\n",
    "#     \"KEGG Pathway\", \"PROGENy Pathway\", \"Cancer-related\",\n",
    "#     \"Disease Type\", \"Disease\"\n",
    "# ]\n",
    "# for col in columns_to_update:\n",
    "#     gene_pair0_copy[col] = gene_pair0_copy[col].str.replace(\n",
    "#         \"</a>\", icon_html, regex=False\n",
    "#     )\n",
    "\n",
    "# Add missing ligand name for mouse-specific\n",
    "mouse_info = pd.read_csv(\"data/MRK_Merged_MGI_DB.tsv\", sep=\"\\t\", dtype=str)\n",
    "mapping_mouse_name = dict(zip(mouse_info['MGI Marker Accession ID'], mouse_info['Marker Name']))\n",
    "mapping_mouse_aliases = dict(zip(mouse_info['MGI Marker Accession ID'], mouse_info['Aliases']))\n",
    "mapping_mouse_uniprot = dict(zip(mouse_info['MGI Marker Accession ID'], mouse_info['UniProt IDs']))\n",
    "mapping_mouse_ncbi = dict(zip(mouse_info['MGI Marker Accession ID'], mouse_info['EntrezGene ID']))\n",
    "\n",
    "# Combine ligand and receptor columns from mouse DB input\n",
    "gene_pair_mouse_card = gene_pair_mouse[gene_pair_mouse[\"LR Pair Card\"].isin(grab_mouse_info)]\n",
    "mouse_id_map = pd.concat([\n",
    "    gene_pair_mouse_card[['MGI ligand', 'ENSEMBL ligand']].rename(columns={\n",
    "        'MGI ligand': 'MGI ID', 'ENSEMBL ligand': 'ENSEMBL ID'\n",
    "    }),\n",
    "    gene_pair_mouse_card[['MGI receptor', 'ENSEMBL receptor']].rename(columns={\n",
    "        'MGI receptor': 'MGI ID', 'ENSEMBL receptor': 'ENSEMBL ID'\n",
    "    })\n",
    "])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "4c884434-df7b-4dd4-9d5e-a9dab6d9e29b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_75207/4290407472.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_pair_mouse_card['Ligand Official Symbol'] = 'NA'\n",
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_75207/4290407472.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_pair_mouse_card['Ligand Official Symbol'] =  gene_pair_mouse_card.apply(\n",
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_75207/4290407472.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_pair_mouse_card['Receptor Official Symbol'] = 'NA'\n",
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_75207/4290407472.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_pair_mouse_card['Receptor Official Symbol'] =  gene_pair_mouse_card.apply(\n"
     ]
    }
   ],
   "source": [
    "gene_pair_mouse_card['Ligand Official Symbol'] = 'NA'\n",
    "gene_pair_mouse_card['Ligand Official Symbol'] =  gene_pair_mouse_card.apply(\n",
    "    lambda row: mapping_mouse_symbol.get(row['MGI ligand'], row['Ligand Official Symbol'])\n",
    "    if pd.notna(row['MGI ligand']) else row['Ligand Official Symbol'],\n",
    "    axis=1\n",
    ")\n",
    "gene_pair_mouse_card['Receptor Official Symbol'] = 'NA'\n",
    "gene_pair_mouse_card['Receptor Official Symbol'] =  gene_pair_mouse_card.apply(\n",
    "    lambda row: mapping_mouse_symbol.get(row['MGI receptor'], row['Receptor Official Symbol'])\n",
    "    if pd.notna(row['MGI receptor']) else row['Receptor Official Symbol'],\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "1467d8c0-8721-4978-81b2-4462b0af9d59",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_75207/3070642398.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_pair_mouse_card[\"same_as_off_lig\"] = (\n",
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_75207/3070642398.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_pair_mouse_card[\"same_as_off_rec\"] = (\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LR_pair_orig</th>\n",
       "      <th>PMID</th>\n",
       "      <th>lig_species</th>\n",
       "      <th>rec_species</th>\n",
       "      <th>ligand_orig</th>\n",
       "      <th>receptor_orig</th>\n",
       "      <th>Pair_species</th>\n",
       "      <th>Mouse evidence</th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>Mouse_ligand</th>\n",
       "      <th>Mouse_receptor</th>\n",
       "      <th>ENSEMBL ligand</th>\n",
       "      <th>ENSEMBL receptor</th>\n",
       "      <th>MGI ligand</th>\n",
       "      <th>MGI receptor</th>\n",
       "      <th>Ligand Official Symbol</th>\n",
       "      <th>Receptor Official Symbol</th>\n",
       "      <th>same_as_off_lig</th>\n",
       "      <th>same_as_off_rec</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>770</th>\n",
       "      <td>Ccl12 Ccr2</td>\n",
       "      <td>16543609</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Ccl12</td>\n",
       "      <td>Ccr2</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>Ccl12 Ccr2</td>\n",
       "      <td>Ccl12</td>\n",
       "      <td>Ccr2</td>\n",
       "      <td>ENSMUSG00000035352</td>\n",
       "      <td>ENSMUSG00000049103</td>\n",
       "      <td>MGI:108224</td>\n",
       "      <td>MGI:106185</td>\n",
       "      <td>Ccl12</td>\n",
       "      <td>Ccr2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>967</th>\n",
       "      <td>Ccl6 Ccr1</td>\n",
       "      <td>31551151</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>Ccl6</td>\n",
       "      <td>Ccr1</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>Ccl6 Ccr1</td>\n",
       "      <td>Ccl6</td>\n",
       "      <td>Ccr1</td>\n",
       "      <td>ENSMUSG00000018927</td>\n",
       "      <td>ENSMUSG00000025804</td>\n",
       "      <td>MGI:98263</td>\n",
       "      <td>MGI:104618</td>\n",
       "      <td>Ccl6</td>\n",
       "      <td>Ccr1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>968</th>\n",
       "      <td>Ccl6 Ccr1</td>\n",
       "      <td>33640900</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Ccl6</td>\n",
       "      <td>Ccr1</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>Ccl6 Ccr1</td>\n",
       "      <td>Ccl6</td>\n",
       "      <td>Ccr1</td>\n",
       "      <td>ENSMUSG00000018927</td>\n",
       "      <td>ENSMUSG00000025804</td>\n",
       "      <td>MGI:98263</td>\n",
       "      <td>MGI:104618</td>\n",
       "      <td>Ccl6</td>\n",
       "      <td>Ccr1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>993</th>\n",
       "      <td>Ccl9 Ccr1</td>\n",
       "      <td>12397598</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Ccl9</td>\n",
       "      <td>Ccr1</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>Ccl9 Ccr1</td>\n",
       "      <td>Ccl9</td>\n",
       "      <td>Ccr1</td>\n",
       "      <td>ENSMUSG00000019122</td>\n",
       "      <td>ENSMUSG00000025804</td>\n",
       "      <td>MGI:104533</td>\n",
       "      <td>MGI:104618</td>\n",
       "      <td>Ccl9</td>\n",
       "      <td>Ccr1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1781</th>\n",
       "      <td>Defb2 Ccr6</td>\n",
       "      <td>11714836</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Defb2</td>\n",
       "      <td>Ccr6</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>Defb2 Ccr6</td>\n",
       "      <td>Defb2</td>\n",
       "      <td>Ccr6</td>\n",
       "      <td>ENSMUSG00000006570</td>\n",
       "      <td>ENSMUSG00000040899</td>\n",
       "      <td>MGI:1338754</td>\n",
       "      <td>MGI:1333797</td>\n",
       "      <td>Defb2</td>\n",
       "      <td>Ccr6</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1782</th>\n",
       "      <td>Defb2 Tlr4</td>\n",
       "      <td>12411706</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Defb2</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>Defb2 Tlr4</td>\n",
       "      <td>Defb2</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>ENSMUSG00000006570</td>\n",
       "      <td>ENSMUSG00000039005</td>\n",
       "      <td>MGI:1338754</td>\n",
       "      <td>MGI:96824</td>\n",
       "      <td>Defb2</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2272</th>\n",
       "      <td>Fcna Tlr4</td>\n",
       "      <td>28844702</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Fcna</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>Fcna Tlr4</td>\n",
       "      <td>Fcna</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>ENSMUSG00000026938</td>\n",
       "      <td>ENSMUSG00000039005</td>\n",
       "      <td>MGI:1340905</td>\n",
       "      <td>MGI:96824</td>\n",
       "      <td>Fcna</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4318</th>\n",
       "      <td>Pcdhb10 Pcdhb10</td>\n",
       "      <td>25171406</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Pcdhb10</td>\n",
       "      <td>Pcdhb10</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>Pcdhb10 Pcdhb10</td>\n",
       "      <td>Pcdhb10</td>\n",
       "      <td>Pcdhb10</td>\n",
       "      <td>ENSMUSG00000045657</td>\n",
       "      <td>ENSMUSG00000045657</td>\n",
       "      <td>MGI:2136745</td>\n",
       "      <td>MGI:2136745</td>\n",
       "      <td>Pcdhb10</td>\n",
       "      <td>Pcdhb10</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4319</th>\n",
       "      <td>Pcdhb11 Pcdhb11</td>\n",
       "      <td>25171406</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Pcdhb11</td>\n",
       "      <td>Pcdhb11</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>Pcdhb11 Pcdhb11</td>\n",
       "      <td>Pcdhb11</td>\n",
       "      <td>Pcdhb11</td>\n",
       "      <td>ENSMUSG00000051486</td>\n",
       "      <td>ENSMUSG00000051486</td>\n",
       "      <td>MGI:2136746</td>\n",
       "      <td>MGI:2136746</td>\n",
       "      <td>Pcdhb11</td>\n",
       "      <td>Pcdhb11</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4321</th>\n",
       "      <td>Pcdhb12 Pcdhb12</td>\n",
       "      <td>25171406</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Pcdhb12</td>\n",
       "      <td>Pcdhb12</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>Pcdhb12 Pcdhb12</td>\n",
       "      <td>Pcdhb12</td>\n",
       "      <td>Pcdhb12</td>\n",
       "      <td>ENSMUSG00000043458</td>\n",
       "      <td>ENSMUSG00000043458</td>\n",
       "      <td>MGI:2136747</td>\n",
       "      <td>MGI:2136747</td>\n",
       "      <td>Pcdhb12</td>\n",
       "      <td>Pcdhb12</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4323</th>\n",
       "      <td>Pcdhb14 Pcdhb14</td>\n",
       "      <td>25171406</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Pcdhb14</td>\n",
       "      <td>Pcdhb14</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>Pcdhb14 Pcdhb14</td>\n",
       "      <td>Pcdhb14</td>\n",
       "      <td>Pcdhb14</td>\n",
       "      <td>ENSMUSG00000044043</td>\n",
       "      <td>ENSMUSG00000044043</td>\n",
       "      <td>MGI:2136749</td>\n",
       "      <td>MGI:2136749</td>\n",
       "      <td>Pcdhb14</td>\n",
       "      <td>Pcdhb14</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4331</th>\n",
       "      <td>Pcdhb21 Pcdhb21</td>\n",
       "      <td>25171406</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Pcdhb21</td>\n",
       "      <td>Pcdhb21</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>Pcdhb21 Pcdhb21</td>\n",
       "      <td>Pcdhb21</td>\n",
       "      <td>Pcdhb21</td>\n",
       "      <td>ENSMUSG00000044022</td>\n",
       "      <td>ENSMUSG00000044022</td>\n",
       "      <td>MGI:2136759</td>\n",
       "      <td>MGI:2136759</td>\n",
       "      <td>Pcdhb21</td>\n",
       "      <td>Pcdhb21</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4334</th>\n",
       "      <td>Pcdhb4 Pcdhb4</td>\n",
       "      <td>25171406</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Pcdhb4</td>\n",
       "      <td>Pcdhb4</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>Pcdhb4 Pcdhb4</td>\n",
       "      <td>Pcdhb4</td>\n",
       "      <td>Pcdhb4</td>\n",
       "      <td>ENSMUSG00000045689</td>\n",
       "      <td>ENSMUSG00000045689</td>\n",
       "      <td>MGI:2136738</td>\n",
       "      <td>MGI:2136738</td>\n",
       "      <td>Pcdhb4</td>\n",
       "      <td>Pcdhb4</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4338</th>\n",
       "      <td>Pcdhb7 Pcdhb7</td>\n",
       "      <td>25171406</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Pcdhb7</td>\n",
       "      <td>Pcdhb7</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>Pcdhb7 Pcdhb7</td>\n",
       "      <td>Pcdhb7</td>\n",
       "      <td>Pcdhb7</td>\n",
       "      <td>ENSMUSG00000045062</td>\n",
       "      <td>ENSMUSG00000045062</td>\n",
       "      <td>MGI:2136741</td>\n",
       "      <td>MGI:2136741</td>\n",
       "      <td>Pcdhb7</td>\n",
       "      <td>Pcdhb7</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4339</th>\n",
       "      <td>Pcdhb8 Pcdhb8</td>\n",
       "      <td>25171406</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>Pcdhb8 Pcdhb8</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>ENSMUSG00000045876</td>\n",
       "      <td>ENSMUSG00000045876</td>\n",
       "      <td>MGI:2136742</td>\n",
       "      <td>MGI:2136742</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4340</th>\n",
       "      <td>Pcdhb8 Pcdhb8</td>\n",
       "      <td>35253643</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>Pcdhb8 Pcdhb8</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>ENSMUSG00000045876</td>\n",
       "      <td>ENSMUSG00000045876</td>\n",
       "      <td>MGI:2136742</td>\n",
       "      <td>MGI:2136742</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4341</th>\n",
       "      <td>Pcdhb9 Pcdhb9</td>\n",
       "      <td>25171406</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Pcdhb9</td>\n",
       "      <td>Pcdhb9</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>Pcdhb9 Pcdhb9</td>\n",
       "      <td>Pcdhb9</td>\n",
       "      <td>Pcdhb9</td>\n",
       "      <td>ENSMUSG00000051242</td>\n",
       "      <td>ENSMUSG00000051242</td>\n",
       "      <td>MGI:2136744</td>\n",
       "      <td>MGI:2136744</td>\n",
       "      <td>Pcdhb9</td>\n",
       "      <td>Pcdhb9</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4367</th>\n",
       "      <td>Pcdhgb8 Pcdhgb8</td>\n",
       "      <td>25171406</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>Pcdhgb8 Pcdhgb8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>ENSMUSG00000103081</td>\n",
       "      <td>ENSMUSG00000103081</td>\n",
       "      <td>MGI:1935200</td>\n",
       "      <td>MGI:1935200</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4903</th>\n",
       "      <td>Saa3 Tlr4</td>\n",
       "      <td>23858030</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Saa3</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>Saa3 Tlr4</td>\n",
       "      <td>Saa3</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>ENSMUSG00000040026</td>\n",
       "      <td>ENSMUSG00000039005</td>\n",
       "      <td>MGI:98223</td>\n",
       "      <td>MGI:96824</td>\n",
       "      <td>Saa3</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "0        LR_pair_orig      PMID        lig_species        rec_species  \\\n",
       "770        Ccl12 Ccr2  16543609       Mus Musculus       Mus Musculus   \n",
       "967         Ccl6 Ccr1  31551151  Rattus norvegicus  Rattus norvegicus   \n",
       "968         Ccl6 Ccr1  33640900       Mus Musculus       Mus Musculus   \n",
       "993         Ccl9 Ccr1  12397598       Mus Musculus       Mus Musculus   \n",
       "1781       Defb2 Ccr6  11714836       Mus Musculus       Mus Musculus   \n",
       "1782       Defb2 Tlr4  12411706       Mus Musculus       Mus Musculus   \n",
       "2272        Fcna Tlr4  28844702       Mus Musculus       Mus Musculus   \n",
       "4318  Pcdhb10 Pcdhb10  25171406       Mus Musculus       Mus Musculus   \n",
       "4319  Pcdhb11 Pcdhb11  25171406       Mus Musculus       Mus Musculus   \n",
       "4321  Pcdhb12 Pcdhb12  25171406       Mus Musculus       Mus Musculus   \n",
       "4323  Pcdhb14 Pcdhb14  25171406       Mus Musculus       Mus Musculus   \n",
       "4331  Pcdhb21 Pcdhb21  25171406       Mus Musculus       Mus Musculus   \n",
       "4334    Pcdhb4 Pcdhb4  25171406       Mus Musculus       Mus Musculus   \n",
       "4338    Pcdhb7 Pcdhb7  25171406       Mus Musculus       Mus Musculus   \n",
       "4339    Pcdhb8 Pcdhb8  25171406       Mus Musculus       Mus Musculus   \n",
       "4340    Pcdhb8 Pcdhb8  35253643       Mus Musculus       Mus Musculus   \n",
       "4341    Pcdhb9 Pcdhb9  25171406       Mus Musculus       Mus Musculus   \n",
       "4367  Pcdhgb8 Pcdhgb8  25171406       Mus Musculus       Mus Musculus   \n",
       "4903        Saa3 Tlr4  23858030       Mus Musculus       Mus Musculus   \n",
       "\n",
       "0    ligand_orig receptor_orig       Pair_species Mouse evidence  \\\n",
       "770        Ccl12          Ccr2       Mus Musculus         DIRECT   \n",
       "967         Ccl6          Ccr1  Rattus norvegicus   CONSERVATION   \n",
       "968         Ccl6          Ccr1       Mus Musculus         DIRECT   \n",
       "993         Ccl9          Ccr1       Mus Musculus         DIRECT   \n",
       "1781       Defb2          Ccr6       Mus Musculus         DIRECT   \n",
       "1782       Defb2          Tlr4       Mus Musculus         DIRECT   \n",
       "2272        Fcna          Tlr4       Mus Musculus         DIRECT   \n",
       "4318     Pcdhb10       Pcdhb10       Mus Musculus         DIRECT   \n",
       "4319     Pcdhb11       Pcdhb11       Mus Musculus         DIRECT   \n",
       "4321     Pcdhb12       Pcdhb12       Mus Musculus         DIRECT   \n",
       "4323     Pcdhb14       Pcdhb14       Mus Musculus         DIRECT   \n",
       "4331     Pcdhb21       Pcdhb21       Mus Musculus         DIRECT   \n",
       "4334      Pcdhb4        Pcdhb4       Mus Musculus         DIRECT   \n",
       "4338      Pcdhb7        Pcdhb7       Mus Musculus         DIRECT   \n",
       "4339      Pcdhb8        Pcdhb8       Mus Musculus         DIRECT   \n",
       "4340      Pcdhb8        Pcdhb8       Mus Musculus         DIRECT   \n",
       "4341      Pcdhb9        Pcdhb9       Mus Musculus         DIRECT   \n",
       "4367     Pcdhgb8       Pcdhgb8       Mus Musculus         DIRECT   \n",
       "4903        Saa3          Tlr4       Mus Musculus         DIRECT   \n",
       "\n",
       "0        LR Pair Card Mouse_ligand Mouse_receptor      ENSEMBL ligand  \\\n",
       "770        Ccl12 Ccr2        Ccl12           Ccr2  ENSMUSG00000035352   \n",
       "967         Ccl6 Ccr1         Ccl6           Ccr1  ENSMUSG00000018927   \n",
       "968         Ccl6 Ccr1         Ccl6           Ccr1  ENSMUSG00000018927   \n",
       "993         Ccl9 Ccr1         Ccl9           Ccr1  ENSMUSG00000019122   \n",
       "1781       Defb2 Ccr6        Defb2           Ccr6  ENSMUSG00000006570   \n",
       "1782       Defb2 Tlr4        Defb2           Tlr4  ENSMUSG00000006570   \n",
       "2272        Fcna Tlr4         Fcna           Tlr4  ENSMUSG00000026938   \n",
       "4318  Pcdhb10 Pcdhb10      Pcdhb10        Pcdhb10  ENSMUSG00000045657   \n",
       "4319  Pcdhb11 Pcdhb11      Pcdhb11        Pcdhb11  ENSMUSG00000051486   \n",
       "4321  Pcdhb12 Pcdhb12      Pcdhb12        Pcdhb12  ENSMUSG00000043458   \n",
       "4323  Pcdhb14 Pcdhb14      Pcdhb14        Pcdhb14  ENSMUSG00000044043   \n",
       "4331  Pcdhb21 Pcdhb21      Pcdhb21        Pcdhb21  ENSMUSG00000044022   \n",
       "4334    Pcdhb4 Pcdhb4       Pcdhb4         Pcdhb4  ENSMUSG00000045689   \n",
       "4338    Pcdhb7 Pcdhb7       Pcdhb7         Pcdhb7  ENSMUSG00000045062   \n",
       "4339    Pcdhb8 Pcdhb8       Pcdhb8         Pcdhb8  ENSMUSG00000045876   \n",
       "4340    Pcdhb8 Pcdhb8       Pcdhb8         Pcdhb8  ENSMUSG00000045876   \n",
       "4341    Pcdhb9 Pcdhb9       Pcdhb9         Pcdhb9  ENSMUSG00000051242   \n",
       "4367  Pcdhgb8 Pcdhgb8      Pcdhgb8        Pcdhgb8  ENSMUSG00000103081   \n",
       "4903        Saa3 Tlr4         Saa3           Tlr4  ENSMUSG00000040026   \n",
       "\n",
       "0       ENSEMBL receptor   MGI ligand MGI receptor Ligand Official Symbol  \\\n",
       "770   ENSMUSG00000049103   MGI:108224   MGI:106185                  Ccl12   \n",
       "967   ENSMUSG00000025804    MGI:98263   MGI:104618                   Ccl6   \n",
       "968   ENSMUSG00000025804    MGI:98263   MGI:104618                   Ccl6   \n",
       "993   ENSMUSG00000025804   MGI:104533   MGI:104618                   Ccl9   \n",
       "1781  ENSMUSG00000040899  MGI:1338754  MGI:1333797                  Defb2   \n",
       "1782  ENSMUSG00000039005  MGI:1338754    MGI:96824                  Defb2   \n",
       "2272  ENSMUSG00000039005  MGI:1340905    MGI:96824                   Fcna   \n",
       "4318  ENSMUSG00000045657  MGI:2136745  MGI:2136745                Pcdhb10   \n",
       "4319  ENSMUSG00000051486  MGI:2136746  MGI:2136746                Pcdhb11   \n",
       "4321  ENSMUSG00000043458  MGI:2136747  MGI:2136747                Pcdhb12   \n",
       "4323  ENSMUSG00000044043  MGI:2136749  MGI:2136749                Pcdhb14   \n",
       "4331  ENSMUSG00000044022  MGI:2136759  MGI:2136759                Pcdhb21   \n",
       "4334  ENSMUSG00000045689  MGI:2136738  MGI:2136738                 Pcdhb4   \n",
       "4338  ENSMUSG00000045062  MGI:2136741  MGI:2136741                 Pcdhb7   \n",
       "4339  ENSMUSG00000045876  MGI:2136742  MGI:2136742                 Pcdhb8   \n",
       "4340  ENSMUSG00000045876  MGI:2136742  MGI:2136742                 Pcdhb8   \n",
       "4341  ENSMUSG00000051242  MGI:2136744  MGI:2136744                 Pcdhb9   \n",
       "4367  ENSMUSG00000103081  MGI:1935200  MGI:1935200                Pcdhgb8   \n",
       "4903  ENSMUSG00000039005    MGI:98223    MGI:96824                   Saa3   \n",
       "\n",
       "0    Receptor Official Symbol  same_as_off_lig  same_as_off_rec  \n",
       "770                      Ccr2             True             True  \n",
       "967                      Ccr1             True             True  \n",
       "968                      Ccr1             True             True  \n",
       "993                      Ccr1             True             True  \n",
       "1781                     Ccr6             True             True  \n",
       "1782                     Tlr4             True             True  \n",
       "2272                     Tlr4             True             True  \n",
       "4318                  Pcdhb10             True             True  \n",
       "4319                  Pcdhb11             True             True  \n",
       "4321                  Pcdhb12             True             True  \n",
       "4323                  Pcdhb14             True             True  \n",
       "4331                  Pcdhb21             True             True  \n",
       "4334                   Pcdhb4             True             True  \n",
       "4338                   Pcdhb7             True             True  \n",
       "4339                   Pcdhb8             True             True  \n",
       "4340                   Pcdhb8             True             True  \n",
       "4341                   Pcdhb9             True             True  \n",
       "4367                  Pcdhgb8             True             True  \n",
       "4903                     Tlr4             True             True  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair_mouse_card[\"same_as_off_lig\"] = (\n",
    "    gene_pair_mouse_card[\"Mouse_ligand\"] == gene_pair_mouse_card[\"Ligand Official Symbol\"]\n",
    ")\n",
    "\n",
    "gene_pair_mouse_card[\"same_as_off_rec\"] = (\n",
    "    gene_pair_mouse_card[\"Mouse_receptor\"] == gene_pair_mouse_card[\"Receptor Official Symbol\"]\n",
    ")\n",
    "\n",
    "gene_pair_mouse_card"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "f034dcc8-807d-4295-b639-1d38d56a6cbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Condition 1: Set 'Ligand Official Symbol' to 'NA' if 'lig_species' is not 'Homo sapiens'\n",
    "# We use .loc for setting values based on a condition to avoid SettingWithCopyWarning\n",
    "gene_pair.loc[gene_pair['lig_species'] != 'Homo sapiens', 'Ligand Official Symbol'] = 'NA'\n",
    "\n",
    "# Condition 2: Set 'Receptor Official Symbol' to 'NA' if 'rec_species' is not 'Homo sapiens'\n",
    "gene_pair.loc[gene_pair['rec_species'] != 'Homo sapiens', 'Receptor Official Symbol'] = 'NA'\n",
    "\n",
    "not_human = gene_pair[(gene_pair['Ligand Official Symbol'] == 'NA') | (gene_pair['Receptor Official Symbol'] == 'NA')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "eb226d47-00ed-4d62-b473-3eada5f5ee76",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Mus Musculus', 'mixed', 'Bos taurus', 'Rattus norvegicus',\n",
       "       'Gallus gallus', 'Danio rerio', 'Xenopus tropicalis',\n",
       "       'Tetraodon nigroviridis', 'Oryctolagus cuniculus'], dtype=object)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_human[\"Pair_species\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "0ee3baa4-244a-4028-9a3b-a295a483bb1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LR Pair</th>\n",
       "      <th>PMID</th>\n",
       "      <th>lig_species</th>\n",
       "      <th>rec_species</th>\n",
       "      <th>ligand_orig</th>\n",
       "      <th>receptor_orig</th>\n",
       "      <th>Pair_species</th>\n",
       "      <th>Human evidence</th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>Ligand</th>\n",
       "      <th>...</th>\n",
       "      <th>Human LR Pair</th>\n",
       "      <th>Ligand Name</th>\n",
       "      <th>Ligand Aliases</th>\n",
       "      <th>Ligand Old symbol</th>\n",
       "      <th>Receptor Name</th>\n",
       "      <th>Receptor Aliases</th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>Receptor Old symbol</th>\n",
       "      <th>Ligand Symbols</th>\n",
       "      <th>Receptor Symbols</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A Atrn</td>\n",
       "      <td>11137996</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>A</td>\n",
       "      <td>Atrn</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>A Atrn</td>\n",
       "      <td>ASIP</td>\n",
       "      <td>...</td>\n",
       "      <td>ASIP ATRN</td>\n",
       "      <td>agouti signaling protein</td>\n",
       "      <td>ASP</td>\n",
       "      <td>AGTIL</td>\n",
       "      <td>attractin</td>\n",
       "      <td>DPPT-L, MGCA</td>\n",
       "      <td>ATRN</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ASIP (AGTIL, ASP)</td>\n",
       "      <td>ATRN (DPPT-L, MGCA)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>12194978</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>A2M</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>A2M</td>\n",
       "      <td>...</td>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>FWP007, S863-7, CPAMD5</td>\n",
       "      <td>N/A</td>\n",
       "      <td>heat shock protein family A (Hsp70) member 5</td>\n",
       "      <td>BiP</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>GRP78</td>\n",
       "      <td>A2M (FWP007, S863-7, CPAMD5)</td>\n",
       "      <td>HSPA5 (GRP78, BiP)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>32541810</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>A2M</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>A2M</td>\n",
       "      <td>...</td>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>FWP007, S863-7, CPAMD5</td>\n",
       "      <td>N/A</td>\n",
       "      <td>heat shock protein family A (Hsp70) member 5</td>\n",
       "      <td>BiP</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>GRP78</td>\n",
       "      <td>A2M (FWP007, S863-7, CPAMD5)</td>\n",
       "      <td>HSPA5 (GRP78, BiP)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>1702392</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>A2M</td>\n",
       "      <td>LRP1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>A2M</td>\n",
       "      <td>...</td>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>FWP007, S863-7, CPAMD5</td>\n",
       "      <td>N/A</td>\n",
       "      <td>LDL receptor related protein 1</td>\n",
       "      <td>LRP, CD91, LRP1A, APOER, IGFBP3R1, IGFBP-3R</td>\n",
       "      <td>LRP1</td>\n",
       "      <td>APR, A2MR</td>\n",
       "      <td>A2M (FWP007, S863-7, CPAMD5)</td>\n",
       "      <td>LRP1 (APR, A2MR, LRP, CD91, LRP1A, APOER, IGFB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>10652313</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>A2M</td>\n",
       "      <td>LRP1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>A2M</td>\n",
       "      <td>...</td>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>FWP007, S863-7, CPAMD5</td>\n",
       "      <td>N/A</td>\n",
       "      <td>LDL receptor related protein 1</td>\n",
       "      <td>LRP, CD91, LRP1A, APOER, IGFBP3R1, IGFBP-3R</td>\n",
       "      <td>LRP1</td>\n",
       "      <td>APR, A2MR</td>\n",
       "      <td>A2M (FWP007, S863-7, CPAMD5)</td>\n",
       "      <td>LRP1 (APR, A2MR, LRP, CD91, LRP1A, APOER, IGFB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6089</th>\n",
       "      <td>ZG16B TLR5</td>\n",
       "      <td>20802527</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ZG16B</td>\n",
       "      <td>TLR5</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>ZG16B TLR5</td>\n",
       "      <td>ZG16B</td>\n",
       "      <td>...</td>\n",
       "      <td>ZG16B TLR5</td>\n",
       "      <td>zymogen granule protein 16B</td>\n",
       "      <td>HRPE773, PRO1567, JCLN2</td>\n",
       "      <td>N/A</td>\n",
       "      <td>toll like receptor 5</td>\n",
       "      <td>TIL3, FLJ10052, MGC126430, MGC126431</td>\n",
       "      <td>TLR5</td>\n",
       "      <td>SLEB1</td>\n",
       "      <td>ZG16B (HRPE773, PRO1567, JCLN2)</td>\n",
       "      <td>TLR5 (SLEB1, TIL3, FLJ10052, MGC126430, MGC126...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6090</th>\n",
       "      <td>ZG16B TLR6</td>\n",
       "      <td>20802527</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ZG16B</td>\n",
       "      <td>TLR6</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>ZG16B TLR6</td>\n",
       "      <td>ZG16B</td>\n",
       "      <td>...</td>\n",
       "      <td>ZG16B TLR6</td>\n",
       "      <td>zymogen granule protein 16B</td>\n",
       "      <td>HRPE773, PRO1567, JCLN2</td>\n",
       "      <td>N/A</td>\n",
       "      <td>toll like receptor 6</td>\n",
       "      <td>CD286</td>\n",
       "      <td>TLR6</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ZG16B (HRPE773, PRO1567, JCLN2)</td>\n",
       "      <td>TLR6 (CD286)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6091</th>\n",
       "      <td>ZG16B Tlr6</td>\n",
       "      <td>20802527</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>ZG16B</td>\n",
       "      <td>Tlr6</td>\n",
       "      <td>mixed</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>ZG16B Tlr6</td>\n",
       "      <td>ZG16B</td>\n",
       "      <td>...</td>\n",
       "      <td>ZG16B TLR6</td>\n",
       "      <td>zymogen granule protein 16B</td>\n",
       "      <td>HRPE773, PRO1567, JCLN2</td>\n",
       "      <td>N/A</td>\n",
       "      <td>toll like receptor 6</td>\n",
       "      <td>CD286</td>\n",
       "      <td>TLR6</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ZG16B (HRPE773, PRO1567, JCLN2)</td>\n",
       "      <td>TLR6 (CD286)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6092</th>\n",
       "      <td>Zp3 Chrna7</td>\n",
       "      <td>22577141</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Zp3</td>\n",
       "      <td>Chrna7</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>Zp3 Chrna7</td>\n",
       "      <td>ZP3</td>\n",
       "      <td>...</td>\n",
       "      <td>ZP3 CHRNA7</td>\n",
       "      <td>zona pellucida glycoprotein 3</td>\n",
       "      <td>ZP3-424, ZP3-372, ZPC</td>\n",
       "      <td>ZP3A, ZP3B</td>\n",
       "      <td>cholinergic receptor nicotinic alpha 7 subunit</td>\n",
       "      <td>N/A</td>\n",
       "      <td>CHRNA7</td>\n",
       "      <td>N/A</td>\n",
       "      <td>ZP3 (ZP3A, ZP3B, ZP3-424, ZP3-372, ZPC)</td>\n",
       "      <td>CHRNA7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6093</th>\n",
       "      <td>Zp3 Zp3R</td>\n",
       "      <td>8188752</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Zp3</td>\n",
       "      <td>Zp3R</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>Zp3 Zp3R</td>\n",
       "      <td>ZP3</td>\n",
       "      <td>...</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>zona pellucida glycoprotein 3</td>\n",
       "      <td>ZP3-424, ZP3-372, ZPC</td>\n",
       "      <td>ZP3A, ZP3B</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>ZP3 (ZP3A, ZP3B, ZP3-424, ZP3-372, ZPC)</td>\n",
       "      <td>no human ortholog</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6094 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         LR Pair      PMID   lig_species   rec_species ligand_orig  \\\n",
       "0         A Atrn  11137996  Mus Musculus  Mus Musculus           A   \n",
       "1      A2M HSPA5  12194978  Homo sapiens  Homo sapiens         A2M   \n",
       "2      A2M HSPA5  32541810  Homo sapiens  Homo sapiens         A2M   \n",
       "3       A2M LRP1   1702392  Homo sapiens  Homo sapiens         A2M   \n",
       "4       A2M LRP1  10652313  Homo sapiens  Homo sapiens         A2M   \n",
       "...          ...       ...           ...           ...         ...   \n",
       "6089  ZG16B TLR5  20802527  Homo sapiens  Homo sapiens       ZG16B   \n",
       "6090  ZG16B TLR6  20802527  Homo sapiens  Homo sapiens       ZG16B   \n",
       "6091  ZG16B Tlr6  20802527  Homo sapiens  Mus Musculus       ZG16B   \n",
       "6092  Zp3 Chrna7  22577141  Mus Musculus  Mus Musculus         Zp3   \n",
       "6093    Zp3 Zp3R   8188752  Mus Musculus  Mus Musculus         Zp3   \n",
       "\n",
       "     receptor_orig  Pair_species   Human evidence LR Pair Card Ligand  ...  \\\n",
       "0             Atrn  Mus Musculus     CONSERVATION       A Atrn   ASIP  ...   \n",
       "1            HSPA5  Homo sapiens           DIRECT    A2M HSPA5    A2M  ...   \n",
       "2            HSPA5  Homo sapiens           DIRECT    A2M HSPA5    A2M  ...   \n",
       "3             LRP1  Homo sapiens           DIRECT     A2M LRP1    A2M  ...   \n",
       "4             LRP1  Homo sapiens           DIRECT     A2M LRP1    A2M  ...   \n",
       "...            ...           ...              ...          ...    ...  ...   \n",
       "6089          TLR5  Homo sapiens           DIRECT   ZG16B TLR5  ZG16B  ...   \n",
       "6090          TLR6  Homo sapiens           DIRECT   ZG16B TLR6  ZG16B  ...   \n",
       "6091          Tlr6         mixed     CONSERVATION   ZG16B Tlr6  ZG16B  ...   \n",
       "6092        Chrna7  Mus Musculus     CONSERVATION   Zp3 Chrna7    ZP3  ...   \n",
       "6093          Zp3R  Mus Musculus  absent in human     Zp3 Zp3R    ZP3  ...   \n",
       "\n",
       "          Human LR Pair                    Ligand Name  \\\n",
       "0             ASIP ATRN       agouti signaling protein   \n",
       "1             A2M HSPA5          alpha-2-macroglobulin   \n",
       "2             A2M HSPA5          alpha-2-macroglobulin   \n",
       "3              A2M LRP1          alpha-2-macroglobulin   \n",
       "4              A2M LRP1          alpha-2-macroglobulin   \n",
       "...                 ...                            ...   \n",
       "6089         ZG16B TLR5    zymogen granule protein 16B   \n",
       "6090         ZG16B TLR6    zymogen granule protein 16B   \n",
       "6091         ZG16B TLR6    zymogen granule protein 16B   \n",
       "6092         ZP3 CHRNA7  zona pellucida glycoprotein 3   \n",
       "6093  no human ortholog  zona pellucida glycoprotein 3   \n",
       "\n",
       "               Ligand Aliases Ligand Old symbol  \\\n",
       "0                         ASP             AGTIL   \n",
       "1      FWP007, S863-7, CPAMD5               N/A   \n",
       "2      FWP007, S863-7, CPAMD5               N/A   \n",
       "3      FWP007, S863-7, CPAMD5               N/A   \n",
       "4      FWP007, S863-7, CPAMD5               N/A   \n",
       "...                       ...               ...   \n",
       "6089  HRPE773, PRO1567, JCLN2               N/A   \n",
       "6090  HRPE773, PRO1567, JCLN2               N/A   \n",
       "6091  HRPE773, PRO1567, JCLN2               N/A   \n",
       "6092    ZP3-424, ZP3-372, ZPC        ZP3A, ZP3B   \n",
       "6093    ZP3-424, ZP3-372, ZPC        ZP3A, ZP3B   \n",
       "\n",
       "                                       Receptor Name  \\\n",
       "0                                          attractin   \n",
       "1       heat shock protein family A (Hsp70) member 5   \n",
       "2       heat shock protein family A (Hsp70) member 5   \n",
       "3                     LDL receptor related protein 1   \n",
       "4                     LDL receptor related protein 1   \n",
       "...                                              ...   \n",
       "6089                            toll like receptor 5   \n",
       "6090                            toll like receptor 6   \n",
       "6091                            toll like receptor 6   \n",
       "6092  cholinergic receptor nicotinic alpha 7 subunit   \n",
       "6093                                             NaN   \n",
       "\n",
       "                                 Receptor Aliases Approved symbol  \\\n",
       "0                                    DPPT-L, MGCA            ATRN   \n",
       "1                                             BiP           HSPA5   \n",
       "2                                             BiP           HSPA5   \n",
       "3     LRP, CD91, LRP1A, APOER, IGFBP3R1, IGFBP-3R            LRP1   \n",
       "4     LRP, CD91, LRP1A, APOER, IGFBP3R1, IGFBP-3R            LRP1   \n",
       "...                                           ...             ...   \n",
       "6089         TIL3, FLJ10052, MGC126430, MGC126431            TLR5   \n",
       "6090                                        CD286            TLR6   \n",
       "6091                                        CD286            TLR6   \n",
       "6092                                          N/A          CHRNA7   \n",
       "6093                                                          NaN   \n",
       "\n",
       "     Receptor Old symbol                           Ligand Symbols  \\\n",
       "0                    N/A                        ASIP (AGTIL, ASP)   \n",
       "1                  GRP78             A2M (FWP007, S863-7, CPAMD5)   \n",
       "2                  GRP78             A2M (FWP007, S863-7, CPAMD5)   \n",
       "3              APR, A2MR             A2M (FWP007, S863-7, CPAMD5)   \n",
       "4              APR, A2MR             A2M (FWP007, S863-7, CPAMD5)   \n",
       "...                  ...                                      ...   \n",
       "6089               SLEB1          ZG16B (HRPE773, PRO1567, JCLN2)   \n",
       "6090                 N/A          ZG16B (HRPE773, PRO1567, JCLN2)   \n",
       "6091                 N/A          ZG16B (HRPE773, PRO1567, JCLN2)   \n",
       "6092                 N/A  ZP3 (ZP3A, ZP3B, ZP3-424, ZP3-372, ZPC)   \n",
       "6093                      ZP3 (ZP3A, ZP3B, ZP3-424, ZP3-372, ZPC)   \n",
       "\n",
       "                                       Receptor Symbols  \n",
       "0                                   ATRN (DPPT-L, MGCA)  \n",
       "1                                    HSPA5 (GRP78, BiP)  \n",
       "2                                    HSPA5 (GRP78, BiP)  \n",
       "3     LRP1 (APR, A2MR, LRP, CD91, LRP1A, APOER, IGFB...  \n",
       "4     LRP1 (APR, A2MR, LRP, CD91, LRP1A, APOER, IGFB...  \n",
       "...                                                 ...  \n",
       "6089  TLR5 (SLEB1, TIL3, FLJ10052, MGC126430, MGC126...  \n",
       "6090                                       TLR6 (CD286)  \n",
       "6091                                       TLR6 (CD286)  \n",
       "6092                                             CHRNA7  \n",
       "6093                                  no human ortholog  \n",
       "\n",
       "[6094 rows x 27 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2620c1fe-9199-4aec-ab46-6a0cf8e1690f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Interaction ID</th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>Human LR Pair</th>\n",
       "      <th>Ligand</th>\n",
       "      <th>Receptor</th>\n",
       "      <th>Ligand Symbols</th>\n",
       "      <th>Receptor Symbols</th>\n",
       "      <th>Ligand Location</th>\n",
       "      <th>Receptor Location</th>\n",
       "      <th>Ligand HGNC ID</th>\n",
       "      <th>Receptor HGNC ID</th>\n",
       "      <th>Perplexity</th>\n",
       "      <th>Human evidence</th>\n",
       "      <th>PMID</th>\n",
       "      <th>Ligand Name</th>\n",
       "      <th>Receptor Name</th>\n",
       "      <th>Mouse_ligand</th>\n",
       "      <th>Mouse_receptor</th>\n",
       "      <th>Ligand MGI ID</th>\n",
       "      <th>Receptor MGI ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4565</th>\n",
       "      <td>CDB04566</td>\n",
       "      <td>CD99 Pilrb1</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>CD99</td>\n",
       "      <td>Pilrb1</td>\n",
       "      <td>&lt;span title=\"CD99 (MIC2, MIC2X, MIC2Y)\"&gt;CD99 (...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on perplexity\"&gt;cell membran...</td>\n",
       "      <td>&lt;span title=\"based on perplexity\"&gt;cell membran...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>CD99 molecule (Xg blood group)</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4566</th>\n",
       "      <td>CDB04567</td>\n",
       "      <td>Ccl12 Ccr2</td>\n",
       "      <td>not conserved CCR2</td>\n",
       "      <td>Ccl12</td>\n",
       "      <td>Ccr2</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;secreted&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>C-C motif chemokine ligand 12</td>\n",
       "      <td>C-C motif chemokine receptor 2</td>\n",
       "      <td>Ccl12</td>\n",
       "      <td>Ccr2</td>\n",
       "      <td>MGI:108224</td>\n",
       "      <td>MGI:106185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4567</th>\n",
       "      <td>CDB04568</td>\n",
       "      <td>Ccl21B CXCR3</td>\n",
       "      <td>not conserved CXCR3</td>\n",
       "      <td>Ccl21B</td>\n",
       "      <td>CXCR3</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"CXCR3 (GPR9, CKR-L2, CMKAR3, IP10...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>nan</td>\n",
       "      <td>C-X-C motif chemokine receptor 3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4568</th>\n",
       "      <td>CDB04569</td>\n",
       "      <td>Ccl21B Ccr7</td>\n",
       "      <td>not conserved CCR7</td>\n",
       "      <td>Ccl21B</td>\n",
       "      <td>Ccr7</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>nan</td>\n",
       "      <td>C-C motif chemokine receptor 7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4569</th>\n",
       "      <td>CDB04570</td>\n",
       "      <td>Ccl21B Cxcr3</td>\n",
       "      <td>not conserved CXCR3</td>\n",
       "      <td>Ccl21B</td>\n",
       "      <td>Cxcr3</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>nan</td>\n",
       "      <td>C-X-C motif chemokine receptor 3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4570</th>\n",
       "      <td>CDB04571</td>\n",
       "      <td>Ccl21C CXCR3</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Ccl21C</td>\n",
       "      <td>CXCR3</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"CXCR3 (GPR9, CKR-L2, CMKAR3, IP10...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>nan</td>\n",
       "      <td>C-X-C motif chemokine receptor 3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4571</th>\n",
       "      <td>CDB04572</td>\n",
       "      <td>Ccl21C Cxcr3</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Ccl21C</td>\n",
       "      <td>Cxcr3</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>nan</td>\n",
       "      <td>C-X-C motif chemokine receptor 3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4572</th>\n",
       "      <td>CDB04573</td>\n",
       "      <td>Ccl6 Ccr1</td>\n",
       "      <td>not conserved CCR1</td>\n",
       "      <td>Ccl6</td>\n",
       "      <td>Ccr1</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;secreted&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>C-C motif chemokine ligand 6</td>\n",
       "      <td>C-C motif chemokine receptor 1</td>\n",
       "      <td>Ccl6</td>\n",
       "      <td>Ccr1</td>\n",
       "      <td>MGI:98263</td>\n",
       "      <td>MGI:104618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4573</th>\n",
       "      <td>CDB04573</td>\n",
       "      <td>Ccl6 Ccr1</td>\n",
       "      <td>not conserved CCR1</td>\n",
       "      <td>Ccl6</td>\n",
       "      <td>Ccr1</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;secreted&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>C-C motif chemokine ligand 6</td>\n",
       "      <td>C-C motif chemokine receptor 1</td>\n",
       "      <td>Ccl6</td>\n",
       "      <td>Ccr1</td>\n",
       "      <td>MGI:98263</td>\n",
       "      <td>MGI:104618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4574</th>\n",
       "      <td>CDB04574</td>\n",
       "      <td>Ccl9 Ccr1</td>\n",
       "      <td>not conserved CCR1</td>\n",
       "      <td>Ccl9</td>\n",
       "      <td>Ccr1</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;secreted&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>C-C motif chemokine ligand 9</td>\n",
       "      <td>C-C motif chemokine receptor 1</td>\n",
       "      <td>Ccl9</td>\n",
       "      <td>Ccr1</td>\n",
       "      <td>MGI:104533</td>\n",
       "      <td>MGI:104618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4575</th>\n",
       "      <td>CDB04575</td>\n",
       "      <td>Cd200 Cd200R3</td>\n",
       "      <td>CD200 not conserved</td>\n",
       "      <td>Cd200</td>\n",
       "      <td>Cd200R3</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;span title=\"based on perplexity\"&gt;cell membran...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>CD200 molecule</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4576</th>\n",
       "      <td>CDB04576</td>\n",
       "      <td>Cd200 Cd200R4</td>\n",
       "      <td>CD200 not conserved</td>\n",
       "      <td>Cd200</td>\n",
       "      <td>Cd200R4</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;span title=\"based on perplexity\"&gt;cell membran...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>CD200 molecule</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4577</th>\n",
       "      <td>CDB04577</td>\n",
       "      <td>Defb2 Ccr6</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Defb2</td>\n",
       "      <td>Ccr6</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;secreted&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>defensin beta 2</td>\n",
       "      <td>C-C motif chemokine receptor 6</td>\n",
       "      <td>Defb2</td>\n",
       "      <td>Ccr6</td>\n",
       "      <td>MGI:1338754</td>\n",
       "      <td>MGI:1333797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4578</th>\n",
       "      <td>CDB04578</td>\n",
       "      <td>Defb2 Tlr4</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Defb2</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;secreted&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>defensin beta 2</td>\n",
       "      <td>toll-like receptor 4</td>\n",
       "      <td>Defb2</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>MGI:1338754</td>\n",
       "      <td>MGI:96824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4579</th>\n",
       "      <td>CDB04579</td>\n",
       "      <td>Fcna Tlr4</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Fcna</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>ficolin A</td>\n",
       "      <td>toll-like receptor 4</td>\n",
       "      <td>Fcna</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>MGI:1340905</td>\n",
       "      <td>MGI:96824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4580</th>\n",
       "      <td>CDB04580</td>\n",
       "      <td>H60A Klrk1</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>H60A</td>\n",
       "      <td>Klrk1</td>\n",
       "      <td>&lt;span title=\"H60A\"&gt;H60A&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>nan</td>\n",
       "      <td>killer cell lectin like receptor K1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4581</th>\n",
       "      <td>CDB04581</td>\n",
       "      <td>H60B Klrk1</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>H60B</td>\n",
       "      <td>Klrk1</td>\n",
       "      <td>&lt;span title=\"H60B\"&gt;H60B&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>nan</td>\n",
       "      <td>killer cell lectin like receptor K1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4582</th>\n",
       "      <td>CDB04582</td>\n",
       "      <td>H60C Klrk1</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>H60C</td>\n",
       "      <td>Klrk1</td>\n",
       "      <td>&lt;span title=\"H60C\"&gt;H60C&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>nan</td>\n",
       "      <td>killer cell lectin like receptor K1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4583</th>\n",
       "      <td>CDB04583</td>\n",
       "      <td>Lair1 Lilrb4A</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Lair1</td>\n",
       "      <td>Lilrb4A</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>leukocyte associated immunoglobulin like recep...</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4584</th>\n",
       "      <td>CDB04584</td>\n",
       "      <td>Pcdhb10 Pcdhb10</td>\n",
       "      <td>not conserved not conserved</td>\n",
       "      <td>Pcdhb10</td>\n",
       "      <td>Pcdhb10</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>protocadherin beta 10</td>\n",
       "      <td>protocadherin beta 10</td>\n",
       "      <td>Pcdhb10</td>\n",
       "      <td>Pcdhb10</td>\n",
       "      <td>MGI:2136745</td>\n",
       "      <td>MGI:2136745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4585</th>\n",
       "      <td>CDB04585</td>\n",
       "      <td>Pcdhb11 Pcdhb11</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Pcdhb11</td>\n",
       "      <td>Pcdhb11</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>protocadherin beta 11</td>\n",
       "      <td>protocadherin beta 11</td>\n",
       "      <td>Pcdhb11</td>\n",
       "      <td>Pcdhb11</td>\n",
       "      <td>MGI:2136746</td>\n",
       "      <td>MGI:2136746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4586</th>\n",
       "      <td>CDB04586</td>\n",
       "      <td>Pcdhb12 Pcdhb12</td>\n",
       "      <td>not conserved not conserved</td>\n",
       "      <td>Pcdhb12</td>\n",
       "      <td>Pcdhb12</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>protocadherin beta 12</td>\n",
       "      <td>protocadherin beta 12</td>\n",
       "      <td>Pcdhb12</td>\n",
       "      <td>Pcdhb12</td>\n",
       "      <td>MGI:2136747</td>\n",
       "      <td>MGI:2136747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4587</th>\n",
       "      <td>CDB04587</td>\n",
       "      <td>Pcdhb14 Pcdhb14</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Pcdhb14</td>\n",
       "      <td>Pcdhb14</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>protocadherin beta 14</td>\n",
       "      <td>protocadherin beta 14</td>\n",
       "      <td>Pcdhb14</td>\n",
       "      <td>Pcdhb14</td>\n",
       "      <td>MGI:2136749</td>\n",
       "      <td>MGI:2136749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4588</th>\n",
       "      <td>CDB04588</td>\n",
       "      <td>Pcdhb21 Pcdhb21</td>\n",
       "      <td>not conserved not conserved</td>\n",
       "      <td>Pcdhb21</td>\n",
       "      <td>Pcdhb21</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>protocadherin beta 21</td>\n",
       "      <td>protocadherin beta 21</td>\n",
       "      <td>Pcdhb21</td>\n",
       "      <td>Pcdhb21</td>\n",
       "      <td>MGI:2136759</td>\n",
       "      <td>MGI:2136759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4589</th>\n",
       "      <td>CDB04589</td>\n",
       "      <td>Pcdhb4 Pcdhb4</td>\n",
       "      <td>not conserved not conserved</td>\n",
       "      <td>Pcdhb4</td>\n",
       "      <td>Pcdhb4</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>protocadherin beta 4</td>\n",
       "      <td>protocadherin beta 4</td>\n",
       "      <td>Pcdhb4</td>\n",
       "      <td>Pcdhb4</td>\n",
       "      <td>MGI:2136738</td>\n",
       "      <td>MGI:2136738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4590</th>\n",
       "      <td>CDB04590</td>\n",
       "      <td>Pcdhb7 Pcdhb7</td>\n",
       "      <td>not conserved not conserved</td>\n",
       "      <td>Pcdhb7</td>\n",
       "      <td>Pcdhb7</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>protocadherin beta 7</td>\n",
       "      <td>protocadherin beta 7</td>\n",
       "      <td>Pcdhb7</td>\n",
       "      <td>Pcdhb7</td>\n",
       "      <td>MGI:2136741</td>\n",
       "      <td>MGI:2136741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4591</th>\n",
       "      <td>CDB04591</td>\n",
       "      <td>Pcdhb8 Pcdhb8</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>protocadherin beta 8</td>\n",
       "      <td>protocadherin beta 8</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>MGI:2136742</td>\n",
       "      <td>MGI:2136742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4592</th>\n",
       "      <td>CDB04591</td>\n",
       "      <td>Pcdhb8 Pcdhb8</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>protocadherin beta 8</td>\n",
       "      <td>protocadherin beta 8</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>MGI:2136742</td>\n",
       "      <td>MGI:2136742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4593</th>\n",
       "      <td>CDB04592</td>\n",
       "      <td>Pcdhb9 Pcdhb9</td>\n",
       "      <td>not conserved not conserved</td>\n",
       "      <td>Pcdhb9</td>\n",
       "      <td>Pcdhb9</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>protocadherin beta 9</td>\n",
       "      <td>protocadherin beta 9</td>\n",
       "      <td>Pcdhb9</td>\n",
       "      <td>Pcdhb9</td>\n",
       "      <td>MGI:2136744</td>\n",
       "      <td>MGI:2136744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4594</th>\n",
       "      <td>CDB04593</td>\n",
       "      <td>Pcdhgb8 Pcdhgb8</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>protocadherin gamma subfamily B, 8</td>\n",
       "      <td>protocadherin gamma subfamily B, 8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>MGI:1935200</td>\n",
       "      <td>MGI:1935200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4595</th>\n",
       "      <td>CDB04594</td>\n",
       "      <td>Saa3 Tlr4</td>\n",
       "      <td>not conserved TLR4</td>\n",
       "      <td>Saa3</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>serum amyloid A 3</td>\n",
       "      <td>toll-like receptor 4</td>\n",
       "      <td>Saa3</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>MGI:98223</td>\n",
       "      <td>MGI:96824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4596</th>\n",
       "      <td>CDB04595</td>\n",
       "      <td>Sema4A Timd2</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Sema4A</td>\n",
       "      <td>Timd2</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>semaphorin 4A</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4597</th>\n",
       "      <td>CDB04596</td>\n",
       "      <td>Zp3 Zp3R</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Zp3</td>\n",
       "      <td>Zp3R</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>zona pellucida glycoprotein 3</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Interaction ID     LR Pair Card                Human LR Pair   Ligand  \\\n",
       "4565       CDB04566      CD99 Pilrb1            no human ortholog     CD99   \n",
       "4566       CDB04567       Ccl12 Ccr2           not conserved CCR2    Ccl12   \n",
       "4567       CDB04568     Ccl21B CXCR3          not conserved CXCR3   Ccl21B   \n",
       "4568       CDB04569      Ccl21B Ccr7           not conserved CCR7   Ccl21B   \n",
       "4569       CDB04570     Ccl21B Cxcr3          not conserved CXCR3   Ccl21B   \n",
       "4570       CDB04571     Ccl21C CXCR3            no human ortholog   Ccl21C   \n",
       "4571       CDB04572     Ccl21C Cxcr3            no human ortholog   Ccl21C   \n",
       "4572       CDB04573        Ccl6 Ccr1           not conserved CCR1     Ccl6   \n",
       "4573       CDB04573        Ccl6 Ccr1           not conserved CCR1     Ccl6   \n",
       "4574       CDB04574        Ccl9 Ccr1           not conserved CCR1     Ccl9   \n",
       "4575       CDB04575    Cd200 Cd200R3          CD200 not conserved    Cd200   \n",
       "4576       CDB04576    Cd200 Cd200R4          CD200 not conserved    Cd200   \n",
       "4577       CDB04577       Defb2 Ccr6            no human ortholog    Defb2   \n",
       "4578       CDB04578       Defb2 Tlr4            no human ortholog    Defb2   \n",
       "4579       CDB04579        Fcna Tlr4            no human ortholog     Fcna   \n",
       "4580       CDB04580       H60A Klrk1            no human ortholog     H60A   \n",
       "4581       CDB04581       H60B Klrk1            no human ortholog     H60B   \n",
       "4582       CDB04582       H60C Klrk1            no human ortholog     H60C   \n",
       "4583       CDB04583    Lair1 Lilrb4A            no human ortholog    Lair1   \n",
       "4584       CDB04584  Pcdhb10 Pcdhb10  not conserved not conserved  Pcdhb10   \n",
       "4585       CDB04585  Pcdhb11 Pcdhb11            no human ortholog  Pcdhb11   \n",
       "4586       CDB04586  Pcdhb12 Pcdhb12  not conserved not conserved  Pcdhb12   \n",
       "4587       CDB04587  Pcdhb14 Pcdhb14            no human ortholog  Pcdhb14   \n",
       "4588       CDB04588  Pcdhb21 Pcdhb21  not conserved not conserved  Pcdhb21   \n",
       "4589       CDB04589    Pcdhb4 Pcdhb4  not conserved not conserved   Pcdhb4   \n",
       "4590       CDB04590    Pcdhb7 Pcdhb7  not conserved not conserved   Pcdhb7   \n",
       "4591       CDB04591    Pcdhb8 Pcdhb8            no human ortholog   Pcdhb8   \n",
       "4592       CDB04591    Pcdhb8 Pcdhb8            no human ortholog   Pcdhb8   \n",
       "4593       CDB04592    Pcdhb9 Pcdhb9  not conserved not conserved   Pcdhb9   \n",
       "4594       CDB04593  Pcdhgb8 Pcdhgb8            no human ortholog  Pcdhgb8   \n",
       "4595       CDB04594        Saa3 Tlr4           not conserved TLR4     Saa3   \n",
       "4596       CDB04595     Sema4A Timd2            no human ortholog   Sema4A   \n",
       "4597       CDB04596         Zp3 Zp3R            no human ortholog      Zp3   \n",
       "\n",
       "     Receptor                                     Ligand Symbols  \\\n",
       "4565   Pilrb1  <span title=\"CD99 (MIC2, MIC2X, MIC2Y)\">CD99 (...   \n",
       "4566     Ccr2  <span title=\"no human ortholog\">no human ortho...   \n",
       "4567    CXCR3  <span title=\"no human ortholog\">no human ortho...   \n",
       "4568     Ccr7  <span title=\"no human ortholog\">no human ortho...   \n",
       "4569    Cxcr3  <span title=\"no human ortholog\">no human ortho...   \n",
       "4570    CXCR3  <span title=\"no human ortholog\">no human ortho...   \n",
       "4571    Cxcr3  <span title=\"no human ortholog\">no human ortho...   \n",
       "4572     Ccr1  <span title=\"no human ortholog\">no human ortho...   \n",
       "4573     Ccr1  <span title=\"no human ortholog\">no human ortho...   \n",
       "4574     Ccr1  <span title=\"no human ortholog\">no human ortho...   \n",
       "4575  Cd200R3  <span title=\"no human ortholog\">no human ortho...   \n",
       "4576  Cd200R4  <span title=\"no human ortholog\">no human ortho...   \n",
       "4577     Ccr6  <span title=\"no human ortholog\">no human ortho...   \n",
       "4578     Tlr4  <span title=\"no human ortholog\">no human ortho...   \n",
       "4579     Tlr4  <span title=\"no human ortholog\">no human ortho...   \n",
       "4580    Klrk1                     <span title=\"H60A\">H60A</span>   \n",
       "4581    Klrk1                     <span title=\"H60B\">H60B</span>   \n",
       "4582    Klrk1                     <span title=\"H60C\">H60C</span>   \n",
       "4583  Lilrb4A  <span title=\"no human ortholog\">no human ortho...   \n",
       "4584  Pcdhb10  <span title=\"no human ortholog\">no human ortho...   \n",
       "4585  Pcdhb11  <span title=\"no human ortholog\">no human ortho...   \n",
       "4586  Pcdhb12  <span title=\"no human ortholog\">no human ortho...   \n",
       "4587  Pcdhb14  <span title=\"no human ortholog\">no human ortho...   \n",
       "4588  Pcdhb21  <span title=\"no human ortholog\">no human ortho...   \n",
       "4589   Pcdhb4  <span title=\"no human ortholog\">no human ortho...   \n",
       "4590   Pcdhb7  <span title=\"no human ortholog\">no human ortho...   \n",
       "4591   Pcdhb8  <span title=\"no human ortholog\">no human ortho...   \n",
       "4592   Pcdhb8  <span title=\"no human ortholog\">no human ortho...   \n",
       "4593   Pcdhb9  <span title=\"no human ortholog\">no human ortho...   \n",
       "4594  Pcdhgb8  <span title=\"no human ortholog\">no human ortho...   \n",
       "4595     Tlr4  <span title=\"no human ortholog\">no human ortho...   \n",
       "4596    Timd2  <span title=\"no human ortholog\">no human ortho...   \n",
       "4597     Zp3R  <span title=\"no human ortholog\">no human ortho...   \n",
       "\n",
       "                                       Receptor Symbols  \\\n",
       "4565  <span title=\"no human ortholog\">no human ortho...   \n",
       "4566  <span title=\"no human ortholog\">no human ortho...   \n",
       "4567  <span title=\"CXCR3 (GPR9, CKR-L2, CMKAR3, IP10...   \n",
       "4568  <span title=\"no human ortholog\">no human ortho...   \n",
       "4569  <span title=\"no human ortholog\">no human ortho...   \n",
       "4570  <span title=\"CXCR3 (GPR9, CKR-L2, CMKAR3, IP10...   \n",
       "4571  <span title=\"no human ortholog\">no human ortho...   \n",
       "4572  <span title=\"no human ortholog\">no human ortho...   \n",
       "4573  <span title=\"no human ortholog\">no human ortho...   \n",
       "4574  <span title=\"no human ortholog\">no human ortho...   \n",
       "4575  <span title=\"no human ortholog\">no human ortho...   \n",
       "4576  <span title=\"no human ortholog\">no human ortho...   \n",
       "4577  <span title=\"no human ortholog\">no human ortho...   \n",
       "4578  <span title=\"no human ortholog\">no human ortho...   \n",
       "4579  <span title=\"no human ortholog\">no human ortho...   \n",
       "4580  <span title=\"no human ortholog\">no human ortho...   \n",
       "4581  <span title=\"no human ortholog\">no human ortho...   \n",
       "4582  <span title=\"no human ortholog\">no human ortho...   \n",
       "4583  <span title=\"no human ortholog\">no human ortho...   \n",
       "4584  <span title=\"no human ortholog\">no human ortho...   \n",
       "4585  <span title=\"no human ortholog\">no human ortho...   \n",
       "4586  <span title=\"no human ortholog\">no human ortho...   \n",
       "4587  <span title=\"no human ortholog\">no human ortho...   \n",
       "4588  <span title=\"no human ortholog\">no human ortho...   \n",
       "4589  <span title=\"no human ortholog\">no human ortho...   \n",
       "4590  <span title=\"no human ortholog\">no human ortho...   \n",
       "4591  <span title=\"no human ortholog\">no human ortho...   \n",
       "4592  <span title=\"no human ortholog\">no human ortho...   \n",
       "4593  <span title=\"no human ortholog\">no human ortho...   \n",
       "4594  <span title=\"no human ortholog\">no human ortho...   \n",
       "4595  <span title=\"no human ortholog\">no human ortho...   \n",
       "4596  <span title=\"no human ortholog\">no human ortho...   \n",
       "4597  <span title=\"no human ortholog\">no human ortho...   \n",
       "\n",
       "                                        Ligand Location  \\\n",
       "4565  <span title=\"based on perplexity\">cell membran...   \n",
       "4566     <span title=\"based on uniprot\">secreted</span>   \n",
       "4567                                            unknown   \n",
       "4568                                            unknown   \n",
       "4569                                            unknown   \n",
       "4570                                            unknown   \n",
       "4571                                            unknown   \n",
       "4572     <span title=\"based on uniprot\">secreted</span>   \n",
       "4573     <span title=\"based on uniprot\">secreted</span>   \n",
       "4574     <span title=\"based on uniprot\">secreted</span>   \n",
       "4575  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4576  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4577     <span title=\"based on uniprot\">secreted</span>   \n",
       "4578     <span title=\"based on uniprot\">secreted</span>   \n",
       "4579  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4580                                            unknown   \n",
       "4581                                            unknown   \n",
       "4582                                            unknown   \n",
       "4583  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4584  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4585  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4586  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4587  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4588  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "4589  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4590  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4591  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4592  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4593  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4594  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "4595  <span title=\"based on perplexity, uniprot\">sec...   \n",
       "4596                                            unknown   \n",
       "4597  <span title=\"based on uniprot\">cell membrane</...   \n",
       "\n",
       "                                      Receptor Location  \\\n",
       "4565  <span title=\"based on perplexity\">cell membran...   \n",
       "4566  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4567  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "4568  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4569  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4570  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "4571  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4572  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4573  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4574  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4575  <span title=\"based on perplexity\">cell membran...   \n",
       "4576  <span title=\"based on perplexity\">cell membran...   \n",
       "4577  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4578  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "4579  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "4580  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "4581  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "4582  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "4583                                            unknown   \n",
       "4584  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4585  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4586  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4587  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4588  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "4589  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4590  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4591  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4592  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4593  <span title=\"based on uniprot\">cell membrane</...   \n",
       "4594  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "4595  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "4596  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "4597                                            unknown   \n",
       "\n",
       "                                         Ligand HGNC ID  \\\n",
       "4565  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4566  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4567  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4568  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4569  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4570  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4571  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4572  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4573  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4574  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4575  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4576  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4577  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4578  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4579  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4580  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4581  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4582  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4583  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4584  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4585  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4586  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4587  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4588  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4589  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4590  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4591  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4592  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4593  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4594  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4595  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4596  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4597  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "\n",
       "                                       Receptor HGNC ID  \\\n",
       "4565  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4566  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4567  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4568  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4569  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4570  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4571  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4572  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4573  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4574  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4575  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4576  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4577  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4578  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4579  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4580  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4581  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4582  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4583  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4584  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4585  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4586  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4587  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4588  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4589  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4590  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4591  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4592  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4593  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4594  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4595  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4596  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4597  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "\n",
       "                                             Perplexity   Human evidence  \\\n",
       "4565  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "4566  <a href=\"https://www.perplexity.ai/search?q=Wh...    not conserved   \n",
       "4567  <a href=\"https://www.perplexity.ai/search?q=Wh...    not conserved   \n",
       "4568  <a href=\"https://www.perplexity.ai/search?q=Wh...    not conserved   \n",
       "4569  <a href=\"https://www.perplexity.ai/search?q=Wh...    not conserved   \n",
       "4570  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "4571  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "4572  <a href=\"https://www.perplexity.ai/search?q=Wh...    not conserved   \n",
       "4573  <a href=\"https://www.perplexity.ai/search?q=Wh...    not conserved   \n",
       "4574  <a href=\"https://www.perplexity.ai/search?q=Wh...    not conserved   \n",
       "4575  <a href=\"https://www.perplexity.ai/search?q=Wh...    not conserved   \n",
       "4576  <a href=\"https://www.perplexity.ai/search?q=Wh...    not conserved   \n",
       "4577  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "4578  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "4579  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "4580  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "4581  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "4582  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "4583  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "4584  <a href=\"https://www.perplexity.ai/search?q=Wh...    not conserved   \n",
       "4585  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "4586  <a href=\"https://www.perplexity.ai/search?q=Wh...    not conserved   \n",
       "4587  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "4588  <a href=\"https://www.perplexity.ai/search?q=Wh...    not conserved   \n",
       "4589  <a href=\"https://www.perplexity.ai/search?q=Wh...    not conserved   \n",
       "4590  <a href=\"https://www.perplexity.ai/search?q=Wh...    not conserved   \n",
       "4591  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "4592  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "4593  <a href=\"https://www.perplexity.ai/search?q=Wh...    not conserved   \n",
       "4594  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "4595  <a href=\"https://www.perplexity.ai/search?q=Wh...    not conserved   \n",
       "4596  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "4597  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "\n",
       "                                                   PMID  \\\n",
       "4565  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4566  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4567  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4568  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4569  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4570  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4571  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4572  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4573  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4574  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4575  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4576  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4577  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4578  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4579  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4580  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4581  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4582  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4583  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4584  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4585  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4586  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4587  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4588  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4589  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4590  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4591  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4592  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4593  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4594  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4595  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4596  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4597  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "\n",
       "                                            Ligand Name  \\\n",
       "4565                     CD99 molecule (Xg blood group)   \n",
       "4566                      C-C motif chemokine ligand 12   \n",
       "4567                                                nan   \n",
       "4568                                                nan   \n",
       "4569                                                nan   \n",
       "4570                                                nan   \n",
       "4571                                                nan   \n",
       "4572                       C-C motif chemokine ligand 6   \n",
       "4573                       C-C motif chemokine ligand 6   \n",
       "4574                       C-C motif chemokine ligand 9   \n",
       "4575                                     CD200 molecule   \n",
       "4576                                     CD200 molecule   \n",
       "4577                                    defensin beta 2   \n",
       "4578                                    defensin beta 2   \n",
       "4579                                          ficolin A   \n",
       "4580                                                nan   \n",
       "4581                                                nan   \n",
       "4582                                                nan   \n",
       "4583  leukocyte associated immunoglobulin like recep...   \n",
       "4584                              protocadherin beta 10   \n",
       "4585                              protocadherin beta 11   \n",
       "4586                              protocadherin beta 12   \n",
       "4587                              protocadherin beta 14   \n",
       "4588                              protocadherin beta 21   \n",
       "4589                               protocadherin beta 4   \n",
       "4590                               protocadherin beta 7   \n",
       "4591                               protocadherin beta 8   \n",
       "4592                               protocadherin beta 8   \n",
       "4593                               protocadherin beta 9   \n",
       "4594                 protocadherin gamma subfamily B, 8   \n",
       "4595                                  serum amyloid A 3   \n",
       "4596                                      semaphorin 4A   \n",
       "4597                      zona pellucida glycoprotein 3   \n",
       "\n",
       "                            Receptor Name Mouse_ligand Mouse_receptor  \\\n",
       "4565                                  nan          NaN            NaN   \n",
       "4566       C-C motif chemokine receptor 2        Ccl12           Ccr2   \n",
       "4567     C-X-C motif chemokine receptor 3          NaN            NaN   \n",
       "4568       C-C motif chemokine receptor 7          NaN            NaN   \n",
       "4569     C-X-C motif chemokine receptor 3          NaN            NaN   \n",
       "4570     C-X-C motif chemokine receptor 3          NaN            NaN   \n",
       "4571     C-X-C motif chemokine receptor 3          NaN            NaN   \n",
       "4572       C-C motif chemokine receptor 1         Ccl6           Ccr1   \n",
       "4573       C-C motif chemokine receptor 1         Ccl6           Ccr1   \n",
       "4574       C-C motif chemokine receptor 1         Ccl9           Ccr1   \n",
       "4575                                  nan          NaN            NaN   \n",
       "4576                                  nan          NaN            NaN   \n",
       "4577       C-C motif chemokine receptor 6        Defb2           Ccr6   \n",
       "4578                 toll-like receptor 4        Defb2           Tlr4   \n",
       "4579                 toll-like receptor 4         Fcna           Tlr4   \n",
       "4580  killer cell lectin like receptor K1          NaN            NaN   \n",
       "4581  killer cell lectin like receptor K1          NaN            NaN   \n",
       "4582  killer cell lectin like receptor K1          NaN            NaN   \n",
       "4583                                  nan          NaN            NaN   \n",
       "4584                protocadherin beta 10      Pcdhb10        Pcdhb10   \n",
       "4585                protocadherin beta 11      Pcdhb11        Pcdhb11   \n",
       "4586                protocadherin beta 12      Pcdhb12        Pcdhb12   \n",
       "4587                protocadherin beta 14      Pcdhb14        Pcdhb14   \n",
       "4588                protocadherin beta 21      Pcdhb21        Pcdhb21   \n",
       "4589                 protocadherin beta 4       Pcdhb4         Pcdhb4   \n",
       "4590                 protocadherin beta 7       Pcdhb7         Pcdhb7   \n",
       "4591                 protocadherin beta 8       Pcdhb8         Pcdhb8   \n",
       "4592                 protocadherin beta 8       Pcdhb8         Pcdhb8   \n",
       "4593                 protocadherin beta 9       Pcdhb9         Pcdhb9   \n",
       "4594   protocadherin gamma subfamily B, 8      Pcdhgb8        Pcdhgb8   \n",
       "4595                 toll-like receptor 4         Saa3           Tlr4   \n",
       "4596                                  nan          NaN            NaN   \n",
       "4597                                  nan          NaN            NaN   \n",
       "\n",
       "     Ligand MGI ID Receptor MGI ID  \n",
       "4565           NaN             NaN  \n",
       "4566    MGI:108224      MGI:106185  \n",
       "4567           NaN             NaN  \n",
       "4568           NaN             NaN  \n",
       "4569           NaN             NaN  \n",
       "4570           NaN             NaN  \n",
       "4571           NaN             NaN  \n",
       "4572     MGI:98263      MGI:104618  \n",
       "4573     MGI:98263      MGI:104618  \n",
       "4574    MGI:104533      MGI:104618  \n",
       "4575           NaN             NaN  \n",
       "4576           NaN             NaN  \n",
       "4577   MGI:1338754     MGI:1333797  \n",
       "4578   MGI:1338754       MGI:96824  \n",
       "4579   MGI:1340905       MGI:96824  \n",
       "4580           NaN             NaN  \n",
       "4581           NaN             NaN  \n",
       "4582           NaN             NaN  \n",
       "4583           NaN             NaN  \n",
       "4584   MGI:2136745     MGI:2136745  \n",
       "4585   MGI:2136746     MGI:2136746  \n",
       "4586   MGI:2136747     MGI:2136747  \n",
       "4587   MGI:2136749     MGI:2136749  \n",
       "4588   MGI:2136759     MGI:2136759  \n",
       "4589   MGI:2136738     MGI:2136738  \n",
       "4590   MGI:2136741     MGI:2136741  \n",
       "4591   MGI:2136742     MGI:2136742  \n",
       "4592   MGI:2136742     MGI:2136742  \n",
       "4593   MGI:2136744     MGI:2136744  \n",
       "4594   MGI:1935200     MGI:1935200  \n",
       "4595     MGI:98223       MGI:96824  \n",
       "4596           NaN             NaN  \n",
       "4597           NaN             NaN  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair0_copy[gene_pair0_copy[\"LR Pair Card\"].isin(grab_mouse_info)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "136ee3b5-740b-48da-9265-53f8a432ccab",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_id_map = pd.concat([\n",
    "    gene_pair_mouse_card[['MGI ligand', 'ENSEMBL ligand']].rename(columns={\n",
    "        'MGI ligand': 'MGI ID', 'ENSEMBL ligand': 'ENSEMBL ID'\n",
    "    }),\n",
    "    gene_pair_mouse_card[['MGI receptor', 'ENSEMBL receptor']].rename(columns={\n",
    "        'MGI receptor': 'MGI ID', 'ENSEMBL receptor': 'ENSEMBL ID'\n",
    "    })\n",
    "])\n",
    "\n",
    "# Drop duplicate rows\n",
    "mapping_mouse_ens = mouse_id_map.drop_duplicates().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8f33711c-9ecf-485e-bced-d52aa80a1c1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_mgi_id(col):\n",
    "    \"\"\"Use regular expression to extract the HGNC ID after 'HGNC:'.\"\"\"\n",
    "    match = re.search(r'MGI:(\\d+)', col)\n",
    "    if match:\n",
    "        return 'MGI:' +str(match.group(1))\n",
    "    return None\n",
    "\n",
    "### For latest DB, no need\n",
    "# gene_pair0_copy['Ligand MGI ID'] = gene_pair0_copy['Ligand MGI ID'].apply(extract_mgi_id)\n",
    "# gene_pair0_copy['Receptor MGI ID'] = gene_pair0_copy['Receptor MGI ID'].apply(extract_mgi_id)\n",
    "\n",
    "gene_pair_mouse_card_lim = gene_pair_mouse_card[[\"LR Pair Card\", \"Mouse_ligand\", \"Mouse_receptor\", \"MGI ligand\", \"MGI receptor\"]]\n",
    "gene_pair0_copy =gene_pair0_copy.merge(gene_pair_mouse_card_lim, how=\"left\", on=\"LR Pair Card\")\n",
    "gene_pair0_copy = gene_pair0_copy.rename(columns={\n",
    "        'MGI ligand': 'Ligand MGI ID',\n",
    "        'MGI receptor': 'Receptor MGI ID'\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0726fdd0-c1b6-4cf8-9886-081d5042cdf6",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_75207/844418864.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m gene_pair0_copy['Ligand Name'] = gene_pair0_copy.apply(\n\u001b[0m\u001b[1;32m      2\u001b[0m     \u001b[0;32mlambda\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mmapping_mouse_name\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Ligand MGI ID'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Ligand Name'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnotna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Ligand MGI ID'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Ligand Name'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, func, axis, raw, result_type, args, by_row, engine, engine_kwargs, **kwargs)\u001b[0m\n\u001b[1;32m  10370\u001b[0m             \u001b[0mengine_kwargs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mengine_kwargs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m  10371\u001b[0m             \u001b[0margs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m  10372\u001b[0m             \u001b[0mkwargs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m  10373\u001b[0m         )\n\u001b[0;32m> 10374\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mop\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__finalize__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"apply\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    912\u001b[0m         \u001b[0;31m# raw\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    913\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mraw\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    914\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_raw\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine_kwargs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine_kwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    915\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 916\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_standard\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1061\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mapply_standard\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1062\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"python\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1063\u001b[0;31m             \u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mres_index\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_series_generator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1064\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1065\u001b[0m             \u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mres_index\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_series_numba\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1066\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1077\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1078\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0moption_context\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"mode.chained_assignment\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1079\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseries_gen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1080\u001b[0m                 \u001b[0;31m# ignore SettingWithCopy here in case the user mutates\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1081\u001b[0;31m                 \u001b[0mresults\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1082\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mABCSeries\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1083\u001b[0m                     \u001b[0;31m# If we have a view on v, we need to make a copy because\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1084\u001b[0m                     \u001b[0;31m#  series_generator will swap out the underlying data\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_75207/844418864.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(row)\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m     \u001b[0;32mlambda\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mmapping_mouse_name\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Ligand MGI ID'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Ligand Name'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1575\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mfinal\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1576\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__nonzero__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mNoReturn\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1577\u001b[0;31m         raise ValueError(\n\u001b[0m\u001b[1;32m   1578\u001b[0m             \u001b[0;34mf\"The truth value of a {type(self).__name__} is ambiguous. \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1579\u001b[0m             \u001b[0;34m\"Use a.empty, a.bool(), a.item(), a.any() or a.all().\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1580\u001b[0m         )\n",
      "\u001b[0;31mValueError\u001b[0m: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all()."
     ]
    }
   ],
   "source": [
    "gene_pair0_copy['Ligand Name'] = gene_pair0_copy.apply(\n",
    "    lambda row: mapping_mouse_name.get(row['Ligand MGI ID'], row['Ligand Name'])\n",
    "    if pd.notna(row['Ligand MGI ID']) else row['Ligand Name'],\n",
    "    axis=1\n",
    ")\n",
    "gene_pair0_copy['Receptor Name'] = gene_pair0_copy.apply(\n",
    "    lambda row: mapping_mouse_name.get(row['Receptor MGI ID'], row['Receptor Name'])\n",
    "    if pd.notna(row['Receptor MGI ID']) else row['Receptor Name'],\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "197fbe4b-b6e8-4930-8cc9-aee8d1501cce",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'is_mouse_specific'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/indexes/base.py:3805\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3804\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3805\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3806\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32mindex.pyx:167\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mindex.pyx:196\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7081\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7089\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'is_mouse_specific'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[28], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mgene_pair_keywords_filtered\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mis_mouse_specific\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/frame.py:4102\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   4100\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   4101\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 4102\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4103\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   4104\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/indexes/base.py:3812\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3807\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[1;32m   3808\u001b[0m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[1;32m   3809\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[1;32m   3810\u001b[0m     ):\n\u001b[1;32m   3811\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[0;32m-> 3812\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3813\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3814\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3815\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3816\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3817\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'is_mouse_specific'"
     ]
    }
   ],
   "source": [
    "gene_pair_keywords_filtered['is_mouse_specific']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "cae8dbd2-92a2-4c1d-bd1c-5a9430d0d130",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "from datetime import datetime\n",
    "import os\n",
    "\n",
    "# Create data folder if it doesn't exist\n",
    "os.makedirs(\"data\", exist_ok=True)\n",
    "\n",
    "# Today's date for versioning\n",
    "today = datetime.now().strftime(\"%Y%m%d\")\n",
    "\n",
    "# URLs and filenames\n",
    "homology_url = \"https://www.informatics.jax.org/downloads/reports/HGNC_AllianceHomology.rpt\"\n",
    "mrk_url = \"https://www.informatics.jax.org/downloads/reports/MRK_Sequence.rpt\"\n",
    "mrk_list_url = \"https://www.informatics.jax.org/downloads/reports/MRK_List2.rpt\"\n",
    "\n",
    "homology_file = f\"data/HGNC_AllianceHomology_{today}.tsv\"\n",
    "mrk_file = f\"data/MRK_Sequence_{today}.tsv\"\n",
    "mrk_list_file = f\"data/MRK_List2_{today}.tsv\"\n",
    "merged_file = f\"data/MRK_Merged_{today}_MGI_DB.tsv\"\n",
    "\n",
    "# Download HGNC_AllianceHomology.rpt\n",
    "response = requests.get(homology_url)\n",
    "response.raise_for_status()\n",
    "with open(homology_file, \"wb\") as f:\n",
    "    f.write(response.content)\n",
    "\n",
    "# Download MRK_Sequence.rpt\n",
    "response = requests.get(mrk_url)\n",
    "response.raise_for_status()\n",
    "with open(mrk_file, \"wb\") as f:\n",
    "    f.write(response.content)\n",
    "\n",
    "# Download MRK_List2.rpt\n",
    "response = requests.get(mrk_list_url)\n",
    "response.raise_for_status()\n",
    "with open(mrk_list_file, \"wb\") as f:\n",
    "    f.write(response.content)\n",
    "\n",
    "# Load files\n",
    "df_mrk = pd.read_csv(mrk_file, sep=\"\\t\", dtype=str)\n",
    "df_homology = pd.read_csv(homology_file, sep=\"\\t\", dtype=str, index_col = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b21940cd-ca0b-4f2c-83e4-1dd2b010ee40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'MGI:1341858': 'DNA segment, 03B03F (Research Genetics)',\n",
       " 'MGI:1341869': 'DNA segment, 03B03R (Research Genetics)',\n",
       " 'MGI:1918911': 'RIKEN cDNA 0610005C13 gene',\n",
       " 'MGI:1923503': 'RIKEN cDNA 0610006L08 gene',\n",
       " 'MGI:1925547': 'RIKEN cDNA 0610008J02 gene',\n",
       " 'MGI:3698435': 'RIKEN cDNA 0610009E02 gene',\n",
       " 'MGI:1918921': 'RIKEN cDNA 0610009F21 gene',\n",
       " 'MGI:1918931': 'RIKEN cDNA 0610009K14 gene',\n",
       " 'MGI:1914088': 'RIKEN cDNA 0610009L18 gene',\n",
       " 'MGI:1918926': 'RIKEN cDNA 0610012D04 gene',\n",
       " 'MGI:1925548': 'RIKEN cDNA 0610012E21 gene',\n",
       " 'MGI:1915618': 'RIKEN cDNA 0610025J13 gene',\n",
       " 'MGI:1915614': 'RIKEN cDNA 0610030E20 gene',\n",
       " 'MGI:1925549': 'RIKEN cDNA 0610031C06 gene',\n",
       " 'MGI:1921093': 'RIKEN cDNA 0610031I08 gene',\n",
       " 'MGI:1915619': 'RIKEN cDNA 0610031O16 gene',\n",
       " 'MGI:1925551': 'RIKEN cDNA 0610033I19 gene',\n",
       " 'MGI:1921343': 'RIKEN cDNA 0610033M10 gene',\n",
       " 'MGI:1925613': 'RIKEN cDNA 0610033P13 gene',\n",
       " 'MGI:1925614': 'RIKEN cDNA 0610037L18 gene',\n",
       " 'MGI:1917595': 'RIKEN cDNA 0610038B21 gene',\n",
       " 'MGI:1922649': 'RIKEN cDNA 0610039G09 gene',\n",
       " 'MGI:1915647': 'RIKEN cDNA 0610039H22 gene',\n",
       " 'MGI:1915636': 'RIKEN cDNA 0610039K10 gene',\n",
       " 'MGI:1915626': 'RIKEN cDNA 0610040A22 gene',\n",
       " 'MGI:1914922': 'RIKEN cDNA 0610040B10 gene',\n",
       " 'MGI:1922644': 'RIKEN cDNA 0610040F04 gene',\n",
       " 'MGI:1918938': 'RIKEN cDNA 0610042B02 gene',\n",
       " 'MGI:1922652': 'RIKEN cDNA 0610042B23 gene',\n",
       " 'MGI:1921345': 'RIKEN cDNA 0610042E11 gene',\n",
       " 'MGI:1915630': 'RIKEN cDNA 0610042G04 gene',\n",
       " 'MGI:1915650': 'RIKEN cDNA 0610043K17 gene',\n",
       " 'MGI:1915652': 'RIKEN cDNA 0710001A04 gene',\n",
       " 'MGI:1914929': 'RIKEN cDNA 0710001D07 gene',\n",
       " 'MGI:1915664': 'RIKEN cDNA 0710001L09 gene',\n",
       " 'MGI:1918942': 'RIKEN cDNA 0710005I21 gene',\n",
       " 'MGI:1915659': 'RIKEN cDNA 0710007G10 gene',\n",
       " 'MGI:1918947': 'RIKEN cDNA 0710008F09 gene',\n",
       " 'MGI:1918945': 'RIKEN cDNA 0910001E24 gene',\n",
       " 'MGI:1922653': 'RIKEN cDNA 1010001B22 gene',\n",
       " 'MGI:1915683': 'RIKEN cDNA 1010001I08 gene',\n",
       " 'MGI:1860915': 'DNA segment, 102g4T7',\n",
       " 'MGI:1860916': 'DNA segment, 102F12R',\n",
       " 'MGI:1336996': 'DNA segment, 10S',\n",
       " 'MGI:1337121': 'DNA segment, 10T',\n",
       " 'MGI:1922655': 'RIKEN cDNA 1100001I12 gene',\n",
       " 'MGI:1920975': 'RIKEN cDNA 1110001D16 gene',\n",
       " 'MGI:1915702': 'RIKEN cDNA 1110001M07 gene',\n",
       " 'MGI:1915066': 'RIKEN cDNA 1110002E22 gene',\n",
       " 'MGI:1915738': 'RIKEN cDNA 1110002J07 gene',\n",
       " 'MGI:1915821': 'RIKEN cDNA 1110002L01 gene',\n",
       " 'MGI:1915733': 'RIKEN cDNA 1110002O04 gene',\n",
       " 'MGI:1915716': 'RIKEN cDNA 1110003F02 gene',\n",
       " 'MGI:1915726': 'RIKEN cDNA 1110003F10 gene',\n",
       " 'MGI:1929274': 'RIKEN cDNA 1110004F10 gene',\n",
       " 'MGI:1915714': 'RIKEN cDNA 1110004M10 gene',\n",
       " 'MGI:1923536': 'RIKEN cDNA 1110006E14 gene',\n",
       " 'MGI:1913373': 'RIKEN cDNA 1110006O24 gene',\n",
       " 'MGI:1915732': 'RIKEN cDNA 1110007E10 gene',\n",
       " 'MGI:1915750': 'RIKEN cDNA 1110008E08 gene',\n",
       " 'MGI:1915772': 'RIKEN cDNA 1110013H19 gene',\n",
       " 'MGI:1915879': 'RIKEN cDNA 1110013I04 gene',\n",
       " 'MGI:1913377': 'RIKEN cDNA 1110014L15 gene',\n",
       " 'MGI:1915766': 'RIKEN cDNA 1110015O18 gene',\n",
       " 'MGI:1915844': 'RIKEN cDNA 1110018F16 gene',\n",
       " 'MGI:1915753': 'RIKEN cDNA 1110018N20 gene',\n",
       " 'MGI:1915779': 'RIKEN cDNA 1110019B22 gene',\n",
       " 'MGI:2443389': 'RIKEN cDNA 1110019B24 gene',\n",
       " 'MGI:1923561': 'RIKEN cDNA 1110019D14 gene',\n",
       " 'MGI:1915781': 'RIKEN cDNA 1110020A21 gene',\n",
       " 'MGI:1915834': 'RIKEN cDNA 1110020C17 gene',\n",
       " 'MGI:1915850': 'RIKEN cDNA 1110021P09 gene',\n",
       " 'MGI:1915885': 'RIKEN cDNA 1110025M09 gene',\n",
       " 'MGI:1915940': 'RIKEN cDNA 1110028F11 gene',\n",
       " 'MGI:1915942': 'RIKEN cDNA 1110028F18 gene',\n",
       " 'MGI:1915975': 'RIKEN cDNA 1110032F04 gene',\n",
       " 'MGI:1915977': 'RIKEN cDNA 1110032L06 gene',\n",
       " 'MGI:2147638': 'RIKEN cDNA 1110034N17 gene',\n",
       " 'MGI:1915934': 'RIKEN cDNA 1110035D15 gene',\n",
       " 'MGI:1915922': 'RIKEN cDNA 1110035E04 gene',\n",
       " 'MGI:1915936': 'RIKEN cDNA 1110035H17 gene',\n",
       " 'MGI:1915947': 'RIKEN cDNA 1110036E04 gene',\n",
       " 'MGI:1916001': 'RIKEN cDNA 1110037P09 gene',\n",
       " 'MGI:1916013': 'RIKEN cDNA 1110038B12 gene',\n",
       " 'MGI:2152337': 'RIKEN cDNA 1110038F14 gene',\n",
       " 'MGI:1916058': 'RIKEN cDNA 1110046J04 gene',\n",
       " 'MGI:1916041': 'RIKEN cDNA 1110050P16 gene',\n",
       " 'MGI:1916031': 'RIKEN cDNA 1110055C04 gene',\n",
       " 'MGI:1916021': 'RIKEN cDNA 1110058D11 gene',\n",
       " 'MGI:1913456': 'RIKEN cDNA 1110059E24 gene',\n",
       " 'MGI:1916036': 'RIKEN cDNA 1110059G02 gene',\n",
       " 'MGI:1913452': 'RIKEN cDNA 1110059G10 gene',\n",
       " 'MGI:1916064': 'RIKEN cDNA 1110060G06 gene',\n",
       " 'MGI:1916143': 'RIKEN cDNA 1110064A23 gene',\n",
       " 'MGI:3026937': 'RIKEN cDNA 1110065F06 gene',\n",
       " 'MGI:1916155': 'RIKEN cDNA 1110065H08 gene',\n",
       " 'MGI:1916170': 'RIKEN cDNA 1110065P20 gene',\n",
       " 'MGI:1913046': 'DNA segment, 1-125i7-S',\n",
       " 'MGI:1913045': 'DNA segment, 1-125i7-T',\n",
       " 'MGI:1891878': 'DNA segment, 1-141A1-S',\n",
       " 'MGI:1891910': 'DNA segment, 1-141A1-T',\n",
       " 'MGI:1891886': 'DNA segment, 1-146A2-S',\n",
       " 'MGI:1926216': 'DNA segment, 1-146A2-T',\n",
       " 'MGI:1930517': 'DNA segment, 1-15-T7',\n",
       " 'MGI:1913032': 'DNA segment, 1-166J1-T',\n",
       " 'MGI:1891892': 'DNA segment, 1-168F5-S',\n",
       " 'MGI:1891879': 'DNA segment, 1-168F5-T',\n",
       " 'MGI:1860917': 'DNA segment, 116m19Sp6',\n",
       " 'MGI:1860918': 'DNA segment, 116m19T7',\n",
       " 'MGI:3057268': 'DNA segment,118co4-Sp6',\n",
       " 'MGI:3057269': 'DNA segment,118co4-T7',\n",
       " 'MGI:2153536': 'DNA segment, 118E21-SP6',\n",
       " 'MGI:2153544': 'DNA segment, 118E21-T7',\n",
       " 'MGI:1916173': 'RIKEN cDNA 1190001L17 gene',\n",
       " 'MGI:1916174': 'RIKEN cDNA 1190001M18 gene',\n",
       " 'MGI:1916168': 'RIKEN cDNA 1190005I06 gene',\n",
       " 'MGI:3036227': 'RIKEN cDNA 1190028D05 gene',\n",
       " 'MGI:1341812': 'DNA segment, 119L04aR (Research Genetics)',\n",
       " 'MGI:1341824': 'DNA segment, 119L04F (Research Genetics)',\n",
       " 'MGI:1337061': 'DNA segment, 11S',\n",
       " 'MGI:1337057': 'DNA segment, 11T',\n",
       " 'MGI:1921369': 'RIKEN cDNA 1200007C13 gene',\n",
       " 'MGI:1915687': 'RIKEN cDNA 1200009P19 gene',\n",
       " 'MGI:1918989': 'RIKEN cDNA 1200015M12 gene',\n",
       " 'MGI:1891888': 'DNA segment, 1-222F18-S',\n",
       " 'MGI:1891906': 'DNA segment, 1-223L5-S',\n",
       " 'MGI:1891870': 'DNA segment, 1-223L5-T',\n",
       " 'MGI:1913013': 'DNA segment, 1-23G6-T',\n",
       " 'MGI:1931105': 'DNA segment, 124P15-Sp6',\n",
       " 'MGI:1931110': 'DNA segment, 124P15-T7',\n",
       " 'MGI:1891900': 'DNA segment, 1-291H19-S',\n",
       " 'MGI:1891875': 'DNA segment, 1-291H19-T',\n",
       " 'MGI:1931096': 'DNA segment, 12H19-T7',\n",
       " 'MGI:1337123': 'DNA segment, 12T',\n",
       " 'MGI:1919001': 'RIKEN cDNA 1300002E11 gene',\n",
       " 'MGI:1921410': 'RIKEN cDNA 1300014J16 gene',\n",
       " 'MGI:1891897': 'DNA segment, 1-318K24-S',\n",
       " 'MGI:1891871': 'DNA segment, 1-318K24-T',\n",
       " 'MGI:1913010': 'DNA segment, 1-328P12-S',\n",
       " 'MGI:1913033': 'DNA segment, 1-328P12-T',\n",
       " 'MGI:1891867': 'DNA segment, 1-331B15-S',\n",
       " 'MGI:1891904': 'DNA segment, 1-331B15-T',\n",
       " 'MGI:1341856': 'DNA segment, 133J02F (Research Genetics)',\n",
       " 'MGI:1341846': 'DNA segment, 133J02R (Research Genetics)',\n",
       " 'MGI:1891909': 'DNA segment, 1-34L23-S',\n",
       " 'MGI:1891874': 'DNA segment, 1-34L23-T',\n",
       " 'MGI:1931080': 'DNA segment, 135O21-T7',\n",
       " 'MGI:1913037': 'DNA segment, 1-367K15-S',\n",
       " 'MGI:1913021': 'DNA segment, 1-367K15-T',\n",
       " 'MGI:1860919': 'DNA segment, 137h23T7',\n",
       " 'MGI:2153540': 'DNA segment, 139J18-SP6',\n",
       " 'MGI:2153545': 'DNA segment, 139J18-T7',\n",
       " 'MGI:1913008': 'DNA segment, 1-39J8-T',\n",
       " 'MGI:1913020': 'DNA segment, 1-402G5-S',\n",
       " 'MGI:1913031': 'DNA segment, 1-402G5-T',\n",
       " 'MGI:1891918': 'DNA segment, 1-439F16-T',\n",
       " 'MGI:1891866': 'DNA segment, 1-442G19-S',\n",
       " 'MGI:1891898': 'DNA segment, 1-442G19-T',\n",
       " 'MGI:2137473': 'DNA segment, 144K13-F',\n",
       " 'MGI:1891908': 'DNA segment, 1-450o10-S',\n",
       " 'MGI:1891882': 'DNA segment, 1-450o10-T',\n",
       " 'MGI:1913071': 'DNA segment, 1-465J6-T',\n",
       " 'MGI:1913035': 'DNA segment, 1-473M22-S',\n",
       " 'MGI:1913011': 'DNA segment, 1-473M22-T',\n",
       " 'MGI:1930528': 'DNA segment, 147J19-SP6',\n",
       " 'MGI:1930530': 'DNA segment, 147J19-T7',\n",
       " 'MGI:1860909': 'DNA segment, 14c5Sp6',\n",
       " 'MGI:1860910': 'DNA segment, 14c5T7',\n",
       " 'MGI:1916196': 'RIKEN cDNA 1500002C15 gene',\n",
       " 'MGI:1923933': 'RIKEN cDNA 1500002F19 gene',\n",
       " 'MGI:1916228': 'RIKEN cDNA 1500002I01 gene',\n",
       " 'MGI:1916195': 'RIKEN cDNA 1500002J14 gene',\n",
       " 'MGI:2442808': 'RIKEN cDNA 1500004A13 gene',\n",
       " 'MGI:1917015': 'RIKEN cDNA 1500004F05 gene',\n",
       " 'MGI:1919269': 'RIKEN cDNA 1500005C15 gene',\n",
       " 'MGI:1916206': 'RIKEN cDNA 1500006G06 gene',\n",
       " 'MGI:1917034': 'RIKEN cDNA 1500009L16 gene',\n",
       " 'MGI:1919052': 'RIKEN cDNA 1500010C09 gene',\n",
       " 'MGI:1913486': 'RIKEN cDNA 1500011B03 gene',\n",
       " 'MGI:1916204': 'RIKEN cDNA 1500012K07 gene',\n",
       " 'MGI:1916232': 'RIKEN cDNA 1500015A07 gene',\n",
       " 'MGI:1916244': 'RIKEN cDNA 1500015L24 gene',\n",
       " 'MGI:1916252': 'RIKEN cDNA 1500026H17 gene',\n",
       " 'MGI:1925555': 'RIKEN cDNA 1500032F14 gene',\n",
       " 'MGI:1916275': 'RIKEN cDNA 1500032P08 gene',\n",
       " 'MGI:1917508': 'RIKEN cDNA 1500035N22 gene',\n",
       " 'MGI:1923940': 'RIKEN cDNA 1500037F05 gene',\n",
       " 'MGI:1913070': 'DNA segment, 1-504F3-T',\n",
       " 'MGI:1924911': 'RIKEN cDNA 1520401O13 gene',\n",
       " 'MGI:1891889': 'DNA segment, 1-521i14-S',\n",
       " 'MGI:1891890': 'DNA segment, 1-521i14-T',\n",
       " 'MGI:1891869': 'DNA segment, 1-539K14-S',\n",
       " 'MGI:1891891': 'DNA segment, 1-539K14-T',\n",
       " 'MGI:1913073': 'DNA segment, 1-53E1-S',\n",
       " 'MGI:1930541': 'DNA segment, 153F6-T7',\n",
       " 'MGI:1891921': 'DNA segment, 1-551G19-S',\n",
       " 'MGI:1913029': 'DNA segment, 1-551G19-T',\n",
       " 'MGI:1913019': 'DNA segment, 1-553F10-S',\n",
       " 'MGI:1860920': 'DNA segment, 156l3T7',\n",
       " 'MGI:1913059': 'DNA segment, 1-575D16-S',\n",
       " 'MGI:1913058': 'DNA segment, 1-575D16-T',\n",
       " 'MGI:2137542': 'DNA segment, 158L23-F',\n",
       " 'MGI:1337020': 'DNA segment, 15S',\n",
       " 'MGI:1337091': 'DNA segment, 15T',\n",
       " 'MGI:1917026': 'RIKEN cDNA 1600002D24 gene',\n",
       " 'MGI:1917033': 'RIKEN cDNA 1600010F14 gene',\n",
       " 'MGI:1917031': 'RIKEN cDNA 1600010M07 gene',\n",
       " 'MGI:1915162': 'RIKEN cDNA 1600012H06 gene',\n",
       " 'MGI:1919275': 'RIKEN cDNA 1600012P17 gene',\n",
       " 'MGI:1919494': 'RIKEN cDNA 1600014C10 gene',\n",
       " 'MGI:1919490': 'RIKEN cDNA 1600014C23 gene',\n",
       " 'MGI:1920498': 'RIKEN cDNA 1600015H23 gene',\n",
       " 'MGI:1917017': 'RIKEN cDNA 1600017G08 gene',\n",
       " 'MGI:1920499': 'RIKEN cDNA 1600017P15 gene',\n",
       " 'MGI:1917018': 'RIKEN cDNA 1600019K03 gene',\n",
       " 'MGI:1919262': 'RIKEN cDNA 1600020E01 gene',\n",
       " 'MGI:1920502': 'RIKEN cDNA 1600022D10 gene',\n",
       " 'MGI:1917038': 'RIKEN cDNA 1600023N17 gene',\n",
       " 'MGI:1919280': 'RIKEN cDNA 1600025M17 gene',\n",
       " 'MGI:1917044': 'RIKEN cDNA 1600027J07 gene',\n",
       " 'MGI:1917047': 'RIKEN cDNA 1600029I14 gene',\n",
       " 'MGI:3045362': 'RIKEN cDNA 1600029O15 gene',\n",
       " 'MGI:1913015': 'DNA segment, 1-652J17-T',\n",
       " 'MGI:1891862': 'DNA segment, 1-65A1-S',\n",
       " 'MGI:1891895': 'DNA segment, 1-65A1-T',\n",
       " 'MGI:1891873': 'DNA segment, 1-670i8-S',\n",
       " 'MGI:1891907': 'DNA segment, 1-670i8-T',\n",
       " 'MGI:1891880': 'DNA segment, 1-671o9-S',\n",
       " 'MGI:1891911': 'DNA segment, 1-671o9-T',\n",
       " 'MGI:1930531': 'DNA segment, 167L16-SP6',\n",
       " 'MGI:1930529': 'DNA segment, 167L17-T7',\n",
       " 'MGI:1891876': 'DNA segment, 1-67N17-S',\n",
       " 'MGI:1891903': 'DNA segment, 1-67N17-T',\n",
       " 'MGI:1337017': 'DNA segment, 16S',\n",
       " 'MGI:1337013': 'DNA segment, 16T',\n",
       " 'MGI:1916554': 'RIKEN cDNA 1700001A13 gene',\n",
       " 'MGI:1916543': 'RIKEN cDNA 1700001C07 gene',\n",
       " 'MGI:1916529': 'RIKEN cDNA 1700001D01 gene',\n",
       " 'MGI:1922688': 'RIKEN cDNA 1700001E04 gene',\n",
       " 'MGI:1919076': 'RIKEN cDNA 1700001F09 gene',\n",
       " 'MGI:1922683': 'RIKEN cDNA 1700001G01 gene',\n",
       " 'MGI:1916553': 'RIKEN cDNA 1700001G11 gene',\n",
       " 'MGI:1914753': 'RIKEN cDNA 1700001G17 gene',\n",
       " 'MGI:1919456': 'RIKEN cDNA 1700001I21 gene',\n",
       " 'MGI:1916532': 'RIKEN cDNA 1700001J03 gene',\n",
       " 'MGI:1919075': 'RIKEN cDNA 1700001J04 gene',\n",
       " 'MGI:1919474': 'RIKEN cDNA 1700001J11 gene',\n",
       " 'MGI:1913573': 'RIKEN cDNA 1700001K19 gene',\n",
       " 'MGI:1916569': 'RIKEN cDNA 1700001K23 gene',\n",
       " 'MGI:1916541': 'RIKEN cDNA 1700001L05 gene',\n",
       " 'MGI:1923629': 'RIKEN cDNA 1700001N15 gene',\n",
       " 'MGI:1923631': 'RIKEN cDNA 1700001O22 gene',\n",
       " 'MGI:1916595': 'RIKEN cDNA 1700003C15 gene',\n",
       " 'MGI:1916588': 'RIKEN cDNA 1700003D09 gene',\n",
       " 'MGI:1919087': 'RIKEN cDNA 1700003E16 gene',\n",
       " 'MGI:1916607': 'RIKEN cDNA 1700003E24 gene',\n",
       " 'MGI:1922730': 'RIKEN cDNA 1700003F12 gene',\n",
       " 'MGI:1916600': 'RIKEN cDNA 1700003G18 gene',\n",
       " 'MGI:1923637': 'RIKEN cDNA 1700003H04 gene',\n",
       " 'MGI:1923949': 'RIKEN cDNA 1700003I16 gene',\n",
       " 'MGI:1916606': 'RIKEN cDNA 1700003I22 gene',\n",
       " 'MGI:1919457': 'RIKEN cDNA 1700003K11 gene',\n",
       " 'MGI:1922743': 'RIKEN cDNA 1700003L19 gene',\n",
       " 'MGI:1919475': 'RIKEN cDNA 1700003M07 gene',\n",
       " 'MGI:1919464': 'RIKEN cDNA 1700003N22 gene',\n",
       " 'MGI:1922904': 'RIKEN cDNA 1700003O08 gene',\n",
       " 'MGI:1922882': 'RIKEN cDNA 1700003O11 gene',\n",
       " 'MGI:1916594': 'RIKEN cDNA 1700003P14 gene',\n",
       " 'MGI:1919074': 'RIKEN cDNA 1700006A11 gene',\n",
       " 'MGI:1922682': 'RIKEN cDNA 1700006F04 gene',\n",
       " 'MGI:1919491': 'RIKEN cDNA 1700006H20 gene',\n",
       " 'MGI:1919507': 'RIKEN cDNA 1700006H21 gene',\n",
       " 'MGI:2447818': 'RIKEN cDNA 1700006J14 gene',\n",
       " 'MGI:1925628': 'RIKEN cDNA 1700006P03 gene',\n",
       " 'MGI:1922883': 'RIKEN cDNA 1700007A02 gene',\n",
       " 'MGI:2149745': 'RIKEN cDNA 1700007E05 gene',\n",
       " 'MGI:1915100': 'RIKEN cDNA 1700007F19 gene',\n",
       " 'MGI:1922734': 'RIKEN cDNA 1700007H22 gene',\n",
       " 'MGI:1916570': 'RIKEN cDNA 1700007J10 gene',\n",
       " 'MGI:1922742': 'RIKEN cDNA 1700007J24 gene',\n",
       " 'MGI:1916581': 'RIKEN cDNA 1700007L15 gene',\n",
       " 'MGI:1919487': 'RIKEN cDNA 1700007M16 gene',\n",
       " 'MGI:1916558': 'RIKEN cDNA 1700007P06 gene',\n",
       " 'MGI:1922748': 'RIKEN cDNA 1700008A07 gene',\n",
       " 'MGI:1922713': 'RIKEN cDNA 1700008A23 gene',\n",
       " 'MGI:1913571': 'RIKEN cDNA 1700008B11 gene',\n",
       " 'MGI:1922695': 'RIKEN cDNA 1700008B15 gene',\n",
       " 'MGI:1922744': 'RIKEN cDNA 1700008C04 gene',\n",
       " 'MGI:1922881': 'RIKEN cDNA 1700008F19 gene',\n",
       " 'MGI:1913575': 'RIKEN cDNA 1700008H02 gene',\n",
       " 'MGI:1916610': 'RIKEN cDNA 1700008J07 gene',\n",
       " 'MGI:1916561': 'RIKEN cDNA 1700008K24 gene',\n",
       " 'MGI:1922692': 'RIKEN cDNA 1700008N11 gene',\n",
       " 'MGI:1922699': 'RIKEN cDNA 1700008N17 gene',\n",
       " 'MGI:1916599': 'RIKEN cDNA 1700008O03 gene',\n",
       " 'MGI:1922888': 'RIKEN cDNA 1700008O09 gene',\n",
       " 'MGI:1916597': 'RIKEN cDNA 1700008P02 gene',\n",
       " 'MGI:1921470': 'RIKEN cDNA 1700009C05 gene',\n",
       " 'MGI:1914938': 'RIKEN cDNA 1700009J07 gene',\n",
       " 'MGI:1922721': 'RIKEN cDNA 1700009N14 gene',\n",
       " 'MGI:1922735': 'RIKEN cDNA 1700010B08 gene',\n",
       " 'MGI:1916591': 'RIKEN cDNA 1700010B09 gene',\n",
       " 'MGI:1922741': 'RIKEN cDNA 1700010B13 gene',\n",
       " 'MGI:1923636': 'RIKEN cDNA 1700010D01 gene',\n",
       " 'MGI:1921467': 'RIKEN cDNA 1700010G06 gene',\n",
       " 'MGI:1922750': 'RIKEN cDNA 1700010H22 gene',\n",
       " 'MGI:1922726': 'RIKEN cDNA 1700010I02 gene',\n",
       " 'MGI:1914181': 'RIKEN cDNA 1700010I14 gene',\n",
       " 'MGI:1916605': 'RIKEN cDNA 1700010J16 gene',\n",
       " 'MGI:1922751': 'RIKEN cDNA 1700010K23 gene',\n",
       " 'MGI:1922690': 'RIKEN cDNA 1700010K24 gene',\n",
       " 'MGI:1916583': 'RIKEN cDNA 1700010L04 gene',\n",
       " 'MGI:1922878': 'RIKEN cDNA 1700010L13 gene',\n",
       " 'MGI:1916590': 'RIKEN cDNA 1700010N08 gene',\n",
       " 'MGI:1922894': 'RIKEN cDNA 1700011B04 gene',\n",
       " 'MGI:1922701': 'RIKEN cDNA 1700011C11 gene',\n",
       " 'MGI:1922714': 'RIKEN cDNA 1700011D18 gene',\n",
       " 'MGI:1916535': 'RIKEN cDNA 1700011L03 gene',\n",
       " 'MGI:1914937': 'RIKEN cDNA 1700011L22 gene',\n",
       " 'MGI:1923632': 'RIKEN cDNA 1700012A03 gene',\n",
       " 'MGI:1916574': 'RIKEN cDNA 1700012B07 gene',\n",
       " 'MGI:1916575': 'RIKEN cDNA 1700012B09 gene',\n",
       " 'MGI:1923635': 'RIKEN cDNA 1700012C08 gene',\n",
       " 'MGI:1916587': 'RIKEN cDNA 1700012C14 gene',\n",
       " 'MGI:1919493': 'RIKEN cDNA 1700012D01 gene',\n",
       " 'MGI:1922739': 'RIKEN cDNA 1700012D16 gene',\n",
       " 'MGI:1923817': 'RIKEN cDNA 1700012E03 gene',\n",
       " 'MGI:1923643': 'RIKEN cDNA 1700012H19 gene',\n",
       " 'MGI:1916584': 'RIKEN cDNA 1700012I11 gene',\n",
       " 'MGI:1922737': 'RIKEN cDNA 1700012J15 gene',\n",
       " 'MGI:1922746': 'RIKEN cDNA 1700012J22 gene',\n",
       " 'MGI:1923644': 'RIKEN cDNA 1700012M20 gene',\n",
       " 'MGI:1922738': 'RIKEN cDNA 1700012O15 gene',\n",
       " 'MGI:1916616': 'RIKEN cDNA 1700013A02 gene',\n",
       " 'MGI:1924171': 'RIKEN cDNA 1700013D24 gene',\n",
       " 'MGI:1916630': 'RIKEN cDNA 1700013G24 gene',\n",
       " 'MGI:1922764': 'RIKEN cDNA 1700013H16 gene',\n",
       " 'MGI:1919497': 'RIKEN cDNA 1700013K18 gene',\n",
       " 'MGI:1922755': 'RIKEN cDNA 1700013M08 gene',\n",
       " 'MGI:1922884': 'RIKEN cDNA 1700013N06 gene',\n",
       " 'MGI:1919498': 'RIKEN cDNA 1700014B07 gene',\n",
       " 'MGI:1924191': 'RIKEN cDNA 1700014D06 gene',\n",
       " 'MGI:1924172': 'RIKEN cDNA 1700014F14 gene',\n",
       " 'MGI:1924173': 'RIKEN cDNA 1700014L14 gene',\n",
       " 'MGI:1922782': 'RIKEN cDNA 1700015C15 gene',\n",
       " 'MGI:1924178': 'RIKEN cDNA 1700015C17 gene',\n",
       " 'MGI:1922777': 'RIKEN cDNA 1700015H07 gene',\n",
       " 'MGI:1922784': 'RIKEN cDNA 1700015I17 gene',\n",
       " 'MGI:1925746': 'RIKEN cDNA 1700015O03 gene',\n",
       " 'MGI:1919105': 'RIKEN cDNA 1700015O11 gene',\n",
       " 'MGI:1919473': 'RIKEN cDNA 1700015P03 gene',\n",
       " 'MGI:1921477': 'RIKEN cDNA 1700016A09 gene',\n",
       " 'MGI:1922890': 'RIKEN cDNA 1700016B01 gene',\n",
       " 'MGI:1916677': 'RIKEN cDNA 1700016B15 gene',\n",
       " 'MGI:1921478': 'RIKEN cDNA 1700016C19 gene',\n",
       " 'MGI:1924181': 'RIKEN cDNA 1700016C23 gene',\n",
       " 'MGI:1922785': 'RIKEN cDNA 1700016D02 gene',\n",
       " 'MGI:1925627': 'RIKEN cDNA 1700016D08 gene',\n",
       " 'MGI:1923668': 'RIKEN cDNA 1700016D18 gene',\n",
       " 'MGI:1916674': 'RIKEN cDNA 1700016F12 gene',\n",
       " 'MGI:1919117': 'RIKEN cDNA 1700016G22 gene',\n",
       " 'MGI:1922903': 'RIKEN cDNA 1700016H03 gene',\n",
       " 'MGI:1921468': 'RIKEN cDNA 1700016H13 gene',\n",
       " 'MGI:1922811': 'RIKEN cDNA 1700016J18 gene',\n",
       " 'MGI:1922794': 'RIKEN cDNA 1700016K05 gene',\n",
       " 'MGI:1916662': 'RIKEN cDNA 1700016L04 gene',\n",
       " 'MGI:1922899': 'RIKEN cDNA 1700016L15 gene',\n",
       " 'MGI:1919458': 'RIKEN cDNA 1700016L21 gene',\n",
       " 'MGI:1922792': 'RIKEN cDNA 1700016P03 gene',\n",
       " 'MGI:1916663': 'RIKEN cDNA 1700016P04 gene',\n",
       " 'MGI:1924180': 'RIKEN cDNA 1700016P22 gene',\n",
       " 'MGI:1921461': 'RIKEN cDNA 1700017B05 gene',\n",
       " 'MGI:1919477': 'RIKEN cDNA 1700017G19 gene',\n",
       " 'MGI:1921481': 'RIKEN cDNA 1700017H01 gene',\n",
       " 'MGI:1922756': 'RIKEN cDNA 1700017I07 gene',\n",
       " 'MGI:1922758': 'RIKEN cDNA 1700017J07 gene',\n",
       " 'MGI:1916620': 'RIKEN cDNA 1700017L05 gene',\n",
       " 'MGI:1923648': 'RIKEN cDNA 1700017M07 gene',\n",
       " 'MGI:1913855': 'RIKEN cDNA 1700017N19 gene',\n",
       " 'MGI:1916631': 'RIKEN cDNA 1700018A04 gene',\n",
       " 'MGI:1916641': 'RIKEN cDNA 1700018A14 gene',\n",
       " 'MGI:1919481': 'RIKEN cDNA 1700018A23 gene',\n",
       " 'MGI:1923655': 'RIKEN cDNA 1700018B08 gene',\n",
       " 'MGI:1913582': 'RIKEN cDNA 1700018B24 gene',\n",
       " 'MGI:1916646': 'RIKEN cDNA 1700018F24 gene',\n",
       " 'MGI:1922889': 'RIKEN cDNA 1700018H16 gene',\n",
       " 'MGI:1914579': 'RIKEN cDNA 1700018L02 gene',\n",
       " 'MGI:1919495': 'RIKEN cDNA 1700018M17 gene',\n",
       " 'MGI:1920580': 'RIKEN cDNA 1700018P08 gene',\n",
       " 'MGI:1916645': 'RIKEN cDNA 1700018P22 gene',\n",
       " 'MGI:1916647': 'RIKEN cDNA 1700019A02 gene',\n",
       " 'MGI:1915470': 'RIKEN cDNA 1700019A23 gene',\n",
       " 'MGI:1922793': 'RIKEN cDNA 1700019C18 gene',\n",
       " 'MGI:1914330': 'RIKEN cDNA 1700019D03 gene',\n",
       " 'MGI:1916655': 'RIKEN cDNA 1700019E08 gene',\n",
       " 'MGI:1916652': 'RIKEN cDNA 1700019F05 gene',\n",
       " 'MGI:1919499': 'RIKEN cDNA 1700019G24 gene',\n",
       " 'MGI:1919462': 'RIKEN cDNA 1700019H22 gene',\n",
       " 'MGI:1923662': 'RIKEN cDNA 1700019J19 gene',\n",
       " 'MGI:1919476': 'RIKEN cDNA 1700019L13 gene',\n",
       " 'MGI:1922799': 'RIKEN cDNA 1700019L22 gene',\n",
       " 'MGI:1916673': 'RIKEN cDNA 1700019M22 gene',\n",
       " 'MGI:1922796': 'RIKEN cDNA 1700019P21 gene',\n",
       " 'MGI:1922906': 'RIKEN cDNA 1700020A23 gene',\n",
       " 'MGI:1919484': 'RIKEN cDNA 1700020B03 gene',\n",
       " 'MGI:1916687': 'RIKEN cDNA 1700020D12 gene',\n",
       " 'MGI:1916688': 'RIKEN cDNA 1700020D14 gene',\n",
       " 'MGI:1921485': 'RIKEN cDNA 1700020G03 gene',\n",
       " 'MGI:1922880': 'RIKEN cDNA 1700020G17 gene',\n",
       " 'MGI:1922804': 'RIKEN cDNA 1700020K04 gene',\n",
       " 'MGI:1922887': 'RIKEN cDNA 1700020L05 gene',\n",
       " 'MGI:1922879': 'RIKEN cDNA 1700020L13 gene',\n",
       " 'MGI:1913580': 'RIKEN cDNA 1700020L24 gene',\n",
       " 'MGI:1922812': 'RIKEN cDNA 1700020M10 gene',\n",
       " 'MGI:1923652': 'RIKEN cDNA 1700020M21 gene',\n",
       " 'MGI:1914942': 'RIKEN cDNA 1700020N01 gene',\n",
       " 'MGI:1922759': 'RIKEN cDNA 1700020N15 gene',\n",
       " 'MGI:1914336': 'RIKEN cDNA 1700020N18 gene',\n",
       " 'MGI:1924174': 'RIKEN cDNA 1700021A07 gene',\n",
       " 'MGI:1922769': 'RIKEN cDNA 1700021F02 gene',\n",
       " 'MGI:1924176': 'RIKEN cDNA 1700021F13 gene',\n",
       " 'MGI:1920506': 'RIKEN cDNA 1700021G15 gene',\n",
       " 'MGI:1919485': 'RIKEN cDNA 1700021J08 gene',\n",
       " 'MGI:1919478': 'RIKEN cDNA 1700021L22 gene',\n",
       " 'MGI:1923660': 'RIKEN cDNA 1700021L23 gene',\n",
       " 'MGI:1924179': 'RIKEN cDNA 1700021N20 gene',\n",
       " 'MGI:1916659': 'RIKEN cDNA 1700021N21 gene',\n",
       " 'MGI:1922771': 'RIKEN cDNA 1700021P04 gene',\n",
       " 'MGI:1919502': 'RIKEN cDNA 1700022A21 gene',\n",
       " 'MGI:1922807': 'RIKEN cDNA 1700022A22 gene',\n",
       " 'MGI:1919460': 'RIKEN cDNA 1700022C07 gene',\n",
       " 'MGI:1922886': 'RIKEN cDNA 1700022E09 gene',\n",
       " 'MGI:1916671': 'RIKEN cDNA 1700022F17 gene',\n",
       " 'MGI:1922789': 'RIKEN cDNA 1700022H01 gene',\n",
       " 'MGI:1916670': 'RIKEN cDNA 1700022H16 gene',\n",
       " 'MGI:1916681': 'RIKEN cDNA 1700022N22 gene',\n",
       " 'MGI:1916624': 'RIKEN cDNA 1700023A20 gene',\n",
       " 'MGI:1920507': 'RIKEN cDNA 1700023C21 gene',\n",
       " 'MGI:1921473': 'RIKEN cDNA 1700023D08 gene',\n",
       " 'MGI:1916695': 'RIKEN cDNA 1700023D09 gene',\n",
       " 'MGI:1916683': 'RIKEN cDNA 1700023F02 gene',\n",
       " 'MGI:1919483': 'RIKEN cDNA 1700023G08 gene',\n",
       " 'MGI:1920531': 'RIKEN cDNA 1700023G09 gene',\n",
       " 'MGI:1916692': 'RIKEN cDNA 1700023H06 gene',\n",
       " 'MGI:1922768': 'RIKEN cDNA 1700024B05 gene',\n",
       " 'MGI:1916643': 'RIKEN cDNA 1700024B18 gene',\n",
       " 'MGI:1919101': 'RIKEN cDNA 1700024F20 gene',\n",
       " 'MGI:1914335': 'RIKEN cDNA 1700024G13 gene',\n",
       " 'MGI:1922772': 'RIKEN cDNA 1700024I08 gene',\n",
       " 'MGI:1919098': 'RIKEN cDNA 1700024J04 gene',\n",
       " 'MGI:1922775': 'RIKEN cDNA 1700024K11 gene',\n",
       " 'MGI:1922770': 'RIKEN cDNA 1700024N05 gene',\n",
       " 'MGI:1916642': 'RIKEN cDNA 1700024P12 gene',\n",
       " 'MGI:1916650': 'RIKEN cDNA 1700025A08 gene',\n",
       " 'MGI:1923659': 'RIKEN cDNA 1700025B11 gene',\n",
       " 'MGI:1919461': 'RIKEN cDNA 1700025C18 gene',\n",
       " 'MGI:1916667': 'RIKEN cDNA 1700025D23 gene',\n",
       " 'MGI:1914331': 'RIKEN cDNA 1700025F24 gene',\n",
       " 'MGI:1916649': 'RIKEN cDNA 1700025G04 gene',\n",
       " 'MGI:1922781': 'RIKEN cDNA 1700025H01 gene',\n",
       " 'MGI:1916657': 'RIKEN cDNA 1700025J12 gene',\n",
       " 'MGI:1913584': 'RIKEN cDNA 1700025K04 gene',\n",
       " 'MGI:1916668': 'RIKEN cDNA 1700025K24 gene',\n",
       " 'MGI:1914183': 'RIKEN cDNA 1700025L06 gene',\n",
       " 'MGI:1919526': 'RIKEN cDNA 1700025M24 gene',\n",
       " 'MGI:1916665': 'RIKEN cDNA 1700025N21 gene',\n",
       " 'MGI:1917236': 'RIKEN cDNA 1700025N23 gene',\n",
       " 'MGI:1917234': 'RIKEN cDNA 1700025O08 gene',\n",
       " 'MGI:1916660': 'RIKEN cDNA 1700025O18 gene',\n",
       " 'MGI:1917621': 'RIKEN cDNA 1700026D11 gene',\n",
       " 'MGI:1916669': 'RIKEN cDNA 1700026F02 gene',\n",
       " 'MGI:1917626': 'RIKEN cDNA 1700026H06 gene',\n",
       " 'MGI:1920505': 'RIKEN cDNA 1700026J12 gene',\n",
       " 'MGI:1916682': 'RIKEN cDNA 1700026J14 gene',\n",
       " 'MGI:1919521': 'RIKEN cDNA 1700026N04 gene',\n",
       " 'MGI:1917246': 'RIKEN cDNA 1700027A07 gene',\n",
       " 'MGI:1916697': 'RIKEN cDNA 1700027A13 gene',\n",
       " 'MGI:1919518': 'RIKEN cDNA 1700027F06 gene',\n",
       " 'MGI:1917249': 'RIKEN cDNA 1700027F09 gene',\n",
       " 'MGI:1919524': 'RIKEN cDNA 1700027H10 gene',\n",
       " 'MGI:1922817': 'RIKEN cDNA 1700027I24 gene',\n",
       " 'MGI:1916693': 'RIKEN cDNA 1700027J07 gene',\n",
       " 'MGI:1916702': 'RIKEN cDNA 1700027M17 gene',\n",
       " 'MGI:1917251': 'RIKEN cDNA 1700028B04 gene',\n",
       " 'MGI:1922819': 'RIKEN cDNA 1700028D13 gene',\n",
       " 'MGI:1916713': 'RIKEN cDNA 1700028E10 gene',\n",
       " 'MGI:1917252': 'RIKEN cDNA 1700028E11 gene',\n",
       " 'MGI:1917253': 'RIKEN cDNA 1700028I16 gene',\n",
       " 'MGI:1917254': 'RIKEN cDNA 1700028J19 gene',\n",
       " 'MGI:1923671': 'RIKEN cDNA 1700028K03 gene',\n",
       " 'MGI:1916710': 'RIKEN cDNA 1700028M03 gene',\n",
       " 'MGI:1920530': 'RIKEN cDNA 1700028N14 gene',\n",
       " 'MGI:1922905': 'RIKEN cDNA 1700028P15 gene',\n",
       " 'MGI:1916745': 'RIKEN cDNA 1700029B22 gene',\n",
       " 'MGI:1919533': 'RIKEN cDNA 1700029B24 gene',\n",
       " 'MGI:1919500': 'RIKEN cDNA 1700029E06 gene',\n",
       " 'MGI:1913729': 'RIKEN cDNA 1700029F12 gene',\n",
       " 'MGI:1913751': 'RIKEN cDNA 1700029H14 gene',\n",
       " 'MGI:1919514': 'RIKEN cDNA 1700029J03 gene',\n",
       " 'MGI:1917257': 'RIKEN cDNA 1700029J08 gene',\n",
       " 'MGI:1920521': 'RIKEN cDNA 1700029K24 gene',\n",
       " 'MGI:1916726': 'RIKEN cDNA 1700029M03 gene',\n",
       " 'MGI:1913597': 'RIKEN cDNA 1700029M20 gene',\n",
       " 'MGI:1916741': 'RIKEN cDNA 1700029N11 gene',\n",
       " 'MGI:1921429': 'RIKEN cDNA 1700030A11 gene',\n",
       " 'MGI:1916763': 'RIKEN cDNA 1700030C10 gene',\n",
       " 'MGI:1919496': 'RIKEN cDNA 1700030C12 gene',\n",
       " 'MGI:1922912': 'RIKEN cDNA 1700030C14 gene',\n",
       " 'MGI:1916765': 'RIKEN cDNA 1700030E10 gene',\n",
       " 'MGI:1919513': 'RIKEN cDNA 1700030F04 gene',\n",
       " 'MGI:1916753': 'RIKEN cDNA 1700030I03 gene',\n",
       " 'MGI:1916778': 'RIKEN cDNA 1700030J22 gene',\n",
       " 'MGI:1919504': 'RIKEN cDNA 1700030K09 gene',\n",
       " 'MGI:1917305': 'RIKEN cDNA 1700030L20 gene',\n",
       " 'MGI:1922826': 'RIKEN cDNA 1700030L22 gene',\n",
       " 'MGI:1919509': 'RIKEN cDNA 1700030M09 gene',\n",
       " 'MGI:1917261': 'RIKEN cDNA 1700030N03 gene',\n",
       " 'MGI:1917266': 'RIKEN cDNA 1700030N18 gene',\n",
       " 'MGI:1919536': 'RIKEN cDNA 1700030O20 gene',\n",
       " 'MGI:1917324': 'RIKEN cDNA 1700030P01 gene',\n",
       " 'MGI:1920536': 'RIKEN cDNA 1700031A10 gene',\n",
       " 'MGI:1920539': 'RIKEN cDNA 1700031C06 gene',\n",
       " 'MGI:1920550': 'RIKEN cDNA 1700031F05 gene',\n",
       " 'MGI:1920567': 'RIKEN cDNA 1700031F10 gene',\n",
       " 'MGI:1914577': 'RIKEN cDNA 1700031L13 gene',\n",
       " 'MGI:1920552': 'RIKEN cDNA 1700031M16 gene',\n",
       " 'MGI:1920528': 'RIKEN cDNA 1700031P21 gene',\n",
       " 'MGI:1920532': 'RIKEN cDNA 1700034B16 gene',\n",
       " 'MGI:1925664': 'RIKEN cDNA 1700034E13 gene',\n",
       " 'MGI:1914575': 'RIKEN cDNA 1700034G24 gene',\n",
       " 'MGI:1921515': 'RIKEN cDNA 1700034H15 gene',\n",
       " 'MGI:1920554': 'RIKEN cDNA 1700034J04 gene',\n",
       " 'MGI:1920594': 'RIKEN cDNA 1700034J05 gene',\n",
       " 'MGI:1925599': 'RIKEN cDNA 1700034K08 gene',\n",
       " 'MGI:1920581': 'RIKEN cDNA 1700034P13 gene',\n",
       " 'MGI:1920548': 'RIKEN cDNA 1700036A12 gene',\n",
       " 'MGI:1920533': 'RIKEN cDNA 1700036G14 gene',\n",
       " 'MGI:1920512': 'RIKEN cDNA 1700036O09 gene',\n",
       " 'MGI:1921513': 'RIKEN cDNA 1700037C06 gene',\n",
       " 'MGI:1920511': 'RIKEN cDNA 1700037C18 gene',\n",
       " 'MGI:1914566': 'RIKEN cDNA 1700037F03 gene',\n",
       " 'MGI:1920535': 'RIKEN cDNA 1700037F24 gene',\n",
       " 'MGI:1920526': 'RIKEN cDNA 1700037J18 gene',\n",
       " 'MGI:1920514': 'RIKEN cDNA 1700037N05 gene',\n",
       " 'MGI:1920555': 'RIKEN cDNA 1700038C09 gene',\n",
       " 'MGI:1920515': 'RIKEN cDNA 1700038P13 gene',\n",
       " 'MGI:1920572': 'RIKEN cDNA 1700039E22 gene',\n",
       " 'MGI:1920596': 'RIKEN cDNA 1700039I01 gene',\n",
       " 'MGI:1914768': 'RIKEN cDNA 1700039M10 gene',\n",
       " 'MGI:1921516': 'RIKEN cDNA 1700039M15 gene',\n",
       " 'MGI:1915458': 'RIKEN cDNA 1700039O17 gene',\n",
       " 'MGI:1915457': 'RIKEN cDNA 1700040A12 gene',\n",
       " 'MGI:1914562': 'RIKEN cDNA 1700040B14 gene',\n",
       " 'MGI:1923852': 'RIKEN cDNA 1700040D17 gene',\n",
       " 'MGI:1924240': 'RIKEN cDNA 1700040E09 gene',\n",
       " 'MGI:1920579': 'RIKEN cDNA 1700040F15 gene',\n",
       " 'MGI:1920584': 'RIKEN cDNA 1700040F17 gene',\n",
       " 'MGI:1923964': 'RIKEN cDNA 1700040G22 gene',\n",
       " 'MGI:1920517': 'RIKEN cDNA 1700040K01 gene',\n",
       " 'MGI:1920544': 'RIKEN cDNA 1700040L08 gene',\n",
       " 'MGI:1920518': 'RIKEN cDNA 1700040P09 gene',\n",
       " 'MGI:1920562': 'RIKEN cDNA 1700041A01 gene',\n",
       " 'MGI:1925873': 'RIKEN cDNA 1700041B01 gene',\n",
       " 'MGI:1914569': 'RIKEN cDNA 1700041C23 gene',\n",
       " 'MGI:1920549': 'RIKEN cDNA 1700041G16 gene',\n",
       " 'MGI:1920523': 'RIKEN cDNA 1700041I07 gene',\n",
       " 'MGI:1920565': 'RIKEN cDNA 1700041L08 gene',\n",
       " 'MGI:1920510': 'RIKEN cDNA 1700041M05 gene',\n",
       " 'MGI:1923851': 'RIKEN cDNA 1700041M19 gene',\n",
       " 'MGI:1920587': 'RIKEN cDNA 1700042D02 gene',\n",
       " 'MGI:1914574': 'RIKEN cDNA 1700042D18 gene',\n",
       " 'MGI:1920599': 'RIKEN cDNA 1700042G15 gene',\n",
       " 'MGI:1920540': 'RIKEN cDNA 1700042O05 gene',\n",
       " 'MGI:1920571': 'RIKEN cDNA 1700042O10 gene',\n",
       " 'MGI:1923855': 'RIKEN cDNA 1700042O13 gene',\n",
       " 'MGI:1925598': 'RIKEN cDNA 1700043A12 gene',\n",
       " 'MGI:1920556': 'RIKEN cDNA 1700044C05 gene',\n",
       " 'MGI:1914571': 'RIKEN cDNA 1700044K03 gene',\n",
       " 'MGI:1920560': 'RIKEN cDNA 1700044K19 gene',\n",
       " 'MGI:1920598': 'RIKEN cDNA 1700045H11 gene',\n",
       " 'MGI:1920600': 'RIKEN cDNA 1700045I11 gene',\n",
       " 'MGI:1920557': 'RIKEN cDNA 1700046C09 gene',\n",
       " 'MGI:1923854': 'RIKEN cDNA 1700047A11 gene',\n",
       " 'MGI:1921578': 'RIKEN cDNA 1700047E10 gene',\n",
       " 'MGI:1920570': 'RIKEN cDNA 1700047F07 gene',\n",
       " 'MGI:1920589': 'RIKEN cDNA 1700047G03 gene',\n",
       " 'MGI:1920573': 'RIKEN cDNA 1700047G07 gene',\n",
       " 'MGI:1920585': 'RIKEN cDNA 1700047K16 gene',\n",
       " 'MGI:1920545': 'RIKEN cDNA 1700047L14 gene',\n",
       " 'MGI:1914580': 'RIKEN cDNA 1700047M11 gene',\n",
       " 'MGI:1923853': 'RIKEN cDNA 1700047N06 gene',\n",
       " 'MGI:1920668': 'RIKEN cDNA 1700047O18 gene',\n",
       " 'MGI:1920715': 'RIKEN cDNA 1700048B10 gene',\n",
       " 'MGI:1920695': 'RIKEN cDNA 1700048C15 gene',\n",
       " 'MGI:1920646': 'RIKEN cDNA 1700048F04 gene',\n",
       " 'MGI:1925874': 'RIKEN cDNA 1700048G13 gene',\n",
       " 'MGI:1920607': 'RIKEN cDNA 1700048J15 gene',\n",
       " 'MGI:1920636': 'RIKEN cDNA 1700048M11 gene',\n",
       " 'MGI:1920683': 'RIKEN cDNA 1700048O14 gene',\n",
       " 'MGI:1920637': 'RIKEN cDNA 1700048O20 gene',\n",
       " 'MGI:1920605': 'RIKEN cDNA 1700048P04 gene',\n",
       " 'MGI:1920644': 'RIKEN cDNA 1700049E15 gene',\n",
       " 'MGI:1920665': 'RIKEN cDNA 1700049E17 gene, gene 1',\n",
       " 'MGI:3796510': 'tandem duplication of RIKEN cDNA 1700049E17 gene, gene 2',\n",
       " 'MGI:1920645': 'RIKEN cDNA 1700049E22 gene',\n",
       " 'MGI:1920615': 'RIKEN cDNA 1700049J03 gene',\n",
       " 'MGI:1920633': 'RIKEN cDNA 1700049L16 gene',\n",
       " 'MGI:1920616': 'RIKEN cDNA 1700051A21 gene',\n",
       " 'MGI:1920602': 'RIKEN cDNA 1700051K13 gene',\n",
       " 'MGI:3704315': 'RIKEN cDNA 1700051O22 Gene',\n",
       " 'MGI:1920712': 'RIKEN cDNA 1700052H01 gene',\n",
       " 'MGI:1914590': 'RIKEN cDNA 1700052I22 gene',\n",
       " 'MGI:1920621': 'RIKEN cDNA 1700052K07 gene',\n",
       " 'MGI:1920681': 'RIKEN cDNA 1700052K11 gene',\n",
       " 'MGI:1920617': 'RIKEN cDNA 1700052M09 gene',\n",
       " 'MGI:1925665': 'RIKEN cDNA 1700052M18 gene',\n",
       " 'MGI:1925707': 'RIKEN cDNA 1700054A03 gene',\n",
       " 'MGI:1925600': 'RIKEN cDNA 1700054K02 gene',\n",
       " 'MGI:1914586': 'RIKEN cDNA 1700054K19 gene',\n",
       " 'MGI:1920608': 'RIKEN cDNA 1700054M17 gene',\n",
       " 'MGI:1920606': 'RIKEN cDNA 1700054O05 gene',\n",
       " 'MGI:1921522': 'RIKEN cDNA 1700054O19 gene',\n",
       " 'MGI:1920647': 'RIKEN cDNA 1700055C04 gene',\n",
       " 'MGI:1920609': 'RIKEN cDNA 1700055D16 gene',\n",
       " 'MGI:1920655': 'RIKEN cDNA 1700055D18 gene',\n",
       " 'MGI:1915475': 'RIKEN cDNA 1700055K11 gene',\n",
       " 'MGI:1920624': 'RIKEN cDNA 1700055L01 gene',\n",
       " 'MGI:1920613': 'RIKEN cDNA 1700056E22 gene',\n",
       " 'MGI:1920643': 'RIKEN cDNA 1700056I18 gene',\n",
       " 'MGI:1920642': 'RIKEN cDNA 1700056N10 gene',\n",
       " 'MGI:1920714': 'RIKEN cDNA 1700057A11 gene',\n",
       " 'MGI:1920691': 'RIKEN cDNA 1700057D03 gene',\n",
       " 'MGI:1925710': 'RIKEN cDNA 1700057H15 gene',\n",
       " 'MGI:1920625': 'RIKEN cDNA 1700057H21 gene',\n",
       " 'MGI:1923857': 'RIKEN cDNA 1700058J15 gene',\n",
       " 'MGI:1920678': 'RIKEN cDNA 1700058M10 gene',\n",
       " 'MGI:1920673': 'RIKEN cDNA 1700058M13 gene',\n",
       " 'MGI:1920627': 'RIKEN cDNA 1700058P15 gene',\n",
       " 'MGI:1920675': 'RIKEN cDNA 1700060C16 gene',\n",
       " 'MGI:1920649': 'RIKEN cDNA 1700060C20 gene',\n",
       " 'MGI:1920641': 'RIKEN cDNA 1700060J05 gene',\n",
       " 'MGI:1920628': 'RIKEN cDNA 1700060L04 gene',\n",
       " 'MGI:1914585': 'RIKEN cDNA 1700060O08 gene',\n",
       " 'MGI:1920664': 'RIKEN cDNA 1700061A03 gene',\n",
       " 'MGI:1920611': 'RIKEN cDNA 1700061D13 gene',\n",
       " 'MGI:1920666': 'RIKEN cDNA 1700061D14 gene',\n",
       " 'MGI:3642196': 'RIKEN cDNA 1700061E17 gene',\n",
       " 'MGI:1924241': 'RIKEN cDNA 1700061E18 gene',\n",
       " 'MGI:1920688': 'RIKEN cDNA 1700061F12 gene',\n",
       " 'MGI:1920711': 'RIKEN cDNA 1700061H18 gene',\n",
       " 'MGI:1920612': 'RIKEN cDNA 1700061I17 gene',\n",
       " 'MGI:1924243': 'RIKEN cDNA 1700061J23 gene',\n",
       " 'MGI:1920682': 'RIKEN cDNA 1700061N14 gene',\n",
       " 'MGI:1925711': 'RIKEN cDNA 1700062A02 gene',\n",
       " 'MGI:1920690': 'RIKEN cDNA 1700062C10 gene',\n",
       " 'MGI:1920704': 'RIKEN cDNA 1700063A18 gene',\n",
       " 'MGI:1925750': 'RIKEN cDNA 1700063D05 gene',\n",
       " 'MGI:1921519': 'RIKEN cDNA 1700063H04 gene',\n",
       " 'MGI:1920614': 'RIKEN cDNA 1700063H06 gene',\n",
       " 'MGI:1925877': 'RIKEN cDNA 1700063I16 gene',\n",
       " 'MGI:1920700': 'RIKEN cDNA 1700063J08 gene',\n",
       " 'MGI:1923860': 'RIKEN cDNA 1700063K16 gene',\n",
       " 'MGI:1920702': 'RIKEN cDNA 1700063O14 gene',\n",
       " 'MGI:1920687': 'RIKEN cDNA 1700064F23 gene',\n",
       " 'MGI:1920674': 'RIKEN cDNA 1700064H15 gene',\n",
       " 'MGI:1920650': 'RIKEN cDNA 1700064J06 gene',\n",
       " 'MGI:1920707': 'RIKEN cDNA 1700064M15 gene',\n",
       " 'MGI:1925708': 'RIKEN cDNA 1700064N11 gene',\n",
       " 'MGI:1920660': 'RIKEN cDNA 1700065D16 gene',\n",
       " 'MGI:1925713': 'RIKEN cDNA 1700065F09 gene',\n",
       " 'MGI:1925712': 'RIKEN cDNA 1700065I16 gene',\n",
       " 'MGI:1920659': 'RIKEN cDNA 1700065J11 gene',\n",
       " 'MGI:1920705': 'RIKEN cDNA 1700065J18 gene',\n",
       " 'MGI:1920658': 'RIKEN cDNA 1700065L07 gene',\n",
       " 'MGI:1923859': 'RIKEN cDNA 1700065O20 gene',\n",
       " 'MGI:1920676': 'RIKEN cDNA 1700066B17 gene',\n",
       " 'MGI:1925904': 'RIKEN cDNA 1700066C05 gene',\n",
       " 'MGI:1915175': 'RIKEN cDNA 1700066D14 gene',\n",
       " 'MGI:1914589': 'RIKEN cDNA 1700066J03 gene',\n",
       " 'MGI:1920717': 'RIKEN cDNA 1700066M21 gene',\n",
       " 'MGI:1920721': 'RIKEN cDNA 1700066N21 gene',\n",
       " 'MGI:1920651': 'RIKEN cDNA 1700066O22 gene',\n",
       " 'MGI:1920656': 'RIKEN cDNA 1700067A10 gene',\n",
       " 'MGI:1920725': 'RIKEN cDNA 1700067G17 gene',\n",
       " 'MGI:1920703': 'RIKEN cDNA 1700067K01 gene',\n",
       " 'MGI:1915474': 'RIKEN cDNA 1700067P10 gene',\n",
       " 'MGI:1923863': 'RIKEN cDNA 1700069B07 gene',\n",
       " 'MGI:1920726': 'RIKEN cDNA 1700069L16 gene',\n",
       " 'MGI:1920781': 'RIKEN cDNA 1700069P05 gene',\n",
       " 'MGI:1923861': 'RIKEN cDNA 1700071A11 gene',\n",
       " 'MGI:1920735': 'RIKEN cDNA 1700071G01 gene',\n",
       " 'MGI:3588264': 'RIKEN cDNA 1700071K01 gene',\n",
       " 'MGI:1920754': 'RIKEN cDNA 1700071M16 gene',\n",
       " 'MGI:1920779': 'RIKEN cDNA 1700072B07 gene',\n",
       " 'MGI:1921528': 'RIKEN cDNA 1700072F12 gene',\n",
       " 'MGI:1920746': 'RIKEN cDNA 1700072G22 gene',\n",
       " 'MGI:1920743': 'RIKEN cDNA 1700072H12 gene',\n",
       " 'MGI:1925879': 'RIKEN cDNA 1700072I22 gene',\n",
       " 'MGI:1920755': 'RIKEN cDNA 1700072O05 gene',\n",
       " 'MGI:1920734': 'RIKEN cDNA 1700073E17 gene',\n",
       " 'MGI:1920733': 'RIKEN cDNA 1700074A11 gene',\n",
       " 'MGI:1920732': 'RIKEN cDNA 1700074A21 gene',\n",
       " 'MGI:1925878': 'RIKEN cDNA 1700074G04 gene',\n",
       " 'MGI:1920730': 'RIKEN cDNA 1700074H08 gene',\n",
       " 'MGI:1920780': 'RIKEN cDNA 1700074I03 gene',\n",
       " 'MGI:1920767': 'RIKEN cDNA 1700074L02 gene',\n",
       " 'MGI:1925882': 'RIKEN cDNA 1700080C20 gene',\n",
       " 'MGI:1921535': 'RIKEN cDNA 1700080F18 gene',\n",
       " 'MGI:1920738': 'RIKEN cDNA 1700080G11 gene',\n",
       " 'MGI:1920783': 'RIKEN cDNA 1700080G18 gene',\n",
       " 'MGI:1920739': 'RIKEN cDNA 1700080N15 gene',\n",
       " 'MGI:1920750': 'RIKEN cDNA 1700081B01 gene',\n",
       " 'MGI:1920727': 'RIKEN cDNA 1700081H04 gene',\n",
       " 'MGI:1920778': 'RIKEN cDNA 1700081N11 gene',\n",
       " 'MGI:1920729': 'RIKEN cDNA 1700082C01 gene',\n",
       " 'MGI:1915477': 'RIKEN cDNA 1700082C02 gene',\n",
       " 'MGI:1920784': 'RIKEN cDNA 1700082M22 gene',\n",
       " 'MGI:1920760': 'RIKEN cDNA 1700082O11 gene',\n",
       " 'MGI:1914598': 'RIKEN cDNA 1700083H02 gene',\n",
       " 'MGI:1923868': 'RIKEN cDNA 1700084C06 gene',\n",
       " 'MGI:1923970': 'RIKEN cDNA 1700084D21 gene',\n",
       " 'MGI:1914600': 'RIKEN cDNA 1700084E18 gene',\n",
       " 'MGI:1921530': 'RIKEN cDNA 1700084F23 gene',\n",
       " 'MGI:1920736': 'RIKEN cDNA 1700084J12 gene',\n",
       " 'MGI:1920770': 'RIKEN cDNA 1700084K02 gene',\n",
       " 'MGI:1920737': 'RIKEN cDNA 1700084M14 gene',\n",
       " 'MGI:1920772': 'RIKEN cDNA 1700084M22 gene',\n",
       " 'MGI:1920748': 'RIKEN cDNA 1700084P21 gene',\n",
       " 'MGI:1923971': 'RIKEN cDNA 1700085B13 gene',\n",
       " 'MGI:1921533': 'RIKEN cDNA 1700085C21 gene',\n",
       " 'MGI:1920741': 'RIKEN cDNA 1700085D07 gene',\n",
       " 'MGI:1920762': 'RIKEN cDNA 1700085D22 gene',\n",
       " 'MGI:1920742': 'RIKEN cDNA 1700085G17 gene',\n",
       " 'MGI:1921532': 'RIKEN cDNA 1700086D15 gene',\n",
       " 'MGI:1920766': 'RIKEN cDNA 1700086O06 gene',\n",
       " 'MGI:1914599': 'RIKEN cDNA 1700086P04 gene',\n",
       " 'MGI:1923869': 'RIKEN cDNA 1700087I21 gene',\n",
       " 'MGI:1925717': 'RIKEN cDNA 1700087M22 gene',\n",
       " 'MGI:1920774': 'RIKEN cDNA 1700088E04 gene',\n",
       " 'MGI:1925880': 'RIKEN cDNA 1700089I07 gene',\n",
       " 'MGI:1920803': 'RIKEN cDNA 1700091H14 gene',\n",
       " 'MGI:1920802': 'RIKEN cDNA 1700091J24 gene',\n",
       " 'MGI:1914607': 'RIKEN cDNA 1700092C02 gene',\n",
       " 'MGI:1920804': 'RIKEN cDNA 1700092C10 gene',\n",
       " 'MGI:1921545': 'RIKEN cDNA 1700092C17 gene',\n",
       " 'MGI:1921541': 'RIKEN cDNA 1700092E16 gene',\n",
       " 'MGI:1920826': 'RIKEN cDNA 1700092E19 gene',\n",
       " 'MGI:1920786': 'RIKEN cDNA 1700092K14 gene',\n",
       " 'MGI:1920794': 'RIKEN cDNA 1700093A15 gene',\n",
       " 'MGI:1923873': 'RIKEN cDNA 1700093C20 gene',\n",
       " 'MGI:1914602': 'RIKEN cDNA 1700093E11 gene',\n",
       " 'MGI:1921546': 'RIKEN cDNA 1700093J21 gene',\n",
       " 'MGI:1914608': 'RIKEN cDNA 1700093K21 gene',\n",
       " 'MGI:1923882': 'RIKEN cDNA 1700093P08 gene',\n",
       " 'MGI:1920798': 'RIKEN cDNA 1700094I16 gene',\n",
       " 'MGI:1914773': 'RIKEN cDNA 1700094J05 gene',\n",
       " 'MGI:1920829': 'RIKEN cDNA 1700094M23 gene',\n",
       " 'MGI:1921554': 'RIKEN cDNA 1700094M24 gene',\n",
       " 'MGI:1924245': 'RIKEN cDNA 1700095A13 gene',\n",
       " 'MGI:1914774': 'RIKEN cDNA 1700095A21 gene',\n",
       " 'MGI:1914610': 'RIKEN cDNA 1700095B10 gene',\n",
       " 'MGI:1921543': 'RIKEN cDNA 1700095J03 gene',\n",
       " 'MGI:1920824': 'RIKEN cDNA 1700095J07 gene',\n",
       " 'MGI:1920813': 'RIKEN cDNA 1700095J12 gene',\n",
       " 'MGI:1915478': 'RIKEN cDNA 1700095K22 gene',\n",
       " 'MGI:1925883': 'RIKEN cDNA 1700096C16 gene',\n",
       " 'MGI:1915177': 'RIKEN cDNA 1700096J18 gene',\n",
       " 'MGI:1920821': 'RIKEN cDNA 1700096K18 gene',\n",
       " 'MGI:1920807': 'RIKEN cDNA 1700096P03 gene',\n",
       " 'MGI:1920811': 'RIKEN cDNA 1700097M23 gene',\n",
       " 'MGI:1914772': 'RIKEN cDNA 1700097N02 gene',\n",
       " 'MGI:1923874': 'RIKEN cDNA 1700099I09 gene',\n",
       " 'MGI:1920812': 'RIKEN cDNA 1700100I10 gene',\n",
       " 'MGI:1920834': 'RIKEN cDNA 1700100L14 gene',\n",
       " 'MGI:1920790': 'RIKEN cDNA 1700100M05 gene',\n",
       " 'MGI:1920836': 'RIKEN cDNA 1700101C01 gene',\n",
       " 'MGI:1915180': 'RIKEN cDNA 1700101I11 gene',\n",
       " 'MGI:1920839': 'RIKEN cDNA 1700101I19 gene',\n",
       " 'MGI:1920801': 'RIKEN cDNA 1700101O05 gene',\n",
       " 'MGI:1920825': 'RIKEN cDNA 1700101O22 gene',\n",
       " 'MGI:1920827': 'RIKEN cDNA 1700102F20 gene',\n",
       " 'MGI:1915480': 'RIKEN cDNA 1700102H20 gene',\n",
       " 'MGI:1921540': 'RIKEN cDNA 1700102J08 gene',\n",
       " 'MGI:1920791': 'RIKEN cDNA 1700102N10 gene',\n",
       " 'MGI:2148248': 'RIKEN cDNA 1700102P08 gene',\n",
       " 'MGI:1925886': 'RIKEN cDNA 1700104B16 gene',\n",
       " 'MGI:1920842': 'RIKEN cDNA 1700104L18 gene',\n",
       " 'MGI:1920788': 'RIKEN cDNA 1700105F02 gene',\n",
       " 'MGI:1920800': 'RIKEN cDNA 1700105G05 gene',\n",
       " 'MGI:1915479': 'RIKEN cDNA 1700105P06 gene',\n",
       " 'MGI:1923871': 'RIKEN cDNA 1700106J12 gene',\n",
       " 'MGI:1920809': 'RIKEN cDNA 1700106O07 gene',\n",
       " 'MGI:1920814': 'RIKEN cDNA 1700108F19 gene',\n",
       " 'MGI:1921548': 'RIKEN cDNA 1700108I11 gene',\n",
       " 'MGI:1915495': 'RIKEN cDNA 1700108J01 gene',\n",
       " 'MGI:1920810': 'RIKEN cDNA 1700108N06 gene',\n",
       " 'MGI:1921551': 'RIKEN cDNA 1700108N11 gene',\n",
       " 'MGI:1914605': 'RIKEN cDNA 1700109G14 gene',\n",
       " 'MGI:1925721': 'RIKEN cDNA 1700109G15 gene',\n",
       " 'MGI:1924286': 'RIKEN cDNA 1700109H08 gene',\n",
       " 'MGI:1920841': 'RIKEN cDNA 1700109I08 gene',\n",
       " 'MGI:1921553': 'RIKEN cDNA 1700109K24 gene',\n",
       " 'MGI:1925749': 'RIKEN cDNA 1700109O10 gene',\n",
       " 'MGI:1923884': 'RIKEN cDNA 1700110C19 gene',\n",
       " 'MGI:1923885': 'RIKEN cDNA 1700110C23 gene',\n",
       " 'MGI:1920799': 'RIKEN cDNA 1700110I01 gene',\n",
       " 'MGI:1924251': 'RIKEN cDNA 1700110I07 gene',\n",
       " 'MGI:1920808': 'RIKEN cDNA 1700110K17 gene',\n",
       " 'MGI:1920835': 'RIKEN cDNA 1700111E14 gene',\n",
       " 'MGI:1921555': 'RIKEN cDNA 1700111N16 gene',\n",
       " 'MGI:1920861': 'RIKEN cDNA 1700112D23 gene',\n",
       " 'MGI:1925720': 'RIKEN cDNA 1700112H04 gene',\n",
       " 'MGI:1920840': 'RIKEN cDNA 1700112H15 gene',\n",
       " 'MGI:3696786': 'RIKEN cDNA 1700112J05 gene',\n",
       " 'MGI:1923878': 'RIKEN cDNA 1700112J16 gene',\n",
       " 'MGI:1920846': 'RIKEN cDNA 1700112K13 gene',\n",
       " 'MGI:1920844': 'RIKEN cDNA 1700112L15 gene',\n",
       " 'MGI:1920833': 'RIKEN cDNA 1700112M02 gene',\n",
       " 'MGI:1920843': 'RIKEN cDNA 1700112N08 gene',\n",
       " 'MGI:1923892': 'RIKEN cDNA 1700113A16 gene',\n",
       " 'MGI:1925888': 'RIKEN cDNA 1700113B09 gene',\n",
       " 'MGI:1920890': 'RIKEN cDNA 1700113B19 gene',\n",
       " 'MGI:1923890': 'RIKEN cDNA 1700113H08 gene',\n",
       " 'MGI:1920847': 'RIKEN cDNA 1700113H21 gene',\n",
       " 'MGI:1921561': 'RIKEN cDNA 1700113P08 gene',\n",
       " 'MGI:1925889': 'RIKEN cDNA 1700116B05 gene',\n",
       " 'MGI:1920872': 'RIKEN cDNA 1700116H05 gene',\n",
       " 'MGI:1923899': 'RIKEN cDNA 1700119H24 gene',\n",
       " 'MGI:1925729': 'RIKEN cDNA 1700119I11 gene',\n",
       " 'MGI:3604190': 'RIKEN cDNA 1700120B22 gene',\n",
       " 'MGI:1920850': 'RIKEN cDNA 1700120C14 gene',\n",
       " 'MGI:1923912': 'RIKEN cDNA 1700120C18 gene',\n",
       " 'MGI:1923916': 'RIKEN cDNA 1700120E02 gene',\n",
       " 'MGI:1920856': 'RIKEN cDNA 1700120E14 gene',\n",
       " 'MGI:1921564': 'RIKEN cDNA 1700120G07 gene',\n",
       " 'MGI:1920849': 'RIKEN cDNA 1700120G11 gene',\n",
       " 'MGI:1915482': 'RIKEN cDNA 1700120K04 gene',\n",
       " 'MGI:1923906': 'RIKEN cDNA 1700120O09 gene',\n",
       " 'MGI:1925891': 'RIKEN cDNA 1700121C08 gene',\n",
       " 'MGI:1914614': 'RIKEN cDNA 1700121C10 gene',\n",
       " 'MGI:1923976': 'RIKEN cDNA 1700121F22 gene',\n",
       " 'MGI:1920883': 'RIKEN cDNA 1700121I08 gene',\n",
       " 'MGI:1923910': 'RIKEN cDNA 1700121L03 gene',\n",
       " 'MGI:1923893': 'RIKEN cDNA 1700121L16 gene',\n",
       " 'MGI:1925727': 'RIKEN cDNA 1700121M21 gene',\n",
       " 'MGI:1923889': 'RIKEN cDNA 1700121N20 gene',\n",
       " 'MGI:1925890': 'RIKEN cDNA 1700122C19 gene',\n",
       " 'MGI:1923911': 'RIKEN cDNA 1700122D07 gene',\n",
       " 'MGI:1920869': 'RIKEN cDNA 1700122E12 gene',\n",
       " 'MGI:1920867': 'RIKEN cDNA 1700122H20 gene',\n",
       " 'MGI:1923901': 'RIKEN cDNA 1700122O11 gene',\n",
       " 'MGI:1924247': 'RIKEN cDNA 1700123C12 gene',\n",
       " 'MGI:1920879': 'RIKEN cDNA 1700123E06 gene',\n",
       " 'MGI:1925731': 'RIKEN cDNA 1700123J03 gene',\n",
       " 'MGI:1923978': 'RIKEN cDNA 1700123J17 gene',\n",
       " 'MGI:1923908': 'RIKEN cDNA 1700123K08 gene',\n",
       " 'MGI:1923915': 'RIKEN cDNA 1700123M08 gene',\n",
       " 'MGI:1923891': 'RIKEN cDNA 1700123N01 gene',\n",
       " 'MGI:1920874': 'RIKEN cDNA 1700123O12 gene',\n",
       " 'MGI:1920893': 'RIKEN cDNA 1700123O20 gene',\n",
       " 'MGI:1915488': 'RIKEN cDNA 1700123O21 gene',\n",
       " 'MGI:1915184': 'RIKEN cDNA 1700124L16 gene',\n",
       " 'MGI:1920855': 'RIKEN cDNA 1700124M09 gene',\n",
       " 'MGI:1920886': 'RIKEN cDNA 1700125C05 gene',\n",
       " 'MGI:1920878': 'RIKEN cDNA 1700125G02 gene',\n",
       " 'MGI:1923907': 'RIKEN cDNA 1700125G22 gene',\n",
       " 'MGI:1920859': 'RIKEN cDNA 1700125H03 gene',\n",
       " 'MGI:1920891': 'RIKEN cDNA 1700125M20 gene',\n",
       " 'MGI:1920852': 'RIKEN cDNA 1700126A01 gene',\n",
       " 'MGI:1923898': 'RIKEN cDNA 1700126G02 gene',\n",
       " 'MGI:1920889': 'RIKEN cDNA 1700126H18 gene',\n",
       " 'MGI:1925892': 'RIKEN cDNA 1700127F24 gene',\n",
       " 'MGI:1920862': 'RIKEN cDNA 1700127G19 gene',\n",
       " 'MGI:1920881': 'RIKEN cDNA 1700128A07 gene',\n",
       " 'MGI:1920868': 'RIKEN cDNA 1700128E19 gene',\n",
       " 'MGI:1923919': 'RIKEN cDNA 1700128F08 gene',\n",
       " 'MGI:1920865': 'RIKEN cDNA 1700128I11 gene',\n",
       " 'MGI:1915182': 'RIKEN cDNA 1700129C05 gene',\n",
       " 'MGI:1920877': 'RIKEN cDNA 1700129L04 gene',\n",
       " 'MGI:1925893': 'RIKEN cDNA 1700129O19 gene',\n",
       " 'MGI:1860921': 'DNA segment, 170n17Sp6',\n",
       " 'MGI:1860922': 'DNA segment, 170n17T7',\n",
       " 'MGI:1931097': 'DNA segment, 171M12-Sp6',\n",
       " 'MGI:1931093': 'DNA segment, 171M12-T7',\n",
       " 'MGI:1930518': 'DNA segment, 17-1-T7',\n",
       " 'MGI:1860923': 'DNA segment, 174g24Sp6',\n",
       " 'MGI:2137545': 'DNA segment, 174L15-R',\n",
       " 'MGI:1913007': 'DNA segment, 1-76C9-S',\n",
       " 'MGI:2652172': 'DNA segment, 176E13-T7',\n",
       " 'MGI:1860924': 'DNA segment, 177i23Sp6',\n",
       " 'MGI:1889686': 'DNA segment, 179k6Sp6, Birc1a - Birc1c intergenic region',\n",
       " 'MGI:1337068': 'DNA segment, 17F4L',\n",
       " 'MGI:1336990': 'DNA segment, 17F4R',\n",
       " 'MGI:1916280': 'RIKEN cDNA 1810006J02 gene',\n",
       " 'MGI:1916304': 'RIKEN cDNA 1810007C17 gene',\n",
       " 'MGI:1916305': 'RIKEN cDNA 1810007D17 gene',\n",
       " 'MGI:1916312': 'RIKEN cDNA 1810008B01 gene',\n",
       " 'MGI:1920875': 'RIKEN cDNA 1810008I18 gene',\n",
       " 'MGI:1916346': 'RIKEN cDNA 1810008N23 gene',\n",
       " 'MGI:1913526': 'RIKEN cDNA 1810009A15 gene',\n",
       " 'MGI:1920876': 'RIKEN cDNA 1810009J06 gene',\n",
       " 'MGI:1916349': 'RIKEN cDNA 1810009N02 gene',\n",
       " 'MGI:1916316': 'RIKEN cDNA 1810010H24 gene',\n",
       " 'MGI:1921419': 'RIKEN cDNA 1810010K12 gene',\n",
       " 'MGI:1916345': 'RIKEN cDNA 1810012K08 gene',\n",
       " 'MGI:1916358': 'RIKEN cDNA 1810012K16 gene',\n",
       " 'MGI:1916300': 'RIKEN cDNA 1810013A23 gene',\n",
       " 'MGI:1916292': 'RIKEN cDNA 1810013D15 gene',\n",
       " 'MGI:1913513': 'RIKEN cDNA 1810014B01 gene',\n",
       " 'MGI:1916313': 'RIKEN cDNA 1810014P07 gene',\n",
       " 'MGI:1916353': 'RIKEN cDNA 1810015A16 gene',\n",
       " 'MGI:1916352': 'RIKEN cDNA 1810015C11 gene',\n",
       " 'MGI:1916294': 'RIKEN cDNA 1810017P11 gene',\n",
       " 'MGI:1921421': 'RIKEN cDNA 1810018F18 gene',\n",
       " 'MGI:1917021': 'RIKEN cDNA 1810019D21 gene',\n",
       " 'MGI:1916336': 'RIKEN cDNA 1810019N24 gene',\n",
       " 'MGI:1916370': 'RIKEN cDNA 1810021B22 gene',\n",
       " 'MGI:1916369': 'RIKEN cDNA 1810021M19 gene',\n",
       " 'MGI:1925560': 'RIKEN cDNA 1810024B03 gene',\n",
       " 'MGI:1916427': 'RIKEN cDNA 1810027L02 gene',\n",
       " 'MGI:1913530': 'RIKEN cDNA 1810028F09 gene',\n",
       " 'MGI:1916405': 'RIKEN cDNA 1810030O07 gene',\n",
       " 'MGI:1913753': 'RIKEN cDNA 1810034E14 gene',\n",
       " 'MGI:1925578': 'RIKEN cDNA 1810035K13 gene',\n",
       " 'MGI:3697821': 'RIKEN cDNA 1810036N14 gene',\n",
       " 'MGI:1916423': 'RIKEN cDNA 1810037O21 gene',\n",
       " 'MGI:1925562': 'RIKEN cDNA 1810041H14 gene',\n",
       " 'MGI:1917048': 'RIKEN cDNA 1810044D09 gene',\n",
       " 'MGI:1924010': 'RIKEN cDNA 1810044K17 gene',\n",
       " 'MGI:1922851': 'RIKEN cDNA 1810049I09 gene',\n",
       " 'MGI:3704453': 'RIKEN cDNA 1810049J17 gene',\n",
       " 'MGI:1917107': 'RIKEN cDNA 1810053B23 gene',\n",
       " 'MGI:1919306': 'RIKEN cDNA 1810055G02 gene',\n",
       " 'MGI:1925738': 'RIKEN cDNA 1810056I18 gene',\n",
       " 'MGI:1919552': 'RIKEN cDNA 1810057I12 gene',\n",
       " 'MGI:1923763': 'RIKEN cDNA 1810058N15 gene',\n",
       " 'MGI:1917068': 'RIKEN cDNA 1810059C17 gene',\n",
       " 'MGI:1917070': 'RIKEN cDNA 1810059H22 gene',\n",
       " 'MGI:1919532': 'RIKEN cDNA 1810062G17 gene',\n",
       " 'MGI:1922852': 'RIKEN cDNA 1810062O18 gene',\n",
       " 'MGI:1917104': 'RIKEN cDNA 1810063I02 gene',\n",
       " 'MGI:1917112': 'RIKEN cDNA 1810064F22 gene',\n",
       " 'MGI:1917114': 'RIKEN cDNA 1810065E05 gene',\n",
       " 'MGI:1924185': 'RIKEN cDNA 1810065K23 gene',\n",
       " 'MGI:1917628': 'RIKEN cDNA 1810073G21 gene',\n",
       " 'MGI:1925748': 'RIKEN cDNA 1810073J14 gene',\n",
       " 'MGI:1919535': 'RIKEN cDNA 1810073O08 gene',\n",
       " 'MGI:1928721': 'DNA segment, 181f4Sp6',\n",
       " 'MGI:1928722': 'DNA segment, 181f4T7',\n",
       " 'MGI:1860925': 'DNA segment, 184a16Sp6',\n",
       " 'MGI:1931087': 'DNA segment, 18B7-T7',\n",
       " 'MGI:2137535': 'DNA segment, 194O7-F',\n",
       " 'MGI:1342271': 'DNA segment, 196D11.T7 (Roswell Park)',\n",
       " 'MGI:3057271': 'DNA segment, 199b22-Sp6',\n",
       " 'MGI:3057272': 'DNA segment, 199b22-T7',\n",
       " 'MGI:1337127': 'DNA segment, 1S',\n",
       " 'MGI:1337001': 'DNA segment, 1T',\n",
       " 'MGI:1860926': 'DNA segment, 200H7R',\n",
       " 'MGI:1923766': 'RIKEN cDNA 2010001A14 gene',\n",
       " 'MGI:1917079': 'RIKEN cDNA 2010001K21 gene',\n",
       " 'MGI:1917063': 'RIKEN cDNA 2010001M07 gene',\n",
       " 'MGI:1917091': 'RIKEN cDNA 2010002M09 gene',\n",
       " 'MGI:1923975': 'RIKEN cDNA 2010003D24 gene',\n",
       " 'MGI:1924019': 'RIKEN cDNA 2010003K10 gene',\n",
       " 'MGI:1917111': 'RIKEN cDNA 2010003K11 gene',\n",
       " 'MGI:1923761': 'RIKEN cDNA 2010004M13 gene',\n",
       " 'MGI:1917101': 'RIKEN cDNA 2010007E15 gene',\n",
       " 'MGI:1917099': 'RIKEN cDNA 2010007H06 gene',\n",
       " 'MGI:1917109': 'RIKEN cDNA 2010008C14 gene',\n",
       " 'MGI:1919590': 'RIKEN cDNA 2010009K17 gene',\n",
       " 'MGI:1913686': 'RIKEN cDNA 2010010A06 gene',\n",
       " 'MGI:1922087': 'RIKEN cDNA 2010010I01 gene',\n",
       " 'MGI:1919329': 'RIKEN cDNA 2010013B24 gene',\n",
       " 'MGI:1916444': 'RIKEN cDNA 2010015B12 gene',\n",
       " 'MGI:1919337': 'RIKEN cDNA 2010015M23 gene',\n",
       " 'MGI:1916456': 'RIKEN cDNA 2010016I18 gene',\n",
       " 'MGI:1917137': 'RIKEN cDNA 2010100M03 gene',\n",
       " 'MGI:1917139': 'RIKEN cDNA 2010103J01 gene',\n",
       " 'MGI:1917501': 'RIKEN cDNA 2010105D22 gene',\n",
       " 'MGI:1919354': 'RIKEN cDNA 2010106C02 gene',\n",
       " 'MGI:1914965': 'RIKEN cDNA 2010106E10 gene',\n",
       " 'MGI:1923783': 'RIKEN cDNA 2010106G04 gene',\n",
       " 'MGI:1917506': 'RIKEN cDNA 2010107C10 gene',\n",
       " 'MGI:1922860': 'RIKEN cDNA 2010109A12 gene',\n",
       " 'MGI:1917517': 'RIKEN cDNA 2010109K09 gene',\n",
       " 'MGI:1913800': 'RIKEN cDNA 2010109N18 gene',\n",
       " 'MGI:1917515': 'RIKEN cDNA 2010109P13 gene',\n",
       " 'MGI:1923941': 'RIKEN cDNA 2010110E17 gene',\n",
       " 'MGI:1917148': 'RIKEN cDNA 2010110G14 gene',\n",
       " 'MGI:1922951': 'RIKEN cDNA 2010110K18 gene',\n",
       " 'MGI:1913797': 'RIKEN cDNA 2010203P06 gene',\n",
       " 'MGI:1922859': 'RIKEN cDNA 2010204K13 gene',\n",
       " 'MGI:1917238': 'RIKEN cDNA 2010205J10 gene',\n",
       " 'MGI:1917134': 'RIKEN cDNA 2010300F17 gene',\n",
       " 'MGI:1919372': 'RIKEN cDNA 2010308F09 gene',\n",
       " 'MGI:1917268': 'RIKEN cDNA 2010309G21 gene',\n",
       " 'MGI:1919381': 'RIKEN cDNA 2010310C07 gene',\n",
       " 'MGI:1919320': 'RIKEN cDNA 2010313P22 gene',\n",
       " 'MGI:1919321': 'RIKEN cDNA 2010315B03 gene',\n",
       " 'MGI:1919342': 'RIKEN cDNA 2010320H13 gene',\n",
       " ...}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_mouse_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "161400b5-b329-4af9-9670-43c9338c8ca6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGI Marker Accession ID</th>\n",
       "      <th>Aliases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGI:1341858</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MGI:1341869</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:1337005</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MGI:1918911</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MGI:1923503</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>655272</th>\n",
       "      <td>MGI:2446208</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>655273</th>\n",
       "      <td>MGI:2685277</td>\n",
       "      <td>LOC242610|1110046I03Rik|D4Mgi23|2810482G21Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>655274</th>\n",
       "      <td>MGI:103072</td>\n",
       "      <td>R75157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>655275</th>\n",
       "      <td>MGI:2444286</td>\n",
       "      <td>8430405D05Rik|C130099L13Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>655276</th>\n",
       "      <td>MGI:1920453</td>\n",
       "      <td>3110065C23Rik|6430567E01Rik</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>655277 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       MGI Marker Accession ID                                        Aliases\n",
       "0                  MGI:1341858                                            NaN\n",
       "1                  MGI:1341869                                            NaN\n",
       "2                  MGI:1337005                                            NaN\n",
       "3                  MGI:1918911                                            NaN\n",
       "4                  MGI:1923503                                            NaN\n",
       "...                        ...                                            ...\n",
       "655272             MGI:2446208                                            NaN\n",
       "655273             MGI:2685277  LOC242610|1110046I03Rik|D4Mgi23|2810482G21Rik\n",
       "655274              MGI:103072                                         R75157\n",
       "655275             MGI:2444286                    8430405D05Rik|C130099L13Rik\n",
       "655276             MGI:1920453                    3110065C23Rik|6430567E01Rik\n",
       "\n",
       "[655277 rows x 2 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "from datetime import datetime\n",
    "import os\n",
    "\n",
    "# Create data folder if it doesn't exist\n",
    "os.makedirs(\"data\", exist_ok=True)\n",
    "\n",
    "# Today's date for versioning\n",
    "today = datetime.now().strftime(\"%Y%m%d\")\n",
    "\n",
    "# URLs and filenames\n",
    "homology_url = \"https://www.informatics.jax.org/downloads/reports/HGNC_AllianceHomology.rpt\"\n",
    "mrk_url = \"https://www.informatics.jax.org/downloads/reports/MRK_Sequence.rpt\"\n",
    "mrk_list_url = \"https://www.informatics.jax.org/downloads/reports/MRK_List2.rpt\"\n",
    "\n",
    "homology_file = f\"data/HGNC_AllianceHomology_{today}.tsv\"\n",
    "mrk_file = f\"data/MRK_Sequence_{today}.tsv\"\n",
    "mrk_list_file = f\"data/MRK_List2_{today}.tsv\"\n",
    "merged_file = f\"data/MRK_Merged_{today}_MGI_DB.tsv\"\n",
    "\n",
    "# Download HGNC_AllianceHomology.rpt\n",
    "response = requests.get(homology_url)\n",
    "response.raise_for_status()\n",
    "with open(homology_file, \"wb\") as f:\n",
    "    f.write(response.content)\n",
    "\n",
    "# Download MRK_Sequence.rpt\n",
    "response = requests.get(mrk_url)\n",
    "response.raise_for_status()\n",
    "with open(mrk_file, \"wb\") as f:\n",
    "    f.write(response.content)\n",
    "\n",
    "# Download MRK_List2.rpt\n",
    "response = requests.get(mrk_list_url)\n",
    "response.raise_for_status()\n",
    "with open(mrk_list_file, \"wb\") as f:\n",
    "    f.write(response.content)\n",
    "\n",
    "# Load files\n",
    "df_mrk = pd.read_csv(mrk_file, sep=\"\\t\", dtype=str)\n",
    "df_homology = pd.read_csv(homology_file, sep=\"\\t\", dtype=str, index_col = False)\n",
    "\n",
    "df_mrk_list = pd.read_csv(mrk_list_file, sep=\"\\t\", dtype=str)\n",
    "df_mrk_list= df_mrk_list[[\"MGI Accession ID\",\"Marker Synonyms (pipe-separated)\"]].rename(columns={\"MGI Accession ID\": \"MGI Marker Accession ID\",\n",
    "                                                                                                  \"Marker Synonyms (pipe-separated)\": \"Aliases\"})\n",
    "df_homology = df_homology.rename(columns={\"MGI Accession ID\": \"MGI Marker Accession ID\"})\n",
    "# Keep only relevant columns from HGNC_AllianceHomology\n",
    "columns_to_keep = [\n",
    "    \"MGI Marker Accession ID\",\n",
    "    \"EntrezGene ID\",\n",
    "    \"CCDS IDs\",\n",
    "    \"HGNC ID\"\n",
    "]\n",
    "df_homology_subset = df_homology[columns_to_keep]\n",
    "# Merge: MRK + homology\n",
    "df_merged = df_mrk.merge(df_homology_subset, how=\"left\", on=\"MGI Marker Accession ID\")\n",
    "\n",
    "# Merge with synonyms\n",
    "df_merged = df_merged.merge(df_mrk_list, how=\"left\", on=\"MGI Marker Accession ID\")\n",
    "\n",
    "# Save result\n",
    "df_merged.to_csv(merged_file, sep=\"\\t\", index=False)\n",
    "print(f\"Merged file saved to: {merged_file}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "ee57451f-a85b-49f3-ba2c-f752dd697545",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "087bf195-eb42-42eb-bdcd-0ff8bd673910",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGI Marker Accession ID</th>\n",
       "      <th>EntrezGene ID</th>\n",
       "      <th>CCDS IDs</th>\n",
       "      <th>HGNC ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGI:87853</td>\n",
       "      <td>50518</td>\n",
       "      <td>CCDS16941.1</td>\n",
       "      <td>HGNC:745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MGI:87854</td>\n",
       "      <td>11287</td>\n",
       "      <td>CCDS39650.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:87855</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MGI:87859</td>\n",
       "      <td>11350</td>\n",
       "      <td>CCDS15901.1|CCDS50563.1</td>\n",
       "      <td>HGNC:76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MGI:87860</td>\n",
       "      <td>11352</td>\n",
       "      <td>CCDS15393.1|CCDS48404.1</td>\n",
       "      <td>HGNC:77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90793</th>\n",
       "      <td>MGI:7834358</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90794</th>\n",
       "      <td>MGI:7834360</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90795</th>\n",
       "      <td>MGI:7834362</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90796</th>\n",
       "      <td>MGI:7855873</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90797</th>\n",
       "      <td>MGI:7858886</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>90798 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      MGI Marker Accession ID EntrezGene ID                 CCDS IDs   HGNC ID\n",
       "0                   MGI:87853         50518              CCDS16941.1  HGNC:745\n",
       "1                   MGI:87854         11287              CCDS39650.1       NaN\n",
       "2                   MGI:87855           NaN                      NaN       NaN\n",
       "3                   MGI:87859         11350  CCDS15901.1|CCDS50563.1   HGNC:76\n",
       "4                   MGI:87860         11352  CCDS15393.1|CCDS48404.1   HGNC:77\n",
       "...                       ...           ...                      ...       ...\n",
       "90793             MGI:7834358           NaN                      NaN       NaN\n",
       "90794             MGI:7834360           NaN                      NaN       NaN\n",
       "90795             MGI:7834362           NaN                      NaN       NaN\n",
       "90796             MGI:7855873           NaN                      NaN       NaN\n",
       "90797             MGI:7858886           NaN                      NaN       NaN\n",
       "\n",
       "[90798 rows x 4 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "0650e348-1e97-4c26-a479-60a581046d02",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merged file saved to: data/MRK_Merged_20250730_MGI_DB.tsv\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "59baf801-222a-40f3-927d-38e413fb963e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "124340    ENSMUST00000048096|ENSMUST00000107365|ENSMUST0...\n",
       "Name: Ensembl transcript IDs, dtype: object"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merged['Ensembl transcript IDs'][df_merged['Marker Symbol'] == \"Tlr4\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "02f5fa04-d1ed-4112-b4ad-ca2e6ebf2aa6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['MGI Marker Accession ID', 'Marker Symbol', 'Status', 'Marker Type',\n",
       "       'Marker Name', 'cM position', 'Chromosome', 'Genome Coordinate Start',\n",
       "       'Genome Coordinate End', 'Strand', 'GenBank IDs',\n",
       "       'RefSeq transcript IDs', 'Ensembl transcript IDs', 'UniProt IDs',\n",
       "       'TrEMBL IDs', 'Ensembl protein IDs', 'RefSeq protein IDs',\n",
       "       'UniGene IDs', 'Feature Type', 'EntrezGene ID', 'CCDS IDs', 'HGNC ID',\n",
       "       'Aliases'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merged.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a0859b6b-c2ed-4dd9-9957-380c56928c72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGI Accession ID</th>\n",
       "      <th>Marker Symbol</th>\n",
       "      <th>Marker Name</th>\n",
       "      <th>Feature Type</th>\n",
       "      <th>EntrezGene ID</th>\n",
       "      <th>NCBI Gene chromosome</th>\n",
       "      <th>NCBI Gene start</th>\n",
       "      <th>NCBI Gene end</th>\n",
       "      <th>NCBI Gene strand</th>\n",
       "      <th>Ensembl Gene ID</th>\n",
       "      <th>Ensembl Gene chromosome</th>\n",
       "      <th>Ensembl Gene start</th>\n",
       "      <th>Ensembl Gene end</th>\n",
       "      <th>Ensembl Gene strand</th>\n",
       "      <th>CCDS IDs</th>\n",
       "      <th>HGNC ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MGI:87853</th>\n",
       "      <td>a</td>\n",
       "      <td>nonagouti</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>50518</td>\n",
       "      <td>2</td>\n",
       "      <td>154792519</td>\n",
       "      <td>154892932</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000027596</td>\n",
       "      <td>2</td>\n",
       "      <td>154633322</td>\n",
       "      <td>154892932</td>\n",
       "      <td>+</td>\n",
       "      <td>CCDS16941.1</td>\n",
       "      <td>HGNC:745</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGI:87854</th>\n",
       "      <td>Pzp</td>\n",
       "      <td>PZP, alpha-2-macroglobulin like</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>11287</td>\n",
       "      <td>6</td>\n",
       "      <td>128460530</td>\n",
       "      <td>128503683</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000030359</td>\n",
       "      <td>6</td>\n",
       "      <td>128460530</td>\n",
       "      <td>128503683</td>\n",
       "      <td>-</td>\n",
       "      <td>CCDS39650.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGI:87855</th>\n",
       "      <td>Aal</td>\n",
       "      <td>active avoidance learning</td>\n",
       "      <td>heritable phenotypic marker</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGI:87859</th>\n",
       "      <td>Abl1</td>\n",
       "      <td>c-abl oncogene 1, non-receptor tyrosine kinase</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>11350</td>\n",
       "      <td>2</td>\n",
       "      <td>31578256</td>\n",
       "      <td>31697105</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000026842</td>\n",
       "      <td>2</td>\n",
       "      <td>31578388</td>\n",
       "      <td>31694239</td>\n",
       "      <td>+</td>\n",
       "      <td>CCDS15901.1|CCDS50563.1</td>\n",
       "      <td>HGNC:76</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGI:87860</th>\n",
       "      <td>Abl2</td>\n",
       "      <td>ABL proto-oncogene 2, non-receptor tyrosine ki...</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>11352</td>\n",
       "      <td>1</td>\n",
       "      <td>156386160</td>\n",
       "      <td>156477189</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000026596</td>\n",
       "      <td>1</td>\n",
       "      <td>156386356</td>\n",
       "      <td>156477138</td>\n",
       "      <td>+</td>\n",
       "      <td>CCDS15393.1|CCDS48404.1</td>\n",
       "      <td>HGNC:77</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGI:7834358</th>\n",
       "      <td>Gm75820</td>\n",
       "      <td>predicted gene, 75820</td>\n",
       "      <td>lncRNA gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000137503</td>\n",
       "      <td>2</td>\n",
       "      <td>164339005</td>\n",
       "      <td>164345988</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGI:7834360</th>\n",
       "      <td>Gm75821</td>\n",
       "      <td>predicted gene, 75821</td>\n",
       "      <td>lncRNA gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000138928</td>\n",
       "      <td>2</td>\n",
       "      <td>109876392</td>\n",
       "      <td>109963249</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGI:7834362</th>\n",
       "      <td>Gm75822</td>\n",
       "      <td>predicted gene, 75822</td>\n",
       "      <td>lncRNA gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000140534</td>\n",
       "      <td>2</td>\n",
       "      <td>52821585</td>\n",
       "      <td>52837904</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGI:7855873</th>\n",
       "      <td>BPL001</td>\n",
       "      <td>bone phosphate low mutant 001</td>\n",
       "      <td>heritable phenotypic marker</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGI:7858886</th>\n",
       "      <td>Ngfras</td>\n",
       "      <td>nerve growth factor receptor, antisense RNA</td>\n",
       "      <td>antisense lncRNA gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>90798 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            MGI Accession ID  \\\n",
       "MGI:87853                  a   \n",
       "MGI:87854                Pzp   \n",
       "MGI:87855                Aal   \n",
       "MGI:87859               Abl1   \n",
       "MGI:87860               Abl2   \n",
       "...                      ...   \n",
       "MGI:7834358          Gm75820   \n",
       "MGI:7834360          Gm75821   \n",
       "MGI:7834362          Gm75822   \n",
       "MGI:7855873           BPL001   \n",
       "MGI:7858886           Ngfras   \n",
       "\n",
       "                                                 Marker Symbol  \\\n",
       "MGI:87853                                            nonagouti   \n",
       "MGI:87854                      PZP, alpha-2-macroglobulin like   \n",
       "MGI:87855                            active avoidance learning   \n",
       "MGI:87859       c-abl oncogene 1, non-receptor tyrosine kinase   \n",
       "MGI:87860    ABL proto-oncogene 2, non-receptor tyrosine ki...   \n",
       "...                                                        ...   \n",
       "MGI:7834358                              predicted gene, 75820   \n",
       "MGI:7834360                              predicted gene, 75821   \n",
       "MGI:7834362                              predicted gene, 75822   \n",
       "MGI:7855873                      bone phosphate low mutant 001   \n",
       "MGI:7858886        nerve growth factor receptor, antisense RNA   \n",
       "\n",
       "                             Marker Name Feature Type EntrezGene ID  \\\n",
       "MGI:87853            protein coding gene        50518             2   \n",
       "MGI:87854            protein coding gene        11287             6   \n",
       "MGI:87855    heritable phenotypic marker          NaN           NaN   \n",
       "MGI:87859            protein coding gene        11350             2   \n",
       "MGI:87860            protein coding gene        11352             1   \n",
       "...                                  ...          ...           ...   \n",
       "MGI:7834358                  lncRNA gene          NaN           NaN   \n",
       "MGI:7834360                  lncRNA gene          NaN           NaN   \n",
       "MGI:7834362                  lncRNA gene          NaN           NaN   \n",
       "MGI:7855873  heritable phenotypic marker          NaN           NaN   \n",
       "MGI:7858886        antisense lncRNA gene          NaN           NaN   \n",
       "\n",
       "            NCBI Gene chromosome NCBI Gene start NCBI Gene end  \\\n",
       "MGI:87853              154792519       154892932             +   \n",
       "MGI:87854              128460530       128503683             -   \n",
       "MGI:87855                    NaN             NaN           NaN   \n",
       "MGI:87859               31578256        31697105             +   \n",
       "MGI:87860              156386160       156477189             +   \n",
       "...                          ...             ...           ...   \n",
       "MGI:7834358                  NaN             NaN           NaN   \n",
       "MGI:7834360                  NaN             NaN           NaN   \n",
       "MGI:7834362                  NaN             NaN           NaN   \n",
       "MGI:7855873                  NaN             NaN           NaN   \n",
       "MGI:7858886                  NaN             NaN           NaN   \n",
       "\n",
       "               NCBI Gene strand Ensembl Gene ID Ensembl Gene chromosome  \\\n",
       "MGI:87853    ENSMUSG00000027596               2               154633322   \n",
       "MGI:87854    ENSMUSG00000030359               6               128460530   \n",
       "MGI:87855                   NaN             NaN                     NaN   \n",
       "MGI:87859    ENSMUSG00000026842               2                31578388   \n",
       "MGI:87860    ENSMUSG00000026596               1               156386356   \n",
       "...                         ...             ...                     ...   \n",
       "MGI:7834358  ENSMUSG00000137503               2               164339005   \n",
       "MGI:7834360  ENSMUSG00000138928               2               109876392   \n",
       "MGI:7834362  ENSMUSG00000140534               2                52821585   \n",
       "MGI:7855873                 NaN             NaN                     NaN   \n",
       "MGI:7858886                 NaN             NaN                     NaN   \n",
       "\n",
       "            Ensembl Gene start Ensembl Gene end      Ensembl Gene strand  \\\n",
       "MGI:87853            154892932                +              CCDS16941.1   \n",
       "MGI:87854            128503683                -              CCDS39650.1   \n",
       "MGI:87855                  NaN              NaN                      NaN   \n",
       "MGI:87859             31694239                +  CCDS15901.1|CCDS50563.1   \n",
       "MGI:87860            156477138                +  CCDS15393.1|CCDS48404.1   \n",
       "...                        ...              ...                      ...   \n",
       "MGI:7834358          164345988                -                      NaN   \n",
       "MGI:7834360          109963249                +                      NaN   \n",
       "MGI:7834362           52837904                +                      NaN   \n",
       "MGI:7855873                NaN              NaN                      NaN   \n",
       "MGI:7858886                NaN              NaN                      NaN   \n",
       "\n",
       "             CCDS IDs HGNC ID  \n",
       "MGI:87853    HGNC:745     NaN  \n",
       "MGI:87854         NaN     NaN  \n",
       "MGI:87855         NaN     NaN  \n",
       "MGI:87859     HGNC:76     NaN  \n",
       "MGI:87860     HGNC:77     NaN  \n",
       "...               ...     ...  \n",
       "MGI:7834358       NaN     NaN  \n",
       "MGI:7834360       NaN     NaN  \n",
       "MGI:7834362       NaN     NaN  \n",
       "MGI:7855873       NaN     NaN  \n",
       "MGI:7858886       NaN     NaN  \n",
       "\n",
       "[90798 rows x 16 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_homology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0ea7808c-13ce-4cbd-ae7c-644ffaa282bd",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df_homology_subset' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[30], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mdf_homology_subset\u001b[49m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'df_homology_subset' is not defined"
     ]
    }
   ],
   "source": [
    "df_homology_subset = df_homology[columns_to_keep]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c6adc6b4-6f75-42bf-ad5e-aee81bfa9495",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"None of [Index(['MGI Marker Accession ID', 'NCBI Gene ID',\\n       'CCDS IDs (comma delimited)'],\\n      dtype='object')] are in the [columns]\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[28], line 42\u001b[0m\n\u001b[1;32m     36\u001b[0m \u001b[38;5;66;03m# Keep only relevant columns from HGNC_AllianceHomology\u001b[39;00m\n\u001b[1;32m     37\u001b[0m columns_to_keep \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m     38\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMGI Marker Accession ID\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     39\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNCBI Gene ID\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     40\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCCDS IDs (comma delimited)\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     41\u001b[0m ]\n\u001b[0;32m---> 42\u001b[0m df_homology_subset \u001b[38;5;241m=\u001b[39m \u001b[43mdf_homology\u001b[49m\u001b[43m[\u001b[49m\u001b[43mcolumns_to_keep\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m     44\u001b[0m \u001b[38;5;66;03m# Merge: keep all rows from MRK (left join)\u001b[39;00m\n\u001b[1;32m     45\u001b[0m df_merged \u001b[38;5;241m=\u001b[39m df_mrk\u001b[38;5;241m.\u001b[39mmerge(df_homology_subset, how\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mleft\u001b[39m\u001b[38;5;124m\"\u001b[39m, on\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMGI Marker Accession ID\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/frame.py:4108\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   4106\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m is_iterator(key):\n\u001b[1;32m   4107\u001b[0m         key \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(key)\n\u001b[0;32m-> 4108\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_indexer_strict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcolumns\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m[\u001b[38;5;241m1\u001b[39m]\n\u001b[1;32m   4110\u001b[0m \u001b[38;5;66;03m# take() does not accept boolean indexers\u001b[39;00m\n\u001b[1;32m   4111\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(indexer, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdtype\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m) \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mbool\u001b[39m:\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/indexes/base.py:6200\u001b[0m, in \u001b[0;36mIndex._get_indexer_strict\u001b[0;34m(self, key, axis_name)\u001b[0m\n\u001b[1;32m   6197\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   6198\u001b[0m     keyarr, indexer, new_indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reindex_non_unique(keyarr)\n\u001b[0;32m-> 6200\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_raise_if_missing\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkeyarr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis_name\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   6202\u001b[0m keyarr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtake(indexer)\n\u001b[1;32m   6203\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(key, Index):\n\u001b[1;32m   6204\u001b[0m     \u001b[38;5;66;03m# GH 42790 - Preserve name from an Index\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/indexes/base.py:6249\u001b[0m, in \u001b[0;36mIndex._raise_if_missing\u001b[0;34m(self, key, indexer, axis_name)\u001b[0m\n\u001b[1;32m   6247\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m nmissing:\n\u001b[1;32m   6248\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m nmissing \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mlen\u001b[39m(indexer):\n\u001b[0;32m-> 6249\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNone of [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mkey\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m] are in the [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00maxis_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m]\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   6251\u001b[0m     not_found \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(ensure_index(key)[missing_mask\u001b[38;5;241m.\u001b[39mnonzero()[\u001b[38;5;241m0\u001b[39m]]\u001b[38;5;241m.\u001b[39munique())\n\u001b[1;32m   6252\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mnot_found\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m not in index\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mKeyError\u001b[0m: \"None of [Index(['MGI Marker Accession ID', 'NCBI Gene ID',\\n       'CCDS IDs (comma delimited)'],\\n      dtype='object')] are in the [columns]\""
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f1929b0c-ef41-41af-97c9-52d2454484a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/sakuramaezono/Library/CloudStorage/OneDrive-YokohamaCityUniversity/Personal/05_Python_repositories/ConnectomeDB/src/createDataTable.py:28: DtypeWarning: Columns (32,34,38,40,50) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  pop_up_info = pd.read_table(\"data/HGNC_gene_info_full.tsv\")\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import jinja2\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "import base64\n",
    "import re\n",
    "\n",
    "# Add the src directory to the path for importing modules\n",
    "sys.path.append(os.path.abspath(\"src\"))\n",
    "\n",
    "# Import necessary modules from your existing src files\n",
    "# Ensure createDataTable and createFunctionalAnnotTable are in your 'src' directory\n",
    "from createDataTable import pop_up_info, gene_pair0, generate_perplexity_links, gene_pair00, is_mouse_specific\n",
    "from createFunctionalAnnotTable import gene_pair_annot_ligand, gene_pair_annot_receptor\n",
    "\n",
    "# Test or all\n",
    "test = False\n",
    "test_genes = [\"H60a Klrk1\", \"H60b Klrk1\", \"VEGFA NRP1\", \"THPO MPL\", \"FGF1 FGFR3\"] # Example genes\n",
    "# --- Paths ---\n",
    "MERGED_TEMPLATE_PATH = 'HTML/mergedCard_tabs.html'\n",
    "OUTPUT_DIR = 'data/cards/' # New output directory for combined files\n",
    "\n",
    "# function for replacing visible text:\n",
    "def update_link_text_with_symbol(html_str, new_symbol):\n",
    "    \"\"\"\n",
    "    Replace the visible text in an anchor tag with the provided symbol.\n",
    "    The HGNC ID is extracted from the href and left unchanged.\n",
    "    \"\"\"\n",
    "    # Only proceed if input is valid\n",
    "    if not isinstance(html_str, str) or not html_str.strip():\n",
    "        return html_str\n",
    "    \n",
    "    return re.sub(r\">(.*?)</a>\", f\">{new_symbol}</a>\", html_str) #<i class='fa-solid fa-arrow-up-right-from-square' style='margin-left:4px;'></i></a>\n",
    "\n",
    "# Create Perplexity link\n",
    "\n",
    "# Recreate Perplexity link\n",
    "def create_url_basic(symbol):\n",
    "          label = \"Perplexity (LLM)\"\n",
    "          query = f\"What diseases is {symbol} implicated in?\"\n",
    "          encoded_query = query.replace(\" \", \"%20\")\n",
    "          output = f'<a href=\"https://www.perplexity.ai/search?q={encoded_query}\" target=\"_blank\">{label}</a>' #<i class=\"fa-solid fa-arrow-up-right-from-square\" style=\"margin-left:4px;\"></i>\n",
    "          return output\n",
    "\n",
    "# --- Load and Preprocess Data (Combined from both scripts) ---\n",
    "\n",
    "# Load PubMed data (from createPMIDpages.py)\n",
    "pubmed_data = pd.read_csv(\"data/pubmed_results.csv\")\n",
    "pubmed_data[\"Year\"] = pubmed_data[\"Year\"].astype(str).str.replace(\".0\", \"\", regex=False).astype(int)\n",
    "pubmed_data[\"PMID\"] = pubmed_data[\"PMID\"].astype(str)\n",
    "pubmed_data = pubmed_data.reset_index(drop=True)\n",
    "\n",
    "# remove for now\n",
    "# # Load LLM results (from createPMIDpages.py)\n",
    "# bio_keywords = pd.read_csv(\"data/llm_results.csv\")\n",
    "\n",
    "# --- Prepare gene_pair00 for PMID section (from createPMIDpages.py) ---\n",
    "# gene_pair00 is used for PMID and Keywords, so it needs the '—' placeholder\n",
    "# Ensure gene_pair00 is a copy to avoid SettingWithCopyWarning later\n",
    "gene_pair00_copy = gene_pair00.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "394502ef-14e4-4b22-895c-e5e5b73dbaa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair000 = gene_pair00.copy()\n",
    "gene_pair000[\"LR Pair Card\"] = gene_pair000[\"LR Pair Card\"].astype(str)\n",
    "# --- Prepare gene_pair0 for Card section (from createCards.py) ---\n",
    "# gene_pair0 is used for card details, it should retain spaces for splitting gene names\n",
    "# Ensure gene_pair0 is a copy to avoid SettingWithCopyWarning later\n",
    "gene_pair0_copy = gene_pair0.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "221d2a89-34da-433e-9c30-a4b4d2b6b0b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# grab the pairs by interaction id that has ligand that is mouse_specific\n",
    "mouse_interaction_ids = gene_pair0_copy[gene_pair0_copy['Ligand'].apply(is_mouse_specific)]['Interaction ID'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8baa68ca-5830-4748-8196-1ebf2a3a9b28",
   "metadata": {},
   "outputs": [],
   "source": [
    "from createDataTable import pop_up_info, gene_pair0, generate_perplexity_links, gene_pair00, is_mouse_specific, grab_mouse_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "bfceeea8-1b1a-4c3b-852a-2e8f8b97b34b",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_interaction_ids = gene_pair0_copy[\"Interaction ID\"][gene_pair0_copy[\"Human evidence\"] == \"absent in human\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "81517f4f-a2a2-4483-b58a-1c713fd4868d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Interaction ID</th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>Human LR Pair</th>\n",
       "      <th>Ligand</th>\n",
       "      <th>Receptor</th>\n",
       "      <th>Ligand Symbols</th>\n",
       "      <th>Receptor Symbols</th>\n",
       "      <th>Ligand Location</th>\n",
       "      <th>Receptor Location</th>\n",
       "      <th>Ligand HGNC ID</th>\n",
       "      <th>Receptor HGNC ID</th>\n",
       "      <th>Perplexity</th>\n",
       "      <th>Human evidence</th>\n",
       "      <th>PMID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3561</th>\n",
       "      <td>CDB03562</td>\n",
       "      <td>CD99 Pilrb1</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>CD99</td>\n",
       "      <td>Pilrb1</td>\n",
       "      <td>&lt;span title=\"CD99 (MIC2, MIC2X, MIC2Y)\"&gt;CD99 (...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on perplexity\"&gt;cell membran...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3562</th>\n",
       "      <td>CDB03563</td>\n",
       "      <td>Ccl21C CXCR3</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Ccl21C</td>\n",
       "      <td>CXCR3</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"CXCR3 (GPR9, CKR-L2, CMKAR3, IP10...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3563</th>\n",
       "      <td>CDB03564</td>\n",
       "      <td>Ccl21C Cxcr3</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Ccl21C</td>\n",
       "      <td>Cxcr3</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3564</th>\n",
       "      <td>CDB03565</td>\n",
       "      <td>Defb2 Ccr6</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Defb2</td>\n",
       "      <td>Ccr6</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;secreted&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3565</th>\n",
       "      <td>CDB03566</td>\n",
       "      <td>Defb2 Tlr4</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Defb2</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;secreted&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3566</th>\n",
       "      <td>CDB03567</td>\n",
       "      <td>Fcna Tlr4</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Fcna</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3567</th>\n",
       "      <td>CDB03568</td>\n",
       "      <td>H60A Klrk1</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>H60A</td>\n",
       "      <td>Klrk1</td>\n",
       "      <td>&lt;span title=\"H60A\"&gt;H60A&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3568</th>\n",
       "      <td>CDB03569</td>\n",
       "      <td>H60B Klrk1</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>H60B</td>\n",
       "      <td>Klrk1</td>\n",
       "      <td>&lt;span title=\"H60B\"&gt;H60B&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3569</th>\n",
       "      <td>CDB03570</td>\n",
       "      <td>H60C Klrk1</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>H60C</td>\n",
       "      <td>Klrk1</td>\n",
       "      <td>&lt;span title=\"H60C\"&gt;H60C&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3570</th>\n",
       "      <td>CDB03571</td>\n",
       "      <td>Lair1 Lilrb4A</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Lair1</td>\n",
       "      <td>Lilrb4A</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3571</th>\n",
       "      <td>CDB03572</td>\n",
       "      <td>Pcdhb11 Pcdhb11</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Pcdhb11</td>\n",
       "      <td>Pcdhb11</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3572</th>\n",
       "      <td>CDB03573</td>\n",
       "      <td>Pcdhb14 Pcdhb14</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Pcdhb14</td>\n",
       "      <td>Pcdhb14</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3573</th>\n",
       "      <td>CDB03574</td>\n",
       "      <td>Pcdhb8 Pcdhb8</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>Pcdhb8</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3574</th>\n",
       "      <td>CDB03575</td>\n",
       "      <td>Pcdhgb8 Pcdhgb8</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3575</th>\n",
       "      <td>CDB03576</td>\n",
       "      <td>Sema4A Timd2</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Sema4A</td>\n",
       "      <td>Timd2</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3576</th>\n",
       "      <td>CDB03577</td>\n",
       "      <td>Zp3 Zp3R</td>\n",
       "      <td>no human ortholog</td>\n",
       "      <td>Zp3</td>\n",
       "      <td>Zp3R</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Interaction ID     LR Pair Card      Human LR Pair   Ligand Receptor  \\\n",
       "3561       CDB03562      CD99 Pilrb1  no human ortholog     CD99   Pilrb1   \n",
       "3562       CDB03563     Ccl21C CXCR3  no human ortholog   Ccl21C    CXCR3   \n",
       "3563       CDB03564     Ccl21C Cxcr3  no human ortholog   Ccl21C    Cxcr3   \n",
       "3564       CDB03565       Defb2 Ccr6  no human ortholog    Defb2     Ccr6   \n",
       "3565       CDB03566       Defb2 Tlr4  no human ortholog    Defb2     Tlr4   \n",
       "3566       CDB03567        Fcna Tlr4  no human ortholog     Fcna     Tlr4   \n",
       "3567       CDB03568       H60A Klrk1  no human ortholog     H60A    Klrk1   \n",
       "3568       CDB03569       H60B Klrk1  no human ortholog     H60B    Klrk1   \n",
       "3569       CDB03570       H60C Klrk1  no human ortholog     H60C    Klrk1   \n",
       "3570       CDB03571    Lair1 Lilrb4A  no human ortholog    Lair1  Lilrb4A   \n",
       "3571       CDB03572  Pcdhb11 Pcdhb11  no human ortholog  Pcdhb11  Pcdhb11   \n",
       "3572       CDB03573  Pcdhb14 Pcdhb14  no human ortholog  Pcdhb14  Pcdhb14   \n",
       "3573       CDB03574    Pcdhb8 Pcdhb8  no human ortholog   Pcdhb8   Pcdhb8   \n",
       "3574       CDB03575  Pcdhgb8 Pcdhgb8  no human ortholog  Pcdhgb8  Pcdhgb8   \n",
       "3575       CDB03576     Sema4A Timd2  no human ortholog   Sema4A    Timd2   \n",
       "3576       CDB03577         Zp3 Zp3R  no human ortholog      Zp3     Zp3R   \n",
       "\n",
       "                                         Ligand Symbols  \\\n",
       "3561  <span title=\"CD99 (MIC2, MIC2X, MIC2Y)\">CD99 (...   \n",
       "3562  <span title=\"no human ortholog\">no human ortho...   \n",
       "3563  <span title=\"no human ortholog\">no human ortho...   \n",
       "3564  <span title=\"no human ortholog\">no human ortho...   \n",
       "3565  <span title=\"no human ortholog\">no human ortho...   \n",
       "3566  <span title=\"no human ortholog\">no human ortho...   \n",
       "3567                     <span title=\"H60A\">H60A</span>   \n",
       "3568                     <span title=\"H60B\">H60B</span>   \n",
       "3569                     <span title=\"H60C\">H60C</span>   \n",
       "3570  <span title=\"no human ortholog\">no human ortho...   \n",
       "3571  <span title=\"no human ortholog\">no human ortho...   \n",
       "3572  <span title=\"no human ortholog\">no human ortho...   \n",
       "3573  <span title=\"no human ortholog\">no human ortho...   \n",
       "3574  <span title=\"no human ortholog\">no human ortho...   \n",
       "3575  <span title=\"no human ortholog\">no human ortho...   \n",
       "3576  <span title=\"no human ortholog\">no human ortho...   \n",
       "\n",
       "                                       Receptor Symbols  \\\n",
       "3561  <span title=\"no human ortholog\">no human ortho...   \n",
       "3562  <span title=\"CXCR3 (GPR9, CKR-L2, CMKAR3, IP10...   \n",
       "3563  <span title=\"no human ortholog\">no human ortho...   \n",
       "3564  <span title=\"no human ortholog\">no human ortho...   \n",
       "3565  <span title=\"no human ortholog\">no human ortho...   \n",
       "3566  <span title=\"no human ortholog\">no human ortho...   \n",
       "3567  <span title=\"no human ortholog\">no human ortho...   \n",
       "3568  <span title=\"no human ortholog\">no human ortho...   \n",
       "3569  <span title=\"no human ortholog\">no human ortho...   \n",
       "3570  <span title=\"no human ortholog\">no human ortho...   \n",
       "3571  <span title=\"no human ortholog\">no human ortho...   \n",
       "3572  <span title=\"no human ortholog\">no human ortho...   \n",
       "3573  <span title=\"no human ortholog\">no human ortho...   \n",
       "3574  <span title=\"no human ortholog\">no human ortho...   \n",
       "3575  <span title=\"no human ortholog\">no human ortho...   \n",
       "3576  <span title=\"no human ortholog\">no human ortho...   \n",
       "\n",
       "                                        Ligand Location  \\\n",
       "3561  <span title=\"based on perplexity\">cell membran...   \n",
       "3562                                            unknown   \n",
       "3563                                            unknown   \n",
       "3564     <span title=\"based on uniprot\">secreted</span>   \n",
       "3565     <span title=\"based on uniprot\">secreted</span>   \n",
       "3566                                            unknown   \n",
       "3567                                            unknown   \n",
       "3568                                            unknown   \n",
       "3569                                            unknown   \n",
       "3570  <span title=\"based on uniprot\">cell membrane</...   \n",
       "3571                                            unknown   \n",
       "3572  <span title=\"based on uniprot\">cell membrane</...   \n",
       "3573                                            unknown   \n",
       "3574  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "3575                                            unknown   \n",
       "3576  <span title=\"based on uniprot\">cell membrane</...   \n",
       "\n",
       "                                      Receptor Location  \\\n",
       "3561                                            unknown   \n",
       "3562  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "3563  <span title=\"based on uniprot\">cell membrane</...   \n",
       "3564  <span title=\"based on uniprot\">cell membrane</...   \n",
       "3565  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "3566  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "3567  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "3568  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "3569  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "3570                                            unknown   \n",
       "3571                                            unknown   \n",
       "3572  <span title=\"based on uniprot\">cell membrane</...   \n",
       "3573                                            unknown   \n",
       "3574  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "3575  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "3576                                            unknown   \n",
       "\n",
       "                                         Ligand HGNC ID  \\\n",
       "3561  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3562  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3563  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3564  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3565  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3566  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3567  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3568  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3569  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3570  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3571  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3572  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3573  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3574  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3575  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3576  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "\n",
       "                                       Receptor HGNC ID  \\\n",
       "3561  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3562  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3563  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3564  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3565  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3566  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3567  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3568  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3569  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3570  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3571  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3572  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3573  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3574  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3575  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3576  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "\n",
       "                                             Perplexity   Human evidence  \\\n",
       "3561  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "3562  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "3563  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "3564  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "3565  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "3566  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "3567  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "3568  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "3569  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "3570  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "3571  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "3572  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "3573  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "3574  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "3575  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "3576  <a href=\"https://www.perplexity.ai/search?q=Wh...  absent in human   \n",
       "\n",
       "                                                   PMID  \n",
       "3561  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...  \n",
       "3562  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...  \n",
       "3563  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...  \n",
       "3564  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...  \n",
       "3565  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...  \n",
       "3566  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...  \n",
       "3567  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...  \n",
       "3568  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...  \n",
       "3569  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...  \n",
       "3570  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...  \n",
       "3571  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...  \n",
       "3572  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...  \n",
       "3573  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...  \n",
       "3574  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...  \n",
       "3575  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...  \n",
       "3576  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair0_copy[gene_pair0_copy[\"Interaction ID\"].isin(mouse_interaction_ids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "eff51740-2682-49d4-a0a1-3375b1a64f23",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import jinja2\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "import base64\n",
    "import re\n",
    "\n",
    "# Add the src directory to the path for importing modules\n",
    "sys.path.append(os.path.abspath(\"src\"))\n",
    "\n",
    "# Import necessary modules from your existing src files\n",
    "# Ensure createDataTable and createFunctionalAnnotTable are in your 'src' directory\n",
    "from createDataTable import pop_up_info, gene_pair0, generate_perplexity_links, gene_pair00, is_mouse_specific\n",
    "from createFunctionalAnnotTable import gene_pair_annot_ligand, gene_pair_annot_receptor\n",
    "\n",
    "# Test or all\n",
    "test = False\n",
    "test_genes = [\"H60a Klrk1\", \"H60b Klrk1\", \"VEGFA NRP1\", \"THPO MPL\", \"FGF1 FGFR3\"] # Example genes\n",
    "# --- Paths ---\n",
    "MERGED_TEMPLATE_PATH = 'HTML/mergedCard_tabs.html'\n",
    "OUTPUT_DIR = 'data/cards/' # New output directory for combined files\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d682ff4d-f829-482b-8fd1-b50850bdd403",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Receptor HGNC ID</th>\n",
       "      <th>Receptor Symbols</th>\n",
       "      <th>Receptor Location</th>\n",
       "      <th>Receptor group</th>\n",
       "      <th>root_group_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;span title=\"&lt;span title=\"HSPA5 (GRP78, BiP)\"&gt;...</td>\n",
       "      <td>&lt;span title=\"based on PMID: 12194978, PMID: 32...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/genegr...</td>\n",
       "      <td>582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;span title=\"&lt;span title=\"LRP1 (APR, A2MR, LRP...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/genegr...</td>\n",
       "      <td>634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;span title=\"&lt;span title=\"LRP1 (APR, A2MR, LRP...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/genegr...</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;span title=\"&lt;span title=\"LRP1 (APR, A2MR, LRP...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/genegr...</td>\n",
       "      <td>1690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;span title=\"&lt;span title=\"BDKRB2 (BK-2)\"&gt;BDKRB...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/genegr...</td>\n",
       "      <td>2054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1513</th>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;span title=\"&lt;span title=\"no human ortholog\"&gt;n...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/genegr...</td>\n",
       "      <td>621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1514</th>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;span title=\"&lt;span title=\"no human ortholog\"&gt;n...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/genegr...</td>\n",
       "      <td>1298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1515</th>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;span title=\"&lt;span title=\"no human ortholog\"&gt;n...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/genegr...</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1516</th>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;span title=\"&lt;span title=\"no human ortholog\"&gt;n...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1517</th>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;span title=\"&lt;span title=\"no human ortholog\"&gt;n...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1518 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       Receptor HGNC ID  \\\n",
       "0     <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "1     <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "2     <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3     <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4     <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "...                                                 ...   \n",
       "1513  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "1514  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "1515  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "1516  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "1517  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "\n",
       "                                       Receptor Symbols  \\\n",
       "0     <span title=\"<span title=\"HSPA5 (GRP78, BiP)\">...   \n",
       "1     <span title=\"<span title=\"LRP1 (APR, A2MR, LRP...   \n",
       "2     <span title=\"<span title=\"LRP1 (APR, A2MR, LRP...   \n",
       "3     <span title=\"<span title=\"LRP1 (APR, A2MR, LRP...   \n",
       "4     <span title=\"<span title=\"BDKRB2 (BK-2)\">BDKRB...   \n",
       "...                                                 ...   \n",
       "1513  <span title=\"<span title=\"no human ortholog\">n...   \n",
       "1514  <span title=\"<span title=\"no human ortholog\">n...   \n",
       "1515  <span title=\"<span title=\"no human ortholog\">n...   \n",
       "1516  <span title=\"<span title=\"no human ortholog\">n...   \n",
       "1517  <span title=\"<span title=\"no human ortholog\">n...   \n",
       "\n",
       "                                      Receptor Location  \\\n",
       "0     <span title=\"based on PMID: 12194978, PMID: 32...   \n",
       "1     <span title=\"based on hpa, perplexity, uniprot...   \n",
       "2     <span title=\"based on hpa, perplexity, uniprot...   \n",
       "3     <span title=\"based on hpa, perplexity, uniprot...   \n",
       "4     <span title=\"based on hpa, perplexity, uniprot...   \n",
       "...                                                 ...   \n",
       "1513  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "1514  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "1515  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "1516  <span title=\"based on uniprot\">cell membrane</...   \n",
       "1517  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "\n",
       "                                         Receptor group root_group_id  \n",
       "0     <a href=\"https://www.genenames.org/data/genegr...           582  \n",
       "1     <a href=\"https://www.genenames.org/data/genegr...           634  \n",
       "2     <a href=\"https://www.genenames.org/data/genegr...           471  \n",
       "3     <a href=\"https://www.genenames.org/data/genegr...          1690  \n",
       "4     <a href=\"https://www.genenames.org/data/genegr...          2054  \n",
       "...                                                 ...           ...  \n",
       "1513  <a href=\"https://www.genenames.org/data/genegr...           621  \n",
       "1514  <a href=\"https://www.genenames.org/data/genegr...          1298  \n",
       "1515  <a href=\"https://www.genenames.org/data/genegr...           471  \n",
       "1516                                            unknown           NaN  \n",
       "1517                                            unknown           NaN  \n",
       "\n",
       "[1518 rows x 5 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair_annot_receptor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8535524-a617-45b2-b6cb-f3a42408e004",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "068ca3ed-7b44-480a-be04-0ed5777f4ebc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing HGNC IDs:\n",
      "  HGNC:607\n",
      "HGNC:13916\n",
      "          \n",
      "      #N/A\n",
      "HGNC:16721\n",
      "HGNC:16722\n",
      "HGNC:16451\n",
      "HGNC:21969\n",
      " HGNC:2475\n",
      "HGNC:20594\n",
      " HGNC:3600\n",
      " HGNC:5474\n",
      " HGNC:6552\n",
      " HGNC:9030\n",
      "HGNC:10468\n",
      "HGNC:10469\n",
      "HGNC:10515\n",
      "HGNC:10740\n",
      "HGNC:10776\n",
      "HGNC:20323\n",
      "HGNC:30668\n",
      "  HGNC:281\n",
      "HGNC:12338\n",
      " HGNC:8535\n",
      "HGNC:12449\n",
      " HGNC:3608\n",
      " HGNC:4471\n",
      "HGNC:14687\n",
      " HGNC:4469\n",
      "HGNC:14901\n",
      "HGNC:10875\n",
      "HGNC:14558\n",
      "HGNC:17867\n",
      " HGNC:6693\n",
      " HGNC:1665\n"
     ]
    }
   ],
   "source": [
    "# Clean the column in the existing gene_group\n",
    "hgnc_id_column = fetchGSheet.gene_group[\"hgnc_id\"].astype(str).str.strip()\n",
    "\n",
    "# Clean the input IDs\n",
    "cleaned_input_ids = pd.Series(hgnc_id).astype(str).str.strip()\n",
    "\n",
    "# Find missing IDs\n",
    "missing_ids = cleaned_input_ids[~cleaned_input_ids.isin(hgnc_id_column)]\n",
    "\n",
    "# Optionally print the whole list without truncation\n",
    "print(\"Missing HGNC IDs:\")\n",
    "print(missing_ids.to_string(index=False))\n",
    "\n",
    "# Save to CSV\n",
    "missing_ids_df = pd.DataFrame({\"missing_hgnc_id\": missing_ids})\n",
    "missing_ids_df.to_csv(\"missing_hgnc_ids.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "beb625f1-b877-4207-92f2-622176e9f2b1",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'str' object has no attribute 'astype'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[149], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m cleaned_hgnc_ids_column \u001b[38;5;241m=\u001b[39m \u001b[43mcleaned_id\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mastype\u001b[49m(\u001b[38;5;28mstr\u001b[39m)\u001b[38;5;241m.\u001b[39mstr\u001b[38;5;241m.\u001b[39mstrip()\n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m# Print all the cleaned IDs without truncation\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;28mprint\u001b[39m(cleaned_hgnc_ids_column\u001b[38;5;241m.\u001b[39mto_string())\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'str' object has no attribute 'astype'"
     ]
    }
   ],
   "source": [
    "cleaned_hgnc_ids_column = cleaned_id.astype(str).str.strip()\n",
    "\n",
    "# Print all the cleaned IDs without truncation\n",
    "print(cleaned_hgnc_ids_column.to_string())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "441bc740-a572-4960-87d9-e9200d151f19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Quick check if there is mouse-specific\n",
    "gene_pair['Ligand'] = gene_pair.apply(\n",
    "    lambda row: \"no human ortholog\" if is_mouse_specific(row['Ligand']) else row['Ligand'],\n",
    "    axis=1\n",
    ")\n",
    "gene_pair['Receptor'] = gene_pair.apply(\n",
    "    lambda row: \"no human ortholog\" if is_mouse_specific(row['Receptor']) else row['Receptor'],\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "# gene symbol\n",
    "gene_pair[\"Ligand\"] = [\n",
    "    f'<span title=\"{ligand_name}\">{ligand_symbol}</span>'\n",
    "    for ligand_name, ligand_symbol in zip(gene_pair[\"Ligand Name\"], \n",
    "                                              gene_pair[\"Ligand\"])\n",
    "]\n",
    "# gene symbol\n",
    "gene_pair[\"Receptor\"] = [\n",
    "    f'<span title=\"{receptor_name}\">{receptor_symbol}</span>'\n",
    "    for receptor_name, receptor_symbol in zip(gene_pair[\"Receptor Name\"], \n",
    "                                              gene_pair[\"Receptor\"])\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "c9418728-d4a2-4e76-a64c-d509137e81c3",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"['Ligand Name', 'Receptor Name'] not found in axis\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[120], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m gene_pair \u001b[38;5;241m=\u001b[39m \u001b[43mgene_pair\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdrop\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mLigand Name\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mReceptor Name\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m# Create the links to the HTML cards\u001b[39;00m\n\u001b[1;32m      5\u001b[0m gene_pair[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mHuman LR Pair\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m      6\u001b[0m     \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m<a href=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhttps://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/cards/\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mlrPairOrig\u001b[38;5;241m.\u001b[39mreplace(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m \u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m-\u001b[39m\u001b[38;5;124m\"\u001b[39m)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.html\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m>\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mlrPair\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m</a>\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m      7\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m lrPairOrig, lrPair \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mzip\u001b[39m(gene_pair0[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mHuman LR Pair\u001b[39m\u001b[38;5;124m\"\u001b[39m], gene_pair[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mHuman LR Pair\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n\u001b[1;32m      8\u001b[0m ]\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/frame.py:5581\u001b[0m, in \u001b[0;36mDataFrame.drop\u001b[0;34m(self, labels, axis, index, columns, level, inplace, errors)\u001b[0m\n\u001b[1;32m   5433\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mdrop\u001b[39m(\n\u001b[1;32m   5434\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   5435\u001b[0m     labels: IndexLabel \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   5442\u001b[0m     errors: IgnoreRaise \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mraise\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m   5443\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m DataFrame \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   5444\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   5445\u001b[0m \u001b[38;5;124;03m    Drop specified labels from rows or columns.\u001b[39;00m\n\u001b[1;32m   5446\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   5579\u001b[0m \u001b[38;5;124;03m            weight  1.0     0.8\u001b[39;00m\n\u001b[1;32m   5580\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 5581\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdrop\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   5582\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlabels\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlabels\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   5583\u001b[0m \u001b[43m        \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   5584\u001b[0m \u001b[43m        \u001b[49m\u001b[43mindex\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mindex\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   5585\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   5586\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlevel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlevel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   5587\u001b[0m \u001b[43m        \u001b[49m\u001b[43minplace\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minplace\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   5588\u001b[0m \u001b[43m        \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   5589\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/generic.py:4788\u001b[0m, in \u001b[0;36mNDFrame.drop\u001b[0;34m(self, labels, axis, index, columns, level, inplace, errors)\u001b[0m\n\u001b[1;32m   4786\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m axis, labels \u001b[38;5;129;01min\u001b[39;00m axes\u001b[38;5;241m.\u001b[39mitems():\n\u001b[1;32m   4787\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m labels \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 4788\u001b[0m         obj \u001b[38;5;241m=\u001b[39m \u001b[43mobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_drop_axis\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlabels\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlevel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlevel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4790\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m inplace:\n\u001b[1;32m   4791\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_update_inplace(obj)\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/generic.py:4830\u001b[0m, in \u001b[0;36mNDFrame._drop_axis\u001b[0;34m(self, labels, axis, level, errors, only_slice)\u001b[0m\n\u001b[1;32m   4828\u001b[0m         new_axis \u001b[38;5;241m=\u001b[39m axis\u001b[38;5;241m.\u001b[39mdrop(labels, level\u001b[38;5;241m=\u001b[39mlevel, errors\u001b[38;5;241m=\u001b[39merrors)\n\u001b[1;32m   4829\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 4830\u001b[0m         new_axis \u001b[38;5;241m=\u001b[39m \u001b[43maxis\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdrop\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlabels\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4831\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m axis\u001b[38;5;241m.\u001b[39mget_indexer(new_axis)\n\u001b[1;32m   4833\u001b[0m \u001b[38;5;66;03m# Case for non-unique axis\u001b[39;00m\n\u001b[1;32m   4834\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/indexes/base.py:7070\u001b[0m, in \u001b[0;36mIndex.drop\u001b[0;34m(self, labels, errors)\u001b[0m\n\u001b[1;32m   7068\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m mask\u001b[38;5;241m.\u001b[39many():\n\u001b[1;32m   7069\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m errors \u001b[38;5;241m!=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mignore\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m-> 7070\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mlabels[mask]\u001b[38;5;241m.\u001b[39mtolist()\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m not found in axis\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   7071\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m indexer[\u001b[38;5;241m~\u001b[39mmask]\n\u001b[1;32m   7072\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdelete(indexer)\n",
      "\u001b[0;31mKeyError\u001b[0m: \"['Ligand Name', 'Receptor Name'] not found in axis\""
     ]
    }
   ],
   "source": [
    "gene_pair = gene_pair.drop(columns=[\"Ligand Name\", \"Receptor Name\"])\n",
    "\n",
    "\n",
    "# Create the links to the HTML cards\n",
    "gene_pair[\"Human LR Pair\"] = [\n",
    "    f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/cards/{lrPairOrig.replace(\" \",\"-\")}.html\">{lrPair}</a>'\n",
    "    for lrPairOrig, lrPair in zip(gene_pair0[\"Human LR Pair\"], gene_pair[\"Human LR Pair\"])\n",
    "]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Add tooltips to the column headers\n",
    "gene_pair.columns = [\n",
    "    f'<span title=\"Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\">{col}</span>' if col == \"Human LR Pair\" else\n",
    "    f'<span title=\"Click the logo below to run Perplexity on the Human LR pair\">{col}&nbsp;</span>' if col == \"Perplexity\" else\n",
    "    f'<span title=\"Official Gene Symbol; Hover on symbols below to show gene names\">{col}&nbsp;&nbsp;&nbsp;</span>' if col in [\"Ligand\", \"Receptor\"] else\n",
    "    f'<span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">{col}&nbsp;&nbsp;</span>' if col in [\"Ligand HGNC ID\", \"Receptor HGNC ID\"] else\n",
    "    f'<span title=\" PubMed IDs (PMID) with Literature Evidence for LR Interaction. Click on the link for more details\">{col}</span>' if col == \"PMID\" else\n",
    "    f'<span title=\"Rat Genome Database (RGD) ID. Click on the link for more details\">{col}</span>' if col in [\"Ligand RGD ID\", \"Receptor RGD ID\"] else\n",
    "    f'<span title=\"Mouse Genome Informatics (MGI) ID. Click on the link for more details\">{col}</span>' if col in [\"Ligand MGI ID\", \"Receptor MGI ID\"]else\n",
    "    f'<span title=\"Zebrafish Information Network (ZFIN) ID. Click on the link for more details\">{col}</span>' if col in [\"Ligand ZFIN ID\", \"Receptor ZFIN ID\"] else\n",
    "    f'<span title=\"Location based on the predicted subcellular localization of the human proteome\">{col}</span>' if col in [\"Ligand Location\", \"Receptor Location\"] else\n",
    "    f'<span title=\"Double-click header of {col} to ensure all values are shown\">{col}&nbsp;</span>'\n",
    "    for col in gene_pair.columns\n",
    "]\n",
    "gene_pair = gene_pair.reset_index(drop=True)  # Remove the index\n",
    "\n",
    "#######################################################################\n",
    "# Identify the column(s) that contain '(PMID)' and temporarily remove for presubmission\n",
    "pmid_cols = [col for col in gene_pair.columns if '(PMID)' in col]\n",
    "gene_pair = gene_pair.drop(columns=pmid_cols)\n",
    "#######################################################################\n",
    "\n",
    "gene_pair000 = gene_pair.copy()\n",
    "\n",
    "keywords_to_modify = [\"Ligand\", \"Receptor\"]\n",
    "exclude_keywords = [\"HGNC ID\", \"Location\", \"Human\"]  # Columns containing this will not be modified\n",
    "\n",
    "# Copy the original columns so we can modify only the first 10\n",
    "new_columns = gene_pair000.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "47d8b607-ff09-4a7c-8a1a-ec7706b3a7de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['<span title=\"Double-click header of LR Pair Card to ensure all values are shown\">LR Pair Card&nbsp;</span>',\n",
       " '<span title=\"Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\">Human LR Pair</span>',\n",
       " '<span title=\"Official Gene Symbol; Hover on symbols below to show gene names\">Ligand&nbsp;&nbsp;&nbsp;</span>',\n",
       " '<span title=\"Official Gene Symbol; Hover on symbols below to show gene names\">Receptor&nbsp;&nbsp;&nbsp;</span>',\n",
       " '<span title=\"Double-click header of Ligand Symbols to ensure all values are shown\">Ligand Symbols&nbsp;</span>',\n",
       " '<span title=\"Double-click header of Receptor Symbols to ensure all values are shown\">Receptor Symbols&nbsp;</span>',\n",
       " '<span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\">Ligand Location</span>',\n",
       " '<span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\">Receptor Location</span>',\n",
       " '<span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Ligand HGNC ID&nbsp;&nbsp;</span>',\n",
       " '<span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Receptor HGNC ID&nbsp;&nbsp;</span>',\n",
       " '<span title=\"Click the logo below to run Perplexity on the Human LR pair\">Perplexity&nbsp;</span>',\n",
       " '<span title=\"Double-click header of Human evidence to ensure all values are shown\">Human evidence&nbsp;</span>']"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "fb4a1091-6016-4ae8-9f4f-d86c2a992df0",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_rows = gene_pair[~(gene_pair[\"Human evidence\"] == \"absent in human\")]\n",
    "mouse_rows = gene_pair[gene_pair[\"Human evidence\"] == \"absent in human\"] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "b1992766-71aa-4b9c-b090-3e86438deecc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Concatenate the DataFrames: rows with IDs first, then rows without IDs\n",
    "gene_pair = pd.concat([human_rows, mouse_rows]).reset_index(drop=True)\n",
    "DBlength = len(gene_pair)\n",
    "gene_pair[\"Interaction ID\"] = [f\"CDB{str(i).zfill(5)}\" for i in range(1, DBlength + 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "8d0eac52-26d8-4b97-8b30-b4c606a4c0b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_columns = [col for col in gene_pair000.columns][:16]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "887f9cff-e2a8-4b39-8303-60126224fc81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['<span title=\"Double-click header of Human evidence to ensure all values are shown\">Human evidence&nbsp;</span>']"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "evidence_cols = [col for col in gene_pair.columns if 'Human evidence' in col]\n",
    "evidence_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "e57a327a-575e-440c-91fa-19e5c2b1a42b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;span title=\"Double-click header of LR Pair Card to ensure all values are shown\"&gt;LR Pair Card&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\"&gt;Human LR Pair&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Official Gene Symbol; Hover on symbols below to show gene names\"&gt;Ligand&amp;nbsp;&amp;nbsp;&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Official Gene Symbol; Hover on symbols below to show gene names\"&gt;Receptor&amp;nbsp;&amp;nbsp;&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Ligand Symbols to ensure all values are shown\"&gt;Ligand Symbols&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Receptor Symbols to ensure all values are shown\"&gt;Receptor Symbols&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\"&gt;Ligand Location&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\"&gt;Receptor Location&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\"&gt;Ligand HGNC ID&amp;nbsp;&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\"&gt;Receptor HGNC ID&amp;nbsp;&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Click the logo below to run Perplexity on the Human LR pair\"&gt;Perplexity&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Human evidence to ensure all values are shown\"&gt;Human evidence&amp;nbsp;&lt;/span&gt;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\"alpha-2-macroglobulin\"&gt;A2M&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"heat shock protein family A (Hsp7...</td>\n",
       "      <td>&lt;span title=\"A2M (FWP007, S863-7, CPAMD5)\"&gt;A2M...</td>\n",
       "      <td>&lt;span title=\"HSPA5 (GRP78, BiP)\"&gt;HSPA5 (GRP78,...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;span title=\"based on PMID: 12194978, PMID: 32...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>DIRECT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\"alpha-2-macroglobulin\"&gt;A2M&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"LDL receptor related protein 1\"&gt;L...</td>\n",
       "      <td>&lt;span title=\"A2M (FWP007, S863-7, CPAMD5)\"&gt;A2M...</td>\n",
       "      <td>&lt;span title=\"LRP1 (APR, A2MR, LRP, CD91, LRP1A...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>DIRECT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACE BDKRB2</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\"angiotensin I converting enzyme\"&gt;...</td>\n",
       "      <td>&lt;span title=\"bradykinin receptor B2\"&gt;BDKRB2&lt;/s...</td>\n",
       "      <td>&lt;span title=\"ACE (DCP1, ACE1, CD143)\"&gt;ACE (DCP...</td>\n",
       "      <td>&lt;span title=\"BDKRB2 (BK-2)\"&gt;BDKRB2 (BK-2)&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>DIRECT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADA DPP4</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\"adenosine deaminase\"&gt;ADA&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"dipeptidyl peptidase 4\"&gt;DPP4&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"ADA (ADA1)\"&gt;ADA (ADA1)&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"DPP4 (CD26, ADCP2, DPPIV)\"&gt;DPP4 (...</td>\n",
       "      <td>&lt;span title=\"based on hpa, uniprot\"&gt;cell membr...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>CONSERVATION</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADAM10 EFNA5</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\"ADAM metallopeptidase domain 10\"&gt;...</td>\n",
       "      <td>&lt;span title=\"ephrin A5\"&gt;EFNA5&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"ADAM10 (kuz, MADM, HsT18717, CD15...</td>\n",
       "      <td>&lt;span title=\"EFNA5 (EPLG7, AF1, LERK7)\"&gt;EFNA5 ...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>CONSERVATION</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3556</th>\n",
       "      <td>not conserved CCR1</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\"nan\"&gt;no human ortholog&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"C-C motif chemokine receptor 1\"&gt;n...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>not conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3557</th>\n",
       "      <td>not conserved CCR2</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\"nan\"&gt;no human ortholog&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"C-C motif chemokine receptor 2\"&gt;n...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>not conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3558</th>\n",
       "      <td>not conserved CCR7</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\"nan\"&gt;no human ortholog&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"C-C motif chemokine receptor 7\"&gt;n...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>not conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3559</th>\n",
       "      <td>not conserved CXCR3</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\"nan\"&gt;no human ortholog&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"C-X-C motif chemokine receptor 3\"...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>not conserved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3560</th>\n",
       "      <td>not conserved not conserved</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\"nan\"&gt;no human ortholog&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"nan\"&gt;no human ortholog&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>not conserved</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3561 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     <span title=\"Double-click header of LR Pair Card to ensure all values are shown\">LR Pair Card&nbsp;</span>  \\\n",
       "0                                             A2M HSPA5                                                           \n",
       "1                                              A2M LRP1                                                           \n",
       "2                                            ACE BDKRB2                                                           \n",
       "3                                              ADA DPP4                                                           \n",
       "4                                          ADAM10 EFNA5                                                           \n",
       "...                                                 ...                                                           \n",
       "3556                                 not conserved CCR1                                                           \n",
       "3557                                 not conserved CCR2                                                           \n",
       "3558                                 not conserved CCR7                                                           \n",
       "3559                                not conserved CXCR3                                                           \n",
       "3560                        not conserved not conserved                                                           \n",
       "\n",
       "     <span title=\"Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\">Human LR Pair</span>  \\\n",
       "0     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                               \n",
       "1     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                               \n",
       "2     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                               \n",
       "3     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                               \n",
       "4     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                               \n",
       "...                                                 ...                                                               \n",
       "3556  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                               \n",
       "3557  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                               \n",
       "3558  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                               \n",
       "3559  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                               \n",
       "3560  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                               \n",
       "\n",
       "     <span title=\"Official Gene Symbol; Hover on symbols below to show gene names\">Ligand&nbsp;&nbsp;&nbsp;</span>  \\\n",
       "0        <span title=\"alpha-2-macroglobulin\">A2M</span>                                                              \n",
       "1        <span title=\"alpha-2-macroglobulin\">A2M</span>                                                              \n",
       "2     <span title=\"angiotensin I converting enzyme\">...                                                              \n",
       "3          <span title=\"adenosine deaminase\">ADA</span>                                                              \n",
       "4     <span title=\"ADAM metallopeptidase domain 10\">...                                                              \n",
       "...                                                 ...                                                              \n",
       "3556         <span title=\"nan\">no human ortholog</span>                                                              \n",
       "3557         <span title=\"nan\">no human ortholog</span>                                                              \n",
       "3558         <span title=\"nan\">no human ortholog</span>                                                              \n",
       "3559         <span title=\"nan\">no human ortholog</span>                                                              \n",
       "3560         <span title=\"nan\">no human ortholog</span>                                                              \n",
       "\n",
       "     <span title=\"Official Gene Symbol; Hover on symbols below to show gene names\">Receptor&nbsp;&nbsp;&nbsp;</span>  \\\n",
       "0     <span title=\"heat shock protein family A (Hsp7...                                                                \n",
       "1     <span title=\"LDL receptor related protein 1\">L...                                                                \n",
       "2     <span title=\"bradykinin receptor B2\">BDKRB2</s...                                                                \n",
       "3      <span title=\"dipeptidyl peptidase 4\">DPP4</span>                                                                \n",
       "4                  <span title=\"ephrin A5\">EFNA5</span>                                                                \n",
       "...                                                 ...                                                                \n",
       "3556  <span title=\"C-C motif chemokine receptor 1\">n...                                                                \n",
       "3557  <span title=\"C-C motif chemokine receptor 2\">n...                                                                \n",
       "3558  <span title=\"C-C motif chemokine receptor 7\">n...                                                                \n",
       "3559  <span title=\"C-X-C motif chemokine receptor 3\"...                                                                \n",
       "3560         <span title=\"nan\">no human ortholog</span>                                                                \n",
       "\n",
       "     <span title=\"Double-click header of Ligand Symbols to ensure all values are shown\">Ligand Symbols&nbsp;</span>  \\\n",
       "0     <span title=\"A2M (FWP007, S863-7, CPAMD5)\">A2M...                                                               \n",
       "1     <span title=\"A2M (FWP007, S863-7, CPAMD5)\">A2M...                                                               \n",
       "2     <span title=\"ACE (DCP1, ACE1, CD143)\">ACE (DCP...                                                               \n",
       "3            <span title=\"ADA (ADA1)\">ADA (ADA1)</span>                                                               \n",
       "4     <span title=\"ADAM10 (kuz, MADM, HsT18717, CD15...                                                               \n",
       "...                                                 ...                                                               \n",
       "3556  <span title=\"no human ortholog\">no human ortho...                                                               \n",
       "3557  <span title=\"no human ortholog\">no human ortho...                                                               \n",
       "3558  <span title=\"no human ortholog\">no human ortho...                                                               \n",
       "3559  <span title=\"no human ortholog\">no human ortho...                                                               \n",
       "3560  <span title=\"no human ortholog\">no human ortho...                                                               \n",
       "\n",
       "     <span title=\"Double-click header of Receptor Symbols to ensure all values are shown\">Receptor Symbols&nbsp;</span>  \\\n",
       "0     <span title=\"HSPA5 (GRP78, BiP)\">HSPA5 (GRP78,...                                                                   \n",
       "1     <span title=\"LRP1 (APR, A2MR, LRP, CD91, LRP1A...                                                                   \n",
       "2      <span title=\"BDKRB2 (BK-2)\">BDKRB2 (BK-2)</span>                                                                   \n",
       "3     <span title=\"DPP4 (CD26, ADCP2, DPPIV)\">DPP4 (...                                                                   \n",
       "4     <span title=\"EFNA5 (EPLG7, AF1, LERK7)\">EFNA5 ...                                                                   \n",
       "...                                                 ...                                                                   \n",
       "3556  <span title=\"no human ortholog\">no human ortho...                                                                   \n",
       "3557  <span title=\"no human ortholog\">no human ortho...                                                                   \n",
       "3558  <span title=\"no human ortholog\">no human ortho...                                                                   \n",
       "3559  <span title=\"no human ortholog\">no human ortho...                                                                   \n",
       "3560  <span title=\"no human ortholog\">no human ortho...                                                                   \n",
       "\n",
       "     <span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\">Ligand Location</span>  \\\n",
       "0     <span title=\"based on perplexity, uniprot\">sec...                                                                                                                        \n",
       "1     <span title=\"based on perplexity, uniprot\">sec...                                                                                                                        \n",
       "2     <span title=\"based on perplexity, uniprot\">cel...                                                                                                                        \n",
       "3     <span title=\"based on hpa, uniprot\">cell membr...                                                                                                                        \n",
       "4     <span title=\"based on hpa, perplexity, uniprot...                                                                                                                        \n",
       "...                                                 ...                                                                                                                        \n",
       "3556                                            unknown                                                                                                                        \n",
       "3557                                            unknown                                                                                                                        \n",
       "3558                                            unknown                                                                                                                        \n",
       "3559                                            unknown                                                                                                                        \n",
       "3560                                            unknown                                                                                                                        \n",
       "\n",
       "     <span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\">Receptor Location</span>  \\\n",
       "0     <span title=\"based on PMID: 12194978, PMID: 32...                                                                                                                          \n",
       "1     <span title=\"based on hpa, perplexity, uniprot...                                                                                                                          \n",
       "2     <span title=\"based on hpa, perplexity, uniprot...                                                                                                                          \n",
       "3     <span title=\"based on perplexity, uniprot\">cel...                                                                                                                          \n",
       "4     <span title=\"based on perplexity, uniprot\">cel...                                                                                                                          \n",
       "...                                                 ...                                                                                                                          \n",
       "3556                                            unknown                                                                                                                          \n",
       "3557                                            unknown                                                                                                                          \n",
       "3558                                            unknown                                                                                                                          \n",
       "3559                                            unknown                                                                                                                          \n",
       "3560                                            unknown                                                                                                                          \n",
       "\n",
       "     <span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Ligand HGNC ID&nbsp;&nbsp;</span>  \\\n",
       "0     <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "1     <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "2     <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "3     <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "4     <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "...                                                 ...                                                                               \n",
       "3556  <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "3557  <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "3558  <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "3559  <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "3560  <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "\n",
       "     <span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Receptor HGNC ID&nbsp;&nbsp;</span>  \\\n",
       "0     <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "1     <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "2     <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "3     <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "4     <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "...                                                 ...                                                                                 \n",
       "3556  <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "3557  <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "3558  <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "3559  <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "3560  <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "\n",
       "     <span title=\"Click the logo below to run Perplexity on the Human LR pair\">Perplexity&nbsp;</span>  \\\n",
       "0     <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "1     <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "2     <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "3     <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "4     <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "...                                                 ...                                                  \n",
       "3556  <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "3557  <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "3558  <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "3559  <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "3560  <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "\n",
       "     <span title=\"Double-click header of Human evidence to ensure all values are shown\">Human evidence&nbsp;</span>  \n",
       "0                                                DIRECT                                                              \n",
       "1                                                DIRECT                                                              \n",
       "2                                                DIRECT                                                              \n",
       "3                                          CONSERVATION                                                              \n",
       "4                                          CONSERVATION                                                              \n",
       "...                                                 ...                                                              \n",
       "3556                                      not conserved                                                              \n",
       "3557                                      not conserved                                                              \n",
       "3558                                      not conserved                                                              \n",
       "3559                                      not conserved                                                              \n",
       "3560                                      not conserved                                                              \n",
       "\n",
       "[3561 rows x 12 columns]"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_gene_pair = gene_pair[~(gene_pair[evidence_cols[0]] == \"absent in human\")]\n",
    "human_gene_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66aac484-af60-42f9-873f-e1bf98157831",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create Perplexity link\n",
    "def create_url_basic(perplexity_col):\n",
    "    query = f\"What is the primary evidence that {perplexity_col} bind-each-other-as-a-ligand-and-receptor-pair. Exclude reviews, uniprot, wiki, genecards, PIPS, iuphar as sources.\"\n",
    "    encoded_query = query.replace(\" \", \"%20\")\n",
    "    return f\"https://www.perplexity.ai/search?q={encoded_query}\"\n",
    "    \n",
    "# Option 2 -- new query all together\n",
    "\n",
    "# def generate_perplexity_link_pmid(row): \n",
    "#     query = f\"What-is-the-biological-relevance-of-the-ligand-and-receptor-pair-{row['Human LR Pair']}-based-on-Pubmed-ID-{row['PMID']}\"\n",
    "#     return (\n",
    "#          f'<a href=\"https://www.perplexity.ai/search?q={query}\" target=\"_blank\">'\n",
    "#         f'<img src=\"https://img.icons8.com/?size=30&id=0NbBuNOxUwps&format=png&color=000000\" alt=\"Perplexity AI\" /></a>'\n",
    "#     )\n",
    "\n",
    "# cannot use perplexity logo\n",
    "def generate_perplexity_link_pmid(row): \n",
    "    query = f\"What-is-the-biological-relevance-of-the-ligand-and-receptor-pair-{row['Human LR Pair']}-based-on-Pubmed-ID-{row['PMID']}\"\n",
    "    return (\n",
    "         f'<a href=\"https://www.perplexity.ai/search?q={query}\" target=\"_blank\" style=\"text-decoration: none;\">&#128269;</a>'\n",
    "    )\n",
    "\n",
    "# Apply function to the DataFrame\n",
    "gene_pair[\"Perplexity\"] = gene_pair.apply(generate_perplexity_link_pmid, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "7c16544c-bd32-4e62-8ba5-0976dc04b2da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ligand\n",
    "gene_pair[\"Ligand HGNC ID\"] = [\n",
    "    '<a href=\"https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/{}\" target=\"_blank\">{}</a>'.format(ligand, ligand)\n",
    "    for ligand in gene_pair[\"Ligand HGNC ID\"]\n",
    "]\n",
    "\n",
    "# receptor\n",
    "gene_pair[\"Receptor HGNC ID\"] = [\n",
    "    '<a href=\"https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/{}\" target=\"_blank\">{}</a>'.format(receptor, receptor)\n",
    "    for receptor in gene_pair[\"Receptor HGNC ID\"]\n",
    "]\n",
    "\n",
    "\n",
    "# Function to generate hyperlinks for the \"PMID support\" column\n",
    "def generate_links_with_doi(df, gene_column, pmid_column, id_column):\n",
    "    def create_link(gene, id_col, sources):\n",
    "        # Replace spaces with \"——\" in the gene name for the link\n",
    "        gene_name = gene.replace(\" \", \"—\")\n",
    "        \n",
    "        if len(sources) == 1:\n",
    "            source = sources[0]\n",
    "            if source.startswith(\"https://www.biorxiv.org/content/\"):\n",
    "                # If the value starts with \"https://doi.org/\", use it as the hyperlink\n",
    "                return f'<a href=\"{source}\" target=\"_blank\">BioRxiv</a>'\n",
    "            else:\n",
    "                # If it's a single PMID, hyperlink the PMID text\n",
    "                return f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/cards/{gene}.html\">{source}</a>'\n",
    "        else:\n",
    "            # If multiple PMIDs, show the count and hyperlink to the page\n",
    "            return f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/cards/{gene}.html\" target=\"_blank\">{len(sources)} PMIDs</a>'\n",
    "\n",
    "    # Process each row to generate the \"PMID\" column # was \"PMID support\"\n",
    "    df[\"PMID\"] = [\n",
    "        create_link(\n",
    "            gene=row[gene_column], \n",
    "            id_col = row[id_column],\n",
    "            sources=[s.strip() for s in row[pmid_column].split(',') if s.strip()]\n",
    "        )\n",
    "        for _, row in df.iterrows()\n",
    "    ]\n",
    "    return df\n",
    "\n",
    "\n",
    "# Generate the links for the \"PMID\" column # was \"PMID support\"\n",
    "gene_pair = generate_links_with_doi(gene_pair, gene_column=\"Human LR Pair\", \n",
    "                                    pmid_column=\"PMID\", id_column= \"Interaction ID\")\n",
    "\n",
    "# for disease type, cancer-related and top pathways, when missing say \"ask Perplexity\"\n",
    "\n",
    "\n",
    "def generate_perplexity_kegglinks(\n",
    "    df,\n",
    "    pathway_col=\"KEGG Pathway\",\n",
    "    default_query_template=\"What-biological or other functional-pathways-is-the-ligand-receptor-{pair}-associated-with\"\n",
    "):\n",
    "    def create_link(row):\n",
    "        value = row.get(pathway_col, \"\")\n",
    "        \n",
    "        if pd.isna(value) or str(value).strip().lower() in [\"nan\", \"none\", \"\", \"unknown\"]:\n",
    "            pair = row[\"Human LR Pair\"]\n",
    "            label = \"ask Perplexity\"\n",
    "            query = default_query_template.format(pair=pair)\n",
    "            encoded_query = urllib.parse.quote(query)\n",
    "            return f'<a href=\"https://www.perplexity.ai/search?q={encoded_query}\" target=\"_blank\">{label}</a>'\n",
    "        else:\n",
    "            return value\n",
    "\n",
    "    df[pathway_col] = df.apply(create_link, axis=1)\n",
    "    return df\n",
    "\n",
    "gene_pair = generate_perplexity_kegglinks(gene_pair, pathway_col=\"KEGG Pathway\")\n",
    "    \n",
    "def generate_perplexity_links(df, pathway_col, default_query_template):\n",
    "    def create_link(row):\n",
    "        pathway_value = str(row[pathway_col]).strip().lower()\n",
    "        pair = row[\"Human LR Pair\"]\n",
    "        \n",
    "        if pd.isna(row[pathway_col]) or pathway_value in [\"nan\", \"none\", \"\", \"unknown\"]:\n",
    "            label = \"ask Perplexity\"\n",
    "            query = default_query_template.format(pair=pair)\n",
    "            output =  f'<a href=\"https://www.perplexity.ai/search?q={query}\" target=\"_blank\">{label}</a>'\n",
    "        else:\n",
    "            label = row[pathway_col]\n",
    "            query = f\"What-is-the-role-of-the-ligand-and-receptor-pair-{pair}-in-{label}\"\n",
    "            output = f'{label} (see <a href=\"https://www.perplexity.ai/search?q={query}\" target=\"_blank\">evidence in Perplexity</a>)'\n",
    "        \n",
    "        return output\n",
    "    \n",
    "    df[pathway_col] = df.apply(create_link, axis=1)\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "3cd0c7d8-1231-4421-999d-7fcce7fd8460",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_geneToolTip(species):\n",
    "    def tooltip_html(symbol, name):\n",
    "        return (\n",
    "            f'<span class=\"tooltip\">{symbol}'\n",
    "            f'<span class=\"tooltiptext\">{name}</span></span>'\n",
    "        )\n",
    "\n",
    "    gene_pair[species + \" Ligand\"] = [\n",
    "        tooltip_html(ligand_symbol, ligand_name)\n",
    "        for ligand_name, ligand_symbol in zip(gene_pair[species + \" Ligand Name\"], gene_pair[species + \" Ligand\"])\n",
    "    ]\n",
    "    gene_pair[species + \" Receptor\"] = [\n",
    "        tooltip_html(receptor_symbol, receptor_name)\n",
    "        for receptor_name, receptor_symbol in zip(gene_pair[species + \" Receptor Name\"], gene_pair[species + \" Receptor\"])\n",
    "    ]\n",
    "\n",
    "### Remove tooltip for name for each species for now as only zebrafish has the proper names ###     \n",
    "# speciesPrime_list = [\"Zebrafish\"]\n",
    "# # Loop through each species and update gene_pair\n",
    "# for species in speciesPrime_list:\n",
    "#    gene_pair = add_geneToolTip(species)\n",
    "\n",
    "mouse_columns = ['Mouse Ligand', 'Mouse Receptor','Ligand MGI ID','Receptor MGI ID'] \n",
    "rat_columns = ['Rat Ligand','Rat Receptor','Ligand RGD ID','Receptor RGD ID']\n",
    "zebrafish_columns = ['Zebrafish Ligand','Zebrafish Receptor','Ligand ZFIN ID','Receptor ZFIN ID']\n",
    "\n",
    "# List of prefixes\n",
    "prefixes = (\"Chimpanzee\", \"Chicken\", \"Pig\", \"Cow\", \"Dog\", \"Horse\", \"Sheep\", \"Marmoset\", \"Macaque\", \"Frog\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "699c061c-29d8-49a3-a26d-ae1625d67c25",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"None of [Index(['LR Pair Card', 'Human LR Pair', 'Ligand', 'Receptor', 'Ligand Symbols',\\n       'Receptor Symbols', 'Ligand Location', 'Receptor Location',\\n       'Ligand HGNC ID', 'Receptor HGNC ID', 'Perplexity', 'Human evidence',\\n       'PMID'],\\n      dtype='object')] are in the [columns]\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[109], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m gene_pair0 \u001b[38;5;241m=\u001b[39m \u001b[43mgene_pair\u001b[49m\u001b[43m[\u001b[49m\u001b[43mfirst_columns\u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mPMID\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m      2\u001b[0m gene_pair \u001b[38;5;241m=\u001b[39m gene_pair[first_columns]\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/frame.py:4108\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   4106\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m is_iterator(key):\n\u001b[1;32m   4107\u001b[0m         key \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(key)\n\u001b[0;32m-> 4108\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_indexer_strict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcolumns\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m[\u001b[38;5;241m1\u001b[39m]\n\u001b[1;32m   4110\u001b[0m \u001b[38;5;66;03m# take() does not accept boolean indexers\u001b[39;00m\n\u001b[1;32m   4111\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(indexer, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdtype\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m) \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mbool\u001b[39m:\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/indexes/base.py:6200\u001b[0m, in \u001b[0;36mIndex._get_indexer_strict\u001b[0;34m(self, key, axis_name)\u001b[0m\n\u001b[1;32m   6197\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   6198\u001b[0m     keyarr, indexer, new_indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reindex_non_unique(keyarr)\n\u001b[0;32m-> 6200\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_raise_if_missing\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkeyarr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis_name\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   6202\u001b[0m keyarr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtake(indexer)\n\u001b[1;32m   6203\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(key, Index):\n\u001b[1;32m   6204\u001b[0m     \u001b[38;5;66;03m# GH 42790 - Preserve name from an Index\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/indexes/base.py:6249\u001b[0m, in \u001b[0;36mIndex._raise_if_missing\u001b[0;34m(self, key, indexer, axis_name)\u001b[0m\n\u001b[1;32m   6247\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m nmissing:\n\u001b[1;32m   6248\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m nmissing \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mlen\u001b[39m(indexer):\n\u001b[0;32m-> 6249\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNone of [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mkey\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m] are in the [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00maxis_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m]\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   6251\u001b[0m     not_found \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(ensure_index(key)[missing_mask\u001b[38;5;241m.\u001b[39mnonzero()[\u001b[38;5;241m0\u001b[39m]]\u001b[38;5;241m.\u001b[39munique())\n\u001b[1;32m   6252\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mnot_found\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m not in index\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mKeyError\u001b[0m: \"None of [Index(['LR Pair Card', 'Human LR Pair', 'Ligand', 'Receptor', 'Ligand Symbols',\\n       'Receptor Symbols', 'Ligand Location', 'Receptor Location',\\n       'Ligand HGNC ID', 'Receptor HGNC ID', 'Perplexity', 'Human evidence',\\n       'PMID'],\\n      dtype='object')] are in the [columns]\""
     ]
    }
   ],
   "source": [
    "gene_pair0 = gene_pair[first_columns+[\"PMID\"]]\n",
    "gene_pair = gene_pair[first_columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f385074e-5b84-4245-bc5b-5a0ccf169a1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Quick check if there is mouse-specific\n",
    "gene_pair['Ligand'] = gene_pair.apply(\n",
    "    lambda row: \"no human ortholog\" if is_mouse_specific(row['Ligand']) else row['Ligand'],\n",
    "    axis=1\n",
    ")\n",
    "gene_pair['Receptor'] = gene_pair.apply(\n",
    "    lambda row: \"no human ortholog\" if is_mouse_specific(row['Receptor']) else row['Receptor'],\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "a2baf05c-c735-43a6-9e80-501b1a7e1850",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene symbol\n",
    "gene_pair[\"Ligand\"] = [\n",
    "    f'<span title=\"{ligand_name}\">{ligand_symbol}</span>'\n",
    "    for ligand_name, ligand_symbol in zip(gene_pair[\"Ligand Name\"], \n",
    "                                              gene_pair[\"Ligand\"])\n",
    "]\n",
    "# gene symbol\n",
    "gene_pair[\"Receptor\"] = [\n",
    "    f'<span title=\"{receptor_name}\">{receptor_symbol}</span>'\n",
    "    for receptor_name, receptor_symbol in zip(gene_pair[\"Receptor Name\"], \n",
    "                                              gene_pair[\"Receptor\"])\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "ff7f96ec-4611-4be6-baf3-c0bf7fb426e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair = gene_pair.drop(columns=[\"Ligand Name\", \"Receptor Name\"])\n",
    "\n",
    "\n",
    "# Create the links to the HTML cards\n",
    "gene_pair[\"Human LR Pair\"] = [\n",
    "    f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/cards/{lrPairOrig.replace(\" \",\"-\")}.html\">{lrPair}</a>'\n",
    "    for lrPairOrig, lrPair in zip(gene_pair0[\"Human LR Pair\"], gene_pair[\"Human LR Pair\"])\n",
    "]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Add tooltips to the column headers\n",
    "gene_pair.columns = [\n",
    "    f'<span title=\"Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\">{col}</span>' if col == \"Human LR Pair\" else\n",
    "    f'<span title=\"Click the logo below to run Perplexity on the Human LR pair\">{col}&nbsp;</span>' if col == \"Perplexity\" else\n",
    "    f'<span title=\"Official Gene Symbol; Hover on symbols below to show gene names\">{col}&nbsp;&nbsp;&nbsp;</span>' if col in [\"Ligand\", \"Receptor\"] else\n",
    "    f'<span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">{col}&nbsp;&nbsp;</span>' if col in [\"Ligand HGNC ID\", \"Receptor HGNC ID\"] else\n",
    "    f'<span title=\" PubMed IDs (PMID) with Literature Evidence for LR Interaction. Click on the link for more details\">{col}</span>' if col == \"PMID\" else\n",
    "    f'<span title=\"Rat Genome Database (RGD) ID. Click on the link for more details\">{col}</span>' if col in [\"Ligand RGD ID\", \"Receptor RGD ID\"] else\n",
    "    f'<span title=\"Mouse Genome Informatics (MGI) ID. Click on the link for more details\">{col}</span>' if col in [\"Ligand MGI ID\", \"Receptor MGI ID\"]else\n",
    "    f'<span title=\"Zebrafish Information Network (ZFIN) ID. Click on the link for more details\">{col}</span>' if col in [\"Ligand ZFIN ID\", \"Receptor ZFIN ID\"] else\n",
    "    f'<span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\">{col}</span>' if col in [\"Ligand Location\", \"Receptor Location\"] else\n",
    "    f'<span title=\"Double-click header of {col} to ensure all values are shown\">{col}&nbsp;</span>'\n",
    "    for col in gene_pair.columns\n",
    "]\n",
    "\n",
    "gene_pair = gene_pair.reset_index(drop=True)  # Remove the index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "e4e2a5a4-abd6-4366-a8c7-04abc3df4a71",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair000 = gene_pair.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "1aea223c-b1bd-4fca-8d00-51c931bcc84d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['<span title=\"Double-click header of LR Pair Card to ensure all values are shown\">LR Pair Card&nbsp;</span>',\n",
       " '<span title=\"Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\">Human LR Pair</span>',\n",
       " '<span title=\"Official Gene Symbol; Hover on symbols below to show gene names\"\">Human Ligand&nbsp;&nbsp;&nbsp;</span</span>',\n",
       " '<span title=\"Official Gene Symbol; Hover on symbols below to show gene names\"\">Human Receptor&nbsp;&nbsp;&nbsp;</span</span>',\n",
       " '<span title=\"Double-click header of Ligand Symbols to ensure all values are shown\"\">Human Ligand Symbols&nbsp;</span</span>',\n",
       " '<span title=\"Double-click header of Receptor Symbols to ensure all values are shown\"\">Human Receptor Symbols&nbsp;</span</span>',\n",
       " '<span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\">Ligand Location</span>',\n",
       " '<span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\">Receptor Location</span>',\n",
       " '<span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Ligand HGNC ID&nbsp;&nbsp;</span>',\n",
       " '<span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Receptor HGNC ID&nbsp;&nbsp;</span>',\n",
       " '<span title=\"Click the logo below to run Perplexity on the Human LR pair\">Perplexity&nbsp;</span>',\n",
       " '<span title=\"Double-click header of Human evidence to ensure all values are shown\">Human evidence&nbsp;</span>',\n",
       " '<span title=\"Double-click header of LR Pair to ensure all values are shown\">LR Pair&nbsp;</span>',\n",
       " '<span title=\"Double-click header of Ligand ENSEMBL ID to ensure all values are shown\">Ligand ENSEMBL ID&nbsp;</span>',\n",
       " '<span title=\"Double-click header of Receptor ENSEMBL ID to ensure all values are shown\">Receptor ENSEMBL ID&nbsp;</span>',\n",
       " '<span title=\"Double-click header of Ligand Aliases to ensure all values are shown\">Ligand Aliases&nbsp;</span>',\n",
       " '<span title=\"Double-click header of Ligand Old symbol to ensure all values are shown\">Ligand Old symbol&nbsp;</span>',\n",
       " '<span title=\"Double-click header of Receptor Aliases to ensure all values are shown\">Receptor Aliases&nbsp;</span>',\n",
       " '<span title=\"Double-click header of Approved symbol to ensure all values are shown\">Approved symbol&nbsp;</span>',\n",
       " '<span title=\"Double-click header of Receptor Old symbol to ensure all values are shown\">Receptor Old symbol&nbsp;</span>',\n",
       " '<span title=\" PubMed IDs (PMID) with Literature Evidence for LR Interaction. Click on the link for more details\">PMID</span>',\n",
       " '<span title=\"Double-click header of Pair_species to ensure all values are shown\">Pair_species&nbsp;</span>',\n",
       " '<span title=\"Double-click header of lig_species to ensure all values are shown\">lig_species&nbsp;</span>',\n",
       " '<span title=\"Double-click header of rec_species to ensure all values are shown\">rec_species&nbsp;</span>',\n",
       " '<span title=\"Double-click header of ligand_orig to ensure all values are shown\">ligand_orig&nbsp;</span>',\n",
       " '<span title=\"Double-click header of receptor_orig to ensure all values are shown\">receptor_orig&nbsp;</span>',\n",
       " '<span title=\"Double-click header of Interaction ID to ensure all values are shown\">Interaction ID&nbsp;</span>',\n",
       " '<span title=\"Double-click header of KEGG Pathway to ensure all values are shown\">KEGG Pathway&nbsp;</span>']"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "keywords_to_modify = [\"Ligand\", \"Receptor\"]\n",
    "exclude_keywords = [\"HGNC ID\", \"Location\", \"Human\"]  # Columns containing this will not be modified\n",
    "\n",
    "# Copy the original columns so we can modify only the first 10\n",
    "new_columns = gene_pair000.columns.tolist()\n",
    "\n",
    "# Modify only the first 10 columns\n",
    "new_columns[:10] = [\n",
    "    f'{col.split(\">\")[0]}\">Human {col.split(\">\")[1]}</span>'\n",
    "    if any(keyword in col for keyword in keywords_to_modify) and not any(exclude in col for exclude in exclude_keywords)\n",
    "    else col\n",
    "    for col in new_columns[:10]\n",
    "]\n",
    "new_columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "a3e23cc9-c4bf-46e2-abcb-2579a236b22a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This is crucial for consistent handling by the function before processing \"N/A\".\n",
    "gene_pair['Receptor'] = gene_pair['Receptor'].fillna('')\n",
    "gene_pair['Receptor Old symbol'] = gene_pair['Receptor Old symbol'].fillna('')\n",
    "gene_pair['Receptor Aliases'] = gene_pair['Receptor Aliases'].fillna('')\n",
    "\n",
    "gene_pair['Receptor Symbols'] = gene_pair.apply(\n",
    "    lambda row: \"no human ortholog\" if is_mouse_specific(row['Receptor']) \n",
    "                else format_symbol_aliases(row['Receptor'], row['Receptor Old symbol'], row['Receptor Aliases']),\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "### tooltips \n",
    "gene_pair[\"Ligand Symbols\"] = [\n",
    "    f'<span title=\"{aliases}\">{aliases}</span>'\n",
    "    for aliases in gene_pair[\"Ligand Symbols\"]\n",
    "]\n",
    "gene_pair[\"Receptor Symbols\"] = [\n",
    "    f'<span title=\"{aliases}\">{aliases}</span>'\n",
    "    for aliases in gene_pair[\"Receptor Symbols\"]\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "3abb9922-3191-4458-9f45-61ef3de28cf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop columns where all values are NA in gene_pair\n",
    "gene_pair = gene_pair.dropna(axis=1, how='all')\n",
    "\n",
    "gene_pair = gene_pair.fillna(\" \")\n",
    "gene_pair = gene_pair[gene_pair['Human LR Pair'] != ' ']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "13ab373e-7366-448b-8c61-9bab29362892",
   "metadata": {},
   "outputs": [],
   "source": [
    "mask = gene_pair['Ligand HGNC ID'].astype(str).str.strip() == ''\n",
    "gene_pair.loc[mask, 'Mouse Ligand'] = gene_pair.loc[mask, 'Ligand']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "3112ad8e-a04f-49a0-9802-7044ae670977",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Ccl21C Cxcr3', 'Ccl21C CXCR3', 'CD99 Pilrb1', 'Defb2 Ccr6',\n",
       "       'Defb2 Tlr4', 'Fcna Tlr4', 'H60A Klrk1', 'H60B Klrk1',\n",
       "       'H60C Klrk1', 'Lair1 Lilrb4A', 'Pcdhb11 Pcdhb11',\n",
       "       'Pcdhb14 Pcdhb14', 'Pcdhb8 Pcdhb8', 'Pcdhgb8 Pcdhgb8',\n",
       "       'Sema4A Timd2', 'Zp3 Zp3R'], dtype=object)"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grab_mouse_info = gene_pair[\"LR Pair Card\"][gene_pair[\"Human evidence\"] == \"absent in human\"]\n",
    "grab_mouse_info = grab_mouse_info.unique()\n",
    "grab_mouse_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "4e22e891-3101-48fe-8f70-86e0291dd48c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LR Pair</th>\n",
       "      <th>PMID</th>\n",
       "      <th>lig_species</th>\n",
       "      <th>rec_species</th>\n",
       "      <th>ligand_orig</th>\n",
       "      <th>receptor_orig</th>\n",
       "      <th>Pair_species</th>\n",
       "      <th>Human evidence</th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>Ligand ENSEMBL ID</th>\n",
       "      <th>...</th>\n",
       "      <th>Ligand Name</th>\n",
       "      <th>Ligand Aliases</th>\n",
       "      <th>Ligand Old symbol</th>\n",
       "      <th>Receptor Name</th>\n",
       "      <th>Receptor Aliases</th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>Receptor Old symbol</th>\n",
       "      <th>Ligand Symbols</th>\n",
       "      <th>Receptor Symbols</th>\n",
       "      <th>Mouse Ligand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A Atrn</td>\n",
       "      <td>11137996</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>A</td>\n",
       "      <td>Atrn</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>ASIP ATRN</td>\n",
       "      <td>ENSG00000101440</td>\n",
       "      <td>...</td>\n",
       "      <td>agouti signaling protein</td>\n",
       "      <td>ASP</td>\n",
       "      <td>AGTIL</td>\n",
       "      <td>attractin</td>\n",
       "      <td>DPPT-L, MGCA</td>\n",
       "      <td>ATRN</td>\n",
       "      <td>N/A</td>\n",
       "      <td>&lt;span title=\"ASIP (AGTIL, ASP)\"&gt;ASIP (AGTIL, A...</td>\n",
       "      <td>&lt;span title=\"ATRN (DPPT-L, MGCA)\"&gt;ATRN (DPPT-L...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>12194978</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>A2M</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>...</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>FWP007, S863-7, CPAMD5</td>\n",
       "      <td>N/A</td>\n",
       "      <td>heat shock protein family A (Hsp70) member 5</td>\n",
       "      <td>BiP</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>GRP78</td>\n",
       "      <td>&lt;span title=\"A2M (FWP007, S863-7, CPAMD5)\"&gt;A2M...</td>\n",
       "      <td>&lt;span title=\"HSPA5 (GRP78, BiP)\"&gt;HSPA5 (GRP78,...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>32541810</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>A2M</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>...</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>FWP007, S863-7, CPAMD5</td>\n",
       "      <td>N/A</td>\n",
       "      <td>heat shock protein family A (Hsp70) member 5</td>\n",
       "      <td>BiP</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>GRP78</td>\n",
       "      <td>&lt;span title=\"A2M (FWP007, S863-7, CPAMD5)\"&gt;A2M...</td>\n",
       "      <td>&lt;span title=\"HSPA5 (GRP78, BiP)\"&gt;HSPA5 (GRP78,...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>1702392</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>A2M</td>\n",
       "      <td>LRP1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>...</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>FWP007, S863-7, CPAMD5</td>\n",
       "      <td>N/A</td>\n",
       "      <td>LDL receptor related protein 1</td>\n",
       "      <td>LRP, CD91, LRP1A, APOER, IGFBP3R1, IGFBP-3R</td>\n",
       "      <td>LRP1</td>\n",
       "      <td>APR, A2MR</td>\n",
       "      <td>&lt;span title=\"A2M (FWP007, S863-7, CPAMD5)\"&gt;A2M...</td>\n",
       "      <td>&lt;span title=\"LRP1 (APR, A2MR, LRP, CD91, LRP1A...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>10652313</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>A2M</td>\n",
       "      <td>LRP1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>...</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>FWP007, S863-7, CPAMD5</td>\n",
       "      <td>N/A</td>\n",
       "      <td>LDL receptor related protein 1</td>\n",
       "      <td>LRP, CD91, LRP1A, APOER, IGFBP3R1, IGFBP-3R</td>\n",
       "      <td>LRP1</td>\n",
       "      <td>APR, A2MR</td>\n",
       "      <td>&lt;span title=\"A2M (FWP007, S863-7, CPAMD5)\"&gt;A2M...</td>\n",
       "      <td>&lt;span title=\"LRP1 (APR, A2MR, LRP, CD91, LRP1A...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6121</th>\n",
       "      <td>ZG16B TLR5</td>\n",
       "      <td>20802527</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ZG16B</td>\n",
       "      <td>TLR5</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>ZG16B TLR5</td>\n",
       "      <td>ENSG00000283056</td>\n",
       "      <td>...</td>\n",
       "      <td>zymogen granule protein 16B</td>\n",
       "      <td>HRPE773, PRO1567, JCLN2</td>\n",
       "      <td>N/A</td>\n",
       "      <td>toll like receptor 5</td>\n",
       "      <td>TIL3, FLJ10052, MGC126430, MGC126431</td>\n",
       "      <td>TLR5</td>\n",
       "      <td>SLEB1</td>\n",
       "      <td>&lt;span title=\"ZG16B (HRPE773, PRO1567, JCLN2)\"&gt;...</td>\n",
       "      <td>&lt;span title=\"TLR5 (SLEB1, TIL3, FLJ10052, MGC1...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6122</th>\n",
       "      <td>ZG16B TLR6</td>\n",
       "      <td>20802527</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ZG16B</td>\n",
       "      <td>TLR6</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>ZG16B TLR6</td>\n",
       "      <td>ENSG00000283056</td>\n",
       "      <td>...</td>\n",
       "      <td>zymogen granule protein 16B</td>\n",
       "      <td>HRPE773, PRO1567, JCLN2</td>\n",
       "      <td>N/A</td>\n",
       "      <td>toll like receptor 6</td>\n",
       "      <td>CD286</td>\n",
       "      <td>TLR6</td>\n",
       "      <td>N/A</td>\n",
       "      <td>&lt;span title=\"ZG16B (HRPE773, PRO1567, JCLN2)\"&gt;...</td>\n",
       "      <td>&lt;span title=\"TLR6 (CD286)\"&gt;TLR6 (CD286)&lt;/span&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6123</th>\n",
       "      <td>ZG16B Tlr6</td>\n",
       "      <td>20802527</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>ZG16B</td>\n",
       "      <td>Tlr6</td>\n",
       "      <td>mixed</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>ZG16B TLR6</td>\n",
       "      <td>ENSG00000283056</td>\n",
       "      <td>...</td>\n",
       "      <td>zymogen granule protein 16B</td>\n",
       "      <td>HRPE773, PRO1567, JCLN2</td>\n",
       "      <td>N/A</td>\n",
       "      <td>toll like receptor 6</td>\n",
       "      <td>CD286</td>\n",
       "      <td>TLR6</td>\n",
       "      <td>N/A</td>\n",
       "      <td>&lt;span title=\"ZG16B (HRPE773, PRO1567, JCLN2)\"&gt;...</td>\n",
       "      <td>&lt;span title=\"TLR6 (CD286)\"&gt;TLR6 (CD286)&lt;/span&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6124</th>\n",
       "      <td>Zp3 Chrna7</td>\n",
       "      <td>22577141</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Zp3</td>\n",
       "      <td>Chrna7</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>CONSERVATION</td>\n",
       "      <td>ZP3 CHRNA7</td>\n",
       "      <td>ENSG00000188372</td>\n",
       "      <td>...</td>\n",
       "      <td>zona pellucida glycoprotein 3</td>\n",
       "      <td>ZP3-424, ZP3-372, ZPC</td>\n",
       "      <td>ZP3A, ZP3B</td>\n",
       "      <td>cholinergic receptor nicotinic alpha 7 subunit</td>\n",
       "      <td>N/A</td>\n",
       "      <td>CHRNA7</td>\n",
       "      <td>N/A</td>\n",
       "      <td>&lt;span title=\"ZP3 (ZP3A, ZP3B, ZP3-424, ZP3-372...</td>\n",
       "      <td>&lt;span title=\"CHRNA7\"&gt;CHRNA7&lt;/span&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6125</th>\n",
       "      <td>Zp3 Zp3R</td>\n",
       "      <td>8188752</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Zp3</td>\n",
       "      <td>Zp3R</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>absent in human</td>\n",
       "      <td>Zp3 Zp3R</td>\n",
       "      <td>ENSG00000188372</td>\n",
       "      <td>...</td>\n",
       "      <td>zona pellucida glycoprotein 3</td>\n",
       "      <td>ZP3-424, ZP3-372, ZPC</td>\n",
       "      <td>ZP3A, ZP3B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>&lt;span title=\"no human ortholog\"&gt;no human ortho...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6126 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         LR Pair      PMID   lig_species   rec_species ligand_orig  \\\n",
       "0         A Atrn  11137996  Mus Musculus  Mus Musculus           A   \n",
       "1      A2M HSPA5  12194978  Homo sapiens  Homo sapiens         A2M   \n",
       "2      A2M HSPA5  32541810  Homo sapiens  Homo sapiens         A2M   \n",
       "3       A2M LRP1   1702392  Homo sapiens  Homo sapiens         A2M   \n",
       "4       A2M LRP1  10652313  Homo sapiens  Homo sapiens         A2M   \n",
       "...          ...       ...           ...           ...         ...   \n",
       "6121  ZG16B TLR5  20802527  Homo sapiens  Homo sapiens       ZG16B   \n",
       "6122  ZG16B TLR6  20802527  Homo sapiens  Homo sapiens       ZG16B   \n",
       "6123  ZG16B Tlr6  20802527  Homo sapiens  Mus Musculus       ZG16B   \n",
       "6124  Zp3 Chrna7  22577141  Mus Musculus  Mus Musculus         Zp3   \n",
       "6125    Zp3 Zp3R   8188752  Mus Musculus  Mus Musculus         Zp3   \n",
       "\n",
       "     receptor_orig  Pair_species   Human evidence LR Pair Card  \\\n",
       "0             Atrn  Mus Musculus     CONSERVATION    ASIP ATRN   \n",
       "1            HSPA5  Homo sapiens           DIRECT    A2M HSPA5   \n",
       "2            HSPA5  Homo sapiens           DIRECT    A2M HSPA5   \n",
       "3             LRP1  Homo sapiens           DIRECT     A2M LRP1   \n",
       "4             LRP1  Homo sapiens           DIRECT     A2M LRP1   \n",
       "...            ...           ...              ...          ...   \n",
       "6121          TLR5  Homo sapiens           DIRECT   ZG16B TLR5   \n",
       "6122          TLR6  Homo sapiens           DIRECT   ZG16B TLR6   \n",
       "6123          Tlr6         mixed     CONSERVATION   ZG16B TLR6   \n",
       "6124        Chrna7  Mus Musculus     CONSERVATION   ZP3 CHRNA7   \n",
       "6125          Zp3R  Mus Musculus  absent in human     Zp3 Zp3R   \n",
       "\n",
       "     Ligand ENSEMBL ID  ...                    Ligand Name  \\\n",
       "0      ENSG00000101440  ...       agouti signaling protein   \n",
       "1      ENSG00000175899  ...          alpha-2-macroglobulin   \n",
       "2      ENSG00000175899  ...          alpha-2-macroglobulin   \n",
       "3      ENSG00000175899  ...          alpha-2-macroglobulin   \n",
       "4      ENSG00000175899  ...          alpha-2-macroglobulin   \n",
       "...                ...  ...                            ...   \n",
       "6121   ENSG00000283056  ...    zymogen granule protein 16B   \n",
       "6122   ENSG00000283056  ...    zymogen granule protein 16B   \n",
       "6123   ENSG00000283056  ...    zymogen granule protein 16B   \n",
       "6124   ENSG00000188372  ...  zona pellucida glycoprotein 3   \n",
       "6125   ENSG00000188372  ...  zona pellucida glycoprotein 3   \n",
       "\n",
       "               Ligand Aliases Ligand Old symbol  \\\n",
       "0                         ASP             AGTIL   \n",
       "1      FWP007, S863-7, CPAMD5               N/A   \n",
       "2      FWP007, S863-7, CPAMD5               N/A   \n",
       "3      FWP007, S863-7, CPAMD5               N/A   \n",
       "4      FWP007, S863-7, CPAMD5               N/A   \n",
       "...                       ...               ...   \n",
       "6121  HRPE773, PRO1567, JCLN2               N/A   \n",
       "6122  HRPE773, PRO1567, JCLN2               N/A   \n",
       "6123  HRPE773, PRO1567, JCLN2               N/A   \n",
       "6124    ZP3-424, ZP3-372, ZPC        ZP3A, ZP3B   \n",
       "6125    ZP3-424, ZP3-372, ZPC        ZP3A, ZP3B   \n",
       "\n",
       "                                       Receptor Name  \\\n",
       "0                                          attractin   \n",
       "1       heat shock protein family A (Hsp70) member 5   \n",
       "2       heat shock protein family A (Hsp70) member 5   \n",
       "3                     LDL receptor related protein 1   \n",
       "4                     LDL receptor related protein 1   \n",
       "...                                              ...   \n",
       "6121                            toll like receptor 5   \n",
       "6122                            toll like receptor 6   \n",
       "6123                            toll like receptor 6   \n",
       "6124  cholinergic receptor nicotinic alpha 7 subunit   \n",
       "6125                                                   \n",
       "\n",
       "                                 Receptor Aliases Approved symbol  \\\n",
       "0                                    DPPT-L, MGCA            ATRN   \n",
       "1                                             BiP           HSPA5   \n",
       "2                                             BiP           HSPA5   \n",
       "3     LRP, CD91, LRP1A, APOER, IGFBP3R1, IGFBP-3R            LRP1   \n",
       "4     LRP, CD91, LRP1A, APOER, IGFBP3R1, IGFBP-3R            LRP1   \n",
       "...                                           ...             ...   \n",
       "6121         TIL3, FLJ10052, MGC126430, MGC126431            TLR5   \n",
       "6122                                        CD286            TLR6   \n",
       "6123                                        CD286            TLR6   \n",
       "6124                                          N/A          CHRNA7   \n",
       "6125                                                                \n",
       "\n",
       "     Receptor Old symbol                                     Ligand Symbols  \\\n",
       "0                    N/A  <span title=\"ASIP (AGTIL, ASP)\">ASIP (AGTIL, A...   \n",
       "1                  GRP78  <span title=\"A2M (FWP007, S863-7, CPAMD5)\">A2M...   \n",
       "2                  GRP78  <span title=\"A2M (FWP007, S863-7, CPAMD5)\">A2M...   \n",
       "3              APR, A2MR  <span title=\"A2M (FWP007, S863-7, CPAMD5)\">A2M...   \n",
       "4              APR, A2MR  <span title=\"A2M (FWP007, S863-7, CPAMD5)\">A2M...   \n",
       "...                  ...                                                ...   \n",
       "6121               SLEB1  <span title=\"ZG16B (HRPE773, PRO1567, JCLN2)\">...   \n",
       "6122                 N/A  <span title=\"ZG16B (HRPE773, PRO1567, JCLN2)\">...   \n",
       "6123                 N/A  <span title=\"ZG16B (HRPE773, PRO1567, JCLN2)\">...   \n",
       "6124                 N/A  <span title=\"ZP3 (ZP3A, ZP3B, ZP3-424, ZP3-372...   \n",
       "6125                      <span title=\"no human ortholog\">no human ortho...   \n",
       "\n",
       "                                       Receptor Symbols Mouse Ligand  \n",
       "0     <span title=\"ATRN (DPPT-L, MGCA)\">ATRN (DPPT-L...          NaN  \n",
       "1     <span title=\"HSPA5 (GRP78, BiP)\">HSPA5 (GRP78,...          NaN  \n",
       "2     <span title=\"HSPA5 (GRP78, BiP)\">HSPA5 (GRP78,...          NaN  \n",
       "3     <span title=\"LRP1 (APR, A2MR, LRP, CD91, LRP1A...          NaN  \n",
       "4     <span title=\"LRP1 (APR, A2MR, LRP, CD91, LRP1A...          NaN  \n",
       "...                                                 ...          ...  \n",
       "6121  <span title=\"TLR5 (SLEB1, TIL3, FLJ10052, MGC1...          NaN  \n",
       "6122     <span title=\"TLR6 (CD286)\">TLR6 (CD286)</span>          NaN  \n",
       "6123     <span title=\"TLR6 (CD286)\">TLR6 (CD286)</span>          NaN  \n",
       "6124                 <span title=\"CHRNA7\">CHRNA7</span>          NaN  \n",
       "6125  <span title=\"no human ortholog\">no human ortho...          NaN  \n",
       "\n",
       "[6126 rows x 28 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fb361d56-d4ec-4167-96ef-400feabad16a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "These 29 titles that have to be manually checked -- possible incomplete titles\n",
      "[28280243.0, 28360196.0, 28646018.0, 28844702.0, 29549127.0, 29741477.0, 30262652.0, 30659054.0, 31300520.0, 31664130.0, 31987794.0, 32541810.0, 32681389.0, 32876567.0, 33497493.0, 33526813.0, 33790888.0, 33962943.0, 33975953.0, 35132089.0, 35605991.0, 35617401.0, 35711472.0, 36790376.0, 37540598.0, 38628398.0, 38758807.0, 38902261.0, nan]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_32608/1870216621.py:125: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  pmids_without_period = df[df['Title'].isna() | ~df['Title'].str.endswith(('.', '?')).fillna(False)]['PMID']\n"
     ]
    }
   ],
   "source": [
    "def fetch_pubmed_data(pmid_list):\n",
    "    base_url = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\"\n",
    "    results = []\n",
    "\n",
    "    # Load existing data if output file exists\n",
    "    if os.path.exists(output_file):\n",
    "        existing_data = pd.read_csv(output_file)\n",
    "    else:\n",
    "        existing_data = pd.DataFrame(columns=[\"PMID\", \"Title\", \"Abstract\", \"Journal\", \"Year\"])\n",
    "\n",
    "    # Split PMIDs into batches\n",
    "    batch_size = 50\n",
    "    pmid_batches = [pmid_list[i:i + batch_size] for i in range(0, len(pmid_list), batch_size)]\n",
    "\n",
    "    # Iterate over the batches\n",
    "    for batch in pmid_batches:\n",
    "        params = {\n",
    "            \"db\": \"pubmed\",\n",
    "            \"id\": \",\".join(batch),  # Join PMIDs as comma-separated\n",
    "            \"retmode\": \"xml\",\n",
    "            \"api_key\": ncbi_api_key\n",
    "        }\n",
    "\n",
    "        try:\n",
    "            response = requests.get(base_url, params=params)\n",
    "            response.raise_for_status()\n",
    "\n",
    "            # Parse the XML response\n",
    "            root = ET.fromstring(response.text)\n",
    "            for article in root.findall(\".//PubmedArticle\"):\n",
    "                # Extract Title and Abstract\n",
    "                title = article.findtext(\".//ArticleTitle\", default=\"N/A\")\n",
    "                abstract = article.findtext(\".//AbstractText\", default=\"No abstract available\")\n",
    "\n",
    "                # Extract Journal Title\n",
    "                journal_tag = article.find(\".//Journal/Title\")\n",
    "                journal = journal_tag.text.strip() if journal_tag is not None and journal_tag.text else \"N/A\"\n",
    "\n",
    "                # Extract Publication Year\n",
    "                pub_date = article.find(\".//PubDate\")\n",
    "                if pub_date is not None:\n",
    "                    year_tag = pub_date.find(\"Year\")\n",
    "                    year = year_tag.text if year_tag is not None else \"N/A\"\n",
    "\n",
    "                    # Fallback to MedlineDate if Year is missing\n",
    "                    if year == \"N/A\":\n",
    "                        medline_date_tag = pub_date.find(\"MedlineDate\")\n",
    "                        year = medline_date_tag.text.split()[0] if medline_date_tag is not None else \"N/A\"\n",
    "                else:\n",
    "                    year = \"N/A\"  # PubDate is completely missing\n",
    "# \n",
    "                # Initialize species as N/A\n",
    "                # species = \"N/A\"\n",
    "\n",
    "                # # Check if the word \"patient\" is detected in title or abstract (assume human)\n",
    "                # if \"patient\" in title.lower() or \"patient\" in abstract.lower():\n",
    "                #     species = \"Homo sapiens\"\n",
    "                # elif \"human\" in title.lower() or \"human\" in abstract.lower():\n",
    "                #     species = \"Homo sapiens\"\n",
    "                # else:\n",
    "                #     # Look for HGNC gene symbols in title or abstract (assume human if found)\n",
    "                #     for gene in hgnc_symbols:\n",
    "                #         if gene in title or gene in abstract:\n",
    "                #             species = \"Homo sapiens\"\n",
    "                #             break\n",
    "                #     else:\n",
    "                #         # Look for MeSH terms related to species\n",
    "                #         for mesh_heading in article.findall(\".//MeshHeadingList/MeshHeading\"):\n",
    "                #             descriptor_name = mesh_heading.findtext(\"DescriptorName\")\n",
    "                #             if descriptor_name:\n",
    "                #                 # Match official species names using the species_dict\n",
    "                #                 for species_term, scientific_name in species_dict.items():\n",
    "                #                     if species_term in descriptor_name.lower():\n",
    "                #                         species = scientific_name\n",
    "                #                         break  # Stop after finding the first match\n",
    "\n",
    "                # Append the result\n",
    "                results.append({\n",
    "                    \"PMID\": article.findtext(\".//MedlineCitation/PMID\"),\n",
    "                    \"Title\": title,\n",
    "                    \"Abstract\": abstract,\n",
    "                    \"Journal\": journal,\n",
    "                    \"Year\": year,\n",
    "                    #\"Species\": species,\n",
    "                })\n",
    "\n",
    "        except Exception as e:\n",
    "            print(f\"Error fetching batch {batch}: {e}\")\n",
    "            # Optionally save the response for debugging\n",
    "            with open(f\"error_batch_{batch[0]}_{batch[-1]}.xml\", \"w\") as f:\n",
    "                f.write(response.text)\n",
    "\n",
    "        # Rate limiting to avoid API overload\n",
    "        time.sleep(1)  # Increase delay for better API compliance\n",
    "\n",
    "    # Save results\n",
    "    new_data = pd.DataFrame(results)\n",
    "    if not new_data.empty:\n",
    "        # Merge existing and new data, updating missing values\n",
    "        updated_data = pd.concat([existing_data, new_data])\n",
    "\n",
    "        # Ensure all PMIDs are strings\n",
    "        updated_data[\"PMID\"] = updated_data[\"PMID\"].astype(str)\n",
    "\n",
    "        # Drop rows with missing PMIDs\n",
    "        updated_data = updated_data.dropna(subset=[\"PMID\"])\n",
    "\n",
    "        # Ensure rows are ordered and remove duplicates\n",
    "        updated_data = (\n",
    "            updated_data.sort_values(by=\"PMID\")  # Ensure rows are ordered\n",
    "            .drop_duplicates(subset=\"PMID\", keep=\"last\")  # Keep the latest data\n",
    "        )\n",
    "        updated_data[\"Journal\"] = updated_data[\"Journal\"].str.split(\" (\", n=1, expand=False, regex=False).str[0]\n",
    "        updated_data.to_csv(output_file, index=False)\n",
    "    else:\n",
    "        print(\"No new data fetched.\")\n",
    "\n",
    "    return results\n",
    "\n",
    "# Fetch PubMed data with your list of PMIDs, output file path, and NCBI API key\n",
    "fetch_pubmed_data(pmid_list)\n",
    "\n",
    "# Filter and print PMIDs where the Title does not end with a period this are the ones that need to be manually edited as title is not complete.\n",
    "df = pd.read_csv(output_file)\n",
    "pmids_without_period = df[df['Title'].isna() | ~df['Title'].str.endswith(('.', '?')).fillna(False)]['PMID']\n",
    "pmid_check= pmids_without_period.tolist()\n",
    "print(\"These \" + str(len(pmid_check)) + \" titles that have to be manually checked -- possible incomplete titles\")\n",
    "print(pmid_check)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "45796695-557d-4d41-b0d4-41b934b978d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "source = np.array(gene_pair_human[\"PMID\"].unique())\n",
    "\n",
    "# Read the API key from a file\n",
    "with open(\"data/ncbi_api_key.txt\", \"r\") as file:\n",
    "    ncbi_api_key = file.read().strip()\n",
    "\n",
    "# File to save the results\n",
    "output_file = \"data/pubmed_results.csv\"\n",
    "\n",
    "# Load your list of PMIDs\n",
    "pmid_list = source"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f9f7b6f-68af-4746-ad54-0f385499893e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_pubmed_data(pmid_list):\n",
    "    base_url = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\"\n",
    "    results = []\n",
    "\n",
    "    # Load existing data if output file exists\n",
    "    if os.path.exists(output_file):\n",
    "        existing_data = pd.read_csv(output_file)\n",
    "    else:\n",
    "        existing_data = pd.DataFrame(columns=[\"PMID\", \"Title\", \"Abstract\", \"Journal\", \"Year\"])\n",
    "\n",
    "    # Split PMIDs into batches\n",
    "    batch_size = 50\n",
    "    pmid_batches = [pmid_list[i:i + batch_size] for i in range(0, len(pmid_list), batch_size)]\n",
    "\n",
    "    # Iterate over the batches\n",
    "    for batch in pmid_batches:\n",
    "        params = {\n",
    "            \"db\": \"pubmed\",\n",
    "            \"id\": \",\".join(batch),  # Join PMIDs as comma-separated\n",
    "            \"retmode\": \"xml\",\n",
    "            \"api_key\": ncbi_api_key\n",
    "        }\n",
    "\n",
    "        try:\n",
    "            response = requests.get(base_url, params=params)\n",
    "            response.raise_for_status()\n",
    "\n",
    "            # Parse the XML response\n",
    "            root = ET.fromstring(response.text)\n",
    "            for article in root.findall(\".//PubmedArticle\"):\n",
    "                # Extract Title and Abstract\n",
    "                title = article.findtext(\".//ArticleTitle\", default=\"N/A\")\n",
    "                abstract = article.findtext(\".//AbstractText\", default=\"No abstract available\")\n",
    "\n",
    "                # Extract Journal Title\n",
    "                journal_tag = article.find(\".//Journal/Title\")\n",
    "                journal = journal_tag.text.strip() if journal_tag is not None and journal_tag.text else \"N/A\"\n",
    "\n",
    "                # Extract Publication Year\n",
    "                pub_date = article.find(\".//PubDate\")\n",
    "                if pub_date is not None:\n",
    "                    year_tag = pub_date.find(\"Year\")\n",
    "                    year = year_tag.text if year_tag is not None else \"N/A\"\n",
    "\n",
    "                    # Fallback to MedlineDate if Year is missing\n",
    "                    if year == \"N/A\":\n",
    "                        medline_date_tag = pub_date.find(\"MedlineDate\")\n",
    "                        year = medline_date_tag.text.split()[0] if medline_date_tag is not None else \"N/A\"\n",
    "                else:\n",
    "                    year = \"N/A\"  # PubDate is completely missing\n",
    "# \n",
    "                # Initialize species as N/A\n",
    "                # species = \"N/A\"\n",
    "\n",
    "                # # Check if the word \"patient\" is detected in title or abstract (assume human)\n",
    "                # if \"patient\" in title.lower() or \"patient\" in abstract.lower():\n",
    "                #     species = \"Homo sapiens\"\n",
    "                # elif \"human\" in title.lower() or \"human\" in abstract.lower():\n",
    "                #     species = \"Homo sapiens\"\n",
    "                # else:\n",
    "                #     # Look for HGNC gene symbols in title or abstract (assume human if found)\n",
    "                #     for gene in hgnc_symbols:\n",
    "                #         if gene in title or gene in abstract:\n",
    "                #             species = \"Homo sapiens\"\n",
    "                #             break\n",
    "                #     else:\n",
    "                #         # Look for MeSH terms related to species\n",
    "                #         for mesh_heading in article.findall(\".//MeshHeadingList/MeshHeading\"):\n",
    "                #             descriptor_name = mesh_heading.findtext(\"DescriptorName\")\n",
    "                #             if descriptor_name:\n",
    "                #                 # Match official species names using the species_dict\n",
    "                #                 for species_term, scientific_name in species_dict.items():\n",
    "                #                     if species_term in descriptor_name.lower():\n",
    "                #                         species = scientific_name\n",
    "                #                         break  # Stop after finding the first match\n",
    "\n",
    "                # Append the result\n",
    "                results.append({\n",
    "                    \"PMID\": article.findtext(\".//MedlineCitation/PMID\"),\n",
    "                    \"Title\": title,\n",
    "                    \"Abstract\": abstract,\n",
    "                    \"Journal\": journal,\n",
    "                    \"Year\": year,\n",
    "                    #\"Species\": species,\n",
    "                })\n",
    "\n",
    "        except Exception as e:\n",
    "            print(f\"Error fetching batch {batch}: {e}\")\n",
    "            # Optionally save the response for debugging\n",
    "            with open(f\"error_batch_{batch[0]}_{batch[-1]}.xml\", \"w\") as f:\n",
    "                f.write(response.text)\n",
    "\n",
    "        # Rate limiting to avoid API overload\n",
    "        time.sleep(1)  # Increase delay for better API compliance\n",
    "\n",
    "    # Save results\n",
    "    new_data = pd.DataFrame(results)\n",
    "    if not new_data.empty:\n",
    "        # Merge existing and new data, updating missing values\n",
    "        updated_data = pd.concat([existing_data, new_data])\n",
    "\n",
    "        # Ensure all PMIDs are strings\n",
    "        updated_data[\"PMID\"] = updated_data[\"PMID\"].astype(str)\n",
    "\n",
    "        # Drop rows with missing PMIDs\n",
    "        updated_data = updated_data.dropna(subset=[\"PMID\"])\n",
    "\n",
    "        # Ensure rows are ordered and remove duplicates\n",
    "        updated_data = (\n",
    "            updated_data.sort_values(by=\"PMID\")  # Ensure rows are ordered\n",
    "            .drop_duplicates(subset=\"PMID\", keep=\"last\")  # Keep the latest data\n",
    "        )\n",
    "        updated_data[\"Journal\"] = updated_data[\"Journal\"].str.split(\" (\", n=1, expand=False, regex=False).str[0]\n",
    "        updated_data.to_csv(output_file, index=False)\n",
    "    else:\n",
    "        print(\"No new data fetched.\")\n",
    "\n",
    "    return results\n",
    "\n",
    "# Fetch PubMed data with your list of PMIDs, output file path, and NCBI API key\n",
    "fetch_pubmed_data(pmid_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d1c4f09a-1cf4-44c4-aece-48c9fd0dd240",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PMID</th>\n",
       "      <th>Title</th>\n",
       "      <th>Abstract</th>\n",
       "      <th>Journal</th>\n",
       "      <th>Year</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Formate assay in body fluids: application in m...</td>\n",
       "      <td>No abstract available</td>\n",
       "      <td>Biochemical medicine</td>\n",
       "      <td>1975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>Digitoxin metabolism by rat liver microsomes.</td>\n",
       "      <td>No abstract available</td>\n",
       "      <td>Biochemical pharmacology</td>\n",
       "      <td>1975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10024503</td>\n",
       "      <td>Reversal of hyperlipidaemia in apolipoprotein ...</td>\n",
       "      <td>We have shown previously that human apolipopro...</td>\n",
       "      <td>The Biochemical journal</td>\n",
       "      <td>1999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10025398</td>\n",
       "      <td>The integrin alpha v beta 6 binds and activate...</td>\n",
       "      <td>Transforming growth factor beta (TGF beta) fam...</td>\n",
       "      <td>Cell</td>\n",
       "      <td>1999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10037686</td>\n",
       "      <td>Identification of a novel activation-inducible...</td>\n",
       "      <td>Among members of the tumor necrosis factor rec...</td>\n",
       "      <td>The Journal of biological chemistry</td>\n",
       "      <td>1999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3393</th>\n",
       "      <td>9972281</td>\n",
       "      <td>Insulin-like growth factor-I receptor signal t...</td>\n",
       "      <td>The insulin-like growth factor-I receptor (IGF...</td>\n",
       "      <td>Comparative biochemistry and physiology. Part ...</td>\n",
       "      <td>1998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3394</th>\n",
       "      <td>9973222</td>\n",
       "      <td>Activation of neurotrophin-3 receptor TrkC ind...</td>\n",
       "      <td>Elevated expression of the neurotrophin-3 (NT-...</td>\n",
       "      <td>Cancer research</td>\n",
       "      <td>1999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3395</th>\n",
       "      <td>9988678</td>\n",
       "      <td>Decorin is a biological ligand for the epiderm...</td>\n",
       "      <td>Ectopic expression of decorin induces profound...</td>\n",
       "      <td>The Journal of biological chemistry</td>\n",
       "      <td>1999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3396</th>\n",
       "      <td>9988761</td>\n",
       "      <td>Identification of residues within the 727-767 ...</td>\n",
       "      <td>Mapping approaches employing blocking antibodi...</td>\n",
       "      <td>The Journal of biological chemistry</td>\n",
       "      <td>1999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3397</th>\n",
       "      <td>9990065</td>\n",
       "      <td>Prototypic G protein-coupled receptor for the ...</td>\n",
       "      <td>Glucagon-like peptide 2 (GLP-2) is a 33-aa pro...</td>\n",
       "      <td>Proceedings of the National Academy of Science...</td>\n",
       "      <td>1999</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3398 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          PMID                                              Title  \\\n",
       "0            1  Formate assay in body fluids: application in m...   \n",
       "1           10      Digitoxin metabolism by rat liver microsomes.   \n",
       "2     10024503  Reversal of hyperlipidaemia in apolipoprotein ...   \n",
       "3     10025398  The integrin alpha v beta 6 binds and activate...   \n",
       "4     10037686  Identification of a novel activation-inducible...   \n",
       "...        ...                                                ...   \n",
       "3393   9972281  Insulin-like growth factor-I receptor signal t...   \n",
       "3394   9973222  Activation of neurotrophin-3 receptor TrkC ind...   \n",
       "3395   9988678  Decorin is a biological ligand for the epiderm...   \n",
       "3396   9988761  Identification of residues within the 727-767 ...   \n",
       "3397   9990065  Prototypic G protein-coupled receptor for the ...   \n",
       "\n",
       "                                               Abstract  \\\n",
       "0                                 No abstract available   \n",
       "1                                 No abstract available   \n",
       "2     We have shown previously that human apolipopro...   \n",
       "3     Transforming growth factor beta (TGF beta) fam...   \n",
       "4     Among members of the tumor necrosis factor rec...   \n",
       "...                                                 ...   \n",
       "3393  The insulin-like growth factor-I receptor (IGF...   \n",
       "3394  Elevated expression of the neurotrophin-3 (NT-...   \n",
       "3395  Ectopic expression of decorin induces profound...   \n",
       "3396  Mapping approaches employing blocking antibodi...   \n",
       "3397  Glucagon-like peptide 2 (GLP-2) is a 33-aa pro...   \n",
       "\n",
       "                                                Journal  Year  \n",
       "0                                  Biochemical medicine  1975  \n",
       "1                              Biochemical pharmacology  1975  \n",
       "2                               The Biochemical journal  1999  \n",
       "3                                                  Cell  1999  \n",
       "4                   The Journal of biological chemistry  1999  \n",
       "...                                                 ...   ...  \n",
       "3393  Comparative biochemistry and physiology. Part ...  1998  \n",
       "3394                                    Cancer research  1999  \n",
       "3395                The Journal of biological chemistry  1999  \n",
       "3396                The Journal of biological chemistry  1999  \n",
       "3397  Proceedings of the National Academy of Science...  1999  \n",
       "\n",
       "[3398 rows x 5 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c0a43c38-7980-416d-959e-7580c6044b0a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "These 23 titles that have to be manually checked -- possible incomplete titles\n",
      "[28280243, 28360196, 28646018, 29549127, 29741477, 30262652, 30659054, 31300520, 31664130, 31987794, 32541810, 32876567, 33497493, 33526813, 33790888, 33962943, 35132089, 35605991, 35617401, 36790376, 37540598, 38628398, 38902261]\n",
      "These 125 PMIDs have abstracts that might be incomplete (do not end with a period or question mark or other indicators):\n",
      "[24701371, 27378688, 27875312, 27982078, 28280243, 28302677, 28360196, 28364041, 28394331, 28408722, 28465413, 28546512, 28646018, 28696225, 28698550, 28733458, 28783682, 28785723, 28794434, 28846098, 28851741, 28893801, 28939773, 28943410, 28973891, 29059156, 29100055, 29178324, 29180449, 29296932, 29348142, 29467366, 29545933, 29549127, 29717114, 29721382, 29741477, 29742426, 29769720, 29777742, 29853539, 29898920, 29904386, 29925589, 29930766, 30041429, 30139742, 30139743, 30262652, 30632962, 30659054, 30712922, 30713770, 30733680, 30819903, 30854241, 30956130, 31098409, 31175175, 31216173, 31300520, 31367043, 31645726, 31664130, 31857654, 31987794, 32144270, 32149455, 32506691, 32541810, 32694578, 32702237, 32820046, 32876567, 32907880, 32948210, 33306155, 33497493, 33503438, 33526813, 33790888, 33852831, 33875597, 33962943, 34108253, 34163464, 34407556, 34508778, 34518695, 34678058, 34910520, 35132089, 35141051, 35165283, 35253643, 35295855, 35414038, 35596683, 35617401, 35739238, 35802072, 36044575, 36205393, 36790376, 37036990, 37097004, 37168680, 37333140, 37356715, 37435859, 37523551, 37540598, 37554323, 37633268, 38147550, 38203798, 38426726, 38628398, 38781210, 38897996, 38902261, 39128984, 39229119, 39261724, 6172602]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_28570/1528775117.py:3: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  pmids_without_period = df[df['Title'].isna() | ~df['Title'].str.endswith(('.', '?')).fillna(False)]['PMID']\n"
     ]
    }
   ],
   "source": [
    "# Filter and print PMIDs where the Title does not end with a period this are the ones that need to be manually edited as title is not complete.\n",
    "df = pd.read_csv(output_file)\n",
    "pmids_without_period = df[df['Title'].isna() | ~df['Title'].str.endswith(('.', '?')).fillna(False)]['PMID']\n",
    "pmid_check= pmids_without_period.tolist()\n",
    "print(\"These \" + str(len(pmid_check)) + \" titles that have to be manually checked -- possible incomplete titles\")\n",
    "print(pmid_check)\n",
    "\n",
    "# Fill NaN values in 'Abstract' column with an empty string\n",
    "df['Abstract'] = df['Abstract'].fillna('')\n",
    "\n",
    "pmids_without_period = df[~df['Abstract'].str.endswith(('.', '?', 'available', ')', 'Review', '...'))]['PMID'] # Corrected the condition to find abstracts that *do not* end with a period or question mark\n",
    "pmid_check = pmids_without_period.tolist()\n",
    "\n",
    "print(\"These \" + str(len(pmid_check)) + \" PMIDs have abstracts that might be incomplete (do not end with a period or question mark or other indicators):\")\n",
    "print(pmid_check)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "21ccd695-05d2-4215-bb35-398fcf90a187",
   "metadata": {},
   "outputs": [],
   "source": [
    "prev_file = \"data/pubmed_results_20250614.csv\"\n",
    "df = pd.read_csv(prev_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f6eb8025-cb90-4172-bc38-910a2d17bbfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "df =df[df[\"PMID\"].isin([28280243, 28360196, 28646018, 29549127, 29741477, 30262652, 30659054, 31300520, 31664130, 31987794, 32541810, 32876567, 33497493, 33526813, 33790888, 33962943, 35132089, 35605991, 35617401, 36790376, 37540598, 38628398, 38902261])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "2b923763-596d-4c65-9b2c-31570ffd536d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pmid_check = [24701371, 27378688, 27875312, 27982078, 28280243, 28302677, 28360196, 28364041, 28394331, 28408722, 28465413, 28546512, 28646018, 28696225, 28698550, 28733458, 28783682, 28785723, 28794434, 28846098, 28851741, 28893801, 28939773, 28943410, 28973891, 29059156, 29100055, 29178324, 29180449, 29296932, 29348142, 29467366, 29545933, 29549127, 29717114, 29721382, 29741477, 29742426, 29769720, 29777742, 29853539, 29898920, 29904386, 29925589, 29930766, 30041429, 30139742, 30139743, 30262652, 30632962, 30659054, 30712922, 30713770, 30733680, 30819903, 30854241, 30956130, 31098409, 31175175, 31216173, 31300520, 31367043, 31645726, 31664130, 31857654, 31987794, 32144270, 32149455, 32506691, 32541810, 32694578, 32702237, 32820046, 32876567, 32907880, 32948210, 33306155, 33497493, 33503438, 33526813, 33790888, 33852831, 33875597, 33962943, 34108253, 34163464, 34407556, 34508778, 34518695, 34678058, 34910520, 35132089, 35141051, 35165283, 35253643, 35295855, 35414038, 35596683, 35617401, 35739238, 35802072, 36044575, 36205393, 36790376, 37036990, 37097004, 37168680, 37333140, 37356715, 37435859, 37523551, 37540598, 37554323, 37633268, 38147550, 38203798, 38426726, 38628398, 38781210, 38897996, 38902261, 39128984, 39229119, 39261724, 6172602]\n",
    "df =df[df[\"PMID\"].isin(pmid_check)]\n",
    "df.to_csv(\"data/missing_PMID_abstract_info.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8ae5d57d-e469-4c6d-9407-619349ec7885",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['A Atrn', 'A2M HSPA5', 'A2M LRP1', ..., 'ZG16B Tlr6', 'Zp3 Chrna7',\n",
       "       'Zp3 Zp3R'], dtype=object)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair_human[\"LR_pair_orig\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "b304cd7f-fbc6-48f0-9389-113e9a0f7dd5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Written 61464 rows to data/GENES_RAT_20250723_RGD_DB.tsv\n"
     ]
    }
   ],
   "source": [
    "# 1. Define URL and dated filename\n",
    "import csv\n",
    "url = \"https://download.rgd.mcw.edu/data_release/GENES_RAT.txt\"\n",
    "out_file = f\"data/GENES_RAT_{today}_RGD_DB.tsv\"\n",
    "\n",
    "# 2. Fetch raw content as text\n",
    "response = requests.get(url)\n",
    "response.raise_for_status()\n",
    "content = response.text\n",
    "\n",
    "# 3. Split into lines, remove first 102 header lines\n",
    "lines = content.splitlines()[102:]\n",
    "\n",
    "# 4. Parse into rows\n",
    "rows = [line.split(\"\\t\") for line in lines]\n",
    "header = rows[1]\n",
    "data_rows = rows[2:]\n",
    "\n",
    "\n",
    "# 5. Write as proper TSV\n",
    "with open(out_file, \"w\", newline=\"\", encoding=\"utf-8\") as f:\n",
    "    writer = csv.writer(f, delimiter=\"\\t\")\n",
    "    writer.writerow(header)\n",
    "    writer.writerows(data_rows)\n",
    "\n",
    "print(f\"✅ Written {len(rows)} rows to {out_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "96eff553-7434-49e5-a6e2-c2ff66b2835a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Merged file saved: data/Zebrafish_merged_20250723_ZFIN_DB.tsv — shape: (240290, 14)\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "from io import StringIO\n",
    "from datetime import datetime\n",
    "\n",
    "# Define file URLs and your exact header specifications\n",
    "files = {\n",
    "    \"aliases\": {\n",
    "        \"url\": \"https://zfin.org/downloads/aliases.txt\",\n",
    "        \"headers\": [\"Current ZFIN ID\", \"Current Name\", \"Current Symbol\",\n",
    "                    \"Previous Name\", \"SO ID\"],\n",
    "        \"key\": \"Current ZFIN ID\"\n",
    "    },\n",
    "    \"gene\": {\n",
    "        \"url\": \"https://zfin.org/downloads/gene.txt\",\n",
    "        \"headers\": [\"ZFIN ID\", \"SO ID\", \"Symbol\", \"NCBI Gene ID\"],\n",
    "        \"key\": \"ZFIN ID\"\n",
    "    },\n",
    "    \"orthos\": {\n",
    "        \"url\": \"https://zfin.org/downloads/human_orthos.txt\",\n",
    "        \"headers\": [\"ZFIN ID\", \"ZFIN Symbol\", \"ZFIN Name\", \"Human Symbol\",\n",
    "                    \"Human Name\", \"OMIM ID\", \"Gene ID\", \"HGNC ID\",\n",
    "                    \"Evidence\", \"Pub ID\", \"ZFIN Abbreviation Name\",\n",
    "                    \"ECO ID\", \"ECO Term Name\"],\n",
    "        \"key\": \"ZFIN ID\"\n",
    "    }\n",
    "}\n",
    "\n",
    "# 1. Download and load each file with your defined headers\n",
    "dfs = {}\n",
    "for name, params in files.items():\n",
    "    r = requests.get(params[\"url\"])\n",
    "    r.raise_for_status()\n",
    "    df = pd.read_csv(\n",
    "        StringIO(r.text),\n",
    "        sep=\"\\t\",\n",
    "        comment=\"#\",\n",
    "        header=None,\n",
    "        names=params[\"headers\"],\n",
    "        dtype=str\n",
    "    )\n",
    "    # Standardize the key column name\n",
    "    df = df.rename(columns={params[\"key\"]: \"ZFIN_ID\"})\n",
    "    dfs[name] = df\n",
    "\n",
    "# 2. Merge on ZFIN_ID using outer join\n",
    "merged = dfs[\"aliases\"]\n",
    "for name in [\"gene\", \"orthos\"]:\n",
    "    merged = merged.merge(dfs[name], on=\"ZFIN_ID\", how=\"outer\", suffixes=(\"\", f\"_{name}\"))\n",
    "\n",
    "# 3. Drop duplicated columns (identical names and contents)\n",
    "merged = merged.loc[:, ~merged.columns.duplicated()]\n",
    "# Also drop perfectly identical-content duplicates\n",
    "columns = merged.columns\n",
    "to_drop = [\n",
    "    col2 for i, col1 in enumerate(columns)\n",
    "    for col2 in columns[i+1:]\n",
    "    if merged[col1].equals(merged[col2])\n",
    "]\n",
    "merged = merged.drop(columns=to_drop)\n",
    "merged = merged.drop(columns=['SO ID_gene', 'Symbol', 'Evidence','Pub ID', 'ZFIN Symbol', 'ZFIN Name'])\n",
    "merged = merged.drop_duplicates()\n",
    "\n",
    "# 4. Save with today's date\n",
    "today = datetime.now().strftime(\"%Y%m%d\")\n",
    "output = f\"data/Zebrafish_merged_{today}_ZFIN_DB.tsv\"\n",
    "merged.to_csv(output, sep=\"\\t\", index=False, encoding=\"utf-8\")\n",
    "\n",
    "print(f\"✅ Merged file saved: {output} — shape: {merged.shape}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "74223a48-ef5f-4958-a898-477bf03c07e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import fetchGSheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "9e2a9375-269d-4eff-aee0-18e8feda859e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting annotation process...\n",
      "Loading frog UniProt mapping from data/uniprotMapping_frog.csv...\n",
      "Loaded 2901 frog UniProt entries.\n",
      "\n",
      "--- Processing human ---\n",
      "Loading InParanoid data from data/frog_centric_inParanoid_human.csv...\n",
      "Loaded 19461 InParanoid entries for human.\n",
      "Loading human UniProt mapping from data/uniprotMapping_human.csv...\n",
      "Loaded 45583 human UniProt entries.\n",
      "Merging frog gene names...\n",
      "Merging human gene names...\n",
      "Saving merged data to data/frog_centric_inParanoid_human_AnnWithUniProt.csv...\n",
      "Successfully saved 45745 annotated entries for human.\n",
      "\n",
      "Annotation process completed for all specified species.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "from io import StringIO\n",
    "from itertools import product\n",
    "\n",
    "### IMPORTANT ####\n",
    "# Warning: First run \"src/convertOrthUniprotToEnsembl.r\" in R\n",
    "##################\n",
    "\n",
    "# === USER INPUT ===\n",
    "data_dir = \"data\"\n",
    "# original species\n",
    "orig_species_input = \"frog\"\n",
    "# Ortholog species\n",
    "species_input = \"human\"  # Options: \"mouse\", \"zebrafish\", # No sheep in inParanoid\n",
    "\n",
    "# species_to_process = [\"mouse\", \"rat\", \"human\", \"zebrafish\", \"chimpanzee\", \"chicken\", \"pig\", \"cow\", \"dog\", \"horse\", \"marmoset\",   \"macaque\"]\n",
    "\n",
    "# === INTERNAL MAPPINGS ===\n",
    "\n",
    "orig_species_info = {\n",
    "    \"mouse\":         {\"taxid\": \"10090\", \"code\": \"mmusculus\"},\n",
    "    \"rat\":           {\"taxid\": \"10116\", \"code\": \"rnorvegicus\"},\n",
    "    \"zebrafish\":     {\"taxid\": \"7955\",  \"code\": \"drerio\"},\n",
    "    \"chimpanzee\":    {\"taxid\": \"9598\",  \"code\": \"ptroglodytes\"},\n",
    "    \"chicken\":       {\"taxid\": \"9031\",  \"code\": \"ggallus\"},\n",
    "    \"pig\":           {\"taxid\": \"9823\",  \"code\": \"sscrofa\"},\n",
    "    \"cow\":           {\"taxid\": \"9913\",  \"code\": \"btaurus\"},\n",
    "    \"dog\":           {\"taxid\": \"9615\",  \"code\": \"clfamiliaris\"},\n",
    "    \"horse\":         {\"taxid\": \"9796\",  \"code\": \"ecaballus\"},\n",
    "    #\"sheep\":         {\"taxid\": \"9940\",  \"code\": \"oarambouillet\"},\n",
    "    \"marmoset\":      {\"taxid\": \"9483\",  \"code\": \"cjacchus\"},\n",
    "    \"macaque\":       {\"taxid\": \"9544\",  \"code\": \"mmulatta\"},\n",
    "    \"frog\":          {\"taxid\": \"8364\",  \"code\": \"xtropicalis\"},\n",
    "    \"rabbit\":        {\"taxid\": \"9986\",  \"code\": \"ocuniculus\"},\n",
    "    \"guineapig\":     {\"taxid\": \"10141\",  \"code\": \"cporcellus\"},\n",
    "    \"pufferfish\":    {\"taxid\": \"99883\",  \"code\": \"tnigroviridis\"},\n",
    "    \"human\":         {\"taxid\": \"9606\",  \"code\": \"hsapiens\"},\n",
    "}\n",
    "\n",
    "\n",
    "species_info = {\n",
    "    \"mouse\":         {\"taxid\": \"10090\", \"code\": \"mmusculus\"},\n",
    "    \"rat\":           {\"taxid\": \"10116\", \"code\": \"rnorvegicus\"},\n",
    "    \"zebrafish\":     {\"taxid\": \"7955\",  \"code\": \"drerio\"},\n",
    "    \"chimpanzee\":    {\"taxid\": \"9598\",  \"code\": \"ptroglodytes\"},\n",
    "    \"chicken\":       {\"taxid\": \"9031\",  \"code\": \"ggallus\"},\n",
    "    \"pig\":           {\"taxid\": \"9823\",  \"code\": \"sscrofa\"},\n",
    "    \"cow\":           {\"taxid\": \"9913\",  \"code\": \"btaurus\"},\n",
    "    \"dog\":           {\"taxid\": \"9615\",  \"code\": \"clfamiliaris\"},\n",
    "    \"horse\":         {\"taxid\": \"9796\",  \"code\": \"ecaballus\"},\n",
    "    #\"sheep\":         {\"taxid\": \"9940\",  \"code\": \"oarambouillet\"},\n",
    "    \"marmoset\":      {\"taxid\": \"9483\",  \"code\": \"cjacchus\"},\n",
    "    \"macaque\":       {\"taxid\": \"9544\",  \"code\": \"mmulatta\"},\n",
    "    \"frog\":          {\"taxid\": \"8364\",  \"code\": \"xtropicalis\"},\n",
    "    \"rabbit\":        {\"taxid\": \"9986\",  \"code\": \"ocuniculus\"},\n",
    "    \"guineapig\":     {\"taxid\": \"10141\",  \"code\": \"cporcellus\"},\n",
    "    \"pufferfish\":    {\"taxid\": \"99883\",  \"code\": \"tnigroviridis\"},\n",
    "    \"human\":         {\"taxid\": \"9606\",  \"code\": \"hsapiens\"},\n",
    "}\n",
    "\n",
    "if orig_species_input not in orig_species_info:\n",
    "    raise ValueError(f\"Species '{orig_species_input}' not supported. Choose from: {list(orig_species_info)}\")\n",
    "\n",
    "if species_input not in species_info:\n",
    "    raise ValueError(f\"Species '{species_input}' not supported. Choose from: {list(species_info)}\")\n",
    "\n",
    "species = species_info[species_input]\n",
    "taxid = species[\"taxid\"]\n",
    "code = species[\"code\"]\n",
    "\n",
    "orig_species = orig_species_info[orig_species_input]\n",
    "orig_taxid = orig_species[\"taxid\"]\n",
    "orig_code = orig_species[\"code\"]\n",
    "\n",
    "# === Step 1: Download inParanoid file for human vs species ===\n",
    "url = f\"https://inparanoidb.sbc.su.se/download/sqltable/{orig_taxid}&{taxid}&prot\"\n",
    "r = requests.get(url)\n",
    "r.raise_for_status()\n",
    "\n",
    "df = pd.read_csv(StringIO(r.text.strip()), sep=\"\\t\", header=None)\n",
    "df.columns = [\"cluster_id\", \"bitscore\", \"source_file\", \"inparalog_score\", \"protein_id\", \"seed_score\"]\n",
    "\n",
    "# === Step 2: Add species labels ===\n",
    "def infer_species(src):\n",
    "    if orig_taxid in src:\n",
    "        return orig_species_input\n",
    "    elif taxid in src:\n",
    "        return species_input\n",
    "    return \"unknown\"\n",
    "\n",
    "df[\"species\"] = df[\"source_file\"].apply(infer_species)\n",
    "\n",
    "# === Step 3: Build ortholog pairs ===\n",
    "records = []\n",
    "for cid, grp in df.groupby(\"cluster_id\"):\n",
    "    orig_spec = grp[grp[\"species\"] == orig_species_input]\n",
    "    others = grp[grp[\"species\"] == species_input]\n",
    "    \n",
    "    for h, o in product(orig_spec.itertuples(index=False), others.itertuples(index=False)):\n",
    "        records.append({\n",
    "            \"cluster_id\": cid,\n",
    "            f\"{orig_species_input}_protein\": h.protein_id,\n",
    "            f\"{orig_species_input}_inparalog_score\": h.inparalog_score,\n",
    "            f\"{orig_species_input}_seed_score\": h.seed_score,\n",
    "            f\"{species_input}_protein\": o.protein_id,\n",
    "            f\"{species_input}_inparalog_score\": o.inparalog_score,\n",
    "            f\"{species_input}_seed_score\": o.seed_score,\n",
    "            \"bitscore\": (h.bitscore + o.bitscore) / 2\n",
    "        })\n",
    "\n",
    "\n",
    "df_orthologs = pd.DataFrame(records)\n",
    "df_orthologs.to_csv(f\"data/{orig_species_input}_centric_inParanoid_{species_input}.csv\", index=False)\n",
    "\n",
    "# === Step 4: UniProt → Gene Name ===\n",
    "print(\"Starting annotation process...\")\n",
    "\n",
    "# --- Load Human UniProt Mapping once ---\n",
    "orig_uniprot_file = os.path.join(data_dir, f\"uniprotMapping_{orig_species_input}.csv\")\n",
    "if not os.path.exists(orig_uniprot_file):\n",
    "    print(f\"Error: {orig_species_input} UniProt mapping file not found at {orig_uniprot_file}. Please run the UniProt fetching script for human first.\")\n",
    "    exit() # Exit if the essential human file is missing\n",
    "\n",
    "print(f\"Loading {orig_species_input} UniProt mapping from {orig_uniprot_file}...\")\n",
    "# We only need 'Entry' (Accession) and 'Gene Names'\n",
    "orig_uniprot_df = pd.read_csv(orig_uniprot_file, usecols=[\"Entry\", \"Gene Names\"])\n",
    "# Rename 'Gene Names' to distinguish it as human gene name\n",
    "orig_uniprot_df = orig_uniprot_df.rename(columns={\"Gene Names\": f\"{orig_species_input}_Gene_Name\"})\n",
    "print(f\"Loaded {len(orig_uniprot_df)} {orig_species_input} UniProt entries.\")\n",
    "\n",
    "# --- Process each species ---\n",
    "print(f\"\\n--- Processing {species_input} ---\")\n",
    "\n",
    "inparanoid_file = os.path.join(data_dir, f\"{orig_species_input}_centric_inParanoid_{species_input}.csv\")\n",
    "\n",
    "species_uniprot_file = os.path.join(data_dir, f\"uniprotMapping_{species_input}.csv\")\n",
    "output_file = os.path.join(data_dir, f\"{orig_species_input}_centric_inParanoid_{species_input}_AnnWithUniProt.csv\")\n",
    "\n",
    "# Check if input files exist\n",
    "if not os.path.exists(inparanoid_file):\n",
    "    print(f\"Skipping {species_input}: InParanoid file not found at {inparanoid_file}\")\n",
    "    if not os.path.exists(species_uniprot_file):\n",
    "        print(f\"Skipping {species_input}: UniProt mapping file not found at {species_uniprot_file}\")\n",
    "\n",
    "print(f\"Loading InParanoid data from {inparanoid_file}...\")\n",
    "df_inparanoid = pd.read_csv(inparanoid_file)\n",
    "print(f\"Loaded {len(df_inparanoid)} InParanoid entries for {species_input}.\")\n",
    "\n",
    "print(f\"Loading {species_input} UniProt mapping from {species_uniprot_file}...\")\n",
    "species_uniprot_df = pd.read_csv(species_uniprot_file, usecols=[\"Entry\", \"Gene Names\"])\n",
    "# Rename 'Gene Names' to distinguish it as the species' gene name\n",
    "species_uniprot_df = species_uniprot_df.rename(columns={\"Gene Names\": f\"{species_input}_Gene_Name\"})\n",
    "print(f\"Loaded {len(species_uniprot_df)} {species_input} UniProt entries.\")\n",
    "\n",
    "# 1. Annotate based on \"orig_species_protein\" with orig UniProt data\n",
    "print(f\"Merging {orig_species_input} gene names...\")\n",
    "# Use left merge to keep all rows from df_inparanoid\n",
    "df_merged = pd.merge(\n",
    "    df_inparanoid,\n",
    "    orig_uniprot_df,\n",
    "    left_on=f\"{orig_species_input}_protein\",\n",
    "    right_on=\"Entry\",\n",
    "    how=\"left\"\n",
    ")\n",
    "    # Drop the redundant 'Entry' column from the merge\n",
    "df_merged = df_merged.drop(columns=[\"Entry\"])\n",
    "\n",
    "# 2. Annotate based on \"{species_input}_protein\" with species UniProt data\n",
    "print(f\"Merging {species_input} gene names...\")\n",
    "df_merged = pd.merge(\n",
    "    df_merged, # Merge into the already merged dataframe\n",
    "    species_uniprot_df,\n",
    "    left_on=f\"{species_input}_protein\",\n",
    "    right_on=\"Entry\",\n",
    "    how=\"left\"\n",
    ")\n",
    "# Drop the redundant 'Entry' column from the second merge\n",
    "df_merged = df_merged.drop(columns=[\"Entry\"])\n",
    "\n",
    "# 3. Save merged data\n",
    "print(f\"Saving merged data to {output_file}...\")\n",
    "df_merged.to_csv(output_file, index=False)\n",
    "print(f\"Successfully saved {len(df_merged)} annotated entries for {species_input}.\")\n",
    "\n",
    "print(\"\\nAnnotation process completed for all specified species.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "bc34340c-6c00-4d1d-848e-3711ead16704",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>mouse_protein</th>\n",
       "      <th>mouse_inparalog_score</th>\n",
       "      <th>mouse_seed_score</th>\n",
       "      <th>cow_protein</th>\n",
       "      <th>cow_inparalog_score</th>\n",
       "      <th>cow_seed_score</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>mouse_Gene_Name</th>\n",
       "      <th>cow_Gene_Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Q9QXZ0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>F1N6H4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10545.0</td>\n",
       "      <td>Macf1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Q9QXZ0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>F1N6H4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10545.0</td>\n",
       "      <td>Acf7</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>Q9QXZ0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>F1N6H4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10545.0</td>\n",
       "      <td>Aclp7</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>Q9QXZ0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>F1N6H4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10545.0</td>\n",
       "      <td>Kiaa0754</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>Q9QXZ0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>F1N6H4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10545.0</td>\n",
       "      <td>Macf</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44484</th>\n",
       "      <td>16645</td>\n",
       "      <td>Q925H3</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A0A3Q1NHC3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>Krtap16-8</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44485</th>\n",
       "      <td>16645</td>\n",
       "      <td>Q925H3</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A0A3Q1NHC3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>Krtap16.8</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44486</th>\n",
       "      <td>16645</td>\n",
       "      <td>O09048</td>\n",
       "      <td>0.667</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A0A3Q1NHC3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44487</th>\n",
       "      <td>16645</td>\n",
       "      <td>O08631</td>\n",
       "      <td>0.241</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A0A3Q1NHC3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44488</th>\n",
       "      <td>16646</td>\n",
       "      <td>Q9D6B9</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A0A3Q1MJX0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>Hrct1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>44489 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       cluster_id mouse_protein  mouse_inparalog_score  mouse_seed_score  \\\n",
       "0               1        Q9QXZ0                  1.000               1.0   \n",
       "1               1        Q9QXZ0                  1.000               1.0   \n",
       "2               1        Q9QXZ0                  1.000               1.0   \n",
       "3               1        Q9QXZ0                  1.000               1.0   \n",
       "4               1        Q9QXZ0                  1.000               1.0   \n",
       "...           ...           ...                    ...               ...   \n",
       "44484       16645        Q925H3                  1.000               1.0   \n",
       "44485       16645        Q925H3                  1.000               1.0   \n",
       "44486       16645        O09048                  0.667               NaN   \n",
       "44487       16645        O08631                  0.241               NaN   \n",
       "44488       16646        Q9D6B9                  1.000               1.0   \n",
       "\n",
       "      cow_protein  cow_inparalog_score  cow_seed_score  bitscore  \\\n",
       "0          F1N6H4                  1.0             1.0   10545.0   \n",
       "1          F1N6H4                  1.0             1.0   10545.0   \n",
       "2          F1N6H4                  1.0             1.0   10545.0   \n",
       "3          F1N6H4                  1.0             1.0   10545.0   \n",
       "4          F1N6H4                  1.0             1.0   10545.0   \n",
       "...           ...                  ...             ...       ...   \n",
       "44484  A0A3Q1NHC3                  1.0             1.0      42.0   \n",
       "44485  A0A3Q1NHC3                  1.0             1.0      42.0   \n",
       "44486  A0A3Q1NHC3                  1.0             1.0      42.0   \n",
       "44487  A0A3Q1NHC3                  1.0             1.0      42.0   \n",
       "44488  A0A3Q1MJX0                  1.0             1.0      42.0   \n",
       "\n",
       "      mouse_Gene_Name cow_Gene_Name  \n",
       "0               Macf1           NaN  \n",
       "1                Acf7           NaN  \n",
       "2               Aclp7           NaN  \n",
       "3            Kiaa0754           NaN  \n",
       "4                Macf           NaN  \n",
       "...               ...           ...  \n",
       "44484       Krtap16-8           NaN  \n",
       "44485       Krtap16.8           NaN  \n",
       "44486             NaN           NaN  \n",
       "44487             NaN           NaN  \n",
       "44488           Hrct1           NaN  \n",
       "\n",
       "[44489 rows x 10 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e81d9c50-0176-4be8-aada-c4adb40e70af",
   "metadata": {},
   "outputs": [],
   "source": [
    "hgnc_df = pd.read_csv(\"data/HGNC_gene_info_full.tsv\", sep=\"\\t\", dtype=str)\n",
    "hgnc_df = hgnc_df.dropna(subset=[\"uniprot_ids\", \"ensembl_gene_id\"])\n",
    "hgnc_exploded = hgnc_df.assign(uniprot_id=hgnc_df[\"uniprot_ids\"].str.split(\",\")).explode(\"uniprot_id\")\n",
    "hgnc_exploded[\"uniprot_id\"] = hgnc_exploded[\"uniprot_id\"].str.strip()\n",
    "\n",
    "df_merged = df_orthologs.merge(\n",
    "    hgnc_exploded[[\"uniprot_id\", \"ensembl_gene_id\", \"symbol\"]],\n",
    "    left_on=\"human_protein\",\n",
    "    right_on=\"uniprot_id\",\n",
    "    how=\"left\"\n",
    ")\n",
    "\n",
    "df_merged = df_merged.rename(columns={\n",
    "    \"symbol\": \"human_gene\",\n",
    "    \"ensembl_gene_id\": \"human_ensembl_gene_id\"\n",
    "}).drop(columns=[\"uniprot_id\"])\n",
    "\n",
    "df_merged = df_merged.dropna(subset=[\"human_ensembl_gene_id\"])\n",
    "df_merged.to_csv(f\"data/{species_input}_inParanoid_withHGNC.tsv\", sep=\"\\t\", index=False)\n",
    "\n",
    "# === Step 5: Optional - Species UniProt → Ensembl mapping ===\n",
    "map_path = f\"data/{species_input}_uniprot_to_ensembl.tsv\"\n",
    "try:\n",
    "    species_map = pd.read_csv(map_path, sep=\"\\t\", dtype=str)\n",
    "    df_merged = df_merged.merge(\n",
    "        species_map,\n",
    "        left_on=f\"{species_input}_protein\",\n",
    "        right_on=\"uniprotswissprot\",\n",
    "        how=\"left\"\n",
    "    ).rename(columns={\"ensembl_gene_id\": f\"{species_input}_ensembl_gene_id\"}) \\\n",
    "     .drop(columns=[\"uniprotswissprot\"])\n",
    "except FileNotFoundError:\n",
    "    print(f\"⚠️  Mapping file not found: {map_path}\")\n",
    "\n",
    "df_merged.to_csv(f\"data/df_merged_with_{species_input}_ensembl.tsv\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ec3af586-31f6-4064-b528-2c1c3e4e3016",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "from io import StringIO\n",
    "from itertools import product\n",
    "\n",
    "### IMPORTANT ####\n",
    "# Warning: First run \"src/convertOrthUniprotToEnsembl.r\" in R\n",
    "##################\n",
    "\n",
    "# === USER INPUT ===\n",
    "# original species\n",
    "orig_species_input = \"mouse\"\n",
    "# Ortholog species\n",
    "species_input = \"human\"  # Options: \"mouse\", \"zebrafish\", # No sheep in inParanoid\n",
    "\n",
    "# === INTERNAL MAPPINGS ===\n",
    "\n",
    "orig_species_info = {\n",
    "    \"mouse\":         {\"taxid\": \"10090\", \"code\": \"mmusculus\"},\n",
    "    \"rat\":           {\"taxid\": \"10116\", \"code\": \"rnorvegicus\"},\n",
    "    \"zebrafish\":     {\"taxid\": \"7955\",  \"code\": \"drerio\"},\n",
    "    \"chimpanzee\":    {\"taxid\": \"9598\",  \"code\": \"ptroglodytes\"},\n",
    "    \"chicken\":       {\"taxid\": \"9031\",  \"code\": \"ggallus\"},\n",
    "    \"pig\":           {\"taxid\": \"9823\",  \"code\": \"sscrofa\"},\n",
    "    \"cow\":           {\"taxid\": \"9913\",  \"code\": \"btaurus\"},\n",
    "    \"dog\":           {\"taxid\": \"9615\",  \"code\": \"clfamiliaris\"},\n",
    "    \"horse\":         {\"taxid\": \"9796\",  \"code\": \"ecaballus\"},\n",
    "    \"sheep\":         {\"taxid\": \"9940\",  \"code\": \"oarambouillet\"},\n",
    "    \"marmoset\":      {\"taxid\": \"9483\",  \"code\": \"cjacchus\"},\n",
    "    \"macaque\":       {\"taxid\": \"9544\",  \"code\": \"mmulatta\"},\n",
    "    \"human\":         {\"taxid\": \"9606\",  \"code\": \"hsapiens\"},\n",
    "}\n",
    "\n",
    "\n",
    "species_info = {\n",
    "    \"mouse\":         {\"taxid\": \"10090\", \"code\": \"mmusculus\"},\n",
    "    \"rat\":           {\"taxid\": \"10116\", \"code\": \"rnorvegicus\"},\n",
    "    \"zebrafish\":     {\"taxid\": \"7955\",  \"code\": \"drerio\"},\n",
    "    \"chimpanzee\":    {\"taxid\": \"9598\",  \"code\": \"ptroglodytes\"},\n",
    "    \"chicken\":       {\"taxid\": \"9031\",  \"code\": \"ggallus\"},\n",
    "    \"pig\":           {\"taxid\": \"9823\",  \"code\": \"sscrofa\"},\n",
    "    \"cow\":           {\"taxid\": \"9913\",  \"code\": \"btaurus\"},\n",
    "    \"dog\":           {\"taxid\": \"9615\",  \"code\": \"clfamiliaris\"},\n",
    "    \"horse\":         {\"taxid\": \"9796\",  \"code\": \"ecaballus\"},\n",
    "    \"sheep\":         {\"taxid\": \"9940\",  \"code\": \"oarambouillet\"},\n",
    "    \"marmoset\":      {\"taxid\": \"9483\",  \"code\": \"cjacchus\"},\n",
    "    \"macaque\":       {\"taxid\": \"9544\",  \"code\": \"mmulatta\"},\n",
    "    \"human\":         {\"taxid\": \"9606\",  \"code\": \"hsapiens\"},\n",
    "}\n",
    "\n",
    "if orig_species_input not in orig_species_info:\n",
    "    raise ValueError(f\"Species '{orig_species_input}' not supported. Choose from: {list(orig_species_info)}\")\n",
    "\n",
    "if species_input not in species_info:\n",
    "    raise ValueError(f\"Species '{species_input}' not supported. Choose from: {list(species_info)}\")\n",
    "\n",
    "species = species_info[species_input]\n",
    "taxid = species[\"taxid\"]\n",
    "code = species[\"code\"]\n",
    "\n",
    "orig_species = orig_species_info[orig_species_input]\n",
    "orig_taxid = orig_species[\"taxid\"]\n",
    "orig_code = orig_species[\"code\"]\n",
    "\n",
    "# === Step 1: Download inParanoid file for human vs species ===\n",
    "url = f\"https://inparanoidb.sbc.su.se/download/sqltable/{orig_taxid}&{taxid}&prot\"\n",
    "r = requests.get(url)\n",
    "r.raise_for_status()\n",
    "\n",
    "df = pd.read_csv(StringIO(r.text.strip()), sep=\"\\t\", header=None)\n",
    "df.columns = [\"cluster_id\", \"bitscore\", \"source_file\", \"inparalog_score\", \"protein_id\", \"seed_score\"]\n",
    "\n",
    "# === Step 2: Add species labels ===\n",
    "def infer_species(src):\n",
    "    if orig_taxid in src:\n",
    "        return orig_species_input\n",
    "    elif taxid in src:\n",
    "        return species_input\n",
    "    return \"unknown\"\n",
    "\n",
    "\n",
    "df[\"species\"] = df[\"source_file\"].apply(infer_species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e06e8517-1eac-4d31-b440-f4bbad8cc94a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>source_file</th>\n",
       "      <th>inparalog_score</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>seed_score</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>60090</td>\n",
       "      <td>10090.fa</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A2ASS6</td>\n",
       "      <td>1.0</td>\n",
       "      <td>mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>60090</td>\n",
       "      <td>9606.fa</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q8WZ42</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>14503</td>\n",
       "      <td>10090.fa</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q6ZWR6</td>\n",
       "      <td>1.0</td>\n",
       "      <td>mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>14503</td>\n",
       "      <td>9606.fa</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q8NF91</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>12156</td>\n",
       "      <td>10090.fa</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A2AAJ9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35953</th>\n",
       "      <td>17096</td>\n",
       "      <td>41</td>\n",
       "      <td>9606.fa</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P0DP42</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35954</th>\n",
       "      <td>17097</td>\n",
       "      <td>41</td>\n",
       "      <td>10090.fa</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A0A0G2JDY8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35955</th>\n",
       "      <td>17097</td>\n",
       "      <td>41</td>\n",
       "      <td>9606.fa</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A0A075B6W3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35956</th>\n",
       "      <td>17098</td>\n",
       "      <td>40</td>\n",
       "      <td>10090.fa</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q64389</td>\n",
       "      <td>1.0</td>\n",
       "      <td>mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35957</th>\n",
       "      <td>17098</td>\n",
       "      <td>40</td>\n",
       "      <td>9606.fa</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P31358</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35958 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       cluster_id  bitscore source_file  inparalog_score  protein_id  \\\n",
       "0               1     60090    10090.fa              1.0      A2ASS6   \n",
       "1               1     60090     9606.fa              1.0      Q8WZ42   \n",
       "2               2     14503    10090.fa              1.0      Q6ZWR6   \n",
       "3               2     14503     9606.fa              1.0      Q8NF91   \n",
       "4               3     12156    10090.fa              1.0      A2AAJ9   \n",
       "...           ...       ...         ...              ...         ...   \n",
       "35953       17096        41     9606.fa              1.0      P0DP42   \n",
       "35954       17097        41    10090.fa              1.0  A0A0G2JDY8   \n",
       "35955       17097        41     9606.fa              1.0  A0A075B6W3   \n",
       "35956       17098        40    10090.fa              1.0      Q64389   \n",
       "35957       17098        40     9606.fa              1.0      P31358   \n",
       "\n",
       "       seed_score species  \n",
       "0             1.0   mouse  \n",
       "1             1.0   human  \n",
       "2             1.0   mouse  \n",
       "3             1.0   human  \n",
       "4             1.0   mouse  \n",
       "...           ...     ...  \n",
       "35953         1.0   human  \n",
       "35954         1.0   mouse  \n",
       "35955         1.0   human  \n",
       "35956         1.0   mouse  \n",
       "35957         1.0   human  \n",
       "\n",
       "[35958 rows x 7 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "8ae34f83-1c81-4b62-8432-42df6f292fcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# === Step 3: Build ortholog pairs ===\n",
    "# Assume 'species' column exists, and 'orig_species_input' is defined\n",
    "records = []\n",
    "for cid, grp in df.groupby(\"cluster_id\"):\n",
    "    orig_spec = grp[grp[\"species\"] == orig_species_input]\n",
    "    others = grp[grp[\"species\"] == species_input]\n",
    "    \n",
    "    for h, o in product(orig_spec.itertuples(index=False), others.itertuples(index=False)):\n",
    "        records.append({\n",
    "            \"cluster_id\": cid,\n",
    "            f\"{orig_species_input}_protein\": h.protein_id,\n",
    "            f\"{orig_species_input}_inparalog_score\": h.inparalog_score,\n",
    "            f\"{orig_species_input}_seed_score\": h.seed_score,\n",
    "            f\"{species_input}_protein\": o.protein_id,\n",
    "            f\"{species_input}_inparalog_score\": o.inparalog_score,\n",
    "            f\"{species_input}_seed_score\": o.seed_score,\n",
    "            \"bitscore\": (h.bitscore + o.bitscore) / 2\n",
    "        })\n",
    "df_orthologs = pd.DataFrame(records)\n",
    "df_orthologs.to_csv(f\"data/{orig_species_input}_centric_inParanoid_{species_input}.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "c867b8f6-e5eb-4ad6-997a-f52c43d19188",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting annotation process...\n",
      "Loading mouse UniProt mapping from data/uniprotMapping_mouse.csv...\n",
      "Loaded 32289 mouse UniProt entries.\n",
      "\n",
      "--- Processing mouse ---\n",
      "Loading InParanoid data from data/inParanoid_mouse.csv...\n",
      "Loaded 20805 InParanoid entries for mouse.\n",
      "Loading mouse UniProt mapping from data/uniprotMapping_mouse.csv...\n",
      "Loaded 32289 mouse UniProt entries.\n",
      "Merging mouse gene names...\n",
      "Merging mouse gene names...\n",
      "Saving merged data to data/mouse_centric_inParanoid_mouse_AnnWithUniProt.csv...\n",
      "Successfully saved 81780 annotated entries for mouse.\n",
      "\n",
      "Annotation process completed for all specified species.\n"
     ]
    }
   ],
   "source": [
    "# === Step 4: UniProt → Gene Name ===\n",
    "print(\"Starting annotation process...\")\n",
    "\n",
    "# --- Load Human UniProt Mapping once ---\n",
    "orig_uniprot_file = os.path.join(data_dir, f\"uniprotMapping_{orig_species_input}.csv\")\n",
    "if not os.path.exists(orig_uniprot_file):\n",
    "    print(f\"Error: {orig_species_input} UniProt mapping file not found at {orig_uniprot_file}. Please run the UniProt fetching script for human first.\")\n",
    "    exit() # Exit if the essential human file is missing\n",
    "\n",
    "print(f\"Loading {orig_species_input} UniProt mapping from {orig_uniprot_file}...\")\n",
    "# We only need 'Entry' (Accession) and 'Gene Names'\n",
    "orig_uniprot_df = pd.read_csv(orig_uniprot_file, usecols=[\"Entry\", \"Gene Names\"])\n",
    "# Rename 'Gene Names' to distinguish it as human gene name\n",
    "orig_uniprot_df = orig_uniprot_df.rename(columns={\"Gene Names\": f\"{orig_species_input}_Gene_Name\"})\n",
    "print(f\"Loaded {len(orig_uniprot_df)} {orig_species_input} UniProt entries.\")\n",
    "\n",
    "# --- Process each species ---\n",
    "for species_input in species_to_process:\n",
    "    print(f\"\\n--- Processing {species_input} ---\")\n",
    "\n",
    "    inparanoid_file = os.path.join(data_dir, f\"inParanoid_{species_input}.csv\")\n",
    "    species_uniprot_file = os.path.join(data_dir, f\"uniprotMapping_{species_input}.csv\")\n",
    "    output_file = os.path.join(data_dir, f\"{orig_species_input}_centric_inParanoid_{species_input}_AnnWithUniProt.csv\")\n",
    "\n",
    "    # Check if input files exist\n",
    "    if not os.path.exists(inparanoid_file):\n",
    "        print(f\"Skipping {species_input}: InParanoid file not found at {inparanoid_file}\")\n",
    "        continue\n",
    "    if not os.path.exists(species_uniprot_file):\n",
    "        print(f\"Skipping {species_input}: UniProt mapping file not found at {species_uniprot_file}\")\n",
    "        continue\n",
    "\n",
    "    print(f\"Loading InParanoid data from {inparanoid_file}...\")\n",
    "    df_inparanoid = pd.read_csv(inparanoid_file)\n",
    "    print(f\"Loaded {len(df_inparanoid)} InParanoid entries for {species_input}.\")\n",
    "\n",
    "    print(f\"Loading {species_input} UniProt mapping from {species_uniprot_file}...\")\n",
    "    species_uniprot_df = pd.read_csv(species_uniprot_file, usecols=[\"Entry\", \"Gene Names\"])\n",
    "    # Rename 'Gene Names' to distinguish it as the species' gene name\n",
    "    species_uniprot_df = species_uniprot_df.rename(columns={\"Gene Names\": f\"{species_input}_Gene_Name\"})\n",
    "    print(f\"Loaded {len(species_uniprot_df)} {species_input} UniProt entries.\")\n",
    "\n",
    "    # 1. Annotate based on \"orig_species_protein\" with orig UniProt data\n",
    "    print(f\"Merging {orig_species_input} gene names...\")\n",
    "    # Use left merge to keep all rows from df_inparanoid\n",
    "    df_merged = pd.merge(\n",
    "        df_inparanoid,\n",
    "        orig_uniprot_df,\n",
    "        left_on=f\"{orig_species_input}_protein\",\n",
    "        right_on=\"Entry\",\n",
    "        how=\"left\"\n",
    "    )\n",
    "    # Drop the redundant 'Entry' column from the merge\n",
    "    df_merged = df_merged.drop(columns=[\"Entry\"])\n",
    "\n",
    "    # 2. Annotate based on \"{species_input}_protein\" with species UniProt data\n",
    "    print(f\"Merging {species_input} gene names...\")\n",
    "    df_merged = pd.merge(\n",
    "        df_merged, # Merge into the already merged dataframe\n",
    "        species_uniprot_df,\n",
    "        left_on=f\"{orig_species_input}_protein\",\n",
    "        right_on=\"Entry\",\n",
    "        how=\"left\"\n",
    "    )\n",
    "    # Drop the redundant 'Entry' column from the second merge\n",
    "    df_merged = df_merged.drop(columns=[\"Entry\"])\n",
    "\n",
    "    # 3. Save merged data\n",
    "    print(f\"Saving merged data to {output_file}...\")\n",
    "    df_merged.to_csv(output_file, index=False)\n",
    "    print(f\"Successfully saved {len(df_merged)} annotated entries for {species_input}.\")\n",
    "\n",
    "print(\"\\nAnnotation process completed for all specified species.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "ac222ef1-da4a-4f33-ad63-0f687241cd44",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting annotation process...\n",
      "Loading mouse UniProt mapping from data/uniprotMapping_mouse.csv...\n",
      "Loaded 32289 mouse UniProt entries.\n"
     ]
    }
   ],
   "source": [
    "# === Step 4: UniProt → Gene Name ===\n",
    "print(\"Starting annotation process...\")\n",
    "\n",
    "# --- Load Human UniProt Mapping once ---\n",
    "orig_uniprot_file = os.path.join(data_dir, f\"uniprotMapping_{orig_species_input}.csv\")\n",
    "if not os.path.exists(orig_uniprot_file):\n",
    "    print(f\"Error: {orig_species_input} UniProt mapping file not found at {orig_uniprot_file}. Please run the UniProt fetching script for human first.\")\n",
    "    exit() # Exit if the essential human file is missing\n",
    "\n",
    "print(f\"Loading {orig_species_input} UniProt mapping from {orig_uniprot_file}...\")\n",
    "# We only need 'Entry' (Accession) and 'Gene Names'\n",
    "orig_uniprot_df = pd.read_csv(orig_uniprot_file, usecols=[\"Entry\", \"Gene Names\"])\n",
    "# Rename 'Gene Names' to distinguish it as human gene name\n",
    "orig_uniprot_df = orig_uniprot_df.rename(columns={\"Gene Names\": f\"{orig_species_input}_Gene_Name\"})\n",
    "print(f\"Loaded {len(orig_uniprot_df)} {orig_species_input} UniProt entries.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "92706520-95d5-47e9-97d2-864ba32d0a97",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_to_process = [\"mouse\", \"rat\", \"human\", \"zebrafish\", \"chimpanzee\", \"chicken\", \"pig\", \"cow\", \"dog\", \"horse\", \"marmoset\",   \"macaque\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "c77b5a96-9069-4b29-a8c2-1a7384e089a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'human'"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "daf29b15-4403-4455-bc21-afed3cd1245e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "from io import StringIO\n",
    "from itertools import product\n",
    "\n",
    "### IMPORTANT ####\n",
    "# Warning: First run \"src/convertOrthUniprotToEnsembl.r\" in R\n",
    "##################\n",
    "\n",
    "# === USER INPUT ===\n",
    "# original species\n",
    "orig_species_input = \"mouse\"\n",
    "# Ortholog species\n",
    "species_input = \"human\"  # Options: \"mouse\", \"zebrafish\", # No sheep in inParanoid\n",
    "\n",
    "species_to_process = [\"mouse\", \"rat\", \"human\", \"zebrafish\", \"chimpanzee\", \"chicken\", \"pig\", \"cow\", \"dog\", \"horse\", \"marmoset\",   \"macaque\"]\n",
    "\n",
    "# === INTERNAL MAPPINGS ===\n",
    "\n",
    "orig_species_info = {\n",
    "    \"mouse\":         {\"taxid\": \"10090\", \"code\": \"mmusculus\"},\n",
    "    \"rat\":           {\"taxid\": \"10116\", \"code\": \"rnorvegicus\"},\n",
    "    \"zebrafish\":     {\"taxid\": \"7955\",  \"code\": \"drerio\"},\n",
    "    \"chimpanzee\":    {\"taxid\": \"9598\",  \"code\": \"ptroglodytes\"},\n",
    "    \"chicken\":       {\"taxid\": \"9031\",  \"code\": \"ggallus\"},\n",
    "    \"pig\":           {\"taxid\": \"9823\",  \"code\": \"sscrofa\"},\n",
    "    \"cow\":           {\"taxid\": \"9913\",  \"code\": \"btaurus\"},\n",
    "    \"dog\":           {\"taxid\": \"9615\",  \"code\": \"clfamiliaris\"},\n",
    "    \"horse\":         {\"taxid\": \"9796\",  \"code\": \"ecaballus\"},\n",
    "    \"sheep\":         {\"taxid\": \"9940\",  \"code\": \"oarambouillet\"},\n",
    "    \"marmoset\":      {\"taxid\": \"9483\",  \"code\": \"cjacchus\"},\n",
    "    \"macaque\":       {\"taxid\": \"9544\",  \"code\": \"mmulatta\"},\n",
    "    \"human\":         {\"taxid\": \"9606\",  \"code\": \"hsapiens\"},\n",
    "}\n",
    "\n",
    "\n",
    "species_info = {\n",
    "    \"mouse\":         {\"taxid\": \"10090\", \"code\": \"mmusculus\"},\n",
    "    \"rat\":           {\"taxid\": \"10116\", \"code\": \"rnorvegicus\"},\n",
    "    \"zebrafish\":     {\"taxid\": \"7955\",  \"code\": \"drerio\"},\n",
    "    \"chimpanzee\":    {\"taxid\": \"9598\",  \"code\": \"ptroglodytes\"},\n",
    "    \"chicken\":       {\"taxid\": \"9031\",  \"code\": \"ggallus\"},\n",
    "    \"pig\":           {\"taxid\": \"9823\",  \"code\": \"sscrofa\"},\n",
    "    \"cow\":           {\"taxid\": \"9913\",  \"code\": \"btaurus\"},\n",
    "    \"dog\":           {\"taxid\": \"9615\",  \"code\": \"clfamiliaris\"},\n",
    "    \"horse\":         {\"taxid\": \"9796\",  \"code\": \"ecaballus\"},\n",
    "    \"sheep\":         {\"taxid\": \"9940\",  \"code\": \"oarambouillet\"},\n",
    "    \"marmoset\":      {\"taxid\": \"9483\",  \"code\": \"cjacchus\"},\n",
    "    \"macaque\":       {\"taxid\": \"9544\",  \"code\": \"mmulatta\"},\n",
    "    \"human\":         {\"taxid\": \"9606\",  \"code\": \"hsapiens\"},\n",
    "}\n",
    "\n",
    "if orig_species_input not in orig_species_info:\n",
    "    raise ValueError(f\"Species '{orig_species_input}' not supported. Choose from: {list(orig_species_info)}\")\n",
    "\n",
    "if species_input not in species_info:\n",
    "    raise ValueError(f\"Species '{species_input}' not supported. Choose from: {list(species_info)}\")\n",
    "\n",
    "species = species_info[species_input]\n",
    "taxid = species[\"taxid\"]\n",
    "code = species[\"code\"]\n",
    "\n",
    "orig_species = orig_species_info[orig_species_input]\n",
    "orig_taxid = orig_species[\"taxid\"]\n",
    "orig_code = orig_species[\"code\"]\n",
    "\n",
    "# === Step 1: Download inParanoid file for human vs species ===\n",
    "url = f\"https://inparanoidb.sbc.su.se/download/sqltable/{orig_taxid}&{taxid}&prot\"\n",
    "r = requests.get(url)\n",
    "r.raise_for_status()\n",
    "\n",
    "df = pd.read_csv(StringIO(r.text.strip()), sep=\"\\t\", header=None)\n",
    "df.columns = [\"cluster_id\", \"bitscore\", \"source_file\", \"inparalog_score\", \"protein_id\", \"seed_score\"]\n",
    "\n",
    "# === Step 2: Add species labels ===\n",
    "def infer_species(src):\n",
    "    if orig_taxid in src:\n",
    "        return orig_species_input\n",
    "    elif taxid in src:\n",
    "        return species_input\n",
    "    return \"unknown\"\n",
    "\n",
    "df[\"species\"] = df[\"source_file\"].apply(infer_species)\n",
    "\n",
    "# === Step 3: Build ortholog pairs ===\n",
    "records = []\n",
    "for cid, grp in df.groupby(\"cluster_id\"):\n",
    "    orig_spec = grp[grp[\"species\"] == orig_species_input]\n",
    "    others = grp[grp[\"species\"] == species_input]\n",
    "    \n",
    "    for h, o in product(orig_spec.itertuples(index=False), others.itertuples(index=False)):\n",
    "        records.append({\n",
    "            \"cluster_id\": cid,\n",
    "            f\"{orig_species_input}_protein\": h.protein_id,\n",
    "            f\"{orig_species_input}_inparalog_score\": h.inparalog_score,\n",
    "            f\"{orig_species_input}_seed_score\": h.seed_score,\n",
    "            f\"{species_input}_protein\": o.protein_id,\n",
    "            f\"{species_input}_inparalog_score\": o.inparalog_score,\n",
    "            f\"{species_input}_seed_score\": o.seed_score,\n",
    "            \"bitscore\": (h.bitscore + o.bitscore) / 2\n",
    "        })\n",
    "\n",
    "\n",
    "df_orthologs = pd.DataFrame(records)\n",
    "df_orthologs.to_csv(f\"data/{orig_species_input}_centric_inParanoid_{species_input}.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "1a8fb241-4fd7-46aa-bbc7-9377bb18d9e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>mouse_protein</th>\n",
       "      <th>mouse_inparalog_score</th>\n",
       "      <th>mouse_seed_score</th>\n",
       "      <th>human_protein</th>\n",
       "      <th>human_inparalog_score</th>\n",
       "      <th>human_seed_score</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A2ASS6</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q8WZ42</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>60090.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>Q6ZWR6</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q8NF91</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>14503.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>A2AAJ9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q5VST9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12156.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>Q9QXZ0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q9UPN3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10869.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>Q91ZU6</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q03001</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10549.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20800</th>\n",
       "      <td>17094</td>\n",
       "      <td>A0A0G2JGM9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A0A075B6W9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>45.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20801</th>\n",
       "      <td>17095</td>\n",
       "      <td>A0A0G2JFW2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A0A075B6W1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>44.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20802</th>\n",
       "      <td>17096</td>\n",
       "      <td>A0A494B9K2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P0DP42</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>41.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20803</th>\n",
       "      <td>17097</td>\n",
       "      <td>A0A0G2JDY8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A0A075B6W3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>41.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20804</th>\n",
       "      <td>17098</td>\n",
       "      <td>Q64389</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P31358</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20805 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       cluster_id mouse_protein  mouse_inparalog_score  mouse_seed_score  \\\n",
       "0               1        A2ASS6                    1.0               1.0   \n",
       "1               2        Q6ZWR6                    1.0               1.0   \n",
       "2               3        A2AAJ9                    1.0               1.0   \n",
       "3               4        Q9QXZ0                    1.0               1.0   \n",
       "4               5        Q91ZU6                    1.0               1.0   \n",
       "...           ...           ...                    ...               ...   \n",
       "20800       17094    A0A0G2JGM9                    1.0               1.0   \n",
       "20801       17095    A0A0G2JFW2                    1.0               1.0   \n",
       "20802       17096    A0A494B9K2                    1.0               1.0   \n",
       "20803       17097    A0A0G2JDY8                    1.0               1.0   \n",
       "20804       17098        Q64389                    1.0               1.0   \n",
       "\n",
       "      human_protein  human_inparalog_score  human_seed_score  bitscore  \n",
       "0            Q8WZ42                    1.0               1.0   60090.0  \n",
       "1            Q8NF91                    1.0               1.0   14503.0  \n",
       "2            Q5VST9                    1.0               1.0   12156.0  \n",
       "3            Q9UPN3                    1.0               1.0   10869.0  \n",
       "4            Q03001                    1.0               1.0   10549.0  \n",
       "...             ...                    ...               ...       ...  \n",
       "20800    A0A075B6W9                    1.0               1.0      45.0  \n",
       "20801    A0A075B6W1                    1.0               1.0      44.0  \n",
       "20802        P0DP42                    1.0               1.0      41.0  \n",
       "20803    A0A075B6W3                    1.0               1.0      41.0  \n",
       "20804        P31358                    1.0               1.0      40.0  \n",
       "\n",
       "[20805 rows x 8 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_orthologs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "c8bc9fc5-9f5f-48bc-9af2-9b8545bf0989",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting annotation process...\n",
      "Loading mouse UniProt mapping from data/uniprotMapping_mouse.csv...\n",
      "Loaded 32289 mouse UniProt entries.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>mouse_Gene_Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A087WPF7</td>\n",
       "      <td>Auts2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A087WPF7</td>\n",
       "      <td>Kiaa0442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A088MLT8</td>\n",
       "      <td>Iqcj-Schip1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A088MLT8</td>\n",
       "      <td>Iqschfp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A088MLT8</td>\n",
       "      <td>Schip1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32284</th>\n",
       "      <td>Q9WUQ7</td>\n",
       "      <td>Dexi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32285</th>\n",
       "      <td>Q9WUQ7</td>\n",
       "      <td>Myle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32286</th>\n",
       "      <td>Q9WVB6</td>\n",
       "      <td>Lenep</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32287</th>\n",
       "      <td>Q9WVB6</td>\n",
       "      <td>Lep503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32288</th>\n",
       "      <td>Q9Z1R4</td>\n",
       "      <td>D17h6s53e</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32289 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Entry mouse_Gene_Name\n",
       "0      A0A087WPF7           Auts2\n",
       "1      A0A087WPF7        Kiaa0442\n",
       "2      A0A088MLT8     Iqcj-Schip1\n",
       "3      A0A088MLT8         Iqschfp\n",
       "4      A0A088MLT8          Schip1\n",
       "...           ...             ...\n",
       "32284      Q9WUQ7            Dexi\n",
       "32285      Q9WUQ7            Myle\n",
       "32286      Q9WVB6           Lenep\n",
       "32287      Q9WVB6          Lep503\n",
       "32288      Q9Z1R4       D17h6s53e\n",
       "\n",
       "[32289 rows x 2 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# === Step 4: UniProt → Gene Name ===\n",
    "print(\"Starting annotation process...\")\n",
    "\n",
    "# --- Load Human UniProt Mapping once ---\n",
    "orig_uniprot_file = os.path.join(data_dir, f\"uniprotMapping_{orig_species_input}.csv\")\n",
    "if not os.path.exists(orig_uniprot_file):\n",
    "    print(f\"Error: {orig_species_input} UniProt mapping file not found at {orig_uniprot_file}. Please run the UniProt fetching script for human first.\")\n",
    "    exit() # Exit if the essential human file is missing\n",
    "\n",
    "print(f\"Loading {orig_species_input} UniProt mapping from {orig_uniprot_file}...\")\n",
    "# We only need 'Entry' (Accession) and 'Gene Names'\n",
    "orig_uniprot_df = pd.read_csv(orig_uniprot_file, usecols=[\"Entry\", \"Gene Names\"])\n",
    "# Rename 'Gene Names' to distinguish it as human gene name\n",
    "orig_uniprot_df = orig_uniprot_df.rename(columns={\"Gene Names\": f\"{orig_species_input}_Gene_Name\"})\n",
    "print(f\"Loaded {len(orig_uniprot_df)} {orig_species_input} UniProt entries.\")\n",
    "orig_uniprot_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "c97e66b6-0054-464e-ad66-d88f4e19dab2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- Processing human ---\n",
      "Loading InParanoid data from data/mouse_centric_inParanoid_human.csv...\n",
      "Loaded 20805 InParanoid entries for human.\n",
      "Loading human UniProt mapping from data/uniprotMapping_human.csv...\n",
      "Loaded 45583 human UniProt entries.\n"
     ]
    }
   ],
   "source": [
    "# --- Process each species ---\n",
    "print(f\"\\n--- Processing {species_input} ---\")\n",
    "\n",
    "inparanoid_file = os.path.join(data_dir, f\"{orig_species_input}_centric_inParanoid_{species_input}.csv\")\n",
    "species_uniprot_file = os.path.join(data_dir, f\"uniprotMapping_{species_input}.csv\")\n",
    "output_file = os.path.join(data_dir, f\"{orig_species_input}_centric_inParanoid_{species_input}_AnnWithUniProt.csv\")\n",
    "\n",
    "# Check if input files exist\n",
    "if not os.path.exists(inparanoid_file):\n",
    "    print(f\"Skipping {species_input}: InParanoid file not found at {inparanoid_file}\")\n",
    "\n",
    "if not os.path.exists(species_uniprot_file):\n",
    "    print(f\"Skipping {species_input}: UniProt mapping file not found at {species_uniprot_file}\")\n",
    "\n",
    "print(f\"Loading InParanoid data from {inparanoid_file}...\")\n",
    "df_inparanoid = pd.read_csv(inparanoid_file)\n",
    "print(f\"Loaded {len(df_inparanoid)} InParanoid entries for {species_input}.\")\n",
    "\n",
    "print(f\"Loading {species_input} UniProt mapping from {species_uniprot_file}...\")\n",
    "species_uniprot_df = pd.read_csv(species_uniprot_file, usecols=[\"Entry\", \"Gene Names\"])\n",
    "# Rename 'Gene Names' to distinguish it as the species' gene name\n",
    "species_uniprot_df = species_uniprot_df.rename(columns={\"Gene Names\": f\"{species_input}_Gene_Name\"})\n",
    "print(f\"Loaded {len(species_uniprot_df)} {species_input} UniProt entries.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "de3ec689-9fb6-4adc-8425-5bafafb142ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merging mouse gene names...\n",
      "Merging human gene names...\n",
      "Saving merged data to data/mouse_centric_inParanoid_human_AnnWithUniProt.csv...\n",
      "Successfully saved 92169 annotated entries for human.\n",
      "\n",
      "Annotation process completed for all specified species.\n"
     ]
    }
   ],
   "source": [
    "# 1. Annotate based on \"orig_species_protein\" with orig UniProt data\n",
    "print(f\"Merging {orig_species_input} gene names...\")\n",
    "# Use left merge to keep all rows from df_inparanoid\n",
    "df_merged = pd.merge(\n",
    "    df_inparanoid,\n",
    "    orig_uniprot_df,\n",
    "    left_on=f\"{orig_species_input}_protein\",\n",
    "    right_on=\"Entry\",\n",
    "    how=\"left\"\n",
    ")\n",
    "    # Drop the redundant 'Entry' column from the merge\n",
    "df_merged = df_merged.drop(columns=[\"Entry\"])\n",
    "\n",
    "# 2. Annotate based on \"{species_input}_protein\" with species UniProt data\n",
    "print(f\"Merging {species_input} gene names...\")\n",
    "df_merged = pd.merge(\n",
    "    df_merged, # Merge into the already merged dataframe\n",
    "    species_uniprot_df,\n",
    "    left_on=f\"{species_input}_protein\",\n",
    "    right_on=\"Entry\",\n",
    "    how=\"left\"\n",
    ")\n",
    "# Drop the redundant 'Entry' column from the second merge\n",
    "df_merged = df_merged.drop(columns=[\"Entry\"])\n",
    "\n",
    "# 3. Save merged data\n",
    "print(f\"Saving merged data to {output_file}...\")\n",
    "df_merged.to_csv(output_file, index=False)\n",
    "print(f\"Successfully saved {len(df_merged)} annotated entries for {species_input}.\")\n",
    "\n",
    "print(\"\\nAnnotation process completed for all specified species.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91502a4c-8214-4f4b-87c0-731f691fb883",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a2be061-9c57-4e5e-adda-d259424dd6ae",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "063c5532-562f-4c1a-ac05-a2eb4f5cffe7",
   "metadata": {},
   "outputs": [],
   "source": [
    "triplicate= gene_pair[['LR_pair', 'triplet', 'ligand_species_ann', 'receptor_species_ann', 'used Human L', 'used Human R', 'notes', 'Remove decision']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "11a69e13-fdc7-4112-80cd-0fc46ad7317f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LR_pair</th>\n",
       "      <th>triplet</th>\n",
       "      <th>ligand_species_ann</th>\n",
       "      <th>receptor_species_ann</th>\n",
       "      <th>used Human L</th>\n",
       "      <th>used Human R</th>\n",
       "      <th>notes</th>\n",
       "      <th>Remove decision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LPL LRP1</td>\n",
       "      <td>1281473 LPL LRP1</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1992</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FGF1 ITGAV</td>\n",
       "      <td>18441324 FGF1 ITGAV</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CALCA RAMP1</td>\n",
       "      <td>18599553 CALCA RAMP1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PODXL2 SELE</td>\n",
       "      <td>18606703 PODXL2 SELE</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TMPRSS6 HJV</td>\n",
       "      <td>18976966 TMPRSS6 HJV</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5411</th>\n",
       "      <td>KISS1 KISS1R</td>\n",
       "      <td>12944565 KISS1 KISS1R</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>REMOVE-Not support</td>\n",
       "      <td>REMOVE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5412</th>\n",
       "      <td>CCL1 CCR8</td>\n",
       "      <td>12967681 CCL1 CCR8</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>REMOVE-Not support</td>\n",
       "      <td>REMOVE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5413</th>\n",
       "      <td>TNF TNFRSF21</td>\n",
       "      <td>9714541 TNF TNFRSF21</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>REMOVE-NOT SUPPORT - Al - Agrees</td>\n",
       "      <td>REMOVE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5414</th>\n",
       "      <td>NTNG1 LRRC4C</td>\n",
       "      <td>14595443 NTNG1 LRRC4C</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>REMOVE-Reversed Interaction</td>\n",
       "      <td>REMOVE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5415</th>\n",
       "      <td>IL16 GRIN2D</td>\n",
       "      <td>10479680 IL16 GRIN2D</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>REMOVE- insufficient evidence</td>\n",
       "      <td>REMOVE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5416 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0          LR_pair                triplet ligand_species_ann  \\\n",
       "0         LPL LRP1       1281473 LPL LRP1         Bos taurus   \n",
       "1       FGF1 ITGAV    18441324 FGF1 ITGAV       Homo sapiens   \n",
       "2      CALCA RAMP1   18599553 CALCA RAMP1       Homo sapiens   \n",
       "3      PODXL2 SELE   18606703 PODXL2 SELE       Homo sapiens   \n",
       "4      TMPRSS6 HJV   18976966 TMPRSS6 HJV       Homo sapiens   \n",
       "...            ...                    ...                ...   \n",
       "5411  KISS1 KISS1R  12944565 KISS1 KISS1R                      \n",
       "5412     CCL1 CCR8     12967681 CCL1 CCR8                      \n",
       "5413  TNF TNFRSF21   9714541 TNF TNFRSF21                      \n",
       "5414  NTNG1 LRRC4C  14595443 NTNG1 LRRC4C                      \n",
       "5415   IL16 GRIN2D   10479680 IL16 GRIN2D                      \n",
       "\n",
       "0    receptor_species_ann used Human L used Human R  \\\n",
       "0            Homo sapiens                             \n",
       "1            Homo sapiens                             \n",
       "2            Homo sapiens                             \n",
       "3            Homo sapiens                             \n",
       "4            Homo sapiens                             \n",
       "...                   ...          ...          ...   \n",
       "5411                                                  \n",
       "5412                                                  \n",
       "5413                                                  \n",
       "5414                                                  \n",
       "5415                                                  \n",
       "\n",
       "0                                notes Remove decision  \n",
       "0                                 1992                  \n",
       "1                                                       \n",
       "2                                                       \n",
       "3                                                       \n",
       "4                                                       \n",
       "...                                ...             ...  \n",
       "5411                REMOVE-Not support          REMOVE  \n",
       "5412                REMOVE-Not support          REMOVE  \n",
       "5413  REMOVE-NOT SUPPORT - Al - Agrees          REMOVE  \n",
       "5414       REMOVE-Reversed Interaction          REMOVE  \n",
       "5415     REMOVE- insufficient evidence          REMOVE  \n",
       "\n",
       "[5416 rows x 8 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "triplicate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "3b15d63f-83af-4514-be93-0357248b3bc3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_39836/1473446994.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  triplicate['notes'] = triplicate['Remove decision'].str.lower()\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LR_pair</th>\n",
       "      <th>triplet</th>\n",
       "      <th>ligand_species_ann</th>\n",
       "      <th>receptor_species_ann</th>\n",
       "      <th>used Human L</th>\n",
       "      <th>used Human R</th>\n",
       "      <th>notes</th>\n",
       "      <th>Remove decision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ROBO1 ROBO2</td>\n",
       "      <td>26186094 ROBO1 ROBO2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>remove</td>\n",
       "      <td>REMOVE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>IL18 IL18RAP</td>\n",
       "      <td>15760905 IL18 IL18RAP</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>remove</td>\n",
       "      <td>REMOVE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>BDNF NTRK1</td>\n",
       "      <td>2157470 BDNF NTRK1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>remove</td>\n",
       "      <td>REMOVE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>IL1A IL1RAP</td>\n",
       "      <td>2950091 IL1A IL1RAP</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>remove</td>\n",
       "      <td>REMOVE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>HSPG2 LRP1</td>\n",
       "      <td>7526899 HSPG2 LRP1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>remove</td>\n",
       "      <td>REMOVE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5376</th>\n",
       "      <td>COL4A3 CD93</td>\n",
       "      <td></td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td>remove</td>\n",
       "      <td>REMOVE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5377</th>\n",
       "      <td>COL4A4 CD93</td>\n",
       "      <td></td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td>remove</td>\n",
       "      <td>REMOVE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5378</th>\n",
       "      <td>COL4A5 CD93</td>\n",
       "      <td></td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td>remove</td>\n",
       "      <td>REMOVE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5379</th>\n",
       "      <td>COL4A6 CD93</td>\n",
       "      <td></td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td>remove</td>\n",
       "      <td>REMOVE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5411</th>\n",
       "      <td>IGSF1 ACVR1B</td>\n",
       "      <td></td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td>remove</td>\n",
       "      <td>REMOVE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>168 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0          LR_pair                triplet ligand_species_ann  \\\n",
       "0      ROBO1 ROBO2   26186094 ROBO1 ROBO2                      \n",
       "6     IL18 IL18RAP  15760905 IL18 IL18RAP                      \n",
       "152     BDNF NTRK1     2157470 BDNF NTRK1                      \n",
       "178    IL1A IL1RAP    2950091 IL1A IL1RAP                      \n",
       "221     HSPG2 LRP1     7526899 HSPG2 LRP1                      \n",
       "...            ...                    ...                ...   \n",
       "5376   COL4A3 CD93                              Homo sapiens   \n",
       "5377   COL4A4 CD93                              Homo sapiens   \n",
       "5378   COL4A5 CD93                              Homo sapiens   \n",
       "5379   COL4A6 CD93                              Homo sapiens   \n",
       "5411  IGSF1 ACVR1B                              Homo sapiens   \n",
       "\n",
       "0    receptor_species_ann used Human L used Human R   notes Remove decision  \n",
       "0                                                    remove          REMOVE  \n",
       "6                                                    remove          REMOVE  \n",
       "152                                                  remove          REMOVE  \n",
       "178                                                  remove          REMOVE  \n",
       "221                                                  remove          REMOVE  \n",
       "...                   ...          ...          ...     ...             ...  \n",
       "5376         Homo sapiens            ✅            ✅  remove          REMOVE  \n",
       "5377         Homo sapiens            ✅            ✅  remove          REMOVE  \n",
       "5378         Homo sapiens            ✅            ✅  remove          REMOVE  \n",
       "5379         Homo sapiens            ✅            ✅  remove          REMOVE  \n",
       "5411         Homo sapiens            ✅            ✅  remove          REMOVE  \n",
       "\n",
       "[168 rows x 8 columns]"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter out rows where 'notes' contains 'remove'\n",
    "triplicate['notes'] = triplicate['Remove decision'].str.lower()\n",
    "\n",
    "# Remove rows where 'notes' contains 'remove'\n",
    "triplicate_with_remove = triplicate[triplicate['notes'].str.contains('remove', na=False)]\n",
    "\n",
    "# Remove rows where 'notes' contains 'remove'\n",
    "triplicate = triplicate[~triplicate['notes'].str.contains('remove', na=False)]\n",
    "\n",
    "triplicate\n",
    "triplicate_with_remove"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8d85fde7-ef2a-4267-be4b-5ef157834f88",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'triplicate_with_remove' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Convert both columns to sets for easy set operations\u001b[39;00m\n\u001b[1;32m      2\u001b[0m set_filtered \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mset\u001b[39m(triplicate[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLR_pair\u001b[39m\u001b[38;5;124m'\u001b[39m])\n\u001b[0;32m----> 3\u001b[0m set_unfiltered \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mset\u001b[39m(\u001b[43mtriplicate_with_remove\u001b[49m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLR_pair\u001b[39m\u001b[38;5;124m'\u001b[39m])\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m# Overlap (present in both)\u001b[39;00m\n\u001b[1;32m      6\u001b[0m overlap \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(set_filtered \u001b[38;5;241m&\u001b[39m set_unfiltered)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'triplicate_with_remove' is not defined"
     ]
    }
   ],
   "source": [
    "# Convert both columns to sets for easy set operations\n",
    "set_filtered = set(triplicate['LR_pair'])\n",
    "set_unfiltered = set(triplicate_with_remove['LR_pair'])\n",
    "\n",
    "# Overlap (present in both)\n",
    "overlap = list(set_filtered & set_unfiltered)\n",
    "\n",
    "# Not covered (in filtered but NOT in unfiltered)\n",
    "not_covered = list(set_unfiltered - set_filtered)\n",
    "\n",
    "print(\"LR_pairs covered (overlap):\", overlap, len(overlap))\n",
    "print(\"LR_pairs not covered:\", not_covered, len(not_covered))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2e7061c-fd11-47e5-93f8-239fbe986cd9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "c2e36002-7b4d-469c-8922-ddd1d9680e76",
   "metadata": {},
   "outputs": [],
   "source": [
    "triplicate_with_remove.to_csv(\"data/rows_with_remove.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bfa1271a-e810-428e-9cd7-16921d999ffd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ligand_species_ann</th>\n",
       "      <th>receptor_species_ann</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>3082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Homo sapiens, Mus Musculus</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>Homo sapiens, Mus Musculus</td>\n",
       "      <td>Homo sapiens, Mus Musculus, Rattus norvegicus,...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>Homo sapiens, Gallus gallus</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Xenopus laevis</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>Xenopus laevis, Homo sapiens</td>\n",
       "      <td>Homo sapiens, Xenopus laevis</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>161 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               ligand_species_ann  \\\n",
       "36                   Homo sapiens   \n",
       "104                  Mus Musculus   \n",
       "50                   Homo sapiens   \n",
       "0                                   \n",
       "101                  Mus Musculus   \n",
       "..                            ...   \n",
       "74     Homo sapiens, Mus Musculus   \n",
       "71     Homo sapiens, Mus Musculus   \n",
       "67    Homo sapiens, Gallus gallus   \n",
       "62                   Homo sapiens   \n",
       "160  Xenopus laevis, Homo sapiens   \n",
       "\n",
       "                                  receptor_species_ann  count  \n",
       "36                                        Homo sapiens   3082  \n",
       "104                                       Mus Musculus    702  \n",
       "50                                        Mus Musculus    181  \n",
       "0                                                         179  \n",
       "101                                       Homo sapiens    176  \n",
       "..                                                 ...    ...  \n",
       "74                                   Rattus norvegicus      1  \n",
       "71   Homo sapiens, Mus Musculus, Rattus norvegicus,...      1  \n",
       "67                                        Homo sapiens      1  \n",
       "62                                      Xenopus laevis      1  \n",
       "160                       Homo sapiens, Xenopus laevis      1  \n",
       "\n",
       "[161 rows x 3 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group by the two species columns and count the number of rows for each combination\n",
    "summary_counts = triplicate.groupby(['ligand_species_ann', 'receptor_species_ann']).size().reset_index(name='count')\n",
    "\n",
    "# Optional: sort by count descending\n",
    "summary_counts = summary_counts.sort_values(by='count', ascending=False)\n",
    "summary_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f77bea3f-245f-412f-a0ab-0ade7eb76424",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_counts.to_csv(\"data/pairs_temp_perPair.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "69311d95-4aaf-42e5-b4b7-93c9aa3a311f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rows where both species columns contain 'Homo sapiens': 3536\n"
     ]
    }
   ],
   "source": [
    "homo_sapiens_in_both = triplicate[\n",
    "    triplicate['ligand_species_ann'].str.contains('Homo sapiens', case=False, na=False) &\n",
    "    triplicate['receptor_species_ann'].str.contains('Homo sapiens', case=False, na=False)\n",
    "]\n",
    "\n",
    "print(\"Rows where both species columns contain 'Homo sapiens':\", homo_sapiens_in_both.shape[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "89a44748-a53a-4f34-8134-329d207e7179",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.6742944317315027"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "homo_sapiens_in_both.shape[0]/len(triplicate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "194cdc41-684a-4218-89bd-c0fa17baa596",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5244"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(triplicate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "60c645d9-db01-4128-9a37-8be1c8912b51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LR_pair</th>\n",
       "      <th>triplet</th>\n",
       "      <th>ligand_species_ann</th>\n",
       "      <th>receptor_species_ann</th>\n",
       "      <th>used Human L</th>\n",
       "      <th>used Human R</th>\n",
       "      <th>notes</th>\n",
       "      <th>Remove decision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ANGPT2 ITGA5</td>\n",
       "      <td>16424009 ANGPT2 ITGA5</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ANGPT2 ITGAV</td>\n",
       "      <td>16424009 ANGPT2 ITGAV</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ANGPT2 ITGB1</td>\n",
       "      <td>16424009 ANGPT2 ITGB1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IL18 CD48</td>\n",
       "      <td>15760905 IL18 CD48</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>IL18 IL18R1</td>\n",
       "      <td>15760905 IL18 IL18R1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5406</th>\n",
       "      <td>LAMA5 ITGA7</td>\n",
       "      <td></td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5407</th>\n",
       "      <td>LAMB2 ITGB4</td>\n",
       "      <td></td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5408</th>\n",
       "      <td>LAMB2 ITGA6</td>\n",
       "      <td></td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5409</th>\n",
       "      <td>LAMB1 ITGB4</td>\n",
       "      <td></td>\n",
       "      <td>Mus Musculus, Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5410</th>\n",
       "      <td>LAMC1 ITGA6</td>\n",
       "      <td></td>\n",
       "      <td>Mus Musculus, Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5244 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0          LR_pair                triplet          ligand_species_ann  \\\n",
       "1     ANGPT2 ITGA5  16424009 ANGPT2 ITGA5                Homo sapiens   \n",
       "2     ANGPT2 ITGAV  16424009 ANGPT2 ITGAV                Homo sapiens   \n",
       "3     ANGPT2 ITGB1  16424009 ANGPT2 ITGB1                Homo sapiens   \n",
       "4        IL18 CD48     15760905 IL18 CD48                Homo sapiens   \n",
       "5      IL18 IL18R1   15760905 IL18 IL18R1                Homo sapiens   \n",
       "...            ...                    ...                         ...   \n",
       "5406   LAMA5 ITGA7                                       Homo sapiens   \n",
       "5407   LAMB2 ITGB4                                       Homo sapiens   \n",
       "5408   LAMB2 ITGA6                                       Homo sapiens   \n",
       "5409   LAMB1 ITGB4                         Mus Musculus, Homo sapiens   \n",
       "5410   LAMC1 ITGA6                         Mus Musculus, Homo sapiens   \n",
       "\n",
       "0    receptor_species_ann used Human L used Human R notes Remove decision  \n",
       "1            Homo sapiens            ✅            ✅                        \n",
       "2            Homo sapiens            ✅            ✅                        \n",
       "3            Homo sapiens            ✅            ✅                        \n",
       "4            Homo sapiens            ✅            ✅                        \n",
       "5            Homo sapiens            ✅            ✅                        \n",
       "...                   ...          ...          ...   ...             ...  \n",
       "5406         Homo sapiens            ✅            ✅                        \n",
       "5407         Homo sapiens            ✅            ✅                        \n",
       "5408         Homo sapiens            ✅            ✅                        \n",
       "5409         Homo sapiens            ✅            ✅                        \n",
       "5410         Homo sapiens            ✅            ✅                        \n",
       "\n",
       "[5244 rows x 8 columns]"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "triplicate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "ff5df646-05bb-4cda-8dd3-73ed4b7640a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LR_pair</th>\n",
       "      <th>triplet</th>\n",
       "      <th>ligand_species_ann</th>\n",
       "      <th>receptor_species_ann</th>\n",
       "      <th>used Human L</th>\n",
       "      <th>used Human R</th>\n",
       "      <th>notes</th>\n",
       "      <th>Remove decision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ANGPT2 ITGA5</td>\n",
       "      <td>16424009 ANGPT2 ITGA5</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ANGPT2 ITGAV</td>\n",
       "      <td>16424009 ANGPT2 ITGAV</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ANGPT2 ITGB1</td>\n",
       "      <td>16424009 ANGPT2 ITGB1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IL18 CD48</td>\n",
       "      <td>15760905 IL18 CD48</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>IL18 IL18R1</td>\n",
       "      <td>15760905 IL18 IL18R1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5406</th>\n",
       "      <td>LAMA5 ITGA7</td>\n",
       "      <td></td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5407</th>\n",
       "      <td>LAMB2 ITGB4</td>\n",
       "      <td></td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5408</th>\n",
       "      <td>LAMB2 ITGA6</td>\n",
       "      <td></td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5409</th>\n",
       "      <td>LAMB1 ITGB4</td>\n",
       "      <td></td>\n",
       "      <td>Mus Musculus, Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5410</th>\n",
       "      <td>LAMC1 ITGA6</td>\n",
       "      <td></td>\n",
       "      <td>Mus Musculus, Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>✅</td>\n",
       "      <td>✅</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3536 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0          LR_pair                triplet          ligand_species_ann  \\\n",
       "1     ANGPT2 ITGA5  16424009 ANGPT2 ITGA5                Homo sapiens   \n",
       "2     ANGPT2 ITGAV  16424009 ANGPT2 ITGAV                Homo sapiens   \n",
       "3     ANGPT2 ITGB1  16424009 ANGPT2 ITGB1                Homo sapiens   \n",
       "4        IL18 CD48     15760905 IL18 CD48                Homo sapiens   \n",
       "5      IL18 IL18R1   15760905 IL18 IL18R1                Homo sapiens   \n",
       "...            ...                    ...                         ...   \n",
       "5406   LAMA5 ITGA7                                       Homo sapiens   \n",
       "5407   LAMB2 ITGB4                                       Homo sapiens   \n",
       "5408   LAMB2 ITGA6                                       Homo sapiens   \n",
       "5409   LAMB1 ITGB4                         Mus Musculus, Homo sapiens   \n",
       "5410   LAMC1 ITGA6                         Mus Musculus, Homo sapiens   \n",
       "\n",
       "0    receptor_species_ann used Human L used Human R notes Remove decision  \n",
       "1            Homo sapiens            ✅            ✅                        \n",
       "2            Homo sapiens            ✅            ✅                        \n",
       "3            Homo sapiens            ✅            ✅                        \n",
       "4            Homo sapiens            ✅            ✅                        \n",
       "5            Homo sapiens            ✅            ✅                        \n",
       "...                   ...          ...          ...   ...             ...  \n",
       "5406         Homo sapiens            ✅            ✅                        \n",
       "5407         Homo sapiens            ✅            ✅                        \n",
       "5408         Homo sapiens            ✅            ✅                        \n",
       "5409         Homo sapiens            ✅            ✅                        \n",
       "5410         Homo sapiens            ✅            ✅                        \n",
       "\n",
       "[3536 rows x 8 columns]"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "homo_sapiens_in_both"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "289e4613-9072-417c-a6b4-a0c45a862647",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LR_pair</th>\n",
       "      <th>triplet</th>\n",
       "      <th>ligand_species_ann</th>\n",
       "      <th>receptor_species_ann</th>\n",
       "      <th>used Human L</th>\n",
       "      <th>used Human R</th>\n",
       "      <th>notes</th>\n",
       "      <th>Remove decision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A2M HSPA5</td>\n",
       "      <td>12194978 A2M HSPA5, 32541810 A2M HSPA5</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>, human α2M, human Grp 78 (HSPA5?)</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A2M LRP1</td>\n",
       "      <td>10652313 A2M LRP1, 12194978 A2M LRP1, 1702392 ...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>, LRP primary accession number Q0754 was perfo...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACE BDKRB2</td>\n",
       "      <td>10748135 ACE BDKRB2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADA DPP4</td>\n",
       "      <td>15213224 ADA DPP4</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADAM10 EFNA5</td>\n",
       "      <td>16239146 ADAM10 EFNA5</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NEW PAIR</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3540</th>\n",
       "      <td>ZG16B TLR2</td>\n",
       "      <td>20802527 ZG16B TLR2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens, Mus Musculus</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>The expression vectors pHA-mTLR2, -dominant mu...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3541</th>\n",
       "      <td>ZG16B TLR4</td>\n",
       "      <td>20802527 ZG16B TLR4</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens, Mus Musculus</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Human TLR1, 2, 3, 4, 5, 8, 9, and 10, CXCR4 an...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3542</th>\n",
       "      <td>ZG16B TLR5</td>\n",
       "      <td>20802527 ZG16B TLR5</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PAUF is a mammalian ligand identified for the ...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3543</th>\n",
       "      <td>ZG16B TLR6</td>\n",
       "      <td>20802527 ZG16B TLR6</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens, Mus Musculus</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3544</th>\n",
       "      <td>ZP3 CHRNA7</td>\n",
       "      <td>22577141 ZP3 CHRNA7</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3545 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0          LR_pair                                            triplet  \\\n",
       "0        A2M HSPA5             12194978 A2M HSPA5, 32541810 A2M HSPA5   \n",
       "1         A2M LRP1  10652313 A2M LRP1, 12194978 A2M LRP1, 1702392 ...   \n",
       "2       ACE BDKRB2                                10748135 ACE BDKRB2   \n",
       "3         ADA DPP4                                  15213224 ADA DPP4   \n",
       "4     ADAM10 EFNA5                              16239146 ADAM10 EFNA5   \n",
       "...            ...                                                ...   \n",
       "3540    ZG16B TLR2                                20802527 ZG16B TLR2   \n",
       "3541    ZG16B TLR4                                20802527 ZG16B TLR4   \n",
       "3542    ZG16B TLR5                                20802527 ZG16B TLR5   \n",
       "3543    ZG16B TLR6                                20802527 ZG16B TLR6   \n",
       "3544    ZP3 CHRNA7                                22577141 ZP3 CHRNA7   \n",
       "\n",
       "0    ligand_species_ann        receptor_species_ann used Human L used Human R  \\\n",
       "0          Homo sapiens                Homo sapiens                             \n",
       "1          Homo sapiens                Homo sapiens                             \n",
       "2          Homo sapiens                Homo sapiens                             \n",
       "3          Homo sapiens                  Bos taurus                             \n",
       "4            Bos taurus                Mus Musculus                             \n",
       "...                 ...                         ...          ...          ...   \n",
       "3540       Homo sapiens  Homo sapiens, Mus Musculus                             \n",
       "3541       Homo sapiens  Homo sapiens, Mus Musculus                             \n",
       "3542       Homo sapiens                Homo sapiens                             \n",
       "3543       Homo sapiens  Homo sapiens, Mus Musculus                             \n",
       "3544       Mus Musculus                Mus Musculus                             \n",
       "\n",
       "0                                                 notes Remove decision  \n",
       "0                    , human α2M, human Grp 78 (HSPA5?)                  \n",
       "1     , LRP primary accession number Q0754 was perfo...                  \n",
       "2                                                                        \n",
       "3                                                                        \n",
       "4                                              NEW PAIR                  \n",
       "...                                                 ...             ...  \n",
       "3540  The expression vectors pHA-mTLR2, -dominant mu...                  \n",
       "3541  Human TLR1, 2, 3, 4, 5, 8, 9, and 10, CXCR4 an...                  \n",
       "3542  PAUF is a mammalian ligand identified for the ...                  \n",
       "3543                                                                     \n",
       "3544                                                                     \n",
       "\n",
       "[3545 rows x 8 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LR_pair_flat = (\n",
    "    triplicate\n",
    "    .groupby('LR_pair', as_index=False)\n",
    "    .agg(lambda x: ', '.join(sorted(set(x.dropna().astype(str)))))\n",
    ")\n",
    "LR_pair_flat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ec6ced58-103c-4721-8e66-3a377227aa8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ligand_species_ann</th>\n",
       "      <th>receptor_species_ann</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>Homo sapiens, Mus Musculus</td>\n",
       "      <td>Homo sapiens, Mus Musculus</td>\n",
       "      <td>175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Mus Musculus</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>Homo sapiens, Homo sapiens, Mus Musculus</td>\n",
       "      <td>Homo sapiens, Mus Musculus, Homo sapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>Homo sapiens, Homo sapiens, Mus Musculus</td>\n",
       "      <td>Homo sapiens, Mus Musculus, Mus Musculus</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>Homo sapiens, Homo sapiens, Mus Musculus</td>\n",
       "      <td>Mus Musculus, Mus Musculus, Homo sapiens</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>Homo sapiens, Homo sapiens, Mus Musculus</td>\n",
       "      <td>Mus Musculus, Rattus norvegicus</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>Xenopus laevis, Homo sapiens</td>\n",
       "      <td>Homo sapiens, Xenopus laevis</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>301 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           ligand_species_ann  \\\n",
       "83                               Homo sapiens   \n",
       "247                              Mus Musculus   \n",
       "168                Homo sapiens, Mus Musculus   \n",
       "111                              Homo sapiens   \n",
       "0                                               \n",
       "..                                        ...   \n",
       "133  Homo sapiens, Homo sapiens, Mus Musculus   \n",
       "134  Homo sapiens, Homo sapiens, Mus Musculus   \n",
       "135  Homo sapiens, Homo sapiens, Mus Musculus   \n",
       "136  Homo sapiens, Homo sapiens, Mus Musculus   \n",
       "300              Xenopus laevis, Homo sapiens   \n",
       "\n",
       "                         receptor_species_ann  count  \n",
       "83                               Homo sapiens   1644  \n",
       "247                              Mus Musculus    407  \n",
       "168                Homo sapiens, Mus Musculus    175  \n",
       "111                              Mus Musculus     93  \n",
       "0                                                 87  \n",
       "..                                        ...    ...  \n",
       "133  Homo sapiens, Mus Musculus, Homo sapiens      1  \n",
       "134  Homo sapiens, Mus Musculus, Mus Musculus      1  \n",
       "135  Mus Musculus, Mus Musculus, Homo sapiens      1  \n",
       "136           Mus Musculus, Rattus norvegicus      1  \n",
       "300              Homo sapiens, Xenopus laevis      1  \n",
       "\n",
       "[301 rows x 3 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group by the two species columns and count the number of rows for each combination\n",
    "summary_counts = LR_pair_flat.groupby(['ligand_species_ann', 'receptor_species_ann']).size().reset_index(name='count')\n",
    "\n",
    "# Optional: sort by count descending\n",
    "summary_counts = summary_counts.sort_values(by='count', ascending=False)\n",
    "summary_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f71e95e9-14d1-499b-9051-9c6699c0f96c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "0b97f3cb-4024-45ad-a62e-452f6d3b859b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rows where both species columns contain 'Homo sapiens': 2444\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.7061542906674372"
      ]
     },
     "execution_count": 178,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "homo_sapiens_in_both = LR_pair_flat[\n",
    "    LR_pair_flat['ligand_species_ann'].str.contains('Homo sapiens', case=False, na=False) &\n",
    "    LR_pair_flat['receptor_species_ann'].str.contains('Homo sapiens', case=False, na=False)\n",
    "]\n",
    "\n",
    "print(\"Rows where both species columns contain 'Homo sapiens':\", homo_sapiens_in_both.shape[0])\n",
    "homo_sapiens_in_both.shape[0]/len(LR_pair_flat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "089b1a0f-ee87-4c97-a68e-78d4068c422b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "529\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_39836/2490221723.py:8: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  result = mus_musculus_in_both[\n"
     ]
    }
   ],
   "source": [
    "mus_musculus_in_both = (\n",
    "    LR_pair_flat[\n",
    "        LR_pair_flat['ligand_species_ann'].str.contains('Mus musculus', case=False, na=False) &\n",
    "        LR_pair_flat['receptor_species_ann'].str.contains('Mus musculus', case=False, na=False)\n",
    "    ]\n",
    ")\n",
    "\n",
    "result = mus_musculus_in_both[\n",
    "    ~(\n",
    "        LR_pair_flat['ligand_species_ann'].str.contains('Homo sapiens', case=False, na=False) &\n",
    "        LR_pair_flat['receptor_species_ann'].str.contains('Homo sapiens', case=False, na=False)\n",
    "    )\n",
    "]\n",
    "\n",
    "print(len(result))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "ec9881fa-2058-4b63-b815-472c4365875c",
   "metadata": {},
   "outputs": [],
   "source": [
    "result.to_csv(\"data/mouse_as_main_pairs.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c5b0eb8a-e38f-4ac1-bb45-49cdb325a3cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_57421/2028544403.py:55: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_pair_df.rename(columns=rename_dict, inplace=True)\n",
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_57421/2028544403.py:119: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  ligand_final = ligand_merge.groupby(\"orig_row\", group_keys=False).apply(resolve_ligand_row).reset_index(drop=True)\n",
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_57421/2028544403.py:138: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  final_result = receptor_merge.groupby(\"orig_row\", group_keys=False).apply(resolve_receptor_row).reset_index(drop=True).drop(columns=[\"orig_row\"])\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import os\n",
    "import pandas as pd\n",
    "import warnings\n",
    "import re\n",
    "from itables import init_notebook_mode, show\n",
    "from IPython.display import display, Javascript\n",
    "import itables.options as opt\n",
    "import createDataTable_perSpecies\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=pd.errors.DtypeWarning)\n",
    "\n",
    "# === Species Parameter === #\n",
    "species = \"horse\"  # Change to \"zebrafish\", \"sheep\", etc.\n",
    "species_file_prefix = {\n",
    "    \"mouse\": \"mmusculus\",\n",
    "    \"rat\": \"rnorvegicus\",\n",
    "    \"zebrafish\": \"drerio\",\n",
    "    \"chimpanzee\": \"ptroglodytes\",\n",
    "    \"chicken\": \"ggallus\",\n",
    "    \"pig\": \"sscrofa\",\n",
    "    \"cow\": \"btaurus\",\n",
    "    \"dog\": \"clfamiliaris\",\n",
    "    \"horse\": \"ecaballus\",\n",
    "    \"marmoset\": \"cjacchus\",\n",
    "    \"macaque\": \"mmulatta\",\n",
    "    \"sheep\": \"oarambouillet\"\n",
    "}[species]\n",
    "\n",
    "# === Load gene pair === #\n",
    "gene_pair_var = f\"{species}_gene_pair1\"\n",
    "gene_pair_df = getattr(createDataTable_perSpecies, gene_pair_var)\n",
    "ligand_ens_id = [col for col in gene_pair_df.columns if \"Ligand Ensembl ID\" in col][0]\n",
    "receptor_ens_id = [col for col in gene_pair_df.columns if \"Receptor Ensembl ID\" in col][0]\n",
    "\n",
    "\n",
    "gene_pair_df = gene_pair_df[[\n",
    "    '<span title=\"Double-click header of Interaction ID to ensure all values are shown\">Interaction ID&nbsp;</span>',\n",
    "    '<span title=\"Double-click header of Ligand to ensure all values are shown\">Ligand&nbsp;</span>',\n",
    "    '<span title=\"Double-click header of Receptor to ensure all values are shown\">Receptor&nbsp;</span>',\n",
    "    'LR Pair Card', f'{species.capitalize()} LR Pair',\n",
    "    '<span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Ligand HGNC ID&nbsp;&nbsp;</span>',\n",
    "    '<span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Receptor HGNC ID&nbsp;&nbsp;</span>',\n",
    "    'Ligand GOC score', 'Ligand WGA coverage',\n",
    "    'Ligand % Identity', 'Ligand Target % Identity',\n",
    "    'Ligand Orthology Confidence', ligand_ens_id,\n",
    "    'Receptor GOC score', 'Receptor WGA coverage', 'Receptor % Identity',\n",
    "    'Receptor Target % Identity', 'Receptor Orthology Confidence', receptor_ens_id]]\n",
    "\n",
    "# Rename columns\n",
    "rename_dict = dict(zip(gene_pair_df.columns[:7], [\n",
    "    \"Interaction ID\", \"Ligand\", \"Receptor\", \"LR Pair Card\",\n",
    "    f\"{species.capitalize()} LR Pair\", \"Ligand HGNC ID\", \"Receptor HGNC ID\"\n",
    "]))\n",
    "gene_pair_df.rename(columns=rename_dict, inplace=True)\n",
    "\n",
    "gene_pair_df = gene_pair_df.rename(columns={\n",
    "    ligand_ens_id: \"Ligand Ensembl ID\",\n",
    "    receptor_ens_id: \"Receptor Ensembl ID\"\n",
    "})\n",
    "    \n",
    "\n",
    "# Load ortholog mapping\n",
    "biomart_df = pd.read_csv(f\"data/{species_file_prefix}_ID_biomart.csv\", dtype=str)\n",
    "biomart_df = biomart_df.dropna(subset=[f\"{species_file_prefix}_homolog_ensembl_gene\", \"ensembl_gene_id\"])\n",
    "\n",
    "# Extract ID from anchor tags\n",
    "def extract_link_text(html_string):\n",
    "    match = re.search(r'<a[^>]*>(.*?)</a>', html_string)\n",
    "    return match.group(1).strip() if match else None\n",
    "\n",
    "def extract_hgnc_id(col):\n",
    "    match = re.search(r'HGNC:(\\d+)', col)\n",
    "    return 'HGNC:' + str(match.group(1)) if match else None\n",
    "def extract_paircard_id(col):\n",
    "    \"\"\"Use regular expression to extract the HGNC ID after 'cards/'.\"\"\"\n",
    "    match = re.search(r'cards/([^/]+)\\.html', col)\n",
    "    if match:\n",
    "        return str(match.group(1))\n",
    "    return None\n",
    "    \n",
    "# Process columns\n",
    "gene_pair_df['LR Pair Card'] = gene_pair_df['LR Pair Card'].apply(extract_paircard_id)\n",
    "gene_pair_df['Ligand HGNC ID'] = gene_pair_df['Ligand HGNC ID'].apply(extract_hgnc_id)\n",
    "gene_pair_df['Receptor HGNC ID'] = gene_pair_df['Receptor HGNC ID'].apply(extract_hgnc_id)\n",
    "\n",
    "# Mapping\n",
    "species_to_human_map = dict(zip(\n",
    "    biomart_df[f\"{species_file_prefix}_homolog_ensembl_gene\"],\n",
    "    biomart_df[\"ensembl_gene_id\"]\n",
    "))\n",
    "\n",
    "gene_pair_df[\"Human Ligand Ensembl ID\"] = gene_pair_df[\"Ligand Ensembl ID\"].map(species_to_human_map)\n",
    "gene_pair_df[\"Human Receptor Ensembl ID\"] = gene_pair_df[\"Receptor Ensembl ID\"].map(species_to_human_map)\n",
    "\n",
    "# Load df_merged\n",
    "merged_df = pd.read_csv(f\"data/df_merged_with_{species}_ensembl.tsv\", sep=\"\\t\")\n",
    "\n",
    "# Index for merge\n",
    "gene_pair_indexed = gene_pair_df.reset_index(drop=False).rename(columns={\"index\": \"orig_row\"})\n",
    "\n",
    "# LIGAND MERGE\n",
    "df_ligand = merged_df.add_prefix(\"Ligand_\")\n",
    "ligand_merge = gene_pair_indexed.merge(\n",
    "    df_ligand,\n",
    "    left_on=\"Human Ligand Ensembl ID\",\n",
    "    right_on=\"Ligand_human_ensembl_gene_id\",\n",
    "    how=\"left\"\n",
    ")\n",
    "if isinstance(ligand_merge.columns, pd.MultiIndex):\n",
    "    ligand_merge.columns = ligand_merge.columns.get_level_values(0)\n",
    "if ligand_merge.columns.duplicated().any():\n",
    "    ligand_merge = ligand_merge.loc[:, ~ligand_merge.columns.duplicated()]\n",
    "\n",
    "def resolve_ligand_row(group):\n",
    "    match = group[group[f\"Ligand_{species}_ensembl_gene_id\"] == group[\"Ligand Ensembl ID\"]]\n",
    "    return match.iloc[[0]] if len(match) else group.iloc[[0]]\n",
    "\n",
    "ligand_final = ligand_merge.groupby(\"orig_row\", group_keys=False).apply(resolve_ligand_row).reset_index(drop=True)\n",
    "\n",
    "# RECEPTOR MERGE\n",
    "df_receptor = merged_df.add_prefix(\"Receptor_\")\n",
    "receptor_merge = ligand_final.merge(\n",
    "    df_receptor,\n",
    "    left_on=\"Human Receptor Ensembl ID\",\n",
    "    right_on=\"Receptor_human_ensembl_gene_id\",\n",
    "    how=\"left\"\n",
    ")\n",
    "if isinstance(receptor_merge.columns, pd.MultiIndex):\n",
    "    receptor_merge.columns = receptor_merge.columns.get_level_values(0)\n",
    "if receptor_merge.columns.duplicated().any():\n",
    "    receptor_merge = receptor_merge.loc[:, ~receptor_merge.columns.duplicated()]\n",
    "\n",
    "def resolve_receptor_row(group):\n",
    "    match = group[group[f\"Receptor_{species}_ensembl_gene_id\"] == group[\"Receptor Ensembl ID\"]]\n",
    "    return match.iloc[[0]] if len(match) else group.iloc[[0]]\n",
    "\n",
    "final_result = receptor_merge.groupby(\"orig_row\", group_keys=False).apply(resolve_receptor_row).reset_index(drop=True).drop(columns=[\"orig_row\"])\n",
    "\n",
    "assert len(final_result) == len(gene_pair_df), f\"Row mismatch: {len(final_result)} != {len(gene_pair_df)}\"\n",
    "\n",
    "final_result.to_csv(f\"data/human_{species}_merged_ensemblBiomaRt_inParanoid.csv\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6a263de6-76b1-4d48-9065-9c8e3012bcd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# summarize counts depending on filter\n",
    "species = \"horse\"  # Change to desired species, e.g., \"zebrafish\", \"sheep\"\n",
    "capital_species = species.capitalize()\n",
    "\n",
    "# Load final result file\n",
    "final_result = pd.read_csv(f\"data/human_{species}_merged_ensemblBiomaRt_inParanoid.csv\")\n",
    "\n",
    "# Define score columns dynamically\n",
    "score_cols = [\n",
    "    f\"Ligand_human_inparalog_score\",\n",
    "    f\"Receptor_human_inparalog_score\",\n",
    "    f\"Ligand_{species}_inparalog_score\",\n",
    "    f\"Receptor_{species}_inparalog_score\",\n",
    "    f\"Ligand_human_seed_score\",\n",
    "    f\"Receptor_human_seed_score\",\n",
    "    f\"Ligand_{species}_seed_score\",\n",
    "    f\"Receptor_{species}_seed_score\",\n",
    "    f\"Ligand_bitscore\",\n",
    "    f\"Receptor_bitscore\"\n",
    "]\n",
    "\n",
    "for col in score_cols:\n",
    "    if col in final_result.columns:\n",
    "        final_result[col] = pd.to_numeric(final_result[col], errors='coerce')\n",
    "\n",
    "# Detect columns\n",
    "confidence_orth_ligand = [col for col in final_result.columns if \"Ligand Orthology Confidence\" in col][0]\n",
    "GOC_col_ligand = [col for col in final_result.columns if \"Ligand GOC\" in col][0]\n",
    "percIdent_col_ligand = [col for col in final_result.columns if \"Ligand % Identity\" in col][0]\n",
    "human_ligand_col = [col for col in final_result.columns if \"Ligand HGNC ID\" in col][0]\n",
    "ligand_col = [col for col in final_result.columns if \"Ligand\" in col][0]\n",
    "ligand_human_inparalog_score = [col for col in final_result.columns if \"Ligand_human_inparalog_score\" in col][0]\n",
    "ligand_species_inparalog_score = [col for col in final_result.columns if f\"Ligand_{species}_inparalog_score\" in col][0]\n",
    "ligand_human_seed_score = [col for col in final_result.columns if \"Ligand_human_seed_score\" in col][0]\n",
    "ligand_species_seed_score = [col for col in final_result.columns if f\"Ligand_{species}_seed_score\" in col][0]\n",
    "ligand_bit_score = [col for col in final_result.columns if \"Ligand_bitscore\" in col][0]\n",
    "\n",
    "human_receptor_col = [col for col in final_result.columns if \"Receptor HGNC ID\" in col][0]\n",
    "confidence_orth_receptor = [col for col in final_result.columns if \"Receptor Orthology Confidence\" in col][0]\n",
    "GOC_col_receptor = [col for col in final_result.columns if \"Receptor GOC\" in col][0]\n",
    "percIdent_col_receptor = [col for col in final_result.columns if \"Receptor % Identity\" in col][0]\n",
    "receptor_col = [col for col in final_result.columns if \"Receptor\" in col][0]\n",
    "receptor_human_inparalog_score = [col for col in final_result.columns if \"Receptor_human_inparalog_score\" in col][0]\n",
    "receptor_species_inparalog_score = [col for col in final_result.columns if f\"Receptor_{species}_inparalog_score\" in col][0]\n",
    "receptor_human_seed_score = [col for col in final_result.columns if \"Receptor_human_seed_score\" in col][0]\n",
    "receptor_species_seed_score = [col for col in final_result.columns if f\"Receptor_{species}_seed_score\" in col][0]\n",
    "receptor_bit_score = [col for col in final_result.columns if \"Receptor_bitscore\" in col][0]\n",
    "\n",
    "# Define function\n",
    "\n",
    "def summarize_orthologs(human_col, species_col, label,\n",
    "                        confidence_orth_col=None, confidence_orth_threshold=None,\n",
    "                        GOC_col=None, GOC_threshold=None,\n",
    "                        perc_identity_col=None, perc_identity_thres=None,\n",
    "                        ligand_human_inparalog_score_col=None, ligand_human_inparalog_score_threshold=None,\n",
    "                        receptor_human_inparalog_score_col=None, receptor_human_inparalog_score_threshold=None,\n",
    "                        ligand_species_inparalog_score_col=None, ligand_species_inparalog_score_threshold=None,\n",
    "                        receptor_species_inparalog_score_col=None, receptor_species_inparalog_score_threshold=None,\n",
    "                        ligand_human_seed_score_col=None, ligand_human_seed_score_threshold=None,\n",
    "                        receptor_human_seed_score_col=None, receptor_human_seed_score_threshold=None,\n",
    "                        ligand_species_seed_score_col=None, ligand_species_seed_score_threshold=None,\n",
    "                        receptor_species_seed_score_col=None, receptor_species_seed_score_threshold=None,\n",
    "                        ligand_bit_score_col=None, ligand_bit_score_threshold=None,\n",
    "                        receptor_bit_score_col=None, receptor_bit_score_threshold=None):\n",
    "\n",
    "    df = final_result.copy()\n",
    "\n",
    "    filters = [\n",
    "        (confidence_orth_col, lambda x: x == confidence_orth_threshold),\n",
    "        (GOC_col, lambda x: x >= GOC_threshold),\n",
    "        (perc_identity_col, lambda x: x >= perc_identity_thres),\n",
    "        (ligand_human_inparalog_score_col, lambda x: x >= ligand_human_inparalog_score_threshold),\n",
    "        (receptor_human_inparalog_score_col, lambda x: x >= receptor_human_inparalog_score_threshold),\n",
    "        (ligand_species_inparalog_score_col, lambda x: x >= ligand_species_inparalog_score_threshold),\n",
    "        (receptor_species_inparalog_score_col, lambda x: x >= receptor_species_inparalog_score_threshold),\n",
    "        (ligand_human_seed_score_col, lambda x: x >= ligand_human_seed_score_threshold),\n",
    "        (receptor_human_seed_score_col, lambda x: x >= receptor_human_seed_score_threshold),\n",
    "        (ligand_species_seed_score_col, lambda x: x >= ligand_species_seed_score_threshold),\n",
    "        (receptor_species_seed_score_col, lambda x: x >= receptor_species_seed_score_threshold),\n",
    "        (ligand_bit_score_col, lambda x: x >= ligand_bit_score_threshold),\n",
    "        (receptor_bit_score_col, lambda x: x >= receptor_bit_score_threshold),\n",
    "    ]\n",
    "\n",
    "    original_rows = df.shape[0]\n",
    "    for col, condition in filters:\n",
    "        if col and condition is not None:\n",
    "            before = df.shape[0]\n",
    "            df = df[df[col].apply(condition)]\n",
    "            after = df.shape[0]\n",
    "            print(f\"Filtered {col}: {before - after} rows removed (remaining: {after})\")\n",
    "\n",
    "    unique_pairs = df[[human_col, species_col]].drop_duplicates()\n",
    "\n",
    "    counts = (\n",
    "        unique_pairs\n",
    "        .groupby(human_col)[species_col]\n",
    "        .count()\n",
    "        .sort_values(ascending=False)\n",
    "        .reset_index(name='count')\n",
    "    )\n",
    "\n",
    "    filter_tag = f\"{label.lower()}\"\n",
    "    if confidence_orth_threshold is not None:\n",
    "        filter_tag += f\"_conf{confidence_orth_threshold}\"\n",
    "    if GOC_threshold is not None:\n",
    "        filter_tag += f\"_GOCge{GOC_threshold}\"\n",
    "    if ligand_human_inparalog_score_threshold is not None:\n",
    "        filter_tag += f\"_LHISge{ligand_human_inparalog_score_threshold}\"\n",
    "    if receptor_bit_score_threshold is not None:\n",
    "        filter_tag += f\"_RBSge{receptor_bit_score_threshold}\"\n",
    "\n",
    "    #counts.to_csv(f\"data/human_{species}_orth_count_{filter_tag}.csv\", index=False)\n",
    "\n",
    "    summary_counts = counts['count'].value_counts().sort_index()\n",
    "    total_human_genes = counts.shape[0]\n",
    "\n",
    "    summary_lines = [\n",
    "        f\"Out of {total_human_genes} unique human {label.lower()} genes:\",\n",
    "        f\" - Filters applied: \" + \"; \".join([\n",
    "            f\"{col} ≥ {threshold}\" for col, threshold in [\n",
    "                (confidence_orth_col, confidence_orth_threshold),\n",
    "                (GOC_col, GOC_threshold),\n",
    "                (perc_identity_col, perc_identity_thres),\n",
    "                (ligand_human_inparalog_score_col, ligand_human_inparalog_score_threshold),\n",
    "                (receptor_human_inparalog_score_col, receptor_human_inparalog_score_threshold),\n",
    "                (ligand_species_inparalog_score_col, ligand_species_inparalog_score_threshold),\n",
    "                (receptor_species_inparalog_score_col, receptor_species_inparalog_score_threshold),\n",
    "                (ligand_human_seed_score_col, ligand_human_seed_score_threshold),\n",
    "                (receptor_human_seed_score_col, receptor_human_seed_score_threshold),\n",
    "                (ligand_species_seed_score_col, ligand_species_seed_score_threshold),\n",
    "                (receptor_species_seed_score_col, receptor_species_seed_score_threshold),\n",
    "                (ligand_bit_score_col, ligand_bit_score_threshold),\n",
    "                (receptor_bit_score_col, receptor_bit_score_threshold),\n",
    "            ] if threshold is not None and col is not None\n",
    "        ])\n",
    "    ]\n",
    "\n",
    "    for orth_count, gene_count in summary_counts.items():\n",
    "        summary_lines.append(\n",
    "            f\" - {gene_count} human {label.lower()} genes had {orth_count} {species} ortholog(s)\"\n",
    "        )\n",
    "\n",
    "    return \"\\n\".join(summary_lines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "76cbb855-1166-4d4a-9615-c251d9fb036d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered Ligand Orthology Confidence: 341 rows removed (remaining: 2650)\n",
      "Filtered Ligand_bitscore: 121 rows removed (remaining: 2529)\n",
      "Out of 731 unique human ligand genes:\n",
      " - Filters applied: Ligand Orthology Confidence ≥ 1; Ligand_bitscore ≥ 40\n",
      " - 727 human ligand genes had 1 horse ortholog(s)\n",
      " - 3 human ligand genes had 2 horse ortholog(s)\n",
      " - 1 human ligand genes had 3 horse ortholog(s)\n",
      "Filtered Receptor Orthology Confidence: 163 rows removed (remaining: 2828)\n",
      "Filtered Receptor_bitscore: 92 rows removed (remaining: 2736)\n",
      "Out of 644 unique human receptor genes:\n",
      " - Filters applied: Receptor Orthology Confidence ≥ 1; Receptor_bitscore ≥ 40\n",
      " - 641 human receptor genes had 1 horse ortholog(s)\n",
      " - 1 human receptor genes had 2 horse ortholog(s)\n",
      " - 2 human receptor genes had 3 horse ortholog(s)\n"
     ]
    }
   ],
   "source": [
    "# Ligand\n",
    "ligand_summary = summarize_orthologs(\n",
    "    human_col=human_ligand_col,\n",
    "    species_col=ligand_col,\n",
    "    label=\"Ligand\",\n",
    "    # GOC_col=GOC_col_ligand,\n",
    "    # GOC_threshold=25,\n",
    "    confidence_orth_col=confidence_orth_ligand,\n",
    "    confidence_orth_threshold=1,\n",
    "    # perc_identity_col = percIdent_col_ligand,\n",
    "    # perc_identity_thres = 60\n",
    "    # ligand_human_inparalog_score_col=ligand_human_inparalog_score,\n",
    "    # ligand_human_inparalog_score_threshold=1,\n",
    "    ligand_bit_score_col=ligand_bit_score,\n",
    "    ligand_bit_score_threshold=40,\n",
    "    # ligand_species_inparalog_score_col=ligand_species_inparalog_score,\n",
    "    # ligand_species_inparalog_score_threshold=1,\n",
    ")\n",
    "print(ligand_summary)\n",
    "\n",
    "# Receptor\n",
    "receptor_summary = summarize_orthologs(\n",
    "    human_col=human_receptor_col,\n",
    "    species_col=receptor_col,\n",
    "    label=\"Receptor\",\n",
    "    # GOC_col=GOC_col_receptor,\n",
    "    # GOC_threshold=25,\n",
    "    confidence_orth_col=confidence_orth_receptor,\n",
    "    confidence_orth_threshold=1,\n",
    "    # perc_identity_col = percIdent_col_receptor,\n",
    "    # perc_identity_thres = 60\n",
    "    # receptor_human_inparalog_score_col=receptor_human_inparalog_score,\n",
    "    # receptor_human_inparalog_score_threshold=1,\n",
    "    receptor_bit_score_col=receptor_bit_score,\n",
    "    receptor_bit_score_threshold=40,\n",
    "    # receptor_species_inparalog_score_col=receptor_species_inparalog_score,\n",
    "    # receptor_species_inparalog_score_threshold=1,\n",
    ")\n",
    "print(receptor_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "aa9e07cf-9ba9-48a9-a30c-50866faafb5c",
   "metadata": {},
   "outputs": [
    {
     "ename": "EmptyDataError",
     "evalue": "No columns to parse from file",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mEmptyDataError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 41\u001b[0m\n\u001b[1;32m     38\u001b[0m r \u001b[38;5;241m=\u001b[39m requests\u001b[38;5;241m.\u001b[39mget(url)\n\u001b[1;32m     39\u001b[0m r\u001b[38;5;241m.\u001b[39mraise_for_status()\n\u001b[0;32m---> 41\u001b[0m df \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mStringIO\u001b[49m\u001b[43m(\u001b[49m\u001b[43mr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtext\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstrip\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mheader\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m     42\u001b[0m df\u001b[38;5;241m.\u001b[39mcolumns \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcluster_id\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbitscore\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msource_file\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124minparalog_score\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mprotein_id\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mseed_score\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m     44\u001b[0m \u001b[38;5;66;03m# === Step 2: Add species labels ===\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1026\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m   1013\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m   1014\u001b[0m     dialect,\n\u001b[1;32m   1015\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1022\u001b[0m     dtype_backend\u001b[38;5;241m=\u001b[39mdtype_backend,\n\u001b[1;32m   1023\u001b[0m )\n\u001b[1;32m   1024\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m-> 1026\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/io/parsers/readers.py:620\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    617\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    619\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 620\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    622\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    623\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1620\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1617\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1619\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1620\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1898\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1895\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(msg)\n\u001b[1;32m   1897\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1898\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mmapping\u001b[49m\u001b[43m[\u001b[49m\u001b[43mengine\u001b[49m\u001b[43m]\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1899\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[1;32m   1900\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/io/parsers/c_parser_wrapper.py:93\u001b[0m, in \u001b[0;36mCParserWrapper.__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m     90\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdtype_backend\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpyarrow\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m     91\u001b[0m     \u001b[38;5;66;03m# Fail here loudly instead of in cython after reading\u001b[39;00m\n\u001b[1;32m     92\u001b[0m     import_optional_dependency(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpyarrow\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m---> 93\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reader \u001b[38;5;241m=\u001b[39m \u001b[43mparsers\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mTextReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43msrc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     95\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39munnamed_cols \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reader\u001b[38;5;241m.\u001b[39munnamed_cols\n\u001b[1;32m     97\u001b[0m \u001b[38;5;66;03m# error: Cannot determine type of 'names'\u001b[39;00m\n",
      "File \u001b[0;32mparsers.pyx:581\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader.__cinit__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mEmptyDataError\u001b[0m: No columns to parse from file"
     ]
    }
   ],
   "source": [
    "\n",
    "df_orthologs.to_csv(f\"data/inParanoid_{species_input}.csv\", index=False)\n",
    "\n",
    "# === Step 4: Human UniProt → HGNC/Ensembl mapping ===\n",
    "hgnc_df = pd.read_csv(\"data/HGNC_gene_info_full.tsv\", sep=\"\\t\", dtype=str)\n",
    "hgnc_df = hgnc_df.dropna(subset=[\"uniprot_ids\", \"ensembl_gene_id\"])\n",
    "hgnc_exploded = hgnc_df.assign(uniprot_id=hgnc_df[\"uniprot_ids\"].str.split(\",\")).explode(\"uniprot_id\")\n",
    "hgnc_exploded[\"uniprot_id\"] = hgnc_exploded[\"uniprot_id\"].str.strip()\n",
    "\n",
    "df_merged = df_orthologs.merge(\n",
    "    hgnc_exploded[[\"uniprot_id\", \"ensembl_gene_id\", \"symbol\"]],\n",
    "    left_on=\"human_protein\",\n",
    "    right_on=\"uniprot_id\",\n",
    "    how=\"left\"\n",
    ")\n",
    "\n",
    "df_merged = df_merged.rename(columns={\n",
    "    \"symbol\": \"human_gene\",\n",
    "    \"ensembl_gene_id\": \"human_ensembl_gene_id\"\n",
    "}).drop(columns=[\"uniprot_id\"])\n",
    "\n",
    "df_merged = df_merged.dropna(subset=[\"human_ensembl_gene_id\"])\n",
    "df_merged.to_csv(f\"data/{species_input}_inParanoid_withHGNC.tsv\", sep=\"\\t\", index=False)\n",
    "\n",
    "# === Step 5: Optional - Species UniProt → Ensembl mapping ===\n",
    "map_path = f\"data/{species_input}_uniprot_to_ensembl.tsv\"\n",
    "try:\n",
    "    species_map = pd.read_csv(map_path, sep=\"\\t\", dtype=str)\n",
    "    df_merged = df_merged.merge(\n",
    "        species_map,\n",
    "        left_on=f\"{species_input}_protein\",\n",
    "        right_on=\"uniprotswissprot\",\n",
    "        how=\"left\"\n",
    "    ).rename(columns={\"ensembl_gene_id\": f\"{species_input}_ensembl_gene_id\"}) \\\n",
    "     .drop(columns=[\"uniprotswissprot\"])\n",
    "except FileNotFoundError:\n",
    "    print(f\"⚠️  Mapping file not found: {map_path}\")\n",
    "\n",
    "df_merged.to_csv(f\"data/df_merged_with_{species_input}_ensembl.tsv\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "id": "0ac4ea30-b41d-42aa-8337-939b5cc4fd2e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered Ligand_bitscore: 1323 rows removed (remaining: 4193)\n",
      "Out of 509 unique human ligand genes:\n",
      " - Filters applied: Ligand_bitscore ≥ 40\n",
      " - 311 human ligand genes had 1 zebrafish ortholog(s)\n",
      " - 178 human ligand genes had 2 zebrafish ortholog(s)\n",
      " - 9 human ligand genes had 3 zebrafish ortholog(s)\n",
      " - 4 human ligand genes had 4 zebrafish ortholog(s)\n",
      " - 4 human ligand genes had 5 zebrafish ortholog(s)\n",
      " - 2 human ligand genes had 7 zebrafish ortholog(s)\n",
      " - 1 human ligand genes had 8 zebrafish ortholog(s)\n",
      "Filtered Receptor_bitscore: 1293 rows removed (remaining: 4223)\n",
      "Out of 438 unique human receptor genes:\n",
      " - Filters applied: Receptor_bitscore ≥ 40\n",
      " - 273 human receptor genes had 1 zebrafish ortholog(s)\n",
      " - 139 human receptor genes had 2 zebrafish ortholog(s)\n",
      " - 12 human receptor genes had 3 zebrafish ortholog(s)\n",
      " - 7 human receptor genes had 4 zebrafish ortholog(s)\n",
      " - 3 human receptor genes had 5 zebrafish ortholog(s)\n",
      " - 3 human receptor genes had 7 zebrafish ortholog(s)\n",
      " - 1 human receptor genes had 10 zebrafish ortholog(s)\n"
     ]
    }
   ],
   "source": [
    "# Ligand\n",
    "ligand_summary = summarize_orthologs(\n",
    "    human_col=human_ligand_col,\n",
    "    species_col=ligand_col,\n",
    "    label=\"Ligand\",\n",
    "    # GOC_col=GOC_col_ligand,\n",
    "    # GOC_threshold=25,\n",
    "    # confidence_orth_col=confidence_orth_ligand,\n",
    "    # confidence_orth_threshold=1,\n",
    "    # perc_identity_col = percIdent_col_ligand,\n",
    "    # perc_identity_thres = 60\n",
    "    # ligand_human_inparalog_score_col=ligand_human_inparalog_score,\n",
    "    # ligand_human_inparalog_score_threshold=1,\n",
    "    ligand_bit_score_col=ligand_bit_score,\n",
    "    ligand_bit_score_threshold=40,\n",
    "    # ligand_species_inparalog_score_col=ligand_species_inparalog_score,\n",
    "    # ligand_species_inparalog_score_threshold=1,\n",
    ")\n",
    "print(ligand_summary)\n",
    "\n",
    "# Receptor\n",
    "receptor_summary = summarize_orthologs(\n",
    "    human_col=human_receptor_col,\n",
    "    species_col=receptor_col,\n",
    "    label=\"Receptor\",\n",
    "    # GOC_col=GOC_col_receptor,\n",
    "    # GOC_threshold=25,\n",
    "    # confidence_orth_col=confidence_orth_receptor,\n",
    "    # confidence_orth_threshold=1,\n",
    "    # perc_identity_col = percIdent_col_receptor,\n",
    "    # perc_identity_thres = 60\n",
    "    # receptor_human_inparalog_score_col=receptor_human_inparalog_score,\n",
    "    # receptor_human_inparalog_score_threshold=1,\n",
    "    receptor_bit_score_col=receptor_bit_score,\n",
    "    receptor_bit_score_threshold=40,\n",
    "    # receptor_species_inparalog_score_col=receptor_species_inparalog_score,\n",
    "    # receptor_species_inparalog_score_threshold=1,\n",
    ")\n",
    "print(receptor_summary)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 322,
   "id": "a841b2ba-ed4f-471e-ba85-262c3309c319",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "def extract_paircard_id(col):\n",
      "    \"\"\"Use regular expression to extract the HGNC ID after 'cards/'.\"\"\"\n",
      "    match = re.search(r'cards/([^/]+)\\.html', col)\n",
      "    if match:\n",
      "        return str(match.group(1))\n",
      "    return None\n",
      "\n"
     ]
    }
   ],
   "source": [
    "extract_link_text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "id": "8fecb75a-95fc-4526-b662-dd70fdd536f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Updating database/other/sheepOrth.qmd for Ovis aries rambouillet ---\n",
      "Successfully updated 'database/other/sheepOrth.qmd' for Ovis aries rambouillet.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#  'Sheep LR Pair' data\n",
    "update_connectomedb_qmd(\n",
    "    qmd_file_path=\"database/other/sheepOrth.qmd\", \n",
    "    lr_pair_data=createDataTable_perSpecies.sheep_gene_pair1[\"Sheep LR Pair\"],\n",
    "    species_name=\"Ovis aries rambouillet\",\n",
    "    species = \"Sheep\",\n",
    "    ortholog = True\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "4e48beea-e1ef-41a6-9c58-397015cb18d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_48719/1509169499.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mouse_gene_pair1['LR Pair Card'] = mouse_gene_pair1['LR Pair Card'].apply(extract_paircard_id)\n",
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_48719/1509169499.py:50: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mouse_gene_pair1[\"Human Ligand Ensembl ID\"] = mouse_gene_pair1[\"Ligand Ensembl ID\"].map(mouse_to_human_map)\n",
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_48719/1509169499.py:53: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mouse_gene_pair1[\"Human Receptor Ensembl ID\"] = mouse_gene_pair1[\"Receptor Ensembl ID\"].map(mouse_to_human_map)\n",
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_48719/1509169499.py:61: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mouse_gene_pair1['Ligand HGNC ID'] = mouse_gene_pair1['Ligand HGNC ID'].apply(extract_hgnc_id)\n",
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_48719/1509169499.py:62: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mouse_gene_pair1['Receptor HGNC ID'] = mouse_gene_pair1['Receptor HGNC ID'].apply(extract_hgnc_id)\n",
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_48719/1509169499.py:95: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(resolve_ligand_row)\n",
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_48719/1509169499.py:126: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(resolve_receptor_row)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Interaction ID</th>\n",
       "      <th>Ligand</th>\n",
       "      <th>Receptor</th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>Mouse LR Pair</th>\n",
       "      <th>Ligand HGNC ID</th>\n",
       "      <th>Receptor HGNC ID</th>\n",
       "      <th>Ligand GOC score</th>\n",
       "      <th>Ligand WGA coverage</th>\n",
       "      <th>Ligand % Identity</th>\n",
       "      <th>...</th>\n",
       "      <th>Receptor_human_protein</th>\n",
       "      <th>Receptor_human_inparalog_score</th>\n",
       "      <th>Receptor_human_seed_score</th>\n",
       "      <th>Receptor_mouse_protein</th>\n",
       "      <th>Receptor_mouse_inparalog_score</th>\n",
       "      <th>Receptor_mouse_seed_score</th>\n",
       "      <th>Receptor_bitscore</th>\n",
       "      <th>Receptor_human_ensembl_gene_id</th>\n",
       "      <th>Receptor_human_gene</th>\n",
       "      <th>Receptor_mouse_ensembl_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDB00001</td>\n",
       "      <td>A2m</td>\n",
       "      <td>Hspa5</td>\n",
       "      <td>A2M-HSPA5</td>\n",
       "      <td>A2m → Hspa5</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>HGNC:5238</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>...</td>\n",
       "      <td>P11021</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P20029</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1246.0</td>\n",
       "      <td>ENSG00000044574</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>ENSMUSG00000026864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDB00002</td>\n",
       "      <td>A2m</td>\n",
       "      <td>Lrp1</td>\n",
       "      <td>A2M-LRP1</td>\n",
       "      <td>A2m → Lrp1</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>HGNC:6692</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>...</td>\n",
       "      <td>Q07954</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q91ZX7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>6853.0</td>\n",
       "      <td>ENSG00000123384</td>\n",
       "      <td>LRP1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDB00003</td>\n",
       "      <td>Ace</td>\n",
       "      <td>Bdkrb2</td>\n",
       "      <td>ACE-BDKRB2</td>\n",
       "      <td>Ace → Bdkrb2</td>\n",
       "      <td>HGNC:2707</td>\n",
       "      <td>HGNC:1030</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>83.3078</td>\n",
       "      <td>...</td>\n",
       "      <td>P30411</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P32299</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>636.0</td>\n",
       "      <td>ENSG00000168398</td>\n",
       "      <td>BDKRB2</td>\n",
       "      <td>ENSMUSG00000021070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDB00004</td>\n",
       "      <td>Ada</td>\n",
       "      <td>Dpp4</td>\n",
       "      <td>ADA-DPP4</td>\n",
       "      <td>Ada → Dpp4</td>\n",
       "      <td>HGNC:186</td>\n",
       "      <td>HGNC:3009</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>80.7163</td>\n",
       "      <td>...</td>\n",
       "      <td>P27487</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P28843</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1365.0</td>\n",
       "      <td>ENSG00000197635</td>\n",
       "      <td>DPP4</td>\n",
       "      <td>ENSMUSG00000035000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDB00005</td>\n",
       "      <td>Adam10</td>\n",
       "      <td>Epha3</td>\n",
       "      <td>ADAM10-EPHA3</td>\n",
       "      <td>Adam10 → Epha3</td>\n",
       "      <td>HGNC:188</td>\n",
       "      <td>HGNC:3387</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>96.1230</td>\n",
       "      <td>...</td>\n",
       "      <td>P29320</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P29319</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>ENSG00000044524</td>\n",
       "      <td>EPHA3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3956</th>\n",
       "      <td>CDB03444</td>\n",
       "      <td>Pcdhb21</td>\n",
       "      <td>Pcdhb21</td>\n",
       "      <td>Pcdhb21-Pcdhb21</td>\n",
       "      <td>Pcdhb21 → Pcdhb21</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3957</th>\n",
       "      <td>CDB03445</td>\n",
       "      <td>Pcdhb22</td>\n",
       "      <td>Pcdhb22</td>\n",
       "      <td>Pcdhb22-Pcdhb22</td>\n",
       "      <td>Pcdhb22 → Pcdhb22</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Q9Y5E8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q91XZ8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.989</td>\n",
       "      <td>1191.0</td>\n",
       "      <td>ENSG00000113248</td>\n",
       "      <td>PCDHB15</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3958</th>\n",
       "      <td>CDB03446</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>Pcdhgb8-Pcdhgb8</td>\n",
       "      <td>Pcdhgb8 → Pcdhgb8</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3959</th>\n",
       "      <td>CDB03447</td>\n",
       "      <td>Saa3</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>Saa3-Tlr4</td>\n",
       "      <td>Saa3 → Tlr4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>O00206</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q9QUK6</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>909.0</td>\n",
       "      <td>ENSG00000136869</td>\n",
       "      <td>TLR4</td>\n",
       "      <td>ENSMUSG00000039005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960</th>\n",
       "      <td>CDB03448</td>\n",
       "      <td>Sema4a</td>\n",
       "      <td>Timd2</td>\n",
       "      <td>Sema4a-Timd2</td>\n",
       "      <td>Sema4a → Timd2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Q96D42</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q5QNS5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>157.0</td>\n",
       "      <td>ENSG00000113249</td>\n",
       "      <td>HAVCR1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3961 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Interaction ID   Ligand Receptor     LR Pair Card      Mouse LR Pair  \\\n",
       "0          CDB00001      A2m    Hspa5        A2M-HSPA5        A2m → Hspa5   \n",
       "1          CDB00002      A2m     Lrp1         A2M-LRP1         A2m → Lrp1   \n",
       "2          CDB00003      Ace   Bdkrb2       ACE-BDKRB2       Ace → Bdkrb2   \n",
       "3          CDB00004      Ada     Dpp4         ADA-DPP4         Ada → Dpp4   \n",
       "4          CDB00005   Adam10    Epha3     ADAM10-EPHA3     Adam10 → Epha3   \n",
       "...             ...      ...      ...              ...                ...   \n",
       "3956       CDB03444  Pcdhb21  Pcdhb21  Pcdhb21-Pcdhb21  Pcdhb21 → Pcdhb21   \n",
       "3957       CDB03445  Pcdhb22  Pcdhb22  Pcdhb22-Pcdhb22  Pcdhb22 → Pcdhb22   \n",
       "3958       CDB03446  Pcdhgb8  Pcdhgb8  Pcdhgb8-Pcdhgb8  Pcdhgb8 → Pcdhgb8   \n",
       "3959       CDB03447     Saa3     Tlr4        Saa3-Tlr4        Saa3 → Tlr4   \n",
       "3960       CDB03448   Sema4a    Timd2     Sema4a-Timd2     Sema4a → Timd2   \n",
       "\n",
       "     Ligand HGNC ID Receptor HGNC ID  Ligand GOC score  Ligand WGA coverage  \\\n",
       "0            HGNC:7        HGNC:5238              75.0                100.0   \n",
       "1            HGNC:7        HGNC:6692              75.0                100.0   \n",
       "2         HGNC:2707        HGNC:1030             100.0                100.0   \n",
       "3          HGNC:186        HGNC:3009             100.0                100.0   \n",
       "4          HGNC:188        HGNC:3387             100.0                100.0   \n",
       "...             ...              ...               ...                  ...   \n",
       "3956           None             None               NaN                  NaN   \n",
       "3957           None             None               NaN                  NaN   \n",
       "3958           None             None               NaN                  NaN   \n",
       "3959           None             None               NaN                  NaN   \n",
       "3960           None             None               NaN                  NaN   \n",
       "\n",
       "      Ligand % Identity  ...  Receptor_human_protein  \\\n",
       "0               72.4559  ...                  P11021   \n",
       "1               72.4559  ...                  Q07954   \n",
       "2               83.3078  ...                  P30411   \n",
       "3               80.7163  ...                  P27487   \n",
       "4               96.1230  ...                  P29320   \n",
       "...                 ...  ...                     ...   \n",
       "3956                NaN  ...                     NaN   \n",
       "3957                NaN  ...                  Q9Y5E8   \n",
       "3958                NaN  ...                     NaN   \n",
       "3959                NaN  ...                  O00206   \n",
       "3960                NaN  ...                  Q96D42   \n",
       "\n",
       "      Receptor_human_inparalog_score Receptor_human_seed_score  \\\n",
       "0                                1.0                       1.0   \n",
       "1                                1.0                       1.0   \n",
       "2                                1.0                       1.0   \n",
       "3                                1.0                       1.0   \n",
       "4                                1.0                       1.0   \n",
       "...                              ...                       ...   \n",
       "3956                             NaN                       NaN   \n",
       "3957                             1.0                       1.0   \n",
       "3958                             NaN                       NaN   \n",
       "3959                             1.0                       1.0   \n",
       "3960                             1.0                       1.0   \n",
       "\n",
       "      Receptor_mouse_protein  Receptor_mouse_inparalog_score  \\\n",
       "0                     P20029                             1.0   \n",
       "1                     Q91ZX7                             1.0   \n",
       "2                     P32299                             1.0   \n",
       "3                     P28843                             1.0   \n",
       "4                     P29319                             1.0   \n",
       "...                      ...                             ...   \n",
       "3956                     NaN                             NaN   \n",
       "3957                  Q91XZ8                             1.0   \n",
       "3958                     NaN                             NaN   \n",
       "3959                  Q9QUK6                             1.0   \n",
       "3960                  Q5QNS5                             1.0   \n",
       "\n",
       "      Receptor_mouse_seed_score  Receptor_bitscore  \\\n",
       "0                         1.000             1246.0   \n",
       "1                         1.000             6853.0   \n",
       "2                         1.000              636.0   \n",
       "3                         1.000             1365.0   \n",
       "4                         1.000             1923.0   \n",
       "...                         ...                ...   \n",
       "3956                        NaN                NaN   \n",
       "3957                      0.989             1191.0   \n",
       "3958                        NaN                NaN   \n",
       "3959                      1.000              909.0   \n",
       "3960                      1.000              157.0   \n",
       "\n",
       "      Receptor_human_ensembl_gene_id Receptor_human_gene  \\\n",
       "0                    ENSG00000044574               HSPA5   \n",
       "1                    ENSG00000123384                LRP1   \n",
       "2                    ENSG00000168398              BDKRB2   \n",
       "3                    ENSG00000197635                DPP4   \n",
       "4                    ENSG00000044524               EPHA3   \n",
       "...                              ...                 ...   \n",
       "3956                             NaN                 NaN   \n",
       "3957                 ENSG00000113248             PCDHB15   \n",
       "3958                             NaN                 NaN   \n",
       "3959                 ENSG00000136869                TLR4   \n",
       "3960                 ENSG00000113249              HAVCR1   \n",
       "\n",
       "     Receptor_mouse_ensembl_gene_id  \n",
       "0                ENSMUSG00000026864  \n",
       "1                               NaN  \n",
       "2                ENSMUSG00000021070  \n",
       "3                ENSMUSG00000035000  \n",
       "4                               NaN  \n",
       "...                             ...  \n",
       "3956                            NaN  \n",
       "3957                            NaN  \n",
       "3958                            NaN  \n",
       "3959             ENSMUSG00000039005  \n",
       "3960                            NaN  \n",
       "\n",
       "[3961 rows x 43 columns]"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys\n",
    "import os\n",
    "import pandas as pd\n",
    "import warnings\n",
    "import re\n",
    "warnings.filterwarnings(\"ignore\", category=pd.errors.DtypeWarning)\n",
    "from itables import init_notebook_mode, show\n",
    "from IPython.display import display, Javascript\n",
    "import itables.options as opt\n",
    "from createDataTable_perSpecies import mouse_gene_pair1\n",
    "\n",
    "mouse_gene_pair1= mouse_gene_pair1[['<span title=\"Double-click header of Interaction ID to ensure all values are shown\">Interaction ID&nbsp;</span>','<span title=\"Double-click header of Ligand to ensure all values are shown\">Ligand&nbsp;</span>',\n",
    "       '<span title=\"Double-click header of Receptor to ensure all values are shown\">Receptor&nbsp;</span>',\n",
    "                  'LR Pair Card', 'Mouse LR Pair','<span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Ligand HGNC ID&nbsp;&nbsp;</span>',\n",
    "       '<span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Receptor HGNC ID&nbsp;&nbsp;</span>','Ligand GOC score', 'Ligand WGA coverage',\n",
    "       'Ligand % Identity', 'Ligand Target % Identity',\n",
    "       'Ligand Orthology Confidence', 'Ligand Ensembl ID',\n",
    "       'Receptor GOC score', 'Receptor WGA coverage', 'Receptor % Identity',\n",
    "       'Receptor Target % Identity', 'Receptor Orthology Confidence','Receptor Ensembl ID']]\n",
    "\n",
    "mouse_gene_pair1.columns = [\n",
    "    \"Interaction ID\",\n",
    "    \"Ligand\",\n",
    "    \"Receptor\",\n",
    "    \"LR Pair Card\", \n",
    "    \"Mouse LR Pair\",\n",
    "    \"Ligand HGNC ID\",\n",
    "    \"Receptor HGNC ID\",\n",
    "    *mouse_gene_pair1.columns[7:]\n",
    "]\n",
    "mousebioM_df = pd.read_csv(\"data/mmusculus_ID_biomart.csv\", dtype=str)\n",
    "mousebioM_df = mousebioM_df.dropna(subset=[\"mmusculus_homolog_ensembl_gene\", \"ensembl_gene_id\"])\n",
    "\n",
    "\n",
    "def extract_link_text(html_string):\n",
    "    \"\"\"Extract visible text from an anchor tag <a>...</a>.\"\"\"\n",
    "    match = re.search(r'<a[^>]*>(.*?)</a>', html_string)\n",
    "    if match:\n",
    "        return match.group(1).strip()\n",
    "    return None\n",
    "\n",
    "mouse_gene_pair1['LR Pair Card'] = mouse_gene_pair1['LR Pair Card'].apply(extract_paircard_id)\n",
    "# Create the mapping dictionary from mouse to human Ensembl gene ID\n",
    "mouse_to_human_map = dict(zip(\n",
    "    mousebioM_df[\"mmusculus_homolog_ensembl_gene\"],\n",
    "    mousebioM_df[\"ensembl_gene_id\"]\n",
    "))\n",
    "\n",
    "# Map Ligand\n",
    "mouse_gene_pair1[\"Human Ligand Ensembl ID\"] = mouse_gene_pair1[\"Ligand Ensembl ID\"].map(mouse_to_human_map)\n",
    "\n",
    "# Map Receptor\n",
    "mouse_gene_pair1[\"Human Receptor Ensembl ID\"] = mouse_gene_pair1[\"Receptor Ensembl ID\"].map(mouse_to_human_map)\n",
    "def extract_hgnc_id(col):\n",
    "    \"\"\"Use regular expression to extract the HGNC ID after 'HGNC:'.\"\"\"\n",
    "    match = re.search(r'HGNC:(\\d+)', col)\n",
    "    if match:\n",
    "        return 'HGNC:' +str(match.group(1))\n",
    "    return None\n",
    "    \n",
    "mouse_gene_pair1['Ligand HGNC ID'] = mouse_gene_pair1['Ligand HGNC ID'].apply(extract_hgnc_id)\n",
    "mouse_gene_pair1['Receptor HGNC ID'] = mouse_gene_pair1['Receptor HGNC ID'].apply(extract_hgnc_id)\n",
    "df_merged =pd.read_csv(\"data/df_merged_with_mouse_ensembl.tsv\",sep=\"\\t\")\n",
    "\n",
    "# Step 0: Add 'orig_row' once, keep it clean\n",
    "mouse_gene_pair1_indexed = mouse_gene_pair1.reset_index(drop=False).rename(columns={\"index\": \"orig_row\"})\n",
    "\n",
    "### === LIGAND MERGE === ###\n",
    "df_ligand = df_merged.add_prefix(\"Ligand_\")\n",
    "ligand_merge = mouse_gene_pair1_indexed.merge(\n",
    "    df_ligand,\n",
    "    left_on=\"Human Ligand Ensembl ID\",\n",
    "    right_on=\"Ligand_human_ensembl_gene_id\",\n",
    "    how=\"left\"\n",
    ")\n",
    "\n",
    "# Ensure orig_row is single column (sometimes merge creates duplicates with suffix)\n",
    "if isinstance(ligand_merge.columns, pd.MultiIndex):\n",
    "    ligand_merge.columns = ligand_merge.columns.get_level_values(0)\n",
    "\n",
    "if ligand_merge.columns.duplicated().any():\n",
    "    ligand_merge = ligand_merge.loc[:, ~ligand_merge.columns.duplicated()]\n",
    "\n",
    "def resolve_ligand_row(group):\n",
    "    match = group[group[\"Ligand_mouse_ensembl_gene_id\"] == group[\"Ligand Ensembl ID\"]]\n",
    "    if len(match) == 1:\n",
    "        return match\n",
    "    elif len(match) > 1:\n",
    "        return match.iloc[[0]]\n",
    "    else:\n",
    "        return group.iloc[[0]]\n",
    "\n",
    "ligand_final = (\n",
    "    ligand_merge.groupby(\"orig_row\", group_keys=False)\n",
    "    .apply(resolve_ligand_row)\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "### === RECEPTOR MERGE === ###\n",
    "df_receptor = df_merged.add_prefix(\"Receptor_\")\n",
    "receptor_merge = ligand_final.merge(\n",
    "    df_receptor,\n",
    "    left_on=\"Human Receptor Ensembl ID\",\n",
    "    right_on=\"Receptor_human_ensembl_gene_id\",\n",
    "    how=\"left\"\n",
    ")\n",
    "\n",
    "# Same cleanup for receptor_merge\n",
    "if isinstance(receptor_merge.columns, pd.MultiIndex):\n",
    "    receptor_merge.columns = receptor_merge.columns.get_level_values(0)\n",
    "\n",
    "if receptor_merge.columns.duplicated().any():\n",
    "    receptor_merge = receptor_merge.loc[:, ~receptor_merge.columns.duplicated()]\n",
    "\n",
    "def resolve_receptor_row(group):\n",
    "    match = group[group[\"Receptor_mouse_ensembl_gene_id\"] == group[\"Receptor Ensembl ID\"]]\n",
    "    if len(match) == 1:\n",
    "        return match\n",
    "    elif len(match) > 1:\n",
    "        return match.iloc[[0]]\n",
    "    else:\n",
    "        return group.iloc[[0]]\n",
    "\n",
    "final_result = (\n",
    "    receptor_merge.groupby(\"orig_row\", group_keys=False)\n",
    "    .apply(resolve_receptor_row)\n",
    "    .reset_index(drop=True)\n",
    "    .drop(columns=[\"orig_row\"])\n",
    ")\n",
    "\n",
    "assert len(final_result) == len(mouse_gene_pair1), f\"Row mismatch: {len(final_result)} != {len(mouse_gene_pair1)}\"\n",
    "final_result\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "8022429c-c21d-43a7-9f96-4c7edcaad092",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_result.to_csv(\"data/human_mouse_merged_ensemblBiomaRt_inParanoid.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "73315d82-2ed6-4f33-b701-357ebb6004df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['<span title=\"Double-click header of Interaction ID to ensure all values are shown\">Interaction ID&nbsp;</span>',\n",
       "       'LR Pair Card', 'Mouse LR Pair',\n",
       "       '<span title=\"Double-click header of Ligand to ensure all values are shown\">Ligand&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Receptor to ensure all values are shown\">Receptor&nbsp;</span>',\n",
       "       '<span title=\"Genome Informatics (MGI) ID. Click on the link for more details\">Ligand MGI ID</span>',\n",
       "       '<span title=\"Genome Informatics (MGI) ID. Click on the link for more details\">Receptor MGI ID</span>',\n",
       "       'Ligand Ensembl ID', 'Ligand GOC score', 'Ligand WGA coverage',\n",
       "       'Ligand % Identity', 'Ligand Target % Identity',\n",
       "       'Ligand Orthology Confidence', 'Ligand Name', 'Receptor Ensembl ID',\n",
       "       'Receptor GOC score', 'Receptor WGA coverage', 'Receptor % Identity',\n",
       "       'Receptor Target % Identity', 'Receptor Orthology Confidence',\n",
       "       'Receptor Name',\n",
       "       '<span title=\"Official Gene Symbol; Hover on symbols below to show gene names\"\">Human Ligand&nbsp;&nbsp;&nbsp;</span</span>',\n",
       "       '<span title=\"Official Gene Symbol; Hover on symbols below to show gene names\"\">Human Receptor&nbsp;&nbsp;&nbsp;</span</span>',\n",
       "       '<span title=\"Double-click header of Ligand Symbols to ensure all values are shown\"\">Human Ligand Symbols&nbsp;</span</span>',\n",
       "       '<span title=\"Double-click header of Receptor Symbols to ensure all values are shown\"\">Human Receptor Symbols&nbsp;</span</span>',\n",
       "       '<span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\">Ligand Location</span>',\n",
       "       '<span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\">Receptor Location</span>',\n",
       "       '<span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Ligand HGNC ID&nbsp;&nbsp;</span>',\n",
       "       '<span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Receptor HGNC ID&nbsp;&nbsp;</span>',\n",
       "       '<span title=\"Click the logo below to run Perplexity on the Human LR pair\">Perplexity&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Database Source to ensure all values are shown\">Database Source&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Binding Location to ensure all values are shown\">Binding Location&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Trans-binding to ensure all values are shown\">Trans-binding&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Bidirectional Signalling to ensure all values are shown\">Bidirectional Signalling&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Interaction Type to ensure all values are shown\">Interaction Type&nbsp;</span>'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse_gene_pair1.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "6aaa4098-f4d0-4298-b36e-b83b540e5d4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_gene_pair1= mouse_gene_pair1[['<span title=\"Double-click header of Interaction ID to ensure all values are shown\">Interaction ID&nbsp;</span>',\n",
    "                  'LR Pair Card', 'Mouse LR Pair','<span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Ligand HGNC ID&nbsp;&nbsp;</span>',\n",
    "       '<span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Receptor HGNC ID&nbsp;&nbsp;</span>','Ligand GOC score', 'Ligand WGA coverage',\n",
    "       'Ligand % Identity', 'Ligand Target % Identity',\n",
    "       'Ligand Orthology Confidence', 'Ligand Ensembl ID',\n",
    "       'Receptor GOC score', 'Receptor WGA coverage', 'Receptor % Identity',\n",
    "       'Receptor Target % Identity', 'Receptor Orthology Confidence','Receptor Ensembl ID']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "a2bb1738-965e-4d9a-96cf-599f8fa89778",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Interaction ID</th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>Mouse LR Pair</th>\n",
       "      <th>Ligand HGNC ID</th>\n",
       "      <th>Receptor HGNC ID</th>\n",
       "      <th>Ligand GOC score</th>\n",
       "      <th>Ligand WGA coverage</th>\n",
       "      <th>Ligand % Identity</th>\n",
       "      <th>Ligand Target % Identity</th>\n",
       "      <th>Ligand Orthology Confidence</th>\n",
       "      <th>Ligand Ensembl ID</th>\n",
       "      <th>Receptor GOC score</th>\n",
       "      <th>Receptor WGA coverage</th>\n",
       "      <th>Receptor % Identity</th>\n",
       "      <th>Receptor Target % Identity</th>\n",
       "      <th>Receptor Orthology Confidence</th>\n",
       "      <th>Receptor Ensembl ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDB00001</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>A2m → Hspa5</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000030111</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>98.6239</td>\n",
       "      <td>98.4733</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000026864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDB00002</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>A2m → Lrp1</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000030111</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>97.9974</td>\n",
       "      <td>97.9758</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000040249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDB00003</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Ace → Bdkrb2</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>83.3078</td>\n",
       "      <td>82.9268</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000020681</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>79.7954</td>\n",
       "      <td>79.5918</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000021070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDB00004</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Ada → Dpp4</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>80.7163</td>\n",
       "      <td>83.2386</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000017697</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>84.4648</td>\n",
       "      <td>85.1316</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000035000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDB00005</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Adam10 → Epha3</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>96.1230</td>\n",
       "      <td>95.9947</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000054693</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>96.5412</td>\n",
       "      <td>96.4431</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000052504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3956</th>\n",
       "      <td>CDB03444</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Pcdhb21 → Pcdhb21</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000044022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000044022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3957</th>\n",
       "      <td>CDB03445</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Pcdhb22 → Pcdhb22</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000073591</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000073591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3958</th>\n",
       "      <td>CDB03446</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Pcdhgb8 → Pcdhgb8</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000103081</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000103081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3959</th>\n",
       "      <td>CDB03447</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Saa3 → Tlr4</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000040026</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000039005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960</th>\n",
       "      <td>CDB03448</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Sema4a → Timd2</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000028064</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000040413</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3961 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Interaction ID                                       LR Pair Card  \\\n",
       "0          CDB00001  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "1          CDB00002  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "2          CDB00003  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "3          CDB00004  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4          CDB00005  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "...             ...                                                ...   \n",
       "3956       CDB03444  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "3957       CDB03445  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "3958       CDB03446  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "3959       CDB03447  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "3960       CDB03448  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "\n",
       "          Mouse LR Pair                                     Ligand HGNC ID  \\\n",
       "0           A2m → Hspa5  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "1            A2m → Lrp1  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "2          Ace → Bdkrb2  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3            Ada → Dpp4  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4        Adam10 → Epha3  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "...                 ...                                                ...   \n",
       "3956  Pcdhb21 → Pcdhb21  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3957  Pcdhb22 → Pcdhb22  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3958  Pcdhgb8 → Pcdhgb8  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3959        Saa3 → Tlr4  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3960     Sema4a → Timd2  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "\n",
       "                                       Receptor HGNC ID  Ligand GOC score  \\\n",
       "0     <a href=\"https://www.genenames.org/data/gene-s...              75.0   \n",
       "1     <a href=\"https://www.genenames.org/data/gene-s...              75.0   \n",
       "2     <a href=\"https://www.genenames.org/data/gene-s...             100.0   \n",
       "3     <a href=\"https://www.genenames.org/data/gene-s...             100.0   \n",
       "4     <a href=\"https://www.genenames.org/data/gene-s...             100.0   \n",
       "...                                                 ...               ...   \n",
       "3956  <a href=\"https://www.genenames.org/data/gene-s...               NaN   \n",
       "3957  <a href=\"https://www.genenames.org/data/gene-s...               NaN   \n",
       "3958  <a href=\"https://www.genenames.org/data/gene-s...               NaN   \n",
       "3959  <a href=\"https://www.genenames.org/data/gene-s...               NaN   \n",
       "3960  <a href=\"https://www.genenames.org/data/gene-s...               NaN   \n",
       "\n",
       "      Ligand WGA coverage  Ligand % Identity  Ligand Target % Identity  \\\n",
       "0                   100.0            72.4559                   72.4559   \n",
       "1                   100.0            72.4559                   72.4559   \n",
       "2                   100.0            83.3078                   82.9268   \n",
       "3                   100.0            80.7163                   83.2386   \n",
       "4                   100.0            96.1230                   95.9947   \n",
       "...                   ...                ...                       ...   \n",
       "3956                  NaN                NaN                       NaN   \n",
       "3957                  NaN                NaN                       NaN   \n",
       "3958                  NaN                NaN                       NaN   \n",
       "3959                  NaN                NaN                       NaN   \n",
       "3960                  NaN                NaN                       NaN   \n",
       "\n",
       "      Ligand Orthology Confidence   Ligand Ensembl ID  Receptor GOC score  \\\n",
       "0                             1.0  ENSMUSG00000030111                75.0   \n",
       "1                             1.0  ENSMUSG00000030111               100.0   \n",
       "2                             1.0  ENSMUSG00000020681               100.0   \n",
       "3                             1.0  ENSMUSG00000017697               100.0   \n",
       "4                             1.0  ENSMUSG00000054693               100.0   \n",
       "...                           ...                 ...                 ...   \n",
       "3956                          NaN  ENSMUSG00000044022                 NaN   \n",
       "3957                          NaN  ENSMUSG00000073591                 NaN   \n",
       "3958                          NaN  ENSMUSG00000103081                 NaN   \n",
       "3959                          NaN  ENSMUSG00000040026                 NaN   \n",
       "3960                          NaN  ENSMUSG00000028064                 NaN   \n",
       "\n",
       "      Receptor WGA coverage  Receptor % Identity  Receptor Target % Identity  \\\n",
       "0                     100.0              98.6239                     98.4733   \n",
       "1                     100.0              97.9974                     97.9758   \n",
       "2                     100.0              79.7954                     79.5918   \n",
       "3                     100.0              84.4648                     85.1316   \n",
       "4                     100.0              96.5412                     96.4431   \n",
       "...                     ...                  ...                         ...   \n",
       "3956                    NaN                  NaN                         NaN   \n",
       "3957                    NaN                  NaN                         NaN   \n",
       "3958                    NaN                  NaN                         NaN   \n",
       "3959                    NaN                  NaN                         NaN   \n",
       "3960                    NaN                  NaN                         NaN   \n",
       "\n",
       "      Receptor Orthology Confidence Receptor Ensembl ID  \n",
       "0                               1.0  ENSMUSG00000026864  \n",
       "1                               1.0  ENSMUSG00000040249  \n",
       "2                               1.0  ENSMUSG00000021070  \n",
       "3                               1.0  ENSMUSG00000035000  \n",
       "4                               1.0  ENSMUSG00000052504  \n",
       "...                             ...                 ...  \n",
       "3956                            NaN  ENSMUSG00000044022  \n",
       "3957                            NaN  ENSMUSG00000073591  \n",
       "3958                            NaN  ENSMUSG00000103081  \n",
       "3959                            NaN  ENSMUSG00000039005  \n",
       "3960                            NaN  ENSMUSG00000040413  \n",
       "\n",
       "[3961 rows x 17 columns]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse_gene_pair1.columns = [\n",
    "    \"Interaction ID\",\n",
    "    \"LR Pair Card\",\n",
    "    \"Mouse LR Pair\",\n",
    "    \"Ligand HGNC ID\",\n",
    "    \"Receptor HGNC ID\",\n",
    "    *mouse_gene_pair1.columns[5:]\n",
    "]\n",
    "mouse_gene_pair1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "59de536e-d225-4b6c-8c70-7d03f64420fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "mousebioM_df = pd.read_csv(\"data/mmusculus_ID_biomart.csv\", dtype=str)\n",
    "mousebioM_df = mousebioM_df.dropna(subset=[\"mmusculus_homolog_ensembl_gene\", \"ensembl_gene_id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "750ac985-b986-42bc-84a1-aa44abe37439",
   "metadata": {},
   "outputs": [],
   "source": [
    "mousebioM_df= mousebioM_df[[\"mmusculus_homolog_ensembl_gene\", \"ensembl_gene_id\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "f68c344d-ceec-4c03-998e-7a50bef6c6b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_48719/3523008023.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mouse_gene_pair1['LR Pair Card'] = mouse_gene_pair1['LR Pair Card'].apply(extract_paircard_id)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Interaction ID</th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>Mouse LR Pair</th>\n",
       "      <th>Ligand HGNC ID</th>\n",
       "      <th>Receptor HGNC ID</th>\n",
       "      <th>Ligand GOC score</th>\n",
       "      <th>Ligand WGA coverage</th>\n",
       "      <th>Ligand % Identity</th>\n",
       "      <th>Ligand Target % Identity</th>\n",
       "      <th>Ligand Orthology Confidence</th>\n",
       "      <th>Ligand Ensembl ID</th>\n",
       "      <th>Receptor GOC score</th>\n",
       "      <th>Receptor WGA coverage</th>\n",
       "      <th>Receptor % Identity</th>\n",
       "      <th>Receptor Target % Identity</th>\n",
       "      <th>Receptor Orthology Confidence</th>\n",
       "      <th>Receptor Ensembl ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDB00001</td>\n",
       "      <td>A2M-HSPA5</td>\n",
       "      <td>A2m → Hspa5</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000030111</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>98.6239</td>\n",
       "      <td>98.4733</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000026864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDB00002</td>\n",
       "      <td>A2M-LRP1</td>\n",
       "      <td>A2m → Lrp1</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000030111</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>97.9974</td>\n",
       "      <td>97.9758</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000040249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDB00003</td>\n",
       "      <td>ACE-BDKRB2</td>\n",
       "      <td>Ace → Bdkrb2</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>83.3078</td>\n",
       "      <td>82.9268</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000020681</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>79.7954</td>\n",
       "      <td>79.5918</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000021070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDB00004</td>\n",
       "      <td>ADA-DPP4</td>\n",
       "      <td>Ada → Dpp4</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>80.7163</td>\n",
       "      <td>83.2386</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000017697</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>84.4648</td>\n",
       "      <td>85.1316</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000035000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDB00005</td>\n",
       "      <td>ADAM10-EPHA3</td>\n",
       "      <td>Adam10 → Epha3</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>96.1230</td>\n",
       "      <td>95.9947</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000054693</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>96.5412</td>\n",
       "      <td>96.4431</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000052504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3956</th>\n",
       "      <td>CDB03444</td>\n",
       "      <td>Pcdhb21-Pcdhb21</td>\n",
       "      <td>Pcdhb21 → Pcdhb21</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000044022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000044022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3957</th>\n",
       "      <td>CDB03445</td>\n",
       "      <td>Pcdhb22-Pcdhb22</td>\n",
       "      <td>Pcdhb22 → Pcdhb22</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000073591</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000073591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3958</th>\n",
       "      <td>CDB03446</td>\n",
       "      <td>Pcdhgb8-Pcdhgb8</td>\n",
       "      <td>Pcdhgb8 → Pcdhgb8</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000103081</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000103081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3959</th>\n",
       "      <td>CDB03447</td>\n",
       "      <td>Saa3-Tlr4</td>\n",
       "      <td>Saa3 → Tlr4</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000040026</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000039005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960</th>\n",
       "      <td>CDB03448</td>\n",
       "      <td>Sema4a-Timd2</td>\n",
       "      <td>Sema4a → Timd2</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000028064</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000040413</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3961 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Interaction ID     LR Pair Card      Mouse LR Pair  \\\n",
       "0          CDB00001        A2M-HSPA5        A2m → Hspa5   \n",
       "1          CDB00002         A2M-LRP1         A2m → Lrp1   \n",
       "2          CDB00003       ACE-BDKRB2       Ace → Bdkrb2   \n",
       "3          CDB00004         ADA-DPP4         Ada → Dpp4   \n",
       "4          CDB00005     ADAM10-EPHA3     Adam10 → Epha3   \n",
       "...             ...              ...                ...   \n",
       "3956       CDB03444  Pcdhb21-Pcdhb21  Pcdhb21 → Pcdhb21   \n",
       "3957       CDB03445  Pcdhb22-Pcdhb22  Pcdhb22 → Pcdhb22   \n",
       "3958       CDB03446  Pcdhgb8-Pcdhgb8  Pcdhgb8 → Pcdhgb8   \n",
       "3959       CDB03447        Saa3-Tlr4        Saa3 → Tlr4   \n",
       "3960       CDB03448     Sema4a-Timd2     Sema4a → Timd2   \n",
       "\n",
       "                                         Ligand HGNC ID  \\\n",
       "0     <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "1     <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "2     <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3     <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4     <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "...                                                 ...   \n",
       "3956  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3957  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3958  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3959  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3960  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "\n",
       "                                       Receptor HGNC ID  Ligand GOC score  \\\n",
       "0     <a href=\"https://www.genenames.org/data/gene-s...              75.0   \n",
       "1     <a href=\"https://www.genenames.org/data/gene-s...              75.0   \n",
       "2     <a href=\"https://www.genenames.org/data/gene-s...             100.0   \n",
       "3     <a href=\"https://www.genenames.org/data/gene-s...             100.0   \n",
       "4     <a href=\"https://www.genenames.org/data/gene-s...             100.0   \n",
       "...                                                 ...               ...   \n",
       "3956  <a href=\"https://www.genenames.org/data/gene-s...               NaN   \n",
       "3957  <a href=\"https://www.genenames.org/data/gene-s...               NaN   \n",
       "3958  <a href=\"https://www.genenames.org/data/gene-s...               NaN   \n",
       "3959  <a href=\"https://www.genenames.org/data/gene-s...               NaN   \n",
       "3960  <a href=\"https://www.genenames.org/data/gene-s...               NaN   \n",
       "\n",
       "      Ligand WGA coverage  Ligand % Identity  Ligand Target % Identity  \\\n",
       "0                   100.0            72.4559                   72.4559   \n",
       "1                   100.0            72.4559                   72.4559   \n",
       "2                   100.0            83.3078                   82.9268   \n",
       "3                   100.0            80.7163                   83.2386   \n",
       "4                   100.0            96.1230                   95.9947   \n",
       "...                   ...                ...                       ...   \n",
       "3956                  NaN                NaN                       NaN   \n",
       "3957                  NaN                NaN                       NaN   \n",
       "3958                  NaN                NaN                       NaN   \n",
       "3959                  NaN                NaN                       NaN   \n",
       "3960                  NaN                NaN                       NaN   \n",
       "\n",
       "      Ligand Orthology Confidence   Ligand Ensembl ID  Receptor GOC score  \\\n",
       "0                             1.0  ENSMUSG00000030111                75.0   \n",
       "1                             1.0  ENSMUSG00000030111               100.0   \n",
       "2                             1.0  ENSMUSG00000020681               100.0   \n",
       "3                             1.0  ENSMUSG00000017697               100.0   \n",
       "4                             1.0  ENSMUSG00000054693               100.0   \n",
       "...                           ...                 ...                 ...   \n",
       "3956                          NaN  ENSMUSG00000044022                 NaN   \n",
       "3957                          NaN  ENSMUSG00000073591                 NaN   \n",
       "3958                          NaN  ENSMUSG00000103081                 NaN   \n",
       "3959                          NaN  ENSMUSG00000040026                 NaN   \n",
       "3960                          NaN  ENSMUSG00000028064                 NaN   \n",
       "\n",
       "      Receptor WGA coverage  Receptor % Identity  Receptor Target % Identity  \\\n",
       "0                     100.0              98.6239                     98.4733   \n",
       "1                     100.0              97.9974                     97.9758   \n",
       "2                     100.0              79.7954                     79.5918   \n",
       "3                     100.0              84.4648                     85.1316   \n",
       "4                     100.0              96.5412                     96.4431   \n",
       "...                     ...                  ...                         ...   \n",
       "3956                    NaN                  NaN                         NaN   \n",
       "3957                    NaN                  NaN                         NaN   \n",
       "3958                    NaN                  NaN                         NaN   \n",
       "3959                    NaN                  NaN                         NaN   \n",
       "3960                    NaN                  NaN                         NaN   \n",
       "\n",
       "      Receptor Orthology Confidence Receptor Ensembl ID  \n",
       "0                               1.0  ENSMUSG00000026864  \n",
       "1                               1.0  ENSMUSG00000040249  \n",
       "2                               1.0  ENSMUSG00000021070  \n",
       "3                               1.0  ENSMUSG00000035000  \n",
       "4                               1.0  ENSMUSG00000052504  \n",
       "...                             ...                 ...  \n",
       "3956                            NaN  ENSMUSG00000044022  \n",
       "3957                            NaN  ENSMUSG00000073591  \n",
       "3958                            NaN  ENSMUSG00000103081  \n",
       "3959                            NaN  ENSMUSG00000039005  \n",
       "3960                            NaN  ENSMUSG00000040413  \n",
       "\n",
       "[3961 rows x 17 columns]"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "def extract_link_text(html_string):\n",
    "    \"\"\"Extract visible text from an anchor tag <a>...</a>.\"\"\"\n",
    "    match = re.search(r'<a[^>]*>(.*?)</a>', html_string)\n",
    "    if match:\n",
    "        return match.group(1).strip()\n",
    "    return None\n",
    "\n",
    "mouse_gene_pair1['LR Pair Card'] = mouse_gene_pair1['LR Pair Card'].apply(extract_paircard_id)\n",
    "mouse_gene_pair1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "29bc8a24-4281-4f3b-9e89-b076cbd9368a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_48719/1451020785.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mouse_gene_pair1[\"Human Ligand Ensembl ID\"] = mouse_gene_pair1[\"Ligand Ensembl ID\"].map(mouse_to_human_map)\n",
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_48719/1451020785.py:11: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mouse_gene_pair1[\"Human Receptor Ensembl ID\"] = mouse_gene_pair1[\"Receptor Ensembl ID\"].map(mouse_to_human_map)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Interaction ID</th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>Mouse LR Pair</th>\n",
       "      <th>Ligand HGNC ID</th>\n",
       "      <th>Receptor HGNC ID</th>\n",
       "      <th>Ligand GOC score</th>\n",
       "      <th>Ligand WGA coverage</th>\n",
       "      <th>Ligand % Identity</th>\n",
       "      <th>Ligand Target % Identity</th>\n",
       "      <th>Ligand Orthology Confidence</th>\n",
       "      <th>Ligand Ensembl ID</th>\n",
       "      <th>Receptor GOC score</th>\n",
       "      <th>Receptor WGA coverage</th>\n",
       "      <th>Receptor % Identity</th>\n",
       "      <th>Receptor Target % Identity</th>\n",
       "      <th>Receptor Orthology Confidence</th>\n",
       "      <th>Receptor Ensembl ID</th>\n",
       "      <th>Human Ligand Ensembl ID</th>\n",
       "      <th>Human Receptor Ensembl ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDB00001</td>\n",
       "      <td>A2M-HSPA5</td>\n",
       "      <td>A2m → Hspa5</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000030111</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>98.6239</td>\n",
       "      <td>98.4733</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000026864</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>ENSG00000044574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDB00002</td>\n",
       "      <td>A2M-LRP1</td>\n",
       "      <td>A2m → Lrp1</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000030111</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>97.9974</td>\n",
       "      <td>97.9758</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000040249</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>ENSG00000123384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDB00003</td>\n",
       "      <td>ACE-BDKRB2</td>\n",
       "      <td>Ace → Bdkrb2</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>83.3078</td>\n",
       "      <td>82.9268</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000020681</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>79.7954</td>\n",
       "      <td>79.5918</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000021070</td>\n",
       "      <td>ENSG00000159640</td>\n",
       "      <td>ENSG00000168398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDB00004</td>\n",
       "      <td>ADA-DPP4</td>\n",
       "      <td>Ada → Dpp4</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>80.7163</td>\n",
       "      <td>83.2386</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000017697</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>84.4648</td>\n",
       "      <td>85.1316</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000035000</td>\n",
       "      <td>ENSG00000196839</td>\n",
       "      <td>ENSG00000197635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDB00005</td>\n",
       "      <td>ADAM10-EPHA3</td>\n",
       "      <td>Adam10 → Epha3</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>96.1230</td>\n",
       "      <td>95.9947</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000054693</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>96.5412</td>\n",
       "      <td>96.4431</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000052504</td>\n",
       "      <td>ENSG00000137845</td>\n",
       "      <td>ENSG00000044524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3956</th>\n",
       "      <td>CDB03444</td>\n",
       "      <td>Pcdhb21-Pcdhb21</td>\n",
       "      <td>Pcdhb21 → Pcdhb21</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000044022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000044022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3957</th>\n",
       "      <td>CDB03445</td>\n",
       "      <td>Pcdhb22-Pcdhb22</td>\n",
       "      <td>Pcdhb22 → Pcdhb22</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000073591</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000073591</td>\n",
       "      <td>ENSG00000113248</td>\n",
       "      <td>ENSG00000113248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3958</th>\n",
       "      <td>CDB03446</td>\n",
       "      <td>Pcdhgb8-Pcdhgb8</td>\n",
       "      <td>Pcdhgb8 → Pcdhgb8</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000103081</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000103081</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3959</th>\n",
       "      <td>CDB03447</td>\n",
       "      <td>Saa3-Tlr4</td>\n",
       "      <td>Saa3 → Tlr4</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000040026</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000039005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000136869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960</th>\n",
       "      <td>CDB03448</td>\n",
       "      <td>Sema4a-Timd2</td>\n",
       "      <td>Sema4a → Timd2</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000028064</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000040413</td>\n",
       "      <td>ENSG00000196189</td>\n",
       "      <td>ENSG00000113249</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3961 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Interaction ID     LR Pair Card      Mouse LR Pair  \\\n",
       "0          CDB00001        A2M-HSPA5        A2m → Hspa5   \n",
       "1          CDB00002         A2M-LRP1         A2m → Lrp1   \n",
       "2          CDB00003       ACE-BDKRB2       Ace → Bdkrb2   \n",
       "3          CDB00004         ADA-DPP4         Ada → Dpp4   \n",
       "4          CDB00005     ADAM10-EPHA3     Adam10 → Epha3   \n",
       "...             ...              ...                ...   \n",
       "3956       CDB03444  Pcdhb21-Pcdhb21  Pcdhb21 → Pcdhb21   \n",
       "3957       CDB03445  Pcdhb22-Pcdhb22  Pcdhb22 → Pcdhb22   \n",
       "3958       CDB03446  Pcdhgb8-Pcdhgb8  Pcdhgb8 → Pcdhgb8   \n",
       "3959       CDB03447        Saa3-Tlr4        Saa3 → Tlr4   \n",
       "3960       CDB03448     Sema4a-Timd2     Sema4a → Timd2   \n",
       "\n",
       "                                         Ligand HGNC ID  \\\n",
       "0     <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "1     <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "2     <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3     <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4     <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "...                                                 ...   \n",
       "3956  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3957  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3958  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3959  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3960  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "\n",
       "                                       Receptor HGNC ID  Ligand GOC score  \\\n",
       "0     <a href=\"https://www.genenames.org/data/gene-s...              75.0   \n",
       "1     <a href=\"https://www.genenames.org/data/gene-s...              75.0   \n",
       "2     <a href=\"https://www.genenames.org/data/gene-s...             100.0   \n",
       "3     <a href=\"https://www.genenames.org/data/gene-s...             100.0   \n",
       "4     <a href=\"https://www.genenames.org/data/gene-s...             100.0   \n",
       "...                                                 ...               ...   \n",
       "3956  <a href=\"https://www.genenames.org/data/gene-s...               NaN   \n",
       "3957  <a href=\"https://www.genenames.org/data/gene-s...               NaN   \n",
       "3958  <a href=\"https://www.genenames.org/data/gene-s...               NaN   \n",
       "3959  <a href=\"https://www.genenames.org/data/gene-s...               NaN   \n",
       "3960  <a href=\"https://www.genenames.org/data/gene-s...               NaN   \n",
       "\n",
       "      Ligand WGA coverage  Ligand % Identity  Ligand Target % Identity  \\\n",
       "0                   100.0            72.4559                   72.4559   \n",
       "1                   100.0            72.4559                   72.4559   \n",
       "2                   100.0            83.3078                   82.9268   \n",
       "3                   100.0            80.7163                   83.2386   \n",
       "4                   100.0            96.1230                   95.9947   \n",
       "...                   ...                ...                       ...   \n",
       "3956                  NaN                NaN                       NaN   \n",
       "3957                  NaN                NaN                       NaN   \n",
       "3958                  NaN                NaN                       NaN   \n",
       "3959                  NaN                NaN                       NaN   \n",
       "3960                  NaN                NaN                       NaN   \n",
       "\n",
       "      Ligand Orthology Confidence   Ligand Ensembl ID  Receptor GOC score  \\\n",
       "0                             1.0  ENSMUSG00000030111                75.0   \n",
       "1                             1.0  ENSMUSG00000030111               100.0   \n",
       "2                             1.0  ENSMUSG00000020681               100.0   \n",
       "3                             1.0  ENSMUSG00000017697               100.0   \n",
       "4                             1.0  ENSMUSG00000054693               100.0   \n",
       "...                           ...                 ...                 ...   \n",
       "3956                          NaN  ENSMUSG00000044022                 NaN   \n",
       "3957                          NaN  ENSMUSG00000073591                 NaN   \n",
       "3958                          NaN  ENSMUSG00000103081                 NaN   \n",
       "3959                          NaN  ENSMUSG00000040026                 NaN   \n",
       "3960                          NaN  ENSMUSG00000028064                 NaN   \n",
       "\n",
       "      Receptor WGA coverage  Receptor % Identity  Receptor Target % Identity  \\\n",
       "0                     100.0              98.6239                     98.4733   \n",
       "1                     100.0              97.9974                     97.9758   \n",
       "2                     100.0              79.7954                     79.5918   \n",
       "3                     100.0              84.4648                     85.1316   \n",
       "4                     100.0              96.5412                     96.4431   \n",
       "...                     ...                  ...                         ...   \n",
       "3956                    NaN                  NaN                         NaN   \n",
       "3957                    NaN                  NaN                         NaN   \n",
       "3958                    NaN                  NaN                         NaN   \n",
       "3959                    NaN                  NaN                         NaN   \n",
       "3960                    NaN                  NaN                         NaN   \n",
       "\n",
       "      Receptor Orthology Confidence Receptor Ensembl ID  \\\n",
       "0                               1.0  ENSMUSG00000026864   \n",
       "1                               1.0  ENSMUSG00000040249   \n",
       "2                               1.0  ENSMUSG00000021070   \n",
       "3                               1.0  ENSMUSG00000035000   \n",
       "4                               1.0  ENSMUSG00000052504   \n",
       "...                             ...                 ...   \n",
       "3956                            NaN  ENSMUSG00000044022   \n",
       "3957                            NaN  ENSMUSG00000073591   \n",
       "3958                            NaN  ENSMUSG00000103081   \n",
       "3959                            NaN  ENSMUSG00000039005   \n",
       "3960                            NaN  ENSMUSG00000040413   \n",
       "\n",
       "     Human Ligand Ensembl ID Human Receptor Ensembl ID  \n",
       "0            ENSG00000175899           ENSG00000044574  \n",
       "1            ENSG00000175899           ENSG00000123384  \n",
       "2            ENSG00000159640           ENSG00000168398  \n",
       "3            ENSG00000196839           ENSG00000197635  \n",
       "4            ENSG00000137845           ENSG00000044524  \n",
       "...                      ...                       ...  \n",
       "3956                     NaN                       NaN  \n",
       "3957         ENSG00000113248           ENSG00000113248  \n",
       "3958                     NaN                       NaN  \n",
       "3959                     NaN           ENSG00000136869  \n",
       "3960         ENSG00000196189           ENSG00000113249  \n",
       "\n",
       "[3961 rows x 19 columns]"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create the mapping dictionary from mouse to human Ensembl gene ID\n",
    "mouse_to_human_map = dict(zip(\n",
    "    mousebioM_df[\"mmusculus_homolog_ensembl_gene\"],\n",
    "    mousebioM_df[\"ensembl_gene_id\"]\n",
    "))\n",
    "\n",
    "# Map Ligand\n",
    "mouse_gene_pair1[\"Human Ligand Ensembl ID\"] = mouse_gene_pair1[\"Ligand Ensembl ID\"].map(mouse_to_human_map)\n",
    "\n",
    "# Map Receptor\n",
    "mouse_gene_pair1[\"Human Receptor Ensembl ID\"] = mouse_gene_pair1[\"Receptor Ensembl ID\"].map(mouse_to_human_map)\n",
    "mouse_gene_pair1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "9edba995-b6e6-4ee3-bafc-76ca9ff6c4bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_48719/1590521635.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mouse_gene_pair1['Ligand HGNC ID'] = mouse_gene_pair1['Ligand HGNC ID'].apply(extract_hgnc_id)\n",
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_48719/1590521635.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mouse_gene_pair1['Receptor HGNC ID'] = mouse_gene_pair1['Receptor HGNC ID'].apply(extract_hgnc_id)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Interaction ID</th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>Mouse LR Pair</th>\n",
       "      <th>Ligand HGNC ID</th>\n",
       "      <th>Receptor HGNC ID</th>\n",
       "      <th>Ligand GOC score</th>\n",
       "      <th>Ligand WGA coverage</th>\n",
       "      <th>Ligand % Identity</th>\n",
       "      <th>Ligand Target % Identity</th>\n",
       "      <th>Ligand Orthology Confidence</th>\n",
       "      <th>Ligand Ensembl ID</th>\n",
       "      <th>Receptor GOC score</th>\n",
       "      <th>Receptor WGA coverage</th>\n",
       "      <th>Receptor % Identity</th>\n",
       "      <th>Receptor Target % Identity</th>\n",
       "      <th>Receptor Orthology Confidence</th>\n",
       "      <th>Receptor Ensembl ID</th>\n",
       "      <th>Human Ligand Ensembl ID</th>\n",
       "      <th>Human Receptor Ensembl ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDB00001</td>\n",
       "      <td>A2M-HSPA5</td>\n",
       "      <td>A2m → Hspa5</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>HGNC:5238</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000030111</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>98.6239</td>\n",
       "      <td>98.4733</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000026864</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>ENSG00000044574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDB00002</td>\n",
       "      <td>A2M-LRP1</td>\n",
       "      <td>A2m → Lrp1</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>HGNC:6692</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000030111</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>97.9974</td>\n",
       "      <td>97.9758</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000040249</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>ENSG00000123384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDB00003</td>\n",
       "      <td>ACE-BDKRB2</td>\n",
       "      <td>Ace → Bdkrb2</td>\n",
       "      <td>HGNC:2707</td>\n",
       "      <td>HGNC:1030</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>83.3078</td>\n",
       "      <td>82.9268</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000020681</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>79.7954</td>\n",
       "      <td>79.5918</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000021070</td>\n",
       "      <td>ENSG00000159640</td>\n",
       "      <td>ENSG00000168398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDB00004</td>\n",
       "      <td>ADA-DPP4</td>\n",
       "      <td>Ada → Dpp4</td>\n",
       "      <td>HGNC:186</td>\n",
       "      <td>HGNC:3009</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>80.7163</td>\n",
       "      <td>83.2386</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000017697</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>84.4648</td>\n",
       "      <td>85.1316</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000035000</td>\n",
       "      <td>ENSG00000196839</td>\n",
       "      <td>ENSG00000197635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDB00005</td>\n",
       "      <td>ADAM10-EPHA3</td>\n",
       "      <td>Adam10 → Epha3</td>\n",
       "      <td>HGNC:188</td>\n",
       "      <td>HGNC:3387</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>96.1230</td>\n",
       "      <td>95.9947</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000054693</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>96.5412</td>\n",
       "      <td>96.4431</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000052504</td>\n",
       "      <td>ENSG00000137845</td>\n",
       "      <td>ENSG00000044524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3956</th>\n",
       "      <td>CDB03444</td>\n",
       "      <td>Pcdhb21-Pcdhb21</td>\n",
       "      <td>Pcdhb21 → Pcdhb21</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000044022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000044022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3957</th>\n",
       "      <td>CDB03445</td>\n",
       "      <td>Pcdhb22-Pcdhb22</td>\n",
       "      <td>Pcdhb22 → Pcdhb22</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000073591</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000073591</td>\n",
       "      <td>ENSG00000113248</td>\n",
       "      <td>ENSG00000113248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3958</th>\n",
       "      <td>CDB03446</td>\n",
       "      <td>Pcdhgb8-Pcdhgb8</td>\n",
       "      <td>Pcdhgb8 → Pcdhgb8</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000103081</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000103081</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3959</th>\n",
       "      <td>CDB03447</td>\n",
       "      <td>Saa3-Tlr4</td>\n",
       "      <td>Saa3 → Tlr4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000040026</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000039005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000136869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960</th>\n",
       "      <td>CDB03448</td>\n",
       "      <td>Sema4a-Timd2</td>\n",
       "      <td>Sema4a → Timd2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000028064</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000040413</td>\n",
       "      <td>ENSG00000196189</td>\n",
       "      <td>ENSG00000113249</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3961 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Interaction ID     LR Pair Card      Mouse LR Pair Ligand HGNC ID  \\\n",
       "0          CDB00001        A2M-HSPA5        A2m → Hspa5         HGNC:7   \n",
       "1          CDB00002         A2M-LRP1         A2m → Lrp1         HGNC:7   \n",
       "2          CDB00003       ACE-BDKRB2       Ace → Bdkrb2      HGNC:2707   \n",
       "3          CDB00004         ADA-DPP4         Ada → Dpp4       HGNC:186   \n",
       "4          CDB00005     ADAM10-EPHA3     Adam10 → Epha3       HGNC:188   \n",
       "...             ...              ...                ...            ...   \n",
       "3956       CDB03444  Pcdhb21-Pcdhb21  Pcdhb21 → Pcdhb21           None   \n",
       "3957       CDB03445  Pcdhb22-Pcdhb22  Pcdhb22 → Pcdhb22           None   \n",
       "3958       CDB03446  Pcdhgb8-Pcdhgb8  Pcdhgb8 → Pcdhgb8           None   \n",
       "3959       CDB03447        Saa3-Tlr4        Saa3 → Tlr4           None   \n",
       "3960       CDB03448     Sema4a-Timd2     Sema4a → Timd2           None   \n",
       "\n",
       "     Receptor HGNC ID  Ligand GOC score  Ligand WGA coverage  \\\n",
       "0           HGNC:5238              75.0                100.0   \n",
       "1           HGNC:6692              75.0                100.0   \n",
       "2           HGNC:1030             100.0                100.0   \n",
       "3           HGNC:3009             100.0                100.0   \n",
       "4           HGNC:3387             100.0                100.0   \n",
       "...               ...               ...                  ...   \n",
       "3956             None               NaN                  NaN   \n",
       "3957             None               NaN                  NaN   \n",
       "3958             None               NaN                  NaN   \n",
       "3959             None               NaN                  NaN   \n",
       "3960             None               NaN                  NaN   \n",
       "\n",
       "      Ligand % Identity  Ligand Target % Identity  \\\n",
       "0               72.4559                   72.4559   \n",
       "1               72.4559                   72.4559   \n",
       "2               83.3078                   82.9268   \n",
       "3               80.7163                   83.2386   \n",
       "4               96.1230                   95.9947   \n",
       "...                 ...                       ...   \n",
       "3956                NaN                       NaN   \n",
       "3957                NaN                       NaN   \n",
       "3958                NaN                       NaN   \n",
       "3959                NaN                       NaN   \n",
       "3960                NaN                       NaN   \n",
       "\n",
       "      Ligand Orthology Confidence   Ligand Ensembl ID  Receptor GOC score  \\\n",
       "0                             1.0  ENSMUSG00000030111                75.0   \n",
       "1                             1.0  ENSMUSG00000030111               100.0   \n",
       "2                             1.0  ENSMUSG00000020681               100.0   \n",
       "3                             1.0  ENSMUSG00000017697               100.0   \n",
       "4                             1.0  ENSMUSG00000054693               100.0   \n",
       "...                           ...                 ...                 ...   \n",
       "3956                          NaN  ENSMUSG00000044022                 NaN   \n",
       "3957                          NaN  ENSMUSG00000073591                 NaN   \n",
       "3958                          NaN  ENSMUSG00000103081                 NaN   \n",
       "3959                          NaN  ENSMUSG00000040026                 NaN   \n",
       "3960                          NaN  ENSMUSG00000028064                 NaN   \n",
       "\n",
       "      Receptor WGA coverage  Receptor % Identity  Receptor Target % Identity  \\\n",
       "0                     100.0              98.6239                     98.4733   \n",
       "1                     100.0              97.9974                     97.9758   \n",
       "2                     100.0              79.7954                     79.5918   \n",
       "3                     100.0              84.4648                     85.1316   \n",
       "4                     100.0              96.5412                     96.4431   \n",
       "...                     ...                  ...                         ...   \n",
       "3956                    NaN                  NaN                         NaN   \n",
       "3957                    NaN                  NaN                         NaN   \n",
       "3958                    NaN                  NaN                         NaN   \n",
       "3959                    NaN                  NaN                         NaN   \n",
       "3960                    NaN                  NaN                         NaN   \n",
       "\n",
       "      Receptor Orthology Confidence Receptor Ensembl ID  \\\n",
       "0                               1.0  ENSMUSG00000026864   \n",
       "1                               1.0  ENSMUSG00000040249   \n",
       "2                               1.0  ENSMUSG00000021070   \n",
       "3                               1.0  ENSMUSG00000035000   \n",
       "4                               1.0  ENSMUSG00000052504   \n",
       "...                             ...                 ...   \n",
       "3956                            NaN  ENSMUSG00000044022   \n",
       "3957                            NaN  ENSMUSG00000073591   \n",
       "3958                            NaN  ENSMUSG00000103081   \n",
       "3959                            NaN  ENSMUSG00000039005   \n",
       "3960                            NaN  ENSMUSG00000040413   \n",
       "\n",
       "     Human Ligand Ensembl ID Human Receptor Ensembl ID  \n",
       "0            ENSG00000175899           ENSG00000044574  \n",
       "1            ENSG00000175899           ENSG00000123384  \n",
       "2            ENSG00000159640           ENSG00000168398  \n",
       "3            ENSG00000196839           ENSG00000197635  \n",
       "4            ENSG00000137845           ENSG00000044524  \n",
       "...                      ...                       ...  \n",
       "3956                     NaN                       NaN  \n",
       "3957         ENSG00000113248           ENSG00000113248  \n",
       "3958                     NaN                       NaN  \n",
       "3959                     NaN           ENSG00000136869  \n",
       "3960         ENSG00000196189           ENSG00000113249  \n",
       "\n",
       "[3961 rows x 19 columns]"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def extract_hgnc_id(col):\n",
    "    \"\"\"Use regular expression to extract the HGNC ID after 'HGNC:'.\"\"\"\n",
    "    match = re.search(r'HGNC:(\\d+)', col)\n",
    "    if match:\n",
    "        return 'HGNC:' +str(match.group(1))\n",
    "    return None\n",
    "    \n",
    "mouse_gene_pair1['Ligand HGNC ID'] = mouse_gene_pair1['Ligand HGNC ID'].apply(extract_hgnc_id)\n",
    "mouse_gene_pair1['Receptor HGNC ID'] = mouse_gene_pair1['Receptor HGNC ID'].apply(extract_hgnc_id)\n",
    "mouse_gene_pair1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "bb6bc879-b368-4743-857f-1341b8578dba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>human_protein</th>\n",
       "      <th>human_inparalog_score</th>\n",
       "      <th>human_seed_score</th>\n",
       "      <th>mouse_protein</th>\n",
       "      <th>mouse_inparalog_score</th>\n",
       "      <th>mouse_seed_score</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>human_ensembl_gene_id</th>\n",
       "      <th>human_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Q8WZ42</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A2ASS6</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>60090.0</td>\n",
       "      <td>ENSG00000155657</td>\n",
       "      <td>TTN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>Q8NF91</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q6ZWR6</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>14503.0</td>\n",
       "      <td>ENSG00000131018</td>\n",
       "      <td>SYNE1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>Q5VST9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A2AAJ9</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12156.0</td>\n",
       "      <td>ENSG00000154358</td>\n",
       "      <td>OBSCN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>Q03001</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q91ZU6</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10549.0</td>\n",
       "      <td>ENSG00000151914</td>\n",
       "      <td>DST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>Q8WXG9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q8VHN7</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10049.0</td>\n",
       "      <td>ENSG00000164199</td>\n",
       "      <td>ADGRV1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20946</th>\n",
       "      <td>17091</td>\n",
       "      <td>A0A1B0GTK5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A0A1B0GSI2</td>\n",
       "      <td>0.212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>46.0</td>\n",
       "      <td>ENSG00000225396</td>\n",
       "      <td>FAM236D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20947</th>\n",
       "      <td>17091</td>\n",
       "      <td>A0A1B0GTK5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A0A1B0GSB3</td>\n",
       "      <td>0.178</td>\n",
       "      <td>NaN</td>\n",
       "      <td>46.0</td>\n",
       "      <td>ENSG00000225396</td>\n",
       "      <td>FAM236D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20948</th>\n",
       "      <td>17092</td>\n",
       "      <td>Q96LM9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>E9Q1X6</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>ENSG00000125975</td>\n",
       "      <td>C20orf173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20952</th>\n",
       "      <td>17096</td>\n",
       "      <td>P0DP42</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A0A494B9K2</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>ENSG00000244219</td>\n",
       "      <td>TMEM225B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20954</th>\n",
       "      <td>17098</td>\n",
       "      <td>P31358</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q64389</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>ENSG00000169442</td>\n",
       "      <td>CD52</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20402 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       cluster_id human_protein  human_inparalog_score  human_seed_score  \\\n",
       "0               1        Q8WZ42                    1.0               1.0   \n",
       "1               2        Q8NF91                    1.0               1.0   \n",
       "2               3        Q5VST9                    1.0               1.0   \n",
       "4               5        Q03001                    1.0               1.0   \n",
       "5               6        Q8WXG9                    1.0               1.0   \n",
       "...           ...           ...                    ...               ...   \n",
       "20946       17091    A0A1B0GTK5                    1.0               1.0   \n",
       "20947       17091    A0A1B0GTK5                    1.0               1.0   \n",
       "20948       17092        Q96LM9                    1.0               1.0   \n",
       "20952       17096        P0DP42                    1.0               1.0   \n",
       "20954       17098        P31358                    1.0               1.0   \n",
       "\n",
       "      mouse_protein  mouse_inparalog_score  mouse_seed_score  bitscore  \\\n",
       "0            A2ASS6                  1.000               1.0   60090.0   \n",
       "1            Q6ZWR6                  1.000               1.0   14503.0   \n",
       "2            A2AAJ9                  1.000               1.0   12156.0   \n",
       "4            Q91ZU6                  1.000               1.0   10549.0   \n",
       "5            Q8VHN7                  1.000               1.0   10049.0   \n",
       "...             ...                    ...               ...       ...   \n",
       "20946    A0A1B0GSI2                  0.212               NaN      46.0   \n",
       "20947    A0A1B0GSB3                  0.178               NaN      46.0   \n",
       "20948        E9Q1X6                  1.000               1.0      46.0   \n",
       "20952    A0A494B9K2                  1.000               1.0      41.0   \n",
       "20954        Q64389                  1.000               1.0      40.0   \n",
       "\n",
       "      human_ensembl_gene_id human_gene  \n",
       "0           ENSG00000155657        TTN  \n",
       "1           ENSG00000131018      SYNE1  \n",
       "2           ENSG00000154358      OBSCN  \n",
       "4           ENSG00000151914        DST  \n",
       "5           ENSG00000164199     ADGRV1  \n",
       "...                     ...        ...  \n",
       "20946       ENSG00000225396    FAM236D  \n",
       "20947       ENSG00000225396    FAM236D  \n",
       "20948       ENSG00000125975  C20orf173  \n",
       "20952       ENSG00000244219   TMEM225B  \n",
       "20954       ENSG00000169442       CD52  \n",
       "\n",
       "[20402 rows x 10 columns]"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hgnc_df = pd.read_csv(\"data/HGNC_gene_info_full.tsv\", sep=\"\\t\", dtype=str)\n",
    "hgnc_df = hgnc_df.dropna(subset=[\"uniprot_ids\", \"ensembl_gene_id\"])\n",
    "# Split uniprot_ids on comma and explode\n",
    "hgnc_exploded = hgnc_df.assign(uniprot_id=hgnc_df[\"uniprot_ids\"].str.split(\",\")).explode(\"uniprot_id\")\n",
    "hgnc_exploded[\"uniprot_id\"] = hgnc_exploded[\"uniprot_id\"].str.strip()\n",
    "\n",
    "uniprot_to_ensembl = hgnc_exploded.set_index(\"uniprot_id\")[\"ensembl_gene_id\"].to_dict()\n",
    "\n",
    "# Left join on human_protein\n",
    "df_merged = df_orthologs.merge(\n",
    "    hgnc_exploded[[\"uniprot_id\", \"ensembl_gene_id\", \"symbol\"]],\n",
    "    left_on=\"human_protein\",\n",
    "    right_on=\"uniprot_id\",\n",
    "    how=\"left\"\n",
    ")\n",
    "\n",
    "# Optionally rename\n",
    "df_merged = df_merged.rename(columns={\n",
    "    \"symbol\": \"human_gene\",\n",
    "    \"ensembl_gene_id\": \"human_ensembl_gene_id\"\n",
    "}).drop(columns=[\"uniprot_id\"])\n",
    "df_merged = df_merged.dropna(subset=[\"human_ensembl_gene_id\"])\n",
    "df_merged.to_csv(\"data/mmusculus_inParanoid_uniProt_withHGNCAnn.tsv\", sep=\"\\t\", index=False)\n",
    "\n",
    "df_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "32fd1c6a-c855-4635-8a9b-365fa5b495ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>human_protein</th>\n",
       "      <th>human_inparalog_score</th>\n",
       "      <th>human_seed_score</th>\n",
       "      <th>mouse_protein</th>\n",
       "      <th>mouse_inparalog_score</th>\n",
       "      <th>mouse_seed_score</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>human_ensembl_gene_id</th>\n",
       "      <th>human_gene</th>\n",
       "      <th>mouse_ensembl_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Q8WZ42</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A2ASS6</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>60090.0</td>\n",
       "      <td>ENSG00000155657</td>\n",
       "      <td>TTN</td>\n",
       "      <td>ENSMUSG00000051747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>Q8NF91</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q6ZWR6</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>14503.0</td>\n",
       "      <td>ENSG00000131018</td>\n",
       "      <td>SYNE1</td>\n",
       "      <td>ENSMUSG00000096054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>Q5VST9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A2AAJ9</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12156.0</td>\n",
       "      <td>ENSG00000154358</td>\n",
       "      <td>OBSCN</td>\n",
       "      <td>ENSMUSG00000061462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>Q03001</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q91ZU6</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10549.0</td>\n",
       "      <td>ENSG00000151914</td>\n",
       "      <td>DST</td>\n",
       "      <td>ENSMUSG00000026131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6</td>\n",
       "      <td>Q8WXG9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q8VHN7</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10049.0</td>\n",
       "      <td>ENSG00000164199</td>\n",
       "      <td>ADGRV1</td>\n",
       "      <td>ENSMUSG00000069170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20920</th>\n",
       "      <td>17091</td>\n",
       "      <td>A0A1B0GTK5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A0A1B0GSI2</td>\n",
       "      <td>0.212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>46.0</td>\n",
       "      <td>ENSG00000225396</td>\n",
       "      <td>FAM236D</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20921</th>\n",
       "      <td>17091</td>\n",
       "      <td>A0A1B0GTK5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A0A1B0GSB3</td>\n",
       "      <td>0.178</td>\n",
       "      <td>NaN</td>\n",
       "      <td>46.0</td>\n",
       "      <td>ENSG00000225396</td>\n",
       "      <td>FAM236D</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20922</th>\n",
       "      <td>17092</td>\n",
       "      <td>Q96LM9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>E9Q1X6</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>ENSG00000125975</td>\n",
       "      <td>C20orf173</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20923</th>\n",
       "      <td>17096</td>\n",
       "      <td>P0DP42</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A0A494B9K2</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>ENSG00000244219</td>\n",
       "      <td>TMEM225B</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20924</th>\n",
       "      <td>17098</td>\n",
       "      <td>P31358</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q64389</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>ENSG00000169442</td>\n",
       "      <td>CD52</td>\n",
       "      <td>ENSMUSG00000000682</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20925 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       cluster_id human_protein  human_inparalog_score  human_seed_score  \\\n",
       "0               1        Q8WZ42                    1.0               1.0   \n",
       "1               2        Q8NF91                    1.0               1.0   \n",
       "2               3        Q5VST9                    1.0               1.0   \n",
       "3               5        Q03001                    1.0               1.0   \n",
       "4               6        Q8WXG9                    1.0               1.0   \n",
       "...           ...           ...                    ...               ...   \n",
       "20920       17091    A0A1B0GTK5                    1.0               1.0   \n",
       "20921       17091    A0A1B0GTK5                    1.0               1.0   \n",
       "20922       17092        Q96LM9                    1.0               1.0   \n",
       "20923       17096        P0DP42                    1.0               1.0   \n",
       "20924       17098        P31358                    1.0               1.0   \n",
       "\n",
       "      mouse_protein  mouse_inparalog_score  mouse_seed_score  bitscore  \\\n",
       "0            A2ASS6                  1.000               1.0   60090.0   \n",
       "1            Q6ZWR6                  1.000               1.0   14503.0   \n",
       "2            A2AAJ9                  1.000               1.0   12156.0   \n",
       "3            Q91ZU6                  1.000               1.0   10549.0   \n",
       "4            Q8VHN7                  1.000               1.0   10049.0   \n",
       "...             ...                    ...               ...       ...   \n",
       "20920    A0A1B0GSI2                  0.212               NaN      46.0   \n",
       "20921    A0A1B0GSB3                  0.178               NaN      46.0   \n",
       "20922        E9Q1X6                  1.000               1.0      46.0   \n",
       "20923    A0A494B9K2                  1.000               1.0      41.0   \n",
       "20924        Q64389                  1.000               1.0      40.0   \n",
       "\n",
       "      human_ensembl_gene_id human_gene mouse_ensembl_gene_id  \n",
       "0           ENSG00000155657        TTN    ENSMUSG00000051747  \n",
       "1           ENSG00000131018      SYNE1    ENSMUSG00000096054  \n",
       "2           ENSG00000154358      OBSCN    ENSMUSG00000061462  \n",
       "3           ENSG00000151914        DST    ENSMUSG00000026131  \n",
       "4           ENSG00000164199     ADGRV1    ENSMUSG00000069170  \n",
       "...                     ...        ...                   ...  \n",
       "20920       ENSG00000225396    FAM236D                   NaN  \n",
       "20921       ENSG00000225396    FAM236D                   NaN  \n",
       "20922       ENSG00000125975  C20orf173                   NaN  \n",
       "20923       ENSG00000244219   TMEM225B                   NaN  \n",
       "20924       ENSG00000169442       CD52    ENSMUSG00000000682  \n",
       "\n",
       "[20925 rows x 11 columns]"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Step 1: Load the mapping file\n",
    "mouse_map = pd.read_csv(\"data/mouse_uniprot_to_ensembl.tsv\", sep=\"\\t\", dtype=str)\n",
    "\n",
    "# Step 2: Merge with df_merged on UniProt ID\n",
    "# Assuming your UniProt column in df_merged is named 'mouse_protein'\n",
    "df_merged = df_merged.merge(\n",
    "    mouse_map,\n",
    "    left_on=\"mouse_protein\",\n",
    "    right_on=\"uniprotswissprot\",\n",
    "    how=\"left\"\n",
    ")\n",
    "\n",
    "# Step 3: Rename the new column for clarity (optional)\n",
    "df_merged = df_merged.rename(columns={\"ensembl_gene_id\": \"mouse_ensembl_gene_id\"})\n",
    "\n",
    "# Step 4: Drop the helper merge column if not needed\n",
    "df_merged = df_merged.drop(columns=[\"uniprotswissprot\"])\n",
    "\n",
    "# Optional: Save\n",
    "df_merged.to_csv(\"data/df_merged_with_mouse_ensembl.tsv\", sep=\"\\t\", index=False)\n",
    "df_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "183330d8-1284-423e-ae06-ff0c6bb46d38",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Interaction ID</th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>Mouse LR Pair</th>\n",
       "      <th>Ligand HGNC ID</th>\n",
       "      <th>Receptor HGNC ID</th>\n",
       "      <th>Ligand GOC score</th>\n",
       "      <th>Ligand WGA coverage</th>\n",
       "      <th>Ligand % Identity</th>\n",
       "      <th>Ligand Target % Identity</th>\n",
       "      <th>Ligand Orthology Confidence</th>\n",
       "      <th>...</th>\n",
       "      <th>Ligand_human_protein</th>\n",
       "      <th>Ligand_human_inparalog_score</th>\n",
       "      <th>Ligand_human_seed_score</th>\n",
       "      <th>Ligand_mouse_protein</th>\n",
       "      <th>Ligand_mouse_inparalog_score</th>\n",
       "      <th>Ligand_mouse_seed_score</th>\n",
       "      <th>Ligand_bitscore</th>\n",
       "      <th>Ligand_human_ensembl_gene_id</th>\n",
       "      <th>Ligand_human_gene</th>\n",
       "      <th>Ligand_mouse_ensembl_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDB00001</td>\n",
       "      <td>A2M-HSPA5</td>\n",
       "      <td>A2m → Hspa5</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>HGNC:5238</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>P01023</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q6GQT1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2065.0</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSMUSG00000030111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDB00002</td>\n",
       "      <td>A2M-LRP1</td>\n",
       "      <td>A2m → Lrp1</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>HGNC:6692</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>P01023</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q6GQT1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2065.0</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSMUSG00000030111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDB00003</td>\n",
       "      <td>ACE-BDKRB2</td>\n",
       "      <td>Ace → Bdkrb2</td>\n",
       "      <td>HGNC:2707</td>\n",
       "      <td>HGNC:1030</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>83.3078</td>\n",
       "      <td>82.9268</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>P12821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P09470</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2342.0</td>\n",
       "      <td>ENSG00000159640</td>\n",
       "      <td>ACE</td>\n",
       "      <td>ENSMUSG00000020681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDB00004</td>\n",
       "      <td>ADA-DPP4</td>\n",
       "      <td>Ada → Dpp4</td>\n",
       "      <td>HGNC:186</td>\n",
       "      <td>HGNC:3009</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>80.7163</td>\n",
       "      <td>83.2386</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>P00813</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P03958</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>620.0</td>\n",
       "      <td>ENSG00000196839</td>\n",
       "      <td>ADA</td>\n",
       "      <td>ENSMUSG00000017697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDB00005</td>\n",
       "      <td>ADAM10-EPHA3</td>\n",
       "      <td>Adam10 → Epha3</td>\n",
       "      <td>HGNC:188</td>\n",
       "      <td>HGNC:3387</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>96.1230</td>\n",
       "      <td>95.9947</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>O14672</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>O35598</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1461.0</td>\n",
       "      <td>ENSG00000137845</td>\n",
       "      <td>ADAM10</td>\n",
       "      <td>ENSMUSG00000054693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9562</th>\n",
       "      <td>CDB03444</td>\n",
       "      <td>Pcdhb21-Pcdhb21</td>\n",
       "      <td>Pcdhb21 → Pcdhb21</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9563</th>\n",
       "      <td>CDB03445</td>\n",
       "      <td>Pcdhb22-Pcdhb22</td>\n",
       "      <td>Pcdhb22 → Pcdhb22</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Q9Y5E8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q91XZ8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.989</td>\n",
       "      <td>1191.0</td>\n",
       "      <td>ENSG00000113248</td>\n",
       "      <td>PCDHB15</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9564</th>\n",
       "      <td>CDB03446</td>\n",
       "      <td>Pcdhgb8-Pcdhgb8</td>\n",
       "      <td>Pcdhgb8 → Pcdhgb8</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9565</th>\n",
       "      <td>CDB03447</td>\n",
       "      <td>Saa3-Tlr4</td>\n",
       "      <td>Saa3 → Tlr4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9566</th>\n",
       "      <td>CDB03448</td>\n",
       "      <td>Sema4a-Timd2</td>\n",
       "      <td>Sema4a → Timd2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Q9H3S1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q62178</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1297.0</td>\n",
       "      <td>ENSG00000196189</td>\n",
       "      <td>SEMA4A</td>\n",
       "      <td>ENSMUSG00000028064</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9567 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Interaction ID     LR Pair Card      Mouse LR Pair Ligand HGNC ID  \\\n",
       "0          CDB00001        A2M-HSPA5        A2m → Hspa5         HGNC:7   \n",
       "1          CDB00002         A2M-LRP1         A2m → Lrp1         HGNC:7   \n",
       "2          CDB00003       ACE-BDKRB2       Ace → Bdkrb2      HGNC:2707   \n",
       "3          CDB00004         ADA-DPP4         Ada → Dpp4       HGNC:186   \n",
       "4          CDB00005     ADAM10-EPHA3     Adam10 → Epha3       HGNC:188   \n",
       "...             ...              ...                ...            ...   \n",
       "9562       CDB03444  Pcdhb21-Pcdhb21  Pcdhb21 → Pcdhb21           None   \n",
       "9563       CDB03445  Pcdhb22-Pcdhb22  Pcdhb22 → Pcdhb22           None   \n",
       "9564       CDB03446  Pcdhgb8-Pcdhgb8  Pcdhgb8 → Pcdhgb8           None   \n",
       "9565       CDB03447        Saa3-Tlr4        Saa3 → Tlr4           None   \n",
       "9566       CDB03448     Sema4a-Timd2     Sema4a → Timd2           None   \n",
       "\n",
       "     Receptor HGNC ID  Ligand GOC score  Ligand WGA coverage  \\\n",
       "0           HGNC:5238              75.0                100.0   \n",
       "1           HGNC:6692              75.0                100.0   \n",
       "2           HGNC:1030             100.0                100.0   \n",
       "3           HGNC:3009             100.0                100.0   \n",
       "4           HGNC:3387             100.0                100.0   \n",
       "...               ...               ...                  ...   \n",
       "9562             None               NaN                  NaN   \n",
       "9563             None               NaN                  NaN   \n",
       "9564             None               NaN                  NaN   \n",
       "9565             None               NaN                  NaN   \n",
       "9566             None               NaN                  NaN   \n",
       "\n",
       "      Ligand % Identity  Ligand Target % Identity  \\\n",
       "0               72.4559                   72.4559   \n",
       "1               72.4559                   72.4559   \n",
       "2               83.3078                   82.9268   \n",
       "3               80.7163                   83.2386   \n",
       "4               96.1230                   95.9947   \n",
       "...                 ...                       ...   \n",
       "9562                NaN                       NaN   \n",
       "9563                NaN                       NaN   \n",
       "9564                NaN                       NaN   \n",
       "9565                NaN                       NaN   \n",
       "9566                NaN                       NaN   \n",
       "\n",
       "      Ligand Orthology Confidence  ... Ligand_human_protein  \\\n",
       "0                             1.0  ...               P01023   \n",
       "1                             1.0  ...               P01023   \n",
       "2                             1.0  ...               P12821   \n",
       "3                             1.0  ...               P00813   \n",
       "4                             1.0  ...               O14672   \n",
       "...                           ...  ...                  ...   \n",
       "9562                          NaN  ...                  NaN   \n",
       "9563                          NaN  ...               Q9Y5E8   \n",
       "9564                          NaN  ...                  NaN   \n",
       "9565                          NaN  ...                  NaN   \n",
       "9566                          NaN  ...               Q9H3S1   \n",
       "\n",
       "      Ligand_human_inparalog_score  Ligand_human_seed_score  \\\n",
       "0                              1.0                      1.0   \n",
       "1                              1.0                      1.0   \n",
       "2                              1.0                      1.0   \n",
       "3                              1.0                      1.0   \n",
       "4                              1.0                      1.0   \n",
       "...                            ...                      ...   \n",
       "9562                           NaN                      NaN   \n",
       "9563                           1.0                      1.0   \n",
       "9564                           NaN                      NaN   \n",
       "9565                           NaN                      NaN   \n",
       "9566                           1.0                      1.0   \n",
       "\n",
       "      Ligand_mouse_protein  Ligand_mouse_inparalog_score  \\\n",
       "0                   Q6GQT1                           1.0   \n",
       "1                   Q6GQT1                           1.0   \n",
       "2                   P09470                           1.0   \n",
       "3                   P03958                           1.0   \n",
       "4                   O35598                           1.0   \n",
       "...                    ...                           ...   \n",
       "9562                   NaN                           NaN   \n",
       "9563                Q91XZ8                           1.0   \n",
       "9564                   NaN                           NaN   \n",
       "9565                   NaN                           NaN   \n",
       "9566                Q62178                           1.0   \n",
       "\n",
       "      Ligand_mouse_seed_score Ligand_bitscore Ligand_human_ensembl_gene_id  \\\n",
       "0                       1.000          2065.0              ENSG00000175899   \n",
       "1                       1.000          2065.0              ENSG00000175899   \n",
       "2                       1.000          2342.0              ENSG00000159640   \n",
       "3                       1.000           620.0              ENSG00000196839   \n",
       "4                       1.000          1461.0              ENSG00000137845   \n",
       "...                       ...             ...                          ...   \n",
       "9562                      NaN             NaN                          NaN   \n",
       "9563                    0.989          1191.0              ENSG00000113248   \n",
       "9564                      NaN             NaN                          NaN   \n",
       "9565                      NaN             NaN                          NaN   \n",
       "9566                    1.000          1297.0              ENSG00000196189   \n",
       "\n",
       "     Ligand_human_gene  Ligand_mouse_ensembl_gene_id  \n",
       "0                  A2M            ENSMUSG00000030111  \n",
       "1                  A2M            ENSMUSG00000030111  \n",
       "2                  ACE            ENSMUSG00000020681  \n",
       "3                  ADA            ENSMUSG00000017697  \n",
       "4               ADAM10            ENSMUSG00000054693  \n",
       "...                ...                           ...  \n",
       "9562               NaN                           NaN  \n",
       "9563           PCDHB15                           NaN  \n",
       "9564               NaN                           NaN  \n",
       "9565               NaN                           NaN  \n",
       "9566            SEMA4A            ENSMUSG00000028064  \n",
       "\n",
       "[9567 rows x 30 columns]"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 1. Prepare df_merged with prefix\n",
    "df_ligand = df_merged.add_prefix(\"Ligand_\")\n",
    "\n",
    "# 2. Merge on Human Ligand Ensembl ID (left join)\n",
    "ligand_merged = mouse_gene_pair1.merge(\n",
    "    df_ligand,\n",
    "    left_on=\"Human Ligand Ensembl ID\",\n",
    "    right_on=\"Ligand_human_ensembl_gene_id\",\n",
    "    how=\"left\",\n",
    "    suffixes=('', '_merged')\n",
    ")\n",
    "ligand_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "d1b77270-b7fc-4afe-b3b8-cbfd55cff8b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_48719/3666339144.py:16: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(resolve_ligand)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Interaction ID</th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>Mouse LR Pair</th>\n",
       "      <th>Ligand HGNC ID</th>\n",
       "      <th>Receptor HGNC ID</th>\n",
       "      <th>Ligand GOC score</th>\n",
       "      <th>Ligand WGA coverage</th>\n",
       "      <th>Ligand % Identity</th>\n",
       "      <th>Ligand Target % Identity</th>\n",
       "      <th>Ligand Orthology Confidence</th>\n",
       "      <th>...</th>\n",
       "      <th>Ligand_human_protein</th>\n",
       "      <th>Ligand_human_inparalog_score</th>\n",
       "      <th>Ligand_human_seed_score</th>\n",
       "      <th>Ligand_mouse_protein</th>\n",
       "      <th>Ligand_mouse_inparalog_score</th>\n",
       "      <th>Ligand_mouse_seed_score</th>\n",
       "      <th>Ligand_bitscore</th>\n",
       "      <th>Ligand_human_ensembl_gene_id</th>\n",
       "      <th>Ligand_human_gene</th>\n",
       "      <th>Ligand_mouse_ensembl_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDB00001</td>\n",
       "      <td>A2M-HSPA5</td>\n",
       "      <td>A2m → Hspa5</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>HGNC:5238</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>P01023</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q6GQT1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2065.0</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSMUSG00000030111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDB00002</td>\n",
       "      <td>A2M-LRP1</td>\n",
       "      <td>A2m → Lrp1</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>HGNC:6692</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>P01023</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q6GQT1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2065.0</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSMUSG00000030111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDB00003</td>\n",
       "      <td>ACE-BDKRB2</td>\n",
       "      <td>Ace → Bdkrb2</td>\n",
       "      <td>HGNC:2707</td>\n",
       "      <td>HGNC:1030</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>83.3078</td>\n",
       "      <td>82.9268</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>P12821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P09470</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2342.0</td>\n",
       "      <td>ENSG00000159640</td>\n",
       "      <td>ACE</td>\n",
       "      <td>ENSMUSG00000020681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDB00004</td>\n",
       "      <td>ADA-DPP4</td>\n",
       "      <td>Ada → Dpp4</td>\n",
       "      <td>HGNC:186</td>\n",
       "      <td>HGNC:3009</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>80.7163</td>\n",
       "      <td>83.2386</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>P00813</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P03958</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>620.0</td>\n",
       "      <td>ENSG00000196839</td>\n",
       "      <td>ADA</td>\n",
       "      <td>ENSMUSG00000017697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDB00005</td>\n",
       "      <td>ADAM10-EPHA3</td>\n",
       "      <td>Adam10 → Epha3</td>\n",
       "      <td>HGNC:188</td>\n",
       "      <td>HGNC:3387</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>96.1230</td>\n",
       "      <td>95.9947</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>O14672</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>O35598</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1461.0</td>\n",
       "      <td>ENSG00000137845</td>\n",
       "      <td>ADAM10</td>\n",
       "      <td>ENSMUSG00000054693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3156</th>\n",
       "      <td>CDB03444</td>\n",
       "      <td>Pcdhb21-Pcdhb21</td>\n",
       "      <td>Pcdhb21 → Pcdhb21</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3157</th>\n",
       "      <td>CDB03445</td>\n",
       "      <td>Pcdhb22-Pcdhb22</td>\n",
       "      <td>Pcdhb22 → Pcdhb22</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Q9Y5E8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q91XZ8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.989</td>\n",
       "      <td>1191.0</td>\n",
       "      <td>ENSG00000113248</td>\n",
       "      <td>PCDHB15</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3158</th>\n",
       "      <td>CDB03446</td>\n",
       "      <td>Pcdhgb8-Pcdhgb8</td>\n",
       "      <td>Pcdhgb8 → Pcdhgb8</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3159</th>\n",
       "      <td>CDB03447</td>\n",
       "      <td>Saa3-Tlr4</td>\n",
       "      <td>Saa3 → Tlr4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3160</th>\n",
       "      <td>CDB03448</td>\n",
       "      <td>Sema4a-Timd2</td>\n",
       "      <td>Sema4a → Timd2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Q9H3S1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q62178</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1297.0</td>\n",
       "      <td>ENSG00000196189</td>\n",
       "      <td>SEMA4A</td>\n",
       "      <td>ENSMUSG00000028064</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3161 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Interaction ID     LR Pair Card      Mouse LR Pair Ligand HGNC ID  \\\n",
       "0          CDB00001        A2M-HSPA5        A2m → Hspa5         HGNC:7   \n",
       "1          CDB00002         A2M-LRP1         A2m → Lrp1         HGNC:7   \n",
       "2          CDB00003       ACE-BDKRB2       Ace → Bdkrb2      HGNC:2707   \n",
       "3          CDB00004         ADA-DPP4         Ada → Dpp4       HGNC:186   \n",
       "4          CDB00005     ADAM10-EPHA3     Adam10 → Epha3       HGNC:188   \n",
       "...             ...              ...                ...            ...   \n",
       "3156       CDB03444  Pcdhb21-Pcdhb21  Pcdhb21 → Pcdhb21           None   \n",
       "3157       CDB03445  Pcdhb22-Pcdhb22  Pcdhb22 → Pcdhb22           None   \n",
       "3158       CDB03446  Pcdhgb8-Pcdhgb8  Pcdhgb8 → Pcdhgb8           None   \n",
       "3159       CDB03447        Saa3-Tlr4        Saa3 → Tlr4           None   \n",
       "3160       CDB03448     Sema4a-Timd2     Sema4a → Timd2           None   \n",
       "\n",
       "     Receptor HGNC ID  Ligand GOC score  Ligand WGA coverage  \\\n",
       "0           HGNC:5238              75.0                100.0   \n",
       "1           HGNC:6692              75.0                100.0   \n",
       "2           HGNC:1030             100.0                100.0   \n",
       "3           HGNC:3009             100.0                100.0   \n",
       "4           HGNC:3387             100.0                100.0   \n",
       "...               ...               ...                  ...   \n",
       "3156             None               NaN                  NaN   \n",
       "3157             None               NaN                  NaN   \n",
       "3158             None               NaN                  NaN   \n",
       "3159             None               NaN                  NaN   \n",
       "3160             None               NaN                  NaN   \n",
       "\n",
       "      Ligand % Identity  Ligand Target % Identity  \\\n",
       "0               72.4559                   72.4559   \n",
       "1               72.4559                   72.4559   \n",
       "2               83.3078                   82.9268   \n",
       "3               80.7163                   83.2386   \n",
       "4               96.1230                   95.9947   \n",
       "...                 ...                       ...   \n",
       "3156                NaN                       NaN   \n",
       "3157                NaN                       NaN   \n",
       "3158                NaN                       NaN   \n",
       "3159                NaN                       NaN   \n",
       "3160                NaN                       NaN   \n",
       "\n",
       "      Ligand Orthology Confidence  ... Ligand_human_protein  \\\n",
       "0                             1.0  ...               P01023   \n",
       "1                             1.0  ...               P01023   \n",
       "2                             1.0  ...               P12821   \n",
       "3                             1.0  ...               P00813   \n",
       "4                             1.0  ...               O14672   \n",
       "...                           ...  ...                  ...   \n",
       "3156                          NaN  ...                  NaN   \n",
       "3157                          NaN  ...               Q9Y5E8   \n",
       "3158                          NaN  ...                  NaN   \n",
       "3159                          NaN  ...                  NaN   \n",
       "3160                          NaN  ...               Q9H3S1   \n",
       "\n",
       "      Ligand_human_inparalog_score  Ligand_human_seed_score  \\\n",
       "0                              1.0                      1.0   \n",
       "1                              1.0                      1.0   \n",
       "2                              1.0                      1.0   \n",
       "3                              1.0                      1.0   \n",
       "4                              1.0                      1.0   \n",
       "...                            ...                      ...   \n",
       "3156                           NaN                      NaN   \n",
       "3157                           1.0                      1.0   \n",
       "3158                           NaN                      NaN   \n",
       "3159                           NaN                      NaN   \n",
       "3160                           1.0                      1.0   \n",
       "\n",
       "      Ligand_mouse_protein  Ligand_mouse_inparalog_score  \\\n",
       "0                   Q6GQT1                           1.0   \n",
       "1                   Q6GQT1                           1.0   \n",
       "2                   P09470                           1.0   \n",
       "3                   P03958                           1.0   \n",
       "4                   O35598                           1.0   \n",
       "...                    ...                           ...   \n",
       "3156                   NaN                           NaN   \n",
       "3157                Q91XZ8                           1.0   \n",
       "3158                   NaN                           NaN   \n",
       "3159                   NaN                           NaN   \n",
       "3160                Q62178                           1.0   \n",
       "\n",
       "      Ligand_mouse_seed_score Ligand_bitscore Ligand_human_ensembl_gene_id  \\\n",
       "0                       1.000          2065.0              ENSG00000175899   \n",
       "1                       1.000          2065.0              ENSG00000175899   \n",
       "2                       1.000          2342.0              ENSG00000159640   \n",
       "3                       1.000           620.0              ENSG00000196839   \n",
       "4                       1.000          1461.0              ENSG00000137845   \n",
       "...                       ...             ...                          ...   \n",
       "3156                      NaN             NaN                          NaN   \n",
       "3157                    0.989          1191.0              ENSG00000113248   \n",
       "3158                      NaN             NaN                          NaN   \n",
       "3159                      NaN             NaN                          NaN   \n",
       "3160                    1.000          1297.0              ENSG00000196189   \n",
       "\n",
       "     Ligand_human_gene  Ligand_mouse_ensembl_gene_id  \n",
       "0                  A2M            ENSMUSG00000030111  \n",
       "1                  A2M            ENSMUSG00000030111  \n",
       "2                  ACE            ENSMUSG00000020681  \n",
       "3                  ADA            ENSMUSG00000017697  \n",
       "4               ADAM10            ENSMUSG00000054693  \n",
       "...                ...                           ...  \n",
       "3156               NaN                           NaN  \n",
       "3157           PCDHB15                           NaN  \n",
       "3158               NaN                           NaN  \n",
       "3159               NaN                           NaN  \n",
       "3160            SEMA4A            ENSMUSG00000028064  \n",
       "\n",
       "[3161 rows x 30 columns]"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def resolve_ligand(group):\n",
    "    if group.shape[0] == 1:\n",
    "        return group\n",
    "    # Try to keep rows where mouse Ensembl ID also matches\n",
    "    match = group[group[\"Ligand_mouse_ensembl_gene_id\"] == group[\"Ligand Ensembl ID\"]]\n",
    "    if len(match) == 1:\n",
    "        return match\n",
    "    elif len(match) > 1:\n",
    "        return match.iloc[[0]]  # arbitrarily pick one if multiple match\n",
    "    else:\n",
    "        return group.iloc[[0]]  # fallback to first\n",
    "\n",
    "# Apply resolver only where duplication exists\n",
    "resolved = (\n",
    "    ligand_merged.groupby(\"Interaction ID\", group_keys=False)\n",
    "    .apply(resolve_ligand)\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "resolved"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "1c048aa1-8a2a-4ecb-ab9d-b2f57bbccee2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mouse_gene_pair1: (3961, 19)\n",
      "df_merged: (20925, 11)\n"
     ]
    }
   ],
   "source": [
    "print(\"mouse_gene_pair1:\", mouse_gene_pair1.shape)\n",
    "print(\"df_merged:\", df_merged.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "201b524e-09f9-45ec-9b8b-ae4de02efe83",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Step 0: Add row ID to preserve original structure\n",
    "mouse_gene_pair1 = mouse_gene_pair1.reset_index(drop=False).rename(columns={\"index\": \"orig_row\"})\n",
    "\n",
    "# Step 1: Add prefix to merged data\n",
    "df_ligand = df_merged.add_prefix(\"Ligand_\")\n",
    "\n",
    "# Step 2: Merge on Human Ligand Ensembl ID (left join, allows multiple matches per row)\n",
    "ligand_merge = mouse_gene_pair1.merge(\n",
    "    df_ligand,\n",
    "    left_on=\"Human Ligand Ensembl ID\",\n",
    "    right_on=\"Ligand_human_ensembl_gene_id\",\n",
    "    how=\"left\",\n",
    "    suffixes=('', '_dup')\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "fc4683e0-8c38-4350-ac94-994f9a96e80b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_48719/1661301700.py:17: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(resolve_per_row)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Interaction ID</th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>Mouse LR Pair</th>\n",
       "      <th>Ligand HGNC ID</th>\n",
       "      <th>Receptor HGNC ID</th>\n",
       "      <th>Ligand GOC score</th>\n",
       "      <th>Ligand WGA coverage</th>\n",
       "      <th>Ligand % Identity</th>\n",
       "      <th>Ligand Target % Identity</th>\n",
       "      <th>Ligand Orthology Confidence</th>\n",
       "      <th>...</th>\n",
       "      <th>Ligand_human_protein</th>\n",
       "      <th>Ligand_human_inparalog_score</th>\n",
       "      <th>Ligand_human_seed_score</th>\n",
       "      <th>Ligand_mouse_protein</th>\n",
       "      <th>Ligand_mouse_inparalog_score</th>\n",
       "      <th>Ligand_mouse_seed_score</th>\n",
       "      <th>Ligand_bitscore</th>\n",
       "      <th>Ligand_human_ensembl_gene_id</th>\n",
       "      <th>Ligand_human_gene</th>\n",
       "      <th>Ligand_mouse_ensembl_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDB00001</td>\n",
       "      <td>A2M-HSPA5</td>\n",
       "      <td>A2m → Hspa5</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>HGNC:5238</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>P01023</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q6GQT1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2065.0</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSMUSG00000030111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDB00002</td>\n",
       "      <td>A2M-LRP1</td>\n",
       "      <td>A2m → Lrp1</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>HGNC:6692</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>P01023</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q6GQT1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2065.0</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSMUSG00000030111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDB00003</td>\n",
       "      <td>ACE-BDKRB2</td>\n",
       "      <td>Ace → Bdkrb2</td>\n",
       "      <td>HGNC:2707</td>\n",
       "      <td>HGNC:1030</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>83.3078</td>\n",
       "      <td>82.9268</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>P12821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P09470</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2342.0</td>\n",
       "      <td>ENSG00000159640</td>\n",
       "      <td>ACE</td>\n",
       "      <td>ENSMUSG00000020681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDB00004</td>\n",
       "      <td>ADA-DPP4</td>\n",
       "      <td>Ada → Dpp4</td>\n",
       "      <td>HGNC:186</td>\n",
       "      <td>HGNC:3009</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>80.7163</td>\n",
       "      <td>83.2386</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>P00813</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P03958</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>620.0</td>\n",
       "      <td>ENSG00000196839</td>\n",
       "      <td>ADA</td>\n",
       "      <td>ENSMUSG00000017697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDB00005</td>\n",
       "      <td>ADAM10-EPHA3</td>\n",
       "      <td>Adam10 → Epha3</td>\n",
       "      <td>HGNC:188</td>\n",
       "      <td>HGNC:3387</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>96.1230</td>\n",
       "      <td>95.9947</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>O14672</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>O35598</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1461.0</td>\n",
       "      <td>ENSG00000137845</td>\n",
       "      <td>ADAM10</td>\n",
       "      <td>ENSMUSG00000054693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3956</th>\n",
       "      <td>CDB03444</td>\n",
       "      <td>Pcdhb21-Pcdhb21</td>\n",
       "      <td>Pcdhb21 → Pcdhb21</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3957</th>\n",
       "      <td>CDB03445</td>\n",
       "      <td>Pcdhb22-Pcdhb22</td>\n",
       "      <td>Pcdhb22 → Pcdhb22</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Q9Y5E8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q91XZ8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.989</td>\n",
       "      <td>1191.0</td>\n",
       "      <td>ENSG00000113248</td>\n",
       "      <td>PCDHB15</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3958</th>\n",
       "      <td>CDB03446</td>\n",
       "      <td>Pcdhgb8-Pcdhgb8</td>\n",
       "      <td>Pcdhgb8 → Pcdhgb8</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3959</th>\n",
       "      <td>CDB03447</td>\n",
       "      <td>Saa3-Tlr4</td>\n",
       "      <td>Saa3 → Tlr4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960</th>\n",
       "      <td>CDB03448</td>\n",
       "      <td>Sema4a-Timd2</td>\n",
       "      <td>Sema4a → Timd2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Q9H3S1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q62178</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1297.0</td>\n",
       "      <td>ENSG00000196189</td>\n",
       "      <td>SEMA4A</td>\n",
       "      <td>ENSMUSG00000028064</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3961 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Interaction ID     LR Pair Card      Mouse LR Pair Ligand HGNC ID  \\\n",
       "0          CDB00001        A2M-HSPA5        A2m → Hspa5         HGNC:7   \n",
       "1          CDB00002         A2M-LRP1         A2m → Lrp1         HGNC:7   \n",
       "2          CDB00003       ACE-BDKRB2       Ace → Bdkrb2      HGNC:2707   \n",
       "3          CDB00004         ADA-DPP4         Ada → Dpp4       HGNC:186   \n",
       "4          CDB00005     ADAM10-EPHA3     Adam10 → Epha3       HGNC:188   \n",
       "...             ...              ...                ...            ...   \n",
       "3956       CDB03444  Pcdhb21-Pcdhb21  Pcdhb21 → Pcdhb21           None   \n",
       "3957       CDB03445  Pcdhb22-Pcdhb22  Pcdhb22 → Pcdhb22           None   \n",
       "3958       CDB03446  Pcdhgb8-Pcdhgb8  Pcdhgb8 → Pcdhgb8           None   \n",
       "3959       CDB03447        Saa3-Tlr4        Saa3 → Tlr4           None   \n",
       "3960       CDB03448     Sema4a-Timd2     Sema4a → Timd2           None   \n",
       "\n",
       "     Receptor HGNC ID  Ligand GOC score  Ligand WGA coverage  \\\n",
       "0           HGNC:5238              75.0                100.0   \n",
       "1           HGNC:6692              75.0                100.0   \n",
       "2           HGNC:1030             100.0                100.0   \n",
       "3           HGNC:3009             100.0                100.0   \n",
       "4           HGNC:3387             100.0                100.0   \n",
       "...               ...               ...                  ...   \n",
       "3956             None               NaN                  NaN   \n",
       "3957             None               NaN                  NaN   \n",
       "3958             None               NaN                  NaN   \n",
       "3959             None               NaN                  NaN   \n",
       "3960             None               NaN                  NaN   \n",
       "\n",
       "      Ligand % Identity  Ligand Target % Identity  \\\n",
       "0               72.4559                   72.4559   \n",
       "1               72.4559                   72.4559   \n",
       "2               83.3078                   82.9268   \n",
       "3               80.7163                   83.2386   \n",
       "4               96.1230                   95.9947   \n",
       "...                 ...                       ...   \n",
       "3956                NaN                       NaN   \n",
       "3957                NaN                       NaN   \n",
       "3958                NaN                       NaN   \n",
       "3959                NaN                       NaN   \n",
       "3960                NaN                       NaN   \n",
       "\n",
       "      Ligand Orthology Confidence  ... Ligand_human_protein  \\\n",
       "0                             1.0  ...               P01023   \n",
       "1                             1.0  ...               P01023   \n",
       "2                             1.0  ...               P12821   \n",
       "3                             1.0  ...               P00813   \n",
       "4                             1.0  ...               O14672   \n",
       "...                           ...  ...                  ...   \n",
       "3956                          NaN  ...                  NaN   \n",
       "3957                          NaN  ...               Q9Y5E8   \n",
       "3958                          NaN  ...                  NaN   \n",
       "3959                          NaN  ...                  NaN   \n",
       "3960                          NaN  ...               Q9H3S1   \n",
       "\n",
       "      Ligand_human_inparalog_score  Ligand_human_seed_score  \\\n",
       "0                              1.0                      1.0   \n",
       "1                              1.0                      1.0   \n",
       "2                              1.0                      1.0   \n",
       "3                              1.0                      1.0   \n",
       "4                              1.0                      1.0   \n",
       "...                            ...                      ...   \n",
       "3956                           NaN                      NaN   \n",
       "3957                           1.0                      1.0   \n",
       "3958                           NaN                      NaN   \n",
       "3959                           NaN                      NaN   \n",
       "3960                           1.0                      1.0   \n",
       "\n",
       "      Ligand_mouse_protein  Ligand_mouse_inparalog_score  \\\n",
       "0                   Q6GQT1                           1.0   \n",
       "1                   Q6GQT1                           1.0   \n",
       "2                   P09470                           1.0   \n",
       "3                   P03958                           1.0   \n",
       "4                   O35598                           1.0   \n",
       "...                    ...                           ...   \n",
       "3956                   NaN                           NaN   \n",
       "3957                Q91XZ8                           1.0   \n",
       "3958                   NaN                           NaN   \n",
       "3959                   NaN                           NaN   \n",
       "3960                Q62178                           1.0   \n",
       "\n",
       "      Ligand_mouse_seed_score Ligand_bitscore Ligand_human_ensembl_gene_id  \\\n",
       "0                       1.000          2065.0              ENSG00000175899   \n",
       "1                       1.000          2065.0              ENSG00000175899   \n",
       "2                       1.000          2342.0              ENSG00000159640   \n",
       "3                       1.000           620.0              ENSG00000196839   \n",
       "4                       1.000          1461.0              ENSG00000137845   \n",
       "...                       ...             ...                          ...   \n",
       "3956                      NaN             NaN                          NaN   \n",
       "3957                    0.989          1191.0              ENSG00000113248   \n",
       "3958                      NaN             NaN                          NaN   \n",
       "3959                      NaN             NaN                          NaN   \n",
       "3960                    1.000          1297.0              ENSG00000196189   \n",
       "\n",
       "     Ligand_human_gene  Ligand_mouse_ensembl_gene_id  \n",
       "0                  A2M            ENSMUSG00000030111  \n",
       "1                  A2M            ENSMUSG00000030111  \n",
       "2                  ACE            ENSMUSG00000020681  \n",
       "3                  ADA            ENSMUSG00000017697  \n",
       "4               ADAM10            ENSMUSG00000054693  \n",
       "...                ...                           ...  \n",
       "3956               NaN                           NaN  \n",
       "3957           PCDHB15                           NaN  \n",
       "3958               NaN                           NaN  \n",
       "3959               NaN                           NaN  \n",
       "3960            SEMA4A            ENSMUSG00000028064  \n",
       "\n",
       "[3961 rows x 30 columns]"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Step 3: For each original row, select best match\n",
    "def resolve_per_row(group):\n",
    "    # Prefer rows where mouse Ensembl also matches\n",
    "    match = group[group[\"Ligand_mouse_ensembl_gene_id\"] == group[\"Ligand Ensembl ID\"]]\n",
    "    if len(match) == 1:\n",
    "        return match\n",
    "    elif len(match) > 1:\n",
    "        return match.iloc[[0]]  # pick first among valid\n",
    "    elif len(group) > 0:\n",
    "        return group.iloc[[0]]  # fallback: first available (human matched)\n",
    "    else:\n",
    "        return pd.DataFrame([group.iloc[0] * pd.NA])  # unlikely fallback\n",
    "\n",
    "# Step 4: Apply per original row (not Interaction ID)\n",
    "ligand_final = (\n",
    "    ligand_merge.groupby(\"orig_row\", group_keys=False)\n",
    "    .apply(resolve_per_row)\n",
    "    .reset_index(drop=True)\n",
    "    .drop(columns=[\"orig_row\"])\n",
    ")\n",
    "ligand_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "51b696a9-c883-4ddf-89c8-99256be4724d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_48719/3348606655.py:32: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(resolve_receptor_per_row)\n"
     ]
    }
   ],
   "source": [
    "# Step 0: Add row ID again if not already there\n",
    "if \"orig_row\" not in mouse_gene_pair1.columns:\n",
    "    mouse_gene_pair1 = mouse_gene_pair1.reset_index(drop=False).rename(columns={\"index\": \"orig_row\"})\n",
    "\n",
    "# Step 1: Add prefix to df_merged for receptor\n",
    "df_receptor = df_merged.add_prefix(\"Receptor_\")\n",
    "\n",
    "# Step 2: Merge on Human Receptor Ensembl ID\n",
    "receptor_merge = mouse_gene_pair1.merge(\n",
    "    df_receptor,\n",
    "    left_on=\"Human Receptor Ensembl ID\",\n",
    "    right_on=\"Receptor_human_ensembl_gene_id\",\n",
    "    how=\"left\",\n",
    "    suffixes=('', '_dup')\n",
    ")\n",
    "\n",
    "# Step 3: Resolve best match per row using mouse Ensembl ID\n",
    "def resolve_receptor_per_row(group):\n",
    "    match = group[group[\"Receptor_mouse_ensembl_gene_id\"] == group[\"Receptor Ensembl ID\"]]\n",
    "    if len(match) == 1:\n",
    "        return match\n",
    "    elif len(match) > 1:\n",
    "        return match.iloc[[0]]\n",
    "    elif len(group) > 0:\n",
    "        return group.iloc[[0]]\n",
    "    else:\n",
    "        return pd.DataFrame([group.iloc[0] * pd.NA])  # safe fallback\n",
    "\n",
    "# Step 4: Apply\n",
    "receptor_final = (\n",
    "    receptor_merge.groupby(\"orig_row\", group_keys=False)\n",
    "    .apply(resolve_receptor_per_row)\n",
    "    .reset_index(drop=True)\n",
    "    .drop(columns=[\"orig_row\"])\n",
    ")\n",
    "\n",
    "# Step 5: Check consistency\n",
    "assert len(receptor_final) == len(mouse_gene_pair1), f\"Row mismatch: {len(receptor_final)} != {len(mouse_gene_pair1)}\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "2216de0f-927c-4861-ab64-2c316b334102",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_row</th>\n",
       "      <th>Interaction ID</th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>Mouse LR Pair</th>\n",
       "      <th>Ligand HGNC ID</th>\n",
       "      <th>Receptor HGNC ID</th>\n",
       "      <th>Ligand GOC score</th>\n",
       "      <th>Ligand WGA coverage</th>\n",
       "      <th>Ligand % Identity</th>\n",
       "      <th>Ligand Target % Identity</th>\n",
       "      <th>Ligand Orthology Confidence</th>\n",
       "      <th>Ligand Ensembl ID</th>\n",
       "      <th>Receptor GOC score</th>\n",
       "      <th>Receptor WGA coverage</th>\n",
       "      <th>Receptor % Identity</th>\n",
       "      <th>Receptor Target % Identity</th>\n",
       "      <th>Receptor Orthology Confidence</th>\n",
       "      <th>Receptor Ensembl ID</th>\n",
       "      <th>Human Ligand Ensembl ID</th>\n",
       "      <th>Human Receptor Ensembl ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>CDB00001</td>\n",
       "      <td>A2M-HSPA5</td>\n",
       "      <td>A2m → Hspa5</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>HGNC:5238</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000030111</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>98.6239</td>\n",
       "      <td>98.4733</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000026864</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>ENSG00000044574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>CDB00002</td>\n",
       "      <td>A2M-LRP1</td>\n",
       "      <td>A2m → Lrp1</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>HGNC:6692</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000030111</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>97.9974</td>\n",
       "      <td>97.9758</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000040249</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>ENSG00000123384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>CDB00003</td>\n",
       "      <td>ACE-BDKRB2</td>\n",
       "      <td>Ace → Bdkrb2</td>\n",
       "      <td>HGNC:2707</td>\n",
       "      <td>HGNC:1030</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>83.3078</td>\n",
       "      <td>82.9268</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000020681</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>79.7954</td>\n",
       "      <td>79.5918</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000021070</td>\n",
       "      <td>ENSG00000159640</td>\n",
       "      <td>ENSG00000168398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>CDB00004</td>\n",
       "      <td>ADA-DPP4</td>\n",
       "      <td>Ada → Dpp4</td>\n",
       "      <td>HGNC:186</td>\n",
       "      <td>HGNC:3009</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>80.7163</td>\n",
       "      <td>83.2386</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000017697</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>84.4648</td>\n",
       "      <td>85.1316</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000035000</td>\n",
       "      <td>ENSG00000196839</td>\n",
       "      <td>ENSG00000197635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>CDB00005</td>\n",
       "      <td>ADAM10-EPHA3</td>\n",
       "      <td>Adam10 → Epha3</td>\n",
       "      <td>HGNC:188</td>\n",
       "      <td>HGNC:3387</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>96.1230</td>\n",
       "      <td>95.9947</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000054693</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>96.5412</td>\n",
       "      <td>96.4431</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSG00000052504</td>\n",
       "      <td>ENSG00000137845</td>\n",
       "      <td>ENSG00000044524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3956</th>\n",
       "      <td>3956</td>\n",
       "      <td>CDB03444</td>\n",
       "      <td>Pcdhb21-Pcdhb21</td>\n",
       "      <td>Pcdhb21 → Pcdhb21</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000044022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000044022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3957</th>\n",
       "      <td>3957</td>\n",
       "      <td>CDB03445</td>\n",
       "      <td>Pcdhb22-Pcdhb22</td>\n",
       "      <td>Pcdhb22 → Pcdhb22</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000073591</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000073591</td>\n",
       "      <td>ENSG00000113248</td>\n",
       "      <td>ENSG00000113248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3958</th>\n",
       "      <td>3958</td>\n",
       "      <td>CDB03446</td>\n",
       "      <td>Pcdhgb8-Pcdhgb8</td>\n",
       "      <td>Pcdhgb8 → Pcdhgb8</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000103081</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000103081</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3959</th>\n",
       "      <td>3959</td>\n",
       "      <td>CDB03447</td>\n",
       "      <td>Saa3-Tlr4</td>\n",
       "      <td>Saa3 → Tlr4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000040026</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000039005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000136869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960</th>\n",
       "      <td>3960</td>\n",
       "      <td>CDB03448</td>\n",
       "      <td>Sema4a-Timd2</td>\n",
       "      <td>Sema4a → Timd2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000028064</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000040413</td>\n",
       "      <td>ENSG00000196189</td>\n",
       "      <td>ENSG00000113249</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3961 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      orig_row Interaction ID     LR Pair Card      Mouse LR Pair  \\\n",
       "0            0       CDB00001        A2M-HSPA5        A2m → Hspa5   \n",
       "1            1       CDB00002         A2M-LRP1         A2m → Lrp1   \n",
       "2            2       CDB00003       ACE-BDKRB2       Ace → Bdkrb2   \n",
       "3            3       CDB00004         ADA-DPP4         Ada → Dpp4   \n",
       "4            4       CDB00005     ADAM10-EPHA3     Adam10 → Epha3   \n",
       "...        ...            ...              ...                ...   \n",
       "3956      3956       CDB03444  Pcdhb21-Pcdhb21  Pcdhb21 → Pcdhb21   \n",
       "3957      3957       CDB03445  Pcdhb22-Pcdhb22  Pcdhb22 → Pcdhb22   \n",
       "3958      3958       CDB03446  Pcdhgb8-Pcdhgb8  Pcdhgb8 → Pcdhgb8   \n",
       "3959      3959       CDB03447        Saa3-Tlr4        Saa3 → Tlr4   \n",
       "3960      3960       CDB03448     Sema4a-Timd2     Sema4a → Timd2   \n",
       "\n",
       "     Ligand HGNC ID Receptor HGNC ID  Ligand GOC score  Ligand WGA coverage  \\\n",
       "0            HGNC:7        HGNC:5238              75.0                100.0   \n",
       "1            HGNC:7        HGNC:6692              75.0                100.0   \n",
       "2         HGNC:2707        HGNC:1030             100.0                100.0   \n",
       "3          HGNC:186        HGNC:3009             100.0                100.0   \n",
       "4          HGNC:188        HGNC:3387             100.0                100.0   \n",
       "...             ...              ...               ...                  ...   \n",
       "3956           None             None               NaN                  NaN   \n",
       "3957           None             None               NaN                  NaN   \n",
       "3958           None             None               NaN                  NaN   \n",
       "3959           None             None               NaN                  NaN   \n",
       "3960           None             None               NaN                  NaN   \n",
       "\n",
       "      Ligand % Identity  Ligand Target % Identity  \\\n",
       "0               72.4559                   72.4559   \n",
       "1               72.4559                   72.4559   \n",
       "2               83.3078                   82.9268   \n",
       "3               80.7163                   83.2386   \n",
       "4               96.1230                   95.9947   \n",
       "...                 ...                       ...   \n",
       "3956                NaN                       NaN   \n",
       "3957                NaN                       NaN   \n",
       "3958                NaN                       NaN   \n",
       "3959                NaN                       NaN   \n",
       "3960                NaN                       NaN   \n",
       "\n",
       "      Ligand Orthology Confidence   Ligand Ensembl ID  Receptor GOC score  \\\n",
       "0                             1.0  ENSMUSG00000030111                75.0   \n",
       "1                             1.0  ENSMUSG00000030111               100.0   \n",
       "2                             1.0  ENSMUSG00000020681               100.0   \n",
       "3                             1.0  ENSMUSG00000017697               100.0   \n",
       "4                             1.0  ENSMUSG00000054693               100.0   \n",
       "...                           ...                 ...                 ...   \n",
       "3956                          NaN  ENSMUSG00000044022                 NaN   \n",
       "3957                          NaN  ENSMUSG00000073591                 NaN   \n",
       "3958                          NaN  ENSMUSG00000103081                 NaN   \n",
       "3959                          NaN  ENSMUSG00000040026                 NaN   \n",
       "3960                          NaN  ENSMUSG00000028064                 NaN   \n",
       "\n",
       "      Receptor WGA coverage  Receptor % Identity  Receptor Target % Identity  \\\n",
       "0                     100.0              98.6239                     98.4733   \n",
       "1                     100.0              97.9974                     97.9758   \n",
       "2                     100.0              79.7954                     79.5918   \n",
       "3                     100.0              84.4648                     85.1316   \n",
       "4                     100.0              96.5412                     96.4431   \n",
       "...                     ...                  ...                         ...   \n",
       "3956                    NaN                  NaN                         NaN   \n",
       "3957                    NaN                  NaN                         NaN   \n",
       "3958                    NaN                  NaN                         NaN   \n",
       "3959                    NaN                  NaN                         NaN   \n",
       "3960                    NaN                  NaN                         NaN   \n",
       "\n",
       "      Receptor Orthology Confidence Receptor Ensembl ID  \\\n",
       "0                               1.0  ENSMUSG00000026864   \n",
       "1                               1.0  ENSMUSG00000040249   \n",
       "2                               1.0  ENSMUSG00000021070   \n",
       "3                               1.0  ENSMUSG00000035000   \n",
       "4                               1.0  ENSMUSG00000052504   \n",
       "...                             ...                 ...   \n",
       "3956                            NaN  ENSMUSG00000044022   \n",
       "3957                            NaN  ENSMUSG00000073591   \n",
       "3958                            NaN  ENSMUSG00000103081   \n",
       "3959                            NaN  ENSMUSG00000039005   \n",
       "3960                            NaN  ENSMUSG00000040413   \n",
       "\n",
       "     Human Ligand Ensembl ID Human Receptor Ensembl ID  \n",
       "0            ENSG00000175899           ENSG00000044574  \n",
       "1            ENSG00000175899           ENSG00000123384  \n",
       "2            ENSG00000159640           ENSG00000168398  \n",
       "3            ENSG00000196839           ENSG00000197635  \n",
       "4            ENSG00000137845           ENSG00000044524  \n",
       "...                      ...                       ...  \n",
       "3956                     NaN                       NaN  \n",
       "3957         ENSG00000113248           ENSG00000113248  \n",
       "3958                     NaN                       NaN  \n",
       "3959                     NaN           ENSG00000136869  \n",
       "3960         ENSG00000196189           ENSG00000113249  \n",
       "\n",
       "[3961 rows x 20 columns]"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse_gene_pair1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "ac4954a8-486e-4bb7-a3d6-828cd6f93a18",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_48719/2123255046.py:31: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(resolve_ligand_row)\n",
      "/var/folders/r4/pjgb1k_x7fv9xgcb_dg70bnm0000gn/T/ipykernel_48719/2123255046.py:62: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  .apply(resolve_receptor_row)\n"
     ]
    }
   ],
   "source": [
    "# Step 0: Add 'orig_row' once, keep it clean\n",
    "mouse_gene_pair1_indexed = mouse_gene_pair1.reset_index(drop=False).rename(columns={\"index\": \"orig_row\"})\n",
    "\n",
    "### === LIGAND MERGE === ###\n",
    "df_ligand = df_merged.add_prefix(\"Ligand_\")\n",
    "ligand_merge = mouse_gene_pair1_indexed.merge(\n",
    "    df_ligand,\n",
    "    left_on=\"Human Ligand Ensembl ID\",\n",
    "    right_on=\"Ligand_human_ensembl_gene_id\",\n",
    "    how=\"left\"\n",
    ")\n",
    "\n",
    "# Ensure orig_row is single column (sometimes merge creates duplicates with suffix)\n",
    "if isinstance(ligand_merge.columns, pd.MultiIndex):\n",
    "    ligand_merge.columns = ligand_merge.columns.get_level_values(0)\n",
    "\n",
    "if ligand_merge.columns.duplicated().any():\n",
    "    ligand_merge = ligand_merge.loc[:, ~ligand_merge.columns.duplicated()]\n",
    "\n",
    "def resolve_ligand_row(group):\n",
    "    match = group[group[\"Ligand_mouse_ensembl_gene_id\"] == group[\"Ligand Ensembl ID\"]]\n",
    "    if len(match) == 1:\n",
    "        return match\n",
    "    elif len(match) > 1:\n",
    "        return match.iloc[[0]]\n",
    "    else:\n",
    "        return group.iloc[[0]]\n",
    "\n",
    "ligand_final = (\n",
    "    ligand_merge.groupby(\"orig_row\", group_keys=False)\n",
    "    .apply(resolve_ligand_row)\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "### === RECEPTOR MERGE === ###\n",
    "df_receptor = df_merged.add_prefix(\"Receptor_\")\n",
    "receptor_merge = ligand_final.merge(\n",
    "    df_receptor,\n",
    "    left_on=\"Human Receptor Ensembl ID\",\n",
    "    right_on=\"Receptor_human_ensembl_gene_id\",\n",
    "    how=\"left\"\n",
    ")\n",
    "\n",
    "# Same cleanup for receptor_merge\n",
    "if isinstance(receptor_merge.columns, pd.MultiIndex):\n",
    "    receptor_merge.columns = receptor_merge.columns.get_level_values(0)\n",
    "\n",
    "if receptor_merge.columns.duplicated().any():\n",
    "    receptor_merge = receptor_merge.loc[:, ~receptor_merge.columns.duplicated()]\n",
    "\n",
    "def resolve_receptor_row(group):\n",
    "    match = group[group[\"Receptor_mouse_ensembl_gene_id\"] == group[\"Receptor Ensembl ID\"]]\n",
    "    if len(match) == 1:\n",
    "        return match\n",
    "    elif len(match) > 1:\n",
    "        return match.iloc[[0]]\n",
    "    else:\n",
    "        return group.iloc[[0]]\n",
    "\n",
    "final_result = (\n",
    "    receptor_merge.groupby(\"orig_row\", group_keys=False)\n",
    "    .apply(resolve_receptor_row)\n",
    "    .reset_index(drop=True)\n",
    "    .drop(columns=[\"orig_row\"])\n",
    ")\n",
    "\n",
    "assert len(final_result) == len(mouse_gene_pair1), f\"Row mismatch: {len(final_result)} != {len(mouse_gene_pair1)}\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "id": "a4b31464-7db9-4606-bade-dccd9b3c081b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Interaction ID</th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>Ligand</th>\n",
       "      <th>Receptor</th>\n",
       "      <th>Mouse LR Pair</th>\n",
       "      <th>Ligand HGNC ID</th>\n",
       "      <th>Receptor HGNC ID</th>\n",
       "      <th>Ligand GOC score</th>\n",
       "      <th>Ligand WGA coverage</th>\n",
       "      <th>Ligand % Identity</th>\n",
       "      <th>...</th>\n",
       "      <th>Receptor_human_protein</th>\n",
       "      <th>Receptor_human_inparalog_score</th>\n",
       "      <th>Receptor_human_seed_score</th>\n",
       "      <th>Receptor_mouse_protein</th>\n",
       "      <th>Receptor_mouse_inparalog_score</th>\n",
       "      <th>Receptor_mouse_seed_score</th>\n",
       "      <th>Receptor_bitscore</th>\n",
       "      <th>Receptor_human_ensembl_gene_id</th>\n",
       "      <th>Receptor_human_gene</th>\n",
       "      <th>Receptor_mouse_ensembl_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDB00001</td>\n",
       "      <td>None</td>\n",
       "      <td>Hspa5</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>A2m → Hspa5</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>HGNC:5238</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>...</td>\n",
       "      <td>P11021</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P20029</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1246.0</td>\n",
       "      <td>ENSG00000044574</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>ENSMUSG00000026864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDB00002</td>\n",
       "      <td>None</td>\n",
       "      <td>Lrp1</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>A2m → Lrp1</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>HGNC:6692</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>72.4559</td>\n",
       "      <td>...</td>\n",
       "      <td>Q07954</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q91ZX7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>6853.0</td>\n",
       "      <td>ENSG00000123384</td>\n",
       "      <td>LRP1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDB00003</td>\n",
       "      <td>None</td>\n",
       "      <td>Bdkrb2</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Ace → Bdkrb2</td>\n",
       "      <td>HGNC:2707</td>\n",
       "      <td>HGNC:1030</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>83.3078</td>\n",
       "      <td>...</td>\n",
       "      <td>P30411</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P32299</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>636.0</td>\n",
       "      <td>ENSG00000168398</td>\n",
       "      <td>BDKRB2</td>\n",
       "      <td>ENSMUSG00000021070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDB00004</td>\n",
       "      <td>None</td>\n",
       "      <td>Dpp4</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Ada → Dpp4</td>\n",
       "      <td>HGNC:186</td>\n",
       "      <td>HGNC:3009</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>80.7163</td>\n",
       "      <td>...</td>\n",
       "      <td>P27487</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P28843</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1365.0</td>\n",
       "      <td>ENSG00000197635</td>\n",
       "      <td>DPP4</td>\n",
       "      <td>ENSMUSG00000035000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDB00005</td>\n",
       "      <td>None</td>\n",
       "      <td>Epha3</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Adam10 → Epha3</td>\n",
       "      <td>HGNC:188</td>\n",
       "      <td>HGNC:3387</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>96.1230</td>\n",
       "      <td>...</td>\n",
       "      <td>P29320</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P29319</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>ENSG00000044524</td>\n",
       "      <td>EPHA3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3956</th>\n",
       "      <td>CDB03444</td>\n",
       "      <td>None</td>\n",
       "      <td>Pcdhb21</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Pcdhb21 → Pcdhb21</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3957</th>\n",
       "      <td>CDB03445</td>\n",
       "      <td>None</td>\n",
       "      <td>Pcdhb22</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Pcdhb22 → Pcdhb22</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Q9Y5E8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q91XZ8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.989</td>\n",
       "      <td>1191.0</td>\n",
       "      <td>ENSG00000113248</td>\n",
       "      <td>PCDHB15</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3958</th>\n",
       "      <td>CDB03446</td>\n",
       "      <td>None</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Pcdhgb8 → Pcdhgb8</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3959</th>\n",
       "      <td>CDB03447</td>\n",
       "      <td>None</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Saa3 → Tlr4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>O00206</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q9QUK6</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>909.0</td>\n",
       "      <td>ENSG00000136869</td>\n",
       "      <td>TLR4</td>\n",
       "      <td>ENSMUSG00000039005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960</th>\n",
       "      <td>CDB03448</td>\n",
       "      <td>None</td>\n",
       "      <td>Timd2</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>Sema4a → Timd2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Q96D42</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q5QNS5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>157.0</td>\n",
       "      <td>ENSG00000113249</td>\n",
       "      <td>HAVCR1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3961 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Interaction ID LR Pair Card   Ligand  \\\n",
       "0          CDB00001         None    Hspa5   \n",
       "1          CDB00002         None     Lrp1   \n",
       "2          CDB00003         None   Bdkrb2   \n",
       "3          CDB00004         None     Dpp4   \n",
       "4          CDB00005         None    Epha3   \n",
       "...             ...          ...      ...   \n",
       "3956       CDB03444         None  Pcdhb21   \n",
       "3957       CDB03445         None  Pcdhb22   \n",
       "3958       CDB03446         None  Pcdhgb8   \n",
       "3959       CDB03447         None     Tlr4   \n",
       "3960       CDB03448         None    Timd2   \n",
       "\n",
       "                                               Receptor      Mouse LR Pair  \\\n",
       "0     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...        A2m → Hspa5   \n",
       "1     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...         A2m → Lrp1   \n",
       "2     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...       Ace → Bdkrb2   \n",
       "3     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...         Ada → Dpp4   \n",
       "4     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...     Adam10 → Epha3   \n",
       "...                                                 ...                ...   \n",
       "3956  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...  Pcdhb21 → Pcdhb21   \n",
       "3957  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...  Pcdhb22 → Pcdhb22   \n",
       "3958  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...  Pcdhgb8 → Pcdhgb8   \n",
       "3959  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...        Saa3 → Tlr4   \n",
       "3960  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...     Sema4a → Timd2   \n",
       "\n",
       "     Ligand HGNC ID Receptor HGNC ID  Ligand GOC score  Ligand WGA coverage  \\\n",
       "0            HGNC:7        HGNC:5238              75.0                100.0   \n",
       "1            HGNC:7        HGNC:6692              75.0                100.0   \n",
       "2         HGNC:2707        HGNC:1030             100.0                100.0   \n",
       "3          HGNC:186        HGNC:3009             100.0                100.0   \n",
       "4          HGNC:188        HGNC:3387             100.0                100.0   \n",
       "...             ...              ...               ...                  ...   \n",
       "3956           None             None               NaN                  NaN   \n",
       "3957           None             None               NaN                  NaN   \n",
       "3958           None             None               NaN                  NaN   \n",
       "3959           None             None               NaN                  NaN   \n",
       "3960           None             None               NaN                  NaN   \n",
       "\n",
       "      Ligand % Identity  ...  Receptor_human_protein  \\\n",
       "0               72.4559  ...                  P11021   \n",
       "1               72.4559  ...                  Q07954   \n",
       "2               83.3078  ...                  P30411   \n",
       "3               80.7163  ...                  P27487   \n",
       "4               96.1230  ...                  P29320   \n",
       "...                 ...  ...                     ...   \n",
       "3956                NaN  ...                     NaN   \n",
       "3957                NaN  ...                  Q9Y5E8   \n",
       "3958                NaN  ...                     NaN   \n",
       "3959                NaN  ...                  O00206   \n",
       "3960                NaN  ...                  Q96D42   \n",
       "\n",
       "      Receptor_human_inparalog_score Receptor_human_seed_score  \\\n",
       "0                                1.0                       1.0   \n",
       "1                                1.0                       1.0   \n",
       "2                                1.0                       1.0   \n",
       "3                                1.0                       1.0   \n",
       "4                                1.0                       1.0   \n",
       "...                              ...                       ...   \n",
       "3956                             NaN                       NaN   \n",
       "3957                             1.0                       1.0   \n",
       "3958                             NaN                       NaN   \n",
       "3959                             1.0                       1.0   \n",
       "3960                             1.0                       1.0   \n",
       "\n",
       "      Receptor_mouse_protein  Receptor_mouse_inparalog_score  \\\n",
       "0                     P20029                             1.0   \n",
       "1                     Q91ZX7                             1.0   \n",
       "2                     P32299                             1.0   \n",
       "3                     P28843                             1.0   \n",
       "4                     P29319                             1.0   \n",
       "...                      ...                             ...   \n",
       "3956                     NaN                             NaN   \n",
       "3957                  Q91XZ8                             1.0   \n",
       "3958                     NaN                             NaN   \n",
       "3959                  Q9QUK6                             1.0   \n",
       "3960                  Q5QNS5                             1.0   \n",
       "\n",
       "      Receptor_mouse_seed_score  Receptor_bitscore  \\\n",
       "0                         1.000             1246.0   \n",
       "1                         1.000             6853.0   \n",
       "2                         1.000              636.0   \n",
       "3                         1.000             1365.0   \n",
       "4                         1.000             1923.0   \n",
       "...                         ...                ...   \n",
       "3956                        NaN                NaN   \n",
       "3957                      0.989             1191.0   \n",
       "3958                        NaN                NaN   \n",
       "3959                      1.000              909.0   \n",
       "3960                      1.000              157.0   \n",
       "\n",
       "      Receptor_human_ensembl_gene_id Receptor_human_gene  \\\n",
       "0                    ENSG00000044574               HSPA5   \n",
       "1                    ENSG00000123384                LRP1   \n",
       "2                    ENSG00000168398              BDKRB2   \n",
       "3                    ENSG00000197635                DPP4   \n",
       "4                    ENSG00000044524               EPHA3   \n",
       "...                              ...                 ...   \n",
       "3956                             NaN                 NaN   \n",
       "3957                 ENSG00000113248             PCDHB15   \n",
       "3958                             NaN                 NaN   \n",
       "3959                 ENSG00000136869                TLR4   \n",
       "3960                 ENSG00000113249              HAVCR1   \n",
       "\n",
       "     Receptor_mouse_ensembl_gene_id  \n",
       "0                ENSMUSG00000026864  \n",
       "1                               NaN  \n",
       "2                ENSMUSG00000021070  \n",
       "3                ENSMUSG00000035000  \n",
       "4                               NaN  \n",
       "...                             ...  \n",
       "3956                            NaN  \n",
       "3957                            NaN  \n",
       "3958                            NaN  \n",
       "3959             ENSMUSG00000039005  \n",
       "3960                            NaN  \n",
       "\n",
       "[3961 rows x 43 columns]"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "id": "b13f430f-a79e-42be-8355-98bfed856ea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_result.to_csv(\"data/human_mouse_merged_ensemblBiomaRt_inParanoid.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "6580b556-1ead-411a-8de0-c758839f29c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Interaction ID</th>\n",
       "      <th>Ligand</th>\n",
       "      <th>Receptor</th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>Mouse LR Pair</th>\n",
       "      <th>Ligand HGNC ID</th>\n",
       "      <th>Receptor HGNC ID</th>\n",
       "      <th>Ligand GOC score</th>\n",
       "      <th>Ligand WGA coverage</th>\n",
       "      <th>...</th>\n",
       "      <th>Receptor_human_protein</th>\n",
       "      <th>Receptor_human_inparalog_score</th>\n",
       "      <th>Receptor_human_seed_score</th>\n",
       "      <th>Receptor_mouse_protein</th>\n",
       "      <th>Receptor_mouse_inparalog_score</th>\n",
       "      <th>Receptor_mouse_seed_score</th>\n",
       "      <th>Receptor_bitscore</th>\n",
       "      <th>Receptor_human_ensembl_gene_id</th>\n",
       "      <th>Receptor_human_gene</th>\n",
       "      <th>Receptor_mouse_ensembl_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>CDB00001</td>\n",
       "      <td>A2m</td>\n",
       "      <td>Hspa5</td>\n",
       "      <td>A2M-HSPA5</td>\n",
       "      <td>A2m → Hspa5</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>HGNC:5238</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>...</td>\n",
       "      <td>P11021</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P20029</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1246.0</td>\n",
       "      <td>ENSG00000044574</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>ENSMUSG00000026864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>CDB00002</td>\n",
       "      <td>A2m</td>\n",
       "      <td>Lrp1</td>\n",
       "      <td>A2M-LRP1</td>\n",
       "      <td>A2m → Lrp1</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>HGNC:6692</td>\n",
       "      <td>75.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Q07954</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q91ZX7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>6853.0</td>\n",
       "      <td>ENSG00000123384</td>\n",
       "      <td>LRP1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>CDB00003</td>\n",
       "      <td>Ace</td>\n",
       "      <td>Bdkrb2</td>\n",
       "      <td>ACE-BDKRB2</td>\n",
       "      <td>Ace → Bdkrb2</td>\n",
       "      <td>HGNC:2707</td>\n",
       "      <td>HGNC:1030</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>...</td>\n",
       "      <td>P30411</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P32299</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>636.0</td>\n",
       "      <td>ENSG00000168398</td>\n",
       "      <td>BDKRB2</td>\n",
       "      <td>ENSMUSG00000021070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>CDB00004</td>\n",
       "      <td>Ada</td>\n",
       "      <td>Dpp4</td>\n",
       "      <td>ADA-DPP4</td>\n",
       "      <td>Ada → Dpp4</td>\n",
       "      <td>HGNC:186</td>\n",
       "      <td>HGNC:3009</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>...</td>\n",
       "      <td>P27487</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P28843</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1365.0</td>\n",
       "      <td>ENSG00000197635</td>\n",
       "      <td>DPP4</td>\n",
       "      <td>ENSMUSG00000035000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>CDB00005</td>\n",
       "      <td>Adam10</td>\n",
       "      <td>Epha3</td>\n",
       "      <td>ADAM10-EPHA3</td>\n",
       "      <td>Adam10 → Epha3</td>\n",
       "      <td>HGNC:188</td>\n",
       "      <td>HGNC:3387</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>...</td>\n",
       "      <td>P29320</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>P29319</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1923.0</td>\n",
       "      <td>ENSG00000044524</td>\n",
       "      <td>EPHA3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3956</th>\n",
       "      <td>3956</td>\n",
       "      <td>CDB03444</td>\n",
       "      <td>Pcdhb21</td>\n",
       "      <td>Pcdhb21</td>\n",
       "      <td>Pcdhb21-Pcdhb21</td>\n",
       "      <td>Pcdhb21 → Pcdhb21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3957</th>\n",
       "      <td>3957</td>\n",
       "      <td>CDB03445</td>\n",
       "      <td>Pcdhb22</td>\n",
       "      <td>Pcdhb22</td>\n",
       "      <td>Pcdhb22-Pcdhb22</td>\n",
       "      <td>Pcdhb22 → Pcdhb22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Q9Y5E8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q91XZ8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.989</td>\n",
       "      <td>1191.0</td>\n",
       "      <td>ENSG00000113248</td>\n",
       "      <td>PCDHB15</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3958</th>\n",
       "      <td>3958</td>\n",
       "      <td>CDB03446</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>Pcdhgb8-Pcdhgb8</td>\n",
       "      <td>Pcdhgb8 → Pcdhgb8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3959</th>\n",
       "      <td>3959</td>\n",
       "      <td>CDB03447</td>\n",
       "      <td>Saa3</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>Saa3-Tlr4</td>\n",
       "      <td>Saa3 → Tlr4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>O00206</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q9QUK6</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>909.0</td>\n",
       "      <td>ENSG00000136869</td>\n",
       "      <td>TLR4</td>\n",
       "      <td>ENSMUSG00000039005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3960</th>\n",
       "      <td>3960</td>\n",
       "      <td>CDB03448</td>\n",
       "      <td>Sema4a</td>\n",
       "      <td>Timd2</td>\n",
       "      <td>Sema4a-Timd2</td>\n",
       "      <td>Sema4a → Timd2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Q96D42</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q5QNS5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>157.0</td>\n",
       "      <td>ENSG00000113249</td>\n",
       "      <td>HAVCR1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3961 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0 Interaction ID   Ligand Receptor     LR Pair Card  \\\n",
       "0              0       CDB00001      A2m    Hspa5        A2M-HSPA5   \n",
       "1              1       CDB00002      A2m     Lrp1         A2M-LRP1   \n",
       "2              2       CDB00003      Ace   Bdkrb2       ACE-BDKRB2   \n",
       "3              3       CDB00004      Ada     Dpp4         ADA-DPP4   \n",
       "4              4       CDB00005   Adam10    Epha3     ADAM10-EPHA3   \n",
       "...          ...            ...      ...      ...              ...   \n",
       "3956        3956       CDB03444  Pcdhb21  Pcdhb21  Pcdhb21-Pcdhb21   \n",
       "3957        3957       CDB03445  Pcdhb22  Pcdhb22  Pcdhb22-Pcdhb22   \n",
       "3958        3958       CDB03446  Pcdhgb8  Pcdhgb8  Pcdhgb8-Pcdhgb8   \n",
       "3959        3959       CDB03447     Saa3     Tlr4        Saa3-Tlr4   \n",
       "3960        3960       CDB03448   Sema4a    Timd2     Sema4a-Timd2   \n",
       "\n",
       "          Mouse LR Pair Ligand HGNC ID Receptor HGNC ID  Ligand GOC score  \\\n",
       "0           A2m → Hspa5         HGNC:7        HGNC:5238              75.0   \n",
       "1            A2m → Lrp1         HGNC:7        HGNC:6692              75.0   \n",
       "2          Ace → Bdkrb2      HGNC:2707        HGNC:1030             100.0   \n",
       "3            Ada → Dpp4       HGNC:186        HGNC:3009             100.0   \n",
       "4        Adam10 → Epha3       HGNC:188        HGNC:3387             100.0   \n",
       "...                 ...            ...              ...               ...   \n",
       "3956  Pcdhb21 → Pcdhb21            NaN              NaN               NaN   \n",
       "3957  Pcdhb22 → Pcdhb22            NaN              NaN               NaN   \n",
       "3958  Pcdhgb8 → Pcdhgb8            NaN              NaN               NaN   \n",
       "3959        Saa3 → Tlr4            NaN              NaN               NaN   \n",
       "3960     Sema4a → Timd2            NaN              NaN               NaN   \n",
       "\n",
       "      Ligand WGA coverage  ...  Receptor_human_protein  \\\n",
       "0                   100.0  ...                  P11021   \n",
       "1                   100.0  ...                  Q07954   \n",
       "2                   100.0  ...                  P30411   \n",
       "3                   100.0  ...                  P27487   \n",
       "4                   100.0  ...                  P29320   \n",
       "...                   ...  ...                     ...   \n",
       "3956                  NaN  ...                     NaN   \n",
       "3957                  NaN  ...                  Q9Y5E8   \n",
       "3958                  NaN  ...                     NaN   \n",
       "3959                  NaN  ...                  O00206   \n",
       "3960                  NaN  ...                  Q96D42   \n",
       "\n",
       "      Receptor_human_inparalog_score  Receptor_human_seed_score  \\\n",
       "0                                1.0                        1.0   \n",
       "1                                1.0                        1.0   \n",
       "2                                1.0                        1.0   \n",
       "3                                1.0                        1.0   \n",
       "4                                1.0                        1.0   \n",
       "...                              ...                        ...   \n",
       "3956                             NaN                        NaN   \n",
       "3957                             1.0                        1.0   \n",
       "3958                             NaN                        NaN   \n",
       "3959                             1.0                        1.0   \n",
       "3960                             1.0                        1.0   \n",
       "\n",
       "     Receptor_mouse_protein  Receptor_mouse_inparalog_score  \\\n",
       "0                    P20029                             1.0   \n",
       "1                    Q91ZX7                             1.0   \n",
       "2                    P32299                             1.0   \n",
       "3                    P28843                             1.0   \n",
       "4                    P29319                             1.0   \n",
       "...                     ...                             ...   \n",
       "3956                    NaN                             NaN   \n",
       "3957                 Q91XZ8                             1.0   \n",
       "3958                    NaN                             NaN   \n",
       "3959                 Q9QUK6                             1.0   \n",
       "3960                 Q5QNS5                             1.0   \n",
       "\n",
       "      Receptor_mouse_seed_score  Receptor_bitscore  \\\n",
       "0                         1.000             1246.0   \n",
       "1                         1.000             6853.0   \n",
       "2                         1.000              636.0   \n",
       "3                         1.000             1365.0   \n",
       "4                         1.000             1923.0   \n",
       "...                         ...                ...   \n",
       "3956                        NaN                NaN   \n",
       "3957                      0.989             1191.0   \n",
       "3958                        NaN                NaN   \n",
       "3959                      1.000              909.0   \n",
       "3960                      1.000              157.0   \n",
       "\n",
       "      Receptor_human_ensembl_gene_id  Receptor_human_gene  \\\n",
       "0                    ENSG00000044574                HSPA5   \n",
       "1                    ENSG00000123384                 LRP1   \n",
       "2                    ENSG00000168398               BDKRB2   \n",
       "3                    ENSG00000197635                 DPP4   \n",
       "4                    ENSG00000044524                EPHA3   \n",
       "...                              ...                  ...   \n",
       "3956                             NaN                  NaN   \n",
       "3957                 ENSG00000113248              PCDHB15   \n",
       "3958                             NaN                  NaN   \n",
       "3959                 ENSG00000136869                 TLR4   \n",
       "3960                 ENSG00000113249               HAVCR1   \n",
       "\n",
       "     Receptor_mouse_ensembl_gene_id  \n",
       "0                ENSMUSG00000026864  \n",
       "1                               NaN  \n",
       "2                ENSMUSG00000021070  \n",
       "3                ENSMUSG00000035000  \n",
       "4                               NaN  \n",
       "...                             ...  \n",
       "3956                            NaN  \n",
       "3957                            NaN  \n",
       "3958                            NaN  \n",
       "3959             ENSMUSG00000039005  \n",
       "3960                            NaN  \n",
       "\n",
       "[3961 rows x 44 columns]"
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_result= pd.read_csv(\"data/human_mouse_merged_ensemblBiomaRt_inParanoid.csv\")\n",
    "final_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "id": "d9bf4718-0852-401f-ab64-0a6582f04b9e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0', 'Interaction ID', 'Ligand', 'Receptor', 'LR Pair Card',\n",
       "       'Mouse LR Pair', 'Ligand HGNC ID', 'Receptor HGNC ID',\n",
       "       'Ligand GOC score', 'Ligand WGA coverage', 'Ligand % Identity',\n",
       "       'Ligand Target % Identity', 'Ligand Orthology Confidence',\n",
       "       'Ligand Ensembl ID', 'Receptor GOC score', 'Receptor WGA coverage',\n",
       "       'Receptor % Identity', 'Receptor Target % Identity',\n",
       "       'Receptor Orthology Confidence', 'Receptor Ensembl ID',\n",
       "       'Human Ligand Ensembl ID', 'Human Receptor Ensembl ID',\n",
       "       'Ligand_cluster_id', 'Ligand_human_protein',\n",
       "       'Ligand_human_inparalog_score', 'Ligand_human_seed_score',\n",
       "       'Ligand_mouse_protein', 'Ligand_mouse_inparalog_score',\n",
       "       'Ligand_mouse_seed_score', 'Ligand_bitscore',\n",
       "       'Ligand_human_ensembl_gene_id', 'Ligand_human_gene',\n",
       "       'Ligand_mouse_ensembl_gene_id', 'Receptor_cluster_id',\n",
       "       'Receptor_human_protein', 'Receptor_human_inparalog_score',\n",
       "       'Receptor_human_seed_score', 'Receptor_mouse_protein',\n",
       "       'Receptor_mouse_inparalog_score', 'Receptor_mouse_seed_score',\n",
       "       'Receptor_bitscore', 'Receptor_human_ensembl_gene_id',\n",
       "       'Receptor_human_gene', 'Receptor_mouse_ensembl_gene_id'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 288,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_result.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "id": "5cd094a1-bd10-4bd9-b0f7-4493b605bc16",
   "metadata": {},
   "outputs": [],
   "source": [
    "score_cols = [\n",
    "    ligand_human_inparalog_score,\n",
    "    receptor_human_inparalog_score,\n",
    "    ligand_mouse_inparalog_score,\n",
    "    receptor_mouse_inparalog_score,\n",
    "    ligand_human_seed_score,\n",
    "    receptor_human_seed_score,\n",
    "    ligand_mouse_seed_score,\n",
    "    receptor_mouse_seed_score,\n",
    "    ligand_bit_score,\n",
    "    receptor_bit_score\n",
    "]\n",
    "\n",
    "for col in score_cols:\n",
    "    if col in final_result.columns:\n",
    "        final_result[col] = pd.to_numeric(final_result[col], errors='coerce')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "id": "50dbf801-97c3-4ff9-b79a-6574ac0d880d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def summarize_orthologs(human_col, species_col, label,\n",
    "                        confidence_orth_col=None, confidence_orth_threshold=None,\n",
    "                        GOC_col=None, GOC_threshold=None,\n",
    "                        perc_identity_col= None, perc_identity_thres= None,\n",
    "                        ligand_human_inparalog_score_col=None, ligand_human_inparalog_score_threshold=None,\n",
    "                        receptor_human_inparalog_score_col=None, receptor_human_inparalog_score_threshold=None,\n",
    "                        ligand_mouse_inparalog_score_col=None, ligand_mouse_inparalog_score_threshold=None,\n",
    "                        receptor_mouse_inparalog_score_col=None, receptor_mouse_inparalog_score_threshold=None,\n",
    "                        ligand_human_seed_score_col=None, ligand_human_seed_score_threshold=None,\n",
    "                        receptor_human_seed_score_col=None, receptor_human_seed_score_threshold=None,\n",
    "                        ligand_mouse_seed_score_col=None, ligand_mouse_seed_score_threshold=None,\n",
    "                        receptor_mouse_seed_score_col=None, receptor_mouse_seed_score_threshold=None,\n",
    "                        ligand_bit_score_col=None, ligand_bit_score_threshold=None,\n",
    "                        receptor_bit_score_col=None, receptor_bit_score_threshold=None):\n",
    "    \n",
    "    df = final_result.copy()\n",
    "\n",
    "    # Apply filters one by one if thresholds are given\n",
    "    filters = [\n",
    "        (confidence_orth_col, lambda x: x == confidence_orth_threshold),\n",
    "        (GOC_col, lambda x: x >= GOC_threshold),\n",
    "        (perc_identity_col, lambda x: x >= perc_identity_thres),\n",
    "        (ligand_human_inparalog_score_col, lambda x: x >= ligand_human_inparalog_score_threshold),\n",
    "        (receptor_human_inparalog_score_col, lambda x: x >= receptor_human_inparalog_score_threshold),\n",
    "        (ligand_mouse_inparalog_score_col, lambda x: x >= ligand_mouse_inparalog_score_threshold),\n",
    "        (receptor_mouse_inparalog_score_col, lambda x: x >= receptor_mouse_inparalog_score_threshold),\n",
    "        (ligand_human_seed_score_col, lambda x: x >= ligand_human_seed_score_threshold),\n",
    "        (receptor_human_seed_score_col, lambda x: x >= receptor_human_seed_score_threshold),\n",
    "        (ligand_mouse_seed_score_col, lambda x: x >= ligand_mouse_seed_score_threshold),\n",
    "        (receptor_mouse_seed_score_col, lambda x: x >= receptor_mouse_seed_score_threshold),\n",
    "        (ligand_bit_score_col, lambda x: x >= ligand_bit_score_threshold),\n",
    "        (receptor_bit_score_col, lambda x: x >= receptor_bit_score_threshold),\n",
    "    ]\n",
    "\n",
    "\n",
    "    # Apply filters and track how many rows were removed\n",
    "    original_rows = df.shape[0]\n",
    "    for col, condition in filters:\n",
    "        if col and condition is not None:\n",
    "            before = df.shape[0]\n",
    "            df = df[df[col].apply(condition)]\n",
    "            after = df.shape[0]\n",
    "            print(f\"Filtered {col}: {before - after} rows removed (remaining: {after})\")\n",
    "\n",
    "    # Compute unique ortholog pairs\n",
    "    unique_pairs = df[[human_col, species_col]].drop_duplicates()\n",
    "\n",
    "    # Count mouse orthologs per human gene\n",
    "    counts = (\n",
    "        unique_pairs\n",
    "        .groupby(human_col)[species_col]\n",
    "        .count()\n",
    "        .sort_values(ascending=False)\n",
    "        .reset_index(name='count')\n",
    "    )\n",
    "\n",
    "    # Build tag from filters\n",
    "    filter_tag = label.lower()\n",
    "    if confidence_orth_threshold is not None:\n",
    "        filter_tag += f\"_conf{confidence_orth_threshold}\"\n",
    "    if GOC_threshold is not None:\n",
    "        filter_tag += f\"_GOCge{GOC_threshold}\"\n",
    "    if ligand_human_inparalog_score_threshold is not None:\n",
    "        filter_tag += f\"_LHISge{ligand_human_inparalog_score_threshold}\"\n",
    "    if receptor_bit_score_threshold is not None:\n",
    "        filter_tag += f\"_RBSge{receptor_bit_score_threshold}\"\n",
    "\n",
    "    counts.to_csv(f\"data/human_mouse_orth_count_{filter_tag}.csv\", index=False)\n",
    "\n",
    "    summary_counts = counts['count'].value_counts().sort_index()\n",
    "    total_human_genes = counts.shape[0]\n",
    "\n",
    "    # Collect active filters for reporting\n",
    "    active_filters = []\n",
    "    if confidence_orth_col and confidence_orth_threshold is not None:\n",
    "        if isinstance(confidence_orth_threshold, (list, set, tuple)):\n",
    "            active_filters.append(f\"{confidence_orth_col} in {sorted(confidence_orth_threshold)}\")\n",
    "        else:\n",
    "            active_filters.append(f\"{confidence_orth_col} equals '{confidence_orth_threshold}'\")\n",
    "    if GOC_col and GOC_threshold is not None:\n",
    "        active_filters.append(f\"{GOC_col} ≥ {GOC_threshold}\")\n",
    "    if perc_identity_col and perc_identity_thres is not None:\n",
    "        active_filters.append(f\"{perc_identity_col} ≥ {perc_identity_thres}\")\n",
    "    if ligand_human_inparalog_score_col and ligand_human_inparalog_score_threshold is not None:\n",
    "        active_filters.append(f\"{ligand_human_inparalog_score_col} ≥ {ligand_human_inparalog_score_threshold}\")\n",
    "    if receptor_human_inparalog_score_col and receptor_human_inparalog_score_threshold is not None:\n",
    "        active_filters.append(f\"{receptor_human_inparalog_score_col} ≥ {receptor_human_inparalog_score_threshold}\")\n",
    "    if ligand_mouse_inparalog_score_col and ligand_mouse_inparalog_score_threshold is not None:\n",
    "        active_filters.append(f\"{ligand_mouse_inparalog_score_col} ≥ {ligand_mouse_inparalog_score_threshold}\")\n",
    "    if receptor_mouse_inparalog_score_col and receptor_mouse_inparalog_score_threshold is not None:\n",
    "        active_filters.append(f\"{receptor_mouse_inparalog_score_col} ≥ {receptor_mouse_inparalog_score_threshold}\")\n",
    "    if ligand_human_seed_score_col and ligand_human_seed_score_threshold is not None:\n",
    "        active_filters.append(f\"{ligand_human_seed_score_col} ≥ {ligand_human_seed_score_threshold}\")\n",
    "    if receptor_human_seed_score_col and receptor_human_seed_score_threshold is not None:\n",
    "        active_filters.append(f\"{receptor_human_seed_score_col} ≥ {receptor_human_seed_score_threshold}\")\n",
    "    if ligand_mouse_seed_score_col and ligand_mouse_seed_score_threshold is not None:\n",
    "        active_filters.append(f\"{ligand_mouse_seed_score_col} ≥ {ligand_mouse_seed_score_threshold}\")\n",
    "    if receptor_mouse_seed_score_col and receptor_mouse_seed_score_threshold is not None:\n",
    "        active_filters.append(f\"{receptor_mouse_seed_score_col} ≥ {receptor_mouse_seed_score_threshold}\")\n",
    "    if ligand_bit_score_col and ligand_bit_score_threshold is not None:\n",
    "        active_filters.append(f\"{ligand_bit_score_col} ≥ {ligand_bit_score_threshold}\")\n",
    "    if receptor_bit_score_col and receptor_bit_score_threshold is not None:\n",
    "        active_filters.append(f\"{receptor_bit_score_col} ≥ {receptor_bit_score_threshold}\")\n",
    "    \n",
    "    filter_text = \"; \".join(active_filters) if active_filters else \"No filters applied\"\n",
    "    \n",
    "    summary_lines = [\n",
    "        f\"Out of {total_human_genes} unique human {label.lower()} genes:\",\n",
    "        f\" - Filters applied: {filter_text}\"\n",
    "    ]\n",
    "\n",
    "    for orth_count, gene_count in summary_counts.items():\n",
    "        summary_lines.append(\n",
    "            f\" - {gene_count} human {label.lower()} genes had {orth_count} mouse ortholog(s)\"\n",
    "        )\n",
    "\n",
    "    return \"\\n\".join(summary_lines)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "id": "313f9aba-0ea2-4d3d-914f-e521ff22914f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Detect columns\n",
    "# Ligand\n",
    "confidence_orth_ligand = [col for col in final_result.columns if \"Ligand Orthology Confidence\" in col][0]\n",
    "GOC_col_ligand = [col for col in final_result.columns if \"Ligand GOC\" in col][0]\n",
    "percIdent_col_ligand = [col for col in final_result.columns if \"Ligand % Identity\" in col][0]\n",
    "human_ligand_col = [col for col in final_result.columns if \"Ligand HGNC ID\" in col][0]\n",
    "ligand_col = [col for col in final_result.columns if \"Ligand\" in col][0]\n",
    "ligand_col = [col for col in final_result.columns if \"Ligand\" in col][0]\n",
    "ligand_human_inparalog_score = [col for col in final_result.columns if \"Ligand_human_inparalog_score\" in col][0]\n",
    "ligand_mouse_inparalog_score = [col for col in final_result.columns if \"Ligand_mouse_inparalog_score\" in col][0]\n",
    "ligand_human_seed_score = [col for col in final_result.columns if \"Ligand_human_seed_score\" in col][0]\n",
    "ligand_mouse_seed_score = [col for col in final_result.columns if \"Ligand_mouse_seed_score\" in col][0]\n",
    "ligand_bit_score = [col for col in final_result.columns if \"Ligand_bitscore\" in col][0]\n",
    "\n",
    "#Receptor\n",
    "human_receptor_col = [col for col in final_result.columns if \"Receptor HGNC ID\" in col][0]\n",
    "confidence_orth_receptor = [col for col in final_result.columns if \"Receptor Orthology Confidence\" in col][0]\n",
    "GOC_col_receptor = [col for col in final_result.columns if \"Receptor GOC\" in col][0]\n",
    "percIdent_col_receptor = [col for col in final_result.columns if \"Receptor % Identity\" in col][0]\n",
    "receptor_col = [col for col in final_result.columns if \"Receptor\" in col][0]\n",
    "receptor_human_inparalog_score = [col for col in final_result.columns if \"Receptor_human_inparalog_score\" in col][0]\n",
    "receptor_mouse_inparalog_score = [col for col in final_result.columns if \"Receptor_mouse_inparalog_score\" in col][0]\n",
    "receptor_human_seed_score = [col for col in final_result.columns if \"Receptor_human_seed_score\" in col][0]\n",
    "receptor_mouse_seed_score = [col for col in final_result.columns if \"Receptor_mouse_seed_score\" in col][0]\n",
    "receptor_bit_score = [col for col in final_result.columns if \"Receptor_bitscore\" in col][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "id": "cf33bf7a-8f0d-4b13-b319-f47da921fc49",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered Ligand Orthology Confidence: 657 rows removed (remaining: 3304)\n",
      "Filtered Ligand % Identity: 246 rows removed (remaining: 3058)\n",
      "Out of 835 unique human ligand genes:\n",
      " - Filters applied: Ligand Orthology Confidence equals '1'; Ligand % Identity ≥ 60\n",
      " - 814 human ligand genes had 1 mouse ortholog(s)\n",
      " - 13 human ligand genes had 2 mouse ortholog(s)\n",
      " - 4 human ligand genes had 3 mouse ortholog(s)\n",
      " - 4 human ligand genes had 5 mouse ortholog(s)\n"
     ]
    }
   ],
   "source": [
    "ligand_summary = summarize_orthologs(\n",
    "    human_col=human_ligand_col,\n",
    "    species_col=ligand_col,\n",
    "    label=\"Ligand\",\n",
    "    confidence_orth_col=confidence_orth_ligand,\n",
    "    confidence_orth_threshold=1,\n",
    "    perc_identity_col = percIdent_col_ligand,\n",
    "    perc_identity_thres = 60\n",
    "    # GOC_col=GOC_col_ligand,\n",
    "    # GOC_threshold=100,\n",
    "    # ligand_human_inparalog_score_col=ligand_human_inparalog_score,\n",
    "    # ligand_human_inparalog_score_threshold=1,\n",
    "    # ligand_human_seed_score_col=ligand_human_seed_score,\n",
    "    # ligand_human_seed_score_threshold=1,\n",
    "    # ligand_mouse_seed_score_col=ligand_mouse_seed_score,\n",
    "    # ligand_mouse_seed_score_threshold=1,\n",
    "    # ligand_bit_score_col=ligand_bit_score,\n",
    "    # ligand_bit_score_threshold=40\n",
    ")\n",
    "\n",
    "print(ligand_summary)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "id": "1c363032-01f7-471b-8d7a-9a9c6392cb75",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered Receptor Orthology Confidence: 779 rows removed (remaining: 3182)\n",
      "Filtered Receptor % Identity: 107 rows removed (remaining: 3075)\n",
      "Out of 684 unique human receptor genes:\n",
      " - Filters applied: Receptor Orthology Confidence equals '1'; Receptor % Identity ≥ 60\n",
      " - 679 human receptor genes had 1 mouse ortholog(s)\n",
      " - 4 human receptor genes had 2 mouse ortholog(s)\n",
      " - 1 human receptor genes had 3 mouse ortholog(s)\n"
     ]
    }
   ],
   "source": [
    "receptor_summary = summarize_orthologs(\n",
    "    human_col=human_receptor_col,\n",
    "    species_col=receptor_col,\n",
    "    label=\"Receptor\",\n",
    "    confidence_orth_col=confidence_orth_receptor,\n",
    "    confidence_orth_threshold=1,\n",
    "    perc_identity_thres = 60,\n",
    "    perc_identity_col = percIdent_col_receptor,\n",
    "    # GOC_col=GOC_col_receptor,\n",
    "    # GOC_threshold=100,\n",
    "    # receptor_human_inparalog_score_col=receptor_human_inparalog_score,\n",
    "    # receptor_human_inparalog_score_threshold=1,\n",
    "    # receptor_human_seed_score_col=receptor_human_seed_score,\n",
    "    # receptor_human_seed_score_threshold=1,\n",
    "    # receptor_mouse_seed_score_col=receptor_mouse_seed_score,\n",
    "    # receptor_mouse_seed_score_threshold=1,\n",
    "    # receptor_bit_score_col=receptor_bit_score,\n",
    "    # receptor_bit_score_threshold=40\n",
    ")\n",
    "\n",
    "print(receptor_summary)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c54f640f-774d-41d6-a7fb-a641372fb452",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c81b9d6-bf9e-448c-a1ef-48819156f5f2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "547914fa-5f15-4456-926b-2dac185f859d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "386822e7-3375-45e0-a9bc-d7fde2716ab4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09dcc17c-2d8e-49bf-a8aa-9d7dcac790d2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "cc26cb71-4186-4ed9-935d-e4ba6ba46fc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "from io import StringIO\n",
    "from itertools import product\n",
    "\n",
    "# Download InParanoid prot table\n",
    "url = \"https://inparanoidb.sbc.su.se/download/sqltable/9606&10090&prot\"\n",
    "r = requests.get(url)\n",
    "r.raise_for_status()\n",
    "\n",
    "df = pd.read_csv(StringIO(r.text.strip()), sep=\"\\t\", header=None)\n",
    "df.columns = [\"cluster_id\", \"bitscore\", \"source_file\", \"inparalog_score\", \"protein_id\", \"seed_score\"]\n",
    "\n",
    "# Tag each row by species\n",
    "def infer_species(src):\n",
    "    if \"9606\" in src:\n",
    "        return \"human\"\n",
    "    elif \"10090\" in src:\n",
    "        return \"mouse\"\n",
    "    return \"unknown\"\n",
    "\n",
    "df[\"species\"] = df[\"source_file\"].apply(infer_species)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "fe44da0a-838a-4e04-90f0-d8aec872dcb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Expand ortholog pairs within each cluster\n",
    "records = []\n",
    "for cid, grp in df.groupby(\"cluster_id\"):\n",
    "    humans = grp[grp[\"species\"] == \"human\"]\n",
    "    mice = grp[grp[\"species\"] == \"mouse\"]\n",
    "    for h, m in product(humans.itertuples(index=False), mice.itertuples(index=False)):\n",
    "        records.append({\n",
    "            \"cluster_id\": cid,\n",
    "            \"human_protein\": h.protein_id,\n",
    "            \"human_inparalog_score\": h.inparalog_score,\n",
    "            \"human_seed_score\": h.seed_score,\n",
    "            \"mouse_protein\": m.protein_id,\n",
    "            \"mouse_inparalog_score\": m.inparalog_score,\n",
    "            \"mouse_seed_score\": m.seed_score,\n",
    "            \"bitscore\": (h.bitscore + m.bitscore) / 2  # average for now\n",
    "        })\n",
    "\n",
    "df_orthologs = pd.DataFrame(records)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "06145c63-a62f-4fac-a197-94ba7f6889ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_orthologs.to_csv(\"data/inParanoid_mmusculus.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "b1bda587-759c-40c9-a03e-40ca65726119",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>locus_group</th>\n",
       "      <th>locus_type</th>\n",
       "      <th>status</th>\n",
       "      <th>location</th>\n",
       "      <th>location_sortable</th>\n",
       "      <th>alias_symbol</th>\n",
       "      <th>alias_name</th>\n",
       "      <th>...</th>\n",
       "      <th>lncrnadb</th>\n",
       "      <th>enzyme_id</th>\n",
       "      <th>intermediate_filament_db</th>\n",
       "      <th>rna_central_id</th>\n",
       "      <th>lncipedia</th>\n",
       "      <th>gtrnadb</th>\n",
       "      <th>agr</th>\n",
       "      <th>mane_select</th>\n",
       "      <th>gencc</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>ENST00000263100.8|NM_130786.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P04217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>10q11.23</td>\n",
       "      <td>10q11.23</td>\n",
       "      <td>ACF|ASP|ACF64|ACF65|APOBEC1CF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>ENST00000373997.8|NM_014576.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q9NQ94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>FWP007|S863-7|CPAMD5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>ENST00000318602.12|NM_000014.6</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>P01023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HGNC:23336</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>alpha-2-macroglobulin like 1</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>FLJ25179|p170</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:23336</td>\n",
       "      <td>ENST00000299698.12|NM_144670.6</td>\n",
       "      <td>HGNC:23336</td>\n",
       "      <td>A8K2U0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HGNC:30005</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>alpha 1,3-galactosyltransferase 2</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1p35.1</td>\n",
       "      <td>01p35.1</td>\n",
       "      <td>IGBS3S|IGB3S</td>\n",
       "      <td>iGb3 synthase|isoglobotriaosylceramide synthase</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:30005</td>\n",
       "      <td>ENST00000442999.3|NM_001080438.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>U3KPV4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44099</th>\n",
       "      <td>HGNC:3556</td>\n",
       "      <td>FABP2</td>\n",
       "      <td>fatty acid binding protein 2</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>4q26</td>\n",
       "      <td>04q26</td>\n",
       "      <td>I-FABP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:3556</td>\n",
       "      <td>ENST00000274024.4|NM_000134.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P12104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44100</th>\n",
       "      <td>HGNC:3557</td>\n",
       "      <td>FABP3</td>\n",
       "      <td>fatty acid binding protein 3</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1p35.2</td>\n",
       "      <td>01p35.2</td>\n",
       "      <td>H-FABP|O-FABP</td>\n",
       "      <td>mammary-derived growth inhibitor</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:3557</td>\n",
       "      <td>ENST00000373713.7|NM_004102.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P05413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44102</th>\n",
       "      <td>HGNC:3559</td>\n",
       "      <td>FABP4</td>\n",
       "      <td>fatty acid binding protein 4</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>8q21.13</td>\n",
       "      <td>08q21.13</td>\n",
       "      <td>A-FABP|aP2</td>\n",
       "      <td>adipocyte fatty acid binding protein</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:3559</td>\n",
       "      <td>ENST00000256104.5|NM_001442.3</td>\n",
       "      <td>HGNC:3559</td>\n",
       "      <td>P15090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44103</th>\n",
       "      <td>HGNC:3560</td>\n",
       "      <td>FABP5</td>\n",
       "      <td>fatty acid binding protein 5</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>8q21.13</td>\n",
       "      <td>08q21.13</td>\n",
       "      <td>E-FABP|PA-FABP|KFABP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:3560</td>\n",
       "      <td>ENST00000297258.11|NM_001444.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q01469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44106</th>\n",
       "      <td>HGNC:22573</td>\n",
       "      <td>FABP5P3</td>\n",
       "      <td>fatty acid binding protein 5 pseudogene 3</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>Approved</td>\n",
       "      <td>7q36.1</td>\n",
       "      <td>07q36.1</td>\n",
       "      <td>TCAG_1781704</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:22573</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A8MUU1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20294 rows × 55 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          hgnc_id   symbol                                       name  \\\n",
       "0          HGNC:5     A1BG                     alpha-1-B glycoprotein   \n",
       "2      HGNC:24086     A1CF             APOBEC1 complementation factor   \n",
       "3          HGNC:7      A2M                      alpha-2-macroglobulin   \n",
       "5      HGNC:23336    A2ML1               alpha-2-macroglobulin like 1   \n",
       "9      HGNC:30005  A3GALT2          alpha 1,3-galactosyltransferase 2   \n",
       "...           ...      ...                                        ...   \n",
       "44099   HGNC:3556    FABP2               fatty acid binding protein 2   \n",
       "44100   HGNC:3557    FABP3               fatty acid binding protein 3   \n",
       "44102   HGNC:3559    FABP4               fatty acid binding protein 4   \n",
       "44103   HGNC:3560    FABP5               fatty acid binding protein 5   \n",
       "44106  HGNC:22573  FABP5P3  fatty acid binding protein 5 pseudogene 3   \n",
       "\n",
       "               locus_group                 locus_type    status  location  \\\n",
       "0      protein-coding gene  gene with protein product  Approved  19q13.43   \n",
       "2      protein-coding gene  gene with protein product  Approved  10q11.23   \n",
       "3      protein-coding gene  gene with protein product  Approved  12p13.31   \n",
       "5      protein-coding gene  gene with protein product  Approved  12p13.31   \n",
       "9      protein-coding gene  gene with protein product  Approved    1p35.1   \n",
       "...                    ...                        ...       ...       ...   \n",
       "44099  protein-coding gene  gene with protein product  Approved      4q26   \n",
       "44100  protein-coding gene  gene with protein product  Approved    1p35.2   \n",
       "44102  protein-coding gene  gene with protein product  Approved   8q21.13   \n",
       "44103  protein-coding gene  gene with protein product  Approved   8q21.13   \n",
       "44106           pseudogene                 pseudogene  Approved    7q36.1   \n",
       "\n",
       "      location_sortable                   alias_symbol  \\\n",
       "0              19q13.43                            NaN   \n",
       "2              10q11.23  ACF|ASP|ACF64|ACF65|APOBEC1CF   \n",
       "3              12p13.31           FWP007|S863-7|CPAMD5   \n",
       "5              12p13.31                  FLJ25179|p170   \n",
       "9               01p35.1                   IGBS3S|IGB3S   \n",
       "...                 ...                            ...   \n",
       "44099             04q26                         I-FABP   \n",
       "44100           01p35.2                  H-FABP|O-FABP   \n",
       "44102          08q21.13                     A-FABP|aP2   \n",
       "44103          08q21.13           E-FABP|PA-FABP|KFABP   \n",
       "44106           07q36.1                   TCAG_1781704   \n",
       "\n",
       "                                            alias_name  ... lncrnadb  \\\n",
       "0                                                  NaN  ...      NaN   \n",
       "2                                                  NaN  ...      NaN   \n",
       "3                                                  NaN  ...      NaN   \n",
       "5                                                  NaN  ...      NaN   \n",
       "9      iGb3 synthase|isoglobotriaosylceramide synthase  ...      NaN   \n",
       "...                                                ...  ...      ...   \n",
       "44099                                              NaN  ...      NaN   \n",
       "44100                 mammary-derived growth inhibitor  ...      NaN   \n",
       "44102             adipocyte fatty acid binding protein  ...      NaN   \n",
       "44103                                              NaN  ...      NaN   \n",
       "44106                                              NaN  ...      NaN   \n",
       "\n",
       "      enzyme_id intermediate_filament_db rna_central_id lncipedia gtrnadb  \\\n",
       "0           NaN                      NaN            NaN       NaN     NaN   \n",
       "2           NaN                      NaN            NaN       NaN     NaN   \n",
       "3           NaN                      NaN            NaN       NaN     NaN   \n",
       "5           NaN                      NaN            NaN       NaN     NaN   \n",
       "9           NaN                      NaN            NaN       NaN     NaN   \n",
       "...         ...                      ...            ...       ...     ...   \n",
       "44099       NaN                      NaN            NaN       NaN     NaN   \n",
       "44100       NaN                      NaN            NaN       NaN     NaN   \n",
       "44102       NaN                      NaN            NaN       NaN     NaN   \n",
       "44103       NaN                      NaN            NaN       NaN     NaN   \n",
       "44106       NaN                      NaN            NaN       NaN     NaN   \n",
       "\n",
       "              agr                       mane_select       gencc uniprot_id  \n",
       "0          HGNC:5     ENST00000263100.8|NM_130786.4         NaN     P04217  \n",
       "2      HGNC:24086     ENST00000373997.8|NM_014576.4         NaN     Q9NQ94  \n",
       "3          HGNC:7    ENST00000318602.12|NM_000014.6      HGNC:7     P01023  \n",
       "5      HGNC:23336    ENST00000299698.12|NM_144670.6  HGNC:23336     A8K2U0  \n",
       "9      HGNC:30005  ENST00000442999.3|NM_001080438.1         NaN     U3KPV4  \n",
       "...           ...                               ...         ...        ...  \n",
       "44099   HGNC:3556     ENST00000274024.4|NM_000134.4         NaN     P12104  \n",
       "44100   HGNC:3557     ENST00000373713.7|NM_004102.5         NaN     P05413  \n",
       "44102   HGNC:3559     ENST00000256104.5|NM_001442.3   HGNC:3559     P15090  \n",
       "44103   HGNC:3560    ENST00000297258.11|NM_001444.3         NaN     Q01469  \n",
       "44106  HGNC:22573                               NaN         NaN     A8MUU1  \n",
       "\n",
       "[20294 rows x 55 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hgnc_df = pd.read_csv(\"data/HGNC_gene_info_full.tsv\", sep=\"\\t\", dtype=str)\n",
    "hgnc_df = hgnc_df.dropna(subset=[\"uniprot_ids\"])\n",
    "# Split uniprot_ids on comma and explode\n",
    "hgnc_exploded = hgnc_df.assign(uniprot_id=hgnc_df[\"uniprot_ids\"].str.split(\",\")).explode(\"uniprot_id\")\n",
    "hgnc_exploded[\"uniprot_id\"] = hgnc_exploded[\"uniprot_id\"].str.strip()\n",
    "hgnc_exploded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d6bdabc2-e96e-4bef-8998-2c309fe82656",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>human_protein</th>\n",
       "      <th>human_inparalog_score</th>\n",
       "      <th>human_seed_score</th>\n",
       "      <th>mouse_protein</th>\n",
       "      <th>mouse_inparalog_score</th>\n",
       "      <th>mouse_seed_score</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>human_hgnc_id</th>\n",
       "      <th>human_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Q8WZ42</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A2ASS6</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>60090.0</td>\n",
       "      <td>HGNC:12403</td>\n",
       "      <td>TTN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>Q8NF91</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q6ZWR6</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>14503.0</td>\n",
       "      <td>HGNC:17089</td>\n",
       "      <td>SYNE1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>Q5VST9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A2AAJ9</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12156.0</td>\n",
       "      <td>HGNC:15719</td>\n",
       "      <td>OBSCN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>Q03001</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q91ZU6</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10549.0</td>\n",
       "      <td>HGNC:1090</td>\n",
       "      <td>DST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>Q8WXG9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q8VHN7</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10049.0</td>\n",
       "      <td>HGNC:17416</td>\n",
       "      <td>ADGRV1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20948</th>\n",
       "      <td>17091</td>\n",
       "      <td>A0A1B0GTK5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A0A1B0GSI2</td>\n",
       "      <td>0.212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>46.0</td>\n",
       "      <td>HGNC:52642</td>\n",
       "      <td>FAM236D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20949</th>\n",
       "      <td>17091</td>\n",
       "      <td>A0A1B0GTK5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A0A1B0GSB3</td>\n",
       "      <td>0.178</td>\n",
       "      <td>NaN</td>\n",
       "      <td>46.0</td>\n",
       "      <td>HGNC:52642</td>\n",
       "      <td>FAM236D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20950</th>\n",
       "      <td>17092</td>\n",
       "      <td>Q96LM9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>E9Q1X6</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>HGNC:16166</td>\n",
       "      <td>C20orf173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20954</th>\n",
       "      <td>17096</td>\n",
       "      <td>P0DP42</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A0A494B9K2</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>HGNC:53075</td>\n",
       "      <td>TMEM225B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20956</th>\n",
       "      <td>17098</td>\n",
       "      <td>P31358</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q64389</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>HGNC:1804</td>\n",
       "      <td>CD52</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20583 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       cluster_id human_protein  human_inparalog_score  human_seed_score  \\\n",
       "0               1        Q8WZ42                    1.0               1.0   \n",
       "1               2        Q8NF91                    1.0               1.0   \n",
       "2               3        Q5VST9                    1.0               1.0   \n",
       "4               5        Q03001                    1.0               1.0   \n",
       "5               6        Q8WXG9                    1.0               1.0   \n",
       "...           ...           ...                    ...               ...   \n",
       "20948       17091    A0A1B0GTK5                    1.0               1.0   \n",
       "20949       17091    A0A1B0GTK5                    1.0               1.0   \n",
       "20950       17092        Q96LM9                    1.0               1.0   \n",
       "20954       17096        P0DP42                    1.0               1.0   \n",
       "20956       17098        P31358                    1.0               1.0   \n",
       "\n",
       "      mouse_protein  mouse_inparalog_score  mouse_seed_score  bitscore  \\\n",
       "0            A2ASS6                  1.000               1.0   60090.0   \n",
       "1            Q6ZWR6                  1.000               1.0   14503.0   \n",
       "2            A2AAJ9                  1.000               1.0   12156.0   \n",
       "4            Q91ZU6                  1.000               1.0   10549.0   \n",
       "5            Q8VHN7                  1.000               1.0   10049.0   \n",
       "...             ...                    ...               ...       ...   \n",
       "20948    A0A1B0GSI2                  0.212               NaN      46.0   \n",
       "20949    A0A1B0GSB3                  0.178               NaN      46.0   \n",
       "20950        E9Q1X6                  1.000               1.0      46.0   \n",
       "20954    A0A494B9K2                  1.000               1.0      41.0   \n",
       "20956        Q64389                  1.000               1.0      40.0   \n",
       "\n",
       "      human_hgnc_id human_gene  \n",
       "0        HGNC:12403        TTN  \n",
       "1        HGNC:17089      SYNE1  \n",
       "2        HGNC:15719      OBSCN  \n",
       "4         HGNC:1090        DST  \n",
       "5        HGNC:17416     ADGRV1  \n",
       "...             ...        ...  \n",
       "20948    HGNC:52642    FAM236D  \n",
       "20949    HGNC:52642    FAM236D  \n",
       "20950    HGNC:16166  C20orf173  \n",
       "20954    HGNC:53075   TMEM225B  \n",
       "20956     HGNC:1804       CD52  \n",
       "\n",
       "[20583 rows x 10 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Left join on human_protein\n",
    "df_merged = df_orthologs.merge(\n",
    "    hgnc_exploded[[\"uniprot_id\", \"hgnc_id\", \"symbol\"]],\n",
    "    left_on=\"human_protein\",\n",
    "    right_on=\"uniprot_id\",\n",
    "    how=\"left\"\n",
    ")\n",
    "\n",
    "# Optionally rename\n",
    "df_merged = df_merged.rename(columns={\n",
    "    \"symbol\": \"human_gene\",\n",
    "    \"hgnc_id\": \"human_hgnc_id\"\n",
    "}).drop(columns=[\"uniprot_id\"])\n",
    "df_merged = df_merged.dropna(subset=[\"human_hgnc_id\"])\n",
    "df_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "9174e69a-ed49-4570-8700-8e3419d1d287",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "65d4d761-f6f1-4e03-a76a-d1bcc37763cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "interaction_id_col = [col for col in mouse_gene_pair1.columns if \"Interaction ID\" in col][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "51b84e7e-7226-4d5f-ae8f-1cfbd8880c5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_ligand_col = [col for col in mouse_gene_pair1.columns if \"Human Ligand\" in col][0]\n",
    "ligand_col = [col for col in mouse_gene_pair1.columns if \"Ligand\" in col][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "ed8a6f42-116f-4529-8a0a-ace43c906c09",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_receptor_col = [col for col in mouse_gene_pair1.columns if \"Human Receptor\" in col][0]\n",
    "receptor_col = [col for col in mouse_gene_pair1.columns if \"Receptor\" in col][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "31170dcf-f34b-41bd-9dbf-a942074ab61d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        75.0\n",
       "1        75.0\n",
       "2       100.0\n",
       "3       100.0\n",
       "4       100.0\n",
       "        ...  \n",
       "3956      NaN\n",
       "3957      NaN\n",
       "3958      NaN\n",
       "3959      NaN\n",
       "3960      NaN\n",
       "Name: Ligand GOC score, Length: 3961, dtype: float64"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GOC_col = [col for col in mouse_gene_pair1.columns if \"GOC\" in col][0]\n",
    "mouse_gene_pair1[GOC_col]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "0afbe13c-c386-4da9-a113-d290f4f39d2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "hgnc_df = pd.read_csv(\"data/HGNC_gene_info_full.tsv\", sep=\"\\t\", dtype=str)\n",
    "hgnc_df = hgnc_df.dropna(subset=[\"uniprot_ids\", \"ensembl_gene_id\"])\n",
    "# Split uniprot_ids on comma and explode\n",
    "hgnc_exploded = hgnc_df.assign(uniprot_id=hgnc_df[\"uniprot_ids\"].str.split(\",\")).explode(\"uniprot_id\")\n",
    "hgnc_exploded[\"uniprot_id\"] = hgnc_exploded[\"uniprot_id\"].str.strip()\n",
    "\n",
    "uniprot_to_ensembl = hgnc_exploded.set_index(\"uniprot_id\")[\"ensembl_gene_id\"].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "6619c6ea-db59-49fd-8006-c7c1b5b72bef",
   "metadata": {},
   "outputs": [],
   "source": [
    "hgnc_df = pd.read_csv(\"data/HGNC_gene_info_full.tsv\", sep=\"\\t\", dtype=str)\n",
    "hgnc_df = hgnc_df.dropna(subset=[\"uniprot_ids\", \"ensembl_gene_id\"])\n",
    "# Split uniprot_ids on comma and explode\n",
    "hgnc_exploded = hgnc_df.assign(uniprot_id=hgnc_df[\"uniprot_ids\"].str.split(\",\")).explode(\"uniprot_id\")\n",
    "hgnc_exploded[\"uniprot_id\"] = hgnc_exploded[\"uniprot_id\"].str.strip()\n",
    "\n",
    "uniprot_to_ensembl = hgnc_exploded.set_index(\"uniprot_id\")[\"ensembl_gene_id\"].to_dict()\n",
    "\n",
    "# Left join on human_protein\n",
    "df_merged = df_orthologs.merge(\n",
    "    hgnc_exploded[[\"uniprot_id\", \"ensembl_gene_id\", \"symbol\"]],\n",
    "    left_on=\"human_protein\",\n",
    "    right_on=\"uniprot_id\",\n",
    "    how=\"left\"\n",
    ")\n",
    "\n",
    "# Optionally rename\n",
    "df_merged = df_merged.rename(columns={\n",
    "    \"symbol\": \"human_gene\",\n",
    "    \"ensembl_gene_id\": \"human_ensembl_gene_id\"\n",
    "}).drop(columns=[\"uniprot_id\"])\n",
    "df_merged = df_merged.dropna(subset=[\"human_ensembl_gene_id\"])\n",
    "df_merged.to_csv(\"data/mmusculus_inParanoid_uniProt_withHGNCAnn.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "390c3405-22ff-4d77-b0a0-aa54f50c222c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>human_protein</th>\n",
       "      <th>human_inparalog_score</th>\n",
       "      <th>human_seed_score</th>\n",
       "      <th>mouse_protein</th>\n",
       "      <th>mouse_inparalog_score</th>\n",
       "      <th>mouse_seed_score</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>human_ensembl_gene_id</th>\n",
       "      <th>human_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Q8WZ42</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A2ASS6</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>60090.0</td>\n",
       "      <td>ENSG00000155657</td>\n",
       "      <td>TTN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>Q8NF91</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q6ZWR6</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>14503.0</td>\n",
       "      <td>ENSG00000131018</td>\n",
       "      <td>SYNE1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>Q5VST9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A2AAJ9</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12156.0</td>\n",
       "      <td>ENSG00000154358</td>\n",
       "      <td>OBSCN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>Q03001</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q91ZU6</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10549.0</td>\n",
       "      <td>ENSG00000151914</td>\n",
       "      <td>DST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>Q8WXG9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q8VHN7</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10049.0</td>\n",
       "      <td>ENSG00000164199</td>\n",
       "      <td>ADGRV1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20946</th>\n",
       "      <td>17091</td>\n",
       "      <td>A0A1B0GTK5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A0A1B0GSI2</td>\n",
       "      <td>0.212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>46.0</td>\n",
       "      <td>ENSG00000225396</td>\n",
       "      <td>FAM236D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20947</th>\n",
       "      <td>17091</td>\n",
       "      <td>A0A1B0GTK5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A0A1B0GSB3</td>\n",
       "      <td>0.178</td>\n",
       "      <td>NaN</td>\n",
       "      <td>46.0</td>\n",
       "      <td>ENSG00000225396</td>\n",
       "      <td>FAM236D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20948</th>\n",
       "      <td>17092</td>\n",
       "      <td>Q96LM9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>E9Q1X6</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>ENSG00000125975</td>\n",
       "      <td>C20orf173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20952</th>\n",
       "      <td>17096</td>\n",
       "      <td>P0DP42</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A0A494B9K2</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>ENSG00000244219</td>\n",
       "      <td>TMEM225B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20954</th>\n",
       "      <td>17098</td>\n",
       "      <td>P31358</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Q64389</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>ENSG00000169442</td>\n",
       "      <td>CD52</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20402 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       cluster_id human_protein  human_inparalog_score  human_seed_score  \\\n",
       "0               1        Q8WZ42                    1.0               1.0   \n",
       "1               2        Q8NF91                    1.0               1.0   \n",
       "2               3        Q5VST9                    1.0               1.0   \n",
       "4               5        Q03001                    1.0               1.0   \n",
       "5               6        Q8WXG9                    1.0               1.0   \n",
       "...           ...           ...                    ...               ...   \n",
       "20946       17091    A0A1B0GTK5                    1.0               1.0   \n",
       "20947       17091    A0A1B0GTK5                    1.0               1.0   \n",
       "20948       17092        Q96LM9                    1.0               1.0   \n",
       "20952       17096        P0DP42                    1.0               1.0   \n",
       "20954       17098        P31358                    1.0               1.0   \n",
       "\n",
       "      mouse_protein  mouse_inparalog_score  mouse_seed_score  bitscore  \\\n",
       "0            A2ASS6                  1.000               1.0   60090.0   \n",
       "1            Q6ZWR6                  1.000               1.0   14503.0   \n",
       "2            A2AAJ9                  1.000               1.0   12156.0   \n",
       "4            Q91ZU6                  1.000               1.0   10549.0   \n",
       "5            Q8VHN7                  1.000               1.0   10049.0   \n",
       "...             ...                    ...               ...       ...   \n",
       "20946    A0A1B0GSI2                  0.212               NaN      46.0   \n",
       "20947    A0A1B0GSB3                  0.178               NaN      46.0   \n",
       "20948        E9Q1X6                  1.000               1.0      46.0   \n",
       "20952    A0A494B9K2                  1.000               1.0      41.0   \n",
       "20954        Q64389                  1.000               1.0      40.0   \n",
       "\n",
       "      human_ensembl_gene_id human_gene  \n",
       "0           ENSG00000155657        TTN  \n",
       "1           ENSG00000131018      SYNE1  \n",
       "2           ENSG00000154358      OBSCN  \n",
       "4           ENSG00000151914        DST  \n",
       "5           ENSG00000164199     ADGRV1  \n",
       "...                     ...        ...  \n",
       "20946       ENSG00000225396    FAM236D  \n",
       "20947       ENSG00000225396    FAM236D  \n",
       "20948       ENSG00000125975  C20orf173  \n",
       "20952       ENSG00000244219   TMEM225B  \n",
       "20954       ENSG00000169442       CD52  \n",
       "\n",
       "[20402 rows x 10 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "b19b9eac-3430-4863-ad89-68e378666524",
   "metadata": {},
   "outputs": [],
   "source": [
    "def summarize_orthologs(human_col, species_col, label,\n",
    "                        confidence_orth_col=None, confidence_orth_threshold=None,\n",
    "                        GOC_col=None, GOC_threshold=None):\n",
    "    df = mouse_gene_pair1.copy()\n",
    "\n",
    "    if confidence_orth_col and confidence_orth_threshold is not None:\n",
    "        df = df[df[confidence_orth_col] == confidence_orth_threshold]\n",
    "\n",
    "    if GOC_col and GOC_threshold is not None:\n",
    "        df = df[df[GOC_col] >= GOC_threshold]  # Use >= instead of ==\n",
    "\n",
    "    unique_pairs = df[[human_col, species_col]].drop_duplicates()\n",
    "\n",
    "    counts = (\n",
    "        unique_pairs\n",
    "        .groupby(human_col)[species_col]\n",
    "        .count()\n",
    "        .sort_values(ascending=False)\n",
    "        .reset_index(name='count')\n",
    "    )\n",
    "\n",
    "    filter_tag = label.lower()\n",
    "    if confidence_orth_threshold is not None:\n",
    "        filter_tag += f\"_conf{confidence_orth_threshold}\"\n",
    "    if GOC_threshold is not None:\n",
    "        filter_tag += f\"_GOCge{GOC_threshold}\"\n",
    "\n",
    "    counts.to_csv(f\"data/human_mouse_orth_count_{filter_tag}.csv\", index=False)\n",
    "\n",
    "    summary_counts = counts['count'].value_counts().sort_index()\n",
    "    total_human_genes = counts.shape[0]\n",
    "\n",
    "    summary_lines = [\n",
    "        f\"Out of {total_human_genes} unique human {label.lower()} genes \"\n",
    "        f\"(Orthology Confidence = {confidence_orth_threshold}, GOC ≥ {GOC_threshold}):\"\n",
    "    ]\n",
    "    for orth_count, gene_count in summary_counts.items():\n",
    "        summary_lines.append(\n",
    "            f\" - {gene_count} human {label.lower()} genes had {orth_count} mouse ortholog(s)\"\n",
    "        )\n",
    "\n",
    "    return \"\\n\".join(summary_lines)\n",
    "\n",
    "\n",
    "# Detect columns\n",
    "confidence_orth_ligand = [col for col in mouse_gene_pair1.columns if \"Ligand Orthology Confidence\" in col][0]\n",
    "GOC_col_ligand = [col for col in mouse_gene_pair1.columns if \"Ligand GOC\" in col][0]\n",
    "\n",
    "confidence_orth_receptor = [col for col in mouse_gene_pair1.columns if \"Receptor Orthology Confidence\" in col][0]\n",
    "GOC_col_receptor = [col for col in mouse_gene_pair1.columns if \"Receptor GOC\" in col][0]\n",
    "\n",
    "# Generate summaries\n",
    "ligand_summary = summarize_orthologs(\n",
    "    human_ligand_col, ligand_col, \"Ligand\",\n",
    "    confidence_orth_col=confidence_orth_ligand, confidence_orth_threshold=None,\n",
    "    GOC_col=GOC_col_ligand, GOC_threshold=0\n",
    ")\n",
    "\n",
    "receptor_summary = summarize_orthologs(\n",
    "    human_receptor_col, receptor_col, \"Receptor\",\n",
    "    confidence_orth_col=confidence_orth_receptor, confidence_orth_threshold=None,\n",
    "    GOC_col=GOC_col_receptor, GOC_threshold=0\n",
    ")\n",
    "\n",
    "# Print\n",
    "print(ligand_summary)\n",
    "print()\n",
    "print(receptor_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "b3c7c890-4529-4234-a3bc-534e7c4c85a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        75.0\n",
       "1        75.0\n",
       "2       100.0\n",
       "3       100.0\n",
       "4       100.0\n",
       "        ...  \n",
       "3956      NaN\n",
       "3957      NaN\n",
       "3958      NaN\n",
       "3959      NaN\n",
       "3960      NaN\n",
       "Name: Ligand GOC score, Length: 3961, dtype: float64"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse_gene_pair1[GOC_col_ligand]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "d5599a62-b278-4bf2-8f60-2c4ce2b3e3f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Out of 985 unique human ligand genes (Orthology Confidence = None, GOC ≥ 0):\n",
      " - 925 human ligand genes had 1 mouse ortholog(s)\n",
      " - 31 human ligand genes had 2 mouse ortholog(s)\n",
      " - 5 human ligand genes had 3 mouse ortholog(s)\n",
      " - 2 human ligand genes had 4 mouse ortholog(s)\n",
      " - 5 human ligand genes had 5 mouse ortholog(s)\n",
      " - 3 human ligand genes had 6 mouse ortholog(s)\n",
      " - 2 human ligand genes had 7 mouse ortholog(s)\n",
      " - 12 human ligand genes had 14 mouse ortholog(s)\n",
      "\n",
      "Out of 780 unique human receptor genes (Orthology Confidence = None, GOC ≥ 0):\n",
      " - 742 human receptor genes had 1 mouse ortholog(s)\n",
      " - 15 human receptor genes had 2 mouse ortholog(s)\n",
      " - 7 human receptor genes had 3 mouse ortholog(s)\n",
      " - 3 human receptor genes had 4 mouse ortholog(s)\n",
      " - 3 human receptor genes had 5 mouse ortholog(s)\n",
      " - 4 human receptor genes had 6 mouse ortholog(s)\n",
      " - 6 human receptor genes had 7 mouse ortholog(s)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Detect columns\n",
    "confidence_orth_ligand = [col for col in mouse_gene_pair1.columns if \"Ligand Orthology Confidence\" in col][0]\n",
    "GOC_col_ligand = [col for col in mouse_gene_pair1.columns if \"Ligand GOC\" in col][0]\n",
    "\n",
    "confidence_orth_receptor = [col for col in mouse_gene_pair1.columns if \"Receptor Orthology Confidence\" in col][0]\n",
    "GOC_col_receptor = [col for col in mouse_gene_pair1.columns if \"Receptor GOC\" in col][0]\n",
    "\n",
    "# Generate summaries\n",
    "ligand_summary = summarize_orthologs(\n",
    "    human_ligand_col, ligand_col, \"Ligand\",\n",
    "    confidence_orth_col=confidence_orth_ligand, confidence_orth_threshold=None,\n",
    "    GOC_col=GOC_col_ligand, GOC_threshold=0\n",
    ")\n",
    "\n",
    "receptor_summary = summarize_orthologs(\n",
    "    human_receptor_col, receptor_col, \"Receptor\",\n",
    "    confidence_orth_col=confidence_orth_receptor, confidence_orth_threshold=None,\n",
    "    GOC_col=GOC_col_receptor, GOC_threshold=0\n",
    ")\n",
    "\n",
    "# Print\n",
    "print(ligand_summary)\n",
    "print()\n",
    "print(receptor_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "780f674a-a73e-46b8-ad8a-c72b77c7ec4b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "781"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(pd.unique(mouse_gene_pair1[human_receptor_col]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36c2d906-7c1b-433a-91a7-f963fcdf9292",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "quarto_llm_env",
   "language": "python",
   "name": "quarto_llm_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
