{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fcc5ddc1-519e-439c-a130-3dd8bcd181bd",
   "metadata": {},
   "source": [
    "# Python Notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d4b960ac-0404-4f97-b786-01c850c43e7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to prepare datatables (cleaning and hyperlinking, adding tool tips, etc) input for the database qmds\n",
    "import sys, os\n",
    "import pandas as pd\n",
    "# Change working directory to ConnectomeDB\n",
    "project_root = \"/Users/sakuramaezono/Library/CloudStorage/OneDrive-YokohamaCityUniversity/Personal/05_Python_repositories/ConnectomeDB\"\n",
    "os.chdir(project_root)\n",
    "sys.path.append(os.path.abspath(\"src\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9c34eae7-a7ec-46bc-885a-af5bac416c01",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to prepare datatables (cleaning and hyperlinking, adding tool tips, etc) input for the database qmds\n",
    "import sys, os\n",
    "from itables import init_notebook_mode\n",
    "import pandas as pd\n",
    "from itables import show\n",
    "from itables import options\n",
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "import fetchGSheet \n",
    "import warnings\n",
    "import urllib.parse\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "a5fbc713-2218-4c57-af1f-67a73dd11e36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Interaction ID</th>\n",
       "      <th>Human LR Pair</th>\n",
       "      <th>Ligand Name</th>\n",
       "      <th>Receptor Name</th>\n",
       "      <th>Ligand MGI ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3444</th>\n",
       "      <td>CDB03445</td>\n",
       "      <td>Pcdhb22 Pcdhb22</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3445</th>\n",
       "      <td>CDB03446</td>\n",
       "      <td>Pcdhgb8 Pcdhgb8</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3446</th>\n",
       "      <td>CDB03447</td>\n",
       "      <td>Saa3 Ly96</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3447</th>\n",
       "      <td>CDB03448</td>\n",
       "      <td>Saa3 Tlr4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3448</th>\n",
       "      <td>CDB03449</td>\n",
       "      <td>Sema4a Timd2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Interaction ID    Human LR Pair Ligand Name Receptor Name  \\\n",
       "3444       CDB03445  Pcdhb22 Pcdhb22                             \n",
       "3445       CDB03446  Pcdhgb8 Pcdhgb8                             \n",
       "3446       CDB03447        Saa3 Ly96                             \n",
       "3447       CDB03448        Saa3 Tlr4                             \n",
       "3448       CDB03449     Sema4a Timd2                             \n",
       "\n",
       "                                          Ligand MGI ID  \n",
       "3444  <a href=\"https://www.informatics.jax.org/marke...  \n",
       "3445  <a href=\"https://www.informatics.jax.org/marke...  \n",
       "3446  <a href=\"https://www.informatics.jax.org/marke...  \n",
       "3447  <a href=\"https://www.informatics.jax.org/marke...  \n",
       "3448  <a href=\"https://www.informatics.jax.org/marke...  "
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair0_copy[['Interaction ID', 'Human LR Pair', 'Ligand Name', 'Receptor Name', 'Ligand MGI ID']].tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "744372a8-0844-4f4f-919b-b71eb470588d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Interaction ID</th>\n",
       "      <th>Human LR Pair</th>\n",
       "      <th>Ligand Name</th>\n",
       "      <th>Receptor Name</th>\n",
       "      <th>Ligand MGI ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3444</th>\n",
       "      <td>CDB03445</td>\n",
       "      <td>Pcdhb22 Pcdhb22</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MGI:2136760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3445</th>\n",
       "      <td>CDB03446</td>\n",
       "      <td>Pcdhgb8 Pcdhgb8</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MGI:1935200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3446</th>\n",
       "      <td>CDB03447</td>\n",
       "      <td>Saa3 Ly96</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MGI:98223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3447</th>\n",
       "      <td>CDB03448</td>\n",
       "      <td>Saa3 Tlr4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MGI:98223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3448</th>\n",
       "      <td>CDB03449</td>\n",
       "      <td>Sema4a Timd2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MGI:107560</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Interaction ID    Human LR Pair Ligand Name Receptor Name Ligand MGI ID\n",
       "3444       CDB03445  Pcdhb22 Pcdhb22                             MGI:2136760\n",
       "3445       CDB03446  Pcdhgb8 Pcdhgb8                             MGI:1935200\n",
       "3446       CDB03447        Saa3 Ly96                               MGI:98223\n",
       "3447       CDB03448        Saa3 Tlr4                               MGI:98223\n",
       "3448       CDB03449     Sema4a Timd2                              MGI:107560"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def extract_mgi_id(col):\n",
    "    \"\"\"Use regular expression to extract the HGNC ID after 'HGNC:'.\"\"\"\n",
    "    match = re.search(r'MGI:(\\d+)', col)\n",
    "    if match:\n",
    "        return 'MGI:' +str(match.group(1))\n",
    "    return None\n",
    "gene_pair0_copy['Ligand MGI ID'] = gene_pair0_copy['Ligand MGI ID'].apply(extract_mgi_id)\n",
    "gene_pair0_copy[['Interaction ID', 'Human LR Pair', 'Ligand Name', 'Receptor Name', 'Ligand MGI ID']].tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "48ecf005-2300-4844-bae3-ace7d53da84b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['MGI:107560', 'MGI:2136754', 'MGI:2136756', 'MGI:2136757',\n",
       "       'MGI:2136758', 'MGI:2136759', 'MGI:2136760', 'MGI:1935200',\n",
       "       'MGI:1306817', 'MGI:3649078', 'MGI:3774845', 'MGI:98223',\n",
       "       'MGI:2159681', 'MGI:1196250', 'MGI:96824'], dtype=object)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Function to convert Mouse (ENSEMBL) to RGD ID via biomart\n",
    "import os\n",
    "import sys\n",
    "from pybiomart import Server\n",
    "import pandas as pd\n",
    "import createDataTable \n",
    "# Full list of unique MGI IDs\n",
    "mgi_ids = createDataTable.mouse_specific_mgi_ids\n",
    "mgi_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0784de1c-15a8-4de0-bb8e-935e743e2e7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Apply the mapping to 'Ligand Name'\n",
    "gene_pair0_copy['Ligand Name'] = gene_pair0_copy.apply(\n",
    "    lambda row: mapping_mouse_name.get(row['Ligand MGI ID'], row['Ligand Name'])\n",
    "    if pd.isna(row['Ligand Name']) or str(row['Ligand Name']).strip() == '' else row['Ligand Name'],\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e1b4c4ac-c415-403c-adaf-e4a665a20167",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Interaction ID</th>\n",
       "      <th>Human LR Pair</th>\n",
       "      <th>Ligand Name</th>\n",
       "      <th>Receptor Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3435</th>\n",
       "      <td>CDB03436</td>\n",
       "      <td>ZP3 CHRNA7</td>\n",
       "      <td>zona pellucida glycoprotein 3</td>\n",
       "      <td>cholinergic receptor nicotinic alpha 7 subunit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3436</th>\n",
       "      <td>CDB03437</td>\n",
       "      <td>H60a Klrk1</td>\n",
       "      <td></td>\n",
       "      <td>killer cell lectin-like receptor subfamily K, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3437</th>\n",
       "      <td>CDB03438</td>\n",
       "      <td>H60b Klrk1</td>\n",
       "      <td>histocompatibility 60b</td>\n",
       "      <td>killer cell lectin-like receptor subfamily K, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3438</th>\n",
       "      <td>CDB03439</td>\n",
       "      <td>H60c Klrk1</td>\n",
       "      <td>histocompatibility 60c</td>\n",
       "      <td>killer cell lectin-like receptor subfamily K, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3439</th>\n",
       "      <td>CDB03440</td>\n",
       "      <td>Pcdhb17 Pcdhb17</td>\n",
       "      <td>protocadherin beta 17</td>\n",
       "      <td>protocadherin beta 17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3440</th>\n",
       "      <td>CDB03441</td>\n",
       "      <td>Pcdhb18 Pcdhb18</td>\n",
       "      <td>protocadherin beta 18</td>\n",
       "      <td>protocadherin beta 18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3441</th>\n",
       "      <td>CDB03442</td>\n",
       "      <td>Pcdhb19 Pcdhb19</td>\n",
       "      <td>protocadherin beta 19</td>\n",
       "      <td>protocadherin beta 19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3442</th>\n",
       "      <td>CDB03443</td>\n",
       "      <td>Pcdhb20 Pcdhb20</td>\n",
       "      <td>protocadherin beta 20</td>\n",
       "      <td>protocadherin beta 20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3443</th>\n",
       "      <td>CDB03444</td>\n",
       "      <td>Pcdhb21 Pcdhb21</td>\n",
       "      <td>protocadherin beta 21</td>\n",
       "      <td>protocadherin beta 21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3444</th>\n",
       "      <td>CDB03445</td>\n",
       "      <td>Pcdhb22 Pcdhb22</td>\n",
       "      <td>protocadherin beta 22</td>\n",
       "      <td>protocadherin beta 22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3445</th>\n",
       "      <td>CDB03446</td>\n",
       "      <td>Pcdhgb8 Pcdhgb8</td>\n",
       "      <td>protocadherin gamma subfamily B, 8</td>\n",
       "      <td>protocadherin gamma subfamily B, 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3446</th>\n",
       "      <td>CDB03447</td>\n",
       "      <td>Saa3 Tlr4</td>\n",
       "      <td>serum amyloid A 3</td>\n",
       "      <td>toll-like receptor 4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3447</th>\n",
       "      <td>CDB03448</td>\n",
       "      <td>Sema4a Timd2</td>\n",
       "      <td>sema domain, immunoglobulin domain (Ig), trans...</td>\n",
       "      <td>T cell immunoglobulin and mucin domain contain...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Interaction ID    Human LR Pair  \\\n",
       "3435       CDB03436       ZP3 CHRNA7   \n",
       "3436       CDB03437       H60a Klrk1   \n",
       "3437       CDB03438       H60b Klrk1   \n",
       "3438       CDB03439       H60c Klrk1   \n",
       "3439       CDB03440  Pcdhb17 Pcdhb17   \n",
       "3440       CDB03441  Pcdhb18 Pcdhb18   \n",
       "3441       CDB03442  Pcdhb19 Pcdhb19   \n",
       "3442       CDB03443  Pcdhb20 Pcdhb20   \n",
       "3443       CDB03444  Pcdhb21 Pcdhb21   \n",
       "3444       CDB03445  Pcdhb22 Pcdhb22   \n",
       "3445       CDB03446  Pcdhgb8 Pcdhgb8   \n",
       "3446       CDB03447        Saa3 Tlr4   \n",
       "3447       CDB03448     Sema4a Timd2   \n",
       "\n",
       "                                            Ligand Name  \\\n",
       "3435                      zona pellucida glycoprotein 3   \n",
       "3436                                                      \n",
       "3437                             histocompatibility 60b   \n",
       "3438                             histocompatibility 60c   \n",
       "3439                              protocadherin beta 17   \n",
       "3440                              protocadherin beta 18   \n",
       "3441                              protocadherin beta 19   \n",
       "3442                              protocadherin beta 20   \n",
       "3443                              protocadherin beta 21   \n",
       "3444                              protocadherin beta 22   \n",
       "3445                 protocadherin gamma subfamily B, 8   \n",
       "3446                                  serum amyloid A 3   \n",
       "3447  sema domain, immunoglobulin domain (Ig), trans...   \n",
       "\n",
       "                                          Receptor Name  \n",
       "3435     cholinergic receptor nicotinic alpha 7 subunit  \n",
       "3436  killer cell lectin-like receptor subfamily K, ...  \n",
       "3437  killer cell lectin-like receptor subfamily K, ...  \n",
       "3438  killer cell lectin-like receptor subfamily K, ...  \n",
       "3439                              protocadherin beta 17  \n",
       "3440                              protocadherin beta 18  \n",
       "3441                              protocadherin beta 19  \n",
       "3442                              protocadherin beta 20  \n",
       "3443                              protocadherin beta 21  \n",
       "3444                              protocadherin beta 22  \n",
       "3445                 protocadherin gamma subfamily B, 8  \n",
       "3446                               toll-like receptor 4  \n",
       "3447  T cell immunoglobulin and mucin domain contain...  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair0_copy[['Interaction ID', 'Human LR Pair', 'Ligand Name', 'Receptor Name']].tail(13)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6e5cc3ae-4e62-48b6-815d-9929d3846d11",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to prepare datatables (cleaning and hyperlinking, adding tool tips, etc) input for the database qmds\n",
    "import sys, os\n",
    "from itables import init_notebook_mode\n",
    "import pandas as pd\n",
    "from itables import show\n",
    "from itables import options\n",
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "import fetchGSheet \n",
    "import warnings\n",
    "import urllib.parse\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e4eec98c-0398-4dfd-8b31-5b8131a2ff89",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/sakuramaezono/Library/CloudStorage/OneDrive-YokohamaCityUniversity/Personal/05_Python_repositories/ConnectomeDB/src/createDataTable.py:28: DtypeWarning: Columns (32,34,38,40,50) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  pop_up_info = pd.read_table(\"data/HGNC_gene_info_full.tsv\")\n"
     ]
    }
   ],
   "source": [
    "## Function to convert Mouse (ENSEMBL) to RGD ID via biomart\n",
    "import os\n",
    "import sys\n",
    "from pybiomart import Server\n",
    "import pandas as pd\n",
    "sys.path.append(os.path.abspath(\"src\"))\n",
    "from createDataTable import mouse_specific_mgi_ids\n",
    "# Full list of unique MGI IDs\n",
    "mgi_ids = mouse_specific_mgi_ids\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "259b8961-69e8-4875-95dc-7a1b3901a4be",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to convert Mouse (ENSEMBL) to RGD ID via biomart\n",
    "import os\n",
    "import sys\n",
    "from pybiomart import Server\n",
    "import pandas as pd\n",
    "sys.path.append(os.path.abspath(\"src\"))\n",
    "from createDataTable import mouse_specific_mgi_ids\n",
    "# Full list of unique MGI IDs\n",
    "mgi_ids = mouse_specific_mgi_ids\n",
    "\n",
    "server = Server(host='http://www.ensembl.org')\n",
    "\n",
    "mouse_dataset = server.marts['ENSEMBL_MART_ENSEMBL'].datasets['mmusculus_gene_ensembl']\n",
    "rat_dataset = server.marts['ENSEMBL_MART_ENSEMBL'].datasets['rnorvegicus_gene_ensembl']\n",
    "\n",
    "# Get mouse Ensembl IDs and rat ortholog Ensembl IDs\n",
    "mouse_genes = mouse_dataset.query(\n",
    "    attributes=[\n",
    "        'ensembl_gene_id',\n",
    "        'mgi_id',\n",
    "        'mgi_symbol',\n",
    "        'mgi_description'\n",
    "    ]\n",
    ")\n",
    "mouse_genes.rename(columns={\n",
    "    'MGI description': 'MGI name'})\n",
    "mouse_genes = mouse_genes[mouse_genes[\"MGI ID\"].isin(mgi_ids)]\n",
    "mouse_genes.to_csv(\"data/mouse_name_mapping.csv\", index=False)\n",
    "# Step 2: Get mouse Ensembl genes with rat ortholog Ensembl gene IDs (no mouse MGI here)\n",
    "mouse_rat_orthologs = mouse_dataset.query(\n",
    "    attributes=[\n",
    "        'ensembl_gene_id',\n",
    "        'rnorvegicus_homolog_ensembl_gene',\n",
    "        'rnorvegicus_homolog_associated_gene_name'\n",
    "    ]\n",
    ")\n",
    "\n",
    "# Step 3: Merge mouse gene info with orthologs on ensembl_gene_id\n",
    "mouse_to_rat = pd.merge(mouse_genes, mouse_rat_orthologs, on='Gene stable ID')\n",
    "# Filter out entries without rat ortholog\n",
    "mouse_to_rat = mouse_to_rat[mouse_to_rat['Norway rat - BN/NHsdMcwi gene stable ID'].notna() & (mouse_to_rat['Norway rat - BN/NHsdMcwi gene stable ID'] != '')]\n",
    "\n",
    "# Step 4: Retrieve RGD IDs for rat ortholog Ensembl gene IDs\n",
    "rat_ensembl_ids = mouse_to_rat['Norway rat - BN/NHsdMcwi gene stable ID'].unique().tolist()\n",
    "rat_ensembl_to_rgd = rat_dataset.query(\n",
    "    attributes=['ensembl_gene_id', 'rgd_id', 'description']\n",
    ")\n",
    "rat_ensembl_to_rgd = rat_ensembl_to_rgd[rat_ensembl_to_rgd[\"Gene stable ID\"].isin(rat_ensembl_ids)]\n",
    "# Step 5: Merge rat RGD IDs back\n",
    "final_mapping = pd.merge(\n",
    "    mouse_to_rat,\n",
    "    rat_ensembl_to_rgd,\n",
    "    left_on='Norway rat - BN/NHsdMcwi gene stable ID',\n",
    "    right_on='Gene stable ID',\n",
    "    how='left'\n",
    ")\n",
    "# Clean column names for clarity\n",
    "final_mapping.rename(columns={\n",
    "    'Gene stable ID_x': 'mouse_ensembl_gene_id',\n",
    "    'Gene stable ID_y': 'rat_ensembl_gene_id',\n",
    "    'Norway rat - BN/NHsdMcwi gene name': 'RGD symbol',\n",
    "    'Gene description': 'RGD name',\n",
    "    'rgd_id': 'RGD ID'\n",
    "}, inplace=True)\n",
    "\n",
    "final_mapping = final_mapping.drop(columns=[\"Norway rat - BN/NHsdMcwi gene stable ID\"])\n",
    "final_mapping['RGD ID'] = final_mapping['RGD ID'].apply(lambda x: f\"RGD:{int(x)}\" if pd.notnull(x) else None)\n",
    "# Save\n",
    "final_mapping.to_csv(\"data/mouse_to_rat_mapping.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "bfcabd2a-1193-484c-9975-fc666efefcaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "rat_dataset.list_attributes().to_csv(\"data/rat_attributes.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0e9f2a10-1e7b-4eb3-bb77-765e0ad6d40e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import jinja2\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "import base64\n",
    "import re\n",
    "\n",
    "# Add the src directory to the path for importing modules\n",
    "sys.path.append(os.path.abspath(\"src\"))\n",
    "\n",
    "# Import necessary modules from your existing src files\n",
    "# Ensure createDataTable and createFunctionalAnnotTable are in your 'src' directory\n",
    "from createDataTable import pop_up_info, gene_pair0, generate_perplexity_links, gene_pair00, is_mouse_specific\n",
    "from createFunctionalAnnotTable import gene_pair_annot_ligand, gene_pair_annot_receptor\n",
    "\n",
    "# Test or all\n",
    "test = True\n",
    "test_genes = [\"H60a Klrk1\", \"H60b Klrk1\", \"VEGFA NRP1\", \"THPO MPL\", \"FGF1 FGFR3\"] # Example genes\n",
    "# --- Paths ---\n",
    "MERGED_TEMPLATE_PATH = 'HTML/mergedCard_tabs.html'\n",
    "OUTPUT_DIR = 'data/cards/' # New output directory for combined files\n",
    "\n",
    "# function for replacing visible text:\n",
    "def update_link_text_with_symbol(html_str, new_symbol):\n",
    "    \"\"\"\n",
    "    Replace the visible text in an anchor tag with the provided symbol.\n",
    "    The HGNC ID is extracted from the href and left unchanged.\n",
    "    \"\"\"\n",
    "    # Only proceed if input is valid\n",
    "    if not isinstance(html_str, str) or not html_str.strip():\n",
    "        return html_str\n",
    "    \n",
    "    return re.sub(r\">(.*?)</a>\", f\">{new_symbol}</a>\", html_str) #<i class='fa-solid fa-arrow-up-right-from-square' style='margin-left:4px;'></i></a>\n",
    "\n",
    "# Create Perplexity link\n",
    "\n",
    "# Recreate Perplexity link\n",
    "def create_url_basic(symbol):\n",
    "          label = \"Perplexity (LLM)\"\n",
    "          query = f\"What diseases is {symbol} implicated in?\"\n",
    "          encoded_query = query.replace(\" \", \"%20\")\n",
    "          output = f'<a href=\"https://www.perplexity.ai/search?q={encoded_query}\" target=\"_blank\">{label}</a>' #<i class=\"fa-solid fa-arrow-up-right-from-square\" style=\"margin-left:4px;\"></i>\n",
    "          return output\n",
    "\n",
    "# --- Load and Preprocess Data (Combined from both scripts) ---\n",
    "\n",
    "# Load PubMed data (from createPMIDpages.py)\n",
    "pubmed_data = pd.read_csv(\"data/pubmed_results.csv\")\n",
    "pubmed_data[\"Year\"] = pubmed_data[\"Year\"].astype(str).str.replace(\".0\", \"\", regex=False).astype(int)\n",
    "pubmed_data[\"PMID\"] = pubmed_data[\"PMID\"].astype(str)\n",
    "pubmed_data = pubmed_data.reset_index(drop=True)\n",
    "\n",
    "# Load LLM results (from createPMIDpages.py)\n",
    "bio_keywords = pd.read_csv(\"data/llm_results.csv\")\n",
    "\n",
    "# --- Prepare gene_pair00 for PMID section (from createPMIDpages.py) ---\n",
    "# gene_pair00 is used for PMID and Keywords, so it needs the '—' placeholder\n",
    "# Ensure gene_pair00 is a copy to avoid SettingWithCopyWarning later\n",
    "gene_pair00_copy = gene_pair00.copy()\n",
    "gene_pair00_copy[\"Human LR Pair\"] = gene_pair00_copy[\"Human LR Pair\"].str.replace(\" \", \"—\")\n",
    "\n",
    "# Merge with LLM results\n",
    "gene_pair000 = gene_pair00_copy.merge(bio_keywords, how='left', left_on=\"Human LR Pair\", right_on='Human LR Pair')\n",
    "gene_pair000[\"Relevance Keywords\"] = gene_pair000[\"Relevance Keywords\"].astype(str)\n",
    "gene_pair000[\"Human LR Pair\"] = gene_pair000[\"Human LR Pair\"].astype(str) # Ensure string type\n",
    "\n",
    "# --- Prepare gene_pair0 for Card section (from createCards.py) ---\n",
    "# gene_pair0 is used for card details, it should retain spaces for splitting gene names\n",
    "# Ensure gene_pair0 is a copy to avoid SettingWithCopyWarning later\n",
    "gene_pair0_copy = gene_pair0.copy()\n",
    "\n",
    "# grab the pairs by interaction id that has ligand that is mouse_specific\n",
    "mouse_interaction_ids = gene_pair0_copy[gene_pair0_copy['Ligand'].apply(is_mouse_specific)]['Interaction ID'].tolist()\n",
    "\n",
    "# Add Disease (specific) to cards\n",
    "df_disease = pd.read_csv(\"data/disease_annotations_per_pair.csv\")\n",
    "df_disease = df_disease.groupby('interaction')['disease'].apply(', '.join).reset_index()\n",
    "mapping_disease = dict(zip(df_disease['interaction'], df_disease['disease']))\n",
    "gene_pair0_copy[\"Disease\"] = gene_pair0_copy['Human LR Pair'].map(mapping_disease).fillna(\"unknown\")\n",
    "\n",
    "gene_pair0_copy = generate_perplexity_links(\n",
    "    gene_pair0_copy,\n",
    "    pathway_col=\"Disease\",\n",
    "    default_query_template=\"What-diseases-is-the-ligand-receptor-pair-{pair}-associated-with\"\n",
    ")\n",
    "\n",
    "# # Hide for now (linking to actual PMID database\n",
    "# gene_pair0_copy[\"Interaction ID\"] = gene_pair0_copy[\"Interaction ID\"].apply(\n",
    "#     lambda x: f\"<a href='https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database/filter/{x}.html'>{x}</a>\"\n",
    "# )\n",
    "\n",
    "# Add external link icon\n",
    "# icon_html = '<i class=\"fa-solid fa-arrow-up-right-from-square\" style=\"margin-left:4px;\"></i></a>'\n",
    "# columns_to_update = [\n",
    "#     \"KEGG Pathway\", \"PROGENy Pathway\", \"Cancer-related\",\n",
    "#     \"Disease Type\", \"Disease\"\n",
    "# ]\n",
    "# for col in columns_to_update:\n",
    "#     gene_pair0_copy[col] = gene_pair0_copy[col].str.replace(\n",
    "#         \"</a>\", icon_html, regex=False\n",
    "#     )\n",
    "\n",
    "# Add missing ligand name for mouse-specific\n",
    "mouse_rat_info = pd.read_csv(\"data/mouse_name_mapping.csv\")\n",
    "mapping_mouse_name = dict(zip(mouse_rat_info['MGI ID'], mouse_rat_info['MGI description']))\n",
    "mouse_rat_info = pd.read_csv(\"data/mouse_to_rat_mapping.csv\")\n",
    "mapping_mr2 = dict(zip(mouse_rat_info['RGD ID'], mouse_rat_info['RGD name']))\n",
    "\n",
    "\n",
    "# extract mgi\n",
    "def extract_mgi_id(col):\n",
    "    \"\"\"Use regular expression to extract the HGNC ID after 'HGNC:'.\"\"\"\n",
    "    match = re.search(r'MGI:(\\d+)', col)\n",
    "    if match:\n",
    "        return 'MGI:' +str(match.group(1))\n",
    "    return None\n",
    "\n",
    "gene_pair0_copy['Ligand MGI ID'] = gene_pair0_copy['Ligand MGI ID'].apply(extract_mgi_id)\n",
    "gene_pair0_copy['Receptor MGI ID'] = gene_pair0_copy['Receptor MGI ID'].apply(extract_mgi_id)\n",
    "\n",
    "# Apply the mapping to 'Ligand Name'\n",
    "gene_pair0_copy['Ligand Name'] = gene_pair0_copy.apply(\n",
    "    lambda row: mapping_mouse_name.get(row['Ligand MGI ID'], row['Ligand Name'])\n",
    "    if pd.isna(row['Ligand Name']) or str(row['Ligand Name']).strip() == '' else row['Ligand Name'],\n",
    "    axis=1\n",
    ")\n",
    "\n",
    " # Apply the mapping to 'Receptor Name'\n",
    "gene_pair0_copy['Receptor Name'] = gene_pair0_copy.apply(\n",
    "    lambda row: mapping_mouse_name.get(row['Receptor MGI ID'], row['Receptor Name'])\n",
    "    if pd.isna(row['Receptor Name']) or str(row['Receptor Name']).strip() == '' else row['Receptor Name'],\n",
    "    axis=1\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0f1e7b03-e617-4300-8ce4-a291d37e464d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Interaction ID</th>\n",
       "      <th>Ligand Name</th>\n",
       "      <th>Receptor Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDB00001</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>heat shock protein family A (Hsp70) member 5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDB00002</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>LDL receptor related protein 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDB00003</td>\n",
       "      <td>angiotensin I converting enzyme</td>\n",
       "      <td>bradykinin receptor B2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDB00004</td>\n",
       "      <td>adenosine deaminase</td>\n",
       "      <td>dipeptidyl peptidase 4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDB00005</td>\n",
       "      <td>ADAM metallopeptidase domain 10</td>\n",
       "      <td>EPH receptor A3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3443</th>\n",
       "      <td>CDB03444</td>\n",
       "      <td>protocadherin beta 21</td>\n",
       "      <td>protocadherin beta 21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3444</th>\n",
       "      <td>CDB03445</td>\n",
       "      <td>protocadherin beta 22</td>\n",
       "      <td>protocadherin beta 22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3445</th>\n",
       "      <td>CDB03446</td>\n",
       "      <td>protocadherin gamma subfamily B, 8</td>\n",
       "      <td>protocadherin gamma subfamily B, 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3446</th>\n",
       "      <td>CDB03447</td>\n",
       "      <td>serum amyloid A 3</td>\n",
       "      <td>toll-like receptor 4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3447</th>\n",
       "      <td>CDB03448</td>\n",
       "      <td>sema domain, immunoglobulin domain (Ig), trans...</td>\n",
       "      <td>T cell immunoglobulin and mucin domain contain...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3448 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Interaction ID                                        Ligand Name  \\\n",
       "0          CDB00001                              alpha-2-macroglobulin   \n",
       "1          CDB00002                              alpha-2-macroglobulin   \n",
       "2          CDB00003                    angiotensin I converting enzyme   \n",
       "3          CDB00004                                adenosine deaminase   \n",
       "4          CDB00005                    ADAM metallopeptidase domain 10   \n",
       "...             ...                                                ...   \n",
       "3443       CDB03444                              protocadherin beta 21   \n",
       "3444       CDB03445                              protocadherin beta 22   \n",
       "3445       CDB03446                 protocadherin gamma subfamily B, 8   \n",
       "3446       CDB03447                                  serum amyloid A 3   \n",
       "3447       CDB03448  sema domain, immunoglobulin domain (Ig), trans...   \n",
       "\n",
       "                                          Receptor Name  \n",
       "0          heat shock protein family A (Hsp70) member 5  \n",
       "1                        LDL receptor related protein 1  \n",
       "2                                bradykinin receptor B2  \n",
       "3                                dipeptidyl peptidase 4  \n",
       "4                                       EPH receptor A3  \n",
       "...                                                 ...  \n",
       "3443                              protocadherin beta 21  \n",
       "3444                              protocadherin beta 22  \n",
       "3445                 protocadherin gamma subfamily B, 8  \n",
       "3446                               toll-like receptor 4  \n",
       "3447  T cell immunoglobulin and mucin domain contain...  \n",
       "\n",
       "[3448 rows x 3 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair0_copy[[\"Interaction ID\", \"Ligand Name\", \"Receptor Name\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "a2c2dee9-f899-473b-a43d-8d082eae1a19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add some missing MGI since they were mouse-specific\n",
    "mouse_rat_info = pd.read_csv(\"data/mouse_to_rat_mapping.csv\")\n",
    "mapping = dict(zip(mouse_rat_info['MGI ID'], mouse_rat_info['RGD ID']))\n",
    "# Function to apply the mapping only if the MGI ID is empty\n",
    "def map_if_empty(row, gene_col, mgi_col, mapping_dict):\n",
    "    if pd.isna(row[mgi_col]) or row[mgi_col] == '':\n",
    "        return mapping_dict.get(row[gene_col], '') # Use .get() to avoid KeyError if symbol not in mapping\n",
    "    return row[mgi_col]\n",
    "\n",
    "# Apply the mapping to 'Receptor MGI ID'\n",
    "gene_pair['Receptor RGD ID'] = gene_pair.apply(\n",
    "    lambda row: map_if_empty(row, 'Receptor MGI ID', 'Receptor RGD ID', mapping),\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "# Apply the mapping to 'Ligand MGI ID'\n",
    "gene_pair['Ligand RGD ID'] = gene_pair.apply(\n",
    "    lambda row: map_if_empty(row, 'Ligand MGI ID', 'Receptor RGD ID', mapping),\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "18d52b20-4145-4400-a558-519d21a3d5e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['MGI:98223', 'MGI:107560', 'MGI:2136754', 'MGI:2136756',\n",
       "       'MGI:2136757', 'MGI:2136758', 'MGI:2136759', 'MGI:2136760',\n",
       "       'MGI:1935200', 'MGI:1306817', 'MGI:3649078', 'MGI:3774845',\n",
       "       'MGI:1860250', 'MGI:1919379', 'MGI:2136755', 'MGI:2136720',\n",
       "       'MGI:1196250', 'MGI:96824'], dtype=object)"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Function to convert Mouse (ENSEMBL) to RGD ID via biomart\n",
    "import os\n",
    "import sys\n",
    "from pybiomart import Server\n",
    "import pandas as pd\n",
    "sys.path.append(os.path.abspath(\"src\"))\n",
    "from createDataTable import mouse_specific_mgi_ids\n",
    "# Full list of unique MGI IDs\n",
    "mgi_ids = mouse_specific_mgi_ids\n",
    "mgi_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "8fae9f7c-47b4-4c60-a278-ba2e65e6cf87",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to convert Mouse (ENSEMBL) to RGD ID via biomart\n",
    "import os\n",
    "import sys\n",
    "from pybiomart import Server\n",
    "import pandas as pd\n",
    "sys.path.append(os.path.abspath(\"src\"))\n",
    "from createDataTable import mouse_specific_mgi_ids\n",
    "# Full list of unique MGI IDs\n",
    "mgi_ids = mouse_specific_mgi_ids\n",
    "\n",
    "server = Server(host='http://www.ensembl.org')\n",
    "\n",
    "mouse_dataset = server.marts['ENSEMBL_MART_ENSEMBL'].datasets['mmusculus_gene_ensembl']\n",
    "rat_dataset = server.marts['ENSEMBL_MART_ENSEMBL'].datasets['rnorvegicus_gene_ensembl']\n",
    "\n",
    "# Get mouse Ensembl IDs and rat ortholog Ensembl IDs\n",
    "mouse_genes = mouse_dataset.query(\n",
    "    attributes=[\n",
    "        'ensembl_gene_id',\n",
    "        'mgi_id',\n",
    "        'mgi_symbol',\n",
    "        'mgi_description'\n",
    "    ]\n",
    ")\n",
    "mouse_genes.rename(columns={\n",
    "    'MGI description': 'MGI name'})\n",
    "\n",
    "mouse_genes = mouse_genes[mouse_genes[\"MGI ID\"].isin(mgi_ids)]\n",
    "mouse_genes.to_csv(\"data/mouse_name.csv\", index=False)\n",
    "# Step 2: Get mouse Ensembl genes with rat ortholog Ensembl gene IDs (no mouse MGI here)\n",
    "mouse_rat_orthologs = mouse_dataset.query(\n",
    "    attributes=[\n",
    "        'ensembl_gene_id',\n",
    "        'rnorvegicus_homolog_ensembl_gene',\n",
    "        'rnorvegicus_homolog_associated_gene_name'\n",
    "    ]\n",
    ")\n",
    "\n",
    "# Step 3: Merge mouse gene info with orthologs on ensembl_gene_id\n",
    "mouse_to_rat = pd.merge(mouse_genes, mouse_rat_orthologs, on='Gene stable ID')\n",
    "# Filter out entries without rat ortholog\n",
    "mouse_to_rat = mouse_to_rat[mouse_to_rat['Norway rat - BN/NHsdMcwi gene stable ID'].notna() & (mouse_to_rat['Norway rat - BN/NHsdMcwi gene stable ID'] != '')]\n",
    "\n",
    "# Step 4: Retrieve RGD IDs for rat ortholog Ensembl gene IDs\n",
    "rat_ensembl_ids = mouse_to_rat['Norway rat - BN/NHsdMcwi gene stable ID'].unique().tolist()\n",
    "rat_ensembl_to_rgd = rat_dataset.query(\n",
    "    attributes=['ensembl_gene_id', 'rgd_id', 'description']\n",
    ")\n",
    "rat_ensembl_to_rgd = rat_ensembl_to_rgd[rat_ensembl_to_rgd[\"Gene stable ID\"].isin(rat_ensembl_ids)]\n",
    "# Step 5: Merge rat RGD IDs back\n",
    "final_mapping = pd.merge(\n",
    "    mouse_to_rat,\n",
    "    rat_ensembl_to_rgd,\n",
    "    left_on='Norway rat - BN/NHsdMcwi gene stable ID',\n",
    "    right_on='Gene stable ID',\n",
    "    how='left'\n",
    ")\n",
    "# Clean column names for clarity\n",
    "final_mapping.rename(columns={\n",
    "    'Gene stable ID_x': 'mouse_ensembl_gene_id',\n",
    "    'Gene stable ID_y': 'rat_ensembl_gene_id',\n",
    "    'Norway rat - BN/NHsdMcwi gene name': 'RGD symbol',\n",
    "    'Gene description': 'RGD name',\n",
    "    'rgd_id': 'RGD ID'\n",
    "}, inplace=True)\n",
    "\n",
    "final_mapping = final_mapping.drop(columns=[\"Norway rat - BN/NHsdMcwi gene stable ID\"])\n",
    "final_mapping['RGD ID'] = final_mapping['RGD ID'].apply(lambda x: f\"RGD:{int(x)}\" if pd.notnull(x) else None)\n",
    "# Save\n",
    "final_mapping.to_csv(\"data/mouse_to_rat_mapping.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "bb656e86-bbc3-48e7-98d9-54a19dfe0c2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>MGI ID</th>\n",
       "      <th>MGI symbol</th>\n",
       "      <th>Norway rat - BN/NHsdMcwi gene stable ID</th>\n",
       "      <th>Norway rat - BN/NHsdMcwi gene name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000039005</td>\n",
       "      <td>MGI:96824</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>ENSRNOG00000010522</td>\n",
       "      <td>Tlr4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000028064</td>\n",
       "      <td>MGI:107560</td>\n",
       "      <td>Sema4a</td>\n",
       "      <td>ENSRNOG00000019737</td>\n",
       "      <td>Sema4a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000091618</td>\n",
       "      <td>MGI:3774845</td>\n",
       "      <td>H60c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000030149</td>\n",
       "      <td>MGI:1196250</td>\n",
       "      <td>Klrk1</td>\n",
       "      <td>ENSRNOG00000061739</td>\n",
       "      <td>Klrk1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000046387</td>\n",
       "      <td>MGI:2136754</td>\n",
       "      <td>Pcdhb17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000048347</td>\n",
       "      <td>MGI:2136756</td>\n",
       "      <td>Pcdhb18</td>\n",
       "      <td>ENSRNOG00000039470</td>\n",
       "      <td>AC131360.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMUSG00000043313</td>\n",
       "      <td>MGI:2136757</td>\n",
       "      <td>Pcdhb19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSMUSG00000046191</td>\n",
       "      <td>MGI:2136758</td>\n",
       "      <td>Pcdhb20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSMUSG00000044022</td>\n",
       "      <td>MGI:2136759</td>\n",
       "      <td>Pcdhb21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSMUSG00000073591</td>\n",
       "      <td>MGI:2136760</td>\n",
       "      <td>Pcdhb22</td>\n",
       "      <td>ENSRNOG00000020028</td>\n",
       "      <td>Pcdhb22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSMUSG00000103081</td>\n",
       "      <td>MGI:1935200</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSMUSG00000075297</td>\n",
       "      <td>MGI:3649078</td>\n",
       "      <td>H60b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSMUSG00000040026</td>\n",
       "      <td>MGI:98223</td>\n",
       "      <td>Saa3</td>\n",
       "      <td>ENSRNOG00000011895</td>\n",
       "      <td>Saal1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSMUSG00000020283</td>\n",
       "      <td>MGI:1919379</td>\n",
       "      <td>Pex13</td>\n",
       "      <td>ENSRNOG00000054896</td>\n",
       "      <td>Pex13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene stable ID       MGI ID MGI symbol  \\\n",
       "0   ENSMUSG00000039005    MGI:96824       Tlr4   \n",
       "1   ENSMUSG00000028064   MGI:107560     Sema4a   \n",
       "2   ENSMUSG00000091618  MGI:3774845       H60c   \n",
       "3   ENSMUSG00000030149  MGI:1196250      Klrk1   \n",
       "4   ENSMUSG00000046387  MGI:2136754    Pcdhb17   \n",
       "5   ENSMUSG00000048347  MGI:2136756    Pcdhb18   \n",
       "6   ENSMUSG00000043313  MGI:2136757    Pcdhb19   \n",
       "7   ENSMUSG00000046191  MGI:2136758    Pcdhb20   \n",
       "8   ENSMUSG00000044022  MGI:2136759    Pcdhb21   \n",
       "9   ENSMUSG00000073591  MGI:2136760    Pcdhb22   \n",
       "10  ENSMUSG00000103081  MGI:1935200    Pcdhgb8   \n",
       "11  ENSMUSG00000075297  MGI:3649078       H60b   \n",
       "12  ENSMUSG00000040026    MGI:98223       Saa3   \n",
       "13  ENSMUSG00000020283  MGI:1919379      Pex13   \n",
       "\n",
       "   Norway rat - BN/NHsdMcwi gene stable ID Norway rat - BN/NHsdMcwi gene name  \n",
       "0                       ENSRNOG00000010522                               Tlr4  \n",
       "1                       ENSRNOG00000019737                             Sema4a  \n",
       "2                                      NaN                                NaN  \n",
       "3                       ENSRNOG00000061739                              Klrk1  \n",
       "4                                      NaN                                NaN  \n",
       "5                       ENSRNOG00000039470                         AC131360.1  \n",
       "6                                      NaN                                NaN  \n",
       "7                                      NaN                                NaN  \n",
       "8                                      NaN                                NaN  \n",
       "9                       ENSRNOG00000020028                            Pcdhb22  \n",
       "10                                     NaN                                NaN  \n",
       "11                                     NaN                                NaN  \n",
       "12                      ENSRNOG00000011895                              Saal1  \n",
       "13                      ENSRNOG00000054896                              Pex13  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Step 2: Get mouse Ensembl genes with rat ortholog Ensembl gene IDs (no mouse MGI here)\n",
    "mouse_rat_orthologs = mouse_dataset.query(\n",
    "    attributes=[\n",
    "        'ensembl_gene_id',\n",
    "        'rnorvegicus_homolog_ensembl_gene',\n",
    "        'rnorvegicus_homolog_associated_gene_name'\n",
    "    ]\n",
    ")\n",
    "\n",
    "# Step 3: Merge mouse gene info with orthologs on ensembl_gene_id\n",
    "mouse_to_rat = pd.merge(mouse_genes, mouse_rat_orthologs, on='Gene stable ID')\n",
    "# Filter out entries without rat ortholog\n",
    "mouse_to_rat = mouse_to_rat[mouse_to_rat['Norway rat - BN/NHsdMcwi gene stable ID'].notna() & (mouse_to_rat['Norway rat - BN/NHsdMcwi gene stable ID'] != '')]\n",
    "\n",
    "# Step 4: Retrieve RGD IDs for rat ortholog Ensembl gene IDs\n",
    "rat_ensembl_ids = mouse_to_rat['Norway rat - BN/NHsdMcwi gene stable ID'].unique().tolist()\n",
    "rat_ensembl_to_rgd = rat_dataset.query(\n",
    "    attributes=['ensembl_gene_id', 'rgd_id']\n",
    ")\n",
    "rat_ensembl_to_rgd = rat_ensembl_to_rgd[rat_ensembl_to_rgd[\"Gene stable ID\"].isin(rat_ensembl_ids)]\n",
    "# Step 5: Merge rat RGD IDs back\n",
    "final_mapping = pd.merge(\n",
    "    mouse_to_rat,\n",
    "    rat_ensembl_to_rgd,\n",
    "    left_on='Norway rat - BN/NHsdMcwi gene stable ID',\n",
    "    right_on='Gene stable ID',\n",
    "    how='left'\n",
    ")\n",
    "# Clean column names for clarity\n",
    "final_mapping.rename(columns={\n",
    "    'Gene stable ID_x': 'mouse_ensembl_gene_id',\n",
    "    'Gene stable ID_y': 'rat_ensembl_gene_id',\n",
    "    'Norway rat - BN/NHsdMcwi gene name': 'RGD symbol',\n",
    "    'RGD ID': 'rgd_id'\n",
    "}, inplace=True)\n",
    "\n",
    "print(final_mapping.head())\n",
    "final_mapping = final_mapping.drop(columns=[\"Norway rat - BN/NHsdMcwi gene stable ID\"])\n",
    "# Save \n",
    "final_mapping.to_csv(\"data/mouse_to_rat_mapping.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "e5910a76-996e-438d-8a3f-fea7acf88411",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "b15e1282-4c2f-4cb5-9b23-06c48117bd91",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ENSRNOG00000010522',\n",
       " 'ENSRNOG00000019737',\n",
       " 'ENSRNOG00000061739',\n",
       " 'ENSRNOG00000039470',\n",
       " 'ENSRNOG00000020028',\n",
       " 'ENSRNOG00000011895',\n",
       " 'ENSRNOG00000054896']"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rat_ensembl_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "207dfdad-ddf3-4922-854a-5181777891ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>RGD ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8417</th>\n",
       "      <td>ENSRNOG00000010522</td>\n",
       "      <td>3870.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19169</th>\n",
       "      <td>ENSRNOG00000039470</td>\n",
       "      <td>15004341.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19170</th>\n",
       "      <td>ENSRNOG00000039470</td>\n",
       "      <td>1549775.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19197</th>\n",
       "      <td>ENSRNOG00000020028</td>\n",
       "      <td>1310328.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19353</th>\n",
       "      <td>ENSRNOG00000054896</td>\n",
       "      <td>1310682.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24859</th>\n",
       "      <td>ENSRNOG00000011895</td>\n",
       "      <td>1587443.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38717</th>\n",
       "      <td>ENSRNOG00000061739</td>\n",
       "      <td>3180.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43545</th>\n",
       "      <td>ENSRNOG00000019737</td>\n",
       "      <td>1304636.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene stable ID      RGD ID\n",
       "8417   ENSRNOG00000010522      3870.0\n",
       "19169  ENSRNOG00000039470  15004341.0\n",
       "19170  ENSRNOG00000039470   1549775.0\n",
       "19197  ENSRNOG00000020028   1310328.0\n",
       "19353  ENSRNOG00000054896   1310682.0\n",
       "24859  ENSRNOG00000011895   1587443.0\n",
       "38717  ENSRNOG00000061739      3180.0\n",
       "43545  ENSRNOG00000019737   1304636.0"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "92876acd-7fbb-43b2-961d-773005a3ac01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID_x</th>\n",
       "      <th>MGI ID</th>\n",
       "      <th>MGI symbol</th>\n",
       "      <th>Norway rat - BN/NHsdMcwi gene stable ID</th>\n",
       "      <th>Norway rat - BN/NHsdMcwi gene name</th>\n",
       "      <th>Gene stable ID_y</th>\n",
       "      <th>RGD ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000039005</td>\n",
       "      <td>MGI:96824</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>ENSRNOG00000010522</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>ENSRNOG00000010522</td>\n",
       "      <td>3870.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000028064</td>\n",
       "      <td>MGI:107560</td>\n",
       "      <td>Sema4a</td>\n",
       "      <td>ENSRNOG00000019737</td>\n",
       "      <td>Sema4a</td>\n",
       "      <td>ENSRNOG00000019737</td>\n",
       "      <td>1304636.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000030149</td>\n",
       "      <td>MGI:1196250</td>\n",
       "      <td>Klrk1</td>\n",
       "      <td>ENSRNOG00000061739</td>\n",
       "      <td>Klrk1</td>\n",
       "      <td>ENSRNOG00000061739</td>\n",
       "      <td>3180.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000048347</td>\n",
       "      <td>MGI:2136756</td>\n",
       "      <td>Pcdhb18</td>\n",
       "      <td>ENSRNOG00000039470</td>\n",
       "      <td>AC131360.1</td>\n",
       "      <td>ENSRNOG00000039470</td>\n",
       "      <td>15004341.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000048347</td>\n",
       "      <td>MGI:2136756</td>\n",
       "      <td>Pcdhb18</td>\n",
       "      <td>ENSRNOG00000039470</td>\n",
       "      <td>AC131360.1</td>\n",
       "      <td>ENSRNOG00000039470</td>\n",
       "      <td>1549775.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000073591</td>\n",
       "      <td>MGI:2136760</td>\n",
       "      <td>Pcdhb22</td>\n",
       "      <td>ENSRNOG00000020028</td>\n",
       "      <td>Pcdhb22</td>\n",
       "      <td>ENSRNOG00000020028</td>\n",
       "      <td>1310328.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMUSG00000040026</td>\n",
       "      <td>MGI:98223</td>\n",
       "      <td>Saa3</td>\n",
       "      <td>ENSRNOG00000011895</td>\n",
       "      <td>Saal1</td>\n",
       "      <td>ENSRNOG00000011895</td>\n",
       "      <td>1587443.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSMUSG00000020283</td>\n",
       "      <td>MGI:1919379</td>\n",
       "      <td>Pex13</td>\n",
       "      <td>ENSRNOG00000054896</td>\n",
       "      <td>Pex13</td>\n",
       "      <td>ENSRNOG00000054896</td>\n",
       "      <td>1310682.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene stable ID_x       MGI ID MGI symbol  \\\n",
       "0  ENSMUSG00000039005    MGI:96824       Tlr4   \n",
       "1  ENSMUSG00000028064   MGI:107560     Sema4a   \n",
       "2  ENSMUSG00000030149  MGI:1196250      Klrk1   \n",
       "3  ENSMUSG00000048347  MGI:2136756    Pcdhb18   \n",
       "4  ENSMUSG00000048347  MGI:2136756    Pcdhb18   \n",
       "5  ENSMUSG00000073591  MGI:2136760    Pcdhb22   \n",
       "6  ENSMUSG00000040026    MGI:98223       Saa3   \n",
       "7  ENSMUSG00000020283  MGI:1919379      Pex13   \n",
       "\n",
       "  Norway rat - BN/NHsdMcwi gene stable ID Norway rat - BN/NHsdMcwi gene name  \\\n",
       "0                      ENSRNOG00000010522                               Tlr4   \n",
       "1                      ENSRNOG00000019737                             Sema4a   \n",
       "2                      ENSRNOG00000061739                              Klrk1   \n",
       "3                      ENSRNOG00000039470                         AC131360.1   \n",
       "4                      ENSRNOG00000039470                         AC131360.1   \n",
       "5                      ENSRNOG00000020028                            Pcdhb22   \n",
       "6                      ENSRNOG00000011895                              Saal1   \n",
       "7                      ENSRNOG00000054896                              Pex13   \n",
       "\n",
       "     Gene stable ID_y      RGD ID  \n",
       "0  ENSRNOG00000010522      3870.0  \n",
       "1  ENSRNOG00000019737   1304636.0  \n",
       "2  ENSRNOG00000061739      3180.0  \n",
       "3  ENSRNOG00000039470  15004341.0  \n",
       "4  ENSRNOG00000039470   1549775.0  \n",
       "5  ENSRNOG00000020028   1310328.0  \n",
       "6  ENSRNOG00000011895   1587443.0  \n",
       "7  ENSRNOG00000054896   1310682.0  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "2fa0ebd1-1dcc-4d47-bb6a-086656fc2e64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  mouse_ensembl_gene_id       MGI ID MGI symbol  \\\n",
      "0    ENSMUSG00000039005    MGI:96824       Tlr4   \n",
      "1    ENSMUSG00000028064   MGI:107560     Sema4a   \n",
      "2    ENSMUSG00000030149  MGI:1196250      Klrk1   \n",
      "3    ENSMUSG00000048347  MGI:2136756    Pcdhb18   \n",
      "4    ENSMUSG00000048347  MGI:2136756    Pcdhb18   \n",
      "\n",
      "  Norway rat - BN/NHsdMcwi gene stable ID  RGD symbol rat_ensembl_gene_id  \\\n",
      "0                      ENSRNOG00000010522        Tlr4  ENSRNOG00000010522   \n",
      "1                      ENSRNOG00000019737      Sema4a  ENSRNOG00000019737   \n",
      "2                      ENSRNOG00000061739       Klrk1  ENSRNOG00000061739   \n",
      "3                      ENSRNOG00000039470  AC131360.1  ENSRNOG00000039470   \n",
      "4                      ENSRNOG00000039470  AC131360.1  ENSRNOG00000039470   \n",
      "\n",
      "       rgd_id  \n",
      "0      3870.0  \n",
      "1   1304636.0  \n",
      "2      3180.0  \n",
      "3  15004341.0  \n",
      "4   1549775.0  \n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "96074bc9-7b21-401d-b1c2-73f8178306ff",
   "metadata": {},
   "outputs": [
    {
     "ename": "BiomartException",
     "evalue": "Unknown filter ensembl_gene_id, check dataset filters for a list of valid filters.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pybiomart/dataset.py:254\u001b[0m, in \u001b[0;36mDataset.query\u001b[0;34m(self, attributes, filters, only_unique, use_attr_names)\u001b[0m\n\u001b[1;32m    253\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 254\u001b[0m     filter_ \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfilters\u001b[49m\u001b[43m[\u001b[49m\u001b[43mname\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m    255\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_add_filter_node(dataset, filter_, value)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'ensembl_gene_id'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mBiomartException\u001b[0m                          Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[44], line 7\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m# Step 4: Retrieve RGD IDs for rat ortholog Ensembl gene IDs\u001b[39;00m\n\u001b[1;32m      5\u001b[0m rat_ensembl_ids \u001b[38;5;241m=\u001b[39m mouse_to_rat[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mNorway rat - BN/NHsdMcwi gene stable ID\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39munique()\u001b[38;5;241m.\u001b[39mtolist()\n\u001b[0;32m----> 7\u001b[0m rat_ensembl_to_rgd \u001b[38;5;241m=\u001b[39m \u001b[43mrat_dataset\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mquery\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m      8\u001b[0m \u001b[43m    \u001b[49m\u001b[43mattributes\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mensembl_gene_id\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mrgd_id\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      9\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfilters\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m{\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mensembl_gene_id\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mrat_ensembl_ids\u001b[49m\u001b[43m}\u001b[49m\n\u001b[1;32m     10\u001b[0m \u001b[43m)\u001b[49m\n\u001b[1;32m     12\u001b[0m \u001b[38;5;66;03m# Step 5: Merge rat RGD IDs back\u001b[39;00m\n\u001b[1;32m     13\u001b[0m final_mapping \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mmerge(\n\u001b[1;32m     14\u001b[0m     mouse_to_rat,\n\u001b[1;32m     15\u001b[0m     rat_ensembl_to_rgd,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     18\u001b[0m     how\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mleft\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m     19\u001b[0m )\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pybiomart/dataset.py:257\u001b[0m, in \u001b[0;36mDataset.query\u001b[0;34m(self, attributes, filters, only_unique, use_attr_names)\u001b[0m\n\u001b[1;32m    255\u001b[0m             \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_add_filter_node(dataset, filter_, value)\n\u001b[1;32m    256\u001b[0m         \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m:\n\u001b[0;32m--> 257\u001b[0m             \u001b[38;5;28;01mraise\u001b[39;00m BiomartException(\n\u001b[1;32m    258\u001b[0m                 \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mUnknown filter \u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m, check dataset filters \u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m    259\u001b[0m                 \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfor a list of valid filters.\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mformat(name))\n\u001b[1;32m    261\u001b[0m \u001b[38;5;66;03m# Fetch response.\u001b[39;00m\n\u001b[1;32m    262\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget(query\u001b[38;5;241m=\u001b[39mElementTree\u001b[38;5;241m.\u001b[39mtostring(root))\n",
      "\u001b[0;31mBiomartException\u001b[0m: Unknown filter ensembl_gene_id, check dataset filters for a list of valid filters."
     ]
    }
   ],
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "0778c5e2-a34e-495e-9819-7fc17dabd809",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ENSRNOG00000030644',\n",
       " 'ENSRNOG00000031033',\n",
       " 'ENSRNOG00000034234',\n",
       " 'ENSRNOG00000030371',\n",
       " 'ENSRNOG00000033299',\n",
       " 'ENSRNOG00000031979',\n",
       " 'ENSRNOG00000030700',\n",
       " 'ENSRNOG00000033615',\n",
       " 'ENSRNOG00000031053',\n",
       " 'ENSRNOG00000029707',\n",
       " 'ENSRNOG00000029971',\n",
       " 'ENSRNOG00000029042',\n",
       " 'ENSRNOG00000031766',\n",
       " 'ENSRNOG00000064035',\n",
       " 'ENSRNOG00000077508',\n",
       " 'ENSRNOG00000043130',\n",
       " 'ENSRNOG00000046931',\n",
       " 'ENSRNOG00000073402',\n",
       " 'ENSRNOG00000089019',\n",
       " 'ENSRNOG00000063444',\n",
       " 'ENSRNOG00000067899',\n",
       " 'ENSRNOG00000088869',\n",
       " 'ENSRNOG00000075698',\n",
       " 'ENSRNOG00000062749',\n",
       " 'ENSRNOG00000083237',\n",
       " 'ENSRNOG00000065680',\n",
       " 'ENSRNOG00000067003',\n",
       " 'ENSRNOG00000063633',\n",
       " 'ENSRNOG00000064574',\n",
       " 'ENSRNOG00000072535',\n",
       " 'ENSRNOG00000084642',\n",
       " 'ENSRNOG00000071611',\n",
       " 'ENSRNOG00000067042',\n",
       " 'ENSRNOG00000073978',\n",
       " 'ENSRNOG00000033747',\n",
       " 'ENSRNOG00000022800',\n",
       " 'ENSRNOG00000017256',\n",
       " 'ENSRNOG00000084963',\n",
       " 'ENSRNOG00000063954',\n",
       " 'ENSRNOG00000063582',\n",
       " 'ENSRNOG00000064887',\n",
       " 'ENSRNOG00000077684',\n",
       " 'ENSRNOG00000076768',\n",
       " 'ENSRNOG00000064259',\n",
       " 'ENSRNOG00000072094',\n",
       " 'ENSRNOG00000067676',\n",
       " 'ENSRNOG00000078153',\n",
       " 'ENSRNOG00000080258',\n",
       " 'ENSRNOG00000082759',\n",
       " 'ENSRNOG00000078859',\n",
       " 'ENSRNOG00000089969',\n",
       " 'ENSRNOG00000075715',\n",
       " 'ENSRNOG00000089859',\n",
       " 'ENSRNOG00000083031',\n",
       " 'ENSRNOG00000074289',\n",
       " 'ENSRNOG00000084670',\n",
       " 'ENSRNOG00000083348',\n",
       " 'ENSRNOG00000077163',\n",
       " 'ENSRNOG00000072719',\n",
       " 'ENSRNOG00000088152',\n",
       " 'ENSRNOG00000089633',\n",
       " 'ENSRNOG00000049619',\n",
       " 'ENSRNOG00000048867',\n",
       " 'ENSRNOG00000083094',\n",
       " 'ENSRNOG00000065734',\n",
       " 'ENSRNOG00000075048',\n",
       " 'ENSRNOG00000075217',\n",
       " 'ENSRNOG00000085874',\n",
       " 'ENSRNOG00000073240',\n",
       " 'ENSRNOG00000068289',\n",
       " 'ENSRNOG00000078247',\n",
       " 'ENSRNOG00000083477',\n",
       " 'ENSRNOG00000084696',\n",
       " 'ENSRNOG00000088275',\n",
       " 'ENSRNOG00000078425',\n",
       " 'ENSRNOG00000079973',\n",
       " 'ENSRNOG00000073718',\n",
       " 'ENSRNOG00000079273',\n",
       " 'ENSRNOG00000085381',\n",
       " 'ENSRNOG00000072683',\n",
       " 'ENSRNOG00000089769',\n",
       " 'ENSRNOG00000075704',\n",
       " 'ENSRNOG00000085197',\n",
       " 'ENSRNOG00000089025',\n",
       " 'ENSRNOG00000074627',\n",
       " 'ENSRNOG00000086556',\n",
       " 'ENSRNOG00000085696',\n",
       " 'ENSRNOG00000088255',\n",
       " 'ENSRNOG00000090248',\n",
       " 'ENSRNOG00000077860',\n",
       " 'ENSRNOG00000084517',\n",
       " 'ENSRNOG00000069646',\n",
       " 'ENSRNOG00000084216',\n",
       " 'ENSRNOG00000073388',\n",
       " 'ENSRNOG00000089042',\n",
       " 'ENSRNOG00000070707',\n",
       " 'ENSRNOG00000076756',\n",
       " 'ENSRNOG00000072522',\n",
       " 'ENSRNOG00000027907',\n",
       " 'ENSRNOG00000015867',\n",
       " 'ENSRNOG00000064736',\n",
       " 'ENSRNOG00000015602',\n",
       " 'ENSRNOG00000014215',\n",
       " 'ENSRNOG00000053926',\n",
       " 'ENSRNOG00000016659',\n",
       " 'ENSRNOG00000017035',\n",
       " 'ENSRNOG00000016677',\n",
       " 'ENSRNOG00000050693',\n",
       " 'ENSRNOG00000004273',\n",
       " 'ENSRNOG00000082217',\n",
       " 'ENSRNOG00000043017',\n",
       " 'ENSRNOG00000027478',\n",
       " 'ENSRNOG00000017159',\n",
       " 'ENSRNOG00000003198',\n",
       " 'ENSRNOG00000034182',\n",
       " 'ENSRNOG00000034289',\n",
       " 'ENSRNOG00000088815',\n",
       " 'ENSRNOG00000065044',\n",
       " 'ENSRNOG00000056248',\n",
       " 'ENSRNOG00000051237',\n",
       " 'ENSRNOG00000023197',\n",
       " 'ENSRNOG00000060963',\n",
       " 'ENSRNOG00000089917',\n",
       " 'ENSRNOG00000086574',\n",
       " 'ENSRNOG00000060522',\n",
       " 'ENSRNOG00000054045',\n",
       " 'ENSRNOG00000051865',\n",
       " 'ENSRNOG00000051703',\n",
       " 'ENSRNOG00000054092',\n",
       " 'ENSRNOG00000057516',\n",
       " 'ENSRNOG00000015225',\n",
       " 'ENSRNOG00000024620',\n",
       " 'ENSRNOG00000002746',\n",
       " 'ENSRNOG00000053854',\n",
       " 'ENSRNOG00000077429',\n",
       " 'ENSRNOG00000085838',\n",
       " 'ENSRNOG00000089964',\n",
       " 'ENSRNOG00000091221',\n",
       " 'ENSRNOG00000078431',\n",
       " 'ENSRNOG00000054892',\n",
       " 'ENSRNOG00000076964',\n",
       " 'ENSRNOG00000036615',\n",
       " 'ENSRNOG00000063750',\n",
       " 'ENSRNOG00000068993',\n",
       " 'ENSRNOG00000053761',\n",
       " 'ENSRNOG00000027151',\n",
       " 'ENSRNOG00000052937',\n",
       " 'ENSRNOG00000004084',\n",
       " 'ENSRNOG00000014891',\n",
       " 'ENSRNOG00000002797',\n",
       " 'ENSRNOG00000035454',\n",
       " 'ENSRNOG00000040386',\n",
       " 'ENSRNOG00000035487',\n",
       " 'ENSRNOG00000036244',\n",
       " 'ENSRNOG00000039969',\n",
       " 'ENSRNOG00000072371',\n",
       " 'ENSRNOG00000086720',\n",
       " 'ENSRNOG00000052779',\n",
       " 'ENSRNOG00000053579',\n",
       " 'ENSRNOG00000040351',\n",
       " 'ENSRNOG00000035634',\n",
       " 'ENSRNOG00000002833',\n",
       " 'ENSRNOG00000014928',\n",
       " 'ENSRNOG00000004110',\n",
       " 'ENSRNOG00000048378',\n",
       " 'ENSRNOG00000035600',\n",
       " 'ENSRNOG00000040406',\n",
       " 'ENSRNOG00000027645',\n",
       " 'ENSRNOG00000061147',\n",
       " 'ENSRNOG00000056025',\n",
       " 'ENSRNOG00000054449',\n",
       " 'ENSRNOG00000036486',\n",
       " 'ENSRNOG00000058007',\n",
       " 'ENSRNOG00000035490',\n",
       " 'ENSRNOG00000035557',\n",
       " 'ENSRNOG00000027444',\n",
       " 'ENSRNOG00000015030',\n",
       " 'ENSRNOG00000012062',\n",
       " 'ENSRNOG00000056258',\n",
       " 'ENSRNOG00000012085',\n",
       " 'ENSRNOG00000005077',\n",
       " 'ENSRNOG00000004696',\n",
       " 'ENSRNOG00000086017',\n",
       " 'ENSRNOG00000090553',\n",
       " 'ENSRNOG00000004841',\n",
       " 'ENSRNOG00000061452',\n",
       " 'ENSRNOG00000021569',\n",
       " 'ENSRNOG00000015213',\n",
       " 'ENSRNOG00000027309',\n",
       " 'ENSRNOG00000015232',\n",
       " 'ENSRNOG00000016853',\n",
       " 'ENSRNOG00000024499',\n",
       " 'ENSRNOG00000068162',\n",
       " 'ENSRNOG00000016505',\n",
       " 'ENSRNOG00000005053',\n",
       " 'ENSRNOG00000058813',\n",
       " 'ENSRNOG00000012094',\n",
       " 'ENSRNOG00000051051',\n",
       " 'ENSRNOG00000025476',\n",
       " 'ENSRNOG00000015406',\n",
       " 'ENSRNOG00000039496',\n",
       " 'ENSRNOG00000055728',\n",
       " 'ENSRNOG00000014645',\n",
       " 'ENSRNOG00000016530',\n",
       " 'ENSRNOG00000033559',\n",
       " 'ENSRNOG00000015522',\n",
       " 'ENSRNOG00000016714',\n",
       " 'ENSRNOG00000002115',\n",
       " 'ENSRNOG00000002104',\n",
       " 'ENSRNOG00000053042',\n",
       " 'ENSRNOG00000071958',\n",
       " 'ENSRNOG00000083862',\n",
       " 'ENSRNOG00000089241',\n",
       " 'ENSRNOG00000004973',\n",
       " 'ENSRNOG00000058285',\n",
       " 'ENSRNOG00000060496',\n",
       " 'ENSRNOG00000002092',\n",
       " 'ENSRNOG00000022364',\n",
       " 'ENSRNOG00000014459',\n",
       " 'ENSRNOG00000023034',\n",
       " 'ENSRNOG00000052381',\n",
       " 'ENSRNOG00000011335',\n",
       " 'ENSRNOG00000061806',\n",
       " 'ENSRNOG00000072432',\n",
       " 'ENSRNOG00000079151',\n",
       " 'ENSRNOG00000052562',\n",
       " 'ENSRNOG00000021555',\n",
       " 'ENSRNOG00000013774',\n",
       " 'ENSRNOG00000021524',\n",
       " 'ENSRNOG00000066332',\n",
       " 'ENSRNOG00000048606',\n",
       " 'ENSRNOG00000002080',\n",
       " 'ENSRNOG00000023013',\n",
       " 'ENSRNOG00000016632',\n",
       " 'ENSRNOG00000054964',\n",
       " 'ENSRNOG00000076101',\n",
       " 'ENSRNOG00000060048',\n",
       " 'ENSRNOG00000004382',\n",
       " 'ENSRNOG00000057880',\n",
       " 'ENSRNOG00000060617',\n",
       " 'ENSRNOG00000013738',\n",
       " 'ENSRNOG00000071760',\n",
       " 'ENSRNOG00000025285',\n",
       " 'ENSRNOG00000016526',\n",
       " 'ENSRNOG00000045695',\n",
       " 'ENSRNOG00000057599',\n",
       " 'ENSRNOG00000050508',\n",
       " 'ENSRNOG00000016275',\n",
       " 'ENSRNOG00000015516',\n",
       " 'ENSRNOG00000065140',\n",
       " 'ENSRNOG00000061428',\n",
       " 'ENSRNOG00000015895',\n",
       " 'ENSRNOG00000006178',\n",
       " 'ENSRNOG00000036487',\n",
       " 'ENSRNOG00000035635',\n",
       " 'ENSRNOG00000013674',\n",
       " 'ENSRNOG00000058376',\n",
       " 'ENSRNOG00000067508',\n",
       " 'ENSRNOG00000053139',\n",
       " 'ENSRNOG00000061657',\n",
       " 'ENSRNOG00000056216',\n",
       " 'ENSRNOG00000066176',\n",
       " 'ENSRNOG00000021589',\n",
       " 'ENSRNOG00000015894',\n",
       " 'ENSRNOG00000066670',\n",
       " 'ENSRNOG00000032973',\n",
       " 'ENSRNOG00000022202',\n",
       " 'ENSRNOG00000007184',\n",
       " 'ENSRNOG00000004723',\n",
       " 'ENSRNOG00000030119',\n",
       " 'ENSRNOG00000057231',\n",
       " 'ENSRNOG00000075783',\n",
       " 'ENSRNOG00000053785',\n",
       " 'ENSRNOG00000016023',\n",
       " 'ENSRNOG00000002790',\n",
       " 'ENSRNOG00000059714',\n",
       " 'ENSRNOG00000058664',\n",
       " 'ENSRNOG00000057382',\n",
       " 'ENSRNOG00000016075',\n",
       " 'ENSRNOG00000057832',\n",
       " 'ENSRNOG00000063275',\n",
       " 'ENSRNOG00000016284',\n",
       " 'ENSRNOG00000015645',\n",
       " 'ENSRNOG00000016301',\n",
       " 'ENSRNOG00000061182',\n",
       " 'ENSRNOG00000053230',\n",
       " 'ENSRNOG00000059230',\n",
       " 'ENSRNOG00000053730',\n",
       " 'ENSRNOG00000036842',\n",
       " 'ENSRNOG00000016948',\n",
       " 'ENSRNOG00000052773',\n",
       " 'ENSRNOG00000002072',\n",
       " 'ENSRNOG00000007262',\n",
       " 'ENSRNOG00000002751',\n",
       " 'ENSRNOG00000004377',\n",
       " 'ENSRNOG00000017002',\n",
       " 'ENSRNOG00000049345',\n",
       " 'ENSRNOG00000051680',\n",
       " 'ENSRNOG00000021400',\n",
       " 'ENSRNOG00000072090',\n",
       " 'ENSRNOG00000082755',\n",
       " 'ENSRNOG00000083264',\n",
       " 'ENSRNOG00000081394',\n",
       " 'ENSRNOG00000079797',\n",
       " 'ENSRNOG00000085288',\n",
       " 'ENSRNOG00000090479',\n",
       " 'ENSRNOG00000083909',\n",
       " 'ENSRNOG00000080005',\n",
       " 'ENSRNOG00000056558',\n",
       " 'ENSRNOG00000025274',\n",
       " 'ENSRNOG00000062552',\n",
       " 'ENSRNOG00000042327',\n",
       " 'ENSRNOG00000043219',\n",
       " 'ENSRNOG00000020350',\n",
       " 'ENSRNOG00000021399',\n",
       " 'ENSRNOG00000014460',\n",
       " 'ENSRNOG00000036841',\n",
       " 'ENSRNOG00000001525',\n",
       " 'ENSRNOG00000014361',\n",
       " 'ENSRNOG00000061379',\n",
       " 'ENSRNOG00000089965',\n",
       " 'ENSRNOG00000050091',\n",
       " 'ENSRNOG00000001271',\n",
       " 'ENSRNOG00000034790',\n",
       " 'ENSRNOG00000082846',\n",
       " 'ENSRNOG00000015418',\n",
       " 'ENSRNOG00000020412',\n",
       " 'ENSRNOG00000033803',\n",
       " 'ENSRNOG00000064545',\n",
       " 'ENSRNOG00000034180',\n",
       " 'ENSRNOG00000029582',\n",
       " 'ENSRNOG00000058193',\n",
       " 'ENSRNOG00000090744',\n",
       " 'ENSRNOG00000060919',\n",
       " 'ENSRNOG00000012799',\n",
       " 'ENSRNOG00000052218',\n",
       " 'ENSRNOG00000019711',\n",
       " 'ENSRNOG00000016544',\n",
       " 'ENSRNOG00000002051',\n",
       " 'ENSRNOG00000019577',\n",
       " 'ENSRNOG00000007400',\n",
       " 'ENSRNOG00000028743',\n",
       " 'ENSRNOG00000022156',\n",
       " 'ENSRNOG00000020455',\n",
       " 'ENSRNOG00000017064',\n",
       " 'ENSRNOG00000002662',\n",
       " 'ENSRNOG00000004412',\n",
       " 'ENSRNOG00000003377',\n",
       " 'ENSRNOG00000002660',\n",
       " 'ENSRNOG00000022429',\n",
       " 'ENSRNOG00000061530',\n",
       " 'ENSRNOG00000069107',\n",
       " 'ENSRNOG00000001829',\n",
       " 'ENSRNOG00000050152',\n",
       " 'ENSRNOG00000007333',\n",
       " 'ENSRNOG00000001833',\n",
       " 'ENSRNOG00000081104',\n",
       " 'ENSRNOG00000019542',\n",
       " 'ENSRNOG00000056703',\n",
       " 'ENSRNOG00000049054',\n",
       " 'ENSRNOG00000014775',\n",
       " 'ENSRNOG00000027317',\n",
       " 'ENSRNOG00000002036',\n",
       " 'ENSRNOG00000061612',\n",
       " 'ENSRNOG00000035547',\n",
       " 'ENSRNOG00000049407',\n",
       " 'ENSRNOG00000004821',\n",
       " 'ENSRNOG00000061510',\n",
       " 'ENSRNOG00000057865',\n",
       " 'ENSRNOG00000071440',\n",
       " 'ENSRNOG00000052892',\n",
       " 'ENSRNOG00000056815',\n",
       " 'ENSRNOG00000019235',\n",
       " 'ENSRNOG00000013217',\n",
       " 'ENSRNOG00000007850',\n",
       " 'ENSRNOG00000054017',\n",
       " 'ENSRNOG00000016671',\n",
       " 'ENSRNOG00000004854',\n",
       " 'ENSRNOG00000036839',\n",
       " 'ENSRNOG00000005054',\n",
       " 'ENSRNOG00000028670',\n",
       " 'ENSRNOG00000020460',\n",
       " 'ENSRNOG00000031661',\n",
       " 'ENSRNOG00000056716',\n",
       " 'ENSRNOG00000020463',\n",
       " 'ENSRNOG00000014264',\n",
       " 'ENSRNOG00000032776',\n",
       " 'ENSRNOG00000025581',\n",
       " 'ENSRNOG00000055256',\n",
       " 'ENSRNOG00000090943',\n",
       " 'ENSRNOG00000013240',\n",
       " 'ENSRNOG00000051408',\n",
       " 'ENSRNOG00000064975',\n",
       " 'ENSRNOG00000020475',\n",
       " 'ENSRNOG00000004958',\n",
       " 'ENSRNOG00000008035',\n",
       " 'ENSRNOG00000036837',\n",
       " 'ENSRNOG00000027564',\n",
       " 'ENSRNOG00000020527',\n",
       " 'ENSRNOG00000005061',\n",
       " 'ENSRNOG00000005417',\n",
       " 'ENSRNOG00000049681',\n",
       " 'ENSRNOG00000054689',\n",
       " 'ENSRNOG00000005551',\n",
       " 'ENSRNOG00000000166',\n",
       " 'ENSRNOG00000036835',\n",
       " 'ENSRNOG00000056574',\n",
       " 'ENSRNOG00000025028',\n",
       " 'ENSRNOG00000020551',\n",
       " 'ENSRNOG00000035525',\n",
       " 'ENSRNOG00000061010',\n",
       " 'ENSRNOG00000072086',\n",
       " 'ENSRNOG00000036833',\n",
       " 'ENSRNOG00000020552',\n",
       " 'ENSRNOG00000017164',\n",
       " 'ENSRNOG00000006103',\n",
       " 'ENSRNOG00000060171',\n",
       " 'ENSRNOG00000014019',\n",
       " 'ENSRNOG00000048101',\n",
       " 'ENSRNOG00000052252',\n",
       " 'ENSRNOG00000058865',\n",
       " 'ENSRNOG00000020782',\n",
       " 'ENSRNOG00000049137',\n",
       " 'ENSRNOG00000075202',\n",
       " 'ENSRNOG00000000167',\n",
       " 'ENSRNOG00000016532',\n",
       " 'ENSRNOG00000024651',\n",
       " 'ENSRNOG00000016207',\n",
       " 'ENSRNOG00000008834',\n",
       " 'ENSRNOG00000059732',\n",
       " 'ENSRNOG00000083010',\n",
       " 'ENSRNOG00000068332',\n",
       " 'ENSRNOG00000057451',\n",
       " 'ENSRNOG00000061367',\n",
       " 'ENSRNOG00000017866',\n",
       " 'ENSRNOG00000060476',\n",
       " 'ENSRNOG00000060806',\n",
       " 'ENSRNOG00000015546',\n",
       " 'ENSRNOG00000089480',\n",
       " 'ENSRNOG00000029220',\n",
       " 'ENSRNOG00000011639',\n",
       " 'ENSRNOG00000016036',\n",
       " 'ENSRNOG00000053050',\n",
       " 'ENSRNOG00000036831',\n",
       " 'ENSRNOG00000023591',\n",
       " 'ENSRNOG00000020567',\n",
       " 'ENSRNOG00000033940',\n",
       " 'ENSRNOG00000066842',\n",
       " 'ENSRNOG00000060334',\n",
       " 'ENSRNOG00000054118',\n",
       " 'ENSRNOG00000005530',\n",
       " 'ENSRNOG00000054386',\n",
       " 'ENSRNOG00000055121',\n",
       " 'ENSRNOG00000051955',\n",
       " 'ENSRNOG00000056883',\n",
       " 'ENSRNOG00000086729',\n",
       " 'ENSRNOG00000011167',\n",
       " 'ENSRNOG00000037851',\n",
       " 'ENSRNOG00000017826',\n",
       " 'ENSRNOG00000058249',\n",
       " 'ENSRNOG00000011949',\n",
       " 'ENSRNOG00000017503',\n",
       " 'ENSRNOG00000027096',\n",
       " 'ENSRNOG00000005650',\n",
       " 'ENSRNOG00000061102',\n",
       " 'ENSRNOG00000021904',\n",
       " 'ENSRNOG00000035517',\n",
       " 'ENSRNOG00000017904',\n",
       " 'ENSRNOG00000020587',\n",
       " 'ENSRNOG00000014503',\n",
       " 'ENSRNOG00000013912',\n",
       " 'ENSRNOG00000001825',\n",
       " 'ENSRNOG00000000165',\n",
       " 'ENSRNOG00000036830',\n",
       " 'ENSRNOG00000022218',\n",
       " 'ENSRNOG00000027422',\n",
       " 'ENSRNOG00000036829',\n",
       " 'ENSRNOG00000025252',\n",
       " 'ENSRNOG00000058061',\n",
       " 'ENSRNOG00000060348',\n",
       " 'ENSRNOG00000068183',\n",
       " 'ENSRNOG00000006467',\n",
       " 'ENSRNOG00000008734',\n",
       " 'ENSRNOG00000049994',\n",
       " 'ENSRNOG00000046468',\n",
       " 'ENSRNOG00000028594',\n",
       " 'ENSRNOG00000069719',\n",
       " 'ENSRNOG00000030639',\n",
       " 'ENSRNOG00000017318',\n",
       " 'ENSRNOG00000007043',\n",
       " 'ENSRNOG00000086696',\n",
       " 'ENSRNOG00000077396',\n",
       " 'ENSRNOG00000061195',\n",
       " 'ENSRNOG00000002451',\n",
       " 'ENSRNOG00000090726',\n",
       " 'ENSRNOG00000069323',\n",
       " 'ENSRNOG00000010645',\n",
       " 'ENSRNOG00000020617',\n",
       " 'ENSRNOG00000025300',\n",
       " 'ENSRNOG00000002032',\n",
       " 'ENSRNOG00000047406',\n",
       " 'ENSRNOG00000014323',\n",
       " 'ENSRNOG00000008934',\n",
       " 'ENSRNOG00000017585',\n",
       " 'ENSRNOG00000017514',\n",
       " 'ENSRNOG00000039994',\n",
       " 'ENSRNOG00000055394',\n",
       " 'ENSRNOG00000008978',\n",
       " 'ENSRNOG00000002028',\n",
       " 'ENSRNOG00000008987',\n",
       " 'ENSRNOG00000061627',\n",
       " 'ENSRNOG00000051898',\n",
       " 'ENSRNOG00000028930',\n",
       " 'ENSRNOG00000032150',\n",
       " 'ENSRNOG00000017840',\n",
       " 'ENSRNOG00000053406',\n",
       " 'ENSRNOG00000009208',\n",
       " 'ENSRNOG00000006699',\n",
       " 'ENSRNOG00000042152',\n",
       " 'ENSRNOG00000053813',\n",
       " 'ENSRNOG00000017951',\n",
       " 'ENSRNOG00000026527',\n",
       " 'ENSRNOG00000006899',\n",
       " 'ENSRNOG00000013011',\n",
       " 'ENSRNOG00000071176',\n",
       " 'ENSRNOG00000010488',\n",
       " 'ENSRNOG00000039001',\n",
       " 'ENSRNOG00000066548',\n",
       " 'ENSRNOG00000040008',\n",
       " 'ENSRNOG00000055506',\n",
       " 'ENSRNOG00000006744',\n",
       " 'ENSRNOG00000020669',\n",
       " 'ENSRNOG00000017604',\n",
       " 'ENSRNOG00000089880',\n",
       " 'ENSRNOG00000053851',\n",
       " 'ENSRNOG00000027115',\n",
       " 'ENSRNOG00000014333',\n",
       " 'ENSRNOG00000017988',\n",
       " 'ENSRNOG00000026114',\n",
       " 'ENSRNOG00000058572',\n",
       " 'ENSRNOG00000079381',\n",
       " 'ENSRNOG00000020700',\n",
       " 'ENSRNOG00000001992',\n",
       " 'ENSRNOG00000018070',\n",
       " 'ENSRNOG00000018092',\n",
       " 'ENSRNOG00000038877',\n",
       " 'ENSRNOG00000007901',\n",
       " 'ENSRNOG00000017674',\n",
       " 'ENSRNOG00000011931',\n",
       " 'ENSRNOG00000060879',\n",
       " 'ENSRNOG00000008015',\n",
       " 'ENSRNOG00000026953',\n",
       " 'ENSRNOG00000011245',\n",
       " 'ENSRNOG00000037845',\n",
       " 'ENSRNOG00000008785',\n",
       " 'ENSRNOG00000014515',\n",
       " 'ENSRNOG00000005344',\n",
       " 'ENSRNOG00000053351',\n",
       " 'ENSRNOG00000060150',\n",
       " 'ENSRNOG00000034007',\n",
       " 'ENSRNOG00000056893',\n",
       " 'ENSRNOG00000009145',\n",
       " 'ENSRNOG00000017568',\n",
       " 'ENSRNOG00000058333',\n",
       " 'ENSRNOG00000029090',\n",
       " 'ENSRNOG00000019695',\n",
       " 'ENSRNOG00000016967',\n",
       " 'ENSRNOG00000005689',\n",
       " 'ENSRNOG00000055960',\n",
       " 'ENSRNOG00000019643',\n",
       " 'ENSRNOG00000059298',\n",
       " 'ENSRNOG00000025726',\n",
       " 'ENSRNOG00000064275',\n",
       " 'ENSRNOG00000010721',\n",
       " 'ENSRNOG00000005790',\n",
       " 'ENSRNOG00000005424',\n",
       " 'ENSRNOG00000014720',\n",
       " 'ENSRNOG00000002429',\n",
       " 'ENSRNOG00000067441',\n",
       " 'ENSRNOG00000019403',\n",
       " 'ENSRNOG00000005825',\n",
       " 'ENSRNOG00000074453',\n",
       " 'ENSRNOG00000002428',\n",
       " 'ENSRNOG00000070513',\n",
       " 'ENSRNOG00000014999',\n",
       " 'ENSRNOG00000011134',\n",
       " 'ENSRNOG00000037839',\n",
       " 'ENSRNOG00000055217',\n",
       " 'ENSRNOG00000054720',\n",
       " 'ENSRNOG00000005492',\n",
       " 'ENSRNOG00000024452',\n",
       " 'ENSRNOG00000001813',\n",
       " 'ENSRNOG00000065823',\n",
       " 'ENSRNOG00000019365',\n",
       " 'ENSRNOG00000026569',\n",
       " 'ENSRNOG00000056917',\n",
       " 'ENSRNOG00000037838',\n",
       " 'ENSRNOG00000058368',\n",
       " 'ENSRNOG00000049827',\n",
       " 'ENSRNOG00000075763',\n",
       " 'ENSRNOG00000076377',\n",
       " 'ENSRNOG00000010155',\n",
       " 'ENSRNOG00000051867',\n",
       " 'ENSRNOG00000012019',\n",
       " 'ENSRNOG00000009310',\n",
       " 'ENSRNOG00000047491',\n",
       " 'ENSRNOG00000046926',\n",
       " 'ENSRNOG00000015566',\n",
       " 'ENSRNOG00000019305',\n",
       " 'ENSRNOG00000021484',\n",
       " 'ENSRNOG00000090541',\n",
       " 'ENSRNOG00000064925',\n",
       " 'ENSRNOG00000058286',\n",
       " 'ENSRNOG00000058841',\n",
       " 'ENSRNOG00000005927',\n",
       " 'ENSRNOG00000011211',\n",
       " 'ENSRNOG00000056492',\n",
       " 'ENSRNOG00000011542',\n",
       " 'ENSRNOG00000015325',\n",
       " 'ENSRNOG00000015215',\n",
       " 'ENSRNOG00000074836',\n",
       " 'ENSRNOG00000045818',\n",
       " 'ENSRNOG00000019217',\n",
       " 'ENSRNOG00000008610',\n",
       " 'ENSRNOG00000011382',\n",
       " 'ENSRNOG00000009431',\n",
       " 'ENSRNOG00000017249',\n",
       " 'ENSRNOG00000008644',\n",
       " 'ENSRNOG00000028504',\n",
       " 'ENSRNOG00000047198',\n",
       " 'ENSRNOG00000028403',\n",
       " 'ENSRNOG00000061533',\n",
       " 'ENSRNOG00000022970',\n",
       " 'ENSRNOG00000038973',\n",
       " 'ENSRNOG00000065348',\n",
       " 'ENSRNOG00000017180',\n",
       " 'ENSRNOG00000051536',\n",
       " 'ENSRNOG00000072406',\n",
       " 'ENSRNOG00000073488',\n",
       " 'ENSRNOG00000086097',\n",
       " 'ENSRNOG00000061486',\n",
       " 'ENSRNOG00000015059',\n",
       " 'ENSRNOG00000008931',\n",
       " 'ENSRNOG00000032970',\n",
       " 'ENSRNOG00000016216',\n",
       " 'ENSRNOG00000019134',\n",
       " 'ENSRNOG00000017272',\n",
       " 'ENSRNOG00000011572',\n",
       " 'ENSRNOG00000011675',\n",
       " 'ENSRNOG00000085848',\n",
       " 'ENSRNOG00000013736',\n",
       " 'ENSRNOG00000002365',\n",
       " 'ENSRNOG00000017428',\n",
       " 'ENSRNOG00000010034',\n",
       " 'ENSRNOG00000016021',\n",
       " 'ENSRNOG00000054044',\n",
       " 'ENSRNOG00000017712',\n",
       " 'ENSRNOG00000036828',\n",
       " 'ENSRNOG00000045918',\n",
       " 'ENSRNOG00000014879',\n",
       " 'ENSRNOG00000039017',\n",
       " 'ENSRNOG00000065480',\n",
       " 'ENSRNOG00000015035',\n",
       " 'ENSRNOG00000060295',\n",
       " 'ENSRNOG00000084862',\n",
       " 'ENSRNOG00000011704',\n",
       " 'ENSRNOG00000027756',\n",
       " 'ENSRNOG00000017746',\n",
       " 'ENSRNOG00000002346',\n",
       " 'ENSRNOG00000028292',\n",
       " 'ENSRNOG00000007323',\n",
       " 'ENSRNOG00000000964',\n",
       " 'ENSRNOG00000014630',\n",
       " 'ENSRNOG00000036827',\n",
       " 'ENSRNOG00000011168',\n",
       " 'ENSRNOG00000024364',\n",
       " 'ENSRNOG00000009388',\n",
       " 'ENSRNOG00000058394',\n",
       " 'ENSRNOG00000036826',\n",
       " 'ENSRNOG00000002323',\n",
       " 'ENSRNOG00000017867',\n",
       " 'ENSRNOG00000006998',\n",
       " 'ENSRNOG00000069097',\n",
       " 'ENSRNOG00000054350',\n",
       " 'ENSRNOG00000009284',\n",
       " 'ENSRNOG00000067167',\n",
       " 'ENSRNOG00000013152',\n",
       " 'ENSRNOG00000011471',\n",
       " 'ENSRNOG00000021186',\n",
       " 'ENSRNOG00000022791',\n",
       " 'ENSRNOG00000013886',\n",
       " 'ENSRNOG00000006978',\n",
       " 'ENSRNOG00000066403',\n",
       " 'ENSRNOG00000040136',\n",
       " 'ENSRNOG00000011873',\n",
       " 'ENSRNOG00000006622',\n",
       " 'ENSRNOG00000012141',\n",
       " 'ENSRNOG00000023637',\n",
       " 'ENSRNOG00000020893',\n",
       " 'ENSRNOG00000067086',\n",
       " 'ENSRNOG00000030871',\n",
       " 'ENSRNOG00000006304',\n",
       " 'ENSRNOG00000004657',\n",
       " 'ENSRNOG00000037827',\n",
       " 'ENSRNOG00000065715',\n",
       " 'ENSRNOG00000014089',\n",
       " 'ENSRNOG00000002449',\n",
       " 'ENSRNOG00000052375',\n",
       " 'ENSRNOG00000084041',\n",
       " 'ENSRNOG00000091132',\n",
       " 'ENSRNOG00000017752',\n",
       " 'ENSRNOG00000054705',\n",
       " 'ENSRNOG00000009001',\n",
       " 'ENSRNOG00000016396',\n",
       " 'ENSRNOG00000060935',\n",
       " 'ENSRNOG00000053891',\n",
       " 'ENSRNOG00000078984',\n",
       " 'ENSRNOG00000079764',\n",
       " 'ENSRNOG00000013180',\n",
       " 'ENSRNOG00000004360',\n",
       " 'ENSRNOG00000032745',\n",
       " 'ENSRNOG00000010259',\n",
       " 'ENSRNOG00000061675',\n",
       " 'ENSRNOG00000029939',\n",
       " 'ENSRNOG00000020923',\n",
       " 'ENSRNOG00000057432',\n",
       " 'ENSRNOG00000058856',\n",
       " 'ENSRNOG00000009550',\n",
       " 'ENSRNOG00000014502',\n",
       " 'ENSRNOG00000027355',\n",
       " 'ENSRNOG00000017953',\n",
       " 'ENSRNOG00000021680',\n",
       " 'ENSRNOG00000027622',\n",
       " 'ENSRNOG00000023053',\n",
       " 'ENSRNOG00000056084',\n",
       " 'ENSRNOG00000028299',\n",
       " 'ENSRNOG00000002291',\n",
       " 'ENSRNOG00000047605',\n",
       " 'ENSRNOG00000012185',\n",
       " 'ENSRNOG00000009585',\n",
       " 'ENSRNOG00000051896',\n",
       " 'ENSRNOG00000012928',\n",
       " 'ENSRNOG00000042692',\n",
       " 'ENSRNOG00000069859',\n",
       " 'ENSRNOG00000040110',\n",
       " 'ENSRNOG00000011341',\n",
       " 'ENSRNOG00000016528',\n",
       " 'ENSRNOG00000002013',\n",
       " 'ENSRNOG00000020478',\n",
       " 'ENSRNOG00000031067',\n",
       " 'ENSRNOG00000086890',\n",
       " 'ENSRNOG00000061629',\n",
       " 'ENSRNOG00000056965',\n",
       " 'ENSRNOG00000067356',\n",
       " 'ENSRNOG00000063726',\n",
       " 'ENSRNOG00000020468',\n",
       " 'ENSRNOG00000068728',\n",
       " 'ENSRNOG00000038743',\n",
       " 'ENSRNOG00000004103',\n",
       " 'ENSRNOG00000018000',\n",
       " 'ENSRNOG00000002509',\n",
       " 'ENSRNOG00000068914',\n",
       " 'ENSRNOG00000011603',\n",
       " 'ENSRNOG00000022505',\n",
       " 'ENSRNOG00000016688',\n",
       " 'ENSRNOG00000012653',\n",
       " 'ENSRNOG00000058981',\n",
       " 'ENSRNOG00000026050',\n",
       " 'ENSRNOG00000016703',\n",
       " 'ENSRNOG00000012860',\n",
       " 'ENSRNOG00000009821',\n",
       " 'ENSRNOG00000084925',\n",
       " 'ENSRNOG00000009826',\n",
       " 'ENSRNOG00000065352',\n",
       " 'ENSRNOG00000017864',\n",
       " 'ENSRNOG00000004061',\n",
       " 'ENSRNOG00000016712',\n",
       " 'ENSRNOG00000061012',\n",
       " 'ENSRNOG00000009792',\n",
       " 'ENSRNOG00000012209',\n",
       " 'ENSRNOG00000018954',\n",
       " 'ENSRNOG00000056005',\n",
       " 'ENSRNOG00000061115',\n",
       " 'ENSRNOG00000068584',\n",
       " 'ENSRNOG00000029078',\n",
       " 'ENSRNOG00000064141',\n",
       " 'ENSRNOG00000065214',\n",
       " 'ENSRNOG00000016783',\n",
       " 'ENSRNOG00000038784',\n",
       " 'ENSRNOG00000014979',\n",
       " 'ENSRNOG00000011026',\n",
       " 'ENSRNOG00000066701',\n",
       " 'ENSRNOG00000068254',\n",
       " 'ENSRNOG00000070574',\n",
       " 'ENSRNOG00000015329',\n",
       " 'ENSRNOG00000069417',\n",
       " 'ENSRNOG00000030058',\n",
       " 'ENSRNOG00000091232',\n",
       " 'ENSRNOG00000075610',\n",
       " 'ENSRNOG00000070375',\n",
       " 'ENSRNOG00000065350',\n",
       " 'ENSRNOG00000011803',\n",
       " 'ENSRNOG00000069303',\n",
       " 'ENSRNOG00000053322',\n",
       " 'ENSRNOG00000009739',\n",
       " 'ENSRNOG00000011784',\n",
       " 'ENSRNOG00000011111',\n",
       " 'ENSRNOG00000039024',\n",
       " 'ENSRNOG00000012721',\n",
       " 'ENSRNOG00000028085',\n",
       " 'ENSRNOG00000057922',\n",
       " 'ENSRNOG00000024178',\n",
       " 'ENSRNOG00000009740',\n",
       " 'ENSRNOG00000009903',\n",
       " 'ENSRNOG00000011285',\n",
       " 'ENSRNOG00000012521',\n",
       " 'ENSRNOG00000005758',\n",
       " 'ENSRNOG00000008053',\n",
       " 'ENSRNOG00000053631',\n",
       " 'ENSRNOG00000024628',\n",
       " 'ENSRNOG00000028077',\n",
       " 'ENSRNOG00000023005',\n",
       " 'ENSRNOG00000019109',\n",
       " 'ENSRNOG00000012342',\n",
       " 'ENSRNOG00000053945',\n",
       " 'ENSRNOG00000010147',\n",
       " 'ENSRNOG00000017579',\n",
       " 'ENSRNOG00000038970',\n",
       " 'ENSRNOG00000016351',\n",
       " 'ENSRNOG00000012934',\n",
       " 'ENSRNOG00000086235',\n",
       " 'ENSRNOG00000070192',\n",
       " 'ENSRNOG00000082337',\n",
       " 'ENSRNOG00000072234',\n",
       " 'ENSRNOG00000091012',\n",
       " 'ENSRNOG00000068499',\n",
       " 'ENSRNOG00000084162',\n",
       " 'ENSRNOG00000087441',\n",
       " 'ENSRNOG00000088946',\n",
       " 'ENSRNOG00000003964',\n",
       " 'ENSRNOG00000002021',\n",
       " 'ENSRNOG00000061286',\n",
       " 'ENSRNOG00000012644',\n",
       " 'ENSRNOG00000016812',\n",
       " 'ENSRNOG00000067918',\n",
       " 'ENSRNOG00000058403',\n",
       " 'ENSRNOG00000073233',\n",
       " 'ENSRNOG00000073793',\n",
       " 'ENSRNOG00000064041',\n",
       " 'ENSRNOG00000011453',\n",
       " 'ENSRNOG00000088307',\n",
       " 'ENSRNOG00000085450',\n",
       " 'ENSRNOG00000082992',\n",
       " 'ENSRNOG00000083547',\n",
       " 'ENSRNOG00000077227',\n",
       " 'ENSRNOG00000072664',\n",
       " 'ENSRNOG00000072494',\n",
       " 'ENSRNOG00000078118',\n",
       " 'ENSRNOG00000078951',\n",
       " 'ENSRNOG00000091139',\n",
       " 'ENSRNOG00000004100',\n",
       " 'ENSRNOG00000030538',\n",
       " 'ENSRNOG00000009149',\n",
       " 'ENSRNOG00000052568',\n",
       " 'ENSRNOG00000002784',\n",
       " 'ENSRNOG00000022975',\n",
       " 'ENSRNOG00000007059',\n",
       " 'ENSRNOG00000043031',\n",
       " 'ENSRNOG00000011193',\n",
       " 'ENSRNOG00000058609',\n",
       " 'ENSRNOG00000011321',\n",
       " 'ENSRNOG00000012843',\n",
       " 'ENSRNOG00000072617',\n",
       " 'ENSRNOG00000079673',\n",
       " 'ENSRNOG00000088515',\n",
       " 'ENSRNOG00000073662',\n",
       " 'ENSRNOG00000070759',\n",
       " 'ENSRNOG00000030888',\n",
       " 'ENSRNOG00000011261',\n",
       " 'ENSRNOG00000005350',\n",
       " 'ENSRNOG00000064917',\n",
       " 'ENSRNOG00000023474',\n",
       " 'ENSRNOG00000016628',\n",
       " 'ENSRNOG00000060694',\n",
       " 'ENSRNOG00000032063',\n",
       " 'ENSRNOG00000000164',\n",
       " 'ENSRNOG00000047493',\n",
       " 'ENSRNOG00000017945',\n",
       " 'ENSRNOG00000007048',\n",
       " 'ENSRNOG00000066432',\n",
       " 'ENSRNOG00000063623',\n",
       " 'ENSRNOG00000031782',\n",
       " 'ENSRNOG00000086182',\n",
       " 'ENSRNOG00000071391',\n",
       " 'ENSRNOG00000013368',\n",
       " 'ENSRNOG00000002041',\n",
       " 'ENSRNOG00000017250',\n",
       " 'ENSRNOG00000004914',\n",
       " 'ENSRNOG00000051459',\n",
       " 'ENSRNOG00000033192',\n",
       " 'ENSRNOG00000050220',\n",
       " 'ENSRNOG00000085300',\n",
       " 'ENSRNOG00000012324',\n",
       " 'ENSRNOG00000063329',\n",
       " 'ENSRNOG00000064085',\n",
       " 'ENSRNOG00000027491',\n",
       " 'ENSRNOG00000070810',\n",
       " 'ENSRNOG00000087029',\n",
       " 'ENSRNOG00000090125',\n",
       " 'ENSRNOG00000089256',\n",
       " 'ENSRNOG00000082826',\n",
       " 'ENSRNOG00000073607',\n",
       " 'ENSRNOG00000077846',\n",
       " 'ENSRNOG00000090224',\n",
       " 'ENSRNOG00000074844',\n",
       " 'ENSRNOG00000091374',\n",
       " 'ENSRNOG00000072848',\n",
       " 'ENSRNOG00000089779',\n",
       " 'ENSRNOG00000087183',\n",
       " 'ENSRNOG00000077570',\n",
       " 'ENSRNOG00000089966',\n",
       " 'ENSRNOG00000036984',\n",
       " 'ENSRNOG00000011636',\n",
       " 'ENSRNOG00000033609',\n",
       " 'ENSRNOG00000000395',\n",
       " 'ENSRNOG00000018067',\n",
       " 'ENSRNOG00000011251',\n",
       " 'ENSRNOG00000059491',\n",
       " 'ENSRNOG00000055183',\n",
       " 'ENSRNOG00000062231',\n",
       " 'ENSRNOG00000004744',\n",
       " 'ENSRNOG00000012742',\n",
       " 'ENSRNOG00000004692',\n",
       " 'ENSRNOG00000072867',\n",
       " 'ENSRNOG00000091185',\n",
       " 'ENSRNOG00000012720',\n",
       " 'ENSRNOG00000069372',\n",
       " 'ENSRNOG00000057970',\n",
       " 'ENSRNOG00000006920',\n",
       " 'ENSRNOG00000031595',\n",
       " 'ENSRNOG00000022932',\n",
       " 'ENSRNOG00000046547',\n",
       " 'ENSRNOG00000075868',\n",
       " 'ENSRNOG00000054221',\n",
       " 'ENSRNOG00000088882',\n",
       " 'ENSRNOG00000052215',\n",
       " 'ENSRNOG00000066388',\n",
       " 'ENSRNOG00000076665',\n",
       " 'ENSRNOG00000066470',\n",
       " 'ENSRNOG00000074947',\n",
       " 'ENSRNOG00000071923',\n",
       " 'ENSRNOG00000068347',\n",
       " 'ENSRNOG00000077085',\n",
       " 'ENSRNOG00000086367',\n",
       " 'ENSRNOG00000067056',\n",
       " 'ENSRNOG00000083108',\n",
       " 'ENSRNOG00000079222',\n",
       " 'ENSRNOG00000074614',\n",
       " 'ENSRNOG00000081990',\n",
       " 'ENSRNOG00000085879',\n",
       " 'ENSRNOG00000063035',\n",
       " 'ENSRNOG00000059999',\n",
       " 'ENSRNOG00000061288',\n",
       " 'ENSRNOG00000020423',\n",
       " 'ENSRNOG00000023809',\n",
       " 'ENSRNOG00000018068',\n",
       " 'ENSRNOG00000002585',\n",
       " 'ENSRNOG00000042309',\n",
       " 'ENSRNOG00000068381',\n",
       " 'ENSRNOG00000065160',\n",
       " 'ENSRNOG00000023407',\n",
       " 'ENSRNOG00000064389',\n",
       " 'ENSRNOG00000043099',\n",
       " 'ENSRNOG00000043350',\n",
       " 'ENSRNOG00000046307',\n",
       " 'ENSRNOG00000010388',\n",
       " 'ENSRNOG00000004962',\n",
       " 'ENSRNOG00000058120',\n",
       " 'ENSRNOG00000026120',\n",
       " 'ENSRNOG00000004500',\n",
       " 'ENSRNOG00000065128',\n",
       " 'ENSRNOG00000033693',\n",
       " 'ENSRNOG00000018945',\n",
       " 'ENSRNOG00000057750',\n",
       " 'ENSRNOG00000057624',\n",
       " 'ENSRNOG00000030894',\n",
       " 'ENSRNOG00000066437',\n",
       " 'ENSRNOG00000030954',\n",
       " 'ENSRNOG00000054576',\n",
       " 'ENSRNOG00000086355',\n",
       " 'ENSRNOG00000000396',\n",
       " 'ENSRNOG00000012918',\n",
       " 'ENSRNOG00000011334',\n",
       " 'ENSRNOG00000016076',\n",
       " 'ENSRNOG00000011696',\n",
       " 'ENSRNOG00000048441',\n",
       " 'ENSRNOG00000025541',\n",
       " 'ENSRNOG00000004813',\n",
       " 'ENSRNOG00000013623',\n",
       " 'ENSRNOG00000025519',\n",
       " ...]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rat_ensembl_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5ef4a15-8448-4964-8a78-b9da4e47ce5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Step 5: Merge rat RGD IDs back\n",
    "final_mapping = pd.merge(\n",
    "    mouse_to_rat,\n",
    "    rat_ensembl_to_rgd,\n",
    "    left_on='Rat homolog Ensembl gene',\n",
    "    right_on='Gene stable ID',\n",
    "    how='left'\n",
    ")\n",
    "\n",
    "# Clean column names for clarity\n",
    "final_mapping.rename(columns={\n",
    "    'Gene stable ID_x': 'mouse_ensembl_gene_id',\n",
    "    'Gene stable ID_y': 'rat_ensembl_gene_id',\n",
    "    'RGD ID': 'rgd_id'\n",
    "}, inplace=True)\n",
    "\n",
    "print(final_mapping.head())\n",
    "\n",
    "# Save if desired\n",
    "final_mapping.to_csv(\"mouse_to_rat_RGD_mapping.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd812a55-0501-43c7-b8ad-86e6cdf255a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "print(mouse_to_rat)\n",
    "\n",
    "# Extract rat Ensembl gene IDs, drop empty\n",
    "rat_ensembl_ids = mouse_to_rat['Rat homolog Ensembl gene'].dropna().unique().tolist()\n",
    "\n",
    "if rat_ensembl_ids:\n",
    "    # Step 2: Map rat Ensembl gene IDs to RGD IDs and symbols\n",
    "    rat_to_rgd = rat_dataset.query(\n",
    "        attributes=['ensembl_gene_id', 'rgd_id', 'rgd_symbol'],\n",
    "        filters={'stable_gene_id': rat_ensembl_ids}\n",
    "    )\n",
    "    print(\"Rat Ensembl to RGD IDs:\")\n",
    "    print(rat_to_rgd)\n",
    "else:\n",
    "    print(\"No rat homologs found for the provided MGI IDs.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b984400-ea70-48e9-bdb9-f1cffef2d4ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "MGI:1196250\n",
    "MGI:1196250\n",
    "MGI:1196250\n",
    "MGI:2136754\n",
    "MGI:2136755\n",
    "MGI:2136756\n",
    "MGI:2136758\n",
    "MGI:2136759\n",
    "MGI:2136760\n",
    "MGI:2136720\n",
    "MGI:1860250\n",
    "MGI:96824\n",
    "MGI:1919379"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3985f620-b591-4c64-a97e-c7f48c3c301a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>MGI ID</th>\n",
       "      <th>MGI symbol</th>\n",
       "      <th>Norway rat - BN/NHsdMcwi gene stable ID</th>\n",
       "      <th>Norway rat - BN/NHsdMcwi gene name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000064341</td>\n",
       "      <td>MGI:101787</td>\n",
       "      <td>mt-Nd1</td>\n",
       "      <td>ENSRNOG00000030644</td>\n",
       "      <td>Mt-nd1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSMUSG00000064345</td>\n",
       "      <td>MGI:102500</td>\n",
       "      <td>mt-Nd2</td>\n",
       "      <td>ENSRNOG00000031033</td>\n",
       "      <td>Mt-nd2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSMUSG00000064351</td>\n",
       "      <td>MGI:102504</td>\n",
       "      <td>mt-Co1</td>\n",
       "      <td>ENSRNOG00000034234</td>\n",
       "      <td>Mt-co1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSMUSG00000064354</td>\n",
       "      <td>MGI:102503</td>\n",
       "      <td>mt-Co2</td>\n",
       "      <td>ENSRNOG00000030371</td>\n",
       "      <td>Mt-co2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENSMUSG00000064356</td>\n",
       "      <td>MGI:99926</td>\n",
       "      <td>mt-Atp8</td>\n",
       "      <td>ENSRNOG00000033299</td>\n",
       "      <td>Mt-atp8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97931</th>\n",
       "      <td>ENSMUSG00000070943</td>\n",
       "      <td>MGI:3030190</td>\n",
       "      <td>Or1ak2</td>\n",
       "      <td>ENSRNOG00000007962</td>\n",
       "      <td>Or1ak2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97932</th>\n",
       "      <td>ENSMUSG00000055838</td>\n",
       "      <td>MGI:3030191</td>\n",
       "      <td>Or1q1</td>\n",
       "      <td>ENSRNOG00000007965</td>\n",
       "      <td>Or1q1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97933</th>\n",
       "      <td>ENSMUSG00000075379</td>\n",
       "      <td>MGI:3030192</td>\n",
       "      <td>Or12k5</td>\n",
       "      <td>ENSRNOG00000079606</td>\n",
       "      <td>Or12k5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97936</th>\n",
       "      <td>ENSMUSG00000083361</td>\n",
       "      <td>MGI:3030194</td>\n",
       "      <td>Or12k7</td>\n",
       "      <td>ENSRNOG00000032102</td>\n",
       "      <td>Or12k7b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97937</th>\n",
       "      <td>ENSMUSG00000083361</td>\n",
       "      <td>MGI:3030194</td>\n",
       "      <td>Or12k7</td>\n",
       "      <td>ENSRNOG00000031803</td>\n",
       "      <td>Or12k7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41893 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene stable ID       MGI ID MGI symbol  \\\n",
       "5      ENSMUSG00000064341   MGI:101787     mt-Nd1   \n",
       "9      ENSMUSG00000064345   MGI:102500     mt-Nd2   \n",
       "15     ENSMUSG00000064351   MGI:102504     mt-Co1   \n",
       "18     ENSMUSG00000064354   MGI:102503     mt-Co2   \n",
       "20     ENSMUSG00000064356    MGI:99926    mt-Atp8   \n",
       "...                   ...          ...        ...   \n",
       "97931  ENSMUSG00000070943  MGI:3030190     Or1ak2   \n",
       "97932  ENSMUSG00000055838  MGI:3030191      Or1q1   \n",
       "97933  ENSMUSG00000075379  MGI:3030192     Or12k5   \n",
       "97936  ENSMUSG00000083361  MGI:3030194     Or12k7   \n",
       "97937  ENSMUSG00000083361  MGI:3030194     Or12k7   \n",
       "\n",
       "      Norway rat - BN/NHsdMcwi gene stable ID  \\\n",
       "5                          ENSRNOG00000030644   \n",
       "9                          ENSRNOG00000031033   \n",
       "15                         ENSRNOG00000034234   \n",
       "18                         ENSRNOG00000030371   \n",
       "20                         ENSRNOG00000033299   \n",
       "...                                       ...   \n",
       "97931                      ENSRNOG00000007962   \n",
       "97932                      ENSRNOG00000007965   \n",
       "97933                      ENSRNOG00000079606   \n",
       "97936                      ENSRNOG00000032102   \n",
       "97937                      ENSRNOG00000031803   \n",
       "\n",
       "      Norway rat - BN/NHsdMcwi gene name  \n",
       "5                                 Mt-nd1  \n",
       "9                                 Mt-nd2  \n",
       "15                                Mt-co1  \n",
       "18                                Mt-co2  \n",
       "20                               Mt-atp8  \n",
       "...                                  ...  \n",
       "97931                             Or1ak2  \n",
       "97932                              Or1q1  \n",
       "97933                             Or12k5  \n",
       "97936                            Or12k7b  \n",
       "97937                             Or12k7  \n",
       "\n",
       "[41893 rows x 5 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse_to_rat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ca397f53-a74e-40d2-83d7-1f4f5852be09",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'Rat homolog Ensembl gene'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/indexes/base.py:3805\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3804\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3805\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3806\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32mindex.pyx:167\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mindex.pyx:196\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7081\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7089\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Rat homolog Ensembl gene'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[4], line 31\u001b[0m\n\u001b[1;32m     28\u001b[0m mouse_to_rat \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mmerge(mouse_genes, mouse_rat_orthologs, on\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGene stable ID\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m     30\u001b[0m \u001b[38;5;66;03m# Filter out entries without rat ortholog\u001b[39;00m\n\u001b[0;32m---> 31\u001b[0m mouse_to_rat \u001b[38;5;241m=\u001b[39m mouse_to_rat[\u001b[43mmouse_to_rat\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mRat homolog Ensembl gene\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mnotna() \u001b[38;5;241m&\u001b[39m (mouse_to_rat[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mRat homolog Ensembl gene\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m!=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m)]\n\u001b[1;32m     33\u001b[0m \u001b[38;5;66;03m# Step 4: Retrieve RGD IDs for rat ortholog Ensembl gene IDs\u001b[39;00m\n\u001b[1;32m     34\u001b[0m rat_ensembl_ids \u001b[38;5;241m=\u001b[39m mouse_to_rat[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mRat homolog Ensembl gene\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39munique()\u001b[38;5;241m.\u001b[39mtolist()\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/frame.py:4102\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   4100\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   4101\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 4102\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4103\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   4104\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/indexes/base.py:3812\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3807\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[1;32m   3808\u001b[0m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[1;32m   3809\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[1;32m   3810\u001b[0m     ):\n\u001b[1;32m   3811\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[0;32m-> 3812\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3813\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3814\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3815\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3816\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3817\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Rat homolog Ensembl gene'"
     ]
    }
   ],
   "source": [
    "\n",
    "# Step 5: Merge rat RGD IDs back\n",
    "final_mapping = pd.merge(\n",
    "    mouse_to_rat,\n",
    "    rat_ensembl_to_rgd,\n",
    "    left_on='Rat homolog Ensembl gene',\n",
    "    right_on='Gene stable ID',\n",
    "    how='left'\n",
    ")\n",
    "\n",
    "# Clean column names for clarity\n",
    "final_mapping.rename(columns={\n",
    "    'Gene stable ID_x': 'mouse_ensembl_gene_id',\n",
    "    'Gene stable ID_y': 'rat_ensembl_gene_id',\n",
    "    'RGD ID': 'rgd_id'\n",
    "}, inplace=True)\n",
    "\n",
    "print(final_mapping.head())\n",
    "\n",
    "# Save if desired\n",
    "final_mapping.to_csv(\"data/mouse_to_rat_mapping.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fa55a887-8dcd-4bd8-b522-2307e99e17f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AsyncClient', 'ChatResponse', 'Client', 'EmbedResponse', 'EmbeddingsResponse', 'GenerateResponse', 'Image', 'ListResponse', 'Message', 'Options', 'ProcessResponse', 'ProgressResponse', 'RequestError', 'ResponseError', 'ShowResponse', 'StatusResponse', 'Tool', '__all__', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', '_client', '_types', '_utils', 'chat', 'copy', 'create', 'delete', 'embed', 'embeddings', 'generate', 'list', 'ps', 'pull', 'push', 'show']\n"
     ]
    }
   ],
   "source": [
    "import ollama\n",
    "print(dir(ollama))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b566a2c2-bf2d-4fe4-a896-e1119cf2865e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;span title=\"Double-click header of Interaction ID to ensure all values are shown\"&gt;Interaction ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\"&gt;Human LR Pair&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Official Gene Symbol; Hover on symbols below to show gene names\"&gt;Ligand&amp;nbsp;&amp;nbsp;&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Official Gene Symbol; Hover on symbols below to show gene names\"&gt;Receptor&amp;nbsp;&amp;nbsp;&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Ligand Symbols to ensure all values are shown\"&gt;Ligand Symbols&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Receptor Symbols to ensure all values are shown\"&gt;Receptor Symbols&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\"&gt;Ligand Location&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\"&gt;Receptor Location&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\"&gt;Ligand HGNC ID&amp;nbsp;&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\"&gt;Receptor HGNC ID&amp;nbsp;&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>...</th>\n",
       "      <th>&lt;span title=\"Double-click header of Sheep Receptor Ensembl ID to ensure all values are shown\"&gt;Sheep Receptor Ensembl ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Sheep Receptor to ensure all values are shown\"&gt;Sheep Receptor&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Marmoset Ligand Ensembl ID to ensure all values are shown\"&gt;Marmoset Ligand Ensembl ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Marmoset Ligand to ensure all values are shown\"&gt;Marmoset Ligand&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Marmoset Receptor Ensembl ID to ensure all values are shown\"&gt;Marmoset Receptor Ensembl ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Marmoset Receptor to ensure all values are shown\"&gt;Marmoset Receptor&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Macaque Ligand Ensembl ID to ensure all values are shown\"&gt;Macaque Ligand Ensembl ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Macaque Ligand to ensure all values are shown\"&gt;Macaque Ligand&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Macaque Receptor Ensembl ID to ensure all values are shown\"&gt;Macaque Receptor Ensembl ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Macaque Receptor to ensure all values are shown\"&gt;Macaque Receptor&amp;nbsp;&lt;/span&gt;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDB00001</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\"alpha-2-macroglobulin\"&gt;A2M&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"heat shock protein family A (Hsp7...</td>\n",
       "      <td>&lt;span title=\"A2M (FWP007, S863-7, CPAMD5)\"&gt;A2M...</td>\n",
       "      <td>&lt;span title=\"HSPA5 (GRP78, BiP)\"&gt;HSPA5 (GRP78,...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;span title=\"based on PMID: 12194978, PMID: 32...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSOARG00020007714</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSCJAG00000006880</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSMMUG00000005217</td>\n",
       "      <td>HSPA5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDB00002</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\"alpha-2-macroglobulin\"&gt;A2M&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"LDL receptor related protein 1\"&gt;L...</td>\n",
       "      <td>&lt;span title=\"A2M (FWP007, S863-7, CPAMD5)\"&gt;A2M...</td>\n",
       "      <td>&lt;span title=\"LRP1 (APR, A2MR, LRP, CD91, LRP1A...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSOARG00020001685</td>\n",
       "      <td>LRP1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSCJAG00000010289</td>\n",
       "      <td>LRP1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSMMUG00000021782</td>\n",
       "      <td>LRP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDB00003</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\"angiotensin I converting enzyme\"&gt;...</td>\n",
       "      <td>&lt;span title=\"bradykinin receptor B2\"&gt;BDKRB2&lt;/s...</td>\n",
       "      <td>&lt;span title=\"ACE (DCP1, ACE1, CD143)\"&gt;ACE (DCP...</td>\n",
       "      <td>&lt;span title=\"BDKRB2 (BK-2)\"&gt;BDKRB2 (BK-2)&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSOARG00020013146</td>\n",
       "      <td>BDKRB2</td>\n",
       "      <td>ENSCJAG00000019755</td>\n",
       "      <td>ACE</td>\n",
       "      <td>ENSCJAG00000006399</td>\n",
       "      <td>BDKRB2</td>\n",
       "      <td>ENSMMUG00000012608</td>\n",
       "      <td>ACE</td>\n",
       "      <td>ENSMMUG00000043275</td>\n",
       "      <td>BDKRB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDB00004</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\"adenosine deaminase\"&gt;ADA&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"dipeptidyl peptidase 4\"&gt;DPP4&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"ADA (ADA1)\"&gt;ADA (ADA1)&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"DPP4 (CD26, ADCP2, DPPIV)\"&gt;DPP4 (...</td>\n",
       "      <td>&lt;span title=\"based on hpa, uniprot\"&gt;cell membr...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSOARG00020012079</td>\n",
       "      <td>DPP4</td>\n",
       "      <td>ENSCJAG00000017705</td>\n",
       "      <td>ADA</td>\n",
       "      <td>ENSCJAG00000005739</td>\n",
       "      <td>DPP4</td>\n",
       "      <td>ENSMMUG00000010603</td>\n",
       "      <td>ADA</td>\n",
       "      <td>ENSMMUG00000008387</td>\n",
       "      <td>DPP4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDB00005</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\"ADAM metallopeptidase domain 10\"&gt;...</td>\n",
       "      <td>&lt;span title=\"EPH receptor A3\"&gt;EPHA3&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"ADAM10 (kuz, MADM, HsT18717, CD15...</td>\n",
       "      <td>&lt;span title=\"EPHA3 (ETK, ETK1, TYRO4, HEK, HEK...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSOARG00020002625</td>\n",
       "      <td>EPHA3</td>\n",
       "      <td>ENSCJAG00000020430</td>\n",
       "      <td>ADAM10</td>\n",
       "      <td>ENSCJAG00000008692</td>\n",
       "      <td>EPHA3</td>\n",
       "      <td>ENSMMUG00000017507</td>\n",
       "      <td>ADAM10</td>\n",
       "      <td>ENSMMUG00000008242</td>\n",
       "      <td>EPHA3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3444</th>\n",
       "      <td>CDB03445</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhb22&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhb22&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"Pcdhb22\"&gt;Pcdhb22&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"Pcdhb22\"&gt;Pcdhb22&lt;/span&gt;</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3445</th>\n",
       "      <td>CDB03446</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhgb8&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhgb8&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"Pcdhgb8\"&gt;Pcdhgb8&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"Pcdhgb8\"&gt;Pcdhgb8&lt;/span&gt;</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3446</th>\n",
       "      <td>CDB03447</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\" \"&gt;Saa3&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\" \"&gt;Ly96&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"Saa3\"&gt;Saa3&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"Ly96\"&gt;Ly96&lt;/span&gt;</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3447</th>\n",
       "      <td>CDB03448</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\" \"&gt;Saa3&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\" \"&gt;Tlr4&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"Saa3\"&gt;Saa3&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"Tlr4\"&gt;Tlr4&lt;/span&gt;</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3448</th>\n",
       "      <td>CDB03449</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\" \"&gt;Sema4a&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\" \"&gt;Timd2&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"Sema4a\"&gt;Sema4a&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"Timd2\"&gt;Timd2&lt;/span&gt;</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3449 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     <span title=\"Double-click header of Interaction ID to ensure all values are shown\">Interaction ID&nbsp;</span>  \\\n",
       "0                                              CDB00001                                                               \n",
       "1                                              CDB00002                                                               \n",
       "2                                              CDB00003                                                               \n",
       "3                                              CDB00004                                                               \n",
       "4                                              CDB00005                                                               \n",
       "...                                                 ...                                                               \n",
       "3444                                           CDB03445                                                               \n",
       "3445                                           CDB03446                                                               \n",
       "3446                                           CDB03447                                                               \n",
       "3447                                           CDB03448                                                               \n",
       "3448                                           CDB03449                                                               \n",
       "\n",
       "     <span title=\"Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\">Human LR Pair</span>  \\\n",
       "0     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                               \n",
       "1     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                               \n",
       "2     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                               \n",
       "3     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                               \n",
       "4     <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                               \n",
       "...                                                 ...                                                               \n",
       "3444  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                               \n",
       "3445  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                               \n",
       "3446  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                               \n",
       "3447  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                               \n",
       "3448  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...                                                               \n",
       "\n",
       "     <span title=\"Official Gene Symbol; Hover on symbols below to show gene names\">Ligand&nbsp;&nbsp;&nbsp;</span>  \\\n",
       "0        <span title=\"alpha-2-macroglobulin\">A2M</span>                                                              \n",
       "1        <span title=\"alpha-2-macroglobulin\">A2M</span>                                                              \n",
       "2     <span title=\"angiotensin I converting enzyme\">...                                                              \n",
       "3          <span title=\"adenosine deaminase\">ADA</span>                                                              \n",
       "4     <span title=\"ADAM metallopeptidase domain 10\">...                                                              \n",
       "...                                                 ...                                                              \n",
       "3444                     <span title=\" \">Pcdhb22</span>                                                              \n",
       "3445                     <span title=\" \">Pcdhgb8</span>                                                              \n",
       "3446                        <span title=\" \">Saa3</span>                                                              \n",
       "3447                        <span title=\" \">Saa3</span>                                                              \n",
       "3448                      <span title=\" \">Sema4a</span>                                                              \n",
       "\n",
       "     <span title=\"Official Gene Symbol; Hover on symbols below to show gene names\">Receptor&nbsp;&nbsp;&nbsp;</span>  \\\n",
       "0     <span title=\"heat shock protein family A (Hsp7...                                                                \n",
       "1     <span title=\"LDL receptor related protein 1\">L...                                                                \n",
       "2     <span title=\"bradykinin receptor B2\">BDKRB2</s...                                                                \n",
       "3      <span title=\"dipeptidyl peptidase 4\">DPP4</span>                                                                \n",
       "4            <span title=\"EPH receptor A3\">EPHA3</span>                                                                \n",
       "...                                                 ...                                                                \n",
       "3444                     <span title=\" \">Pcdhb22</span>                                                                \n",
       "3445                     <span title=\" \">Pcdhgb8</span>                                                                \n",
       "3446                        <span title=\" \">Ly96</span>                                                                \n",
       "3447                        <span title=\" \">Tlr4</span>                                                                \n",
       "3448                       <span title=\" \">Timd2</span>                                                                \n",
       "\n",
       "     <span title=\"Double-click header of Ligand Symbols to ensure all values are shown\">Ligand Symbols&nbsp;</span>  \\\n",
       "0     <span title=\"A2M (FWP007, S863-7, CPAMD5)\">A2M...                                                               \n",
       "1     <span title=\"A2M (FWP007, S863-7, CPAMD5)\">A2M...                                                               \n",
       "2     <span title=\"ACE (DCP1, ACE1, CD143)\">ACE (DCP...                                                               \n",
       "3            <span title=\"ADA (ADA1)\">ADA (ADA1)</span>                                                               \n",
       "4     <span title=\"ADAM10 (kuz, MADM, HsT18717, CD15...                                                               \n",
       "...                                                 ...                                                               \n",
       "3444               <span title=\"Pcdhb22\">Pcdhb22</span>                                                               \n",
       "3445               <span title=\"Pcdhgb8\">Pcdhgb8</span>                                                               \n",
       "3446                     <span title=\"Saa3\">Saa3</span>                                                               \n",
       "3447                     <span title=\"Saa3\">Saa3</span>                                                               \n",
       "3448                 <span title=\"Sema4a\">Sema4a</span>                                                               \n",
       "\n",
       "     <span title=\"Double-click header of Receptor Symbols to ensure all values are shown\">Receptor Symbols&nbsp;</span>  \\\n",
       "0     <span title=\"HSPA5 (GRP78, BiP)\">HSPA5 (GRP78,...                                                                   \n",
       "1     <span title=\"LRP1 (APR, A2MR, LRP, CD91, LRP1A...                                                                   \n",
       "2      <span title=\"BDKRB2 (BK-2)\">BDKRB2 (BK-2)</span>                                                                   \n",
       "3     <span title=\"DPP4 (CD26, ADCP2, DPPIV)\">DPP4 (...                                                                   \n",
       "4     <span title=\"EPHA3 (ETK, ETK1, TYRO4, HEK, HEK...                                                                   \n",
       "...                                                 ...                                                                   \n",
       "3444               <span title=\"Pcdhb22\">Pcdhb22</span>                                                                   \n",
       "3445               <span title=\"Pcdhgb8\">Pcdhgb8</span>                                                                   \n",
       "3446                     <span title=\"Ly96\">Ly96</span>                                                                   \n",
       "3447                     <span title=\"Tlr4\">Tlr4</span>                                                                   \n",
       "3448                   <span title=\"Timd2\">Timd2</span>                                                                   \n",
       "\n",
       "     <span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\">Ligand Location</span>  \\\n",
       "0     <span title=\"based on perplexity, uniprot\">sec...                                                                                                                        \n",
       "1     <span title=\"based on perplexity, uniprot\">sec...                                                                                                                        \n",
       "2     <span title=\"based on perplexity, uniprot\">cel...                                                                                                                        \n",
       "3     <span title=\"based on hpa, uniprot\">cell membr...                                                                                                                        \n",
       "4     <span title=\"based on hpa, perplexity, uniprot...                                                                                                                        \n",
       "...                                                 ...                                                                                                                        \n",
       "3444                                            unknown                                                                                                                        \n",
       "3445                                            unknown                                                                                                                        \n",
       "3446                                            unknown                                                                                                                        \n",
       "3447                                            unknown                                                                                                                        \n",
       "3448                                            unknown                                                                                                                        \n",
       "\n",
       "     <span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\">Receptor Location</span>  \\\n",
       "0     <span title=\"based on PMID: 12194978, PMID: 32...                                                                                                                          \n",
       "1     <span title=\"based on hpa, perplexity, uniprot...                                                                                                                          \n",
       "2     <span title=\"based on hpa, perplexity, uniprot...                                                                                                                          \n",
       "3     <span title=\"based on perplexity, uniprot\">cel...                                                                                                                          \n",
       "4     <span title=\"based on hpa, perplexity, uniprot...                                                                                                                          \n",
       "...                                                 ...                                                                                                                          \n",
       "3444                                            unknown                                                                                                                          \n",
       "3445                                            unknown                                                                                                                          \n",
       "3446                                            unknown                                                                                                                          \n",
       "3447                                            unknown                                                                                                                          \n",
       "3448                                            unknown                                                                                                                          \n",
       "\n",
       "     <span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Ligand HGNC ID&nbsp;&nbsp;</span>  \\\n",
       "0     <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "1     <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "2     <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "3     <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "4     <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "...                                                 ...                                                                               \n",
       "3444  <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "3445  <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "3446  <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "3447  <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "3448  <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "\n",
       "     <span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Receptor HGNC ID&nbsp;&nbsp;</span>  \\\n",
       "0     <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "1     <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "2     <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "3     <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "4     <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "...                                                 ...                                                                                 \n",
       "3444  <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "3445  <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "3446  <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "3447  <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "3448  <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "\n",
       "      ...  \\\n",
       "0     ...   \n",
       "1     ...   \n",
       "2     ...   \n",
       "3     ...   \n",
       "4     ...   \n",
       "...   ...   \n",
       "3444  ...   \n",
       "3445  ...   \n",
       "3446  ...   \n",
       "3447  ...   \n",
       "3448  ...   \n",
       "\n",
       "     <span title=\"Double-click header of Sheep Receptor Ensembl ID to ensure all values are shown\">Sheep Receptor Ensembl ID&nbsp;</span>  \\\n",
       "0                                    ENSOARG00020007714                                                                                     \n",
       "1                                    ENSOARG00020001685                                                                                     \n",
       "2                                    ENSOARG00020013146                                                                                     \n",
       "3                                    ENSOARG00020012079                                                                                     \n",
       "4                                    ENSOARG00020002625                                                                                     \n",
       "...                                                 ...                                                                                     \n",
       "3444                                                                                                                                        \n",
       "3445                                                                                                                                        \n",
       "3446                                                                                                                                        \n",
       "3447                                                                                                                                        \n",
       "3448                                                                                                                                        \n",
       "\n",
       "     <span title=\"Double-click header of Sheep Receptor to ensure all values are shown\">Sheep Receptor&nbsp;</span>  \\\n",
       "0                                                 HSPA5                                                               \n",
       "1                                                  LRP1                                                               \n",
       "2                                                BDKRB2                                                               \n",
       "3                                                  DPP4                                                               \n",
       "4                                                 EPHA3                                                               \n",
       "...                                                 ...                                                               \n",
       "3444                                                                                                                  \n",
       "3445                                                                                                                  \n",
       "3446                                                                                                                  \n",
       "3447                                                                                                                  \n",
       "3448                                                                                                                  \n",
       "\n",
       "     <span title=\"Double-click header of Marmoset Ligand Ensembl ID to ensure all values are shown\">Marmoset Ligand Ensembl ID&nbsp;</span>  \\\n",
       "0                                                                                                                                             \n",
       "1                                                                                                                                             \n",
       "2                                    ENSCJAG00000019755                                                                                       \n",
       "3                                    ENSCJAG00000017705                                                                                       \n",
       "4                                    ENSCJAG00000020430                                                                                       \n",
       "...                                                 ...                                                                                       \n",
       "3444                                                                                                                                          \n",
       "3445                                                                                                                                          \n",
       "3446                                                                                                                                          \n",
       "3447                                                                                                                                          \n",
       "3448                                                                                                                                          \n",
       "\n",
       "     <span title=\"Double-click header of Marmoset Ligand to ensure all values are shown\">Marmoset Ligand&nbsp;</span>  \\\n",
       "0                                                                                                                       \n",
       "1                                                                                                                       \n",
       "2                                                   ACE                                                                 \n",
       "3                                                   ADA                                                                 \n",
       "4                                                ADAM10                                                                 \n",
       "...                                                 ...                                                                 \n",
       "3444                                                                                                                    \n",
       "3445                                                                                                                    \n",
       "3446                                                                                                                    \n",
       "3447                                                                                                                    \n",
       "3448                                                                                                                    \n",
       "\n",
       "     <span title=\"Double-click header of Marmoset Receptor Ensembl ID to ensure all values are shown\">Marmoset Receptor Ensembl ID&nbsp;</span>  \\\n",
       "0                                    ENSCJAG00000006880                                                                                           \n",
       "1                                    ENSCJAG00000010289                                                                                           \n",
       "2                                    ENSCJAG00000006399                                                                                           \n",
       "3                                    ENSCJAG00000005739                                                                                           \n",
       "4                                    ENSCJAG00000008692                                                                                           \n",
       "...                                                 ...                                                                                           \n",
       "3444                                                                                                                                              \n",
       "3445                                                                                                                                              \n",
       "3446                                                                                                                                              \n",
       "3447                                                                                                                                              \n",
       "3448                                                                                                                                              \n",
       "\n",
       "     <span title=\"Double-click header of Marmoset Receptor to ensure all values are shown\">Marmoset Receptor&nbsp;</span>  \\\n",
       "0                                                 HSPA5                                                                     \n",
       "1                                                  LRP1                                                                     \n",
       "2                                                BDKRB2                                                                     \n",
       "3                                                  DPP4                                                                     \n",
       "4                                                 EPHA3                                                                     \n",
       "...                                                 ...                                                                     \n",
       "3444                                                                                                                        \n",
       "3445                                                                                                                        \n",
       "3446                                                                                                                        \n",
       "3447                                                                                                                        \n",
       "3448                                                                                                                        \n",
       "\n",
       "     <span title=\"Double-click header of Macaque Ligand Ensembl ID to ensure all values are shown\">Macaque Ligand Ensembl ID&nbsp;</span>  \\\n",
       "0                                                                                                                                           \n",
       "1                                                                                                                                           \n",
       "2                                    ENSMMUG00000012608                                                                                     \n",
       "3                                    ENSMMUG00000010603                                                                                     \n",
       "4                                    ENSMMUG00000017507                                                                                     \n",
       "...                                                 ...                                                                                     \n",
       "3444                                                                                                                                        \n",
       "3445                                                                                                                                        \n",
       "3446                                                                                                                                        \n",
       "3447                                                                                                                                        \n",
       "3448                                                                                                                                        \n",
       "\n",
       "     <span title=\"Double-click header of Macaque Ligand to ensure all values are shown\">Macaque Ligand&nbsp;</span>  \\\n",
       "0                                                                                                                     \n",
       "1                                                                                                                     \n",
       "2                                                   ACE                                                               \n",
       "3                                                   ADA                                                               \n",
       "4                                                ADAM10                                                               \n",
       "...                                                 ...                                                               \n",
       "3444                                                                                                                  \n",
       "3445                                                                                                                  \n",
       "3446                                                                                                                  \n",
       "3447                                                                                                                  \n",
       "3448                                                                                                                  \n",
       "\n",
       "     <span title=\"Double-click header of Macaque Receptor Ensembl ID to ensure all values are shown\">Macaque Receptor Ensembl ID&nbsp;</span>  \\\n",
       "0                                    ENSMMUG00000005217                                                                                         \n",
       "1                                    ENSMMUG00000021782                                                                                         \n",
       "2                                    ENSMMUG00000043275                                                                                         \n",
       "3                                    ENSMMUG00000008387                                                                                         \n",
       "4                                    ENSMMUG00000008242                                                                                         \n",
       "...                                                 ...                                                                                         \n",
       "3444                                                                                                                                            \n",
       "3445                                                                                                                                            \n",
       "3446                                                                                                                                            \n",
       "3447                                                                                                                                            \n",
       "3448                                                                                                                                            \n",
       "\n",
       "     <span title=\"Double-click header of Macaque Receptor to ensure all values are shown\">Macaque Receptor&nbsp;</span>  \n",
       "0                                                 HSPA5                                                                  \n",
       "1                                                  LRP1                                                                  \n",
       "2                                                BDKRB2                                                                  \n",
       "3                                                  DPP4                                                                  \n",
       "4                                                 EPHA3                                                                  \n",
       "...                                                 ...                                                                  \n",
       "3444                                                                                                                     \n",
       "3445                                                                                                                     \n",
       "3446                                                                                                                     \n",
       "3447                                                                                                                     \n",
       "3448                                                                                                                     \n",
       "\n",
       "[3449 rows x 64 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species = \"Mouse\"\n",
    "id_prefix = \"MGI\"\n",
    "gene_pair000_df = gene_pair000\n",
    "gene_pair_df = gene_pair\n",
    "gene_pair_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b3e93cc4-e859-4633-a99d-f0b07fc73bfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "    # Identify relevant columns for the species\n",
    "    if species in [\"Mouse\", \"Rat\", \"Zebrafish\"]:\n",
    "        species_columns = [col for col in gene_pair_df.columns if id_prefix in col or species in col]\n",
    "    else:\n",
    "        species_columns = [col for col in gene_pair_df.columns if species in col]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "6ca09adb-2e29-4aac-9c55-0b2a5c898631",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;span title=\"Double-click header of Mouse Ligand to ensure all values are shown\"&gt;Mouse Ligand&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Mouse Receptor to ensure all values are shown\"&gt;Mouse Receptor&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Mouse Genome Informatics (MGI) ID. Click on the link for more details\"&gt;Ligand MGI ID&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Mouse Genome Informatics (MGI) ID. Click on the link for more details\"&gt;Receptor MGI ID&lt;/span&gt;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A2m</td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A2m</td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ace</td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ada</td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Adam10</td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Adam10</td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Adam11</td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Adam12</td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Adam12</td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Adam12</td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Adam15</td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Adam15</td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Adam15</td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Adam15</td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Adam15</td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   <span title=\"Double-click header of Mouse Ligand to ensure all values are shown\">Mouse Ligand&nbsp;</span>  \\\n",
       "0                                                 A2m                                                           \n",
       "1                                                 A2m                                                           \n",
       "2                                                 Ace                                                           \n",
       "3                                                 Ada                                                           \n",
       "4                                              Adam10                                                           \n",
       "5                                              Adam10                                                           \n",
       "6                                              Adam11                                                           \n",
       "7                                              Adam12                                                           \n",
       "8                                              Adam12                                                           \n",
       "9                                              Adam12                                                           \n",
       "10                                             Adam15                                                           \n",
       "11                                             Adam15                                                           \n",
       "12                                             Adam15                                                           \n",
       "13                                             Adam15                                                           \n",
       "14                                             Adam15                                                           \n",
       "\n",
       "   <span title=\"Double-click header of Mouse Receptor to ensure all values are shown\">Mouse Receptor&nbsp;</span>  \\\n",
       "0                                                                                                                   \n",
       "1                                                                                                                   \n",
       "2                                                                                                                   \n",
       "3                                                                                                                   \n",
       "4                                                                                                                   \n",
       "5                                                                                                                   \n",
       "6                                                                                                                   \n",
       "7                                                                                                                   \n",
       "8                                                                                                                   \n",
       "9                                                                                                                   \n",
       "10                                                                                                                  \n",
       "11                                                                                                                  \n",
       "12                                                                                                                  \n",
       "13                                                                                                                  \n",
       "14                                                                                                                  \n",
       "\n",
       "   <span title=\"Mouse Genome Informatics (MGI) ID. Click on the link for more details\">Ligand MGI ID</span>  \\\n",
       "0   <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "1   <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "2   <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "3   <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "4   <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "5   <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "6   <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "7   <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "8   <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "9   <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "10  <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "11  <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "12  <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "13  <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "14  <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "\n",
       "   <span title=\"Mouse Genome Informatics (MGI) ID. Click on the link for more details\">Receptor MGI ID</span>  \n",
       "0   <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "1   <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "2   <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "3   <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "4   <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "5   <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "6   <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "7   <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "8   <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "9   <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "10  <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "11  <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "12  <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "13  <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "14  <a href=\"https://www.informatics.jax.org/marke...                                                          "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair000_df[species_columns].head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b34393cf-b409-4da6-9b39-2006f3b1dfae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;span title=\"Double-click header of Mouse Ligand to ensure all values are shown\"&gt;Mouse Ligand&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Mouse Receptor to ensure all values are shown\"&gt;Mouse Receptor&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Mouse Genome Informatics (MGI) ID. Click on the link for more details\"&gt;Ligand MGI ID&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Mouse Genome Informatics (MGI) ID. Click on the link for more details\"&gt;Receptor MGI ID&lt;/span&gt;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A2m</td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A2m</td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ace</td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ada</td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Adam10</td>\n",
       "      <td></td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3444</th>\n",
       "      <td>Pcdhb22</td>\n",
       "      <td>Pcdhb22</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3445</th>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3446</th>\n",
       "      <td>Saa3</td>\n",
       "      <td>Ly96</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3447</th>\n",
       "      <td>Saa3</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3448</th>\n",
       "      <td>Sema4a</td>\n",
       "      <td>Timd2</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3449 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     <span title=\"Double-click header of Mouse Ligand to ensure all values are shown\">Mouse Ligand&nbsp;</span>  \\\n",
       "0                                                   A2m                                                           \n",
       "1                                                   A2m                                                           \n",
       "2                                                   Ace                                                           \n",
       "3                                                   Ada                                                           \n",
       "4                                                Adam10                                                           \n",
       "...                                                 ...                                                           \n",
       "3444                                            Pcdhb22                                                           \n",
       "3445                                            Pcdhgb8                                                           \n",
       "3446                                               Saa3                                                           \n",
       "3447                                               Saa3                                                           \n",
       "3448                                             Sema4a                                                           \n",
       "\n",
       "     <span title=\"Double-click header of Mouse Receptor to ensure all values are shown\">Mouse Receptor&nbsp;</span>  \\\n",
       "0                                                                                                                     \n",
       "1                                                                                                                     \n",
       "2                                                                                                                     \n",
       "3                                                                                                                     \n",
       "4                                                                                                                     \n",
       "...                                                 ...                                                               \n",
       "3444                                            Pcdhb22                                                               \n",
       "3445                                            Pcdhgb8                                                               \n",
       "3446                                               Ly96                                                               \n",
       "3447                                               Tlr4                                                               \n",
       "3448                                              Timd2                                                               \n",
       "\n",
       "     <span title=\"Mouse Genome Informatics (MGI) ID. Click on the link for more details\">Ligand MGI ID</span>  \\\n",
       "0     <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "1     <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "2     <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "3     <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "4     <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "...                                                 ...                                                         \n",
       "3444  <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "3445  <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "3446  <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "3447  <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "3448  <a href=\"https://www.informatics.jax.org/marke...                                                         \n",
       "\n",
       "     <span title=\"Mouse Genome Informatics (MGI) ID. Click on the link for more details\">Receptor MGI ID</span>  \n",
       "0     <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "1     <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "2     <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "3     <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "4     <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "...                                                 ...                                                          \n",
       "3444  <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "3445  <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "3446  <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "3447  <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "3448  <a href=\"https://www.informatics.jax.org/marke...                                                          \n",
       "\n",
       "[3449 rows x 4 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "    # Filter rows where all species-specific columns are not empty\n",
    "    species_gene_pair = gene_pair000_df[(gene_pair000_df[species_columns].map(str.strip) != \"\").all(axis=1)]\n",
    "    \n",
    "    # Rename columns to remove species name\n",
    "    species_gene_pair.columns = [\n",
    "        col.replace(f\"{species} \", \"\").strip() if \"Ligand\" in col or \"Receptor\" in col else col\n",
    "        for col in species_gene_pair.columns\n",
    "    ]\n",
    "gene_pair000_df[species_columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1bc5935b-efc3-4b06-a098-21dd3d390be2",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'None of [Index([\\'<span title=\"Double-click header of Mouse Ligand to ensure all values are shown\">Mouse Ligand&nbsp;</span>\\',\\n       \\'<span title=\"Double-click header of Mouse Receptor to ensure all values are shown\">Mouse Receptor&nbsp;</span>\\',\\n       \\'<span title=\"Mouse Genome Informatics (MGI) ID. Click on the link for more details\">Ligand MGI ID</span>\\',\\n       \\'<span title=\"Mouse Genome Informatics (MGI) ID. Click on the link for more details\">Receptor MGI ID</span>\\'],\\n      dtype=\\'object\\')] are in the [columns]'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[20], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mspecies_gene_pair\u001b[49m\u001b[43m[\u001b[49m\u001b[43mspecies_columns\u001b[49m\u001b[43m]\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/frame.py:4108\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   4106\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m is_iterator(key):\n\u001b[1;32m   4107\u001b[0m         key \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(key)\n\u001b[0;32m-> 4108\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_indexer_strict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcolumns\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m[\u001b[38;5;241m1\u001b[39m]\n\u001b[1;32m   4110\u001b[0m \u001b[38;5;66;03m# take() does not accept boolean indexers\u001b[39;00m\n\u001b[1;32m   4111\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(indexer, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdtype\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m) \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mbool\u001b[39m:\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/indexes/base.py:6200\u001b[0m, in \u001b[0;36mIndex._get_indexer_strict\u001b[0;34m(self, key, axis_name)\u001b[0m\n\u001b[1;32m   6197\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   6198\u001b[0m     keyarr, indexer, new_indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reindex_non_unique(keyarr)\n\u001b[0;32m-> 6200\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_raise_if_missing\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkeyarr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis_name\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   6202\u001b[0m keyarr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtake(indexer)\n\u001b[1;32m   6203\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(key, Index):\n\u001b[1;32m   6204\u001b[0m     \u001b[38;5;66;03m# GH 42790 - Preserve name from an Index\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/quarto_llm_env/lib/python3.11/site-packages/pandas/core/indexes/base.py:6249\u001b[0m, in \u001b[0;36mIndex._raise_if_missing\u001b[0;34m(self, key, indexer, axis_name)\u001b[0m\n\u001b[1;32m   6247\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m nmissing:\n\u001b[1;32m   6248\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m nmissing \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mlen\u001b[39m(indexer):\n\u001b[0;32m-> 6249\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNone of [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mkey\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m] are in the [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00maxis_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m]\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   6251\u001b[0m     not_found \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(ensure_index(key)[missing_mask\u001b[38;5;241m.\u001b[39mnonzero()[\u001b[38;5;241m0\u001b[39m]]\u001b[38;5;241m.\u001b[39munique())\n\u001b[1;32m   6252\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mnot_found\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m not in index\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'None of [Index([\\'<span title=\"Double-click header of Mouse Ligand to ensure all values are shown\">Mouse Ligand&nbsp;</span>\\',\\n       \\'<span title=\"Double-click header of Mouse Receptor to ensure all values are shown\">Mouse Receptor&nbsp;</span>\\',\\n       \\'<span title=\"Mouse Genome Informatics (MGI) ID. Click on the link for more details\">Ligand MGI ID</span>\\',\\n       \\'<span title=\"Mouse Genome Informatics (MGI) ID. Click on the link for more details\">Receptor MGI ID</span>\\'],\\n      dtype=\\'object\\')] are in the [columns]'"
     ]
    }
   ],
   "source": [
    "species_gene_pair[species_columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4b80a629-494c-45b2-9ebe-06008097634b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to prepare datatables (cleaning and hyperlinking, adding tool tips, etc) input for the database qmds\n",
    "import sys, os\n",
    "from itables import init_notebook_mode\n",
    "import pandas as pd\n",
    "from itables import show\n",
    "from itables import options\n",
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "from bs4 import BeautifulSoup\n",
    "from createDataTable import gene_pair, gene_pair000, human_columns, lrPairsCount\n",
    "import warnings\n",
    "\n",
    "# Suppress SettingWithCopyWarning\n",
    "warnings.simplefilter(\"ignore\", category=UserWarning)\n",
    "\n",
    "# for mouse-specific, fill out missing info\n",
    "\n",
    "\n",
    "\n",
    "# Other vertebrates\n",
    "species_list = [\n",
    "    \"ptroglodytes\", \"ggallus\", \"sscrofa\", \"btaurus\", \n",
    "    \"clfamiliaris\", \"ecaballus\", \"oarambouillet\", \"cjacchus\",\n",
    "    \"mmulatta\"\n",
    "]\n",
    "\n",
    "def process_species(gene_pair_df, gene_pair000_df, species, id_prefix, ligand_index, receptor_index):\n",
    "    \"\"\"\n",
    "    Processes ligand-receptor interactions for a given species.\n",
    "\n",
    "    Parameters:\n",
    "        gene_pair_df (pd.DataFrame): The main gene pair DataFrame.\n",
    "        gene_pair000_df (pd.DataFrame): The filtered gene pair DataFrame.\n",
    "        species (str): The species name (e.g., \"Mouse\", \"Rat\", \"Zebrafish\").\n",
    "        id_prefix (str): The identifier prefix for the species (e.g., \"MGI\" for Mouse, \"RGD\" for Rat).\n",
    "        ligand_index (int): Index for selecting the ligand column.\n",
    "        receptor_index (int): Index for selecting the receptor column.\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: Processed DataFrame for the species.\n",
    "    \"\"\"\n",
    "    # Identify relevant columns for the species\n",
    "    if species in [\"Mouse\", \"Rat\", \"Zebrafish\"]:\n",
    "        species_columns = [col for col in gene_pair_df.columns if id_prefix in col or species in col]\n",
    "    else:\n",
    "        species_columns = [col for col in gene_pair_df.columns if species in col]\n",
    "\n",
    "    # Filter rows where all species-specific columns are not empty\n",
    "    species_gene_pair = gene_pair000_df[(gene_pair000_df[species_columns].map(str.strip) != \"\").all(axis=1)]\n",
    "    \n",
    "    # Rename columns to remove species name\n",
    "    species_gene_pair.columns = [\n",
    "        col.replace(f\"{species} \", \"\").strip() if \"Ligand\" in col or \"Receptor\" in col else col\n",
    "        for col in species_gene_pair.columns\n",
    "    ]\n",
    "\n",
    "    # Identify ligand and receptor columns dynamically\n",
    "    ligand_col = [col for col in species_gene_pair.columns if \"Ligand&nbsp;\" in col][ligand_index]\n",
    "    receptor_col = [col for col in species_gene_pair.columns if \"Receptor&nbsp;\" in col][receptor_index]\n",
    "    ligand_Location = [col for col in species_gene_pair.columns if \"Ligand Location\" in col][0]\n",
    "    receptor_Location = [col for col in species_gene_pair.columns if \"Receptor Location\" in col][0]\n",
    "\n",
    "    if id_prefix == \"Ensembl\":\n",
    "        # Extract species names dynamically\n",
    "        all_species = {\"Chimpanzee\", \"Pig\", \"Dog\", \"Cow\", \n",
    "                       \"Chicken\", \"Horse\", \"Sheep\", \"Marmoset\", \"Macaque\"}\n",
    "        # Function to clean column names by removing HTML tags\n",
    "        def clean_column_name(col):\n",
    "            # Remove HTML tags using BeautifulSoup\n",
    "            cleaned_name = BeautifulSoup(col, \"html.parser\").get_text()\n",
    "            return cleaned_name.strip()\n",
    "        \n",
    "        # Step 1: Identify columns that contain any species from all_species\n",
    "        cleaned_columns = [clean_column_name(col) for col in species_gene_pair.columns]\n",
    "        \n",
    "        # Identify columns where the species from all_species is in the cleaned name\n",
    "        columns_to_remove = [\n",
    "            species_gene_pair.columns[i] for i, cleaned_name in enumerate(cleaned_columns)\n",
    "            if any(species in cleaned_name for species in all_species)\n",
    "        ]\n",
    "        \n",
    "        # Step 2: Remove those columns from the DataFrame\n",
    "        species_gene_pair = species_gene_pair.drop(columns=columns_to_remove)\n",
    "        \n",
    "        \n",
    "            # Function to format ligand-receptor pairs\n",
    "    def format_lr_pair(row):\n",
    "        if row[ligand_Location] in ['secreted', '']:\n",
    "            return f\"{row[ligand_col]} <span style='font-size: 15px;'>○</span> <span style='font-size: 24px;'>⤚</span> {row[receptor_col]}\"\n",
    "        elif row[receptor_Location] == 'plasma membrane':\n",
    "            return f\"{row[ligand_col]} <span style='font-size: 24px;'>⤙</span> <span style='font-size: 24px;'>⤚</span> {row[receptor_col]}\"\n",
    "        else:\n",
    "            return f\"{row[ligand_col]} \\u2192 {row[receptor_col]}\"\n",
    "\n",
    "    # Apply the formatting function\n",
    "    species_gene_pair1 = species_gene_pair.copy()\n",
    "    species_gene_pair1.loc[:, f\"{species} LR Pair\"] = species_gene_pair1.apply(format_lr_pair, axis=1)\n",
    "\n",
    "    # Identify relevant columns for the species\n",
    "    species_columns = [col for col in species_gene_pair1.columns if id_prefix in col]\n",
    "    new_order = [human_columns[0]]+ [f\"{species} LR Pair\", ligand_col, receptor_col] + species_columns + human_columns[1:]\n",
    "    species_gene_pair1 = species_gene_pair1[new_order].reset_index(drop=True)\n",
    "    # Turn Interaction ID into clickable links\n",
    "    # species_gene_pair1[species_gene_pair1.columns[0]] = species_gene_pair1[species_gene_pair1.columns[0]].apply(\n",
    "    #     lambda x: f\"<a href='https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database/filter/{x}.html'>{x}</a>\"\n",
    "    # )\n",
    "    col_to_rename = [col for col in species_gene_pair1.columns if \"Human LR Pair\" in col][0]\n",
    "    species_gene_pair1.rename(columns={col_to_rename: \"LR Pair Card\"}, inplace=True)\n",
    "\n",
    "    return species_gene_pair1\n",
    "\n",
    "\n",
    "# Process each species\n",
    "# Mouse\n",
    "mouse_gene_pair1 = process_species(gene_pair, gene_pair000, \"Mouse\", \"MGI\", 1, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "517450d9-72d5-4caa-aba8-844aa4e90544",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;span title=\"Double-click header of Interaction ID to ensure all values are shown\"&gt;Interaction ID&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>Mouse LR Pair</th>\n",
       "      <th>&lt;span title=\"Double-click header of Ligand to ensure all values are shown\"&gt;Ligand&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Receptor to ensure all values are shown\"&gt;Receptor&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Genome Informatics (MGI) ID. Click on the link for more details\"&gt;Ligand MGI ID&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Genome Informatics (MGI) ID. Click on the link for more details\"&gt;Receptor MGI ID&lt;/span&gt;</th>\n",
       "      <th>LR Pair Card</th>\n",
       "      <th>&lt;span title=\"Official Gene Symbol; Hover on symbols below to show gene names\"\"&gt;Human Ligand&amp;nbsp;&amp;nbsp;&amp;nbsp;&lt;/span&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Official Gene Symbol; Hover on symbols below to show gene names\"\"&gt;Human Receptor&amp;nbsp;&amp;nbsp;&amp;nbsp;&lt;/span&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Ligand Symbols to ensure all values are shown\"\"&gt;Human Ligand Symbols&amp;nbsp;&lt;/span&lt;/span&gt;</th>\n",
       "      <th>...</th>\n",
       "      <th>&lt;span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\"&gt;Ligand Location&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\"&gt;Receptor Location&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\"&gt;Ligand HGNC ID&amp;nbsp;&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\"&gt;Receptor HGNC ID&amp;nbsp;&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Click the logo below to run Perplexity on the Human LR pair\"&gt;Perplexity&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Database Source to ensure all values are shown\"&gt;Database Source&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Binding Location to ensure all values are shown\"&gt;Binding Location&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Trans-binding to ensure all values are shown\"&gt;Trans-binding&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Bidirectional Signalling to ensure all values are shown\"&gt;Bidirectional Signalling&amp;nbsp;&lt;/span&gt;</th>\n",
       "      <th>&lt;span title=\"Double-click header of Interaction Type to ensure all values are shown\"&gt;Interaction Type&amp;nbsp;&lt;/span&gt;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDB03437</td>\n",
       "      <td>H60a → Klrk1</td>\n",
       "      <td>H60a</td>\n",
       "      <td>Klrk1</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\" \"&gt;H60a&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\" \"&gt;Klrk1&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"H60a\"&gt;H60a&lt;/span&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>&lt;span title=\"connectomeDB2025 🆕\"&gt;connectomeDB2...</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>No</td>\n",
       "      <td>non-covalent binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDB03438</td>\n",
       "      <td>H60b → Klrk1</td>\n",
       "      <td>H60b</td>\n",
       "      <td>Klrk1</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\" \"&gt;H60b&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\" \"&gt;Klrk1&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"H60b\"&gt;H60b&lt;/span&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>&lt;span title=\"connectomeDB2025 🆕\"&gt;connectomeDB2...</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>No</td>\n",
       "      <td>non-covalent binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDB03439</td>\n",
       "      <td>H60c → Klrk1</td>\n",
       "      <td>H60c</td>\n",
       "      <td>Klrk1</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\" \"&gt;H60c&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\" \"&gt;Klrk1&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"H60c\"&gt;H60c&lt;/span&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>&lt;span title=\"connectomeDB2025 🆕\"&gt;connectomeDB2...</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>No</td>\n",
       "      <td>non-covalent binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDB03440</td>\n",
       "      <td>Pcdhb17 → Pcdhb17</td>\n",
       "      <td>Pcdhb17</td>\n",
       "      <td>Pcdhb17</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhb17&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhb17&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"Pcdhb17\"&gt;Pcdhb17&lt;/span&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>&lt;span title=\"connectomeDB2025 🆕\"&gt;connectomeDB2...</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>unknown</td>\n",
       "      <td>non-covalent binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDB03441</td>\n",
       "      <td>Pcdhb18 → Pcdhb18</td>\n",
       "      <td>Pcdhb18</td>\n",
       "      <td>Pcdhb18</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhb18&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhb18&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"Pcdhb18\"&gt;Pcdhb18&lt;/span&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>&lt;span title=\"connectomeDB2025 🆕\"&gt;connectomeDB2...</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>cis</td>\n",
       "      <td>unknown</td>\n",
       "      <td>non-covalent binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CDB03442</td>\n",
       "      <td>Pcdhb19 → Pcdhb19</td>\n",
       "      <td>Pcdhb19</td>\n",
       "      <td>Pcdhb19</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhb19&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhb19&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"Pcdhb19\"&gt;Pcdhb19&lt;/span&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>&lt;span title=\"connectomeDB2025 🆕\"&gt;connectomeDB2...</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>cis</td>\n",
       "      <td>unknown</td>\n",
       "      <td>non-covalent binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CDB03443</td>\n",
       "      <td>Pcdhb20 → Pcdhb20</td>\n",
       "      <td>Pcdhb20</td>\n",
       "      <td>Pcdhb20</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhb20&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhb20&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"Pcdhb20\"&gt;Pcdhb20&lt;/span&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>&lt;span title=\"connectomeDB2025 🆕\"&gt;connectomeDB2...</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>cis</td>\n",
       "      <td>unknown</td>\n",
       "      <td>non-covalent binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CDB03444</td>\n",
       "      <td>Pcdhb21 → Pcdhb21</td>\n",
       "      <td>Pcdhb21</td>\n",
       "      <td>Pcdhb21</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhb21&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhb21&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"Pcdhb21\"&gt;Pcdhb21&lt;/span&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>&lt;span title=\"connectomeDB2025 🆕\"&gt;connectomeDB2...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CDB03445</td>\n",
       "      <td>Pcdhb22 → Pcdhb22</td>\n",
       "      <td>Pcdhb22</td>\n",
       "      <td>Pcdhb22</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhb22&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhb22&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"Pcdhb22\"&gt;Pcdhb22&lt;/span&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>&lt;span title=\"connectomeDB2025 🆕\"&gt;connectomeDB2...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CDB03446</td>\n",
       "      <td>Pcdhgb8 → Pcdhgb8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhgb8&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\" \"&gt;Pcdhgb8&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"Pcdhgb8\"&gt;Pcdhgb8&lt;/span&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>&lt;span title=\"connectomeDB2025 🆕\"&gt;connectomeDB2...</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>unknown</td>\n",
       "      <td>non-covalent binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CDB03447</td>\n",
       "      <td>Saa3 → Ly96</td>\n",
       "      <td>Saa3</td>\n",
       "      <td>Ly96</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\" \"&gt;Saa3&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\" \"&gt;Ly96&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"Saa3\"&gt;Saa3&lt;/span&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>&lt;span title=\"connectomeDB2025 🆕\"&gt;connectomeDB2...</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>No</td>\n",
       "      <td>non-covalent binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CDB03448</td>\n",
       "      <td>Saa3 → Tlr4</td>\n",
       "      <td>Saa3</td>\n",
       "      <td>Tlr4</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\" \"&gt;Saa3&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\" \"&gt;Tlr4&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"Saa3\"&gt;Saa3&lt;/span&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>&lt;span title=\"connectomeDB2025 🆕\"&gt;connectomeDB2...</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>No</td>\n",
       "      <td>non-covalent binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CDB03449</td>\n",
       "      <td>Sema4a → Timd2</td>\n",
       "      <td>Sema4a</td>\n",
       "      <td>Timd2</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://www.informatics.jax.org/marke...</td>\n",
       "      <td>&lt;a href=\"https://comp.med.yokohama-cu.ac.jp/co...</td>\n",
       "      <td>&lt;span title=\" \"&gt;Sema4a&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\" \"&gt;Timd2&lt;/span&gt;</td>\n",
       "      <td>&lt;span title=\"Sema4a\"&gt;Sema4a&lt;/span&gt;</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;a href=\"https://www.perplexity.ai/search?q=Wh...</td>\n",
       "      <td>&lt;span title=\"connectomeDB2025 🆕\"&gt;connectomeDB2...</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>No</td>\n",
       "      <td>non-covalent binding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   <span title=\"Double-click header of Interaction ID to ensure all values are shown\">Interaction ID&nbsp;</span>  \\\n",
       "0                                            CDB03437                                                               \n",
       "1                                            CDB03438                                                               \n",
       "2                                            CDB03439                                                               \n",
       "3                                            CDB03440                                                               \n",
       "4                                            CDB03441                                                               \n",
       "5                                            CDB03442                                                               \n",
       "6                                            CDB03443                                                               \n",
       "7                                            CDB03444                                                               \n",
       "8                                            CDB03445                                                               \n",
       "9                                            CDB03446                                                               \n",
       "10                                           CDB03447                                                               \n",
       "11                                           CDB03448                                                               \n",
       "12                                           CDB03449                                                               \n",
       "\n",
       "        Mouse LR Pair  \\\n",
       "0        H60a → Klrk1   \n",
       "1        H60b → Klrk1   \n",
       "2        H60c → Klrk1   \n",
       "3   Pcdhb17 → Pcdhb17   \n",
       "4   Pcdhb18 → Pcdhb18   \n",
       "5   Pcdhb19 → Pcdhb19   \n",
       "6   Pcdhb20 → Pcdhb20   \n",
       "7   Pcdhb21 → Pcdhb21   \n",
       "8   Pcdhb22 → Pcdhb22   \n",
       "9   Pcdhgb8 → Pcdhgb8   \n",
       "10        Saa3 → Ly96   \n",
       "11        Saa3 → Tlr4   \n",
       "12     Sema4a → Timd2   \n",
       "\n",
       "   <span title=\"Double-click header of Ligand to ensure all values are shown\">Ligand&nbsp;</span>  \\\n",
       "0                                                H60a                                               \n",
       "1                                                H60b                                               \n",
       "2                                                H60c                                               \n",
       "3                                             Pcdhb17                                               \n",
       "4                                             Pcdhb18                                               \n",
       "5                                             Pcdhb19                                               \n",
       "6                                             Pcdhb20                                               \n",
       "7                                             Pcdhb21                                               \n",
       "8                                             Pcdhb22                                               \n",
       "9                                             Pcdhgb8                                               \n",
       "10                                               Saa3                                               \n",
       "11                                               Saa3                                               \n",
       "12                                             Sema4a                                               \n",
       "\n",
       "   <span title=\"Double-click header of Receptor to ensure all values are shown\">Receptor&nbsp;</span>  \\\n",
       "0                                               Klrk1                                                   \n",
       "1                                               Klrk1                                                   \n",
       "2                                               Klrk1                                                   \n",
       "3                                             Pcdhb17                                                   \n",
       "4                                             Pcdhb18                                                   \n",
       "5                                             Pcdhb19                                                   \n",
       "6                                             Pcdhb20                                                   \n",
       "7                                             Pcdhb21                                                   \n",
       "8                                             Pcdhb22                                                   \n",
       "9                                             Pcdhgb8                                                   \n",
       "10                                               Ly96                                                   \n",
       "11                                               Tlr4                                                   \n",
       "12                                              Timd2                                                   \n",
       "\n",
       "   <span title=\"Genome Informatics (MGI) ID. Click on the link for more details\">Ligand MGI ID</span>  \\\n",
       "0   <a href=\"https://www.informatics.jax.org/marke...                                                   \n",
       "1   <a href=\"https://www.informatics.jax.org/marke...                                                   \n",
       "2   <a href=\"https://www.informatics.jax.org/marke...                                                   \n",
       "3   <a href=\"https://www.informatics.jax.org/marke...                                                   \n",
       "4   <a href=\"https://www.informatics.jax.org/marke...                                                   \n",
       "5   <a href=\"https://www.informatics.jax.org/marke...                                                   \n",
       "6   <a href=\"https://www.informatics.jax.org/marke...                                                   \n",
       "7   <a href=\"https://www.informatics.jax.org/marke...                                                   \n",
       "8   <a href=\"https://www.informatics.jax.org/marke...                                                   \n",
       "9   <a href=\"https://www.informatics.jax.org/marke...                                                   \n",
       "10  <a href=\"https://www.informatics.jax.org/marke...                                                   \n",
       "11  <a href=\"https://www.informatics.jax.org/marke...                                                   \n",
       "12  <a href=\"https://www.informatics.jax.org/marke...                                                   \n",
       "\n",
       "   <span title=\"Genome Informatics (MGI) ID. Click on the link for more details\">Receptor MGI ID</span>  \\\n",
       "0   <a href=\"https://www.informatics.jax.org/marke...                                                     \n",
       "1   <a href=\"https://www.informatics.jax.org/marke...                                                     \n",
       "2   <a href=\"https://www.informatics.jax.org/marke...                                                     \n",
       "3   <a href=\"https://www.informatics.jax.org/marke...                                                     \n",
       "4   <a href=\"https://www.informatics.jax.org/marke...                                                     \n",
       "5   <a href=\"https://www.informatics.jax.org/marke...                                                     \n",
       "6   <a href=\"https://www.informatics.jax.org/marke...                                                     \n",
       "7   <a href=\"https://www.informatics.jax.org/marke...                                                     \n",
       "8   <a href=\"https://www.informatics.jax.org/marke...                                                     \n",
       "9   <a href=\"https://www.informatics.jax.org/marke...                                                     \n",
       "10  <a href=\"https://www.informatics.jax.org/marke...                                                     \n",
       "11  <a href=\"https://www.informatics.jax.org/marke...                                                     \n",
       "12  <a href=\"https://www.informatics.jax.org/marke...                                                     \n",
       "\n",
       "                                         LR Pair Card  \\\n",
       "0   <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "1   <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "2   <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "3   <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "4   <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "5   <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "6   <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "7   <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "8   <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "9   <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "10  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "11  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "12  <a href=\"https://comp.med.yokohama-cu.ac.jp/co...   \n",
       "\n",
       "   <span title=\"Official Gene Symbol; Hover on symbols below to show gene names\"\">Human Ligand&nbsp;&nbsp;&nbsp;</span</span>  \\\n",
       "0                         <span title=\" \">H60a</span>                                                                           \n",
       "1                         <span title=\" \">H60b</span>                                                                           \n",
       "2                         <span title=\" \">H60c</span>                                                                           \n",
       "3                      <span title=\" \">Pcdhb17</span>                                                                           \n",
       "4                      <span title=\" \">Pcdhb18</span>                                                                           \n",
       "5                      <span title=\" \">Pcdhb19</span>                                                                           \n",
       "6                      <span title=\" \">Pcdhb20</span>                                                                           \n",
       "7                      <span title=\" \">Pcdhb21</span>                                                                           \n",
       "8                      <span title=\" \">Pcdhb22</span>                                                                           \n",
       "9                      <span title=\" \">Pcdhgb8</span>                                                                           \n",
       "10                        <span title=\" \">Saa3</span>                                                                           \n",
       "11                        <span title=\" \">Saa3</span>                                                                           \n",
       "12                      <span title=\" \">Sema4a</span>                                                                           \n",
       "\n",
       "   <span title=\"Official Gene Symbol; Hover on symbols below to show gene names\"\">Human Receptor&nbsp;&nbsp;&nbsp;</span</span>  \\\n",
       "0                        <span title=\" \">Klrk1</span>                                                                             \n",
       "1                        <span title=\" \">Klrk1</span>                                                                             \n",
       "2                        <span title=\" \">Klrk1</span>                                                                             \n",
       "3                      <span title=\" \">Pcdhb17</span>                                                                             \n",
       "4                      <span title=\" \">Pcdhb18</span>                                                                             \n",
       "5                      <span title=\" \">Pcdhb19</span>                                                                             \n",
       "6                      <span title=\" \">Pcdhb20</span>                                                                             \n",
       "7                      <span title=\" \">Pcdhb21</span>                                                                             \n",
       "8                      <span title=\" \">Pcdhb22</span>                                                                             \n",
       "9                      <span title=\" \">Pcdhgb8</span>                                                                             \n",
       "10                        <span title=\" \">Ly96</span>                                                                             \n",
       "11                        <span title=\" \">Tlr4</span>                                                                             \n",
       "12                       <span title=\" \">Timd2</span>                                                                             \n",
       "\n",
       "   <span title=\"Double-click header of Ligand Symbols to ensure all values are shown\"\">Human Ligand Symbols&nbsp;</span</span>  \\\n",
       "0                      <span title=\"H60a\">H60a</span>                                                                            \n",
       "1                      <span title=\"H60b\">H60b</span>                                                                            \n",
       "2                      <span title=\"H60c\">H60c</span>                                                                            \n",
       "3                <span title=\"Pcdhb17\">Pcdhb17</span>                                                                            \n",
       "4                <span title=\"Pcdhb18\">Pcdhb18</span>                                                                            \n",
       "5                <span title=\"Pcdhb19\">Pcdhb19</span>                                                                            \n",
       "6                <span title=\"Pcdhb20\">Pcdhb20</span>                                                                            \n",
       "7                <span title=\"Pcdhb21\">Pcdhb21</span>                                                                            \n",
       "8                <span title=\"Pcdhb22\">Pcdhb22</span>                                                                            \n",
       "9                <span title=\"Pcdhgb8\">Pcdhgb8</span>                                                                            \n",
       "10                     <span title=\"Saa3\">Saa3</span>                                                                            \n",
       "11                     <span title=\"Saa3\">Saa3</span>                                                                            \n",
       "12                 <span title=\"Sema4a\">Sema4a</span>                                                                            \n",
       "\n",
       "    ...  \\\n",
       "0   ...   \n",
       "1   ...   \n",
       "2   ...   \n",
       "3   ...   \n",
       "4   ...   \n",
       "5   ...   \n",
       "6   ...   \n",
       "7   ...   \n",
       "8   ...   \n",
       "9   ...   \n",
       "10  ...   \n",
       "11  ...   \n",
       "12  ...   \n",
       "\n",
       "   <span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\">Ligand Location</span>  \\\n",
       "0                                             unknown                                                                                                                        \n",
       "1                                             unknown                                                                                                                        \n",
       "2                                             unknown                                                                                                                        \n",
       "3                                             unknown                                                                                                                        \n",
       "4                                             unknown                                                                                                                        \n",
       "5                                             unknown                                                                                                                        \n",
       "6                                             unknown                                                                                                                        \n",
       "7                                             unknown                                                                                                                        \n",
       "8                                             unknown                                                                                                                        \n",
       "9                                             unknown                                                                                                                        \n",
       "10                                            unknown                                                                                                                        \n",
       "11                                            unknown                                                                                                                        \n",
       "12                                            unknown                                                                                                                        \n",
       "\n",
       "   <span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\">Receptor Location</span>  \\\n",
       "0                                             unknown                                                                                                                          \n",
       "1                                             unknown                                                                                                                          \n",
       "2                                             unknown                                                                                                                          \n",
       "3                                             unknown                                                                                                                          \n",
       "4                                             unknown                                                                                                                          \n",
       "5                                             unknown                                                                                                                          \n",
       "6                                             unknown                                                                                                                          \n",
       "7                                             unknown                                                                                                                          \n",
       "8                                             unknown                                                                                                                          \n",
       "9                                             unknown                                                                                                                          \n",
       "10                                            unknown                                                                                                                          \n",
       "11                                            unknown                                                                                                                          \n",
       "12                                            unknown                                                                                                                          \n",
       "\n",
       "   <span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Ligand HGNC ID&nbsp;&nbsp;</span>  \\\n",
       "0   <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "1   <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "2   <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "3   <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "4   <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "5   <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "6   <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "7   <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "8   <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "9   <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "10  <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "11  <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "12  <a href=\"https://www.genenames.org/data/gene-s...                                                                               \n",
       "\n",
       "   <span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Receptor HGNC ID&nbsp;&nbsp;</span>  \\\n",
       "0   <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "1   <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "2   <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "3   <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "4   <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "5   <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "6   <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "7   <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "8   <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "9   <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "10  <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "11  <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "12  <a href=\"https://www.genenames.org/data/gene-s...                                                                                 \n",
       "\n",
       "   <span title=\"Click the logo below to run Perplexity on the Human LR pair\">Perplexity&nbsp;</span>  \\\n",
       "0   <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "1   <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "2   <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "3   <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "4   <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "5   <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "6   <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "7   <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "8   <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "9   <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "10  <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "11  <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "12  <a href=\"https://www.perplexity.ai/search?q=Wh...                                                  \n",
       "\n",
       "   <span title=\"Double-click header of Database Source to ensure all values are shown\">Database Source&nbsp;</span>  \\\n",
       "0   <span title=\"connectomeDB2025 🆕\">connectomeDB2...                                                                 \n",
       "1   <span title=\"connectomeDB2025 🆕\">connectomeDB2...                                                                 \n",
       "2   <span title=\"connectomeDB2025 🆕\">connectomeDB2...                                                                 \n",
       "3   <span title=\"connectomeDB2025 🆕\">connectomeDB2...                                                                 \n",
       "4   <span title=\"connectomeDB2025 🆕\">connectomeDB2...                                                                 \n",
       "5   <span title=\"connectomeDB2025 🆕\">connectomeDB2...                                                                 \n",
       "6   <span title=\"connectomeDB2025 🆕\">connectomeDB2...                                                                 \n",
       "7   <span title=\"connectomeDB2025 🆕\">connectomeDB2...                                                                 \n",
       "8   <span title=\"connectomeDB2025 🆕\">connectomeDB2...                                                                 \n",
       "9   <span title=\"connectomeDB2025 🆕\">connectomeDB2...                                                                 \n",
       "10  <span title=\"connectomeDB2025 🆕\">connectomeDB2...                                                                 \n",
       "11  <span title=\"connectomeDB2025 🆕\">connectomeDB2...                                                                 \n",
       "12  <span title=\"connectomeDB2025 🆕\">connectomeDB2...                                                                 \n",
       "\n",
       "   <span title=\"Double-click header of Binding Location to ensure all values are shown\">Binding Location&nbsp;</span>  \\\n",
       "0                                       extracellular                                                                   \n",
       "1                                       extracellular                                                                   \n",
       "2                                       extracellular                                                                   \n",
       "3                                       extracellular                                                                   \n",
       "4                                       extracellular                                                                   \n",
       "5                                       extracellular                                                                   \n",
       "6                                       extracellular                                                                   \n",
       "7                                             unknown                                                                   \n",
       "8                                             unknown                                                                   \n",
       "9                                       extracellular                                                                   \n",
       "10                                      extracellular                                                                   \n",
       "11                                      extracellular                                                                   \n",
       "12                                      extracellular                                                                   \n",
       "\n",
       "   <span title=\"Double-click header of Trans-binding to ensure all values are shown\">Trans-binding&nbsp;</span>  \\\n",
       "0                                               trans                                                             \n",
       "1                                               trans                                                             \n",
       "2                                               trans                                                             \n",
       "3                                               trans                                                             \n",
       "4                                                 cis                                                             \n",
       "5                                                 cis                                                             \n",
       "6                                                 cis                                                             \n",
       "7                                             unknown                                                             \n",
       "8                                             unknown                                                             \n",
       "9                                               trans                                                             \n",
       "10                                              trans                                                             \n",
       "11                                              trans                                                             \n",
       "12                                              trans                                                             \n",
       "\n",
       "   <span title=\"Double-click header of Bidirectional Signalling to ensure all values are shown\">Bidirectional Signalling&nbsp;</span>  \\\n",
       "0                                                  No                                                                                   \n",
       "1                                                  No                                                                                   \n",
       "2                                                  No                                                                                   \n",
       "3                                             unknown                                                                                   \n",
       "4                                             unknown                                                                                   \n",
       "5                                             unknown                                                                                   \n",
       "6                                             unknown                                                                                   \n",
       "7                                             unknown                                                                                   \n",
       "8                                             unknown                                                                                   \n",
       "9                                             unknown                                                                                   \n",
       "10                                                 No                                                                                   \n",
       "11                                                 No                                                                                   \n",
       "12                                                 No                                                                                   \n",
       "\n",
       "   <span title=\"Double-click header of Interaction Type to ensure all values are shown\">Interaction Type&nbsp;</span>  \n",
       "0                                non-covalent binding                                                                  \n",
       "1                                non-covalent binding                                                                  \n",
       "2                                non-covalent binding                                                                  \n",
       "3                                non-covalent binding                                                                  \n",
       "4                                non-covalent binding                                                                  \n",
       "5                                non-covalent binding                                                                  \n",
       "6                                non-covalent binding                                                                  \n",
       "7                                             unknown                                                                  \n",
       "8                                             unknown                                                                  \n",
       "9                                non-covalent binding                                                                  \n",
       "10                               non-covalent binding                                                                  \n",
       "11                               non-covalent binding                                                                  \n",
       "12                               non-covalent binding                                                                  \n",
       "\n",
       "[13 rows x 21 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse_gene_pair1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "cb3d4edd-d209-4388-978f-fb3a106f6df8",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_info = pd.read_csv(\"data/mouse_info.csv\")\n",
    "mapping = dict(zip(mouse_info['symbol'], mouse_info['mgi']))\n",
    "gene_pair['Receptor MGI ID'] = gene_pair['Receptor'].replace(mapping)\n",
    "gene_pair['Ligand MGI ID'] = gene_pair['Ligand'].replace(mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "c1ec9693-1ef8-49ad-9564-238b29dcf8a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping = dict(zip(mouse_info['symbol'], mouse_info['mgi']))\n",
    "gene_pair['Receptor MGI ID'] = gene_pair['Receptor'].replace(mapping)\n",
    "gene_pair['Ligand MGI ID'] = gene_pair['Ligand'].replace(mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "91155095-0f15-44e8-bd92-391361a46242",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Human LR Pair</th>\n",
       "      <th>Ligand</th>\n",
       "      <th>Ligand HGNC ID</th>\n",
       "      <th>Receptor</th>\n",
       "      <th>Receptor HGNC ID</th>\n",
       "      <th>Perplexity</th>\n",
       "      <th>PMID</th>\n",
       "      <th>Binding Location</th>\n",
       "      <th>Trans-binding</th>\n",
       "      <th>Bidirectional Signalling</th>\n",
       "      <th>...</th>\n",
       "      <th>Rat Ligand</th>\n",
       "      <th>Ligand ZFIN ID</th>\n",
       "      <th>Zebrafish Ligand</th>\n",
       "      <th>Zebrafish Ligand Name</th>\n",
       "      <th>Receptor Name</th>\n",
       "      <th>Receptor MGI ID</th>\n",
       "      <th>Receptor RGD ID</th>\n",
       "      <th>Receptor Aliases</th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>Receptor Old symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>APOA1 ABCA1</td>\n",
       "      <td>APOA1</td>\n",
       "      <td>HGNC:600</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>HGNC:29</td>\n",
       "      <td>APOA1 ABCA1</td>\n",
       "      <td>12084722</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>Yes</td>\n",
       "      <td>...</td>\n",
       "      <td>Apoa1</td>\n",
       "      <td>ZDB-GENE-050302-172</td>\n",
       "      <td>apoa1b</td>\n",
       "      <td>apolipoprotein A-Ib</td>\n",
       "      <td>ATP binding cassette subfamily A member 1</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>RGD:631344</td>\n",
       "      <td>TGD</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>ABC1, HDLDT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>APOA1 ABCA1</td>\n",
       "      <td>APOA1</td>\n",
       "      <td>HGNC:600</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>HGNC:29</td>\n",
       "      <td>APOA1 ABCA1</td>\n",
       "      <td>12869555</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>Yes</td>\n",
       "      <td>...</td>\n",
       "      <td>Apoa1</td>\n",
       "      <td>ZDB-GENE-050302-172</td>\n",
       "      <td>apoa1b</td>\n",
       "      <td>apolipoprotein A-Ib</td>\n",
       "      <td>ATP binding cassette subfamily A member 1</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>RGD:631344</td>\n",
       "      <td>TGD</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>ABC1, HDLDT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>APOA1 ABCA1</td>\n",
       "      <td>APOA1</td>\n",
       "      <td>HGNC:600</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>HGNC:29</td>\n",
       "      <td>APOA1 ABCA1</td>\n",
       "      <td>22271762</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>Yes</td>\n",
       "      <td>...</td>\n",
       "      <td>Apoa1</td>\n",
       "      <td>ZDB-GENE-050302-172</td>\n",
       "      <td>apoa1b</td>\n",
       "      <td>apolipoprotein A-Ib</td>\n",
       "      <td>ATP binding cassette subfamily A member 1</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>RGD:631344</td>\n",
       "      <td>TGD</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>ABC1, HDLDT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>APOA1 ABCA1</td>\n",
       "      <td>APOA1</td>\n",
       "      <td>HGNC:600</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>HGNC:29</td>\n",
       "      <td>APOA1 ABCA1</td>\n",
       "      <td>30458687</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>Yes</td>\n",
       "      <td>...</td>\n",
       "      <td>Apoa1</td>\n",
       "      <td>ZDB-GENE-050302-172</td>\n",
       "      <td>apoa1b</td>\n",
       "      <td>apolipoprotein A-Ib</td>\n",
       "      <td>ATP binding cassette subfamily A member 1</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>RGD:631344</td>\n",
       "      <td>TGD</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>ABC1, HDLDT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>APOE ABCA1</td>\n",
       "      <td>APOE</td>\n",
       "      <td>HGNC:613</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>HGNC:29</td>\n",
       "      <td>APOE ABCA1</td>\n",
       "      <td>14754908</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>Apoe</td>\n",
       "      <td>ZDB-GENE-010724-18</td>\n",
       "      <td>apoea</td>\n",
       "      <td>apolipoprotein Ea</td>\n",
       "      <td>ATP binding cassette subfamily A member 1</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>RGD:631344</td>\n",
       "      <td>TGD</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>ABC1, HDLDT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5146</th>\n",
       "      <td>Pcdhgb8 Pcdhgb8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td></td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td></td>\n",
       "      <td>Pcdhgb8 Pcdhgb8</td>\n",
       "      <td>25171406</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGI:2136720</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5147</th>\n",
       "      <td>H60a Klrk1</td>\n",
       "      <td>H60a</td>\n",
       "      <td></td>\n",
       "      <td>Klrk1</td>\n",
       "      <td></td>\n",
       "      <td>H60a Klrk1</td>\n",
       "      <td>10894171</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGI:1196250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5148</th>\n",
       "      <td>H60b Klrk1</td>\n",
       "      <td>H60b</td>\n",
       "      <td></td>\n",
       "      <td>Klrk1</td>\n",
       "      <td></td>\n",
       "      <td>H60b Klrk1</td>\n",
       "      <td>10894171</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGI:1196250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5149</th>\n",
       "      <td>H60c Klrk1</td>\n",
       "      <td>H60c</td>\n",
       "      <td></td>\n",
       "      <td>Klrk1</td>\n",
       "      <td></td>\n",
       "      <td>H60c Klrk1</td>\n",
       "      <td>10894171</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGI:1196250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5150</th>\n",
       "      <td>Saa3 Tlr4</td>\n",
       "      <td>Saa3</td>\n",
       "      <td></td>\n",
       "      <td>Tlr4</td>\n",
       "      <td></td>\n",
       "      <td>Saa3 Tlr4</td>\n",
       "      <td>23858030</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGI:96824</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5151 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Human LR Pair   Ligand Ligand HGNC ID Receptor Receptor HGNC ID  \\\n",
       "0         APOA1 ABCA1    APOA1       HGNC:600    ABCA1          HGNC:29   \n",
       "1         APOA1 ABCA1    APOA1       HGNC:600    ABCA1          HGNC:29   \n",
       "2         APOA1 ABCA1    APOA1       HGNC:600    ABCA1          HGNC:29   \n",
       "3         APOA1 ABCA1    APOA1       HGNC:600    ABCA1          HGNC:29   \n",
       "4          APOE ABCA1     APOE       HGNC:613    ABCA1          HGNC:29   \n",
       "...               ...      ...            ...      ...              ...   \n",
       "5146  Pcdhgb8 Pcdhgb8  Pcdhgb8                 Pcdhgb8                    \n",
       "5147       H60a Klrk1     H60a                   Klrk1                    \n",
       "5148       H60b Klrk1     H60b                   Klrk1                    \n",
       "5149       H60c Klrk1     H60c                   Klrk1                    \n",
       "5150        Saa3 Tlr4     Saa3                    Tlr4                    \n",
       "\n",
       "           Perplexity      PMID Binding Location Trans-binding  \\\n",
       "0         APOA1 ABCA1  12084722    extracellular         trans   \n",
       "1         APOA1 ABCA1  12869555    extracellular         trans   \n",
       "2         APOA1 ABCA1  22271762    extracellular         trans   \n",
       "3         APOA1 ABCA1  30458687    extracellular         trans   \n",
       "4          APOE ABCA1  14754908    extracellular         trans   \n",
       "...               ...       ...              ...           ...   \n",
       "5146  Pcdhgb8 Pcdhgb8  25171406    extracellular         trans   \n",
       "5147       H60a Klrk1  10894171    extracellular         trans   \n",
       "5148       H60b Klrk1  10894171    extracellular         trans   \n",
       "5149       H60c Klrk1  10894171    extracellular         trans   \n",
       "5150        Saa3 Tlr4  23858030    extracellular         trans   \n",
       "\n",
       "     Bidirectional Signalling  ... Rat Ligand       Ligand ZFIN ID  \\\n",
       "0                         Yes  ...      Apoa1  ZDB-GENE-050302-172   \n",
       "1                         Yes  ...      Apoa1  ZDB-GENE-050302-172   \n",
       "2                         Yes  ...      Apoa1  ZDB-GENE-050302-172   \n",
       "3                         Yes  ...      Apoa1  ZDB-GENE-050302-172   \n",
       "4                          No  ...       Apoe   ZDB-GENE-010724-18   \n",
       "...                       ...  ...        ...                  ...   \n",
       "5146                  unknown  ...        NaN                  NaN   \n",
       "5147                       No  ...        NaN                  NaN   \n",
       "5148                       No  ...        NaN                  NaN   \n",
       "5149                       No  ...        NaN                  NaN   \n",
       "5150                       No  ...        NaN                  NaN   \n",
       "\n",
       "     Zebrafish Ligand Zebrafish Ligand Name  \\\n",
       "0              apoa1b   apolipoprotein A-Ib   \n",
       "1              apoa1b   apolipoprotein A-Ib   \n",
       "2              apoa1b   apolipoprotein A-Ib   \n",
       "3              apoa1b   apolipoprotein A-Ib   \n",
       "4               apoea     apolipoprotein Ea   \n",
       "...               ...                   ...   \n",
       "5146              NaN                   NaN   \n",
       "5147              NaN                   NaN   \n",
       "5148              NaN                   NaN   \n",
       "5149              NaN                   NaN   \n",
       "5150              NaN                   NaN   \n",
       "\n",
       "                                  Receptor Name Receptor MGI ID  \\\n",
       "0     ATP binding cassette subfamily A member 1           ABCA1   \n",
       "1     ATP binding cassette subfamily A member 1           ABCA1   \n",
       "2     ATP binding cassette subfamily A member 1           ABCA1   \n",
       "3     ATP binding cassette subfamily A member 1           ABCA1   \n",
       "4     ATP binding cassette subfamily A member 1           ABCA1   \n",
       "...                                         ...             ...   \n",
       "5146                                        NaN     MGI:2136720   \n",
       "5147                                        NaN     MGI:1196250   \n",
       "5148                                        NaN     MGI:1196250   \n",
       "5149                                        NaN     MGI:1196250   \n",
       "5150                                        NaN       MGI:96824   \n",
       "\n",
       "     Receptor RGD ID Receptor Aliases Approved symbol Receptor Old symbol  \n",
       "0         RGD:631344              TGD           ABCA1        ABC1, HDLDT1  \n",
       "1         RGD:631344              TGD           ABCA1        ABC1, HDLDT1  \n",
       "2         RGD:631344              TGD           ABCA1        ABC1, HDLDT1  \n",
       "3         RGD:631344              TGD           ABCA1        ABC1, HDLDT1  \n",
       "4         RGD:631344              TGD           ABCA1        ABC1, HDLDT1  \n",
       "...              ...              ...             ...                 ...  \n",
       "5146             NaN              NaN             NaN                 NaN  \n",
       "5147             NaN              NaN             NaN                 NaN  \n",
       "5148             NaN              NaN             NaN                 NaN  \n",
       "5149             NaN              NaN             NaN                 NaN  \n",
       "5150             NaN              NaN             NaN                 NaN  \n",
       "\n",
       "[5151 rows x 34 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "70e1ff13-1a8d-40b3-9833-cb86b2e5038b",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Step 1: Fill NaN (and None) values in the relevant columns with empty strings\n",
    "# This is crucial for consistent handling by the function before processing \"N/A\".\n",
    "gene_pair['Ligand'] = gene_pair['Ligand'].fillna('')\n",
    "gene_pair['Ligand Old symbol'] = gene_pair['Ligand Old symbol'].fillna('')\n",
    "gene_pair['Ligand Aliases'] = gene_pair['Ligand Aliases'].fillna('')\n",
    "\n",
    "# Step 2: Apply the modified format_symbol_aliases function\n",
    "gene_pair['Ligand Symbols'] = gene_pair.apply(\n",
    "    lambda row: format_symbol_aliases(\n",
    "        row['Ligand'],         # Mapped to 'symbol' in the function\n",
    "        row['Ligand Old symbol'],\n",
    "        row['Ligand Aliases']\n",
    "    ),\n",
    "    axis=1\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e485c01f-9bf8-4b6d-a87e-2e6ad281bca6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Human LR Pair</th>\n",
       "      <th>Ligand</th>\n",
       "      <th>Ligand HGNC ID</th>\n",
       "      <th>Receptor</th>\n",
       "      <th>Receptor HGNC ID</th>\n",
       "      <th>Perplexity</th>\n",
       "      <th>PMID</th>\n",
       "      <th>Binding Location</th>\n",
       "      <th>Trans-binding</th>\n",
       "      <th>Bidirectional Signalling</th>\n",
       "      <th>...</th>\n",
       "      <th>Ligand ZFIN ID</th>\n",
       "      <th>Zebrafish Ligand</th>\n",
       "      <th>Zebrafish Ligand Name</th>\n",
       "      <th>Receptor Name</th>\n",
       "      <th>Receptor MGI ID</th>\n",
       "      <th>Receptor RGD ID</th>\n",
       "      <th>Receptor Aliases</th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>Receptor Old symbol</th>\n",
       "      <th>Ligand Symbols</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>APOA1 ABCA1</td>\n",
       "      <td>APOA1</td>\n",
       "      <td>HGNC:600</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>HGNC:29</td>\n",
       "      <td>APOA1 ABCA1</td>\n",
       "      <td>12084722</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>Yes</td>\n",
       "      <td>...</td>\n",
       "      <td>ZDB-GENE-050302-172</td>\n",
       "      <td>apoa1b</td>\n",
       "      <td>apolipoprotein A-Ib</td>\n",
       "      <td>ATP binding cassette subfamily A member 1</td>\n",
       "      <td>MGI:99607</td>\n",
       "      <td>RGD:631344</td>\n",
       "      <td>TGD</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>ABC1, HDLDT1</td>\n",
       "      <td>APOA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>APOA1 ABCA1</td>\n",
       "      <td>APOA1</td>\n",
       "      <td>HGNC:600</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>HGNC:29</td>\n",
       "      <td>APOA1 ABCA1</td>\n",
       "      <td>12869555</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>Yes</td>\n",
       "      <td>...</td>\n",
       "      <td>ZDB-GENE-050302-172</td>\n",
       "      <td>apoa1b</td>\n",
       "      <td>apolipoprotein A-Ib</td>\n",
       "      <td>ATP binding cassette subfamily A member 1</td>\n",
       "      <td>MGI:99607</td>\n",
       "      <td>RGD:631344</td>\n",
       "      <td>TGD</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>ABC1, HDLDT1</td>\n",
       "      <td>APOA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>APOA1 ABCA1</td>\n",
       "      <td>APOA1</td>\n",
       "      <td>HGNC:600</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>HGNC:29</td>\n",
       "      <td>APOA1 ABCA1</td>\n",
       "      <td>22271762</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>Yes</td>\n",
       "      <td>...</td>\n",
       "      <td>ZDB-GENE-050302-172</td>\n",
       "      <td>apoa1b</td>\n",
       "      <td>apolipoprotein A-Ib</td>\n",
       "      <td>ATP binding cassette subfamily A member 1</td>\n",
       "      <td>MGI:99607</td>\n",
       "      <td>RGD:631344</td>\n",
       "      <td>TGD</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>ABC1, HDLDT1</td>\n",
       "      <td>APOA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>APOA1 ABCA1</td>\n",
       "      <td>APOA1</td>\n",
       "      <td>HGNC:600</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>HGNC:29</td>\n",
       "      <td>APOA1 ABCA1</td>\n",
       "      <td>30458687</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>Yes</td>\n",
       "      <td>...</td>\n",
       "      <td>ZDB-GENE-050302-172</td>\n",
       "      <td>apoa1b</td>\n",
       "      <td>apolipoprotein A-Ib</td>\n",
       "      <td>ATP binding cassette subfamily A member 1</td>\n",
       "      <td>MGI:99607</td>\n",
       "      <td>RGD:631344</td>\n",
       "      <td>TGD</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>ABC1, HDLDT1</td>\n",
       "      <td>APOA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>APOE ABCA1</td>\n",
       "      <td>APOE</td>\n",
       "      <td>HGNC:613</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>HGNC:29</td>\n",
       "      <td>APOE ABCA1</td>\n",
       "      <td>14754908</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>ZDB-GENE-010724-18</td>\n",
       "      <td>apoea</td>\n",
       "      <td>apolipoprotein Ea</td>\n",
       "      <td>ATP binding cassette subfamily A member 1</td>\n",
       "      <td>MGI:99607</td>\n",
       "      <td>RGD:631344</td>\n",
       "      <td>TGD</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>ABC1, HDLDT1</td>\n",
       "      <td>APOE (AD2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5146</th>\n",
       "      <td>Pcdhgb8 Pcdhgb8</td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td></td>\n",
       "      <td>Pcdhgb8</td>\n",
       "      <td></td>\n",
       "      <td>Pcdhgb8 Pcdhgb8</td>\n",
       "      <td>25171406</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pcdhgb8 (, )</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5147</th>\n",
       "      <td>H60a Klrk1</td>\n",
       "      <td>H60a</td>\n",
       "      <td></td>\n",
       "      <td>Klrk1</td>\n",
       "      <td></td>\n",
       "      <td>H60a Klrk1</td>\n",
       "      <td>10894171</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H60a (, )</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5148</th>\n",
       "      <td>H60b Klrk1</td>\n",
       "      <td>H60b</td>\n",
       "      <td></td>\n",
       "      <td>Klrk1</td>\n",
       "      <td></td>\n",
       "      <td>H60b Klrk1</td>\n",
       "      <td>10894171</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H60b (, )</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5149</th>\n",
       "      <td>H60c Klrk1</td>\n",
       "      <td>H60c</td>\n",
       "      <td></td>\n",
       "      <td>Klrk1</td>\n",
       "      <td></td>\n",
       "      <td>H60c Klrk1</td>\n",
       "      <td>10894171</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H60c (, )</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5150</th>\n",
       "      <td>Saa3 Tlr4</td>\n",
       "      <td>Saa3</td>\n",
       "      <td></td>\n",
       "      <td>Tlr4</td>\n",
       "      <td></td>\n",
       "      <td>Saa3 Tlr4</td>\n",
       "      <td>23858030</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>trans</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Saa3 (, )</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5151 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Human LR Pair   Ligand Ligand HGNC ID Receptor Receptor HGNC ID  \\\n",
       "0         APOA1 ABCA1    APOA1       HGNC:600    ABCA1          HGNC:29   \n",
       "1         APOA1 ABCA1    APOA1       HGNC:600    ABCA1          HGNC:29   \n",
       "2         APOA1 ABCA1    APOA1       HGNC:600    ABCA1          HGNC:29   \n",
       "3         APOA1 ABCA1    APOA1       HGNC:600    ABCA1          HGNC:29   \n",
       "4          APOE ABCA1     APOE       HGNC:613    ABCA1          HGNC:29   \n",
       "...               ...      ...            ...      ...              ...   \n",
       "5146  Pcdhgb8 Pcdhgb8  Pcdhgb8                 Pcdhgb8                    \n",
       "5147       H60a Klrk1     H60a                   Klrk1                    \n",
       "5148       H60b Klrk1     H60b                   Klrk1                    \n",
       "5149       H60c Klrk1     H60c                   Klrk1                    \n",
       "5150        Saa3 Tlr4     Saa3                    Tlr4                    \n",
       "\n",
       "           Perplexity      PMID Binding Location Trans-binding  \\\n",
       "0         APOA1 ABCA1  12084722    extracellular         trans   \n",
       "1         APOA1 ABCA1  12869555    extracellular         trans   \n",
       "2         APOA1 ABCA1  22271762    extracellular         trans   \n",
       "3         APOA1 ABCA1  30458687    extracellular         trans   \n",
       "4          APOE ABCA1  14754908    extracellular         trans   \n",
       "...               ...       ...              ...           ...   \n",
       "5146  Pcdhgb8 Pcdhgb8  25171406    extracellular         trans   \n",
       "5147       H60a Klrk1  10894171    extracellular         trans   \n",
       "5148       H60b Klrk1  10894171    extracellular         trans   \n",
       "5149       H60c Klrk1  10894171    extracellular         trans   \n",
       "5150        Saa3 Tlr4  23858030    extracellular         trans   \n",
       "\n",
       "     Bidirectional Signalling  ...       Ligand ZFIN ID Zebrafish Ligand  \\\n",
       "0                         Yes  ...  ZDB-GENE-050302-172           apoa1b   \n",
       "1                         Yes  ...  ZDB-GENE-050302-172           apoa1b   \n",
       "2                         Yes  ...  ZDB-GENE-050302-172           apoa1b   \n",
       "3                         Yes  ...  ZDB-GENE-050302-172           apoa1b   \n",
       "4                          No  ...   ZDB-GENE-010724-18            apoea   \n",
       "...                       ...  ...                  ...              ...   \n",
       "5146                  unknown  ...                  NaN              NaN   \n",
       "5147                       No  ...                  NaN              NaN   \n",
       "5148                       No  ...                  NaN              NaN   \n",
       "5149                       No  ...                  NaN              NaN   \n",
       "5150                       No  ...                  NaN              NaN   \n",
       "\n",
       "     Zebrafish Ligand Name                              Receptor Name  \\\n",
       "0      apolipoprotein A-Ib  ATP binding cassette subfamily A member 1   \n",
       "1      apolipoprotein A-Ib  ATP binding cassette subfamily A member 1   \n",
       "2      apolipoprotein A-Ib  ATP binding cassette subfamily A member 1   \n",
       "3      apolipoprotein A-Ib  ATP binding cassette subfamily A member 1   \n",
       "4        apolipoprotein Ea  ATP binding cassette subfamily A member 1   \n",
       "...                    ...                                        ...   \n",
       "5146                   NaN                                        NaN   \n",
       "5147                   NaN                                        NaN   \n",
       "5148                   NaN                                        NaN   \n",
       "5149                   NaN                                        NaN   \n",
       "5150                   NaN                                        NaN   \n",
       "\n",
       "     Receptor MGI ID Receptor RGD ID Receptor Aliases Approved symbol  \\\n",
       "0          MGI:99607      RGD:631344              TGD           ABCA1   \n",
       "1          MGI:99607      RGD:631344              TGD           ABCA1   \n",
       "2          MGI:99607      RGD:631344              TGD           ABCA1   \n",
       "3          MGI:99607      RGD:631344              TGD           ABCA1   \n",
       "4          MGI:99607      RGD:631344              TGD           ABCA1   \n",
       "...              ...             ...              ...             ...   \n",
       "5146             NaN             NaN              NaN             NaN   \n",
       "5147             NaN             NaN              NaN             NaN   \n",
       "5148             NaN             NaN              NaN             NaN   \n",
       "5149             NaN             NaN              NaN             NaN   \n",
       "5150             NaN             NaN              NaN             NaN   \n",
       "\n",
       "     Receptor Old symbol Ligand Symbols  \n",
       "0           ABC1, HDLDT1          APOA1  \n",
       "1           ABC1, HDLDT1          APOA1  \n",
       "2           ABC1, HDLDT1          APOA1  \n",
       "3           ABC1, HDLDT1          APOA1  \n",
       "4           ABC1, HDLDT1     APOE (AD2)  \n",
       "...                  ...            ...  \n",
       "5146                 NaN   Pcdhgb8 (, )  \n",
       "5147                 NaN      H60a (, )  \n",
       "5148                 NaN      H60b (, )  \n",
       "5149                 NaN      H60c (, )  \n",
       "5150                 NaN      Saa3 (, )  \n",
       "\n",
       "[5151 rows x 35 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "14448206-dc1d-4269-9b39-0c38c8125bdf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "These 1 PMIDs have abstracts that might be incomplete (do not end with a period or question mark):\n",
      "[6172602]\n"
     ]
    }
   ],
   "source": [
    "# Fill NaN values in 'Abstract' column with an empty string\n",
    "df['Abstract'] = df['Abstract'].fillna('')\n",
    "\n",
    "pmids_without_period = df[~df['Abstract'].str.endswith(('.', '?', 'available', ')', 'Review', '...'))]['PMID'] # Corrected the condition to find abstracts that *do not* end with a period or question mark\n",
    "pmid_check = pmids_without_period.tolist()\n",
    "\n",
    "print(\"These \" + str(len(pmid_check)) + \" PMIDs have abstracts that might be incomplete (do not end with a period or question mark):\")\n",
    "print(pmid_check)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "7971d050-bef1-4197-b78d-e3670b589aeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Abstract'] = df['Abstract'].fillna('')\n",
    "pmids_without_period = df[~df['Abstract'].str.startswith(('No abstract available'))]['PMID']\n",
    "pmid_check = pmids_without_period.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "b9bcaa52-d81c-4561-91d2-e14d74734dfa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[10025398,\n",
       " 10037686,\n",
       " 10037743,\n",
       " 10050855,\n",
       " 10051567,\n",
       " 10066262,\n",
       " 10073957,\n",
       " 10074428,\n",
       " 10077672,\n",
       " 10085134,\n",
       " 10100920,\n",
       " 10102268,\n",
       " 10103110,\n",
       " 10187774,\n",
       " 10188995,\n",
       " 10189055,\n",
       " 10190900,\n",
       " 10190906,\n",
       " 10193788,\n",
       " 10194432,\n",
       " 10196157,\n",
       " 10196161,\n",
       " 10196194,\n",
       " 10196234,\n",
       " 10196235,\n",
       " 10196546,\n",
       " 10202040,\n",
       " 10206645,\n",
       " 10207021,\n",
       " 10209034,\n",
       " 10212223,\n",
       " 10214951,\n",
       " 10225955,\n",
       " 10229797,\n",
       " 10231374,\n",
       " 10233762,\n",
       " 10233851,\n",
       " 10318773,\n",
       " 10318826,\n",
       " 10318834,\n",
       " 10318947,\n",
       " 10330424,\n",
       " 10336501,\n",
       " 10339405,\n",
       " 10342833,\n",
       " 10342881,\n",
       " 10342886,\n",
       " 10347172,\n",
       " 10347248,\n",
       " 10348342,\n",
       " 10352278,\n",
       " 10358030,\n",
       " 10364174,\n",
       " 10364178,\n",
       " 10364234,\n",
       " 10366627,\n",
       " 10366629,\n",
       " 10368033,\n",
       " 10369126,\n",
       " 10369464,\n",
       " 10371504,\n",
       " 10377245,\n",
       " 10381577,\n",
       " 10381815,\n",
       " 10381885,\n",
       " 10382758,\n",
       " 10385705,\n",
       " 10390077,\n",
       " 10395410,\n",
       " 10395669,\n",
       " 10397731,\n",
       " 10399917,\n",
       " 10399920,\n",
       " 10400673,\n",
       " 10409677,\n",
       " 10419462,\n",
       " 10421367,\n",
       " 10421368,\n",
       " 10421785,\n",
       " 10422787,\n",
       " 10426993,\n",
       " 10429675,\n",
       " 10433822,\n",
       " 10438490,\n",
       " 10438732,\n",
       " 10438935,\n",
       " 10446903,\n",
       " 10454496,\n",
       " 10461027,\n",
       " 10464311,\n",
       " 10470109,\n",
       " 10471841,\n",
       " 10475064,\n",
       " 10477718,\n",
       " 10479680,\n",
       " 10487979,\n",
       " 10488147,\n",
       " 10490030,\n",
       " 10493100,\n",
       " 10493512,\n",
       " 10493819,\n",
       " 10495959,\n",
       " 10497191,\n",
       " 10498628,\n",
       " 10498827,\n",
       " 10499587,\n",
       " 10504300,\n",
       " 10504396,\n",
       " 10514495,\n",
       " 10515610,\n",
       " 10516308,\n",
       " 10518536,\n",
       " 10518929,\n",
       " 10519430,\n",
       " 10520994,\n",
       " 10520995,\n",
       " 10521347,\n",
       " 10525310,\n",
       " 10528211,\n",
       " 10529171,\n",
       " 10531456,\n",
       " 10537356,\n",
       " 10540231,\n",
       " 10544247,\n",
       " 10545285,\n",
       " 10549288,\n",
       " 10551863,\n",
       " 10551867,\n",
       " 10557084,\n",
       " 10570056,\n",
       " 10571240,\n",
       " 10571241,\n",
       " 10572074,\n",
       " 10572088,\n",
       " 10574949,\n",
       " 10575006,\n",
       " 10581185,\n",
       " 10583729,\n",
       " 10585296,\n",
       " 10593335,\n",
       " 10593924,\n",
       " 10593938,\n",
       " 10595537,\n",
       " 10598579,\n",
       " 10598591,\n",
       " 10598788,\n",
       " 10600606,\n",
       " 10601865,\n",
       " 10602324,\n",
       " 10604470,\n",
       " 10607706,\n",
       " 10608878,\n",
       " 10617103,\n",
       " 10617658,\n",
       " 10618369,\n",
       " 10628755,\n",
       " 10631182,\n",
       " 10632583,\n",
       " 10634209,\n",
       " 10641437,\n",
       " 10644723,\n",
       " 10647992,\n",
       " 10648835,\n",
       " 10652271,\n",
       " 10652313,\n",
       " 10652336,\n",
       " 10655506,\n",
       " 10657301,\n",
       " 10660125,\n",
       " 10662782,\n",
       " 10662798,\n",
       " 10666376,\n",
       " 10669731,\n",
       " 10681521,\n",
       " 10684262,\n",
       " 10689365,\n",
       " 10692489,\n",
       " 10692589,\n",
       " 10698266,\n",
       " 10702270,\n",
       " 10706119,\n",
       " 10706604,\n",
       " 10706668,\n",
       " 10713099,\n",
       " 10713152,\n",
       " 10716993,\n",
       " 10721994,\n",
       " 10725346,\n",
       " 10725696,\n",
       " 10725697,\n",
       " 10727522,\n",
       " 10734053,\n",
       " 10734104,\n",
       " 10736564,\n",
       " 10741396,\n",
       " 10744716,\n",
       " 10746731,\n",
       " 10747918,\n",
       " 10747940,\n",
       " 10748121,\n",
       " 10748135,\n",
       " 10749879,\n",
       " 10749887,\n",
       " 10749942,\n",
       " 10751172,\n",
       " 10770212,\n",
       " 10770925,\n",
       " 10772816,\n",
       " 10772929,\n",
       " 10772961,\n",
       " 10777510,\n",
       " 10780976,\n",
       " 10781587,\n",
       " 10788520,\n",
       " 10799510,\n",
       " 10801128,\n",
       " 10801840,\n",
       " 10801846,\n",
       " 10801853,\n",
       " 10806482,\n",
       " 10809759,\n",
       " 10811223,\n",
       " 10811630,\n",
       " 10811850,\n",
       " 10812968,\n",
       " 10814828,\n",
       " 10816592,\n",
       " 10816663,\n",
       " 10820003,\n",
       " 10820396,\n",
       " 10825303,\n",
       " 10829012,\n",
       " 10835423,\n",
       " 10837024,\n",
       " 10837471,\n",
       " 10842354,\n",
       " 10845107,\n",
       " 10845911,\n",
       " 10846180,\n",
       " 10848816,\n",
       " 10851242,\n",
       " 10852818,\n",
       " 10854424,\n",
       " 10854439,\n",
       " 10854442,\n",
       " 10856295,\n",
       " 10862746,\n",
       " 10866201,\n",
       " 10866830,\n",
       " 10871287,\n",
       " 10875937,\n",
       " 10877843,\n",
       " 10880230,\n",
       " 10880444,\n",
       " 10881198,\n",
       " 10882063,\n",
       " 10882119,\n",
       " 10887190,\n",
       " 10894171,\n",
       " 10894543,\n",
       " 10894944,\n",
       " 10896774,\n",
       " 10899166,\n",
       " 10908309,\n",
       " 10908663,\n",
       " 10910894,\n",
       " 10917520,\n",
       " 10917521,\n",
       " 10918299,\n",
       " 10918572,\n",
       " 10922494,\n",
       " 10930533,\n",
       " 10934193,\n",
       " 10934197,\n",
       " 10934324,\n",
       " 10935626,\n",
       " 10939329,\n",
       " 10942578,\n",
       " 10942781,\n",
       " 10944520,\n",
       " 10946254,\n",
       " 10948192,\n",
       " 10950949,\n",
       " 10950950,\n",
       " 10956646,\n",
       " 10958687,\n",
       " 10958785,\n",
       " 10958799,\n",
       " 10961879,\n",
       " 10962009,\n",
       " 10964907,\n",
       " 10966616,\n",
       " 10966654,\n",
       " 10969083,\n",
       " 10973284,\n",
       " 10975800,\n",
       " 10978167,\n",
       " 10986460,\n",
       " 10990458,\n",
       " 10993067,\n",
       " 10993080,\n",
       " 10995872,\n",
       " 10997905,\n",
       " 11003391,\n",
       " 11006133,\n",
       " 11009104,\n",
       " 11010812,\n",
       " 11011067,\n",
       " 11011154,\n",
       " 11012895,\n",
       " 11015293,\n",
       " 11015447,\n",
       " 11015614,\n",
       " 11017100,\n",
       " 11017109,\n",
       " 11019817,\n",
       " 11023820,\n",
       " 11024015,\n",
       " 11024046,\n",
       " 11024295,\n",
       " 11027527,\n",
       " 11027676,\n",
       " 11029007,\n",
       " 11030354,\n",
       " 11030716,\n",
       " 11032250,\n",
       " 11033437,\n",
       " 11034352,\n",
       " 11035029,\n",
       " 11035066,\n",
       " 11039935,\n",
       " 11041375,\n",
       " 11048721,\n",
       " 11053440,\n",
       " 11054877,\n",
       " 11055978,\n",
       " 11058597,\n",
       " 11062498,\n",
       " 11066092,\n",
       " 11069186,\n",
       " 11073102,\n",
       " 11081635,\n",
       " 11085751,\n",
       " 11086056,\n",
       " 11087996,\n",
       " 11091218,\n",
       " 11092808,\n",
       " 11097863,\n",
       " 11099506,\n",
       " 11101306,\n",
       " 11106601,\n",
       " 11107061,\n",
       " 11110785,\n",
       " 11112777,\n",
       " 11114375,\n",
       " 11114383,\n",
       " 11116144,\n",
       " 11116158,\n",
       " 11116211,\n",
       " 11120746,\n",
       " 11123370,\n",
       " 11124255,\n",
       " 11133776,\n",
       " 11137709,\n",
       " 11137996,\n",
       " 11141076,\n",
       " 11146556,\n",
       " 11154210,\n",
       " 11156415,\n",
       " 11157754,\n",
       " 11158588,\n",
       " 11158855,\n",
       " 11159507,\n",
       " 11159842,\n",
       " 11160457,\n",
       " 11162558,\n",
       " 11162621,\n",
       " 11163076,\n",
       " 11163236,\n",
       " 11168387,\n",
       " 11169396,\n",
       " 11170631,\n",
       " 11172728,\n",
       " 11201742,\n",
       " 11208866,\n",
       " 11226328,\n",
       " 11226387,\n",
       " 11227737,\n",
       " 11228164,\n",
       " 11238599,\n",
       " 11239433,\n",
       " 11239445,\n",
       " 11241215,\n",
       " 11248553,\n",
       " 11248808,\n",
       " 11250200,\n",
       " 11250867,\n",
       " 11254912,\n",
       " 11257118,\n",
       " 11257134,\n",
       " 11259768,\n",
       " 11263668,\n",
       " 11264152,\n",
       " 11266181,\n",
       " 11266516,\n",
       " 11274220,\n",
       " 11278538,\n",
       " 11278614,\n",
       " 11278629,\n",
       " 11278667,\n",
       " 11278720,\n",
       " 11278724,\n",
       " 11278916,\n",
       " 11279011,\n",
       " 11279101,\n",
       " 11279107,\n",
       " 11282024,\n",
       " 11283023,\n",
       " 11285233,\n",
       " 11286993,\n",
       " 11287419,\n",
       " 11287427,\n",
       " 11290339,\n",
       " 11290797,\n",
       " 11294841,\n",
       " 11294867,\n",
       " 11297552,\n",
       " 11300760,\n",
       " 11309369,\n",
       " 11313993,\n",
       " 11314756,\n",
       " 11316806,\n",
       " 11323700,\n",
       " 11325787,\n",
       " 11327697,\n",
       " 11327878,\n",
       " 11329063,\n",
       " 11331280,\n",
       " 11332691,\n",
       " 11336643,\n",
       " 11336703,\n",
       " 11336704,\n",
       " 11336787,\n",
       " 11337500,\n",
       " 11342471,\n",
       " 11342595,\n",
       " 11342658,\n",
       " 11344130,\n",
       " 11346656,\n",
       " 11353828,\n",
       " 11356126,\n",
       " 11356985,\n",
       " 11357136,\n",
       " 11359774,\n",
       " 11359786,\n",
       " 11369591,\n",
       " 11373281,\n",
       " 11375253,\n",
       " 11375980,\n",
       " 11382746,\n",
       " 11385580,\n",
       " 11390394,\n",
       " 11390428,\n",
       " 11390453,\n",
       " 11401330,\n",
       " 11402070,\n",
       " 11403934,\n",
       " 11404363,\n",
       " 11404457,\n",
       " 11408607,\n",
       " 11416136,\n",
       " 11416224,\n",
       " 11416225,\n",
       " 11418113,\n",
       " 11418238,\n",
       " 11418623,\n",
       " 11418668,\n",
       " 11423128,\n",
       " 11423542,\n",
       " 11432858,\n",
       " 11432874,\n",
       " 11433302,\n",
       " 11435317,\n",
       " 11435703,\n",
       " 11437656,\n",
       " 11446740,\n",
       " 11447115,\n",
       " 11448771,\n",
       " 11449371,\n",
       " 11457843,\n",
       " 11459935,\n",
       " 11460491,\n",
       " 11461706,\n",
       " 11461914,\n",
       " 11466363,\n",
       " 11470772,\n",
       " 11470830,\n",
       " 11470831,\n",
       " 11470832,\n",
       " 11478907,\n",
       " 11481330,\n",
       " 11481429,\n",
       " 11485736,\n",
       " 11485994,\n",
       " 11486033,\n",
       " 11487023,\n",
       " 11493674,\n",
       " 11498506,\n",
       " 11501527,\n",
       " 11509692,\n",
       " 11511084,\n",
       " 11511101,\n",
       " 11518718,\n",
       " 11519828,\n",
       " 11520208,\n",
       " 11520797,\n",
       " 11533490,\n",
       " 11535811,\n",
       " 11544254,\n",
       " 11554781,\n",
       " 11555649,\n",
       " 11557679,\n",
       " 11559354,\n",
       " 11560943,\n",
       " 11560994,\n",
       " 11562786,\n",
       " 11564763,\n",
       " 11564780,\n",
       " 11574465,\n",
       " 11581298,\n",
       " 11588013,\n",
       " 11590146,\n",
       " 11590190,\n",
       " 11600206,\n",
       " 11602681,\n",
       " 11606131,\n",
       " 11665980,\n",
       " 11693189,\n",
       " 11696542,\n",
       " 11698684,\n",
       " 11700046,\n",
       " 11700073,\n",
       " 11705989,\n",
       " 11706020,\n",
       " 11709756,\n",
       " 11713248,\n",
       " 11714836,\n",
       " 11717514,\n",
       " 11723247,\n",
       " 11724772,\n",
       " 11724793,\n",
       " 11726197,\n",
       " 11728309,\n",
       " 11728344,\n",
       " 11733378,\n",
       " 11733492,\n",
       " 11734558,\n",
       " 11735113,\n",
       " 11738243,\n",
       " 11739188,\n",
       " 11742004,\n",
       " 11743735,\n",
       " 11747427,\n",
       " 11748263,\n",
       " 11751331,\n",
       " 11751872,\n",
       " 11754357,\n",
       " 11772027,\n",
       " 11773063,\n",
       " 11780069,\n",
       " 11781320,\n",
       " 11785979,\n",
       " 11786020,\n",
       " 11786503,\n",
       " 11801679,\n",
       " 11801734,\n",
       " 11805102,\n",
       " 11809971,\n",
       " 11811984,\n",
       " 11812005,\n",
       " 11812992,\n",
       " 11821050,\n",
       " 11821900,\n",
       " 11825645,\n",
       " 11825873,\n",
       " 11825898,\n",
       " 11830493,\n",
       " 11834704,\n",
       " 11839274,\n",
       " 11839762,\n",
       " 11839764,\n",
       " 11846885,\n",
       " 11850634,\n",
       " 11850637,\n",
       " 11854294,\n",
       " 11855816,\n",
       " 11856343,\n",
       " 11857925,\n",
       " 11858939,\n",
       " 11861531,\n",
       " 11865065,\n",
       " 11875495,\n",
       " 11876253,\n",
       " 11877390,\n",
       " 11886876,\n",
       " 11888273,\n",
       " 11889121,\n",
       " 11897681,\n",
       " 11903042,\n",
       " 11907075,\n",
       " 11909953,\n",
       " 11911831,\n",
       " 11912201,\n",
       " 11931347,\n",
       " 11937491,\n",
       " 11940489,\n",
       " 11952095,\n",
       " 11952836,\n",
       " 11956183,\n",
       " 11970960,\n",
       " 11970971,\n",
       " 11976263,\n",
       " 11978642,\n",
       " 11980922,\n",
       " 11981027,\n",
       " 11982704,\n",
       " 11986215,\n",
       " 11986311,\n",
       " 11994433,\n",
       " 11994538,\n",
       " 12006486,\n",
       " 12006574,\n",
       " 12007532,\n",
       " 12007534,\n",
       " 12011057,\n",
       " 12015308,\n",
       " 12023369,\n",
       " 12024206,\n",
       " 12029087,\n",
       " 12031600,\n",
       " 12032146,\n",
       " 12036876,\n",
       " 12036962,\n",
       " 12044914,\n",
       " 12045255,\n",
       " 12045258,\n",
       " 12050162,\n",
       " 12050670,\n",
       " 12051776,\n",
       " 12054506,\n",
       " 12054613,\n",
       " 12068084,\n",
       " 12068310,\n",
       " 12070135,\n",
       " 12070757,\n",
       " 12081195,\n",
       " 12081481,\n",
       " 12084722,\n",
       " 12084975,\n",
       " 12087105,\n",
       " 12093292,\n",
       " 12095980,\n",
       " 12097820,\n",
       " 12105206,\n",
       " 12114498,\n",
       " 12115612,\n",
       " 12118011,\n",
       " 12118248,\n",
       " 12121226,\n",
       " 12121660,\n",
       " 12122009,\n",
       " 12123670,\n",
       " 12130646,\n",
       " 12133828,\n",
       " 12134113,\n",
       " 12135884,\n",
       " 12138091,\n",
       " 12138115,\n",
       " 12139915,\n",
       " 12149192,\n",
       " 12149255,\n",
       " 12160746,\n",
       " 12165521,\n",
       " 12165561,\n",
       " 12167620,\n",
       " 12183630,\n",
       " 12189157,\n",
       " 12194978,\n",
       " 12195704,\n",
       " 12196113,\n",
       " 12196561,\n",
       " 12198496,\n",
       " 12208495,\n",
       " 12208882,\n",
       " 12210842,\n",
       " 12212791,\n",
       " 12213892,\n",
       " 12218106,\n",
       " 12220741,\n",
       " 12221720,\n",
       " 12223469,\n",
       " 12234182,\n",
       " 12238917,\n",
       " 12242085,\n",
       " 12244066,\n",
       " 12270118,\n",
       " 12270937,\n",
       " 12297049,\n",
       " 12297050,\n",
       " 12297674,\n",
       " 12325120,\n",
       " 12351624,\n",
       " 12356317,\n",
       " 12358597,\n",
       " 12361948,\n",
       " 12368905,\n",
       " 12374741,\n",
       " 12374801,\n",
       " 12374982,\n",
       " 12377945,\n",
       " 12379593,\n",
       " 12381680,\n",
       " 12383247,\n",
       " 12383251,\n",
       " 12392763,\n",
       " 12401809,\n",
       " 12409230,\n",
       " 12411706,\n",
       " 12414726,\n",
       " 12415008,\n",
       " 12417094,\n",
       " 12417446,\n",
       " 12419820,\n",
       " 12421602,\n",
       " 12427015,\n",
       " 12427764,\n",
       " 12429744,\n",
       " 12434297,\n",
       " 12446593,\n",
       " 12454284,\n",
       " 12454288,\n",
       " 12466190,\n",
       " 12469119,\n",
       " 12477859,\n",
       " 12482954,\n",
       " 12486116,\n",
       " 12486241,\n",
       " 12493742,\n",
       " 12496762,\n",
       " 12504588,\n",
       " 12506116,\n",
       " 12509413,\n",
       " 12519790,\n",
       " 12522143,\n",
       " 12522212,\n",
       " 12527209,\n",
       " 12527360,\n",
       " 12529173,\n",
       " 12529935,\n",
       " 12529937,\n",
       " 12547834,\n",
       " 12548712,\n",
       " 12554737,\n",
       " 12560218,\n",
       " 12563219,\n",
       " 12565884,\n",
       " 12571234,\n",
       " 12588872,\n",
       " 12591959,\n",
       " 12601009,\n",
       " 12603837,\n",
       " 12603839,\n",
       " 12604689,\n",
       " 12606401,\n",
       " 12606605,\n",
       " 12606621,\n",
       " 12606634,\n",
       " 12609745,\n",
       " 12611880,\n",
       " 12620237,\n",
       " 12627975,\n",
       " 12636918,\n",
       " 12643274,\n",
       " 12645948,\n",
       " 12649175,\n",
       " 12649481,\n",
       " 12651901,\n",
       " 12652295,\n",
       " 12660315,\n",
       " 12665544,\n",
       " 12667445,\n",
       " 12670700,\n",
       " 12671732,\n",
       " 12672054,\n",
       " 12682249,\n",
       " 12682293,\n",
       " 12682303,\n",
       " 12684503,\n",
       " 12695522,\n",
       " 12695525,\n",
       " 12700180,\n",
       " 12702704,\n",
       " 12704343,\n",
       " 12707269,\n",
       " 12709393,\n",
       " 12714578,\n",
       " 12714592,\n",
       " 12716896,\n",
       " 12718728,\n",
       " 12719537,\n",
       " 12721620,\n",
       " 12724320,\n",
       " 12724354,\n",
       " 12725245,\n",
       " 12727971,\n",
       " 12727981,\n",
       " 12729750,\n",
       " 12732206,\n",
       " 12732212,\n",
       " 12734373,\n",
       " 12736251,\n",
       " 12740392,\n",
       " 12755684,\n",
       " 12761559,\n",
       " 12761826,\n",
       " 12775584,\n",
       " 12782713,\n",
       " 12782716,\n",
       " 12787023,\n",
       " 12787562,\n",
       " 12794140,\n",
       " 12796776,\n",
       " 12799371,\n",
       " 12802337,\n",
       " 12807887,\n",
       " 12809600,\n",
       " 12810690,\n",
       " 12810718,\n",
       " 1281473,\n",
       " 12814948,\n",
       " 12815165,\n",
       " 1281789,\n",
       " 12818164,\n",
       " 12821648,\n",
       " 12826663,\n",
       " 12829785,\n",
       " 12829792,\n",
       " 12832759,\n",
       " 12833155,\n",
       " 12837245,\n",
       " 12839624,\n",
       " 12839991,\n",
       " 12842087,\n",
       " 12847263,\n",
       " 12851328,\n",
       " 12853576,\n",
       " 12857724,\n",
       " 12869555,\n",
       " 12870663,\n",
       " 12874224,\n",
       " 12879005,\n",
       " 12881528,\n",
       " 12882840,\n",
       " 12884299,\n",
       " 12887919,\n",
       " 12887920,\n",
       " 12889478,\n",
       " 12893276,\n",
       " 12897152,\n",
       " 12897781,\n",
       " 12899622,\n",
       " 12899627,\n",
       " 12900447,\n",
       " 12902511,\n",
       " 12907451,\n",
       " 12907757,\n",
       " 12909644,\n",
       " 12915402,\n",
       " 12939143,\n",
       " 12943195,\n",
       " 12944565,\n",
       " 12947048,\n",
       " 12950167,\n",
       " 12954451,\n",
       " 12958167,\n",
       " 12960329,\n",
       " 12967681,\n",
       " 12967935,\n",
       " 12972292,\n",
       " 12974481,\n",
       " 12975355,\n",
       " 1309330,\n",
       " 1311270,\n",
       " 1312036,\n",
       " 13130507,\n",
       " 1313946,\n",
       " 1315572,\n",
       " 1331689,\n",
       " 1333056,\n",
       " 1335692,\n",
       " 1337145,\n",
       " 1346068,\n",
       " 1346555,\n",
       " 1350381,\n",
       " 13679391,\n",
       " 1369684,\n",
       " 1372909,\n",
       " 1374165,\n",
       " 1374383,\n",
       " 1374385,\n",
       " 1375232,\n",
       " 1376732,\n",
       " 1377218,\n",
       " 1377404,\n",
       " 1378450,\n",
       " 1379593,\n",
       " 1379600,\n",
       " 1385111,\n",
       " 1385156,\n",
       " 1386409,\n",
       " 1427896,\n",
       " 14506282,\n",
       " 14507735,\n",
       " 14508119,\n",
       " 14514671,\n",
       " 14514732,\n",
       " 14522967,\n",
       " 14522968,\n",
       " 14522992,\n",
       " 14527339,\n",
       " 14530218,\n",
       " 14530373,\n",
       " 14530377,\n",
       " 14531999,\n",
       " 14550283,\n",
       " 14550305,\n",
       " 14557255,\n",
       " 14559902,\n",
       " 14567914,\n",
       " 14568956,\n",
       " 14580208,\n",
       " 14585398,\n",
       " 14592966,\n",
       " 14595443,\n",
       " 14596620,\n",
       " 14599850,\n",
       " 14607932,\n",
       " 14610079,\n",
       " 14615490,\n",
       " 14623906,\n",
       " 14625285,\n",
       " 14631378,\n",
       " 14637022,\n",
       " 14642436,\n",
       " 1464627,\n",
       " 14656470,\n",
       " 14659083,\n",
       " 14668209,\n",
       " 14675762,\n",
       " 14684734,\n",
       " 14684852,\n",
       " 14691055,\n",
       " 14691232,\n",
       " 14692877,\n",
       " 14697332,\n",
       " 14699013,\n",
       " 14701797,\n",
       " 14701821,\n",
       " 14712479,\n",
       " 14714274,\n",
       " 14715956,\n",
       " 14716313,\n",
       " 14718570,\n",
       " 14722252,\n",
       " 14730967,\n",
       " 14732692,\n",
       " 14734551,\n",
       " 14734748,\n",
       " 14739301,\n",
       " 14743216,\n",
       " 14754908,\n",
       " 14764580,\n",
       " 14764606,\n",
       " 14764663,\n",
       " 14966521,\n",
       " 14970208,\n",
       " 14970227,\n",
       " 14976204,\n",
       " 14977921,\n",
       " 14980518,\n",
       " 14983046,\n",
       " 14985763,\n",
       " 14991768,\n",
       " 15001579,\n",
       " 15001692,\n",
       " 15004192,\n",
       " 1500859,\n",
       " 1502153,\n",
       " 15023143,\n",
       " 15031111,\n",
       " ...]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pmid_check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bc6c369f-7969-4b70-b9c5-b58f5c484808",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0          A2M HSPA5\n",
       "1           A2M LRP1\n",
       "2         ACE BDKRB2\n",
       "3           ADA DPP4\n",
       "4       ADAM10 EPHA3\n",
       "            ...     \n",
       "3431      ZG16B TLR2\n",
       "3432      ZG16B TLR4\n",
       "3433      ZG16B TLR5\n",
       "3434      ZG16B TLR6\n",
       "3435      ZP3 CHRNA7\n",
       "Name: Human LR Pair, Length: 3436, dtype: object"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair0[\"Human LR Pair\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "40369c2d-26f6-4a2b-b95b-d10df30254fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'<a href=\"https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:2707\" target=\"_blank\">HGNC:2707</a>'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair0[\"Ligand HGNC ID\"][2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "17d85b4c-a48d-4dcd-929e-aded1c8c85db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['<span title=\"Double-click header of Interaction ID to ensure all values are shown\">Interaction ID&nbsp;</span>',\n",
       "       '<span title=\"Ligand-Receptor Interacting Pair, as described in Liu et al. (PMID: XXXXXX)\">Human LR Pair</span>',\n",
       "       '<span title=\"Official Gene Symbol; Hover on symbols below to show gene names\">Ligand&nbsp;&nbsp;&nbsp;</span>',\n",
       "       '<span title=\"Official Gene Symbol; Hover on symbols below to show gene names\">Receptor&nbsp;&nbsp;&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Ligand Symbols to ensure all values are shown\">Ligand Symbols&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Receptor Symbols to ensure all values are shown\">Receptor Symbols&nbsp;</span>',\n",
       "       '<span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\">Ligand Location</span>',\n",
       "       '<span title=\"Location based on the predicted subcellular localization of the human proteome, as described in Ramilowski et al. (PMID: 26198319)\">Receptor Location</span>',\n",
       "       '<span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Ligand HGNC ID&nbsp;&nbsp;</span>',\n",
       "       '<span title=\"HUGO Gene Nomenclature Committee (HGNC) ID. Click on the link for more details\">Receptor HGNC ID&nbsp;&nbsp;</span>',\n",
       "       '<span title=\"Click the logo below to run Perplexity on the Human LR pair\">Perplexity&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Database Source to ensure all values are shown\">Database Source&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Binding Location to ensure all values are shown\">Binding Location&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Trans-binding to ensure all values are shown\">Trans-binding&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Bidirectional Signalling to ensure all values are shown\">Bidirectional Signalling&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Interaction Type to ensure all values are shown\">Interaction Type&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Mouse Ligand to ensure all values are shown\">Mouse Ligand&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Mouse Receptor to ensure all values are shown\">Mouse Receptor&nbsp;</span>',\n",
       "       '<span title=\"Mouse Genome Informatics (MGI) ID. Click on the link for more details\">Ligand MGI ID</span>',\n",
       "       '<span title=\"Mouse Genome Informatics (MGI) ID. Click on the link for more details\">Receptor MGI ID</span>',\n",
       "       '<span title=\"Double-click header of Rat Ligand to ensure all values are shown\">Rat Ligand&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Rat Receptor to ensure all values are shown\">Rat Receptor&nbsp;</span>',\n",
       "       '<span title=\"Rat Genome Database (RGD) ID. Click on the link for more details\">Ligand RGD ID</span>',\n",
       "       '<span title=\"Rat Genome Database (RGD) ID. Click on the link for more details\">Receptor RGD ID</span>',\n",
       "       '<span title=\"Double-click header of Zebrafish Ligand to ensure all values are shown\">Zebrafish Ligand&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Zebrafish Receptor to ensure all values are shown\">Zebrafish Receptor&nbsp;</span>',\n",
       "       '<span title=\"Zebrafish Information Network (ZFIN) ID. Click on the link for more details\">Ligand ZFIN ID</span>',\n",
       "       '<span title=\"Zebrafish Information Network (ZFIN) ID. Click on the link for more details\">Receptor ZFIN ID</span>',\n",
       "       '<span title=\"Double-click header of Chimpanzee Ligand Ensembl ID to ensure all values are shown\">Chimpanzee Ligand Ensembl ID&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Chimpanzee Ligand to ensure all values are shown\">Chimpanzee Ligand&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Chimpanzee Receptor Ensembl ID to ensure all values are shown\">Chimpanzee Receptor Ensembl ID&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Chimpanzee Receptor to ensure all values are shown\">Chimpanzee Receptor&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Chicken Ligand Ensembl ID to ensure all values are shown\">Chicken Ligand Ensembl ID&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Chicken Ligand to ensure all values are shown\">Chicken Ligand&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Chicken Receptor Ensembl ID to ensure all values are shown\">Chicken Receptor Ensembl ID&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Chicken Receptor to ensure all values are shown\">Chicken Receptor&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Pig Ligand Ensembl ID to ensure all values are shown\">Pig Ligand Ensembl ID&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Pig Ligand to ensure all values are shown\">Pig Ligand&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Pig Receptor Ensembl ID to ensure all values are shown\">Pig Receptor Ensembl ID&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Pig Receptor to ensure all values are shown\">Pig Receptor&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Cow Ligand Ensembl ID to ensure all values are shown\">Cow Ligand Ensembl ID&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Cow Ligand to ensure all values are shown\">Cow Ligand&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Cow Receptor Ensembl ID to ensure all values are shown\">Cow Receptor Ensembl ID&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Cow Receptor to ensure all values are shown\">Cow Receptor&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Dog Ligand Ensembl ID to ensure all values are shown\">Dog Ligand Ensembl ID&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Dog Ligand to ensure all values are shown\">Dog Ligand&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Dog Receptor Ensembl ID to ensure all values are shown\">Dog Receptor Ensembl ID&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Dog Receptor to ensure all values are shown\">Dog Receptor&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Horse Ligand Ensembl ID to ensure all values are shown\">Horse Ligand Ensembl ID&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Horse Ligand to ensure all values are shown\">Horse Ligand&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Horse Receptor Ensembl ID to ensure all values are shown\">Horse Receptor Ensembl ID&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Horse Receptor to ensure all values are shown\">Horse Receptor&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Sheep Ligand Ensembl ID to ensure all values are shown\">Sheep Ligand Ensembl ID&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Sheep Ligand to ensure all values are shown\">Sheep Ligand&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Sheep Receptor Ensembl ID to ensure all values are shown\">Sheep Receptor Ensembl ID&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Sheep Receptor to ensure all values are shown\">Sheep Receptor&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Marmoset Ligand Ensembl ID to ensure all values are shown\">Marmoset Ligand Ensembl ID&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Marmoset Ligand to ensure all values are shown\">Marmoset Ligand&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Marmoset Receptor Ensembl ID to ensure all values are shown\">Marmoset Receptor Ensembl ID&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Marmoset Receptor to ensure all values are shown\">Marmoset Receptor&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Rhesus Monkey Ligand Ensembl ID to ensure all values are shown\">Rhesus Monkey Ligand Ensembl ID&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Rhesus Monkey Ligand to ensure all values are shown\">Rhesus Monkey Ligand&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Rhesus Monkey Receptor Ensembl ID to ensure all values are shown\">Rhesus Monkey Receptor Ensembl ID&nbsp;</span>',\n",
       "       '<span title=\"Double-click header of Rhesus Monkey Receptor to ensure all values are shown\">Rhesus Monkey Receptor&nbsp;</span>'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Identify the column(s) that contain '(PMID)' and temporarily remove\n",
    "pmid_cols = [col for col in gene_pair.columns if '(PMID)' in col]\n",
    "gene_pair = gene_pair.drop(columns=pmid_cols)\n",
    "gene_pair.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7bb47f9e-bba7-4783-bb11-24df07dbe56e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Temporarily remove the column(s)\n",
    "gene_pair = gene_pair.drop(columns=pmid_cols)\n",
    "gene_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "d54b3c04-d721-4c60-81d7-3d89e4bb39b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Approved symbol', 'Alias symbol', 'Previous symbol',\n",
       "       'Date symbol changed', 'ensembl_gene_id', 'uniprot_ids', 'omim_id',\n",
       "       'Other Symbols', 'ensembl_peptide_id'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pop_up_info_lim.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "6e542eca-e0e3-4196-8757-6a17e891d48c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HGNC ID</th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>Approved name</th>\n",
       "      <th>locus_group</th>\n",
       "      <th>locus_type</th>\n",
       "      <th>status</th>\n",
       "      <th>location</th>\n",
       "      <th>location_sortable</th>\n",
       "      <th>Alias symbol</th>\n",
       "      <th>alias_name</th>\n",
       "      <th>...</th>\n",
       "      <th>cd</th>\n",
       "      <th>lncrnadb</th>\n",
       "      <th>enzyme_id</th>\n",
       "      <th>intermediate_filament_db</th>\n",
       "      <th>rna_central_id</th>\n",
       "      <th>lncipedia</th>\n",
       "      <th>gtrnadb</th>\n",
       "      <th>agr</th>\n",
       "      <th>mane_select</th>\n",
       "      <th>gencc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>N/A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>ENST00000263100.8|NM_130786.4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HGNC:37133</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>A1BG antisense RNA 1</td>\n",
       "      <td>non-coding RNA</td>\n",
       "      <td>RNA, long non-coding</td>\n",
       "      <td>Approved</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>FLJ23569</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>URS00007E4F6E</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:37133</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>10q11.23</td>\n",
       "      <td>10q11.23</td>\n",
       "      <td>ACF, ASP, ACF64, ACF65, APOBEC1CF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>ENST00000373997.8|NM_014576.4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>FWP007, S863-7, CPAMD5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>ENST00000318602.12|NM_000014.6</td>\n",
       "      <td>HGNC:7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HGNC:27057</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>A2M antisense RNA 1</td>\n",
       "      <td>non-coding RNA</td>\n",
       "      <td>RNA, long non-coding</td>\n",
       "      <td>Approved</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>N/A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>URS00001F234A</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:27057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44079</th>\n",
       "      <td>HGNC:3562</td>\n",
       "      <td>FABP7</td>\n",
       "      <td>fatty acid binding protein 7</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>6q22.31</td>\n",
       "      <td>06q22.31</td>\n",
       "      <td>B-FABP, BLBP</td>\n",
       "      <td>brain lipid binding protein</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:3562</td>\n",
       "      <td>ENST00000368444.8|NM_001446.5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44080</th>\n",
       "      <td>HGNC:41951</td>\n",
       "      <td>FABP7P1</td>\n",
       "      <td>fatty acid binding protein 7 pseudogene 1</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1q44</td>\n",
       "      <td>01q44</td>\n",
       "      <td>N/A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:41951</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44081</th>\n",
       "      <td>HGNC:41952</td>\n",
       "      <td>FABP7P2</td>\n",
       "      <td>fatty acid binding protein 7 pseudogene 2</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>Approved</td>\n",
       "      <td>2q11.1</td>\n",
       "      <td>02q11.1</td>\n",
       "      <td>N/A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:41952</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44082</th>\n",
       "      <td>HGNC:3563</td>\n",
       "      <td>FABP9</td>\n",
       "      <td>fatty acid binding protein 9</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>8q21.13</td>\n",
       "      <td>08q21.13</td>\n",
       "      <td>PERF, T-FABP, PERF15</td>\n",
       "      <td>testis fatty acid binding protein</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:3563</td>\n",
       "      <td>ENST00000379071.4|NM_001080526.2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44083</th>\n",
       "      <td>HGNC:34524</td>\n",
       "      <td>FABP12</td>\n",
       "      <td>fatty acid binding protein 12</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>8q21.13</td>\n",
       "      <td>08q21.13</td>\n",
       "      <td>N/A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:34524</td>\n",
       "      <td>ENST00000360464.6|NM_001105281.6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>44084 rows × 54 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          HGNC ID Approved symbol                              Approved name  \\\n",
       "0          HGNC:5            A1BG                     alpha-1-B glycoprotein   \n",
       "1      HGNC:37133        A1BG-AS1                       A1BG antisense RNA 1   \n",
       "2      HGNC:24086            A1CF             APOBEC1 complementation factor   \n",
       "3          HGNC:7             A2M                      alpha-2-macroglobulin   \n",
       "4      HGNC:27057         A2M-AS1                        A2M antisense RNA 1   \n",
       "...           ...             ...                                        ...   \n",
       "44079   HGNC:3562           FABP7               fatty acid binding protein 7   \n",
       "44080  HGNC:41951         FABP7P1  fatty acid binding protein 7 pseudogene 1   \n",
       "44081  HGNC:41952         FABP7P2  fatty acid binding protein 7 pseudogene 2   \n",
       "44082   HGNC:3563           FABP9               fatty acid binding protein 9   \n",
       "44083  HGNC:34524          FABP12              fatty acid binding protein 12   \n",
       "\n",
       "               locus_group                 locus_type    status  location  \\\n",
       "0      protein-coding gene  gene with protein product  Approved  19q13.43   \n",
       "1           non-coding RNA       RNA, long non-coding  Approved  19q13.43   \n",
       "2      protein-coding gene  gene with protein product  Approved  10q11.23   \n",
       "3      protein-coding gene  gene with protein product  Approved  12p13.31   \n",
       "4           non-coding RNA       RNA, long non-coding  Approved  12p13.31   \n",
       "...                    ...                        ...       ...       ...   \n",
       "44079  protein-coding gene  gene with protein product  Approved   6q22.31   \n",
       "44080           pseudogene                 pseudogene  Approved      1q44   \n",
       "44081           pseudogene                 pseudogene  Approved    2q11.1   \n",
       "44082  protein-coding gene  gene with protein product  Approved   8q21.13   \n",
       "44083  protein-coding gene  gene with protein product  Approved   8q21.13   \n",
       "\n",
       "      location_sortable                       Alias symbol  \\\n",
       "0              19q13.43                                N/A   \n",
       "1              19q13.43                           FLJ23569   \n",
       "2              10q11.23  ACF, ASP, ACF64, ACF65, APOBEC1CF   \n",
       "3              12p13.31             FWP007, S863-7, CPAMD5   \n",
       "4              12p13.31                                N/A   \n",
       "...                 ...                                ...   \n",
       "44079          06q22.31                       B-FABP, BLBP   \n",
       "44080             01q44                                N/A   \n",
       "44081           02q11.1                                N/A   \n",
       "44082          08q21.13               PERF, T-FABP, PERF15   \n",
       "44083          08q21.13                                N/A   \n",
       "\n",
       "                              alias_name  ...   cd lncrnadb enzyme_id  \\\n",
       "0                                    NaN  ...  NaN      NaN       NaN   \n",
       "1                                    NaN  ...  NaN      NaN       NaN   \n",
       "2                                    NaN  ...  NaN      NaN       NaN   \n",
       "3                                    NaN  ...  NaN      NaN       NaN   \n",
       "4                                    NaN  ...  NaN      NaN       NaN   \n",
       "...                                  ...  ...  ...      ...       ...   \n",
       "44079        brain lipid binding protein  ...  NaN      NaN       NaN   \n",
       "44080                                NaN  ...  NaN      NaN       NaN   \n",
       "44081                                NaN  ...  NaN      NaN       NaN   \n",
       "44082  testis fatty acid binding protein  ...  NaN      NaN       NaN   \n",
       "44083                                NaN  ...  NaN      NaN       NaN   \n",
       "\n",
       "      intermediate_filament_db rna_central_id lncipedia gtrnadb         agr  \\\n",
       "0                          NaN            NaN       NaN     NaN      HGNC:5   \n",
       "1                          NaN  URS00007E4F6E  A1BG-AS1     NaN  HGNC:37133   \n",
       "2                          NaN            NaN       NaN     NaN  HGNC:24086   \n",
       "3                          NaN            NaN       NaN     NaN      HGNC:7   \n",
       "4                          NaN  URS00001F234A   A2M-AS1     NaN  HGNC:27057   \n",
       "...                        ...            ...       ...     ...         ...   \n",
       "44079                      NaN            NaN       NaN     NaN   HGNC:3562   \n",
       "44080                      NaN            NaN       NaN     NaN  HGNC:41951   \n",
       "44081                      NaN            NaN       NaN     NaN  HGNC:41952   \n",
       "44082                      NaN            NaN       NaN     NaN   HGNC:3563   \n",
       "44083                      NaN            NaN       NaN     NaN  HGNC:34524   \n",
       "\n",
       "                            mane_select   gencc  \n",
       "0         ENST00000263100.8|NM_130786.4     NaN  \n",
       "1                                   NaN     NaN  \n",
       "2         ENST00000373997.8|NM_014576.4     NaN  \n",
       "3        ENST00000318602.12|NM_000014.6  HGNC:7  \n",
       "4                                   NaN     NaN  \n",
       "...                                 ...     ...  \n",
       "44079     ENST00000368444.8|NM_001446.5     NaN  \n",
       "44080                               NaN     NaN  \n",
       "44081                               NaN     NaN  \n",
       "44082  ENST00000379071.4|NM_001080526.2     NaN  \n",
       "44083  ENST00000360464.6|NM_001105281.6     NaN  \n",
       "\n",
       "[44084 rows x 54 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pop_up_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "00779318-fe54-4b2a-a5b0-a916f06fdc26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HGNC ID</th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>Approved name</th>\n",
       "      <th>locus_group</th>\n",
       "      <th>locus_type</th>\n",
       "      <th>status</th>\n",
       "      <th>location</th>\n",
       "      <th>location_sortable</th>\n",
       "      <th>Alias symbol</th>\n",
       "      <th>alias_name</th>\n",
       "      <th>...</th>\n",
       "      <th>cd</th>\n",
       "      <th>lncrnadb</th>\n",
       "      <th>enzyme_id</th>\n",
       "      <th>intermediate_filament_db</th>\n",
       "      <th>rna_central_id</th>\n",
       "      <th>lncipedia</th>\n",
       "      <th>gtrnadb</th>\n",
       "      <th>agr</th>\n",
       "      <th>mane_select</th>\n",
       "      <th>gencc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HGNC:100</td>\n",
       "      <td>ASIC1</td>\n",
       "      <td>acid sensing ion channel subunit 1</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>12q13.12</td>\n",
       "      <td>12q13.12</td>\n",
       "      <td>BNaC2, hBNaC2</td>\n",
       "      <td>nan</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>HGNC:100</td>\n",
       "      <td>ENST00000447966.7|NM_001095.4</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HGNC:10000</td>\n",
       "      <td>RGS4</td>\n",
       "      <td>regulator of G protein signaling 4</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1q23.3</td>\n",
       "      <td>01q23.3</td>\n",
       "      <td>N/A</td>\n",
       "      <td>nan</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>HGNC:10000</td>\n",
       "      <td>ENST00000367909.11|NM_005613.6</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HGNC:10001</td>\n",
       "      <td>RGS5</td>\n",
       "      <td>regulator of G protein signaling 5</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1q23.3</td>\n",
       "      <td>01q23.3</td>\n",
       "      <td>N/A</td>\n",
       "      <td>nan</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>HGNC:10001</td>\n",
       "      <td>ENST00000313961.10|NM_003617.4</td>\n",
       "      <td>HGNC:10001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HGNC:10002</td>\n",
       "      <td>RGS6</td>\n",
       "      <td>regulator of G protein signaling 6</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>14q24.2</td>\n",
       "      <td>14q24.2</td>\n",
       "      <td>N/A</td>\n",
       "      <td>nan</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>HGNC:10002</td>\n",
       "      <td>ENST00000553525.6|NM_001204424.2</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HGNC:10003</td>\n",
       "      <td>RGS7</td>\n",
       "      <td>regulator of G protein signaling 7</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1q43</td>\n",
       "      <td>01q43</td>\n",
       "      <td>N/A</td>\n",
       "      <td>nan</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>HGNC:10003</td>\n",
       "      <td>ENST00000440928.6|NM_001364886.1</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44079</th>\n",
       "      <td>HGNC:9995</td>\n",
       "      <td>RGS13</td>\n",
       "      <td>regulator of G protein signaling 13</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1q31.2</td>\n",
       "      <td>01q31.2</td>\n",
       "      <td>N/A</td>\n",
       "      <td>nan</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>HGNC:9995</td>\n",
       "      <td>ENST00000391995.7|NM_002927.5</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44080</th>\n",
       "      <td>HGNC:9996</td>\n",
       "      <td>RGS14</td>\n",
       "      <td>regulator of G protein signaling 14</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>5q35.3</td>\n",
       "      <td>05q35.3</td>\n",
       "      <td>N/A</td>\n",
       "      <td>nan</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>HGNC:9996</td>\n",
       "      <td>ENST00000408923.8|NM_006480.5</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44081</th>\n",
       "      <td>HGNC:9997</td>\n",
       "      <td>RGS16</td>\n",
       "      <td>regulator of G protein signaling 16</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1q25.3</td>\n",
       "      <td>01q25.3</td>\n",
       "      <td>A28-RGS14, RGS-r</td>\n",
       "      <td>nan</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>HGNC:9997</td>\n",
       "      <td>ENST00000367558.6|NM_002928.4</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44082</th>\n",
       "      <td>HGNC:9998</td>\n",
       "      <td>RGS2</td>\n",
       "      <td>regulator of G protein signaling 2</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1q31.2</td>\n",
       "      <td>01q31.2</td>\n",
       "      <td>N/A</td>\n",
       "      <td>nan</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>HGNC:9998</td>\n",
       "      <td>ENST00000235382.7|NM_002923.4</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44083</th>\n",
       "      <td>HGNC:9999</td>\n",
       "      <td>RGS3</td>\n",
       "      <td>regulator of G protein signaling 3</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>9q32</td>\n",
       "      <td>09q32</td>\n",
       "      <td>C2PA, FLJ20370, PDZ-RGS3</td>\n",
       "      <td>nan</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>HGNC:9999</td>\n",
       "      <td>ENST00000695401.1|NM_001394167.1</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>44084 rows × 54 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          HGNC ID Approved symbol                        Approved name  \\\n",
       "0        HGNC:100           ASIC1   acid sensing ion channel subunit 1   \n",
       "1      HGNC:10000            RGS4   regulator of G protein signaling 4   \n",
       "2      HGNC:10001            RGS5   regulator of G protein signaling 5   \n",
       "3      HGNC:10002            RGS6   regulator of G protein signaling 6   \n",
       "4      HGNC:10003            RGS7   regulator of G protein signaling 7   \n",
       "...           ...             ...                                  ...   \n",
       "44079   HGNC:9995           RGS13  regulator of G protein signaling 13   \n",
       "44080   HGNC:9996           RGS14  regulator of G protein signaling 14   \n",
       "44081   HGNC:9997           RGS16  regulator of G protein signaling 16   \n",
       "44082   HGNC:9998            RGS2   regulator of G protein signaling 2   \n",
       "44083   HGNC:9999            RGS3   regulator of G protein signaling 3   \n",
       "\n",
       "               locus_group                 locus_type    status  location  \\\n",
       "0      protein-coding gene  gene with protein product  Approved  12q13.12   \n",
       "1      protein-coding gene  gene with protein product  Approved    1q23.3   \n",
       "2      protein-coding gene  gene with protein product  Approved    1q23.3   \n",
       "3      protein-coding gene  gene with protein product  Approved   14q24.2   \n",
       "4      protein-coding gene  gene with protein product  Approved      1q43   \n",
       "...                    ...                        ...       ...       ...   \n",
       "44079  protein-coding gene  gene with protein product  Approved    1q31.2   \n",
       "44080  protein-coding gene  gene with protein product  Approved    5q35.3   \n",
       "44081  protein-coding gene  gene with protein product  Approved    1q25.3   \n",
       "44082  protein-coding gene  gene with protein product  Approved    1q31.2   \n",
       "44083  protein-coding gene  gene with protein product  Approved      9q32   \n",
       "\n",
       "      location_sortable              Alias symbol alias_name  ...   cd  \\\n",
       "0              12q13.12             BNaC2, hBNaC2        nan  ...  nan   \n",
       "1               01q23.3                       N/A        nan  ...  nan   \n",
       "2               01q23.3                       N/A        nan  ...  nan   \n",
       "3               14q24.2                       N/A        nan  ...  nan   \n",
       "4                 01q43                       N/A        nan  ...  nan   \n",
       "...                 ...                       ...        ...  ...  ...   \n",
       "44079           01q31.2                       N/A        nan  ...  nan   \n",
       "44080           05q35.3                       N/A        nan  ...  nan   \n",
       "44081           01q25.3          A28-RGS14, RGS-r        nan  ...  nan   \n",
       "44082           01q31.2                       N/A        nan  ...  nan   \n",
       "44083             09q32  C2PA, FLJ20370, PDZ-RGS3        nan  ...  nan   \n",
       "\n",
       "      lncrnadb enzyme_id intermediate_filament_db rna_central_id lncipedia  \\\n",
       "0          nan       nan                      nan            nan       nan   \n",
       "1          nan       nan                      nan            nan       nan   \n",
       "2          nan       nan                      nan            nan       nan   \n",
       "3          nan       nan                      nan            nan       nan   \n",
       "4          nan       nan                      nan            nan       nan   \n",
       "...        ...       ...                      ...            ...       ...   \n",
       "44079      nan       nan                      nan            nan       nan   \n",
       "44080      nan       nan                      nan            nan       nan   \n",
       "44081      nan       nan                      nan            nan       nan   \n",
       "44082      nan       nan                      nan            nan       nan   \n",
       "44083      nan       nan                      nan            nan       nan   \n",
       "\n",
       "      gtrnadb         agr                       mane_select       gencc  \n",
       "0         nan    HGNC:100     ENST00000447966.7|NM_001095.4         nan  \n",
       "1         nan  HGNC:10000    ENST00000367909.11|NM_005613.6         nan  \n",
       "2         nan  HGNC:10001    ENST00000313961.10|NM_003617.4  HGNC:10001  \n",
       "3         nan  HGNC:10002  ENST00000553525.6|NM_001204424.2         nan  \n",
       "4         nan  HGNC:10003  ENST00000440928.6|NM_001364886.1         nan  \n",
       "...       ...         ...                               ...         ...  \n",
       "44079     nan   HGNC:9995     ENST00000391995.7|NM_002927.5         nan  \n",
       "44080     nan   HGNC:9996     ENST00000408923.8|NM_006480.5         nan  \n",
       "44081     nan   HGNC:9997     ENST00000367558.6|NM_002928.4         nan  \n",
       "44082     nan   HGNC:9998     ENST00000235382.7|NM_002923.4         nan  \n",
       "44083     nan   HGNC:9999  ENST00000695401.1|NM_001394167.1         nan  \n",
       "\n",
       "[44084 rows x 54 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pop_up_info = pop_up_info.groupby('HGNC ID').agg(agg_func).reset_index()\n",
    "pop_up_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "121bc7b7-0516-4ed9-86ec-46b9e3b35b3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ligand HGNC ID</th>\n",
       "      <th>Ligand Symbols</th>\n",
       "      <th>Ligand Location</th>\n",
       "      <th>Ligand group</th>\n",
       "      <th>root_group_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;span title=\"&lt;span title=\"A2M (FWP007, S863-7,...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/genegr...</td>\n",
       "      <td>1234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;span title=\"&lt;span title=\"ACE (DCP1, ACE1, CD1...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/genegr...</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;span title=\"&lt;span title=\"ADA (ADA1)\"&gt;ADA (ADA...</td>\n",
       "      <td>&lt;span title=\"based on hpa, uniprot\"&gt;cell membr...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/genegr...</td>\n",
       "      <td>1302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;span title=\"&lt;span title=\"ADAM10 (kuz, MADM, H...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/genegr...</td>\n",
       "      <td>2104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;span title=\"&lt;span title=\"ADAM10 (kuz, MADM, H...</td>\n",
       "      <td>&lt;span title=\"based on hpa, perplexity, uniprot...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/genegr...</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1583</th>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;span title=\"&lt;span title=\"YBX1 (NSEP1, YB-1, Y...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/genegr...</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1584</th>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;span title=\"&lt;span title=\"YBX1 (NSEP1, YB-1, Y...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/genegr...</td>\n",
       "      <td>1520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1585</th>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;span title=\"&lt;span title=\"YBX1 (NSEP1, YB-1, Y...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;sec...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/genegr...</td>\n",
       "      <td>2045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1586</th>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;span title=\"&lt;span title=\"ZG16B (HRPE773, PRO1...</td>\n",
       "      <td>&lt;span title=\"based on perplexity, uniprot\"&gt;cel...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/genegr...</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1587</th>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/gene-s...</td>\n",
       "      <td>&lt;span title=\"&lt;span title=\"ZP3 (ZP3A, ZP3B, ZP3...</td>\n",
       "      <td>&lt;span title=\"based on uniprot\"&gt;cell membrane&lt;/...</td>\n",
       "      <td>&lt;a href=\"https://www.genenames.org/data/genegr...</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1588 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Ligand HGNC ID  \\\n",
       "0     <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "1     <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "2     <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "3     <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "4     <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "...                                                 ...   \n",
       "1583  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "1584  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "1585  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "1586  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "1587  <a href=\"https://www.genenames.org/data/gene-s...   \n",
       "\n",
       "                                         Ligand Symbols  \\\n",
       "0     <span title=\"<span title=\"A2M (FWP007, S863-7,...   \n",
       "1     <span title=\"<span title=\"ACE (DCP1, ACE1, CD1...   \n",
       "2     <span title=\"<span title=\"ADA (ADA1)\">ADA (ADA...   \n",
       "3     <span title=\"<span title=\"ADAM10 (kuz, MADM, H...   \n",
       "4     <span title=\"<span title=\"ADAM10 (kuz, MADM, H...   \n",
       "...                                                 ...   \n",
       "1583  <span title=\"<span title=\"YBX1 (NSEP1, YB-1, Y...   \n",
       "1584  <span title=\"<span title=\"YBX1 (NSEP1, YB-1, Y...   \n",
       "1585  <span title=\"<span title=\"YBX1 (NSEP1, YB-1, Y...   \n",
       "1586  <span title=\"<span title=\"ZG16B (HRPE773, PRO1...   \n",
       "1587  <span title=\"<span title=\"ZP3 (ZP3A, ZP3B, ZP3...   \n",
       "\n",
       "                                        Ligand Location  \\\n",
       "0     <span title=\"based on perplexity, uniprot\">sec...   \n",
       "1     <span title=\"based on perplexity, uniprot\">cel...   \n",
       "2     <span title=\"based on hpa, uniprot\">cell membr...   \n",
       "3     <span title=\"based on hpa, perplexity, uniprot...   \n",
       "4     <span title=\"based on hpa, perplexity, uniprot...   \n",
       "...                                                 ...   \n",
       "1583  <span title=\"based on perplexity, uniprot\">sec...   \n",
       "1584  <span title=\"based on perplexity, uniprot\">sec...   \n",
       "1585  <span title=\"based on perplexity, uniprot\">sec...   \n",
       "1586  <span title=\"based on perplexity, uniprot\">cel...   \n",
       "1587  <span title=\"based on uniprot\">cell membrane</...   \n",
       "\n",
       "                                           Ligand group root_group_id  \n",
       "0     <a href=\"https://www.genenames.org/data/genegr...          1234  \n",
       "1     <a href=\"https://www.genenames.org/data/genegr...           471  \n",
       "2     <a href=\"https://www.genenames.org/data/genegr...          1302  \n",
       "3     <a href=\"https://www.genenames.org/data/genegr...          2104  \n",
       "4     <a href=\"https://www.genenames.org/data/genegr...           471  \n",
       "...                                                 ...           ...  \n",
       "1583  <a href=\"https://www.genenames.org/data/genegr...           106  \n",
       "1584  <a href=\"https://www.genenames.org/data/genegr...          1520  \n",
       "1585  <a href=\"https://www.genenames.org/data/genegr...          2045  \n",
       "1586  <a href=\"https://www.genenames.org/data/genegr...            NA  \n",
       "1587  <a href=\"https://www.genenames.org/data/genegr...             8  \n",
       "\n",
       "[1588 rows x 5 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pair_annot_ligand[\"Ligand group\"] = gene_pair_annot_ligand[\"Ligand group\"].fillna(\"unknown\")\n",
    "gene_pair_annot_ligand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e44ec81f-5abb-44b2-a7f9-c1c070281875",
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop duplicates \n",
    "gene_pair_annot_ligand = gene_pair_annot_ligand.drop_duplicates().reset_index(drop=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c02e520b-f361-490d-abdd-ec0b6e5792cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_annot_ligand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d56b4303-a4f8-4708-8a47-c0563f33dc8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Receptor group merge and tooltip\n",
    "gene_pair_annot_receptor = gene_pair_annot2[['Receptor HGNC ID', \n",
    "                                           'Receptor Symbols',\n",
    "                                           'receptor_hgnc_id',\n",
    "                                           'Receptor Location']].copy()\n",
    "# drop duplicates \n",
    "gene_pair_annot_receptor = gene_pair_annot_receptor.merge(gene_group_lim, how='left', left_on='receptor_hgnc_id', right_on='hgnc_id')\n",
    "gene_pair_annot_receptor = gene_pair_annot_receptor.rename(columns={\"root_group_name\": \"Receptor group\"}).drop(columns=[\"hgnc_id\",\"receptor_hgnc_id\"])\n",
    "\n",
    "gene_pair_annot_receptor[\"Receptor group\"] = gene_pair_annot_receptor[\"Receptor group\"].fillna(\"unknown\")\n",
    "# gene_pair_annot_receptor[\"Receptor group\"] = gene_pair_annot_receptor[\"Receptor group\"].apply(\n",
    "#     lambda x: f'<span title=\"{x}\">{x}</span>' if pd.notna(x) else \"unknown\"\n",
    "# )\n",
    "gene_pair_annot_receptor = gene_pair_annot_receptor.drop_duplicates().reset_index(drop=True)\n",
    "gene_pair_annot_receptor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40e05d84-508d-4d69-a16f-aca85028c17f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_annot_ligand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a8dbbef-2c0c-498c-aaa6-8e90dde2c696",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import jinja2\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "import base64\n",
    "import re\n",
    "\n",
    "# Add the src directory to the path for importing modules\n",
    "sys.path.append(os.path.abspath(\"src\"))\n",
    "\n",
    "# Import necessary modules from your existing src files\n",
    "# Ensure createDataTable and createFunctionalAnnotTable are in your 'src' directory\n",
    "from createDataTable import pop_up_info, gene_pair0, generate_perplexity_links, gene_pair00\n",
    "from createFunctionalAnnotTable import gene_pair_annot_ligand, gene_pair_annot_receptor\n",
    "gene_pair_annot_ligand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34ed0a5b-a068-4fee-ac7a-e513005a4628",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Test or all\n",
    "test = True\n",
    "test_genes = [\"VEGFA ITGB1\", \"VEGFA KDR\", \"VEGFA NRP1\", \"THPO MPL\", \"FGF1 FGFR3\"] # Example genes\n",
    "# --- Paths ---\n",
    "MERGED_TEMPLATE_PATH = 'HTML/mergedCardWithPMIDTemplate.html'\n",
    "OUTPUT_DIR = 'data/cards/' # New output directory for combined files\n",
    "\n",
    "# --- Load and Preprocess Data (Combined from both scripts) ---\n",
    "\n",
    "# Load PubMed data (from createPMIDpages.py)\n",
    "pubmed_data = pd.read_csv(\"data/pubmed_results.csv\")\n",
    "pubmed_data[\"Year\"] = pubmed_data[\"Year\"].astype(str).str.replace(\".0\", \"\", regex=False).astype(int)\n",
    "pubmed_data[\"PMID\"] = pubmed_data[\"PMID\"].astype(str)\n",
    "pubmed_data = pubmed_data.reset_index(drop=True)\n",
    "\n",
    "# Load LLM results (from createPMIDpages.py)\n",
    "bio_keywords = pd.read_csv(\"data/llm_results.csv\")\n",
    "\n",
    "# --- Prepare gene_pair00 for PMID section (from createPMIDpages.py) ---\n",
    "# gene_pair00 is used for PMID and Keywords, so it needs the '——' placeholder\n",
    "# Ensure gene_pair00 is a copy to avoid SettingWithCopyWarning later\n",
    "gene_pair00_copy = gene_pair00.copy()\n",
    "gene_pair00_copy[\"Human LR Pair\"] = gene_pair00_copy[\"Human LR Pair\"].str.replace(\" \", \"——\")\n",
    "\n",
    "# Merge with LLM results\n",
    "gene_pair000 = gene_pair00_copy.merge(bio_keywords, how='left', left_on=\"Human LR Pair\", right_on='Human LR Pair')\n",
    "gene_pair000[\"Relevance Keywords\"] = gene_pair000[\"Relevance Keywords\"].astype(str)\n",
    "gene_pair000[\"Human LR Pair\"] = gene_pair000[\"Human LR Pair\"].astype(str) # Ensure string type\n",
    "\n",
    "# --- Prepare gene_pair0 for Card section (from createCards.py) ---\n",
    "# gene_pair0 is used for card details, it should retain spaces for splitting gene names\n",
    "# Ensure gene_pair0 is a copy to avoid SettingWithCopyWarning later\n",
    "gene_pair0_copy = gene_pair0.copy()\n",
    "\n",
    "# Add Disease (specific) to cards\n",
    "df_disease = pd.read_csv(\"data/disease_annotations_per_pair.csv\")\n",
    "df_disease = df_disease.groupby('interaction')['disease'].apply(', '.join).reset_index()\n",
    "mapping_disease = dict(zip(df_disease['interaction'], df_disease['disease']))\n",
    "gene_pair0_copy[\"Disease\"] = gene_pair0_copy['Human LR Pair'].map(mapping_disease).fillna(\"unknown\")\n",
    "\n",
    "gene_pair0_copy = generate_perplexity_links(\n",
    "    gene_pair0_copy,\n",
    "    pathway_col=\"Disease\",\n",
    "    default_query_template=\"What-diseases-is-the-ligand-receptor-pair-{pair}-associated-with\"\n",
    ")\n",
    "\n",
    "gene_pair0_copy[\"Interaction ID\"] = gene_pair0_copy[\"Interaction ID\"].apply(\n",
    "    lambda x: f\"<a href='https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database/filter/{x}.html'>{x}</a>\"\n",
    ")\n",
    "# Add external link icon\n",
    "icon_html = '<i class=\"fa-solid fa-arrow-up-right-from-square\" style=\"margin-left:4px;\"></i></a>'\n",
    "columns_to_update = [\n",
    "    \"KEGG Pathway\", \"PROGENy Pathway\", \"Cancer-related\",\n",
    "    \"Disease Type\", \"Disease\"\n",
    "]\n",
    "for col in columns_to_update:\n",
    "    gene_pair0_copy[col] = gene_pair0_copy[col].str.replace(\n",
    "        \"</a>\", icon_html, regex=False\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77017801-a1eb-4cf6-b610-be483f528498",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_annot_ligand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39718c1c-4b1d-4c5c-99bf-ebc3cb3a6cde",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import pandas as pd\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\", category=pd.errors.DtypeWarning)\n",
    "from itables import init_notebook_mode, show\n",
    "from IPython.display import display, Javascript\n",
    "import itables.options as opt\n",
    "# Change working directory to ConnectomeDB\n",
    "project_root = os.path.dirname(os.getcwd())\n",
    "os.chdir(project_root)\n",
    "sys.path.append(os.path.abspath(\"src\"))\n",
    "from createDataTable_perSpecies import mouse_gene_pair1\n",
    "\n",
    "mouse_gene_pair1.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e92cc566-d480-40e3-9593-6d6ccc6faeab",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_ID = dict(zip(gene_pair0['Human LR Pair'], gene_pair0['Interaction ID']))\n",
    "gene_pair_PMID = fetchGSheet.gene_pair.dropna(axis=1, how='all')\n",
    "gene_pair_PMID = gene_pair_PMID[gene_pair_PMID['LR pair'] != '']\n",
    "gene_pair_PMID= gene_pair_PMID[[\"LR pair\", \"PMID\", \"original source\"]]\n",
    "gene_pair_PMID\n",
    "\n",
    "# Mapping for replacements\n",
    "mapping = dict(zip(fetchGSheet.src_info['original source'], fetchGSheet.src_info['shortname']))\n",
    "# Replace values in the column based on the mapping\n",
    "gene_pair_PMID['Database Source'] = gene_pair_PMID['original source'].replace(mapping)\n",
    "\n",
    "# Replace values in the column based on the mapping\n",
    "gene_pair_PMID[\"Interaction ID\"] = gene_pair_PMID['LR pair'].replace(mapping_ID)\n",
    "df_pub = pd.read_csv(\"data/pubmed_results.csv\", usecols=[0,1,3,4,5])\n",
    "gene_pair_PMID[\"PMID\"] = gene_pair_PMID[\"PMID\"].astype(str)\n",
    "df_pub[\"PMID\"] = df_pub[\"PMID\"].astype(str)\n",
    "gene_pair_trip = pd.merge(gene_pair_PMID, df_pub, how='left', on='PMID')\n",
    "### patch for the BioRxiV ###\n",
    "gene_pair_trip[\"Year\"] = gene_pair_trip[\"Year\"].apply(\n",
    "    lambda x: \"2024\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"Species Specificity\", \"NaN\", \"\"] else x\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69c017c9-4e4c-4da6-93e4-d5113b155977",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair.columns[11]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9336cd4-c110-4d50-a480-169a7ed42f4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_trip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46a4672c-7ac5-404e-aaf1-938720902b11",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_trip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e2d9abe-8ec1-4127-a7ae-7b12dc8c5a85",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_group_lim['root_group_name']  =[\n",
    "        f'<a href=\"https://www.genenames.org/data/genegroup/#!/group/{root_group_id}\" target=\"_blank\">{root_group_name}</a>' \n",
    "        if pd.notna(root_group_id) and root_group_id.strip() else \"\" \n",
    "        for root_group_id, root_group_name in zip(gene_group_lim[\"root_group_id\"], gene_group_lim[\"root_group_name\"])\n",
    "    ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dac8dafa-b780-42d4-8935-acf14f17bda4",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_group_lim[gene_group_lim[\"root_group_id\"]==\"NA\"]\n",
    "gene_group_lim[\"root_group_name\"] = gene_group_lim[\"root_group_name\"].apply(\n",
    "    lambda x: \"unknown\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"na\", \"\"] else x\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a56c2385-b0e5-4719-849a-59a0a5ccf3af",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_group_lim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7052fe33-214c-4316-b406-f92f5df5a5f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_group_lim[['root_group_name']]  =[\n",
    "        f'<a href=\"https://www.genenames.org/data/genegroup/#!/group/{root_group_id}\" target=\"_blank\">{root_group_name}</a>' \n",
    "        if pd.notna(root_group_id) and root_group_id.strip() else \"\" \n",
    "        for root_group_id, root_group_name in zip(gene_group_lim[\"root_group_id\"], gene_group_lim[\"root_group_name\"])\n",
    "    ]\n",
    "gene_group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a68e94b7-094a-414e-b51d-730e695a5cee",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71aa1780-a341-42d7-9e9d-1ceaa976991d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_annot_ligand[\"Ligand group\"]\n",
    "ligand_mapping = dict(zip(gene_pair_annot_ligand['Ligand HGNC ID'], gene_pair_annot_ligand['Ligand group']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eda63edf-512a-4918-aec1-9b8c8da14172",
   "metadata": {},
   "outputs": [],
   "source": [
    "ligand_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "902249cd-66bf-478d-96a8-bff46eb5826e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add Disease Category per pair\n",
    "df= pd.read_csv(\"data/disease_annotations_per_pair.csv\")\n",
    "df_cat=pd.read_csv(\"data/disease_categories.csv\")\n",
    "mapping = dict(zip(df_cat['Disease Name'], df_cat['Category']))\n",
    "# Replace values in the column based on the mapping\n",
    "df[\"Disease Type\"] = df['disease'].replace(mapping)\n",
    "df = df[[\"interaction\", \"Disease Type\"]].drop_duplicates()\n",
    "df['Disease Type'] = df['Disease Type'].astype(str)\n",
    "df = df.sort_values(by='Disease Type', ascending=True)\n",
    "# Group by 'col1' and combine 'col2' values with ', '\n",
    "df = df.groupby('interaction')['Disease Type'].apply(', '.join).reset_index()\n",
    "# Create \"Cancer-related\" column based on whether \"Cancers & Neoplasms\" is in col2\n",
    "df['Cancer-related'] = df['Disease Type'].apply(lambda x: 'Yes' if 'Cancer' in x else 'No')\n",
    "disease_df = df[df[\"interaction\"].isin(LR_pairs)]\n",
    "# Function to update the \"Cancer-related\" column and modify \"col2\" if needed\n",
    "\n",
    "gene_pair = gene_pair.merge(disease_df, how='left', left_on='Human LR Pair', right_on='interaction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6fb7f07-e57c-4f26-bb95-0936d461b675",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to prepare datatables (cleaning and hyperlinking, adding tool tips, etc) input for the database qmds\n",
    "import sys, os\n",
    "from itables import init_notebook_mode\n",
    "import pandas as pd\n",
    "from itables import show\n",
    "from itables import options\n",
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "import fetchGSheet \n",
    "import warnings\n",
    "import urllib.parse\n",
    "\n",
    "# Suppress SettingWithCopyWarning\n",
    "warnings.simplefilter(\"ignore\", category=UserWarning)\n",
    "\n",
    "\n",
    "# Other vertebrates\n",
    "species_list = [\n",
    "    \"ptroglodytes\", \"ggallus\", \"sscrofa\", \"btaurus\", \n",
    "    \"clfamiliaris\", \"ecaballus\", \"oarambouillet\",\n",
    "    \"cjacchus\", \"mmulatta\"\n",
    "]\n",
    "\n",
    "# Select only the relevant columns from pop_up_info\n",
    "cols_to_keep = cols_to_keep = list(range(0, 30)) \n",
    "# Step 3: Load file using only the desired columns\n",
    "df = pd.read_table(\"data/HGNC_gene_info_full.tsv\", usecols=cols_to_keep)\n",
    "pop_up_info = pd.read_table(\"data/HGNC_gene_info_full.tsv\")\n",
    "pop_up_info = pop_up_info.rename(columns={\"hgnc_id\": \"HGNC ID\", \n",
    "                                          \"name\": \"Approved name\",\n",
    "                                          \"symbol\": \"Approved symbol\",\n",
    "                                          \"rgd_id\": \"RGD ID\",\n",
    "                                          \"mgd_id\": \"MGI ID\", \n",
    "                                          \"rgd_id\": \"RGD ID\",\n",
    "                                          \"alias_symbol\": \"Alias symbol\", # add to table\n",
    "                                          \"prev_symbol\": \"Previous symbol\", # add to table\n",
    "                                          \"date_symbol_changed\": \"Date symbol changed\"\n",
    "                                         })\n",
    "\n",
    "# Keep only first MGI/RGD ID\n",
    "pop_up_info[\"MGI ID\"] = pop_up_info[\"MGI ID\"].str.split(\"|\").str[0]\n",
    "pop_up_info[\"RGD ID\"] = pop_up_info[\"RGD ID\"].str.split(\"|\").str[0]\n",
    "\n",
    "pop_up_info[\"Alias symbol\"] = pop_up_info[\"Alias symbol\"].apply(\n",
    "    lambda x: \"N/A\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\"] else x\n",
    ")\n",
    "\n",
    "pop_up_info[\"Previous symbol\"] = pop_up_info[\"Previous symbol\"].apply(\n",
    "    lambda x: \"N/A\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\"] else x\n",
    ")\n",
    "\n",
    "# Replace \"|\" with \", \"\n",
    "pop_up_info[\"Alias symbol\"] = [value.replace(\"|\", \", \") for value in pop_up_info[\"Alias symbol\"]]\n",
    "pop_up_info[\"Previous symbol\"] = [value.replace(\"|\", \", \") for value in pop_up_info[\"Previous symbol\"]]\n",
    "\n",
    "pop_up_info[\"Date symbol changed\"] = pop_up_info[\"Date symbol changed\"].apply(\n",
    "    lambda x: \"N/A\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\"] else x\n",
    ")\n",
    "\n",
    "\n",
    "pop_up_info_lim = pop_up_info[[\"HGNC ID\", \"Approved name\", \"MGI ID\", \"RGD ID\", \"Alias symbol\",\n",
    "                               \"Approved symbol\", \"Previous symbol\"]] # rm \"Approved symbol\" for now\n",
    "pop_up_info_lim = pop_up_info_lim.drop_duplicates(subset=\"HGNC ID\", keep=\"first\")\n",
    "\n",
    "# Drop columns where all values are NA in gene_pair\n",
    "gene_pair = fetchGSheet.gene_pair.dropna(axis=1, how='all')\n",
    "gene_pair = gene_pair[gene_pair['LR pair'] != '']\n",
    "# for now set source count as triplicates\n",
    "sourceCount = len(gene_pair[['LR pair']])\n",
    "\n",
    "# for now, keep only the following columns\n",
    "gene_pair = gene_pair[['LR pair', 'Ligand', 'Ligand.HGNC', 'Receptor', 'Receptor.HGNC',\n",
    "                       'perplexity link', 'PMID', 'binding location', \n",
    "                       'bind in trans?', 'bidirectional signalling?',\n",
    "                       'interaction type', 'original source']]\n",
    "\n",
    "gene_pair = gene_pair.dropna(subset=['LR pair'])\n",
    "\n",
    "# some PMIDs kick in with \",\" so replace\n",
    "gene_pair[\"PMID\"] = [value.replace(\",\", \"\") for value in gene_pair[\"PMID\"]]\n",
    "gene_pair = gene_pair.dropna(subset=['PMID'])\n",
    "\n",
    "# Mapping for replacements\n",
    "mapping = dict(zip(fetchGSheet.src_info['original source'], fetchGSheet.src_info['shortname']))\n",
    "# Replace values in the column based on the mapping\n",
    "gene_pair['original source'] = gene_pair['original source'].replace(mapping)\n",
    "\n",
    "## add Ligand/Receptor Location\n",
    "def dedup_locations(loc_str):\n",
    "    # Split, strip, deduplicate, and sort\n",
    "    parts = [loc.strip() for loc in loc_str.split(',') if loc.strip()]\n",
    "    unique_sorted = sorted(set(parts), key=str.lower)  # case-insensitive sort\n",
    "    return unique_sorted\n",
    "\n",
    "def generate_LocToolTip(row, geneloc, loc_col):\n",
    "    ligand = row[loc_col]\n",
    "    original_locations = [loc.strip() for loc in row[\"location\"].split(',')]\n",
    "    original_sources = [src.strip() for src in row[\"source\"].split(',')]\n",
    "\n",
    "    # Get deduplicated locations\n",
    "    unique_locations = dedup_locations(row[\"location\"])\n",
    "\n",
    "    if len(unique_locations) == 1:\n",
    "        # Single tooltip case\n",
    "        location = unique_locations[0]\n",
    "        matching_rows = geneloc[(geneloc[loc_col] == ligand) & (geneloc[\"location\"].str.contains(location))]\n",
    "        all_sources = matching_rows[\"source\"].unique()\n",
    "        sources_str = \", \".join(sorted(set(all_sources)))\n",
    "        return f'<span title=\"based on {sources_str}\">{location}</span>'\n",
    "    else:\n",
    "        # Multiple tooltips — find each (ligand, location) match in original df\n",
    "        spans = []\n",
    "        for loc in unique_locations:\n",
    "            matching_rows = geneloc[\n",
    "                (geneloc[loc_col] == ligand) &\n",
    "                (geneloc[\"location\"].str.contains(loc))\n",
    "            ]\n",
    "            all_sources = matching_rows[\"source\"].unique()\n",
    "            sources_str = \", \".join(sorted(set(all_sources)))\n",
    "            spans.append(f'<span title=\"based on {sources_str}\">{loc}</span>')\n",
    "        return \", \".join(spans)\n",
    "\n",
    "\n",
    "# Group the original loc_info by Ligand\n",
    "ligand_loc = fetchGSheet.ligand_loc.dropna(axis=1, how='all')\n",
    "grouped = ligand_loc.groupby(\"Ligand\").agg({\n",
    "    \"location\": lambda x: ', '.join(x),\n",
    "    \"source\": lambda x: ', '.join(x)\n",
    "}).reset_index()\n",
    "\n",
    "# Generate tooltips\n",
    "grouped[\"Ligand location\"] = grouped.apply(lambda row: generate_LocToolTip(row, ligand_loc,loc_col=\"Ligand\"), axis=1)\n",
    "# create dict\n",
    "mapping_loc = dict(zip(grouped['Ligand'], grouped['Ligand location'])) \n",
    "gene_pair['Ligand location'] = gene_pair['Ligand'].replace(mapping_loc)\n",
    "\n",
    "\n",
    "# Group the original loc_info by Receptor\n",
    "receptor_loc = fetchGSheet.receptor_loc.dropna(axis=1, how='all')\n",
    "grouped = receptor_loc.groupby(\"Receptor\").agg({\n",
    "    \"location\": lambda x: ', '.join(x),\n",
    "    \"source\": lambda x: ', '.join(x)\n",
    "}).reset_index()\n",
    "\n",
    "# Generate tooltips\n",
    "grouped[\"Receptor location\"] = grouped.apply(lambda row: generate_LocToolTip(row, receptor_loc,loc_col=\"Receptor\"), axis=1)\n",
    "# create dict\n",
    "mapping_loc = dict(zip(grouped['Receptor'], grouped['Receptor location'])) \n",
    "gene_pair['Receptor location'] = gene_pair['Receptor'].replace(mapping_loc)\n",
    "\n",
    "\n",
    "# Set missing mappings to 'unknown'\n",
    "gene_pair.loc[gene_pair['Ligand location'] == gene_pair['Ligand'], 'Ligand location'] = 'unknown'\n",
    "gene_pair.loc[gene_pair['Receptor location'] == gene_pair['Receptor'], 'Receptor location'] = 'unknown'\n",
    "# Set \"n/a\" to unknown\n",
    "gene_pair['Ligand location'] = [value.replace(\"n/a\", \"unknown\") for value in gene_pair['Ligand location']]\n",
    "gene_pair['Receptor location'] = [value.replace(\"n/a\", \"unknown\") for value in gene_pair['Receptor location']]\n",
    "\n",
    "# Fetch species IDs from the dataset\n",
    "hgnc_id = [col for col in gene_pair.columns if \"HGNC\" in col]\n",
    "hgnc_id = pd.concat([gene_pair[col] for col in hgnc_id]).unique()\n",
    "\n",
    "# Rename columns for better clarity\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "    \"LR pair\": \"Human LR Pair\",\n",
    "    \"Ligand.HGNC\": \"Ligand HGNC ID\",\n",
    "    \"Receptor.HGNC\": \"Receptor HGNC ID\",\n",
    "    \"perplexity link\": \"Perplexity\", # will be replaced with actual link later\n",
    "    \"original source\": \"Database Source\",\n",
    "    \"Ligand location\": \"Ligand Location\",\n",
    "    \"Receptor location\": \"Receptor Location\",\n",
    "    \"binding location\": \"Binding Location\",\n",
    "    \"bind in trans?\" : \"Trans-binding\", \n",
    "    \"bidirectional signalling?\": \"Bidirectional Signalling\",\n",
    "    \"interaction type\" : \"Interaction Type\"\n",
    "    #\"PMID\": \"PMID support\" # was PMID support\n",
    "})\n",
    "\n",
    "\n",
    "# Merge gene_pair with pop_up_info_lim for Ligand(L)\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Ligand HGNC ID', right_on='HGNC ID')\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Ligand name\", \n",
    "                                     \"MGI ID\": \"Ligand MGI ID\",\n",
    "                                     \"RGD ID\": \"Ligand RGD ID\",\n",
    "                                      \"Alias symbol\": \"Ligand Aliases\",\n",
    "                                      \"Previous symbol\": \"Ligand Old symbol\",\n",
    "                                     },\n",
    "                            )\n",
    "gene_pair = gene_pair.drop(columns=[\"HGNC ID\", \"Approved symbol\"])\n",
    "# Add top pathway per pair\n",
    "LR_pairs = gene_pair[\"Human LR Pair\"].unique()\n",
    "df= pd.read_csv(\"data/pathway_annotations_per_pair.csv\")\n",
    "#df = df[df[\"interaction\"].isin(LR_pairs)]\n",
    "# Sort by absolute value of 'weight', descending (larger abs(weight) first)\n",
    "df_sorted = df.reindex(df['weight'].abs().sort_values(ascending=False).index)\n",
    "# Keep only the first occurrence for each unique 'interaction'\n",
    "#df_unique = df_sorted.drop_duplicates(subset='interaction', keep='first')\n",
    "#Keep ALL\n",
    "df = df_sorted.reset_index(drop=True)\n",
    "top_pathway_df = df[[\"interaction\", \"source\"]]\n",
    "top_pathway_df = top_pathway_df.groupby('interaction')['source'].apply(', '.join).reset_index()\n",
    "top_pathway_df = top_pathway_df.rename(columns={\n",
    "                                      \"source\": \"PROGENy Pathway\"\n",
    "})\n",
    "top_pathway_df[\"interaction\"] = [value.replace(\"^\", \" \") for value in top_pathway_df[\"interaction\"]]\n",
    "gene_pair = gene_pair.merge(top_pathway_df, how='left', left_on='Human LR Pair', right_on='interaction')\n",
    "gene_pair = gene_pair.drop(columns=[\"interaction\"])\n",
    "#df = df_unique.reset_index(drop=True)\n",
    "top_pathway_df=fetchGSheet.kegg_pathway_info[[\"LR Pair\", \"kegg_pathway_id\", \"kegg_relationship\", \"kegg_pathway_name\"]].copy()\n",
    "# add link to kegg_pathway_name\n",
    "top_pathway_df[\"kegg_pathway_name\"] = [\n",
    "    f'<a href=\"https://www.kegg.jp/pathway/{kegg_id}\" target=\"_blank\">{name}</a>'\n",
    "    for kegg_id, name in zip(top_pathway_df[\"kegg_pathway_id\"], top_pathway_df[\"kegg_pathway_name\"])\n",
    "]\n",
    "\n",
    "# link to kegg_pathway_id\n",
    "top_pathway_df[\"kegg_pathway_id\"] = [\n",
    "    f'<a href=\"https://www.kegg.jp/pathway/{id}\" target=\"_blank\">{id}</a>'\n",
    "    for id in top_pathway_df[\"kegg_pathway_id\"]]\n",
    "\n",
    "\n",
    "top_pathway_df = top_pathway_df.rename(columns={\n",
    "                                      \"kegg_pathway_name\": \"KEGG Pathway\",\n",
    "                                      \"kegg_relationship\": \"KEGG relationship\",\n",
    "                                      \"kegg_pathway_id\": \"KEGG Pathway ID\"\n",
    "    \n",
    "})\n",
    "top_pathway_df1 = top_pathway_df[[\"LR Pair\", \"KEGG Pathway\"]].drop_duplicates()\n",
    "top_pathway_df1 = top_pathway_df1.groupby('LR Pair')['KEGG Pathway'].apply(', '.join).reset_index()\n",
    "gene_pair = gene_pair.merge(top_pathway_df1, how='left', left_on='Human LR Pair', right_on='LR Pair')\n",
    "gene_pair = gene_pair.drop(columns=[\"LR Pair\"])\n",
    "\n",
    "# Add Disease Category per pair\n",
    "df= pd.read_csv(\"data/disease_annotations_per_pair.csv\")\n",
    "df_cat=pd.read_csv(\"data/disease_categories.csv\")\n",
    "mapping = dict(zip(df_cat['Disease Name'], df_cat['Category']))\n",
    "# Replace values in the column based on the mapping\n",
    "df[\"Disease Type\"] = df['disease'].replace(mapping)\n",
    "df = df[[\"interaction\", \"Disease Type\"]].drop_duplicates()\n",
    "df['Disease Type'] = df['Disease Type'].astype(str)\n",
    "df = df.sort_values(by='Disease Type', ascending=True)\n",
    "# Group by 'col1' and combine 'col2' values with ', '\n",
    "df = df.groupby('interaction')['Disease Type'].apply(', '.join).reset_index()\n",
    "# Create \"Cancer-related\" column based on whether \"Cancers & Neoplasms\" is in col2\n",
    "df['Cancer-related'] = df['Disease Type'].apply(lambda x: 'Yes' if 'Cancer' in x else 'No')\n",
    "disease_df = df[df[\"interaction\"].isin(LR_pairs)]\n",
    "# Function to update the \"Cancer-related\" column and modify \"col2\" if needed\n",
    "\n",
    "gene_pair = gene_pair.merge(disease_df, how='left', left_on='Human LR Pair', right_on='interaction')\n",
    "\n",
    "# Add MGI annotation\n",
    "MGI_info = pd.read_csv(\"data/MGI_ID_biomart.csv\")\n",
    "gene_pair = gene_pair.merge(MGI_info, how='left', left_on='Ligand MGI ID', right_on='MGI ID')\n",
    "\n",
    "# Find rows where Ligand HGNC ID is missing & copy Ligand to MGI name for those rows\n",
    "mask = gene_pair['Ligand HGNC ID'].astype(str).str.strip() == ''\n",
    "gene_pair.loc[mask, 'MGI name'] = gene_pair.loc[mask, 'Ligand']\n",
    "# Map MGI ID using the MGI_info table\n",
    "gene_pair = gene_pair.merge(MGI_info, left_on='MGI name', right_on='MGI name', how='left', suffixes=('', '_from_info'))\n",
    "# Fill missing 'MGI ID' only where it was previously missing\n",
    "gene_pair['Ligand MGI ID'] = gene_pair['Ligand MGI ID'].combine_first(gene_pair['MGI ID_from_info'])\n",
    "gene_pair = gene_pair.drop(columns=['MGI ID_from_info'])\n",
    "\n",
    "# Add RGD annotation\n",
    "RGD_info = pd.read_csv(\"data/RGD_ID_biomart.csv\")\n",
    "RGD_info['RGD ID'] = \"RGD:\" + RGD_info['RGD ID'].astype(str)\n",
    "gene_pair = gene_pair.merge(RGD_info, how='left', left_on='Ligand RGD ID', right_on='RGD ID')\n",
    "\n",
    "# Add ZFIN id and symbol\n",
    "ZFIN_info = pd.read_csv(\"data/ZFIN_ID_human_orthos.txt\", sep=\"\\t\", skiprows=1)\n",
    "ZFIN_info = ZFIN_info[['ZFIN ID', 'ZFIN Symbol', 'ZFIN Name', 'HGNC ID']]\n",
    "\n",
    "ZFIN_info = ZFIN_info.dropna(subset=['HGNC ID'])\n",
    "ZFIN_info = ZFIN_info.drop_duplicates(subset=['HGNC ID'])\n",
    "ZFIN_info['HGNC ID'] = ZFIN_info['HGNC ID'].apply(lambda x: f'HGNC:{int(x)}')\n",
    "gene_pair = gene_pair.merge(ZFIN_info, how='left', left_on='Ligand HGNC ID', right_on='HGNC ID')\n",
    "\n",
    "gene_pair = gene_pair.drop(columns=[\"RGD ID\", \"MGI ID\", \"HGNC ID\", \"interaction\"])\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "                                     \"MGI name\": \"Mouse Ligand\", \n",
    "                                     \"RGD name\": \"Rat Ligand\",\n",
    "                                     \"ZFIN ID\": \"Ligand ZFIN ID\",\n",
    "                                     \"ZFIN Symbol\": \"Zebrafish Ligand\",\n",
    "                                     \"ZFIN Name\": \"Zebrafish Ligand name\"}\n",
    "                            )\n",
    "\n",
    "gene_pair = gene_pair.merge(pop_up_info_lim, how='left', left_on='Receptor HGNC ID', right_on='HGNC ID')\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\"Approved name\": \"Receptor name\",\n",
    "                                      \"MGI ID\": \"Receptor MGI ID\",\n",
    "                                      \"RGD ID\": \"Receptor RGD ID\",\n",
    "                                      \"Alias symbol\": \"Receptor Aliases\",\n",
    "                                      \"Previous symbol\": \"Receptor Old symbol\",}\n",
    "                            )\n",
    "\n",
    "\n",
    "gene_pair = gene_pair.drop(columns=[\"HGNC ID\"])\n",
    "\n",
    "# Add new columns where all Ligand Symbol & Aliases and Receptor Symbol & Aliases merged in one column\n",
    "def format_symbol_aliases(symbol, old_symbol, aliases):\n",
    "    # Filter out \"N/A\" values\n",
    "    parts = [p for p in (old_symbol, aliases) if p != \"N/A\"]\n",
    "    # Return just the symbol if no valid aliases or old symbols\n",
    "    return f\"{symbol} ({', '.join(parts)})\" if parts else symbol\n",
    "\n",
    "gene_pair['Ligand Symbol & Aliases'] = gene_pair.apply(\n",
    "    lambda row: format_symbol_aliases(row['Ligand'], row['Ligand Old symbol'], row['Ligand Aliases']),\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "gene_pair['Receptor Symbol & Aliases'] = gene_pair.apply(\n",
    "    lambda row: format_symbol_aliases(row['Receptor'], row['Receptor Old symbol'], row['Receptor Aliases']),\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "### tooltips \n",
    "gene_pair[\"Ligand Symbol & Aliases\"] = [\n",
    "    f'<span title=\"{aliases}\">{aliases}</span>'\n",
    "    for aliases in gene_pair[\"Ligand Symbol & Aliases\"]\n",
    "]\n",
    "gene_pair[\"Receptor Symbol & Aliases\"] = [\n",
    "    f'<span title=\"{aliases}\">{aliases}</span>'\n",
    "    for aliases in gene_pair[\"Receptor Symbol & Aliases\"]\n",
    "]\n",
    "\n",
    "\n",
    "# Add MGI name\n",
    "gene_pair = gene_pair.merge(MGI_info, how='left', left_on='Receptor MGI ID', right_on='MGI ID')\n",
    "# Find rows where Receptor HGNC ID is missing & copy Receptor to MGI name for those rows\n",
    "mask = gene_pair['Ligand HGNC ID'].astype(str).str.strip() == ''\n",
    "gene_pair.loc[mask, 'MGI name'] = gene_pair.loc[mask, 'Receptor']\n",
    "# Map MGI ID using the MGI_info table\n",
    "gene_pair = gene_pair.merge(MGI_info, left_on='MGI name', right_on='MGI name', how='left', suffixes=('', '_from_info'))\n",
    "# Fill missing 'MGI ID' only where it was previously missing\n",
    "gene_pair['Receptor MGI ID'] = gene_pair['Receptor MGI ID'].combine_first(gene_pair['MGI ID_from_info'])\n",
    "gene_pair = gene_pair.drop(columns=['MGI ID_from_info'])\n",
    "\n",
    "gene_pair = gene_pair.merge(RGD_info, how='left', left_on='Receptor RGD ID', right_on='RGD ID')\n",
    "gene_pair = gene_pair.merge(ZFIN_info, how='left', left_on='Receptor HGNC ID', right_on='HGNC ID')\n",
    "gene_pair = gene_pair.drop(columns=[\"RGD ID\", \"MGI ID\", \"HGNC ID\"])\n",
    "\n",
    "gene_pair = gene_pair.rename(columns={\n",
    "                                     \"MGI name\": \"Mouse Receptor\", \n",
    "                                     \"RGD name\": \"Rat Receptor\",\n",
    "                                     \"ZFIN ID\": \"Receptor ZFIN ID\",\n",
    "                                     \"ZFIN Symbol\": \"Zebrafish Receptor\",\n",
    "                                     \"ZFIN Name\": \"Zebrafish Receptor name\"}\n",
    "                            )\n",
    "\n",
    "#gene_pair = gene_pair.drop(columns=[\"Approved symbol_x\", \"Approved symbol_y\"])\n",
    "\n",
    "# Function to add species-specific species Enseml ID and symbol for all other species except for mouse, rat, and zebrafish\n",
    "def appendOtherSpeciesInfo(species, origDF):\n",
    "    species_name = {\n",
    "    \"ptroglodytes\": \"Chimpanzee\",\n",
    "    \"ggallus\": \"Chicken\",\n",
    "    \"sscrofa\": \"Pig\",\n",
    "    \"btaurus\": \"Cow\",\n",
    "    \"clfamiliaris\": \"Dog\",\n",
    "    \"ecaballus\": \"Horse\",\n",
    "    \"oarambouillet\": \"Sheep\",\n",
    "    \"cjacchus\": \"Marmoset\",\n",
    "    \"mmulatta\": \"Rhesus Monkey\"    \n",
    "    }.get(species, \"Unknown species\")\n",
    "    \n",
    "    # Load species-specific data\n",
    "    species_info = pd.read_csv(f\"data/{species}_ID_biomart.csv\")\n",
    "\n",
    "    # Keep relevant columns\n",
    "    species_info = species_info[[f\"{species}_homolog_ensembl_gene\", \n",
    "                                 f\"{species}_homolog_associated_gene_name\", \n",
    "                                 'hgnc_id']]\n",
    "\n",
    "    # Remove rows where 'hgnc_id' is NaN and drop duplicates\n",
    "    species_info = species_info.dropna(subset=['hgnc_id'])\n",
    "    species_info = species_info.drop_duplicates(subset=['hgnc_id'])\n",
    "\n",
    "    # Merge with ligand data\n",
    "    origDF = origDF.merge(species_info, how='left', \n",
    "                           left_on='Ligand HGNC ID', right_on='hgnc_id')\n",
    "    \n",
    "    # Rename columns for ligand info\n",
    "    origDF = origDF.rename(columns={\n",
    "        f\"{species}_homolog_associated_gene_name\": f\"{species_name} Ligand\", \n",
    "        f\"{species}_homolog_ensembl_gene\": f\"{species_name} Ligand Ensembl ID\"\n",
    "    })\n",
    "\n",
    "    # Drop duplicate 'hgnc_id' column\n",
    "    origDF = origDF.drop(columns=['hgnc_id'])\n",
    "\n",
    "    # Merge with receptor data\n",
    "    origDF = origDF.merge(species_info, how='left', \n",
    "                           left_on='Receptor HGNC ID', right_on='hgnc_id')\n",
    "\n",
    "    # Rename columns for receptor info\n",
    "    origDF = origDF.rename(columns={\n",
    "        f\"{species}_homolog_associated_gene_name\": f\"{species_name} Receptor\", \n",
    "        f\"{species}_homolog_ensembl_gene\": f\"{species_name} Receptor Ensembl ID\"\n",
    "    })\n",
    "\n",
    "        # Drop duplicate 'hgnc_id' column\n",
    "    origDF = origDF.drop(columns=['hgnc_id'])\n",
    "\n",
    "    # Drop columns where all values are NaN\n",
    "    origDF = origDF.dropna(axis=1, how='all')\n",
    "\n",
    "    return origDF\n",
    "\n",
    "\n",
    "# Loop through each species and update gene_pair\n",
    "for species in species_list:\n",
    "    gene_pair = appendOtherSpeciesInfo(species, gene_pair)\n",
    "\n",
    "# Drop columns where all values are NA in gene_pair\n",
    "gene_pair = gene_pair.dropna(axis=1, how='all')\n",
    "\n",
    "gene_pair = gene_pair.fillna(\" \")\n",
    "gene_pair = gene_pair[gene_pair['Human LR Pair'] != ' ']\n",
    "\n",
    "# if \"PMID link\" in gene_pair.columns:\n",
    "#    gene_pair = gene_pair.drop(columns=[\"PMID link\"])\n",
    "\n",
    "# Add\n",
    "first_columns=['Human LR Pair', 'Ligand', 'Receptor', 'Database Source']\n",
    "\n",
    "#end_columns=['HGNC L R', 'sanity check', 'curator', 'secondary source?']\n",
    "#gene_pair = gene_pair[first_columns + [col for col in gene_pair.columns if col not in first_columns + end_columns] + end_columns]\n",
    "gene_pair = gene_pair[first_columns + [col for col in gene_pair.columns if col not in first_columns]]\n",
    "\n",
    "# number of unique vars\n",
    "\n",
    "lrPairsCount = len(gene_pair[\"Human LR Pair\"].unique())\n",
    "\n",
    "ligandCount = len(gene_pair[\"Ligand\"].unique())\n",
    "\n",
    "receptorCount = len(gene_pair[\"Receptor\"].unique())\n",
    "\n",
    "# Mouse Orthologue\n",
    "MouseLigandCount = len(gene_pair[\"Ligand MGI ID\"].unique())\n",
    "\n",
    "MouseReceptorCount = len(gene_pair[\"Receptor MGI ID\"].unique())\n",
    "\n",
    "# Rat Orthologue\n",
    "RatLigandCount = len(gene_pair[\"Ligand RGD ID\"].unique())\n",
    "\n",
    "RatReceptorCount = len(gene_pair[\"Receptor RGD ID\"].unique())\n",
    "\n",
    "gene_pair[\"PMID\"] = [value.replace(\" \", \"\") for value in gene_pair[\"PMID\"]] # was'PMID support'\n",
    "\n",
    "\n",
    "source = np.array(gene_pair[\"PMID\"].unique())\n",
    "source = source.astype(str)\n",
    "source = \",\".join(sorted(set(filter(lambda x: x.lower() != 'nan', source))))\n",
    "# Split the string into individual elements, filter out empty strings, and get unique values\n",
    "source = sorted(\n",
    "    set(filter(lambda x: x.strip() and x.strip().lower() != 'nan', source.split(',')))\n",
    ")\n",
    "source = [value.replace(\" \", \"\") for value in source]\n",
    "\n",
    "\n",
    "\n",
    "# Function to join unique sorted values\n",
    "agg_func = lambda x: ', '.join(sorted(set(map(str, x))))\n",
    "\n",
    "# Group and aggregate all columns except 'LR pair'\n",
    "gene_pair = gene_pair.groupby('Human LR Pair').agg(agg_func).reset_index()\n",
    "gene_pair = gene_pair[gene_pair['Human LR Pair'] != '']\n",
    "DBlength = len(gene_pair)\n",
    "gene_pair[\"Interaction ID\"] = [f\"CDB{str(i).zfill(5)}\" for i in range(1, DBlength + 1)]\n",
    "\n",
    "# for creating PMIDs\n",
    "gene_pair00 = gene_pair[['Human LR Pair', 'PMID']] # was \"PMID support\"\n",
    "\n",
    "# Recreate Perplexity link\n",
    "def create_url_basic(perplexity_col):\n",
    "    query = f\"What is the primary evidence that {perplexity_col} bind-each-other-as-a-ligand-and-receptor-pair. Exclude reviews, uniprot, wiki, genecards, PIPS, iuphar as sources.\"\n",
    "    encoded_query = query.replace(\" \", \"%20\")\n",
    "    return f\"https://www.perplexity.ai/search?q={encoded_query}\"\n",
    "# Option 2 -- new query all together\n",
    "def generate_perplexity_link_pmid(row): \n",
    "    query = f\"What-is-the-biological-relevance-of-the-ligand-and-receptor-pair-{row['Human LR Pair']}-based-on-Pubmed-ID-{row['PMID']}\"\n",
    "    return (\n",
    "        f'<a href=\"https://www.perplexity.ai/search?q={query}\" target=\"_blank\">'\n",
    "        f'<img src=\"https://img.icons8.com/?size=30&id=0NbBuNOxUwps&format=png&color=000000\" alt=\"Perplexity AI\" /></a>'\n",
    "    )\n",
    "# Apply function to the DataFrame\n",
    "gene_pair[\"Perplexity\"] = gene_pair.apply(generate_perplexity_link_pmid, axis=1)\n",
    "\n",
    "# create URLs for the HGNC IDs\n",
    "\n",
    "# ligand\n",
    "gene_pair[\"Ligand HGNC ID\"] = [\n",
    "    '<a href=\"https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/{}\" target=\"_blank\">{}</a>'.format(ligand, ligand)\n",
    "    for ligand in gene_pair[\"Ligand HGNC ID\"]\n",
    "]\n",
    "\n",
    "# receptor\n",
    "gene_pair[\"Receptor HGNC ID\"] = [\n",
    "    '<a href=\"https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/{}\" target=\"_blank\">{}</a>'.format(receptor, receptor)\n",
    "    for receptor in gene_pair[\"Receptor HGNC ID\"]\n",
    "]\n",
    "\n",
    "\n",
    "# Function to generate hyperlinks for the \"PMID support\" column\n",
    "def generate_links_with_doi(df, gene_column, pmid_column, id_column):\n",
    "    def create_link(gene, id_col, sources):\n",
    "        # Replace spaces with \"——\" in the gene name for the link\n",
    "        gene_name = gene.replace(\" \", \"——\")\n",
    "        \n",
    "        if len(sources) == 1:\n",
    "            source = sources[0]\n",
    "            if source.startswith(\"https://www.biorxiv.org/content/\"):\n",
    "                # If the value starts with \"https://doi.org/\", use it as the hyperlink\n",
    "                return f'<a href=\"{source}\" target=\"_blank\">BioRxiv</a>'\n",
    "            else:\n",
    "                # If it's a single PMID, hyperlink the PMID text\n",
    "                return f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/cards/{gene}_{id_col}.html\">{source}</a>'\n",
    "        else:\n",
    "            # If multiple PMIDs, show the count and hyperlink to the page\n",
    "            return f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/cards/{gene}_{id_col}.html\" target=\"_blank\">{len(sources)} PMIDs</a>'\n",
    "\n",
    "    # Process each row to generate the \"PMID\" column # was \"PMID support\"\n",
    "    df[\"PMID\"] = [\n",
    "        create_link(\n",
    "            gene=row[gene_column], \n",
    "            id_col = row[id_column],\n",
    "            sources=[s.strip() for s in row[pmid_column].split(',') if s.strip()]\n",
    "        )\n",
    "        for _, row in df.iterrows()\n",
    "    ]\n",
    "    return df\n",
    "\n",
    "\n",
    "# Generate the links for the \"PMID\" column # was \"PMID support\"\n",
    "gene_pair = generate_links_with_doi(gene_pair, gene_column=\"Human LR Pair\", \n",
    "                                    pmid_column=\"PMID\", id_column= \"Interaction ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c13ae01-fe8b-4465-ade7-11aef1f1b4b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.unique(gene_pair[\"KEGG Pathway\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e21bf58-16de-447c-b168-0a4a9d44ec62",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43bf58b6-9e52-499f-9846-3d2bcc832b47",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group and aggregate all columns except 'LR pair'\n",
    "agg_func = lambda x: ','.join(sorted(set(map(str, x))))\n",
    "gene_pair1 = gene_pair.groupby('Human LR Pair').agg(agg_func).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "215ecc99-0987-4346-a809-824ea9d8b444",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b37d689-e6b2-4bfb-a8c6-aafe40a66d52",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_keywords_df=gene_pair000\n",
    "template=template\n",
    "interaction_card_df=interaction_card\n",
    "ligand_card_1_df=ligand_card_1\n",
    "receptor_card_1_df=receptor_card_1\n",
    "ligand_card_2_df=ligand_card_2\n",
    "receptor_card_2_df=receptor_card_2\n",
    "pubmed_data_df=pubmed_data\n",
    "gene_pair_main_df=gene_pair0\n",
    "output_dir=OUTPUT_DIR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a60de164-ab57-47ba-888f-a06cc1607657",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair0_copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ae702db-0832-4962-af56-9c6da22d0dbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use the 'Human LR Pair' from the gene_pair_keywords_df for iteration\n",
    "for idx, row in gene_pair_keywords_df.iterrows():\n",
    "    lr_pair_name = row[\"Human LR Pair\"] # This is the gene_name (e.g., VEGFA——KDR)\n",
    "    keywords = row[\"Relevance Keywords\"] # Keywords from llm_results\n",
    "    pmids_str = row[\"PMID\"] # PMIDs associated with this LR Pair\n",
    "\n",
    "    # Split LR Pair for file naming and data filtering\n",
    "    value1, value2 = lr_pair_name.replace(\"——\", \" \").split()\n",
    "\n",
    "    # Data rows for each section, filtered by the current LR Pair\n",
    "    row0 = interaction_card_df[interaction_card_df['Human LR Pair'] == lr_pair_name]\n",
    "    row1 = ligand_card_1_df[ligand_card_1_df['Human LR Pair'] == lr_pair_name]\n",
    "    row2 = receptor_card_1_df[receptor_card_1_df['Human LR Pair'] == lr_pair_name]\n",
    "    row3 = ligand_card_2_df[ligand_card_2_df['Human LR Pair'] == lr_pair_name]\n",
    "    row4 = receptor_card_2_df[receptor_card_2_df['Human LR Pair'] == lr_pair_name]\n",
    "\n",
    "    # make the pair cards here\n",
    "    def convert_pair_url(df_pairs):\n",
    "        df_pairs[\"Human LR Pair\"] = [\n",
    "            f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/cards/{lrpair}.html\" target=\"_blank\" '\n",
    "            f'title=\"Open {lrpair} card\" style=\"color: #0000EE; text-decoration: underline;\">'\n",
    "            f'{lrpair}</a>'\n",
    "            if pd.notna(lrpair) and lrpair.strip() else \"\"\n",
    "            for lrpair in df_pairs[\"Human LR Pair\"]\n",
    "        ]\n",
    "        return df_pairs\n",
    "\n",
    "    # Related ligand pairs\n",
    "    ligand_pairs = gene_pair_main_df[gene_pair_main_df['Ligand'] == value1]\n",
    "    ligand_pairs = ligand_pairs[ligand_pairs[\"Human LR Pair\"].str.replace(\" \", \"——\") != lr_pair_name] # Ensure comparison with placeholder\n",
    "    ligand_pairs = convert_pair_url(ligand_pairs[[\"Human LR Pair\"]])\n",
    "    ligand_pairs_str = ' ・ '.join([btn for btn in ligand_pairs[\"Human LR Pair\"] if btn])\n",
    "\n",
    "    # Related receptor pairs\n",
    "    receptor_pairs = gene_pair_main_df[gene_pair_main_df['Receptor'] == value2]\n",
    "    receptor_pairs = receptor_pairs[receptor_pairs[\"Human LR Pair\"].str.replace(\" \", \"——\") != lr_pair_name] # Ensure comparison with placeholder\n",
    "    receptor_pairs = convert_pair_url(receptor_pairs[[\"Human LR Pair\"]])\n",
    "    receptor_pairs_str = ' ・ '.join([btn for btn in receptor_pairs[\"Human LR Pair\"] if btn])\n",
    "    \n",
    "    # Prepare table data (convert to dict)\n",
    "    table0_data = row0.drop('Human LR Pair', axis=1).to_dict(orient='records')[0] if not row0.empty else {}\n",
    "    table1_data = row1.drop('Human LR Pair', axis=1).to_dict(orient='records')[0] if not row1.empty else {}\n",
    "    table2_data = row2.drop('Human LR Pair', axis=1).to_dict(orient='records')[0] if not row2.empty else {}\n",
    "    table3_data = row3.drop('Human LR Pair', axis=1).to_dict(orient='records')[0] if not row3.empty else {}\n",
    "    table4_data = row4.drop('Human LR Pair', axis=1).to_dict(orient='records')[0] if not row4.empty else {}\n",
    "\n",
    "    # PMID stuff\n",
    "    tab_headers = []\n",
    "    tab_contents = []\n",
    "    sources = [pmid.strip() for pmid in str(pmids_str).split(',') if pmid.strip()]\n",
    "\n",
    "    if sources:\n",
    "        for i, pmid in enumerate(sources):\n",
    "            pubmed_row = pubmed_data_df[pubmed_data_df[\"PMID\"] == pmid]\n",
    "\n",
    "            if not pubmed_row.empty:\n",
    "                title = pubmed_row[\"Title\"].values[0]\n",
    "                abstract = pubmed_row[\"Abstract\"].values[0]\n",
    "                journal = pubmed_row[\"Journal\"].values[0]\n",
    "                year = pubmed_row[\"Year\"].values[0]\n",
    "            else:\n",
    "                title = \"No Title Found\"\n",
    "                abstract = \"No Abstract Found\"\n",
    "                journal = \"Journal Unknown\"\n",
    "                year = \"Year Unknown\"\n",
    "\n",
    "            active_class = \"active\" if i == 0 else \"\"\n",
    "            tab_headers.append(f'<button class=\"tablinks {active_class}\" onclick=\"openTab(event, \\'tab{pmid}\\')\">{pmid}</button>')\n",
    "            tab_contents.append(f\"\"\"\n",
    "            <div id=\"tab{pmid}\" class=\"tabcontent {active_class}\">\n",
    "                <h2>{title}</h2>\n",
    "                <p><strong>{journal}, {year}; <a href=\"https://pubmed.ncbi.nlm.nih.gov/{pmid}/\" target=\"_blank\">For more details, see PubMed</a></strong></p>\n",
    "                <p>{abstract}</p>\n",
    "            </div>\n",
    "            \"\"\")\n",
    "\n",
    "    # Assuming you have images for ligand and receptor or they are not needed in the template\n",
    "    # If you intend to use them, define them here, e.g.:\n",
    "    ligand_image = encode_image(\"path/to/ligand_image.png\") # Replace with actual path\n",
    "    receptor_image = encode_image(\"path/to/receptor_image.png\") # Replace with actual path\n",
    "\n",
    "    # Render the template with all data\n",
    "    rendered_content = template.render(\n",
    "        gene_name=lr_pair_name.replace(\"——\", \" \"), # Pass the human-readable pair name\n",
    "        value1=value1,\n",
    "        value2=value2,\n",
    "        table0_data=table0_data,\n",
    "        table1_data=table1_data,\n",
    "        table2_data=table2_data,\n",
    "        table3_data=table3_data,\n",
    "        table4_data=table4_data,\n",
    "        ligand_image=ligand_image,\n",
    "        receptor_image=receptor_image,\n",
    "        ligand_pairs=ligand_pairs_str,\n",
    "        receptor_pairs=receptor_pairs_str,\n",
    "        tab_headers=\"\".join(tab_headers),\n",
    "        tab_contents=\"\".join(tab_contents),\n",
    "        keywords=keywords\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f150993a-a981-430b-a699-39d7cf1ff948",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import jinja2\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "import base64\n",
    "import re\n",
    "\n",
    "# Add the src directory to the path for importing modules\n",
    "sys.path.append(os.path.abspath(\"src\"))\n",
    "\n",
    "# Import necessary modules from your existing src files\n",
    "# Ensure createDataTable and createFunctionalAnnotTable are in your 'src' directory\n",
    "from createDataTable import pop_up_info, gene_pair0, generate_perplexity_links, gene_pair00\n",
    "from createFunctionalAnnotTable import gene_pair_annot_ligand, gene_pair_annot_receptor\n",
    "\n",
    "# --- Paths ---\n",
    "MERGED_TEMPLATE_PATH = 'HTML/mergedCardWithPMIDTemplate.html'\n",
    "OUTPUT_DIR = 'data/cards/' # New output directory for combined files\n",
    "\n",
    "# --- Load and Preprocess Data (Combined from both scripts) ---\n",
    "\n",
    "# Load PubMed data (from createPMIDpages.py)\n",
    "pubmed_data = pd.read_csv(\"data/pubmed_results.csv\")\n",
    "pubmed_data[\"Year\"] = pubmed_data[\"Year\"].astype(str).str.replace(\".0\", \"\", regex=False).astype(int)\n",
    "pubmed_data[\"PMID\"] = pubmed_data[\"PMID\"].astype(str)\n",
    "pubmed_data = pubmed_data.reset_index(drop=True)\n",
    "\n",
    "# Load LLM results (from createPMIDpages.py)\n",
    "bio_keywords = pd.read_csv(\"data/llm_results.csv\")\n",
    "\n",
    "# --- Prepare gene_pair00 for PMID section (from createPMIDpages.py) ---\n",
    "# gene_pair00 is used for PMID and Keywords, so it needs the '——' placeholder\n",
    "# Ensure gene_pair00 is a copy to avoid SettingWithCopyWarning later\n",
    "gene_pair00_copy = gene_pair00.copy()\n",
    "gene_pair00_copy[\"Human LR Pair\"] = gene_pair00_copy[\"Human LR Pair\"].str.replace(\" \", \"——\")\n",
    "\n",
    "# Merge with LLM results\n",
    "gene_pair000 = gene_pair00_copy.merge(bio_keywords, how='left', left_on=\"Human LR Pair\", right_on='Human LR Pair')\n",
    "gene_pair000[\"Relevance Keywords\"] = gene_pair000[\"Relevance Keywords\"].astype(str)\n",
    "gene_pair000[\"Human LR Pair\"] = gene_pair000[\"Human LR Pair\"].astype(str) # Ensure string type\n",
    "\n",
    "# --- Prepare gene_pair0 for Card section (from createCards.py) ---\n",
    "# gene_pair0 is used for card details, it should retain spaces for splitting gene names\n",
    "# Ensure gene_pair0 is a copy to avoid SettingWithCopyWarning later\n",
    "gene_pair0_copy = gene_pair0.copy()\n",
    "\n",
    "# Add Disease (specific) to cards\n",
    "df_disease = pd.read_csv(\"data/disease_annotations_per_pair.csv\")\n",
    "df_disease = df_disease.groupby('interaction')['disease'].apply(', '.join).reset_index()\n",
    "mapping_disease = dict(zip(df_disease['interaction'], df_disease['disease']))\n",
    "gene_pair0_copy[\"Disease\"] = gene_pair0_copy['Human LR Pair'].map(mapping_disease).fillna(\"unknown\")\n",
    "\n",
    "gene_pair0_copy = generate_perplexity_links(\n",
    "    gene_pair0_copy,\n",
    "    pathway_col=\"Disease\",\n",
    "    default_query_template=\"What-diseases-is-the-ligand-receptor-pair-{pair}-associated-with\"\n",
    ")\n",
    "\n",
    "gene_pair0_copy[\"Interaction ID\"] = gene_pair0_copy[\"Interaction ID\"].apply(\n",
    "    lambda x: f\"<a href='https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database/filter/{x}.html'>{x}</a>\"\n",
    ")\n",
    "# Add external link icon\n",
    "icon_html = '<i class=\"fa-solid fa-arrow-up-right-from-square\" style=\"margin-left:4px;\"></i></a>'\n",
    "columns_to_update = [\n",
    "    \"KEGG Pathway\", \"PROGENy Pathway\", \"Cancer-related\",\n",
    "    \"Disease Type\", \"Disease\"\n",
    "]\n",
    "for col in columns_to_update:\n",
    "    gene_pair0_copy[col] = gene_pair0_copy[col].str.replace(\n",
    "        \"</a>\", icon_html, regex=False\n",
    "    )\n",
    "\n",
    "# Add Ligand/Receptor group info\n",
    "agg_func = lambda x: ', '.join(sorted(set(map(str, x))))\n",
    "gene_pair_annot_ligand = gene_pair_annot_ligand.groupby('Ligand HGNC ID').agg(agg_func).reset_index()\n",
    "ligand_mapping = dict(zip(gene_pair_annot_ligand['Ligand HGNC ID'], gene_pair_annot_ligand['Ligand group']))\n",
    "\n",
    "gene_pair_annot_receptor = gene_pair_annot_receptor.groupby('Receptor HGNC ID').agg(agg_func).reset_index()\n",
    "receptor_mapping = dict(zip(gene_pair_annot_receptor['Receptor HGNC ID'], gene_pair_annot_receptor['Receptor group']))\n",
    "\n",
    "\n",
    "# --- Helper Functions (Combined and adjusted) ---\n",
    "\n",
    "def load_template(template_path):\n",
    "    \"\"\"Load Jinja2 template from a file.\"\"\"\n",
    "    with open(template_path, 'r', encoding='utf-8') as file:\n",
    "        return jinja2.Template(file.read())\n",
    "\n",
    "def encode_image(image_path):\n",
    "    \"\"\"Encode an image to base64. (Not used in this version, but kept for reference)\"\"\"\n",
    "    try:\n",
    "        with open(image_path, \"rb\") as image_file:\n",
    "            return base64.b64encode(image_file.read()).decode('utf-8')\n",
    "    except FileNotFoundError:\n",
    "        return None\n",
    "\n",
    "def extract_hgnc_id(col):\n",
    "    \"\"\"Use regular expression to extract the HGNC ID after 'HGNC:'.\"\"\"\n",
    "    match = re.search(r'HGNC:(\\d+)', col)\n",
    "    if match:\n",
    "        return match.group(1)\n",
    "    return None\n",
    "\n",
    "def convert_hgnc_url(col):\n",
    "    hgnc_id = extract_hgnc_id(col)\n",
    "    if hgnc_id:\n",
    "        visible_text = 'GeneCards <i class=\"fa-solid fa-arrow-up-right-from-square\" style=\"margin-left: 4px;\"></i>'\n",
    "        new_link = (\n",
    "            f'<a href=\"https://www.genecards.org/cgi-bin/carddisp.pl?id_type=hgnc&id={hgnc_id}\" '\n",
    "            f'target=\"_blank\" style=\"color: #0000EE; text-decoration: underline;\">{visible_text}</a>'\n",
    "        )\n",
    "        return new_link\n",
    "    return None\n",
    "\n",
    "def convert_hgnc_url_disease(col):\n",
    "    hgnc_id = extract_hgnc_id(col)\n",
    "    if hgnc_id:\n",
    "        visible_text = 'MalaCards <i class=\"fa-solid fa-arrow-up-right-from-square\" style=\"margin-left: 4px;\"></i>'\n",
    "        new_link = f'<a href=\"https://www.genecards.org/cgi-bin/carddisp.pl?id_type=hgnc&id={hgnc_id}#diseases\" target=\"_blank\">{visible_text}</a>'\n",
    "        return new_link\n",
    "    return None\n",
    "\n",
    "def convert_hgnc_url_exp(col):\n",
    "    hgnc_id = extract_hgnc_id(col)\n",
    "    if hgnc_id:\n",
    "        visible_text = 'mRNA expression in normal human tissues <i class=\"fa-solid fa-arrow-up-right-from-square\" style=\"margin-left: 4px;\"></i>'\n",
    "        new_link = f'<a href=\"https://www.genecards.org/cgi-bin/carddisp.pl?id_type=hgnc&id={hgnc_id}#expression\" target=\"_blank\">{visible_text}</a>'\n",
    "        return new_link\n",
    "    return None\n",
    "\n",
    "def prepare_card_dataframes(gene_pair_input_df):\n",
    "    \"\"\"Prepare interaction, ligand, and receptor dataframes for the card section.\"\"\"\n",
    "    # Ensure gene_pair_input_df is a copy to avoid SettingWithCopyWarning\n",
    "    gene_pair_input_df = gene_pair_input_df.copy()\n",
    "\n",
    "    gene_pair_input_df[\"Interaction Type\"] = [\n",
    "        f'{ligand} {ligandLocation} ligand binds to {receptor} {receptorLocation} receptor'\n",
    "        for ligand, ligandLocation, receptor, receptorLocation in zip(\n",
    "            gene_pair_input_df[\"Ligand\"], gene_pair_input_df[\"Ligand Location\"],\n",
    "            gene_pair_input_df[\"Receptor\"], gene_pair_input_df[\"Receptor Location\"]\n",
    "        )\n",
    "    ]\n",
    "    interaction_card = gene_pair_input_df[[\"Interaction ID\", \"Human LR Pair\", \"Interaction Type\", \"Perplexity\", \"PMID\", \"KEGG Pathway\",  \"PROGENy Pathway\", \"Cancer-related\", \"Disease Type\", \"Disease\"]]\n",
    "    interaction_card[\"Perplexity\"] = interaction_card[\"Perplexity\"].str.replace('size=30', 'size=80')\n",
    "\n",
    "    pop_up_info_lim = pop_up_info[\n",
    "        [\"Approved symbol\", \"Alias symbol\", \"Previous symbol\", \"Date symbol changed\"]\n",
    "    ].drop_duplicates(subset=\"Approved symbol\", keep=\"first\")\n",
    "\n",
    "    def format_symbol_aliases(old_symbol, aliases):\n",
    "        parts = [p for p in (old_symbol, aliases) if p != \"N/A\"]\n",
    "        return f\"{', '.join(parts)}\" if parts else aliases\n",
    "\n",
    "    pop_up_info_lim['Other Symbols'] = pop_up_info_lim.apply(\n",
    "        lambda row: format_symbol_aliases(row[\"Previous symbol\"], row[\"Alias symbol\"]),\n",
    "        axis=1\n",
    "    )\n",
    "\n",
    "    ligand_card = gene_pair_input_df[[\"Human LR Pair\", \"Ligand\", \"Ligand name\", \"Ligand HGNC ID\", \"Ligand MGI ID\", \"Ligand RGD ID\", \"Ligand Location\"]].merge(\n",
    "        pop_up_info_lim, how='left', left_on='Ligand', right_on='Approved symbol'\n",
    "    ).drop_duplicates(subset='Human LR Pair', keep=\"first\").drop(columns=[\"Ligand\", \"Approved symbol\"])\n",
    "\n",
    "    ligand_card_1 = ligand_card[[\"Human LR Pair\", \"Ligand name\", \"Other Symbols\" ]]\n",
    "    ligand_card_2 = ligand_card[[\"Human LR Pair\", \"Ligand HGNC ID\", \"Ligand Location\"]]\n",
    "    ligand_card_2[\"HGNC gene card\"] = ligand_card_2[\"Ligand HGNC ID\"].apply(convert_hgnc_url)\n",
    "    ligand_card_2[\"Disease relevance\"] = ligand_card_2[\"Ligand HGNC ID\"].apply(convert_hgnc_url_disease)\n",
    "    ligand_card_2[\"Expression Profile\"] = ligand_card_2[\"Ligand HGNC ID\"].apply(convert_hgnc_url_exp)\n",
    "    ligand_card_2[\"Lineage group\"] = ligand_card_2['Ligand HGNC ID'].map(ligand_mapping).fillna(\"none\")\n",
    "    icon_html_card = '<i class=\"fa-solid fa-arrow-up-right-from-square\" style=\"margin-left:4px;\"></i></a>' # Use a different name to avoid conflict\n",
    "    for col in [\"Ligand HGNC ID\"]:\n",
    "        ligand_card_2[col] = ligand_card_2[col].str.replace(\n",
    "            \"</a>\", icon_html_card, regex=False\n",
    "        )\n",
    "    ligand_card_2 = ligand_card_2[[\"Human LR Pair\", \"Ligand HGNC ID\", \"HGNC gene card\", \"Ligand Location\", \"Lineage group\", \"Disease relevance\", \"Expression Profile\"]]\n",
    "\n",
    "\n",
    "    receptor_card = gene_pair_input_df[[\"Human LR Pair\", \"Receptor\", \"Receptor name\", \"Receptor HGNC ID\", \"Receptor MGI ID\", \"Receptor RGD ID\", \"Receptor Location\"]].merge(\n",
    "        pop_up_info_lim, how='left', left_on='Receptor', right_on='Approved symbol'\n",
    "    ).drop_duplicates(subset='Human LR Pair', keep=\"first\").drop(columns=[\"Receptor\", \"Approved symbol\"])\n",
    "\n",
    "    receptor_card_1 = receptor_card[[\"Human LR Pair\", \"Receptor name\", \"Other Symbols\"]]\n",
    "    receptor_card_2 = receptor_card[[\"Human LR Pair\", \"Receptor HGNC ID\", \"Receptor Location\"]]\n",
    "    receptor_card_2[\"HGNC gene card\"] = receptor_card_2[\"Receptor HGNC ID\"].apply(convert_hgnc_url)\n",
    "    receptor_card_2[\"Disease relevance\"] = receptor_card_2[\"Receptor HGNC ID\"].apply(convert_hgnc_url_disease)\n",
    "    receptor_card_2[\"Expression Profile\"] = receptor_card_2[\"Receptor HGNC ID\"].apply(convert_hgnc_url_exp)\n",
    "    receptor_card_2[\"Lineage group\"] = receptor_card_2['Receptor HGNC ID'].map(receptor_mapping).fillna(\"none\")\n",
    "    for col in [\"Receptor HGNC ID\"]:\n",
    "        receptor_card_2[col] = receptor_card_2[col].str.replace(\n",
    "            \"</a>\", icon_html_card, regex=False\n",
    "        )\n",
    "    receptor_card_2 = receptor_card_2[[\"Human LR Pair\", \"Receptor HGNC ID\",  \"HGNC gene card\", \"Receptor Location\", \"Lineage group\", \"Disease relevance\", \"Expression Profile\" ]]\n",
    "\n",
    "    return interaction_card, ligand_card_1, ligand_card_2, receptor_card_1, receptor_card_2\n",
    "\n",
    "# Define test genes - these should be in the 'space' format for gene_pair0\n",
    "    # and will be converted to '——' for gene_pair000 internally.\n",
    "test_genes = [\"VEGFA ITGB1\", \"VEGFA KDR\", \"VEGFA NRP1\"] # Example genes\n",
    "\n",
    "    # Filter gene_pair0 for the test genes to be used in prepare_card_dataframes\n",
    "    # This gene_pair_input should have space-separated LR pairs\n",
    "gene_pair_input = gene_pair0_copy[gene_pair0_copy[\"Human LR Pair\"].isin(test_genes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26e9a48e-6b65-4ab7-969a-705b3fb1c31d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_main_df=gene_pair0_copy\n",
    "gene_pair_main_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "210eb5cd-cd7f-4855-aa26-d618b377838b",
   "metadata": {},
   "outputs": [],
   "source": [
    "page_navigation_map = []\n",
    "    # Ensure df_for_navigation is sorted consistently to guarantee correct prev/next order.\n",
    "df_for_navigation = gene_pair_main_df.sort_values(\n",
    "        by=\"Interaction ID\",\n",
    "        key=lambda series: series.apply(\n",
    "            lambda x: int(re.search(r'CDB(\\d+)</a>', x).group(1))\n",
    "            if isinstance(x, str) and re.search(r'CDB(\\d+)</a>', x)\n",
    "            else -1 # Return -1 (or 0) for sorting if the pattern is not found or if x is not a string\n",
    "        )\n",
    "    ).reset_index(drop=True)\n",
    "df_for_navigation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "198fdea8-4c9f-436c-9742-1a06c8d2e7a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_main_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f1732a5-e2ee-4873-803c-8ec95f7106d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_for_rendering_and_navigation = cleaned_main_df.sort_values(\n",
    "        by=\"Clean Interaction ID\",\n",
    "        key=lambda series: series.str[3:].astype(int) # Now it's just \"CDBXXXXX\" so simple slicing works\n",
    "    ).reset_index(drop=True)\n",
    "df_for_rendering_and_navigation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc178dfe-3096-471c-843e-70bc27eeb3b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "page_navigation_map = []\n",
    "    # Ensure df_for_navigation is sorted consistently to guarantee correct prev/next order.\n",
    "df_for_navigation = gene_pair_main_df.sort_values(by=\"Interaction ID\", key=lambda x: x.str[3:].astype(int)).reset_index(drop=True)\n",
    "for idx, row in df_for_navigation.iterrows():\n",
    "        interaction_id = row[\"Interaction ID\"] # e.g., 'CDB00001'\n",
    "        human_lr_pair = row[\"Human LR Pair\"] # e.g., 'VEGFA KDR'\n",
    "\n",
    "        # Construct the filename with interaction_id appended\n",
    "value1, value2 = human_lr_pair.split(\" \")\n",
    "        # UPDATED FILENAME CONSTRUCTION\n",
    "filename = f\"{value1.strip()} —— {value2.strip()}_{interaction_id}.html\"\n",
    "full_url_path = os.path.join(\"/data/cards/\", filename) # Path relative to web root\n",
    "\n",
    "page_navigation_map.append({\n",
    "            \"interaction_id\": interaction_id,\n",
    "            \"url\": full_url_path,\n",
    "            \"filename\": filename # Store filename for saving later\n",
    "        })\n",
    "page_navigation_map\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5727ab12-f0c3-4c01-8787-c241284ab9b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, page_info in enumerate(page_navigation_map):\n",
    "        current_interaction_id = page_info[\"interaction_id\"]\n",
    "        current_filename_to_save = page_info[\"filename\"]\n",
    "            # Retrieve the original human_lr_pair for data filtering, if needed\n",
    "        current_human_lr_pair = df_for_navigation.loc[df_for_navigation[\"Interaction ID\"] == current_interaction_id, \"Human LR Pair\"].iloc[0]\n",
    "\n",
    "        # Determine previous page info\n",
    "        prev_page_info = None\n",
    "        if i > 0:\n",
    "            prev_page_info = page_navigation_map[i - 1]\n",
    "        print(prev_page_info)\n",
    "page_navigation_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4fe8873-9b65-47bc-b4bb-a68d8089a226",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_interaction_ids = gene_pair_main_df[\"Interaction ID\"].str[3:].astype(int).max()\n",
    "total_interaction_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c89b4924-2e4d-4d59-b05e-ba3700b75145",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_main_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e670a607-3979-44f8-ac76-af1c5f474771",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace spaces in \"Human LR Pair\" with a placeholder\n",
    "gene_pair00[\"Human LR Pair\"] = gene_pair00[\"Human LR Pair\"].str.replace(\" \", \"——\")\n",
    "gene_pair000 = gene_pair00.merge(bio_keywords, how='left', left_on=\"Human LR Pair\", right_on='Human LR Pair')\n",
    "gene_pair000[\"Relevance Keywords\"] = gene_pair000[\"Relevance Keywords\"].astype(str)\n",
    "gene_pair000[\"Human LR Pair\"]  = gene_pair000[\"Human LR Pair\"].astype(str)\n",
    "pubmed_data = pubmed_data.reset_index(drop=True)  # Remove the index\n",
    "gene_pair000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad176181-14ba-4123-ae2b-4701380b85ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_genes = [\"VEGFA KDR\", \"ADAM17 IL6R\"]\n",
    "test_genes.replace(\"——\", \" \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b618faa7-252b-4873-8f45-2c0ba6f637d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09930bab-a2e6-4314-9523-123da518d74d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import jinja2\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "import base64\n",
    "import re\n",
    "\n",
    "\n",
    "sys.path.append(os.path.abspath(\"src\"))  \n",
    "import fetchGSheet\n",
    "from createDataTable import pop_up_info, gene_pair0, generate_perplexity_links\n",
    "from createFunctionalAnnotTable import gene_pair_annot_ligand, gene_pair_annot_receptor\n",
    "gene_pair0\n",
    "value = \"CD80 CD274\"\n",
    "value1 = \"CD80\"\n",
    "ligand_pairs = gene_pair0[gene_pair0['Ligand'] == value1]\n",
    "# print out all Human LR Pair values except for value\n",
    "ligand_pairs = ligand_pairs[ligand_pairs[\"Human LR Pair\"] != value]\n",
    "ligand_pairs = ligand_pairs[[\"Human LR Pair\"]]\n",
    "# Create the links to the HTML cards\n",
    "ligand_pairs[\"Human LR Pair\"] = [\n",
    "            f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/cards/{lrpair}.html\" target=\"_blank\" '\n",
    "            f'role=\"button\" title=\"Open {lrpair} card\" class=\"btn btn-outline-primary\" '\n",
    "            f'style=\"background-color: #3498db; color: white; border-color: #2980b9; font-size: 16px; '\n",
    "            f'padding: 8px 12px; margin: 4px; text-decoration: none; border-radius: 4px;\">'\n",
    "            f'{lrpair} Card</a>'\n",
    "            if pd.notna(lrpair) and lrpair.strip() else \"\"\n",
    "            for lrpair in ligand_pairs[\"Human LR Pair\"]\n",
    "        ]\n",
    "        \n",
    "        # Aggregate into one value separated by space\n",
    "ligand_pairs = ' '.join([btn for btn in ligand_pairs[\"Human LR Pair\"] if btn])\n",
    "ligand_pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e99d101-6664-4e76-a802-24d6692a357c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_trip.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1153ab51-39a9-4479-abae-8af7300606da",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_trip_cat = gene_pair_trip.iloc[:, [8,5,12]]\n",
    "cat_cols_trip = gene_pair_trip_cat.select_dtypes(include='object').columns\n",
    "cat_cols_trip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eed05322-b23b-4a48-8a41-6c2dc798b0e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to prepare functional annotation datatable\n",
    "import sys, os\n",
    "from itables import init_notebook_mode\n",
    "import pandas as pd\n",
    "from itables import show\n",
    "from itables import options\n",
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "from bs4 import BeautifulSoup\n",
    "from createDataTable import gene_pair, gene_pair0, generate_perplexity_link_pmid\n",
    "import warnings\n",
    "import fetchGSheet \n",
    "import string\n",
    "\n",
    "def make_ids_unique(series):\n",
    "    return [\n",
    "        f\"{id_val}{letter}\"\n",
    "        #if count > 1 else id_val -- UNCOMMENT SO UNIQUE ONES ARE ALSO WITH APPENDED LETTER [\"A\"]\n",
    "        for id_val, count, letter in zip(\n",
    "            series,\n",
    "            series.groupby(series).transform('count'),\n",
    "            series.groupby(series).cumcount().map(lambda i: string.ascii_uppercase[i] if i < 26 else f\"_{i}\")\n",
    "        )\n",
    "    ]\n",
    "\n",
    "gene_pair0 = gene_pair0.dropna(subset=['Human LR Pair'])\n",
    "mapping_ID = dict(zip(gene_pair0['Human LR Pair'], gene_pair0['Interaction ID']))\n",
    "gene_pair_PMID = fetchGSheet.gene_pair.dropna(axis=1, how='all')\n",
    "gene_pair_PMID = gene_pair_PMID[gene_pair_PMID['LR pair'] != '']\n",
    "gene_pair_PMID= gene_pair_PMID[[\"LR pair\", \"PMID\", \"original source\"]]\n",
    "# Mapping for replacements\n",
    "mapping = dict(zip(fetchGSheet.src_info['original source'], fetchGSheet.src_info['shortname']))\n",
    "# Replace values in the column based on the mapping\n",
    "gene_pair_PMID['Database Source'] = gene_pair_PMID['original source'].replace(mapping)\n",
    "\n",
    "# Replace values in the column based on the mapping\n",
    "gene_pair_PMID[\"Interaction ID\"] = gene_pair_PMID['LR pair'].replace(mapping_ID)\n",
    "df_pub = pd.read_csv(\"data/pubmed_results.csv\", usecols=[0,1,3,4,5])\n",
    "gene_pair_PMID[\"PMID\"] = gene_pair_PMID[\"PMID\"].astype(str)\n",
    "df_pub[\"PMID\"] = df_pub[\"PMID\"].astype(str)\n",
    "gene_pair_trip = pd.merge(gene_pair_PMID, df_pub, how='left', on='PMID')\n",
    "### patch for the BioRxiV ###\n",
    "gene_pair_trip[\"Year\"] = gene_pair_trip[\"Year\"].apply(\n",
    "    lambda x: \"2024\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"Species Specificity\", \"NaN\", \"\"] else x\n",
    ")\n",
    "###\n",
    "gene_pair_trip[\"Year\"] = pd.to_numeric(gene_pair_trip[\"Year\"], errors=\"coerce\").astype(\"Int64\")\n",
    "gene_pair_trip = gene_pair_trip.merge(gene_pair, how='left', left_on='Interaction ID', right_on=gene_pair.columns[0])\n",
    "gene_pair_trip = gene_pair_trip.drop(columns=[\"Interaction ID\", gene_pair.columns[2], gene_pair.columns[5],gene_pair.columns[6]])\n",
    "gene_pair_trip = gene_pair_trip.drop_duplicates()\n",
    "gene_pair_trip = gene_pair_trip.reset_index(drop=True)  \n",
    "\n",
    "# Add perplexity query\n",
    "gene_pair_trip = gene_pair_trip.rename(columns={\"LR pair\": \"Human LR Pair\"})\n",
    "gene_pair_trip[\"Perplexity\"] = gene_pair_trip.apply(generate_perplexity_link_pmid, axis=1)\n",
    "gene_pair_trip = gene_pair_trip.drop(columns=[\"Human LR Pair\", \"original source\"])\n",
    "\n",
    "gene_pair_trip[\"Species\"] = gene_pair_trip[\"Species\"].apply(\n",
    "    lambda x: \"unknown\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"Species Specificity\", \"NaN\", \"\"] else x\n",
    ")\n",
    "\n",
    "### patch for the BioRxiV ###\n",
    "gene_pair_trip[\"Title\"] = gene_pair_trip[\"Title\"].apply(\n",
    "    lambda x: \"ACKR5/GPR182 is a scavenger receptor for the atypical chemokine CXCL17, GPR15L and various endogenous peptides.\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"NaN\", \"\"] else x\n",
    ")\n",
    "\n",
    "### Pop-up for title\n",
    "gene_pair_trip[\"Title\"] = [\n",
    "    f'<span title=\"{title}\">{title}</span>'\n",
    "    for title in gene_pair_trip[\"Title\"]\n",
    "]\n",
    "\n",
    "gene_pair_trip[\"Journal\"] = gene_pair_trip[\"Journal\"].apply(\n",
    "    lambda x: \"bioRxiv\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"NaN\", \"\"] else x\n",
    ")\n",
    "\n",
    "\n",
    "gene_pair_trip[\"PMID\"] = [\n",
    "    f'<a href=\"https://pubmed.ncbi.nlm.nih.gov/{pmid}\" target=\"_blank\">{pmid}</a>'\n",
    "    if pd.notna(pmid) and str(pmid).isdigit()\n",
    "    else f'<a href=\"{pmid}\" target=\"_blank\">BioRxiv</a>'\n",
    "    for pmid in gene_pair_trip[\"PMID\"]\n",
    "]\n",
    "####\n",
    "\n",
    "df_annot=pd.read_csv(\"data/journal_abbv.csv\")\n",
    "mapping = dict(zip(df_annot['Journal Name'], df_annot['Abbreviation']))\n",
    "# Replace values in the column based on the mapping\n",
    "gene_pair_trip[\"JournalAbbv\"] = gene_pair_trip['Journal'].replace(mapping)\n",
    "gene_pair_trip[\"Journal\"] = [\n",
    "    f'<span title=\"{Journal}\">{JournalAbbv}</span>'\n",
    "    for Journal, JournalAbbv in zip(gene_pair_trip[\"Journal\"], gene_pair_trip[\"JournalAbbv\"])\n",
    "    ]\n",
    "gene_pair_trip = gene_pair_trip.drop(columns=['JournalAbbv'])\n",
    "\n",
    "# Make ID unique\n",
    "gene_pair_trip = gene_pair_trip.sort_values(by='Year', ascending=True)\n",
    "gene_pair_trip[gene_pair_trip.columns[6]] = make_ids_unique(gene_pair_trip[gene_pair_trip.columns[6]])\n",
    "gene_pair_trip = gene_pair_trip.sort_values(by='Year', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d23bcf2f-d2b7-48b1-b47d-656d4e8ce5a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to prepare functional annotation datatable\n",
    "import sys, os\n",
    "import re\n",
    "from itables import init_notebook_mode\n",
    "import pandas as pd\n",
    "from itables import show\n",
    "from itables import options\n",
    "from IPython.display import HTML, display\n",
    "import numpy as np\n",
    "from bs4 import BeautifulSoup\n",
    "from createDataTable import gene_pair0, gene_pair, top_pathway_df\n",
    "from fetchGSheet import gene_group\n",
    "import warnings\n",
    "\n",
    "gene_pair_annot = gene_pair0[[\"Interaction ID\", \"Human LR Pair\", \"Cancer-related\", \"Ligand symbol and aliases\",  \"Receptor symbol and aliases\"]].copy()\n",
    "# Diseases\n",
    "df= pd.read_csv(\"data/disease_annotations_per_pair.csv\")\n",
    "df_cat=pd.read_csv(\"data/disease_categories.csv\")\n",
    "mapping = dict(zip(df_cat['Disease Name'], df_cat['Category']))\n",
    "# Replace values in the column based on the mapping\n",
    "df[\"Disease Type\"] = df['disease'].replace(mapping)\n",
    "gene_pair_annot = gene_pair_annot.merge(df, how='left', left_on='Human LR Pair', right_on='interaction')\n",
    "gene_pair_annot = gene_pair_annot.drop(columns=[\"interaction\"])\n",
    "# PROGENy Pathway retrieved via LIANA+\n",
    "df= pd.read_csv(\"data/pathway_annotations_per_pair.csv\") \n",
    "gene_pair_annot = gene_pair_annot.merge(df, how='left', left_on='Human LR Pair', right_on='interaction')\n",
    "gene_pair_annot = gene_pair_annot.drop(columns=[\"interaction\", \"weight\"])\n",
    "\n",
    "gene_pair_annot = gene_pair_annot.rename(columns={\n",
    "                                     \"disease\": \"Disease\", \n",
    "                                     \"source\": \"PROGENy Pathway\"}\n",
    "                            )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c8c6511-3fad-409c-8ff5-bbe2a16936e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bring in KEGG Pathways from AU side\n",
    "gene_pair_annot = gene_pair_annot.merge(top_pathway_df, how='left', left_on='Human LR Pair', right_on='LR Pair')\n",
    "\n",
    "# reorder\n",
    "gene_pair_annot = gene_pair_annot[[\"Interaction ID\", \"Human LR Pair\", \"Disease\", \"Disease Type\", \"Cancer-related\", \"KEGG Pathway ID\", \"KEGG Pathway\", \"KEGG relationship\", \"PROGENy Pathway\", \"Ligand symbol and aliases\",  \"Receptor symbol and aliases\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afce45de-acd6-4cc8-8cfb-bcd873d99c13",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_annot[\"Disease\"] = gene_pair_annot[\"Disease\"].apply(\n",
    "    lambda x: \"unknown\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\"] else x)\n",
    "gene_pair_annot[\"Disease Type\"] = gene_pair_annot[\"Disease Type\"].apply(\n",
    "    lambda x: \"unknown\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\"] else x)\n",
    "gene_pair_annot[\"PROGENy Pathway\"] = gene_pair_annot[\"PROGENy Pathway\"].apply(\n",
    "    lambda x: \"unknown\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\"] else x)\n",
    "gene_pair_annot[\"KEGG Pathway\"] = gene_pair_annot[\"KEGG Pathway\"].apply(\n",
    "    lambda x: \"unknown\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\"] else x)\n",
    "\n",
    "gene_pair_annot[\"KEGG Pathway ID\"] = gene_pair_annot[\"KEGG Pathway ID\"].apply(\n",
    "    lambda x: \"unknown\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\"] else x)\n",
    "\n",
    "gene_pair_annot[\"KEGG relationship\"] = gene_pair_annot[\"KEGG relationship\"].apply(\n",
    "    lambda x: \"unknown\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\"] else x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "deba0fc8-50fc-424e-a84f-c72646695130",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "gene_pair_annot = gene_pair_annot.reset_index(drop=True).copy()\n",
    "gene_pair_annot[\"Interaction ID\"] = gene_pair_annot[\"Interaction ID\"].apply(\n",
    "    lambda x: f\"<a href='https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database/filter/{x}.html'>{x}</a>\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf90b9ea-94eb-437c-9c1c-f1b19d711452",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Separate Disease and Pathway and then rm duplicates\n",
    "gene_pair_disease = gene_pair_annot[[\"Interaction ID\", \"Human LR Pair\", \"Disease\", \"Disease Type\", \"Cancer-related\", \"Ligand symbol and aliases\",  \"Receptor symbol and aliases\"]]\n",
    "gene_pair_disease = gene_pair_disease.drop_duplicates()\n",
    "gene_pair_disease=gene_pair_disease.reset_index(drop=True)  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "747e0727-acc6-4337-95ac-ba06281f0f33",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_perplexity_link(row):\n",
    "    if pd.isna(row[\"Disease\"]) or row[\"Disease\"] == \"unknown\":\n",
    "        query = f\"What-disease-is-the-{row['Human LR Pair']}-associated-with\"\n",
    "    else:\n",
    "        query = f\"What-is-the-role-of-the-ligand-and-receptor-pair-{row['Human LR Pair']}-in-{row['Disease']}\"\n",
    "    \n",
    "    return (\n",
    "        f'<a href=\"https://www.perplexity.ai/search?q={query}\" target=\"_blank\">'\n",
    "        f'<img src=\"https://img.icons8.com/?size=30&id=0NbBuNOxUwps&format=png&color=000000\" alt=\"Perplexity AI\" /></a>'\n",
    "    )\n",
    "\n",
    "gene_pair_disease[\"Perplexity\"] = gene_pair_disease.apply(generate_perplexity_link, axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f8a8e10-241c-471c-a1ec-b67c1626a816",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the links to the HTML cards\n",
    "gene_pair_disease[\"Human LR Pair\"] = [\n",
    "    f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/cards/{lrPairOrig}.html\">{lrPair}</a>'\n",
    "    for lrPairOrig, lrPair in zip(gene_pair_disease[\"Human LR Pair\"], gene_pair_disease[\"Human LR Pair\"])\n",
    "]\n",
    "\n",
    "### Pop-up for disease, disease type\n",
    "gene_pair_disease[\"Disease\"] = [\n",
    "    f'<span title=\"{disease}\">{disease}</span>'\n",
    "    for disease in gene_pair_disease[\"Disease\"]\n",
    "]\n",
    "\n",
    "gene_pair_disease[\"Disease Type\"] = [\n",
    "    f'<span title=\"{disease}\">{disease}</span>'\n",
    "    for disease in gene_pair_disease[\"Disease Type\"]\n",
    "]\n",
    "\n",
    "gene_pair_pathway = gene_pair_annot[[\"Interaction ID\", \"Human LR Pair\",  \"KEGG Pathway ID\", \"KEGG Pathway\", \"KEGG relationship\", \"PROGENy Pathway\", \"Ligand symbol and aliases\",  \"Receptor symbol and aliases\"]]\n",
    "gene_pair_pathway = gene_pair_pathway.drop_duplicates()\n",
    "gene_pair_pathway=gene_pair_pathway.reset_index(drop=True)  \n",
    "def generate_perplexity_link_pathway(row):\n",
    "    if pd.isna(row[\"KEGG Pathway\"]) or row[\"KEGG Pathway\"] == \"unknown\":\n",
    "        if pd.isna(row[\"PROGENy Pathway\"]) or row[\"PROGENy Pathway\"] == \"unknown\":\n",
    "            query = f\"What-biological-pathway-is-the-{row['Human LR Pair']}-associated-with\"\n",
    "        else:\n",
    "            query = f\"What-is-the-role-of-the-ligand-and-receptor-pair-{row['Human LR Pair']}-in-{row['PROGENy Pathway']}\"\n",
    "    else:\n",
    "        query = f\"What-is-the-role-of-the-ligand-and-receptor-pair-{row['Human LR Pair']}-in-{row['KEGG Pathway']}\"\n",
    "    \n",
    "    return (\n",
    "        f'<a href=\"https://www.perplexity.ai/search?q={query}\" target=\"_blank\">'\n",
    "        f'<img src=\"https://img.icons8.com/?size=30&id=0NbBuNOxUwps&format=png&color=000000\" alt=\"Perplexity AI\" /></a>'\n",
    "    )\n",
    "\n",
    "gene_pair_pathway[\"Perplexity\"] = gene_pair_pathway.apply(generate_perplexity_link_pathway, axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fafb58e-8b9d-4e47-9f9e-eaf9ee33ede2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the links to the HTML cards\n",
    "gene_pair_pathway[\"Human LR Pair\"] = [\n",
    "    f'<a href=\"https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/cards/{lrPairOrig}.html\">{lrPair}</a>'\n",
    "    for lrPairOrig, lrPair in zip(gene_pair_pathway[\"Human LR Pair\"], gene_pair_pathway[\"Human LR Pair\"])\n",
    "]\n",
    "\n",
    "### Pop-up for KEGG Pathway\n",
    "gene_pair_pathway[\"KEGG Pathway\"] = [\n",
    "    f'<span title=\"{path}\">{path}</span>'\n",
    "    for path in gene_pair_pathway[\"KEGG Pathway\"]\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ad31da1-02e0-45bf-8588-4a25e6e88e45",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_annot2 = gene_pair0[[\n",
    "    'Ligand HGNC ID', 'Receptor HGNC ID', \n",
    "    \"Ligand symbol and aliases\",  \n",
    "    \"Receptor symbol and aliases\",\n",
    "    'Ligand location', 'Receptor location',\n",
    "]].copy()\n",
    "\n",
    "# Extract HGNC IDs cleanly using regex only if string is valid\n",
    "def extract_hgnc_id(text):\n",
    "    if pd.isna(text): return None\n",
    "    match = re.search(r'(HGNC:\\d+)', str(text))\n",
    "    return match.group(1) if match else None\n",
    "\n",
    "gene_pair_annot2[\"ligand_hgnc_id\"] = gene_pair_annot2[\"Ligand HGNC ID\"].apply(extract_hgnc_id)\n",
    "gene_pair_annot2[\"receptor_hgnc_id\"] = gene_pair_annot2[\"Receptor HGNC ID\"].apply(extract_hgnc_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "852032aa-607d-4a1d-9e00-8e587b5e5cb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pair_annot2.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68cb4d7a-b757-4c92-bfd4-3569610a4998",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add Disease (specific) to cards\n",
    "df= pd.read_csv(\"data/disease_annotations_per_pair.csv\")\n",
    "df = df.groupby('interaction')['disease'].apply(', '.join).reset_index()\n",
    "mapping = dict(zip(df['interaction'],df['disease']))\n",
    "gene_pair0[\"Disease\"] = gene_pair0['Human LR Pair'].replace(mapping)\n",
    "\n",
    "gene_pair0[\"Disease\"] = gene_pair0[\"Disease\"].apply(\n",
    "    lambda x: \"ask Perplexity\" if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\", \"unknown\"] else x)\n",
    "gene_pair0[\"Disease Type\"] = gene_pair0[\"Disease Type\"].apply(\n",
    "    lambda x: \"ask Perplexity\"  if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\",\"unknown\"] else x)\n",
    "gene_pair0[\"PROGENy Pathway\"] = gene_pair0[\"PROGENy Pathway\"].apply(\n",
    "    lambda x: \"ask Perplexity\"  if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\", \"unknown\"] else x)\n",
    "gene_pair0[\"KEGG Pathway\"] = gene_pair0[\"KEGG Pathway\"].apply(\n",
    "    lambda x: \"ask Perplexity\"  if pd.isna(x) or str(x).strip().lower() in [\"nan\", \"none\", \"\", \"unknown\"] else x)\n",
    "\n",
    "# if only one replace gene_pair0 to e.g. \n",
    "gene_pair_input = gene_pair0[gene_pair0[\"Human LR Pair\"] == \"A2M HSPA5\"]\n",
    "gene_pair_input\n",
    "gene_pair_input = gene_pair_input = gene_pair0[gene_pair0[\"Human LR Pair\"].isin([\"A2M HSPA5\", \"APOE LRP1\"])]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3783b9c0-01b1-41f8-9f01-d6f5a44f3a18",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Diseases\n",
    "df= pd.read_csv(\"data/disease_annotations_per_pair.csv\")\n",
    "df = df.groupby('interaction')['disease'].apply(', '.join).reset_index()\n",
    "mapping = dict(zip(df['interaction'],df['disease']))\n",
    "gene_pair0[\"Disease\"]  = gene_pair0['Human LR Pair'].replace(mapping)\n",
    "gene_pair0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "891833f5-e2a6-4425-89de-ff15f0b9dd4e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2eaf849-854b-4f88-ae74-340a592ccec6",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "human_gene_pair = human_gene_pair[[\"Interaction ID\", \"Human LR Pair\", \"Ligand\", \n",
    "                                   \"Ligand symbol and aliases\", \"Ligand HGNC ID\",\n",
    "                                   \"Ligand location\", \"Receptor\", \n",
    "                                   \"Receptor symbol and aliases\", \"Receptor HGNC ID\",\n",
    "                                   \"Receptor location\", \"PMID\", \"Perplexity\",\n",
    "                                   \"Database Source\", \"binding location\",\n",
    "                                   \"bind in trans?\", \"bidirectional signalling?\",\n",
    "                                   \"interaction type\", \"KEGG Pathway\", \"Cancer-related\",\n",
    "                                   \"Disease Type\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96fd9edb-eecc-40ab-878d-798007fc5975",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import anndata\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from tqdm import tqdm\n",
    "import requests\n",
    "import scanpy as sc\n",
    "import re\n",
    "\n",
    "output_dir = \"data/tabula_sapiens/\"\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "os.makedirs(output_dir +\"umap\" , exist_ok=True)\n",
    "os.makedirs(output_dir +\"heatmap\" , exist_ok=True)\n",
    "\n",
    "output_file = \"data/tissue_dataset.h5ad\"\n",
    "\n",
    "if os.path.exists(output_file):\n",
    "    # Do only if the file exists\n",
    "    print(\"File exists. Proceeding with the task.\")\n",
    "\n",
    "    # Rest of the logic that should only run if file exists\n",
    "else:\n",
    "    url = \"https://datasets.cellxgene.cziscience.com/9daa676b-07ec-4cea-80aa-daa49200aa64.h5ad\"\n",
    "    #Tabula Sapiens is a benchmark, first-draft human cell atlas of over 1.1M cells from 28 organs of 24 normal human subjects. This work is the product of the Tabula Sapiens Consortium. Taking the organs from the same individual controls for genetic background, age, environment, and epigenetic effects, and allows detailed analysis and comparison of cell types that are shared between tissues.\n",
    "    # Get file size for progress bar\n",
    "    response = requests.head(url)\n",
    "    total_size = int(response.headers.get('Content-Length', 0))\n",
    "    \n",
    "    # Stream download with tqdm\n",
    "    with requests.get(url, stream=True) as r:\n",
    "        r.raise_for_status()\n",
    "        with open(output_file, 'wb') as f, tqdm(\n",
    "            total=total_size, unit='B', unit_scale=True, desc=output_file\n",
    "        ) as pbar:\n",
    "            for chunk in r.iter_content(chunk_size=1024 * 1024):\n",
    "                if chunk:\n",
    "                    f.write(chunk)\n",
    "                    pbar.update(len(chunk))\n",
    "                    \n",
    "# Load with scanpy\n",
    "adata = sc.read_h5ad(output_file)\n",
    "print(adata)\n",
    "print(adata.obs.columns) \n",
    "print(adata.var_names)    # Gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6045c03-45f7-4bd6-a3c1-92c1a6ba89f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "from createDataTable import gene_pair0, pop_up_info\n",
    "\n",
    "# Build HGNC-to-Ensembl mapping\n",
    "ensembl_id = dict(zip(pop_up_info['HGNC ID'], pop_up_info['ensembl_gene_id']))\n",
    "\n",
    "# Define HGNC ID extractor\n",
    "def extract_hgnc_id(text):\n",
    "    if pd.isna(text): return None\n",
    "    match = re.search(r'(HGNC:\\d+)', str(text))\n",
    "    return match.group(1) if match else None\n",
    "\n",
    "# --- Prepare ligand dataframe ---\n",
    "ligand_df = gene_pair0[['Ligand', 'Ligand HGNC ID']].copy()\n",
    "ligand_df.columns = ['gene_symbol', 'hgnc_id']  # Standardize column names\n",
    "\n",
    "# --- Prepare receptor dataframe ---\n",
    "receptor_df = gene_pair0[['Receptor', 'Receptor HGNC ID']].copy()\n",
    "receptor_df.columns = ['gene_symbol', 'hgnc_id']\n",
    "\n",
    "# Combine both\n",
    "gene_pair_input = pd.concat([ligand_df, receptor_df], ignore_index=True)\n",
    "\n",
    "# Extract clean HGNC IDs\n",
    "gene_pair_input['hgnc_id'] = gene_pair_input['hgnc_id'].apply(extract_hgnc_id)\n",
    "\n",
    "# Map to Ensembl IDs\n",
    "gene_pair_input['ensembl_id'] = gene_pair_input['hgnc_id'].map(ensembl_id)\n",
    "\n",
    "# Drop duplicates and NaNs if needed\n",
    "gene_pair_input = gene_pair_input.drop_duplicates().dropna(subset=['ensembl_id'])\n",
    "\n",
    "# Final output\n",
    "print(gene_pair_input.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f8c86bb-25b5-41a0-8086-8b515a7689a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- Imports ---\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.sparse import issparse, csr_matrix\n",
    "import plotly.express as px\n",
    "from joblib import Parallel, delayed\n",
    "from tqdm import tqdm\n",
    "# --- Ensure Output Directories Exist ---\n",
    "os.makedirs(f\"{output_dir}heatmap\", exist_ok=True)\n",
    "os.makedirs(f\"{output_dir}umap\", exist_ok=True)\n",
    "\n",
    "# --- Annotations ---\n",
    "tissues = adata.obs[\"tissue_in_publication\"].astype(str).values\n",
    "cell_types = adata.obs[\"cell_type\"].astype(str).values\n",
    "assert len(tissues) == adata.n_obs, \"Tissue annotation length mismatch\"\n",
    "assert len(cell_types) == adata.n_obs, \"Cell type annotation length mismatch\"\n",
    "\n",
    "# --- Gene Lists and Labels ---\n",
    "gene_id_list = gene_pair_input[\"ensembl_id\"].tolist()\n",
    "\n",
    "# --- Limit to First Few for Testing ---\n",
    "test_mode = False  # ← change to True if testing\n",
    "max_genes = 3 if test_mode else None\n",
    "if test_mode:\n",
    "    gene_id_list = gene_id_list[:max_genes]\n",
    "\n",
    "# Strip versions from gene IDs if needed\n",
    "def strip_version(gid):\n",
    "    return gid.split(\".\")[0] if \".\" in gid else gid\n",
    "\n",
    "gene_id_list_stripped = [strip_version(g) for g in gene_id_list]\n",
    "gene_label_map = dict(zip(gene_id_list_stripped,\n",
    "                          gene_pair_input.set_index(\"ensembl_id\").loc[gene_id_list][\"gene_symbol\"].values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d58c8487-db6d-4f42-946f-a2813a7d2667",
   "metadata": {},
   "outputs": [],
   "source": [
    "# If you want to save the scaled data again\n",
    "# --- Precompute Scaled Expression ---\n",
    "def precompute_scaled_expression(adata, gene_id_list):\n",
    "    if \"scale_data\" not in adata.layers:\n",
    "        raise ValueError(\"Layer 'scale_data' not found in adata.\")\n",
    "\n",
    "    X = adata.layers[\"scale_data\"]  # shape: cells x genes\n",
    "    var_names_stripped = [strip_version(g) for g in adata.var_names]\n",
    "    gene_indices = {gene_id: i for i, gene_id in enumerate(var_names_stripped)}\n",
    "\n",
    "    gene_expr_map = {}\n",
    "    missing_genes = []\n",
    "\n",
    "    for gene_id in tqdm(gene_id_list, desc=\"Precomputing scaled expression\"):\n",
    "        if gene_id not in gene_indices:\n",
    "            missing_genes.append(gene_id)\n",
    "            continue\n",
    "        i = gene_indices[gene_id]\n",
    "        x = X[:, i]  # ✅ not transposed\n",
    "        expr = x.toarray().flatten() if hasattr(x, \"toarray\") else x.flatten()\n",
    "        gene_expr_map[gene_id] = expr\n",
    "\n",
    "    print(f\"[✓] Precomputed expression for {len(gene_expr_map)} genes.\")\n",
    "    if missing_genes:\n",
    "        print(f\"[!] Skipped {len(missing_genes)} genes not found in adata.var_names.\")\n",
    "\n",
    "    return gene_expr_map, missing_genes\n",
    "\n",
    "# uncomment for now so it is not accidentally ran\n",
    "# gene_expr_map, _ = precompute_scaled_expression(adata, gene_id_list_stripped)\n",
    "\n",
    "# --- Plot Function ---\n",
    "def plot_gene_umap_expression(adata, gene_id, gene_name, output_dir, gene_expr_map=None):\n",
    "    try:\n",
    "        if gene_expr_map and gene_id in gene_expr_map:\n",
    "            expr = gene_expr_map[gene_id]\n",
    "        else:\n",
    "            if gene_id not in adata.var_names:\n",
    "                print(f\"[!] {gene_id} not in adata.var_names\")\n",
    "                return\n",
    "            expr_data = adata[:, gene_id].X\n",
    "            expr = expr_data.toarray().flatten() if issparse(expr_data) else expr_data.flatten()\n",
    "\n",
    "        df = pd.DataFrame(adata.obsm[\"X_umap\"], columns=[\"UMAP1\", \"UMAP2\"])\n",
    "        df[\"expression\"] = expr\n",
    "        df[\"tissue\"] = tissues\n",
    "        df[\"cell_type\"] = cell_types\n",
    "\n",
    "        fig = px.scatter(\n",
    "            df,\n",
    "            x=\"UMAP1\", y=\"UMAP2\",\n",
    "            color=\"expression\",\n",
    "            color_continuous_scale=\"viridis\",\n",
    "            title=f\"{gene_name} Expression on UMAP\",\n",
    "            width=600, height=600,\n",
    "            render_mode=\"webgl\",\n",
    "            custom_data=[\"tissue\", \"cell_type\"]\n",
    "        )\n",
    "\n",
    "        fig.update_traces(marker=dict(size=3), hovertemplate=\"<br>\".join([\n",
    "            \"Tissue: %{customdata[0]}\",\n",
    "            \"Cell type: %{customdata[1]}\",\n",
    "            \"Expression: %{marker.color:.2f}\"\n",
    "        ]))\n",
    "\n",
    "        fig.update_layout(\n",
    "            plot_bgcolor='white',\n",
    "            paper_bgcolor='white',\n",
    "            legend_title=\"Expr\",\n",
    "            xaxis=dict(showgrid=False, zeroline=False, showticklabels=False),\n",
    "            yaxis=dict(showgrid=False, zeroline=False, showticklabels=False)\n",
    "        )\n",
    "\n",
    "        out_path = os.path.join(output_dir, \"umap\", f\"{gene_name}.html\")\n",
    "        fig.write_html(out_path, include_plotlyjs=\"cdn\")\n",
    "        del fig  # free memory\n",
    "        print(f\"[✓] UMAP written: {out_path}\")\n",
    "    except Exception as e:\n",
    "        print(f\"[✗] Error plotting {gene_name} ({gene_id}): {e}\")\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bec182b-e011-4dae-adc6-9f84a2af4d0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- Wrapper for Parallel Execution ---\n",
    "def generate_umap_wrapper(gene_id_original):\n",
    "    gene_id = strip_version(gene_id_original)\n",
    "    gene_label = gene_label_map.get(gene_id, gene_id)\n",
    "    plot_gene_umap_expression(adata, gene_id, gene_label, output_dir, gene_expr_map)\n",
    "# --- Run in Parallel ---\n",
    "Parallel(n_jobs=16, backend=\"threading\")(\n",
    "    delayed(generate_umap_wrapper)(gene_id)\n",
    "    for gene_id in tqdm(gene_id_list, desc=\"Generating UMAPs\")\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2c9ec79-d72a-4760-8fbc-9e02e7fb8b1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- Chunking Logic --- MAKE SURE you remove the gene_expr_map line of code from above\n",
    "chunk_size = 424\n",
    "num_genes = len(gene_id_list_stripped)\n",
    "\n",
    "for i in range(0, num_genes, chunk_size):\n",
    "    print(f\"\\n[INFO] Processing chunk {i // chunk_size + 1} of {(num_genes - 1) // chunk_size + 1}\")\n",
    "    \n",
    "    # Get the current batch\n",
    "    batch_gene_ids = gene_id_list_stripped[i:i + chunk_size]\n",
    "    \n",
    "    # Precompute expression for this chunk\n",
    "    gene_expr_map, _ = precompute_scaled_expression(adata, batch_gene_ids)\n",
    "\n",
    "    # Wrapper for plotting in parallel\n",
    "    def generate_umap_wrapper(gene_id):\n",
    "        gene_label = gene_label_map.get(gene_id, gene_id)\n",
    "        plot_gene_umap_expression(adata, gene_id, gene_label, output_dir, gene_expr_map)\n",
    "\n",
    "    # Plot in parallel\n",
    "    Parallel(n_jobs=16, backend=\"threading\")(\n",
    "        delayed(generate_umap_wrapper)(gene_id)\n",
    "        for gene_id in tqdm(batch_gene_ids, desc=\"Generating UMAPs\")\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f48ff76-1161-4f80-8a18-35f1911e18ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# normalize all data and log\n",
    "# sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "# sc.pp.log1p(adata)\n",
    "# sc.pp.pca(adata)\n",
    "# sc.pp.neighbors(adata)\n",
    "# sc.tl.umap(adata)\n",
    "# UMAP colored by gene expression\n",
    "#sc.pl.umap(adata, color=gene_pair_input[\"ensembl_id\"][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26054f7e-9d10-44c6-8b86-536cfddaace5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test_genes = [\"ENSG00000141510\", \"ENSG00000148773\"]  # Replace with your own test gene IDs\n",
    "\n",
    "# gene_id_list = test_genes  # or your full list of Ensembl IDs\n",
    "# gene_expr_map, missing_genes = precompute_scaled_expression(adata, gene_id_list)\n",
    "# for gene_id in test_genes:\n",
    "#     gene_name = gene_label_map.get(gene_id, gene_id)\n",
    "#     plot_gene_umap_expression(adata, gene_id, gene_name, output_dir, gene_expr_map)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6856c0e-9007-4cde-ae08-4a3789099e00",
   "metadata": {},
   "outputs": [],
   "source": [
    "X = adata.layers[\"scale_data\"]  # shape: cells x genes\n",
    "var_names_stripped = [strip_version(g) for g in adata.var_names]\n",
    "gene_indices = {gene_id: i for i, gene_id in enumerate(var_names_stripped)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "588a63e1-6968-4759-b7d1-5ee813125fd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- Normalize Ensembl IDs and build label map ---\n",
    "def strip_version(ensembl_id):\n",
    "    return ensembl_id.split('.')[0]\n",
    "\n",
    "# Rebuild gene_label_map using stripped Ensembl IDs\n",
    "gene_label_map = {\n",
    "    strip_version(row[\"ensembl_id\"]): row[\"gene_symbol\"]\n",
    "    for _, row in gene_pair_input.iterrows()\n",
    "}\n",
    "\n",
    "# Strip version from gene_id_list\n",
    "gene_id_list = [strip_version(g) for g in gene_id_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb3a7e6e-93ea-46a3-8553-41a1e20de1ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- Filter to valid genes ---\n",
    "valid_gene_ids = [g for g in gene_id_list if g in gene_indices]\n",
    "rows = [gene_indices[g] for g in valid_gene_ids]\n",
    "missing_genes = [g for g in gene_id_list if g not in gene_indices]\n",
    "print(f\"[!] Skipped {len(missing_genes)} genes not found in adata.var_names.\")\n",
    "len(valid_gene_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b432c1f-0050-48c4-a5c9-97be35164e1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "tissues = adata.obs[\"tissue_in_publication\"].astype(str).values\n",
    "np.unique(tissues)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6e00ec6-6c7f-4672-9793-518ef3170342",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_ont = adata.obs[\"broad_cell_class\"].astype(str).values\n",
    "np.unique(cell_type_ont)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b841edc3-b4aa-4242-a1d6-759dfc960af9",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_cat = pd.read_csv(\"data/grouping_for_heatmap.csv\")\n",
    "mapping = dict(zip(cell_cat[\"cell_type\"], cell_cat[\"category\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6466f2af-dffd-4caf-848f-e9d52bbc8f0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "# If you want to save the scaled data again\n",
    "# with open(\"data/gene_expr_map_scaled.pkl\", \"wb\") as f:\n",
    "#     pickle.dump(gene_expr_map, f)\n",
    "\n",
    "with open(\"data/gene_expr_map_scaled.pkl\", \"rb\") as f:\n",
    "    gene_expr_map = pickle.load(f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "287ea2f4-3a18-4293-b94f-0abd18bb9b0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_expr_map['ENSG00000175899'].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc5b204d-0fa8-4d22-9bf7-9966b79bd6de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Enable test mode\n",
    "test_mode = False\n",
    "max_genes = 3 if test_mode else None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8e6d1cf-f166-4b3f-aa3e-a047848def84",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import plotly.graph_objects as go\n",
    "import plotly.express as px\n",
    "from joblib import Parallel, delayed\n",
    "from tqdm import tqdm\n",
    "from plotly.subplots import make_subplots\n",
    "from scipy.cluster.hierarchy import linkage, leaves_list\n",
    "from scipy.spatial.distance import pdist\n",
    "\n",
    "if test_mode:\n",
    "    valid_gene_ids = valid_gene_ids[:max_genes]\n",
    "colorscale_theme = [\n",
    "    [0.0, \"blue\"],\n",
    "    [0.5, \"lightgray\"],\n",
    "    [1.0, \"red\"]\n",
    "]\n",
    "\n",
    "# Map cell_type → cell_class → color\n",
    "unique_classes = sorted(set(cell_type_ont))\n",
    "# Define color palette\n",
    "color_palette = px.colors.qualitative.Set2\n",
    "\n",
    "# Map: cell_type → cell_class\n",
    "cell_type_to_class = dict(zip(cell_types, cell_type_ont))\n",
    "\n",
    "# Sort cell types by cell class and name\n",
    "sorted_cell_types = sorted(set(cell_types), key=lambda ct: (cell_type_to_class[ct], ct))\n",
    "\n",
    "# Extract cell class order from sorted y-axis\n",
    "sorted_classes_from_yaxis = []\n",
    "seen_classes = set()\n",
    "for ct in sorted_cell_types:\n",
    "    cls = cell_type_to_class[ct]\n",
    "    if cls not in seen_classes:\n",
    "        sorted_classes_from_yaxis.append(cls)\n",
    "        seen_classes.add(cls)\n",
    "\n",
    "# Assign color per class using y-axis order\n",
    "class_to_color = {\n",
    "    cls: color_palette[i % len(color_palette)]\n",
    "    for i, cls in enumerate(sorted_classes_from_yaxis)\n",
    "}\n",
    "\n",
    "# Assign color to cell types based on class\n",
    "cell_type_to_color = {ct: class_to_color[cell_type_to_class[ct]] for ct in set(cell_types)}\n",
    "\n",
    "\n",
    "n_cell = str(len(np.unique(cell_types))) \n",
    "n_cellGroup = str(len(np.unique(cell_type_ont)))\n",
    "n_tissue = str(len(np.unique(tissues)))\n",
    "\n",
    "cell_cat = pd.read_csv(\"data/grouping_for_heatmap.csv\")\n",
    "mapping = dict(zip(cell_cat[\"cell_type\"], cell_cat[\"category\"]))\n",
    "\n",
    "def generate_heatmap(gene_id, gene_label, expr):\n",
    "    # --- Prepare data ---\n",
    "    df = pd.DataFrame({\n",
    "        \"expression\": expr,\n",
    "        \"tissue\": tissues,\n",
    "        \"cell_type\": cell_types,\n",
    "        \"cell_class\": cell_type_ont\n",
    "    })\n",
    "\n",
    "    df['cell_panel'] = df['cell_class'].replace(mapping)\n",
    "\n",
    "    # Ordered categorical cell types\n",
    "    sorted_cell_types = sorted(set(cell_types), key=lambda ct: (cell_type_to_class[ct], ct))\n",
    "    df[\"cell_type\"] = pd.Categorical(df[\"cell_type\"], categories=sorted_cell_types, ordered=True)\n",
    "\n",
    "    # --- Global pivot ---\n",
    "    pivot_df = df.groupby([\"tissue\", \"cell_type\"]).agg(\n",
    "        mean_expression=(\"expression\", \"mean\"),\n",
    "        cell_class=(\"cell_class\", \"first\")\n",
    "    ).reset_index()\n",
    "\n",
    "    global_heatmap = pivot_df.pivot(index=\"cell_type\", columns=\"tissue\", values=\"mean_expression\")\n",
    "\n",
    "    # Global zmin/zmax\n",
    "    global_zmin = global_heatmap.min().min()\n",
    "    global_zmax = global_heatmap.max().max()\n",
    "\n",
    "    # Mapping for customdata\n",
    "    cell_type_class_map = df.drop_duplicates(\"cell_type\").set_index(\"cell_type\")[\"cell_class\"].to_dict()\n",
    "\n",
    "    # --- Tissue clustering ---\n",
    "    clustered_tissues = global_heatmap.T.fillna(0)\n",
    "    linkage_matrix = linkage(pdist(clustered_tissues, metric='euclidean'), method='average')\n",
    "    clustered_tissue_order = clustered_tissues.index[leaves_list(linkage_matrix)]\n",
    "    global_heatmap = global_heatmap[clustered_tissue_order]\n",
    "\n",
    "    # --- Panels setup ---\n",
    "    panel_names = sorted(df['cell_panel'].dropna().unique())\n",
    "    n_panels = len(panel_names)\n",
    "\n",
    "    # Panel size logic\n",
    "    row_heights = []\n",
    "    for panel in panel_names:\n",
    "        n = len(df[df['cell_panel'] == panel][\"cell_type\"].unique())\n",
    "        row_heights.append(n)\n",
    "    total_rows = sum(row_heights)\n",
    "    normalized_heights = [h / total_rows for h in row_heights]\n",
    "\n",
    "    # Only 1 column (heatmap), not 2\n",
    "    fig = make_subplots(\n",
    "        rows=n_panels, cols=1,\n",
    "        shared_xaxes=False,\n",
    "        shared_yaxes=True,\n",
    "        vertical_spacing=0.035,\n",
    "        row_heights=normalized_heights,\n",
    "        specs=[[{\"type\": \"heatmap\"}]] * n_panels\n",
    "    )\n",
    "\n",
    "    for i, panel in enumerate(panel_names, start=1):\n",
    "        panel_cell_types = df[df['cell_panel'] == panel][\"cell_type\"].unique()\n",
    "        panel_cell_types = sorted(panel_cell_types, key=lambda ct: cell_type_class_map.get(ct, \"\"))\n",
    "        heatmap_data = global_heatmap.loc[global_heatmap.index.intersection(panel_cell_types)]\n",
    "        y_labels = heatmap_data.index.astype(str).tolist()\n",
    "\n",
    "        hover_class_data = np.array([[cell_type_class_map[ct]] * len(heatmap_data.columns) for ct in y_labels])\n",
    "\n",
    "        # --- Heatmap only ---\n",
    "        base_y_pos = 1 - sum(normalized_heights[:i-1])\n",
    "        y_pos = base_y_pos\n",
    "        if i == 2:\n",
    "            y_pos = base_y_pos - 0.010\n",
    "        if i == 3:\n",
    "            y_pos = base_y_pos - 0.035\n",
    "        elif i == 4:\n",
    "            y_pos = base_y_pos - 0.023\n",
    "        elif i == 5:\n",
    "            y_pos = base_y_pos - 0.026\n",
    "        fig.add_trace(go.Heatmap(\n",
    "            z=heatmap_data.values,\n",
    "            x=heatmap_data.columns,\n",
    "            y=y_labels,\n",
    "            customdata=hover_class_data,\n",
    "            colorscale=colorscale_theme,\n",
    "            zmin=global_zmin,\n",
    "            zmax=global_zmax,\n",
    "            showscale=True,\n",
    "            colorbar=dict(\n",
    "                title=\"Log(Gene Expression)\",\n",
    "                orientation=\"h\",\n",
    "                x=-0.7,\n",
    "                xanchor=\"left\",\n",
    "                y=y_pos,\n",
    "                len=0.9,\n",
    "                thickness=10\n",
    "            ),\n",
    "            hovertemplate=\"Tissue: %{x}<br>Cell type: %{y}<br>Class: %{customdata}<br>Expr: %{z:.2f}<extra></extra>\"\n",
    "        ), row=i, col=1)\n",
    "\n",
    "        # Axes\n",
    "        fig.update_xaxes(#title_text=panel,\n",
    "                         showticklabels=True, \n",
    "                         tickangle=270, \n",
    "                         tickfont=dict(size=9), row=i, col=1)\n",
    "        fig.update_yaxes(\n",
    "            autorange=\"reversed\",\n",
    "            tickfont=dict(size=9),\n",
    "            title_text=panel,\n",
    "            title_font=dict(size=13, color=\"black\"),\n",
    "            title_standoff=5,\n",
    "            row=i,\n",
    "            col=1\n",
    "        )\n",
    "        # Compute the center y-position of this panel (in data coordinates)\n",
    "        mid_index = len(y_labels) // 2\n",
    "        mid_cell = y_labels[mid_index] if y_labels else \"\"\n",
    "        \n",
    "        # Compute the center y-position of this panel (in data coordinates)\n",
    "        mid_index = len(y_labels) // 2\n",
    "        mid_cell = y_labels[mid_index] if y_labels else \"\"\n",
    "        \n",
    "        fig.add_annotation(\n",
    "            text=f\"<b>{panel}</b>\",\n",
    "            xref=f'x{i}',  # panel's x-axis\n",
    "            yref=f'y{i}',  # panel's y-axis\n",
    "            x=heatmap_data.columns[len(heatmap_data.columns) // 2],  # center tissue\n",
    "            y=mid_cell,  # center cell type\n",
    "            showarrow=False,\n",
    "            font=dict(size=14, color='black'),\n",
    "            xanchor=\"center\",\n",
    "            yanchor=\"middle\"\n",
    "        )\n",
    "\n",
    "\n",
    "    # Layout\n",
    "    fig.update_layout(\n",
    "        width=585,\n",
    "        height=total_rows * 20,\n",
    "        plot_bgcolor=\"white\",\n",
    "        paper_bgcolor=\"white\",\n",
    "        margin=dict(l=1, r=1, t=80, b=100),\n",
    "        title=dict(\n",
    "            text=f\"{gene_label} Expression\",\n",
    "            x=0.7,\n",
    "            y= 0.995,\n",
    "            xanchor=\"center\",\n",
    "            yanchor=\"top\",\n",
    "            font=dict(size=18, color=\"black\")\n",
    "        ),\n",
    "        annotations=[\n",
    "            dict(\n",
    "                text=f\"{gene_label} Expression\",\n",
    "                x=0.5,\n",
    "                y=-0.03,\n",
    "                xref=\"paper\",\n",
    "                yref=\"paper\",\n",
    "                xanchor=\"center\",\n",
    "                yanchor=\"bottom\",\n",
    "                showarrow=False,\n",
    "                font=dict(size=18, color=\"black\")\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "    # Add panel titles globally above each subplot\n",
    "    for i, panel in enumerate(panel_names, start=1):\n",
    "        # same as above\n",
    "        base_y_pos = 1 - sum(normalized_heights[:i-1])\n",
    "        y_pos = base_y_pos\n",
    "        if i == 2:\n",
    "            y_pos = base_y_pos - 0.010\n",
    "        if i == 3:\n",
    "            y_pos = base_y_pos - 0.035\n",
    "        elif i == 4:\n",
    "            y_pos = base_y_pos - 0.023\n",
    "        elif i == 5:\n",
    "            y_pos = base_y_pos - 0.026\n",
    "    \n",
    "        fig.add_annotation(\n",
    "            text=f\"<b>{panel}</b>\",\n",
    "            x=0.5,\n",
    "            y=y_pos,\n",
    "            xref=\"paper\",\n",
    "            yref=\"paper\",\n",
    "            showarrow=False,\n",
    "            font=dict(size=14),\n",
    "            xanchor=\"center\",\n",
    "            yanchor=\"bottom\"\n",
    "        )\n",
    "\n",
    "\n",
    "\n",
    "    # Save\n",
    "    output_path = f\"{output_dir}heatmap/{gene_label}.html\"\n",
    "    fig.write_html(output_path, include_plotlyjs=\"cdn\")\n",
    "    print(f\"[✓] Heatmap written: {output_path}\")\n",
    "\n",
    "\n",
    "\n",
    "def generate_heatmap_wrapper(gene_id):\n",
    "    try:\n",
    "        gene_label = gene_label_map[gene_id]\n",
    "        expr = gene_expr_map[gene_id]\n",
    "        generate_heatmap(gene_id, gene_label, expr)\n",
    "    except Exception as e:\n",
    "        print(f\"[!] Error with {gene_id}: {e}\")\n",
    "\n",
    "Parallel(n_jobs=16, backend=\"threading\")(\n",
    "    delayed(generate_heatmap_wrapper)(gene_id)\n",
    "    for gene_id in tqdm(valid_gene_ids, desc=\"Generating heatmaps\")\n",
    ")\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09a9fa51-c0df-40f9-8a35-faf0ed1c20c7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74067451-a374-4bd5-9e7d-91e7576fe0a0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbba7f33-71d8-4552-aeb1-a59cdc3fdc10",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38ceb0b4-9be8-4c53-b93c-e266bc97c03f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.express as px\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "def plot_gene_umap_expression(adata, gene_id, gene_name, output_dir, gene_expr_map=None):\n",
    "    \"\"\"\n",
    "    Plot UMAP expression of a single gene with tissue and cell type in hover.\n",
    "    \n",
    "    Parameters:\n",
    "    - adata: AnnData object\n",
    "    - gene_id: Ensembl gene ID\n",
    "    - gene_name: gene symbol (or fallback name)\n",
    "    - output_dir: where to save HTML\n",
    "    - gene_expr_map: optional dict of {gene_id: precomputed expression array}\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # Ensure output folder exists\n",
    "        os.makedirs(f\"{output_dir}/umap\", exist_ok=True)\n",
    "\n",
    "        # UMAP coordinates\n",
    "        df = pd.DataFrame(adata.obsm[\"X_umap\"], columns=[\"UMAP1\", \"UMAP2\"])\n",
    "\n",
    "        # Expression: use precomputed if available\n",
    "        if gene_expr_map and gene_id in gene_expr_map:\n",
    "            expr = gene_expr_map[gene_id]\n",
    "        else:\n",
    "            # Check existence\n",
    "            if gene_id not in adata.var_names:\n",
    "                print(f\"[!] Skipping {gene_id}: not found in adata.var_names.\")\n",
    "                return\n",
    "            expr_data = adata[:, gene_id].X\n",
    "            expr = expr_data.toarray().flatten() if hasattr(expr_data, \"toarray\") else expr_data.flatten()\n",
    "\n",
    "        # Add expression and metadata\n",
    "        df[\"expression\"] = expr\n",
    "        df[\"tissue\"] = adata.obs[\"tissue\"].astype(str).values\n",
    "        df[\"cell_type\"] = adata.obs[\"cell_type\"].astype(str).values\n",
    "\n",
    "        # Plot\n",
    "        fig = px.scatter(\n",
    "            df,\n",
    "            x=\"UMAP1\", y=\"UMAP2\",\n",
    "            color=\"expression\",\n",
    "            color_continuous_scale=\"viridis\",\n",
    "            title=f\"{gene_name} Expression on UMAP\",\n",
    "            width=600, height=600,\n",
    "            render_mode=\"webgl\",\n",
    "            custom_data=[\"tissue\", \"cell_type\"]\n",
    "        )\n",
    "\n",
    "        fig.update_traces(\n",
    "            marker=dict(size=3),\n",
    "            hovertemplate=\"<br>\".join([\n",
    "                \"Tissue: %{customdata[0]}\",\n",
    "                \"Cell type: %{customdata[1]}\",\n",
    "                \"Expression: %{marker.color:.2f}\"\n",
    "            ])\n",
    "        )\n",
    "\n",
    "        fig.update_layout(\n",
    "            plot_bgcolor='white',\n",
    "            paper_bgcolor='white',\n",
    "            legend_title=\"Expr\",\n",
    "            xaxis=dict(showgrid=False, zeroline=False, showticklabels=False, title=None),\n",
    "            yaxis=dict(showgrid=False, zeroline=False, showticklabels=False, title=None)\n",
    "        )\n",
    "\n",
    "        # Save HTML\n",
    "        fig.write_html(f\"{output_dir}/umap/{gene_name}.html\", include_plotlyjs=\"cdn\")\n",
    "        print(f\"[✓] UMAP saved: {output_dir}/umap/{gene_name}.html\")\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"[✗] Failed for {gene_name} ({gene_id}): {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88aab264-ed57-45d6-8b72-5f21f9c06d32",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f4a933a-e9de-430d-8b1c-9f6433e26f00",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6bc3291-45d8-454d-a777-6dedb90035ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = gene_pair_input[\"ensembl id\"][0]\n",
    "\n",
    "# Build dataframe\n",
    "df = pd.DataFrame(adata.obsm[\"X_umap\"], columns=[\"UMAP1\", \"UMAP2\"])\n",
    "df[\"expression\"] = adata[:, gene].X.toarray().flatten() if hasattr(adata[:, gene].X, \"toarray\") else adata[:, gene].X.flatten()\n",
    "df[\"tissue\"] = adata.obs[\"tissue\"].astype(str).values  # ensure it’s an array\n",
    "df[\"cell_type\"] = adata.obs[\"cell_type\"].astype(str).values  # ensure it’s an array\n",
    "# Ensure `custom_data` is passed correctly as a list of columns\n",
    "fig = px.scatter(\n",
    "    df,\n",
    "    x=\"UMAP1\", y=\"UMAP2\",\n",
    "    color=\"expression\",\n",
    "    color_continuous_scale=\"viridis\",\n",
    "    title=f\"{gene_name[0]} Expression on UMAP\",\n",
    "    width=600, height=600,\n",
    "    render_mode=\"webgl\",  # <-- faster for many points\n",
    "    custom_data=[\"tissue\", \"cell_type\"]  # required for hovertemplate\n",
    ")\n",
    "\n",
    "# Update hover to show tissue properly\n",
    "fig.update_traces(\n",
    "    marker=dict(size=3),\n",
    "    hovertemplate=\"<br>\".join([\n",
    "        \"Tissue: %{customdata[0]}\",\n",
    "        \"Cell type: %{customdata[1]}\",\n",
    "        \"Expression: %{marker.color:.2f}\"\n",
    "    ])\n",
    ")\n",
    "fig.update_layout(\n",
    "    plot_bgcolor='white',\n",
    "    paper_bgcolor='white',\n",
    "    legend_title=\"Expr\",\n",
    "    xaxis=dict(\n",
    "        showgrid=False,\n",
    "        zeroline=False,\n",
    "        showticklabels=False,\n",
    "        title=None\n",
    "    ),\n",
    "    yaxis=dict(\n",
    "        showgrid=False,\n",
    "        zeroline=False,\n",
    "        showticklabels=False,\n",
    "        title=None\n",
    "    )\n",
    ")\n",
    "\n",
    "\n",
    "fig.show()\n",
    "fig.write_html(f\"{output_dir}umap/{gene_name[0]}.html\", include_plotlyjs=\"cdn\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7853fce9-185c-4579-afe8-f0fe32e05ad9",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color='tissue_in_publication',\n",
    "    legend_loc='right margin',  # or 'right margin' for large datasets\n",
    "    size=0.1,  # reduce for large datasets\n",
    "    title='UMAP by Tissue',\n",
    "    frameon=False\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a22a407-63a8-4650-86ff-29a45aae2fcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color='cell_type',\n",
    "    legend_loc='right margin',  # or 'right margin' for large datasets\n",
    "    size=0.1,  # reduce for large datasets\n",
    "    title='UMAP by Cell Type',\n",
    "    frameon=False\n",
    ")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "quarto_llm_env",
   "language": "python",
   "name": "quarto_llm_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
