{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b2788c71-e0c4-4846-8f10-ac693892fb64",
   "metadata": {},
   "source": [
    "# Notebook for R (Biomart Homolog download)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08823232-7311-4f98-b0a7-69db50a7ac0a",
   "metadata": {},
   "source": [
    "# R Notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7c705768-ddd2-4722-a439-7949d29f86f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'05-Jun-2025'"
      ],
      "text/latex": [
       "'05-Jun-2025'"
      ],
      "text/markdown": [
       "'05-Jun-2025'"
      ],
      "text/plain": [
       "[1] \"05-Jun-2025\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "format(Sys.Date(), \"%d-%b-%Y\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5eae9b64-a549-4754-8767-0a1414a501f3",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in library(tidyverse):  ‘tidyverse’ という名前のパッケージはありません \n",
     "output_type": "error",
     "traceback": [
      "Error in library(tidyverse):  ‘tidyverse’ という名前のパッケージはありません \nTraceback:\n",
      "1. stop(packageNotFoundError(package, lib.loc, sys.call()))"
     ]
    }
   ],
   "source": [
    "library(biomaRt)\n",
    "library(tidyverse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3bfb073f-c9be-4f83-95f6-358c39188e96",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 0 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>ensembl_gene_id</th><th scope=col>ensembl_peptide_id</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 0 × 2\n",
       "\\begin{tabular}{ll}\n",
       " ensembl\\_gene\\_id & ensembl\\_peptide\\_id\\\\\n",
       " <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 0 × 2\n",
       "\n",
       "| ensembl_gene_id &lt;chr&gt; | ensembl_peptide_id &lt;chr&gt; |\n",
       "|---|---|\n",
       "\n"
      ],
      "text/plain": [
       "     ensembl_gene_id ensembl_peptide_id"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Connect to Ensembl BioMart\n",
    "ensembl <- useMart(\"ensembl\", dataset = \"hsapiens_gene_ensembl\")\n",
    "human_genes <- getBM(\n",
    "        attributes = c(\"ensembl_gene_id\", \"ensembl_peptide_id\"),\n",
    "        mart = ensembl\n",
    "    )\n",
    "human_genes <- unique(human_genes)\n",
    "human_genes[human_genes$ensembl_gene_id==\"\",]\n",
    "human_genes <- human_genes[human_genes$ensembl_peptide_id != \"\", ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "4bd932d1-aa28-4e48-b5a9-4cadbdaa3970",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 123887 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ensembl_gene_id</th><th scope=col>ensembl_peptide_id</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>6</th><td>ENSG00000198888</td><td>ENSP00000354687</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>ENSG00000198763</td><td>ENSP00000355046</td></tr>\n",
       "\t<tr><th scope=row>16</th><td>ENSG00000198804</td><td>ENSP00000354499</td></tr>\n",
       "\t<tr><th scope=row>19</th><td>ENSG00000198712</td><td>ENSP00000354876</td></tr>\n",
       "\t<tr><th scope=row>21</th><td>ENSG00000228253</td><td>ENSP00000355265</td></tr>\n",
       "\t<tr><th scope=row>22</th><td>ENSG00000198899</td><td>ENSP00000354632</td></tr>\n",
       "\t<tr><th scope=row>23</th><td>ENSG00000198938</td><td>ENSP00000354982</td></tr>\n",
       "\t<tr><th scope=row>25</th><td>ENSG00000198840</td><td>ENSP00000355206</td></tr>\n",
       "\t<tr><th scope=row>27</th><td>ENSG00000212907</td><td>ENSP00000354728</td></tr>\n",
       "\t<tr><th scope=row>28</th><td>ENSG00000198886</td><td>ENSP00000354961</td></tr>\n",
       "\t<tr><th scope=row>32</th><td>ENSG00000198786</td><td>ENSP00000354813</td></tr>\n",
       "\t<tr><th scope=row>33</th><td>ENSG00000198695</td><td>ENSP00000354665</td></tr>\n",
       "\t<tr><th scope=row>35</th><td>ENSG00000198727</td><td>ENSP00000354554</td></tr>\n",
       "\t<tr><th scope=row>41</th><td>ENSG00000271254</td><td>ENSP00000484923</td></tr>\n",
       "\t<tr><th scope=row>42</th><td>ENSG00000271254</td><td>ENSP00000478665</td></tr>\n",
       "\t<tr><th scope=row>45</th><td>ENSG00000278704</td><td>ENSP00000484918</td></tr>\n",
       "\t<tr><th scope=row>60</th><td>ENSG00000277196</td><td>ENSP00000481127</td></tr>\n",
       "\t<tr><th scope=row>61</th><td>ENSG00000277196</td><td>ENSP00000482462</td></tr>\n",
       "\t<tr><th scope=row>75</th><td>ENSG00000278817</td><td>ENSP00000482514</td></tr>\n",
       "\t<tr><th scope=row>76</th><td>ENSG00000276017</td><td>ENSP00000481613</td></tr>\n",
       "\t<tr><th scope=row>89</th><td>ENSG00000277630</td><td>ENSP00000479139</td></tr>\n",
       "\t<tr><th scope=row>90</th><td>ENSG00000277630</td><td>ENSP00000477677</td></tr>\n",
       "\t<tr><th scope=row>98</th><td>ENSG00000276256</td><td>ENSP00000484596</td></tr>\n",
       "\t<tr><th scope=row>126</th><td>ENSG00000268674</td><td>ENSP00000473163</td></tr>\n",
       "\t<tr><th scope=row>127</th><td>ENSG00000277475</td><td>ENSP00000478882</td></tr>\n",
       "\t<tr><th scope=row>135</th><td>ENSG00000273748</td><td>ENSP00000479921</td></tr>\n",
       "\t<tr><th scope=row>141</th><td>ENSG00000278633</td><td>ENSP00000484693</td></tr>\n",
       "\t<tr><th scope=row>150</th><td>ENSG00000276345</td><td>ENSP00000478035</td></tr>\n",
       "\t<tr><th scope=row>165</th><td>ENSG00000277400</td><td>ENSP00000483280</td></tr>\n",
       "\t<tr><th scope=row>166</th><td>ENSG00000274847</td><td>ENSP00000478910</td></tr>\n",
       "\t<tr><th scope=row>⋮</th><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><th scope=row>201391</th><td>ENSG00000198198</td><td>ENSP00000489255</td></tr>\n",
       "\t<tr><th scope=row>201392</th><td>ENSG00000198198</td><td>ENSP00000492385</td></tr>\n",
       "\t<tr><th scope=row>201393</th><td>ENSG00000198198</td><td>ENSP00000492726</td></tr>\n",
       "\t<tr><th scope=row>201397</th><td>ENSG00000142949</td><td>ENSP00000483557</td></tr>\n",
       "\t<tr><th scope=row>201398</th><td>ENSG00000142949</td><td>ENSP00000521028</td></tr>\n",
       "\t<tr><th scope=row>201399</th><td>ENSG00000142949</td><td>ENSP00000521029</td></tr>\n",
       "\t<tr><th scope=row>201400</th><td>ENSG00000142949</td><td>ENSP00000353030</td></tr>\n",
       "\t<tr><th scope=row>201401</th><td>ENSG00000142949</td><td>ENSP00000398822</td></tr>\n",
       "\t<tr><th scope=row>201402</th><td>ENSG00000142949</td><td>ENSP00000413306</td></tr>\n",
       "\t<tr><th scope=row>201403</th><td>ENSG00000142949</td><td>ENSP00000391764</td></tr>\n",
       "\t<tr><th scope=row>201404</th><td>ENSG00000142949</td><td>ENSP00000408952</td></tr>\n",
       "\t<tr><th scope=row>201405</th><td>ENSG00000142949</td><td>ENSP00000416343</td></tr>\n",
       "\t<tr><th scope=row>201406</th><td>ENSG00000142949</td><td>ENSP00000361484</td></tr>\n",
       "\t<tr><th scope=row>201407</th><td>ENSG00000066135</td><td>ENSP00000361473</td></tr>\n",
       "\t<tr><th scope=row>201408</th><td>ENSG00000066135</td><td>ENSP00000493741</td></tr>\n",
       "\t<tr><th scope=row>201409</th><td>ENSG00000066135</td><td>ENSP00000496362</td></tr>\n",
       "\t<tr><th scope=row>201420</th><td>ENSG00000271425</td><td>ENSP00000463957</td></tr>\n",
       "\t<tr><th scope=row>201421</th><td>ENSG00000271425</td><td>ENSP00000483006</td></tr>\n",
       "\t<tr><th scope=row>201424</th><td>ENSG00000188157</td><td>ENSP00000368678</td></tr>\n",
       "\t<tr><th scope=row>201425</th><td>ENSG00000188157</td><td>ENSP00000484607</td></tr>\n",
       "\t<tr><th scope=row>201426</th><td>ENSG00000188157</td><td>ENSP00000499046</td></tr>\n",
       "\t<tr><th scope=row>201427</th><td>ENSG00000188157</td><td>ENSP00000498543</td></tr>\n",
       "\t<tr><th scope=row>201430</th><td>ENSG00000152092</td><td>ENSP00000356629</td></tr>\n",
       "\t<tr><th scope=row>201431</th><td>ENSG00000152092</td><td>ENSP00000521041</td></tr>\n",
       "\t<tr><th scope=row>201432</th><td>ENSG00000152092</td><td>ENSP00000354536</td></tr>\n",
       "\t<tr><th scope=row>201433</th><td>ENSG00000152092</td><td>ENSP00000395041</td></tr>\n",
       "\t<tr><th scope=row>201447</th><td>ENSG00000011021</td><td>ENSP00000308367</td></tr>\n",
       "\t<tr><th scope=row>201448</th><td>ENSG00000011021</td><td>ENSP00000234488</td></tr>\n",
       "\t<tr><th scope=row>201450</th><td>ENSG00000011021</td><td>ENSP00000496938</td></tr>\n",
       "\t<tr><th scope=row>201451</th><td>ENSG00000011021</td><td>ENSP00000365679</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 123887 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & ensembl\\_gene\\_id & ensembl\\_peptide\\_id\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t6 & ENSG00000198888 & ENSP00000354687\\\\\n",
       "\t10 & ENSG00000198763 & ENSP00000355046\\\\\n",
       "\t16 & ENSG00000198804 & ENSP00000354499\\\\\n",
       "\t19 & ENSG00000198712 & ENSP00000354876\\\\\n",
       "\t21 & ENSG00000228253 & ENSP00000355265\\\\\n",
       "\t22 & ENSG00000198899 & ENSP00000354632\\\\\n",
       "\t23 & ENSG00000198938 & ENSP00000354982\\\\\n",
       "\t25 & ENSG00000198840 & ENSP00000355206\\\\\n",
       "\t27 & ENSG00000212907 & ENSP00000354728\\\\\n",
       "\t28 & ENSG00000198886 & ENSP00000354961\\\\\n",
       "\t32 & ENSG00000198786 & ENSP00000354813\\\\\n",
       "\t33 & ENSG00000198695 & ENSP00000354665\\\\\n",
       "\t35 & ENSG00000198727 & ENSP00000354554\\\\\n",
       "\t41 & ENSG00000271254 & ENSP00000484923\\\\\n",
       "\t42 & ENSG00000271254 & ENSP00000478665\\\\\n",
       "\t45 & ENSG00000278704 & ENSP00000484918\\\\\n",
       "\t60 & ENSG00000277196 & ENSP00000481127\\\\\n",
       "\t61 & ENSG00000277196 & ENSP00000482462\\\\\n",
       "\t75 & ENSG00000278817 & ENSP00000482514\\\\\n",
       "\t76 & ENSG00000276017 & ENSP00000481613\\\\\n",
       "\t89 & ENSG00000277630 & ENSP00000479139\\\\\n",
       "\t90 & ENSG00000277630 & ENSP00000477677\\\\\n",
       "\t98 & ENSG00000276256 & ENSP00000484596\\\\\n",
       "\t126 & ENSG00000268674 & ENSP00000473163\\\\\n",
       "\t127 & ENSG00000277475 & ENSP00000478882\\\\\n",
       "\t135 & ENSG00000273748 & ENSP00000479921\\\\\n",
       "\t141 & ENSG00000278633 & ENSP00000484693\\\\\n",
       "\t150 & ENSG00000276345 & ENSP00000478035\\\\\n",
       "\t165 & ENSG00000277400 & ENSP00000483280\\\\\n",
       "\t166 & ENSG00000274847 & ENSP00000478910\\\\\n",
       "\t⋮ & ⋮ & ⋮\\\\\n",
       "\t201391 & ENSG00000198198 & ENSP00000489255\\\\\n",
       "\t201392 & ENSG00000198198 & ENSP00000492385\\\\\n",
       "\t201393 & ENSG00000198198 & ENSP00000492726\\\\\n",
       "\t201397 & ENSG00000142949 & ENSP00000483557\\\\\n",
       "\t201398 & ENSG00000142949 & ENSP00000521028\\\\\n",
       "\t201399 & ENSG00000142949 & ENSP00000521029\\\\\n",
       "\t201400 & ENSG00000142949 & ENSP00000353030\\\\\n",
       "\t201401 & ENSG00000142949 & ENSP00000398822\\\\\n",
       "\t201402 & ENSG00000142949 & ENSP00000413306\\\\\n",
       "\t201403 & ENSG00000142949 & ENSP00000391764\\\\\n",
       "\t201404 & ENSG00000142949 & ENSP00000408952\\\\\n",
       "\t201405 & ENSG00000142949 & ENSP00000416343\\\\\n",
       "\t201406 & ENSG00000142949 & ENSP00000361484\\\\\n",
       "\t201407 & ENSG00000066135 & ENSP00000361473\\\\\n",
       "\t201408 & ENSG00000066135 & ENSP00000493741\\\\\n",
       "\t201409 & ENSG00000066135 & ENSP00000496362\\\\\n",
       "\t201420 & ENSG00000271425 & ENSP00000463957\\\\\n",
       "\t201421 & ENSG00000271425 & ENSP00000483006\\\\\n",
       "\t201424 & ENSG00000188157 & ENSP00000368678\\\\\n",
       "\t201425 & ENSG00000188157 & ENSP00000484607\\\\\n",
       "\t201426 & ENSG00000188157 & ENSP00000499046\\\\\n",
       "\t201427 & ENSG00000188157 & ENSP00000498543\\\\\n",
       "\t201430 & ENSG00000152092 & ENSP00000356629\\\\\n",
       "\t201431 & ENSG00000152092 & ENSP00000521041\\\\\n",
       "\t201432 & ENSG00000152092 & ENSP00000354536\\\\\n",
       "\t201433 & ENSG00000152092 & ENSP00000395041\\\\\n",
       "\t201447 & ENSG00000011021 & ENSP00000308367\\\\\n",
       "\t201448 & ENSG00000011021 & ENSP00000234488\\\\\n",
       "\t201450 & ENSG00000011021 & ENSP00000496938\\\\\n",
       "\t201451 & ENSG00000011021 & ENSP00000365679\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 123887 × 2\n",
       "\n",
       "| <!--/--> | ensembl_gene_id &lt;chr&gt; | ensembl_peptide_id &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 6 | ENSG00000198888 | ENSP00000354687 |\n",
       "| 10 | ENSG00000198763 | ENSP00000355046 |\n",
       "| 16 | ENSG00000198804 | ENSP00000354499 |\n",
       "| 19 | ENSG00000198712 | ENSP00000354876 |\n",
       "| 21 | ENSG00000228253 | ENSP00000355265 |\n",
       "| 22 | ENSG00000198899 | ENSP00000354632 |\n",
       "| 23 | ENSG00000198938 | ENSP00000354982 |\n",
       "| 25 | ENSG00000198840 | ENSP00000355206 |\n",
       "| 27 | ENSG00000212907 | ENSP00000354728 |\n",
       "| 28 | ENSG00000198886 | ENSP00000354961 |\n",
       "| 32 | ENSG00000198786 | ENSP00000354813 |\n",
       "| 33 | ENSG00000198695 | ENSP00000354665 |\n",
       "| 35 | ENSG00000198727 | ENSP00000354554 |\n",
       "| 41 | ENSG00000271254 | ENSP00000484923 |\n",
       "| 42 | ENSG00000271254 | ENSP00000478665 |\n",
       "| 45 | ENSG00000278704 | ENSP00000484918 |\n",
       "| 60 | ENSG00000277196 | ENSP00000481127 |\n",
       "| 61 | ENSG00000277196 | ENSP00000482462 |\n",
       "| 75 | ENSG00000278817 | ENSP00000482514 |\n",
       "| 76 | ENSG00000276017 | ENSP00000481613 |\n",
       "| 89 | ENSG00000277630 | ENSP00000479139 |\n",
       "| 90 | ENSG00000277630 | ENSP00000477677 |\n",
       "| 98 | ENSG00000276256 | ENSP00000484596 |\n",
       "| 126 | ENSG00000268674 | ENSP00000473163 |\n",
       "| 127 | ENSG00000277475 | ENSP00000478882 |\n",
       "| 135 | ENSG00000273748 | ENSP00000479921 |\n",
       "| 141 | ENSG00000278633 | ENSP00000484693 |\n",
       "| 150 | ENSG00000276345 | ENSP00000478035 |\n",
       "| 165 | ENSG00000277400 | ENSP00000483280 |\n",
       "| 166 | ENSG00000274847 | ENSP00000478910 |\n",
       "| ⋮ | ⋮ | ⋮ |\n",
       "| 201391 | ENSG00000198198 | ENSP00000489255 |\n",
       "| 201392 | ENSG00000198198 | ENSP00000492385 |\n",
       "| 201393 | ENSG00000198198 | ENSP00000492726 |\n",
       "| 201397 | ENSG00000142949 | ENSP00000483557 |\n",
       "| 201398 | ENSG00000142949 | ENSP00000521028 |\n",
       "| 201399 | ENSG00000142949 | ENSP00000521029 |\n",
       "| 201400 | ENSG00000142949 | ENSP00000353030 |\n",
       "| 201401 | ENSG00000142949 | ENSP00000398822 |\n",
       "| 201402 | ENSG00000142949 | ENSP00000413306 |\n",
       "| 201403 | ENSG00000142949 | ENSP00000391764 |\n",
       "| 201404 | ENSG00000142949 | ENSP00000408952 |\n",
       "| 201405 | ENSG00000142949 | ENSP00000416343 |\n",
       "| 201406 | ENSG00000142949 | ENSP00000361484 |\n",
       "| 201407 | ENSG00000066135 | ENSP00000361473 |\n",
       "| 201408 | ENSG00000066135 | ENSP00000493741 |\n",
       "| 201409 | ENSG00000066135 | ENSP00000496362 |\n",
       "| 201420 | ENSG00000271425 | ENSP00000463957 |\n",
       "| 201421 | ENSG00000271425 | ENSP00000483006 |\n",
       "| 201424 | ENSG00000188157 | ENSP00000368678 |\n",
       "| 201425 | ENSG00000188157 | ENSP00000484607 |\n",
       "| 201426 | ENSG00000188157 | ENSP00000499046 |\n",
       "| 201427 | ENSG00000188157 | ENSP00000498543 |\n",
       "| 201430 | ENSG00000152092 | ENSP00000356629 |\n",
       "| 201431 | ENSG00000152092 | ENSP00000521041 |\n",
       "| 201432 | ENSG00000152092 | ENSP00000354536 |\n",
       "| 201433 | ENSG00000152092 | ENSP00000395041 |\n",
       "| 201447 | ENSG00000011021 | ENSP00000308367 |\n",
       "| 201448 | ENSG00000011021 | ENSP00000234488 |\n",
       "| 201450 | ENSG00000011021 | ENSP00000496938 |\n",
       "| 201451 | ENSG00000011021 | ENSP00000365679 |\n",
       "\n"
      ],
      "text/plain": [
       "       ensembl_gene_id ensembl_peptide_id\n",
       "6      ENSG00000198888 ENSP00000354687   \n",
       "10     ENSG00000198763 ENSP00000355046   \n",
       "16     ENSG00000198804 ENSP00000354499   \n",
       "19     ENSG00000198712 ENSP00000354876   \n",
       "21     ENSG00000228253 ENSP00000355265   \n",
       "22     ENSG00000198899 ENSP00000354632   \n",
       "23     ENSG00000198938 ENSP00000354982   \n",
       "25     ENSG00000198840 ENSP00000355206   \n",
       "27     ENSG00000212907 ENSP00000354728   \n",
       "28     ENSG00000198886 ENSP00000354961   \n",
       "32     ENSG00000198786 ENSP00000354813   \n",
       "33     ENSG00000198695 ENSP00000354665   \n",
       "35     ENSG00000198727 ENSP00000354554   \n",
       "41     ENSG00000271254 ENSP00000484923   \n",
       "42     ENSG00000271254 ENSP00000478665   \n",
       "45     ENSG00000278704 ENSP00000484918   \n",
       "60     ENSG00000277196 ENSP00000481127   \n",
       "61     ENSG00000277196 ENSP00000482462   \n",
       "75     ENSG00000278817 ENSP00000482514   \n",
       "76     ENSG00000276017 ENSP00000481613   \n",
       "89     ENSG00000277630 ENSP00000479139   \n",
       "90     ENSG00000277630 ENSP00000477677   \n",
       "98     ENSG00000276256 ENSP00000484596   \n",
       "126    ENSG00000268674 ENSP00000473163   \n",
       "127    ENSG00000277475 ENSP00000478882   \n",
       "135    ENSG00000273748 ENSP00000479921   \n",
       "141    ENSG00000278633 ENSP00000484693   \n",
       "150    ENSG00000276345 ENSP00000478035   \n",
       "165    ENSG00000277400 ENSP00000483280   \n",
       "166    ENSG00000274847 ENSP00000478910   \n",
       "⋮      ⋮               ⋮                 \n",
       "201391 ENSG00000198198 ENSP00000489255   \n",
       "201392 ENSG00000198198 ENSP00000492385   \n",
       "201393 ENSG00000198198 ENSP00000492726   \n",
       "201397 ENSG00000142949 ENSP00000483557   \n",
       "201398 ENSG00000142949 ENSP00000521028   \n",
       "201399 ENSG00000142949 ENSP00000521029   \n",
       "201400 ENSG00000142949 ENSP00000353030   \n",
       "201401 ENSG00000142949 ENSP00000398822   \n",
       "201402 ENSG00000142949 ENSP00000413306   \n",
       "201403 ENSG00000142949 ENSP00000391764   \n",
       "201404 ENSG00000142949 ENSP00000408952   \n",
       "201405 ENSG00000142949 ENSP00000416343   \n",
       "201406 ENSG00000142949 ENSP00000361484   \n",
       "201407 ENSG00000066135 ENSP00000361473   \n",
       "201408 ENSG00000066135 ENSP00000493741   \n",
       "201409 ENSG00000066135 ENSP00000496362   \n",
       "201420 ENSG00000271425 ENSP00000463957   \n",
       "201421 ENSG00000271425 ENSP00000483006   \n",
       "201424 ENSG00000188157 ENSP00000368678   \n",
       "201425 ENSG00000188157 ENSP00000484607   \n",
       "201426 ENSG00000188157 ENSP00000499046   \n",
       "201427 ENSG00000188157 ENSP00000498543   \n",
       "201430 ENSG00000152092 ENSP00000356629   \n",
       "201431 ENSG00000152092 ENSP00000521041   \n",
       "201432 ENSG00000152092 ENSP00000354536   \n",
       "201433 ENSG00000152092 ENSP00000395041   \n",
       "201447 ENSG00000011021 ENSP00000308367   \n",
       "201448 ENSG00000011021 ENSP00000234488   \n",
       "201450 ENSG00000011021 ENSP00000496938   \n",
       "201451 ENSG00000011021 ENSP00000365679   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "human_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "a010c869-62ca-4f2d-b8c7-10d1b812fbe4",
   "metadata": {},
   "outputs": [],
   "source": [
    "readr::write_csv(human_genes, \"data/ensg_ensp_biomart.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3217174b-41a9-411d-8ebc-0cdd726e6c37",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 3170 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>name</th><th scope=col>description</th><th scope=col>page</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>ensembl_gene_id                </td><td>Gene stable ID                              </td><td>feature_page</td></tr>\n",
       "\t<tr><td>ensembl_gene_id_version        </td><td>Gene stable ID version                      </td><td>feature_page</td></tr>\n",
       "\t<tr><td>ensembl_transcript_id          </td><td>Transcript stable ID                        </td><td>feature_page</td></tr>\n",
       "\t<tr><td>ensembl_transcript_id_version  </td><td>Transcript stable ID version                </td><td>feature_page</td></tr>\n",
       "\t<tr><td>ensembl_peptide_id             </td><td>Protein stable ID                           </td><td>feature_page</td></tr>\n",
       "\t<tr><td>ensembl_peptide_id_version     </td><td>Protein stable ID version                   </td><td>feature_page</td></tr>\n",
       "\t<tr><td>ensembl_exon_id                </td><td>Exon stable ID                              </td><td>feature_page</td></tr>\n",
       "\t<tr><td>description                    </td><td>Gene description                            </td><td>feature_page</td></tr>\n",
       "\t<tr><td>chromosome_name                </td><td>Chromosome/scaffold name                    </td><td>feature_page</td></tr>\n",
       "\t<tr><td>start_position                 </td><td>Gene start (bp)                             </td><td>feature_page</td></tr>\n",
       "\t<tr><td>end_position                   </td><td>Gene end (bp)                               </td><td>feature_page</td></tr>\n",
       "\t<tr><td>strand                         </td><td>Strand                                      </td><td>feature_page</td></tr>\n",
       "\t<tr><td>band                           </td><td>Karyotype band                              </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcript_start               </td><td>Transcript start (bp)                       </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcript_end                 </td><td>Transcript end (bp)                         </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcription_start_site       </td><td>Transcription start site (TSS)              </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcript_length              </td><td>Transcript length (including UTRs and CDS)  </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcript_tsl                 </td><td>Transcript support level (TSL)              </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcript_gencode_basic       </td><td>GENCODE basic annotation                    </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcript_primary_basic       </td><td>GENCODE primary annotation                  </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcript_appris              </td><td>APPRIS annotation                           </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcript_is_canonical        </td><td>Ensembl Canonical                           </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcript_mane_select         </td><td>RefSeq match transcript (MANE Select)       </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcript_mane_plus_clinical  </td><td>RefSeq match transcript (MANE Plus Clinical)</td><td>feature_page</td></tr>\n",
       "\t<tr><td>external_gene_name             </td><td>Gene name                                   </td><td>feature_page</td></tr>\n",
       "\t<tr><td>external_gene_source           </td><td>Source of gene name                         </td><td>feature_page</td></tr>\n",
       "\t<tr><td>external_transcript_name       </td><td>Transcript name                             </td><td>feature_page</td></tr>\n",
       "\t<tr><td>external_transcript_source_name</td><td>Source of transcript name                   </td><td>feature_page</td></tr>\n",
       "\t<tr><td>transcript_count               </td><td>Transcript count                            </td><td>feature_page</td></tr>\n",
       "\t<tr><td>percentage_gene_gc_content     </td><td>Gene % GC content                           </td><td>feature_page</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>5_utr_end                    </td><td>5' UTR end                                </td><td>sequences</td></tr>\n",
       "\t<tr><td>3_utr_start                  </td><td>3' UTR start                              </td><td>sequences</td></tr>\n",
       "\t<tr><td>3_utr_end                    </td><td>3' UTR end                                </td><td>sequences</td></tr>\n",
       "\t<tr><td>ensembl_transcript_id        </td><td>Transcript stable ID                      </td><td>sequences</td></tr>\n",
       "\t<tr><td>ensembl_transcript_id_version</td><td>Transcript stable ID version              </td><td>sequences</td></tr>\n",
       "\t<tr><td>ensembl_peptide_id           </td><td>Protein stable ID                         </td><td>sequences</td></tr>\n",
       "\t<tr><td>ensembl_peptide_id_version   </td><td>Protein stable ID version                 </td><td>sequences</td></tr>\n",
       "\t<tr><td>transcript_biotype           </td><td>Transcript type                           </td><td>sequences</td></tr>\n",
       "\t<tr><td>transcript_version           </td><td>Version (transcript)                      </td><td>sequences</td></tr>\n",
       "\t<tr><td>peptide_version              </td><td>Version (protein)                         </td><td>sequences</td></tr>\n",
       "\t<tr><td>strand                       </td><td>Strand                                    </td><td>sequences</td></tr>\n",
       "\t<tr><td>transcript_start             </td><td>Transcript start (bp)                     </td><td>sequences</td></tr>\n",
       "\t<tr><td>transcript_end               </td><td>Transcript end (bp)                       </td><td>sequences</td></tr>\n",
       "\t<tr><td>transcription_start_site     </td><td>Transcription start site (TSS)            </td><td>sequences</td></tr>\n",
       "\t<tr><td>transcript_length            </td><td>Transcript length (including UTRs and CDS)</td><td>sequences</td></tr>\n",
       "\t<tr><td>cds_length                   </td><td>CDS Length                                </td><td>sequences</td></tr>\n",
       "\t<tr><td>cds_start                    </td><td>CDS start                                 </td><td>sequences</td></tr>\n",
       "\t<tr><td>cds_end                      </td><td>CDS end                                   </td><td>sequences</td></tr>\n",
       "\t<tr><td>ensembl_exon_id              </td><td>Exon stable ID                            </td><td>sequences</td></tr>\n",
       "\t<tr><td>exon_chrom_start             </td><td>Exon region start (bp)                    </td><td>sequences</td></tr>\n",
       "\t<tr><td>exon_chrom_end               </td><td>Exon region end (bp)                      </td><td>sequences</td></tr>\n",
       "\t<tr><td>strand                       </td><td>Strand                                    </td><td>sequences</td></tr>\n",
       "\t<tr><td>rank                         </td><td>Exon rank in transcript                   </td><td>sequences</td></tr>\n",
       "\t<tr><td>phase                        </td><td>Start phase                               </td><td>sequences</td></tr>\n",
       "\t<tr><td>end_phase                    </td><td>End phase                                 </td><td>sequences</td></tr>\n",
       "\t<tr><td>cdna_coding_start            </td><td>cDNA coding start                         </td><td>sequences</td></tr>\n",
       "\t<tr><td>cdna_coding_end              </td><td>cDNA coding end                           </td><td>sequences</td></tr>\n",
       "\t<tr><td>genomic_coding_start         </td><td>Genomic coding start                      </td><td>sequences</td></tr>\n",
       "\t<tr><td>genomic_coding_end           </td><td>Genomic coding end                        </td><td>sequences</td></tr>\n",
       "\t<tr><td>is_constitutive              </td><td>Constitutive exon                         </td><td>sequences</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3170 × 3\n",
       "\\begin{tabular}{lll}\n",
       " name & description & page\\\\\n",
       " <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t ensembl\\_gene\\_id                 & Gene stable ID                               & feature\\_page\\\\\n",
       "\t ensembl\\_gene\\_id\\_version         & Gene stable ID version                       & feature\\_page\\\\\n",
       "\t ensembl\\_transcript\\_id           & Transcript stable ID                         & feature\\_page\\\\\n",
       "\t ensembl\\_transcript\\_id\\_version   & Transcript stable ID version                 & feature\\_page\\\\\n",
       "\t ensembl\\_peptide\\_id              & Protein stable ID                            & feature\\_page\\\\\n",
       "\t ensembl\\_peptide\\_id\\_version      & Protein stable ID version                    & feature\\_page\\\\\n",
       "\t ensembl\\_exon\\_id                 & Exon stable ID                               & feature\\_page\\\\\n",
       "\t description                     & Gene description                             & feature\\_page\\\\\n",
       "\t chromosome\\_name                 & Chromosome/scaffold name                     & feature\\_page\\\\\n",
       "\t start\\_position                  & Gene start (bp)                              & feature\\_page\\\\\n",
       "\t end\\_position                    & Gene end (bp)                                & feature\\_page\\\\\n",
       "\t strand                          & Strand                                       & feature\\_page\\\\\n",
       "\t band                            & Karyotype band                               & feature\\_page\\\\\n",
       "\t transcript\\_start                & Transcript start (bp)                        & feature\\_page\\\\\n",
       "\t transcript\\_end                  & Transcript end (bp)                          & feature\\_page\\\\\n",
       "\t transcription\\_start\\_site        & Transcription start site (TSS)               & feature\\_page\\\\\n",
       "\t transcript\\_length               & Transcript length (including UTRs and CDS)   & feature\\_page\\\\\n",
       "\t transcript\\_tsl                  & Transcript support level (TSL)               & feature\\_page\\\\\n",
       "\t transcript\\_gencode\\_basic        & GENCODE basic annotation                     & feature\\_page\\\\\n",
       "\t transcript\\_primary\\_basic        & GENCODE primary annotation                   & feature\\_page\\\\\n",
       "\t transcript\\_appris               & APPRIS annotation                            & feature\\_page\\\\\n",
       "\t transcript\\_is\\_canonical         & Ensembl Canonical                            & feature\\_page\\\\\n",
       "\t transcript\\_mane\\_select          & RefSeq match transcript (MANE Select)        & feature\\_page\\\\\n",
       "\t transcript\\_mane\\_plus\\_clinical   & RefSeq match transcript (MANE Plus Clinical) & feature\\_page\\\\\n",
       "\t external\\_gene\\_name              & Gene name                                    & feature\\_page\\\\\n",
       "\t external\\_gene\\_source            & Source of gene name                          & feature\\_page\\\\\n",
       "\t external\\_transcript\\_name        & Transcript name                              & feature\\_page\\\\\n",
       "\t external\\_transcript\\_source\\_name & Source of transcript name                    & feature\\_page\\\\\n",
       "\t transcript\\_count                & Transcript count                             & feature\\_page\\\\\n",
       "\t percentage\\_gene\\_gc\\_content      & Gene \\% GC content                            & feature\\_page\\\\\n",
       "\t ⋮ & ⋮ & ⋮\\\\\n",
       "\t 5\\_utr\\_end                     & 5' UTR end                                 & sequences\\\\\n",
       "\t 3\\_utr\\_start                   & 3' UTR start                               & sequences\\\\\n",
       "\t 3\\_utr\\_end                     & 3' UTR end                                 & sequences\\\\\n",
       "\t ensembl\\_transcript\\_id         & Transcript stable ID                       & sequences\\\\\n",
       "\t ensembl\\_transcript\\_id\\_version & Transcript stable ID version               & sequences\\\\\n",
       "\t ensembl\\_peptide\\_id            & Protein stable ID                          & sequences\\\\\n",
       "\t ensembl\\_peptide\\_id\\_version    & Protein stable ID version                  & sequences\\\\\n",
       "\t transcript\\_biotype            & Transcript type                            & sequences\\\\\n",
       "\t transcript\\_version            & Version (transcript)                       & sequences\\\\\n",
       "\t peptide\\_version               & Version (protein)                          & sequences\\\\\n",
       "\t strand                        & Strand                                     & sequences\\\\\n",
       "\t transcript\\_start              & Transcript start (bp)                      & sequences\\\\\n",
       "\t transcript\\_end                & Transcript end (bp)                        & sequences\\\\\n",
       "\t transcription\\_start\\_site      & Transcription start site (TSS)             & sequences\\\\\n",
       "\t transcript\\_length             & Transcript length (including UTRs and CDS) & sequences\\\\\n",
       "\t cds\\_length                    & CDS Length                                 & sequences\\\\\n",
       "\t cds\\_start                     & CDS start                                  & sequences\\\\\n",
       "\t cds\\_end                       & CDS end                                    & sequences\\\\\n",
       "\t ensembl\\_exon\\_id               & Exon stable ID                             & sequences\\\\\n",
       "\t exon\\_chrom\\_start              & Exon region start (bp)                     & sequences\\\\\n",
       "\t exon\\_chrom\\_end                & Exon region end (bp)                       & sequences\\\\\n",
       "\t strand                        & Strand                                     & sequences\\\\\n",
       "\t rank                          & Exon rank in transcript                    & sequences\\\\\n",
       "\t phase                         & Start phase                                & sequences\\\\\n",
       "\t end\\_phase                     & End phase                                  & sequences\\\\\n",
       "\t cdna\\_coding\\_start             & cDNA coding start                          & sequences\\\\\n",
       "\t cdna\\_coding\\_end               & cDNA coding end                            & sequences\\\\\n",
       "\t genomic\\_coding\\_start          & Genomic coding start                       & sequences\\\\\n",
       "\t genomic\\_coding\\_end            & Genomic coding end                         & sequences\\\\\n",
       "\t is\\_constitutive               & Constitutive exon                          & sequences\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3170 × 3\n",
       "\n",
       "| name &lt;chr&gt; | description &lt;chr&gt; | page &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| ensembl_gene_id                 | Gene stable ID                               | feature_page |\n",
       "| ensembl_gene_id_version         | Gene stable ID version                       | feature_page |\n",
       "| ensembl_transcript_id           | Transcript stable ID                         | feature_page |\n",
       "| ensembl_transcript_id_version   | Transcript stable ID version                 | feature_page |\n",
       "| ensembl_peptide_id              | Protein stable ID                            | feature_page |\n",
       "| ensembl_peptide_id_version      | Protein stable ID version                    | feature_page |\n",
       "| ensembl_exon_id                 | Exon stable ID                               | feature_page |\n",
       "| description                     | Gene description                             | feature_page |\n",
       "| chromosome_name                 | Chromosome/scaffold name                     | feature_page |\n",
       "| start_position                  | Gene start (bp)                              | feature_page |\n",
       "| end_position                    | Gene end (bp)                                | feature_page |\n",
       "| strand                          | Strand                                       | feature_page |\n",
       "| band                            | Karyotype band                               | feature_page |\n",
       "| transcript_start                | Transcript start (bp)                        | feature_page |\n",
       "| transcript_end                  | Transcript end (bp)                          | feature_page |\n",
       "| transcription_start_site        | Transcription start site (TSS)               | feature_page |\n",
       "| transcript_length               | Transcript length (including UTRs and CDS)   | feature_page |\n",
       "| transcript_tsl                  | Transcript support level (TSL)               | feature_page |\n",
       "| transcript_gencode_basic        | GENCODE basic annotation                     | feature_page |\n",
       "| transcript_primary_basic        | GENCODE primary annotation                   | feature_page |\n",
       "| transcript_appris               | APPRIS annotation                            | feature_page |\n",
       "| transcript_is_canonical         | Ensembl Canonical                            | feature_page |\n",
       "| transcript_mane_select          | RefSeq match transcript (MANE Select)        | feature_page |\n",
       "| transcript_mane_plus_clinical   | RefSeq match transcript (MANE Plus Clinical) | feature_page |\n",
       "| external_gene_name              | Gene name                                    | feature_page |\n",
       "| external_gene_source            | Source of gene name                          | feature_page |\n",
       "| external_transcript_name        | Transcript name                              | feature_page |\n",
       "| external_transcript_source_name | Source of transcript name                    | feature_page |\n",
       "| transcript_count                | Transcript count                             | feature_page |\n",
       "| percentage_gene_gc_content      | Gene % GC content                            | feature_page |\n",
       "| ⋮ | ⋮ | ⋮ |\n",
       "| 5_utr_end                     | 5' UTR end                                 | sequences |\n",
       "| 3_utr_start                   | 3' UTR start                               | sequences |\n",
       "| 3_utr_end                     | 3' UTR end                                 | sequences |\n",
       "| ensembl_transcript_id         | Transcript stable ID                       | sequences |\n",
       "| ensembl_transcript_id_version | Transcript stable ID version               | sequences |\n",
       "| ensembl_peptide_id            | Protein stable ID                          | sequences |\n",
       "| ensembl_peptide_id_version    | Protein stable ID version                  | sequences |\n",
       "| transcript_biotype            | Transcript type                            | sequences |\n",
       "| transcript_version            | Version (transcript)                       | sequences |\n",
       "| peptide_version               | Version (protein)                          | sequences |\n",
       "| strand                        | Strand                                     | sequences |\n",
       "| transcript_start              | Transcript start (bp)                      | sequences |\n",
       "| transcript_end                | Transcript end (bp)                        | sequences |\n",
       "| transcription_start_site      | Transcription start site (TSS)             | sequences |\n",
       "| transcript_length             | Transcript length (including UTRs and CDS) | sequences |\n",
       "| cds_length                    | CDS Length                                 | sequences |\n",
       "| cds_start                     | CDS start                                  | sequences |\n",
       "| cds_end                       | CDS end                                    | sequences |\n",
       "| ensembl_exon_id               | Exon stable ID                             | sequences |\n",
       "| exon_chrom_start              | Exon region start (bp)                     | sequences |\n",
       "| exon_chrom_end                | Exon region end (bp)                       | sequences |\n",
       "| strand                        | Strand                                     | sequences |\n",
       "| rank                          | Exon rank in transcript                    | sequences |\n",
       "| phase                         | Start phase                                | sequences |\n",
       "| end_phase                     | End phase                                  | sequences |\n",
       "| cdna_coding_start             | cDNA coding start                          | sequences |\n",
       "| cdna_coding_end               | cDNA coding end                            | sequences |\n",
       "| genomic_coding_start          | Genomic coding start                       | sequences |\n",
       "| genomic_coding_end            | Genomic coding end                         | sequences |\n",
       "| is_constitutive               | Constitutive exon                          | sequences |\n",
       "\n"
      ],
      "text/plain": [
       "     name                           \n",
       "1    ensembl_gene_id                \n",
       "2    ensembl_gene_id_version        \n",
       "3    ensembl_transcript_id          \n",
       "4    ensembl_transcript_id_version  \n",
       "5    ensembl_peptide_id             \n",
       "6    ensembl_peptide_id_version     \n",
       "7    ensembl_exon_id                \n",
       "8    description                    \n",
       "9    chromosome_name                \n",
       "10   start_position                 \n",
       "11   end_position                   \n",
       "12   strand                         \n",
       "13   band                           \n",
       "14   transcript_start               \n",
       "15   transcript_end                 \n",
       "16   transcription_start_site       \n",
       "17   transcript_length              \n",
       "18   transcript_tsl                 \n",
       "19   transcript_gencode_basic       \n",
       "20   transcript_primary_basic       \n",
       "21   transcript_appris              \n",
       "22   transcript_is_canonical        \n",
       "23   transcript_mane_select         \n",
       "24   transcript_mane_plus_clinical  \n",
       "25   external_gene_name             \n",
       "26   external_gene_source           \n",
       "27   external_transcript_name       \n",
       "28   external_transcript_source_name\n",
       "29   transcript_count               \n",
       "30   percentage_gene_gc_content     \n",
       "⋮    ⋮                              \n",
       "3141 5_utr_end                      \n",
       "3142 3_utr_start                    \n",
       "3143 3_utr_end                      \n",
       "3144 ensembl_transcript_id          \n",
       "3145 ensembl_transcript_id_version  \n",
       "3146 ensembl_peptide_id             \n",
       "3147 ensembl_peptide_id_version     \n",
       "3148 transcript_biotype             \n",
       "3149 transcript_version             \n",
       "3150 peptide_version                \n",
       "3151 strand                         \n",
       "3152 transcript_start               \n",
       "3153 transcript_end                 \n",
       "3154 transcription_start_site       \n",
       "3155 transcript_length              \n",
       "3156 cds_length                     \n",
       "3157 cds_start                      \n",
       "3158 cds_end                        \n",
       "3159 ensembl_exon_id                \n",
       "3160 exon_chrom_start               \n",
       "3161 exon_chrom_end                 \n",
       "3162 strand                         \n",
       "3163 rank                           \n",
       "3164 phase                          \n",
       "3165 end_phase                      \n",
       "3166 cdna_coding_start              \n",
       "3167 cdna_coding_end                \n",
       "3168 genomic_coding_start           \n",
       "3169 genomic_coding_end             \n",
       "3170 is_constitutive                \n",
       "     description                                  page        \n",
       "1    Gene stable ID                               feature_page\n",
       "2    Gene stable ID version                       feature_page\n",
       "3    Transcript stable ID                         feature_page\n",
       "4    Transcript stable ID version                 feature_page\n",
       "5    Protein stable ID                            feature_page\n",
       "6    Protein stable ID version                    feature_page\n",
       "7    Exon stable ID                               feature_page\n",
       "8    Gene description                             feature_page\n",
       "9    Chromosome/scaffold name                     feature_page\n",
       "10   Gene start (bp)                              feature_page\n",
       "11   Gene end (bp)                                feature_page\n",
       "12   Strand                                       feature_page\n",
       "13   Karyotype band                               feature_page\n",
       "14   Transcript start (bp)                        feature_page\n",
       "15   Transcript end (bp)                          feature_page\n",
       "16   Transcription start site (TSS)               feature_page\n",
       "17   Transcript length (including UTRs and CDS)   feature_page\n",
       "18   Transcript support level (TSL)               feature_page\n",
       "19   GENCODE basic annotation                     feature_page\n",
       "20   GENCODE primary annotation                   feature_page\n",
       "21   APPRIS annotation                            feature_page\n",
       "22   Ensembl Canonical                            feature_page\n",
       "23   RefSeq match transcript (MANE Select)        feature_page\n",
       "24   RefSeq match transcript (MANE Plus Clinical) feature_page\n",
       "25   Gene name                                    feature_page\n",
       "26   Source of gene name                          feature_page\n",
       "27   Transcript name                              feature_page\n",
       "28   Source of transcript name                    feature_page\n",
       "29   Transcript count                             feature_page\n",
       "30   Gene % GC content                            feature_page\n",
       "⋮    ⋮                                            ⋮           \n",
       "3141 5' UTR end                                   sequences   \n",
       "3142 3' UTR start                                 sequences   \n",
       "3143 3' UTR end                                   sequences   \n",
       "3144 Transcript stable ID                         sequences   \n",
       "3145 Transcript stable ID version                 sequences   \n",
       "3146 Protein stable ID                            sequences   \n",
       "3147 Protein stable ID version                    sequences   \n",
       "3148 Transcript type                              sequences   \n",
       "3149 Version (transcript)                         sequences   \n",
       "3150 Version (protein)                            sequences   \n",
       "3151 Strand                                       sequences   \n",
       "3152 Transcript start (bp)                        sequences   \n",
       "3153 Transcript end (bp)                          sequences   \n",
       "3154 Transcription start site (TSS)               sequences   \n",
       "3155 Transcript length (including UTRs and CDS)   sequences   \n",
       "3156 CDS Length                                   sequences   \n",
       "3157 CDS start                                    sequences   \n",
       "3158 CDS end                                      sequences   \n",
       "3159 Exon stable ID                               sequences   \n",
       "3160 Exon region start (bp)                       sequences   \n",
       "3161 Exon region end (bp)                         sequences   \n",
       "3162 Strand                                       sequences   \n",
       "3163 Exon rank in transcript                      sequences   \n",
       "3164 Start phase                                  sequences   \n",
       "3165 End phase                                    sequences   \n",
       "3166 cDNA coding start                            sequences   \n",
       "3167 cDNA coding end                              sequences   \n",
       "3168 Genomic coding start                         sequences   \n",
       "3169 Genomic coding end                           sequences   \n",
       "3170 Constitutive exon                            sequences   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "listAttributes(ensembl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "051a490c-6c6a-4813-8e9d-19d8237085ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to retrieve orthologs for different species\n",
    "get_species_orthologs <- function(species_name) {\n",
    "  # Connect to Ensembl BioMart\n",
    "  ensembl <- useMart(\"ensembl\", dataset = \"hsapiens_gene_ensembl\")\n",
    "\n",
    "  # Retrieve all human genes with their Ensembl IDs and HGNC symbols\n",
    "    human_genes <- getBM(\n",
    "        attributes = c(\"hgnc_id\", \"hgnc_symbol\", \"ensembl_gene_id\"),\n",
    "        mart = ensembl\n",
    "    )\n",
    "  \n",
    "  # Define species-specific attributes dynamically based on species name\n",
    "  species_column <- paste0(species_name, \"_homolog_ensembl_gene\")\n",
    "  species_gene_name <- paste0(species_name, \"_homolog_associated_gene_name\")\n",
    "  \n",
    "  # Get orthologs for the specified species\n",
    "  orthologs <- getBM(\n",
    "    attributes = c(\"ensembl_gene_id\", \n",
    "                   species_column, \n",
    "                   species_gene_name),\n",
    "    mart = ensembl\n",
    "  )\n",
    "  final_result <- merge(human_genes, orthologs, by = \"ensembl_gene_id\", all.x = TRUE)\n",
    "  readr::write_csv(final_result, paste0(\"data/\",species_name, \"_ID_biomart.csv\"))\n",
    "  return(final_result)\n",
    "}\n",
    "\n",
    "# Function to list available attributes for a given Ensembl species dataset\n",
    "list_species_attributes <- function(species_dataset) {\n",
    "  ensembl <- useMart(\"ensembl\", dataset = species_dataset)\n",
    "  attributes <- listAttributes(ensembl)\n",
    "  return(attributes)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fdc0b84f-f1c9-49f7-a05b-6b526b57db6b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 89264 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>ensembl_gene_id</th><th scope=col>hgnc_id</th><th scope=col>hgnc_symbol</th><th scope=col>cjacchus_homolog_ensembl_gene</th><th scope=col>cjacchus_homolog_associated_gene_name</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>ENSG00000000003</td><td>HGNC:11858</td><td>TSPAN6 </td><td>ENSCJAG00000004288</td><td>TSPAN6 </td></tr>\n",
       "\t<tr><td>ENSG00000000005</td><td>HGNC:17757</td><td>TNMD   </td><td>ENSCJAG00000004270</td><td>TNMD   </td></tr>\n",
       "\t<tr><td>ENSG00000000419</td><td>HGNC:3005 </td><td>DPM1   </td><td>ENSCJAG00000000551</td><td>DPM1   </td></tr>\n",
       "\t<tr><td>ENSG00000000457</td><td>HGNC:19285</td><td>SCYL3  </td><td>ENSCJAG00000010050</td><td>SCYL3  </td></tr>\n",
       "\t<tr><td>ENSG00000000460</td><td>HGNC:25565</td><td>FIRRM  </td><td>ENSCJAG00000010088</td><td>FIRRM  </td></tr>\n",
       "\t<tr><td>ENSG00000000938</td><td>HGNC:3697 </td><td>FGR    </td><td>ENSCJAG00000009293</td><td>FGR    </td></tr>\n",
       "\t<tr><td>ENSG00000000971</td><td>HGNC:4883 </td><td>CFH    </td><td>ENSCJAG00000008415</td><td>CFH    </td></tr>\n",
       "\t<tr><td>ENSG00000001036</td><td>HGNC:4008 </td><td>FUCA2  </td><td>ENSCJAG00000019246</td><td>FUCA2  </td></tr>\n",
       "\t<tr><td>ENSG00000001084</td><td>HGNC:4311 </td><td>GCLC   </td><td>ENSCJAG00000011063</td><td>GCLC   </td></tr>\n",
       "\t<tr><td>ENSG00000001167</td><td>HGNC:7804 </td><td>NFYA   </td><td>ENSCJAG00000004074</td><td>NFYA   </td></tr>\n",
       "\t<tr><td>ENSG00000001460</td><td>HGNC:28070</td><td>STPG1  </td><td>ENSCJAG00000020857</td><td>STPG1  </td></tr>\n",
       "\t<tr><td>ENSG00000001461</td><td>HGNC:25233</td><td>NIPAL3 </td><td>ENSCJAG00000020848</td><td>NIPAL3 </td></tr>\n",
       "\t<tr><td>ENSG00000001497</td><td>HGNC:25726</td><td>LAS1L  </td><td>ENSCJAG00000001434</td><td>LAS1L  </td></tr>\n",
       "\t<tr><td>ENSG00000001561</td><td>HGNC:3359 </td><td>ENPP4  </td><td>ENSCJAG00000012114</td><td>ENPP4  </td></tr>\n",
       "\t<tr><td>ENSG00000001617</td><td>HGNC:10728</td><td>SEMA3F </td><td>ENSCJAG00000004851</td><td>SEMA3F </td></tr>\n",
       "\t<tr><td>ENSG00000001626</td><td>HGNC:1884 </td><td>CFTR   </td><td>ENSCJAG00000004895</td><td>CFTR   </td></tr>\n",
       "\t<tr><td>ENSG00000001629</td><td>HGNC:22215</td><td>ANKIB1 </td><td>ENSCJAG00000013964</td><td>ANKIB1 </td></tr>\n",
       "\t<tr><td>ENSG00000001630</td><td>HGNC:2649 </td><td>CYP51A1</td><td>ENSCJAG00000014019</td><td>CYP51A1</td></tr>\n",
       "\t<tr><td>ENSG00000001631</td><td>HGNC:1573 </td><td>KRIT1  </td><td>ENSCJAG00000014000</td><td>KRIT1  </td></tr>\n",
       "\t<tr><td>ENSG00000002016</td><td>HGNC:9824 </td><td>RAD52  </td><td>ENSCJAG00000000631</td><td>RAD52  </td></tr>\n",
       "\t<tr><td>ENSG00000002330</td><td>HGNC:936  </td><td>BAD    </td><td>ENSCJAG00000018074</td><td>BAD    </td></tr>\n",
       "\t<tr><td>ENSG00000002549</td><td>HGNC:18449</td><td>LAP3   </td><td>ENSCJAG00000012680</td><td>LAP3   </td></tr>\n",
       "\t<tr><td>ENSG00000002586</td><td>HGNC:7082 </td><td>CD99   </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000002587</td><td>HGNC:5194 </td><td>HS3ST1 </td><td>ENSCJAG00000012119</td><td>HS3ST1 </td></tr>\n",
       "\t<tr><td>ENSG00000002726</td><td>HGNC:80   </td><td>AOC1   </td><td>ENSCJAG00000020136</td><td>AOC1   </td></tr>\n",
       "\t<tr><td>ENSG00000002745</td><td>HGNC:16267</td><td>WNT16  </td><td>ENSCJAG00000039822</td><td>WNT16  </td></tr>\n",
       "\t<tr><td>ENSG00000002746</td><td>HGNC:22195</td><td>HECW1  </td><td>ENSCJAG00000007768</td><td>HECW1  </td></tr>\n",
       "\t<tr><td>ENSG00000002822</td><td>HGNC:6762 </td><td>MAD1L1 </td><td>ENSCJAG00000001463</td><td>MAD1L1 </td></tr>\n",
       "\t<tr><td>ENSG00000002834</td><td>HGNC:6513 </td><td>LASP1  </td><td>ENSCJAG00000008328</td><td>LASP1  </td></tr>\n",
       "\t<tr><td>ENSG00000002919</td><td>HGNC:14975</td><td>SNX11  </td><td>ENSCJAG00000039549</td><td>SNX11  </td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>ENSG00000310537</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310539</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310540</td><td>HGNC:56008</td><td>LINC02968</td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310541</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310542</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310543</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310544</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310545</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310546</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310547</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310548</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310549</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310550</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310551</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310552</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310553</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310554</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310555</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310556</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310557</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310558</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310559</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310560</td><td>HGNC:27849</td><td>PAXX     </td><td>ENSCJAG00000052807</td><td>PAXX    </td></tr>\n",
       "\t<tr><td>ENSG00000310561</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310562</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310564</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310566</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310567</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "\t<tr><td>ENSG00000310576</td><td>HGNC:26073</td><td>FAM174C  </td><td>ENSCJAG00000005541</td><td>C19orf24</td></tr>\n",
       "\t<tr><td>ENSG00000310577</td><td>          </td><td>         </td><td>                  </td><td>        </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 89264 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " ensembl\\_gene\\_id & hgnc\\_id & hgnc\\_symbol & cjacchus\\_homolog\\_ensembl\\_gene & cjacchus\\_homolog\\_associated\\_gene\\_name\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t ENSG00000000003 & HGNC:11858 & TSPAN6  & ENSCJAG00000004288 & TSPAN6 \\\\\n",
       "\t ENSG00000000005 & HGNC:17757 & TNMD    & ENSCJAG00000004270 & TNMD   \\\\\n",
       "\t ENSG00000000419 & HGNC:3005  & DPM1    & ENSCJAG00000000551 & DPM1   \\\\\n",
       "\t ENSG00000000457 & HGNC:19285 & SCYL3   & ENSCJAG00000010050 & SCYL3  \\\\\n",
       "\t ENSG00000000460 & HGNC:25565 & FIRRM   & ENSCJAG00000010088 & FIRRM  \\\\\n",
       "\t ENSG00000000938 & HGNC:3697  & FGR     & ENSCJAG00000009293 & FGR    \\\\\n",
       "\t ENSG00000000971 & HGNC:4883  & CFH     & ENSCJAG00000008415 & CFH    \\\\\n",
       "\t ENSG00000001036 & HGNC:4008  & FUCA2   & ENSCJAG00000019246 & FUCA2  \\\\\n",
       "\t ENSG00000001084 & HGNC:4311  & GCLC    & ENSCJAG00000011063 & GCLC   \\\\\n",
       "\t ENSG00000001167 & HGNC:7804  & NFYA    & ENSCJAG00000004074 & NFYA   \\\\\n",
       "\t ENSG00000001460 & HGNC:28070 & STPG1   & ENSCJAG00000020857 & STPG1  \\\\\n",
       "\t ENSG00000001461 & HGNC:25233 & NIPAL3  & ENSCJAG00000020848 & NIPAL3 \\\\\n",
       "\t ENSG00000001497 & HGNC:25726 & LAS1L   & ENSCJAG00000001434 & LAS1L  \\\\\n",
       "\t ENSG00000001561 & HGNC:3359  & ENPP4   & ENSCJAG00000012114 & ENPP4  \\\\\n",
       "\t ENSG00000001617 & HGNC:10728 & SEMA3F  & ENSCJAG00000004851 & SEMA3F \\\\\n",
       "\t ENSG00000001626 & HGNC:1884  & CFTR    & ENSCJAG00000004895 & CFTR   \\\\\n",
       "\t ENSG00000001629 & HGNC:22215 & ANKIB1  & ENSCJAG00000013964 & ANKIB1 \\\\\n",
       "\t ENSG00000001630 & HGNC:2649  & CYP51A1 & ENSCJAG00000014019 & CYP51A1\\\\\n",
       "\t ENSG00000001631 & HGNC:1573  & KRIT1   & ENSCJAG00000014000 & KRIT1  \\\\\n",
       "\t ENSG00000002016 & HGNC:9824  & RAD52   & ENSCJAG00000000631 & RAD52  \\\\\n",
       "\t ENSG00000002330 & HGNC:936   & BAD     & ENSCJAG00000018074 & BAD    \\\\\n",
       "\t ENSG00000002549 & HGNC:18449 & LAP3    & ENSCJAG00000012680 & LAP3   \\\\\n",
       "\t ENSG00000002586 & HGNC:7082  & CD99    &                    &        \\\\\n",
       "\t ENSG00000002587 & HGNC:5194  & HS3ST1  & ENSCJAG00000012119 & HS3ST1 \\\\\n",
       "\t ENSG00000002726 & HGNC:80    & AOC1    & ENSCJAG00000020136 & AOC1   \\\\\n",
       "\t ENSG00000002745 & HGNC:16267 & WNT16   & ENSCJAG00000039822 & WNT16  \\\\\n",
       "\t ENSG00000002746 & HGNC:22195 & HECW1   & ENSCJAG00000007768 & HECW1  \\\\\n",
       "\t ENSG00000002822 & HGNC:6762  & MAD1L1  & ENSCJAG00000001463 & MAD1L1 \\\\\n",
       "\t ENSG00000002834 & HGNC:6513  & LASP1   & ENSCJAG00000008328 & LASP1  \\\\\n",
       "\t ENSG00000002919 & HGNC:14975 & SNX11   & ENSCJAG00000039549 & SNX11  \\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t ENSG00000310537 &            &           &                    &         \\\\\n",
       "\t ENSG00000310539 &            &           &                    &         \\\\\n",
       "\t ENSG00000310540 & HGNC:56008 & LINC02968 &                    &         \\\\\n",
       "\t ENSG00000310541 &            &           &                    &         \\\\\n",
       "\t ENSG00000310542 &            &           &                    &         \\\\\n",
       "\t ENSG00000310543 &            &           &                    &         \\\\\n",
       "\t ENSG00000310544 &            &           &                    &         \\\\\n",
       "\t ENSG00000310545 &            &           &                    &         \\\\\n",
       "\t ENSG00000310546 &            &           &                    &         \\\\\n",
       "\t ENSG00000310547 &            &           &                    &         \\\\\n",
       "\t ENSG00000310548 &            &           &                    &         \\\\\n",
       "\t ENSG00000310549 &            &           &                    &         \\\\\n",
       "\t ENSG00000310550 &            &           &                    &         \\\\\n",
       "\t ENSG00000310551 &            &           &                    &         \\\\\n",
       "\t ENSG00000310552 &            &           &                    &         \\\\\n",
       "\t ENSG00000310553 &            &           &                    &         \\\\\n",
       "\t ENSG00000310554 &            &           &                    &         \\\\\n",
       "\t ENSG00000310555 &            &           &                    &         \\\\\n",
       "\t ENSG00000310556 &            &           &                    &         \\\\\n",
       "\t ENSG00000310557 &            &           &                    &         \\\\\n",
       "\t ENSG00000310558 &            &           &                    &         \\\\\n",
       "\t ENSG00000310559 &            &           &                    &         \\\\\n",
       "\t ENSG00000310560 & HGNC:27849 & PAXX      & ENSCJAG00000052807 & PAXX    \\\\\n",
       "\t ENSG00000310561 &            &           &                    &         \\\\\n",
       "\t ENSG00000310562 &            &           &                    &         \\\\\n",
       "\t ENSG00000310564 &            &           &                    &         \\\\\n",
       "\t ENSG00000310566 &            &           &                    &         \\\\\n",
       "\t ENSG00000310567 &            &           &                    &         \\\\\n",
       "\t ENSG00000310576 & HGNC:26073 & FAM174C   & ENSCJAG00000005541 & C19orf24\\\\\n",
       "\t ENSG00000310577 &            &           &                    &         \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 89264 × 5\n",
       "\n",
       "| ensembl_gene_id &lt;chr&gt; | hgnc_id &lt;chr&gt; | hgnc_symbol &lt;chr&gt; | cjacchus_homolog_ensembl_gene &lt;chr&gt; | cjacchus_homolog_associated_gene_name &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| ENSG00000000003 | HGNC:11858 | TSPAN6  | ENSCJAG00000004288 | TSPAN6  |\n",
       "| ENSG00000000005 | HGNC:17757 | TNMD    | ENSCJAG00000004270 | TNMD    |\n",
       "| ENSG00000000419 | HGNC:3005  | DPM1    | ENSCJAG00000000551 | DPM1    |\n",
       "| ENSG00000000457 | HGNC:19285 | SCYL3   | ENSCJAG00000010050 | SCYL3   |\n",
       "| ENSG00000000460 | HGNC:25565 | FIRRM   | ENSCJAG00000010088 | FIRRM   |\n",
       "| ENSG00000000938 | HGNC:3697  | FGR     | ENSCJAG00000009293 | FGR     |\n",
       "| ENSG00000000971 | HGNC:4883  | CFH     | ENSCJAG00000008415 | CFH     |\n",
       "| ENSG00000001036 | HGNC:4008  | FUCA2   | ENSCJAG00000019246 | FUCA2   |\n",
       "| ENSG00000001084 | HGNC:4311  | GCLC    | ENSCJAG00000011063 | GCLC    |\n",
       "| ENSG00000001167 | HGNC:7804  | NFYA    | ENSCJAG00000004074 | NFYA    |\n",
       "| ENSG00000001460 | HGNC:28070 | STPG1   | ENSCJAG00000020857 | STPG1   |\n",
       "| ENSG00000001461 | HGNC:25233 | NIPAL3  | ENSCJAG00000020848 | NIPAL3  |\n",
       "| ENSG00000001497 | HGNC:25726 | LAS1L   | ENSCJAG00000001434 | LAS1L   |\n",
       "| ENSG00000001561 | HGNC:3359  | ENPP4   | ENSCJAG00000012114 | ENPP4   |\n",
       "| ENSG00000001617 | HGNC:10728 | SEMA3F  | ENSCJAG00000004851 | SEMA3F  |\n",
       "| ENSG00000001626 | HGNC:1884  | CFTR    | ENSCJAG00000004895 | CFTR    |\n",
       "| ENSG00000001629 | HGNC:22215 | ANKIB1  | ENSCJAG00000013964 | ANKIB1  |\n",
       "| ENSG00000001630 | HGNC:2649  | CYP51A1 | ENSCJAG00000014019 | CYP51A1 |\n",
       "| ENSG00000001631 | HGNC:1573  | KRIT1   | ENSCJAG00000014000 | KRIT1   |\n",
       "| ENSG00000002016 | HGNC:9824  | RAD52   | ENSCJAG00000000631 | RAD52   |\n",
       "| ENSG00000002330 | HGNC:936   | BAD     | ENSCJAG00000018074 | BAD     |\n",
       "| ENSG00000002549 | HGNC:18449 | LAP3    | ENSCJAG00000012680 | LAP3    |\n",
       "| ENSG00000002586 | HGNC:7082  | CD99    | <!----> | <!----> |\n",
       "| ENSG00000002587 | HGNC:5194  | HS3ST1  | ENSCJAG00000012119 | HS3ST1  |\n",
       "| ENSG00000002726 | HGNC:80    | AOC1    | ENSCJAG00000020136 | AOC1    |\n",
       "| ENSG00000002745 | HGNC:16267 | WNT16   | ENSCJAG00000039822 | WNT16   |\n",
       "| ENSG00000002746 | HGNC:22195 | HECW1   | ENSCJAG00000007768 | HECW1   |\n",
       "| ENSG00000002822 | HGNC:6762  | MAD1L1  | ENSCJAG00000001463 | MAD1L1  |\n",
       "| ENSG00000002834 | HGNC:6513  | LASP1   | ENSCJAG00000008328 | LASP1   |\n",
       "| ENSG00000002919 | HGNC:14975 | SNX11   | ENSCJAG00000039549 | SNX11   |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| ENSG00000310537 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310539 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310540 | HGNC:56008 | LINC02968 | <!----> | <!----> |\n",
       "| ENSG00000310541 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310542 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310543 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310544 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310545 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310546 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310547 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310548 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310549 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310550 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310551 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310552 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310553 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310554 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310555 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310556 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310557 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310558 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310559 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310560 | HGNC:27849 | PAXX      | ENSCJAG00000052807 | PAXX     |\n",
       "| ENSG00000310561 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310562 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310564 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310566 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310567 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310576 | HGNC:26073 | FAM174C   | ENSCJAG00000005541 | C19orf24 |\n",
       "| ENSG00000310577 | <!----> | <!----> | <!----> | <!----> |\n",
       "\n"
      ],
      "text/plain": [
       "      ensembl_gene_id hgnc_id    hgnc_symbol cjacchus_homolog_ensembl_gene\n",
       "1     ENSG00000000003 HGNC:11858 TSPAN6      ENSCJAG00000004288           \n",
       "2     ENSG00000000005 HGNC:17757 TNMD        ENSCJAG00000004270           \n",
       "3     ENSG00000000419 HGNC:3005  DPM1        ENSCJAG00000000551           \n",
       "4     ENSG00000000457 HGNC:19285 SCYL3       ENSCJAG00000010050           \n",
       "5     ENSG00000000460 HGNC:25565 FIRRM       ENSCJAG00000010088           \n",
       "6     ENSG00000000938 HGNC:3697  FGR         ENSCJAG00000009293           \n",
       "7     ENSG00000000971 HGNC:4883  CFH         ENSCJAG00000008415           \n",
       "8     ENSG00000001036 HGNC:4008  FUCA2       ENSCJAG00000019246           \n",
       "9     ENSG00000001084 HGNC:4311  GCLC        ENSCJAG00000011063           \n",
       "10    ENSG00000001167 HGNC:7804  NFYA        ENSCJAG00000004074           \n",
       "11    ENSG00000001460 HGNC:28070 STPG1       ENSCJAG00000020857           \n",
       "12    ENSG00000001461 HGNC:25233 NIPAL3      ENSCJAG00000020848           \n",
       "13    ENSG00000001497 HGNC:25726 LAS1L       ENSCJAG00000001434           \n",
       "14    ENSG00000001561 HGNC:3359  ENPP4       ENSCJAG00000012114           \n",
       "15    ENSG00000001617 HGNC:10728 SEMA3F      ENSCJAG00000004851           \n",
       "16    ENSG00000001626 HGNC:1884  CFTR        ENSCJAG00000004895           \n",
       "17    ENSG00000001629 HGNC:22215 ANKIB1      ENSCJAG00000013964           \n",
       "18    ENSG00000001630 HGNC:2649  CYP51A1     ENSCJAG00000014019           \n",
       "19    ENSG00000001631 HGNC:1573  KRIT1       ENSCJAG00000014000           \n",
       "20    ENSG00000002016 HGNC:9824  RAD52       ENSCJAG00000000631           \n",
       "21    ENSG00000002330 HGNC:936   BAD         ENSCJAG00000018074           \n",
       "22    ENSG00000002549 HGNC:18449 LAP3        ENSCJAG00000012680           \n",
       "23    ENSG00000002586 HGNC:7082  CD99                                     \n",
       "24    ENSG00000002587 HGNC:5194  HS3ST1      ENSCJAG00000012119           \n",
       "25    ENSG00000002726 HGNC:80    AOC1        ENSCJAG00000020136           \n",
       "26    ENSG00000002745 HGNC:16267 WNT16       ENSCJAG00000039822           \n",
       "27    ENSG00000002746 HGNC:22195 HECW1       ENSCJAG00000007768           \n",
       "28    ENSG00000002822 HGNC:6762  MAD1L1      ENSCJAG00000001463           \n",
       "29    ENSG00000002834 HGNC:6513  LASP1       ENSCJAG00000008328           \n",
       "30    ENSG00000002919 HGNC:14975 SNX11       ENSCJAG00000039549           \n",
       "⋮     ⋮               ⋮          ⋮           ⋮                            \n",
       "89235 ENSG00000310537                                                     \n",
       "89236 ENSG00000310539                                                     \n",
       "89237 ENSG00000310540 HGNC:56008 LINC02968                                \n",
       "89238 ENSG00000310541                                                     \n",
       "89239 ENSG00000310542                                                     \n",
       "89240 ENSG00000310543                                                     \n",
       "89241 ENSG00000310544                                                     \n",
       "89242 ENSG00000310545                                                     \n",
       "89243 ENSG00000310546                                                     \n",
       "89244 ENSG00000310547                                                     \n",
       "89245 ENSG00000310548                                                     \n",
       "89246 ENSG00000310549                                                     \n",
       "89247 ENSG00000310550                                                     \n",
       "89248 ENSG00000310551                                                     \n",
       "89249 ENSG00000310552                                                     \n",
       "89250 ENSG00000310553                                                     \n",
       "89251 ENSG00000310554                                                     \n",
       "89252 ENSG00000310555                                                     \n",
       "89253 ENSG00000310556                                                     \n",
       "89254 ENSG00000310557                                                     \n",
       "89255 ENSG00000310558                                                     \n",
       "89256 ENSG00000310559                                                     \n",
       "89257 ENSG00000310560 HGNC:27849 PAXX        ENSCJAG00000052807           \n",
       "89258 ENSG00000310561                                                     \n",
       "89259 ENSG00000310562                                                     \n",
       "89260 ENSG00000310564                                                     \n",
       "89261 ENSG00000310566                                                     \n",
       "89262 ENSG00000310567                                                     \n",
       "89263 ENSG00000310576 HGNC:26073 FAM174C     ENSCJAG00000005541           \n",
       "89264 ENSG00000310577                                                     \n",
       "      cjacchus_homolog_associated_gene_name\n",
       "1     TSPAN6                               \n",
       "2     TNMD                                 \n",
       "3     DPM1                                 \n",
       "4     SCYL3                                \n",
       "5     FIRRM                                \n",
       "6     FGR                                  \n",
       "7     CFH                                  \n",
       "8     FUCA2                                \n",
       "9     GCLC                                 \n",
       "10    NFYA                                 \n",
       "11    STPG1                                \n",
       "12    NIPAL3                               \n",
       "13    LAS1L                                \n",
       "14    ENPP4                                \n",
       "15    SEMA3F                               \n",
       "16    CFTR                                 \n",
       "17    ANKIB1                               \n",
       "18    CYP51A1                              \n",
       "19    KRIT1                                \n",
       "20    RAD52                                \n",
       "21    BAD                                  \n",
       "22    LAP3                                 \n",
       "23                                         \n",
       "24    HS3ST1                               \n",
       "25    AOC1                                 \n",
       "26    WNT16                                \n",
       "27    HECW1                                \n",
       "28    MAD1L1                               \n",
       "29    LASP1                                \n",
       "30    SNX11                                \n",
       "⋮     ⋮                                    \n",
       "89235                                      \n",
       "89236                                      \n",
       "89237                                      \n",
       "89238                                      \n",
       "89239                                      \n",
       "89240                                      \n",
       "89241                                      \n",
       "89242                                      \n",
       "89243                                      \n",
       "89244                                      \n",
       "89245                                      \n",
       "89246                                      \n",
       "89247                                      \n",
       "89248                                      \n",
       "89249                                      \n",
       "89250                                      \n",
       "89251                                      \n",
       "89252                                      \n",
       "89253                                      \n",
       "89254                                      \n",
       "89255                                      \n",
       "89256                                      \n",
       "89257 PAXX                                 \n",
       "89258                                      \n",
       "89259                                      \n",
       "89260                                      \n",
       "89261                                      \n",
       "89262                                      \n",
       "89263 C19orf24                             \n",
       "89264                                      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#  Marmoset (Callithrix jacchus)\n",
    "get_species_orthologs(\"cjacchus\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3ee88b30-50ba-4ca1-9760-4da4af919b0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 87850 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>ensembl_gene_id</th><th scope=col>hgnc_id</th><th scope=col>hgnc_symbol</th><th scope=col>mmulatta_homolog_ensembl_gene</th><th scope=col>mmulatta_homolog_associated_gene_name</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>ENSG00000000003</td><td>HGNC:11858</td><td>TSPAN6 </td><td>                  </td><td>      </td></tr>\n",
       "\t<tr><td>ENSG00000000005</td><td>HGNC:17757</td><td>TNMD   </td><td>ENSMMUG00000008212</td><td>TNMD  </td></tr>\n",
       "\t<tr><td>ENSG00000000419</td><td>HGNC:3005 </td><td>DPM1   </td><td>ENSMMUG00000002759</td><td>DPM1  </td></tr>\n",
       "\t<tr><td>ENSG00000000457</td><td>HGNC:19285</td><td>SCYL3  </td><td>ENSMMUG00000016583</td><td>SCYL3 </td></tr>\n",
       "\t<tr><td>ENSG00000000460</td><td>HGNC:25565</td><td>FIRRM  </td><td>ENSMMUG00000016582</td><td>FIRRM </td></tr>\n",
       "\t<tr><td>ENSG00000000938</td><td>HGNC:3697 </td><td>FGR    </td><td>ENSMMUG00000014434</td><td>FGR   </td></tr>\n",
       "\t<tr><td>ENSG00000000971</td><td>HGNC:4883 </td><td>CFH    </td><td>ENSMMUG00000045497</td><td>CFH   </td></tr>\n",
       "\t<tr><td>ENSG00000001036</td><td>HGNC:4008 </td><td>FUCA2  </td><td>ENSMMUG00000005366</td><td>FUCA2 </td></tr>\n",
       "\t<tr><td>ENSG00000001084</td><td>HGNC:4311 </td><td>GCLC   </td><td>ENSMMUG00000008684</td><td>GCLC  </td></tr>\n",
       "\t<tr><td>ENSG00000001167</td><td>HGNC:7804 </td><td>NFYA   </td><td>ENSMMUG00000021173</td><td>NFYA  </td></tr>\n",
       "\t<tr><td>ENSG00000001460</td><td>HGNC:28070</td><td>STPG1  </td><td>ENSMMUG00000054255</td><td>STPG1 </td></tr>\n",
       "\t<tr><td>ENSG00000001461</td><td>HGNC:25233</td><td>NIPAL3 </td><td>ENSMMUG00000004466</td><td>NIPAL3</td></tr>\n",
       "\t<tr><td>ENSG00000001497</td><td>HGNC:25726</td><td>LAS1L  </td><td>ENSMMUG00000003922</td><td>LAS1L </td></tr>\n",
       "\t<tr><td>ENSG00000001561</td><td>HGNC:3359 </td><td>ENPP4  </td><td>ENSMMUG00000042260</td><td>ENPP4 </td></tr>\n",
       "\t<tr><td>ENSG00000001617</td><td>HGNC:10728</td><td>SEMA3F </td><td>ENSMMUG00000010337</td><td>SEMA3F</td></tr>\n",
       "\t<tr><td>ENSG00000001626</td><td>HGNC:1884 </td><td>CFTR   </td><td>ENSMMUG00000011269</td><td>CFTR  </td></tr>\n",
       "\t<tr><td>ENSG00000001629</td><td>HGNC:22215</td><td>ANKIB1 </td><td>ENSMMUG00000005120</td><td>ANKIB1</td></tr>\n",
       "\t<tr><td>ENSG00000001630</td><td>HGNC:2649 </td><td>CYP51A1</td><td>                  </td><td>      </td></tr>\n",
       "\t<tr><td>ENSG00000001631</td><td>HGNC:1573 </td><td>KRIT1  </td><td>ENSMMUG00000014907</td><td>KRIT1 </td></tr>\n",
       "\t<tr><td>ENSG00000002016</td><td>HGNC:9824 </td><td>RAD52  </td><td>ENSMMUG00000018654</td><td>RAD52 </td></tr>\n",
       "\t<tr><td>ENSG00000002330</td><td>HGNC:936  </td><td>BAD    </td><td>ENSMMUG00000060596</td><td>BAD   </td></tr>\n",
       "\t<tr><td>ENSG00000002549</td><td>HGNC:18449</td><td>LAP3   </td><td>ENSMMUG00000005732</td><td>LAP3  </td></tr>\n",
       "\t<tr><td>ENSG00000002586</td><td>HGNC:7082 </td><td>CD99   </td><td>ENSMMUG00000018687</td><td>CD99  </td></tr>\n",
       "\t<tr><td>ENSG00000002586</td><td>HGNC:7082 </td><td>CD99   </td><td>ENSMMUG00000055976</td><td>      </td></tr>\n",
       "\t<tr><td>ENSG00000002587</td><td>HGNC:5194 </td><td>HS3ST1 </td><td>ENSMMUG00000001341</td><td>HS3ST1</td></tr>\n",
       "\t<tr><td>ENSG00000002726</td><td>HGNC:80   </td><td>AOC1   </td><td>ENSMMUG00000020435</td><td>AOC1  </td></tr>\n",
       "\t<tr><td>ENSG00000002745</td><td>HGNC:16267</td><td>WNT16  </td><td>ENSMMUG00000002439</td><td>WNT16 </td></tr>\n",
       "\t<tr><td>ENSG00000002746</td><td>HGNC:22195</td><td>HECW1  </td><td>ENSMMUG00000019015</td><td>HECW1 </td></tr>\n",
       "\t<tr><td>ENSG00000002822</td><td>HGNC:6762 </td><td>MAD1L1 </td><td>ENSMMUG00000011659</td><td>MAD1L1</td></tr>\n",
       "\t<tr><td>ENSG00000002834</td><td>HGNC:6513 </td><td>LASP1  </td><td>ENSMMUG00000007799</td><td>LASP1 </td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>ENSG00000310537</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310539</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310540</td><td>HGNC:56008</td><td>LINC02968</td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310541</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310542</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310543</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310544</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310545</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310546</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310547</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310548</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310549</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310550</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310551</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310552</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310553</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310554</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310555</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310556</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310557</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310558</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310559</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310560</td><td>HGNC:27849</td><td>PAXX     </td><td>ENSMMUG00000021432</td><td>PAXX   </td></tr>\n",
       "\t<tr><td>ENSG00000310561</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310562</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310564</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310566</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310567</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "\t<tr><td>ENSG00000310576</td><td>HGNC:26073</td><td>FAM174C  </td><td>ENSMMUG00000022104</td><td>FAM174C</td></tr>\n",
       "\t<tr><td>ENSG00000310577</td><td>          </td><td>         </td><td>                  </td><td>       </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 87850 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " ensembl\\_gene\\_id & hgnc\\_id & hgnc\\_symbol & mmulatta\\_homolog\\_ensembl\\_gene & mmulatta\\_homolog\\_associated\\_gene\\_name\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t ENSG00000000003 & HGNC:11858 & TSPAN6  &                    &       \\\\\n",
       "\t ENSG00000000005 & HGNC:17757 & TNMD    & ENSMMUG00000008212 & TNMD  \\\\\n",
       "\t ENSG00000000419 & HGNC:3005  & DPM1    & ENSMMUG00000002759 & DPM1  \\\\\n",
       "\t ENSG00000000457 & HGNC:19285 & SCYL3   & ENSMMUG00000016583 & SCYL3 \\\\\n",
       "\t ENSG00000000460 & HGNC:25565 & FIRRM   & ENSMMUG00000016582 & FIRRM \\\\\n",
       "\t ENSG00000000938 & HGNC:3697  & FGR     & ENSMMUG00000014434 & FGR   \\\\\n",
       "\t ENSG00000000971 & HGNC:4883  & CFH     & ENSMMUG00000045497 & CFH   \\\\\n",
       "\t ENSG00000001036 & HGNC:4008  & FUCA2   & ENSMMUG00000005366 & FUCA2 \\\\\n",
       "\t ENSG00000001084 & HGNC:4311  & GCLC    & ENSMMUG00000008684 & GCLC  \\\\\n",
       "\t ENSG00000001167 & HGNC:7804  & NFYA    & ENSMMUG00000021173 & NFYA  \\\\\n",
       "\t ENSG00000001460 & HGNC:28070 & STPG1   & ENSMMUG00000054255 & STPG1 \\\\\n",
       "\t ENSG00000001461 & HGNC:25233 & NIPAL3  & ENSMMUG00000004466 & NIPAL3\\\\\n",
       "\t ENSG00000001497 & HGNC:25726 & LAS1L   & ENSMMUG00000003922 & LAS1L \\\\\n",
       "\t ENSG00000001561 & HGNC:3359  & ENPP4   & ENSMMUG00000042260 & ENPP4 \\\\\n",
       "\t ENSG00000001617 & HGNC:10728 & SEMA3F  & ENSMMUG00000010337 & SEMA3F\\\\\n",
       "\t ENSG00000001626 & HGNC:1884  & CFTR    & ENSMMUG00000011269 & CFTR  \\\\\n",
       "\t ENSG00000001629 & HGNC:22215 & ANKIB1  & ENSMMUG00000005120 & ANKIB1\\\\\n",
       "\t ENSG00000001630 & HGNC:2649  & CYP51A1 &                    &       \\\\\n",
       "\t ENSG00000001631 & HGNC:1573  & KRIT1   & ENSMMUG00000014907 & KRIT1 \\\\\n",
       "\t ENSG00000002016 & HGNC:9824  & RAD52   & ENSMMUG00000018654 & RAD52 \\\\\n",
       "\t ENSG00000002330 & HGNC:936   & BAD     & ENSMMUG00000060596 & BAD   \\\\\n",
       "\t ENSG00000002549 & HGNC:18449 & LAP3    & ENSMMUG00000005732 & LAP3  \\\\\n",
       "\t ENSG00000002586 & HGNC:7082  & CD99    & ENSMMUG00000018687 & CD99  \\\\\n",
       "\t ENSG00000002586 & HGNC:7082  & CD99    & ENSMMUG00000055976 &       \\\\\n",
       "\t ENSG00000002587 & HGNC:5194  & HS3ST1  & ENSMMUG00000001341 & HS3ST1\\\\\n",
       "\t ENSG00000002726 & HGNC:80    & AOC1    & ENSMMUG00000020435 & AOC1  \\\\\n",
       "\t ENSG00000002745 & HGNC:16267 & WNT16   & ENSMMUG00000002439 & WNT16 \\\\\n",
       "\t ENSG00000002746 & HGNC:22195 & HECW1   & ENSMMUG00000019015 & HECW1 \\\\\n",
       "\t ENSG00000002822 & HGNC:6762  & MAD1L1  & ENSMMUG00000011659 & MAD1L1\\\\\n",
       "\t ENSG00000002834 & HGNC:6513  & LASP1   & ENSMMUG00000007799 & LASP1 \\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t ENSG00000310537 &            &           &                    &        \\\\\n",
       "\t ENSG00000310539 &            &           &                    &        \\\\\n",
       "\t ENSG00000310540 & HGNC:56008 & LINC02968 &                    &        \\\\\n",
       "\t ENSG00000310541 &            &           &                    &        \\\\\n",
       "\t ENSG00000310542 &            &           &                    &        \\\\\n",
       "\t ENSG00000310543 &            &           &                    &        \\\\\n",
       "\t ENSG00000310544 &            &           &                    &        \\\\\n",
       "\t ENSG00000310545 &            &           &                    &        \\\\\n",
       "\t ENSG00000310546 &            &           &                    &        \\\\\n",
       "\t ENSG00000310547 &            &           &                    &        \\\\\n",
       "\t ENSG00000310548 &            &           &                    &        \\\\\n",
       "\t ENSG00000310549 &            &           &                    &        \\\\\n",
       "\t ENSG00000310550 &            &           &                    &        \\\\\n",
       "\t ENSG00000310551 &            &           &                    &        \\\\\n",
       "\t ENSG00000310552 &            &           &                    &        \\\\\n",
       "\t ENSG00000310553 &            &           &                    &        \\\\\n",
       "\t ENSG00000310554 &            &           &                    &        \\\\\n",
       "\t ENSG00000310555 &            &           &                    &        \\\\\n",
       "\t ENSG00000310556 &            &           &                    &        \\\\\n",
       "\t ENSG00000310557 &            &           &                    &        \\\\\n",
       "\t ENSG00000310558 &            &           &                    &        \\\\\n",
       "\t ENSG00000310559 &            &           &                    &        \\\\\n",
       "\t ENSG00000310560 & HGNC:27849 & PAXX      & ENSMMUG00000021432 & PAXX   \\\\\n",
       "\t ENSG00000310561 &            &           &                    &        \\\\\n",
       "\t ENSG00000310562 &            &           &                    &        \\\\\n",
       "\t ENSG00000310564 &            &           &                    &        \\\\\n",
       "\t ENSG00000310566 &            &           &                    &        \\\\\n",
       "\t ENSG00000310567 &            &           &                    &        \\\\\n",
       "\t ENSG00000310576 & HGNC:26073 & FAM174C   & ENSMMUG00000022104 & FAM174C\\\\\n",
       "\t ENSG00000310577 &            &           &                    &        \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 87850 × 5\n",
       "\n",
       "| ensembl_gene_id &lt;chr&gt; | hgnc_id &lt;chr&gt; | hgnc_symbol &lt;chr&gt; | mmulatta_homolog_ensembl_gene &lt;chr&gt; | mmulatta_homolog_associated_gene_name &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| ENSG00000000003 | HGNC:11858 | TSPAN6  | <!----> | <!----> |\n",
       "| ENSG00000000005 | HGNC:17757 | TNMD    | ENSMMUG00000008212 | TNMD   |\n",
       "| ENSG00000000419 | HGNC:3005  | DPM1    | ENSMMUG00000002759 | DPM1   |\n",
       "| ENSG00000000457 | HGNC:19285 | SCYL3   | ENSMMUG00000016583 | SCYL3  |\n",
       "| ENSG00000000460 | HGNC:25565 | FIRRM   | ENSMMUG00000016582 | FIRRM  |\n",
       "| ENSG00000000938 | HGNC:3697  | FGR     | ENSMMUG00000014434 | FGR    |\n",
       "| ENSG00000000971 | HGNC:4883  | CFH     | ENSMMUG00000045497 | CFH    |\n",
       "| ENSG00000001036 | HGNC:4008  | FUCA2   | ENSMMUG00000005366 | FUCA2  |\n",
       "| ENSG00000001084 | HGNC:4311  | GCLC    | ENSMMUG00000008684 | GCLC   |\n",
       "| ENSG00000001167 | HGNC:7804  | NFYA    | ENSMMUG00000021173 | NFYA   |\n",
       "| ENSG00000001460 | HGNC:28070 | STPG1   | ENSMMUG00000054255 | STPG1  |\n",
       "| ENSG00000001461 | HGNC:25233 | NIPAL3  | ENSMMUG00000004466 | NIPAL3 |\n",
       "| ENSG00000001497 | HGNC:25726 | LAS1L   | ENSMMUG00000003922 | LAS1L  |\n",
       "| ENSG00000001561 | HGNC:3359  | ENPP4   | ENSMMUG00000042260 | ENPP4  |\n",
       "| ENSG00000001617 | HGNC:10728 | SEMA3F  | ENSMMUG00000010337 | SEMA3F |\n",
       "| ENSG00000001626 | HGNC:1884  | CFTR    | ENSMMUG00000011269 | CFTR   |\n",
       "| ENSG00000001629 | HGNC:22215 | ANKIB1  | ENSMMUG00000005120 | ANKIB1 |\n",
       "| ENSG00000001630 | HGNC:2649  | CYP51A1 | <!----> | <!----> |\n",
       "| ENSG00000001631 | HGNC:1573  | KRIT1   | ENSMMUG00000014907 | KRIT1  |\n",
       "| ENSG00000002016 | HGNC:9824  | RAD52   | ENSMMUG00000018654 | RAD52  |\n",
       "| ENSG00000002330 | HGNC:936   | BAD     | ENSMMUG00000060596 | BAD    |\n",
       "| ENSG00000002549 | HGNC:18449 | LAP3    | ENSMMUG00000005732 | LAP3   |\n",
       "| ENSG00000002586 | HGNC:7082  | CD99    | ENSMMUG00000018687 | CD99   |\n",
       "| ENSG00000002586 | HGNC:7082  | CD99    | ENSMMUG00000055976 | <!----> |\n",
       "| ENSG00000002587 | HGNC:5194  | HS3ST1  | ENSMMUG00000001341 | HS3ST1 |\n",
       "| ENSG00000002726 | HGNC:80    | AOC1    | ENSMMUG00000020435 | AOC1   |\n",
       "| ENSG00000002745 | HGNC:16267 | WNT16   | ENSMMUG00000002439 | WNT16  |\n",
       "| ENSG00000002746 | HGNC:22195 | HECW1   | ENSMMUG00000019015 | HECW1  |\n",
       "| ENSG00000002822 | HGNC:6762  | MAD1L1  | ENSMMUG00000011659 | MAD1L1 |\n",
       "| ENSG00000002834 | HGNC:6513  | LASP1   | ENSMMUG00000007799 | LASP1  |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| ENSG00000310537 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310539 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310540 | HGNC:56008 | LINC02968 | <!----> | <!----> |\n",
       "| ENSG00000310541 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310542 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310543 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310544 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310545 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310546 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310547 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310548 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310549 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310550 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310551 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310552 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310553 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310554 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310555 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310556 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310557 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310558 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310559 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310560 | HGNC:27849 | PAXX      | ENSMMUG00000021432 | PAXX    |\n",
       "| ENSG00000310561 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310562 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310564 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310566 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310567 | <!----> | <!----> | <!----> | <!----> |\n",
       "| ENSG00000310576 | HGNC:26073 | FAM174C   | ENSMMUG00000022104 | FAM174C |\n",
       "| ENSG00000310577 | <!----> | <!----> | <!----> | <!----> |\n",
       "\n"
      ],
      "text/plain": [
       "      ensembl_gene_id hgnc_id    hgnc_symbol mmulatta_homolog_ensembl_gene\n",
       "1     ENSG00000000003 HGNC:11858 TSPAN6                                   \n",
       "2     ENSG00000000005 HGNC:17757 TNMD        ENSMMUG00000008212           \n",
       "3     ENSG00000000419 HGNC:3005  DPM1        ENSMMUG00000002759           \n",
       "4     ENSG00000000457 HGNC:19285 SCYL3       ENSMMUG00000016583           \n",
       "5     ENSG00000000460 HGNC:25565 FIRRM       ENSMMUG00000016582           \n",
       "6     ENSG00000000938 HGNC:3697  FGR         ENSMMUG00000014434           \n",
       "7     ENSG00000000971 HGNC:4883  CFH         ENSMMUG00000045497           \n",
       "8     ENSG00000001036 HGNC:4008  FUCA2       ENSMMUG00000005366           \n",
       "9     ENSG00000001084 HGNC:4311  GCLC        ENSMMUG00000008684           \n",
       "10    ENSG00000001167 HGNC:7804  NFYA        ENSMMUG00000021173           \n",
       "11    ENSG00000001460 HGNC:28070 STPG1       ENSMMUG00000054255           \n",
       "12    ENSG00000001461 HGNC:25233 NIPAL3      ENSMMUG00000004466           \n",
       "13    ENSG00000001497 HGNC:25726 LAS1L       ENSMMUG00000003922           \n",
       "14    ENSG00000001561 HGNC:3359  ENPP4       ENSMMUG00000042260           \n",
       "15    ENSG00000001617 HGNC:10728 SEMA3F      ENSMMUG00000010337           \n",
       "16    ENSG00000001626 HGNC:1884  CFTR        ENSMMUG00000011269           \n",
       "17    ENSG00000001629 HGNC:22215 ANKIB1      ENSMMUG00000005120           \n",
       "18    ENSG00000001630 HGNC:2649  CYP51A1                                  \n",
       "19    ENSG00000001631 HGNC:1573  KRIT1       ENSMMUG00000014907           \n",
       "20    ENSG00000002016 HGNC:9824  RAD52       ENSMMUG00000018654           \n",
       "21    ENSG00000002330 HGNC:936   BAD         ENSMMUG00000060596           \n",
       "22    ENSG00000002549 HGNC:18449 LAP3        ENSMMUG00000005732           \n",
       "23    ENSG00000002586 HGNC:7082  CD99        ENSMMUG00000018687           \n",
       "24    ENSG00000002586 HGNC:7082  CD99        ENSMMUG00000055976           \n",
       "25    ENSG00000002587 HGNC:5194  HS3ST1      ENSMMUG00000001341           \n",
       "26    ENSG00000002726 HGNC:80    AOC1        ENSMMUG00000020435           \n",
       "27    ENSG00000002745 HGNC:16267 WNT16       ENSMMUG00000002439           \n",
       "28    ENSG00000002746 HGNC:22195 HECW1       ENSMMUG00000019015           \n",
       "29    ENSG00000002822 HGNC:6762  MAD1L1      ENSMMUG00000011659           \n",
       "30    ENSG00000002834 HGNC:6513  LASP1       ENSMMUG00000007799           \n",
       "⋮     ⋮               ⋮          ⋮           ⋮                            \n",
       "87821 ENSG00000310537                                                     \n",
       "87822 ENSG00000310539                                                     \n",
       "87823 ENSG00000310540 HGNC:56008 LINC02968                                \n",
       "87824 ENSG00000310541                                                     \n",
       "87825 ENSG00000310542                                                     \n",
       "87826 ENSG00000310543                                                     \n",
       "87827 ENSG00000310544                                                     \n",
       "87828 ENSG00000310545                                                     \n",
       "87829 ENSG00000310546                                                     \n",
       "87830 ENSG00000310547                                                     \n",
       "87831 ENSG00000310548                                                     \n",
       "87832 ENSG00000310549                                                     \n",
       "87833 ENSG00000310550                                                     \n",
       "87834 ENSG00000310551                                                     \n",
       "87835 ENSG00000310552                                                     \n",
       "87836 ENSG00000310553                                                     \n",
       "87837 ENSG00000310554                                                     \n",
       "87838 ENSG00000310555                                                     \n",
       "87839 ENSG00000310556                                                     \n",
       "87840 ENSG00000310557                                                     \n",
       "87841 ENSG00000310558                                                     \n",
       "87842 ENSG00000310559                                                     \n",
       "87843 ENSG00000310560 HGNC:27849 PAXX        ENSMMUG00000021432           \n",
       "87844 ENSG00000310561                                                     \n",
       "87845 ENSG00000310562                                                     \n",
       "87846 ENSG00000310564                                                     \n",
       "87847 ENSG00000310566                                                     \n",
       "87848 ENSG00000310567                                                     \n",
       "87849 ENSG00000310576 HGNC:26073 FAM174C     ENSMMUG00000022104           \n",
       "87850 ENSG00000310577                                                     \n",
       "      mmulatta_homolog_associated_gene_name\n",
       "1                                          \n",
       "2     TNMD                                 \n",
       "3     DPM1                                 \n",
       "4     SCYL3                                \n",
       "5     FIRRM                                \n",
       "6     FGR                                  \n",
       "7     CFH                                  \n",
       "8     FUCA2                                \n",
       "9     GCLC                                 \n",
       "10    NFYA                                 \n",
       "11    STPG1                                \n",
       "12    NIPAL3                               \n",
       "13    LAS1L                                \n",
       "14    ENPP4                                \n",
       "15    SEMA3F                               \n",
       "16    CFTR                                 \n",
       "17    ANKIB1                               \n",
       "18                                         \n",
       "19    KRIT1                                \n",
       "20    RAD52                                \n",
       "21    BAD                                  \n",
       "22    LAP3                                 \n",
       "23    CD99                                 \n",
       "24                                         \n",
       "25    HS3ST1                               \n",
       "26    AOC1                                 \n",
       "27    WNT16                                \n",
       "28    HECW1                                \n",
       "29    MAD1L1                               \n",
       "30    LASP1                                \n",
       "⋮     ⋮                                    \n",
       "87821                                      \n",
       "87822                                      \n",
       "87823                                      \n",
       "87824                                      \n",
       "87825                                      \n",
       "87826                                      \n",
       "87827                                      \n",
       "87828                                      \n",
       "87829                                      \n",
       "87830                                      \n",
       "87831                                      \n",
       "87832                                      \n",
       "87833                                      \n",
       "87834                                      \n",
       "87835                                      \n",
       "87836                                      \n",
       "87837                                      \n",
       "87838                                      \n",
       "87839                                      \n",
       "87840                                      \n",
       "87841                                      \n",
       "87842                                      \n",
       "87843 PAXX                                 \n",
       "87844                                      \n",
       "87845                                      \n",
       "87846                                      \n",
       "87847                                      \n",
       "87848                                      \n",
       "87849 FAM174C                              \n",
       "87850                                      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#  Rhesus macaque\n",
    "get_species_orthologs(\"mmulatta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d47d6e6-b104-4345-8f44-f5dd0ff6ab43",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ccd7dcc-1f8f-428c-9a39-de919ea5df1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Chimp\n",
    "get_species_orthologs(\"ptroglodytes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "343a679c-7a5d-4701-8492-cf1389b98a66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Chicken (Gallus gallus)\n",
    "chicken_orthologs <- get_species_orthologs(\"ggallus\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da977b53-d8a2-43c7-8c3b-490b78338879",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pig (Sus scrofa)\n",
    "pig_orthologs <- get_species_orthologs(\"sscrofa\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ee45d8c-a1b2-4ff1-a6ec-e85644344cc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cow (Bos taurus)\n",
    "cow_orthologs <- get_species_orthologs(\"btaurus\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "844dab5f-efa4-431d-b4d7-c0a4e7e73129",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dog (Canis lupus familiaris)\n",
    "dog_orthologs <- get_species_orthologs(\"clfamiliaris\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df5e988b-bc26-4197-80ad-5f192676979a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Horse (Equus caballus)\n",
    "horse_orthologs <- get_species_orthologs(\"ecaballus\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bc689e3-1d42-4854-800e-6f3599d6dab0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sheep (Ovis aries rambouillet)\n",
    "sheep_orthologs <- get_species_orthologs(\"oarambouillet\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbd8a107-1754-48b5-9ccb-91b0402661a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mouse (Mus musculus) # test\n",
    "mouse_orthologs <- get_species_orthologs(\"mmusculus\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24993537-5679-4488-aed4-bb39dc2f8ff2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rat (Rattus norvegicus) # test\n",
    "rat_orthologs <- get_species_orthologs(\"rnorvegicus\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35ba4f14-904c-4a51-9786-28283a8e5a81",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Zebrafish (Danio rerio) # test\n",
    "zebra_orthologs <- get_species_orthologs(\"drerio\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c019b291-7814-4b04-8bed-e4c1cfe8ba9a",
   "metadata": {},
   "source": [
    "# Check homologs available"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61af6dd1-d3c0-433d-8dd6-f43597514ca5",
   "metadata": {},
   "outputs": [],
   "source": [
    "ensembl_human <- useMart(\"ensembl\", dataset =\"hsapiens_gene_ensembl\")\n",
    "attributes_human <- listAttributes(ensembl_human)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18e81d43-180d-43c5-90b0-b847c2193313",
   "metadata": {},
   "outputs": [],
   "source": [
    "homologs_available <- attributes_human[grep(\"homolog_ensembl_gene\", attributes_human$name), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ab9c1ca-a2fc-41d3-b491-556d13b99bfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "readr::write_csv(homologs_available, paste0(\"data/\",\"human_homologs\", \"_biomart.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "eabcf439-9a7e-45cf-b8e0-6540a1043bb8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>name</th><th scope=col>description</th><th scope=col>page</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ensembl_gene_id              </td><td>Gene stable ID              </td><td>feature_page</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ensembl_gene_id_version      </td><td>Gene stable ID version      </td><td>feature_page</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>ensembl_transcript_id        </td><td>Transcript stable ID        </td><td>feature_page</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>ensembl_transcript_id_version</td><td>Transcript stable ID version</td><td>feature_page</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>ensembl_peptide_id           </td><td>Protein stable ID           </td><td>feature_page</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>ensembl_peptide_id_version   </td><td>Protein stable ID version   </td><td>feature_page</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & name & description & page\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & ensembl\\_gene\\_id               & Gene stable ID               & feature\\_page\\\\\n",
       "\t2 & ensembl\\_gene\\_id\\_version       & Gene stable ID version       & feature\\_page\\\\\n",
       "\t3 & ensembl\\_transcript\\_id         & Transcript stable ID         & feature\\_page\\\\\n",
       "\t4 & ensembl\\_transcript\\_id\\_version & Transcript stable ID version & feature\\_page\\\\\n",
       "\t5 & ensembl\\_peptide\\_id            & Protein stable ID            & feature\\_page\\\\\n",
       "\t6 & ensembl\\_peptide\\_id\\_version    & Protein stable ID version    & feature\\_page\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 3\n",
       "\n",
       "| <!--/--> | name &lt;chr&gt; | description &lt;chr&gt; | page &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | ensembl_gene_id               | Gene stable ID               | feature_page |\n",
       "| 2 | ensembl_gene_id_version       | Gene stable ID version       | feature_page |\n",
       "| 3 | ensembl_transcript_id         | Transcript stable ID         | feature_page |\n",
       "| 4 | ensembl_transcript_id_version | Transcript stable ID version | feature_page |\n",
       "| 5 | ensembl_peptide_id            | Protein stable ID            | feature_page |\n",
       "| 6 | ensembl_peptide_id_version    | Protein stable ID version    | feature_page |\n",
       "\n"
      ],
      "text/plain": [
       "  name                          description                  page        \n",
       "1 ensembl_gene_id               Gene stable ID               feature_page\n",
       "2 ensembl_gene_id_version       Gene stable ID version       feature_page\n",
       "3 ensembl_transcript_id         Transcript stable ID         feature_page\n",
       "4 ensembl_transcript_id_version Transcript stable ID version feature_page\n",
       "5 ensembl_peptide_id            Protein stable ID            feature_page\n",
       "6 ensembl_peptide_id_version    Protein stable ID version    feature_page"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "human_attributes <- list_species_attributes(\"hsapiens_gene_ensembl\")\n",
    "head(human_attributes)  # Show first few attributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "af952140-94ca-4f05-9121-adb28507b946",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 54</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>hgnc_id</th><th scope=col>symbol</th><th scope=col>name</th><th scope=col>locus_group</th><th scope=col>locus_type</th><th scope=col>status</th><th scope=col>location</th><th scope=col>location_sortable</th><th scope=col>alias_symbol</th><th scope=col>alias_name</th><th scope=col>⋯</th><th scope=col>cd</th><th scope=col>lncrnadb</th><th scope=col>enzyme_id</th><th scope=col>intermediate_filament_db</th><th scope=col>rna_central_id</th><th scope=col>lncipedia</th><th scope=col>gtrnadb</th><th scope=col>agr</th><th scope=col>mane_select</th><th scope=col>gencc</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>HGNC:5    </td><td>A1BG    </td><td>alpha-1-B glycoprotein        </td><td>protein-coding gene</td><td>gene with protein product</td><td>Approved</td><td>19q13.43</td><td>19q13.43</td><td>                             </td><td></td><td>⋯</td><td></td><td></td><td></td><td>NA</td><td>             </td><td>        </td><td></td><td>HGNC:5    </td><td>ENST00000263100.8|NM_130786.4 </td><td>          </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>HGNC:37133</td><td>A1BG-AS1</td><td>A1BG antisense RNA 1          </td><td>non-coding RNA     </td><td>RNA, long non-coding     </td><td>Approved</td><td>19q13.43</td><td>19q13.43</td><td>FLJ23569                     </td><td></td><td>⋯</td><td></td><td></td><td></td><td>NA</td><td>URS00007E4F6E</td><td>A1BG-AS1</td><td></td><td>HGNC:37133</td><td>                              </td><td>          </td></tr>\n",
       "\t<tr><th scope=row>3</th><td>HGNC:24086</td><td>A1CF    </td><td>APOBEC1 complementation factor</td><td>protein-coding gene</td><td>gene with protein product</td><td>Approved</td><td>10q11.23</td><td>10q11.23</td><td>ACF|ASP|ACF64|ACF65|APOBEC1CF</td><td></td><td>⋯</td><td></td><td></td><td></td><td>NA</td><td>             </td><td>        </td><td></td><td>HGNC:24086</td><td>ENST00000373997.8|NM_014576.4 </td><td>          </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>HGNC:7    </td><td>A2M     </td><td>alpha-2-macroglobulin         </td><td>protein-coding gene</td><td>gene with protein product</td><td>Approved</td><td>12p13.31</td><td>12p13.31</td><td>FWP007|S863-7|CPAMD5         </td><td></td><td>⋯</td><td></td><td></td><td></td><td>NA</td><td>             </td><td>        </td><td></td><td>HGNC:7    </td><td>ENST00000318602.12|NM_000014.6</td><td>HGNC:7    </td></tr>\n",
       "\t<tr><th scope=row>5</th><td>HGNC:27057</td><td>A2M-AS1 </td><td>A2M antisense RNA 1           </td><td>non-coding RNA     </td><td>RNA, long non-coding     </td><td>Approved</td><td>12p13.31</td><td>12p13.31</td><td>                             </td><td></td><td>⋯</td><td></td><td></td><td></td><td>NA</td><td>URS00001F234A</td><td>A2M-AS1 </td><td></td><td>HGNC:27057</td><td>                              </td><td>          </td></tr>\n",
       "\t<tr><th scope=row>6</th><td>HGNC:23336</td><td>A2ML1   </td><td>alpha-2-macroglobulin like 1  </td><td>protein-coding gene</td><td>gene with protein product</td><td>Approved</td><td>12p13.31</td><td>12p13.31</td><td>FLJ25179|p170                </td><td></td><td>⋯</td><td></td><td></td><td></td><td>NA</td><td>             </td><td>        </td><td></td><td>HGNC:23336</td><td>ENST00000299698.12|NM_144670.6</td><td>HGNC:23336</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 54\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & hgnc\\_id & symbol & name & locus\\_group & locus\\_type & status & location & location\\_sortable & alias\\_symbol & alias\\_name & ⋯ & cd & lncrnadb & enzyme\\_id & intermediate\\_filament\\_db & rna\\_central\\_id & lncipedia & gtrnadb & agr & mane\\_select & gencc\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & ⋯ & <chr> & <chr> & <chr> & <lgl> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & HGNC:5     & A1BG     & alpha-1-B glycoprotein         & protein-coding gene & gene with protein product & Approved & 19q13.43 & 19q13.43 &                               &  & ⋯ &  &  &  & NA &               &          &  & HGNC:5     & ENST00000263100.8\\textbar{}NM\\_130786.4  &           \\\\\n",
       "\t2 & HGNC:37133 & A1BG-AS1 & A1BG antisense RNA 1           & non-coding RNA      & RNA, long non-coding      & Approved & 19q13.43 & 19q13.43 & FLJ23569                      &  & ⋯ &  &  &  & NA & URS00007E4F6E & A1BG-AS1 &  & HGNC:37133 &                                &           \\\\\n",
       "\t3 & HGNC:24086 & A1CF     & APOBEC1 complementation factor & protein-coding gene & gene with protein product & Approved & 10q11.23 & 10q11.23 & ACF\\textbar{}ASP\\textbar{}ACF64\\textbar{}ACF65\\textbar{}APOBEC1CF &  & ⋯ &  &  &  & NA &               &          &  & HGNC:24086 & ENST00000373997.8\\textbar{}NM\\_014576.4  &           \\\\\n",
       "\t4 & HGNC:7     & A2M      & alpha-2-macroglobulin          & protein-coding gene & gene with protein product & Approved & 12p13.31 & 12p13.31 & FWP007\\textbar{}S863-7\\textbar{}CPAMD5          &  & ⋯ &  &  &  & NA &               &          &  & HGNC:7     & ENST00000318602.12\\textbar{}NM\\_000014.6 & HGNC:7    \\\\\n",
       "\t5 & HGNC:27057 & A2M-AS1  & A2M antisense RNA 1            & non-coding RNA      & RNA, long non-coding      & Approved & 12p13.31 & 12p13.31 &                               &  & ⋯ &  &  &  & NA & URS00001F234A & A2M-AS1  &  & HGNC:27057 &                                &           \\\\\n",
       "\t6 & HGNC:23336 & A2ML1    & alpha-2-macroglobulin like 1   & protein-coding gene & gene with protein product & Approved & 12p13.31 & 12p13.31 & FLJ25179\\textbar{}p170                 &  & ⋯ &  &  &  & NA &               &          &  & HGNC:23336 & ENST00000299698.12\\textbar{}NM\\_144670.6 & HGNC:23336\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 54\n",
       "\n",
       "| <!--/--> | hgnc_id &lt;chr&gt; | symbol &lt;chr&gt; | name &lt;chr&gt; | locus_group &lt;chr&gt; | locus_type &lt;chr&gt; | status &lt;chr&gt; | location &lt;chr&gt; | location_sortable &lt;chr&gt; | alias_symbol &lt;chr&gt; | alias_name &lt;chr&gt; | ⋯ ⋯ | cd &lt;chr&gt; | lncrnadb &lt;chr&gt; | enzyme_id &lt;chr&gt; | intermediate_filament_db &lt;lgl&gt; | rna_central_id &lt;chr&gt; | lncipedia &lt;chr&gt; | gtrnadb &lt;chr&gt; | agr &lt;chr&gt; | mane_select &lt;chr&gt; | gencc &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | HGNC:5     | A1BG     | alpha-1-B glycoprotein         | protein-coding gene | gene with protein product | Approved | 19q13.43 | 19q13.43 | <!----> | <!----> | ⋯ | <!----> | <!----> | <!----> | NA | <!----> | <!----> | <!----> | HGNC:5     | ENST00000263100.8|NM_130786.4  | <!----> |\n",
       "| 2 | HGNC:37133 | A1BG-AS1 | A1BG antisense RNA 1           | non-coding RNA      | RNA, long non-coding      | Approved | 19q13.43 | 19q13.43 | FLJ23569                      | <!----> | ⋯ | <!----> | <!----> | <!----> | NA | URS00007E4F6E | A1BG-AS1 | <!----> | HGNC:37133 | <!----> | <!----> |\n",
       "| 3 | HGNC:24086 | A1CF     | APOBEC1 complementation factor | protein-coding gene | gene with protein product | Approved | 10q11.23 | 10q11.23 | ACF|ASP|ACF64|ACF65|APOBEC1CF | <!----> | ⋯ | <!----> | <!----> | <!----> | NA | <!----> | <!----> | <!----> | HGNC:24086 | ENST00000373997.8|NM_014576.4  | <!----> |\n",
       "| 4 | HGNC:7     | A2M      | alpha-2-macroglobulin          | protein-coding gene | gene with protein product | Approved | 12p13.31 | 12p13.31 | FWP007|S863-7|CPAMD5          | <!----> | ⋯ | <!----> | <!----> | <!----> | NA | <!----> | <!----> | <!----> | HGNC:7     | ENST00000318602.12|NM_000014.6 | HGNC:7     |\n",
       "| 5 | HGNC:27057 | A2M-AS1  | A2M antisense RNA 1            | non-coding RNA      | RNA, long non-coding      | Approved | 12p13.31 | 12p13.31 | <!----> | <!----> | ⋯ | <!----> | <!----> | <!----> | NA | URS00001F234A | A2M-AS1  | <!----> | HGNC:27057 | <!----> | <!----> |\n",
       "| 6 | HGNC:23336 | A2ML1    | alpha-2-macroglobulin like 1   | protein-coding gene | gene with protein product | Approved | 12p13.31 | 12p13.31 | FLJ25179|p170                 | <!----> | ⋯ | <!----> | <!----> | <!----> | NA | <!----> | <!----> | <!----> | HGNC:23336 | ENST00000299698.12|NM_144670.6 | HGNC:23336 |\n",
       "\n"
      ],
      "text/plain": [
       "  hgnc_id    symbol   name                           locus_group        \n",
       "1 HGNC:5     A1BG     alpha-1-B glycoprotein         protein-coding gene\n",
       "2 HGNC:37133 A1BG-AS1 A1BG antisense RNA 1           non-coding RNA     \n",
       "3 HGNC:24086 A1CF     APOBEC1 complementation factor protein-coding gene\n",
       "4 HGNC:7     A2M      alpha-2-macroglobulin          protein-coding gene\n",
       "5 HGNC:27057 A2M-AS1  A2M antisense RNA 1            non-coding RNA     \n",
       "6 HGNC:23336 A2ML1    alpha-2-macroglobulin like 1   protein-coding gene\n",
       "  locus_type                status   location location_sortable\n",
       "1 gene with protein product Approved 19q13.43 19q13.43         \n",
       "2 RNA, long non-coding      Approved 19q13.43 19q13.43         \n",
       "3 gene with protein product Approved 10q11.23 10q11.23         \n",
       "4 gene with protein product Approved 12p13.31 12p13.31         \n",
       "5 RNA, long non-coding      Approved 12p13.31 12p13.31         \n",
       "6 gene with protein product Approved 12p13.31 12p13.31         \n",
       "  alias_symbol                  alias_name ⋯ cd lncrnadb enzyme_id\n",
       "1                                          ⋯                      \n",
       "2 FLJ23569                                 ⋯                      \n",
       "3 ACF|ASP|ACF64|ACF65|APOBEC1CF            ⋯                      \n",
       "4 FWP007|S863-7|CPAMD5                     ⋯                      \n",
       "5                                          ⋯                      \n",
       "6 FLJ25179|p170                            ⋯                      \n",
       "  intermediate_filament_db rna_central_id lncipedia gtrnadb agr       \n",
       "1 NA                                                        HGNC:5    \n",
       "2 NA                       URS00007E4F6E  A1BG-AS1          HGNC:37133\n",
       "3 NA                                                        HGNC:24086\n",
       "4 NA                                                        HGNC:7    \n",
       "5 NA                       URS00001F234A  A2M-AS1           HGNC:27057\n",
       "6 NA                                                        HGNC:23336\n",
       "  mane_select                    gencc     \n",
       "1 ENST00000263100.8|NM_130786.4            \n",
       "2                                          \n",
       "3 ENST00000373997.8|NM_014576.4            \n",
       "4 ENST00000318602.12|NM_000014.6 HGNC:7    \n",
       "5                                          \n",
       "6 ENST00000299698.12|NM_144670.6 HGNC:23336"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# download HGNC complete txt file\n",
    "download.file(\n",
    "  url = \"https://storage.googleapis.com/public-download-files/hgnc/tsv/tsv/hgnc_complete_set.txt\",\n",
    "  destfile = \"hgnc_complete_set.txt\",\n",
    "  method = \"auto\"\n",
    ")\n",
    "\n",
    "# Optional: read into R\n",
    "hgnc_data <- read.delim(\"hgnc_complete_set.txt\", sep = \"\\t\", header = TRUE)\n",
    "head(hgnc_data)\n",
    "# Save to file\n",
    "write.table(hgnc_data, file = \"data/HGNC_gene_info_full.tsv\", sep = \"\\t\", row.names = FALSE, quote = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fd7e16a5-09c8-4e8b-857b-08c695cd9d2b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'hgnc_id'</li><li>'symbol'</li><li>'name'</li><li>'locus_group'</li><li>'locus_type'</li><li>'status'</li><li>'location'</li><li>'location_sortable'</li><li>'alias_symbol'</li><li>'alias_name'</li><li>'prev_symbol'</li><li>'prev_name'</li><li>'gene_group'</li><li>'gene_group_id'</li><li>'date_approved_reserved'</li><li>'date_symbol_changed'</li><li>'date_name_changed'</li><li>'date_modified'</li><li>'entrez_id'</li><li>'ensembl_gene_id'</li><li>'vega_id'</li><li>'ucsc_id'</li><li>'ena'</li><li>'refseq_accession'</li><li>'ccds_id'</li><li>'uniprot_ids'</li><li>'pubmed_id'</li><li>'mgd_id'</li><li>'rgd_id'</li><li>'lsdb'</li><li>'cosmic'</li><li>'omim_id'</li><li>'mirbase'</li><li>'homeodb'</li><li>'snornabase'</li><li>'bioparadigms_slc'</li><li>'orphanet'</li><li>'pseudogene.org'</li><li>'horde_id'</li><li>'merops'</li><li>'imgt'</li><li>'iuphar'</li><li>'kznf_gene_catalog'</li><li>'mamit.trnadb'</li><li>'cd'</li><li>'lncrnadb'</li><li>'enzyme_id'</li><li>'intermediate_filament_db'</li><li>'rna_central_id'</li><li>'lncipedia'</li><li>'gtrnadb'</li><li>'agr'</li><li>'mane_select'</li><li>'gencc'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'hgnc\\_id'\n",
       "\\item 'symbol'\n",
       "\\item 'name'\n",
       "\\item 'locus\\_group'\n",
       "\\item 'locus\\_type'\n",
       "\\item 'status'\n",
       "\\item 'location'\n",
       "\\item 'location\\_sortable'\n",
       "\\item 'alias\\_symbol'\n",
       "\\item 'alias\\_name'\n",
       "\\item 'prev\\_symbol'\n",
       "\\item 'prev\\_name'\n",
       "\\item 'gene\\_group'\n",
       "\\item 'gene\\_group\\_id'\n",
       "\\item 'date\\_approved\\_reserved'\n",
       "\\item 'date\\_symbol\\_changed'\n",
       "\\item 'date\\_name\\_changed'\n",
       "\\item 'date\\_modified'\n",
       "\\item 'entrez\\_id'\n",
       "\\item 'ensembl\\_gene\\_id'\n",
       "\\item 'vega\\_id'\n",
       "\\item 'ucsc\\_id'\n",
       "\\item 'ena'\n",
       "\\item 'refseq\\_accession'\n",
       "\\item 'ccds\\_id'\n",
       "\\item 'uniprot\\_ids'\n",
       "\\item 'pubmed\\_id'\n",
       "\\item 'mgd\\_id'\n",
       "\\item 'rgd\\_id'\n",
       "\\item 'lsdb'\n",
       "\\item 'cosmic'\n",
       "\\item 'omim\\_id'\n",
       "\\item 'mirbase'\n",
       "\\item 'homeodb'\n",
       "\\item 'snornabase'\n",
       "\\item 'bioparadigms\\_slc'\n",
       "\\item 'orphanet'\n",
       "\\item 'pseudogene.org'\n",
       "\\item 'horde\\_id'\n",
       "\\item 'merops'\n",
       "\\item 'imgt'\n",
       "\\item 'iuphar'\n",
       "\\item 'kznf\\_gene\\_catalog'\n",
       "\\item 'mamit.trnadb'\n",
       "\\item 'cd'\n",
       "\\item 'lncrnadb'\n",
       "\\item 'enzyme\\_id'\n",
       "\\item 'intermediate\\_filament\\_db'\n",
       "\\item 'rna\\_central\\_id'\n",
       "\\item 'lncipedia'\n",
       "\\item 'gtrnadb'\n",
       "\\item 'agr'\n",
       "\\item 'mane\\_select'\n",
       "\\item 'gencc'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'hgnc_id'\n",
       "2. 'symbol'\n",
       "3. 'name'\n",
       "4. 'locus_group'\n",
       "5. 'locus_type'\n",
       "6. 'status'\n",
       "7. 'location'\n",
       "8. 'location_sortable'\n",
       "9. 'alias_symbol'\n",
       "10. 'alias_name'\n",
       "11. 'prev_symbol'\n",
       "12. 'prev_name'\n",
       "13. 'gene_group'\n",
       "14. 'gene_group_id'\n",
       "15. 'date_approved_reserved'\n",
       "16. 'date_symbol_changed'\n",
       "17. 'date_name_changed'\n",
       "18. 'date_modified'\n",
       "19. 'entrez_id'\n",
       "20. 'ensembl_gene_id'\n",
       "21. 'vega_id'\n",
       "22. 'ucsc_id'\n",
       "23. 'ena'\n",
       "24. 'refseq_accession'\n",
       "25. 'ccds_id'\n",
       "26. 'uniprot_ids'\n",
       "27. 'pubmed_id'\n",
       "28. 'mgd_id'\n",
       "29. 'rgd_id'\n",
       "30. 'lsdb'\n",
       "31. 'cosmic'\n",
       "32. 'omim_id'\n",
       "33. 'mirbase'\n",
       "34. 'homeodb'\n",
       "35. 'snornabase'\n",
       "36. 'bioparadigms_slc'\n",
       "37. 'orphanet'\n",
       "38. 'pseudogene.org'\n",
       "39. 'horde_id'\n",
       "40. 'merops'\n",
       "41. 'imgt'\n",
       "42. 'iuphar'\n",
       "43. 'kznf_gene_catalog'\n",
       "44. 'mamit.trnadb'\n",
       "45. 'cd'\n",
       "46. 'lncrnadb'\n",
       "47. 'enzyme_id'\n",
       "48. 'intermediate_filament_db'\n",
       "49. 'rna_central_id'\n",
       "50. 'lncipedia'\n",
       "51. 'gtrnadb'\n",
       "52. 'agr'\n",
       "53. 'mane_select'\n",
       "54. 'gencc'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"hgnc_id\"                  \"symbol\"                  \n",
       " [3] \"name\"                     \"locus_group\"             \n",
       " [5] \"locus_type\"               \"status\"                  \n",
       " [7] \"location\"                 \"location_sortable\"       \n",
       " [9] \"alias_symbol\"             \"alias_name\"              \n",
       "[11] \"prev_symbol\"              \"prev_name\"               \n",
       "[13] \"gene_group\"               \"gene_group_id\"           \n",
       "[15] \"date_approved_reserved\"   \"date_symbol_changed\"     \n",
       "[17] \"date_name_changed\"        \"date_modified\"           \n",
       "[19] \"entrez_id\"                \"ensembl_gene_id\"         \n",
       "[21] \"vega_id\"                  \"ucsc_id\"                 \n",
       "[23] \"ena\"                      \"refseq_accession\"        \n",
       "[25] \"ccds_id\"                  \"uniprot_ids\"             \n",
       "[27] \"pubmed_id\"                \"mgd_id\"                  \n",
       "[29] \"rgd_id\"                   \"lsdb\"                    \n",
       "[31] \"cosmic\"                   \"omim_id\"                 \n",
       "[33] \"mirbase\"                  \"homeodb\"                 \n",
       "[35] \"snornabase\"               \"bioparadigms_slc\"        \n",
       "[37] \"orphanet\"                 \"pseudogene.org\"          \n",
       "[39] \"horde_id\"                 \"merops\"                  \n",
       "[41] \"imgt\"                     \"iuphar\"                  \n",
       "[43] \"kznf_gene_catalog\"        \"mamit.trnadb\"            \n",
       "[45] \"cd\"                       \"lncrnadb\"                \n",
       "[47] \"enzyme_id\"                \"intermediate_filament_db\"\n",
       "[49] \"rna_central_id\"           \"lncipedia\"               \n",
       "[51] \"gtrnadb\"                  \"agr\"                     \n",
       "[53] \"mane_select\"              \"gencc\"                   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colnames(hgnc_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a5b0f369-9ded-4d34-858a-dedc2d360b0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#listAttributes(ensembl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "324bb4ac-7732-4ed1-9eec-07cdabbe928e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save to file\n",
    "write.table(hgnc_data, file = \"data/HGNC_gene_info_full.tsv\", sep = \"\\t\", row.names = FALSE, quote = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8badc343-6b0f-4b82-8fb5-7f0e0d36d439",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
