<!DOCTYPE html>
<html>
<meta charset="UTF-8">
<head>
    <title>C3——LRP1 - PMID Details</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
            color: #333;
        }

        h1 {
            text-align: center;
            font-size: 2.5em;
            color: #2c3e50;
            margin-bottom: 20px;
        }

        /* Navbar Styles */
        nav {
            background-color: #2c3e50;
            padding: 10px 20px;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            color: white;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        nav .navbar-left, nav .navbar-right {
            display: flex;
            align-items: center;
        }

        nav a {
            color: white;
            text-decoration: none;
            margin: 0 15px;
            font-size: 1.1em;
            transition: color 0.3s ease;
        }

        nav a:hover {
            color: #f39c12;
        }

        nav .divider {
            color: #ecf0f1;
            margin: 0 10px;
        }

        /* Page Footer */
        footer {
            margin-top: 40px;
            padding: 10px 20px;
            background-color: #f7f7f7;
            color: #333;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        /* Tab and Content Styles */
        .tab {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin-bottom: 20px;
        }

        .tab button {
            padding: 10px 20px;
            margin: 5px;
            margin-bottom: 10px;
            cursor: pointer;
            background-color: #f4f4f4;
            border: 2px solid #ddd;
            border-radius: 5px;
            font-size: 1.1em;
            transition: background-color 0.3s ease;
            flex-basis: calc(10% - 10px);
            box-sizing: border-box;
        }

        .tab button:hover {
            background-color: #ecf0f1;
        }

        .tab button.active {
            background-color: #3498db;
            color: white;
            border-color: #2980b9;
        }

        .tabcontent {
            display: none;
            padding: 20px;
            margin-top: 10px;
            background-color: #ecf0f1;
            border-radius: 5px;
            box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
        }

        .tabcontent.active {
            display: block;
        }

        .tabcontent h2 {
            font-size: 2em;
            color: #2980b9;
        }

        .tabcontent p {
            font-size: 1.2em;
            line-height: 1.8;
        }

        /* Emphasize Title and Abstract */
        .title, .abstract {
            background-color: #f9f9f9;
            padding: 20px;
            margin: 10px 0;
            border-left: 5px solid #2980b9;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
        }

        .abstract {
            border-left: 5px solid #e74c3c;
        }
    </style>
    <script>
        function adjustTabStyle() {
            var tabCount = document.querySelectorAll('.tab button').length;
            var tabContainer = document.querySelector('.tab');

            if (tabCount > 10) {
                tabContainer.classList.add('small-tabs');
            }
        }

        function openTab(evt, tabId) {
            document.querySelectorAll(".tabcontent").forEach(tab => tab.style.display = "none");
            document.querySelectorAll(".tablinks").forEach(link => link.classList.remove("active"));
            document.getElementById(tabId).style.display = "block";
            evt.currentTarget.classList.add("active");
        }

        window.onload = adjustTabStyle;
    </script>
</head>
<body>
    <nav>
        <div class="navbar-left">
            <a href="">ConnectomeDB</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/">Home</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database.html">Database</a>
        </div>
        <div class="navbar-right">
            <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/">YCU Bioinformatics Lab</a>
            <span class="divider">|</span>
            <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/">Perkins Systems Biology & Genomics Lab</a>
        </div>
    </nav>

    <h1>PMID Details for C3——LRP1</h1>

    <div class="tab">
        <button class="tablinks active" onclick="openTab(event, 'tab10608878')">10608878</button><button class="tablinks " onclick="openTab(event, 'tab10608878')">10608878</button>
    </div>

    <div class="tabcontent" id="titleTab">
        <div class="title">
            <h2>Title</h2>
            <p>{{TITLE}}</p>
            <i> Journal:{{JOURNAL}}; Year Published:{{YEAR}}</i>   
        </div>
    </div>

    <div class="tabcontent" id="abstractTab">
        <div class="abstract">
            <h2>Abstract</h2>
            <p>{{ABSTRACT}}</p>
        </div>
    </div>

    
                <div id="tab10608878" class="tabcontent active">
                    <h2>Metabolism of activated complement component C3 is mediated by the low density lipoprotein receptor-related protein/alpha(2)-macroglobulin receptor.</h2>
                    <h3>The Journal of biological chemistry; 1999</h3>
                    <p>Complement component 3 (C3) and alpha(2)-macroglobulin evolved from a common, evolutionarily old, ancestor gene. Low density lipoprotein-receptor-related protein/alpha(2)-macroglobulin receptor (LRP/alpha(2)MR), a member of the low density lipoprotein receptor family, is responsible for the clearance of alpha(2)-macroglobulin-protease complexes. In this study, we examined whether C3 has conserved affinity for LRP/alpha(2)MR. Ligand blot experiments with human (125)I-C3 on endosomal proteins show binding to a 600-kDa protein, indistinguishable from LRP/alpha(2)MR by the following criteria: it is competed by receptor-associated protein (the 39-kDa receptor-associated protein that impairs binding of all ligands to LRP/alpha(2)MR) and by lactoferrin and Pseudomonas exotoxin, other well known ligands of the multifunctional receptor. Binding of C3 is sensitive to reduction of the receptor and is Ca(2+)-dependent. All these features are typical for cysteine-rich binding repeats of the low density lipoprotein receptor family. In LRP/alpha(2)MR, they are found in four cassettes (2, 8, 10, and 11 repeats). Ligand blotting to chicken LR8 demonstrates that a single 8-fold repeat is sufficient for binding. Confocal microscopy visualizes initial surface labeling of human fibroblasts incubated with fluorescent labeled C3, which changes after 5 min to an intracellular vesicular staining pattern that is abolished in the presence of receptor-associated protein. Cell uptake is abolished in mouse fibroblasts deficient in LRP/alpha(2)MR. Native plasma C3 is not internalized. We demonstrate that the capacity to internalize C3 is saturable and exhibits a K(D) value of 17 nM. After intravenous injection, rat hepatocytes accumulate C3 in sedimentable vesicles with a density typical for endosomes. In conclusion, our ligand blot and uptake studies demonstrate the competence of the LRP/alpha(2)MR to bind and endocytose C3 and provide evidence for an LRP/alpha(2)MR-mediated system participating in C3 metabolism.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/10608878/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab10608878" class="tabcontent ">
                    <h2>Metabolism of activated complement component C3 is mediated by the low density lipoprotein receptor-related protein/alpha(2)-macroglobulin receptor.</h2>
                    <h3>The Journal of biological chemistry; 1999</h3>
                    <p>Complement component 3 (C3) and alpha(2)-macroglobulin evolved from a common, evolutionarily old, ancestor gene. Low density lipoprotein-receptor-related protein/alpha(2)-macroglobulin receptor (LRP/alpha(2)MR), a member of the low density lipoprotein receptor family, is responsible for the clearance of alpha(2)-macroglobulin-protease complexes. In this study, we examined whether C3 has conserved affinity for LRP/alpha(2)MR. Ligand blot experiments with human (125)I-C3 on endosomal proteins show binding to a 600-kDa protein, indistinguishable from LRP/alpha(2)MR by the following criteria: it is competed by receptor-associated protein (the 39-kDa receptor-associated protein that impairs binding of all ligands to LRP/alpha(2)MR) and by lactoferrin and Pseudomonas exotoxin, other well known ligands of the multifunctional receptor. Binding of C3 is sensitive to reduction of the receptor and is Ca(2+)-dependent. All these features are typical for cysteine-rich binding repeats of the low density lipoprotein receptor family. In LRP/alpha(2)MR, they are found in four cassettes (2, 8, 10, and 11 repeats). Ligand blotting to chicken LR8 demonstrates that a single 8-fold repeat is sufficient for binding. Confocal microscopy visualizes initial surface labeling of human fibroblasts incubated with fluorescent labeled C3, which changes after 5 min to an intracellular vesicular staining pattern that is abolished in the presence of receptor-associated protein. Cell uptake is abolished in mouse fibroblasts deficient in LRP/alpha(2)MR. Native plasma C3 is not internalized. We demonstrate that the capacity to internalize C3 is saturable and exhibits a K(D) value of 17 nM. After intravenous injection, rat hepatocytes accumulate C3 in sedimentable vesicles with a density typical for endosomes. In conclusion, our ligand blot and uptake studies demonstrate the competence of the LRP/alpha(2)MR to bind and endocytose C3 and provide evidence for an LRP/alpha(2)MR-mediated system participating in C3 metabolism.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/10608878/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                

    <footer>
        <span>© Copyright 2024, YCU Bioinformatics Laboratory</span>
        <span>Page built with <a href="https://quarto.org/">Quarto</a> by バイオインフォマティクス解析センター</span>
    </footer>
</body>
</html>
