<!DOCTYPE html>
<html>
<meta charset="UTF-8">
<head>
    <title>CCL2——CCR2 - PMID Details</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
            color: #333;
        }

        h1 {
            text-align: center;
            font-size: 2.5em;
            color: #2c3e50;
            margin-bottom: 20px;
        }

        /* Navbar Styles */
        nav {
            background-color: #2c3e50;
            padding: 10px 20px;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            color: white;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        nav .navbar-left, nav .navbar-right {
            display: flex;
            align-items: center;
        }

        nav a {
            color: white;
            text-decoration: none;
            margin: 0 15px;
            font-size: 1.1em;
            transition: color 0.3s ease;
        }

        nav a:hover {
            color: #f39c12;
        }

        nav .divider {
            color: #ecf0f1;
            margin: 0 10px;
        }

        /* Page Footer */
        footer {
            margin-top: 40px;
            padding: 10px 20px;
            background-color: #f7f7f7;
            color: #333;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        /* Tab and Content Styles */
        .tab {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin-bottom: 20px;
        }

        .tab button {
            padding: 10px 20px;
            margin: 5px;
            margin-bottom: 10px;
            cursor: pointer;
            background-color: #f4f4f4;
            border: 2px solid #ddd;
            border-radius: 5px;
            font-size: 1.1em;
            transition: background-color 0.3s ease;
            flex-basis: calc(10% - 10px);
            box-sizing: border-box;
        }

        .tab button:hover {
            background-color: #ecf0f1;
        }

        .tab button.active {
            background-color: #3498db;
            color: white;
            border-color: #2980b9;
        }

        .tabcontent {
            display: none;
            padding: 20px;
            margin-top: 10px;
            background-color: #ecf0f1;
            border-radius: 5px;
            box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
        }

        .tabcontent.active {
            display: block;
        }

        .tabcontent h2 {
            font-size: 2em;
            color: #2980b9;
        }

        .tabcontent p {
            font-size: 1.2em;
            line-height: 1.8;
        }

        /* Emphasize Title and Abstract */
        .title, .abstract {
            background-color: #f9f9f9;
            padding: 20px;
            margin: 10px 0;
            border-left: 5px solid #2980b9;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
        }

        .abstract {
            border-left: 5px solid #e74c3c;
        }
    </style>
    <script>
        function adjustTabStyle() {
            var tabCount = document.querySelectorAll('.tab button').length;
            var tabContainer = document.querySelector('.tab');

            if (tabCount > 10) {
                tabContainer.classList.add('small-tabs');
            }
        }

        function openTab(evt, tabId) {
            document.querySelectorAll(".tabcontent").forEach(tab => tab.style.display = "none");
            document.querySelectorAll(".tablinks").forEach(link => link.classList.remove("active"));
            document.getElementById(tabId).style.display = "block";
            evt.currentTarget.classList.add("active");
        }

        window.onload = adjustTabStyle;
    </script>
</head>
<body>
    <nav>
        <div class="navbar-left">
            <a href="">ConnectomeDB</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/">Home</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database.html">Database</a>
        </div>
        <div class="navbar-right">
            <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/">YCU Bioinformatics Lab</a>
            <span class="divider">|</span>
            <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/">Perkins Systems Biology & Genomics Lab</a>
        </div>
    </nav>

    <h1>PMID Details for CCL2——CCR2</h1>

    <div class="tab">
        <button class="tablinks active" onclick="openTab(event, 'tab9287323')">9287323</button><button class="tablinks " onclick="openTab(event, 'tab9276730')">9276730</button><button class="tablinks " onclick="openTab(event, 'tab9346309')">9346309</button><button class="tablinks " onclick="openTab(event, 'tab12554737')">12554737</button><button class="tablinks " onclick="openTab(event, 'tab15207250')">15207250</button><button class="tablinks " onclick="openTab(event, 'tab10770925')">10770925</button><button class="tablinks " onclick="openTab(event, 'tab10529171')">10529171</button><button class="tablinks " onclick="openTab(event, 'tab9287323')">9287323</button>
    </div>

    <div class="tabcontent" id="titleTab">
        <div class="title">
            <h2>Title</h2>
            <p>{{TITLE}}</p>
            <i> Journal:{{JOURNAL}}; Year Published:{{YEAR}}</i>   
        </div>
    </div>

    <div class="tabcontent" id="abstractTab">
        <div class="abstract">
            <h2>Abstract</h2>
            <p>{{ABSTRACT}}</p>
        </div>
    </div>

    
                <div id="tab9287323" class="tabcontent active">
                    <h2>The amino-terminal domain of CCR2 is both necessary and sufficient for high affinity binding of monocyte chemoattractant protein 1. Receptor activation by a pseudo-tethered ligand.</h2>
                    <h3>The Journal of biological chemistry; 1997</h3>
                    <p>High affinity binding of monocyte chemoattractant protein 1 (MCP-1) requires the presence of the amino-terminal domain of CCR2, the MCP-1 receptor. Here we report that the 35 amino-terminal residues of CCR2, expressed as a membrane-bound fusion protein, bound MCP-1 with an affinity similar to that of the intact, wild-type receptor. Furthermore, the amino-terminal fusion protein enhanced, in trans, agonist-dependent activation of a CCR2 variant that was engineered to lack the high affinity binding sites for MCP-1. Mutation of highly conserved cysteines in the amino-terminal domain and third extracellular loop of CCR2, but not in the fusion protein, resulted in a dramatic loss of MCP-1 binding, suggesting the existence of a critical intramolecular disulfide bond that positions the amino-terminal protein for ligand interaction. These data indicate that the amino-terminal region of CCR2 is both necessary and sufficient for the high affinity binding of MCP-1 and provide the first direct evidence for activation of a chemokine receptor by a pseudo-tethered ligand. In this model, high affinity binding by the relatively short amino-terminal domain of CCR2 serves to tether MCP-1 and enhance low affinity interactions with distal regions of the receptor.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/9287323/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab9276730" class="tabcontent ">
                    <h2>High expression of the chemokine receptor CCR3 in human blood basophils. Role in activation by eotaxin, MCP-4, and other chemokines.</h2>
                    <h3>The Journal of clinical investigation; 1997</h3>
                    <p>Eosinophil leukocytes express high numbers of the chemokine receptor CCR3 which binds eotaxin, monocyte chemotactic protein (MCP)-4, and some other CC chemokines. In this paper we show that CCR3 is also highly expressed on human blood basophils, as indicated by Northern blotting and flow cytometry, and mediates mainly chemotaxis. Eotaxin and MCP-4 elicited basophil migration in vitro with similar efficacy as regulated upon activation normal T cells expressed and secreted (RANTES) and MCP-3. They also induced the release of histamine and leukotrienes in IL-3-primed basophils, but their efficacy was lower than that of MCP-1 and MCP-3, which were the most potent stimuli of exocytosis. Pretreatment of the basophils with a CCR3-blocking antibody abrogated the migration induced by eotaxin, RANTES, and by low to optimal concentrations of MCP-4, but decreased only minimally the response to MCP-3. The CCR3-blocking antibody also affected exocytosis: it abrogated histamine and leukotriene release induced by eotaxin, and partially inhibited the response to RANTES and MCP-4. In contrast, the antibody did not affect the responses induced by MCP-1, MCP-3, and macrophage inflammatory protein-1alpha, which may depend on CCR1 and CCR2, two additional receptors detected by Northern blotting with basophil RNA. This study demonstrates that CCR3 is the major receptor for eotaxin, RANTES, and MCP-4 in human basophils, and suggests that basophils and eosinophils, which are the characteristic effector cells of allergic inflammation, depend largely on CCR3 for migration towards different chemokines into inflamed tissues.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/9276730/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab9346309" class="tabcontent ">
                    <h2>Characterisation of macrophage inflammatory protein-5/human CC cytokine-2, a member of the macrophage-inflammatory-protein family of chemokines.</h2>
                    <h3>European journal of biochemistry; 1997</h3>
                    <p>A human monocyte-activating CC chemokine has been identified based on sequences in an expressed sequence tag (EST) cDNA database. The protein shows highest sequence identity to the macrophage inflammatory protein (MIP) group of chemokines, particularly MIP-3 (76.7%) and MIP-1alpha (75.4%), and has been named MIP-5. Model building confirms that the protein has a similar three dimensional structure to other chemokines, but has an additional third disulphide bond. Northern blot analysis and reverse-transcriptase PCR show that the mRNA for MIP-5 is expressed at a high levels in liver, intestine and in lung leukocytes. MIP-5 induces chemotaxis of human monocytes, T-lymphocytes and, to a lesser degree, eosinophils at nanomolar concentrations; it has no effect on neutrophil migration. In receptor-binding assays, MIP-5 shows IC50 values of 12 nM for competition with 125I-MIP-1alpha for binding to CC-chemokine receptor (CCR)1, and 2.5 nM for competition with 125I-MCP-3 for binding to CCR3. It shows no ability to compete with ligand for binding to the two interleukin (IL)-8 receptors (CXC-chemokine receptors 1 and 2) or to CCR2, CCR4 or CCR5. Consistent with this binding data, MIP-5 was only able to induce calcium fluxes in CHO cells stably transfected with CCR1 or CCR3.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/9346309/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab12554737" class="tabcontent ">
                    <h2>A highly selective CCR2 chemokine agonist encoded by human herpesvirus 6.</h2>
                    <h3>The Journal of biological chemistry; 2003</h3>
                    <p>The chemokine-like, secreted protein product of the U83 gene from human herpesvirus 6, here named vCCL4, was chemically synthesized to be characterized in a complete library of the 18 known human chemokine receptors expressed individually in stably transfected cell lines. vCCL4 was found to cause calcium mobilization as efficiently as the endogenous chemokine ligand CCL2 through the CCR2 receptor, whereas the virally encoded chemokine did not affect any of the other 17 human chemokine receptors tested. Mutual cross-desensitization between CCL2 and vCCL4 was demonstrated in the CCR2-transfected cells. The affinity of vCCL4 for the CCR2 receptor was 79 nm as determined in competition binding against radioactively labeled CCL2. In the murine pre-B lymphocyte cell line L1.2 stably transfected with the CCR2 receptor, vCCL4 acted as a relatively low potency but highly efficacious chemoattractant being equally or more efficacious in causing cell migration than CCL2 and CCL7 and considerably more efficacious than CCL8 and CCL13. It is concluded that human herpesvirus 6 encodes a highly selective and efficacious CCR2 agonist, which will attract CCR2 expressing cells, for example macrophages and monocytes, conceivably for the virus to infect and to establish latency in. It is suggested that vCCL4 during reactivation of the virus in for example monocyte-derived microglia could perhaps be involved in the pathogenesis of the CCR2-dependent disease, multiple sclerosis.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/12554737/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab15207250" class="tabcontent ">
                    <h2>High level expression, activation, and antagonism of CC chemokine receptors CCR2 and CCR3 in Chinese hamster ovary cells.</h2>
                    <h3>Cytokine; 2004</h3>
                    <p>The specificity of leukocyte trafficking in inflammation is controlled by the interactions of chemokines with chemokine receptors. Reliable structure-function studies of chemokine-receptor interactions would benefit from cell lines that express consistent high levels of chemokine receptors. We describe herein two new Chinese hamster ovary (CHO) cell lines in which the genes for chemokine receptors CCR2 and CCR3 have been incorporated into identical positions in the host genome. CCR2 is the primary receptor for the chemokine monocyte chemoattractant protein-1 (MCP-1) whereas CCR3 is the primary receptor for the chemokines eotaxin-1, eotaxin-2 and eotaxin-3. Both receptors are expressed at >5,000,000 copies per cell, substantially higher levels than in previous cell lines, and both are competent for binding and activation by the cognate chemokines for these receptors. Using these cell lines we confirm that eotaxin-1 and eotaxin-3 can act as an agonist and an antagonist, respectively, of CCR2. In addition, we show that eotaxin-2 is an antagonist of CCR2 and MCP-1 is an agonist of CCR3. Comparison of the chemokine sequences reveals several positions that are identical in MCP-1 and eotaxin-1 but different in eotaxin-2 and eotaxin-3, suggesting that these amino acids play a role in CCR2 activation.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/15207250/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab10770925" class="tabcontent ">
                    <h2>Identification of the binding site for a novel class of CCR2b chemokine receptor antagonists: binding to a common chemokine receptor motif within the helical bundle.</h2>
                    <h3>The Journal of biological chemistry; 2000</h3>
                    <p>Monocyte chemoattracant-1 (MCP-1) stimulates leukocyte chemotaxis to inflammatory sites, such as rheumatoid arthritis, atherosclerosis, and asthma, by use of the MCP-1 receptor, CCR2, a member of the G-protein-coupled seven-transmembrane receptor superfamily. These studies identified a family of antagonists, spiropiperidines. One of the more potent compounds blocks MCP-1 binding to CCR2 with a K(d) of 60 nm, but it is unable to block binding to CXCR1, CCR1, or CCR3. These compounds were effective inhibitors of chemotaxis toward MCP-1 but were very poor inhibitors of CCR1-mediated chemotaxis. The compounds are effective blockers of MCP-1-driven inhibition of adenylate cyclase and MCP-1- and MCP-3-driven cytosolic calcium influx; the compounds are not agonists for these pathways. We showed that glutamate 291 (Glu(291)) of CCR2 is a critical residue for high affinity binding and that this residue contributes little to MCP-1 binding to CCR2. The basic nitrogen present in the spiropiperidine compounds may be the interaction partner for Glu(291), because the basicity of this nitrogen was essential for affinity; furthermore, a different class of antagonists, a class that does not have a basic nitrogen (2-carboxypyrroles), were not affected by mutations of Glu(291). In addition to the CCR2 receptor, spiropiperidine compounds have affinity for several biogenic amine receptors. Receptor models indicate that the acidic residue, Glu(291), from transmembrane-7 of CCR2 is in a position similar to the acidic residue contributed from transmembrane-3 of biogenic amine receptors, which may account for the shared affinity of spiropiperidines for these two receptor classes. The models suggest that the acid-base pair, Glu(291) to piperidine nitrogen, anchors the spiropiperidine compound within the transmembrane ovoid bundle. This binding site may overlap with the space required by MCP-1 during binding and signaling; thus the small molecule ligands act as antagonists. An acidic residue in transmembrane region 7 is found in most chemokine receptors and is rare in other serpentine receptors. The model of the binding site may suggest ways to make new small molecule chemokine receptor antagonists, and it may rationalize the design of more potent and selective antagonists.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/10770925/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab10529171" class="tabcontent ">
                    <h2>Identification of residues in the monocyte chemotactic protein-1 that contact the MCP-1 receptor, CCR2.</h2>
                    <h3>Biochemistry; 1999</h3>
                    <p>The CC chemokine, MCP-1, has been identified as a major chemoattractant for T cells and monocytes, and plays a significant role in the pathology of inflammatory diseases. To identify the regions of MCP-1 that contact its receptor, CCR2, we substituted all surface-exposed residues with alanine. Some residues were also mutated to other amino acids to identify the importance of charge, hydrophobicity, or aromaticity at specific positions. The binding affinity of each mutant for CCR2 was assayed with THP-1 and CCR2-transfected CHL cells. The majority of point mutations had no effect. Residues at the N-terminus of the protein, known to be crucial for signaling, contribute less than a factor of 10 to the binding affinity. However, two clusters of primarily basic residues (R24, K35, K38, K49, and Y13), separated by a 35 A hydrophobic groove, reduced the level of binding by 15-100-fold. A peptide fragment encompassing residues 13-35 recapitulated some of the mutational data derived from the intact protein. It exhibited modest binding as a linear peptide and dramatically improved affinity when the region which adopts a single turn of a 3(10)-helix in the protein, which includes R24, was constrained by a disulfide bond. Additional constraints at the ends of the peptide, corresponding to the disulfide between the first and third cysteines in MCP-1, yielded further improvements in affinity. Together, these data suggest a model in which a large surface area of MCP-1 contacts the receptor, and the accumulation of a number of weak interactions results in the 35 pM affinity observed for the wild-type (WT) protein. The receptor binding site of MCP-1 also is significantly different from the binding sites of RANTES and IL-8, providing insight into the issue of receptor specificity. It was previously shown that the N-terminus of CCR2 is critical for binding MCP-1 [Monteclaro, F. S., and Charo, I. F. (1996) J. Biol. Chem. 271, 19084-92; Monteclaro, F. S., and Charo, I. F. (1997) J. Biol. Chem. 272, 23186-90]. Point mutations of six acidic residues in this region of the receptor were made to test their role in ligand binding. This identified D25 and D27 of the DYDY motif as being important. On the basis of our data, we propose a model in which the receptor N-terminus lies along the hydrophobic groove in an extended fashion, placing the DYDY motif near the basic cluster involving R24 and K49 of MCP-1. This in turn orients the signaling residues (Y13 and the N-terminus) for productive interaction with the receptor.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/10529171/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab9287323" class="tabcontent ">
                    <h2>The amino-terminal domain of CCR2 is both necessary and sufficient for high affinity binding of monocyte chemoattractant protein 1. Receptor activation by a pseudo-tethered ligand.</h2>
                    <h3>The Journal of biological chemistry; 1997</h3>
                    <p>High affinity binding of monocyte chemoattractant protein 1 (MCP-1) requires the presence of the amino-terminal domain of CCR2, the MCP-1 receptor. Here we report that the 35 amino-terminal residues of CCR2, expressed as a membrane-bound fusion protein, bound MCP-1 with an affinity similar to that of the intact, wild-type receptor. Furthermore, the amino-terminal fusion protein enhanced, in trans, agonist-dependent activation of a CCR2 variant that was engineered to lack the high affinity binding sites for MCP-1. Mutation of highly conserved cysteines in the amino-terminal domain and third extracellular loop of CCR2, but not in the fusion protein, resulted in a dramatic loss of MCP-1 binding, suggesting the existence of a critical intramolecular disulfide bond that positions the amino-terminal protein for ligand interaction. These data indicate that the amino-terminal region of CCR2 is both necessary and sufficient for the high affinity binding of MCP-1 and provide the first direct evidence for activation of a chemokine receptor by a pseudo-tethered ligand. In this model, high affinity binding by the relatively short amino-terminal domain of CCR2 serves to tether MCP-1 and enhance low affinity interactions with distal regions of the receptor.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/9287323/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                

    <footer>
        <span>© Copyright 2024, YCU Bioinformatics Laboratory</span>
        <span>Page built with <a href="https://quarto.org/">Quarto</a> by バイオインフォマティクス解析センター</span>
    </footer>
</body>
</html>
