<!DOCTYPE html>
<html>
<meta charset="UTF-8">
<head>
    <title>CCL5——CCR1 - PMID Details</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
            color: #333;
        }

        h1 {
            text-align: center;
            font-size: 2.5em;
            color: #2c3e50;
            margin-bottom: 20px;
        }

        /* Navbar Styles */
        nav {
            background-color: #2c3e50;
            padding: 10px 20px;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            color: white;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        nav .navbar-left, nav .navbar-right {
            display: flex;
            align-items: center;
        }

        nav a {
            color: white;
            text-decoration: none;
            margin: 0 15px;
            font-size: 1.1em;
            transition: color 0.3s ease;
        }

        nav a:hover {
            color: #f39c12;
        }

        nav .divider {
            color: #ecf0f1;
            margin: 0 10px;
        }

        /* Page Footer */
        footer {
            margin-top: 40px;
            padding: 10px 20px;
            background-color: #f7f7f7;
            color: #333;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        /* Tab and Content Styles */
        .tab {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin-bottom: 20px;
        }

        .tab button {
            padding: 10px 20px;
            margin: 5px;
            margin-bottom: 10px;
            cursor: pointer;
            background-color: #f4f4f4;
            border: 2px solid #ddd;
            border-radius: 5px;
            font-size: 1.1em;
            transition: background-color 0.3s ease;
            flex-basis: calc(10% - 10px);
            box-sizing: border-box;
        }

        .tab button:hover {
            background-color: #ecf0f1;
        }

        .tab button.active {
            background-color: #3498db;
            color: white;
            border-color: #2980b9;
        }

        .tabcontent {
            display: none;
            padding: 20px;
            margin-top: 10px;
            background-color: #ecf0f1;
            border-radius: 5px;
            box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
        }

        .tabcontent.active {
            display: block;
        }

        .tabcontent h2 {
            font-size: 2em;
            color: #2980b9;
        }

        .tabcontent p {
            font-size: 1.2em;
            line-height: 1.8;
        }

        /* Emphasize Title and Abstract */
        .title, .abstract {
            background-color: #f9f9f9;
            padding: 20px;
            margin: 10px 0;
            border-left: 5px solid #2980b9;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
        }

        .abstract {
            border-left: 5px solid #e74c3c;
        }
    </style>
    <script>
        function adjustTabStyle() {
            var tabCount = document.querySelectorAll('.tab button').length;
            var tabContainer = document.querySelector('.tab');

            if (tabCount > 10) {
                tabContainer.classList.add('small-tabs');
            }
        }

        function openTab(evt, tabId) {
            document.querySelectorAll(".tabcontent").forEach(tab => tab.style.display = "none");
            document.querySelectorAll(".tablinks").forEach(link => link.classList.remove("active"));
            document.getElementById(tabId).style.display = "block";
            evt.currentTarget.classList.add("active");
        }

        window.onload = adjustTabStyle;
    </script>
</head>
<body>
    <nav>
        <div class="navbar-left">
            <a href="">ConnectomeDB</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/">Home</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database.html">Database</a>
        </div>
        <div class="navbar-right">
            <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/">YCU Bioinformatics Lab</a>
            <span class="divider">|</span>
            <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/">Perkins Systems Biology & Genomics Lab</a>
        </div>
    </nav>

    <h1>PMID Details for CCL5——CCR1</h1>

    <div class="tab">
        <button class="tablinks active" onclick="openTab(event, 'tab9624164')">9624164</button><button class="tablinks " onclick="openTab(event, 'tab8530354')">8530354</button><button class="tablinks " onclick="openTab(event, 'tab9289016')">9289016</button><button class="tablinks " onclick="openTab(event, 'tab11116158')">11116158</button><button class="tablinks " onclick="openTab(event, 'tab14637022')">14637022</button><button class="tablinks " onclick="openTab(event, 'tab12270118')">12270118</button><button class="tablinks " onclick="openTab(event, 'tab7545673')">7545673</button>
    </div>

    <div class="tabcontent" id="titleTab">
        <div class="title">
            <h2>Title</h2>
            <p>{{TITLE}}</p>
            <i> Journal:{{JOURNAL}}; Year Published:{{YEAR}}</i>   
        </div>
    </div>

    <div class="tabcontent" id="abstractTab">
        <div class="abstract">
            <h2>Abstract</h2>
            <p>{{ABSTRACT}}</p>
        </div>
    </div>

    
                <div id="tab9624164" class="tabcontent active">
                    <h2>Identification and characterization of small molecule functional antagonists of the CCR1 chemokine receptor.</h2>
                    <h3>The Journal of biological chemistry; 1998</h3>
                    <p>The CC chemokines macrophage inflammatory protein-1alpha (MIP-1alpha) and RANTES (regulated on activation normal T cell expressed) have been implicated in rheumatoid arthritis and multiple sclerosis. Since their effects are mediated through the CCR1 chemokine receptor, we set up a small molecule CCR1 antagonist program to search for inhibitors. Through high capacity screening we discovered a number of 4-hydroxypiperidine compounds with CCR1 antagonist activity and report their synthesis and in vitro pharmacology here. Scatchard analysis of the competition binding data revealed that the compounds had Ki values ranging from 40 to 4000 nM. The pharmacological profile of the most potent member of this series, compound 1 (2-2-diphenyl-5-(4-chlorophenyl)piperidin-lyl)valeronitri te), was further evaluated. Compound 1 showed concentration-dependent inhibition of MIP-1alpha-induced extracellular acidification and Ca2+ mobilization demonstrating functional antagonism. When given alone, the compound did not elicit any responses, indicating the absence of intrinsic agonist activity. Compound 1 inhibited MIP-1alpha- and RANTES-induced migration in peripheral blood mononuclear cells in a dose-responsive manner. Selectivity testing against a panel of seven transmembrane domain receptors indicated that compound 1 is inactive on a number of receptors at concentrations up to 10 microM. This is the first description of CCR1 receptor antagonists that may be useful in the treatment of chronic inflammatory diseases involving MIP-1alpha, RANTES, and CCR1.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/9624164/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab8530354" class="tabcontent ">
                    <h2>Monocyte chemoattractant protein-3 is a functional ligand for CC chemokine receptors 1 and 2B.</h2>
                    <h3>The Journal of biological chemistry; 1995</h3>
                    <p>The CC chemokine monocyte chemoattractant protein-3 (MCP-3) activates human monocytes, lymphocytes, basophils, and eosinophils. MCP-3 has been reported to induce [Ca2+]i changes in cells transfected with the monocyte-selective MCP-1 receptor 2B (CC CKR2B) and competes for 125I-MCP-1 binding on CC CKR2B, suggesting that it may mediate monocyte responses to MCP-3. However, we now show that MCP-3 is a ligand and potent agonist for the macrophage inflammatory protein-1 alpha (MIP-1 alpha)/regulated on activation, normal T expressed, and secreted protein (RANTES) receptor CC CKR1 (rank order for [Ca2+]i changes = MIP-1 alpha > MCP-3 > RANTES), which is expressed in monocytes > neutrophils > eosinophils. 125I-MCP-3 bound directly to CC CKR1 and CC CKR2B (Ki = 8 and 7 nM, respectively). Binding to CC CKR1 was competed by all CC chemokines tested except MCP-1. In contrast, binding to CC CKR2B was competed only by MCP-3 and MCP-1. Both MCP-1 and MCP-3 were equipotent agonists (EC50 = 10 nM for [Ca2+]i changes). Thus, MCP-3 is a functional ligand for both CC CKR1 and CC CKR2B, which otherwise have distinct selectivities for CC chemokines. These data suggest that monocyte responses to MCP-3 could be mediated by both CC CKR2B and CC CKR1, whereas eosinophil responses to MCP-3 could be mediated by CC CKR1.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/8530354/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab9289016" class="tabcontent ">
                    <h2>Distinct but overlapping epitopes for the interaction of a CC-chemokine with CCR1, CCR3 and CCR5.</h2>
                    <h3>Biochemistry; 1997</h3>
                    <p>Chemokines play an important role in inflammation. The mechanism via which they bind to more than one receptor and activate them is not well understood. The chemokines are thought to interact with their receptors via two distinct sites, one necessary for binding and the other for activation of signal transduction. In this study we have used alanine scanning mutagenesis to identify residues on RANTES that specifically interact with its receptors CCR1, CCR3, and CCR5 for binding and activation. Residues within a potential receptor binding site known as the N-loop (residues 12-20) and near the N-terminus of RANTES were individually mutated to alanine. The results of this study show that, within the N-loop, the side chain of R17 is necessary for RANTES binding to CCR1, F12 for binding to CCR3, and F12 and I15 for binding to CCR5, thus forming distinct but overlapping binding epitopes. In addition, our finding that P2 is necessary for binding to CCR5 is the first to show that a residue near the N-terminus of a CC-chemokine is involved in binding to a receptor. We have also found that P2, D6, and T7 near the N-terminus are involved in activating signal transduction via CCR1, P2 and Y3 via CCR3, and Y3 and D6 via CCR5. These results indicate that RANTES interacts with each of its receptors in a distinct and specific manner and provide further evidence to support the two-site model of interaction between chemokines and their receptors.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/9289016/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab11116158" class="tabcontent ">
                    <h2>The BBXB motif of RANTES is the principal site for heparin binding and controls receptor selectivity.</h2>
                    <h3>The Journal of biological chemistry; 2001</h3>
                    <p>The chemokine RANTES (regulated on activation normal T cell expressed and secreted; CCL5) binds selectively to glycosaminoglycans (GAGs) such as heparin, chondroitin sulfate, and dermatan sulfate. The primary sequence of RANTES contains two clusters of basic residues, (44)RKNR(47) and (55)KKWVR(59). The first is a BBXB motif common in heparin-binding proteins, and the second is located in the loop directly preceding the C-terminal helix. We have mutated these residues to alanine, both as point mutations as well as triple mutations of the 40s and 50s clusters. Using a binding assay to heparin beads with radiolabeled proteins, the (44)AANA(47) mutant demonstrated an 80% reduction in its capacity to bind heparin, whereas the (55)AAWVA(59) mutant retained full binding capacity. Mutation of the (44)RKNR(47) site reduced the selectivity of RANTES binding to different GAGs. The mutants were tested for their integrity by receptor binding assays on CCR1 and CCR5 as well as their ability to induce chemotaxis in vitro. In all assays the single point mutations and the triple 50s cluster mutation caused no significant difference in activity compared with the wild type sequence. However, the triple 40s mutant showed a 80-fold reduction in affinity for CCR1, despite normal binding to CCR5. It was only able to induce monocyte chemotaxis at micromolar concentrations. The triple 40s mutant was also able to inhibit HIV-1 infectivity, but consistent with its abrogated GAG binding capacity, it no longer induced enhanced infectivity at high concentrations.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/11116158/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab14637022" class="tabcontent ">
                    <h2>Interaction of RANTES with syndecan-1 and syndecan-4 expressed by human primary macrophages.</h2>
                    <h3>Biochimica et biophysica acta; 2003</h3>
                    <p>Interaction of RANTES with its membrane ligands or receptors transduces multiple intracellular signals. Whether RANTES uses proteoglycans (PGs) belonging to the syndecan family to attach to primary cells expressing RANTES G-protein-coupled receptors (GPCRs) was investigated. We demonstrate that RANTES specifically binds to high and low affinity binding sites on human monocyte-derived macrophages (MDM). We show by co-immunoprecipitation experiments that RANTES is associated on these cells with syndecan-1 and syndecan-4, but neither with syndecan-2 nor with betaglycan, in addition to CD44 and its GPCRs, CCR5 and CCR1. Glycosaminidases pre-treatment of the monocyte derived-macrophages strongly decreases the binding of RANTES to syndecan-1 and syndecan-4 and also to CCR5, and abolishes RANTES binding to CD44. This suggests that glycosaminoglycans (GAGs) are involved in RANTES binding to the PGs and that such bindings facilitate the subsequent interaction of RANTES with CCR5, on the MDM, characterized by low membrane expression of CCR5. The role of these interactions in the pathophysiology of RANTES deserves further study.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/14637022/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab12270118" class="tabcontent ">
                    <h2>Selective CCL5/RANTES-induced mast cell migration through interactions with chemokine receptors CCR1 and CCR4.</h2>
                    <h3>Biochemical and biophysical research communications; 2002</h3>
                    <p>Mast cells (MCs) accumulate at sites of allergic mucosal inflammation where they act as central effector and regulatory cells. Because chemokines are of vital importance in directing inflammatory leukocytes to the sites of inflammations, we have investigated the expression and function of CC-chemokine receptor (CCR) on human MCs. Two previously unrecognized MC-chemokine receptors, CCR1 and CCR4, could be identified on cord blood-derived MCs (CBMCs). CCR1 and CCR4 expressed on CBMCs exhibited a unique response profile. Of seven CCR1 and CCR4 agonists tested, only CCL5/RANTES act as an agonist inducing chemotaxis. The migration could be partially blocked by specific antibodies against CCR1 or CCR4, while a complete inhibition was achieved when both CCR1 and CCR4 were blocked. These results demonstrate that both CCR1 and CCR4 are functional receptors on human mast cells with capacity to mediate migration towards CCL5.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/12270118/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab7545673" class="tabcontent ">
                    <h2>Monocyte chemotactic protein-3 (MCP3) interacts with multiple leukocyte receptors. C-C CKR1, a receptor for macrophage inflammatory protein-1 alpha/Rantes, is also a functional receptor for MCP3.</h2>
                    <h3>The Journal of biological chemistry; 1995</h3>
                    <p>Monocyte chemotactic protein-3 (MCP3) is recently identified and molecularly cloned C-C chemokine that is chemotactic for and activates a great variety of inflammatory cell types. MCP3 has been reported to interact with several C-C chemokine receptors, which can be simultaneously or selectively expressed on leukocyte subpopulations. In order to isolate receptor(s) for MCP3, a cDNA library was constructed using mRNA from a human NK-like cell line, YT. These cells showed high affinity binding sites for 125I-MCP3 and migrated in response to MCP3. A chemokine receptor cDNA clone, designated YT4, was sequenced and found to be identical to the known C-C CKR1 or macrophage inflammatory protein-1 alpha (MIP1 alpha)/Rantes receptor. YT4 cDNA was subcloned into a mammalian expression vector, and stable transfectants were prepared using the embryonic kidney cell line 293. The transfectants (YT4/293) showed high affinity binding for 125I-MCP3 in addition to specifically binding 125I-MIP1 alpha and 125I-Rantes. All three C-C chemokines were able to cross-compete for binding sites on YT4/293 cells and induced directional migration of YT4/293 cells in vitro, with MCP3 being the most potent chemoattractant. MCP3, MIP1 alpha, and Rantes were equally able to cross-attenuate the migratory response of YT4/293 cells to one another. In contrast, MCP1 and MIP1 beta had very limited capacity to compete for MCP3 binding on YT4/293 cells and had only a minor attenuating effect on MCP3-induced migration. Since MCP3 has been reported to use MCP1 receptor(s), our results with transfected 293 cells expressing only C-C CKR1 clearly establish that C-C CKR1 is also a functional receptor for MCP3.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/7545673/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                

    <footer>
        <span>© Copyright 2024, YCU Bioinformatics Laboratory</span>
        <span>Page built with <a href="https://quarto.org/">Quarto</a> by バイオインフォマティクス解析センター</span>
    </footer>
</body>
</html>
