<!DOCTYPE html>
<html>
<meta charset="UTF-8">
<head>
    <title>CXCL10——CXCR3 - PMID Details</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
            color: #333;
        }

        h1 {
            text-align: center;
            font-size: 2.5em;
            color: #2c3e50;
            margin-bottom: 20px;
        }

        /* Navbar Styles */
        nav {
            background-color: #2c3e50;
            padding: 10px 20px;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            color: white;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        nav .navbar-left, nav .navbar-right {
            display: flex;
            align-items: center;
        }

        nav a {
            color: white;
            text-decoration: none;
            margin: 0 15px;
            font-size: 1.1em;
            transition: color 0.3s ease;
        }

        nav a:hover {
            color: #f39c12;
        }

        nav .divider {
            color: #ecf0f1;
            margin: 0 10px;
        }

        /* Page Footer */
        footer {
            margin-top: 40px;
            padding: 10px 20px;
            background-color: #f7f7f7;
            color: #333;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        /* Tab and Content Styles */
        .tab {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin-bottom: 20px;
        }

        .tab button {
            padding: 10px 20px;
            margin: 5px;
            margin-bottom: 10px;
            cursor: pointer;
            background-color: #f4f4f4;
            border: 2px solid #ddd;
            border-radius: 5px;
            font-size: 1.1em;
            transition: background-color 0.3s ease;
            flex-basis: calc(10% - 10px);
            box-sizing: border-box;
        }

        .tab button:hover {
            background-color: #ecf0f1;
        }

        .tab button.active {
            background-color: #3498db;
            color: white;
            border-color: #2980b9;
        }

        .tabcontent {
            display: none;
            padding: 20px;
            margin-top: 10px;
            background-color: #ecf0f1;
            border-radius: 5px;
            box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
        }

        .tabcontent.active {
            display: block;
        }

        .tabcontent h2 {
            font-size: 2em;
            color: #2980b9;
        }

        .tabcontent p {
            font-size: 1.2em;
            line-height: 1.8;
        }

        /* Emphasize Title and Abstract */
        .title, .abstract {
            background-color: #f9f9f9;
            padding: 20px;
            margin: 10px 0;
            border-left: 5px solid #2980b9;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
        }

        .abstract {
            border-left: 5px solid #e74c3c;
        }
    </style>
    <script>
        function adjustTabStyle() {
            var tabCount = document.querySelectorAll('.tab button').length;
            var tabContainer = document.querySelector('.tab');

            if (tabCount > 10) {
                tabContainer.classList.add('small-tabs');
            }
        }

        function openTab(evt, tabId) {
            document.querySelectorAll(".tabcontent").forEach(tab => tab.style.display = "none");
            document.querySelectorAll(".tablinks").forEach(link => link.classList.remove("active"));
            document.getElementById(tabId).style.display = "block";
            evt.currentTarget.classList.add("active");
        }

        window.onload = adjustTabStyle;
    </script>
</head>
<body>
    <nav>
        <div class="navbar-left">
            <a href="">ConnectomeDB</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/">Home</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database.html">Database</a>
        </div>
        <div class="navbar-right">
            <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/">YCU Bioinformatics Lab</a>
            <span class="divider">|</span>
            <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/">Perkins Systems Biology & Genomics Lab</a>
        </div>
    </nav>

    <h1>PMID Details for CXCL10——CXCR3</h1>

    <div class="tab">
        <button class="tablinks active" onclick="openTab(event, 'tab9064356')">9064356</button><button class="tablinks " onclick="openTab(event, 'tab12884299')">12884299</button><button class="tablinks " onclick="openTab(event, 'tab9660793')">9660793</button><button class="tablinks " onclick="openTab(event, 'tab15761110')">15761110</button><button class="tablinks " onclick="openTab(event, 'tab12571234')">12571234</button><button class="tablinks " onclick="openTab(event, 'tab10233762')">10233762</button>
    </div>

    <div class="tabcontent" id="titleTab">
        <div class="title">
            <h2>Title</h2>
            <p>{{TITLE}}</p>
            <i> Journal:{{JOURNAL}}; Year Published:{{YEAR}}</i>   
        </div>
    </div>

    <div class="tabcontent" id="abstractTab">
        <div class="abstract">
            <h2>Abstract</h2>
            <p>{{ABSTRACT}}</p>
        </div>
    </div>

    
                <div id="tab9064356" class="tabcontent active">
                    <h2>Chemokine receptor specific for IP10 and mig: structure, function, and expression in activated T-lymphocytes.</h2>
                    <h3>The Journal of experimental medicine; 1996</h3>
                    <p>A human receptor that is selective for the CXC chemokines IP10 and Mig was cloned and characterized. The receptor cDNA has an open reading frame of 1104-bp encoding a protein of 368 amino acids with a molecular mass of 40,659 dalton. The sequence includes seven putative transmembrane segments characteristic of G-protein coupled receptors. It shares 40.9 and 40.3% identical amino acids with the two IL-8 receptors, and 34.2-36.9% identity with the five known CC chemokine receptors. The IP10/Mig receptor is highly expressed in IL-2-activated T lymphocytes, but is not detectable in resting T lymphocytes. B lymphocytes, monocytes and granulocytes. It mediates Ca2+ mobilization and chemotaxis in response to IP10 and Mig, but does not recognize the CXC-chemokines IL-8, GRO alpha, NAP-2, GCP-2. ENA78, PF4, the CC-chemokines MCP-1, MCP-2, MCP-3, MCP-4, MIP-1 alpha, MIP-1 beta. RANTES, 1309, eotaxin, nor lymphotactin. The exclusive expression in activated T-lymphocytes is of high interest since the receptors for chemokines which have been shown so far to attract lymphocytes, e.g., MCP-1, MCP-2, MCP-3, MIP-1 alpha, MIP-1 beta, and RANTES, are also found in monocytes and granulocytes. The present observations suggest that the IP10/Mig receptor is involved in the selective recruitment of effector T cells.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/9064356/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab12884299" class="tabcontent ">
                    <h2>CCR3 functional responses are regulated by both CXCR3 and its ligands CXCL9, CXCL10 and CXCL11.</h2>
                    <h3>European journal of immunology; 2003</h3>
                    <p>The chemokine receptor CXCR3 is predominantly expressed on T lymphocytes, and its agonists CXCL9, CXCL10 and CXCL11 are IFN-gamma-inducible chemokines that promote Th1 responses. In contrast, the CCR3 agonists CCL11, CCL24 and CCL26 are involved in the recruitment of cells such as eosinophils and basophils during Th2 responses. Here, we report that although CCL11, CCL24 and CCL26 are neither agonists nor antagonists of CXCR3, CCL11 binds with high affinity to CXCR3. This suggests that, in vivo, CXCR3 may act as a decoy receptor, sequestering locally produced CCL11. We also demonstrate that the CXCR3 ligands inhibit CCR3-mediated functional responses of both human eosinophils and CCR3 transfectants induced by all three eotaxins, with CXCL11 being the most efficacious antagonist. The examination of CCR3-CCR1 chimeric constructs revealed that CCL11 and CXCL11 share overlapping binding sites contained within the CCR3 extracellular loops, a region that was previously shown to be essential for effective receptor-activation. Hence, eosinophil responses mediated by chemokines acting at CCR3 may be regulated by two distinct mechanisms: the antagonistic effects of CXCR3 ligands and the sequestration of CCL11 by CXCR3-expressing cells. Such interplay may serve to finely tune inflammatory responses in vivo.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/12884299/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab9660793" class="tabcontent ">
                    <h2>Binding and functional properties of recombinant and endogenous CXCR3 chemokine receptors.</h2>
                    <h3>The Journal of biological chemistry; 1998</h3>
                    <p>IP10 and MIG are two members of the CXC branch of the chemokine superfamily whose expression is dramatically up-regulated by interferon (IFN)-gamma. The proteins act largely on natural killer (NK)-cells and activated T-cells and have been implicated in mediating some of the effects of IFN-gamma and lipopolysaccharides (LPSs), as well as T-cell-dependent anti-tumor responses. Recently both chemokines have been shown to be functional agonists of the same G-protein-coupled receptor, CXCR3. We now report the pharmacological characterization of CXCR3 and find that, when heterologously expressed, CXCR3 binds IP10 and MIG with Ki values of 0.14 and 4.9 nM, respectively. The receptor has very modest affinity for SDF-1alpha and little or no affinity for other CXC-chemokines. The properties of the endogenous receptor expressed on activated T-cells are similar. Surprisingly, several CC-chemokines, particularly eotaxin and MCP-4, also compete with moderate affinity for the binding of IP10 to CXCR3. Eotaxin does not activate CXCR3 but, in CXCR3-transfected cells, can block IP10-mediated receptor activation. Eotaxin, therefore, may be a natural CXCR3 antagonist.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/9660793/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab15761110" class="tabcontent ">
                    <h2>Pharmacological characterization of CXC chemokine receptor 3 ligands and a small molecule antagonist.</h2>
                    <h3>The Journal of pharmacology and experimental therapeutics; 2005</h3>
                    <p>The CXC chemokine receptor 3 (CXCR3) is predominantly expressed on T helper type 1 (Th1) cells that are involved in inflammatory diseases. The three CXCR3 ligands CXCL9, CXCL10, and CXCL11 are produced at sites of inflammation and elicit migration of pathological Th1 cells. Here, we are the first to characterize the pharmacological potencies and specificity of a CXCR3 antagonist, N-1R-[3-(4-ethoxy-phenyl)-4-oxo-3,4-dihydro-pyrido[2,3-d]pyrimidin-2-yl]-ethyl-N-pyridin-3-ylmethyl-2-(4-fluoro-3-trifluoromethyl-phenyl)-acetamide (NBI-74330), from the T487 small molecule series. NBI-74330 demonstrated potent inhibition of [(125)I]CXCL10 and [(125)I]CXCL11 specific binding (K(i) of 1.5 and 3.2 nM, respectively) and of functional responses mediated by CXCR3, such as ligand-induced guanosine 5'-O-(3-[(35)S]thio)triphosphate ([(35)S]GTPgammaS) binding, calcium mobilization, and cellular chemotaxis (IC(50) of 7 to 18 nM). NBI-74330 was selective for CXCR3 because it showed no significant inhibition of chemotactic responses to other chemokines and did not inhibit radioligand binding to a panel of nonchemokine G-protein coupled receptors. There was a striking difference in potencies among the three CXCR3 ligands, with CXCL11 >> CXCL10 > CXCL9. A comparison of the rank order of K(i) values with the rank order of monocyte production levels of these three ligands revealed a precise inverse correlation, suggesting that the weaker receptor affinities of CXCL9 and CXCL10 were physiologically compensated for by an elevated expression, perhaps to maintain effectiveness of each ligand under physiological conditions.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/15761110/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab12571234" class="tabcontent ">
                    <h2>CXCR3 and heparin binding sites of the chemokine IP-10 (CXCL10).</h2>
                    <h3>The Journal of biological chemistry; 2003</h3>
                    <p>The chemokine IP-10 (interferon-inducible protein of 10 kDa, CXCL10) binds the G protein-coupled receptor CXCR3, which is found mainly on activated T cells and NK cells, and plays an important role in Th1-type inflammatory diseases. IP-10 also binds to glycosaminoglycans (GAGs), an interaction thought to be important for its sequestration on endothelial and other cells. In this study, we performed an extensive mutational analysis to identify the CXCR3 and heparin binding sites of murine IP-10. The mutants were characterized for heparin binding, CXCR3 binding, and the ability to induce chemotaxis, Ca(2+) flux, and CXCR3 internalization. Double mutations neutralizing adjacent basic residues at the C terminus did not lead to a significant reduction in heparin binding, indicating that the main heparin binding site of IP-10 is not along the C-terminal alpha helix. Alanine exchange of Arg-22 had the largest effect on heparin binding, with residues Arg-20, Ile-24, Lys-26, Lys-46, and Lys-47 further contributing to heparin binding. A charge change mutation of Arg-22 resulted in further reduction in heparin binding. The N-terminal residue Arg-8, preceding the first cysteine, was critical for CXCR3 signaling. Mutations of charged and uncharged residues in the loop regions of residues 20-24 and 46-47, which caused reduced heparin binding, also resulted in reduced CXCR3 binding and signaling. CXCR3 expressing GAG-deficient Chinese hamster ovary cells revealed that GAG binding was not required for IP-10 binding and signaling through CXCR3, which suggests that the CXCR3 and heparin binding sites of IP-10 are partially overlapping.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/12571234/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab10233762" class="tabcontent ">
                    <h2>Human IP-9: A keratinocyte-derived high affinity CXC-chemokine ligand for the IP-10/Mig receptor (CXCR3).</h2>
                    <h3>The Journal of investigative dermatology; 1999</h3>
                    <p>Chemokines and their receptors play a crucial part in the recruitment of leukocytes into inflammatory sites. The CXC chemokines IP-10 and Mig are selective attractants for activated (memory) T cells, the predominant cell type in skin infiltrates in many inflammatory dermatoses. The selectivity for activated T cells can be explained by the fact that both chemokines exert their effects through a common receptor, CXCR3, which is nearly exclusively expressed on activated T cells. The aim of this study was to identify biologically active CXCR3 ligands produced by keratinocytes. To that end, Chinese hamster ovary cells expressing a cDNA encoding CXCR3 were challenged with proteins obtained from interferon-gamma stimulated keratinocytes and subsequently monitored for effects on second messenger systems. By this approach we were able to isolate IP-10 and Mig, and in addition identified a novel highly potent ligand for the CXCR3 receptor, designated interferon-gamma-inducible protein-9, which proved to be chemotactic for activated T cells expressing CXCR3. Protein sequence and mass spectrometric analysis followed by molecular cloning of the cDNA encoding interferon-gamma-inducible protein-9, revealed that interferon-gamma-inducible protein-9 is a CXC chemokine with a molecular mass of 8303 Da. From a GenBank database query it became clear that interferon-gamma-inducible protein-9 is in fact the protein encoded by the cDNA sequence also known as beta-R1, H174 or I-TAC. In situ hybridization experiments showed that interferon-gamma-inducible protein-9 mRNA is expressed by basal layer keratinocytes in a variety of skin disorders, including allergic contact dermatitis, lichen planus, and mycosis fungoides suggesting a functional role for this chemokine in skin immune responses.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/10233762/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                

    <footer>
        <span>© Copyright 2024, YCU Bioinformatics Laboratory</span>
        <span>Page built with <a href="https://quarto.org/">Quarto</a> by バイオインフォマティクス解析センター</span>
    </footer>
</body>
</html>
