<!DOCTYPE html>
<html>
<meta charset="UTF-8">
<head>
    <title>CXCL11——CXCR3 - PMID Details</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
            color: #333;
        }

        h1 {
            text-align: center;
            font-size: 2.5em;
            color: #2c3e50;
            margin-bottom: 20px;
        }

        /* Navbar Styles */
        nav {
            background-color: #2c3e50;
            padding: 10px 20px;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            color: white;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        nav .navbar-left, nav .navbar-right {
            display: flex;
            align-items: center;
        }

        nav a {
            color: white;
            text-decoration: none;
            margin: 0 15px;
            font-size: 1.1em;
            transition: color 0.3s ease;
        }

        nav a:hover {
            color: #f39c12;
        }

        nav .divider {
            color: #ecf0f1;
            margin: 0 10px;
        }

        /* Page Footer */
        footer {
            margin-top: 40px;
            padding: 10px 20px;
            background-color: #f7f7f7;
            color: #333;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        /* Tab and Content Styles */
        .tab {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin-bottom: 20px;
        }

        .tab button {
            padding: 10px 20px;
            margin: 5px;
            margin-bottom: 10px;
            cursor: pointer;
            background-color: #f4f4f4;
            border: 2px solid #ddd;
            border-radius: 5px;
            font-size: 1.1em;
            transition: background-color 0.3s ease;
            flex-basis: calc(10% - 10px);
            box-sizing: border-box;
        }

        .tab button:hover {
            background-color: #ecf0f1;
        }

        .tab button.active {
            background-color: #3498db;
            color: white;
            border-color: #2980b9;
        }

        .tabcontent {
            display: none;
            padding: 20px;
            margin-top: 10px;
            background-color: #ecf0f1;
            border-radius: 5px;
            box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
        }

        .tabcontent.active {
            display: block;
        }

        .tabcontent h2 {
            font-size: 2em;
            color: #2980b9;
        }

        .tabcontent p {
            font-size: 1.2em;
            line-height: 1.8;
        }

        /* Emphasize Title and Abstract */
        .title, .abstract {
            background-color: #f9f9f9;
            padding: 20px;
            margin: 10px 0;
            border-left: 5px solid #2980b9;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
        }

        .abstract {
            border-left: 5px solid #e74c3c;
        }
    </style>
    <script>
        function adjustTabStyle() {
            var tabCount = document.querySelectorAll('.tab button').length;
            var tabContainer = document.querySelector('.tab');

            if (tabCount > 10) {
                tabContainer.classList.add('small-tabs');
            }
        }

        function openTab(evt, tabId) {
            document.querySelectorAll(".tabcontent").forEach(tab => tab.style.display = "none");
            document.querySelectorAll(".tablinks").forEach(link => link.classList.remove("active"));
            document.getElementById(tabId).style.display = "block";
            evt.currentTarget.classList.add("active");
        }

        window.onload = adjustTabStyle;
    </script>
</head>
<body>
    <nav>
        <div class="navbar-left">
            <a href="">ConnectomeDB</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/">Home</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database.html">Database</a>
        </div>
        <div class="navbar-right">
            <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/">YCU Bioinformatics Lab</a>
            <span class="divider">|</span>
            <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/">Perkins Systems Biology & Genomics Lab</a>
        </div>
    </nav>

    <h1>PMID Details for CXCL11——CXCR3</h1>

    <div class="tab">
        <button class="tablinks active" onclick="openTab(event, 'tab12884299')">12884299</button><button class="tablinks " onclick="openTab(event, 'tab15761110')">15761110</button><button class="tablinks " onclick="openTab(event, 'tab15273303')">15273303</button><button class="tablinks " onclick="openTab(event, 'tab9625760')">9625760</button>
    </div>

    <div class="tabcontent" id="titleTab">
        <div class="title">
            <h2>Title</h2>
            <p>{{TITLE}}</p>
            <i> Journal:{{JOURNAL}}; Year Published:{{YEAR}}</i>   
        </div>
    </div>

    <div class="tabcontent" id="abstractTab">
        <div class="abstract">
            <h2>Abstract</h2>
            <p>{{ABSTRACT}}</p>
        </div>
    </div>

    
                <div id="tab12884299" class="tabcontent active">
                    <h2>CCR3 functional responses are regulated by both CXCR3 and its ligands CXCL9, CXCL10 and CXCL11.</h2>
                    <h3>European journal of immunology; 2003</h3>
                    <p>The chemokine receptor CXCR3 is predominantly expressed on T lymphocytes, and its agonists CXCL9, CXCL10 and CXCL11 are IFN-gamma-inducible chemokines that promote Th1 responses. In contrast, the CCR3 agonists CCL11, CCL24 and CCL26 are involved in the recruitment of cells such as eosinophils and basophils during Th2 responses. Here, we report that although CCL11, CCL24 and CCL26 are neither agonists nor antagonists of CXCR3, CCL11 binds with high affinity to CXCR3. This suggests that, in vivo, CXCR3 may act as a decoy receptor, sequestering locally produced CCL11. We also demonstrate that the CXCR3 ligands inhibit CCR3-mediated functional responses of both human eosinophils and CCR3 transfectants induced by all three eotaxins, with CXCL11 being the most efficacious antagonist. The examination of CCR3-CCR1 chimeric constructs revealed that CCL11 and CXCL11 share overlapping binding sites contained within the CCR3 extracellular loops, a region that was previously shown to be essential for effective receptor-activation. Hence, eosinophil responses mediated by chemokines acting at CCR3 may be regulated by two distinct mechanisms: the antagonistic effects of CXCR3 ligands and the sequestration of CCL11 by CXCR3-expressing cells. Such interplay may serve to finely tune inflammatory responses in vivo.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/12884299/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab15761110" class="tabcontent ">
                    <h2>Pharmacological characterization of CXC chemokine receptor 3 ligands and a small molecule antagonist.</h2>
                    <h3>The Journal of pharmacology and experimental therapeutics; 2005</h3>
                    <p>The CXC chemokine receptor 3 (CXCR3) is predominantly expressed on T helper type 1 (Th1) cells that are involved in inflammatory diseases. The three CXCR3 ligands CXCL9, CXCL10, and CXCL11 are produced at sites of inflammation and elicit migration of pathological Th1 cells. Here, we are the first to characterize the pharmacological potencies and specificity of a CXCR3 antagonist, N-1R-[3-(4-ethoxy-phenyl)-4-oxo-3,4-dihydro-pyrido[2,3-d]pyrimidin-2-yl]-ethyl-N-pyridin-3-ylmethyl-2-(4-fluoro-3-trifluoromethyl-phenyl)-acetamide (NBI-74330), from the T487 small molecule series. NBI-74330 demonstrated potent inhibition of [(125)I]CXCL10 and [(125)I]CXCL11 specific binding (K(i) of 1.5 and 3.2 nM, respectively) and of functional responses mediated by CXCR3, such as ligand-induced guanosine 5'-O-(3-[(35)S]thio)triphosphate ([(35)S]GTPgammaS) binding, calcium mobilization, and cellular chemotaxis (IC(50) of 7 to 18 nM). NBI-74330 was selective for CXCR3 because it showed no significant inhibition of chemotactic responses to other chemokines and did not inhibit radioligand binding to a panel of nonchemokine G-protein coupled receptors. There was a striking difference in potencies among the three CXCR3 ligands, with CXCL11 >> CXCL10 > CXCL9. A comparison of the rank order of K(i) values with the rank order of monocyte production levels of these three ligands revealed a precise inverse correlation, suggesting that the weaker receptor affinities of CXCL9 and CXCL10 were physiologically compensated for by an elevated expression, perhaps to maintain effectiveness of each ligand under physiological conditions.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/15761110/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab15273303" class="tabcontent ">
                    <h2>NMR structure of CXCR3 binding chemokine CXCL11 (ITAC).</h2>
                    <h3>Protein science : a publication of the Protein Society; 2004</h3>
                    <p>CXCL11 (ITAC) is one of three chemokines known to bind the receptor CXCR3, the two others being CXCL9 (Mig) and CXCL10 (IP-10). CXCL11 differs from the other CXCR3 ligands in both the strength and the particularities of its receptor interactions: It has a higher affinity, is a stronger agonist, and behaves differently when critical N-terminal residues are deleted. The structure of CXCL11 was determined using solution NMR to allow comparison with that of CXCL10 and help elucidate the source of the differences. CXCL11 takes on the canonical chemokine fold but exhibits greater conformational flexibility than has been observed for related chemokines under the same sample conditions. Unlike related chemokines such as IP-10 and IL-8, ITAC does not appear to form dimers at millimolar concentrations. The origin for this behavior can be found in the solution structure, which indicates a beta-bulge in beta-strand 1 that distorts the dimerization interface used by other CXC chemokines.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/15273303/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab9625760" class="tabcontent ">
                    <h2>Interferon-inducible T cell alpha chemoattractant (I-TAC): a novel non-ELR CXC chemokine with potent activity on activated T cells through selective high affinity binding to CXCR3.</h2>
                    <h3>The Journal of experimental medicine; 1998</h3>
                    <p>Chemokines are essential mediators of normal leukocyte trafficking as well as of leukocyte recruitment during inflammation. We describe here a novel non-ELR CXC chemokine identified through sequence analysis of cDNAs derived from cytokine-activated primary human astrocytes. This novel chemokine, referred to as I-TAC (interferon-inducible T cell alpha chemoattractant), is regulated by interferon (IFN) and has potent chemoattractant activity for interleukin (IL)-2-activated T cells, but not for freshly isolated unstimulated T cells, neutrophils, or monocytes. I-TAC interacts selectively with CXCR3, which is the receptor for two other IFN-inducible chemokines, the IFN-gamma-inducible 10-kD protein (IP-10) and IFN-gamma- induced human monokine (HuMig), but with a significantly higher affinity. In addition, higher potency and efficacy of I-TAC over IP-10 and HuMig is demonstrated by transient mobilization of intracellular calcium as well as chemotactic migration in both activated T cells and transfected cell lines expressing CXCR3. Stimulation of astrocytes with IFN-gamma and IL-1 together results in an approximately 400,000-fold increase in I-TAC mRNA expression, whereas stimulating monocytes with either of the cytokines alone or in combination results in only a 100-fold increase in the level of I-TAC transcript. Moderate expression is also observed in pancreas, lung, thymus, and spleen. The high level of expression in IFN- and IL-1-stimulated astrocytes suggests that I-TAC could be a major chemoattractant for effector T cells involved in the pathophysiology of neuroinflammatory disorders, although I-TAC may also play a role in the migration of activated T cells during IFN-dominated immune responses.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/9625760/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                

    <footer>
        <span>© Copyright 2024, YCU Bioinformatics Laboratory</span>
        <span>Page built with <a href="https://quarto.org/">Quarto</a> by バイオインフォマティクス解析センター</span>
    </footer>
</body>
</html>
