<!DOCTYPE html>
<html>
<meta charset="UTF-8">
<head>
    <title>EGF——EGFR - PMID Details</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
            color: #333;
        }

        h1 {
            text-align: center;
            font-size: 2.5em;
            color: #2c3e50;
            margin-bottom: 20px;
        }

        /* Navbar Styles */
        nav {
            background-color: #2c3e50;
            padding: 10px 20px;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            color: white;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        nav .navbar-left, nav .navbar-right {
            display: flex;
            align-items: center;
        }

        nav a {
            color: white;
            text-decoration: none;
            margin: 0 15px;
            font-size: 1.1em;
            transition: color 0.3s ease;
        }

        nav a:hover {
            color: #f39c12;
        }

        nav .divider {
            color: #ecf0f1;
            margin: 0 10px;
        }

        /* Page Footer */
        footer {
            margin-top: 40px;
            padding: 10px 20px;
            background-color: #f7f7f7;
            color: #333;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        /* Tab and Content Styles */
        .tab {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin-bottom: 20px;
        }

        .tab button {
            padding: 10px 20px;
            margin: 5px;
            margin-bottom: 10px;
            cursor: pointer;
            background-color: #f4f4f4;
            border: 2px solid #ddd;
            border-radius: 5px;
            font-size: 1.1em;
            transition: background-color 0.3s ease;
            flex-basis: calc(10% - 10px);
            box-sizing: border-box;
        }

        .tab button:hover {
            background-color: #ecf0f1;
        }

        .tab button.active {
            background-color: #3498db;
            color: white;
            border-color: #2980b9;
        }

        .tabcontent {
            display: none;
            padding: 20px;
            margin-top: 10px;
            background-color: #ecf0f1;
            border-radius: 5px;
            box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
        }

        .tabcontent.active {
            display: block;
        }

        .tabcontent h2 {
            font-size: 2em;
            color: #2980b9;
        }

        .tabcontent p {
            font-size: 1.2em;
            line-height: 1.8;
        }

        /* Emphasize Title and Abstract */
        .title, .abstract {
            background-color: #f9f9f9;
            padding: 20px;
            margin: 10px 0;
            border-left: 5px solid #2980b9;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
        }

        .abstract {
            border-left: 5px solid #e74c3c;
        }
    </style>
    <script>
        function adjustTabStyle() {
            var tabCount = document.querySelectorAll('.tab button').length;
            var tabContainer = document.querySelector('.tab');

            if (tabCount > 10) {
                tabContainer.classList.add('small-tabs');
            }
        }

        function openTab(evt, tabId) {
            document.querySelectorAll(".tabcontent").forEach(tab => tab.style.display = "none");
            document.querySelectorAll(".tablinks").forEach(link => link.classList.remove("active"));
            document.getElementById(tabId).style.display = "block";
            evt.currentTarget.classList.add("active");
        }

        window.onload = adjustTabStyle;
    </script>
</head>
<body>
    <nav>
        <div class="navbar-left">
            <a href="">ConnectomeDB</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/">Home</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database.html">Database</a>
        </div>
        <div class="navbar-right">
            <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/">YCU Bioinformatics Lab</a>
            <span class="divider">|</span>
            <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/">Perkins Systems Biology & Genomics Lab</a>
        </div>
    </nav>

    <h1>PMID Details for EGF——EGFR</h1>

    <div class="tab">
        <button class="tablinks active" onclick="openTab(event, 'tab6289330')">6289330</button><button class="tablinks " onclick="openTab(event, 'tab12093292')">12093292</button><button class="tablinks " onclick="openTab(event, 'tab12620237')">12620237</button><button class="tablinks " onclick="openTab(event, 'tab10788520')">10788520</button><button class="tablinks " onclick="openTab(event, 'tab16274239')">16274239</button><button class="tablinks " onclick="openTab(event, 'tab15620700')">15620700</button><button class="tablinks " onclick="openTab(event, 'tab12297050')">12297050</button>
    </div>

    <div class="tabcontent" id="titleTab">
        <div class="title">
            <h2>Title</h2>
            <p>{{TITLE}}</p>
            <i> Journal:{{JOURNAL}}; Year Published:{{YEAR}}</i>   
        </div>
    </div>

    <div class="tabcontent" id="abstractTab">
        <div class="abstract">
            <h2>Abstract</h2>
            <p>{{ABSTRACT}}</p>
        </div>
    </div>

    
                <div id="tab6289330" class="tabcontent active">
                    <h2>Identity of human epidermal growth factor (EGF) receptor with glycoprotein SA-7: evidence for differential phosphorylation of the two components of the EGF receptor from A431 cells.</h2>
                    <h3>Proceedings of the National Academy of Sciences of the United States of America; 1982</h3>
                    <p>A 165-kilodalton (kDal) surface glycoprotein encoded by human chromosome 7 (SA-7) had been characterized by using antisera raised against human chromosome 7-containing somatic cell hybrids. We now present evidence that SA-7 is the human receptor for epidermal growth factor (EGF) and that these antisera recognize human-specific determinants. The gene coding for the human EGF receptor is localized to the p12 to p22 region of chromosome 7. We have characterized the 145-kDal/165-kDal EGF receptor doublet of A431 cells after immunoprecipitation of radiolabeled cell extracts with these antisera. We find that a protein with endogenous kinase activity copurifies with the A431 receptor doublet and that both components of the doublet contain phosphotyrosine and phosphothreonine and the 165-kDal component contains phosphoserine as well. Further, although each component of the receptor doublet has an average pI of 7, both display charge heterogeneity and appear to have unique charge isomers. The relationship between the two components of the A431 EGF receptor is discussed.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/6289330/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab12093292" class="tabcontent ">
                    <h2>Role of the N-terminus of epidermal growth factor in ErbB-2/ErbB-3 binding studied by phage display.</h2>
                    <h3>Biochemistry; 2002</h3>
                    <p>Epidermal growth factor (EGF) binds with high affinity to the EGF receptor, also known as ErbB-1, but upon replacement of the N-terminal linear region by neuregulin (NRG) 1 or transforming growth factor (TGF) alpha sequences it gains in addition high affinity for ErbB-2/ErbB-3 heterodimers. However, these chimeras weakly bind to ErbB-3 alone. To further dissect the ligand binding selectivity of the ErbB network, we have applied the phage display technique to examine the role of the linear N-terminal region in EGF for interaction with ErbB-2/ErbB-3 heterodimers. A library of EGF variants was constructed in which residues 2, 3, and 4 were randomly mutated, followed by selection for binding to intact MDA-MB-453 cells that overexpress ErbB-2 and ErbB-3 but lack ErbB-1. Analysis of the selected phage EGF variants revealed clones with high binding affinity to ErbB-2/ErbB-3 while maintaining high affinity to ErbB-1. In these variants, Trp (or alternatively His) was almost exclusively present at position 2, while specific combinations of hydrophobic, basic, and small residues were found at positions 3 and 4. The mitogenic activity of the phage EGF variants corresponded with their relative binding affinity. Two of the selected EGF variants, EGF/WVS and EGF/WRS, were further characterized as recombinant proteins. In contrast to previously characterized chimeras of EGF with NRG-1 or TGF-alpha, these variants did not only show high binding affinity for ErbB-2/ErbB-3 heterodimers but also for ErbB-3 alone. These data show that the linear N-terminal region of EGF-like growth factors is directly involved in binding to ErbB-3.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/12093292/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab12620237" class="tabcontent ">
                    <h2>EGF activates its receptor by removing interactions that autoinhibit ectodomain dimerization.</h2>
                    <h3>Molecular cell; 2003</h3>
                    <p>Epidermal growth factor (EGF) receptor is the prototype of the ErbB (HER) family receptor tyrosine kinases (RTKs), which regulate cell growth and differentiation and are implicated in many human cancers. EGF activates its receptor by inducing dimerization of the 621 amino acid EGF receptor extracellular region. We describe the 2.8 A resolution crystal structure of this entire extracellular region (sEGFR) in an unactivated state. The structure reveals an autoinhibited configuration, where the dimerization interface recently identified in activated sEGFR structures is completely occluded by intramolecular interactions. To activate the receptor, EGF binding must promote a large domain rearrangement that exposes this dimerization interface. This contrasts starkly with other RTK activation mechanisms and suggests new approaches for designing ErbB receptor antagonists.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/12620237/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab10788520" class="tabcontent ">
                    <h2>Polyubiquitination of the epidermal growth factor receptor occurs at the plasma membrane upon ligand-induced activation.</h2>
                    <h3>The Journal of biological chemistry; 2000</h3>
                    <p>We have previously shown that, although overexpression of mutant dynamin inhibits clathrin-dependent endocytosis and disrupts high affinity binding of epidermal growth factor (EGF) to the EGF receptor (EGFR), it does not inhibit ligand-induced translocation of the EGFR into clathrin-coated pits. In the present study, we demonstrate that, upon ligand binding and incubation at 37 degrees C, the EGFR was polyubiquitinated regardless of overexpression of mutant dynamin. In cells not overexpressing mutant dynamin, the EGFR was rapidly internalized and deubiquitinated. In cells being endocytosis-deficient, due to overexpression of mutant dynamin, however, the EGFR was upon prolonged chase first found in deeply invaginated coated pits, and then eventually moved out of the coated pits and back onto the smooth plasma membrane. Polyubiquitination occurred equally efficiently in cells with or without intact clathrin-dependent endocytosis, while the kinetics of ubiquitination and deubiquitination was somewhat different. We further found that the EGF-induced ubiquitination of Eps15 occurred both in the absence and presence of endocytosis with the same kinetics as polyubiquitination of the EGFR, but that the EGF-induced monoubiquitination of Eps15 was somewhat reduced upon overexpression of mutant dynamin. Our data show that EGF-induced polyubiquitination of the EGFR occurs at the plasma membrane.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/10788520/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab16274239" class="tabcontent ">
                    <h2>Functional effects of glycosylation at Asn-579 of the epidermal growth factor receptor.</h2>
                    <h3>Biochemistry; 2005</h3>
                    <p>We have investigated functional effects of glycosylation at N(579) of the epidermal growth factor receptor (EGFR). Our previous study showed that the population of cell-surface expressed EGFRs in A431 cells, a human epidermoid carcinoma cell line, is composed of two subpopulations that differ by glycosylation at N(579) [Zhen et al. (2003) Biochemistry 42, 5478-5492]. To characterize the subpopulation of receptors not glycosylated at N(579), we established a 32D cell line expressing a point mutant of the EGFR (N579Q), which cannot be glycosylated at this position. Analysis of epitope accessibility suggests that the lack of glycosylation at N(579) weakens auto-inhibitory tether interactions, and cross-linking experiments suggest a somewhat elevated level of preformed N579Q-EGFR dimers in the absence of ligand relative to wild-type EGFR (WT-EGFR). However, ligand drives the majority of N579Q-EGFR dimerization, suggesting that untethering, while necessary, is not sufficient to drive dimerization. Ligand-binding experiments reveal a much greater fraction of N579Q-EGFRs in a high-affinity state compared to the fraction of WT-EGFRs in a high-affinity state. However, differences in the kinetic association and dissociation rates indicate that the high-affinity states of the WT and the N579Q receptors are distinct. EGF-stimulated phosphorylation in cells expressing N579Q-EGFRs results in notable differences in the pattern of tyrosine phosphorylated proteins compared with that obtained in cells expressing WT-EGFRs. Moreover, although WT-EGFRs confer cell survival in 32D cells in the absence of interleukin-3 and EGF, we found that receptors lacking glycosylation at N(579) do not. This is the first study of which we are aware to show that selective glycosylation of a specific N-glycosylation site can produce two functionally distinct receptors.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/16274239/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab15620700" class="tabcontent ">
                    <h2>Activation of tyrosine kinase of EGFR induces Gbetagamma-dependent GRK-EGFR complex formation.</h2>
                    <h3>FEBS letters; 2005</h3>
                    <p>This study demonstrated that activation of tyrosine kinase of epidermal growth factor receptor (EGFR) induces its association with G protein-coupled receptor kinase 2 (GRK2). Immunoprecipitation experiments showed that EGF stimulation increased GRK2 binding to EGFR complex in HEK293 cells coexpressing EGFR and GRK2. The EGF-induced GRK2-EGFR complex formation was greatly reduced by perturbation of EGFR and Src tyrosine kinase activity. Furthermore, studies with GRK2 mutants showed that neither catalytic activity nor the N-terminal domain of GRK2 was required for EGF-induced GRK2-EGFR complex formation. However, overexpression of Gbetagamma scavengers blocked EGF-induced formation of GRK2-EGFR complex.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/15620700/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab12297050" class="tabcontent ">
                    <h2>Crystal structure of the complex of human epidermal growth factor and receptor extracellular domains.</h2>
                    <h3>Cell; 2002</h3>
                    <p>Epidermal growth factor (EGF) regulates cell proliferation and differentiation by binding to the EGF receptor (EGFR) extracellular region, comprising domains I-IV, with the resultant dimerization of the receptor tyrosine kinase. In this study, the crystal structure of a 2:2 complex of human EGF and the EGFR extracellular region has been determined at 3.3 A resolution. EGFR domains I-III are arranged in a C shape, and EGF is docked between domains I and III. The 1:1 EGF*EGFR complex dimerizes through a direct receptor*receptor interaction, in which a protruding beta-hairpin arm of each domain II holds the body of the other. The unique "receptor-mediated dimerization" was verified by EGFR mutagenesis.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/12297050/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                

    <footer>
        <span>© Copyright 2024, YCU Bioinformatics Laboratory</span>
        <span>Page built with <a href="https://quarto.org/">Quarto</a> by バイオインフォマティクス解析センター</span>
    </footer>
</body>
</html>
