<!DOCTYPE html>
<html>
<meta charset="UTF-8">
<head>
    <title>F2——F2R - PMID Details</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
            color: #333;
        }

        h1 {
            text-align: center;
            font-size: 2.5em;
            color: #2c3e50;
            margin-bottom: 20px;
        }

        /* Navbar Styles */
        nav {
            background-color: #2c3e50;
            padding: 10px 20px;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            color: white;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        nav .navbar-left, nav .navbar-right {
            display: flex;
            align-items: center;
        }

        nav a {
            color: white;
            text-decoration: none;
            margin: 0 15px;
            font-size: 1.1em;
            transition: color 0.3s ease;
        }

        nav a:hover {
            color: #f39c12;
        }

        nav .divider {
            color: #ecf0f1;
            margin: 0 10px;
        }

        /* Page Footer */
        footer {
            margin-top: 40px;
            padding: 10px 20px;
            background-color: #f7f7f7;
            color: #333;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        /* Tab and Content Styles */
        .tab {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin-bottom: 20px;
        }

        .tab button {
            padding: 10px 20px;
            margin: 5px;
            margin-bottom: 10px;
            cursor: pointer;
            background-color: #f4f4f4;
            border: 2px solid #ddd;
            border-radius: 5px;
            font-size: 1.1em;
            transition: background-color 0.3s ease;
            flex-basis: calc(10% - 10px);
            box-sizing: border-box;
        }

        .tab button:hover {
            background-color: #ecf0f1;
        }

        .tab button.active {
            background-color: #3498db;
            color: white;
            border-color: #2980b9;
        }

        .tabcontent {
            display: none;
            padding: 20px;
            margin-top: 10px;
            background-color: #ecf0f1;
            border-radius: 5px;
            box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
        }

        .tabcontent.active {
            display: block;
        }

        .tabcontent h2 {
            font-size: 2em;
            color: #2980b9;
        }

        .tabcontent p {
            font-size: 1.2em;
            line-height: 1.8;
        }

        /* Emphasize Title and Abstract */
        .title, .abstract {
            background-color: #f9f9f9;
            padding: 20px;
            margin: 10px 0;
            border-left: 5px solid #2980b9;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
        }

        .abstract {
            border-left: 5px solid #e74c3c;
        }
    </style>
    <script>
        function adjustTabStyle() {
            var tabCount = document.querySelectorAll('.tab button').length;
            var tabContainer = document.querySelector('.tab');

            if (tabCount > 10) {
                tabContainer.classList.add('small-tabs');
            }
        }

        function openTab(evt, tabId) {
            document.querySelectorAll(".tabcontent").forEach(tab => tab.style.display = "none");
            document.querySelectorAll(".tablinks").forEach(link => link.classList.remove("active"));
            document.getElementById(tabId).style.display = "block";
            evt.currentTarget.classList.add("active");
        }

        window.onload = adjustTabStyle;
    </script>
</head>
<body>
    <nav>
        <div class="navbar-left">
            <a href="">ConnectomeDB</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/">Home</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database.html">Database</a>
        </div>
        <div class="navbar-right">
            <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/">YCU Bioinformatics Lab</a>
            <span class="divider">|</span>
            <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/">Perkins Systems Biology & Genomics Lab</a>
        </div>
    </nav>

    <h1>PMID Details for F2——F2R</h1>

    <div class="tab">
        <button class="tablinks active" onclick="openTab(event, 'tab11356985')">11356985</button><button class="tablinks " onclick="openTab(event, 'tab10644723')">10644723</button><button class="tablinks " onclick="openTab(event, 'tab12563219')">12563219</button><button class="tablinks " onclick="openTab(event, 'tab9058715')">9058715</button><button class="tablinks " onclick="openTab(event, 'tab10978167')">10978167</button>
    </div>

    <div class="tabcontent" id="titleTab">
        <div class="title">
            <h2>Title</h2>
            <p>{{TITLE}}</p>
            <i> Journal:{{JOURNAL}}; Year Published:{{YEAR}}</i>   
        </div>
    </div>

    <div class="tabcontent" id="abstractTab">
        <div class="abstract">
            <h2>Abstract</h2>
            <p>{{ABSTRACT}}</p>
        </div>
    </div>

    
                <div id="tab11356985" class="tabcontent active">
                    <h2>Proteinase-activated receptors.</h2>
                    <h3>Pharmacological reviews; 2001</h3>
                    <p>Proteinase-activated receptors are a recently described, novel family of seven-transmembrane G-protein-coupled receptors. Rather then being stimulated through ligand receptor occupancy, activation is initiated by cleavage of the N terminus of the receptor by a serine protease resulting in the generation of a new tethered ligand that interacts with the receptor within extracellular loop-2. To date, four proteinase-activated receptors (PARs) have been identified, with distinct N-terminal cleavage sites and tethered ligand pharmacology. In addition to the progress in the generation of PAR-1 antagonists, we describe the role of thrombin in such processes as wound healing and the evidence implicating PAR-1 in vascular disorders and cancer. We also identify advances in the understanding of PAR-1-mediated intracellular signaling and receptor desensitization. The cellular functions, signaling events, and desensitization processes involved in PAR-2 activation are also assessed. However, other major aspects of PAR-2 are highlighted, in particular the ability of several serine protease enzymes, in addition to trypsin, to function as activators of PAR-2. The likely physiological and pathophysiological roles for PAR-2 in skin, intestine, blood vessels, and the peripheral nervous system are considered in the context of PAR-2 activation by multiple serine proteases. The recent discovery of PAR-3 and PAR-4 as additional thrombin-sensitive PARs further highlights the complexity in assessing the effects of thrombin in several different systems, an issue that remains to be fully addressed. These discoveries have also highlighted possible PAR-PAR interactions at both functional and molecular levels. The future identification of other PARs and their modes of activation are an important future direction for this expanding field of study.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/11356985/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab10644723" class="tabcontent ">
                    <h2>PAR1 thrombin receptor-G protein interactions. Separation of binding and coupling determinants in the galpha subunit.</h2>
                    <h3>The Journal of biological chemistry; 2000</h3>
                    <p>Signal transfer between the protease-activated PAR1 thrombin receptor and membrane-associated heterotrimeric G proteins is mediated by protein-protein interactions. We constructed a yeast signaling system that resolves domain-specific functions of binding from coupling in the Galpha subunit. The endogenous yeast Galpha subunit, Gpa1, does not bind to PAR1 and served as a null structural template. N- and C-terminal portions of mammalian G(i2) and G(16) were substituted back into the Gpa1 template and gain-of-function assessed. The C-terminal third of G(16), but not of G(i2), provides sufficient interactions for coupling to occur with PAR1. The N-terminal two-thirds of G(i2) also contains sufficient determinants to bind and couple to PAR1 and overcome the otherwise negative or missing interactions supplied by the C-terminal third of Gpa1. Replacement of the N-terminal alpha-helix of G(i2), residues 1-34, with those of Gpa1 abolishes coupling but not binding to PAR1 or to betagamma subunits. These data support a model that the N-terminal alphaN helix of the Galpha subunit is physically interposed between PAR1 and the Gbeta subunit and directly assists in transferring the signal between agonist-activated receptor and G protein.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/10644723/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab12563219" class="tabcontent ">
                    <h2>The cleaved peptide of PAR1 is a more potent stimulant of platelet-endothelial cell adhesion than is thrombin.</h2>
                    <h3>Journal of vascular surgery; 2003</h3>
                    <p>Platelet-endothelial cell adhesion is an important pathologic response to vessel injury or inflammation. On binding to its endothelial or platelet G protein-linked seven-transmembrane domain receptor, protease-activated receptor-1 (PAR1), thrombin releases a 41-amino acid peptide (TR(1-41)). We examined the effect of TR(1-41) on platelet activation and on platelet-endothelial cell adhesion.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/12563219/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab9058715" class="tabcontent ">
                    <h2>Specific inhibition of thrombin-induced cell activation by the neutrophil proteinases elastase, cathepsin G, and proteinase 3: evidence for distinct cleavage sites within the aminoterminal domain of the thrombin receptor.</h2>
                    <h3>Blood; 1997</h3>
                    <p>The aim of this study was to investigate the inhibitory effects of human leukocyte elastase (HLE), cathepsin G (Cat G), and proteinase 3 (PR3) on the activation of endothelial cells (ECs) and platelets by thrombin and to elucidate the underlying mechanisms. Although preincubation of ECs with HLE or Cat G prevented cytosolic calcium mobilization and prostacyclin synthesis induced by thrombin, these cell responses were not affected when triggered by TRAP42-55, a synthetic peptide corresponding to the sequence of the tethered ligand (Ser42-Phe55) unmasked by thrombin on cleavage of its receptor. Using IIaR-A, a monoclonal antibody directed against the sequence encompassing this cleavage site, flow cytometry analysis showed that the surface expression of this epitope was abolished after incubation of ECs with HLE or Cat G. Further experiments conducted with platelets indicated that not only HLE and Cat G but also PR3 inhibited cell activation induced by thrombin, although they were again ineffective when TRAP42-55 was the agonist. Similar to that for ECs, the epitope for IIaR-A disappeared on treatment of platelets with either proteinase. These results suggested that the neutrophil enzymes proteolyzed the thrombin receptor downstream of the thrombin cleavage site (Arg41-Ser42) but left intact the TRAP42-55 binding site (Gln83-Ser93) within the extracellular aminoterminal domain. The capacity of these proteinases to cleave five overlapping synthetic peptides mapping the portion of the receptor from Asn35 to Pro85 was then investigated. Mass spectrometry studies showed several distinct cleavage sites, i.e., two for HLE (Val(72)-Ser73 and Ile74-Asn75), three for Cat G (Arg41-Ser42, Phe55-Trp56 and Tyr69-Arg70), and one for PR3 (Val(72)-Ser73). We conclude that this singular susceptibility of the thrombin receptor to proteolysis accounts for the ability of neutrophil proteinases to inhibit cell responses to thrombin.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/9058715/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab10978167" class="tabcontent ">
                    <h2>Proteolysis of the exodomain of recombinant protease-activated receptors: prediction of receptor activation or inactivation by MALDI mass spectrometry.</h2>
                    <h3>Biochemistry; 2000</h3>
                    <p>Protease-activated receptors (PARs) mediate cell activation after proteolytic cleavage of their extracellular amino terminus. Thrombin selectively cleaves PAR1, PAR3, and PAR4 to induce activation of platelets and vascular cells, while PAR2 is preferentially cleaved by trypsin. In pathological situations, other proteolytic enzymes may be generated in the circulation and could modify the responses of PARs by cleaving their extracellular domains. To assess the ability of such proteases to activate or inactivate PARs, we designed a strategy for locating cleavage sites on the exofacial NH(2)-terminal fragments of the receptors. The first extracellular segments of PAR1 (PAR1E) and PAR2 (PAR2E) expressed as recombinant proteins in Escherichia coli were incubated with a series of proteases likely to be encountered in the circulation during thrombosis or inflammation. Kinetic and dose-response studies were performed, and the cleavage products were analyzed by MALDI-TOF mass spectrometry. Thrombin cleaved PAR1E at the Arg41-Ser42 activation site at concentrations known to induce cellular activation, supporting a native conformation of the recombinant polypeptide. Plasmin, calpain and leukocyte elastase, cathepsin G, and proteinase 3 cleaved at multiple sites and would be expected to disable PAR1 by cleaving COOH-terminal to the activation site. Cleavage specificities were further confirmed using activation site defective PAR1E S42P mutant polypeptides. Surface plasmon resonance studies on immobilized PAR1E or PAR1E S42P were consistent with cleavage results obtained in solution and allowed us to determine affinities of PAR1E-thrombin binding. FACS analyses of intact platelets confirmed the cleavage of PAR1 downstream of the Arg41-Ser42 site. Mass spectrometry studies of PAR2E predicted activation of PAR2 by trypsin through cleavage at the Arg36-Ser37 site, no effect of thrombin, and inactivation of the receptor by plasmin, calpain and leukocyte elastase, cathepsin G, and proteinase 3. The inhibitory effect of elastase was confirmed on native PAR1 and PAR2 on the basis of Ca(2+) signaling studies in endothelial cells. It was concluded that none of the main proteases generated during fibrinolysis or inflammation appears to be able to signal through PAR1 or PAR2. This strategy provides results which can be extended to the native receptor to predict its activation or inactivation, and it could likewise be used to study other PARs or protease-dependent processes.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/10978167/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                

    <footer>
        <span>© Copyright 2024, YCU Bioinformatics Laboratory</span>
        <span>Page built with <a href="https://quarto.org/">Quarto</a> by バイオインフォマティクス解析センター</span>
    </footer>
</body>
</html>
