<!DOCTYPE html>
<html>
<meta charset="UTF-8">
<head>
    <title>FASLG——FAS - PMID Details</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
            color: #333;
        }

        h1 {
            text-align: center;
            font-size: 2.5em;
            color: #2c3e50;
            margin-bottom: 20px;
        }

        /* Navbar Styles */
        nav {
            background-color: #2c3e50;
            padding: 10px 20px;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            color: white;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        nav .navbar-left, nav .navbar-right {
            display: flex;
            align-items: center;
        }

        nav a {
            color: white;
            text-decoration: none;
            margin: 0 15px;
            font-size: 1.1em;
            transition: color 0.3s ease;
        }

        nav a:hover {
            color: #f39c12;
        }

        nav .divider {
            color: #ecf0f1;
            margin: 0 10px;
        }

        /* Page Footer */
        footer {
            margin-top: 40px;
            padding: 10px 20px;
            background-color: #f7f7f7;
            color: #333;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        /* Tab and Content Styles */
        .tab {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin-bottom: 20px;
        }

        .tab button {
            padding: 10px 20px;
            margin: 5px;
            margin-bottom: 10px;
            cursor: pointer;
            background-color: #f4f4f4;
            border: 2px solid #ddd;
            border-radius: 5px;
            font-size: 1.1em;
            transition: background-color 0.3s ease;
            flex-basis: calc(10% - 10px);
            box-sizing: border-box;
        }

        .tab button:hover {
            background-color: #ecf0f1;
        }

        .tab button.active {
            background-color: #3498db;
            color: white;
            border-color: #2980b9;
        }

        .tabcontent {
            display: none;
            padding: 20px;
            margin-top: 10px;
            background-color: #ecf0f1;
            border-radius: 5px;
            box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
        }

        .tabcontent.active {
            display: block;
        }

        .tabcontent h2 {
            font-size: 2em;
            color: #2980b9;
        }

        .tabcontent p {
            font-size: 1.2em;
            line-height: 1.8;
        }

        /* Emphasize Title and Abstract */
        .title, .abstract {
            background-color: #f9f9f9;
            padding: 20px;
            margin: 10px 0;
            border-left: 5px solid #2980b9;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
        }

        .abstract {
            border-left: 5px solid #e74c3c;
        }
    </style>
    <script>
        function adjustTabStyle() {
            var tabCount = document.querySelectorAll('.tab button').length;
            var tabContainer = document.querySelector('.tab');

            if (tabCount > 10) {
                tabContainer.classList.add('small-tabs');
            }
        }

        function openTab(evt, tabId) {
            document.querySelectorAll(".tabcontent").forEach(tab => tab.style.display = "none");
            document.querySelectorAll(".tablinks").forEach(link => link.classList.remove("active"));
            document.getElementById(tabId).style.display = "block";
            evt.currentTarget.classList.add("active");
        }

        window.onload = adjustTabStyle;
    </script>
</head>
<body>
    <nav>
        <div class="navbar-left">
            <a href="">ConnectomeDB</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/">Home</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database.html">Database</a>
        </div>
        <div class="navbar-right">
            <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/">YCU Bioinformatics Lab</a>
            <span class="divider">|</span>
            <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/">Perkins Systems Biology & Genomics Lab</a>
        </div>
    </nav>

    <h1>PMID Details for FASLG——FAS</h1>

    <div class="tab">
        <button class="tablinks active" onclick="openTab(event, 'tab9126929')">9126929</button><button class="tablinks " onclick="openTab(event, 'tab9228058')">9228058</button><button class="tablinks " onclick="openTab(event, 'tab9126929')">9126929</button>
    </div>

    <div class="tabcontent" id="titleTab">
        <div class="title">
            <h2>Title</h2>
            <p>{{TITLE}}</p>
            <i> Journal:{{JOURNAL}}; Year Published:{{YEAR}}</i>   
        </div>
    </div>

    <div class="tabcontent" id="abstractTab">
        <div class="abstract">
            <h2>Abstract</h2>
            <p>{{ABSTRACT}}</p>
        </div>
    </div>

    
                <div id="tab9126929" class="tabcontent active">
                    <h2>Identification of amino acid residues important for ligand binding to Fas.</h2>
                    <h3>The Journal of experimental medicine; 1997</h3>
                    <p>The interaction of Fas (CD95), a member of the tumor necrosis factor receptor (TNFR) family, and its ligand (FasL) triggers programmed cell death (apoptosis) and is involved in the regulation of immune responses. Although the Fas-FasL interaction is conserved across species barriers, little is currently known about the molecular details of this interaction. Our aim was to identify residues in Fas that are important for ligand binding. With the aid of a Fas molecular model, candidate amino acid residues were selected in the Fas extracellular domain 2 (D2) and D3 and subjected to serine-scanning mutagenesis to produce mutant Fas molecules in the form of Ig fusion proteins. The effects of these mutations on FasL binding was examined by measuring the ability of these proteins to inhibit FasL-mediated apoptosis of Jurkat cells and bind FasL in ELISA and BIAcore assays. Mutation of two amino acids, R86 and R87 (D2), to serine totally abolished the ability of Fas to interact with its ligand, whereas mutants K84S, L90S, E93S (D2), or H126S (D3) showed reduced binding compared with wild-type Fas. Two mutants (K78S and H95S) bound FasL comparably to wild type. Therefore, the binding of FasL involves residues in two domains that correspond to positions critical for ligand binding in other family members (TNFR and CD40) but are conserved between murine and human Fas.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/9126929/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab9228058" class="tabcontent ">
                    <h2>Characterization of Fas (Apo-1, CD95)-Fas ligand interaction.</h2>
                    <h3>The Journal of biological chemistry; 1997</h3>
                    <p>The death-inducing receptor Fas is activated when cross-linked by the type II membrane protein Fas ligand (FasL). When human soluble FasL (sFasL, containing the extracellular portion) was expressed in human embryo kidney 293 cells, the three N-linked glycans of each FasL monomer were found to be essential for efficient secretion. Based on the structure of the closely related lymphotoxin alpha-tumor necrosis factor receptor I complex, a molecular model of the FasL homotrimer bound to three Fas molecules was generated using knowledge-based protein modeling methods. Point mutations of amino acid residues predicted to affect the receptor-ligand interaction were introduced at three sites. The F275L mutant, mimicking the loss of function murine gld mutation, exhibited a high propensity for aggregation and was unable to bind to Fas. Mutants P206R, P206D, and P206F displayed reduced cytotoxicity toward Fas-positive cells with a concomitant decrease in the binding affinity for the recombinant Fas-immunoglobulin Fc fusion proteins. Although the cytotoxic activity of mutant Y218D was unaltered, mutant Y218R was inactive, correlating with the prediction that Tyr-218 of FasL interacts with a cluster of three basic amino acid side chains of Fas. Interestingly, mutant Y218F could induce apoptosis in murine, but not human cells.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/9228058/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab9126929" class="tabcontent ">
                    <h2>Identification of amino acid residues important for ligand binding to Fas.</h2>
                    <h3>The Journal of experimental medicine; 1997</h3>
                    <p>The interaction of Fas (CD95), a member of the tumor necrosis factor receptor (TNFR) family, and its ligand (FasL) triggers programmed cell death (apoptosis) and is involved in the regulation of immune responses. Although the Fas-FasL interaction is conserved across species barriers, little is currently known about the molecular details of this interaction. Our aim was to identify residues in Fas that are important for ligand binding. With the aid of a Fas molecular model, candidate amino acid residues were selected in the Fas extracellular domain 2 (D2) and D3 and subjected to serine-scanning mutagenesis to produce mutant Fas molecules in the form of Ig fusion proteins. The effects of these mutations on FasL binding was examined by measuring the ability of these proteins to inhibit FasL-mediated apoptosis of Jurkat cells and bind FasL in ELISA and BIAcore assays. Mutation of two amino acids, R86 and R87 (D2), to serine totally abolished the ability of Fas to interact with its ligand, whereas mutants K84S, L90S, E93S (D2), or H126S (D3) showed reduced binding compared with wild-type Fas. Two mutants (K78S and H95S) bound FasL comparably to wild type. Therefore, the binding of FasL involves residues in two domains that correspond to positions critical for ligand binding in other family members (TNFR and CD40) but are conserved between murine and human Fas.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/9126929/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                

    <footer>
        <span>© Copyright 2024, YCU Bioinformatics Laboratory</span>
        <span>Page built with <a href="https://quarto.org/">Quarto</a> by バイオインフォマティクス解析センター</span>
    </footer>
</body>
</html>
