<!DOCTYPE html>
<html>
<meta charset="UTF-8">
<head>
    <title>FGF1——FGFR1 - PMID Details</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
            color: #333;
        }

        h1 {
            text-align: center;
            font-size: 2.5em;
            color: #2c3e50;
            margin-bottom: 20px;
        }

        /* Navbar Styles */
        nav {
            background-color: #2c3e50;
            padding: 10px 20px;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            color: white;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        nav .navbar-left, nav .navbar-right {
            display: flex;
            align-items: center;
        }

        nav a {
            color: white;
            text-decoration: none;
            margin: 0 15px;
            font-size: 1.1em;
            transition: color 0.3s ease;
        }

        nav a:hover {
            color: #f39c12;
        }

        nav .divider {
            color: #ecf0f1;
            margin: 0 10px;
        }

        /* Page Footer */
        footer {
            margin-top: 40px;
            padding: 10px 20px;
            background-color: #f7f7f7;
            color: #333;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        /* Tab and Content Styles */
        .tab {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin-bottom: 20px;
        }

        .tab button {
            padding: 10px 20px;
            margin: 5px;
            margin-bottom: 10px;
            cursor: pointer;
            background-color: #f4f4f4;
            border: 2px solid #ddd;
            border-radius: 5px;
            font-size: 1.1em;
            transition: background-color 0.3s ease;
            flex-basis: calc(10% - 10px);
            box-sizing: border-box;
        }

        .tab button:hover {
            background-color: #ecf0f1;
        }

        .tab button.active {
            background-color: #3498db;
            color: white;
            border-color: #2980b9;
        }

        .tabcontent {
            display: none;
            padding: 20px;
            margin-top: 10px;
            background-color: #ecf0f1;
            border-radius: 5px;
            box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
        }

        .tabcontent.active {
            display: block;
        }

        .tabcontent h2 {
            font-size: 2em;
            color: #2980b9;
        }

        .tabcontent p {
            font-size: 1.2em;
            line-height: 1.8;
        }

        /* Emphasize Title and Abstract */
        .title, .abstract {
            background-color: #f9f9f9;
            padding: 20px;
            margin: 10px 0;
            border-left: 5px solid #2980b9;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
        }

        .abstract {
            border-left: 5px solid #e74c3c;
        }
    </style>
    <script>
        function adjustTabStyle() {
            var tabCount = document.querySelectorAll('.tab button').length;
            var tabContainer = document.querySelector('.tab');

            if (tabCount > 10) {
                tabContainer.classList.add('small-tabs');
            }
        }

        function openTab(evt, tabId) {
            document.querySelectorAll(".tabcontent").forEach(tab => tab.style.display = "none");
            document.querySelectorAll(".tablinks").forEach(link => link.classList.remove("active"));
            document.getElementById(tabId).style.display = "block";
            evt.currentTarget.classList.add("active");
        }

        window.onload = adjustTabStyle;
    </script>
</head>
<body>
    <nav>
        <div class="navbar-left">
            <a href="">ConnectomeDB</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/">Home</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database.html">Database</a>
        </div>
        <div class="navbar-right">
            <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/">YCU Bioinformatics Lab</a>
            <span class="divider">|</span>
            <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/">Perkins Systems Biology & Genomics Lab</a>
        </div>
    </nav>

    <h1>PMID Details for FGF1——FGFR1</h1>

    <div class="tab">
        <button class="tablinks active" onclick="openTab(event, 'tab8663044')">8663044</button><button class="tablinks " onclick="openTab(event, 'tab8756477')">8756477</button><button class="tablinks " onclick="openTab(event, 'tab11032250')">11032250</button><button class="tablinks " onclick="openTab(event, 'tab10336501')">10336501</button>
    </div>

    <div class="tabcontent" id="titleTab">
        <div class="title">
            <h2>Title</h2>
            <p>{{TITLE}}</p>
            <i> Journal:{{JOURNAL}}; Year Published:{{YEAR}}</i>   
        </div>
    </div>

    <div class="tabcontent" id="abstractTab">
        <div class="abstract">
            <h2>Abstract</h2>
            <p>{{ABSTRACT}}</p>
        </div>
    </div>

    
                <div id="tab8663044" class="tabcontent active">
                    <h2>Receptor specificity of the fibroblast growth factor family.</h2>
                    <h3>The Journal of biological chemistry; 1996</h3>
                    <p>Fibroblast growth factors (FGFs) are essential molecules for mammalian development. The nine known FGF ligands and the four signaling FGF receptors (and their alternatively spliced variants) are expressed in specific spatial and temporal patterns. The activity of this signaling pathway is regulated by ligand binding specificity, heparan sulfate proteoglycans, and the differential signaling capacity of individual FGF receptors. To determine potentially relevant ligand-receptor pairs we have engineered mitogenically responsive cell lines expressing the major splice variants of all the known FGF receptors. We have assayed the mitogenic activity of the nine known FGF ligands on these cell lines. These studies demonstrate that FGF 1 is the only FGF that can activate all FGF receptor splice variants. Using FGF 1 as an internal standard we have determined the relative activity of all the other members of the FGF family. These data should serve as a biochemical foundation for determining developmental, physiological, and pathophysiological processes that involve FGF signaling pathways.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/8663044/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab8756477" class="tabcontent ">
                    <h2>A natural kinase-deficient variant of fibroblast growth factor receptor 1.</h2>
                    <h3>Biochemistry; 1996</h3>
                    <p>A fibroblast growth factor receptor 1 variant missing 37 amino acids from the carboxy-terminal tyrosine kinase catalytic domain was discovered in human lung fibroblasts and several other human cell lines. The receptor variant binds specifically to acidic fibroblast growth factor but has no tyrosine kinase activity. It was found that cellular transfectants expressing the fibroblast growth factor receptor 1 variant are mitogenically inactive and ligand binding to the receptor causes neither receptor autophosphorylation nor phospholipase C-gamma transphosphorylation. The fibroblast growth factor receptor 1 variant therefore represents an inactive receptor for acidic fibroblast growth factor. Since both kinase and kinase-deficient receptor forms are expressed in cells, it is conceivable that the kinase-deficient receptor plays an important role in regulating cellular responses elicited by acidic fibroblast growth factor stimulation.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/8756477/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab11032250" class="tabcontent ">
                    <h2>Selection of peptide ligands that bind to acid fibroblast growth factor.</h2>
                    <h3>IUBMB life; 2000</h3>
                    <p>Acid fibroblast growth factor (aFGF) binds to its cell-surface receptors in a heparin-dependent manner. In an attempt to define the aFGF recognition site on fibroblast growth factor receptor 1 (FGFR1), we developed a screening strategy for identifying FGF ligands that bind to the receptor-binding region of FGF. To retain the natural conformation of aFGF during screening, we used biotinylated heparin to immobilize aFGF on a streptavidin-coated dish. A 15-mer phage display peptide library was then screened in the dish and a group of related peptide sequences was identified. These peptide sequences contain two conserved motifs, SSG and VPS, corresponding to two protein sequences of the immunoglobulin-like (Ig-like) domain II of FGFR1 at amino acids 180-182 and 221-223 (CPSSG-VPSDKGNYTC). Further experiments demonstrate that the phage displaying these sequences can specifically bind to aFGF and that the synthesized peptide corresponding in sequence can inhibit mitogenic activity of aFGF. These sequences may thus constitute part of the aFGF-binding region on FGFR1, and the synthesized peptide has the potential to become a therapeutic agent as an aFGF antagonist.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/11032250/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab10336501" class="tabcontent ">
                    <h2>Specificity for fibroblast growth factors determined by heparan sulfate in a binary complex with the receptor kinase.</h2>
                    <h3>The Journal of biological chemistry; 1999</h3>
                    <p>A divalent cation-dependent association between heparin or heparan sulfate and the ectodomain of the FGF receptor kinase (FGFR) restricts FGF-independent trans-phosphorylation and supports the binding of activating FGF to self-associated FGFR. Here we show that in contrast to heparin, cellular heparan sulfate forms a binary complex with FGFR that discriminates between FGF-1 and FGF-2. FGFR type 4 (FGFR4) in liver parenchymal cells binds only FGF-1, whereas FGFR1 binds FGF-1 and FGF-2 equally. Cell-free complexes of heparin and recombinant FGFR4 bound FGF-1 and FGF-2 equally. However, in contrast to FGFR1, when recombinant FGFR4 was expressed back in epithelial cells by transfection, it failed to bind FGF-2 unless heparan sulfate was depressed by chlorate or heparinase treatment. Isolated heparan sulfate proteoglycan (HSPG) from liver cells in cell-free complexes with FGFR4 restored the specificity for FGF-1 and supported the binding of both FGF-1 and FGF-2 when complexed with FGFR1. In contrast, FGF-2 bound equally well to complexes of both FGFR1 and FGFR4 formed with endothelial cell-derived HSPG, but the endothelial HSPG was deficient for the binding of FGF-1 to both FGFR complexes. These data suggest that a heparan sulfate subunit is a cell type- and FGFR-specific determinant of the selectivity of the FGFR signaling complex for FGF. In a physiological context, the heparan sulfate subunit may limit the redundancy among the current 18 FGF polypeptides for the 4 known FGFR.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/10336501/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                

    <footer>
        <span>© Copyright 2024, YCU Bioinformatics Laboratory</span>
        <span>Page built with <a href="https://quarto.org/">Quarto</a> by バイオインフォマティクス解析センター</span>
    </footer>
</body>
</html>
