<!DOCTYPE html>
<html>
<meta charset="UTF-8">
<head>
    <title>GNAI2——MTNR1A - PMID Details</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
            color: #333;
        }

        h1 {
            text-align: center;
            font-size: 2.5em;
            color: #2c3e50;
            margin-bottom: 20px;
        }

        /* Navbar Styles */
        nav {
            background-color: #2c3e50;
            padding: 10px 20px;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            color: white;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        nav .navbar-left, nav .navbar-right {
            display: flex;
            align-items: center;
        }

        nav a {
            color: white;
            text-decoration: none;
            margin: 0 15px;
            font-size: 1.1em;
            transition: color 0.3s ease;
        }

        nav a:hover {
            color: #f39c12;
        }

        nav .divider {
            color: #ecf0f1;
            margin: 0 10px;
        }

        /* Page Footer */
        footer {
            margin-top: 40px;
            padding: 10px 20px;
            background-color: #f7f7f7;
            color: #333;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        /* Tab and Content Styles */
        .tab {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin-bottom: 20px;
        }

        .tab button {
            padding: 10px 20px;
            margin: 5px;
            margin-bottom: 10px;
            cursor: pointer;
            background-color: #f4f4f4;
            border: 2px solid #ddd;
            border-radius: 5px;
            font-size: 1.1em;
            transition: background-color 0.3s ease;
            flex-basis: calc(10% - 10px);
            box-sizing: border-box;
        }

        .tab button:hover {
            background-color: #ecf0f1;
        }

        .tab button.active {
            background-color: #3498db;
            color: white;
            border-color: #2980b9;
        }

        .tabcontent {
            display: none;
            padding: 20px;
            margin-top: 10px;
            background-color: #ecf0f1;
            border-radius: 5px;
            box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
        }

        .tabcontent.active {
            display: block;
        }

        .tabcontent h2 {
            font-size: 2em;
            color: #2980b9;
        }

        .tabcontent p {
            font-size: 1.2em;
            line-height: 1.8;
        }

        /* Emphasize Title and Abstract */
        .title, .abstract {
            background-color: #f9f9f9;
            padding: 20px;
            margin: 10px 0;
            border-left: 5px solid #2980b9;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
        }

        .abstract {
            border-left: 5px solid #e74c3c;
        }
    </style>
    <script>
        function adjustTabStyle() {
            var tabCount = document.querySelectorAll('.tab button').length;
            var tabContainer = document.querySelector('.tab');

            if (tabCount > 10) {
                tabContainer.classList.add('small-tabs');
            }
        }

        function openTab(evt, tabId) {
            document.querySelectorAll(".tabcontent").forEach(tab => tab.style.display = "none");
            document.querySelectorAll(".tablinks").forEach(link => link.classList.remove("active"));
            document.getElementById(tabId).style.display = "block";
            evt.currentTarget.classList.add("active");
        }

        window.onload = adjustTabStyle;
    </script>
</head>
<body>
    <nav>
        <div class="navbar-left">
            <a href="">ConnectomeDB</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/">Home</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database.html">Database</a>
        </div>
        <div class="navbar-right">
            <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/">YCU Bioinformatics Lab</a>
            <span class="divider">|</span>
            <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/">Perkins Systems Biology & Genomics Lab</a>
        </div>
    </nav>

    <h1>PMID Details for GNAI2——MTNR1A</h1>

    <div class="tab">
        <button class="tablinks active" onclick="openTab(event, 'tab10598579')">10598579</button><button class="tablinks " onclick="openTab(event, 'tab17215244')">17215244</button>
    </div>

    <div class="tabcontent" id="titleTab">
        <div class="title">
            <h2>Title</h2>
            <p>{{TITLE}}</p>
            <i> Journal:{{JOURNAL}}; Year Published:{{YEAR}}</i>   
        </div>
    </div>

    <div class="tabcontent" id="abstractTab">
        <div class="abstract">
            <h2>Abstract</h2>
            <p>{{ABSTRACT}}</p>
        </div>
    </div>

    
                <div id="tab10598579" class="tabcontent active">
                    <h2>Dual signaling of human Mel1a melatonin receptors via G(i2), G(i3), and G(q/11) proteins.</h2>
                    <h3>Molecular endocrinology (Baltimore, Md.); 1999</h3>
                    <p>Mel 1a melatonin receptors belong to the super-family of guanine nucleotide-binding regulatory protein (G protein)-coupled receptors. So far, interest in Mel 1a receptor signaling has focused mainly on the modulation of the adenylyl cyclase pathway via pertussis toxin (PTX)-sensitive G proteins. To further investigate signaling of the human Mel 1a receptor, we have developed an antibody directed against the C terminus of this receptor. This antibody detected the Mel 1a receptor as a protein with an apparent molecular mass of approximately 60 kDa in immunoblots after separation by SDS-PAGE. It also specifically precipitated the 2-[125I]iodomelatonin (125I-Mel)-labeled receptor from Mel 1a-transfected HEK 293 cells. Coprecipitation experiments showed that G(i2), G(i3), and G(q/11) proteins couple to the Mel 1a receptor in an agonist-dependent and guanine nucleotide-sensitive manner. Coupling was selective since other G proteins present in HEK 293 cells, (G(i1), G(o), G(s), G(z), and G12) were not detected in receptor complexes. Coupling of the Mel 1a receptor to G(i) and G(q) was confirmed by inhibition of high-affinity 125I-Mel binding to receptors with subtype-selective G protein alpha-subunit antibodies. G(i2) and/or G(i3) mediated adenylyl cyclase inhibition while G(q/11) induced a transient elevation in cytosolic calcium concentrations in HEK 293 cells stably expressing Mel 1a receptors. Melatonin-induced cytosolic calcium mobilization via PTX-insensitive G proteins was confirmed in primary cultures of ovine pars tuberalis cells endogenously expressing Mel 1a receptors. In conclusion, we report the development of the first antibody recognizing the cloned human Mel 1a melatonin receptor protein. We show that Mel 1a receptors functionally couple to both PTX-sensitive and PTX-insensitive G proteins. The previously unknown signaling of Mel 1a receptors through G(q/11) widens the spectrum of potential targets for melatonin.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/10598579/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab17215244" class="tabcontent ">
                    <h2>Purification and identification of G protein-coupled receptor protein complexes under native conditions.</h2>
                    <h3>Molecular & cellular proteomics : MCP; 2007</h3>
                    <p>G protein-coupled receptors (GPCRs) constitute the largest family of membrane receptors and are of major therapeutic importance. The identification of GPCR-associated proteins is an important step toward a better understanding of these receptors. However, current methods are not satisfying as only isolated receptor domains (intracellular loops or carboxyl-terminal tails) can be used as "bait." We report here a method based on tandem affinity purification coupled to mass spectrometry that overcomes these limitations as the entire receptor is used to identify protein complexes formed in living mammalian cells. The human MT(1) and MT(2) melatonin receptors were chosen as model GPCRs. Both receptors were tagged with the tandem affinity purification tag at their carboxyl-terminal tails and expressed in human embryonic kidney 293 cells. Receptor solubilization and purification conditions were optimized. The method was validated by the co-purification of G(i) proteins, which are well known GPCR interaction partners but which are difficult to identify with current protein-protein interaction assays. Several new and functionally relevant MT(1)- and MT(2)-associated proteins were identified; some of them were common to both receptors, and others were specific for each subtype. Taken together, our protocol allowed for the first time the purification of GPCR-associated proteins under native conditions in quantities suitable for mass spectrometry analysis.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/17215244/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                

    <footer>
        <span>© Copyright 2024, YCU Bioinformatics Laboratory</span>
        <span>Page built with <a href="https://quarto.org/">Quarto</a> by バイオインフォマティクス解析センター</span>
    </footer>
</body>
</html>
