<!DOCTYPE html>
<html>
<meta charset="UTF-8">
<head>
    <title>HLA-G——LILRB1 - PMID Details</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
            color: #333;
        }

        h1 {
            text-align: center;
            font-size: 2.5em;
            color: #2c3e50;
            margin-bottom: 20px;
        }

        /* Navbar Styles */
        nav {
            background-color: #2c3e50;
            padding: 10px 20px;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            color: white;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        nav .navbar-left, nav .navbar-right {
            display: flex;
            align-items: center;
        }

        nav a {
            color: white;
            text-decoration: none;
            margin: 0 15px;
            font-size: 1.1em;
            transition: color 0.3s ease;
        }

        nav a:hover {
            color: #f39c12;
        }

        nav .divider {
            color: #ecf0f1;
            margin: 0 10px;
        }

        /* Page Footer */
        footer {
            margin-top: 40px;
            padding: 10px 20px;
            background-color: #f7f7f7;
            color: #333;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        /* Tab and Content Styles */
        .tab {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin-bottom: 20px;
        }

        .tab button {
            padding: 10px 20px;
            margin: 5px;
            margin-bottom: 10px;
            cursor: pointer;
            background-color: #f4f4f4;
            border: 2px solid #ddd;
            border-radius: 5px;
            font-size: 1.1em;
            transition: background-color 0.3s ease;
            flex-basis: calc(10% - 10px);
            box-sizing: border-box;
        }

        .tab button:hover {
            background-color: #ecf0f1;
        }

        .tab button.active {
            background-color: #3498db;
            color: white;
            border-color: #2980b9;
        }

        .tabcontent {
            display: none;
            padding: 20px;
            margin-top: 10px;
            background-color: #ecf0f1;
            border-radius: 5px;
            box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
        }

        .tabcontent.active {
            display: block;
        }

        .tabcontent h2 {
            font-size: 2em;
            color: #2980b9;
        }

        .tabcontent p {
            font-size: 1.2em;
            line-height: 1.8;
        }

        /* Emphasize Title and Abstract */
        .title, .abstract {
            background-color: #f9f9f9;
            padding: 20px;
            margin: 10px 0;
            border-left: 5px solid #2980b9;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
        }

        .abstract {
            border-left: 5px solid #e74c3c;
        }
    </style>
    <script>
        function adjustTabStyle() {
            var tabCount = document.querySelectorAll('.tab button').length;
            var tabContainer = document.querySelector('.tab');

            if (tabCount > 10) {
                tabContainer.classList.add('small-tabs');
            }
        }

        function openTab(evt, tabId) {
            document.querySelectorAll(".tabcontent").forEach(tab => tab.style.display = "none");
            document.querySelectorAll(".tablinks").forEach(link => link.classList.remove("active"));
            document.getElementById(tabId).style.display = "block";
            evt.currentTarget.classList.add("active");
        }

        window.onload = adjustTabStyle;
    </script>
</head>
<body>
    <nav>
        <div class="navbar-left">
            <a href="">ConnectomeDB</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/">Home</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database.html">Database</a>
        </div>
        <div class="navbar-right">
            <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/">YCU Bioinformatics Lab</a>
            <span class="divider">|</span>
            <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/">Perkins Systems Biology & Genomics Lab</a>
        </div>
    </nav>

    <h1>PMID Details for HLA-G——LILRB1</h1>

    <div class="tab">
        <button class="tablinks active" onclick="openTab(event, 'tab15718280')">15718280</button><button class="tablinks " onclick="openTab(event, 'tab9933109')">9933109</button><button class="tablinks " onclick="openTab(event, 'tab12853576')">12853576</button><button class="tablinks " onclick="openTab(event, 'tab12874224')">12874224</button>
    </div>

    <div class="tabcontent" id="titleTab">
        <div class="title">
            <h2>Title</h2>
            <p>{{TITLE}}</p>
            <i> Journal:{{JOURNAL}}; Year Published:{{YEAR}}</i>   
        </div>
    </div>

    <div class="tabcontent" id="abstractTab">
        <div class="abstract">
            <h2>Abstract</h2>
            <p>{{ABSTRACT}}</p>
        </div>
    </div>

    
                <div id="tab15718280" class="tabcontent active">
                    <h2>Crystal structure of HLA-G: a nonclassical MHC class I molecule expressed at the fetal-maternal interface.</h2>
                    <h3>Proceedings of the National Academy of Sciences of the United States of America; 2005</h3>
                    <p>HLA-G is a nonclassical major histocompatibility complex class I (MHC-I) molecule that is primarily expressed at the fetal-maternal interface, where it is thought to play a role in protecting the fetus from the maternal immune response. HLA-G binds a limited repertoire of peptides and interacts with the inhibitory leukocyte Ig-like receptors LIR-1 and LIR-2 and possibly with certain natural killer cell receptors. To gain further insights into HLA-G function, we determined the 1.9-A structure of a monomeric HLA-G complexed to a natural endogenous peptide ligand from histone H2A (RIIPRHLQL). An extensive network of contacts between the peptide and the antigen-binding cleft reveal a constrained mode of binding reminiscent of the nonclassical HLA-E molecule, thereby providing a structural basis for the limited peptide repertoire of HLA-G. The alpha3 domain of HLA-G, a candidate binding site for the LIR-1 and -2 inhibitory receptors, is structurally distinct from the alpha3 domains of classical MHC-I molecules, providing a rationale for the observed affinity differences for these ligands. The structural data suggest a head-to-tail mode of dimerization, mediated by an intermolecular disulfide bond, that is consistent with the observation of HLA-G dimers on the cell surface.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/15718280/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab9933109" class="tabcontent ">
                    <h2>The ILT2(LIR1) and CD94/NKG2A NK cell receptors respectively recognize HLA-G1 and HLA-E molecules co-expressed on target cells.</h2>
                    <h3>European journal of immunology; 1999</h3>
                    <p>Previous studies on NK recognition of HLA-G1 employed as targets 721.221 transfectants (.221-G1) that unknowingly co-expressed the HLA-E molecule, subsequently found to be a major ligand for the CD94/NKG2 receptors. In the present study we re-evaluated the relative role played by CD94/NKG2 and ILT2(LIR1) molecules in recognition of HLA-G1 by NK clones. We employed as targets .221-G1 cells and a surface HLA-E-negative transfectant, .221-G1(Eneg), generated by site-directed mutagenesis of the HLA-G1 leader sequence. The antagonistic effects of receptor- (ie. CD94/NKG2A, ILT2) and ligand-specific mAb (i.e. HLA-G, HLA-E) were assessed. In addition, binding of an ILT2-Ig fusion protein to the .221-AEH, expressing only HLA-E, and the .221-G1(Eneg) transfectants was analyzed. Our data demonstrate that NK recognition of cells expressing HLA-G1 involves at least two non-overlapping receptor-ligand systems: the CD94/NKG2 interaction with HLA-E, and the engagement of the ILT2(LIR1) receptor by HLA-G1 molecules.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/9933109/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab12853576" class="tabcontent ">
                    <h2>Human inhibitory receptors Ig-like transcript 2 (ILT2) and ILT4 compete with CD8 for MHC class I binding and bind preferentially to HLA-G.</h2>
                    <h3>Proceedings of the National Academy of Sciences of the United States of America; 2003</h3>
                    <p>Ig-like transcript 4 (ILT4) (also known as leukocyte Ig-like receptor 2, CD85d, and LILRB2) is a cell surface receptor expressed mainly on myelomonocytic cells, whereas ILT2 (also known as leukocyte Ig-like receptor 1, CD85j, and LILRB1) is expressed on a wider range of immune cells including subsets of natural killer and T cells. Both ILTs contain immunoreceptor tyrosine-based inhibitory receptor motifs in their cytoplasmic tails that inhibit cellular responses by recruiting phosphatases such as SHP-1 (Src homology 2 domain containing tyrosine phosphatase 1). Although these ILTs have been shown to recognize a broad range of classical and nonclassical human MHC class I molecules (MHCIs), their precise binding properties remain controversial. We have used surface plasmon resonance to analyze the interaction of soluble forms of ILT4 and ILT2 with several MHCIs. Although the range of affinities measured was quite broad (Kd = 2-45 microM), some interesting differences were observed. ILT2 generally bound with a 2- to 3-fold higher affinity than ILT4 to the same MHCI. Furthermore, ILT2 and ILT4 bound to HLA-G with a 3- to 4-fold higher affinity than to classical MHCIs, suggesting that ILT/HLA-G recognition may play a dominant role in the regulation of natural killer, T, and myelomonocytic cell activation. Finally, we show that ILT2 and ILT4 effectively compete with CD8 for MHCI binding, raising the possibility that ILT2 modulates CD8+ T cell activation by blocking the CD8 binding as well as by recruiting inhibitory molecules through its immunoreceptor tyrosine-based inhibitory receptor motif.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/12853576/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab12874224" class="tabcontent ">
                    <h2>Complexes of HLA-G protein on the cell surface are important for leukocyte Ig-like receptor-1 function.</h2>
                    <h3>Journal of immunology (Baltimore, Md. : 1950); 2003</h3>
                    <p>The nonclassical class I MHC molecule HLA-G is selectively expressed on extravillous cytotrophoblast cells at the maternal-fetal interface during pregnancy. HLA-G can inhibit the killing mediated by NK cells via interaction with the inhibitory NK cell receptor, leukocyte Ig-like receptor-1 (LIR-1). Comparison of the sequence of the HLA-G molecule to other class I MHC proteins revealed two unique cysteine residues located in positions 42 and 147. Mutating these cysteine residues resulted in a dramatic decrease in LIR-1 Ig binding. Accordingly, the mutated HLA-G transfectants were less effective in the inhibition of NK killing and RBL/LIR-1 induced serotonin release. Immunoprecipitation experiments demonstrated the involvement of the cysteine residues in the formation of HLA-G protein oligomers on the cell surface. The cysteine residue located at position 42 is shown to be critical for the expression of such complexes. These oligomers, unique among the class I MHC proteins, probably bind to LIR-1 with increased avidity, resulting in an enhanced inhibitory function of LIR-1 and an impaired killing function of NK cells.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/12874224/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                

    <footer>
        <span>© Copyright 2024, YCU Bioinformatics Laboratory</span>
        <span>Page built with <a href="https://quarto.org/">Quarto</a> by バイオインフォマティクス解析センター</span>
    </footer>
</body>
</html>
