<!DOCTYPE html>
<html>
<meta charset="UTF-8">
<head>
    <title>IFNA2——IFNAR2 - PMID Details</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
            color: #333;
        }

        h1 {
            text-align: center;
            font-size: 2.5em;
            color: #2c3e50;
            margin-bottom: 20px;
        }

        /* Navbar Styles */
        nav {
            background-color: #2c3e50;
            padding: 10px 20px;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            color: white;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        nav .navbar-left, nav .navbar-right {
            display: flex;
            align-items: center;
        }

        nav a {
            color: white;
            text-decoration: none;
            margin: 0 15px;
            font-size: 1.1em;
            transition: color 0.3s ease;
        }

        nav a:hover {
            color: #f39c12;
        }

        nav .divider {
            color: #ecf0f1;
            margin: 0 10px;
        }

        /* Page Footer */
        footer {
            margin-top: 40px;
            padding: 10px 20px;
            background-color: #f7f7f7;
            color: #333;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        /* Tab and Content Styles */
        .tab {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin-bottom: 20px;
        }

        .tab button {
            padding: 10px 20px;
            margin: 5px;
            margin-bottom: 10px;
            cursor: pointer;
            background-color: #f4f4f4;
            border: 2px solid #ddd;
            border-radius: 5px;
            font-size: 1.1em;
            transition: background-color 0.3s ease;
            flex-basis: calc(10% - 10px);
            box-sizing: border-box;
        }

        .tab button:hover {
            background-color: #ecf0f1;
        }

        .tab button.active {
            background-color: #3498db;
            color: white;
            border-color: #2980b9;
        }

        .tabcontent {
            display: none;
            padding: 20px;
            margin-top: 10px;
            background-color: #ecf0f1;
            border-radius: 5px;
            box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
        }

        .tabcontent.active {
            display: block;
        }

        .tabcontent h2 {
            font-size: 2em;
            color: #2980b9;
        }

        .tabcontent p {
            font-size: 1.2em;
            line-height: 1.8;
        }

        /* Emphasize Title and Abstract */
        .title, .abstract {
            background-color: #f9f9f9;
            padding: 20px;
            margin: 10px 0;
            border-left: 5px solid #2980b9;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
        }

        .abstract {
            border-left: 5px solid #e74c3c;
        }
    </style>
    <script>
        function adjustTabStyle() {
            var tabCount = document.querySelectorAll('.tab button').length;
            var tabContainer = document.querySelector('.tab');

            if (tabCount > 10) {
                tabContainer.classList.add('small-tabs');
            }
        }

        function openTab(evt, tabId) {
            document.querySelectorAll(".tabcontent").forEach(tab => tab.style.display = "none");
            document.querySelectorAll(".tablinks").forEach(link => link.classList.remove("active"));
            document.getElementById(tabId).style.display = "block";
            evt.currentTarget.classList.add("active");
        }

        window.onload = adjustTabStyle;
    </script>
</head>
<body>
    <nav>
        <div class="navbar-left">
            <a href="">ConnectomeDB</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/">Home</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database.html">Database</a>
        </div>
        <div class="navbar-right">
            <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/">YCU Bioinformatics Lab</a>
            <span class="divider">|</span>
            <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/">Perkins Systems Biology & Genomics Lab</a>
        </div>
    </nav>

    <h1>PMID Details for IFNA2——IFNAR2</h1>

    <div class="tab">
        <button class="tablinks active" onclick="openTab(event, 'tab10339405')">10339405</button><button class="tablinks " onclick="openTab(event, 'tab11888273')">11888273</button>
    </div>

    <div class="tabcontent" id="titleTab">
        <div class="title">
            <h2>Title</h2>
            <p>{{TITLE}}</p>
            <i> Journal:{{JOURNAL}}; Year Published:{{YEAR}}</i>   
        </div>
    </div>

    <div class="tabcontent" id="abstractTab">
        <div class="abstract">
            <h2>Abstract</h2>
            <p>{{ABSTRACT}}</p>
        </div>
    </div>

    
                <div id="tab10339405" class="tabcontent active">
                    <h2>Biophysical analysis of the interaction of human ifnar2 expressed in E. coli with IFNalpha2.</h2>
                    <h3>Journal of molecular biology; 1999</h3>
                    <p>Type I interferons are cytokines which activate an anti-viral response by binding to two specific cell surface receptors, ifnar1 and ifnar2. Here, we report purification and refolding of the extracellular part of human ifnar2 (ifnar2-EC) expressed in Escherichia coli and its characterization with respect to its interaction with interferon alpha2 (IFNalpha2). The 25 kDa, non-glycosylated ifnar2-EC is a stable, fully active protein, which inhibits antiviral activity of IFNalpha2. The stoichiometry of binding IFNalpha2 is 1:1, as determined by gel filtration, chemical cross-linking and solid-phase detection. The affinity of this interaction is 10 nM, which is similar to the affinity measured for the cell surface-bound ifnar2 receptor. No difference in affinity was found throughout various assays using optical detection as BIAcore or reflectometric interference spectorscopy. However, the binding kinetics as measured in homogeneous phase by fluorescence de-quenching was about three times faster than that measured on a sensor surface. The rate of complex formation is relatively high compared to other cytokine-receptor interactions. The salt dependence of the association kinetics suggest a limited but significant contribution of electrostatic forces towards the rate of complex formation. The dissociation constant increases with decreasing pH according to the protonation of a base with a pKa of 6.7. The surface properties of the IFNalpha2 binding surface on ifnar2 were interpreted according to the pH and salt dependence of the interaction.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/10339405/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab11888273" class="tabcontent ">
                    <h2>The human interferon receptor: NMR-based modeling, mapping of the IFN-alpha 2 binding site, and observed ligand-induced tightening.</h2>
                    <h3>Biochemistry; 2002</h3>
                    <p>The human interferon receptor (IFNAR) mediates the antiviral and antiproliferative activities of type I interferons (IFNs). This receptor is comprised of subunits IFNAR1 and IFNAR2, the latter exhibiting nanomolar affinity for IFNs. Here the extracellular domain of IFNAR2 (IFNAR2-EC), a soluble 25 kDa IFN-binding polypeptide, and its complex with IFN-alpha 2 were studied using multidimensional NMR. IFNAR2-EC is comprised of two fibronectin-III (FN-III) domains connected by a helical hinge region. The deduced global fold was utilized to improve the alignment of IFNAR2-EC against structurally related receptors and to model its structure. A striking feature of IFNAR2-EC is the limited and localized deviations in chemical shifts exhibited upon ligand binding, observed for only 15% of its backbone (1)H and (15)N nuclei. Analysis of these deviations maps the IFN-alpha 2 binding site upon IFNAR2-EC to a contiguous surface on the N-terminal domain, including the S3-S4 loop (residues 44-53), the S5-S6 loop and S6 beta-strand (residues 74-82), and the S7 beta-strand and the hinge region (residues 95-105). The C-terminal domain contributes only marginally to ligand binding, and no change in the hypothesized interdomain interface is observed. The proposed binding domain encompasses all residues implicated by mutagenesis studies in IFN binding, and suggests adjacent residues cooperate in forming the binding surface. D(2)O-exchange experiments indicate that binding of IFN-alpha2 induces tightening of the N-terminal domain of IFNAR2-EC. This increase in receptor rigidity may play an important role in initiating the intracellular stage of the IFN signaling cascade.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/11888273/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                

    <footer>
        <span>© Copyright 2024, YCU Bioinformatics Laboratory</span>
        <span>Page built with <a href="https://quarto.org/">Quarto</a> by バイオインフォマティクス解析センター</span>
    </footer>
</body>
</html>
