<!DOCTYPE html>
<html>
<meta charset="UTF-8">
<head>
    <title>IL10——IL10RA - PMID Details</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
            color: #333;
        }

        h1 {
            text-align: center;
            font-size: 2.5em;
            color: #2c3e50;
            margin-bottom: 20px;
        }

        /* Navbar Styles */
        nav {
            background-color: #2c3e50;
            padding: 10px 20px;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            color: white;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        nav .navbar-left, nav .navbar-right {
            display: flex;
            align-items: center;
        }

        nav a {
            color: white;
            text-decoration: none;
            margin: 0 15px;
            font-size: 1.1em;
            transition: color 0.3s ease;
        }

        nav a:hover {
            color: #f39c12;
        }

        nav .divider {
            color: #ecf0f1;
            margin: 0 10px;
        }

        /* Page Footer */
        footer {
            margin-top: 40px;
            padding: 10px 20px;
            background-color: #f7f7f7;
            color: #333;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        /* Tab and Content Styles */
        .tab {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin-bottom: 20px;
        }

        .tab button {
            padding: 10px 20px;
            margin: 5px;
            margin-bottom: 10px;
            cursor: pointer;
            background-color: #f4f4f4;
            border: 2px solid #ddd;
            border-radius: 5px;
            font-size: 1.1em;
            transition: background-color 0.3s ease;
            flex-basis: calc(10% - 10px);
            box-sizing: border-box;
        }

        .tab button:hover {
            background-color: #ecf0f1;
        }

        .tab button.active {
            background-color: #3498db;
            color: white;
            border-color: #2980b9;
        }

        .tabcontent {
            display: none;
            padding: 20px;
            margin-top: 10px;
            background-color: #ecf0f1;
            border-radius: 5px;
            box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
        }

        .tabcontent.active {
            display: block;
        }

        .tabcontent h2 {
            font-size: 2em;
            color: #2980b9;
        }

        .tabcontent p {
            font-size: 1.2em;
            line-height: 1.8;
        }

        /* Emphasize Title and Abstract */
        .title, .abstract {
            background-color: #f9f9f9;
            padding: 20px;
            margin: 10px 0;
            border-left: 5px solid #2980b9;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
        }

        .abstract {
            border-left: 5px solid #e74c3c;
        }
    </style>
    <script>
        function adjustTabStyle() {
            var tabCount = document.querySelectorAll('.tab button').length;
            var tabContainer = document.querySelector('.tab');

            if (tabCount > 10) {
                tabContainer.classList.add('small-tabs');
            }
        }

        function openTab(evt, tabId) {
            document.querySelectorAll(".tabcontent").forEach(tab => tab.style.display = "none");
            document.querySelectorAll(".tablinks").forEach(link => link.classList.remove("active"));
            document.getElementById(tabId).style.display = "block";
            evt.currentTarget.classList.add("active");
        }

        window.onload = adjustTabStyle;
    </script>
</head>
<body>
    <nav>
        <div class="navbar-left">
            <a href="">ConnectomeDB</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/">Home</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database.html">Database</a>
        </div>
        <div class="navbar-right">
            <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/">YCU Bioinformatics Lab</a>
            <span class="divider">|</span>
            <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/">Perkins Systems Biology & Genomics Lab</a>
        </div>
    </nav>

    <h1>PMID Details for IL10——IL10RA</h1>

    <div class="tab">
        <button class="tablinks active" onclick="openTab(event, 'tab12974481')">12974481</button><button class="tablinks " onclick="openTab(event, 'tab11485736')">11485736</button><button class="tablinks " onclick="openTab(event, 'tab8248239')">8248239</button><button class="tablinks " onclick="openTab(event, 'tab8897595')">8897595</button><button class="tablinks " onclick="openTab(event, 'tab11717514')">11717514</button><button class="tablinks " onclick="openTab(event, 'tab10231374')">10231374</button><button class="tablinks " onclick="openTab(event, 'tab7759550')">7759550</button>
    </div>

    <div class="tabcontent" id="titleTab">
        <div class="title">
            <h2>Title</h2>
            <p>{{TITLE}}</p>
            <i> Journal:{{JOURNAL}}; Year Published:{{YEAR}}</i>   
        </div>
    </div>

    <div class="tabcontent" id="abstractTab">
        <div class="abstract">
            <h2>Abstract</h2>
            <p>{{ABSTRACT}}</p>
        </div>
    </div>

    
                <div id="tab12974481" class="tabcontent active">
                    <h2>Class II cytokine receptors and their ligands: key antiviral and inflammatory modulators.</h2>
                    <h3>Nature reviews. Immunology; 2003</h3>
                    <p>Class II cytokine receptors were originally defined on the basis of sequence homologies in the extracellular domains of receptors for interferons (IFNs) and interleukin-10 (IL-10), and the ligands, known as class II cytokines, also have a common structure. More recently, a series of new receptors and cytokines that belong to this family have been discovered. The therapeutic potential of the 'old' members of this family, IFNs and IL-1, is recognized in the clinic, and the existence of structurally related molecules is raising expectations for additional clinical applications. In this review, I discuss both structural and biological data that are emerging about this family of receptors and ligands, to highlight the potential applications of modulating the activity of these cytokines.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/12974481/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab11485736" class="tabcontent ">
                    <h2>Crystal structure of the IL-10/IL-10R1 complex reveals a shared receptor binding site.</h2>
                    <h3>Immunity; 2001</h3>
                    <p>Interleukin 10 (IL-10) is a dimeric cytokine that plays a central role in suppressing inflammatory responses. These activities are dependent on the interaction of IL-10 with its high-affinity receptor (IL-10R1). This intermediate complex must subsequently recruit the low-affinity IL-10R2 chain before cell signaling can occur. Here we report the 2.9 A crystal structure of IL-10 bound to a soluble form of IL-10R1 (sIL-10R1). The complex consists of two IL-10s and four sIL-10R1 molecules. Several residues in the IL-10/sIL-10R1 interface are conserved in all IL-10 homologs and their receptors. The data suggests that formation of the active IL-10 signaling complex occurs by a novel molecular recognition paradigm where IL-10R1 and IL-10R2 both recognize the same binding site on IL-10.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/11485736/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab8248239" class="tabcontent ">
                    <h2>A receptor for interleukin 10 is related to interferon receptors.</h2>
                    <h3>Proceedings of the National Academy of Sciences of the United States of America; 1993</h3>
                    <p>We isolated cDNAs encoding a mouse interleukin 10 receptor (mIL-10R) from mouse mast cell and macrophage cell lines. The two cDNAs are substantially identical and express an approximately 110-kDa polypeptide in COS7 cells, which binds mIL-10 specifically. A mouse pro-B-cell line (Ba/F3) expressing transfected recombinant mIL-10R binds IL-10 with high affinity (approximately 70 pM) and proliferates in response to mIL-10. mIL-10R is structurally related to interferon receptors (IFNRs). Since IL-10 inhibits macrophage activation by IFN-gamma, a possible implication of this relationship interaction of IL-10R and IFN-gamma R or their signaling pathways.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/8248239/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab8897595" class="tabcontent ">
                    <h2>Crystal structure of human interleukin-10 at 1.6 A resolution and a model of a complex with its soluble receptor.</h2>
                    <h3>Protein science : a publication of the Protein Society; 1996</h3>
                    <p>The crystal structure of human interleukin-10 (IL-10) was refined at 1.6 A resolution against X-ray diffraction data collected at 100 K with the use of synchrotron radiation. Although similar to the IL-10 structure determined previously at room temperature, this low-temperature IL-10 structure contains, in addition, four N-terminal residues, three sulfate anions, and 175 extra water molecules. Whereas the main-chain conformation is preserved, about 30% of the side chains, most of them on the protein surface, assume different conformations. A computer model of a complex of IL-10 with its two soluble receptors was generated based on the topological similarity of IL-10 to interferon-gamma. The contact region between the cytokine and each receptor shows excellent complementarity of polar and hydrophobic interactions, suggesting that the model is generally correct and should be useful in guiding mutagenesis experiments.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/8897595/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab11717514" class="tabcontent ">
                    <h2>Purification, crystallization and preliminary X-ray diffraction of a complex between IL-10 and soluble IL-10R1.</h2>
                    <h3>Acta crystallographica. Section D, Biological crystallography; 2001</h3>
                    <p>A complex between interleukin-10 and the extracellular domain of its high-affinity receptor (sIL-10R1) has been crystallized from polyethylene glycol solutions. Crystals suitable for diffraction analysis required the modification of the NXS/T glycosylation sites on sIL-10R1 by site-directed mutagenesis and inclusion of the detergent cyclohexyl-methyl-beta-D-maltopyranoside in the crystallization experiments. The crystals belong to space group P3(2)12 or its enantimorph P3(1)12, with unit-cell parameters a = b = 46.23, c = 307.78 A, alpha = beta = 90, gamma = 120 degrees, and diffract X-rays to approximately 2.9 A. The IL-10 dimer is positioned on a crystallographic twofold, resulting in one IL-10 chain and one sIL-10R1 chain in the asymmetric unit, which corresponds to a solvent content of approximately 44%.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/11717514/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab10231374" class="tabcontent ">
                    <h2>Purification of receptor complexes of interleukin-10 stoichiometry and the importance of deglycosylation in their crystallization.</h2>
                    <h3>European journal of biochemistry; 1999</h3>
                    <p>Interleukin-10 (IL-10) is a pleiotropic immunosuppressive cytokine that has a wide range of effects in controlling inflammatory responses. Viral IL-10 (vIL-10) is a homologue of human IL-10 (hIL-10) produced by Epstein-Barr virus (EBV). Both hIL-10 and vIL-10 bind to the soluble extracellular fragment of the cytokine receptor IL-10R1 (shIL-10R1). The stoichiometry of the vIL-10 : shIL-10R1 complex has been found to be the same as hIL-10 : shIL-10R1, with two vIL-10 dimers binding to four shIL-10R1 monomers. Complexes of both hIL-10 and vIL-10 with glycosylated shIL-10R1 could not be crystallized. Controlled deglycosylation using peptide : N-glycosidase F and endo-beta-N-acetylglucosaminidase F3 resulted in the formation of crystals of both hIL-10 : shIL-10R1 and vIL-10 : shIL-10R1 complexes, indicating that the difficulty in the crystal formation was largely due to the presence of complex carbohydrate side chains. The availability of the structure of the ligand-receptor complexes should facilitate our understanding of the basis of the interaction between IL-10 and the IL-10 receptor.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/10231374/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab7759550" class="tabcontent ">
                    <h2>Characterization of recombinant extracellular domain of human interleukin-10 receptor.</h2>
                    <h3>The Journal of biological chemistry; 1995</h3>
                    <p>The extracellular region of the human interleukin-10 (hIL-10) receptor was expressed using a myeloma cell line and was purified to homogeneity by ligand-affinity chromatography. SDS-polyacrylamide gel electrophoresis analysis indicated that the soluble receptor is glycosylated and has an apparent molecular mass of 35,000-45,000. Under native conditions, soluble hIL-10 receptor was determined by gel filtration to be a monomeric protein. Soluble hIL-10 receptor was able to inhibit the binding of 125I-hIL-10 to the full-length receptor and was able to antagonize the effect of human IL-10 in cell proliferation and cytokine synthesis inhibition. The apparent dissociation constant (Kd) of soluble hIL-10 receptor was determined to be 563 +/- 59 pM, approximately 2- to 10-fold higher than that found on intact cells (Tan, J. C., Indelicato, S. R., Narula, S. K., Zavodny, P. J., and Chou, C.-C. (1993) J. Biol. Chem. 268, 21053-21059; Liu, Y., Wei, S. H.-Y., Ho, A. S.-Y., de Waal Malefyt, R., and Moore, K. W. (1994) J. Immunol. 152, 1821-1829). When hIL-10 binds soluble hIL-10 receptor in solution, a single complex was detected by gel filtration, and the complex was found to consist of two hIL-10 dimers and four soluble receptor monomers, suggesting that hIL-10 may induce a novel mode of oligomerization of the receptor upon binding.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/7759550/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                

    <footer>
        <span>© Copyright 2024, YCU Bioinformatics Laboratory</span>
        <span>Page built with <a href="https://quarto.org/">Quarto</a> by バイオインフォマティクス解析センター</span>
    </footer>
</body>
</html>
