<!DOCTYPE html>
<html>
<meta charset="UTF-8">
<head>
    <title>INHBA——ACVR1B - PMID Details</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
            color: #333;
        }

        h1 {
            text-align: center;
            font-size: 2.5em;
            color: #2c3e50;
            margin-bottom: 20px;
        }

        /* Navbar Styles */
        nav {
            background-color: #2c3e50;
            padding: 10px 20px;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            color: white;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        nav .navbar-left, nav .navbar-right {
            display: flex;
            align-items: center;
        }

        nav a {
            color: white;
            text-decoration: none;
            margin: 0 15px;
            font-size: 1.1em;
            transition: color 0.3s ease;
        }

        nav a:hover {
            color: #f39c12;
        }

        nav .divider {
            color: #ecf0f1;
            margin: 0 10px;
        }

        /* Page Footer */
        footer {
            margin-top: 40px;
            padding: 10px 20px;
            background-color: #f7f7f7;
            color: #333;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        /* Tab and Content Styles */
        .tab {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin-bottom: 20px;
        }

        .tab button {
            padding: 10px 20px;
            margin: 5px;
            margin-bottom: 10px;
            cursor: pointer;
            background-color: #f4f4f4;
            border: 2px solid #ddd;
            border-radius: 5px;
            font-size: 1.1em;
            transition: background-color 0.3s ease;
            flex-basis: calc(10% - 10px);
            box-sizing: border-box;
        }

        .tab button:hover {
            background-color: #ecf0f1;
        }

        .tab button.active {
            background-color: #3498db;
            color: white;
            border-color: #2980b9;
        }

        .tabcontent {
            display: none;
            padding: 20px;
            margin-top: 10px;
            background-color: #ecf0f1;
            border-radius: 5px;
            box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
        }

        .tabcontent.active {
            display: block;
        }

        .tabcontent h2 {
            font-size: 2em;
            color: #2980b9;
        }

        .tabcontent p {
            font-size: 1.2em;
            line-height: 1.8;
        }

        /* Emphasize Title and Abstract */
        .title, .abstract {
            background-color: #f9f9f9;
            padding: 20px;
            margin: 10px 0;
            border-left: 5px solid #2980b9;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
        }

        .abstract {
            border-left: 5px solid #e74c3c;
        }
    </style>
    <script>
        function adjustTabStyle() {
            var tabCount = document.querySelectorAll('.tab button').length;
            var tabContainer = document.querySelector('.tab');

            if (tabCount > 10) {
                tabContainer.classList.add('small-tabs');
            }
        }

        function openTab(evt, tabId) {
            document.querySelectorAll(".tabcontent").forEach(tab => tab.style.display = "none");
            document.querySelectorAll(".tablinks").forEach(link => link.classList.remove("active"));
            document.getElementById(tabId).style.display = "block";
            evt.currentTarget.classList.add("active");
        }

        window.onload = adjustTabStyle;
    </script>
</head>
<body>
    <nav>
        <div class="navbar-left">
            <a href="">ConnectomeDB</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/">Home</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database.html">Database</a>
        </div>
        <div class="navbar-right">
            <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/">YCU Bioinformatics Lab</a>
            <span class="divider">|</span>
            <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/">Perkins Systems Biology & Genomics Lab</a>
        </div>
    </nav>

    <h1>PMID Details for INHBA——ACVR1B</h1>

    <div class="tab">
        <button class="tablinks active" onclick="openTab(event, 'tab8167568')">8167568</button><button class="tablinks " onclick="openTab(event, 'tab8397373')">8397373</button>
    </div>

    <div class="tabcontent" id="titleTab">
        <div class="title">
            <h2>Title</h2>
            <p>{{TITLE}}</p>
            <i> Journal:{{JOURNAL}}; Year Published:{{YEAR}}</i>   
        </div>
    </div>

    <div class="tabcontent" id="abstractTab">
        <div class="abstract">
            <h2>Abstract</h2>
            <p>{{ABSTRACT}}</p>
        </div>
    </div>

    
                <div id="tab8167568" class="tabcontent active">
                    <h2>Molecular and functional characterization of activin receptors.</h2>
                    <h3>Receptor; 1993</h3>
                    <p>Activins are multifunctional proteins with effects on a broad spectrum of cells and tissues. They are structurally related to a large family of growth and differentiation factors that includes the inhibins, the transforming growth factors b (TGFb), the bone morphogenetic proteins (BMP), Mullerian inhibitory substance, and a number of gene products that control the development of Drosophila and Xenopus. Although the cellular signaling mechanisms of these factors remain unclear, cDNAs encoding cell surface receptors for activin have been cloned. Those receptors are transmembrane serine kinases, suggesting a novel form of signaling. Overexpression of activin receptors in Xenopus embryos indicates that these molecules are functionally involved in the transmission of the activin signal.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/8167568/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab8397373" class="tabcontent ">
                    <h2>Activin receptor-like kinases: a novel subclass of cell-surface receptors with predicted serine/threonine kinase activity.</h2>
                    <h3>Oncogene; 1993</h3>
                    <p>Human cDNA clones encoding four novel putative transmembrane protein serine/threonine kinases, denoted activin receptor-like kinase (ALK) -1, -2, -3 and -4, were obtained using a polymerase chain reaction (PCR)-based strategy. The PCR primers were designed based upon the sequence similarity between the activin receptor type II and Daf-1. The cDNA clones for ALK-1, -2 and -3 encode complete proteins of 503, 509 and 532 amino acids respectively. The ALK-4 cDNA is incomplete and the predicted protein of 383 amino acids has a truncated extracellular domain. The ALKs share similar domain structures, comprising predicted signal sequences at the N-terminals, followed by hydrophilic cysteine-rich ligand-binding domains, single hydrophobic transmembrane regions and C-terminal intracellular portions that consist almost entirely of putative serine/threonine kinase domains. The ALKs have approximately 40% sequence identity to activin receptors type II and IIB, transforming growth factor-beta (TGF-beta) type II receptor and Daf-1 in the kinase domains. However, the sequence identities are higher (60-79%) between ALK-1, -2, -3 and -4, suggesting that they form a subfamily among the putative receptor serine/threonine kinases. The extracellular domains of ALKs show only little sequence identity to other putative receptor serine/threonine kinases, but the cysteine residues are conserved. Their structural properties suggest that ALK-1 to -4 are receptors that may bind ligands that are members of the TGF-beta superfamily. The expression of mRNA in human tissues varied for the different ALKs; ALK-2 and ALK-4 showed ubiquitous tissue expression patterns, whereas the distribution of ALK-1 and ALK-3 varied strongly between different tissues with more restricted expression patterns. These results suggest that each ALK may have different in vivo functions.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/8397373/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                

    <footer>
        <span>© Copyright 2024, YCU Bioinformatics Laboratory</span>
        <span>Page built with <a href="https://quarto.org/">Quarto</a> by バイオインフォマティクス解析センター</span>
    </footer>
</body>
</html>
