<!DOCTYPE html>
<html>
<meta charset="UTF-8">
<head>
    <title>JAG1——NOTCH3 - PMID Details</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
            color: #333;
        }

        h1 {
            text-align: center;
            font-size: 2.5em;
            color: #2c3e50;
            margin-bottom: 20px;
        }

        /* Navbar Styles */
        nav {
            background-color: #2c3e50;
            padding: 10px 20px;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            color: white;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        nav .navbar-left, nav .navbar-right {
            display: flex;
            align-items: center;
        }

        nav a {
            color: white;
            text-decoration: none;
            margin: 0 15px;
            font-size: 1.1em;
            transition: color 0.3s ease;
        }

        nav a:hover {
            color: #f39c12;
        }

        nav .divider {
            color: #ecf0f1;
            margin: 0 10px;
        }

        /* Page Footer */
        footer {
            margin-top: 40px;
            padding: 10px 20px;
            background-color: #f7f7f7;
            color: #333;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        /* Tab and Content Styles */
        .tab {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin-bottom: 20px;
        }

        .tab button {
            padding: 10px 20px;
            margin: 5px;
            margin-bottom: 10px;
            cursor: pointer;
            background-color: #f4f4f4;
            border: 2px solid #ddd;
            border-radius: 5px;
            font-size: 1.1em;
            transition: background-color 0.3s ease;
            flex-basis: calc(10% - 10px);
            box-sizing: border-box;
        }

        .tab button:hover {
            background-color: #ecf0f1;
        }

        .tab button.active {
            background-color: #3498db;
            color: white;
            border-color: #2980b9;
        }

        .tabcontent {
            display: none;
            padding: 20px;
            margin-top: 10px;
            background-color: #ecf0f1;
            border-radius: 5px;
            box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
        }

        .tabcontent.active {
            display: block;
        }

        .tabcontent h2 {
            font-size: 2em;
            color: #2980b9;
        }

        .tabcontent p {
            font-size: 1.2em;
            line-height: 1.8;
        }

        /* Emphasize Title and Abstract */
        .title, .abstract {
            background-color: #f9f9f9;
            padding: 20px;
            margin: 10px 0;
            border-left: 5px solid #2980b9;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
        }

        .abstract {
            border-left: 5px solid #e74c3c;
        }
    </style>
    <script>
        function adjustTabStyle() {
            var tabCount = document.querySelectorAll('.tab button').length;
            var tabContainer = document.querySelector('.tab');

            if (tabCount > 10) {
                tabContainer.classList.add('small-tabs');
            }
        }

        function openTab(evt, tabId) {
            document.querySelectorAll(".tabcontent").forEach(tab => tab.style.display = "none");
            document.querySelectorAll(".tablinks").forEach(link => link.classList.remove("active"));
            document.getElementById(tabId).style.display = "block";
            evt.currentTarget.classList.add("active");
        }

        window.onload = adjustTabStyle;
    </script>
</head>
<body>
    <nav>
        <div class="navbar-left">
            <a href="">ConnectomeDB</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/">Home</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database.html">Database</a>
        </div>
        <div class="navbar-right">
            <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/">YCU Bioinformatics Lab</a>
            <span class="divider">|</span>
            <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/">Perkins Systems Biology & Genomics Lab</a>
        </div>
    </nav>

    <h1>PMID Details for JAG1——NOTCH3</h1>

    <div class="tab">
        <button class="tablinks active" onclick="openTab(event, 'tab11006133')">11006133</button><button class="tablinks " onclick="openTab(event, 'tab14714274')">14714274</button><button class="tablinks " onclick="openTab(event, 'tab12482954')">12482954</button>
    </div>

    <div class="tabcontent" id="titleTab">
        <div class="title">
            <h2>Title</h2>
            <p>{{TITLE}}</p>
            <i> Journal:{{JOURNAL}}; Year Published:{{YEAR}}</i>   
        </div>
    </div>

    <div class="tabcontent" id="abstractTab">
        <div class="abstract">
            <h2>Abstract</h2>
            <p>{{ABSTRACT}}</p>
        </div>
    </div>

    
                <div id="tab11006133" class="tabcontent active">
                    <h2>Physical interaction of Delta1, Jagged1, and Jagged2 with Notch1 and Notch3 receptors.</h2>
                    <h3>Biochemical and biophysical research communications; 2000</h3>
                    <p>The Delta/Serrate/LAG-2 (DSL) domain-containing proteins, Delta1, Jagged1, and Jagged2, are considered to be ligands for Notch receptors. However, the physical interaction between the three DSL proteins and respective Notch receptors remained largely unknown. In this study, we investigated this issue through the targeting of Notch1 and Notch3 in two experimental systems using fusion proteins comprising their extracellular portions. Cell-binding assays showed that soluble forms of Notch1 and Notch3 proteins physically bound to the three DSL proteins on the cell surface. In solid-phase binding assays using immobilized soluble Notch1 and Notch3 proteins, it was revealed that each DSL protein directly bound to the soluble Notch proteins with different affinities. All interactions between the DSL proteins and soluble Notch proteins were dependent on Ca(2+). Taken together, these results suggest that Delta1, Jagged1, and Jagged2 are ligands for Notch1 and Notch3 receptors.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/11006133/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab14714274" class="tabcontent ">
                    <h2>Pathogenic mutations associated with cerebral autosomal dominant arteriopathy with subcortical infarcts and leukoencephalopathy differently affect Jagged1 binding and Notch3 activity via the RBP/JK signaling Pathway.</h2>
                    <h3>American journal of human genetics; 2004</h3>
                    <p>Cerebral autosomal dominant arteriopathy with subcortical infarcts and leukoencephalopathy (CADASIL) is an inherited vascular dementia characterized by the degeneration of smooth-muscle cells in small cerebral arteries. CADASIL is caused by mutations in NOTCH3, one of the four mammalian homologs to the Drosophila melanogaster NOTCH gene. Disease-associated mutations are distributed throughout the 34 epidermal growth factor-like repeats (EGFRs) that compose the extracellular domain of the Notch3 receptor and result in a loss or a gain of a cysteine residue in one of these EGFRs. In human adults, Notch3 expression is highly restricted to vascular smooth-muscle cells. In patients with CADASIL, there is an abnormal accumulation of Notch3 in the vessel. Molecular pathways linking NOTCH3 mutations to degeneration of vascular smooth-muscle cells are as yet poorly understood. In this study, we investigated the effect of CADASIL mutations on Notch3 activity. We studied five naturally occurring mutations: R90C and C212S, located in the previously identified mutational hotspot EGFR2-5; C428S, shown in this study to be located in the ligand-binding domain EGFR10-11; and C542Y and R1006C, located in EGFR13 and EGFR26, respectively. All five mutant proteins were correctly processed. The C428S and C542Y mutant receptors exhibited a significant reduction in Jagged1-induced transcriptional activity of a RBP/JK responsive luciferase reporter, relative to wild-type Notch3. Impaired signaling activity of these two mutants arose through different mechanisms; the C428S mutant lost its Jagged1-binding ability, whereas C542Y retained it but exhibited an impaired presentation to the cell surface. In contrast, the R90C, C212S, and R1006C mutants retained the ability to bind Jagged1 and were associated with apparently normal levels of signaling activity. We conclude that mutations in Notch3 differently affect Jagged1 binding and Notch3 signaling via the RBP/JK pathway.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/14714274/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab12482954" class="tabcontent ">
                    <h2>A CADASIL-mutated Notch 3 receptor exhibits impaired intracellular trafficking and maturation but normal ligand-induced signaling.</h2>
                    <h3>Proceedings of the National Academy of Sciences of the United States of America; 2002</h3>
                    <p>Notch receptors are single transmembrane receptors that contain a large number of epidermal growth factor-like repeats (EGF repeats) in their extracellular domains. Mutations in the EGF repeats of the human Notch 3 receptor lead to the vascular dementia disease Cerebral Autosomal Dominant Arteriopathy with Subcortical Infarcts and Leukoencephalopathy (CADASIL). The vast majority of CADASIL mutations are missense mutations removing or inserting cysteine residues in the EGF repeats, but it is not yet clear whether these mutations primarily affect receptor trafficking, maturation, andor signaling. To address this issue, we have generated and analyzed stable cell lines expressing either wild-type murine Notch 3 (mNotch 3) or the mutant mNotch 3(R142C), which corresponds to the prevalent CADASIL form of Notch 3, Notch 3(R141C) in humans. We find that a lower proportion of mNotch 3(R142C) is expressed in the site 1-cleaved configuration, and that reduced amounts of mNotch 3(R142C) appear at the cell surface, as compared with wild-type mNotch 3. This observation is accompanied by a higher propensity for mNotch 3(R142C) to form intracellular aggregates, which may be a result of increased accumulation or slowed transport in the secretory pathway. In contrast to the impaired cell surface expression, mNotch 3(R142C) signals equally well in response to Delta 1 and Jagged 1 as wild-type mNotch 3. Taken together, these data suggest that trafficking and localization rather than signaling of mNotch 3 are affected in mNotch 3(R142C).</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/12482954/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                

    <footer>
        <span>© Copyright 2024, YCU Bioinformatics Laboratory</span>
        <span>Page built with <a href="https://quarto.org/">Quarto</a> by バイオインフォマティクス解析センター</span>
    </footer>
</body>
</html>
