<!DOCTYPE html>
<html>
<meta charset="UTF-8">
<head>
    <title>PLAU——PLAUR - PMID Details</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
            color: #333;
        }

        h1 {
            text-align: center;
            font-size: 2.5em;
            color: #2c3e50;
            margin-bottom: 20px;
        }

        /* Navbar Styles */
        nav {
            background-color: #2c3e50;
            padding: 10px 20px;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            color: white;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        nav .navbar-left, nav .navbar-right {
            display: flex;
            align-items: center;
        }

        nav a {
            color: white;
            text-decoration: none;
            margin: 0 15px;
            font-size: 1.1em;
            transition: color 0.3s ease;
        }

        nav a:hover {
            color: #f39c12;
        }

        nav .divider {
            color: #ecf0f1;
            margin: 0 10px;
        }

        /* Page Footer */
        footer {
            margin-top: 40px;
            padding: 10px 20px;
            background-color: #f7f7f7;
            color: #333;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        /* Tab and Content Styles */
        .tab {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin-bottom: 20px;
        }

        .tab button {
            padding: 10px 20px;
            margin: 5px;
            margin-bottom: 10px;
            cursor: pointer;
            background-color: #f4f4f4;
            border: 2px solid #ddd;
            border-radius: 5px;
            font-size: 1.1em;
            transition: background-color 0.3s ease;
            flex-basis: calc(10% - 10px);
            box-sizing: border-box;
        }

        .tab button:hover {
            background-color: #ecf0f1;
        }

        .tab button.active {
            background-color: #3498db;
            color: white;
            border-color: #2980b9;
        }

        .tabcontent {
            display: none;
            padding: 20px;
            margin-top: 10px;
            background-color: #ecf0f1;
            border-radius: 5px;
            box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
        }

        .tabcontent.active {
            display: block;
        }

        .tabcontent h2 {
            font-size: 2em;
            color: #2980b9;
        }

        .tabcontent p {
            font-size: 1.2em;
            line-height: 1.8;
        }

        /* Emphasize Title and Abstract */
        .title, .abstract {
            background-color: #f9f9f9;
            padding: 20px;
            margin: 10px 0;
            border-left: 5px solid #2980b9;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
        }

        .abstract {
            border-left: 5px solid #e74c3c;
        }
    </style>
    <script>
        function adjustTabStyle() {
            var tabCount = document.querySelectorAll('.tab button').length;
            var tabContainer = document.querySelector('.tab');

            if (tabCount > 10) {
                tabContainer.classList.add('small-tabs');
            }
        }

        function openTab(evt, tabId) {
            document.querySelectorAll(".tabcontent").forEach(tab => tab.style.display = "none");
            document.querySelectorAll(".tablinks").forEach(link => link.classList.remove("active"));
            document.getElementById(tabId).style.display = "block";
            evt.currentTarget.classList.add("active");
        }

        window.onload = adjustTabStyle;
    </script>
</head>
<body>
    <nav>
        <div class="navbar-left">
            <a href="">ConnectomeDB</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/">Home</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database.html">Database</a>
        </div>
        <div class="navbar-right">
            <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/">YCU Bioinformatics Lab</a>
            <span class="divider">|</span>
            <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/">Perkins Systems Biology & Genomics Lab</a>
        </div>
    </nav>

    <h1>PMID Details for PLAU——PLAUR</h1>

    <div class="tab">
        <button class="tablinks active" onclick="openTab(event, 'tab16456079')">16456079</button><button class="tablinks " onclick="openTab(event, 'tab3031025')">3031025</button><button class="tablinks " onclick="openTab(event, 'tab11501527')">11501527</button><button class="tablinks " onclick="openTab(event, 'tab3023326')">3023326</button>
    </div>

    <div class="tabcontent" id="titleTab">
        <div class="title">
            <h2>Title</h2>
            <p>{{TITLE}}</p>
            <i> Journal:{{JOURNAL}}; Year Published:{{YEAR}}</i>   
        </div>
    </div>

    <div class="tabcontent" id="abstractTab">
        <div class="abstract">
            <h2>Abstract</h2>
            <p>{{ABSTRACT}}</p>
        </div>
    </div>

    
                <div id="tab16456079" class="tabcontent active">
                    <h2>Structure of human urokinase plasminogen activator in complex with its receptor.</h2>
                    <h3>Science (New York, N.Y.); 2006</h3>
                    <p>The urokinase plasminogen activator binds to its cellular receptor with high affinity and initiates signaling cascades that are implicated in pathological processes including tumor growth, metastasis, and inflammation. We report the crystal structure at 1.9 angstroms of the urokinase receptor complexed with the urokinase amino-terminal fragment and an antibody against the receptor. The three domains of urokinase receptor form a concave shape with a central cone-shaped cavity where the urokinase fragment inserts. The structure provides insight into the flexibility of the urokinase receptor that enables its interaction with a wide variety of ligands and a basis for the design of urokinase-urokinase receptor antagonists.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/16456079/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab3031025" class="tabcontent ">
                    <h2>The receptor-binding sequence of urokinase. A biological function for the growth-factor module of proteases.</h2>
                    <h3>The Journal of biological chemistry; 1987</h3>
                    <p>Previous studies have shown that the region of human urokinase-type plasminogen activator (uPA) responsible for receptor binding resides in the amino-terminal fragment (ATF, residues 1-135) (Stoppelli, M.P., Corti, A., Soffientini, A., Cassani, G., Blasi, F., and Assoian, R.K. (1985) Proc. Natl. Acad. Sci. U.S. A. 82, 4939-4943). The area within ATF responsible for specific receptor binding has now been identified by the ability of different synthetic peptides corresponding to different regions of the amino terminus of uPA to inhibit receptor binding of 125I-labeled ATF. A peptide corresponding to human [Ala19]uPA-(12-32) resulted in 50% inhibition of ATF binding at 100 nM. Peptides uPA-(18-32) and [Ala13]uPA-(9-20) inhibit at 100 and 2000 microM, respectively. The human peptide uPA-(1-14) and the mouse peptide [Ala20]uPA-(13-33) have no effect on ATF receptor binding. This region of uPA is referred to as the growth factor module since it shares partial amino acid sequence homology (residues 14-33) to epidermal growth factor (EGF). Furthermore, this region of EGF is responsible for binding of EGF to its receptor (Komoriya, A. Hortsch, M., Meyers, C., Smith, M., Kanety, H., and Schlessinger, J. (1984) Proc. Natl. Acad. Sci. U.S.A. 81, 1351-1355). However, EGF does not inhibit ATF receptor binding. Comparison of the sequences responsible for receptor binding of uPA and EGF indicate that the region of highest homology is between residues 13-19 and 14-20 of human uPA and EGF, respectively. In addition, there is a conservation of the spacings of four cysteines in this module whereas there is no homology between residues 20-30 and 21-33 of uPA and EGF. Thus, residues 20-30 of uPA apparently confer receptor binding specificity, and residues 13-19 provide the proper conformation to the adjacent binding region.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/3031025/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab11501527" class="tabcontent ">
                    <h2>Ligand binding regions in the receptor for urokinase-type plasminogen activator.</h2>
                    <h3>The Journal of biological chemistry; 2001</h3>
                    <p>The interaction between urokinase plasminogen activator (uPA) and its cellular receptor (uPAR) is a key event in cell surface-associated plasminogen activation, relevant for cell migration and invasion. In order to define receptor recognition sites for uPA, we have expressed uPAR fragments as fusion products with the minor coat protein on the surface of M13 bacteriophages. Sequence analysis of cDNA fragments encoding uPA-binding peptides indicated the existence of a composite uPA-binding structure including all three uPAR domains. This finding was confirmed by experiments using an overlapping 15-mer peptide array covering the entire uPAR molecule. Four regions within the uPAR sequence were found to directly bind to uPA: two distinct regions containing amino acids 13--20 and amino acids 74--84 of the uPAR domain I, and regions in the putative loop 3 of the domains II and III. All the uPA-binding fragments from the three domains were shown to have an agonistic effect on uPA binding to immobilized uPAR. Furthermore, uPAR-(154--176) increased uPAR-transfected BAF3-cell adhesion on vitronectin in the presence of uPA, whereas uPAR-(247--276) stimulated the cell adhesion both in the absence or presence of uPA. The latter fragment was also able to augment the binding of vitronectin to uPAR in a purified system, thereby mimicking the effect of uPA on this interaction. These results indicate that uPA binding can take place to particular part(s) on several uPAR molecules and that direct uPAR-uPAR contacts may contribute to receptor activation and ligand binding.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/11501527/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab3023326" class="tabcontent ">
                    <h2>Binding of single-chain prourokinase to the urokinase receptor of human U937 cells.</h2>
                    <h3>The Journal of biological chemistry; 1986</h3>
                    <p>The single-chain form of human urokinase plasminogen activator (uPA) is the major form of the enzyme found in cells, tissues, and extracellular fluids. The protein, called pro-uPA, has high (Kd = 0.5 nM) affinity for the specific uPA receptor of U937 human monocyte-like cells. Its conversion to two-chain uPA by plasmin does not appreciably change the binding parameters. In addition, conversion of pro-uPA to uPA occurs with receptor-bound pro-uPA and does not lead to dissociation from the membrane. These data show that secreted pro-uPA can find its way to the specific surface receptor without previous conversion to the two-chain form and that, once bound, can be activated by plasmin.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/3023326/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                

    <footer>
        <span>© Copyright 2024, YCU Bioinformatics Laboratory</span>
        <span>Page built with <a href="https://quarto.org/">Quarto</a> by バイオインフォマティクス解析センター</span>
    </footer>
</body>
</html>
