<!DOCTYPE html>
<html>
<meta charset="UTF-8">
<head>
    <title>PLG——F2RL1 - PMID Details</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
            color: #333;
        }

        h1 {
            text-align: center;
            font-size: 2.5em;
            color: #2c3e50;
            margin-bottom: 20px;
        }

        /* Navbar Styles */
        nav {
            background-color: #2c3e50;
            padding: 10px 20px;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            color: white;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        nav .navbar-left, nav .navbar-right {
            display: flex;
            align-items: center;
        }

        nav a {
            color: white;
            text-decoration: none;
            margin: 0 15px;
            font-size: 1.1em;
            transition: color 0.3s ease;
        }

        nav a:hover {
            color: #f39c12;
        }

        nav .divider {
            color: #ecf0f1;
            margin: 0 10px;
        }

        /* Page Footer */
        footer {
            margin-top: 40px;
            padding: 10px 20px;
            background-color: #f7f7f7;
            color: #333;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        /* Tab and Content Styles */
        .tab {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin-bottom: 20px;
        }

        .tab button {
            padding: 10px 20px;
            margin: 5px;
            margin-bottom: 10px;
            cursor: pointer;
            background-color: #f4f4f4;
            border: 2px solid #ddd;
            border-radius: 5px;
            font-size: 1.1em;
            transition: background-color 0.3s ease;
            flex-basis: calc(10% - 10px);
            box-sizing: border-box;
        }

        .tab button:hover {
            background-color: #ecf0f1;
        }

        .tab button.active {
            background-color: #3498db;
            color: white;
            border-color: #2980b9;
        }

        .tabcontent {
            display: none;
            padding: 20px;
            margin-top: 10px;
            background-color: #ecf0f1;
            border-radius: 5px;
            box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
        }

        .tabcontent.active {
            display: block;
        }

        .tabcontent h2 {
            font-size: 2em;
            color: #2980b9;
        }

        .tabcontent p {
            font-size: 1.2em;
            line-height: 1.8;
        }

        /* Emphasize Title and Abstract */
        .title, .abstract {
            background-color: #f9f9f9;
            padding: 20px;
            margin: 10px 0;
            border-left: 5px solid #2980b9;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
        }

        .abstract {
            border-left: 5px solid #e74c3c;
        }
    </style>
    <script>
        function adjustTabStyle() {
            var tabCount = document.querySelectorAll('.tab button').length;
            var tabContainer = document.querySelector('.tab');

            if (tabCount > 10) {
                tabContainer.classList.add('small-tabs');
            }
        }

        function openTab(evt, tabId) {
            document.querySelectorAll(".tabcontent").forEach(tab => tab.style.display = "none");
            document.querySelectorAll(".tablinks").forEach(link => link.classList.remove("active"));
            document.getElementById(tabId).style.display = "block";
            evt.currentTarget.classList.add("active");
        }

        window.onload = adjustTabStyle;
    </script>
</head>
<body>
    <nav>
        <div class="navbar-left">
            <a href="">ConnectomeDB</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/">Home</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database.html">Database</a>
        </div>
        <div class="navbar-right">
            <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/">YCU Bioinformatics Lab</a>
            <span class="divider">|</span>
            <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/">Perkins Systems Biology & Genomics Lab</a>
        </div>
    </nav>

    <h1>PMID Details for PLG——F2RL1</h1>

    <div class="tab">
        <button class="tablinks active" onclick="openTab(event, 'tab12604689')">12604689</button><button class="tablinks " onclick="openTab(event, 'tab10978167')">10978167</button><button class="tablinks " onclick="openTab(event, 'tab9020112')">9020112</button>
    </div>

    <div class="tabcontent" id="titleTab">
        <div class="title">
            <h2>Title</h2>
            <p>{{TITLE}}</p>
            <i> Journal:{{JOURNAL}}; Year Published:{{YEAR}}</i>   
        </div>
    </div>

    <div class="tabcontent" id="abstractTab">
        <div class="abstract">
            <h2>Abstract</h2>
            <p>{{ABSTRACT}}</p>
        </div>
    </div>

    
                <div id="tab12604689" class="tabcontent active">
                    <h2>Selective tryptic cleavage at the tethered ligand site of the amino terminal domain of proteinase-activated receptor-2 in intact cells.</h2>
                    <h3>The Journal of pharmacology and experimental therapeutics; 2003</h3>
                    <p>In intact cells, trypsin activates proteinase-activated receptor-2 (PAR(2)) by hydrolysis at residues R(36)/S(37) (amino acids are abbreviated by their one-letter code), revealing an active tethered ligand sequence. We sought to determine whether in intact cells, the tryptic cleavage/activation of PAR(2) might also be accompanied by hydrolysis at other potential N-terminal cleavage sites, like residues K(34), R(41), K(51), and K(72), as implied by the tryptic cleavage in vitro at these residues of Escherichia coli-expressed human N-terminal PAR(2)R(31)-P(79). To this end, four PAR(2) mutants with altered tryptic cleavage sites were prepared (PAR(2)R(36)A, PAR(2)S(37)P, PAR(2)R(41)A, and PAR(2)R(36)AR(41)A), expressed in Kirsten virus-transformed rat kidney cells and were evaluated together with the wild-type PAR(2)-expressing cells for 1) activation (Ca(2+) signaling) by trypsin and the receptor-activating peptide SLIGRL-NH(2) (SL-NH(2)) and 2) the tryptic release of two antigenic receptor determinants, one N-terminal to the R(36)/S(37) cleavage/activation site detected by SLAW-A antibody and the second (detected by antibody, B5), N-terminal to residues K(51), K(72). None of the mutants resistant to cleavage at R(36) were activated by trypsin, yet all retained reactivity to B5 and all were activated by SL-NH(2). In contrast, trypsin activated both wild-type and PAR(2)R(41)A, leading to a disappearance of SLAW-A but not B5 reactivity. We conclude that, as opposed to the E. coli-expressed PAR(2) N-terminal polypeptide, PAR(2) expressed in intact cells displays selective tryptic cleavage at the R(36)/S(37) activation site, without cleaving downstream. Thus, in intact cells, trypsin activation does not concurrently "disarm" rat PAR(2), but leaves the "tethered ligand" persistently attached to the body of the receptor.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/12604689/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab10978167" class="tabcontent ">
                    <h2>Proteolysis of the exodomain of recombinant protease-activated receptors: prediction of receptor activation or inactivation by MALDI mass spectrometry.</h2>
                    <h3>Biochemistry; 2000</h3>
                    <p>Protease-activated receptors (PARs) mediate cell activation after proteolytic cleavage of their extracellular amino terminus. Thrombin selectively cleaves PAR1, PAR3, and PAR4 to induce activation of platelets and vascular cells, while PAR2 is preferentially cleaved by trypsin. In pathological situations, other proteolytic enzymes may be generated in the circulation and could modify the responses of PARs by cleaving their extracellular domains. To assess the ability of such proteases to activate or inactivate PARs, we designed a strategy for locating cleavage sites on the exofacial NH(2)-terminal fragments of the receptors. The first extracellular segments of PAR1 (PAR1E) and PAR2 (PAR2E) expressed as recombinant proteins in Escherichia coli were incubated with a series of proteases likely to be encountered in the circulation during thrombosis or inflammation. Kinetic and dose-response studies were performed, and the cleavage products were analyzed by MALDI-TOF mass spectrometry. Thrombin cleaved PAR1E at the Arg41-Ser42 activation site at concentrations known to induce cellular activation, supporting a native conformation of the recombinant polypeptide. Plasmin, calpain and leukocyte elastase, cathepsin G, and proteinase 3 cleaved at multiple sites and would be expected to disable PAR1 by cleaving COOH-terminal to the activation site. Cleavage specificities were further confirmed using activation site defective PAR1E S42P mutant polypeptides. Surface plasmon resonance studies on immobilized PAR1E or PAR1E S42P were consistent with cleavage results obtained in solution and allowed us to determine affinities of PAR1E-thrombin binding. FACS analyses of intact platelets confirmed the cleavage of PAR1 downstream of the Arg41-Ser42 site. Mass spectrometry studies of PAR2E predicted activation of PAR2 by trypsin through cleavage at the Arg36-Ser37 site, no effect of thrombin, and inactivation of the receptor by plasmin, calpain and leukocyte elastase, cathepsin G, and proteinase 3. The inhibitory effect of elastase was confirmed on native PAR1 and PAR2 on the basis of Ca(2+) signaling studies in endothelial cells. It was concluded that none of the main proteases generated during fibrinolysis or inflammation appears to be able to signal through PAR1 or PAR2. This strategy provides results which can be extended to the native receptor to predict its activation or inactivation, and it could likewise be used to study other PARs or protease-dependent processes.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/10978167/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab9020112" class="tabcontent ">
                    <h2>Interactions of mast cell tryptase with thrombin receptors and PAR-2.</h2>
                    <h3>The Journal of biological chemistry; 1997</h3>
                    <p>Tryptase is a serine protease secreted by mast cells that is able to activate other cells. In the present studies we have tested whether these responses could be mediated by thrombin receptors or PAR-2, two G-protein-coupled receptors that are activated by proteolysis. When added to a peptide corresponding to the N terminus of PAR-2, tryptase cleaved the peptide at the activating site, but at higher concentrations it also cleaved downstream, as did trypsin, a known activator of PAR-2. Thrombin, factor Xa, plasmin, urokinase, plasma kallikrein, and tissue kallikrein had no effect. Tryptase also cleaved the analogous thrombin receptor peptide at the activating site but less efficiently. When added to COS-1 cells expressing either receptor, tryptase stimulated phosphoinositide hydrolysis. With PAR-2, this response was half-maximal at 1 nM tryptase and could be inhibited by the tryptase inhibitor, APC366, or by antibodies to tryptase and PAR-2. When added to human endothelial cells, which normally express PAR-2 and thrombin receptors, or keratinocytes, which express only PAR-2, tryptase caused an increase in cytosolic Ca2+. However, when added to platelets or CHRF-288 cells, which express thrombin receptors but not PAR-2, tryptase caused neither aggregation nor increased Ca2+. These results show that 1) tryptase has the potential to activate both PAR-2 and thrombin receptors; 2) for PAR-2, this potential is realized, although cleavage at secondary sites may limit activation, particularly at higher tryptase concentrations; and 3) in contrast, although tryptase clearly activates thrombin receptors in COS-1 cells, it does not appear to cleave endogenous thrombin receptors in platelets or CHRF-288 cells. These distinctions correlate with the observed differences in the rate of cleavage of the PAR-2 and thrombin receptor peptides by tryptase. Tryptase is the first protease other than trypsin that has been shown to activate human PAR-2. Its presence within mast cell granules places it in tissues where PAR-2 is expressed but trypsin is unlikely to reach.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/9020112/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                

    <footer>
        <span>© Copyright 2024, YCU Bioinformatics Laboratory</span>
        <span>Page built with <a href="https://quarto.org/">Quarto</a> by バイオインフォマティクス解析センター</span>
    </footer>
</body>
</html>
