<!DOCTYPE html>
<html>
<meta charset="UTF-8">
<head>
    <title>POMC——MC4R - PMID Details</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
            color: #333;
        }

        h1 {
            text-align: center;
            font-size: 2.5em;
            color: #2c3e50;
            margin-bottom: 20px;
        }

        /* Navbar Styles */
        nav {
            background-color: #2c3e50;
            padding: 10px 20px;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            color: white;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        nav .navbar-left, nav .navbar-right {
            display: flex;
            align-items: center;
        }

        nav a {
            color: white;
            text-decoration: none;
            margin: 0 15px;
            font-size: 1.1em;
            transition: color 0.3s ease;
        }

        nav a:hover {
            color: #f39c12;
        }

        nav .divider {
            color: #ecf0f1;
            margin: 0 10px;
        }

        /* Page Footer */
        footer {
            margin-top: 40px;
            padding: 10px 20px;
            background-color: #f7f7f7;
            color: #333;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        /* Tab and Content Styles */
        .tab {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin-bottom: 20px;
        }

        .tab button {
            padding: 10px 20px;
            margin: 5px;
            margin-bottom: 10px;
            cursor: pointer;
            background-color: #f4f4f4;
            border: 2px solid #ddd;
            border-radius: 5px;
            font-size: 1.1em;
            transition: background-color 0.3s ease;
            flex-basis: calc(10% - 10px);
            box-sizing: border-box;
        }

        .tab button:hover {
            background-color: #ecf0f1;
        }

        .tab button.active {
            background-color: #3498db;
            color: white;
            border-color: #2980b9;
        }

        .tabcontent {
            display: none;
            padding: 20px;
            margin-top: 10px;
            background-color: #ecf0f1;
            border-radius: 5px;
            box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
        }

        .tabcontent.active {
            display: block;
        }

        .tabcontent h2 {
            font-size: 2em;
            color: #2980b9;
        }

        .tabcontent p {
            font-size: 1.2em;
            line-height: 1.8;
        }

        /* Emphasize Title and Abstract */
        .title, .abstract {
            background-color: #f9f9f9;
            padding: 20px;
            margin: 10px 0;
            border-left: 5px solid #2980b9;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
        }

        .abstract {
            border-left: 5px solid #e74c3c;
        }
    </style>
    <script>
        function adjustTabStyle() {
            var tabCount = document.querySelectorAll('.tab button').length;
            var tabContainer = document.querySelector('.tab');

            if (tabCount > 10) {
                tabContainer.classList.add('small-tabs');
            }
        }

        function openTab(evt, tabId) {
            document.querySelectorAll(".tabcontent").forEach(tab => tab.style.display = "none");
            document.querySelectorAll(".tablinks").forEach(link => link.classList.remove("active"));
            document.getElementById(tabId).style.display = "block";
            evt.currentTarget.classList.add("active");
        }

        window.onload = adjustTabStyle;
    </script>
</head>
<body>
    <nav>
        <div class="navbar-left">
            <a href="">ConnectomeDB</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/">Home</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database.html">Database</a>
        </div>
        <div class="navbar-right">
            <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/">YCU Bioinformatics Lab</a>
            <span class="divider">|</span>
            <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/">Perkins Systems Biology & Genomics Lab</a>
        </div>
    </nav>

    <h1>PMID Details for POMC——MC4R</h1>

    <div class="tab">
        <button class="tablinks active" onclick="openTab(event, 'tab8884876')">8884876</button><button class="tablinks " onclick="openTab(event, 'tab10493100')">10493100</button><button class="tablinks " onclick="openTab(event, 'tab11101306')">11101306</button><button class="tablinks " onclick="openTab(event, 'tab12165561')">12165561</button><button class="tablinks " onclick="openTab(event, 'tab11101306')">11101306</button>
    </div>

    <div class="tabcontent" id="titleTab">
        <div class="title">
            <h2>Title</h2>
            <p>{{TITLE}}</p>
            <i> Journal:{{JOURNAL}}; Year Published:{{YEAR}}</i>   
        </div>
    </div>

    <div class="tabcontent" id="abstractTab">
        <div class="abstract">
            <h2>Abstract</h2>
            <p>{{ABSTRACT}}</p>
        </div>
    </div>

    
                <div id="tab8884876" class="tabcontent active">
                    <h2>Characterisation of the melanocortin 4 receptor by radioligand binding.</h2>
                    <h3>Pharmacology & toxicology; 1996</h3>
                    <p>The DNA encoding the human melanocortin 4 receptor was expressed in COS (CV-1 origin, SV 40) cells and its radioligand binding properties was tested by using the [125I][Nle4, D-Phe7] alpha-melanocyte stimulating hormone (MSH). The radioligand was found to bind to a single saturable site with a Kd of 3.84 +/- 0.57 nmol/l in the MC4 receptor expressing cells. The order of potency of a number of substance competing for the [125I][Nle4, D-Phe7] alpha-MSH binding was the following; [Nle4, D-Phe7] alpha-MSH > [Nle4]-alpha-MSH > beta-MSH > desacetyl-alpha-MSH > alpha-MSH > ACTH (1-39) > ACTH (4-10) > gamma 1-MSH > gamma 2-MSH. This order of potency is unique for the melanocortin 4 receptor when compared to our previously published data for the other melanocortin receptor subtypes. Most notably the melanocortin 4 receptor shows highest affinity for beta-MSH, among the endogenous MSH-peptides. Furthermore the melanocortin 4 receptor shows very low affinity for the gamma-MSH peptides. This distinguishes the melanocortin 4 receptor from the melanocortin 3 receptor, which is the other major central nervous system melanocortin-receptor, as melanocortin 3 receptor shows high affinity for gamma-MSH. Our finding might indicate a specific role for beta-MSH for the melanocortin 4 receptor.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/8884876/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab10493100" class="tabcontent ">
                    <h2>Characterization of melanocortin receptor ligands on cloned brain melanocortin receptors and on grooming behavior in the rat.</h2>
                    <h3>European journal of pharmacology; 1999</h3>
                    <p>Since the melanocortin MC3 and melanocortin MC4 receptors are the main melanocortin receptor subtypes expressed in rat brain, we characterized the activity and affinity of nine melanocortin receptor ligands using these receptors in vitro, as well as their activity in a well-defined melanocortin-induced behavior in the rat: grooming behavior. We report here that [D-Tyr4]melanotan-II and RMI-2001 (Ac-cyclo-[Cys4, Gly5, D-Phe7, Cys10]alpha-MSH-NH2) have significantly higher affinity and potency on the rat melanocortin MC4 receptor as compared to the rat melanocortin MC3 receptor. Nle-gamma-MSH (melanocyte-stimulating hormone) was the only ligand with higher affinity and potency on the rat melanocortin MC3 receptor. The potency order of melanocortin MC4 receptor agonists, but not that of melanocortin MC3 receptor agonists, fitted with the potency of these ligands to stimulate grooming behavior, when administered intracerebroventricularly. SHU9119 (Ac-cyclo-[Nle4, Asp5, D-Nal(2)7, Lys10]alpha-MSH-(4-10)-NH2) and RMI-2005 (Ac-cyclo-[Cys4, Gly5, D-Na](2)7, Nal(2)9, Cys10]alpha-MSH-(4-10)-NH2) were able to inhibit alpha-MSH-induced melanocortin receptor activity in vitro, as well as alpha-MSH-induced grooming behavior. Melanotan-II, [Nle4-D-Phe7]alpha-MSH and RMI-2001 were also effective in inducing grooming behavior when administered intravenously. In the absence of purely selective melanocortin MC(3/4) receptor ligands, we demonstrated that careful comparison of ligand potencies in vitro with ligand potencies in vivo, could identify which melanocortin receptor subtype mediated alpha-MSH-induced grooming behavior. Furthermore, blockade of novelty-induced grooming behavior by SHU9119 demonstrated that this physiological stress response is mediated via activation of the melanocortin system.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/10493100/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab11101306" class="tabcontent ">
                    <h2>Molecular determinants of ligand binding to the human melanocortin-4 receptor.</h2>
                    <h3>Biochemistry; 2000</h3>
                    <p>To elucidate the molecular basis for the interaction of ligands with the human melanocortin-4 receptor (hMC4R), agonist structure-activity studies and receptor point mutagenesis were performed. Structure-activity studies of [Nle(4), D-Phe(7)]-alpha-melanocyte stimulating hormone (NDP-MSH) identified D-Phe7-Arg8-Trp9 as the minimal NDP-MSH fragment that possesses full agonist efficacy at the hMC4R. In an effort to identify receptor residues that might interact with amino acids in this tripeptide sequence 24 hMC4R transmembrane (TM) residues were mutated (the rationale for choosing specific receptor residues for mutation is outlined in the Results section). Mutation of TM3 residues D122 and D126 and TM6 residues F261 and H264 decreased the binding affinity of NDP-MSH 5-fold or greater, thereby identifying these receptor residues as sites potentially involved in the sought after ligand-receptor interactions. By examination of the binding affinities and potencies of substituted NDP-MSH peptides at receptor mutants, evidence was found that core melanocortin peptide residue Arg8 interacts at a molecular level with hMC4R TM3 residue D122. TM3 mutations were also observed to decrease the binding of hMC4R antagonists. Notably, mutation of TM3 residue D126 to alanine decreased the binding affinity of AGRP (87-132), a C-terminal derivative of the endogenous melanocortin antagonist, 8-fold, and simultaneous mutations D122A/D126A completely abolished AGRP (87-132) binding. In addition, mutation of TM3 residue D122 or D126 decreased the binding affinity of hMC4R antagonist SHU 9119. These results provide further insight into the molecular determinants of hMC4R ligand binding.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/11101306/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab12165561" class="tabcontent ">
                    <h2>A missense mutation disrupting a dibasic prohormone processing site in pro-opiomelanocortin (POMC) increases susceptibility to early-onset obesity through a novel molecular mechanism.</h2>
                    <h3>Human molecular genetics; 2002</h3>
                    <p>The functional loss of both alleles of the human pro-opiomelanocortin (POMC) gene leads to a very rare syndrome of hypoadrenalism, red hair and early-onset obesity. In order to examine whether more subtle genetic variants in POMC might contribute to early-onset obesity, the coding region of the gene was sequenced in 262 Caucasian subjects with a history of severe obesity from childhood. Two children were found to be heterozygous for a missense mutation, R236G, which disrupts the dibasic cleavage site between beta melanocyte-stimulating hormone (beta-MSH) and beta-endorphin. Beta-TC3 cells transfected with the mutant POMC cDNA produced a mutant beta-MSH/beta-endorphin fusion protein. This fusion protein bound to the human melanocortin-4 receptor (hMC4R) with an affinity similar to its natural ligands, but had a markedly reduced ability to activate the receptor. This variant co-segregated with early-onset obesity over three generations in one family and was absent in 412 normal weight UK Caucasian controls. Combining the results in UK Caucasians with a new case-control study in French subjects and three previously published reports, mutations disrupting this processing site were present in 0.88% of subjects with early-onset obesity and 0.22% of normal-weight controls. These results suggest that the R236G mutation may confer an inherited susceptibility to obesity through the production of an aberrant fusion protein that has the capacity to interfere with central melanocortin signalling.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/12165561/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab11101306" class="tabcontent ">
                    <h2>Molecular determinants of ligand binding to the human melanocortin-4 receptor.</h2>
                    <h3>Biochemistry; 2000</h3>
                    <p>To elucidate the molecular basis for the interaction of ligands with the human melanocortin-4 receptor (hMC4R), agonist structure-activity studies and receptor point mutagenesis were performed. Structure-activity studies of [Nle(4), D-Phe(7)]-alpha-melanocyte stimulating hormone (NDP-MSH) identified D-Phe7-Arg8-Trp9 as the minimal NDP-MSH fragment that possesses full agonist efficacy at the hMC4R. In an effort to identify receptor residues that might interact with amino acids in this tripeptide sequence 24 hMC4R transmembrane (TM) residues were mutated (the rationale for choosing specific receptor residues for mutation is outlined in the Results section). Mutation of TM3 residues D122 and D126 and TM6 residues F261 and H264 decreased the binding affinity of NDP-MSH 5-fold or greater, thereby identifying these receptor residues as sites potentially involved in the sought after ligand-receptor interactions. By examination of the binding affinities and potencies of substituted NDP-MSH peptides at receptor mutants, evidence was found that core melanocortin peptide residue Arg8 interacts at a molecular level with hMC4R TM3 residue D122. TM3 mutations were also observed to decrease the binding of hMC4R antagonists. Notably, mutation of TM3 residue D126 to alanine decreased the binding affinity of AGRP (87-132), a C-terminal derivative of the endogenous melanocortin antagonist, 8-fold, and simultaneous mutations D122A/D126A completely abolished AGRP (87-132) binding. In addition, mutation of TM3 residue D122 or D126 decreased the binding affinity of hMC4R antagonist SHU 9119. These results provide further insight into the molecular determinants of hMC4R ligand binding.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/11101306/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                

    <footer>
        <span>© Copyright 2024, YCU Bioinformatics Laboratory</span>
        <span>Page built with <a href="https://quarto.org/">Quarto</a> by バイオインフォマティクス解析センター</span>
    </footer>
</body>
</html>
