<!DOCTYPE html>
<html>
<meta charset="UTF-8">
<head>
    <title>PROC——THBD - PMID Details</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
            color: #333;
        }

        h1 {
            text-align: center;
            font-size: 2.5em;
            color: #2c3e50;
            margin-bottom: 20px;
        }

        /* Navbar Styles */
        nav {
            background-color: #2c3e50;
            padding: 10px 20px;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            color: white;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        nav .navbar-left, nav .navbar-right {
            display: flex;
            align-items: center;
        }

        nav a {
            color: white;
            text-decoration: none;
            margin: 0 15px;
            font-size: 1.1em;
            transition: color 0.3s ease;
        }

        nav a:hover {
            color: #f39c12;
        }

        nav .divider {
            color: #ecf0f1;
            margin: 0 10px;
        }

        /* Page Footer */
        footer {
            margin-top: 40px;
            padding: 10px 20px;
            background-color: #f7f7f7;
            color: #333;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        /* Tab and Content Styles */
        .tab {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin-bottom: 20px;
        }

        .tab button {
            padding: 10px 20px;
            margin: 5px;
            margin-bottom: 10px;
            cursor: pointer;
            background-color: #f4f4f4;
            border: 2px solid #ddd;
            border-radius: 5px;
            font-size: 1.1em;
            transition: background-color 0.3s ease;
            flex-basis: calc(10% - 10px);
            box-sizing: border-box;
        }

        .tab button:hover {
            background-color: #ecf0f1;
        }

        .tab button.active {
            background-color: #3498db;
            color: white;
            border-color: #2980b9;
        }

        .tabcontent {
            display: none;
            padding: 20px;
            margin-top: 10px;
            background-color: #ecf0f1;
            border-radius: 5px;
            box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
        }

        .tabcontent.active {
            display: block;
        }

        .tabcontent h2 {
            font-size: 2em;
            color: #2980b9;
        }

        .tabcontent p {
            font-size: 1.2em;
            line-height: 1.8;
        }

        /* Emphasize Title and Abstract */
        .title, .abstract {
            background-color: #f9f9f9;
            padding: 20px;
            margin: 10px 0;
            border-left: 5px solid #2980b9;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
        }

        .abstract {
            border-left: 5px solid #e74c3c;
        }
    </style>
    <script>
        function adjustTabStyle() {
            var tabCount = document.querySelectorAll('.tab button').length;
            var tabContainer = document.querySelector('.tab');

            if (tabCount > 10) {
                tabContainer.classList.add('small-tabs');
            }
        }

        function openTab(evt, tabId) {
            document.querySelectorAll(".tabcontent").forEach(tab => tab.style.display = "none");
            document.querySelectorAll(".tablinks").forEach(link => link.classList.remove("active"));
            document.getElementById(tabId).style.display = "block";
            evt.currentTarget.classList.add("active");
        }

        window.onload = adjustTabStyle;
    </script>
</head>
<body>
    <nav>
        <div class="navbar-left">
            <a href="">ConnectomeDB</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/">Home</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database.html">Database</a>
        </div>
        <div class="navbar-right">
            <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/">YCU Bioinformatics Lab</a>
            <span class="divider">|</span>
            <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/">Perkins Systems Biology & Genomics Lab</a>
        </div>
    </nav>

    <h1>PMID Details for PROC——THBD</h1>

    <div class="tab">
        <button class="tablinks active" onclick="openTab(event, 'tab11106601')">11106601</button><button class="tablinks " onclick="openTab(event, 'tab12588872')">12588872</button>
    </div>

    <div class="tabcontent" id="titleTab">
        <div class="title">
            <h2>Title</h2>
            <p>{{TITLE}}</p>
            <i> Journal:{{JOURNAL}}; Year Published:{{YEAR}}</i>   
        </div>
    </div>

    <div class="tabcontent" id="abstractTab">
        <div class="abstract">
            <h2>Abstract</h2>
            <p>{{ABSTRACT}}</p>
        </div>
    </div>

    
                <div id="tab11106601" class="tabcontent active">
                    <h2>Modeling zymogen protein C.</h2>
                    <h3>Biophysical journal; 2000</h3>
                    <p>A solution structure for the complete zymogen form of human coagulation protein C is modeled. The initial core structure is based on the x-ray crystallographic structure of the gamma-carboxyglutamic acid (Gla)-domainless activated form. The Gla domain (residues 1-48) is modeled from the x-ray crystal coordinates of the factor VII(a)/tissue factor complex and oriented with the epidermal growth factor-1 domain to yield an initial orientation consistent with the x-ray crystal structure of porcine factor IX(a). The missing C-terminal residues in the light chain (residues 147-157) and the activation peptide residues 158-169 were introduced using homology modeling so that the activation peptide residues directly interact with the residues in the calcium binding loop. Molecular dynamics simulations (Amber-particle-mesh-Ewald) are used to obtain the complete calcium-complexed solution structure. The individual domain structures of protein C in solution are largely unaffected by solvation, whereas the Gla-epidermal growth factor-1 orientation evolves to a form different from both factors VII(a) and IX(a). The solution structure of the zymogen protein C is compared with the crystal structures of the existing zymogen serine proteases: chymotrypsinogen, proproteinase, and prethrombin-2. Calculated electrostatic potential surfaces support the involvement of the serine protease calcium ion binding loop in providing a suitable electrostatic environment around the scissile bond for II(a)/thrombomodulin interaction.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/11106601/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab12588872" class="tabcontent ">
                    <h2>Mutations in autolytic loop-2 and at Asp554 of human prothrombin that enhance protein C activation by meizothrombin.</h2>
                    <h3>The Journal of biological chemistry; 2003</h3>
                    <p>Thrombin acts on many protein substrates during the hemostatic process. Its specificity for these substrates is modulated through interactions at regions remote from the active site of the thrombin molecule, designated exosites. Exosite interactions can be with the substrate, cofactors such as thrombomodulin, or fragments from prothrombin. The relative activity of alpha-thrombin for fibrinogen is 10 times greater than that for protein C. However, the relative activity of meizothrombin for protein C is 14 times greater than that for fibrinogen. Modulation of thrombin specificity is linked to its Na(+)-binding site and residues in autolytic loop-2 that interact with the Na(+)-binding site. Recombinant prothrombins that yield recombinant meizothrombin (rMT) and rMT des-fragment 1 (rMT(desF1)) enable comparisons of the effects of mutations at the Na(+)-binding residue (Asp(554)) and deletion of loop-2 (Glu(466)-Thr(469)) on the relative activity of meizothrombin for several substrates. Hydrolysis of t-butoxycarbonyl-VPR-p-nitroanilide by alpha-thrombin, recombinant alpha-thrombin, or rMT(desF1) was almost identical, but that by rMT was only 40% of that by alpha-thrombin. Clotting of fibrinogen by rMT and rMT(desF1) was 12-16% of that by alpha-thrombin, as already known. Strikingly, however, although meizothrombins modified by substitution of Asp(554) with either Ala or Leu or by deletion of loop-2 had 6-8 and <1%, respectively, of the clotting activity of alpha-thrombin, the activity of these meizothrombins for protein C was increased to >10 times that of alpha-thrombin. It is proposed that interactions within thrombin that involve autolytic loop-2 and the Na(+)-binding site primarily enhance thrombin action on fibrinogen, but impair thrombin action on protein C.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/12588872/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                

    <footer>
        <span>© Copyright 2024, YCU Bioinformatics Laboratory</span>
        <span>Page built with <a href="https://quarto.org/">Quarto</a> by バイオインフォマティクス解析センター</span>
    </footer>
</body>
</html>
