<!DOCTYPE html>
<html>
<meta charset="UTF-8">
<head>
    <title>PROC——PROCR - PMID Details</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
            color: #333;
        }

        h1 {
            text-align: center;
            font-size: 2.5em;
            color: #2c3e50;
            margin-bottom: 20px;
        }

        /* Navbar Styles */
        nav {
            background-color: #2c3e50;
            padding: 10px 20px;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            color: white;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        nav .navbar-left, nav .navbar-right {
            display: flex;
            align-items: center;
        }

        nav a {
            color: white;
            text-decoration: none;
            margin: 0 15px;
            font-size: 1.1em;
            transition: color 0.3s ease;
        }

        nav a:hover {
            color: #f39c12;
        }

        nav .divider {
            color: #ecf0f1;
            margin: 0 10px;
        }

        /* Page Footer */
        footer {
            margin-top: 40px;
            padding: 10px 20px;
            background-color: #f7f7f7;
            color: #333;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        /* Tab and Content Styles */
        .tab {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin-bottom: 20px;
        }

        .tab button {
            padding: 10px 20px;
            margin: 5px;
            margin-bottom: 10px;
            cursor: pointer;
            background-color: #f4f4f4;
            border: 2px solid #ddd;
            border-radius: 5px;
            font-size: 1.1em;
            transition: background-color 0.3s ease;
            flex-basis: calc(10% - 10px);
            box-sizing: border-box;
        }

        .tab button:hover {
            background-color: #ecf0f1;
        }

        .tab button.active {
            background-color: #3498db;
            color: white;
            border-color: #2980b9;
        }

        .tabcontent {
            display: none;
            padding: 20px;
            margin-top: 10px;
            background-color: #ecf0f1;
            border-radius: 5px;
            box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
        }

        .tabcontent.active {
            display: block;
        }

        .tabcontent h2 {
            font-size: 2em;
            color: #2980b9;
        }

        .tabcontent p {
            font-size: 1.2em;
            line-height: 1.8;
        }

        /* Emphasize Title and Abstract */
        .title, .abstract {
            background-color: #f9f9f9;
            padding: 20px;
            margin: 10px 0;
            border-left: 5px solid #2980b9;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
        }

        .abstract {
            border-left: 5px solid #e74c3c;
        }
    </style>
    <script>
        function adjustTabStyle() {
            var tabCount = document.querySelectorAll('.tab button').length;
            var tabContainer = document.querySelector('.tab');

            if (tabCount > 10) {
                tabContainer.classList.add('small-tabs');
            }
        }

        function openTab(evt, tabId) {
            document.querySelectorAll(".tabcontent").forEach(tab => tab.style.display = "none");
            document.querySelectorAll(".tablinks").forEach(link => link.classList.remove("active"));
            document.getElementById(tabId).style.display = "block";
            evt.currentTarget.classList.add("active");
        }

        window.onload = adjustTabStyle;
    </script>
</head>
<body>
    <nav>
        <div class="navbar-left">
            <a href="">ConnectomeDB</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/">Home</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database.html">Database</a>
        </div>
        <div class="navbar-right">
            <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/">YCU Bioinformatics Lab</a>
            <span class="divider">|</span>
            <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/">Perkins Systems Biology & Genomics Lab</a>
        </div>
    </nav>

    <h1>PMID Details for PROC——PROCR</h1>

    <div class="tab">
        <button class="tablinks active" onclick="openTab(event, 'tab7929370')">7929370</button><button class="tablinks " onclick="openTab(event, 'tab11099506')">11099506</button><button class="tablinks " onclick="openTab(event, 'tab10681521')">10681521</button>
    </div>

    <div class="tabcontent" id="titleTab">
        <div class="title">
            <h2>Title</h2>
            <p>{{TITLE}}</p>
            <i> Journal:{{JOURNAL}}; Year Published:{{YEAR}}</i>   
        </div>
    </div>

    <div class="tabcontent" id="abstractTab">
        <div class="abstract">
            <h2>Abstract</h2>
            <p>{{ABSTRACT}}</p>
        </div>
    </div>

    
                <div id="tab7929370" class="tabcontent active">
                    <h2>Identification, cloning, and regulation of a novel endothelial cell protein C/activated protein C receptor.</h2>
                    <h3>The Journal of biological chemistry; 1994</h3>
                    <p>Human protein C and activated protein C are shown to bind to endothelium specifically, selectively and saturably (Kd = 30 nM, 7000 sites per cell) in a Ca(2+)-dependent fashion. Expression cloning revealed a 1.3-kilobase pair cDNA that coded for a novel type 1 transmembrane glycoprotein capable of binding protein C. This protein appears to be a member of the CD1/major histocompatibility complex superfamily. Like thrombomodulin, the receptor involved in protein C activation, the endothelial cell protein C receptor function and message are both down-regulated by exposure of endothelium to tumor necrosis factor. Identification of endothelial cell protein C receptor as a member of the CD1/major histocompatibility complex superfamily provides insights into the role of protein C in regulating the inflammatory response.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/7929370/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab11099506" class="tabcontent ">
                    <h2>Identification of the protein C/activated protein C binding sites on the endothelial cell protein C receptor. Implications for a novel mode of ligand recognition by a major histocompatibility complex class 1-type receptor.</h2>
                    <h3>Journal of biological chemistry; 2001</h3>
                    <p>The endothelial cell protein C receptor (EPCR) is an endothelial cell-specific transmembrane protein that binds both protein C and activated protein C (APC). EPCR regulates the protein C anticoagulant pathway by binding protein C and augmenting protein C activation by the thrombin-thrombomodulin complex. EPCR is homologous to the MHC class 1/CD1 family, members of which contain two alpha-helices that sit upon an 8-stranded beta-sheet platform. In this study, we identified 10 residues that, when mutated to alanine, result in the loss of protein C/APC binding (Arg-81, Leu-82, Val-83, Glu-86, Arg-87, Phe-146, Tyr-154, Thr-157, Arg-158, and Glu-160). Glutamine substitutions at the four N-linked carbohydrate attachment sites of EPCR have little affect on APC binding, suggesting that the carbohydrate moieties of EPCR are not critical for ligand recognition. We then mapped the epitopes for four anti-human EPCR monoclonal antibodies (mAbs), two of which block EPCR/Fl-APC (APC labeled at the active site with fluorescein) interactions, whereas two do not. These epitopes were localized by generating human-mouse EPCR chimeric proteins, since the mAbs under investigation do not recognize mouse EPCR. We found that 5 of the 10 candidate residues for protein C/APC binding (Arg-81, Leu-82, Val-83, Glu-86, Arg-87) colocalize with the epitope for one of the blocking mAbs. Three-dimensional molecular modeling of EPCR indicates that the 10 protein C/APC binding candidate residues are clustered at the distal end of the two alpha-helical segments. Protein C activation studies on 293 cells that coexpress EPCR variants and thrombomodulin demonstrate that protein C binding to EPCR is necessary for the EPCR-dependent enhancement in protein activation by the thrombin-thrombomodulin complex. These studies indicate that EPCR has exploited the MHC class 1 fold for an alternative and possibly novel mode of ligand recognition. These studies are also the first to identify the protein C/APC binding region of EPCR and may provide useful information about molecular defects in EPCR that could contribute to cardiovascular disease susceptibility.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/11099506/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab10681521" class="tabcontent ">
                    <h2>Mechanisms by which soluble endothelial cell protein C receptor modulates protein C and activated protein C function.</h2>
                    <h3>The Journal of biological chemistry; 2000</h3>
                    <p>The endothelial cell protein C receptor (EPCR) functions as an important regulator of the protein C anticoagulant pathway by binding protein C and enhancing activation by the thrombin-thrombomodulin complex. EPCR binds to both protein C and activated protein C (APC) with high affinity. A soluble form of EPCR (sEPCR) circulates in plasma and inhibits APC anticoagulant activity. In this study, we investigate the mechanisms by which sEPCR modulates APC function. Soluble EPCR inhibited the inactivation of factor Va by APC only in the presence of phospholipid vesicles. By using flow cytometric analysis in the presence of 3 mM CaCl(2) and 0. 6 mM MgCl(2), sEPCR inhibited the binding of protein C and APC to phospholipid vesicles (K(i) = 40 +/- 7 and 33 +/- 4 nM, respectively). Without MgCl(2), the K(i) values increased approximately 4-fold. Double label flow cytometric analysis using fluorescein-APC and Texas Red-sEPCR indicated that the APC.sEPCR complex does not interact with phospholipid vesicles. By using surface plasmon resonance, we found that sEPCR also inhibited binding of protein C to phospholipid in a dose-dependent fashion (K(i) = 32 nM). To explore the possibility that sEPCR evokes structural changes in APC, fluorescence spectroscopy studies were performed to monitor sEPCR/Fl-APC interactions. sEPCR binds saturably to Fl-APC (K(d) = 27 +/- 13 nM) with a maximum decrease in Fl-APC fluorescence of 10.8 +/- 0.6%. sEPCR also stimulated the amidolytic activity of APC toward synthetic substrates. We conclude that sEPCR binding to APC blocks phospholipid interaction and alters the active site of APC.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/10681521/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                

    <footer>
        <span>© Copyright 2024, YCU Bioinformatics Laboratory</span>
        <span>Page built with <a href="https://quarto.org/">Quarto</a> by バイオインフォマティクス解析センター</span>
    </footer>
</body>
</html>
