<!DOCTYPE html>
<html>
<meta charset="UTF-8">
<head>
    <title>TNF——TNFRSF1A - PMID Details</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
            color: #333;
        }

        h1 {
            text-align: center;
            font-size: 2.5em;
            color: #2c3e50;
            margin-bottom: 20px;
        }

        /* Navbar Styles */
        nav {
            background-color: #2c3e50;
            padding: 10px 20px;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            color: white;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        nav .navbar-left, nav .navbar-right {
            display: flex;
            align-items: center;
        }

        nav a {
            color: white;
            text-decoration: none;
            margin: 0 15px;
            font-size: 1.1em;
            transition: color 0.3s ease;
        }

        nav a:hover {
            color: #f39c12;
        }

        nav .divider {
            color: #ecf0f1;
            margin: 0 10px;
        }

        /* Page Footer */
        footer {
            margin-top: 40px;
            padding: 10px 20px;
            background-color: #f7f7f7;
            color: #333;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }

        /* Tab and Content Styles */
        .tab {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
            margin-bottom: 20px;
        }

        .tab button {
            padding: 10px 20px;
            margin: 5px;
            margin-bottom: 10px;
            cursor: pointer;
            background-color: #f4f4f4;
            border: 2px solid #ddd;
            border-radius: 5px;
            font-size: 1.1em;
            transition: background-color 0.3s ease;
            flex-basis: calc(10% - 10px);
            box-sizing: border-box;
        }

        .tab button:hover {
            background-color: #ecf0f1;
        }

        .tab button.active {
            background-color: #3498db;
            color: white;
            border-color: #2980b9;
        }

        .tabcontent {
            display: none;
            padding: 20px;
            margin-top: 10px;
            background-color: #ecf0f1;
            border-radius: 5px;
            box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
        }

        .tabcontent.active {
            display: block;
        }

        .tabcontent h2 {
            font-size: 2em;
            color: #2980b9;
        }

        .tabcontent p {
            font-size: 1.2em;
            line-height: 1.8;
        }

        /* Emphasize Title and Abstract */
        .title, .abstract {
            background-color: #f9f9f9;
            padding: 20px;
            margin: 10px 0;
            border-left: 5px solid #2980b9;
            box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
        }

        .abstract {
            border-left: 5px solid #e74c3c;
        }
    </style>
    <script>
        function adjustTabStyle() {
            var tabCount = document.querySelectorAll('.tab button').length;
            var tabContainer = document.querySelector('.tab');

            if (tabCount > 10) {
                tabContainer.classList.add('small-tabs');
            }
        }

        function openTab(evt, tabId) {
            document.querySelectorAll(".tabcontent").forEach(tab => tab.style.display = "none");
            document.querySelectorAll(".tablinks").forEach(link => link.classList.remove("active"));
            document.getElementById(tabId).style.display = "block";
            evt.currentTarget.classList.add("active");
        }

        window.onload = adjustTabStyle;
    </script>
</head>
<body>
    <nav>
        <div class="navbar-left">
            <a href="">ConnectomeDB</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/">Home</a>
            <a href="https://comp.med.yokohama-cu.ac.jp/collab/connectomeDB/database.html">Database</a>
        </div>
        <div class="navbar-right">
            <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/">YCU Bioinformatics Lab</a>
            <span class="divider">|</span>
            <a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/">Perkins Systems Biology & Genomics Lab</a>
        </div>
    </nav>

    <h1>PMID Details for TNF——TNFRSF1A</h1>

    <div class="tab">
        <button class="tablinks active" onclick="openTab(event, 'tab7852363')">7852363</button><button class="tablinks " onclick="openTab(event, 'tab10634209')">10634209</button><button class="tablinks " onclick="openTab(event, 'tab14743216')">14743216</button><button class="tablinks " onclick="openTab(event, 'tab19084540')">19084540</button>
    </div>

    <div class="tabcontent" id="titleTab">
        <div class="title">
            <h2>Title</h2>
            <p>{{TITLE}}</p>
            <i> Journal:{{JOURNAL}}; Year Published:{{YEAR}}</i>   
        </div>
    </div>

    <div class="tabcontent" id="abstractTab">
        <div class="abstract">
            <h2>Abstract</h2>
            <p>{{ABSTRACT}}</p>
        </div>
    </div>

    
                <div id="tab7852363" class="tabcontent active">
                    <h2>Mapping the domain(s) critical for the binding of human tumor necrosis factor-alpha to its two receptors.</h2>
                    <h3>The Journal of biological chemistry; 1995</h3>
                    <p>The extracellular domains of the two human tumor necrosis factor (TNF) receptors critical for binding TNF-alpha were examined by deletion mapping. The ligand binding capability of full-length and truncated recombinant soluble TNF receptors (TNFRs) was assessed by ligand blot analysis and their binding affinity determined by Scatchard analysis. The results showed that deletion of the fourth cysteine-rich domain of the p55 receptor (TNFR-1) did not alter ligand binding affinity significantly. Deletion of domains 3 and 4 of TNFR-1 resulted in no ligand binding, suggesting that domain 3, but not 4, of TNFR-1 binds directly to ligand. Deletion of domain 4 of TNFR-2 resulted in drastically reduced protein yield and 3-fold reduction in ligand binding affinity, while deletion of both domains 4 and 3 yielded no protein. Thus, the domain 4 of TNFR-2, but not that of TNFR-1, appears to be involved directly in binding TNF, although it is also possible that the domain 4 of TNFR-2 is involved in the correct folding of other domains. These results suggest that the modes of interaction between TNF-alpha and its dual receptors are different, providing opportunity to modulate each receptor specifically for research and therapeutic purposes.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/7852363/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab10634209" class="tabcontent ">
                    <h2>TNF-induced signaling in apoptosis.</h2>
                    <h3>Journal of clinical immunology; 1999</h3>
                    <p>Out of the almost 17 members of the TNF superfamily, TNF is probably the most potent inducer of apoptosis. TNF activates both cell-survival and cell-death mechanisms simultaneously. Activation of NF-kB-dependent genes regulates the survival and proliferative effects pf TNF, whereas activation of caspases regulates the apoptotic effects. TNF-induced apoptosis is mediated primarily through the activation of type I receptors, the death domain of which recruits more than a dozen different signaling proteins, which together are considered part of an apoptotic cascade. This cascade does not, however, account for the role of reactive oxygen intermediates, ceramide, phospholipases, and serine proteases which are also implicated in TNF-induced apoptosis. This cascade also does not explain how type II TNF receptors which lack the death domain, induce apoptosis. Nevertheless, this review of apoptosis signaling will be limited to those proteins that makeup the cascade.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/10634209/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab14743216" class="tabcontent ">
                    <h2>A physical and functional map of the human TNF-alpha/NF-kappa B signal transduction pathway.</h2>
                    <h3>Nature cell biology; 2004</h3>
                    <p>Signal transduction pathways are modular composites of functionally interdependent sets of proteins that act in a coordinated fashion to transform environmental information into a phenotypic response. The pro-inflammatory cytokine tumour necrosis factor (TNF)-alpha triggers a signalling cascade, converging on the activation of the transcription factor NF-kappa B, which forms the basis for numerous physiological and pathological processes. Here we report the mapping of a protein interaction network around 32 known and candidate TNF-alpha/NF-kappa B pathway components by using an integrated approach comprising tandem affinity purification, liquid-chromatography tandem mass spectrometry, network analysis and directed functional perturbation studies using RNA interference. We identified 221 molecular associations and 80 previously unknown interactors, including 10 new functional modulators of the pathway. This systems approach provides significant insight into the logic of the TNF-alpha/NF-kappa B pathway and is generally applicable to other pathways relevant to human disease.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/14743216/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                
                <div id="tab19084540" class="tabcontent ">
                    <h2>Structure-function relationship of tumor necrosis factor (TNF) and its receptor interaction based on 3D structural analysis of a fully active TNFR1-selective TNF mutant.</h2>
                    <h3>Journal of molecular biology; 2009</h3>
                    <p>Tumor necrosis factor (TNF) is an important cytokine that suppresses carcinogenesis and excludes infectious pathogens to maintain homeostasis. TNF activates its two receptors [TNF receptor (TNFR) 1 and TNFR2], but the contribution of each receptor to various host defense functions and immunologic surveillance is not yet clear. Here, we used phage display techniques to generate receptor-selective TNF mutants that activate only one TNFR. These TNF mutants will be useful in the functional analysis of TNFR. Six amino acids in the receptor binding interface (near TNF residues 30, 80, and 140) were randomly mutated by polymerase chain reaction. Two phage libraries comprising over 5 million TNF mutants were constructed. By selecting the mutants without affinity for TNFR1 or TNFR2, we successfully isolated 4 TNFR2-selective candidates and 16 TNFR1-selective candidates, respectively. The TNFR1-selective candidates were highly mutated near residue 30, whereas TNFR2-selective candidates were highly mutated near residue 140, although both had conserved sequences near residues 140 and 30, respectively. This finding suggested that the phage display technique was suitable for identifying important regions for the TNF interaction with TNFR1 and TNFR2. Purified clone R1-6, a TNFR1-selective candidate, remained fully bioactive and had full affinity for TNFR1 without activating TNFR2, indicating the usefulness of the R1-6 TNF mutant in analyzing TNFR1 receptor function. To further elucidate the receptor selectivity of R1-6, we examined the structure of R1-6 by X-ray crystallography. The results suggested that R31A and R32G mutations strongly influenced electrostatic interaction with TNFR2, and that L29K mutation contributed to the binding of R1-6 to TNFR1. This phage display technique can be used to efficiently construct functional mutants for analysis of the TNF structure-function relationship, which might facilitate in silico drug design based on receptor selectivity.</p>
                    <p><a href="https://pubmed.ncbi.nlm.nih.gov/19084540/" target="_blank">[For more details, see PubMed]</a></p>
                </div>
                

    <footer>
        <span>© Copyright 2024, YCU Bioinformatics Laboratory</span>
        <span>Page built with <a href="https://quarto.org/">Quarto</a> by バイオインフォマティクス解析センター</span>
    </footer>
</body>
</html>
