---
title: "{{< fa circle-info >}} Introduction"
execute:
  echo: false
  warnings: false  # Add this line to suppress warnings globally
header-includes: <script src='/js/keepDropdownMenuGold.js'></script>
---

<hr>

```{python}
import os
import sys
import pandas as pd
import warnings
warnings.filterwarnings("ignore", category=pd.errors.SettingWithCopyWarning)
warnings.filterwarnings("ignore", category=pd.errors.DtypeWarning)
project_root = os.path.dirname(os.getcwd())
os.chdir(project_root)
sys.path.append(os.path.abspath("src"))
from createDataTable import (
  lrPairsCount,
  ligandCount,
  receptorCount,
  numOfMouseOrth
)
```

<div style="border: 1.5px solid #F5F5F5; border-radius: 10px; padding: 8px 10px 0px 10px; background-color: #F5F5F5; width: 100%; box-sizing: border-box; margin: 0; line-height: 1.25;">
  <strong>ConnectomeDB</strong> is a comprehensive and ongoing project that provides a high-quality manually curated database of interacting ligand-receptor pairs for use in cell-cell communication analysis. First released in 2015 ([Ramilowski, *et al.*](https://doi.org/10.1038/ncomms8866){target="_blank"}), and subsequently updated in 2020 ([Hou, *et al.*](https://www.nature.com/articles/s41467-020-18873-z){target="_blank"}) and 2025 (Liu, Maezono *et al.*), it aims to enhance the understanding of cell-cell communication in humans and other mammals, supporting biological and medical research.<span class="half-break"></span></div>

<span class="half-break"></span>

### **ConnectomeDB2025**

<div style="margin-top: 1em;"></div>

- **`{python} f"{lrPairsCount:,}"` Manually Curated Ligand–Receptor Pairs** (Primarily Human; includes `{python} f"{numOfMouseOrth:,}"` Mouse-specific): comprehensive and most accurate with primary literature support\
- **`{python} f"{ligandCount:,}"` Ligands** & **`{python} f"{receptorCount:,}"` Receptors**: metadata and functional annotations\
- **Homologous Interactions**: ~97% LR pairs in mouse and in 12 other vertebrate species\

<hr>

### **Online Resource**

Accesing all parts of the online resource requires no username and no password.

**Software**: ConnectomeDB2025 is built using a hybrid frontend. <a href='https://quarto.org/' target='_blank'>Quarto 1.7+</a> is used for interactive content and documentation, while certain modules rely on standalone HTML with Python/Jinja2 backends. Users can explore the data via two access modes: (1) species-specific pair list views, offering dynamic search and filtering by gene symbol or annotation fields, and (2) ligand–receptor pair cards, which provide detailed annotations per interaction, including ortholog coverage, aliases, subcellular localization, and literature support.

To ensure responsive performance, all data tables are pre-rendered as JSON using Python, allowing fast in-browser filtering without requiring server-side computation. This setup supports static HTML generation and makes the site usable on a wide range of devices, including mobile.

Some images were created using <a href='https://www.biorender.com/' target='_blank'> BioRender</a>.

**OS & Browsers**: Windows 10/11 (Google Chrome, Microsoft Edge), macOS 15+ (Google Chrome, Safari), mobile devices (Google Chrome, Safari).

<hr>

<div style="margin-top: 1em;"></div>

### **Developers**

This resource was developed and is maintained by <a href="https://www-user.yokohama-cu.ac.jp/~bioinfo/hp/" target="_blank" >YCU Bioinformatics Lab</a>
and 
<a href="https://perkins.org.au/research/labs/cancer-program/systems-biology-and-genomics-overview/" target="_blank">Perkins Systems Biology &amp; Genomics Lab</a>\
**GitHub**: <https://github.com/bioinfo-YCU/ConnectomeDB>

<hr>

<div style="margin-top: 1em;"></div>

### **Cite Us**

{{< include citeus.qmd >}}

